<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/14/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-逃离dplyr_ 不使用group_by和arrange实现分组排序-sort-by-group-without-dplyr" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/01/09/%E9%80%83%E7%A6%BBdplyr_%20%E4%B8%8D%E4%BD%BF%E7%94%A8group_by%E5%92%8Carrange%E5%AE%9E%E7%8E%B0%E5%88%86%E7%BB%84%E6%8E%92%E5%BA%8F-sort-by-group-without-dplyr/" class="article-date">
  <time class="dt-published" datetime="2020-01-09T16:07:03.766Z" itemprop="datePublished">2020-01-09</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/01/09/%E9%80%83%E7%A6%BBdplyr_%20%E4%B8%8D%E4%BD%BF%E7%94%A8group_by%E5%92%8Carrange%E5%AE%9E%E7%8E%B0%E5%88%86%E7%BB%84%E6%8E%92%E5%BA%8F-sort-by-group-without-dplyr/">逃离dplyr:不使用group_by和arrange实现分组排序</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>今天在写代码的时候，发现项目中出现了一些重复的代码，所以要把他们封装成一个单个函数。在封装的过程中，我遇到了一个让我头疼的问题。</p>
<p>在使用dplyr的时候，你可能会注意到一个非常有趣的细节，那就是你不用<code>&quot;&quot;</code>来区别变量和字符串，<code>dplyr</code>能够帮你好这个事情。举个例子，下面的代码都是让<code>iris</code>数据集按照”Sepal.Length”进行排序。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">group_by<span class="punctuation">(</span>iris<span class="punctuation">,</span> Sepal.Length<span class="punctuation">)</span></span><br><span class="line">group_by<span class="punctuation">(</span>iris<span class="punctuation">,</span> <span class="string">&quot;Sepal.Length&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这时候，让我们思考一个问题，如果在之前命名了一个<code>group.by &lt;- &quot;Sepal.Length&quot;</code>，那么运行<code>group_by(iris, group.by)</code>的时候， 这个group.by会被替换成Sepal.Length吗？下面的代码会报错吗？大家可以思考一下，然后往下看。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">group.by <span class="operator">&lt;-</span> <span class="string">&quot;Sepal.Length&quot;</span></span><br><span class="line">group_by<span class="punctuation">(</span>iris<span class="punctuation">,</span> group.by<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>实际上，运行上面的代码，你会得到一个报错”Error: Column <code>group.by</code> is unknown”. <code>group_by</code>没有替换掉你的变量名。</p>
<p>为什么会出现这个情况？这个就涉及到dplyr编程的内容，具体可以参考<a target="_blank" rel="noopener" href="https://cran.r-project.org/web/packages/dplyr/vignettes/programming.html">https://cran.r-project.org/web/packages/dplyr/vignettes/programming.html</a></p>
<p>参考dplyr的教程，如果要让上面的代码能够运行，我们需要需要在变量名前加上<code>!!</code>或者调用<code>UQ</code>函数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">group_by<span class="punctuation">(</span>iris<span class="punctuation">,</span> <span class="operator">!</span><span class="operator">!</span>group.by<span class="punctuation">)</span></span><br><span class="line">group_by<span class="punctuation">(</span>iris<span class="punctuation">,</span> UQ<span class="punctuation">(</span>group.by<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>假如你要创造一个函数，要用到<code>group_by</code>，那么你应该怎么写呢？我们的直觉就是下面的代码</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">my_group_by <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>df<span class="punctuation">,</span> group.by<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  df <span class="operator">&lt;-</span> group_by<span class="punctuation">(</span>df<span class="punctuation">,</span>group.by <span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>根据前面的铺垫，你应该知道，运行<code>my_group_by(iris, Sepal.Length)</code>会出现报错，报错信息为” Error: Column <code>group.by</code> is unknown”。 于是你试着之前的解决方法加上了”UQ” </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">my_group_by <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>df<span class="punctuation">,</span> group.by<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  df <span class="operator">&lt;-</span> group_by<span class="punctuation">(</span>df<span class="punctuation">,</span> UQ<span class="punctuation">(</span>group.by<span class="punctuation">)</span> <span class="punctuation">)</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>df<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>思考下，<code>my_group_by(iris, Sepal.Length)</code>能够得到结果吗？</p>
<p>很遗憾，代码报错了</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Error in splice(dot_call(capture_dots, frame_env = frame_env, named = named,  : </span><br><span class="line"> object &#x27;Sepal.Length&#x27; not found </span><br></pre></td></tr></table></figure>

<p>正确的调用方法是<code>my_group_by(iris, &quot;Sepal.Length&quot;)</code>.  当然由于你用习惯了dplyr，你希望是<code>my_group_by(iris, Sepal.Length)</code>调用代码，那么你的函数需要怎么写呢？为了解决这个问题，你可能要仔细阅读<a target="_blank" rel="noopener" href="https://cran.r-project.org/web/packages/dplyr/vignettes/programming.html">https://cran.r-project.org/web/packages/dplyr/vignettes/programming.html</a>才行，理解什么叫做”Quasiquotation”</p>
<p>经过一波折腾之后，我受够了这种将dplyr代码改成函数时的不一致性（也可能是我不熟练），我决定还是只用R基础代码来实现我的功能，不就是分组排序吗，为啥一定要用<code>group_by</code> 和 <code>arrange</code>呢.</p>
<p>无非就是先利用因子将数据库分成多个列表(split)，然后对每个列表按照某一列进行排序(lapply)，而这里排序过程就是获取从最大到最小的索引(order)，最后按行进行合并(do.call, rbind)而已呀。如下是实现的代码</p>
<pre><code class="r">my.group.by &lt;- function(df, group.by, sort.by,
                        decreasing = TRUE)&#123;
  
  df.split &lt;- split(df, df[[group.by]])
  
  df.split.sort &lt;- lapply(df.split, function(x)&#123;
    x.order &lt;- order(x[[sort.by]],decreasing = decreasing)
    x &lt;- x[x.order,]
    x
  &#125;)
   df &lt;- do.call(rbind, df.split.sort)
  return(df)
&#125;

my.group.by(iris, group.by = &quot;Species&quot;, sort.by = &quot;Sepal.Length&quot;)
</code></pre>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/01/09/%E9%80%83%E7%A6%BBdplyr_%20%E4%B8%8D%E4%BD%BF%E7%94%A8group_by%E5%92%8Carrange%E5%AE%9E%E7%8E%B0%E5%88%86%E7%BB%84%E6%8E%92%E5%BA%8F-sort-by-group-without-dplyr/" data-id="cllszh66w00gve2qef3sp97tw" data-title="逃离dplyr:不使用group_by和arrange实现分组排序" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-使用EDTA进行TE注释-annotate-te-with-edta-pipeline" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/01/08/%E4%BD%BF%E7%94%A8EDTA%E8%BF%9B%E8%A1%8CTE%E6%B3%A8%E9%87%8A-annotate-te-with-edta-pipeline/" class="article-date">
  <time class="dt-published" datetime="2020-01-08T16:16:47.328Z" itemprop="datePublished">2020-01-08</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/01/08/%E4%BD%BF%E7%94%A8EDTA%E8%BF%9B%E8%A1%8CTE%E6%B3%A8%E9%87%8A-annotate-te-with-edta-pipeline/">使用EDTA进行TE注释</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>一句话评价：重复序列注释用EDTA就完事了。</p>
<h2 id="简介"><a href="#简介" class="headerlink" title="简介"></a>简介</h2><p>EDTA, 全称是 Extensive de-novo TE Annotator, 一个综合性的流程工具，它整合了目前LTR预测工具结果，TIR预测工具结果，MITE预测工具结果，Helitrons预测工具结果, 从而构建出一高可信，非冗余的TE数据库，用做基因组的注释。流程图如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/1/image-450f4b170c1f42eb9450676772fb5a5a.png" alt="EDTA流程"></p>
<h2 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h2><p>如果没有管理员权限，可以用conda进行安装。如果有管理员权限，可以尝试用docker或者singularity进行安装。</p>
<p>PS: 如果之前用过EDTA的话，可以更新一下版本，因为从1.7.0开始，EDTA把intact TE和homology based注释结合在一起，最终产生了很高质量的gff，合并了所有注释；1.7.1版把包含基因的intact也去掉了，进一步过滤。（来自于作者的建议）</p>
<h3 id="conda"><a href="#conda" class="headerlink" title="conda"></a>conda</h3><p>使用conda的安装方法如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">conda create -n EDTA</span><br><span class="line">conda activate EDTA</span><br><span class="line">python2 -m pip install --user numpy==1.14.3 biopython==1.74 pp</span><br><span class="line">conda config --<span class="built_in">env</span> --add channels anaconda --add channels conda-forge --add channels biocore --add channels bioconda --add channels cyclus</span><br><span class="line">conda install -n EDTA -y cd-hit repeatmodeler muscle mdust repeatmasker=4.0.9_p2 blast-legacy java-jdk perl perl-text-soundex multiprocess regex tensorflow=1.14.0 keras=2.2.4 scikit-learn=0.19.0 biopython pandas glob2 python=3.6 trf</span><br><span class="line">git <span class="built_in">clone</span> https://github.com/oushujun/EDTA</span><br><span class="line">./EDTA/EDTA.pl</span><br></pre></td></tr></table></figure>

<p>需要注意的一点是，bioconda建议添加国内镜像站点, 否则可能会下载失败。</p>
<h3 id="singularity"><a href="#singularity" class="headerlink" title="singularity"></a>singularity</h3><p>方法1: 使用EDTA上提供的docker镜像，以singularity进行安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">singularity build edta.sif docker://kapeel/edta</span><br></pre></td></tr></table></figure>

<p>在使用时，有一点需要注意，需要用<code>-B</code>将外部的RepeatMasker的Libraries绑定的Libraries，否则可能会在检查依赖这一步失败。</p>
<p>这里，我参考的是<a target="_blank" rel="noopener" href="https://github.com/lfaino/LoReAn/">LoReAN</a>的方法</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ~ <span class="comment"># 切换到家目录</span></span><br><span class="line"><span class="built_in">mkdir</span> -p LoReAn &amp;&amp; <span class="built_in">cd</span> LoReAn</span><br><span class="line">wget https://github.com/lfaino/LoReAn/raw/noIPRS/third_party/software/RepeatMasker.Libraries.tar.gz &amp;&amp; tar -zxvf RepeatMasker.Libraries.tar.gz</span><br><span class="line">singularity <span class="built_in">exec</span> \</span><br><span class="line">    -B /home/xzg/LoReAn/Libraries/:/opt/conda/share/RepeatMasker/Libraries/ \</span><br><span class="line">    /home/xzg/edta.sif /EDTA/EDTA.pl -h</span><br></pre></td></tr></table></figure>

<p>上面的<code>/home/xzg/</code>是我的家目录，需要根据实际情况进行选择</p>
<p>方法2: 使用 @wangshun1121 构建的docker镜像, 他解决了需要<code>-B</code>进行挂载的问题。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">singularity build edta.sif docker://registry.cn-hangzhou.aliyuncs.com/wangshun1121/edta</span><br></pre></td></tr></table></figure>

<h2 id="实战"><a href="#实战" class="headerlink" title="实战"></a>实战</h2><p>我们以拟南芥的第一条染色体为例，进行介绍</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># singularity</span></span><br><span class="line">singularity <span class="built_in">exec</span> ~/LoReAn/edta.sif /EDTA/EDTA.pl \</span><br><span class="line">    -genome chr1.fa -species others -step all -t 20</span><br><span class="line"><span class="comment"># conda</span></span><br><span class="line"><span class="comment"># 我的EDTA在我的家目录下</span></span><br><span class="line">~/EDTA/EDTA.pl -genome chr1.fa -species others -step all -t 20</span><br></pre></td></tr></table></figure>

<p>这里的参数比较简单,</p>
<ul>
<li><code>-genome</code>: 输入的基因组序列</li>
<li><code>-species</code>: 物种名，Rice, Maize和others三个可选</li>
<li><code>-step</code>: 运行步骤,  <code>all|filter|final|anno</code>, 根据具体情况选择</li>
<li><code>-t</code>: 线程数，默认是4</li>
</ul>
<p>此外还有几个参数可以关注下</p>
<ul>
<li><p><code>-cds</code>: 提供已有的CDS序列（不能包括内含子和UTR），用于过滤.这个值也比较重要，建议提供下，否则会降低busco值（来自于作者的推荐）</p>
</li>
<li><p><code>-sensitive</code>: 是否用<code>RepeatModeler</code>分析剩下的TE，默认是0，也就是不要。<code>RepeatModeler</code>运行时间比较久，量力而信。</p>
</li>
<li><p><code>-anno</code>: 是否在构建TE文库后进行全基因组预测，默认是0.</p>
</li>
<li><p><code>-evalues 1</code>: 默认是0，需要同时设置-anno 1才能使用。建议加上，它能够查看注释质量，是非常不错的功能哦（来自于作者的推荐）<br>运行结束之后，会在当前目录下留下运行时的中间文件，保证你程序中断之后，能够断点续跑</p>
</li>
<li><p>xxx.EDTA.raw</p>
</li>
<li><p>xxx.EDTA.combine</p>
</li>
<li><p>xxx.EDTA.final</p>
</li>
</ul>
<p>以及你关注的<code>xxx.EDTA.TElib.fa</code>, 这就是最终的TE文库。</p>
<p><strong>需要注意</strong>的是，在实际运行的时候，你不能单条染色体的运行，这不是程序设计的目的，我们这里用一条染色体仅仅是为了演示，测试程序能否顺利运行。</p>
<p>而在实际项目中，一定要用所有的染色体或者scaffold</p>
<h2 id="可能问题"><a href="#可能问题" class="headerlink" title="可能问题"></a>可能问题</h2><p>我在使用EDTA时，就遇到了两个问题。一个是singularity的EDTA直接使用时无法通过依赖检测，解决方法已经在安装部分提过，这里不在赘述。</p>
<p>另一个问题我在”Identify TIR candidates from scratch”这一步出现下面的报错</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">what():  terminate called after throwing an instance of <span class="string">&#x27;Resource temporarily unavailable std::system_error&#x27;</span></span><br><span class="line">what():  Resource temporarily unavailable</span><br><span class="line">terminate called after throwing an instance of <span class="string">&#x27;std::system_error&#x27;</span></span><br></pre></td></tr></table></figure>

<p>我对这个报错进行了分析，找到了对应代码，即<code>sh $TIR_Learner -g $genome -s $species -t $threads -l $maxint</code>. 用实际内容替换变量后，即下面这行代码</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sh /EDTA/bin/TIR-Learner2.4/TIR-Learner2.4.sh -g chr.fa -s others -t 20 -l 5000</span><br></pre></td></tr></table></figure>

<p>更具体一点，可以将问题定位到脚本的<code>Module 3, Step 3: Get dataset</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">genomeFile=/data/xzg_data/1800_assembly/annotation/repeatAnnotation/chr.fa <span class="comment">#基因组文件的实际路径</span></span><br><span class="line">genomeName=TIR-Learner</span><br><span class="line">tir_path=/EDTA/bin/TIR-Learner2.4 <span class="comment"># TIR-Learner2.4的路径</span></span><br><span class="line">t=1</span><br><span class="line"><span class="built_in">dir</span>=`<span class="built_in">pwd</span>`</span><br><span class="line"><span class="built_in">export</span> OMP_NUM_THREADS=1</span><br><span class="line">python3 <span class="variable">$tir_path</span>/Module3_New/getDataset.py -g <span class="variable">$genomeFile</span> -name <span class="variable">$genomeName</span> -p <span class="variable">$tir_path</span> -t <span class="variable">$t</span> -d <span class="variable">$dir</span><span class="string">&quot;/Module3_New&quot;</span></span><br></pre></td></tr></table></figure>

<p>将线程数设置为1后，该代码顺利跑通。进一步，我定位<code>getDataset.py</code>的出问题的地方实际是<code>predict</code>函数。当然接着执行后续的代码，发现改动这一参数并不影响下面代码的运行。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">echo</span> <span class="string">&quot;Module 3, Step 4: Check TIR/TSD&quot;</span></span><br><span class="line">python3 <span class="variable">$path</span>/Module3_New/CheckTIRTSD_M3.py -name <span class="variable">$genomeName</span> -p <span class="variable">$path</span> -t <span class="variable">$t</span> -d <span class="variable">$dir</span><span class="string">&quot;/Module3_New&quot;</span></span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;Module 3, Step 5: Write to Gff&quot;</span></span><br><span class="line">python3 <span class="variable">$path</span>/Module3/WriteToGff_M3.py -name <span class="variable">$genomeName</span> -p <span class="variable">$path</span> -t <span class="variable">$t</span> -d <span class="variable">$dir</span><span class="string">&quot;/Module3_New&quot;</span></span><br></pre></td></tr></table></figure>

<p>我发现<code>predict</code>函数涉及到了Python的多进程调用，最终在偶然间找到问题真正所在，即Linux系统对用户的资源限制，可以通过<code>ulimit -a</code>查看。</p>
<p>最终我通过设置<code>ulimit -u 9000</code>，提高允许运行的总程序数，将问题解决。</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li><a target="_blank" rel="noopener" href="https://github.com/oushujun/EDTA">EDTA官方文档</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/oushujun/EDTA/issues/11">RMblast的问题</a></li>
<li><a target="_blank" rel="noopener" href="https://www.linuxquestions.org/questions/linux-server-73/bash-fork-retry-resource-temporarily-unavailable-limit-issue-4175618262/">一个关于fork资源不够的解决过程</a></li>
<li>Ou, S., Su, W., Liao, Y., Chougule, K., Agda, J.R.A., Hellinga, A.J., Lugo, C.S.B., Elliott, T.A., Ware, D., Peterson, T., et al. (2019). Benchmarking transposable element annotation methods for creation of a streamlined, comprehensive pipeline. Genome Biology 20, 275.</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/01/08/%E4%BD%BF%E7%94%A8EDTA%E8%BF%9B%E8%A1%8CTE%E6%B3%A8%E9%87%8A-annotate-te-with-edta-pipeline/" data-id="cllszh65400bbe2qe6cgi7b7v" data-title="使用EDTA进行TE注释" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何创造并尝试处理一个segmentation fault (core dumped)-how-to-create-and-solve-segementation-fault" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/01/08/%E5%A6%82%E4%BD%95%E5%88%9B%E9%80%A0%E5%B9%B6%E5%B0%9D%E8%AF%95%E5%A4%84%E7%90%86%E4%B8%80%E4%B8%AAsegmentation%20fault%20(core%20dumped)-how-to-create-and-solve-segementation-fault/" class="article-date">
  <time class="dt-published" datetime="2020-01-08T10:59:32.812Z" itemprop="datePublished">2020-01-08</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/01/08/%E5%A6%82%E4%BD%95%E5%88%9B%E9%80%A0%E5%B9%B6%E5%B0%9D%E8%AF%95%E5%A4%84%E7%90%86%E4%B8%80%E4%B8%AAsegmentation%20fault%20(core%20dumped)-how-to-create-and-solve-segementation-fault/">如何创造并尝试处理一个segmentation fault (core dumped)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p><code>segmentation fault (core dumped)</code>是一个曾经让我非常头疼的报错，当然它还有一种中文形式，**段错误(核心已转移)**。</p>
<p>这种错误可能会在各种程序中出现，每次我都用软件名加报错信息去搜索，对于一些常见软件，我能够找到一些解决方案，于是我能按照它们提供的方式进行解决。而对于一些比较小众的软件，就几乎无解了。</p>
<p>直到我学习「c和指针」，我自己创造出了一个段错误，我终于不再害怕这个报错（当然也仅限于不在害怕而已）。</p>
<p>下面这行代码，我觉得和<code>printf(&quot;hello world!\n&quot;)</code>一样重要。”hello world!”让你开始入门编程，而理解下面的代码则让你深入编程。</p>
<figure class="highlight c"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;stdio.h&gt;</span></span></span><br><span class="line"><span class="type">int</span></span><br><span class="line"><span class="title function_">main</span><span class="params">()</span></span><br><span class="line">&#123;</span><br><span class="line">    <span class="type">int</span> *a ;</span><br><span class="line">    *a = <span class="number">12</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>我们可以将其复制到到一个<code>error.c</code>中，然后编译运行</p>
<figure class="highlight c"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">gcc -o error error.c</span><br><span class="line">./error</span><br><span class="line">[<span class="number">2</span>]    <span class="number">110633</span> segmentation <span class="title function_">fault</span> <span class="params">(core dumped)</span> ./error</span><br></pre></td></tr></table></figure>

<p>在代码中，我们创建一个叫做a的指针变量，后面赋值语句则是把12存储在a所指向的内存地址，从逻辑上看好像没啥问题。其背后的问题，且听我慢慢道来。</p>
<p>a是一个指针变量，它和普通变量一样，它也有特定的内存地址，在那个内存递上也存放着值，该值是一个地址。对于一个普通变量而言，如果它们在声明的时候没有初始化，那么它们会有一个默认值，比如说<code>int i;</code>在我的机器上的结果是0。那么对于一个指针变量而言，如果没有在声明的时候没有初始化，它的值是什么呢？我用下面的代码测试了下</p>
<figure class="highlight c"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;stdio.h&gt;</span></span></span><br><span class="line"></span><br><span class="line"><span class="type">int</span></span><br><span class="line"><span class="title function_">main</span><span class="params">()</span>&#123;</span><br><span class="line">    <span class="type">int</span> i;</span><br><span class="line">    <span class="type">int</span> *a;</span><br><span class="line"></span><br><span class="line">    <span class="built_in">printf</span>(<span class="string">&quot;i:%d\n&quot;</span>,i);</span><br><span class="line">    <span class="built_in">printf</span>(<span class="string">&quot;a:%p\n&quot;</span>,a);</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>在三次运行过程中，i的值始终为0，而a的值则是0x7ffee20e7dc0, 0x7ffc1d202d10,0x7fff50972560.这些都对应着内存上的一个地址。对<code>*a</code>进行赋值的过程，也就是对它对应的内存地址进行赋值，这就带来了一个问题。如果这些内存是合法能被我们操作的，那么这个操作将会顺利进行，不会出现报错，但是可能会出现一些你完全想象不到的问题。但如果内存是受到保护或者对应地址不存在（大部分情况），那么在Linux系统上就会跳出错误。</p>
<p>现在我们知道了，段错误(core dumped)是程序访问&#x2F;使用它不能使用的内存地址所导致的，那么今后面对程序出现这个报错时，我们应该做呢？</p>
<ol>
<li>使用软件名 + 段错误对应的英文(core dumped)进行检索,  比如说 “STAR core dumped”, 相对而言，用英文检索的资料会比较</li>
<li>检查自己的服务器内存是否充足，对于一些比对软件或者组装软件而言，它们比较占用内存，因此内存不够就会出现错误</li>
<li>检查自己的输入是否合法，这可能是程序没有合理的错误捕捉机制，导致遇到错误格式的输入依旧运行了代码，结果中间指针没有正确初始化，最终出错</li>
<li>更新软件：新的软件可能会修复之前代码错误</li>
<li>换个软件：有些软件的版本可能就停留在它发表文章那一天，那么这个工具可能也不太可</li>
</ol>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/01/08/%E5%A6%82%E4%BD%95%E5%88%9B%E9%80%A0%E5%B9%B6%E5%B0%9D%E8%AF%95%E5%A4%84%E7%90%86%E4%B8%80%E4%B8%AAsegmentation%20fault%20(core%20dumped)-how-to-create-and-solve-segementation-fault/" data-id="cllszh67f00i6e2qebdpifhjw" data-title="如何创造并尝试处理一个segmentation fault (core dumped)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-我的2019年-my-2019" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/12/31/%E6%88%91%E7%9A%842019%E5%B9%B4-my-2019/" class="article-date">
  <time class="dt-published" datetime="2019-12-31T23:49:23.156Z" itemprop="datePublished">2019-12-31</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/12/31/%E6%88%91%E7%9A%842019%E5%B9%B4-my-2019/">我的2019年</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>当我开始写2019年总结的时候，我一时间不知道自己在这一年中做了些什么，总感觉自己忙忙碌碌，但是一无所获。</p>
<p>好在，我记录了，虽然不是以日记的形式，而是以一篇篇学习笔记的形式发布在网络上。当我将时间线拉回到2019年1月1日，然后慢慢向现在拉近时，我知道我这一年还是有一些进步的。</p>
<p>第一件值得开心的事，应该是我的个人博客xuzhougeng.top在7月25日的时候又上线了。这是我在2019年年初就想做的事情，我想要把自己简书上的内容完全由自己掌控，不再被无故404，目前已经有3万多人次访问。</p>
<p>其实搭建网站是我很小时候就想做的一件事情，那时候看到一篇文章写到用花生壳1元做自己的网站，就觉得这是一件很酷的事情。只不过无人指导，所以我一直在走弯路，但终于是把路走完了。</p>
<p>第二个事情，应该是我的编程水平的提高。我从年初就决定去学习数据结构和算法，只不过断断续续，最终也就是知道了大概，LeetCode上面的题目刷的还是不够。我尝试在R语言中使用Cpp提高代码运行速度，希望克服对<code>C++</code>的恐惧，从熟悉的语言切入慢慢掌握这门复杂的语言。我还挣扎在R语言绘图系统中，不断的看Y叔代码和ggplot2的源码，最终在翻译了ggproto相关内容后，对ggplot2和图形语法有所理解。</p>
<p>第三个应该是我的数据处理能力上的进步，因为踩了更多的坑，所以有了更多的项目经验 </p>
<ul>
<li>学习使用Nanopore做基因组组装，把相关的工具都学了一遍，Canu&#x2F;NECAT&#x2F;NextDenovo&#x2F;FLYE</li>
<li>学习用HiC辅助基因组组装，相关工具为ALLHiC&#x2F;3D-DNA&#x2F;SALSA</li>
<li>学习BioNano相关的知识点，发现了BioNano在处理杂合基因组上的不足，并且尝试去解决它。</li>
<li>从年初开始的miRNA-seq项目，终于在年末和师兄的不断讨论中，理解了原理，并发现之前存在的问题。</li>
<li>学了Shiny制作网页工具，用在了和师兄合作的单细胞文章数据展示上</li>
<li>在简化基因组鉴定SNP和使用SNP进行下游分析上，我还在不断的思考</li>
<li>学习了使用ANOVA进行QTL定位，当然还有质量性状的定位方法。也写了几篇BSA的教程。</li>
</ul>
<p>除了学习上和技术外，我还坚持锻炼，去健身房撸铁。这需要感谢我的是一个师弟，在他的督促下，我保证了每周起码两次的撸铁，否则我更喜欢在电脑前发呆。目前肚子上还有一层肉，在新的一年里可能要多跑跑步，少喝奶茶才行了。</p>
<!--最重要的一件事情，就是遇到了你，在iPhone制作的年度回忆里全是你，虽然相处的绝对时间上不多，但是每分每秒都很快乐-->

<p>今年差不多就是这些，对于明年，我估计也会做上面这些事情，不断的学习，不断的写作。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/12/31/%E6%88%91%E7%9A%842019%E5%B9%B4-my-2019/" data-id="cllszh68m00kne2qe8fpsfswo" data-title="我的2019年" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-服务器磁盘配额管理-server-quota" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/12/27/%E6%9C%8D%E5%8A%A1%E5%99%A8%E7%A3%81%E7%9B%98%E9%85%8D%E9%A2%9D%E7%AE%A1%E7%90%86-server-quota/" class="article-date">
  <time class="dt-published" datetime="2019-12-27T10:06:48.616Z" itemprop="datePublished">2019-12-27</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/12/27/%E6%9C%8D%E5%8A%A1%E5%99%A8%E7%A3%81%E7%9B%98%E9%85%8D%E9%A2%9D%E7%AE%A1%E7%90%86-server-quota/">服务器磁盘配额管理</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h2 id="硬盘配额"><a href="#硬盘配额" class="headerlink" title="硬盘配额"></a>硬盘配额</h2><p>Quota在是针对一个文件系统进行限制，而不是针对某个具体目录。也就是如果设置了<code>/dev/sda1</code>的quota，那么一个用户在这个<code>/dev/sda1</code>使用的数据量就会受限。如果你还有另外的<code>/dev/sda2</code>, 用户可以在超过<code>/dev/sda1</code>后继续使用<code>/dev/sda2</code>。</p>
<p>Quota针对整个文件系统的<strong>限制项目</strong>分为如下几个部分</p>
<ul>
<li>容量限制或文件数目限制(block或inode)</li>
<li>soft&#x2F;hard: 超过soft时，用户会被警告，但是依然能操作，但是一旦超过hard，那么直接无法操作硬盘</li>
<li>宽限时间: 超过soft后，给用户一定时间进行数据转移</li>
</ul>
<p>对于比较新的系统（例如CentOS7），建议在创建文件系统的时候就使用xfs，采用最新的quota策略。</p>
<h3 id="Ext4文件系统的Quota配置"><a href="#Ext4文件系统的Quota配置" class="headerlink" title="Ext4文件系统的Quota配置"></a>Ext4文件系统的Quota配置</h3><p>第一步。增加文件系统支持</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">mount | grep /data2</span><br><span class="line"><span class="comment"># dev/sdb2 on /data2 type ext4 (rw,relatime,seclabel,stripe=64,data=ordered)</span></span><br><span class="line">mount -o remount,usrquota,grpquota /data2</span><br><span class="line"><span class="comment"># /dev/sdb2 on /data2 type ext4 (rw,relatime,seclabel,quota,usrquota,grpquota,stripe=64,data=ordered)</span></span><br></pre></td></tr></table></figure>

<p>第二步: 扫描文件系统并新建Quota的配置文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">quotacheck -avug</span><br><span class="line"><span class="comment"># -a: 扫描所有在`/etc/mtab`内含有quota支持的文件系统</span></span><br><span class="line"><span class="comment"># -v: 显示扫描过程</span></span><br><span class="line"><span class="comment"># -u: 新建aquota.user，按照用户进行扫描记录</span></span><br><span class="line"><span class="comment"># -g: 新建aquota.group, 按照用户组进行扫描记录</span></span><br></pre></td></tr></table></figure>

<p>这一步可能出现的提示信息和解决方法</p>
<p>如果出现下面提示，说明内核支持xfs, 建议使用xfs的quota管理。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">quotacheck: Your kernel probably supports journaled quota but you are not using it. Consider switching to journaled quota to avoid running quotacheck after an unclean shutdown.</span><br></pre></td></tr></table></figure>

<p>如果出现下面提示，说明该磁盘正在独写，使用<code>lsof/fuser</code>查找可能的进程并关闭，之后重新运行。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">quotacheck: Cannot remount filesystem mounted on /data2 read-only so counted values might not be right.</span><br><span class="line">Please stop all programs writing to filesystem or use -m flag to force checking.</span><br></pre></td></tr></table></figure>

<p>下面的提示会在首次执行时，可以忽略</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">quotacheck: Scanning /dev/sdb1 [/data3] <span class="keyword">done</span></span><br><span class="line">quotacheck: Cannot <span class="built_in">stat</span> old user quota file /data3/aquota.user: No such file or directory. Usage will not be subtracted.</span><br><span class="line">quotacheck: Cannot <span class="built_in">stat</span> old group quota file /data3/aquota.group: No such file or directory. Usage will not be subtracted.</span><br><span class="line">quotacheck: Cannot <span class="built_in">stat</span> old user quota file /data3/aquota.user: No such file or directory. Usage will not be subtracted.</span><br><span class="line">quotacheck: Cannot <span class="built_in">stat</span> old group quota file /data3/aquota.group: No such file or directory. Usage will not be subtracted.</span><br><span class="line">quotacheck: Checked 44 directories and 23 files</span><br><span class="line">quotacheck: Old file not found.</span><br><span class="line">quotacheck: Old file not found.</span><br></pre></td></tr></table></figure>

<p>第三步： 启动&#x2F;关闭quota</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">quotaon -vug /data3</span><br><span class="line">quotaoff -ug /data3</span><br></pre></td></tr></table></figure>

<p>第四步: 编辑账户&#x2F;用户组的限值和宽限时间</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">edquota -u 用户名</span><br></pre></td></tr></table></figure>

<p>编辑表分为7个字段，修改的是第三个和第四个用于限制使用数据量(单位为KB)，修改第六个和第七个限制使用的文件数目(不常用)，0表示无限制。</p>
<p>和quota相关的Linux命令如下:</p>
<ul>
<li>quotactl</li>
<li>quotaon&#x2F;quotaoff</li>
<li>edquota</li>
<li>quotacheck</li>
<li>repquota</li>
<li>warnquota</li>
<li>setquota</li>
</ul>
<h2 id="XFS文件系统的Quota配置"><a href="#XFS文件系统的Quota配置" class="headerlink" title="XFS文件系统的Quota配置"></a>XFS文件系统的Quota配置</h2><p>如果使用xfs作为文件系统，那么就不能或者说没有必要采用上面的方法。xfs文件系统下，quota是文件系统的元信息，使用journaling提供更高层次的一致性保证，因此在管理上就会存在差别。</p>
<ol>
<li><code>quotacheck</code>对XFS系统没有效果。在mount时打开quota后，XFS会自动运行quotacheck做配额核算（quota accounting）</li>
<li>没有必要在XFS文件系统下创建quota文件。aquota.user, aquota.group</li>
<li>配额核算和限制增强之间存在区别。配额核算<strong>必须</strong>在mount的时候打开，但是限制增强可以在配额核算打开之后随意关闭和开启</li>
<li>建议使用state监控XFS quota系统</li>
</ol>
<p>挂载的时候打开quota</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">umount /mnt/disk1</span><br><span class="line">mount -o uquota,gquota /dev/sbd1 /mnt/disk1</span><br><span class="line"><span class="built_in">chmod</span> 1777 /mnt/disk1</span><br></pre></td></tr></table></figure>

<p>为了保证开机运行，还得修改<code>/etc/fstab</code>的第四列的<code>defaults</code>成<code>defaults,uquota,gquota</code></p>
<blockquote>
<p>ext4可以通过<code>fstransform</code>转换成xfs</p>
</blockquote>
<p>直接输入<code>xfs_quota</code>会进入交互模式</p>
<ul>
<li>help: 查看帮助信息, 例如help quota</li>
<li>quota: 查看配额信息</li>
<li>print: 列出设备</li>
<li>df&#x2F;free: 查看-bir(block,inode,realtime)信息，显示方式为 <code>-hN</code>,  <code>-f</code>指定文件</li>
<li>quit&#x2F;q： 退出交互模式</li>
</ul>
<p>使用<code>-x</code>会进入专家模式， 之后通过<code>-c &quot;命令&quot;</code>执行命令, 举几个例子</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 显示使用情况</span></span><br><span class="line">xfs_quota -x -c <span class="string">&#x27;report -ugibh&#x27;</span> /mnt/disk1</span><br><span class="line"><span class="comment"># 限制用户的文件数</span></span><br><span class="line">xfs_quota -x -c <span class="string">&#x27;limit -u isoft=5 ihard=10 user1&#x27;</span> /mnt/disk1</span><br><span class="line"><span class="comment"># 限制用户的数据量</span></span><br><span class="line">xfs_quota -x -c <span class="string">&#x27;limit -u bsoft=1m bhard=2m user1&#x27;</span> /mnt/disk1</span><br></pre></td></tr></table></figure>

<p>其他可用命令</p>
<ul>
<li>path: 显示设备</li>
<li>timer: 宽限时间，默认7天</li>
<li>enable&#x2F;disable: 开启或关闭配额增强</li>
<li>off: 永久性关闭</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/12/27/%E6%9C%8D%E5%8A%A1%E5%99%A8%E7%A3%81%E7%9B%98%E9%85%8D%E9%A2%9D%E7%AE%A1%E7%90%86-server-quota/" data-id="cllszh66c00ese2qe1w7wfl1l" data-title="服务器磁盘配额管理" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-注意编程语言中浮点运算-be-care-of-float-in-program-language" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/12/19/%E6%B3%A8%E6%84%8F%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80%E4%B8%AD%E6%B5%AE%E7%82%B9%E8%BF%90%E7%AE%97-be-care-of-float-in-program-language/" class="article-date">
  <time class="dt-published" datetime="2019-12-19T21:18:17.270Z" itemprop="datePublished">2019-12-19</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/12/19/%E6%B3%A8%E6%84%8F%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80%E4%B8%AD%E6%B5%AE%E7%82%B9%E8%BF%90%E7%AE%97-be-care-of-float-in-program-language/">注意编程语言中浮点运算</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>我们可能都知道，计算机以二进制的方式存储数字，举两个例子：</p>
<p>对于整型的125，在十进制里是1*100 + 2*10+ 5*1，而对应的二进制形式是1111101(64 + 32 + 16 + 8 + 4 + 0 + 1)。</p>
<p>对于浮点型的0.125，在十进制里是1&#x2F;10 + 2&#x2F;100 + 5&#x2F;1000，对应二进制下的0.001，即0&#x2F;2 + 0&#x2F;4 + 1&#x2F;8。</p>
<p>在不溢出的情况下，基本上所有十进制上的整型都能无损转成二进制。但是对于浮点型，就不能保证所有十进制下的小数都能无损转成二进制的表示方式。举个例子，假如你要在二进制下表示0.1(1&#x2F;10)，已知0.1小于我们之前的0.125，因此它不能是0.001, 那么它是0.0001吗？然而0.0001对应的是1&#x2F;16，也就是0.0625，比0.1小。0.00011对应0.9375， 稍微靠近了0.1，0.000111对应0.109375比0.1大，0.0001101对应0.1015625依旧比0.1大，0.00011001对应0.09765625又靠近了0.1。也就是说，0.1在二进制中这是一个无限循环小数。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="number">0.0001100110011001100110011001100110011001100110011</span>...</span><br></pre></td></tr></table></figure>

<p>这就类似于试图用小数形式描述十进制下的1&#x2F;3，只可能是0.3333333…。</p>
<p>循环是无限的，但是计算机存储是有限的，不能用有限的空间来存放无限的循环。因此对于不同编程语言，它们都会为不同的数据类型分配不同的内存大小。对于浮点数运算而言，基本上所有的编程语言都用的是IEEE 754标准。</p>
<p>例如Python而言的<a target="_blank" rel="noopener" href="https://www.python.org/dev/peps/pep-0754/">PEP 754</a>就是关于它的浮点型存放方式。在R里面用<code>?as.double</code>查看帮助文档时也能发现R也遵从IEEE 754。这里不在拓展介绍这个标准，只是为了说明我们只能用有限的空间来存储无限循环的小数。因此，大部分的小数都可能不是你看到的样子。大部分的编程语言，其实都会在浮点型运算结果中给你体现出这种差异，比方说Python</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="number">0.1</span> + <span class="number">0.2</span></span><br><span class="line"><span class="number">0.30000000000000004</span></span><br></pre></td></tr></table></figure>

<p>看到这个结果，你就会好奇为啥0.1+0.2的结果不是0.3，不会好奇为啥03不等于0.3。但是在R语言中，你会好奇为啥0.3不等于0.3， 因为明明看起来一样啊</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> <span class="number">0.1</span> <span class="operator">+</span> <span class="number">0.2</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span> <span class="number">0.3</span></span><br><span class="line"><span class="operator">&gt;</span> <span class="punctuation">(</span><span class="number">0.1</span><span class="operator">+</span><span class="number">0.2</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="number">0.3</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span> <span class="literal">FALSE</span></span><br></pre></td></tr></table></figure>

<p>如果你想看到它的实际值，就需要借助于<code>sprint</code>这个函数了</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sprintf<span class="punctuation">(</span><span class="string">&quot;%.20f&quot;</span><span class="punctuation">,</span> <span class="number">0.1</span> <span class="operator">+</span> <span class="number">0.2</span> <span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span> <span class="string">&quot;0.30000000000000004441&quot;</span></span><br></pre></td></tr></table></figure>

<p>因此，在R语言做浮点运算时，一定要谨慎，可以考虑用Hadley开发的<code>dplyr</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>dplyr<span class="punctuation">)</span></span><br><span class="line">near<span class="punctuation">(</span><span class="number">0.1</span><span class="operator">+</span><span class="number">0.2</span><span class="punctuation">,</span> <span class="number">0.3</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/12/19/%E6%B3%A8%E6%84%8F%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80%E4%B8%AD%E6%B5%AE%E7%82%B9%E8%BF%90%E7%AE%97-be-care-of-float-in-program-language/" data-id="cllszh66e00eze2qecucycufu" data-title="注意编程语言中浮点运算" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-单细胞测序数据整合中的批次效应-batch-effect-in-single-cell-rna-seq" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/12/18/%E5%8D%95%E7%BB%86%E8%83%9E%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88%E4%B8%AD%E7%9A%84%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94-batch-effect-in-single-cell-rna-seq/" class="article-date">
  <time class="dt-published" datetime="2019-12-18T15:15:11.455Z" itemprop="datePublished">2019-12-18</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/12/18/%E5%8D%95%E7%BB%86%E8%83%9E%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88%E4%B8%AD%E7%9A%84%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94-batch-effect-in-single-cell-rna-seq/">单细胞测序数据整合中的批次效应</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>在数据分析的时候，我们的目标是找到样本之间真实的生物学差异。但是这种真实的生物学因素往往会受到各种因素影响，举几个场景</p>
<ul>
<li>不同样本</li>
<li>同一样本的生物学重复</li>
<li>同一样本的技术重复</li>
<li>同一样本在同一个实验室由同一团队在不同时间点处理</li>
<li>同一细胞系&#x2F;小鼠在不同实验室</li>
<li>不同建库策略，10X平台，Drop-seq, SMART2-seq</li>
<li>不同测序平台，BGI&#x2F;Illumina</li>
<li>不同分析流程（甚至一个工具的多个版本，如salmon，CellRanger）</li>
</ul>
<p>这些因素之间有些是生物学真实的差异，有些是抽样时的随机波动。有些是系统性因素，比如说批次效应(batch effect)。</p>
<p>从直觉上讲，最好是分析的过程中过滤掉这种批次效应。但是即便是同一个人对同一个样本做的相同实验，也有可能因为时间差异导致批次效应，我们需要对这种数据集进行批次效应校正吗？我们对批次效应进行校正的同时也会引入新的问题，它很有可能将生物学本身的差异视为批次效应，然后将其去除。因此解决同一种实验的批次效应最好的方法就是搞一套比较好的实验设计（例如将样本分开在不同的实验批次中，(Kang et al., 2018)）。但是受限于实验条件，基本上做不到这些，而且这类批次效应可能也没有那么大的影响。</p>
<p>我们更关注的一类批次效应，其实是不同实验室，不同建库手段，不同测序平台所引起的批次效应。当我们希望通过合并同一组织数据挖掘出更有意义的信息时，就不可避免的会发现，明明是同个组织的数据，表达量就是存在明显的差异(PCA, t-SNE降维可视化)。</p>
<p>之前有人用bulk RNA-seq的方法（limma, ComBat,RUVseq, svaseq）对单细胞数据进行校正，但是这些工具的基本假设都是”bulk RNA-seq数据中的细胞组成相似”，可能适用于一些数据集，但是可推广性不强(Haghverdi et al., 2018)。于是就有一些专门用于单细胞转录组批次校正的工具，这里对这些软件做一个罗列</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://satijalab.org/seurat/v3.1/integration.html">Seurat&#x2F;Integration</a></li>
<li><a target="_blank" rel="noopener" href="https://bioconductor.org/packages/3.11/bioc/html/batchelor.html">batchelor&#x2F;FastMNN</a></li>
<li><a target="_blank" rel="noopener" href="https://bioconductor.org/packages/release/bioc/html/scran.html">scran&#x2F;mnnCorrect</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/Teichlab/bbknn">Python&#x2F;BBKNN</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/jumphone/BEER">BBER</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/hms-dbmi/conos">Conos</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/MacoskoLab/liger">LIGER</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/immunogenomics/harmony">Harmony</a></li>
</ul>
<p>最后对于不同的实验设计，仅从主观上推荐的批次矫正方法</p>
<ul>
<li>技术重复: ComBat</li>
<li>细胞系，生物学重复: ComBat</li>
<li>同一个人的癌症和癌旁组织: 不校正&#x2F;Harmony</li>
<li>不同实验室的同一组织: Harmony</li>
<li>同一个实验室做的不同人的样本: 不校正&#x2F;Harmony</li>
</ul>
<blockquote>
<p>Harmony就是给soft k-mean（GMM）加了一个KL divergence，然后这个惩罚项最小化可以使得不同batch的cells尽可能分配到每个cluster内部，然后可以给每个cell分配一个到cluster的probability，那么每个cluster都会有一个和细胞数维度一样的细胞分配到这个cluster的probability，其实就是后续一个（batch number+1）维度的向量到embedding出来向量的一个回归方程的样本权（权重线性回归），简单而言它认为将不同batch的cell embedding的向量是由batch的variation和cell type的variation所决定，只要把cell type这部分找到（其实就是残差）找到就可以去除batch的影响。但是要定义样本在每个cluster里面的entropy，而且要使这个值尽可能大，所以我不知道这个文章有没有做rare celltype相关的一些测试<br>-@汪伟旭(复旦大学) </p>
</blockquote>
<p>对于批次效应的检查，可以试试<a target="_blank" rel="noopener" href="https://github.com/theislab/kBET">kBET</a>，但是它的内存和运算时间和数据集大小密切相关，很有可能跑不下去。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/12/image-d3353ffbe55846a0acf207d2233ab1ec.png" alt="批次效应"></p>
<h2 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h2><ul>
<li>Haghverdi, L., Lun, A.T.L., Morgan, M.D., and Marioni, J.C. (2018). Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors. Nat Biotechnol 36, 421–427.</li>
<li>Büttner, M., Miao, Z., Wolf, F.A., Teichmann, S.A., and Theis, F.J. (2019). A test metric for assessing single-cell RNA-seq batch correction. Nat Methods 16, 43–49.</li>
<li>Kang, H.M., Subramaniam, M., Targ, S., Nguyen, M., Maliskova, L., McCarthy, E., Wan, E., Wong, S., Byrnes, L., Lanata, C.M., et al. (2018). Multiplexed droplet single-cell RNA-sequencing using natural genetic variation. Nat Biotechnol 36, 89–94.</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/12/18/%E5%8D%95%E7%BB%86%E8%83%9E%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88%E4%B8%AD%E7%9A%84%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94-batch-effect-in-single-cell-rna-seq/" data-id="cllszh66600e6e2qe5pssf4i4" data-title="单细胞测序数据整合中的批次效应" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-Seurat的normalization和scaling-seurat-normalization-and-scaling" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/12/17/Seurat%E7%9A%84normalization%E5%92%8Cscaling-seurat-normalization-and-scaling/" class="article-date">
  <time class="dt-published" datetime="2019-12-17T16:07:37.785Z" itemprop="datePublished">2019-12-17</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/12/17/Seurat%E7%9A%84normalization%E5%92%8Cscaling-seurat-normalization-and-scaling/">Seurat的normalization和scaling</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>Seurat的分析流程有两步, 对数据的normalization和scaling. 两种的作用不同，前者是为了处理每个细胞的总count不同的问题，而后者则是让每个基因的表达量的均值为0，方差为1.</p>
<p>normlization对应的函数是<code>NormalizeData</code>，通过数据进行一些列变换，消除<strong>文库大小</strong>的影响。 它有三种方法, LogNormalize, CLR, RC</p>
<p>默认方法方式是<strong>LogNormalize</strong>， 即对于每个细胞，将每个基因的count除以总数，然后乘以一个scale.factor, 之后以自然对数进行转换。为了提高效率，Seurat编写了<code>C++</code>代码用于加速</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// [[Rcpp::export]]</span></span><br><span class="line"><span class="function">Eigen::SparseMatrix&lt;<span class="type">double</span>&gt; <span class="title">LogNorm</span><span class="params">(Eigen::SparseMatrix&lt;<span class="type">double</span>&gt; data, <span class="type">int</span> scale_factor, <span class="type">bool</span> display_progress = <span class="literal">true</span>)</span></span>&#123;</span><br><span class="line">  <span class="function">Progress <span class="title">p</span><span class="params">(data.outerSize(), display_progress)</span></span>;</span><br><span class="line">  Eigen::VectorXd colSums = data.<span class="built_in">transpose</span>() * Eigen::VectorXd::<span class="built_in">Ones</span>(data.<span class="built_in">rows</span>());</span><br><span class="line">  <span class="keyword">for</span> (<span class="type">int</span> k=<span class="number">0</span>; k &lt; data.<span class="built_in">outerSize</span>(); ++k)&#123;</span><br><span class="line">    p.<span class="built_in">increment</span>();</span><br><span class="line">    <span class="keyword">for</span> (Eigen::SparseMatrix&lt;<span class="type">double</span>&gt;::InnerIterator <span class="built_in">it</span>(data, k); it; ++it)&#123;</span><br><span class="line">      it.<span class="built_in">valueRef</span>() = <span class="built_in">log1p</span>(<span class="built_in">double</span>(it.<span class="built_in">value</span>()) / colSums[k] * scale_factor);</span><br><span class="line">    &#125;</span><br><span class="line">  &#125;</span><br><span class="line">  <span class="keyword">return</span> data;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>也就是说，如果你不追求效率，我们是可以用纯R代码实现。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">mat <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span>data <span class="operator">=</span> rbinom<span class="punctuation">(</span>n <span class="operator">=</span> <span class="number">25</span><span class="punctuation">,</span> size <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> prob <span class="operator">=</span> <span class="number">0.2</span><span class="punctuation">)</span><span class="punctuation">,</span> nrow <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br><span class="line">mat_norm <span class="operator">&lt;-</span> LogNormalize<span class="punctuation">(</span>data <span class="operator">=</span> mat<span class="punctuation">)</span></span><br><span class="line"><span class="comment"># LogNormalize等价于</span></span><br><span class="line">log1p<span class="punctuation">(</span>t<span class="punctuation">(</span>t<span class="punctuation">(</span>mat<span class="punctuation">)</span> <span class="operator">/</span> colSums<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">*</span> <span class="number">10000</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>Scale这一步对应的函数是<code>ScaleData</code>， 在它处理之后，使得每个基因在所有样本的均值是0，而方差是1。这让每个基因在下游分析中的具有相同的权重，使得高表达基因不那么显著。</p>
<p>而如果需要移除数据集中不需要的变异来源(unwanted sources of variation),  <code>ScaleData</code>需要设置额外的参数<code>vars.to.regress</code>，但是作者更加推荐使用<code>SCTransform</code>。</p>
<p>尽管<code>ScaleData</code>对应的源代码非常的长，但是和数据处理相关的就是如下这几条</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Currently, RegressOutMatrix will do nothing if latent.data = NULL</span></span><br><span class="line">data.scale <span class="operator">&lt;-</span> scale.function<span class="punctuation">(</span></span><br><span class="line">  mat <span class="operator">=</span> object<span class="punctuation">[</span>features<span class="punctuation">[</span>block<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span><span class="operator">:</span>block<span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">,</span> split.cells<span class="punctuation">[[</span>group<span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">,</span> drop <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">  scale <span class="operator">=</span> do.scale<span class="punctuation">,</span></span><br><span class="line">  center <span class="operator">=</span> do.center<span class="punctuation">,</span></span><br><span class="line">  scale_max <span class="operator">=</span> scale.max<span class="punctuation">,</span></span><br><span class="line">  display_progress <span class="operator">=</span> <span class="literal">FALSE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>其中<code>scale.function</code>是根据数据类型来决定是<code>FastRowScale</code>还是<code>FastSparseRowScale</code>, 而这两个代码Seurat也是通过<code>C++</code>进行提速了。因此，如果不追求效率，还是可以用纯R代码实现。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">pbmc_small <span class="operator">&lt;-</span> ScaleData<span class="punctuation">(</span>pbmc_small<span class="punctuation">)</span></span><br><span class="line">GetAssayData<span class="punctuation">(</span>pbmc_small<span class="punctuation">,</span> <span class="string">&quot;scale.data&quot;</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;IGLL5&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">2</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># 等价于</span></span><br><span class="line">scaled_data <span class="operator">&lt;-</span> t<span class="punctuation">(</span>scale<span class="punctuation">(</span>t<span class="punctuation">(</span>GetAssayData<span class="punctuation">(</span>pbmc_small<span class="punctuation">,</span> <span class="string">&quot;data&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">scaled_data<span class="punctuation">[</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;IGLL5&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">2</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<blockquote>
<p>默认R语言的<code>scale</code>是按列处理，对于行为基因，列为样本的数据，可以直接套用。但是Seurat的输入数据是行为基因，列为样本，那么就需要按行scale。换句话说，如果你需要对行进行scale，那么你可以通过<code>Seurat:::FastRowScale()</code>的方式调用Seurat写的<code>FastRowScale</code>函数。此外Seurat编写了大量的<code>Fast*</code>函数，都可以尝试用在代码中。</p>
</blockquote>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/12/17/Seurat%E7%9A%84normalization%E5%92%8Cscaling-seurat-normalization-and-scaling/" data-id="cllszh63c004te2qe0i932rxa" data-title="Seurat的normalization和scaling" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-写在2018年的第一天-写在2018年的第一天" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/12/17/%E5%86%99%E5%9C%A82018%E5%B9%B4%E7%9A%84%E7%AC%AC%E4%B8%80%E5%A4%A9-%E5%86%99%E5%9C%A82018%E5%B9%B4%E7%9A%84%E7%AC%AC%E4%B8%80%E5%A4%A9/" class="article-date">
  <time class="dt-published" datetime="2019-12-17T09:04:52.856Z" itemprop="datePublished">2019-12-17</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/12/17/%E5%86%99%E5%9C%A82018%E5%B9%B4%E7%9A%84%E7%AC%AC%E4%B8%80%E5%A4%A9-%E5%86%99%E5%9C%A82018%E5%B9%B4%E7%9A%84%E7%AC%AC%E4%B8%80%E5%A4%A9/">写在2018年的第一天</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h2 id="写在2018年的第一天"><a href="#写在2018年的第一天" class="headerlink" title="写在2018年的第一天"></a>写在2018年的第一天</h2><p>先说一件比较尴尬的事，昨天晚上肚子不舒服，于是蹲了好久，等出来才发现一年就这样过去了。时间节点，如情人节、除夕夜，自己的生日，对我而言并没有多大意义（女朋友的生日除外）。</p>
<p>我们并不需要专门找一天来下定决心重新开始，如果这件事情非常重要你早已经开始，这件事情无关紧要你就会一推再推。但是我们也的确需要这么一天来回顾过去，展望未来。大脑需要一种仪式感，一种暗示，或者说是一种欺骗吧。</p>
<h2 id="我的2017年"><a href="#我的2017年" class="headerlink" title="我的2017年"></a>我的2017年</h2><p>2017年刚开始时，我还是研一新生，还在迷惘未来何去何从。当时的情况是一轮轮转的老板要离开原单位去其他学校，先是告诉我可以继续留下来，后来却否定了这一可能性。二轮的老板看我情况太复杂，也不知道该不该留我，让我和所里继续抗争着。后来我去做第三轮，跑到一个院士下面说搞生信，院士让我先留下学习这2周。实际情况是，我不太懂如何和别人打交道，一直觉得自己格格不入。他们的主力基本走完了，留下的工作人员大部分工作是在改图，原本负责带我的老师正在准备投递文章。我每天早点7点30出发，挤地铁过去，边学边翻译bioconductor的教程。由于那边只供应午饭，所以每天晚上7点左右就回来，半路点外卖，到寝室开吃，然后晚上和大学同学打打DOTA2。有一天，我听着《春风十里》，出了地铁口，看着夜空，吹过的峰让我又紧了紧外衣，不知何去何从。最后我没有在院士那里留下来，我知道问题在我。我得感谢现在的老板，给了我一个去处，又因为实验室还不够大，所以让我继续去自学生信。</p>
<p>为了搞定师姐的一个EMS诱变群体的基因定位项目，我接触了生信媛的幕后大大，也是一位自学生信的师姐，开始在生信媛发我的学习笔记，成为生信媛的一名小编。在2017年春季的某一天，生信媛的大部分小伙伴跑去幕后大大的家里蹭饭，通过那次聚会中我才认识了那些数字信息后的物理意义上的人。或许是因为我稳定的更新频率，或许是我把统计学的作业发在生信媛，生信媛的订阅迎来了第一个1000人。后来也没有刻意转发到各种生信相关群里，每天增加的人也特别的稳定，到10月份达到了5000人。</p>
<p>当然生信媛的微信推送文章一般都会同步到生信技能树论坛，也就顺理成章加入生信技能树论坛的作者群。某一天在承包下生信工作资讯速递这个栏目后，我接触了生信技能树的Jimmy。又由于自己写的比较勤快，让Jimmy看到自己年轻时候的样子，于是把我拉入了生信技能树公众号的小编群，认识了更多的人，如果子、思考问题的熊、Panda姐等。Jimmy有很多想法，比如说写书，做视频课，搞论坛，结果还是先开展了武汉的线下培训。由于线下培训，我又认识了一批人，比如说武汉的线下团队。并且还让Jimmy从澳门给我带来了一台macbook pro，圆了我多年的梦想，也感谢我那个土豪师兄借钱。</p>
<p>因此，在物理世界中，我不懂如何和陌生人打交道。在网络世界中，其实也没有好到哪里去。我唯一交流的方式，就是把自己的学习过程记录下来，写成笔记发到互联网上，当别人看到我的文字时，就知道我差不多是什么样的人。我和陌生人的寒暄方式就是提问和解答，有时候我问别人，有时候别人问我。如果有幸能够持续交流下去，那么在我的世界里就多了一个认识的人。如果这些人还刚好能在现实世界里碰面，那么我或许就会多几个朋友。</p>
<h2 id="我的2018年目标"><a href="#我的2018年目标" class="headerlink" title="我的2018年目标"></a>我的2018年目标</h2><p>在2017年的年初时候，我曾经许下很多目标</p>
<ul>
<li>得到的通往财富自由之路每周总结</li>
<li>天天用英语每日笔记</li>
<li>读书做笔记</li>
<li>做完200道生信编程题</li>
<li>跟随jimmy的直播我的基因组</li>
</ul>
<p>实际情况是这些目标一个我都没有完成。当然这可以从《意志力》这本书找到答案，设下的目标越多，意志力在决定目标的时候的消耗就越多，于是目标就很不可能完成。于是2018年我就放弃制定那么多目标了，我仅仅只有一个目标，就是在2018年争取发一篇文章，发在bioinformatics上。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/12/17/%E5%86%99%E5%9C%A82018%E5%B9%B4%E7%9A%84%E7%AC%AC%E4%B8%80%E5%A4%A9-%E5%86%99%E5%9C%A82018%E5%B9%B4%E7%9A%84%E7%AC%AC%E4%B8%80%E5%A4%A9/" data-id="cllszh66400dwe2qe7bdf8ap3" data-title="写在2018年的第一天" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-从typora转移到VNOTE-typora-to-vnote" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/12/16/%E4%BB%8Etypora%E8%BD%AC%E7%A7%BB%E5%88%B0VNOTE-typora-to-vnote/" class="article-date">
  <time class="dt-published" datetime="2019-12-16T15:39:13.562Z" itemprop="datePublished">2019-12-16</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/12/16/%E4%BB%8Etypora%E8%BD%AC%E7%A7%BB%E5%88%B0VNOTE-typora-to-vnote/">从typora转移到VNOTE</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>因为之前一直用的是Typora进行写作，笔记管理的方式就是新建文件夹，也是下图这种情况。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/12/image-254b1d8ed925480f9163a5c80f7cd0c1.png" alt="笔记本目录"></p>
<p>从轻度写作而言，Typora基本上够用了，但是它并不是一个笔记管理工具。熊的原话是，就本地文件的管理而言，Typora和VNote之间相差一到两个印象笔记。并且从LTF(List, Tag, Filter)角度来看，我只有list，而没有tag，更不要说filter了，因此查找笔记全靠记忆或者靠搜索引擎（我的学习笔记基本上都发布在互联网了）。</p>
<p>为了减少我转移笔记的手工操作，我分析了下VNOTE的笔记管理模式，发现核心文件就两个</p>
<ul>
<li>_v_imaes: 用于存放本地图片</li>
<li>_vnote.json: 用于记录文件信息</li>
</ul>
<p>由于并且我的图片存储习惯和VNOTE一致，也就是为不同笔记本建立单独的图片目录，我的是assets，而VNote是<code>_v_images</code>，因此我在添加笔记本的时候，只做了如下简单的更改。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/12/image-6f2c53dfb35c4e31ab182b92590393de.png" alt="修改配置"></p>
<p>最终结果如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/12/image-d7b5eabbe57d4bc598fbc110d2b1df94.png" alt="最终效果"></p>
<p>因此，对于使用相对路径的我而言，我并不是从typora转移到VNote，而只是让我的之前的笔记本能够被VNote管理而已。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/12/16/%E4%BB%8Etypora%E8%BD%AC%E7%A7%BB%E5%88%B0VNOTE-typora-to-vnote/" data-id="cllszh64h008te2qehw27fhr0" data-title="从typora转移到VNOTE" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/typora/" rel="tag">typora</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/13/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/12/">12</a><a class="page-number" href="/page/13/">13</a><span class="page-number current">14</span><a class="page-number" href="/page/15/">15</a><a class="page-number" href="/page/16/">16</a><span class="space">&hellip;</span><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/15/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
          <li>
            <a href="/2023/06/22/%E5%9C%A8Windows%E4%B8%AD%E6%90%AD%E5%BB%BA%E4%B8%80%E4%B8%AA%E5%9F%BA%E4%BA%8EWSL%E7%9A%84Python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6%E7%8E%AF%E5%A2%83-building-a-wsl-based-python-data-science-environment-in-windows/">在Windows中搭建一个基于WSL的Python数据科学环境</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>