<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/16/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-PBMC简介-A-Brief-Review-of-PBMC" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/11/25/PBMC%E7%AE%80%E4%BB%8B-A-Brief-Review-of-PBMC/" class="article-date">
  <time class="dt-published" datetime="2019-11-25T15:20:32.709Z" itemprop="datePublished">2019-11-25</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/11/25/PBMC%E7%AE%80%E4%BB%8B-A-Brief-Review-of-PBMC/">PBMC简介</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>外周血单核细胞(Peripheral blood monoculear cell, PBMC), 包括淋巴细胞(T细胞，B细胞和自然杀伤(NK)细胞)和单核细胞。而红细胞和血小板没有细胞核，而粒细胞(granulocytes) 包括中性粒细胞、嗜碱性粒细胞和嗜酸性粒细胞，有多叶核，所以不包括在PBMC中。</p>
<p>人类的PBMC主要由淋巴细胞组成，之后是单核细胞以及一小部分的树突状细胞(dendritic cells， DC).</p>
<p>PBMC的提取方法: 利用Ficoll将全血分层，之后进行梯度离心， 血液就会分离成一层血浆，然后是一层PBMCs和一层多形核细胞(polymorphonuclear )，如中性粒细胞和嗜酸性粒细胞，和底层红细胞。通过溶解红细胞可以进一步分离多形核细胞。嗜碱性粒细胞有时出现在密度较高的组分中和PBMC组分中。</p>
<p><strong>注</strong>: Ficoll: 一种亲水性多糖(hydrophillic  polysaccharide )</p>
<h2 id="PBMC的细胞发育过程"><a href="#PBMC的细胞发育过程" class="headerlink" title="PBMC的细胞发育过程"></a>PBMC的细胞发育过程</h2><p> PBMC起源于位于骨髓(bone marrow)的造血干细胞( hematopoietic stem cells, HSCs)。在造血干细胞的造血过程中，它们会产生髓系细胞（单核细胞、巨噬细胞、粒细胞、巨核细胞、树突状细胞、红细胞）和淋巴系细胞（T细胞、B细胞、NK细胞）。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/AdobeStock_171095311-1-d3a4c9537baf4ffdb85cc1acaf1bd5f2.jpeg" alt="peripheral blood mononuclear cells hematopoiesis"></p>
<p>图解:</p>
<ul>
<li>Haematopoietic stem cell: 造血干细胞</li>
<li>common myeloid progenitor: 髓系祖细胞</li>
<li>common lymphoid progenitor: 淋巴祖细胞</li>
<li>myeloblast: 成髓细胞, 骨髓母细胞</li>
<li>Lymphoblast: 成淋巴细胞, 淋巴母细胞</li>
<li>lymphocytes: 淋巴细胞</li>
<li>erythrocytes: 红细胞</li>
<li>plateletes: 血小板</li>
<li>basophils: 嗜碱性粒细胞</li>
<li>neutrophils: 中性粒细胞</li>
<li>eosinophils: 嗜酸性粒细胞</li>
<li>monocytes: 单核细胞</li>
</ul>
<h2 id="人类PBMC的不同细胞类型的比例"><a href="#人类PBMC的不同细胞类型的比例" class="headerlink" title="人类PBMC的不同细胞类型的比例"></a>人类PBMC的不同细胞类型的比例</h2><p>PBMC中大部分都是淋巴细胞，包括B细胞和T细胞，其中CD3+ T细胞又占了淋巴细胞中绝大部分(45-70%)。这些T细胞都处于初始(naive)状态，即已经成熟了但没有受到抗原刺激。 在正常人中，只有非常小的一部分初始T细胞会因抗原识别而被激活。同样的B细胞也都处于初始状态。</p>
<p>相对于淋巴细胞，单核细胞的比例在10-30%，收到刺激之后会发育成树突状细胞(DC)或巨噬细胞。</p>
<p>干细胞的比例非常低，只有0.1-0.2%，因此很难从全血样本中分离到。分子标记是CD34+。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/PBMC-Frequency-99b5863ef1af4b06bddfa276a4c394db.jpg" alt="Frequency of Peripheral Blood Mononuclear Cells"></p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><p> <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Peripheral_blood_mononuclear_cell">https://en.wikipedia.org/wiki/Peripheral_blood_mononuclear_cell</a> </p>
<p> <a target="_blank" rel="noopener" href="https://www.stemexpress.com/blogs/peripheral-blood-mononuclear-cells/">https://www.stemexpress.com/blogs/peripheral-blood-mononuclear-cells/</a> </p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/11/25/PBMC%E7%AE%80%E4%BB%8B-A-Brief-Review-of-PBMC/" data-id="clm1z22fg008aehn58vfi128k" data-title="PBMC简介" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如果你需要我回答你的问题-how-to-ask-a-question" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/11/22/%E5%A6%82%E6%9E%9C%E4%BD%A0%E9%9C%80%E8%A6%81%E6%88%91%E5%9B%9E%E7%AD%94%E4%BD%A0%E7%9A%84%E9%97%AE%E9%A2%98-how-to-ask-a-question/" class="article-date">
  <time class="dt-published" datetime="2019-11-22T15:00:51.987Z" itemprop="datePublished">2019-11-22</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/11/22/%E5%A6%82%E6%9E%9C%E4%BD%A0%E9%9C%80%E8%A6%81%E6%88%91%E5%9B%9E%E7%AD%94%E4%BD%A0%E7%9A%84%E9%97%AE%E9%A2%98-how-to-ask-a-question/">如果你需要我回答你的问题</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>我的生物信息学技能基本上都是自学的，在刚开始的阶段我也和大部分的人一样，会遇到很多问题，然后想着去一些群里提问希望得到解答，当然结果也是和大部分人的一样，问题不会得到答案。</p>
<p>后来，我通过看书和搜索慢慢的有了一些基础，通过项目实践对生信有了一点了解，加上自己也喜欢把自己的学习笔记公开出来，就逐渐有人开始向我提问。有些问题我会想去回答，而有些问题会被我直接忽视。如同当年我提出的问题一样，有些能够有人解答，而有些则石沉大海。我稍微的总结了一些，如果你希望让我或者别人回答你的问题，或许可以参考一下。</p>
<h2 id="我不回答能够被搜索到的问题"><a href="#我不回答能够被搜索到的问题" class="headerlink" title="我不回答能够被搜索到的问题"></a>我不回答能够被搜索到的问题</h2><p>以我自己为例吧，我每天都会遇到很多问题，大部分的问题可以通过直接复制报错搜索的方式解决。有一些问题不是普遍性问题，我会通过提取报错中的关键信息进行检索。由于搜索的次数多了，我看到别人提问的时候，我差不多就能判断出这个问题需要经过几次搜索才能找到答案。对于能够通过简单搜索找到答案的问题，我就无视掉了。当然如果你在提问中指出你做了哪些搜索上的尝试，那么我会提供一些关键字，或者我会帮你去搜索下，如果发现是软件本身写的太差导致搜索不到的问题，那么我会建议你换个软件。</p>
<h2 id="我不回答我不擅长的问题"><a href="#我不回答我不擅长的问题" class="headerlink" title="我不回答我不擅长的问题"></a>我不回答我不擅长的问题</h2><p>术业有专攻，没有经验就贸然回答，也是对提问人的不负责。比方说，我没有任何芯片数据处理的经验，也没有挖掘过GEO数据库或TCGA数据库。因此，当你问我如何合并不同平台的芯片数据时，我也只会说我不会，这不是推脱，是真的不会。</p>
<h2 id="我不回答我之后遇不到的问题"><a href="#我不回答我之后遇不到的问题" class="headerlink" title="我不回答我之后遇不到的问题"></a>我不回答我之后遇不到的问题</h2><p>比方说一些工具类的报错，这些工具要么是我不会用到，要么是这个问题是某些平台所特有的，我很难重复出这个问题，那么我也不会花很多时间去重复这个问题。如果我以后遇到了这个问题，那么我会去解决，并给出解决方案。但是当下并不是我所在乎的，所以我就不会管它。</p>
<h2 id="我喜欢在邮件上回答问题，而不是即时通讯类工具（例如微信）"><a href="#我喜欢在邮件上回答问题，而不是即时通讯类工具（例如微信）" class="headerlink" title="我喜欢在邮件上回答问题，而不是即时通讯类工具（例如微信）"></a>我喜欢在邮件上回答问题，而不是即时通讯类工具（例如微信）</h2><p>因为我一般会用手机看微信，如果我要回答你的问题，我就需要用手机打字。但问题手机的输入不如电脑的方便，而且容易出错，我会觉得很麻烦，就懒回答问题了。而我收邮件一般在电脑端，那么我用键盘打字就会很舒畅，当有些问题我也不太会的时候，也很容易搜索，我就不会感觉回答问题成为了一个负担，那么就会倾向于回答问题。</p>
<h2 id="我喜欢回答一句话或者几个字能够解决的问题"><a href="#我喜欢回答一句话或者几个字能够解决的问题" class="headerlink" title="我喜欢回答一句话或者几个字能够解决的问题"></a>我喜欢回答一句话或者几个字能够解决的问题</h2><p>这类问题一个特征就是，它是一个选择题而不是一道简答题或者论述题。而且这类问题的描述一般会比较多，把问题的背景和自己的解答过程都写了出来，然后问你这样子做，合理吗？如果我觉得没问题，就只要回复一个“没问题”。如果我觉得有问题，那么我还会指出哪里不对，应该如何做。</p>
<h2 id="我会回答提供了测试数据集的问题"><a href="#我会回答提供了测试数据集的问题" class="headerlink" title="我会回答提供了测试数据集的问题"></a>我会回答提供了测试数据集的问题</h2><p>以R语言为例吧，当你在某个问题出错时，如果能把代码和数据集都整理成一个压缩包，并且在出问题的地方，还加上提示，那么我就能很容易定位到问题，从而解决问题。</p>
<p>还有一类提问方式，就是知识付费形式的提问，这类问题大家都喜欢，你用自己辛辛苦苦学到的知识改善了自己的生活，这会让你感觉自己的努力没有白费。如果你以后在群里提问的时候附上红包，那么这个问题应该会有人看到吧。</p>
<p>最后，有问题的话可以发送邮件向我提问，至于我的邮箱地址，可以在我的简书上找到。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/11/22/%E5%A6%82%E6%9E%9C%E4%BD%A0%E9%9C%80%E8%A6%81%E6%88%91%E5%9B%9E%E7%AD%94%E4%BD%A0%E7%9A%84%E9%97%AE%E9%A2%98-how-to-ask-a-question/" data-id="clm1z22hq00ggehn5164h0ha9" data-title="如果你需要我回答你的问题" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-如何在Mac或Linux的RStudio上切换R版本-Using-different-R-version-in-RStudion-linux-macOS" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/11/22/%E5%A6%82%E4%BD%95%E5%9C%A8Mac%E6%88%96Linux%E7%9A%84RStudio%E4%B8%8A%E5%88%87%E6%8D%A2R%E7%89%88%E6%9C%AC-Using-different-R-version-in-RStudion-linux-macOS/" class="article-date">
  <time class="dt-published" datetime="2019-11-22T13:04:06.818Z" itemprop="datePublished">2019-11-22</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/11/22/%E5%A6%82%E4%BD%95%E5%9C%A8Mac%E6%88%96Linux%E7%9A%84RStudio%E4%B8%8A%E5%88%87%E6%8D%A2R%E7%89%88%E6%9C%AC-Using-different-R-version-in-RStudion-linux-macOS/">如何在MacOS/Linux的RStudio上切换R版本</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>在MacOS或Linux上使用RStudio有一点不足，就是不能方便在RStudio上切换R版本。</p>
<p>虽然不方便，但还是可以切换。</p>
<p>首先，我们需要通过编译到方式安装不同版本的R，参考<a href="/archives/Install-R-in-Linux">如何在服务器上安装最新的R</a></p>
<p>我习惯将软件安装在<code>/opt/sysoft</code>下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sudo <span class="built_in">mkdir</span> -p /opt/sysoft</span><br><span class="line">sudo <span class="built_in">chown</span> `<span class="built_in">whoami</span>` /opt/sysoft</span><br></pre></td></tr></table></figure>

<p>下载源代码，并安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">wget https://mirrors.tuna.tsinghua.edu.cn/CRAN/src/base/R-3/R-3.5.3.tar.gz</span><br><span class="line">tar xf R-3.5.3.tar.gz</span><br><span class="line"><span class="built_in">cd</span> R-3.5.3</span><br><span class="line">./configure --enable-R-shlib --prefix=/opt/sysoft/R-3.5.3 --with-x=no</span><br><span class="line">make -j 8</span><br><span class="line">make install</span><br><span class="line"><span class="comment">#</span></span><br><span class="line">wget https://mirrors.tuna.tsinghua.edu.cn/CRAN/src/base/R-3/R-3.6.1.tar.gz</span><br><span class="line">tar xf R-3.6.1.tar.gz</span><br><span class="line"><span class="built_in">cd</span> R-3.6.1</span><br><span class="line">./configure --enable-R-shlib --prefix=/opt/sysoft/R-3.6.1 --with-x=no</span><br><span class="line">make -j 8</span><br><span class="line">make install</span><br></pre></td></tr></table></figure>

<p>在安装完成之后，之后切换R语言，只需要用不同版本的R覆盖<code>/usr/local/bin</code>下面的R即可。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">ln</span> -sf /opt/sysoft/R-3.5.3/bin/R /usr/local/bin</span><br><span class="line"><span class="built_in">ln</span> -sf /opt/sysoft/R-3.6.1/bin/R /usr/local/bin</span><br></pre></td></tr></table></figure>

<p>之后打开RStudio就是更换版本后的R</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/11/22/%E5%A6%82%E4%BD%95%E5%9C%A8Mac%E6%88%96Linux%E7%9A%84RStudio%E4%B8%8A%E5%88%87%E6%8D%A2R%E7%89%88%E6%9C%AC-Using-different-R-version-in-RStudion-linux-macOS/" data-id="clm1z22h400egehn56nbw1c92" data-title="如何在MacOS/Linux的RStudio上切换R版本" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-使用inferCNV分析单细胞转录组中拷贝数变异-Single-Cell-RNA-seq-CNV-analysis-with-inferCNV" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/11/18/%E4%BD%BF%E7%94%A8inferCNV%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E4%B8%AD%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82-Single-Cell-RNA-seq-CNV-analysis-with-inferCNV/" class="article-date">
  <time class="dt-published" datetime="2019-11-18T16:25:09.924Z" itemprop="datePublished">2019-11-18</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/11/18/%E4%BD%BF%E7%94%A8inferCNV%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E4%B8%AD%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82-Single-Cell-RNA-seq-CNV-analysis-with-inferCNV/">使用inferCNV分析单细胞转录组中拷贝数变异</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p><code>inferCNV</code>用与探索肿瘤单细胞RNA-seq数据，分析其中的体细胞大规模染色体拷贝数变化(copy number  alterations, CNA), 例如整条染色体或大片段染色体的增加或丢失(gain or deletions)。工作原理是，以一组”正常”细胞作为参考，分析肿瘤基因组上各个位置的基因表达量强度变化. 通过热图的形式展示每条染色体上的基因相对表达量，相对于正常细胞，肿瘤基因组总会过表达或者低表达。</p>
<p><code>inferCNV</code>提供了一些过滤参数，通过调整参数来降低噪音，更好的揭示支持CNA的信号。此外<code>inferCNV</code>还包括预测CNA区间的方法以及根据异质性模式定义细胞类群的方法。</p>
<h2 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h2><p>尽管<code>inferCNV</code>是一个R包，但是在安装inferCNV之前还需要先下载安装<a target="_blank" rel="noopener" href="https://sourceforge.net/projects/mcmc-jags/files/JAGS/4.x/">JAGS</a> ，好在它有Windows，MacOS和Linux版本，所以<code>inferCNV</code>在各个平台都能用。</p>
<p>Windows和MacOS的JAGS容易安装，而Linux的JAGS需要编译</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 手动安装BLAS和LAPACK不推荐</span></span><br><span class="line"><span class="comment"># yum install blas-devel lapack-devel</span></span><br><span class="line">tar xf JAGS-4.3.0.tar.gz </span><br><span class="line"><span class="built_in">cd</span> JAGS-4.3.0</span><br><span class="line">./configure --libdir=/usr/local/lib64</span><br><span class="line">make -j 20 &amp;&amp; make install </span><br></pre></td></tr></table></figure>

<p>安装R包</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">install.packages<span class="punctuation">(</span><span class="string">&quot;rjags&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>requireNamespace<span class="punctuation">(</span><span class="string">&quot;BiocManager&quot;</span><span class="punctuation">,</span> quietly <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">     install.packages<span class="punctuation">(</span><span class="string">&quot;BiocManager&quot;</span><span class="punctuation">)</span></span><br><span class="line">BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&quot;infercnv&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>测试安装</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>infercnv<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">infercnv_obj <span class="operator">=</span> CreateInfercnvObject<span class="punctuation">(</span>raw_counts_matrix<span class="operator">=</span>system.file<span class="punctuation">(</span><span class="string">&quot;extdata&quot;</span><span class="punctuation">,</span> <span class="string">&quot;oligodendroglioma_expression_downsampled.counts.matrix.gz&quot;</span><span class="punctuation">,</span> package <span class="operator">=</span> <span class="string">&quot;infercnv&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">                                    annotations_file<span class="operator">=</span>system.file<span class="punctuation">(</span><span class="string">&quot;extdata&quot;</span><span class="punctuation">,</span> <span class="string">&quot;oligodendroglioma_annotations_downsampled.txt&quot;</span><span class="punctuation">,</span> package <span class="operator">=</span> <span class="string">&quot;infercnv&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">                                    delim<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">,</span></span><br><span class="line">                                    gene_order_file<span class="operator">=</span>system.file<span class="punctuation">(</span><span class="string">&quot;extdata&quot;</span><span class="punctuation">,</span> <span class="string">&quot;gencode_downsampled.EXAMPLE_ONLY_DONT_REUSE.txt&quot;</span><span class="punctuation">,</span> package <span class="operator">=</span> <span class="string">&quot;infercnv&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">                                    ref_group_names<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;Microglia/Macrophage&quot;</span><span class="punctuation">,</span><span class="string">&quot;Oligodendrocytes (non-malignant)&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> </span><br><span class="line"></span><br><span class="line">infercnv_obj <span class="operator">=</span> infercnv<span class="operator">::</span>run<span class="punctuation">(</span>infercnv_obj<span class="punctuation">,</span></span><br><span class="line">                             cutoff<span class="operator">=</span><span class="number">1</span><span class="punctuation">,</span> <span class="comment"># cutoff=1 works well for Smart-seq2, and cutoff=0.1 works well for 10x Genomics</span></span><br><span class="line">                             out_dir<span class="operator">=</span>tempfile<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">                             cluster_by_groups<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">,</span> </span><br><span class="line">                             denoise<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">,</span></span><br><span class="line">                             HMM<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>如果没有报错，就说明安装成功。</p>
<h2 id="软件使用"><a href="#软件使用" class="headerlink" title="软件使用"></a>软件使用</h2><h3 id="准备输入文件"><a href="#准备输入文件" class="headerlink" title="准备输入文件"></a>准备输入文件</h3><p>需要准备3个输入数据</p>
<ol>
<li>单细胞RNA-seq表达量的原始矩阵</li>
<li>注释文件，记录肿瘤和正常细胞</li>
<li>基因或染色体位置文件</li>
</ol>
<p>第一个是Genes x Cells的表达矩阵(matrix)，行名是基因，列名是细胞编号。</p>
<table>
<thead>
<tr>
<th>MGH54_P16_F12</th>
<th>MGH54_P12_C10</th>
<th>MGH54_P11_C11</th>
<th>MGH54_P15_D06</th>
<th>MGH54_P16_A03</th>
<th>…</th>
<th></th>
</tr>
</thead>
<tbody><tr>
<td>A2M</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>…</td>
</tr>
<tr>
<td>A4GALT</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>…</td>
</tr>
<tr>
<td>AAAS</td>
<td>0</td>
<td>37</td>
<td>30</td>
<td>21</td>
<td>0</td>
<td>…</td>
</tr>
<tr>
<td>AACS</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>2</td>
<td>…</td>
</tr>
<tr>
<td>AADAT</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>…</td>
</tr>
<tr>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
</tr>
</tbody></table>
<p>第二个是样本注释信息文件，命名为”cellAnnotations.txt”。一共两列，第一列是对应第一个文件的列名，第二列是细胞的分组</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">MGH54_P2_C12    Microglia<span class="operator">/</span>Macrophage</span><br><span class="line">MGH36_P6_F03    Microglia<span class="operator">/</span>Macrophage</span><br><span class="line">MGH54_P16_F12   Oligodendrocytes <span class="punctuation">(</span>non<span class="operator">-</span>malignant<span class="punctuation">)</span></span><br><span class="line">MGH54_P12_C10   Oligodendrocytes <span class="punctuation">(</span>non<span class="operator">-</span>malignant<span class="punctuation">)</span></span><br><span class="line">MGH36_P1_B02    malignant_MGH36</span><br><span class="line">MGH36_P1_H10    malignant_MGH36</span><br></pre></td></tr></table></figure>

<p>第三个是基因位置信息文件，命名为”geneOrderingFile.txt”。一共四列，第一列对应第一个文件的行名，其余三列则是基因的位置。<strong>注</strong>：基因名不能有重复</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">WASH7P  chr1    <span class="number">14363</span>   <span class="number">29806</span></span><br><span class="line">LINC00115       chr1    <span class="number">761586</span>  <span class="number">762902</span></span><br><span class="line">NOC2L   chr1    <span class="number">879584</span>  <span class="number">894689</span></span><br><span class="line">MIR200A chr1    <span class="number">1103243</span> <span class="number">1103332</span></span><br><span class="line">SDF4    chr1    <span class="number">1152288</span> <span class="number">1167411</span></span><br><span class="line">UBE2J2  chr1    <span class="number">1189289</span> <span class="number">1209265</span></span><br></pre></td></tr></table></figure>

<h3 id="两步法"><a href="#两步法" class="headerlink" title="两步法"></a>两步法</h3><p>最复杂的工作就是准备输入文件，而一旦上述三个文件已经创建完成，那么分析只要两步以及根据结果对参数进行调整。</p>
<p>第一步，根据上述的三个文件创建<code>inferCNV</code>对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">infercnv_obj <span class="operator">=</span> CreateInfercnvObject<span class="punctuation">(</span>raw_counts_matrix<span class="operator">=</span>matrix<span class="punctuation">,</span> <span class="comment"># 可以直接提供矩阵对象</span></span><br><span class="line">                                    annotations_file<span class="operator">=</span><span class="string">&quot;cellAnnotations.txt&quot;</span><span class="punctuation">,</span></span><br><span class="line">                                    delim<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">,</span></span><br><span class="line">                                    gene_order_file<span class="operator">=</span><span class="string">&quot;gene_ordering_file.txt&quot;</span><span class="punctuation">,</span></span><br><span class="line">                                    ref_group_names<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;normal&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这一步的一个关键参数是<code>ref_group_name</code>, 用于设置参考组。假如你并不知道哪个组是正常，哪个组不正常，那么设置为<code>ref_group_name=NULL</code>, 那么<code>inferCNV</code>会以全局平均值作为基线，这适用于有足够细胞存在差异的情况。此外，这里的<code>raw_count_matrix</code>是排除了低质量细胞的count矩阵。</p>
<p>第二步，运行标准的<code>inferCNV</code>流程。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># perform infercnv operations to reveal cnv signal</span></span><br><span class="line">infercnv_obj <span class="operator">=</span> infercnv<span class="operator">::</span>run<span class="punctuation">(</span>infercnv_obj<span class="punctuation">,</span></span><br><span class="line">                             cutoff<span class="operator">=</span><span class="number">1</span><span class="punctuation">,</span>  <span class="comment"># use 1 for smart-seq, 0.1 for 10x-genomics</span></span><br><span class="line">                             out_dir<span class="operator">=</span><span class="string">&quot;output_dir&quot;</span><span class="punctuation">,</span>  <span class="comment"># 输出文件夹</span></span><br><span class="line">                             cluster_by_groups<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span>   <span class="comment"># 聚类</span></span><br><span class="line">                             denoise<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span> <span class="comment">#去噪</span></span><br><span class="line">                             HMM<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">)</span> <span class="comment"># 是否基于HMM预测CNV</span></span><br></pre></td></tr></table></figure>

<p>关键参数是<code>cutoff</code>, 用于选择哪些基因会被用于分析（在所有细胞的平均表达量需要大于某个阈值）。这个需要根据具体的测序深度来算，官方教程建议10X设置为0.1，smart-seq设置为1。你可以先评估下不同阈值下的保留基因数，决定具体值。<code>cluster_by_groups</code>用于声明是否根据细胞注释文件的分组对肿瘤细胞进行分群。</p>
<p>最终会输出很多文件在<code>out_dir</code>的目录下，而实际有用的是下面几个</p>
<ul>
<li>infercnv.preliminary.png : 初步的inferCNV展示结果（未经去噪或HMM预测）</li>
<li>infercnv.png : 最终inferCNV产生的去噪后的热图.</li>
<li>infercnv.references.txt : 正常细胞矩阵.</li>
<li>infercnv.observations.txt : 肿瘤细胞矩阵.</li>
<li>infercnv.observation_groupings.txt : 肿瘤细胞聚类后的分组关系.</li>
<li>infercnv.observations_dendrogram.txt : NEWICK格式，展示细胞间的层次关系.</li>
</ul>
<h3 id="参数说明"><a href="#参数说明" class="headerlink" title="参数说明"></a>参数说明</h3><p><code>Infercnv::run</code>的参数非常之多，总体上分为如下几类</p>
<ul>
<li>基本设置</li>
<li>平滑参数</li>
<li>基于肿瘤亚群的下游分析(HMM或non-DE-masking)</li>
<li>根据 BayesNet P(Normal) 过滤低可信度HMM预测结果</li>
<li>肿瘤亚群细分</li>
<li>去噪参数</li>
<li>离群值修剪</li>
<li>其他选项</li>
<li>实验性参数（不稳定）</li>
<li>差异表达分析的实验性参数</li>
</ul>
<p>你可以按照具体的需求修改不同步骤的参数，例如聚类默认<code>cluster_by_groups=FALSE</code>会根据<code>k_obs_groups</code>聚类成指定的组数，而层次聚类方法用于计算组间相似度的参数则是<code>hclust_method</code>.</p>
<p>此外，设置<code>HMM=TRUE</code> 的计算时间会长于<code>HMM=FALSE</code>,因此可以先设置<code>HMM=FALSE</code>快速查看结果。</p>
<p>在运行过程中它会显示每个步骤的信息，官方文档给出了示意图帮助理解。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/InferCNV_procedure-e29ea5dab7b042f58bdc0600cb7ff489.png" alt="InferCNV_procedure"></p>
<h3 id="提取信息"><a href="#提取信息" class="headerlink" title="提取信息"></a>提取信息</h3><p><code>inferCNV</code>会输出一个” map_metadata_from_infercnv .txt”文件用于记录每个细胞的元信息，所有信息都可以从该文件中进行提取。或者使用<code>infercnv::add_to_seurat</code>将信息直接增加到原来的seurat对象中。</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li>软件安装: <a target="_blank" rel="noopener" href="https://github.com/broadinstitute/inferCNV/wiki/Installing-infercnv">https://github.com/broadinstitute/inferCNV/wiki/Installing-infercnv</a></li>
<li>文件定义:  <a target="_blank" rel="noopener" href="https://github.com/broadinstitute/inferCNV/wiki/File-Definitions">https://github.com/broadinstitute/inferCNV/wiki/File-Definitions</a> </li>
<li>运行inferCNV: <a target="_blank" rel="noopener" href="https://github.com/broadinstitute/inferCNV/wiki/Running-InferCNV">https://github.com/broadinstitute/inferCNV/wiki/Running-InferCNV</a> </li>
<li><strong>引用</strong>: inferCNV of the Trinity CTAT Project.  <a target="_blank" rel="noopener" href="https://github.com/broadinstitute/inferCNV">https://github.com/broadinstitute/inferCNV</a></li>
</ul>
<p>关于<code>inferCNV</code>的算法原理在如下几篇文章中有说明</p>
<ul>
<li>Anoop P. Patel, Itay Tirosh, et al. Single-cell RNA-seq highlights intratumoral heterogeneity in primary glioblastoma. Science. 2014 Jun 20: 1396-1401</li>
<li>Tirosh I et al. Dissecting the multicellular ecosystem of metastatic melanoma by single-cell RNA-seq. Science. 2016 Apr 8;352(6282):189-96</li>
<li>Tirosh I et al. Single-cell RNA-seq supports a developmental hierarchy in human oligodendroglioma. Nature. 2016 Nov 10;539(7628):309-313. PubMed PMID: 27806376; PubMed Central PMCID: PMC5465819.</li>
<li>Venteicher AS, Tirosh I, et al. Decoupling genetics, lineages, and microenvironment in IDH-mutant gliomas by single-cell RNA-seq. Science. 2017 Mar 31;355(6332).PubMed PMID: 28360267; PubMed Central PMCID: PMC5519096.</li>
<li>Puram SV, Tirosh I, et al. Single-Cell Transcriptomic Analysis of Primary and Metastatic Tumor Ecosystems in Head and Neck Cancer. Cell. 2017 Dec 14;171(7):1611-1624.e24. PubMed PMID: 29198524; PubMed Central PMCID: PMC5878932.</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/11/18/%E4%BD%BF%E7%94%A8inferCNV%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E4%B8%AD%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82-Single-Cell-RNA-seq-CNV-analysis-with-inferCNV/" data-id="clm1z22gc00bnehn5dza4aq51" data-title="使用inferCNV分析单细胞转录组中拷贝数变异" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用mclust进行聚类分析-Cluster-data-with-mclust" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/11/09/%E4%BD%BF%E7%94%A8mclust%E8%BF%9B%E8%A1%8C%E8%81%9A%E7%B1%BB%E5%88%86%E6%9E%90-Cluster-data-with-mclust/" class="article-date">
  <time class="dt-published" datetime="2019-11-09T16:05:10.766Z" itemprop="datePublished">2019-11-09</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/11/09/%E4%BD%BF%E7%94%A8mclust%E8%BF%9B%E8%A1%8C%E8%81%9A%E7%B1%BB%E5%88%86%E6%9E%90-Cluster-data-with-mclust/">使用mclust进行聚类分析</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p><code>mclust</code>(Model-based clustering) 能够基于高斯有限混合模型进行聚类，分类以及密度估计(density estimation)。对于具有各种协方差结构的高斯混合模型，它提供了根据EM算法的参数预测函数。它也提供了根据模型进行模拟的函数。还提供了一类函数，整合了基于模型的层次聚类，混合估计的EM算法，用于聚类、密度估计和判别分析中综合性策略的贝叶斯信息判别标准。最后还有一类函数能够对聚类，分类和密度估计结果中的拟合模型进行可视化展示。</p>
<p>简而言之，<code>mclust</code>在R语言上实现了<strong>基于高斯有限混合模型</strong>的<strong>聚类</strong>，<strong>分类</strong>和<strong>密度估计</strong>分析，并且还有专门的可视化函数展示分析结果。</p>
<blockquote>
<p>和<code>mclust</code>功能相似的其他R包: ‘Rmixmod’, ‘mixture’, ‘EMCluster’, ‘mixtools’, ‘bgmm’, ‘flexmix’</p>
</blockquote>
<h2 id="安装和加载"><a href="#安装和加载" class="headerlink" title="安装和加载"></a>安装和加载</h2><p>在已有的R语言的基础上，只需要运行如下代码即可</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 安装</span></span><br><span class="line">install.packages<span class="punctuation">(</span><span class="string">&quot;mclust&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 加载</span></span><br><span class="line">library<span class="punctuation">(</span>mclust<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="聚类实战"><a href="#聚类实战" class="headerlink" title="聚类实战"></a>聚类实战</h2><p>以一个例子来介绍一下如何使用<code>mclust</code>进行聚类分析。我们得要先加载一个来自于R包’gclus’的数据集’wine’，该数据集有178行，分别是不同区域的品种葡萄， 14列，其中后13列是化学分析的测量值。我们的目标是将其进行分类。</p>
<p>第一步: 加载数据集</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">install.packages<span class="punctuation">(</span><span class="string">&quot;gclus&quot;</span><span class="punctuation">)</span></span><br><span class="line">data<span class="punctuation">(</span><span class="string">&quot;wine&quot;</span><span class="punctuation">,</span> package <span class="operator">=</span> <span class="string">&quot;gclus&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="built_in">dim</span><span class="punctuation">(</span>wine<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>第二步 : 使用<code>Mclust</code>做聚类分析. <code>Mclust</code>主要功能就是分析当前的提供的数据是由什么统计模型</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 第一列和聚类无关</span></span><br><span class="line">X <span class="operator">&lt;-</span> data.matrix<span class="punctuation">(</span>wine<span class="punctuation">[</span><span class="punctuation">,</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">mod <span class="operator">&lt;-</span> Mclust<span class="punctuation">(</span>X<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>直接在交互行输入<code>mod</code>会得到如下信息</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="string">&#x27;Mclust&#x27;</span> model object<span class="operator">:</span> <span class="punctuation">(</span>VVE<span class="punctuation">,</span><span class="number">3</span><span class="punctuation">)</span> </span><br><span class="line"></span><br><span class="line">Available components<span class="operator">:</span> </span><br><span class="line"> <span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span> <span class="string">&quot;call&quot;</span>           <span class="string">&quot;data&quot;</span>           <span class="string">&quot;modelName&quot;</span>     </span><br><span class="line"> <span class="punctuation">[</span><span class="number">4</span><span class="punctuation">]</span> <span class="string">&quot;n&quot;</span>              <span class="string">&quot;d&quot;</span>              <span class="string">&quot;G&quot;</span>             </span><br><span class="line"> <span class="punctuation">[</span><span class="number">7</span><span class="punctuation">]</span> <span class="string">&quot;BIC&quot;</span>            <span class="string">&quot;bic&quot;</span>            <span class="string">&quot;loglik&quot;</span>        </span><br><span class="line"><span class="punctuation">[</span><span class="number">10</span><span class="punctuation">]</span> <span class="string">&quot;df&quot;</span>             <span class="string">&quot;hypvol&quot;</span>         <span class="string">&quot;parameters&quot;</span>    </span><br><span class="line"><span class="punctuation">[</span><span class="number">13</span><span class="punctuation">]</span> <span class="string">&quot;z&quot;</span>              <span class="string">&quot;classification&quot;</span> <span class="string">&quot;uncertainty&quot;</span> </span><br></pre></td></tr></table></figure>

<p>这里需要对结果稍作解释，第一行告诉我们’Mclust’以<code>VVE</code>模型将数据分为3类。第3行开始，它告诉我们’Mclust’的输出结果中包含了如下内容，我们可以通过<code>$</code>来提取。举个例子，我们提取<code>Mclust</code>的聚类结果和已知结果进行比较</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">table<span class="punctuation">(</span>wine<span class="operator">$</span>Class<span class="punctuation">,</span> mod<span class="operator">$</span>classification<span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 如下是输出信息</span></span><br><span class="line">     <span class="number">1</span>  <span class="number">2</span>  <span class="number">3</span></span><br><span class="line">  <span class="number">1</span> <span class="number">59</span>  <span class="number">0</span>  <span class="number">0</span></span><br><span class="line">  <span class="number">2</span>  <span class="number">0</span> <span class="number">69</span>  <span class="number">2</span></span><br><span class="line">  <span class="number">3</span>  <span class="number">0</span>  <span class="number">0</span> <span class="number">48</span></span><br><span class="line"><span class="comment"># adjustedRandIndex:评估聚类效果</span></span><br><span class="line">adjustedRandIndex<span class="punctuation">(</span>wine<span class="operator">$</span>Class<span class="punctuation">,</span> mod<span class="operator">$</span>classification<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>从结果中，我们发现仅有2例没有正确聚类，说明<code>Mclust</code>的效果很好。但是随之而来的问题是，<code>Mclust</code>如何挑选模型以及它为什么认为聚成3类比较合适呢？我们可以根据什么信息进行模型选择呢？</p>
<h2 id="模型选择"><a href="#模型选择" class="headerlink" title="模型选择"></a>模型选择</h2><p>为了解答上面的问题，我们需要稍微了解点<code>Mclust</code>的原理。和其他基于模型的方法类似，<code>Mclust</code>假设观测数据是一个或多个混合高斯分布的抽样结果，<code>Mclust</code>就需要根据现有数据去推断最优可能的模型参数，以及是由 q几组分布抽样而成。<code>mclust</code>一共提供了14种模型（见下表），可以用<code>?mclustModelNames</code>可以查看<code>mclust</code>提供的所有模型。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/image-20191109125815419-9f69302312944a0f92e0c245c6a39f0f.png" alt="模型总结"></p>
<p>为了对模型有一个直观的理解，<code>mclust</code>提供了这些模型数据分为三组前提下在二维中的形状。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/image-20191109125852039-00b71d5c1283487c8084b329c61fb7aa.png" alt="二维的模型形状"></p>
<p>继续回到之前的问题，<code>Mclust</code>如何确定模型和确定分组数目。之前我们调用<code>Mclust</code>时，除了必须设置的输入参数，没有修改其他参数。其实<code>Mclust</code>可以设置的参数不少，和问题直接相关的是如下两个参数</p>
<ul>
<li>G: 分组数，默认情况下是<code>1:9</code></li>
<li>modelNames: 待拟合的模型，默认使用所有14种。</li>
</ul>
<p>也就是，<code>Mclust</code>默认得到14种模型1到9组的分析结果，然后根据一定的标准选择最终的模型和分组数。</p>
<p><code>Mclust</code>提供了两种方法用于评估不同模型在不同分组下的可能性</p>
<ul>
<li>BIC( Bayesian Information Criterion ): 贝叶斯信息判别标准</li>
<li>ICL( integrated complete-data likelihood ): 综合完全数据可能性</li>
</ul>
<p><code>Mclust</code>默认用的就是<code>BIC</code>，因此我们可以用<code>plot.Mclust</code>绘制其中BIC变化曲线</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">plot.Mclust<span class="punctuation">(</span>mod<span class="punctuation">,</span> what <span class="operator">=</span> <span class="string">&quot;BIC&quot;</span><span class="punctuation">,</span> </span><br><span class="line">     ylim <span class="operator">=</span> <span class="built_in">range</span><span class="punctuation">(</span>mod<span class="operator">$</span>BIC<span class="punctuation">[</span><span class="punctuation">,</span><span class="operator">-</span><span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">2</span><span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">,</span> na.rm <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">     legendArgs <span class="operator">=</span> <span class="built_in">list</span><span class="punctuation">(</span>x <span class="operator">=</span> <span class="string">&quot;bottomleft&quot;</span><span class="punctuation">,</span> cex <span class="operator">=</span><span class="number">0.7</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/image-20191109154734889-6583692493584d749ae401eab1d6889a.png" alt="BIC曲线"></p>
<p><code>Mclucst</code>会选择其中BIC最大的模型和分组作为最终的结果。</p>
<p>此外我们可以用<code>MclustBIC</code>和<code>MclustICL</code>分别进行计算</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">par<span class="punctuation">(</span>mfrow<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">BIC <span class="operator">&lt;-</span> mclustBIC<span class="punctuation">(</span>X<span class="punctuation">)</span></span><br><span class="line">ICL <span class="operator">&lt;-</span> mclustICL<span class="punctuation">(</span>X<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>从中选择最佳的模型分组和模型作为输入</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mod2 <span class="operator">&lt;-</span> Mclust<span class="punctuation">(</span>X<span class="punctuation">,</span> G <span class="operator">=</span> <span class="number">3</span><span class="punctuation">,</span> modelNames <span class="operator">=</span> <span class="string">&quot;VVE&quot;</span><span class="punctuation">,</span> x<span class="operator">=</span>BIC<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="可视化展示"><a href="#可视化展示" class="headerlink" title="可视化展示"></a>可视化展示</h2><p><code>mclust</code>为不同的输出都提供了对应的泛型函数用于可视化，你需要用<code>plot</code>就能得到结果。例如对之前的聚类结果在二维空间展示</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">drmod <span class="operator">&lt;-</span> MclustDR<span class="punctuation">(</span>mod<span class="punctuation">,</span> lambda <span class="operator">=</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">plot<span class="punctuation">(</span>drmod<span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 会提供一些列选项让你选择, 展示不同的结果</span></span><br><span class="line"><span class="comment"># Dimension reduction for model-based clustering and classification plots: </span></span><br><span class="line"></span><br><span class="line"><span class="number">1</span><span class="operator">:</span> scatterplot</span><br><span class="line"><span class="number">2</span><span class="operator">:</span> contour</span><br><span class="line"><span class="number">3</span><span class="operator">:</span> classification</span><br><span class="line"><span class="number">4</span><span class="operator">:</span> boundaries</span><br><span class="line"><span class="number">5</span><span class="operator">:</span> density</span><br><span class="line"><span class="number">6</span><span class="operator">:</span> evalues</span><br><span class="line"><span class="comment"># 以1为例</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/image-20191109155624344-40f3e047fc6644c586e932b9196fcce3.png" alt="plot"></p>
<p><code>mclust</code>还有很多其他功能，例如密度估计，自举推断(Bootstrap inference)，这些内容建议阅读”mclust 5: Clustering, Classification and Density Estimation Using Gaussian Finite Mixture Models “</p>
<h2 id="推荐阅读"><a href="#推荐阅读" class="headerlink" title="推荐阅读"></a>推荐阅读</h2><p>想要更好的学习这个R包的使用，还需要去学习如下概念</p>
<ul>
<li>EM算法( expectation–maximization algorithm )</li>
<li>BIC</li>
<li>MLE(maximum likelihood estimator)</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/11/09/%E4%BD%BF%E7%94%A8mclust%E8%BF%9B%E8%A1%8C%E8%81%9A%E7%B1%BB%E5%88%86%E6%9E%90-Cluster-data-with-mclust/" data-id="clm1z22ge00buehn5c2nq2bfy" data-title="使用mclust进行聚类分析" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-线性代数学习笔记-利用最小二乘法做线性回归-Linear-algebra-linear-regression-with-least-square-method" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/11/07/%E7%BA%BF%E6%80%A7%E4%BB%A3%E6%95%B0%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0-%E5%88%A9%E7%94%A8%E6%9C%80%E5%B0%8F%E4%BA%8C%E4%B9%98%E6%B3%95%E5%81%9A%E7%BA%BF%E6%80%A7%E5%9B%9E%E5%BD%92-Linear-algebra-linear-regression-with-least-square-method/" class="article-date">
  <time class="dt-published" datetime="2019-11-07T16:39:29.231Z" itemprop="datePublished">2019-11-07</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/11/07/%E7%BA%BF%E6%80%A7%E4%BB%A3%E6%95%B0%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0-%E5%88%A9%E7%94%A8%E6%9C%80%E5%B0%8F%E4%BA%8C%E4%B9%98%E6%B3%95%E5%81%9A%E7%BA%BF%E6%80%A7%E5%9B%9E%E5%BD%92-Linear-algebra-linear-regression-with-least-square-method/">线性代数学习笔记-利用最小二乘法做线性回归</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h2 id="背景知识"><a href="#背景知识" class="headerlink" title="背景知识"></a>背景知识</h2><p>在学习线性代数的时候，我们都会先从线性方程组入手。求解一个线性方程组是否存在解，如果存在解，那么有多少解，比方说求解下面这个方程组</p>
<p>$$<br>x_1 - 2x_2 &#x3D; -1<br>$$<br>$$<br>-x1 - 2x_2 &#x3D; 3<br>$$</p>
<p>先别急着掏出纸和笔进行运算。在线性代数中, 这类线性方程组更常见的表述方式为<br>$$<br>Ax &#x3D; b<br>$$<br>其中，A是系数矩阵，x和b都是向量。在R语言中，这个方程组可以通过<code>solve</code>函数求解</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">A <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="operator">-</span><span class="number">2</span><span class="punctuation">,</span><span class="operator">-</span><span class="number">2</span><span class="punctuation">)</span><span class="punctuation">,</span> ncol <span class="operator">=</span> <span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">b <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">)</span></span><br><span class="line">solve<span class="punctuation">(</span>A<span class="punctuation">,</span>b<span class="punctuation">)</span> </span><br><span class="line"><span class="comment"># -2.0 -0.5</span></span><br></pre></td></tr></table></figure>

<p>如果方程组有解或者无数解，那么我们称方程组是<strong>相容的</strong>，反之则是<strong>不相容</strong>。</p>
<p>对于不相容的方程，我们无法求解出一个x使得等式成立，也就是下图的Ax和b不在一个平面上。因此只能找到一个x，让Ax尽可能b和接近，即求解一个x使得$||Ax-b||$最小。也就是$||Ax-b||^2$最小,  也就是$(Ax-b)\cdot(Ax-b)$最小， 这也就是术语<strong>最小二乘法</strong>( <strong>least squares method</strong>, 也称之为最小平方法 )的来源。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/image-20191107133503046-c48ec22789924735875f586647b188c5.png" alt="几何意义"></p>
<p>那么如何求解出x呢？跳过教科书的证明部分，这里直接提供定理，用于判定在什么条件下，方程$Ax&#x3D;b$的最小二乘解是唯一的。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/image-20191107135914705-dabeb5d3ec954a80bef9ebd988c4f494.png" alt="定理"></p>
<p>如果已知A的列是线性无关，那么对于方程$Ax&#x3D;b$，取$A&#x3D;QR$， 那么唯一的最小二乘解为$\hat x &#x3D; R^{-1}Q^Tb$,</p>
<blockquote>
<p>$A&#x3D;QR$称之为QR分解，也就是将 $m \times n$ 矩阵A分解为两个矩阵相乘的形式。其中Q是 $m \times n$ 矩阵，其列形成$Col\ A$的一个标准正交基，R是一个 $n \times n$ 上三角可逆矩阵且在对角线元素为正数。</p>
</blockquote>
<h2 id="最小二乘直线"><a href="#最小二乘直线" class="headerlink" title="最小二乘直线"></a>最小二乘直线</h2><p>那么最小二乘有什么用呢？为了方便讨论实际的应用问题，我们用科学和工程数据中更常见的统计分析记号$X\beta&#x3D;y$ 来替代$Ax&#x3D;b$. 其中$X$为<strong>设计矩阵</strong>，$\beta$为<strong>参数向量</strong>，$y$为<strong>观测向量</strong>。</p>
<p>我们都知道身高和体重存在一定关系，但不是完美对应(数据集来自于R语言的women)</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/image-20191107142255926-23d0c820e06b4c779c062c827a8d3a95.png" alt="身高-体重关系"></p>
<p>我们可以找到一条曲线去完美拟合已有的数据，但通常而言越复杂的模型就越容易出现过拟合的情况，不具有普适性，最好是找到一个简单模型，能对当前的数据做一个拟合，并且能预测未来。</p>
<p>两个变量最简单的关系就是线性方程$y&#x3D;\beta_0 +\beta_1 x$。我们希望能够确定 $\beta_0$ 和 $\beta_1$使得预测值和观测值尽可能的接近，也就是让直线和数据尽可能的接近，最常用的度量方法就是<strong>余差平方之和</strong>最小。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/image-20191107145640604-1c2f0f704cc64b669e6135a682b68f1c.png" alt="直线拟合"></p>
<p><strong>最小二乘直线</strong>$y&#x3D;\beta_0 +\beta_1 x$是余差平方之和最小的，这条直线也被称之为y对x的回归直线。直线的系数 $ \beta_0$ 和 $\beta_1$称为线性回归系数。</p>
<p>如果数据点都落在直线上，那么参数 $ \beta_0$ 和 $\beta_1$满足如下方程，那么</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/image-20191107153306341-3531e3cb9b6f415f91709864e7650035.png" alt="线性方程组"></p>
<p>但是当数据点不在直线上时，这就意味着 $X\beta&#x3D;y$ 没有解，那么问题其实是$Ax&#x3D;b$的最小二乘问题，这两者仅仅记法不同。并且由于X的列线性无关，那么唯一的最小二乘解为$\hat \beta &#x3D; R^{-1}Q^Ty$, 取$X&#x3D;QR$，</p>
<p>接下来，我们就可以通过R语言的矩阵函数计算$\beta$</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 加载数据</span></span><br><span class="line">data<span class="punctuation">(</span><span class="string">&quot;women&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置X和Y</span></span><br><span class="line">X <span class="operator">&lt;-</span> cbind<span class="punctuation">(</span><span class="built_in">rep</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="built_in">length</span><span class="punctuation">(</span>women<span class="operator">$</span>height<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> women<span class="operator">$</span>height<span class="punctuation">)</span></span><br><span class="line">y <span class="operator">&lt;-</span> women<span class="operator">$</span>weight</span><br><span class="line"></span><br><span class="line"><span class="comment"># QR分解</span></span><br><span class="line">qr <span class="operator">&lt;-</span> qr<span class="punctuation">(</span>X<span class="punctuation">)</span></span><br><span class="line">Q <span class="operator">&lt;-</span> qr.Q<span class="punctuation">(</span>qr<span class="punctuation">)</span></span><br><span class="line">R <span class="operator">&lt;-</span> qr.R<span class="punctuation">(</span>qr<span class="punctuation">)</span> </span><br><span class="line">R.inv <span class="operator">&lt;-</span> solve<span class="punctuation">(</span>R<span class="punctuation">)</span> <span class="comment"># 求逆矩阵</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 计算</span></span><br><span class="line">beta <span class="operator">&lt;-</span> R.inv <span class="operator">%*%</span> t<span class="punctuation">(</span>Q<span class="punctuation">)</span> <span class="operator">%*%</span> y</span><br></pre></td></tr></table></figure>

<p>根据求解结果，系数分别是-87.52和3.45, 我们来绘图展示下</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">plot<span class="punctuation">(</span>women<span class="operator">$</span>height<span class="punctuation">,</span> women<span class="operator">$</span>weight<span class="punctuation">,</span></span><br><span class="line">     xlab <span class="operator">=</span> <span class="string">&quot;height&quot;</span><span class="punctuation">,</span> ylab<span class="operator">=</span><span class="string">&quot;weigth&quot;</span><span class="punctuation">)</span></span><br><span class="line">x <span class="operator">&lt;-</span> 58<span class="operator">:</span><span class="number">72</span></span><br><span class="line">y <span class="operator">&lt;-</span> 3.45 <span class="operator">*</span> x <span class="operator">-</span> <span class="number">87.52</span></span><br><span class="line">lines<span class="punctuation">(</span>x<span class="operator">=</span>x<span class="punctuation">,</span>y<span class="operator">=</span>y<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/image-20191106233134871-d6a6a6d2ad8d40a59a4dd0539f2ea6f2.png" alt="weight-vs-height"></p>
<p>关于线性回归，R语言提供了<code>lm</code>函数，之前只是无脑调用，现在我把它的黑匣子打开了，它和我们之前做的事情类似，都有一步QR分解，只不过调用了不同的底层函数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># lm</span></span><br><span class="line">z <span class="operator">&lt;-</span> .Call<span class="punctuation">(</span>C_Cdqrls<span class="punctuation">,</span> ...</span><br><span class="line"><span class="comment"># qr</span></span><br><span class="line"><span class="keyword">if</span><span class="punctuation">(</span>LAPACK<span class="punctuation">)</span></span><br><span class="line">     <span class="built_in">return</span><span class="punctuation">(</span>structure<span class="punctuation">(</span>.Internal<span class="punctuation">(</span>La_qr<span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> useLAPACK <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> <span class="built_in">class</span> <span class="operator">=</span> <span class="string">&quot;qr&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">...           </span><br><span class="line">res <span class="operator">&lt;-</span> .Fortran<span class="punctuation">(</span>.F_dqrdc2<span class="punctuation">,</span>...</span><br></pre></td></tr></table></figure>

<p><code>lm</code>的输出信息有很多，比如说残差，系数的显著性水平，这些可以用<code>summary.lm</code>查看。</p>
<blockquote>
<p>之前看「R语言实战」的回归章节时，总是不理解，看完就记住了几个函数而已。现在在线性代数理论基础上对这个部分有所理解了。</p>
</blockquote>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><p>如果对线性代数的最小二乘理论感兴趣，推荐阅读「线性代数及其应用」，戴维，原书第五版，第六章。</p>
<p>关于线性回归的R语言实现，推荐阅读「R语言实战」，第二版，第8章。</p>
<p>此外，维基百科的 <a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E6%9C%80%E5%B0%8F%E4%BA%8C%E4%B9%98%E6%B3%95">最小二乘法</a>页面也提供了比较好的解释。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/11/07/%E7%BA%BF%E6%80%A7%E4%BB%A3%E6%95%B0%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0-%E5%88%A9%E7%94%A8%E6%9C%80%E5%B0%8F%E4%BA%8C%E4%B9%98%E6%B3%95%E5%81%9A%E7%BA%BF%E6%80%A7%E5%9B%9E%E5%BD%92-Linear-algebra-linear-regression-with-least-square-method/" data-id="clm1z22io00iuehn59p4j98kw" data-title="线性代数学习笔记-利用最小二乘法做线性回归" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-数据预处理之变量变换-Data-Preprocess-Variable-Transformation" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/11/02/%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86%E4%B9%8B%E5%8F%98%E9%87%8F%E5%8F%98%E6%8D%A2-Data-Preprocess-Variable-Transformation/" class="article-date">
  <time class="dt-published" datetime="2019-11-02T18:37:55.904Z" itemprop="datePublished">2019-11-02</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/11/02/%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86%E4%B9%8B%E5%8F%98%E9%87%8F%E5%8F%98%E6%8D%A2-Data-Preprocess-Variable-Transformation/">数据预处理之变量变换</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>在学习「数据挖掘导论」的<strong>数据预处理</strong>时，里面谈到了变量变换，我联想到了在基因表达量分析时的常见操作，例如FPKM，TPM，CPM，log对数变换。 比如说在文章里面会见到如下的描述</p>
<blockquote>
<p> The size factor of each cell was computed using a pooling strategy implemented in the R function <code>computeSumFactors</code>. Normalized counts were then computed by dividing the counts for each cell by the size factor for that cell. A log2 transformation was applied to normalized counts. </p>
</blockquote>
<p>变量转换有什么好处，需要注意些什么呢？「数据挖掘导论」讨论了两种重要的变量变换类型: 简单函数变换和规范化。</p>
<p><strong>简单变换</strong>是使用一个简单数学函数分别作用于每一个值，例如log转换，求绝对值，求倒数等。统计学中，变量变换（例如log转换）常用于将不具有高斯（正态）分布的数据变换成具有高斯（正态）分布的数据。在数据挖掘领域可以用来进行数据压缩。这类变换需要我们了解数据在变化前后的后果，例如负数取倒数之后的大小关系会逆转。</p>
<p><strong>标准化(standardization)或规范化(normalization)<strong>的</strong>目标</strong>是使整个值的集合具有<strong>特定的属性</strong>。使用这两个术语需要特别注意使用这两个词的上下文。书中提到”在数据挖掘界，这两个术语常常可互换，然而，在统计学中，术语规范化可能与使得变量<strong>正态</strong>化相混淆”。虽然从中文的角度来看，规范化和正态化明显不一样，但是从英文的角度看，正态分布翻译自normal distribution, 很容易从normal联想到normalization。当然也有将normalization翻译成归一化，即将原来数据值通过$(x - min(X) &#x2F; (max(X) - min(X))$转换到0,1之间。统计学的变量标准化指的就是对原来的数据基于均值和标准差计算z-score(公式如下), 不过考虑到均值和标准差受到离群点波动很大，可以用中位数替代均值，用绝对标准差替代标准差。</p>
<p>$$<br>x’ &#x3D; \frac{x-\bar{x}}{\sigma}<br>$$</p>
<p>R语言中做标准化常用到一个函数<code>scale</code>，它的功能是对矩阵的列进行中心化和(或)缩放</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">scale(x, center = TRUE, scale = TRUE)</span><br></pre></td></tr></table></figure>

<p>参数就3个，基本上我用的时候就提供第一个输入，举个例子</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> x <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">,</span><span class="number">5</span><span class="punctuation">)</span></span><br><span class="line"><span class="operator">&gt;</span> y <span class="operator">&lt;-</span> scale<span class="punctuation">(</span>x<span class="punctuation">)</span></span><br><span class="line"><span class="operator">&gt;</span> y</span><br><span class="line">           <span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">,</span><span class="punctuation">]</span> <span class="operator">-</span><span class="number">1.2649111</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">2</span><span class="punctuation">,</span><span class="punctuation">]</span> <span class="operator">-</span><span class="number">0.6324555</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">3</span><span class="punctuation">,</span><span class="punctuation">]</span>  <span class="number">0.0000000</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">4</span><span class="punctuation">,</span><span class="punctuation">]</span>  <span class="number">0.6324555</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">5</span><span class="punctuation">,</span><span class="punctuation">]</span>  <span class="number">1.2649111</span></span><br><span class="line"><span class="built_in">attr</span><span class="punctuation">(</span><span class="punctuation">,</span><span class="string">&quot;scaled:center&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span> <span class="number">3</span></span><br><span class="line"><span class="built_in">attr</span><span class="punctuation">(</span><span class="punctuation">,</span><span class="string">&quot;scaled:scale&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span> <span class="number">1.581139</span></span><br></pre></td></tr></table></figure>

<p>之前我都是这样子用的，也没有思考过这到底它到底做了些什么。最近为了彻底搞懂它的两个参数，我翻了它的源代码。</p>
<p>代码总共38行，第一部分是关于<code>center</code>参数，也就是中心化。它先判断<code>center</code>是否是逻辑值还是数值向量。如果是逻辑值，则当<code>center=TRUE</code>时它会计算每一列的均值，然后调用<code>sweep</code>按列分别减去均值。如果判定<code>center</code>是一个数值向量时，就是为<code>sweep</code>手动指定每一列的中心值。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="built_in">is.logical</span><span class="punctuation">(</span>center<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  <span class="keyword">if</span> <span class="punctuation">(</span>center<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    center <span class="operator">&lt;-</span> colMeans<span class="punctuation">(</span>x<span class="punctuation">,</span> na.rm<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">    x <span class="operator">&lt;-</span> sweep<span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="number">2L</span><span class="punctuation">,</span> center<span class="punctuation">,</span> check.margin<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"><span class="keyword">else</span> <span class="punctuation">&#123;</span></span><br><span class="line">  <span class="keyword">if</span><span class="punctuation">(</span><span class="operator">!</span><span class="built_in">is.numeric</span><span class="punctuation">(</span>center<span class="punctuation">)</span><span class="punctuation">)</span> center <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>center<span class="punctuation">)</span></span><br><span class="line">  <span class="keyword">if</span> <span class="punctuation">(</span><span class="built_in">length</span><span class="punctuation">(</span>center<span class="punctuation">)</span> <span class="operator">==</span> nc<span class="punctuation">)</span></span><br><span class="line">    x <span class="operator">&lt;-</span> sweep<span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="number">2L</span><span class="punctuation">,</span> center<span class="punctuation">,</span> check.margin<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">  <span class="keyword">else</span></span><br><span class="line">    stop<span class="punctuation">(</span><span class="string">&quot;length of &#x27;center&#x27; must equal the number of columns of &#x27;x&#x27;&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>第二部分是关于<code>scale</code>,决定数据如何缩放。如果是逻辑值且为真时，默认用下面的函数计算每一列的缩放系数，然后用<code>sweep</code>按列除以每列系数</p>
<p>$$<br>\sqrt{\frac{\sum_{i&#x3D;1}^{n}x{‘}_i^2}{n-1}}<br>$$</p>
<blockquote>
<p>这个计算的结果就是样本标准差，因为$x’_i&#x3D;x_i-\bar x$</p>
</blockquote>
<p>如果是数值向量，则用<code>sweep</code>以给定数值按列相除。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="built_in">is.logical</span><span class="punctuation">(</span>scale<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  <span class="keyword">if</span> <span class="punctuation">(</span>scale<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    f <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>v<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">      v <span class="operator">&lt;-</span> v<span class="punctuation">[</span><span class="operator">!</span><span class="built_in">is.na</span><span class="punctuation">(</span>v<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">      <span class="built_in">sqrt</span><span class="punctuation">(</span><span class="built_in">sum</span><span class="punctuation">(</span>v<span class="operator">^</span><span class="number">2</span><span class="punctuation">)</span> <span class="operator">/</span> <span class="built_in">max</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">(</span>v<span class="punctuation">)</span> <span class="operator">-</span> <span class="number">1L</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    <span class="punctuation">&#125;</span></span><br><span class="line">    scale <span class="operator">&lt;-</span> apply<span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="number">2L</span><span class="punctuation">,</span> f<span class="punctuation">)</span></span><br><span class="line">    x <span class="operator">&lt;-</span> sweep<span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="number">2L</span><span class="punctuation">,</span> scale<span class="punctuation">,</span> <span class="string">&quot;/&quot;</span><span class="punctuation">,</span> check.margin<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"><span class="keyword">else</span> <span class="punctuation">&#123;</span></span><br><span class="line">  <span class="keyword">if</span><span class="punctuation">(</span><span class="operator">!</span><span class="built_in">is.numeric</span><span class="punctuation">(</span>scale<span class="punctuation">)</span><span class="punctuation">)</span> scale <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>scale<span class="punctuation">)</span></span><br><span class="line">  <span class="keyword">if</span> <span class="punctuation">(</span><span class="built_in">length</span><span class="punctuation">(</span>scale<span class="punctuation">)</span> <span class="operator">==</span> nc<span class="punctuation">)</span></span><br><span class="line">    x <span class="operator">&lt;-</span> sweep<span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="number">2L</span><span class="punctuation">,</span> scale<span class="punctuation">,</span> <span class="string">&quot;/&quot;</span><span class="punctuation">,</span> check.margin<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">  <span class="keyword">else</span></span><br><span class="line">    stop<span class="punctuation">(</span><span class="string">&quot;length of &#x27;scale&#x27; must equal the number of columns of &#x27;x&#x27;&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>最后一部分是设置输出结果的属性。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">is.numeric</span><span class="punctuation">(</span>center<span class="punctuation">)</span><span class="punctuation">)</span> <span class="built_in">attr</span><span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="string">&quot;scaled:center&quot;</span><span class="punctuation">)</span> <span class="operator">&lt;-</span> center</span><br><span class="line"><span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">is.numeric</span><span class="punctuation">(</span>scale <span class="punctuation">)</span><span class="punctuation">)</span> <span class="built_in">attr</span><span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="string">&quot;scaled:scale&quot;</span> <span class="punctuation">)</span> <span class="operator">&lt;-</span> scale</span><br></pre></td></tr></table></figure>

<p>那么默认参数下，我们就是对矩阵按列进行z-score的标准化。检验标准很简单，计算标准化的数据的均值和标准差，因为z-score的结果就是通过线性变换让数据的均值为0，标准差为1（不会改变原来的数据分布）。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> mean<span class="punctuation">(</span>y<span class="punctuation">)</span></span><br><span class="line"><span class="number">0</span></span><br><span class="line"><span class="operator">&gt;</span> sd<span class="punctuation">(</span>y<span class="punctuation">)</span></span><br><span class="line"><span class="number">1</span></span><br></pre></td></tr></table></figure>

<p>这解决了我多年的疑惑，为啥我在R语言中就是找不到<code>zscore</code>这个函数，因为<code>scale</code>默认情况下就是实现<code>z-score</code>的变换。当然R语言取名<code>scale</code>也是有它的道理，比方说数据中存在离群值，我们应该选择不容易受离群值影响的中位数替代均值，选择绝对平均偏差, 中位数绝对偏差或四分位数极差来替换方差。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 默认参数</span></span><br><span class="line"><span class="operator">&gt;</span> x <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">,</span><span class="number">5</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1000</span><span class="punctuation">)</span><span class="punctuation">,</span> ncol<span class="operator">=</span><span class="number">2</span><span class="punctuation">)</span></span><br><span class="line"><span class="operator">&gt;</span> scale<span class="punctuation">(</span>x<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">           <span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span>       <span class="punctuation">[</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">]</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">,</span><span class="punctuation">]</span> <span class="operator">-</span><span class="number">1.2649111</span> <span class="operator">-</span><span class="number">0.4505747</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">2</span><span class="punctuation">,</span><span class="punctuation">]</span> <span class="operator">-</span><span class="number">0.6324555</span> <span class="operator">-</span><span class="number">0.4483330</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">3</span><span class="punctuation">,</span><span class="punctuation">]</span>  <span class="number">0.0000000</span> <span class="operator">-</span><span class="number">0.4460914</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">4</span><span class="punctuation">,</span><span class="punctuation">]</span>  <span class="number">0.6324555</span> <span class="operator">-</span><span class="number">0.4438497</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">5</span><span class="punctuation">,</span><span class="punctuation">]</span>  <span class="number">1.2649111</span>  <span class="number">1.7888488</span></span><br><span class="line">...</span><br><span class="line"><span class="comment"># 替换后</span></span><br><span class="line"><span class="operator">&gt;</span> x <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">,</span><span class="number">5</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1000</span><span class="punctuation">)</span><span class="punctuation">,</span> ncol<span class="operator">=</span><span class="number">2</span><span class="punctuation">)</span></span><br><span class="line"><span class="operator">&gt;</span> scale<span class="punctuation">(</span>x<span class="punctuation">,</span> </span><br><span class="line">      center <span class="operator">=</span> apply<span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> median<span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">      scale <span class="operator">=</span> apply<span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">        <span class="built_in">sum</span><span class="punctuation">(</span><span class="built_in">abs</span><span class="punctuation">(</span>x <span class="operator">-</span> mean<span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">/</span> <span class="punctuation">(</span><span class="built_in">length</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">      <span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line">      <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">           <span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span>         <span class="punctuation">[</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">]</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">,</span><span class="punctuation">]</span> <span class="operator">-</span><span class="number">1.3333333</span> <span class="operator">-</span><span class="number">0.005012531</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">2</span><span class="punctuation">,</span><span class="punctuation">]</span> <span class="operator">-</span><span class="number">0.6666667</span> <span class="operator">-</span><span class="number">0.002506266</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">3</span><span class="punctuation">,</span><span class="punctuation">]</span>  <span class="number">0.0000000</span>  <span class="number">0.000000000</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">4</span><span class="punctuation">,</span><span class="punctuation">]</span>  <span class="number">0.6666667</span>  <span class="number">0.002506266</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">5</span><span class="punctuation">,</span><span class="punctuation">]</span>  <span class="number">1.3333333</span>  <span class="number">2.498746867</span></span><br><span class="line">...</span><br></pre></td></tr></table></figure>

<p>此外，它还能实现其他形式的数据转换，例如归一化，即将数据的范围缩放到0到1之间</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">x <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">,</span><span class="number">5</span><span class="punctuation">,</span><span class="number">6</span><span class="punctuation">)</span><span class="punctuation">,</span> ncol<span class="operator">=</span><span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">y <span class="operator">&lt;-</span> scale<span class="punctuation">(</span>x<span class="punctuation">,</span> center <span class="operator">=</span> apply<span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> <span class="built_in">min</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">           scale <span class="operator">=</span> apply<span class="punctuation">(</span>x<span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span><span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> <span class="punctuation">&#123;</span><span class="built_in">max</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="operator">-</span><span class="built_in">min</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line">           <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>最后总结一下:</p>
<ul>
<li>数据的变量转换有两类，简单变换和标准化&#x2F;规范化，无论是何种变换都要看它的具体公式</li>
<li>R语言的<code>scale</code>默认就是计算原始数据的z-score, 通过调整<code>center</code>和<code>scale</code>可以实现多种形式的数据转换。</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/11/02/%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86%E4%B9%8B%E5%8F%98%E9%87%8F%E5%8F%98%E6%8D%A2-Data-Preprocess-Variable-Transformation/" data-id="clm1z22i900hfehn50tcegbi1" data-title="数据预处理之变量变换" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用LACHESIS利用HiC数据进行基因组组装-scaffolding-contig-with-LACHESIS" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/11/01/%E4%BD%BF%E7%94%A8LACHESIS%E5%88%A9%E7%94%A8HiC%E6%95%B0%E6%8D%AE%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-scaffolding-contig-with-LACHESIS/" class="article-date">
  <time class="dt-published" datetime="2019-11-01T23:44:28.716Z" itemprop="datePublished">2019-11-01</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/11/01/%E4%BD%BF%E7%94%A8LACHESIS%E5%88%A9%E7%94%A8HiC%E6%95%B0%E6%8D%AE%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-scaffolding-contig-with-LACHESIS/">使用LACHESIS利用HiC数据进行基因组组装</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/11/image-540cd58f9f79426ea648f39202d1b2d5.png" alt="LACHESIS"></p>
<p>LACHESIS是一个比较早利用HiC数据辅助基因组组装的工具，文章发表在<a target="_blank" rel="noopener" href="https://doi.org/10.1038/nbt.2727">Nature Biotechnology</a>上.但是这款软件在2年前就不在维护了，GitHub主页上也推荐使用<a target="_blank" rel="noopener" href="https://github.com/theaidenlab">https://github.com/theaidenlab</a>里的工具进行组装。</p>
<p>不过目前依旧有很多公司在用这个软件做基因组组装，所以我也想试试看。</p>
<h2 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h2><p>LACHESIS需要安装在Linux系统系统，最低内存不低于16GB，要求最小堆栈大小为10MB，可以用<code>ulimit -s</code>查看，通过<code>ulimit -s 10240</code>。同时依赖于以下软件</p>
<ul>
<li>gcc</li>
<li>zlib</li>
<li>boost C++ : &gt; 1.52.0, &lt; 1.67.0, 老版本只有1.57能被我下载</li>
<li>SAMtools: &lt; 0.1.19</li>
</ul>
<p>以没有root权限为例，安装SAMtools和Boost这两个依赖库</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> ~/src</span><br><span class="line"><span class="built_in">mkdir</span> -p ~/opt/&#123;biosoft,sysoft&#125;</span><br><span class="line"><span class="comment"># 安装boost C++, 这一步会很久</span></span><br><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget http://sourceforge.net/projects/boost/files/boost/1.57.0/boost_1_57_0.tar.bz2</span><br><span class="line">tar xf boost_1_57_0.tar.bz2</span><br><span class="line"><span class="built_in">cd</span> boost_1_57_0</span><br><span class="line">./bootstrap.sh --prefix=~/opt/sysoft/boost_1_57_0 --with-libraries=all --with-toolset=gcc</span><br><span class="line">./b2 toolset=gcc</span><br><span class="line">./b2 install</span><br><span class="line">./bjam install</span><br><span class="line"><span class="comment"># 安装SAMtools</span></span><br><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget https://astuteinternet.dl.sourceforge.net/project/samtools/samtools/0.1.19/samtools-0.1.19.tar.bz2</span><br><span class="line">tar xf samtools-0.1.19.tar.bz2</span><br><span class="line"><span class="built_in">cd</span> samtools-0.1.19</span><br><span class="line">make -j 20</span><br></pre></td></tr></table></figure>

<p>然后是下载并解压缩</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ~/opt/biosoft/</span><br><span class="line">curl -o LACHESIS.zip https://codeload.github.com/shendurelab/LACHESIS/legacy.zip/master</span><br><span class="line">unzip LACHESIS.zip</span><br><span class="line"><span class="built_in">mv</span> shendurelab-LACHESIS-2e27abb LACHESIS</span><br><span class="line"><span class="built_in">cd</span> LACHESIS</span><br></pre></td></tr></table></figure>

<p>由于之前安装的samtools并不是安装在<code>/usr</code>目录下，因此要修改<code>src/include/gtools/</code>目录下的<code>SAMStepper.h</code>和<code>SAMStepper.cc</code>中的<code>#include &lt;bam/sam.h&gt;</code>, 指向实际地址。例如我的是<code>#include &quot;/home/xzg/src/samtools-0.1.19/sam.h&quot;</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">export</span> LACHESIS_BOOST_DIR=<span class="variable">$HOME</span>/src/boost_1_57_0</span><br><span class="line"><span class="built_in">export</span> LACHESIS_SAMTOOLS_DIR=<span class="variable">$HOME</span>/src/samtools-0.1.19</span><br><span class="line">./configure --with-samtools=<span class="variable">$HOME</span>/src/samtools-0.1.19 \</span><br><span class="line">    --with-boost=<span class="variable">$HOME</span>/src/boost_1_57_0/ </span><br><span class="line">    </span><br><span class="line">./configure --with-samtools=/data/src/samtools-0.1.19 \</span><br><span class="line">    --with-boost=/data/src/boost_1_57_0/ </span><br><span class="line">make -j 20</span><br><span class="line"><span class="built_in">mv</span> src/bin .</span><br><span class="line"><span class="built_in">mv</span> src/Lachesis bin/</span><br></pre></td></tr></table></figure>

<blockquote>
<p>如果服务器管理员帮你安装了一个高版本的boost，你需要让他进行卸载，否则默认会线用系统的boost，然后就会出错。关于更多的报错，参考<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/51dc11d57ab4">HiC软件安装篇之Lachesis</a></p>
</blockquote>
<p>如果你Perl版本不是 5.14.2 ，那么我们需要打开bin下面的perl脚本，删除如下信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">///////////////////////////////////////////////////////////////////////////////</span><br><span class="line">//                                                                           //</span><br><span class="line">// This software and its documentation are copyright (c) 2014-2015 by Joshua //</span><br><span class="line">// N. Burton and the University of Washington.  All rights are reserved.     //</span><br><span class="line">//                                                                           //</span><br><span class="line">// THIS SOFTWARE IS PROVIDED <span class="string">&quot;AS IS&quot;</span>, WITHOUT WARRANTY OF ANY KIND, EXPRESS  //</span><br><span class="line">// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF                //</span><br><span class="line">// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.  //</span><br><span class="line">// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY      //</span><br><span class="line">// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT //</span><br><span class="line">// OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR  //</span><br><span class="line">// THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                //</span><br><span class="line">//                                                                           //</span><br><span class="line">///////////////////////////////////////////////////////////////////////////////</span><br></pre></td></tr></table></figure>

<p>而且还得将第一行替换成<code>#!/usr/bin/perl -w</code></p>
<p>之后需要将LACHESIS加入环境变量</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">export</span> PATH=<span class="variable">$PATH</span>:~/biosoft/LACHESIS/bin </span><br></pre></td></tr></table></figure>

<p>为了保证软件在后续不会出错，我们还要测试下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> bin</span><br><span class="line">./Lachesis INIs/test_case.ini</span><br></pre></td></tr></table></figure>

<p>当你看到最后结果是<code>: Done!</code>就表明运行成功了，最终结果输出在当前目录下<code>outs</code>中</p>
<h2 id="软件使用"><a href="#软件使用" class="headerlink" title="软件使用"></a>软件使用</h2><p>实际的软件使用要求你需要提供至少两类输入文件</p>
<ul>
<li>HiC数据，双端fastq格式</li>
<li>初始组装，fasta格式</li>
</ul>
<p>参考<a href="/archives/Assemble-genome-using-ALLHiC-with-HiC-Data">使用ALLHiC基于HiC数据辅助基因组组装</a>的第一步，第二步和第三步，使用bwa-aln + bwa-sampe将fastq回帖到参考基因组，然后对数据进行过滤。那么在后续的流程都假设你有如下两个文件</p>
<ul>
<li>draft.asm.fasta: 初步组装结果</li>
<li>sample.clean.bam: HiC数据比对的预处理结果</li>
</ul>
<p>然后我们需要拷贝一个配置运行文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cp</span> ~/opt/biosoft/LACHESIS/bin/INIs/test_case.ini sample.ini</span><br></pre></td></tr></table></figure>

<p>接着去修改参数，由于每个参数都会有说明，这里就展示我编辑过的几个参数</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">SPECIES = <span class="built_in">test</span> <span class="comment"># 写实际的物种名即可</span></span><br><span class="line">DRAFT_ASSEMBLY_FASTA = draft.asm.fasta <span class="comment"># 待组装序列的实际位置</span></span><br><span class="line">SAM_DIR = . <span class="comment">#表示当前目录下查找文件</span></span><br><span class="line">SAM_FILES = sample.clean.bam <span class="comment">#bam文件名</span></span><br><span class="line">RE_SITE_SEQ = AAGCTT <span class="comment">#酶切识别序列</span></span><br><span class="line">USE_REFERENCE = 0 <span class="comment">#不使用参考序列</span></span><br><span class="line">BLAST_FILE_HEAD = . <span class="comment"># BLAST的输出结果</span></span><br><span class="line">CLUSTER_N = 16 <span class="comment"># 最终聚类数目</span></span><br></pre></td></tr></table></figure>

<p>上面这些参数的修改属于基本操作，而下面的参数可能需要反复修改</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># contig中最小的酶切位点数,</span></span><br><span class="line">CLUSTER_MIN_RE_SITES=25</span><br><span class="line"><span class="comment"># contig中最大的link密度, 也就是一个区域与多个contig存在信号</span></span><br><span class="line"><span class="comment"># 可能是异染色质或重复序列</span></span><br><span class="line">CLUSTER_MAX_LINK_DENSITY=2</span><br><span class="line"><span class="comment"># 对于CLUSTER_MIN_RE_SITES过滤掉的contig在初步聚类后还有机会加入已有的分组中</span></span><br><span class="line"><span class="comment"># 如果它加入其中的信号是加入另一组信号的3倍</span></span><br><span class="line">CLUSTER_NONINFORMATIVE_RATIO=3</span><br><span class="line"><span class="comment"># 允许成组的最小酶切数</span></span><br><span class="line">ORDER_MIN_N_RES_IN_TRUNK=15</span><br></pre></td></tr></table></figure>

<p>最终运行即可</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">ulimit</span> -s 10240</span><br><span class="line">Lachesis sample.ini</span><br></pre></td></tr></table></figure>

<p>我们来构建最终的组装fasta</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">CreateScaffoldedFasta.pl draft.asm.fasta out/test_case</span><br></pre></td></tr></table></figure>

<p>最终结果输出在<code>out/test_case/Lachesis_assembly.fasta</code></p>
<p>使用体验:</p>
<ul>
<li>软件安装有很多问题</li>
<li>无法处理多倍体</li>
<li>已经许久不维护</li>
<li>尝试了自己的物种，结果在排序步骤出现了问题</li>
</ul>
<blockquote>
<p>目前的想法: 这个软件我不会再用第二次了。</p>
</blockquote>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/11/01/%E4%BD%BF%E7%94%A8LACHESIS%E5%88%A9%E7%94%A8HiC%E6%95%B0%E6%8D%AE%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-scaffolding-contig-with-LACHESIS/" data-id="clm1z22fz00aaehn5ariqgpc8" data-title="使用LACHESIS利用HiC数据进行基因组组装" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-ALLHiC_ 辅助组装简单的二倍体基因组-ALLHiC-scaffolding-of-a-simple-diploid-genome" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/31/ALLHiC_%20%E8%BE%85%E5%8A%A9%E7%BB%84%E8%A3%85%E7%AE%80%E5%8D%95%E7%9A%84%E4%BA%8C%E5%80%8D%E4%BD%93%E5%9F%BA%E5%9B%A0%E7%BB%84-ALLHiC-scaffolding-of-a-simple-diploid-genome/" class="article-date">
  <time class="dt-published" datetime="2019-10-31T10:40:27.040Z" itemprop="datePublished">2019-10-31</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/31/ALLHiC_%20%E8%BE%85%E5%8A%A9%E7%BB%84%E8%A3%85%E7%AE%80%E5%8D%95%E7%9A%84%E4%BA%8C%E5%80%8D%E4%BD%93%E5%9F%BA%E5%9B%A0%E7%BB%84-ALLHiC-scaffolding-of-a-simple-diploid-genome/">ALLHiC:辅助组装简单的二倍体基因组</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="ALLHiC-辅助组装简单的二倍体基因组"><a href="#ALLHiC-辅助组装简单的二倍体基因组" class="headerlink" title="ALLHiC: 辅助组装简单的二倍体基因组"></a>ALLHiC: 辅助组装简单的二倍体基因组</h1><p>ALLHiC除了能够解决复杂基因组（高杂合，多倍体）的基因组组装问题，还能够搞定简单的基因组组装，毕竟本质上都是利用HiC的信号。</p>
<p>ALLHiC大概有两个优势：</p>
<ol>
<li>对同源多倍体是一种解决方案</li>
<li>当contig组装比较短的时候，大部分情况allhic的排序优于其他软件(如果contig很长的话，可以优先考虑下3d-dna打断，过长的contig或许是错误组装造成)</li>
</ol>
<p>参考<a href="/archives/Assemble-genome-using-ALLHiC-with-HiC-Data">使用ALLHiC基于HiC数据辅助基因组组装</a>完成软件安装和前三步的数据预处理，将Fastq文件转成BAM文件。</p>
<p>下面假定你的初步组装结果是<code>draft.asm.fasta</code>, Fastq预处理后的文件是<code>sample.clean.bam</code></p>
<p>之后，我们将contig根据相互之间的信号强弱关系进行分组，至于分成多少组要根据物种具体的染色体数目而定。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">ALLHiC_partition -b sample.clean.bam -r draft.asm.fasta -e AAGCTT -k 16</span><br><span class="line"><span class="comment"># -k: 聚类组数</span></span><br><span class="line"><span class="comment"># -e: 酶切类型, 预设HindIII: AAGCTT; MboI: GATC</span></span><br><span class="line"><span class="comment"># -m: 每个contig最少的酶切位点, 默认25</span></span><br></pre></td></tr></table></figure>

<p>要根据自己的实际建库方法选择酶切位点的识别序列，<strong>注</strong>, DpnI, DpnII, MboI, Sau3AI 这些酶都是识别相同的序列，仅仅是对甲基化敏感度不同。</p>
<p>接着，提取CLM和每个contig的酶切数</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">allhic extract sample.clean.bam draft.asm.fasta --RE AAGCTT</span><br></pre></td></tr></table></figure>

<p>然后是对每个组的contig位置和方向进行优化</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#  如下命令可以通过循环同时运行</span></span><br><span class="line">allhic optimize sample.clean.counts_AAGCTT.16g1.txt sample.clean.clm</span><br><span class="line">allhic optimize sample.clean.counts_AAGCTT.16g2.txt sample.clean.clm</span><br><span class="line">...</span><br><span class="line">allhic optimize sample.clean.counts_AAGCTT.16g16.txt sample.clean.clm</span><br></pre></td></tr></table></figure>

<p>最终获取染色体级别的组装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ALLHiC_build draft.asm.fasta</span><br></pre></td></tr></table></figure>

<p>绘制热图对组装进行评估</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 获取染色体长度</span></span><br><span class="line">perl getFaLen.pl -i groups.asm.fasta -o len.txt</span><br><span class="line"><span class="comment"># 构建chrn.list</span></span><br><span class="line">grep <span class="string">&#x27;merge.clean.counts_GATC&#x27;</span> len.txt &gt; chrn.list</span><br><span class="line"><span class="comment"># 绘图, 500k表示分辨率</span></span><br><span class="line">ALLHiC_plot sample.clean.bam groups.agp chrn.list 500k pdf</span><br></pre></td></tr></table></figure>

<p>其中<a target="_blank" rel="noopener" href="https://github.com/tangerzhang/my_script/blob/master/getFaLen.pl">getFaLen.pl</a>可以从<a target="_blank" rel="noopener" href="https://github.com/tangerzhang/my_script">https://github.com/tangerzhang/my_script</a>获取</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li><a target="_blank" rel="noopener" href="https://github.com/tangerzhang/ALLHiC/wiki/ALLHiC:-scaffolding-of-a-simple-diploid-genome">https://github.com/tangerzhang/ALLHiC/wiki/ALLHiC:-scaffolding-of-a-simple-diploid-genome</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/31/ALLHiC_%20%E8%BE%85%E5%8A%A9%E7%BB%84%E8%A3%85%E7%AE%80%E5%8D%95%E7%9A%84%E4%BA%8C%E5%80%8D%E4%BD%93%E5%9F%BA%E5%9B%A0%E7%BB%84-ALLHiC-scaffolding-of-a-simple-diploid-genome/" data-id="clm1z22ep005aehn5b6ze8wgl" data-title="ALLHiC:辅助组装简单的二倍体基因组" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-ALLHiC续_ 如何构建Allele.ctg.table-How-to-build-allele-ctg-table" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/30/ALLHiC%E7%BB%AD_%20%E5%A6%82%E4%BD%95%E6%9E%84%E5%BB%BAAllele.ctg.table-How-to-build-allele-ctg-table/" class="article-date">
  <time class="dt-published" datetime="2019-10-30T23:25:18.251Z" itemprop="datePublished">2019-10-30</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/30/ALLHiC%E7%BB%AD_%20%E5%A6%82%E4%BD%95%E6%9E%84%E5%BB%BAAllele.ctg.table-How-to-build-allele-ctg-table/">ALLHiC续:如何构建Allele.ctg.table</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>Allele.ctg.table是<a href="/archives/Assemble-genome-using-ALLHiC-with-HiC-Data">ALLHiC</a>用于过滤多倍体基因组中因等位序列相似引起的HiC噪音的必要输入。</p>
<p>构建的方法有很多种，这里列举两个方法。</p>
<h2 id="方法一-基于BLAST"><a href="#方法一-基于BLAST" class="headerlink" title="方法一: 基于BLAST"></a>方法一: 基于BLAST</h2><p>基于BLAST的方法需要你提供一个染色体级别组装的近缘物种。需要提供4个文件，近缘物种的CDS和GFF文件，多倍体物种的CDS和GFF文件。最重要的是，你得保证CDS序列中的编号和GFF文件的基因编号一致，这也是后续处理至关重要的一步。</p>
<p><strong>问题一</strong>: 我们如何快速获取待组装物种的CDS和GFF文件？最快速的方法就是利用转录组序列基于STAR + StringTie的组合进行拼接，并利用gffread从中提取出cdna序列。</p>
<p>假设你的多倍体文件名为<code>contig.fa</code>, 转录组数据为<code>read_R1.fq.gz</code>, <code>read_R2.fq.gz</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p STAR</span><br><span class="line">ref=contig.fa</span><br><span class="line">THREADS=40</span><br><span class="line">fq1=read_R1.fq.gz</span><br><span class="line">fq2=read_R2.fq.gz</span><br><span class="line">prefix=sample</span><br><span class="line"><span class="comment"># 建立索引</span></span><br><span class="line">STAR \</span><br><span class="line">  --runThreadN <span class="variable">$THREADS</span> \</span><br><span class="line">  --runMode genomeGenerate \</span><br><span class="line">  --genomeDir STAR \</span><br><span class="line">  --genomeFastaFiles <span class="variable">$ref</span></span><br><span class="line"><span class="comment">#比对</span></span><br><span class="line">STAR \</span><br><span class="line">	--genomeDir STAR \</span><br><span class="line">	--runThreadN <span class="variable">$THREADS</span> \</span><br><span class="line">	--readFilesIn <span class="variable">$fq1</span> <span class="variable">$fq2</span> \</span><br><span class="line">	--readFilesCommand zcat \</span><br><span class="line">	--outFileNamePrefix <span class="variable">$prefix</span>\</span><br><span class="line">	--outSAMtype BAM SortedByCoordinate \</span><br><span class="line">	--outBAMsortingThreadN <span class="variable">$THREADS</span> \</span><br><span class="line">	--outSAMstrandField intronMotif \</span><br><span class="line">	--outFilterIntronMotifs RemoveNoncanonical</span><br><span class="line"><span class="comment"># 拼接转录本</span></span><br><span class="line">stringtie <span class="variable">$&#123;prefix&#125;</span>Aligned.sortedByCoord.out.bam -p <span class="variable">$THREADS</span> -o qry.gtf</span><br><span class="line"><span class="comment"># GTF 转成 GFF3</span></span><br><span class="line">gffread qry.gtf -o qry.gff</span><br><span class="line"><span class="comment"># 提取cdna</span></span><br><span class="line">gffread qry.gff -g <span class="variable">$ref</span> -w qry.fa</span><br></pre></td></tr></table></figure>

<p><strong>问题二</strong>:  后续的处理脚本要求输入序列的编号和GFF的基因编号相同，并且编号在GFF文件中的格式为<code>NAME=xxx</code>。</p>
<p>对于我们上一步处理的qry.gff和qry.fa，我们只需要简单粗暴地替换qry.gff里的信息即可。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sed -e <span class="string">&#x27;s/transcript/gene/&#x27;</span> -e <span class="string">&#x27;s/ID/Name/&#x27;</span> qry.gff &gt; qry_gene.gff</span><br></pre></td></tr></table></figure>

<p>对于近缘物种的cds序列和gff文件，考虑到后续的脚本只认gene所在的行，而cds里面存放的是转录本序列，我们可以提取mRNA所在行，然后将其改名成gene。假设参考序列是<code>ref.fa</code>,  参考的GFF文件名是<code>ref.gff</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">ref=ref.gff</span><br><span class="line">grep <span class="string">&#x27;[[:blank:]]mRNA[[:blank:]]&#x27;</span> <span class="variable">$ref</span> | sed -e <span class="string">&#x27;s/mRNA/gene/&#x27;</span> -e <span class="string">&#x27;s/ID/Name/&#x27;</span> &gt; ref_gene.gff</span><br></pre></td></tr></table></figure>

<p>综上你得到了如下四个文件, ref.fa, ref_gene.gff, qry.fa, qry_gene.gff</p>
<p>下面的操作，参考自<a target="_blank" rel="noopener" href="https://github.com/tangerzhang/ALLHiC/wiki/ALLHiC:-identify-allelic-contigs">官方教程</a>完成，</p>
<p><strong>第一步</strong>： 建立BLAST索引</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">makeblastdb -<span class="keyword">in</span> ref.fa -dbtype nucl</span><br></pre></td></tr></table></figure>

<p><strong>第二步</strong>: 将我们物种序列比对到近缘物种</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">blastn -query qry.fa -db ref.fa -out qry_vs_ref.blast.out -evalue 0.001 -outfmt 6 -num_threads 4 -num_alignments 1</span><br></pre></td></tr></table></figure>

<p><strong>第三步</strong>: 过滤低质量的HIT，此处标准是相似度低于60%, 覆盖度低于80%</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">blastn_parse.pl -i qry_vs_ref.blast.out -o Eblast.out -q qry.fa -b 1 -c 0.6 -d 0.8 </span><br></pre></td></tr></table></figure>

<p><strong>第四步</strong>: 基于BLAST结果鉴定等位序列</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">classify.pl -i Eblast.out -p 4 -r ref_gene.gff -g qry_gene.gff</span><br><span class="line"><span class="comment"># -p 是等位基因的数目</span></span><br></pre></td></tr></table></figure>

<p>最终输出的Allele.ctg.table就是后续ALLHiC的prune输入。</p>
<h2 id="方法二-基于GMAP"><a href="#方法二-基于GMAP" class="headerlink" title="方法二:  基于GMAP"></a>方法二:  基于GMAP</h2><p>第二个方法比较简单，只需要提供多倍体基因组的基因组序列和近缘物种的cds序列即可</p>
<p><strong>第一步</strong>: 运行GMAP获取GFF3文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">qry=target.genome  <span class="comment"># 你的多倍体序列</span></span><br><span class="line">ref_cds=reference.cds.fasta <span class="comment"># 同源物种的cds</span></span><br><span class="line">N=4 <span class="comment">#你的基因组倍性, 4表示的是4倍体</span></span><br><span class="line">gmap_build -D . -d DB <span class="variable">$qry</span></span><br><span class="line">gmap -D . -d DB -t 12 -f 2 -n <span class="variable">$N</span>  <span class="variable">$ref_cds</span> &gt; gmap.gff3</span><br></pre></td></tr></table></figure>

<p><strong>第二步</strong>: 获取  allelic.ctg.table </p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl gmap2AlleleTable.pl referenece.gff3</span><br></pre></td></tr></table></figure>

<p> gmap2AlleleTable.pl脚本是ALLHiC的一部分。</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li><a target="_blank" rel="noopener" href="https://github.com/tangerzhang/ALLHiC/issues/16">https://github.com/tangerzhang/ALLHiC/issues/16</a> </li>
<li><a target="_blank" rel="noopener" href="https://github.com/tangerzhang/ALLHiC/wiki/ALLHiC:-identify-allelic-contigs">https://github.com/tangerzhang/ALLHiC/wiki/ALLHiC:-identify-allelic-contigs</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/30/ALLHiC%E7%BB%AD_%20%E5%A6%82%E4%BD%95%E6%9E%84%E5%BB%BAAllele.ctg.table-How-to-build-allele-ctg-table/" data-id="clm1z22eq005dehn5d39zadym" data-title="ALLHiC续:如何构建Allele.ctg.table" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/15/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/14/">14</a><a class="page-number" href="/page/15/">15</a><span class="page-number current">16</span><a class="page-number" href="/page/17/">17</a><a class="page-number" href="/page/18/">18</a><span class="space">&hellip;</span><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/17/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/09/">九月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/09/02/%E5%A6%82%E4%BD%95%E8%AE%A9%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8D%E5%86%8D%E9%BE%9F%E9%80%9F%E4%B8%8B%E8%BD%BDgithub%E7%9A%84%E6%95%B0%E6%8D%AE/">如何让服务器不再龟速下载github的数据</a>
          </li>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>