<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/17/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-使用ALLHiC基于HiC数据辅助基因组组装-Assemble-genome-using-ALLHiC-with-HiC-Data" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/28/%E4%BD%BF%E7%94%A8ALLHiC%E5%9F%BA%E4%BA%8EHiC%E6%95%B0%E6%8D%AE%E8%BE%85%E5%8A%A9%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-Assemble-genome-using-ALLHiC-with-HiC-Data/" class="article-date">
  <time class="dt-published" datetime="2019-10-28T22:30:59.259Z" itemprop="datePublished">2019-10-28</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/28/%E4%BD%BF%E7%94%A8ALLHiC%E5%9F%BA%E4%BA%8EHiC%E6%95%B0%E6%8D%AE%E8%BE%85%E5%8A%A9%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-Assemble-genome-using-ALLHiC-with-HiC-Data/">使用ALLHiC基于HiC数据辅助基因组组装</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>基因组组装大致可以分为三步（1）根据序列之间的重叠情况构建出contig，（2）基于二代的mate pair文库或光学图谱将contig搭建成scaffold，（3）对scaffold进行排序和调整方向得到最终的准染色体级别的基因组。</p>
<p>目前的三代测序组装能够搞定第一步和第二步。而在将contig&#x2F;scaffold提升至准染色体水平上，有4种方案可选。一种是基于遗传图谱，一种是利用BioNano DLS光学图谱，一种是利用近缘物种的染色体同源性，还有一种就是HiC。其中HiC技术是三者中较为简单的一个，不需要高质量的DNA文库，也不需要一个很大的群体，结果也比较准确可信。</p>
<p>HiC的文库构建示意图如下，我们所需要的就是最终双端测序的两端序列之间的距离关系。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/hi-c-3c-seq-capture-c-9f0946eb49b74725af32b32e0e57bad5.png" alt="图片来自于Illumina"></p>
<p>目前利用HiC数据进行组装软件有 LACHESIS, HiRise, SALSA1, 3D-DNA等，这些软件在动物基因组上和简单植物基因组上表现都不错，但是不太适合直接用于多倍体物种和高杂合物种的组装上。主要原因就是等位基因序列的相似性，使得不同套染色体之间的contig出现了假信号，最终错误地将不同套染色体的contig连在了一起。最近在Nature Plants发表的ALLHiC流程就是用来解决多倍体物种和高杂合度基因组的HiC组装难题。</p>
<h2 id="ALLHiC流程一览"><a href="#ALLHiC流程一览" class="headerlink" title="ALLHiC流程一览"></a>ALLHiC流程一览</h2><p>ALLHiC一共分为五步(见下图，Zhang et al., 2019)，pruning, partition, rescue, optimization, building，要求的输入文件为HiC数据比对后的BAM和一个Allele.ctg.table。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/image-20191028095337641-8e3338626c334a8cb9795cb96f020d62.png" alt="算法过程"></p>
<p>其中pruning步骤是ALLHiC区别于其他软件的关键一步。因此我专门将其挑选出来进行介绍，红色实线是潜在的坍缩区域（组装时因为序列高度相似而没有拆分），而其他颜色实线则是不同的单倍型（我用浅灰色椭圆进行区分）。粉红色虚线指的是等位基因间的HiC信号，而黑色虚线则是坍缩区域和未坍缩区域的HiC信号。</p>
<p>ALLHiC在这一步会根据提供的Allele.ctg.table过滤BAM文件中等位基因间的HiC信号，同时筛选出坍缩区域和未坍缩区域的HiC信号。这些信号会用于Rescue步骤，将未锚定contig分配到已分组的contigs群。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/image-20191028132758309-53e1f2a9f127465ba8a902441cd69d02.png" alt="Pruning"></p>
<h2 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h2><p>ALLHiC的安装非常简单，按照习惯，我将软件安装在<code>~/opt/biosoft</code>下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p ~/opt/biosoft &amp;&amp; <span class="built_in">cd</span> ~/opt/biosoft </span><br><span class="line">git <span class="built_in">clone</span> https://github.com/tangerzhang/ALLHiC</span><br><span class="line"><span class="built_in">cd</span> ALLHiC</span><br><span class="line"><span class="built_in">mv</span> allhic.v0.9.8 bin/allhic</span><br><span class="line"><span class="built_in">chmod</span> +x bin/*</span><br><span class="line"><span class="built_in">chmod</span> +x scripts/*  </span><br><span class="line"><span class="comment"># 添加到环境变量</span></span><br><span class="line">PATH=<span class="variable">$HOME</span>/opt/biosoft/ALLHiC/scripts/:<span class="variable">$HOME</span>/opt/biosoft/ALLHiC/bin/:<span class="variable">$PATH</span></span><br><span class="line"><span class="built_in">export</span> PATH</span><br></pre></td></tr></table></figure>

<p>此外ALLHiC还依赖于samtools(v1.9), bedtools 和 Python 3环境的matplotlib(v2.0+)，这些可以通过conda一步搞定。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda create -y -n allhic python=3.7 samtools bedtools matplotlib</span><br></pre></td></tr></table></figure>

<p>之后检查下是否成功安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ conda activate allhic</span><br><span class="line">$ allhic -v</span><br><span class="line">$ ALLHiC_prune</span><br></pre></td></tr></table></figure>

<p>你可能会遇到如下的报错</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ALLHiC_prune: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21<span class="string">&#x27; not found (required by ALLHiC_prune)</span></span><br></pre></td></tr></table></figure>

<p>这是GLIBC过低导致，但是不要尝试动手去升级GLIBC（你承担不起后果的），conda提供了一个比较新的动态库，因此可以通过如下方法来解决问题</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">export</span> LD_LIBRARY_PATH=<span class="variable">$LD_LIBRARY_PATH</span>:~/opt/miniconda3/lib</span><br></pre></td></tr></table></figure>

<h2 id="ALLHiC分析实战"><a href="#ALLHiC分析实战" class="headerlink" title="ALLHiC分析实战"></a>ALLHiC分析实战</h2><p>感谢张兴坦老师提供的测试数据，我将其中的BAM转成了原始FastQ格式，以便从头讲解。数据在<a target="_blank" rel="noopener" href="https://pan.baidu.com/s/1_7VI127eCmYaO_3ukfN67Q">百度网盘</a>,提取码：17yk</p>
<p>输入文件需要有4个，contig序列，等位contig信息和两个双端测序数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">ls</span> -1</span><br><span class="line">Allele.ctg.table</span><br><span class="line">draft.asm.fasta</span><br><span class="line">reads_R1.fastq.gz</span><br><span class="line">reads_R2.fastq.gz</span><br></pre></td></tr></table></figure>

<p>第一步: 建立索引</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">samtools faidx draft.asm.fasta </span><br><span class="line">bwa index -a bwtsw draft.asm.fasta  </span><br></pre></td></tr></table></figure>

<p>第二步: 序列回贴。这一步的限速步骤是<code>bwa sampe</code>，因为它没有多线程参数。如果数据量很大，可以先将原始的fastq数据进行拆分，分别比对后分别执行<code>bwa sampe</code>，最后合并成单个文件。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">bwa aln -t 24 draft.asm.fasta reads_R1.fastq.gz &gt; reads_R1.sai  </span><br><span class="line">bwa aln -t 24 draft.asm.fasta reads_R2.fastq.gz &gt; reads_R2.sai  </span><br><span class="line">bwa sampe draft.asm.fasta reads_R1.sai reads_R2.sai reads_R1.fastq.gz reads_R2.fastq.gz &gt; sample.bwa_aln.sam  </span><br></pre></td></tr></table></figure>

<p>第三步: SAM预处理，移除冗余和低质量信号，提高处理效率</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">PreprocessSAMs.pl sample.bwa_aln.sam draft.asm.fasta MBOI</span><br><span class="line"><span class="comment"># 如果已有BAM文件</span></span><br><span class="line"><span class="comment"># PreprocessSAMs.pl sample.bwa_aln.bam draft.asm.fasta MBOI</span></span><br><span class="line">filterBAM_forHiC.pl sample.bwa_aln.REduced.paired_only.bam sample.clean.sam</span><br><span class="line">samtools view -bt draft.asm.fasta.fai sample.clean.sam &gt; sample.clean.bam</span><br></pre></td></tr></table></figure>

<p>其中<code>filterBAM_forHiC.pl</code>的过滤标准是比对质量高于30(MQ), 只保留唯一比对(XT:A:U), 编辑距离(NM)低于5, 错误匹配低于(XM)4,  不能有超过2个的gap(XO,XG)</p>
<p>第四步(可选):  对于多倍体或者是高杂合的基因组，因为等位基因的序列相似性高，那么很有可能会在不同套基因组间出现假信号，因此需要<a target="_blank" rel="noopener" href="https://github.com/tangerzhang/ALLHiC/wiki/ALLHiC:-identify-allelic-contigs">构建Allele.ctg.table</a>, 用于过滤这种假信号。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ALLHiC_prune -i Allele.ctg.table -b sample.clean.bam -r draft.asm.fasta  </span><br></pre></td></tr></table></figure>

<p>这一步生成<code>prunning.bam</code>用于后续分析</p>
<p>第五步: 这一步是根据HiC信号将不同的contig进行分组，分组数目由<code>-k</code>控制。如果跳过了第四步，那么可以直接用第三步的结果<code>sample.clean.bam</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ALLHiC_partition -b prunning.bam -r draft.asm.fasta -e AAGCTT -k 16  </span><br></pre></td></tr></table></figure>

<p>这一步会生成一系列以<code>prunning</code>开头的文件</p>
<ul>
<li>分组信息: prunning.clusters.txt</li>
<li>每个分组对应的contig: prunning.counts_AAGCTT.XXgYY.txt: </li>
<li>每个contig长度和count数: prunning.counts_AAGCTT.txt</li>
</ul>
<p>第六步: 将未锚定的contig分配已有的分组中。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ALLHiC_rescue -b sample.clean.bam -r draft.asm.fasta \</span><br><span class="line">    -c prunning.clusters.txt \</span><br><span class="line">    -i prunning.counts_AAGCTT.txt</span><br></pre></td></tr></table></figure>

<p>这一步根据之前prunning.counts_AAGCTT.XXgYY.txt对应的groupYY.txt</p>
<p>第七步: 优化每一组中的contig的排序和方向</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 生成.clm文件</span></span><br><span class="line">allhic extract sample.clean.bam draft.asm.fasta --RE AAGCTT  </span><br><span class="line"><span class="comment"># 优化</span></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> group*.txt; <span class="keyword">do</span></span><br><span class="line">    allhic optimize <span class="variable">$i</span> sample.clean.clm</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<p>这一步会基于groupYY.txt生成对应的groupYY.tour</p>
<p>第八步: 将tour格式转成fasta格式，并生成对应的agp。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ALLHiC_build draft.asm.fasta  </span><br></pre></td></tr></table></figure>

<p>这一步生成两个文件，groups.asm.fasta和groups.agp。其中groups.asm.fasta就是我们需要的结果。</p>
<p>第九步: 构建染色质交互矩阵，根据热图评估结果</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">samtools faidx groups.asm.fasta</span><br><span class="line"><span class="built_in">cut</span> -f 1,2 groups.asm.fasta.fai  &gt; chrn.list</span><br><span class="line">ALLHiC_plot sample.clean.bam groups.agp chrn.list 500k pdf</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/image-20191028191408460-bd640b324fbc49a5a1044e56970d664c.png" alt="heatmap"></p>
<p>使用ALLHiC的几个注意事项:</p>
<ol>
<li>ALLHiC依赖于初始的contig，如果嵌合序列和坍缩序列比例过高，那么ALLHiC结果也会不准确。根据文章，ALLHiC能够处理<code>~10%</code>的嵌合比例，<code>~20%</code>的坍缩比例。因此最好是用类似于Canu这种能够区分单倍型的组装软件。</li>
<li>单倍型之间序列相似度不能太高，否则会出现大量非唯一比对，降低可用的HiC信号</li>
<li>构建Allele.ctg.table需要一个比较近缘的高质量基因组</li>
<li>不要用过短的contig，因为短的contig信号少，很容易放到错误的区域</li>
<li>K值的设置要根据实际的基因组数目设置，如果你发现输出结果中某些group过大，可以适当增大k值</li>
</ol>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li><a target="_blank" rel="noopener" href="https://github.com/tanghaibao/allhic">https://github.com/tanghaibao/allhic</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/tangerzhang/ALLHiC/wiki">https://github.com/tangerzhang/ALLHiC/wiki</a></li>
<li>Zhang, X., Zhang, S., Zhao, Q., Ming, R., and Tang, H. (2019). Assembly of allele-aware, chromosomal-scale autopolyploid genomes based on Hi-C data. Nat. Plants <em>5</em>, 833–845.</li>
<li>Zhang, J., Zhang, X., Tang, H., Zhang, Q., Hua, X., Ma, X., Zhu, F., Jones, T., Zhu, X., Bowers, J., et al. (2018). Allele-defined genome of the autopolyploid sugarcane Saccharum spontaneum L. Nature Genetics <em>50</em>, 1565.</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/28/%E4%BD%BF%E7%94%A8ALLHiC%E5%9F%BA%E4%BA%8EHiC%E6%95%B0%E6%8D%AE%E8%BE%85%E5%8A%A9%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-Assemble-genome-using-ALLHiC-with-HiC-Data/" data-id="clm1ywdje00b0ecokfhqg32nh" data-title="使用ALLHiC基于HiC数据辅助基因组组装" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用nextpolish对三代组装进行polish-Polish-Assembly-Generated-From-TGS-Assembler-with-NextPolish" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/25/%E4%BD%BF%E7%94%A8nextpolish%E5%AF%B9%E4%B8%89%E4%BB%A3%E7%BB%84%E8%A3%85%E8%BF%9B%E8%A1%8Cpolish-Polish-Assembly-Generated-From-TGS-Assembler-with-NextPolish/" class="article-date">
  <time class="dt-published" datetime="2019-10-25T11:25:33.711Z" itemprop="datePublished">2019-10-25</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/25/%E4%BD%BF%E7%94%A8nextpolish%E5%AF%B9%E4%B8%89%E4%BB%A3%E7%BB%84%E8%A3%85%E8%BF%9B%E8%A1%8Cpolish-Polish-Assembly-Generated-From-TGS-Assembler-with-NextPolish/">使用nextpolish对三代组装进行polish</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>NextPolish是武汉未来组开发的一个三代基因组polish工具（另外一个常用软件是Pilon）。NextPolish可以使用二代短读序列或者三代序列或者两者结合去纠正三代长读长序列在组装时导致的碱基错误(SNV&#x2F;Indel)。由于它是专为polish设计，因此在运行速度和内存使用上都优与Pilon。</p>
<h2 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h2><p>先确保自己的服务器上安装了Python2.7， 且有Shutil和Signal，或者你可以利用conda新建一个python2.7的环境。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># shell</span></span><br><span class="line">python -V</span><br><span class="line">Python 2.7.15</span><br><span class="line"><span class="comment"># Python 交互命令行</span></span><br><span class="line">import shutil</span><br><span class="line">import signal</span><br></pre></td></tr></table></figure>

<p>之后到<a target="_blank" rel="noopener" href="https://github.com/Nextomics/NextPolish/releases">https://github.com/Nextomics/NextPolish/releases</a>找最新的版本，写这篇时的最新版本就是1.05</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p ~/opt/biosoft</span><br><span class="line"><span class="built_in">cd</span> ~/opt/biosoft</span><br><span class="line">wget https://github.com/Nextomics/NextPolish/releases/download/v1.0.5/NextPolish.tgz</span><br><span class="line">tar -zxvf NextPolish.tgz</span><br><span class="line"><span class="comment"># 编译软件</span></span><br><span class="line"><span class="built_in">cd</span> NextPolish &amp;&amp; make -j 10</span><br><span class="line"><span class="comment"># 加入到.bashrc或.zshrc</span></span><br><span class="line"><span class="built_in">export</span> PATH=~/opt/biosoft/NextPolish:<span class="variable">$PATH</span></span><br></pre></td></tr></table></figure>

<h2 id="软件使用"><a href="#软件使用" class="headerlink" title="软件使用"></a>软件使用</h2><p><strong>注意</strong>：如果你的基因组用的是miniasm这类缺少consensus步骤的组装软件，那么你需要先用运行如下命令，或者是运行<code>racon</code>利用三代序列进行polish。否则，由于基因组上存在过高的错误率，导致二代序列错误比对，影响polish效果。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">threads=20  </span><br><span class="line">genome=input.genome.fa <span class="comment"># 组装的基因组</span></span><br><span class="line">lgsreads=input.lgs.reads.fq.gz <span class="comment"># 三代长度序列</span></span><br><span class="line"><span class="comment"># 将三代回帖到参考基因组</span></span><br><span class="line">minimap2 -a -t <span class="variable">$&#123;threads&#125;</span> -x map-ont/map-pb <span class="variable">$&#123;genome&#125;</span> <span class="variable">$&#123;lgsreads&#125;</span>| \</span><br><span class="line">    samtools view -F 0x4 -b - | \</span><br><span class="line">    samtools <span class="built_in">sort</span> - -m 2g -@ <span class="variable">$&#123;threads&#125;</span> -o genome.lgs.bam  </span><br><span class="line"><span class="comment">#建立索引</span></span><br><span class="line">samtools index -@ <span class="variable">$&#123;threads&#125;</span> genome.lgs.bam</span><br><span class="line">samtools faidx <span class="variable">$&#123;genome&#125;</span></span><br><span class="line"><span class="comment"># 使用nextPolish.py 进行polish</span></span><br><span class="line">python ~/opt/biosoft/NextPolish/lib/nextPolish.py \</span><br><span class="line">    -g <span class="variable">$&#123;genome&#125;</span> -t 5 --bam_lgs genome.lgs.bam -p <span class="variable">$&#123;threads&#125;</span> &gt; genome.lgspolish.fa</span><br></pre></td></tr></table></figure>

<p>生成的genome.lgspolish.fa才能用于后续的二代polish步骤。</p>
<p>NextPolish要求我们需要准备两个文件：</p>
<ul>
<li><code>run.cfg</code>: 配置文件，</li>
<li><code>sgs.fofn</code>: 二代测序文件的位置信息</li>
</ul>
<p>以<a href="/archives/Assembly-nanopore-with-NextDenovo">使用NextDenovo组装Nanopore数据</a>文章组装的结果为例进行介绍。在分析目录下有三个文件。</p>
<ul>
<li>三代组装结果: nextgraph.assembly.contig.fasta</li>
<li>二代序列: ERR2173372_1.fastq,ERR2173372_2.fastq</li>
</ul>
<p>第一步：创建一个文件，用于记录二代序列的位置信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">realpath</span> ERR2173372_1.fastq ERR2173372_2.fastq  &gt; sgs.fofn</span><br></pre></td></tr></table></figure>

<p>第二步：配置run.cfg文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 从NextPolish目录下复制配置文件</span></span><br><span class="line"><span class="built_in">cp</span> ~/opt/biosoft/NextPolish/doc/run.cfg run2.cfg</span><br></pre></td></tr></table></figure>

<p>修改配置文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">[General]</span><br><span class="line">job_type = <span class="built_in">local</span></span><br><span class="line">job_prefix = nextPolish</span><br><span class="line">task = default</span><br><span class="line">rewrite = 1212</span><br><span class="line">rerun = 3</span><br><span class="line">parallel_jobs = 2</span><br><span class="line">multithread_jobs = 10</span><br><span class="line">genome = ./nextgraph.assembly.contig.fasta</span><br><span class="line">genome_size = auto</span><br><span class="line">workdir = ./01_rundir</span><br><span class="line">polish_options = -p &#123;multithread_jobs&#125;</span><br><span class="line"></span><br><span class="line">[sgs_option]</span><br><span class="line">sgs_fofn = ./sgs.fofn</span><br><span class="line">sgs_options = -max_depth 100</span><br></pre></td></tr></table></figure>

<p>其中需要修改的参数为，其余参数查看<a target="_blank" rel="noopener" href="https://github.com/Nextomics/NextPolish/blob/master/doc/OPTION.md">官方的参数配置说明</a>:</p>
<ul>
<li><code>job_type</code>: 任务类型，local表示单个节点运行。由于NextPolish使用DRMAA进行任务投递，因此还支持，SGE, PBS和SLURM</li>
<li><code>task</code>: 任务类型， 用12,1212,121212,12121212来设置polish的轮数，建议迭代2轮就可以了。</li>
<li><code>parallel_jobs</code>和<code>multithread_jobs</code>表示同时投递的任务数和每个任务的线程数，此处2 X 10&#x3D;20</li>
<li><code>genome</code>: 表示组装基因组的位置</li>
<li><code>workdir</code>: 输出文件所在目录</li>
<li><code>sgs_options</code>: 该选项设置二代测序polish的参数，包括-use_duplicate_reads, -unpaired, -max_depth, -bwa, -minimap2(默认使用)</li>
</ul>
<p>运行方法</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">nextPolish run2.cfg &amp;</span><br></pre></td></tr></table></figure>

<p>在最后输出日志中，会提示最终存放的文件在什么位置，然后将这些文件合并到单个文件即可。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/25/%E4%BD%BF%E7%94%A8nextpolish%E5%AF%B9%E4%B8%89%E4%BB%A3%E7%BB%84%E8%A3%85%E8%BF%9B%E8%A1%8Cpolish-Polish-Assembly-Generated-From-TGS-Assembler-with-NextPolish/" data-id="clm1ywdkm00ekecokhbxmfhz1" data-title="使用nextpolish对三代组装进行polish" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-「热图」ComplexHeatmap展示单细胞聚类-ComplexHeatmap-for-single-cell-visualization" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/24/%E3%80%8C%E7%83%AD%E5%9B%BE%E3%80%8DComplexHeatmap%E5%B1%95%E7%A4%BA%E5%8D%95%E7%BB%86%E8%83%9E%E8%81%9A%E7%B1%BB-ComplexHeatmap-for-single-cell-visualization/" class="article-date">
  <time class="dt-published" datetime="2019-10-24T19:49:27.742Z" itemprop="datePublished">2019-10-24</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/24/%E3%80%8C%E7%83%AD%E5%9B%BE%E3%80%8DComplexHeatmap%E5%B1%95%E7%A4%BA%E5%8D%95%E7%BB%86%E8%83%9E%E8%81%9A%E7%B1%BB-ComplexHeatmap-for-single-cell-visualization/">「热图」ComplexHeatmap展示单细胞聚类</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>实用Seurat自带的热图函数<code>DoHeatmap</code>绘制的热图，感觉有点不上档次，于是我尝试使用<code>ComplexHeatmap</code>这个R包来对结果进行展示。</p>
<p>个人觉得好的热图有三个要素</p>
<ul>
<li>聚类: 能够让别人一眼就看到模式</li>
<li>注释: 附加注释能提供更多信息</li>
<li>配色: 要符合直觉，比如说大部分都会认为红色是高表达，蓝色是低表达</li>
</ul>
<p>在正式开始之前，我们需要先获取一下pbmc的数据，Seurat提供了R包<code>SeuratData</code>专门用于获取数据</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">devtools<span class="operator">::</span>install_github<span class="punctuation">(</span><span class="string">&#x27;satijalab/seurat-data&#x27;</span><span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>SeuratData<span class="punctuation">)</span></span><br><span class="line">InstallData<span class="punctuation">(</span><span class="string">&quot;pbmc3k&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>加载数据并进行数据预处理，获取绘制热图所需的数据</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>SeuratData<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>Seurat<span class="punctuation">)</span></span><br><span class="line">data<span class="punctuation">(</span><span class="string">&quot;pbmc3k&quot;</span><span class="punctuation">)</span></span><br><span class="line">pbmc <span class="operator">&lt;-</span> pbmc3k</span><br><span class="line">pbmc<span class="punctuation">[[</span><span class="string">&quot;percent.mt&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> PercentageFeatureSet<span class="punctuation">(</span>pbmc<span class="punctuation">,</span> pattern <span class="operator">=</span> <span class="string">&quot;^MT-&quot;</span><span class="punctuation">)</span></span><br><span class="line">pbmc <span class="operator">&lt;-</span> subset<span class="punctuation">(</span>pbmc<span class="punctuation">,</span> subset <span class="operator">=</span> nFeature_RNA <span class="operator">&gt;</span> <span class="number">200</span> <span class="operator">&amp;</span> nFeature_RNA <span class="operator">&lt;</span> <span class="number">2500</span> <span class="operator">&amp;</span> percent.mt <span class="operator">&lt;</span> <span class="number">5</span><span class="punctuation">)</span></span><br><span class="line">pbmc <span class="operator">&lt;-</span> NormalizeData<span class="punctuation">(</span>pbmc<span class="punctuation">,</span> normalization.method <span class="operator">=</span> <span class="string">&quot;LogNormalize&quot;</span><span class="punctuation">,</span> scale.factor <span class="operator">=</span> <span class="number">10000</span><span class="punctuation">)</span></span><br><span class="line">pbmc <span class="operator">&lt;-</span> FindVariableFeatures<span class="punctuation">(</span>pbmc<span class="punctuation">,</span> selection.method <span class="operator">=</span> <span class="string">&quot;vst&quot;</span><span class="punctuation">,</span> nfeatures <span class="operator">=</span> <span class="number">2000</span><span class="punctuation">)</span></span><br><span class="line">all.genes <span class="operator">&lt;-</span> rownames<span class="punctuation">(</span>pbmc<span class="punctuation">)</span></span><br><span class="line">pbmc <span class="operator">&lt;-</span> ScaleData<span class="punctuation">(</span>pbmc<span class="punctuation">,</span> features <span class="operator">=</span> all.genes<span class="punctuation">)</span></span><br><span class="line">pbmc <span class="operator">&lt;-</span> RunPCA<span class="punctuation">(</span>pbmc<span class="punctuation">,</span> features <span class="operator">=</span> VariableFeatures<span class="punctuation">(</span>object <span class="operator">=</span> pbmc<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">pbmc <span class="operator">&lt;-</span> FindNeighbors<span class="punctuation">(</span>pbmc<span class="punctuation">,</span> dims <span class="operator">=</span> <span class="number">1</span><span class="operator">:</span><span class="number">10</span><span class="punctuation">)</span></span><br><span class="line">pbmc <span class="operator">&lt;-</span> FindClusters<span class="punctuation">(</span>pbmc<span class="punctuation">,</span> resolution <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">pbmc.markers <span class="operator">&lt;-</span> FindAllMarkers<span class="punctuation">(</span>pbmc<span class="punctuation">,</span> </span><br><span class="line">                               only.pos <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> </span><br><span class="line">                               min.pct <span class="operator">=</span> <span class="number">0.25</span><span class="punctuation">,</span> </span><br><span class="line">                               logfc.threshold <span class="operator">=</span> <span class="number">0.25</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>先感受下Seurat自带热图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">top10 <span class="operator">&lt;-</span> pbmc.markers <span class="operator">%&gt;%</span> group_by<span class="punctuation">(</span>cluster<span class="punctuation">)</span> <span class="operator">%&gt;%</span> top_n<span class="punctuation">(</span>n <span class="operator">=</span> <span class="number">10</span><span class="punctuation">,</span> wt <span class="operator">=</span> avg_logFC<span class="punctuation">)</span></span><br><span class="line">DoHeatmap<span class="punctuation">(</span>pbmc<span class="punctuation">,</span> features <span class="operator">=</span> top10<span class="operator">$</span>gene<span class="punctuation">)</span> <span class="operator">+</span> NoLegend<span class="punctuation">(</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571906906510-5666cad1b38f473b8b8896f9ebde28d3.png" alt="Seurat-heatmap"></p>
<p>下面则是介绍如何用R包<code>ComplexHeatmap</code>进行组图，虽然这个R包名带着Complex，但是并不是说这个R包很复杂，这个Complex应该翻译成复合，也就是说这个R包能在热图的基础上整合很多信息。</p>
<p>先安装并加载R包。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&quot;ComplexHeatmap&quot;</span><span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>ComplexHeatmap<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>为了手动绘制一个热图，要从Seurat对象中提取所需要的表达量矩阵。我提取的是原始的count值，然后用<code>log2(count + 1)</code>的方式进行标准化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">mat <span class="operator">&lt;-</span> GetAssayData<span class="punctuation">(</span>pbmc<span class="punctuation">,</span> slot <span class="operator">=</span> <span class="string">&quot;counts&quot;</span><span class="punctuation">)</span></span><br><span class="line">mat <span class="operator">&lt;-</span> log2<span class="punctuation">(</span>mat <span class="operator">+</span> <span class="number">1</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>获取基因和细胞聚类信息</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">gene_features <span class="operator">&lt;-</span> top10</span><br><span class="line">cluster_info <span class="operator">&lt;-</span> sort<span class="punctuation">(</span>pbmc<span class="operator">$</span>seurat_annotations<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>对表达量矩阵进行排序和筛选</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mat <span class="operator">&lt;-</span> as.matrix<span class="punctuation">(</span>mat<span class="punctuation">[</span>top10<span class="operator">$</span>gene<span class="punctuation">,</span> <span class="built_in">names</span><span class="punctuation">(</span>cluster_info<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>用<code>Heatmap</code>绘制热图。对于单细胞这种数据，一定要设置如下4个参数</p>
<ul>
<li><code>cluster_rows= FALSE</code>: 不作行聚类</li>
<li><code>cluster_columns= FALSE</code>: 不作列聚类</li>
<li><code>show_column_names=FALSE</code>: 不展示列名</li>
<li><code>show_row_names=FALSE</code>: 不展示行名，基因数目不多时候可以考虑设置为TRUE</li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">Heatmap<span class="punctuation">(</span>mat<span class="punctuation">,</span></span><br><span class="line">        cluster_rows <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        cluster_columns <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_column_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_row_names <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571911526690-f312b4bf617f4e68bb539556872fe405.png" alt="Heatmap-1"></p>
<p>从图中，我们可以发现以下几个问题：</p>
<ul>
<li>长宽比不合理，当然这和绘图函数无关，可以在保存时修改长宽比</li>
<li>基因名重叠，考虑调整大小，或者不展示，或者只展示重要的基因</li>
<li>颜色可以调整</li>
<li>缺少聚类信息</li>
</ul>
<p>这些问题，我们可以通过在<a target="_blank" rel="noopener" href="https://jokergoo.github.io/ComplexHeatmap-reference/book/">ComplexHeatmap Complete Reference</a>查找对应信息来解决。</p>
<h2 id="配色方案"><a href="#配色方案" class="headerlink" title="配色方案"></a>配色方案</h2><p>在热图中会涉及到两类配色，一种用来表示表达量的连续性变化，一种则是展示聚类。有一个神奇的R包就是用于处理配色，他的Github地址为<a target="_blank" rel="noopener" href="https://github.com/caleblareau/BuenColors">https://github.com/caleblareau/BuenColors</a>。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">devtools::install_github(<span class="string">&quot;caleblareau/BuenColors&quot;</span>)</span><br><span class="line">library(<span class="string">&quot;BuenColors&quot;</span>)</span><br></pre></td></tr></table></figure>

<p>它提供了一些列预设的颜色，比方说<code>jdb_color_maps</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">      HSC       MPP      LMPP       CMP       CLP       MEP       GMP </span><br><span class="line"><span class="string">&quot;#00441B&quot;</span> <span class="string">&quot;#46A040&quot;</span> <span class="string">&quot;#00AF99&quot;</span> <span class="string">&quot;#FFC179&quot;</span> <span class="string">&quot;#98D9E9&quot;</span> <span class="string">&quot;#F6313E&quot;</span> <span class="string">&quot;#FFA300&quot;</span> </span><br><span class="line">      pDC      mono     GMP<span class="operator">-</span>A     GMP<span class="operator">-</span>B     GMP<span class="operator">-</span>C       Ery       CD4 </span><br><span class="line"><span class="string">&quot;#C390D4&quot;</span> <span class="string">&quot;#FF5A00&quot;</span> <span class="string">&quot;#AFAFAF&quot;</span> <span class="string">&quot;#7D7D7D&quot;</span> <span class="string">&quot;#4B4B4B&quot;</span> <span class="string">&quot;#8F1336&quot;</span> <span class="string">&quot;#0081C9&quot;</span> </span><br><span class="line">      CD8        NK         B </span><br><span class="line"><span class="string">&quot;#001588&quot;</span> <span class="string">&quot;#490C65&quot;</span> <span class="string">&quot;#BA7FD0&quot;</span></span><br></pre></td></tr></table></figure>

<p>这些颜色就能用于命名单细胞的类群，比如说我选择了前9个</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">col <span class="operator">&lt;-</span> jdb_color_maps<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">9</span><span class="punctuation">]</span></span><br><span class="line"><span class="built_in">names</span><span class="punctuation">(</span>col<span class="punctuation">)</span> <span class="operator">&lt;-</span> levels<span class="punctuation">(</span>cluster_info<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="增加列聚类信息"><a href="#增加列聚类信息" class="headerlink" title="增加列聚类信息"></a>增加列聚类信息</h2><p><code>Heatmap</code>的<code>row_split</code>和<code>column_split</code>参数可以通过设置分类变量对热图进行分隔。更多对热图进行拆分，可以参考<a target="_blank" rel="noopener" href="https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#heatmap-split">Heatmap split</a></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Heatmap<span class="punctuation">(</span>mat<span class="punctuation">,</span></span><br><span class="line">        cluster_rows <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        cluster_columns <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_column_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_row_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        column_split <span class="operator">=</span> cluster_info<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571911799187-41f8f8f51c7f4ee297fe3e29238daad1.png" alt="Heatmap-2"></p>
<p>只用文字描述可能不够好看，最好是带有颜色的分块图，其中里面的颜色和t-SNE或UMAP聚类颜色一致，才能更好的展示信息。</p>
<p>为了增加聚类注释，我们需要用到<code>HeatmapAnnotation</code>函数，它对细胞的列进行注释，而<code>rowAnnotation</code>函数可以对行进行注释。这两个函数能够增加各种类型的注释，包括条形图，点图，折线图，箱线图，密度图等等，这些函数的特征是<code>anno_xxx</code>，例如<code>anno_block</code>就用来绘制区块图。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">top_anno <span class="operator">&lt;-</span> HeatmapAnnotation<span class="punctuation">(</span></span><br><span class="line">  cluster <span class="operator">=</span> anno_block<span class="punctuation">(</span>gp <span class="operator">=</span> gpar<span class="punctuation">(</span>fill <span class="operator">=</span> col<span class="punctuation">)</span><span class="punctuation">,</span> <span class="comment"># 设置填充色</span></span><br><span class="line">                       labels <span class="operator">=</span> levels<span class="punctuation">(</span>cluster_info<span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">                       labels_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>cex <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span> col <span class="operator">=</span> <span class="string">&quot;white&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment"># 设置字体</span></span><br></pre></td></tr></table></figure>

<p>其中<code>anno_block</code>中的<code>gp</code>参数用于设置各类<strong>图形参数</strong>，<code>labels</code>设置标签，<code>labels_gp</code>设置和标签相关的<strong>图形参数</strong>。可以用<code>?gp</code>来了解有哪些<strong>图形参数</strong>。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">Heatmap<span class="punctuation">(</span>mat<span class="punctuation">,</span></span><br><span class="line">        cluster_rows <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        cluster_columns <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_column_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_row_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        column_split <span class="operator">=</span> cluster_info<span class="punctuation">,</span></span><br><span class="line">        top_annotation <span class="operator">=</span> top_anno<span class="punctuation">,</span> <span class="comment"># 在热图上边增加注释</span></span><br><span class="line">        column_title <span class="operator">=</span> <span class="literal">NULL</span> <span class="punctuation">)</span> <span class="comment"># 不需要列标题</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571914037258-69b71edb4e354f759b9f07674e714f13.png" alt="Heatmap-3"></p>
<h2 id="突出重要基因"><a href="#突出重要基因" class="headerlink" title="突出重要基因"></a>突出重要基因</h2><p>由于基因很多直接展示出来，根本看不清，我们可以强调几个标记基因。用到两个函数是<code>rowAnnotation</code>和<code>anno_mark</code></p>
<p>已知不同类群的标记基因如下</p>
<table>
<thead>
<tr>
<th align="left">Cluster ID</th>
<th align="left">Markers</th>
<th align="left">Cell Type</th>
</tr>
</thead>
<tbody><tr>
<td align="left">0</td>
<td align="left">IL7R, CCR7</td>
<td align="left">Naive CD4+ T</td>
</tr>
<tr>
<td align="left">1</td>
<td align="left">IL7R, S100A4</td>
<td align="left">Memory CD4+</td>
</tr>
<tr>
<td align="left">2</td>
<td align="left">CD14, LYZ</td>
<td align="left">CD14+ Mono</td>
</tr>
<tr>
<td align="left">3</td>
<td align="left">MS4A1</td>
<td align="left">B</td>
</tr>
<tr>
<td align="left">4</td>
<td align="left">CD8A</td>
<td align="left">CD8+ T</td>
</tr>
<tr>
<td align="left">5</td>
<td align="left">FCGR3A, MS4A7</td>
<td align="left">FCGR3A+ Mono</td>
</tr>
<tr>
<td align="left">6</td>
<td align="left">GNLY, NKG7</td>
<td align="left">NK</td>
</tr>
<tr>
<td align="left">7</td>
<td align="left">FCER1A, CST3</td>
<td align="left">DC</td>
</tr>
<tr>
<td align="left">8</td>
<td align="left">PPBP</td>
<td align="left">Platelet</td>
</tr>
</tbody></table>
<p>我们需要给<code>anno_mark</code>提供基因所在行即可。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">mark_gene <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;IL7R&quot;</span><span class="punctuation">,</span><span class="string">&quot;CCR7&quot;</span><span class="punctuation">,</span><span class="string">&quot;IL7R&quot;</span><span class="punctuation">,</span><span class="string">&quot;S100A4&quot;</span><span class="punctuation">,</span><span class="string">&quot;CD14&quot;</span><span class="punctuation">,</span><span class="string">&quot;LYZ&quot;</span><span class="punctuation">,</span><span class="string">&quot;MS4A1&quot;</span><span class="punctuation">,</span><span class="string">&quot;CD8A&quot;</span><span class="punctuation">,</span><span class="string">&quot;FCGR3A&quot;</span><span class="punctuation">,</span><span class="string">&quot;MS4A7&quot;</span><span class="punctuation">,</span><span class="string">&quot;GNLY&quot;</span><span class="punctuation">,</span><span class="string">&quot;NKG7&quot;</span><span class="punctuation">,</span><span class="string">&quot;FCER1A&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CST3&quot;</span><span class="punctuation">,</span><span class="string">&quot;PPBP&quot;</span><span class="punctuation">)</span></span><br><span class="line">gene_pos <span class="operator">&lt;-</span> which<span class="punctuation">(</span>rownames<span class="punctuation">(</span>mat<span class="punctuation">)</span> <span class="operator">%in%</span> mark_gene<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">row_anno <span class="operator">&lt;-</span>  rowAnnotation<span class="punctuation">(</span>mark_gene <span class="operator">=</span> anno_mark<span class="punctuation">(</span>at <span class="operator">=</span> gene_pos<span class="punctuation">,</span> </span><br><span class="line">                                     labels <span class="operator">=</span> mark_gene<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>接着绘制热图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">Heatmap<span class="punctuation">(</span>mat<span class="punctuation">,</span></span><br><span class="line">        cluster_rows <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        cluster_columns <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_column_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_row_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        column_split <span class="operator">=</span> cluster_info<span class="punctuation">,</span></span><br><span class="line">        top_annotation <span class="operator">=</span> top_anno<span class="punctuation">,</span></span><br><span class="line">        right_annotation <span class="operator">=</span> row_anno<span class="punctuation">,</span></span><br><span class="line">        column_title <span class="operator">=</span> <span class="literal">NULL</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571914905222-c332ebab06ee4100838180f2cb0f698a.png" alt="Heatmap-4"></p>
<p>关于如何增加标记注释，参考<a target="_blank" rel="noopener" href="https://jokergoo.github.io/ComplexHeatmap-reference/book/heatmap-annotations.html#mark-annotation">mark-annotation</a></p>
<h2 id="调增图例位置"><a href="#调增图例位置" class="headerlink" title="调增图例位置"></a>调增图例位置</h2><p>目前的热图还有一个问题，也就是表示表达量范围的图例太占位置了，有两种解决方法</p>
<ul>
<li>参数设置<code>show_heatmap_legend=FALSE</code>直接删掉。</li>
<li>利用<code>heatmap_legend_param</code>参数更改样式</li>
</ul>
<p>我们根据<a target="_blank" rel="noopener" href="https://jokergoo.github.io/ComplexHeatmap-reference/book/legends.html">legends</a>这一节的内容进行一些调整</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">Heatmap<span class="punctuation">(</span>mat<span class="punctuation">,</span></span><br><span class="line">        cluster_rows <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        cluster_columns <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_column_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_row_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        column_split <span class="operator">=</span> cluster_info<span class="punctuation">,</span></span><br><span class="line">        top_annotation <span class="operator">=</span> top_anno<span class="punctuation">,</span></span><br><span class="line">        right_annotation <span class="operator">=</span> row_anno<span class="punctuation">,</span></span><br><span class="line">        column_title <span class="operator">=</span> <span class="literal">NULL</span><span class="punctuation">,</span></span><br><span class="line">        heatmap_legend_param <span class="operator">=</span> <span class="built_in">list</span><span class="punctuation">(</span></span><br><span class="line">          title <span class="operator">=</span> <span class="string">&quot;log2(count+1)&quot;</span><span class="punctuation">,</span></span><br><span class="line">          title_position <span class="operator">=</span> <span class="string">&quot;leftcenter-rot&quot;</span></span><br><span class="line">        <span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571915704401-c349679da08d48918a452ba49adbd6cb.png" alt="heatmap-5"></p>
<p>因为ComplextHeatmap是基于Grid图形系统，因此可以先绘制热图，然后再用<code>grid::draw</code>绘制图例，从而实现将条形图的位置移动到图中的任意位置。</p>
<p>先获取绘制热图的对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">p <span class="operator">&lt;-</span> Heatmap<span class="punctuation">(</span>mat<span class="punctuation">,</span></span><br><span class="line">        cluster_rows <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        cluster_columns <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_column_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        show_row_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">        column_split <span class="operator">=</span> cluster_info<span class="punctuation">,</span></span><br><span class="line">        top_annotation <span class="operator">=</span> top_anno<span class="punctuation">,</span></span><br><span class="line">        right_annotation <span class="operator">=</span> row_anno<span class="punctuation">,</span></span><br><span class="line">        column_title <span class="operator">=</span> <span class="literal">NULL</span><span class="punctuation">,</span></span><br><span class="line">        show_heatmap_legend <span class="operator">=</span> <span class="literal">FALSE</span></span><br><span class="line">        <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>根据<code>p@matrix_color_mapping</code>获取图例的颜色的设置，然后用<code>Legend</code>构建图例</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">col_fun  <span class="operator">&lt;-</span> circlize<span class="operator">::</span>colorRamp2<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> <span class="number">2</span> <span class="punctuation">,</span><span class="number">3</span><span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">                                 <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;#0000FFFF&quot;</span><span class="punctuation">,</span> <span class="string">&quot;#9A70FBFF&quot;</span><span class="punctuation">,</span> <span class="string">&quot;#D8C6F3FF&quot;</span><span class="punctuation">,</span> <span class="string">&quot;#FFC8B9FF&quot;</span><span class="punctuation">,</span> <span class="string">&quot;#FF7D5DFF&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">lgd <span class="operator">&lt;-</span>  Legend<span class="punctuation">(</span>col_fun <span class="operator">=</span> col_fun<span class="punctuation">,</span> </span><br><span class="line">               title <span class="operator">=</span> <span class="string">&quot;log2(count+1)&quot;</span><span class="punctuation">,</span> </span><br><span class="line">               title_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>col<span class="operator">=</span><span class="string">&quot;white&quot;</span><span class="punctuation">,</span> cex <span class="operator">=</span> <span class="number">0.75</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">               title_position <span class="operator">=</span> <span class="string">&quot;leftcenter-rot&quot;</span><span class="punctuation">,</span></span><br><span class="line">               <span class="comment">#direction = &quot;horizontal&quot;</span></span><br><span class="line">               at <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">               labels <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;low&quot;</span><span class="punctuation">,</span> <span class="string">&quot;median&quot;</span><span class="punctuation">,</span> <span class="string">&quot;high&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">               labels_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>col<span class="operator">=</span><span class="string">&quot;white&quot;</span><span class="punctuation">)</span></span><br><span class="line">               <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>绘制图形</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">grid.newpage<span class="punctuation">(</span><span class="punctuation">)</span> <span class="comment">#新建画布</span></span><br><span class="line">draw<span class="punctuation">(</span>p<span class="punctuation">)</span> <span class="comment"># 绘制热图</span></span><br><span class="line">draw<span class="punctuation">(</span>lgd<span class="punctuation">,</span> x <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">0.05</span><span class="punctuation">,</span> <span class="string">&quot;npc&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">     y <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">0.05</span><span class="punctuation">,</span> <span class="string">&quot;npc&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">     just <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;left&quot;</span><span class="punctuation">,</span> <span class="string">&quot;bottom&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment"># 绘制图形</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571916860482-418dc81dde834ac3a2ab4e65fc65a005.png" alt="heatmap-6"></p>
<p>ComplexHeatmap绘制热图非常强大的工具，大部分我想要的功能它都有，甚至我没有想到的它也有，这个教程只是展示其中一小部分功能而已，还有很多功能要慢慢探索。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/24/%E3%80%8C%E7%83%AD%E5%9B%BE%E3%80%8DComplexHeatmap%E5%B1%95%E7%A4%BA%E5%8D%95%E7%BB%86%E8%83%9E%E8%81%9A%E7%B1%BB-ComplexHeatmap-for-single-cell-visualization/" data-id="clm1ywdiv0098ecok1lpu3oxi" data-title="「热图」ComplexHeatmap展示单细胞聚类" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-「LeetCode」递归题目之第N个Tribonacci数-LeetCode-N-th Tribonacci Number" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/22/%E3%80%8CLeetCode%E3%80%8D%E9%80%92%E5%BD%92%E9%A2%98%E7%9B%AE%E4%B9%8B%E7%AC%ACN%E4%B8%AATribonacci%E6%95%B0-LeetCode-N-th%20Tribonacci%20Number/" class="article-date">
  <time class="dt-published" datetime="2019-10-22T21:21:35.374Z" itemprop="datePublished">2019-10-22</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/22/%E3%80%8CLeetCode%E3%80%8D%E9%80%92%E5%BD%92%E9%A2%98%E7%9B%AE%E4%B9%8B%E7%AC%ACN%E4%B8%AATribonacci%E6%95%B0-LeetCode-N-th%20Tribonacci%20Number/">「LeetCode」递归题目之第N个Tribonacci数</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>Tribonacci序列Tn定义:</p>
<p>T0&#x3D;0, T1&#x3D;1, T2&#x3D;1, n&gt;&#x3D;0时，Tn+3 &#x3D; Tn + Tn+1 + Tn+2</p>
<p>限制条件是: 0&lt;&#x3D;n&lt;&#x3D;37, 32位整型。</p>
<p>我直接用C++撸了下面的代码，</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;iostream&gt;</span></span></span><br><span class="line"></span><br><span class="line"><span class="keyword">using</span> <span class="keyword">namespace</span> std;</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">Solution</span> &#123;</span><br><span class="line"><span class="keyword">public</span>:</span><br><span class="line">    <span class="function"><span class="type">int</span> <span class="title">tribonacci</span><span class="params">(<span class="type">int</span> n)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">if</span> (n ==  <span class="number">2</span>)&#123;</span><br><span class="line">            <span class="keyword">return</span> <span class="number">1</span>;</span><br><span class="line">        &#125; </span><br><span class="line">        <span class="keyword">if</span> (n == <span class="number">1</span>)&#123;</span><br><span class="line">            <span class="keyword">return</span> <span class="number">1</span>;</span><br><span class="line">        &#125; </span><br><span class="line">        <span class="keyword">if</span> (n == <span class="number">0</span>)&#123;</span><br><span class="line">            <span class="keyword">return</span> <span class="number">0</span>;</span><br><span class="line">        &#125; </span><br><span class="line">        <span class="type">int</span> result = <span class="built_in">tribonacci</span>(n<span class="number">-3</span>) + <span class="built_in">tribonacci</span>(n<span class="number">-2</span>) + <span class="built_in">tribonacci</span>(n<span class="number">-1</span>);</span><br><span class="line"></span><br><span class="line">        <span class="keyword">return</span> result;</span><br><span class="line">        </span><br><span class="line">    &#125;</span><br><span class="line">&#125;;</span><br><span class="line"></span><br><span class="line"><span class="comment">// 下面是自己电脑编译测试用的代码</span></span><br><span class="line"><span class="function"><span class="type">int</span> <span class="title">main</span><span class="params">()</span></span>&#123;</span><br><span class="line"></span><br><span class="line">	Solution sol;</span><br><span class="line">	<span class="keyword">for</span> (<span class="type">int</span> i = <span class="number">0</span>; i &lt;= <span class="number">37</span>; i++)&#123;</span><br><span class="line">	    cout &lt;&lt; <span class="string">&quot;tribonacci &quot;</span> &lt;&lt; i &lt;&lt; <span class="string">&quot; is &quot;</span> &lt;&lt; sol.<span class="built_in">tribonacci</span>(i) &lt;&lt; endl;</span><br><span class="line">	&#125;</span><br><span class="line"></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>提交答案后，提示”The Limit Exceeded”. </p>
<p>在自己服务器测试时，也发现当n&#x3D;31之后的速度下降的非常快。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 编译</span></span><br><span class="line">g++ -o trib trib.cpp</span><br><span class="line">./trib</span><br></pre></td></tr></table></figure>

<p>原因是当n越大时，直接使用递归会产生大量的重复计算，导致计算效率下降。为了解决该问题，需要维护一个已计算结果的字典，避免重复运算。</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;iostream&gt;</span></span></span><br><span class="line"></span><br><span class="line"><span class="keyword">using</span> <span class="keyword">namespace</span> std;</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">Solution</span> &#123;</span><br><span class="line"><span class="keyword">public</span>:</span><br><span class="line">    <span class="type">int</span> tribo[<span class="number">100</span>];</span><br><span class="line">    <span class="function"><span class="type">int</span> <span class="title">tribonacci</span><span class="params">(<span class="type">int</span> n)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">if</span> (n ==  <span class="number">2</span>)&#123;</span><br><span class="line">            <span class="keyword">return</span> <span class="number">1</span>;</span><br><span class="line">        &#125; </span><br><span class="line">        <span class="keyword">if</span> (n == <span class="number">1</span>)&#123;</span><br><span class="line">            <span class="keyword">return</span> <span class="number">1</span>;</span><br><span class="line">        &#125; </span><br><span class="line">        <span class="keyword">if</span> (n == <span class="number">0</span>)&#123;</span><br><span class="line">            <span class="keyword">return</span> <span class="number">0</span>;</span><br><span class="line">        &#125; </span><br><span class="line">        </span><br><span class="line">        <span class="keyword">if</span> (tribo[n] &gt; <span class="number">0</span>)&#123;</span><br><span class="line">            <span class="keyword">return</span> tribo[n]<span class="number">-1</span>;</span><br><span class="line">        &#125;</span><br><span class="line">        </span><br><span class="line">        <span class="type">int</span> result = <span class="built_in">tribonacci</span>(n<span class="number">-3</span>) + <span class="built_in">tribonacci</span>(n<span class="number">-2</span>) + <span class="built_in">tribonacci</span>(n<span class="number">-1</span>);</span><br><span class="line">        tribo[n] = result + <span class="number">1</span>;</span><br><span class="line"></span><br><span class="line">        <span class="keyword">return</span> result;</span><br><span class="line">        </span><br><span class="line">    &#125;</span><br><span class="line">&#125;;</span><br><span class="line"></span><br><span class="line"><span class="comment">// 下面是自己电脑编译测试用的代码</span></span><br><span class="line"><span class="function"><span class="type">int</span> <span class="title">main</span><span class="params">()</span></span>&#123;</span><br><span class="line"></span><br><span class="line">	Solution sol;</span><br><span class="line">	<span class="keyword">for</span> (<span class="type">int</span> i = <span class="number">0</span>; i &lt;= <span class="number">37</span>; i++)&#123;</span><br><span class="line">	    cout &lt;&lt; <span class="string">&quot;tribonacci &quot;</span> &lt;&lt; i &lt;&lt; <span class="string">&quot; is &quot;</span> &lt;&lt; sol.<span class="built_in">tribonacci</span>(i) &lt;&lt; endl;</span><br><span class="line">	&#125;</span><br><span class="line"></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>新的代码中，我预先定义了一个大小为100的数组，主要是因为题目限制n的取值，如果n没有限制取值，我会考虑使用vector容器或者map词典。其次考虑到n&#x3D;0时，结果为0，并且数组初始化值为0，为了能够通过数组值是否为0来判断是否已经计算了相应的tribonacci值，我对结果加上了一个pseducount。</p>
<p>最后这段代码运行速度是0ms.</p>
<p>如果要用哈希字典，代码如下</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;map&gt;</span></span></span><br><span class="line"><span class="comment">//一定要引入map头文件</span></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">Solution</span> &#123;</span><br><span class="line"><span class="keyword">public</span>:</span><br><span class="line">    <span class="comment">//定义一个哈希字典</span></span><br><span class="line">    map&lt;<span class="type">int</span>, <span class="type">int</span>&gt; num;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="type">int</span> <span class="title">tribonacci</span><span class="params">(<span class="type">int</span> n)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">if</span>(n &lt; <span class="number">2</span>) <span class="keyword">return</span> n;</span><br><span class="line">        <span class="keyword">if</span>(n == <span class="number">2</span>) <span class="keyword">return</span> <span class="number">1</span>;</span><br><span class="line">        <span class="comment">//如果字典中有已经计算的结果,直接返回结果</span></span><br><span class="line">        <span class="keyword">if</span>(num.<span class="built_in">find</span>(n) != num.<span class="built_in">end</span>()) <span class="keyword">return</span> num[n];</span><br><span class="line">        <span class="type">int</span> sum = <span class="built_in">tribonacci</span>(n<span class="number">-1</span>) + <span class="built_in">tribonacci</span>(n<span class="number">-2</span>) + <span class="built_in">tribonacci</span>(n<span class="number">-3</span>);</span><br><span class="line">        num[n] = sum;</span><br><span class="line">        <span class="keyword">return</span> sum;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;;</span><br></pre></td></tr></table></figure>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/22/%E3%80%8CLeetCode%E3%80%8D%E9%80%92%E5%BD%92%E9%A2%98%E7%9B%AE%E4%B9%8B%E7%AC%ACN%E4%B8%AATribonacci%E6%95%B0-LeetCode-N-th%20Tribonacci%20Number/" data-id="clm1ywdij0089ecokfvu7bmup" data-title="「LeetCode」递归题目之第N个Tribonacci数" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-使用NextDenovo组装Nanopore数据-Assembly-nanopore-with-NextDenovo" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/21/%E4%BD%BF%E7%94%A8NextDenovo%E7%BB%84%E8%A3%85Nanopore%E6%95%B0%E6%8D%AE-Assembly-nanopore-with-NextDenovo/" class="article-date">
  <time class="dt-published" datetime="2019-10-21T10:59:24.003Z" itemprop="datePublished">2019-10-21</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/21/%E4%BD%BF%E7%94%A8NextDenovo%E7%BB%84%E8%A3%85Nanopore%E6%95%B0%E6%8D%AE-Assembly-nanopore-with-NextDenovo/">使用NextDenovo组装Nanopore数据</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>NextDenovo是武汉未来组胡江博士团队开发的一个三代组装工具，能够用于PacBio和Nanopore数据的组装。但是从官方的介绍而言，此工具在组装Nanopore上优势更大一些。</p>
<p>NextDenovo包括两个模块，NextCorrect用于原始数据纠错，NextGraph能够基于纠错后的进行组装。使用修改版的minimap2进行序列间相互比对。v2.0-beta.1版中在处理高度重复序列上可能存在错误组装，可以通过HiC和BioNano进行纠错。</p>
<h2 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h2><p>NextDenovo的软件安装非常简单, 下载解压缩即可使用。考虑到NextDenovo需要用Python2.7，我们可以用conda新建一个环境</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">conda create -n python2 python=2.7</span><br><span class="line">conda activate python2</span><br><span class="line">pip install psutil</span><br></pre></td></tr></table></figure>

<p>然后下载解压缩（我习惯把软件放在<code>~/opt/bisofot</code>下)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p ~/opt/biosoft/</span><br><span class="line"><span class="built_in">cd</span> ~/opt/biosoft/</span><br><span class="line">wget https://github.com/Nextomics/NextDenovo/releases/download/v2.0-beta.1/NextDenovo.tgz</span><br><span class="line">tar -zxvf NextDenovo.tgz</span><br></pre></td></tr></table></figure>

<p>测试下软件是否可以使用</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/NextDenovo/nextDenovo -h</span><br></pre></td></tr></table></figure>

<h2 id="实战"><a href="#实战" class="headerlink" title="实战"></a>实战</h2><p>以发表在NC上的拟南芥数据为例,  简单介绍下软件的使用</p>
<p><strong>第一步</strong>: 新建一个分析项目</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> NEXT &amp;&amp; <span class="built_in">cd</span> NEXT</span><br></pre></td></tr></table></figure>

<p>然后从EBI上下载该数据，在<code>run.fofn</code>中记录文件的实际位置。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 三代测序</span></span><br><span class="line">wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR217/003/ERR2173373/ERR2173373.fastq.gz</span><br><span class="line"><span class="built_in">realpath</span> ERR2173373.fastq.gz &gt; run.fofn</span><br></pre></td></tr></table></figure>

<p><strong>第二步</strong>: 复制和修改配置文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cp</span> ~/opt/biosoft/NextDenovo/doc/run.cfg .</span><br></pre></td></tr></table></figure>

<p>我的配置文件修改如下，参数说明参考<a target="_blank" rel="noopener" href="https://github.com/Nextomics/NextDenovo/blob/master/doc/OPTION.md">官方文档</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">[General]</span><br><span class="line">job_type = <span class="built_in">local</span></span><br><span class="line">job_prefix = nextDenovo</span><br><span class="line">task = all <span class="comment"># &#x27;all&#x27;, &#x27;correct&#x27;, &#x27;assemble&#x27;</span></span><br><span class="line">rewrite = <span class="built_in">yes</span> <span class="comment"># yes/no</span></span><br><span class="line">deltmp = <span class="built_in">yes</span></span><br><span class="line">rerun = 3</span><br><span class="line">parallel_jobs = 5</span><br><span class="line">input_type = raw</span><br><span class="line">input_fofn = input.fofn</span><br><span class="line">workdir = 01_rundir</span><br><span class="line"><span class="comment"># cluster_options = -l vf=&#123;vf&#125; -q all.q -pe smp &#123;cpu&#125; -S &#123;bash&#125; -w n</span></span><br><span class="line"></span><br><span class="line">[correct_option]</span><br><span class="line">read_cutoff = 1k</span><br><span class="line">seed_cutoff = 3k</span><br><span class="line">blocksize = 3g</span><br><span class="line">pa_correction = 20</span><br><span class="line">seed_cutfiles = 20</span><br><span class="line">sort_options = -m 20g -t 8 -k 40</span><br><span class="line">minimap2_options_raw = -x ava-ont -t 8</span><br><span class="line">correction_options = -p 8</span><br><span class="line"></span><br><span class="line">[assemble_option]</span><br><span class="line">random_round = 20</span><br><span class="line">minimap2_options_cns = -x ava-ont -t 8 -k17 -w17</span><br><span class="line">nextgraph_options = -a 1</span><br></pre></td></tr></table></figure>

<p>配置文件的几个重要参数说明（v2.0-beta.1）</p>
<ul>
<li>job_type 设置运行环境，可以使用（local， sge， pbs等）</li>
<li>运行线程数设置，线程数计算为<code>parallel_jobs</code>分别与<code>sort_option</code>, <code>minimap_options_*</code>的<code>-t</code>数乘积，和<code>correction_options</code>的<code>-p</code>的乘积，量力而行。</li>
<li>seed_cutfiles 如果在集群上运行，建议设置为可用的节点数，同时设置<code>correction_options</code>的<code>-p</code>为各个节点可用的核数，保证每个节点只有一个correction任务，减少运行时的内存和IO。 如果local上运行， 建议设置为总可用的核除以<code>correction_options</code>的<code>-p</code>值. </li>
<li>parallel_jobs建议设置<strong>至少</strong>要大于<code>pa_correction</code>。</li>
<li>blocksize 是将小于seed_cutfiles的数据拆分成的多个文件时单个文件的大小， 总的比对任务数等于基于该参数切分的文件数乘以<code>seed_cutfiles + seed_cutfiles * (seed_cutfiles - 1)/2</code>， 因此对于10g以内的数据量， 建议设置小于1g， 避免总的任务数小于parallel_jobs的值。</li>
<li>测序数据类型相关: 对于PacBio而言，要修改<code>minimap2_options_*</code>中的<code>-x ava-ont</code>为<code>-x ava-pb</code></li>
<li>数据量相关参数: <code>read_cutoff = 1k</code>过滤原始数据中低于1k的read，<code>seed_cutoff = 30k</code>则是选择大于30k以上的数据来矫正。关于<code>seed_cutoff</code>的设置，可以通过<code>~/opt/biosoft/NextDenovo/bin/seq_stat</code>来获取参考值，<strong>不建议</strong>直接使用默认值，因为改值会受到测序深度和测序长度影响，而且一个不合适的值会显著降低组装质量。对于基因组大于200m以上的物种，-d建议默认。</li>
<li>correction_options中的-dbuf可以显著降低矫正时的内存，但会显著降低矫正速度。</li>
<li>random_round参数，建议设置20-100. 该参数是设置随机组装参数的数量，nextGraph会基于每一套随机参数做一次组装， 避免默认参数效果不好。</li>
</ul>
<p><code>seq_stat</code>能够根据物种大小和预期用于组装的深度确定<code>seed_cutoff</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/NextDenovo/bin/seq_stat -g 110Mb -d 30 input.fofn</span><br></pre></td></tr></table></figure>

<p><strong>第三步</strong>: 运行NextDenovo</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/NextDenovo/nextDenovo run.cfg &amp;</span><br></pre></td></tr></table></figure>

<p>运行时间如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">real	64m5.356s</span><br><span class="line">user	1827m37.890s</span><br><span class="line">sys	264m48.246s</span><br></pre></td></tr></table></figure>

<p>默认参数结果是存放在<code>01_rundir/03.ctg_graph/01.ctg_graph.sh.work/ctg_graph00</code>, 可以将其复制到当前目录，用于后续的分析。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cat</span> 01_rundir/03.ctg_graph/01.ctg_graph.sh.work/ctg_graph00/nextgraph.assembly.contig.fasta &gt; nextgraph.assembly.contig.fasta</span><br></pre></td></tr></table></figure>

<p>但是在<code>01.ctg_graph.sh.work</code>目录下除了<code>ctg_graph00</code>以外，还有其他随机参数的在组装结果。随机参数结果只输出了统计结果，用户如需要输出组装序列，可以修改<code>01_rundir/03.ctg_graph/01.ctg_graph.sh</code>，将里面的<code>-a 0</code>替换成<code>-a 1</code>。</p>
<p>每个目录下都有shell输出，可以挑选基于nextDenovo.sh.e这里面的结果挑选组装指标较好的，再输出序列，比如说比较下N50</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">grep N50  01_rundir/03.ctg_graph/01.ctg_graph.sh.work/ctg_graph*/*.e</span><br></pre></td></tr></table></figure>

<p>默认情况下，最终组装出20条contig，总大小116M，N50 12M.</p>
<p>使用minimap2将组装结果和比对到TAIR10上，用dotplotly进行可视化</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">minimap2 -t 100 -x asm5 Athaliana.fa nextgraph.assembly.contig.fasta &gt; next.paf</span><br><span class="line">dotPlotly/pafCoordsDotPlotly.R  -i next.paf -o next  -l -p 6 -k 5</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571625271329-58bef12cf9584d90959127db3919895e.png" alt="共线性"></p>
<p>不难发现，两者存在高度的共线性。大部分TAIR10的染色体对应的都是2条或者3条contig。</p>
<p>此外这篇NC的拟南芥提供了BioNano光学图谱，我使用BioNano Hyrbrid Scaffold 流程进行了混合组装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cp</span> /opt/biosoft/Solve3.3_10252018/HybridScaffold/10252018/hybridScaffold_config.xml .</span><br><span class="line"><span class="comment"># 修改xml中fasta2cmap的enzyme为BSPQI</span></span><br><span class="line">perl /opt/biosoft/Solve3.3_10252018/HybridScaffold/10252018/hybridScaffold.pl \</span><br><span class="line">    -n nextgraph.assembly.contig.fasta \</span><br><span class="line">    -b kbs-mac-74_bng_contigs2017.cmap \</span><br><span class="line">    -c hybridScaffold_config.xml \</span><br><span class="line">    -r /opt/biosoft/Solve3.3_10252018/RefAligner/7915.7989rel/RefAligner \</span><br><span class="line">    -o nextgraph \</span><br><span class="line">    -B 2 -N 2 \</span><br><span class="line">    -f</span><br></pre></td></tr></table></figure>

<p>组装结果如下，从原来的20的contig下降到了16个contig。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">Count  = 16</span><br><span class="line">Min length (Mbp) = 0.026</span><br><span class="line">Median length (Mbp) = 7.224</span><br><span class="line">Mean length (Mbp) = 7.301</span><br><span class="line">N50 length (Mbp) = 13.013</span><br><span class="line">Max length (Mbp) = 14.965</span><br><span class="line">Total length (Mbp) = 116.811</span><br></pre></td></tr></table></figure>

<p>此外还通过BioNano Access进行可视化，以其中一个结果为例。光学图谱和NextDenovo的组装结果存在很高的一致性。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571626377015-874040eb90674dcbb44a2f15fbfc07b5.png" alt="示例"></p>
<p>综上，在Nanopore上组装上，我们又多了一个比较好用的工具。</p>
<blockquote>
<p>最后非常感谢胡江博士对于本文的指导！</p>
</blockquote>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/21/%E4%BD%BF%E7%94%A8NextDenovo%E7%BB%84%E8%A3%85Nanopore%E6%95%B0%E6%8D%AE-Assembly-nanopore-with-NextDenovo/" data-id="clm1ywdk200d5ecok3spn7gm5" data-title="使用NextDenovo组装Nanopore数据" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-「BioNano系列」如何进行cmap之间的比对-BioNano-CMAP-Comparison" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/21/%E3%80%8CBioNano%E7%B3%BB%E5%88%97%E3%80%8D%E5%A6%82%E4%BD%95%E8%BF%9B%E8%A1%8Ccmap%E4%B9%8B%E9%97%B4%E7%9A%84%E6%AF%94%E5%AF%B9-BioNano-CMAP-Comparison/" class="article-date">
  <time class="dt-published" datetime="2019-10-21T10:25:08.151Z" itemprop="datePublished">2019-10-21</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/21/%E3%80%8CBioNano%E7%B3%BB%E5%88%97%E3%80%8D%E5%A6%82%E4%BD%95%E8%BF%9B%E8%A1%8Ccmap%E4%B9%8B%E9%97%B4%E7%9A%84%E6%AF%94%E5%AF%B9-BioNano-CMAP-Comparison/">「BioNano系列」如何进行cmap之间的比对</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>BioNano以cmap格式存放光学图谱，为了评估基因组的组装质量或者了解光学图谱中冗余情况(高杂合基因组组装结果偏大)，我们就需要进行cmap之间的比较。</p>
<h2 id="CMAP间比对"><a href="#CMAP间比对" class="headerlink" title="CMAP间比对"></a>CMAP间比对</h2><p>Solve套件提供了runCharacterize.py脚本封装了RefAligner，用于进行CMAP之间的比对。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">python2.7 runCharacterize.py \</span><br><span class="line">    -t RefAligner的二进制文件路径 \</span><br><span class="line">    -q 用于比对的CMAP \</span><br><span class="line">    -r 参考CMAP \</span><br><span class="line">    -p Pipeline文件路径\</span><br><span class="line">    -a 参数配置文件.xml \</span><br><span class="line">    -n 线程数,默认4</span><br></pre></td></tr></table></figure>

<p>需要注意的是<code>-p</code>和<code>-a</code>参数的设置。<code>-p</code>是Pipeline的文件位置，比如说我的<code>Solve</code>安装在<code>/opt/biosoft/Solve3.4_06042019a</code>，那么参数设置为<code> -p /opt/biosoft/Solve3.4_06042019a/Pipeline/06042019</code>。 而<code>-a</code>则是要在<code>/opt/biosoft/Solve3.4_06042019a/RefAligner/8949.9232rel/</code>目录下选择合适的xml文件。比如你的CMAP是Irys平台，那么你可以考虑用optArguments_nonhaplotype_irys.xml.</p>
<p>以最新发表的辣椒的光学图谱为例，该物种有比较高的杂合度，组装结果偏大，我们可以通过自比对来寻找冗余区域, </p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 下载CMAP</span></span><br><span class="line">wget https://submit.ncbi.nlm.nih.gov/ft/byid/o62junnn/piper_nigrum_no_rcmap_refinefinal1.cmap</span><br><span class="line"><span class="comment"># 自比对</span></span><br><span class="line">python /opt/biosoft/Solve3.4_06042019a/Pipeline/06042019/runCharacterize.py \</span><br><span class="line">	-t /opt/biosoft/Solve3.4_06042019a/RefAligner/8949.9232rel/RefAligner \</span><br><span class="line">	-q piper_nigrum_no_rcmap_refinefinal1.cmap \</span><br><span class="line">	-r piper_nigrum_no_rcmap_refinefinal1.cmap \</span><br><span class="line">	-p /opt/biosoft/Solve3.4_06042019a/Pipeline/06042019 \</span><br><span class="line">	-a /opt/biosoft/Solve3.4_06042019a/RefAligner/8949.9232rel/optArguments_nonhaplotype_saphyr.xml -n 64</span><br></pre></td></tr></table></figure>

<p>最终会在当前文件下生成一个alignRef文件夹，其中结果是q.cmap,r.cmap和xmap的文件可以用于上传到BioNano Access上进行展示。下图就是一个冗余实例，可以把图中较短的图谱删掉</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571624415320-5ae131f556f9408b868d750c99933b5e.png" alt="冗余"></p>
<h2 id="基因组回帖"><a href="#基因组回帖" class="headerlink" title="基因组回帖"></a>基因组回帖</h2><p>为了将基因组回帖到CMAP上，需要先将基因组的fasta格式转成CMAP格式，参数如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl fa2cmap_multi_color.pl -i 输入FASTA -e 酶1 通道1 [酶2 通道2]</span><br></pre></td></tr></table></figure>

<p>其中一个最重要的参数就是<strong>酶切类型</strong>。例如我需要将序列回帖到用Nt.BspQI酶切组装的光学图谱上，因此运行参数如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl /opt/biosoft/Solve3.4_06042019a/HybridScaffold/06042019/scripts/fa2cmap_multi_color.pl -i athaliana.fa -e BspQI 1</span><br></pre></td></tr></table></figure>

<p>最后的athaliana_BSPQI_0kb_0labels.cmap就是模拟酶切的CMAP序列。</p>
<p>之后将模拟酶切的结果回帖到实际的CMAP</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">python /opt/biosoft/Solve3.4_06042019a/Pipeline/06042019/runCharacterize.py \</span><br><span class="line">    -t /opt/biosoft/Solve3.4_06042019a/RefAligner/8949.9232rel/RefAligner \</span><br><span class="line">    -q athaliana_BSPQI_0kb_0labels.cmap \</span><br><span class="line">    -r kbs-mac-74_bng_contigs2017.cmap \</span><br><span class="line">    -p /opt/biosoft/Solve3.4_06042019a/Pipeline/06042019 \</span><br><span class="line">    -a /opt/biosoft/Solve3.4_06042019a/RefAligner/8949.9232rel/optArguments_nonhaplotype_saphyr.xml \</span><br><span class="line">    -n 64</span><br></pre></td></tr></table></figure>

<p>最终会在当前文件下生成一个alignRef文件夹，其中结果是q.cmap,r.cmap和xmap的文件.</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/21/%E3%80%8CBioNano%E7%B3%BB%E5%88%97%E3%80%8D%E5%A6%82%E4%BD%95%E8%BF%9B%E8%A1%8Ccmap%E4%B9%8B%E9%97%B4%E7%9A%84%E6%AF%94%E5%AF%B9-BioNano-CMAP-Comparison/" data-id="clm1ywdie007yecok2xjg45bk" data-title="「BioNano系列」如何进行cmap之间的比对" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用SCALE分析单细胞ATAC-seq数据-Analysis-sc-ATAC-seq-with-SCALE" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/17/%E4%BD%BF%E7%94%A8SCALE%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE-Analysis-sc-ATAC-seq-with-SCALE/" class="article-date">
  <time class="dt-published" datetime="2019-10-17T12:24:29.682Z" itemprop="datePublished">2019-10-17</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/17/%E4%BD%BF%E7%94%A8SCALE%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE-Analysis-sc-ATAC-seq-with-SCALE/">使用SCALE分析单细胞ATAC-seq数据</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>SCALE全称是Single-Cell ATAC-seq analysis vie Latent feature Extraction, 从名字中就能知道这个软件是通过隐特征提取的方式分析单细胞ATAC-seq数据。</p>
<p>在文章中，作者从开发者的角度列出了目前的scATAC-seq分析软件，chromVAR, scABC, cisTopic, scVI，发现每个软件都有一定的不足之处，而从我们软件使用者的角度，其实可以考虑都试试这些工具。</p>
<p>SCALE结合了深度生成模型(Depp Generative Models)变分自动编码器框架(Variational Autoencoder, VAE)与概率高斯混合模型(Gaussian Mixture Model, GMM)去学习隐特征，用于准确地鉴定scATAC-seq数据中的特征。</p>
<p>文章通过一张图来解释了软件的工作机制：</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571238440863-636159eb005d4840b35544bcbb8c1205.png" alt="SCALE框架"></p>
<p>SCALE将sc-ATAC-seq的输入数据x(Cells-by-Peaks矩阵)建模成一个联合分布，p(x,z,c)，c是GMM组件中对应的预定义的K个聚类，z是一个隐变量，是细胞在所有peak中实际可能的值，用于后续的聚类和可视化。z通过$z &#x3D; \mu_z + \sigma_Z \times \epsilon$ 计算而得，公式里面的 $\mu_z$  $\sigma_z$ 是编码器网络从x中学习而得，$\epsilon$ 则是从 $\mathbb{N}(0,\mathbf{I})$ 抽样而成。 </p>
<p>从公式中我们还可以发现z其实和GMM的c有关，所以p(x,z,c)也可以写成P(x|z)p(z|c)p(c)，而p(c)是K个预定义聚类分布的离散概率分布，p(z|c)服从混合高斯分布，而p(x|z)则是服从多变量伯努利分布(multivartiable Bernoulli distribution), 通过解码者网络建模而成。</p>
<p>当然从一个软件使用者的角度而言，我们不会去关心代码，也不会关心原理，我们更关心的是这个工具能做什么。SCALE能做以下的分析</p>
<ul>
<li>SCALE可以对隐特征聚类识别细胞类群</li>
<li>SCALE可以降噪，恢复缺失的peak</li>
<li>SCALE能够区分批次效应和生物学细胞类群之间的差异</li>
</ul>
<h2 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h2><p>推荐使用conda的方式进行软件安装（我测试过了，运行没有问题）</p>
<p>第一步：创建一个环境，名字就是<code>SCALE</code>，并且启动该环境</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">conda create -n SCALE python=3.6 pytorch</span><br><span class="line">conda activate SCALE</span><br></pre></td></tr></table></figure>

<p>第二步：从GitHub上克隆该项目</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git <span class="built_in">clone</span> git://github.com/jsxlei/SCALE.git</span><br></pre></td></tr></table></figure>

<p>第三步：安装SCALE</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> SCALE</span><br><span class="line">python setup.py install</span><br></pre></td></tr></table></figure>

<p>之后分析的时候，只需要通过<code>conda activate SCALE</code>就能启动分析环境。</p>
<p>考虑后续要交互的读取数据和可视化，那么建议再安装一个Jupyter</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install jupyter</span><br></pre></td></tr></table></figure>

<h2 id="软件使用"><a href="#软件使用" class="headerlink" title="软件使用"></a>软件使用</h2><p>SCALE支持两类输入文件：</p>
<ul>
<li>count矩阵，行为peak，列为barcode</li>
<li>10X输出文件: count.mtx.gz, peak.tsv, barcode.tsv</li>
</ul>
<p>我们以官方提供的Forebrain数据集为例进行介绍，因为这个数据相对于另外一个数据集Mouse Atlas小多了。</p>
<p>我们在服务器上新建一个文件夹，用于存放从<a target="_blank" rel="noopener" href="https://cloud.tsinghua.edu.cn/d/bda0332212154163a647/">https://cloud.tsinghua.edu.cn/d/bda0332212154163a647/</a>下载的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> Forebrain</span><br></pre></td></tr></table></figure>

<p>保证<code>Forebrain</code>有下载好的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">ls</span> Forebrain </span><br><span class="line">data.txt</span><br></pre></td></tr></table></figure>

<p>之后运行程序</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">SCALE.py -d Forebrain/data.txt -k 8 --impute</span><br></pre></td></tr></table></figure>

<p>软件运行步骤为：</p>
<ul>
<li>加载数据: Loading  data</li>
<li>模型训练: Training Model</li>
<li>输出结果: Saving imputed data</li>
</ul>
<p>其中模型训练这一步时间比较久，可以尝试用GPU加速（我是普通CPU服务器没有办法）。最终会在当前文件夹看到一个<code>output</code>文件夹，里面有如下内容:</p>
<ul>
<li>imputed_data.txt: 每个细胞在每个特征的推断值，建议用<code>--binary</code>保存二进制格式</li>
<li>model.pt: 用于重复结果的模型文件，<code>--pretrain</code>参数能够读取该模型</li>
<li>feature.txt: 每个细胞的隐特征，用于聚类和可视化</li>
<li>cluster_assignments.txt: 两列，barcode和所属类群</li>
<li>tsne.txt, tsne.pdf: tSNE的坐标和PDF文件，坐标文件可以导入到R语言进行可视化</li>
</ul>
<p>上面是命令行部分，下面则是Python环境进行交互式操作，输入<code>jupyter notebook</code>，之后在网页上打开</p>
<p>首先是导入各种Python库</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">from</span> sklearn.metrics <span class="keyword">import</span> confusion_matrix</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sns</span><br><span class="line"><span class="keyword">from</span> scale.plot <span class="keyword">import</span> plot_embedding, plot_heatmap</span><br></pre></td></tr></table></figure>

<p>然后加载分析结果，包括聚类信息和特征信息</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">y = pd.read_csv(<span class="string">&#x27;output/cluster_assignments.txt&#x27;</span>, sep=<span class="string">&#x27;\t&#x27;</span>, index_col=<span class="number">0</span>, header=<span class="literal">None</span>)[<span class="number">1</span>].values</span><br><span class="line">feature = pd.read_csv(<span class="string">&#x27;output/feature.txt&#x27;</span>, sep=<span class="string">&#x27;\t&#x27;</span>, index_col=<span class="number">0</span>, header=<span class="literal">None</span>)</span><br></pre></td></tr></table></figure>

<p>通过热图展示不同聚类细胞之间的差异图</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">plot_heatmap(feature.T, y, </span><br><span class="line">             figsize=(<span class="number">8</span>, <span class="number">3</span>), cmap=<span class="string">&#x27;RdBu_r&#x27;</span>, vmax=<span class="number">8</span>, vmin=-<span class="number">8</span>, center=<span class="number">0</span>,</span><br><span class="line">             ylabel=<span class="string">&#x27;Feature dimension&#x27;</span>, yticklabels=np.arange(<span class="number">10</span>)+<span class="number">1</span>, </span><br><span class="line">             cax_title=<span class="string">&#x27;Feature value&#x27;</span>, legend_font=<span class="number">6</span>, ncol=<span class="number">1</span>,</span><br><span class="line">             bbox_to_anchor=(<span class="number">1.1</span>, <span class="number">1.1</span>), position=(<span class="number">0.92</span>, <span class="number">0.15</span>, <span class="number">.08</span>, <span class="number">.04</span>))</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571285853797-58bd09803a9e4d2bae1f36ac5cbe0623.png" alt="heatmap"></p>
<blockquote>
<p>如果要矫正批次效应，可以通过根据feature的heatmap，去掉和batch相关的feature来实现</p>
</blockquote>
<p>我们可以展示SCALE对原始数据纠正后的值(imputed data), 该结果也能提高chromVAR鉴定motif的效果</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">imputed = pd.read_csv(<span class="string">&#x27;output/imputed_data.txt&#x27;</span>, sep=<span class="string">&#x27;\t&#x27;</span>, index_col=<span class="number">0</span>)</span><br></pre></td></tr></table></figure>

<p>展示聚类特异性的peak， 分析由<code>mat_specificity_score</code>和<code>cluster_specific</code>完成</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> scale.specifity <span class="keyword">import</span> cluster_specific, mat_specificity_score</span><br><span class="line"></span><br><span class="line">score_mat = mat_specificity_score(imputed, y)</span><br><span class="line">peak_index, peak_labels = cluster_specific(score_mat, np.unique(y), top=<span class="number">200</span>)</span><br><span class="line"></span><br><span class="line">plot_heatmap(imputed.iloc[peak_index], y=y, row_labels=peak_labels, ncol=<span class="number">3</span>, cmap=<span class="string">&#x27;Reds&#x27;</span>, </span><br><span class="line">             vmax=<span class="number">1</span>, row_cluster=<span class="literal">False</span>, legend_font=<span class="number">6</span>, cax_title=<span class="string">&#x27;Peak Value&#x27;</span>,</span><br><span class="line">             figsize=(<span class="number">8</span>, <span class="number">10</span>), bbox_to_anchor=(<span class="number">0.4</span>, <span class="number">1.2</span>), position=(<span class="number">0.8</span>, <span class="number">0.76</span>, <span class="number">0.1</span>, <span class="number">0.015</span>))</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571286462453-1032ee7e22304701a52a8e81af3b0020.png" alt="聚类特异性peak"></p>
<h2 id="参数介绍"><a href="#参数介绍" class="headerlink" title="参数介绍"></a>参数介绍</h2><p>通过<code>SCALE.py -h</code>可以输出SCALE的所有可用参数</p>
<ul>
<li><code>-d/--dataset</code>: 单个文件矩阵应该指定文件路径，10X输出的多个文件则是文件目录</li>
<li><code>-k</code>: 设定输出结果的聚类数</li>
<li><code>-o</code>: 输出文件路径</li>
<li><code>--pretrain</code>: 读取之前训练的模型</li>
<li><code>--lr</code>: 修改起始学习速率, 默认是0.002，和模型训练有关</li>
<li><code>--batch_size</code>: 批处理大小， 默认就行，不需要修改（和批次效应处理无关）</li>
<li><code>-g GPU</code>:  选择GPU设备数目，非GPU服务器用不到</li>
<li><code>--seed</code>: 初始随机数种子，通常在遇到nan缺失时考虑修改</li>
<li><code>-encode_dim</code>, <code>-decode_dim</code>: 编码器和解码器的维度，通常也不需要修改</li>
<li><code>-latent</code> 隐藏层维度</li>
<li><code>--low</code>, <code> --high</code>: 过滤低质量的peak, 即出现比例高于或者低于某个阈值的peak，默认是0.01和0.9。作者推荐保留1万-3万的peak用于SCALE分析。如果数据质量很高，且peak数不多于10万，那么可以不过滤。</li>
<li><code> --min_peaks</code>: 过滤低质量细胞，如果该细胞的peak低于阈值</li>
<li><code>log_transform</code>: <code>log2(x+1)</code>的变换</li>
<li><code>--max_iter</code>: 最大迭代数，默认是30000, 可以观察损失收敛的情况来修改，也就是训练模型这一步输出的信息</li>
<li><code>-weight_decay</code>: 没有说明</li>
<li><code>--impute</code>: 保存推断数据，默认开启</li>
<li><code>--binary</code>: 推荐加上该参数，减少imputed data占用空间</li>
<li><code>--no_tsne</code>:  不需要保存t-SNE结果</li>
<li><code>--reference</code>: 参考细胞类型</li>
<li><code>-t</code>:  如果输出矩阵是列为peak，行为barcode，用该参数进行转置</li>
</ul>
<p>对于使用者而言，我们一般只用修改<code>-k</code>更改最后的聚类数，<code>--low</code>, <code>--high</code>, <code>---min_peaks</code>来对原始数据进行过滤，以及加上<code>--binary</code>节约空间。</p>
<p>假如在训练模型阶段，发现输出信息为<code>loss=nan recon_loss=nan kl_loss=nan</code>,十有八九最终会报错退出， 可以如下的参数调整</p>
<ul>
<li>更改<code>--seed</code></li>
<li>用更加严格的条件过滤peak，例如<code>-x 0.04</code> 或  <code>-x 0.06</code></li>
<li>降低初始的学习速率，<code>--lr 0.0002</code></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/17/%E4%BD%BF%E7%94%A8SCALE%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE-Analysis-sc-ATAC-seq-with-SCALE/" data-id="clm1ywdkd00drecokhdupf2w8" data-title="使用SCALE分析单细胞ATAC-seq数据" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-「文献」杂合基因组的策略思路-Assembly-heterozygous-genome-by-interspecific-hybrids" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/16/%E3%80%8C%E6%96%87%E7%8C%AE%E3%80%8D%E6%9D%82%E5%90%88%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%9A%84%E7%AD%96%E7%95%A5%E6%80%9D%E8%B7%AF-Assembly-heterozygous-genome-by-interspecific-hybrids/" class="article-date">
  <time class="dt-published" datetime="2019-10-16T18:48:45.776Z" itemprop="datePublished">2019-10-16</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/16/%E3%80%8C%E6%96%87%E7%8C%AE%E3%80%8D%E6%9D%82%E5%90%88%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%9A%84%E7%AD%96%E7%95%A5%E6%80%9D%E8%B7%AF-Assembly-heterozygous-genome-by-interspecific-hybrids/">「文献」杂合基因组的策略思路</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>文献出处: <a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41438-019-0139-1">Sequencing a Juglans regia × J. microcarpa hybrid yields high-quality genome assemblies of parental specie</a></p>
<p>文章的亮点在于通过对一个F1子代进行三代测序，之后利用BioNano组装出两个亲本的光学图谱，最后根据光学图谱从F1中拆分出两套单倍型。</p>
<p>杂合基因组组装的时候，通常会通过构建近交系，花粉加倍(DH)来降低杂合度。文章采用了种间杂交的方式来避免杂合。这是因为种间杂交的基因组通常包括亲本的两个单倍型，因此能够从中分离出两个亲本的单倍型基因组。种间杂交的一个优势在于，它比较容易构建。</p>
<p>分析策略流程图如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571214701466-aa197a4f932a4d26b1283283d7604c63.png" alt="流程图"><br>文字版讲解如下：</p>
<p>Step1: 使用两个近源物种（都有基因组草图）进行杂交得到F1(MS1-56)，使用PacBio进行测序组装（460条contig, 1.05G, 两倍基因组大小）</p>
<p>Step2: 使用BioNano构建F1和两个亲本的光学图谱。</p>
<p>Step3: 通过自我比对（self-alignment），鉴定haploid和diploid。 亲本的光学图谱包含两个部分，一个是haploid（两个单倍体因为相似而坍缩成一个)，另一个是diploid(两个单倍体因为不够相似被拆分， 也就是分型(phased))。大概是下面这个效果，a表示Serr ctg4上存在两套单倍型的情况.</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/1571221028944-3d69ebf2957849ab8a0c8fb1955bd2a5.png" alt="光学图谱自我比对"></p>
<p>Step4: 将contig回帖到杂合的光学图谱上，组装成scaffolds，大小为1.06G。之后根据杂合后代光学图谱和亲本光学图谱的关系，对scaffold进行拆分。上图b,c表示的是将F1后代的光学图谱和双亲的光学图谱进行比较，来鉴定出子代中属于一方亲本的单倍型。</p>
<p>Step5: 上述因为太短而无法回帖到光学图谱的contig，比对到illumina组装的双亲基因组，进行区分。</p>
<p>Step6: 根据高密度遗传图谱对scaffold进行排序和确认方向，最终得到准染色体级别的基因组。</p>
<p>最后基因组还可以补洞和纠错，来提高质量，比如说10X Genomics Chicago 测序。</p>
<blockquote>
<p>最后一点感想：文章利用光学图谱进行分型的思路的确很棒。但是我想的是，为啥不直接测亲本基因组，用HiC进行分型呢？</p>
</blockquote>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/16/%E3%80%8C%E6%96%87%E7%8C%AE%E3%80%8D%E6%9D%82%E5%90%88%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%9A%84%E7%AD%96%E7%95%A5%E6%80%9D%E8%B7%AF-Assembly-heterozygous-genome-by-interspecific-hybrids/" data-id="clm1ywdis008zecok2ei7hmel" data-title="「文献」杂合基因组的策略思路" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用Rcpp提高性能之入门篇-Import-R-performance-with-Rcpp" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/15/%E4%BD%BF%E7%94%A8Rcpp%E6%8F%90%E9%AB%98%E6%80%A7%E8%83%BD%E4%B9%8B%E5%85%A5%E9%97%A8%E7%AF%87-Import-R-performance-with-Rcpp/" class="article-date">
  <time class="dt-published" datetime="2019-10-15T22:11:19.460Z" itemprop="datePublished">2019-10-15</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/15/%E4%BD%BF%E7%94%A8Rcpp%E6%8F%90%E9%AB%98%E6%80%A7%E8%83%BD%E4%B9%8B%E5%85%A5%E9%97%A8%E7%AF%87-Import-R-performance-with-Rcpp/">使用Rcpp提高性能之入门篇</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p><code>C++</code>能解决的瓶颈问题有:</p>
<ul>
<li>由于迭代依赖于之前结果，循环难以简便的向量化运算</li>
<li>递归函数，或者是需要对同一个函数运算成千上万次</li>
<li>R语言缺少一些高级数据结构和算法</li>
</ul>
<p>我们只需要在代码中写一部分<code>C++</code>代码来就可以处理上面这些问题。后续操作在Windows下进行，你需要安装Rtools，用<code>install.packages(&quot;Rcpp&quot;)</code>安装新版的Rcpp，最重要一点，你需要保证你R语言时不能是<code>C:/Program Files/R/R-3.5.1/</code>这种形式，否则会报错。</p>
<p>后续操作会用到<code>microbenchmark</code>包来评估R代码和RCPP的效率差异，用<code>install.packages(&#39;microbenchmark)</code>安装</p>
<h2 id="RCPP入门"><a href="#RCPP入门" class="headerlink" title="RCPP入门"></a>RCPP入门</h2><p>先从一个简单的<code>add</code>函数开始，学习如何用<code>cppFunction</code>在R里面写<code>C++</code>代码</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>Rcpp<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cppFunction<span class="punctuation">(</span><span class="string">&#x27;int add(int x, int y, int z) &#123;</span></span><br><span class="line"><span class="string">  int sum = x + y + z;</span></span><br><span class="line"><span class="string">  return sum;</span></span><br><span class="line"><span class="string">&#125;&#x27;</span><span class="punctuation">)</span></span><br><span class="line">add</span><br><span class="line"><span class="comment"># function (x, y) </span></span><br><span class="line"><span class="comment"># .Call(&lt;pointer: 0x0000000063c015a0&gt;, x, y)</span></span><br></pre></td></tr></table></figure>

<p>Rcpp将会编译<code>C++</code>代码， 然后构建能够连接到C++函数的R函数。后续将会介绍如何将一些R代码改写成<code>C++</code>代码。</p>
<ul>
<li>标量输入，标量输出</li>
<li>向量输入，标量输出</li>
<li>向量输入，向量输出</li>
<li>矩阵输入，向量输出</li>
</ul>
<h3 id="没有输入，标量输出"><a href="#没有输入，标量输出" class="headerlink" title="没有输入，标量输出"></a>没有输入，标量输出</h3><p>最简单的函数就是不提供任何输出，返回一个输出，比如说</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">one <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span><span class="punctuation">)</span> <span class="number">1L</span></span><br></pre></td></tr></table></figure>

<p>等价的<code>C++</code>代码是</p>
<figure class="highlight cpp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="type">int</span> <span class="title">one</span><span class="params">()</span></span>&#123;</span><br><span class="line">    <span class="keyword">return</span> <span class="number">1</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>那么将这段<code>C++</code>代码在R用<code>cppFunction</code>中改写就是如下</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cppFunction<span class="punctuation">(</span><span class="string">&#x27;int one()&#123;</span></span><br><span class="line"><span class="string">  return 1;</span></span><br><span class="line"><span class="string">&#125;&#x27;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>上面这段函数就展示了R和<code>C++</code>之间一些重要区别:</p>
<ul>
<li><code>C++</code>写代码不是<code> 函数名 &lt;- function(参数)&#123;&#125;</code> 而是 <code>函数名(函数参数)&#123;&#125;</code></li>
<li><code>C++</code>中必须声明返回类型，<code>ini</code>就是标量整数。C++对应R语言常用向量的类是: <code>NumericVector</code>,<code>IntegerVector</code>, <code>CharacterVector</code>和<code>LogicalVector</code>.</li>
<li>R语言没有标量，全是向量。而<code>C++</code>有向量和标量之分，标量的数据类型是<code>double</code>, <code>int</code>, <code>String</code>和<code>bool</code></li>
<li><code>C++</code>你必须要用到<code>return</code>声明要返回的数据</li>
<li>每段代码后要跟着<code>;</code></li>
</ul>
<h3 id="标量输入，标量输出"><a href="#标量输入，标量输出" class="headerlink" title="标量输入，标量输出"></a>标量输入，标量输出</h3><p>我们可以写一个函数，<code>sign</code>，他的功能就是把一个负数转成正数，正数不变</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">signR <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  <span class="keyword">if</span> <span class="punctuation">(</span>x <span class="operator">&gt;</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    x</span><br><span class="line">  <span class="punctuation">&#125;</span> <span class="keyword">else</span> <span class="keyword">if</span> <span class="punctuation">(</span>x <span class="operator">==</span> <span class="number">0</span> <span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    <span class="number">0</span></span><br><span class="line">  <span class="punctuation">&#125;</span> <span class="keyword">else</span><span class="punctuation">&#123;</span></span><br><span class="line">    <span class="operator">-</span>x</span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">cppFunction<span class="punctuation">(</span><span class="string">&#x27;int signC(int x)&#123;</span></span><br><span class="line"><span class="string">            if( x &gt;0 )&#123;</span></span><br><span class="line"><span class="string">              return x;</span></span><br><span class="line"><span class="string">            &#125; else if (x == 0)&#123;</span></span><br><span class="line"><span class="string">              return 0;</span></span><br><span class="line"><span class="string">            &#125; else &#123;</span></span><br><span class="line"><span class="string">              return -x;</span></span><br><span class="line"><span class="string">            &#125;</span></span><br><span class="line"><span class="string">&#125;&#x27;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这个例子中要注意两件事情</p>
<ul>
<li>在<code>C++</code>中，你需要声明输入的数据类型</li>
<li><code>C++</code>和R的条件语句长得一样。</li>
</ul>
<h3 id="向量输入，标量输出"><a href="#向量输入，标量输出" class="headerlink" title="向量输入，标量输出"></a>向量输入，标量输出</h3><p>R和<code>C++</code>一大区别就是R的循环效率很低。因此在R语言要尽量避免使用显示的循环语句，尽量向量化运算函数。而<code>C++</code>的循环花销特别小，所以可以放心大胆的用。</p>
<p>让我们用R代码写一个求和函数<code>sum</code> 以及 <code>C++</code>的求和函数，然后比较下效率</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">sumR <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  total <span class="operator">&lt;-</span> 0</span><br><span class="line">  <span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    total <span class="operator">&lt;-</span> total <span class="operator">+</span> x<span class="punctuation">[</span>i<span class="punctuation">]</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  total</span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">cppFunction<span class="punctuation">(</span><span class="string">&#x27;int sumC(NumericVector x )&#123; </span></span><br><span class="line"><span class="string">            int n = x.size();</span></span><br><span class="line"><span class="string">            double total = 0;</span></span><br><span class="line"><span class="string">            for(int i = 0; i &lt; n; ++i)&#123;</span></span><br><span class="line"><span class="string">              total += x[i];  </span></span><br><span class="line"><span class="string">            &#125;</span></span><br><span class="line"><span class="string">            return total;</span></span><br><span class="line"><span class="string">            &#125;&#x27;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><code>C++</code>版本和R版本的逻辑相同，但是有如下不同</p>
<ul>
<li>用<code>.size()</code>确认向量的长度</li>
<li><code>for</code>的写法为<code>for(初始值; 判断语句; 递增)</code></li>
<li><strong>记住</strong>: <code>C++</code>的向量索引从0开始，R是从1开始</li>
<li>向量赋值是<code>=</code>而不是<code>&lt;-</code></li>
<li><code>total += x[i]</code>等价于<code>total = total + x[i]</code>, 类似的符号还有<code>-=</code>, <code>*=</code>, <code>/=</code></li>
</ul>
<p>最后用<code>microbenchmark</code>比较下,R自带求和函数和我们自己写的两个版本的差异</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">x <span class="operator">&lt;-</span> runif<span class="punctuation">(</span><span class="number">1000</span><span class="punctuation">)</span></span><br><span class="line">microbenchmark<span class="punctuation">(</span></span><br><span class="line">  <span class="built_in">sum</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  sumC<span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  sumR<span class="punctuation">(</span>x<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>最快的是高度优化过的内置函数，最差的就是<code>sumR()</code>, 速度会比<code>sumC()</code>慢10倍以上。 </p>
<h3 id="向量输入，向量输出"><a href="#向量输入，向量输出" class="headerlink" title="向量输入，向量输出"></a>向量输入，向量输出</h3><p>R中比较常见的操作就是向量间运算，尤其R还会自动补齐。自动补齐某些时候会造成一些问题，但是C++不存在这个问题。我们可以写一个RCPP的<code>+</code>函数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">cppFunction<span class="punctuation">(</span><span class="string">&#x27;NumericVector addC(NumericVector x, NumericVector y)&#123;</span></span><br><span class="line"><span class="string">  int xn = x.size();</span></span><br><span class="line"><span class="string">  int yn = y.size();</span></span><br><span class="line"><span class="string">  </span></span><br><span class="line"><span class="string">  if (xn != yn)&#123;</span></span><br><span class="line"><span class="string">    stop(&quot;input should be same length&quot;);</span></span><br><span class="line"><span class="string">  &#125;</span></span><br><span class="line"><span class="string">  NumericVector out(xn);</span></span><br><span class="line"><span class="string">  for(int i=0; i&lt; xn; ++i)&#123;</span></span><br><span class="line"><span class="string">    out[i] = x[i] + y[i];</span></span><br><span class="line"><span class="string">  &#125;</span></span><br><span class="line"><span class="string">  return out;</span></span><br><span class="line"><span class="string">&#125;&#x27;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">x <span class="operator">&lt;-</span> runif<span class="punctuation">(</span><span class="number">1e6</span><span class="punctuation">)</span></span><br><span class="line">y <span class="operator">&lt;-</span> runif<span class="punctuation">(</span><span class="number">1e6</span><span class="punctuation">)</span></span><br><span class="line">microbenchmark<span class="punctuation">(</span>addC<span class="punctuation">(</span>x<span class="punctuation">,</span>y<span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">               x<span class="operator">+</span>y<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="矩阵输入，向量输出"><a href="#矩阵输入，向量输出" class="headerlink" title="矩阵输入，向量输出"></a>矩阵输入，向量输出</h3><p>每个向量类型都有矩阵等价类，<code>NumericMatrix</code>, <code>IntegerMatirx</code>, <code>CharacterMatirx</code>, <code>LogicalMatirx</code>. 让我们尝试写一个<code>rowSums()</code>函数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">cppFunction<span class="punctuation">(</span><span class="string">&#x27;NumericVector rowSumsC(NumericMatrix x)&#123;</span></span><br><span class="line"><span class="string">  int nrow = x.nrow(), ncol = x.ncol();</span></span><br><span class="line"><span class="string">  NumericVector out(nrow);</span></span><br><span class="line"><span class="string">  </span></span><br><span class="line"><span class="string">  for(int i = 0; i &lt; nrow; i++)&#123;</span></span><br><span class="line"><span class="string">    double total =0;</span></span><br><span class="line"><span class="string">    for(int j =0; j&lt; ncol; j++)&#123;</span></span><br><span class="line"><span class="string">      total += x(i,j);</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">    out[i] = total;</span></span><br><span class="line"><span class="string">  &#125;</span></span><br><span class="line"><span class="string">  return out;</span></span><br><span class="line"><span class="string">&#125;&#x27;</span><span class="punctuation">)</span></span><br><span class="line">set.seed<span class="punctuation">(</span><span class="number">1024</span><span class="punctuation">)</span></span><br><span class="line">x <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span>sample<span class="punctuation">(</span><span class="number">100</span><span class="punctuation">)</span><span class="punctuation">,</span> nrow <span class="operator">=</span> <span class="number">10</span><span class="punctuation">)</span></span><br><span class="line">rowSumsC<span class="punctuation">(</span>x<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这里注意有两点不同，在<code>C++</code>中，你用<code>()</code>对矩阵取值，而不是<code>[]</code></p>
<p>尽管看起来<code>C++</code>的代码运行起来比R语言快多了，比如说R要一分钟，RCPP只要一秒，但是如果算上我们写代码的时间和调试代码的时间，刚开始不熟练估计要10分钟，那么总体来看，还是直接上手写R代码比较合适。</p>
<p>但是如果有一些代码要不断复用，那么写<code>C++</code>代码还是很划算。这个时候就建议将代码写到专门的文本中，用<code>sourceCpp()</code>加载，而不是<code>cppFunction()</code>函数</p>
<p>在Rsutdio中可以创建一个<code>C++</code>模板文件，代码写完之后还可以进行debug。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/image-17ff0fcaeda34cf5bbc0c7676ee3aa11.png" alt="创建模板"></p>
<p>比如说在里面写上面的<code>rowSumsC</code>函数，分为如下几个部分</p>
<p>导入头文件，加载Rcpp到命名空间中，类似于<code>library()</code></p>
<figure class="highlight cpp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;Rcpp.h&gt;</span></span></span><br><span class="line"><span class="keyword">using</span> <span class="keyword">namespace</span> Rcpp;</span><br></pre></td></tr></table></figure>

<p>使用<code>// [[Rcpp::export]]</code>说明这里的函数会被R使用</p>
<figure class="highlight cpp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// [[Rcpp::export]]</span></span><br><span class="line"><span class="function">NumericVector <span class="title">rowSumsC</span><span class="params">(NumericMatrix x)</span></span>&#123;</span><br><span class="line">  <span class="type">int</span> ncol = x.<span class="built_in">ncol</span>(), nrow = x.<span class="built_in">nrow</span>();</span><br><span class="line">  <span class="function">NumericVector <span class="title">out</span><span class="params">(nrow)</span></span>;</span><br><span class="line">  </span><br><span class="line">  <span class="keyword">for</span> (<span class="type">int</span> i =<span class="number">0</span>; i &lt; nrow; i++ )&#123;</span><br><span class="line">    <span class="type">double</span> total = <span class="number">0</span>;</span><br><span class="line">    <span class="keyword">for</span> (<span class="type">int</span> j =<span class="number">0</span> ;j &lt; ncol; j++)&#123;</span><br><span class="line">      total += <span class="built_in">x</span>(i,j);</span><br><span class="line">    &#125;</span><br><span class="line">    out[i] = total;</span><br><span class="line">  &#125;</span><br><span class="line">  <span class="keyword">return</span> out;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>下面部分会在<code>sourceCpp()</code>加载后自动运行</p>
<figure class="highlight cpp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">/*** R</span></span><br><span class="line"><span class="comment">library(microbenchmark)</span></span><br><span class="line"><span class="comment">set.seed(1014)</span></span><br><span class="line"><span class="comment">x &lt;- matrix(sample(100), 10)</span></span><br><span class="line"><span class="comment">microbenchmark(</span></span><br><span class="line"><span class="comment">  rowSumsC(x),</span></span><br><span class="line"><span class="comment">  Matrix::rowSums(x)</span></span><br><span class="line"><span class="comment">)</span></span><br><span class="line"><span class="comment">*/</span></span><br></pre></td></tr></table></figure>

<p>将文件保存成<code>rowSumsC.cpp</code>, 之后在R里用<code>sourceCpp(file = &quot;rowSumsC.cpp&quot;)</code>。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/15/%E4%BD%BF%E7%94%A8Rcpp%E6%8F%90%E9%AB%98%E6%80%A7%E8%83%BD%E4%B9%8B%E5%85%A5%E9%97%A8%E7%AF%87-Import-R-performance-with-Rcpp/" data-id="clm1ywdkb00dkecokf4g90w93" data-title="使用Rcpp提高性能之入门篇" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-Rcpp学习笔记之数据结构-Rcpp_learn_note_data_structure" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/15/Rcpp%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%E4%B9%8B%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-Rcpp_learn_note_data_structure/" class="article-date">
  <time class="dt-published" datetime="2019-10-15T22:01:05.852Z" itemprop="datePublished">2019-10-15</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/15/Rcpp%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%E4%B9%8B%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-Rcpp_learn_note_data_structure/">Rcpp学习笔记之数据结构</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <blockquote>
<p>在使用R语言多年以后，我终于开始去学习Rcpp，利用C++来提高运行速度。其实当你能熟练的使用一门语言后，再去学一门新的语言，并没有想象中的那么难，更何况Rcpp把很多脏活累活都给包办了，在里面调用C++还是挺方便。</p>
</blockquote>
<p>C++是一门静态编译面向对象的编程语言，R是动态解释性面向对象语言，那么有一个不同就在于，你需要先声明一个变量，才能调用该变量。而在声明变量的时候，你就会遇到一个R语言中不怎么思考的问题，我这个变量要存放什么样的数据呢？</p>
<h2 id="数据类型"><a href="#数据类型" class="headerlink" title="数据类型"></a>数据类型</h2><p>R的基本数据类型有六种 “logical”, “integer”, “numeric” (等价于”double”), “complex”, “character” 和 “raw”，但常用的就四种， “integer”, “numeric”,”logical”和”character”。</p>
<p>对于数值数据而言，<code>C++</code>的整型和浮点型可以定义多种精度，而R语言则简化成两种”integer”(32 bit)和”numeric”(53 bits)，并且默认情况下数值都是浮点型。</p>
<p>在数据结构上，R语言提供了向量(vector)，矩阵(matrix)，数组(array)，数据框(data.frame)和列表(list)，唯独没有标量。其中向量可以认为是R语言的基本单位，是其他数据结构的基础。</p>
<p>原本能够让<code>C++</code>处理R对象，以及将<code>C++</code>处理完结果转成R能识别的对象，你需要了解R的内部结构，但是<code>Rcpp</code>通过定义了一系列数据类型（如下），使得我们能够非常容易地在R和<code>C++</code>之间交换对象。</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// Rcpp/include/Rcpp/vector/instantiation.h</span></span><br><span class="line"><span class="keyword">namespace</span> Rcpp&#123;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">typedef</span> Vector&lt;CPLXSXP&gt; ComplexVector ;</span><br><span class="line">    <span class="keyword">typedef</span> Vector&lt;INTSXP&gt; IntegerVector ;</span><br><span class="line">    <span class="keyword">typedef</span> Vector&lt;LGLSXP&gt; LogicalVector ;</span><br><span class="line">    <span class="keyword">typedef</span> Vector&lt;REALSXP&gt; NumericVector ;</span><br><span class="line">    <span class="keyword">typedef</span> Vector&lt;REALSXP&gt; DoubleVector ;</span><br><span class="line">    <span class="keyword">typedef</span> Vector&lt;RAWSXP&gt; RawVector ;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">typedef</span> Vector&lt;STRSXP&gt; CharacterVector ;</span><br><span class="line">    <span class="keyword">typedef</span> Vector&lt;STRSXP&gt; StringVector ;</span><br><span class="line">    <span class="keyword">typedef</span> Vector&lt;VECSXP&gt; GenericVector ;</span><br><span class="line">    <span class="keyword">typedef</span> Vector&lt;VECSXP&gt; List ;</span><br><span class="line">    <span class="keyword">typedef</span> Vector&lt;EXPRSXP&gt; ExpressionVector ;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">typedef</span> Matrix&lt;CPLXSXP&gt; ComplexMatrix ;</span><br><span class="line">    <span class="keyword">typedef</span> Matrix&lt;INTSXP&gt; IntegerMatrix ;</span><br><span class="line">    <span class="keyword">typedef</span> Matrix&lt;LGLSXP&gt; LogicalMatrix ;</span><br><span class="line">    <span class="keyword">typedef</span> Matrix&lt;REALSXP&gt; NumericMatrix ;</span><br><span class="line">    <span class="keyword">typedef</span> Matrix&lt;RAWSXP&gt; RawMatrix ;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">typedef</span> Matrix&lt;STRSXP&gt; CharacterMatrix ;</span><br><span class="line">    <span class="keyword">typedef</span> Matrix&lt;STRSXP&gt; StringMatrix ;</span><br><span class="line">    <span class="keyword">typedef</span> Matrix&lt;VECSXP&gt; GenericMatrix ;</span><br><span class="line">    <span class="keyword">typedef</span> Matrix&lt;VECSXP&gt; ListMatrix ;</span><br><span class="line">    <span class="keyword">typedef</span> Matrix&lt;EXPRSXP&gt; ExpressionMatrix ;</span><br><span class="line"></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>以R语言的向量为例，Rcpp考虑到R语言所有可能的数据类型，提供了9种Vector。比如说”NumericVector”, “IntegerVector”, “LogicalVector”,” CharacterVector”就是对应着”integer”, “numeric”,”logical”和”character”这四种数据类型。因此，你可以根据你输入向量的数据类型来进行选择。</p>
<p>我们以一个简单的求和函数作为例子</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">double sumC<span class="punctuation">(</span>NumericVector v<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line"></span><br><span class="line">    double total <span class="operator">=</span> <span class="number">0</span>;</span><br><span class="line">    int num <span class="operator">=</span> v.size<span class="punctuation">(</span><span class="punctuation">)</span>;</span><br><span class="line">    <span class="keyword">for</span> <span class="punctuation">(</span>int i <span class="operator">=</span> <span class="number">0</span>; i<span class="operator">&lt;</span> num; i<span class="operator">+</span><span class="operator">+</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">        total <span class="operator">+</span><span class="operator">=</span> v<span class="punctuation">[</span>i<span class="punctuation">]</span>;</span><br><span class="line">    <span class="punctuation">&#125;</span></span><br><span class="line">    <span class="built_in">return</span> total;</span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p><code>sumC</code>这个函数读取一个数值向量，对其进行结合，返回一个浮点型标量。当然这个浮点型标量在R语言中就表现为一个长度为1的向量。</p>
<p>由于R语言数据类型和<code>C++</code>的不是一一对应，因此在处理过程中，有些时候需要对数据进行转换</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">std::string(x[i])</span><br></pre></td></tr></table></figure>

<h2 id="变量创建"><a href="#变量创建" class="headerlink" title="变量创建"></a>变量创建</h2><p>我们可以通过下面这些方式创建向量</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// v &lt;- rep(0, 3)</span></span><br><span class="line"><span class="function">NumericVector <span class="title">v</span> <span class="params">(<span class="number">3</span>)</span></span>;</span><br><span class="line"><span class="comment">// v &lt;- rep(1, 3)</span></span><br><span class="line"><span class="function">NumericVector <span class="title">v</span> <span class="params">(<span class="number">3</span>,<span class="number">1</span>)</span></span>;</span><br><span class="line"><span class="comment">// v &lt;- c(1,2,3) </span></span><br><span class="line"><span class="comment">// C++11 Initializer list</span></span><br><span class="line">NumericVector v = &#123;<span class="number">1</span>,<span class="number">2</span>,<span class="number">3</span>&#125;; </span><br><span class="line"><span class="comment">// v &lt;- c(1,2,3)</span></span><br><span class="line">NumericVector v = NumericVector::<span class="built_in">create</span>(<span class="number">1</span>,<span class="number">2</span>,<span class="number">3</span>);</span><br><span class="line"><span class="comment">// v &lt;- c(x=1, y=2, z=3)</span></span><br><span class="line">NumericVector v = NumericVector::<span class="built_in">create</span>(<span class="built_in">Named</span>(<span class="string">&quot;x&quot;</span>,<span class="number">1</span>), <span class="built_in">Named</span>(<span class="string">&quot;y&quot;</span>)=<span class="number">2</span> , _[<span class="string">&quot;z&quot;</span>]=<span class="number">3</span>);</span><br></pre></td></tr></table></figure>

<p>创建矩阵的方法如下</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// m &lt;- matrix(0, nrow=2, ncol=2)</span></span><br><span class="line"><span class="function">NumericMatrix <span class="title">m1</span><span class="params">( <span class="number">2</span> )</span></span>;</span><br><span class="line"><span class="comment">// m &lt;- matrix(0, nrow=2, ncol=3)</span></span><br><span class="line"><span class="function">NumericMatrix <span class="title">m2</span><span class="params">( <span class="number">2</span> , <span class="number">3</span> )</span></span>;</span><br><span class="line"><span class="comment">// m &lt;- matrix(v, nrow=2, ncol=3)</span></span><br><span class="line"><span class="function">NumericMatrix <span class="title">m3</span><span class="params">( <span class="number">2</span> , <span class="number">3</span> , v.begin() )</span></span>;</span><br></pre></td></tr></table></figure>

<p>数据框和列表的创建依赖于已有向量，和R语言中创建形式相似。</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// Creating DataFrame df from Vector v1, v2</span></span><br><span class="line">DataFrame df = DataFrame::<span class="built_in">create</span>(v1, v2);</span><br><span class="line"><span class="comment">// When giving names to columns</span></span><br><span class="line">DataFrame df = DataFrame::<span class="built_in">create</span>( <span class="built_in">Named</span>(<span class="string">&quot;V1&quot;</span>) = v1 , _[<span class="string">&quot;V2&quot;</span>] = v2 );</span><br><span class="line"></span><br><span class="line"><span class="comment">// Create list L from vector v1, v2</span></span><br><span class="line">List L = List::<span class="built_in">create</span>(v1, v2);</span><br><span class="line"><span class="comment">// When giving names to elements</span></span><br><span class="line">List L = List::<span class="built_in">create</span>(<span class="built_in">Named</span>(<span class="string">&quot;name1&quot;</span>) = v1 , _[<span class="string">&quot;name2&quot;</span>] = v2);a</span><br></pre></td></tr></table></figure>

<h2 id="元素获取和赋值"><a href="#元素获取和赋值" class="headerlink" title="元素获取和赋值"></a>元素获取和赋值</h2><p>在选取元素之前一定要注意，<code>C++</code>是以0为基，而R是以1为基。</p>
<p>我们可以用<code>[]</code>和<code>()</code>进行数据选取向量中的元素并复制，支持利用数值向量或者逻辑向量来选取多个数据</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// 新建向量</span></span><br><span class="line">NumericVector v  &#123;<span class="number">10</span>,<span class="number">20</span>,<span class="number">30</span>,<span class="number">40</span>,<span class="number">50</span>&#125;;</span><br><span class="line"><span class="comment">// 设置向两名</span></span><br><span class="line">v.<span class="built_in">names</span>() = <span class="built_in">CharacterVector</span>(&#123;<span class="string">&quot;A&quot;</span>,<span class="string">&quot;B&quot;</span>,<span class="string">&quot;C&quot;</span>,<span class="string">&quot;D&quot;</span>,<span class="string">&quot;E&quot;</span>&#125;);</span><br><span class="line"><span class="comment">// 准备用于获取数据的向量</span></span><br><span class="line">NumericVector   numeric = &#123;<span class="number">1</span>,<span class="number">3</span>&#125;;</span><br><span class="line">IntegerVector   integer = &#123;<span class="number">1</span>,<span class="number">3</span>&#125;;</span><br><span class="line">CharacterVector character = &#123;<span class="string">&quot;B&quot;</span>,<span class="string">&quot;D&quot;</span>&#125;;</span><br><span class="line">LogicalVector   logical = &#123;<span class="literal">false</span>, <span class="literal">true</span>, <span class="literal">false</span>, <span class="literal">true</span>, <span class="literal">false</span>&#125;;</span><br><span class="line"><span class="comment">// 获取数据</span></span><br><span class="line"><span class="type">double</span> x1 = v[<span class="number">0</span>];</span><br><span class="line"><span class="type">double</span> x2 = v[<span class="string">&quot;A&quot;</span>];</span><br><span class="line">NumericVector res1 = v[numeric];</span><br><span class="line">NumericVector res2 = v[integer];</span><br><span class="line">NumericVector res3 = v[character];</span><br><span class="line">NumericVector res4 = v[logical];</span><br><span class="line"><span class="comment">// 赋值</span></span><br><span class="line">v[<span class="number">0</span>]   = <span class="number">100</span>;</span><br><span class="line">v[<span class="string">&quot;A&quot;</span>] = <span class="number">100</span>;</span><br><span class="line">NumericVector v2 &#123;<span class="number">100</span>,<span class="number">200</span>&#125;;</span><br><span class="line">v[numeric]   = v2;</span><br><span class="line">v[integer]   = v2;</span><br><span class="line">v[character] = v2;</span><br><span class="line">v[logical]   = v2;</span><br></pre></td></tr></table></figure>

<p>对于矩阵而言，建议只用<code>()</code>进行数据选取，因为没有<code>[row_index, col_index]</code>的操作。</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// 创建一个5x5的矩阵</span></span><br><span class="line"><span class="function">NumericMatrix <span class="title">m</span><span class="params">( <span class="number">5</span>, <span class="number">5</span> )</span></span>;</span><br><span class="line"><span class="comment">// 获取0,2的元素</span></span><br><span class="line"><span class="type">double</span> x = <span class="built_in">m</span>( <span class="number">0</span> , <span class="number">2</span> );</span><br><span class="line"><span class="comment">// 选择第1行</span></span><br><span class="line">NumericVector v = <span class="built_in">m</span>( <span class="number">0</span> , _ );</span><br><span class="line"><span class="comment">// 选择第3列</span></span><br><span class="line">NumericVector v = <span class="built_in">m</span>( _ , <span class="number">2</span> );</span><br><span class="line"><span class="comment">// 选择第1-2行，3-4列</span></span><br><span class="line">NumericMatrix m2 = <span class="built_in">m</span>( <span class="built_in">Range</span>(<span class="number">0</span>,<span class="number">1</span>) , <span class="built_in">Range</span>(<span class="number">2</span>,<span class="number">3</span>) );</span><br></pre></td></tr></table></figure>

<p>对于DataFrame和List，两则都只支持<code>[]</code>选择其中向量。</p>
<h2 id="成员函数"><a href="#成员函数" class="headerlink" title="成员函数"></a>成员函数</h2><p>成员函数(member function)是<code>C++</code>面向对象编程中的一个概念，通过调用成员函数可以对类进行操作。</p>
<p>举个例子，对于R语言而言，向量是可以有名字的，例如</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">x <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span>a <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span> b <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> <span class="built_in">c</span> <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span>	</span><br></pre></td></tr></table></figure>

<p>R语言是一个面向对象的编程语言，所以这种命名向量其实认为是一个向量拥有了一个名为<code>names</code>的属性而已。</p>
<p>在R语言中<code>attr</code>函数能够增加对象的属性，因此上面代码等价于</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">y <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">)</span></span><br><span class="line"><span class="built_in">attr</span><span class="punctuation">(</span>y<span class="punctuation">,</span> <span class="string">&#x27;names&#x27;</span><span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;a&#x27;</span><span class="punctuation">,</span><span class="string">&#x27;b&#x27;</span><span class="punctuation">,</span><span class="string">&#x27;c&#x27;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>而<code>C++</code>中的向量和矩阵拥有一个成员函数<code>.name()</code>就可以获取&#x2F;修改&#x2F;增加变量命名。</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="function">NumericVector <span class="title">addname2</span><span class="params">(NumericVector x, CharacterVector y)</span></span>&#123;</span><br><span class="line">    x.<span class="built_in">names</span>() = y;</span><br><span class="line">    <span class="keyword">return</span> x;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>更通用的是<code>.attr</code>函数,它能够修改和增加任意属性</p>
<p>不同数据类型拥有不同的成员函数，参考</p>
<ul>
<li>向量:  <a target="_blank" rel="noopener" href="https://teuder.github.io/rcpp4everyone_en/080_vector.html#member-functions">https://teuder.github.io/rcpp4everyone_en/080_vector.html#member-functions</a></li>
<li>矩阵: <a target="_blank" rel="noopener" href="https://teuder.github.io/rcpp4everyone_en/100_matrix.html#member-functions-1">https://teuder.github.io/rcpp4everyone_en/100_matrix.html#member-functions-1</a></li>
<li>数据框:  <a target="_blank" rel="noopener" href="https://teuder.github.io/rcpp4everyone_en/140_dataframe.html#member-functions-2">https://teuder.github.io/rcpp4everyone_en/140_dataframe.html#member-functions-2</a></li>
<li>列表: <a target="_blank" rel="noopener" href="https://teuder.github.io/rcpp4everyone_en/150_list.html#member-functions-3">https://teuder.github.io/rcpp4everyone_en/150_list.html#member-functions-3</a></li>
<li>字符串:  <a target="_blank" rel="noopener" href="https://teuder.github.io/rcpp4everyone_en/170_string.html#member-functions-4">https://teuder.github.io/rcpp4everyone_en/170_string.html#member-functions-4</a></li>
<li>日期: <a target="_blank" rel="noopener" href="https://teuder.github.io/rcpp4everyone_en/180_date.html#member-functions-5">https://teuder.github.io/rcpp4everyone_en/180_date.html#member-functions-5</a></li>
<li>时间: <a target="_blank" rel="noopener" href="https://teuder.github.io/rcpp4everyone_en/190_datetime.html#member-functions-6">https://teuder.github.io/rcpp4everyone_en/190_datetime.html#member-functions-6</a></li>
</ul>
<h2 id="缺失值"><a href="#缺失值" class="headerlink" title="缺失值"></a>缺失值</h2><blockquote>
<p>同R一样，缺失值之间以及缺失值和正常值之间的比较没有意义，因此在写代码过程时要提前考虑到缺失值的可能，或者是用R语言处理完缺失值，把不含缺失值的数据作为<code>C++</code>函数的输入。</p>
</blockquote>
<p>R语言采用比特模式对每一种数据类型进行标注，也就是针对每一种数据类型（Integer，Numeric，Character，Logical）保留一个比特作为缺失数据的标签值。</p>
<p>为了在C++中处理R的缺失值，Rcpp提供了四个变量对应R的四种数据类型</p>
<ul>
<li>NA_INTEGER:  整型，Integer</li>
<li>NA_STRING: 字符串，Character</li>
<li>NA_LOGICAL：逻辑性，Logical</li>
<li>NA_REAL：双精度浮点型， Numeric</li>
</ul>
<p>写一个函数进行讲解下缺失值的使用。</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="type">double</span> <span class="title">sumC</span><span class="params">(NumericVector v, <span class="type">bool</span> na_rm)</span></span>&#123;</span><br><span class="line"></span><br><span class="line">    <span class="type">double</span> total = <span class="number">0</span>;</span><br><span class="line">    <span class="keyword">for</span> (<span class="type">int</span> i = <span class="number">0</span>; i&lt; v.<span class="built_in">size</span>(); i++)&#123;</span><br><span class="line">        <span class="keyword">if</span>( ! NumericVector::<span class="built_in">is_na</span>(v[i]))&#123;</span><br><span class="line">            total += <span class="built_in">double</span>(v[i]);</span><br><span class="line">        &#125; <span class="keyword">else</span> <span class="keyword">if</span> (NumericVector::<span class="built_in">is_na</span>(v[i]) &amp; na_rm)&#123;</span><br><span class="line">            <span class="keyword">continue</span> ;</span><br><span class="line">        &#125; <span class="keyword">else</span> &#123;</span><br><span class="line">            <span class="keyword">return</span> NA_REAL;</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">return</span> total;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><code>NumericVector::is_na()</code>用于判断是否为缺失值，不同的数据类型有不同的<code>is_na</code>。如果<code>na_rm=FALSE</code>，那么返回缺失值，也就是<code>NA_REAL</code>.</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li><a target="_blank" rel="noopener" href="https://stackoverflow.com/questions/23282849/rcpp-charactervector-why-arent-individual-elements-stdstrings">CharacterVector元素转string</a></li>
<li><a target="_blank" rel="noopener" href="https://teuder.github.io/rcpp4everyone_en">https://teuder.github.io/rcpp4everyone_en</a></li>
<li><a target="_blank" rel="noopener" href="https://adv-r.hadley.nz/rcpp.html#rcpp-intro">https://adv-r.hadley.nz/rcpp.html#rcpp-intro</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/15/Rcpp%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%E4%B9%8B%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-Rcpp_learn_note_data_structure/" data-id="clm1ywdg50055ecokha9bey3z" data-title="Rcpp学习笔记之数据结构" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/16/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/15/">15</a><a class="page-number" href="/page/16/">16</a><span class="page-number current">17</span><a class="page-number" href="/page/18/">18</a><a class="page-number" href="/page/19/">19</a><span class="space">&hellip;</span><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/18/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/09/">九月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/09/02/%E5%A6%82%E4%BD%95%E8%AE%A9%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8D%E5%86%8D%E9%BE%9F%E9%80%9F%E4%B8%8B%E8%BD%BDgithub%E7%9A%84%E6%95%B0%E6%8D%AE/">如何让服务器不再龟速下载github的数据</a>
          </li>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>