<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/9/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-如何用WGDI进行共线性分析(一)-comparative-genome-analysis-with-wgdi-1" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/02/01/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%B8%80)-comparative-genome-analysis-with-wgdi-1/" class="article-date">
  <time class="dt-published" datetime="2021-02-01T13:57:32.701Z" itemprop="datePublished">2021-02-01</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/02/01/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%B8%80)-comparative-genome-analysis-with-wgdi-1/">如何用WGDI进行共线性分析(一)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>多倍化以及后续的基因丢失和二倍化现象存在于大部分的物种中, 是物种进化的重要动力。如果一个物种在演化过程中发生过多倍化，那么在基因组上就会存在一些共线性区域（即两个区域之间的基因是旁系同源基因，其基因的排布顺序基本一致）。</p>
<p>例如拟南芥经历了3次古多倍化，包括2次二倍化，一次3倍化 (Tang. et.al 2008 Science).  </p>
<blockquote>
<p>..For example,  Arabidopsis thaliana (thale cress) has undergone   three paleo-polyploidies, including two doublings  (5) and one tripling (12), resulting in ~12 copies of  its ancestral chromosome set in a ~160-Mb genome…</p>
</blockquote>
<p>当然之前只是记住了结论，现在我想的是，如何复现出这个分析结果呢？</p>
<h2 id="前期准备"><a href="#前期准备" class="headerlink" title="前期准备"></a>前期准备</h2><p>首先，我们需要使用conda安装wgdi。 我一般会新建一个环境避免新装软件和其他软件产生冲突</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 安装软件</span></span><br><span class="line">conda create -c bioconda -c conda-forge -n wgdi wgdi</span><br><span class="line"><span class="comment"># 启动环境</span></span><br><span class="line">conda activate wgdi </span><br></pre></td></tr></table></figure>

<p>之后，我们从ENSEMBL上下载基因组，CDS, 蛋白和GFF文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#Athaliana</span></span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz</span><br><span class="line">gunzip Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/arabidopsis_thaliana/cds/Arabidopsis_thaliana.TAIR10.cds.all.fa.gz</span><br><span class="line">gunzip Arabidopsis_thaliana.TAIR10.cds.all.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/arabidopsis_thaliana/pep/Arabidopsis_thaliana.TAIR10.pep.all.fa.gz</span><br><span class="line">gunzip Arabidopsis_thaliana.TAIR10.pep.all.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.44.gff3.gz</span><br><span class="line">gunzip Arabidopsis_thaliana.TAIR10.44.gff3.gz</span><br></pre></td></tr></table></figure>

<p>注意: cDNA和CDS并不相同，CDS只包括起始密码子到终止密码子之间的序列，而cDNA还会包括UTR.</p>
<h2 id="数据预处理"><a href="#数据预处理" class="headerlink" title="数据预处理"></a>数据预处理</h2><p>数据的预处理是用时最长的步骤，因为软件要求的输入格式并非是你手头所拥有的格式，你通常都需要进行一些转换才能得到它所需的形式。</p>
<p>wgdi需要三种信息，分别是BLAST, 基因的位置信息和染色体长度信息，要求格式如下</p>
<ol>
<li><p>BLAST 的 <code>-outfmt 6</code>输出的文件</p>
</li>
<li><p>基因的位置信息: 以tab分隔，分别为chr，id，start，end，strand，order，old_id。（并非真正意义上的GFF格式）</p>
</li>
<li><p>染色体长度信息和染色体上的基因个数，格式为 chr, length, gene number</p>
</li>
</ol>
<p>同时，对于每个基因我们只需要一个转录本，我通常使用最长的转录本作为该基因的代表。之前我写过一个脚本 (get_the_longest_transcripts.py) 提取每个基因的最长转录本，我在此处上写了一个新的脚本用来根据参考基因组和注释的GFF文件生成wgdi的两个输入文件，脚本地址为<a target="_blank" rel="noopener" href="https://github.com/xuzhougeng/myscripts/blob/master/comparative/generate_conf.py">https://github.com/xuzhougeng/myscripts/blob/master/comparative/generate_conf.py</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">pytho generate_conf.py -p ath Arabidopsis_thaliana.TAIR10.dna.toplevel.fa Arabidopsis_thaliana.TAIR10.44.gff3</span><br></pre></td></tr></table></figure>

<p>输出文件是ath.gff 和 ath.len.</p>
<p>由于ENSEMBL上GFF的ID编号和pep.fa 和cds.fa 不太一致，简单的说就是，编号之前中有一个 “gene:” 和 “transcript:” . 如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">head</span> ath.gff </span><br><span class="line">1  gene:AT1G01010  3631  5899  +  1  transcript:AT1G01010.1</span><br><span class="line">1  gene:AT1G01020  6788  9130  -  2  transcript:AT1G01020.1</span><br><span class="line">1  gene:AT1G01030  11649  13714  -  3  transcript:AT1G01030.1</span><br><span class="line">1  gene:AT1G01040  23416  31120  +  4  transcript:AT1G01040.2</span><br><span class="line">1  gene:AT1G01050  31170  33171  -  5  transcript:AT1G01050.1</span><br><span class="line">1  gene:AT1G01060  33379  37757  -  6  transcript:AT1G01060.3</span><br><span class="line">1  gene:AT1G01070  38444  41017  -  7  transcript:AT1G01070.1</span><br><span class="line">1  gene:AT1G01080  45296  47019  -  8  transcript:AT1G01080.2</span><br><span class="line">1  gene:AT1G01090  47234  49304  -  9  transcript:AT1G01090.1</span><br><span class="line">1  gene:AT1G01100  49909  51210  -  10  transcript:AT1G01100.2</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>于是我用 sed 删除了这些信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sed -i -e <span class="string">&#x27;s/gene://&#x27;</span> -e <span class="string">&#x27;s/transcript://&#x27;</span> ath.gff</span><br></pre></td></tr></table></figure>

<p>另外，pep.fa, cds.fa里面包含所有的基因，且命名特别的长， 同时我们也不需要基因中的 “.1”, “.2” 这部分信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">head</span> -n 5 Arabidopsis_thaliana.TAIR10.cds.all.fa </span><br><span class="line">&gt;AT3G05780.1 cds chromosome:TAIR10:3:1714941:1719608:-1 gene:AT3G05780 gene_biotype:protein_coding transcript_biotype:protein_coding gene_symbol:LON3 description:Lon protease homolog 3, mitochondrial [Source:UniProtKB/Swiss-Prot;Acc:Q9M9L8]</span><br><span class="line">ATGATGCCTAAACGGTTTAACACGAGTGGCTTTGACACGACTCTTCGTCTACCTTCGTAC</span><br><span class="line">TACGGTTTCTTGCATCTTACACAAAGCTTAACCCTAAATTCCCGCGTTTTCTACGGTGCC</span><br><span class="line">CGCCATGTGACTCCTCCGGCTATTCGGATCGGATCCAATCCGGTTCAGAGTCTACTACTC</span><br><span class="line">TTCAGGGCACCGACTCAGCTTACCGGATGGAACCGGAGTTCTCGCGATTTATTGGGTCGTb</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>借助于seqkit,  我对原来的数据进行了筛选和重命名</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">seqkit grep -f &lt;(<span class="built_in">cut</span> -f 7 ath.gff ) Arabidopsis_thaliana.TAIR10.cds.all.fa | seqkit <span class="built_in">seq</span> --id-regexp <span class="string">&quot;^(.*?)\\.\\d&quot;</span> -i &gt; ath.cds.fa</span><br><span class="line">seqkit grep -f &lt;(<span class="built_in">cut</span> -f 7 ath.gff ) Arabidopsis_thaliana.TAIR10.pep.all.fa | seqkit <span class="built_in">seq</span> --id-regexp <span class="string">&quot;^(.*?)\\.\\d&quot;</span> -i &gt; ath.pep.fa  </span><br></pre></td></tr></table></figure>

<p>通过NCBI的BLASTP 或者DIAMOND 进行蛋白之间相互比对，输出格式为 <code>-outfmt 6</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">makeblastdb -<span class="keyword">in</span> ath.pep.fa -dbtype prot</span><br><span class="line">blastp -num_threads 50 -db ath.pep.fa -query ath.pep.fa -outfmt 6 -evalue 1e-5 -num_alignments 20  -out ath.blastp.txt &amp; </span><br></pre></td></tr></table></figure>

<h2 id="共线性分析"><a href="#共线性分析" class="headerlink" title="共线性分析"></a>共线性分析</h2><h3 id="绘制点阵图"><a href="#绘制点阵图" class="headerlink" title="绘制点阵图"></a>绘制点阵图</h3><p>在上面步骤结束后，其实绘制点阵图就非常简单了，只需要创建配置文件，然后修改配置文件，最后运行wgdi。</p>
<blockquote>
<p>基本wgdi的其他分析也都是三部曲，创建配置，修改配置，运行程序。</p>
</blockquote>
<p>第一步，创建配置文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wgdi -d \? &gt; ath.conf</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>配置文件的信息如下，</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="section">[dotplot]</span></span><br><span class="line"><span class="attr">blast</span> = blast file</span><br><span class="line"><span class="attr">gff1</span> =  gff1 file</span><br><span class="line"><span class="attr">gff2</span> =  gff2 file</span><br><span class="line"><span class="attr">lens1</span> = lens1 file</span><br><span class="line"><span class="attr">lens2</span> = lens2 file</span><br><span class="line"><span class="attr">genome1_name</span> =  Genome1 name</span><br><span class="line"><span class="attr">genome2_name</span> =  Genome2 name</span><br><span class="line"><span class="attr">multiple</span>  = <span class="number">1</span>   <span class="comment"># 最好的同源基因数, 用输出结果中会用红点表示</span></span><br><span class="line"><span class="attr">score</span> = <span class="number">100</span>     <span class="comment"># blast输出的score 过滤 </span></span><br><span class="line"><span class="attr">evalue</span> = <span class="number">1</span>e-<span class="number">5</span>   <span class="comment"># blast输出的evalue 过滤 </span></span><br><span class="line"><span class="attr">repeat_number</span> = <span class="number">20</span>  <span class="comment"># genome2相对于genome1的最多同源基因数</span></span><br><span class="line"><span class="attr">position</span> = order</span><br><span class="line"><span class="attr">blast_reverse</span> = <span class="literal">false</span></span><br><span class="line"><span class="attr">ancestor_left</span> = none</span><br><span class="line"><span class="attr">ancestor_top</span> = none</span><br><span class="line"><span class="attr">markersize</span> = <span class="number">0.5</span>  <span class="comment"># 点的大小</span></span><br><span class="line"><span class="attr">figsize</span> = <span class="number">10</span>,<span class="number">10</span>   <span class="comment"># 图片大小</span></span><br><span class="line"><span class="attr">savefig</span> = savefile(.png,.pdf)</span><br></pre></td></tr></table></figure>

<p>第二步，修改配置文件。因为是自我比对，所以 gff1和gff2内容一样， lens1和lens2内容一样</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">[dotplot]</span><br><span class="line">blast = ath.blastp.txt</span><br><span class="line">gff1 =  ath.gff</span><br><span class="line">gff2 =  ath.gff</span><br><span class="line">lens1 = ath.len</span><br><span class="line">lens2 = ath.len</span><br><span class="line">genome1_name =  A. thaliana</span><br><span class="line">genome2_name =  A. thaliana</span><br><span class="line">multiple  = 1</span><br><span class="line">score = 100</span><br><span class="line">evalue = 1e-5</span><br><span class="line">repeat_number = 5 </span><br><span class="line">position = order</span><br><span class="line">blast_reverse = <span class="literal">false</span></span><br><span class="line">ancestor_left = none</span><br><span class="line">ancestor_top = none</span><br><span class="line">markersize = 0.5</span><br><span class="line">figsize = 10,10</span><br><span class="line">savefig = ath.dot.png</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>最后运行如下命令</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgdi -d ath.conf</span><br></pre></td></tr></table></figure>

<p>输出的png如下所示</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-aca03c4f647d421a9523c285ff8752c6.png" alt="点阵图"></p>
<p>从图中，我们很容易地就能观察到三种颜色，分别是红色，蓝色，和灰色。在WGDI中，红色表示genome2的基因在genome1中的最优同源（相似度最高）的匹配，次好的四个基因是蓝色，而剩余部分是为灰色。图中对角线出现的片段并非是自身比对，因为WGDI已经过滤掉自身比自身的结果。那么问题来了，这些基因是什么呢？这个问题会在后续的分析中进行解答。</p>
<p>如果基因组存在加倍事件，那么对于一个基因组的某个区域可能会存在另一个区域与其有着相似的基因（同源基因），并且这些基因的排布顺序较为一致。反应到点阵图上，就是能够观察到有多个点所组成的”线”。这里的线需要有一个引号，因为当你放大之后，你会发现其实还是点而已。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-799f63357317488d98484b761446b1ea.png" alt="部分区域放大"></p>
<p>由于进一步鉴定共线性区域就建立在这些”点”的基础上，那么影响这些点是否为同源基因的参数就非常重要了，即配置文件中的score,evalue,repeat_number。</p>
<h3 id="共线性分析和Ks计算"><a href="#共线性分析和Ks计算" class="headerlink" title="共线性分析和Ks计算"></a>共线性分析和Ks计算</h3><p>对于同线性(synteny)和共线性(collinearity)，一般认为同线性指的是两个区域有一定数量的同源基因，对基因顺序排布无要求。共线性是同线性的特殊形式，要求同源基因的排布顺序也相似。</p>
<p>wgdi开发了<code>-icl</code>模块（<strong>I</strong>mproved version of <strong>C</strong>olinear<strong>S</strong>can）用于进行共线性分析，使用起来非常简单，也是先建立配置文件。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgdi -icl \? &gt;&gt; ath.conf</span><br></pre></td></tr></table></figure>

<p>然后对配置文件进行修改.</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="section">[collinearity]</span></span><br><span class="line"><span class="attr">gff1</span> = ath.gff</span><br><span class="line"><span class="attr">gff2</span> = ath.gff</span><br><span class="line"><span class="attr">lens1</span> = ath.len</span><br><span class="line"><span class="attr">lens2</span> = ath.len</span><br><span class="line"><span class="attr">blast</span> = ath.blastp.txt</span><br><span class="line"><span class="attr">blast_reverse</span> = <span class="literal">false</span></span><br><span class="line"><span class="attr">multiple</span>  = <span class="number">1</span></span><br><span class="line"><span class="attr">process</span> = <span class="number">8</span></span><br><span class="line"><span class="attr">evalue</span> = <span class="number">1</span>e-<span class="number">5</span></span><br><span class="line"><span class="attr">score</span> = <span class="number">100</span></span><br><span class="line"><span class="attr">grading</span> = <span class="number">50</span>,<span class="number">40</span>,<span class="number">25</span></span><br><span class="line"><span class="attr">mg</span> = <span class="number">40</span>,<span class="number">40</span></span><br><span class="line"><span class="attr">pvalue</span> = <span class="number">0.2</span></span><br><span class="line"><span class="attr">repeat_number</span> = <span class="number">10</span></span><br><span class="line"><span class="attr">positon</span> = order</span><br><span class="line"><span class="attr">savefile</span> = ath.collinearity.txt</span><br></pre></td></tr></table></figure>

<p>其中的evalue, score和BLAST文件的过滤有关，用于筛选同源基因对。multiple确定最佳比对基因，repeat_number表示最多允许多少基因是潜在的同源基因, grading根据同源基因的匹配程度进行打分, mg 指的是共线性区域中所允许最大空缺基因数。</p>
<p>运行结果后，会得到ath.collinearity.txt，记录共线性区域。以 # 起始的行记录共线性区域的元信息，例如得分(score)，显著性(pvalue)，基因对数(N)等。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grep <span class="string">&#x27;^#&#x27;</span> ath.collinearity.txt | <span class="built_in">tail</span> -n 1</span><br><span class="line"><span class="comment"># Alignment 959: score=778.0 pvalue=0.0003 N=16 Pt&amp;Pt minus </span></span><br></pre></td></tr></table></figure>

<p>紧接着，我们就可以根据共线性结果来计算Ks。</p>
<p>同样的也是先生成配置文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgdi -ks \? &gt;&gt; ath.conf</span><br></pre></td></tr></table></figure>

<p>然后修改配置文件。我们需要提供cds和pep文件，共线性分析输出文件(支持MCScanX的共线性分析结果）。WGDI会用muscle根据蛋白序列进行联配，然后使用pal2pal.pl 基于cds序列将蛋白联配转成密码子联配，最后用paml中的yn00计算ka和ks。</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="section">[ks]</span></span><br><span class="line"><span class="attr">cds_file</span> =   ath.cds.fa</span><br><span class="line"><span class="attr">pep_file</span> =   ath.pep.fa</span><br><span class="line"><span class="attr">align_software</span> = muscle</span><br><span class="line"><span class="attr">pairs_file</span> = ath.collinearity.txt</span><br><span class="line"><span class="attr">ks_file</span> = ath.ks</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>ks计算需要一段时间，那么不妨在计算时了解下Ka和Ks。Ka和Ks分别指的是非同义替换位点数，和同义替换位点数。根据中性演化假说，基因的变化大多是中性突变，不会影响生物的生存，那么当一对同源基因分开的时间越早，不影响生存的碱基替换数就越多(Ks)，反之越多。</p>
<p>运行结果后，输出的Ks文件共有6列，对应的是每个基因对的Ka和Ks值。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">head</span> ath.ks</span><br><span class="line">id1  id2  ka_NG86  ks_NG86  ka_YN00  ks_YN00</span><br><span class="line">AT1G72300  AT1G17240  0.1404  0.629  0.1467  0.5718</span><br><span class="line">AT1G72330  AT1G17290  0.0803  0.5442  0.079  0.6386</span><br><span class="line">AT1G72350  AT1G17310  0.3709  0.9228  0.3745  1.071</span><br><span class="line">AT1G72410  AT1G17360  0.2636  0.875  0.2634  1.1732</span><br><span class="line">AT1G72450  AT1G17380  0.2828  1.1068  0.2857  1.5231</span><br><span class="line">AT1G72490  AT1G17400  0.255  1.2113  0.2597  2.0862</span><br><span class="line">AT1G72520  AT1G17420  0.1006  0.9734  0.1025  1.0599</span><br><span class="line">AT1G72620  AT1G17430  0.1284  0.7328  0.1375  0.603</span><br><span class="line">AT1G72630  AT1G17455  0.0643  0.7608  0.0613  1.2295</span><br></pre></td></tr></table></figure>

<p>鉴于共线性和Ks值输出结果信息量太大，不方便使用，更好的方法是将两者进行聚合，得到关于各个共线性区的汇总信息。</p>
<p>WGDI提供 <code>-bi</code> 参数帮助我们进行数据整合。同样也是生成配置文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgdi -bi ? &gt;&gt; ath.conf</span><br></pre></td></tr></table></figure>

<p>修改配置文件. 其中collinearity 和ks是我们前两步的输出文件，而 ks_col 则是声明使用ks文件里的哪一列。</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="section">[blockinfo]</span></span><br><span class="line"><span class="attr">blast</span> = ath.blastp.txt</span><br><span class="line"><span class="attr">gff1</span> =  ath.gff</span><br><span class="line"><span class="attr">gff2</span> =  ath.gff</span><br><span class="line"><span class="attr">lens1</span> = ath.len</span><br><span class="line"><span class="attr">lens2</span> = ath.len</span><br><span class="line"><span class="attr">collinearity</span> = ath.collinearity.txt</span><br><span class="line"><span class="attr">score</span> = <span class="number">100</span></span><br><span class="line"><span class="attr">evalue</span> = <span class="number">1</span>e-<span class="number">5</span></span><br><span class="line"><span class="attr">repeat_number</span> = <span class="number">20</span></span><br><span class="line"><span class="attr">position</span> = order</span><br><span class="line"><span class="attr">ks</span> = ath.ks</span><br><span class="line"><span class="attr">ks_col</span> = ks_NG86</span><br><span class="line"><span class="attr">savefile</span> = ath_block_information.csv</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>运行即可</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wgdi -bi ath.conf</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>输出文件以csv格式进行存放，因此可以用EXCEL直接打开，共11列。</p>
<ol>
<li><code>id</code> 即共线性的结果的唯一标识</li>
<li><code>chr1</code>,<code>start1</code>,<code>end1</code> 即参考基因组（点图的左边）的共线性范围</li>
<li><code>chr2</code>,<code>start2</code>,<code>end2</code> 即参考基因组（点图的上边）的共线性范围</li>
<li><code>pvalue</code> 即共线性结果评估，常常认为小于0.01的更合理些</li>
<li><code>length</code> 即共线性片段长度</li>
<li><code>ks_median</code> 即共线性片段上所有基因对<code>ks</code>的中位数（主要用来评判ks分布的）</li>
<li><code>ks_average</code> 即共线性片段上所有基因对<code>ks</code>的平均值</li>
<li><code>homo1</code>,<code>homo2</code>,<code>homo3</code>,<code>homo4</code>,<code>homo5</code>与<code>multiple</code>参数有关，即一共有homo+multiple列</li>
</ol>
<p>  主要规则是：基因对如果在点图中为红色，赋值为1，蓝色赋值为0，灰色赋值为-1（颜色参考前述 <code>wgdi</code> 点图 相关推文）。共线性片段上所有基因对赋值后求平均值，这样就赋予该共线性一个-1,1的值。如果共线性的点大部分为红色，那么该值接近于1。可以作为共线性片段的筛选。</p>
<ol>
<li><code>block1</code>,<code>block2</code>分别为共线性片段上基因<code>order</code>的位置。</li>
<li><code>ks</code>共线性片段的上基因对的<code>ks</code>值</li>
<li><code>density1</code>,<code>density2</code> 共线性片段的基因分布密集程度。值越小表示稀疏</li>
</ol>
<p>这个表格是后续探索分析的一个重点，比如说我们可以用它来分析我们点图中出现在对角线上的基因。下面的代码就是提取了第一个共线性区域的所有基因，然后进行富集分析，绘制了一个气泡图。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">df <span class="operator">&lt;-</span> read.csv<span class="punctuation">(</span><span class="string">&quot;ath_block_information.csv&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">tandem_length <span class="operator">&lt;-</span> 200</span><br><span class="line"></span><br><span class="line">df<span class="operator">$</span>start <span class="operator">&lt;-</span> df<span class="operator">$</span>start2 <span class="operator">-</span> df<span class="operator">$</span>start1</span><br><span class="line">df<span class="operator">$</span>end <span class="operator">&lt;-</span> df<span class="operator">$</span>end2 <span class="operator">-</span> df<span class="operator">$</span>end1</span><br><span class="line"></span><br><span class="line">df_tandem <span class="operator">&lt;-</span> df<span class="punctuation">[</span><span class="built_in">abs</span><span class="punctuation">(</span>df<span class="operator">$</span>start<span class="punctuation">)</span> <span class="operator">&lt;=</span> tandem_length <span class="operator">|</span></span><br><span class="line">     <span class="built_in">abs</span><span class="punctuation">(</span>df<span class="operator">$</span>end<span class="punctuation">)</span> <span class="operator">&lt;=</span> tandem_length<span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">gff <span class="operator">&lt;-</span> read.table<span class="punctuation">(</span><span class="string">&quot;ath.gff&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">syn_block1 <span class="operator">&lt;-</span>  df_tandem<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;block1&quot;</span><span class="punctuation">,</span> <span class="string">&quot;block2&quot;</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">gene_order <span class="operator">&lt;-</span> unlist<span class="punctuation">(</span></span><br><span class="line">  lapply<span class="punctuation">(</span>syn_block1<span class="punctuation">,</span> strsplit<span class="punctuation">,</span> split<span class="operator">=</span><span class="string">&quot;_&quot;</span><span class="punctuation">,</span> fixed<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">,</span> useBytes<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">gene_order <span class="operator">&lt;-</span> unique<span class="punctuation">(</span><span class="built_in">as.numeric</span><span class="punctuation">(</span>gene_order<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">syn_block1_gene <span class="operator">&lt;-</span> gff<span class="punctuation">[</span> gff<span class="operator">$</span>V6 <span class="operator">%in%</span> gene_order<span class="punctuation">,</span> <span class="string">&quot;V2&quot;</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#BiocManager::install(&quot;org.At.tair.db&quot;)</span></span><br><span class="line">library<span class="punctuation">(</span>org.At.tair.db<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>clusterProfiler<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ego <span class="operator">&lt;-</span> enrichGO<span class="punctuation">(</span>syn_block1_gene<span class="punctuation">,</span> </span><br><span class="line">                OrgDb <span class="operator">=</span> org.At.tair.db<span class="punctuation">,</span> </span><br><span class="line">                keyType <span class="operator">=</span> <span class="string">&quot;TAIR&quot;</span><span class="punctuation">,</span>ont <span class="operator">=</span> <span class="string">&quot;BP&quot;</span></span><br><span class="line">                <span class="punctuation">)</span></span><br><span class="line">dotplot<span class="punctuation">(</span>ego<span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-4a707a066ab745c2b566d55ab28b765d.png" alt="气泡图"></p>
<p>进一步观察这些基因，可以发现这些基因的编号都是前后相连，说明这个基因簇和花粉识别有一定关系。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> as.data.frame<span class="punctuation">(</span>ego<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line">                   ID           Description GeneRatio  BgRatio       pvalue</span><br><span class="line">GO<span class="operator">:</span><span class="number">0048544</span> GO<span class="operator">:</span><span class="number">0048544</span> recognition of pollen    <span class="number">11</span><span class="operator">/</span><span class="number">560</span> <span class="number">43</span><span class="operator">/</span><span class="number">21845</span> <span class="number">7.794138e-09</span></span><br><span class="line">               p.adjust      qvalue</span><br><span class="line">GO<span class="operator">:</span><span class="number">0048544</span> <span class="number">8.121073e-06</span> <span class="number">7.95418e-06</span></span><br><span class="line">                                                                                                                  geneID</span><br><span class="line">GO<span class="operator">:</span><span class="number">0048544</span> AT1G61360<span class="operator">/</span>AT1G61370<span class="operator">/</span>AT1G61380<span class="operator">/</span>AT1G61390<span class="operator">/</span>AT1G61400<span class="operator">/</span>AT1G61420<span class="operator">/</span>AT1G61440<span class="operator">/</span>AT1G61480<span class="operator">/</span>AT1G61490<span class="operator">/</span>AT1G61500<span class="operator">/</span>AT1G61550</span><br><span class="line">           Count</span><br><span class="line">GO<span class="operator">:</span><span class="number">0048544</span>    <span class="number">11</span></span><br></pre></td></tr></table></figure>

<blockquote>
<p>A <strong>gene cluster</strong> is a group of two or more genes found within an organism’s DNA that encode similar polypeptides, or proteins, which collectively share a generalized function and are **often located within a few thousand base pairs **of each other. –<a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Gene_cluster">维基百科</a></p>
</blockquote>
<p>接下来，我们还可以根据这个表格中的Ks对点阵图进行上色， 以及正确的计算Ks峰值。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/02/01/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%B8%80)-comparative-genome-analysis-with-wgdi-1/" data-id="clm4a30sc00hme6nke8k5cjh7" data-title="如何用WGDI进行共线性分析(一)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-「小技巧」如何让IGV更快的加载GTF和GFF注释文件-trick-load-gtf-and-gff-to-igv-more-quickily" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/01/23/%E3%80%8C%E5%B0%8F%E6%8A%80%E5%B7%A7%E3%80%8D%E5%A6%82%E4%BD%95%E8%AE%A9IGV%E6%9B%B4%E5%BF%AB%E7%9A%84%E5%8A%A0%E8%BD%BDGTF%E5%92%8CGFF%E6%B3%A8%E9%87%8A%E6%96%87%E4%BB%B6-trick-load-gtf-and-gff-to-igv-more-quickily/" class="article-date">
  <time class="dt-published" datetime="2021-01-23T19:03:03.850Z" itemprop="datePublished">2021-01-23</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/01/23/%E3%80%8C%E5%B0%8F%E6%8A%80%E5%B7%A7%E3%80%8D%E5%A6%82%E4%BD%95%E8%AE%A9IGV%E6%9B%B4%E5%BF%AB%E7%9A%84%E5%8A%A0%E8%BD%BDGTF%E5%92%8CGFF%E6%B3%A8%E9%87%8A%E6%96%87%E4%BB%B6-trick-load-gtf-and-gff-to-igv-more-quickily/">「小技巧」如何让IGV更快的加载GTF和GFF注释文件</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>很简单，就下面3行命令</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">gff=</span><br><span class="line">(grep ^<span class="string">&quot;#&quot;</span> <span class="variable">$gff</span>; grep -v ^<span class="string">&quot;#&quot;</span> <span class="variable">$gff</span> | <span class="built_in">sort</span> -k1,1 -k4,4n) | bgzip &gt; sorted.gff.gz;</span><br><span class="line">tabix -p gff sorted.gff.gz;</span><br></pre></td></tr></table></figure>

<p>第一行的gff是定义输入文件。第二行是对GFF文件进行排序。第三行是利用HTSLIB中的tabix工具建立索引，得到一个sorted.gff.gz.tbi 索引文件。</p>
<p>后续IGV在加载文件时，会根据索引文件直接读取给定区域的信息，不但可以降低内存的占用，还可以提高加载速度。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/01/23/%E3%80%8C%E5%B0%8F%E6%8A%80%E5%B7%A7%E3%80%8D%E5%A6%82%E4%BD%95%E8%AE%A9IGV%E6%9B%B4%E5%BF%AB%E7%9A%84%E5%8A%A0%E8%BD%BDGTF%E5%92%8CGFF%E6%B3%A8%E9%87%8A%E6%96%87%E4%BB%B6-trick-load-gtf-and-gff-to-igv-more-quickily/" data-id="clm4a30l50059e6nk3i07ct7f" data-title="「小技巧」如何让IGV更快的加载GTF和GFF注释文件" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何给bioconda贡献recipes-how-to-contribute-to-bioconda" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/01/20/%E5%A6%82%E4%BD%95%E7%BB%99bioconda%E8%B4%A1%E7%8C%AErecipes-how-to-contribute-to-bioconda/" class="article-date">
  <time class="dt-published" datetime="2021-01-20T20:11:49.962Z" itemprop="datePublished">2021-01-20</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Python/">Python</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/01/20/%E5%A6%82%E4%BD%95%E7%BB%99bioconda%E8%B4%A1%E7%8C%AErecipes-how-to-contribute-to-bioconda/">如何给bioconda贡献recipes</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>bioconda是conda中专门用来管理生物信息相关软件的频道。</p>
<p>绝大部分的用户都是利用bioconda安装软件，绝大部分的教程也都是教大家如何去使用bioconda安装软件。但是bioconda的便利性并不是从天而降的，而是无数的软件开发者持续的不断的贡献和更新一些软件的recipe (菜谱) 的结果。这篇教程就是教大家如何在一个软件没有bioconda的安装方式时，自己动手写一个菜谱。</p>
<h2 id="准备工作"><a href="#准备工作" class="headerlink" title="准备工作"></a>准备工作</h2><p>首先，我们在<a target="_blank" rel="noopener" href="https://github.com/bioconda/bioconda-recipes">https://github.com/bioconda/bioconda-recipes</a> 上fork该项目到自己的GitHub项目下，然后将其clone到本地</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">git <span class="built_in">clone</span> https://github.com/xuzhougeng/bioconda-recipes.git</span><br><span class="line"><span class="built_in">cd</span> bioconda-recipes</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>如果一不小心克隆使用了官方的项目或者我的项目，也可以通过如下命令进行修改</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">git remote <span class="built_in">rm</span> origin</span><br><span class="line">git remote add origin 你的github地址</span><br></pre></td></tr></table></figure>

<p>由于bioconda是建立在conda基础上的频道，因此我们还需要预先安装好 conda，关于conda 的安装、配置和使用可以看我上传到<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1s4411F761">哔哩哔哩的视频</a>。之后安装 conda-build, 用于创建conda软件包的框架。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 安装conda-build, 后续用于创建模板</span></span><br><span class="line">conda install conda-build</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<h2 id="实际流程"><a href="#实际流程" class="headerlink" title="实际流程"></a>实际流程</h2><p>向bioconda贡献一个软件的菜谱大致分为6步:</p>
<ol>
<li><p>创建分支</p>
</li>
<li><p>编辑recipes</p>
</li>
<li><p>推送修改</p>
</li>
<li><p>创建pull请求</p>
</li>
<li><p>删除你的分支</p>
</li>
<li><p>安装你的包</p>
</li>
</ol>
<h2 id="第一步-创建分支"><a href="#第一步-创建分支" class="headerlink" title="第一步: 创建分支"></a>第一步: 创建分支</h2><p>我们需要先切换到主分支(master), 并保证你的主分支处于最新状态</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">git checkout master</span><br><span class="line">git pull upstream master</span><br><span class="line">git push origin master</span><br></pre></td></tr></table></figure>

<p>接下来，我们创建新的分支，用于处理当前的菜谱，避免影响主分支</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git checkout -b update_My_recipes</span><br></pre></td></tr></table></figure>

<p>我建议这里的 <code>update_My_recipes</code> 修改成具体的软件名，这样更有标识性。</p>
<h2 id="第二步-编辑recipe"><a href="#第二步-编辑recipe" class="headerlink" title="第二步: 编辑recipe"></a>第二步: 编辑recipe</h2><p>我们使用 <code>conda skeleton</code> 创建一个软件包框架， 这里pypi指的是软件托管在pypi上，wgdi则是软件名。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># bioconda-recipes目录下</span></span><br><span class="line"><span class="built_in">cd</span> recipes</span><br><span class="line">conda skeleton pypi wgdi </span><br></pre></td></tr></table></figure>

<p>因为wgdi存在一些依赖环境，如PAML, MAFFT, MUSCLE, PAL2NAL,  所以我们需要修改 <code>wgdi/meta.yaml</code> 文件，在其中加入这些软件。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">run:</span><br><span class="line">  - biopython</span><br><span class="line">  - matplotlib</span><br><span class="line">  - numpy</span><br><span class="line">  - pandas &gt;=1.1.0</span><br><span class="line">  - python </span><br><span class="line">  - scipy</span><br><span class="line">  - paml</span><br><span class="line">  - mafft</span><br><span class="line">  - muscle</span><br><span class="line">  - pal2nal</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>事实上，conda skeleton 只是做了一小部分事情，我目前的meta.yaml 还不完整，直到后续的递交中修改了数十次才知道出问题的地方以及如何修改。</p>
<p>因此建议后还需要查阅 <a target="_blank" rel="noopener" href="https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html">https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html</a> 了解meta.yaml的更多知识。</p>
<h2 id="第三步-向GitHub上推送更新"><a href="#第三步-向GitHub上推送更新" class="headerlink" title="第三步: 向GitHub上推送更新"></a>第三步: 向GitHub上推送更新</h2><p>这里就是用git命令将我们的修改推送到GitHub上</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># bioconda-recipes目录下</span></span><br><span class="line">git add recipes/wgdi</span><br><span class="line">git commit -m <span class="string">&quot;add wgdi recipes&quot;</span></span><br><span class="line">git push origin update_My_recipes</span><br></pre></td></tr></table></figure>

<h2 id="第四步-创建pull-请求"><a href="#第四步-创建pull-请求" class="headerlink" title="第四步: 创建pull 请求"></a>第四步: 创建pull 请求</h2><p>此时，让我们回到GitHub 中我们fork的bioconda-recipes页面。我们可以看到该修改已经被推送到GitHub上。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-784347f793c94193b23914e59dc625fb.png" alt="compare &amp; pull request"></p>
<p>我们点击 Compare &amp; pull request, 接着就会弹出一个页面，介绍如何书写提交申请。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-6b3654a9dff44ff5b82cf828e96e794e.png" alt="Add title"></p>
<p>这里写了4点要求，分别为</p>
<ol>
<li><p>在PR中加入Add或者update 作为用于说明目的，必须是第一个词</p>
</li>
<li><p>如果该项目和生物学领域无关，建议换个地方。</p>
</li>
<li><p>一旦通过了审核，需要在后续的issue中加入 <code>@BiocondaBot please add label</code>, 具体区别在后面的命令说明中有提到。</p>
</li>
<li><p>如果有问题，你可以到Gitter或者在评论中用 <code>@bioconda/core</code> 提出自己的问题。</p>
</li>
</ol>
<p>当你提出PR之后，bioconda 的构建系统就会测试我们提交的菜谱。如果通过了测试，我们就可以跳转到下一步，否则需要重复第二~四步，直到通过为止。</p>
<p>我们发现在提交之后，构建系统就提示了一个问题。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-782631bcdea541fbaedca10e2ace568f.png" alt="lint problem"></p>
<p>点开Details，会出现如下的界面。我们可以点击View more details on CircleCI Checks 去聊了解更加详细的信息。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-490eb0c1569043ee89bb8b87e088b267.png" alt="view problem"></p>
<p>lint问题的修改建议可在bioconda的对应页面为 <a target="_blank" rel="noopener" href="https://bioconda.github.io/contributor/linting.html">https://bioconda.github.io/contributor/linting.html</a> 进行查找。我把自己遇到的问题放在文末，毕竟是第一次，对于自动化测试系统并不了解，所以我反复修改了十多次。</p>
<p>当我们的菜谱最终通过了自动化测试后，接下来就可以在我们的issue中留言 <code>@bioconda-bot add label</code> ，让机器人增加标签，然后等待管理员review同意你的PR，当然也需要你进行修改才行。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-eaa78ee47eff4337b333aa22c49e6cbd.png" alt="add label"></p>
<h2 id="第五步-删除分支"><a href="#第五步-删除分支" class="headerlink" title="第五步: 删除分支"></a>第五步: 删除分支</h2><p>当我们的Pull Request被合并之后，我们就可以在GitHub上删除我们的分支，或者在本地删除</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Delete local branch</span></span><br><span class="line">git branch -D update_My_recipes</span><br><span class="line"><span class="comment"># Delete branch in your fork via the remote named &quot;origin&quot;</span></span><br><span class="line">git push origin -d update_My_recipes</span><br></pre></td></tr></table></figure>

<h2 id="第六步-安装你的包"><a href="#第六步-安装你的包" class="headerlink" title="第六步: 安装你的包"></a>第六步: 安装你的包</h2><p>经过前面五步之后，你需要再等待一会，才能让你的包被加入到bioconda频道中。一旦被加入到bioconda频道，那么所有人都可以非常容易的通过conda来安装你得工具了。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install -c bioconda wgdi</span><br></pre></td></tr></table></figure>

<h2 id="其他"><a href="#其他" class="headerlink" title="其他"></a>其他</h2><p>在我配置 wgdi包的时候，遇到了如下几种错误</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ERROR: recipes/wgdi/meta.yaml:0: uses_matplotlib: The recipe uses `matplotlib`, but `matplotlib-base` is recommended</span><br><span class="line">ERROR: recipes/wgdi/meta.yaml:16: should_be_noarch_generic: The recipe should be build as `noarch`</span><br><span class="line">ERROR: recipes/wgdi/meta.yaml:4: folder_and_package_name_must_match: The recipe folder and package name <span class="keyword">do</span> not match.</span><br></pre></td></tr></table></figure>

<p>根据我的错误，我在原来的yaml文件中做了如下的更改</p>
<ul>
<li><p>uses_matplotlib: 将原来的matplotlib 替换成matplobtlib-base</p>
</li>
<li><p>should_be_noarch_generic:  在build下增加 <code>noarch: generic</code> （这一步有问题，请继续往后阅读找到正确答案，此处保留时为了记录我的思考过程）</p>
</li>
<li><p>folder_and_package_name_must_match: 在配置文件开头增加 <code>&#123;% set name = "wgdi" %&#125;</code></p>
</li>
</ul>
<p>之后在test-linu时，我遇到了 <code>ModuleNotFoundError: No module named &#39;wgdi&#39;</code>报错。这个问题和之前的 should_be_noarch_generic密切相关。noarch指的是平台无关，也就是这个软件不需要额外的编译就可以使用。一般这类软件通常是Java编译后的包，而非Python包。我最初根据提示的 <code>should_be_noarch_generic</code> 增加的 <code>noarch: generic</code> 实际并不正确，实际应该是 <code>noarch: python</code>, 也就是表示我们包依赖的Python平台。</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料:"></a>参考资料:</h2><ul>
<li><p>meta.yaml: <a target="_blank" rel="noopener" href="https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html">https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html</a></p>
</li>
<li><p>lint: <a target="_blank" rel="noopener" href="https://bioconda.github.io/contributor/linting.html">https://bioconda.github.io/contributor/linting.html</a></p>
</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/01/20/%E5%A6%82%E4%BD%95%E7%BB%99bioconda%E8%B4%A1%E7%8C%AErecipes-how-to-contribute-to-bioconda/" data-id="clm4a30so00i9e6nk33fy9aly" data-title="如何给bioconda贡献recipes" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-通过X11转发在服务器上用IGV-using-x11-in-server-by-x11-forwarding" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/01/18/%E9%80%9A%E8%BF%87X11%E8%BD%AC%E5%8F%91%E5%9C%A8%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E7%94%A8IGV-using-x11-in-server-by-x11-forwarding/" class="article-date">
  <time class="dt-published" datetime="2021-01-18T10:07:16.696Z" itemprop="datePublished">2021-01-18</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/01/18/%E9%80%9A%E8%BF%87X11%E8%BD%AC%E5%8F%91%E5%9C%A8%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E7%94%A8IGV-using-x11-in-server-by-x11-forwarding/">通过X11转发在服务器上用IGV</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>通常情况下，我们无法是直接在服务器上打开IGV，因为它既要求服务器上<strong>安装X服务</strong>，本地也要安装<strong>X服务</strong>，才能将服务器上的图形信息转发到本地。</p>
<blockquote>
<p>事实上是要配置好X11转发，就可以在服务器上运行其他的图形界面程序。</p>
</blockquote>
<p>第一步: 在服务器上配置Xserver(需要管理员权限)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sudo yum install  xorg-x11-server-Xorg xorg-x11-xauth xorg-x11-apps -y</span><br></pre></td></tr></table></figure>

<p>第二步: 在windows上配置X服务</p>
<p>我们从 <a target="_blank" rel="noopener" href="https://sourceforge.net/projects/xming/">https://sourceforge.net/projects/xming/</a> 下载Xming</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-bbb12dad77a0414d83c71915da50655c.png" alt="Xming"></p>
<p>安装完成后，右下角有Xming Server的地址信息，一般会是0.0或者1.0</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-a4f1114d2222410e9be8277e971a1923.png" alt="Xnubg Server"></p>
<p>第三步: 设置Xshell或Putty</p>
<p>!配置Xshell](<a target="_blank" rel="noopener" href="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-8dbd12df088e4b688b71f4b72f86ecd2.png">https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-8dbd12df088e4b688b71f4b72f86ecd2.png</a>)</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-9ee5bdc814074cac90183d424e0e6f55.png" alt="配置Putty"></p>
<p>第四步: 下载IGV并解压缩</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wget https://data.broadinstitute.org/igv/projects/downloads/2.8/IGV_Linux_2.8.13_WithJava.zip</span><br><span class="line">unzip IGV_Linux_2.8.13_WithJava.zip </span><br></pre></td></tr></table></figure>

<p>运行该目录下的 <code>igv_hidpi.sh </code> 或 <code>igv.sh</code>即可</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/01/18/%E9%80%9A%E8%BF%87X11%E8%BD%AC%E5%8F%91%E5%9C%A8%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E7%94%A8IGV-using-x11-in-server-by-x11-forwarding/" data-id="clm4a30ul00l7e6nkazow7ymm" data-title="通过X11转发在服务器上用IGV" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第十六章)-shi-yong-a-r-c-h-r-fen-xi-dan-xi-bao-a-t-a-c---s-e-q-shu-ju--di-shi-liu-zhang-" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E5%85%AD%E7%AB%A0)-shi-yong-a-r-c-h-r-fen-xi-dan-xi-bao-a-t-a-c---s-e-q-shu-ju--di-shi-liu-zhang-/" class="article-date">
  <time class="dt-published" datetime="2021-01-17T16:38:08.778Z" itemprop="datePublished">2021-01-17</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E5%85%AD%E7%AB%A0)-shi-yong-a-r-c-h-r-fen-xi-dan-xi-bao-a-t-a-c---s-e-q-shu-ju--di-shi-liu-zhang-/">使用ArchR分析单细胞ATAC-seq数据(第十六章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第16章-使用ArchR进行轨迹分析"><a href="#第16章-使用ArchR进行轨迹分析" class="headerlink" title="第16章 使用ArchR进行轨迹分析"></a>第16章 使用ArchR进行轨迹分析</h1><p>ArchR在<code>ArchRProject</code>中低维子空间对细胞进行排序来创建细胞轨迹，实现伪时间中细胞排序。我们之前已经在二维UMAP子空间中执行了这种排序，但是ArchR改进了这种方法，使其能够在N维子空间（即LSI）内进行对齐。</p>
<ul>
<li>首先，ArchR需要用户提供一个轨迹骨架描述细胞分组&#x2F;聚类的大致分化方向。例如，用户提供的细胞编号为HSC,  GMP, Monocyte，这表示细胞以干细胞作为起点，然后是祖细胞，最后是分化细胞。这种分化方向信息来自于已有的生物学相关的细胞轨迹发育的背景知识。</li>
<li>其次，ArchR计算每个聚类在N维空间的平均坐标，对于聚类里的每个细胞，计算其相对于平均指标的欧几里得距离，最后只保留前5%的细胞。</li>
<li>然后，ArchR根据轨迹计算<strong>cluster i</strong>里的每个细胞和<strong>cluster i+1</strong>平均坐标的距离，根据距离计算拟时间向量。之后依次遍历所有的cluster执行运算。根据每个细胞相对于细胞分组&#x2F;聚类平均坐标的欧几里得距离，ArchR能够为每个细胞确定N维坐标和它的拟时间值从而成为轨迹的一部分。</li>
<li>然后，ArchR使用<code>smooth.spline</code>函数根据拟时间值讲连续的轨迹拟合到N维坐标中。</li>
<li>接着，ArchR根据欧氏距离最近细胞沿着流形依次将所有细胞对齐到轨迹上。</li>
<li>最后，ArchR将对其后的值缩放到100，并将其保存在<code>ArchRProject</code>中用于下游分析。</li>
</ul>
<p>ArchR可以根据Arrow文件里的特征创建拟时间趋势矩阵。例如，ArchR可以分析随着拟时间发生变化的TF deviation, 基因得分，整合基因表达量，从而鉴定随着细胞轨迹动态变化的调节因子或调控元素。</p>
<ul>
<li>首先，ARchR将细胞沿着细胞轨迹进行分组，组数由用户定义（默认是1&#x2F;100）</li>
<li>然后，ArchR使用用户定义平滑滑窗(默认是9&#x2F;100)对矩阵的特征进行平滑处理(调用<code>data.table::frollmean</code>函数)</li>
<li>最后，ArchR会返回一个<code>SummarizedExperiment</code>对象，它是一个平滑后的拟时间 X 特征矩阵，用于下游分析。</li>
</ul>
<p>ArchR还可以根据以下信息对任意两个平滑后拟时间 X 特征矩阵进行关联分析，例如匹配的命名(如chromVAR TF deviations的正向调控因子和基因得分&#x2F;整合谱)，之前章节提到的利用低重叠细胞聚集的基因组位置重叠方法(例如 peak-to-gene 关联)。最终，ArchR能为细胞轨迹的整合分析提供帮助，解释多组学之间动态调控关系。</p>
<h2 id="16-1-髓系轨迹-单核细胞分化"><a href="#16-1-髓系轨迹-单核细胞分化" class="headerlink" title="16.1 髓系轨迹-单核细胞分化"></a>16.1 髓系轨迹-单核细胞分化</h2><p>这一节，我们会创建一个细胞估计用以模拟HSC发育成分化后单核细胞的过程。 在开始之前，我们先来检查保存在<code>cellColData</code>中的聚类和之前定义的细胞类型，分别是”Clusters”和”Clusters2”。我们通过UMAP来进行展示，从中找到我们感兴趣的细胞类型。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Clusters&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">)</span></span><br><span class="line">p2 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">)</span></span><br><span class="line">ggAlignPlots<span class="punctuation">(</span>p1<span class="punctuation">,</span> p2<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;h&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-ba88f4f2bb9341249d70b11256b8b5e3.png" alt="Plot-UMAP-Clusters12-Combined"></p>
<h3 id="16-1-1-拟时间UMAP和特征单独作图"><a href="#16-1-1-拟时间UMAP和特征单独作图" class="headerlink" title="16.1.1 拟时间UMAP和特征单独作图"></a>16.1.1 拟时间UMAP和特征单独作图</h3><p>我们接下来会用到”Clusters2”里面定义的细胞类型。和之前提到的那样，我们会以干细胞(“Progenitor”)-骨髓祖细胞(“GMP”)-单核细胞(“Mono”)这个分化过程创建轨迹。</p>
<p>第一步，我们需要以用一个向量记录轨迹骨干，里面按顺序记录发育各过程中细胞分组的标签。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">trajectory <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;Progenitor&quot;</span><span class="punctuation">,</span> <span class="string">&quot;GMP&quot;</span><span class="punctuation">,</span> <span class="string">&quot;Mono&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>第二步，我们使用<code>addTrajectory()</code>创建轨迹，并加入到<code>ArchRProject</code>中。我们将其命名为”MyeloidU”。该函数的作用就是在<code>cellColData</code>中新建一列，命名为”MyeloidU”，然后记录每个细胞在轨迹中拟时间值。不在轨迹中的细胞记做<code>NA</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addTrajectory<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;MyeloidU&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span></span><br><span class="line">    trajectory <span class="operator">=</span> trajectory<span class="punctuation">,</span> </span><br><span class="line">    embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    force <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们可以检查下该信息，你会发现每个细胞都有一个唯一值，值的取值范围是0-100。我们将<code>NA</code>的细胞过滤掉，因为它们不属于轨迹。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">head<span class="punctuation">(</span>projHeme5<span class="operator">$</span>MyeloidU<span class="punctuation">[</span><span class="operator">!</span><span class="built_in">is.na</span><span class="punctuation">(</span>projHeme5<span class="operator">$</span>MyeloidU<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># [1] 46.07479 53.75027 44.82834 43.18828 47.49617 43.21015</span></span><br></pre></td></tr></table></figure>

<p>我们使用<code>plotTrajectory()</code>函数绘制轨迹，它会用UMAP进行可视化，用拟时间值进行上色，箭头标识轨迹的发育方向。非轨迹的细胞会以灰色标识。在这个例子中，我们使用<code>colorBy = &quot;cellColData&quot;</code>和<code>name=&quot;MyeloidU&quot;</code>让ArchR使用<code>cellColData</code>里”MyeloidU”作为拟时间轨迹。参数中<code>trajectory</code>和<code>name</code>都是”MyeloidU”，可能不太容易理解，ArchR根据<code>trajectory</code>来提取细胞，根据<code>name</code>来对细胞进行着色。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">p <span class="operator">&lt;-</span> plotTrajectory<span class="punctuation">(</span>projHeme5<span class="punctuation">,</span> trajectory <span class="operator">=</span> <span class="string">&quot;MyeloidU&quot;</span><span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;MyeloidU&quot;</span><span class="punctuation">)</span></span><br><span class="line">p<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-fc24879b4b72453cb6355202ad40328e.png" alt="Plot-MyeloidU-Traj-UMAP"></p>
<p>使用<code>plotPDF()</code>保存为可编辑的矢量图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-MyeloidU-Traj-UMAP.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们还可以在UMAP嵌入图中展示轨迹和其他特征。这可以用来展示和我们轨迹有关的特征。</p>
<p>如果你还没有在<code>projHeme5</code>项目中加入推测权重，请先运行下面的代码</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addImputeWeights<span class="punctuation">(</span>projHeme5<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>然后我们绘制”MyeloidU”轨迹，只不过颜色来自于”CEBPB”的基因得分，这是一个已知的调控单核细胞功能的基因，会随着发育而活跃。我们通过<code>colorBy</code>参数来指定表达量矩阵，使用<code>name</code>来指定特征</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotTrajectory<span class="punctuation">(</span>projHeme5<span class="punctuation">,</span> trajectory <span class="operator">=</span> <span class="string">&quot;MyeloidU&quot;</span><span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;CEBPB&quot;</span><span class="punctuation">,</span> continuousSet <span class="operator">=</span> <span class="string">&quot;horizonExtra&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>重复上面的代码，只不过这次使用的是<code>GeneIntegrationMatrix</code>作为基因表达量矩阵</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p2 <span class="operator">&lt;-</span> plotTrajectory<span class="punctuation">(</span>projHeme5<span class="punctuation">,</span> trajectory <span class="operator">=</span> <span class="string">&quot;MyeloidU&quot;</span><span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;CEBPB&quot;</span><span class="punctuation">,</span> continuousSet <span class="operator">=</span> <span class="string">&quot;blueYellow&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><code>plotTrajectory</code>函数实际上返回了一系列图。列表中第一个图是UMAP嵌入图，根据函数调用进行上色。</p>
<p>通过比较基因得分和基因表达量的UMAP图，不难发现CEBPB基因在单核细胞中特异性表达，出现在你时序轨迹的后半部分。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ggAlignPlots<span class="punctuation">(</span>p1<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">,</span> p2<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;h&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-89ce542676424e0ca9258892bda772e9.png" alt="Plot-UMAP-CEBPB-Combined"></p>
<p><code>plotTrajectory()</code>函数返回的是一个点图，x轴是拟时间，y轴是对应特征的值，此处是CEBPB的基因得分或基因表达量。下图中的细胞以拟时间进行着色</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ggAlignPlots<span class="punctuation">(</span>p1<span class="punctuation">[[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">,</span> p2<span class="punctuation">[[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;h&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-09551c7058fa46b0b2e83796cca94abe.png" alt="Plot-UMAP-CEBPB-Combined2"></p>
<h3 id="16-1-2-拟时间热图"><a href="#16-1-2-拟时间热图" class="headerlink" title="16.1.2 拟时间热图"></a>16.1.2 拟时间热图</h3><p>我们可以用热图的形式对特征随着拟时间变化进行可视化。首先我们用<code>getTrajectory()</code>函数从<code>ArchRProject</code>中提取感兴趣的轨迹，它会返回一个<code>SummarizedExperiment</code>对象。通过设置<code>useMatrix</code>参数，我们可以用motif，基因得分，基因表达量，peak开放性来创建拟时间热图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">trajMM  <span class="operator">&lt;-</span> getTrajectory<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;MyeloidU&quot;</span><span class="punctuation">,</span> useMatrix <span class="operator">=</span> <span class="string">&quot;MotifMatrix&quot;</span><span class="punctuation">,</span> log2Norm <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>接着将<code>SummarizedExperiment</code>传给<code>plotTrajectoryHeatmap()</code>函数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotTrajectoryHeatmap<span class="punctuation">(</span>trajMM<span class="punctuation">,</span> pal <span class="operator">=</span> paletteContinuous<span class="punctuation">(</span>set <span class="operator">=</span> <span class="string">&quot;solarExtra&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-67f55bc9c4e04c05b0d2135da247f489.png" alt="Plot-MyeloidU-Traj-Heatmaps"></p>
<p>同样的分析可以使用基因得分绘制拟时间热图，只需要设置<code>useMatrix = &quot;GeneScoreMatrix&quot;</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">trajGSM <span class="operator">&lt;-</span> getTrajectory<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;MyeloidU&quot;</span><span class="punctuation">,</span> useMatrix <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span> log2Norm <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">p2 <span class="operator">&lt;-</span> trajectoryHeatmap<span class="punctuation">(</span>trajGSM<span class="punctuation">,</span>  pal <span class="operator">=</span> paletteContinuous<span class="punctuation">(</span>set <span class="operator">=</span> <span class="string">&quot;horizonExtra&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-63aae9631cb64d6e99b9bd07ace45e24.png" alt="Plot-MyeloidU-Traj-Heatmaps_2"></p>
<p>同样的，还可以绘制基因表达量的拟时间热图，设置<code>useMatrix = &quot;GeneIntegrationMatrix&quot;</code>.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">trajGIM <span class="operator">&lt;-</span> getTrajectory<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;MyeloidU&quot;</span><span class="punctuation">,</span> useMatrix <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span> log2Norm <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">p3 <span class="operator">&lt;-</span> plotTrajectoryHeatmap<span class="punctuation">(</span>trajGIM<span class="punctuation">,</span>  pal <span class="operator">=</span> paletteContinuous<span class="punctuation">(</span>set <span class="operator">=</span> <span class="string">&quot;blueYellow&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-48ae7a3c79624cb0be3a8f0333ad2284.png" alt="Plot-MyeloidU-Traj-Heatmaps_3"></p>
<p>最后，我们可以通过设置<code>useMatrix = &quot;PeakMatrix&quot;</code>绘制peak开放性的拟时间热图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">trajPM  <span class="operator">&lt;-</span> getTrajectory<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;MyeloidU&quot;</span><span class="punctuation">,</span> useMatrix <span class="operator">=</span> <span class="string">&quot;PeakMatrix&quot;</span><span class="punctuation">,</span> log2Norm <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">p4 <span class="operator">&lt;-</span> plotTrajectoryHeatmap<span class="punctuation">(</span>trajPM<span class="punctuation">,</span> pal <span class="operator">=</span> paletteContinuous<span class="punctuation">(</span>set <span class="operator">=</span> <span class="string">&quot;solarExtra&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-9cf068582c9e442b9438c9fcdcf2629e.png" alt="Plot-MyeloidU-Traj-Heatmaps_4"></p>
<p>使用<code>plotPDF()</code>保存可编辑的矢量图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span> p2<span class="punctuation">,</span> p3<span class="punctuation">,</span> p4<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-MyeloidU-Traj-Heatmaps.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">6</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">8</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="16-1-3-整合拟时间分析"><a href="#16-1-3-整合拟时间分析" class="headerlink" title="16.1.3 整合拟时间分析"></a>16.1.3 整合拟时间分析</h3><p>我们也可以进行整合分析，例如整合基因得分&#x2F;基因表达量和motif开发状态，鉴定随拟时间变化的正向TF调控因子。这是一个非常有效的手段，例如鉴定分化相关的驱动因子。我们可以用<code>correlateTrajectories()</code>函数进行该分析，它接受两个<code>SummarizedExperiment</code>对象，该对象可以用<code>getTrajectories()</code>函数获取。</p>
<p>首先，让我们找到开放状态随拟时间和TF基因的基因得分相关的motif</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">corGSM_MM <span class="operator">&lt;-</span> correlateTrajectories<span class="punctuation">(</span>trajGSM<span class="punctuation">,</span> trajMM<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><code>correlateTrajectories()</code>函数的主要输出是一个列表，第一个是<code>DataFrame</code>对象。<code>DataFrame</code>的列名分别是<code>idx1</code>, <code>matchname1</code>, <code>name1</code>和<code>VarAssay1</code>, 对应索引，匹配名，原名，传递给<code>correlateTrajectories()</code>函数的第一个轨迹的特征的方差分位数(variance quantile)。方差分位数是给定特征的标准化值，能用于比较不同来源assay里的相关值。该<code>DataFrame</code>包括所有符合<code>correlateTrajectories()</code>函数里阈值的特征。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">corGSM_MM<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment">## DataFrame with 36 rows and 12 columns</span></span><br><span class="line"><span class="comment">## idx1 idx2 matchname1 matchname2 name1 name2</span></span><br><span class="line"><span class="comment">##</span></span><br><span class="line"><span class="comment">## 1 82 1081 PRDM16 PRDM16 chr1:PRDM16 z:PRDM16_211</span></span><br><span class="line"><span class="comment">## … … … … … … …</span></span><br><span class="line"><span class="comment">## 36 22097 1565 RXRA RXRA chr9:RXRA z:RXRA_695</span></span><br><span class="line"><span class="comment">## Correlation VarAssay1 VarAssay2 TStat</span></span><br><span class="line"><span class="comment">##</span></span><br><span class="line"><span class="comment">## 1 0.859588836579798 0.999783802481948 0.860344827586207 16.6530781620713</span></span><br><span class="line"><span class="comment">## … … … … …</span></span><br><span class="line"><span class="comment">## 36 0.59509316489084 0.88640982401522 0.890229885057471 7.33039570155334</span></span><br><span class="line"><span class="comment">## Pval FDR</span></span><br><span class="line"><span class="comment">##</span></span><br><span class="line"><span class="comment">## 1 2.47466579579935e-30 4.04855324192774e-27</span></span><br><span class="line"><span class="comment">## … … …</span></span><br><span class="line"><span class="comment">## 36 6.60626612679289e-11 1.25672690505037e-09</span></span><br></pre></td></tr></table></figure>

<p>我们然后从<code>SummarizedExperiment</code>提取相应的轨迹，</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">trajGSM2 <span class="operator">&lt;-</span> trajGSM<span class="punctuation">[</span>corGSM_MM<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="operator">$</span>name1<span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">trajMM2 <span class="operator">&lt;-</span> trajMM<span class="punctuation">[</span>corGSM_MM<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="operator">$</span>name2<span class="punctuation">,</span> <span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<p>为了更好的对这些特征排序，我们创建新的轨迹，它的值来自于原先的轨迹的相乘。这可以让我们的热图根据行进行排序</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">trajCombined <span class="operator">&lt;-</span> trajGSM2</span><br><span class="line">assay<span class="punctuation">(</span>trajCombined<span class="punctuation">)</span> <span class="operator">&lt;-</span> t<span class="punctuation">(</span>apply<span class="punctuation">(</span>assay<span class="punctuation">(</span>trajGSM2<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> scale<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span> t<span class="punctuation">(</span>apply<span class="punctuation">(</span>assay<span class="punctuation">(</span>trajMM2<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> scale<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们可以从<code>plotTrajectoryHeatmap()</code>函数返回的结果提取最优的行序。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">combinedMat <span class="operator">&lt;-</span> plotTrajectoryHeatmap<span class="punctuation">(</span>trajCombined<span class="punctuation">,</span> returnMat <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> varCutOff <span class="operator">=</span> <span class="number">0</span><span class="punctuation">)</span></span><br><span class="line">rowOrder <span class="operator">&lt;-</span> match<span class="punctuation">(</span>rownames<span class="punctuation">(</span>combinedMat<span class="punctuation">)</span><span class="punctuation">,</span> rownames<span class="punctuation">(</span>trajGSM2<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>有了这个之后，我们就可以创建配对热图。</p>
<p>首先，我们根据基因得分轨迹创建热图。在<code>rowOrder</code>参数中设置给定的行序</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ht1 <span class="operator">&lt;-</span> plotTrajectoryHeatmap<span class="punctuation">(</span>trajGSM2<span class="punctuation">,</span>  pal <span class="operator">=</span> paletteContinuous<span class="punctuation">(</span>set <span class="operator">=</span> <span class="string">&quot;horizonExtra&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>  varCutOff <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> rowOrder <span class="operator">=</span> rowOrder<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>然后，我们根据motif轨迹创建热图。在<code>rowOrder</code>参数中设置给定的行序</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ht2 <span class="operator">&lt;-</span> plotTrajectoryHeatmap<span class="punctuation">(</span>trajMM2<span class="punctuation">,</span> pal <span class="operator">=</span> paletteContinuous<span class="punctuation">(</span>set <span class="operator">=</span> <span class="string">&quot;solarExtra&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> varCutOff <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> rowOrder <span class="operator">=</span> rowOrder<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>通过将两个热图并排排列，我们可以看到两个热图的行是匹配的。你可能会发现该分析同时捕获了GATA3和GATA3-AS1（一个GATA3的反义转录本）。这是由于特征匹配所导致的，我们需要在后续分析时进行手动过滤。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ht1 <span class="operator">+</span> ht2</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-8f5b5d4dcc5340519d53f39cfcff940f.png" alt="Plot-MyeloidU-GS-TF"></p>
<p>我们可以重复以上的分析，但这次使用<code>GeneIntegrationMatrix</code>而非基因得分。因为这是相同的分析流程，因此代码就不再过多解释</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">corGIM_MM <span class="operator">&lt;-</span> correlateTrajectories<span class="punctuation">(</span>trajGIM<span class="punctuation">,</span> trajMM<span class="punctuation">)</span></span><br><span class="line">trajGIM2 <span class="operator">&lt;-</span> trajGIM<span class="punctuation">[</span>corGIM_MM<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="operator">$</span>name1<span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">trajMM2 <span class="operator">&lt;-</span> trajMM<span class="punctuation">[</span>corGIM_MM<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="operator">$</span>name2<span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">trajCombined <span class="operator">&lt;-</span> trajGIM2</span><br><span class="line">assay<span class="punctuation">(</span>trajCombined<span class="punctuation">)</span> <span class="operator">&lt;-</span> t<span class="punctuation">(</span>apply<span class="punctuation">(</span>assay<span class="punctuation">(</span>trajGIM2<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> scale<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span> t<span class="punctuation">(</span>apply<span class="punctuation">(</span>assay<span class="punctuation">(</span>trajMM2<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> scale<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">combinedMat <span class="operator">&lt;-</span> plotTrajectoryHeatmap<span class="punctuation">(</span>trajCombined<span class="punctuation">,</span> returnMat <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> varCutOff <span class="operator">=</span> <span class="number">0</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">rowOrder <span class="operator">&lt;-</span> match<span class="punctuation">(</span>rownames<span class="punctuation">(</span>combinedMat<span class="punctuation">)</span><span class="punctuation">,</span> rownames<span class="punctuation">(</span>trajGIM2<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ht1 <span class="operator">&lt;-</span> plotTrajectoryHeatmap<span class="punctuation">(</span>trajGIM2<span class="punctuation">,</span>  pal <span class="operator">=</span> paletteContinuous<span class="punctuation">(</span>set <span class="operator">=</span> <span class="string">&quot;blueYellow&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>  varCutOff <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> rowOrder <span class="operator">=</span> rowOrder<span class="punctuation">)</span></span><br><span class="line">ht2 <span class="operator">&lt;-</span> plotTrajectoryHeatmap<span class="punctuation">(</span>trajMM2<span class="punctuation">,</span> pal <span class="operator">=</span> paletteContinuous<span class="punctuation">(</span>set <span class="operator">=</span> <span class="string">&quot;solarExtra&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> varCutOff <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> rowOrder <span class="operator">=</span> rowOrder<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ht1 <span class="operator">+</span> ht2</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-26c58ea6eb7e45dd9665147a35630d88.png" alt="Plot-MyeloidU-GEx-T"></p>
<h2 id="16-2-淋巴轨迹-B细胞分化"><a href="#16-2-淋巴轨迹-B细胞分化" class="headerlink" title="16.2 淋巴轨迹-B细胞分化"></a>16.2 淋巴轨迹-B细胞分化</h2><p>第二个例子会根据祖细胞-淋巴祖细胞-前B细胞-分化后B细胞来构建B细胞的分化轨迹。由于该分析从本质上是重复之前单核细胞轨迹分析，因此这里只介绍关键性的代码不同。</p>
<p>第一处是<code>trajectory</code>变量的赋值。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">trajectory <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;Progenitor&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CLP&quot;</span><span class="punctuation">,</span> <span class="string">&quot;PreB&quot;</span><span class="punctuation">,</span> <span class="string">&quot;B&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>第二处是使用<code>addTrajectory()</code>创建轨迹，参数<code>name=LymphoidU</code>。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addTrajectory<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;LymphoidU&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span></span><br><span class="line">    trajectory <span class="operator">=</span> trajectory<span class="punctuation">,</span> </span><br><span class="line">    embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    force <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>第三处, 我们需要注意<code>plotTrajectory</code>里的参数<code>trajectory</code>和<code>name</code>需要改成”LymphoidU”</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p <span class="operator">&lt;-</span> plotTrajectory<span class="punctuation">(</span>projHeme5<span class="punctuation">,</span> trajectory <span class="operator">=</span> <span class="string">&quot;LymphoidU&quot;</span><span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;LymphoidU&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>后续的代码基本上可以照搬上一节。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E5%85%AD%E7%AB%A0)-shi-yong-a-r-c-h-r-fen-xi-dan-xi-bao-a-t-a-c---s-e-q-shu-ju--di-shi-liu-zhang-/" data-id="clm4a30pz00e5e6nk2i2agri4" data-title="使用ArchR分析单细胞ATAC-seq数据(第十六章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第十五章)-analysis-sc-atac-seq-with-archr-chapter15" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%BA%94%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter15/" class="article-date">
  <time class="dt-published" datetime="2021-01-17T16:34:56.985Z" itemprop="datePublished">2021-01-17</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%BA%94%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter15/">使用ArchR分析单细胞ATAC-seq数据(第十五章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第15章-使用ArchR进行整合分析"><a href="#第15章-使用ArchR进行整合分析" class="headerlink" title="第15章 使用ArchR进行整合分析"></a>第15章 使用ArchR进行整合分析</h1><p>ArchR的一个优势能够整合多种水平的信息从而提供新的洞见。我们可以只用ATAC-seq数据进行分析，如识别peak之间的共开放性来预测调控相互作用，或整合scRNA-seq数据，如通过peak-基因的连锁分析预测增性子活性。无论是哪种情况，ArchR都可以很容易地从scATAC seq数据中获得更深入的见解。</p>
<h2 id="15-1-创建细胞低重叠聚集"><a href="#15-1-创建细胞低重叠聚集" class="headerlink" title="15.1 创建细胞低重叠聚集"></a>15.1 创建细胞低重叠聚集</h2><p>ArchR能方便许多特征间相关性的分析。在这些相关分析中，使用稀疏的单细胞数据进行这些计算会导致大量的噪声。为规避这一挑战，我们采用了一种由<a target="_blank" rel="noopener" href="https://cole-trapnell-lab.github.io/cicero-release/">Cicero</a>引入的方法，在这些分析之前创建单细胞的低重叠聚集。我们过滤与任何其他聚集重叠超过80%的聚集以减少偏差。同时为提升该方法的速度，我们开发了一个优化的迭代重叠检查流程，借由”Rcpp”包通过<code>C++</code>实现了快速特征相关性运算。在ArchR中，这些优化方法被用于计算peak共开放性、peak-基因连锁和其他连锁分析。这些低重叠聚集运算都是在内部完成，为清楚起见，我们先在这里对其介绍。</p>
<h2 id="15-2-ArchR共开放分析"><a href="#15-2-ArchR共开放分析" class="headerlink" title="15.2 ArchR共开放分析"></a>15.2 ArchR共开放分析</h2><p>共开放是多个单细胞种的两个peak之间开放性的相关性。换句话说，当peak A在某个单细胞中是开放状态时，peak B通常也是开放状态。下图可以直观地说明了这一概念，说明增强子E3通常与启动子P是共开放的。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-0631fdf68e6a4ce19cf81b93afaf2cb9.png" alt="ArchR_Coaccessibility"></p>
<p>有一点需要注意，共开放分析找到的peak通常都是细胞类型特异的peak。这是因为这些peak在一种细胞类型中都是开放的，而另一种细胞类型通常都是关闭的。虽然这些peak之间有很强的相关性，但是不意味着这些peak之间存在调控关系。</p>
<p>在ArchR中，我们使用<code>addCoAccessibility()</code>函数计算共开放性，计算结束得到的开放性信息保存在<code>ArchRProject</code>中</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addCoAccessibility<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>使用<code>getCoAccessibility()</code>函数可以从<code>ArchRProject</code>对象中提取共开放性信息，当设置<code>returnLoops=FALSE</code>时会返回一个<code>DataFrame</code>对象。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cA <span class="operator">&lt;-</span> getCoAccessibility<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">FALSE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>该<code>DataFrame</code>对象包含几个重要的信息。<code>queryHits</code>和<code>subjectHits</code>列记录的是两个存在相关性的peak的索引。<code>correlation</code>则是两个peak之间的开放状态相关的数值。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">cA</span><br><span class="line"><span class="comment"># DataFrame with 64824 rows and 4 columns</span></span><br><span class="line"><span class="comment"># queryHits subjectHits seqnames correlation</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 5 10 chr1 0.63855416236086</span></span><br><span class="line"><span class="comment"># … … … … …</span></span><br><span class="line"><span class="comment"># 64823 143977 143978 chrX 0.550319862774772</span></span><br></pre></td></tr></table></figure>

<p>这个共开放<code>DataFrame</code>还有一个元数据成员，包含一个相关peak的<code>GRanges</code>对象。上面提到的<code>queryHits</code>和<code>subjectHits</code>的索引适用于这个<code>GRanges</code>对象。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">metadata<span class="punctuation">(</span>cA<span class="punctuation">)</span><span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 144009 ranges and 0 metadata columns:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># Mono chr1 752499-752999 *</span></span><br><span class="line"><span class="comment"># NK chr1 762651-763151 *</span></span><br><span class="line"><span class="comment"># B chr1 801006-801506 *</span></span><br><span class="line"><span class="comment"># B chr1 805039-805539 *</span></span><br><span class="line"><span class="comment"># CLP chr1 845325-845825 *</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># Erythroid chrX 154664540-154665040 *</span></span><br><span class="line"><span class="comment"># NK chrX 154807324-154807824 *</span></span><br><span class="line"><span class="comment"># PreB chrX 154840785-154841285 *</span></span><br><span class="line"><span class="comment"># PreB chrX 154842404-154842904 *</span></span><br><span class="line"><span class="comment"># NK chrX 154862017-154862517 *</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>如果我们设置<code>retureLoops=TRUE</code>, 那么<code>getCoAccessibility()</code>会以loop track的形式返回共开放数据。在这个<code>GRanges</code>对象中，<code>IRanges</code>的起始和结束对应的是每个存在互作关系中两个共开放peak的位置。<code>resolution</code>参数设置这些loop的碱基对分辨率。当<code>resolution=1</code>时，它会输出连接每个peak中心的loop.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cA <span class="operator">&lt;-</span> getCoAccessibility<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们比较下<code>GRanges</code>和<code>DataFrame</code>这两个对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">cA<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 32412 ranges and 1 metadata column:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># [1] chr1 845575-856640 * | 0.63855416236086</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [32412] chrX 153980218-153990364 * | 0.550319862774772</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>如果我们将loops的分辨率调整为<code>resolution = 1000</code>, 这能够避免绘制过多的共开放互作事件。从下面的输出的<code>GRanges</code>对象中，我们可以看到条目少了很多</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">cA <span class="operator">&lt;-</span> getCoAccessibility<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1000</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cA<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 30997 ranges and 1 metadata column:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># |</span></span><br><span class="line"><span class="comment"># [1] chr1 845500-856500 * | 0.63855416236086</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [30997] chrX 153980500-153990500 * | 0.550319862774772</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>同样的，如果我们进一步降低分辨率<code>resolution = 10000</code>，我们会得到更少共开放交互事件。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">cA <span class="operator">&lt;-</span> getCoAccessibility<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">10000</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cA<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 21142 ranges and 1 metadata column:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># |</span></span><br><span class="line"><span class="comment"># [1] chr1 845000-855000 * | 0.63855416236086</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [21142] chrX 153985000-153995000 * | 0.550319862774772</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<h3 id="15-2-1-在browser-track中绘制共开放"><a href="#15-2-1-在browser-track中绘制共开放" class="headerlink" title="15.2.1 在browser track中绘制共开放"></a>15.2.1 在browser track中绘制共开放</h3><p>当我们在<code>ArchRProject</code>里增加共开放信息后，我们可以用这个信息在基因组浏览器里绘制loop track。具体做法就是在<code>plotBrowserTrack()</code>函数中设置<code>loops</code>参数。我们这里以<code>getCoAccessibility()</code>默认参数输出结果为例，即<code>corCutOff = 0.5</code>, <code>resolution = 1000</code>, <code>returnLoops = TRUE</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;CD34&quot;</span><span class="punctuation">,</span> <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span><span class="punctuation">,</span> <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MS4A1&quot;</span><span class="punctuation">,</span> <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span><span class="punctuation">,</span> <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD8A&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TBX21&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IL7R&quot;</span> <span class="comment">#TCells</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p <span class="operator">&lt;-</span> plotBrowserTrack<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    geneSymbol <span class="operator">=</span> markerGenes<span class="punctuation">,</span> </span><br><span class="line">    upstream <span class="operator">=</span> <span class="number">50000</span><span class="punctuation">,</span></span><br><span class="line">    downstream <span class="operator">=</span> <span class="number">50000</span><span class="punctuation">,</span></span><br><span class="line">    loops <span class="operator">=</span> getCoAccessibility<span class="punctuation">(</span>projHeme5<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>最后用<code>grid.draw</code>函数绘制结果，通过<code>$</code>选择特定的标记基因。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grid<span class="operator">::</span>grid.newpage<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">grid<span class="operator">::</span>grid.draw<span class="punctuation">(</span>p<span class="operator">$</span>CD14<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-02a7ffc28a7d4065b4dd774a14279192.png" alt="Plot-Tracks-Marker-Genes-with-CoAccessibility"></p>
<p>使用<code>plotPDF()</code>保存可编辑的矢量图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>plotList <span class="operator">=</span> p<span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;Plot-Tracks-Marker-Genes-with-CoAccessibility.pdf&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="15-3-ArchR-Peak2GeneLinkages分析"><a href="#15-3-ArchR-Peak2GeneLinkages分析" class="headerlink" title="15.3 ArchR Peak2GeneLinkages分析"></a>15.3 ArchR Peak2GeneLinkages分析</h2><p>和共开放分析类似，ArchR也能分析所谓的”peak-to-gene”关联, 也就是分析peak和基因的相关性。共开放分析和peak-to-gene关联分析的主要不同在于，共开放分析只需要用到ATAC-seq数据，寻找的是peak之间的共开放关系，而peak-to-gene关联分析则会整合scRNA-seq数据，寻找peak开放状态和基因表达量的关系。这其实是相似问题的两种方法。只不过因为peak-to-gene关联分析使用的是scATAC-seq和scRNA-seq数据，我们会认为这个关联更能反应基因调控关系。</p>
<p>在ArchR中，我们使用<code>addPeak2GeneLinks()</code>函数鉴定peak-to-gene的关联。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addPeak2GeneLinks<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们以上一节类似方式提取这些peak-to-gene的关联信息，只不过这里用的是<code>getPeak2GeneLinks()</code>函数。和之前一样，用户可以设置相关性阈值和关联之间的分辨率。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">p2g <span class="operator">&lt;-</span> getPeak2GeneLinks<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.45</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">FALSE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>当<code>returnLoops=FALSE</code>, 该函数返回一个<code>DataFrame</code>对象，和之前<code>getCoAccessibility()</code>返回的<code>DataFrame</code>相似。主要的不同在于scATAC-seq的索引 peak存放在<code>idxATAC</code>列，而scRNA-seq基因的索引是存放在<code>idxRNA</code>列中</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">p2g</span><br><span class="line"><span class="comment"># DataFrame with 43754 rows and 6 columns</span></span><br><span class="line"><span class="comment"># idxATAC idxRNA Correlation FDR</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 47 5 0.549552663393716 1.34094093110629e-38</span></span><br><span class="line"><span class="comment"># 2 3 6 0.487418258348982 2.1460798658766e-29</span></span><br><span class="line">...</span><br><span class="line"><span class="comment"># VarQATAC VarQRNA</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 0.948753202924817 0.793290683296597</span></span><br><span class="line"><span class="comment"># 2 0.253206396822421 0.445890005913661</span></span><br></pre></td></tr></table></figure>

<p>peak-to-gene关联<code>DataFrame</code>对象同样也有一个<code>GRanges</code>对象记录对应peak的元信息。上面提到的<code>idxATAC</code>索引能用于<code>GRanges</code>对象中。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">metadata<span class="punctuation">(</span>p2g<span class="punctuation">)</span><span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 144009 ranges and 0 metadata columns: ## seqnames ranges strand</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># [1] chr1 752499-752999 *</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># [144009] chrX 154862017-154862517 *</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>如果们设置<code>returnLoop=TRUE</code>，那么<code>getPeak2GeneLinks()</code>会返回一个记录loop track的<code>GRanges</code>对象，链接peak和基因。和之前共开放分析一样，<code>IRanges</code>对象里记录着关联着的peak和基因的位置。当<code>resolution=1</code>, 这连接的是peak的中心和基因的单碱基TSS。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">p2g <span class="operator">&lt;-</span> getPeak2GeneLinks<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.45</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p2g<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 43695 ranges and 2 metadata columns: ## seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># |</span></span><br><span class="line"><span class="comment"># [1] chr1 762901-948847 * | 0.533350285896763</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [43695] chrX 154444701-154664790 * | 0.493389088498317</span></span><br><span class="line"><span class="comment"># FDR</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># [1] 5.20045729958651e-36</span></span><br><span class="line"><span class="comment"># … …</span></span><br><span class="line"><span class="comment"># [43695] 3.37643947034054e-30</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>我们可以将loops的分辨率调整为<code>resolution = 1000</code>, 输出结果主要用于在基因组浏览器中绘制这些连接。因为相同的基因对应的peak过多时，可视化效果会很差。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">p2g <span class="operator">&lt;-</span> getPeak2GeneLinks<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.45</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1000</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p2g<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 42126 ranges and 2 metadata columns:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># |</span></span><br><span class="line"><span class="comment"># [1] chr1 762500-948500 * | 0.533350285896763</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [42126] chrX 154444500-154664500 * | 0.493389088498317</span></span><br><span class="line"><span class="comment"># FDR</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># [1] 5.20045729958651e-36</span></span><br><span class="line"><span class="comment"># … …</span></span><br><span class="line"><span class="comment"># [42126] 3.37643947034054e-30</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>进一步降低分辨率，也会降低peak-to-gene连接的总数。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">p2g <span class="operator">&lt;-</span> getPeak2GeneLinks<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.45</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">10000</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p2g<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 33645 ranges and 2 metadata columns:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># |</span></span><br><span class="line"><span class="comment"># [1] chr1 765000-945000 * | 0.533350285896763</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [33645] chrX 154445000-154665000 * | 0.493389088498317</span></span><br><span class="line"><span class="comment"># FDR</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># [1] 5.20045729958651e-36</span></span><br><span class="line"><span class="comment"># … …</span></span><br><span class="line"><span class="comment"># [33645] 3.37643947034054e-30</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<h3 id="15-3-1-在browser-track中绘制peak-to-gene连接"><a href="#15-3-1-在browser-track中绘制peak-to-gene连接" class="headerlink" title="15.3.1 在browser track中绘制peak-to-gene连接"></a>15.3.1 在browser track中绘制peak-to-gene连接</h3><p>我们可以采用之前共开放分析流程中的相同方法，使用<code>plotBrowserTrack()</code>函数在基因组浏览器上绘制peak-to-gene 连接。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;CD34&quot;</span><span class="punctuation">,</span> <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span><span class="punctuation">,</span> <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MS4A1&quot;</span><span class="punctuation">,</span> <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span><span class="punctuation">,</span> <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD8A&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TBX21&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IL7R&quot;</span> <span class="comment">#TCells</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p <span class="operator">&lt;-</span> plotBrowserTrack<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    geneSymbol <span class="operator">=</span> markerGenes<span class="punctuation">,</span> </span><br><span class="line">    upstream <span class="operator">=</span> <span class="number">50000</span><span class="punctuation">,</span></span><br><span class="line">    downstream <span class="operator">=</span> <span class="number">50000</span><span class="punctuation">,</span></span><br><span class="line">    loops <span class="operator">=</span> getPeak2GeneLinks<span class="punctuation">(</span>projHeme5<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>用<code>grid.draw</code>函数绘制结果，通过<code>$</code>选择特定的标记基因。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grid<span class="operator">::</span>grid.newpage<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">grid<span class="operator">::</span>grid.draw<span class="punctuation">(</span>p<span class="operator">$</span>CD14<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-bf1a61a336e444f89bd51d621d4080da.png" alt="Plot-Tracks-Marker-Genes-with-Peak2GeneLinks"></p>
<p>使用<code>plotPDF()</code>保存可编辑的矢量图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>plotList <span class="operator">=</span> p<span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;Plot-Tracks-Marker-Genes-with-Peak2GeneLinks.pdf&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="15-3-2-绘制Peak-to-gene连接热图"><a href="#15-3-2-绘制Peak-to-gene连接热图" class="headerlink" title="15.3.2 绘制Peak-to-gene连接热图"></a>15.3.2 绘制Peak-to-gene连接热图</h3><p>我们还可以通过绘制peak-to-gene热图的方式展示我们的peak-to-gene连接。热图分为两个部分，一个是scATAC-seq数据，一个是scRNA-seq数据。绘图使用<code>plotPeak2GeneHeatmap()</code>函数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p <span class="operator">&lt;-</span> plotPeak2GeneHeatmap<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>热图的行是k-means聚类结果，其中k由我们进行设置，默认是25.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-2d84a27f73194fd28f3839eb319404f6.png" alt="Peak2GeneHeatmap"></p>
<h2 id="15-4-正向TF-调控因子鉴定"><a href="#15-4-正向TF-调控因子鉴定" class="headerlink" title="15.4 正向TF-调控因子鉴定"></a>15.4 正向TF-调控因子鉴定</h2><p>ATAC-seq还可以无偏好的鉴定TF，这些TF所对应的染色质区域(存在该TF结合的motif DNA序列 )表现出很大改变。此外，从聚集的位置权重矩阵(Position Weight Matrices, PWM)中也能看到这些TF家族在其集合位点上有相似的特征(例如 GATA因子)。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-5381cc535f49440cb162687e3420637b.png" alt="GATA_Motif_PWMs"></p>
<p>这种motif的相似性使得识别特定的TF充满了挑战性，这些TF可能在其预测的结合位点驱动染色质开放状态发生改变。为了规避这一挑战，我们之前用ATAC-seq和RNA-seq来鉴定TF，这些TF的基因表达和对应motif的开放状态呈正相关。我们将这些TF命名为”正向调节因子”。然而，这种分析依赖于匹配的基因表达数据，而这些数据并不是所有实验中都有。ArchR可以识别其推断基因得分与其chromVAR TF偏差z-score相关的TF来克服这种依赖性。实现方式为，ArchR利用低重叠细胞聚集将TF motif的chromVAR deviation z-scroe和TF基因的基因活跃得分相关联。当ArchR整合scRNA-seq时，TF的基因表达可以代替推测的基因活性评分。</p>
<h3 id="15-4-1-第一步-鉴定偏离TF-motif"><a href="#15-4-1-第一步-鉴定偏离TF-motif" class="headerlink" title="15.4.1 第一步: 鉴定偏离TF motif"></a>15.4.1 第一步: 鉴定偏离TF motif</h3><p>鉴定正向TF调控因子的第一步就是鉴定存在偏离的TF motif。我们在之前的章节进行过该分析，为所有的motif创建了一个<code>MotifMatrix</code>，记录着chromVAR deviations 和deviation z-scores。使用<code>getGroupSE()</code>函数可以获取基于聚类平均后的数据，返回的是一个<code>SummarizedExperiment</code>.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">seGroupMotif <span class="operator">&lt;-</span> getGroupSE<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> useMatrix <span class="operator">=</span> <span class="string">&quot;MotifMatrix&quot;</span><span class="punctuation">,</span> groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>因为<code>SummarizedExperiment</code>对象来自于<code>MotifMatrix</code>, 所以它有两个<code>seqnames</code>, “deviation”和”z”，对应chromVAR里原始的偏离值和偏离值的z-scores.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">seGroupMotif</span><br><span class="line"><span class="comment"># class: SummarizedExperiment</span></span><br><span class="line"><span class="comment"># dim: 1740 11</span></span><br><span class="line"><span class="comment"># metadata(0):</span></span><br><span class="line"><span class="comment"># assays(1): MotifMatrix</span></span><br><span class="line"><span class="comment"># rownames(1740): f1 f2 … f1739 f1740</span></span><br><span class="line"><span class="comment"># rowData names(3): seqnames idx name</span></span><br><span class="line"><span class="comment"># colnames(11): B CD4.M … PreB Progenitor</span></span><br><span class="line"><span class="comment"># colData names(16): TSSEnrichment ReadsInTSS … FRIP nCells</span></span><br></pre></td></tr></table></figure>

<p>我们可以从<code>SummarizedExperiment</code>中只提取deviation z-scores.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">seZ <span class="operator">&lt;-</span> seGroupMotif<span class="punctuation">[</span>rowData<span class="punctuation">(</span>seGroupMotif<span class="punctuation">)</span><span class="operator">$</span>seqnames<span class="operator">==</span><span class="string">&quot;z&quot;</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<p>然后，我们计算该z-score在所有聚类中最大变化值。根据motif在不同聚类中观察到的差异程度，我们可以对其分层。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">rowData<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="operator">$</span>maxDelta <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span><span class="built_in">seq_len</span><span class="punctuation">(</span>ncol<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  rowMaxs<span class="punctuation">(</span>assay<span class="punctuation">(</span>seZ<span class="punctuation">)</span> <span class="operator">-</span> assay<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span>x<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> Reduce<span class="punctuation">(</span><span class="string">&quot;cbind&quot;</span><span class="punctuation">,</span> .<span class="punctuation">)</span> <span class="operator">%&gt;%</span> rowMaxs</span><br></pre></td></tr></table></figure>

<h3 id="15-4-2-第二步-鉴定相关的TF-motif和TF基因得分-表达值"><a href="#15-4-2-第二步-鉴定相关的TF-motif和TF基因得分-表达值" class="headerlink" title="15.4.2 第二步: 鉴定相关的TF motif和TF基因得分&#x2F;表达值"></a>15.4.2 第二步: 鉴定相关的TF motif和TF基因得分&#x2F;表达值</h3><p>接着我们使用<code>correlateMatrices()</code>函数来获取motif开放状态矩阵和基因活跃矩阵(基因得分或基因表达量)，在这里就是<code>GeneScoreMatrix</code>和<code>MotifMatrix</code>. 它们将用于鉴定一类motif开放状态和其自身基因活跃相关的TF。正如之前所提到，我们使用许多低重叠细胞聚集在低维空间里(对应<code>reducedDims</code>参数)来计算相关性。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">corGSM_MM <span class="operator">&lt;-</span> correlateMatrices<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    useMatrix1 <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    useMatrix2 <span class="operator">=</span> <span class="string">&quot;MotifMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>该函数返回一个<code>DataFrame</code>对象，记录着<code>GeneScoreMatrix</code>和<code>MotifMatrix</code>，以及它们在低重叠细胞聚类中的相关性。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">corGSM_MM</span><br><span class="line"><span class="comment"># DataFrame with 825 rows and 14 columns</span></span><br><span class="line"><span class="comment"># GeneScoreMatrix_name MotifMatrix_name cor</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 HES4 HES4_95 0.154455056755803</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># 825 MECP2 MECP2_645 0.0629645419544807</span></span><br><span class="line"><span class="comment"># padj pval GeneScoreMatrix_seqnames</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 0.48577773775473 0.000593860315103582 chr1</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># 825 1 0.163611488297825 chrX</span></span><br><span class="line"><span class="comment"># GeneScoreMatrix_start GeneScoreMatrix_end GeneScoreMatrix_strand</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 935552 934342 2</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># 825 153363188 153287264 2</span></span><br><span class="line"><span class="comment"># GeneScoreMatrix_idx GeneScoreMatrix_matchName MotifMatrix_seqnames</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 15 HES4 z</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># 825 874 MECP2 z</span></span><br><span class="line"><span class="comment"># MotifMatrix_idx MotifMatrix_matchName</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 95 HES4</span></span><br><span class="line"><span class="comment"># … … …</span></span><br><span class="line"><span class="comment"># 825 645 MECP2</span></span><br></pre></td></tr></table></figure>

<p>同样的分析还可以使用<code>GeneIntegrationMatrix</code>, 用于替代<code>GeneScoreMatrix</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">corGIM_MM <span class="operator">&lt;-</span> correlateMatrices<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    useMatrix1 <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    useMatrix2 <span class="operator">=</span> <span class="string">&quot;MotifMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="15-4-3-第三步-在相关性DataFrame中添加极大偏差值"><a href="#15-4-3-第三步-在相关性DataFrame中添加极大偏差值" class="headerlink" title="15.4.3 第三步: 在相关性DataFrame中添加极大偏差值"></a>15.4.3 第三步: 在相关性DataFrame中添加极大偏差值</h3><p>对于每个相关性分析，我们使用第一步计算的极大偏差值(maximum delta )来注释每个motif</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">corGSM_MM<span class="operator">$</span>maxDelta <span class="operator">&lt;-</span> rowData<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="punctuation">[</span>match<span class="punctuation">(</span>corGSM_MM<span class="operator">$</span>MotifMatrix_name<span class="punctuation">,</span> rowData<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="operator">$</span>name<span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;maxDelta&quot;</span><span class="punctuation">]</span></span><br><span class="line">corGIM_MM<span class="operator">$</span>maxDelta <span class="operator">&lt;-</span> rowData<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="punctuation">[</span>match<span class="punctuation">(</span>corGIM_MM<span class="operator">$</span>MotifMatrix_name<span class="punctuation">,</span> rowData<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="operator">$</span>name<span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;maxDelta&quot;</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<h3 id="15-4-4-第四步-鉴定正向TF调控因子"><a href="#15-4-4-第四步-鉴定正向TF调控因子" class="headerlink" title="15.4.4 第四步: 鉴定正向TF调控因子"></a>15.4.4 第四步: 鉴定正向TF调控因子</h3><p>我们可以利用所有这些信息来识别正向TF调控因子。在下面的例子中，正向调控因子的标准为</p>
<ul>
<li>motif和基因得分（或基因表达）之间的相关性大于0.5</li>
<li>调整后的p值小于0.01</li>
<li>deviation z-score的最大组间差异位于前四分位。</li>
</ul>
<p>我们应用这些过滤标准，并且做一些文本调整来分离TF名字。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">corGSM_MM <span class="operator">&lt;-</span> corGSM_MM<span class="punctuation">[</span>order<span class="punctuation">(</span><span class="built_in">abs</span><span class="punctuation">(</span>corGSM_MM<span class="operator">$</span>cor<span class="punctuation">)</span><span class="punctuation">,</span> decreasing <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">corGSM_MM <span class="operator">&lt;-</span> corGSM_MM<span class="punctuation">[</span>which<span class="punctuation">(</span><span class="operator">!</span>duplicated<span class="punctuation">(</span>gsub<span class="punctuation">(</span><span class="string">&quot;\\-.*&quot;</span><span class="punctuation">,</span><span class="string">&quot;&quot;</span><span class="punctuation">,</span>corGSM_MM<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;MotifMatrix_name&quot;</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">corGSM_MM<span class="operator">$</span>TFRegulator <span class="operator">&lt;-</span> <span class="string">&quot;NO&quot;</span></span><br><span class="line">corGSM_MM<span class="operator">$</span>TFRegulator<span class="punctuation">[</span>which<span class="punctuation">(</span>corGSM_MM<span class="operator">$</span>cor <span class="operator">&gt;</span> <span class="number">0.5</span> <span class="operator">&amp;</span> corGSM_MM<span class="operator">$</span>padj <span class="operator">&lt;</span> <span class="number">0.01</span> <span class="operator">&amp;</span> corGSM_MM<span class="operator">$</span>maxDelta <span class="operator">&gt;</span> quantile<span class="punctuation">(</span>corGSM_MM<span class="operator">$</span>maxDelta<span class="punctuation">,</span> <span class="number">0.75</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&quot;YES&quot;</span></span><br><span class="line">sort<span class="punctuation">(</span>corGSM_MM<span class="punctuation">[</span>corGSM_MM<span class="operator">$</span>TFRegulator<span class="operator">==</span><span class="string">&quot;YES&quot;</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># [1] “ATOH1” “BCL11A” “CEBPA-DT” “CEBPB” “CEBPD” “CREB1”</span></span><br><span class="line"><span class="comment"># [7] “CREB3L4” “EBF1” “EGR2” “EOMES” “ERF” “ESR1”</span></span><br><span class="line"><span class="comment"># [13] “ETS1” “ETV3” “FUBP1” “GATA1” “GATA2” “GATA5”</span></span><br><span class="line"><span class="comment"># [19] “GATA6” “IRF1” “JDP2” “KLF11” “KLF2” “LYL1”</span></span><br><span class="line"><span class="comment"># [25] “MECOM” “MITF” “NFE2” “NFIA” “NFIB” “NFIC”</span></span><br><span class="line"><span class="comment"># [31] “NFIX” “NHLH1” “POU2F1” “RUNX2” “SIX5” “SMAD1”</span></span><br><span class="line"><span class="comment"># [37] “SMAD9” “SP4” “SPI1” “SPIB” “TAL1” “TCF15”</span></span><br><span class="line"><span class="comment"># [43] “TCF23” “TCF4” “TFAP2C” “TWIST1” “TWIST2” “YY1”</span></span><br><span class="line"><span class="comment"># [49] “ZEB1-AS1”</span></span><br></pre></td></tr></table></figure>

<p>在从基因得分和motif deviation z-scores鉴定到正向TF调控因子后，我们可以在点图中突出展示</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">p <span class="operator">&lt;-</span> ggplot<span class="punctuation">(</span>data.frame<span class="punctuation">(</span>corGSM_MM<span class="punctuation">)</span><span class="punctuation">,</span> aes<span class="punctuation">(</span>cor<span class="punctuation">,</span> maxDelta<span class="punctuation">,</span> color <span class="operator">=</span> TFRegulator<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  geom_point<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  theme_ArchR<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  geom_vline<span class="punctuation">(</span>xintercept <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> lty <span class="operator">=</span> <span class="string">&quot;dashed&quot;</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  scale_color_manual<span class="punctuation">(</span>values <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;NO&quot;</span><span class="operator">=</span><span class="string">&quot;darkgrey&quot;</span><span class="punctuation">,</span> <span class="string">&quot;YES&quot;</span><span class="operator">=</span><span class="string">&quot;firebrick3&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  xlab<span class="punctuation">(</span><span class="string">&quot;Correlation To Gene Score&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  ylab<span class="punctuation">(</span><span class="string">&quot;Max TF Motif Delta&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  scale_y_continuous<span class="punctuation">(</span></span><br><span class="line">    expand <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">    limits <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="built_in">max</span><span class="punctuation">(</span>corGSM_MM<span class="operator">$</span>maxDelta<span class="punctuation">)</span><span class="operator">*</span><span class="number">1.05</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-3ce86da565734c6696875e677cf1e21e.png" alt="TF-Regulators-GeneScores"></p>
<p>使用相同的分析策略处理<code>GeneIntegrationMatrix</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">corGIM_MM <span class="operator">&lt;-</span> corGIM_MM<span class="punctuation">[</span>order<span class="punctuation">(</span><span class="built_in">abs</span><span class="punctuation">(</span>corGIM_MM<span class="operator">$</span>cor<span class="punctuation">)</span><span class="punctuation">,</span> decreasing <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">corGIM_MM <span class="operator">&lt;-</span> corGIM_MM<span class="punctuation">[</span>which<span class="punctuation">(</span><span class="operator">!</span>duplicated<span class="punctuation">(</span>gsub<span class="punctuation">(</span><span class="string">&quot;\\-.*&quot;</span><span class="punctuation">,</span><span class="string">&quot;&quot;</span><span class="punctuation">,</span>corGIM_MM<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;MotifMatrix_name&quot;</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">corGIM_MM<span class="operator">$</span>TFRegulator <span class="operator">&lt;-</span> <span class="string">&quot;NO&quot;</span></span><br><span class="line">corGIM_MM<span class="operator">$</span>TFRegulator<span class="punctuation">[</span>which<span class="punctuation">(</span>corGIM_MM<span class="operator">$</span>cor <span class="operator">&gt;</span> <span class="number">0.5</span> <span class="operator">&amp;</span> corGIM_MM<span class="operator">$</span>padj <span class="operator">&lt;</span> <span class="number">0.01</span> <span class="operator">&amp;</span> corGIM_MM<span class="operator">$</span>maxDelta <span class="operator">&gt;</span> quantile<span class="punctuation">(</span>corGIM_MM<span class="operator">$</span>maxDelta<span class="punctuation">,</span> <span class="number">0.75</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&quot;YES&quot;</span></span><br><span class="line">sort<span class="punctuation">(</span>corGIM_MM<span class="punctuation">[</span>corGIM_MM<span class="operator">$</span>TFRegulator<span class="operator">==</span><span class="string">&quot;YES&quot;</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p <span class="operator">&lt;-</span> ggplot<span class="punctuation">(</span>data.frame<span class="punctuation">(</span>corGIM_MM<span class="punctuation">)</span><span class="punctuation">,</span> aes<span class="punctuation">(</span>cor<span class="punctuation">,</span> maxDelta<span class="punctuation">,</span> color <span class="operator">=</span> TFRegulator<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  geom_point<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  theme_ArchR<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  geom_vline<span class="punctuation">(</span>xintercept <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> lty <span class="operator">=</span> <span class="string">&quot;dashed&quot;</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  scale_color_manual<span class="punctuation">(</span>values <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;NO&quot;</span><span class="operator">=</span><span class="string">&quot;darkgrey&quot;</span><span class="punctuation">,</span> <span class="string">&quot;YES&quot;</span><span class="operator">=</span><span class="string">&quot;firebrick3&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  xlab<span class="punctuation">(</span><span class="string">&quot;Correlation To Gene Expression&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  ylab<span class="punctuation">(</span><span class="string">&quot;Max TF Motif Delta&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  scale_y_continuous<span class="punctuation">(</span></span><br><span class="line">    expand <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">    limits <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="built_in">max</span><span class="punctuation">(</span>corGIM_MM<span class="operator">$</span>maxDelta<span class="punctuation">)</span><span class="operator">*</span><span class="number">1.05</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-82eb8df58b6848ab9a63c70efb798251.png" alt="TF-Regulators-GeneExpression"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%BA%94%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter15/" data-id="clm4a30f50000e6nk11j4bzr1" data-title="使用ArchR分析单细胞ATAC-seq数据(第十五章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第十四章)-analysis-sc-atac-seq-with-archr-chapter14" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E5%9B%9B%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter14/" class="article-date">
  <time class="dt-published" datetime="2021-01-17T16:29:32.657Z" itemprop="datePublished">2021-01-17</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E5%9B%9B%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter14/">使用ArchR分析单细胞ATAC-seq数据(第十四章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第十四章-ArchR的足迹分析"><a href="#第十四章-ArchR的足迹分析" class="headerlink" title="第十四章 ArchR的足迹分析"></a>第十四章 ArchR的足迹分析</h1><p>转录因子(Transcripts factor, TF)足迹分析使得我们能够预测特定位点中TF的精确结合位置。这是因为该位置被TF结合避免了转座酶的切割，而TF结合位点的邻近位置处于开放状态。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-9c35998c679f4262b1d3f7fe73533199.png" alt="image.png"></p>
<p>理想情况下，TF足迹分析需要在单个位置上分析从而确定TF的准确结合位置。但实际上，这需要非常高的测序深度，甚至超过混池ATAC-seq或者scATAC-seq的所有数据。为了解决这个问题，我们可以把和待预测的TF结合相关的Tn5插入位置进行合并。例如，我们可以提取所有包含CTCF motif的peak，制作一个全基因组的CTCF的聚合TF足迹。</p>
<p>为了保证足迹的可靠性，我们需要确保能够可靠的预测出目标TF所对应的结合位点。ArchR使用自带的<code>addMotifAnnotations()</code>函数对peak区域进行搜索，寻找能够匹配的DNA序列。考虑到motif的简并性，无法保证每个motif都有足够的peak。添加到<code>ArchRProject</code>的motif注释以二值矩阵表示(0&#x3D;无motif, 1&#x3D;有motif)。一旦你有了这些motif注释，ArchR使用<code>getFootPrints()</code>函数分析足迹，它以一个<code>ArchRProject</code>对象和一个<code>GenomicRanges</code>对象(记录motif的位置)作为输入。可以使用<code>getPositions()</code>函数从<code>ArchRProject</code>中提取这些位置。之后足迹可以使用<code>plotFootprints()</code>函数可视化。</p>
<p>或许更重要的是，ArchR的足迹分析能够抵消已知的Tn5插入序列偏好性。ArchR使用一个hexmer位置频率矩阵和一个目标Tn5插入位置上的k-mer频率矩阵来实现该功能。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-1c49ea2ad32e4f7b933132461b8233d0.png" alt="foot printing Methods"></p>
<p>最终，该流程输出考虑到Tn5插入偏好性的足迹图。</p>
<p>ArchR支持motif足迹分析和用户提供特征的足迹分析，在后续都会讨论。</p>
<h2 id="14-1-motif足迹分析"><a href="#14-1-motif足迹分析" class="headerlink" title="14.1 motif足迹分析"></a>14.1 motif足迹分析</h2><p>由于教程用到的数据集比较小，因此利用该数据集得到足迹并不是那么的清晰。使用更大的数据会得到较小变异的足迹。</p>
<p>在分析足迹时，我们要先获取和motif相关的所有位置。这可以通过<code>getPositions</code>函数完成。该函数有一个可选参数， <code>name</code>，用于传入<code>peakAnnotation</code>对象中我们想要获取位置的变量名。如果<code>name=NULL</code>, 那么ArchR会使用<code>peakAnnotation</code>槽(slot)的第一个条目(entry)。在下面的例子中，我们没有指定<code>name</code>, ArchR使用的第一个条目为CIS-BP motifs.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">motifPositions <span class="operator">&lt;-</span> getPositions<span class="punctuation">(</span>projHeme5<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这会创建一个<code>GRangesList</code>对象，每个TF motif以不同的<code>GRanges</code>对象进行区分。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">motifPositions</span><br><span class="line"><span class="comment"># GRangesList object of length 870:</span></span><br><span class="line"><span class="comment"># $TFAP2B_1</span></span><br><span class="line"><span class="comment"># GRanges object with 16773 ranges and 1 metadata column:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand | score</span></span><br><span class="line"><span class="comment"># |</span></span><br><span class="line"><span class="comment"># [1] chr1 852468-852479 + | 8.17731199746359</span></span><br><span class="line"><span class="comment"># [2] chr1 873916-873927 + | 8.32673820065588</span></span><br><span class="line"><span class="comment"># [3] chr1 873916-873927 - | 8.32673820065588</span></span><br><span class="line"><span class="comment"># [4] chr1 896671-896682 + | 9.96223327271814</span></span><br><span class="line"><span class="comment"># [5] chr1 896671-896682 - | 8.92408377606486</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [16769] chrX 153991101-153991112 + | 8.39549159740639</span></span><br><span class="line"><span class="comment"># [16770] chrX 154299568-154299579 + | 8.90119825654299</span></span><br><span class="line"><span class="comment"># [16771] chrX 154664929-154664940 - | 8.16690864294221</span></span><br><span class="line"><span class="comment"># [16772] chrX 154807684-154807695 + | 9.57636587154549</span></span><br><span class="line"><span class="comment"># [16773] chrX 154807684-154807695 - | 10.6117355833828</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># …</span></span><br><span class="line"><span class="comment"># &lt;869 more elements&gt;</span></span><br></pre></td></tr></table></figure>

<p>我们提取部分感兴趣的TF motifs用于展示。在我们提取”EBF1”会附带”SREBF1”, 因此我们需要用<code>%ni%</code>显式将其过滤。<code>%ni%</code>函数是R自带函数<code>%in%</code>的相反函数。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">motifs <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;GATA1&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CEBPA&quot;</span><span class="punctuation">,</span> <span class="string">&quot;EBF1&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IRF4&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TBX21&quot;</span><span class="punctuation">,</span> <span class="string">&quot;PAX5&quot;</span><span class="punctuation">)</span></span><br><span class="line">markerMotifs <span class="operator">&lt;-</span> unlist<span class="punctuation">(</span>lapply<span class="punctuation">(</span>motifs<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> grep<span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="built_in">names</span><span class="punctuation">(</span>motifPositions<span class="punctuation">)</span><span class="punctuation">,</span> value <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">markerMotifs <span class="operator">&lt;-</span> markerMotifs<span class="punctuation">[</span>markerMotifs <span class="operator">%ni%</span> <span class="string">&quot;SREBF1_22&quot;</span><span class="punctuation">]</span></span><br><span class="line">markerMotifs</span><br><span class="line"><span class="comment"># [1] “GATA1_383” “CEBPA_155” “EBF1_67” “IRF4_632” “TBX21_780” “PAX5_709”</span></span><br></pre></td></tr></table></figure>

<p>为了准确找到TF足迹，我们需要大量的reads。因此，细胞需要进行分组生成拟混池ATAC-seq谱才能用于TF足迹分析。这些拟混池谱之前在peak鉴定时就已经保存为分组覆盖文件。 如果没有在<code>ArchRProject</code>添加分组覆盖信息，则运行如下命令</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addGroupCoverages<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>在计算分组覆盖度后，我们可以为之前<code>getFootprints()</code>挑选的一组标记motif计算足迹。即便ArchR已经优化了足迹分析流程，我们也建议先对一部分motif分析足迹，而不是直接分析所有motif。 我们通过<code>positions</code>参数来选择motif。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">seFoot <span class="operator">&lt;-</span> getFootprints<span class="punctuation">(</span></span><br><span class="line">  ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">  positions <span class="operator">=</span> motifPositions<span class="punctuation">[</span>markerMotifs<span class="punctuation">]</span><span class="punctuation">,</span> </span><br><span class="line">  groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>当我们获取了这些足迹，我们可以使用<code>plotFootprints()</code>函数进行展示。该函数能够同时以多种方式对足迹进行标准化。下一节，我们会讨论标准化和实际的足迹图。</p>
<h2 id="14-2-Tn5偏好的足迹标准化"><a href="#14-2-Tn5偏好的足迹标准化" class="headerlink" title="14.2 Tn5偏好的足迹标准化"></a>14.2 Tn5偏好的足迹标准化</h2><p>使用ATAC-seq数据分析TF足迹的一大挑战就是Tn5转座酶的插入序列偏好性，这会导致TF足迹的错误分类。为了降低Tn5插入偏好性的影响，ArchR识别每个Tn5插入位置附近的k-mer序列(k由用户提供，默认是6). </p>
<p>对于该项分析，ArchR为每个拟混池识别单碱基分辨率的Tn5插入位点，将这些1-bp位点调整为k-bp窗口（-k&#x2F;2和+（k&#x2F;2-1）bp），然后使用<code>Biostrings</code>包中的<code>oligonucleotidefrequency(w=k, simplify.as=&quot;collapse&quot;)</code>函数创建k-mer频率表。然后，ArchR使用与<code>BSgenome</code>相关的基因组文件，以相同的函数计算出全基因组范围预期的k-mers。</p>
<p>为了计算拟混池足迹的插入偏差，ArchR创建了一个k-mer频率矩阵，该矩阵表示为从motif中心到窗口+&#x2F;-N bp（用户定义，默认为250 bp）的所有可能k-mer。然后，遍历每个motif位点，ArchR将定位的k-mer填充到k-mer频率矩阵中。然后在全基因组范围内计算每个motif位置。利用样本的k-mer频率表，ArchR可以通过将k-mer位置频率表乘以观察&#x2F;期望 Tn5 k-mer频率来计算预期的Tn5插入。</p>
<p>以上所有这些发生在<code>plotFootprints()</code>函数中。</p>
<h3 id="14-2-1-减去Tn5偏好"><a href="#14-2-1-减去Tn5偏好" class="headerlink" title="14.2.1 减去Tn5偏好"></a>14.2.1 减去Tn5偏好</h3><p>一个标准化方式就是从足迹信号中减去Tn5偏好。该标准化方法通过设置<code>plotFootprints()</code>的<code>normMethod = &quot;Subtract&quot;</code>实现</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">plotFootprints<span class="punctuation">(</span></span><br><span class="line">  seFoot <span class="operator">=</span> seFoot<span class="punctuation">,</span></span><br><span class="line">  ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">  normMethod <span class="operator">=</span> <span class="string">&quot;Subtract&quot;</span><span class="punctuation">,</span></span><br><span class="line">  plotName <span class="operator">=</span> <span class="string">&quot;Footprints-Subtract-Bias&quot;</span><span class="punctuation">,</span></span><br><span class="line">  addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">  smoothWindow <span class="operator">=</span> <span class="number">5</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>默认，这些图保存在<code>ArchRProjec</code>t的<code>outputDirectory</code>。如果你需要绘制所有motif, 可以将其返回为<code>ggplot2</code>对象，需要注意这个<code>ggplot</code>对象会非常大。下面是一个从motif足迹中减去Tn5偏好信号的结果</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-9445c244d46f4c60b77b93eaf5209d14.png" alt="Footprints-Subtract-Bias"></p>
<h3 id="14-2-2-除以Tn5偏好"><a href="#14-2-2-除以Tn5偏好" class="headerlink" title="14.2.2 除以Tn5偏好"></a>14.2.2 除以Tn5偏好</h3><p>第二种标准化方法就是将足迹除以Tn5偏好信号。该标准化方法通过设置<code>plotFootprints()</code>的<code>normMethod = &quot;Divide&quot;</code>实现</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">plotFootprints<span class="punctuation">(</span></span><br><span class="line">  seFoot <span class="operator">=</span> seFoot<span class="punctuation">,</span></span><br><span class="line">  ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">  normMethod <span class="operator">=</span> <span class="string">&quot;Divide&quot;</span><span class="punctuation">,</span></span><br><span class="line">  plotName <span class="operator">=</span> <span class="string">&quot;Footprints-Divide-Bias&quot;</span><span class="punctuation">,</span></span><br><span class="line">  addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">  smoothWindow <span class="operator">=</span> <span class="number">5</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>下面是一个从motif足迹中除以Tn5偏好信号的结果</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-16a63ea23fac4d31abf71b7b09d9a091.png" alt="Footprints-Divide-Bias"></p>
<h3 id="14-2-3-无Tn5偏好标准化的足迹"><a href="#14-2-3-无Tn5偏好标准化的足迹" class="headerlink" title="14.2.3 无Tn5偏好标准化的足迹"></a>14.2.3 无Tn5偏好标准化的足迹</h3><p>尽管我们高度推荐将足迹根据Tn5序列插入偏好性进行标准化，当然你可以通过设置<code>plotFootprints()</code>的<code>normMethod = &quot;None&quot;</code>来省去标准化。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">plotFootprints<span class="punctuation">(</span></span><br><span class="line">  seFoot <span class="operator">=</span> seFoot<span class="punctuation">,</span></span><br><span class="line">  ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">  normMethod <span class="operator">=</span> <span class="string">&quot;None&quot;</span><span class="punctuation">,</span></span><br><span class="line">  plotName <span class="operator">=</span> <span class="string">&quot;Footprints-No-Normalization&quot;</span><span class="punctuation">,</span></span><br><span class="line">  addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">  smoothWindow <span class="operator">=</span> <span class="number">5</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>


<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-7fcb07f110c747c4ac55c3a71bde7f96.png" alt="Footprints-No-Normalization"></p>
<h2 id="14-3-特征足迹"><a href="#14-3-特征足迹" class="headerlink" title="14.3 特征足迹"></a>14.3 特征足迹</h2><p>除了motif足迹分析，ArchR还允许用户分析任意定义的特征数据集。为了对功能进行说明，我们将会使用<code>plotFootprints()</code>函数创建TSS插入谱(在之前数据质量控制一节中引入)。一个TSS插入谱本质上就是特殊的足迹。</p>
<p>我们在之前小节讨论过，足迹会用到来源于拟混池重复的分组覆盖文件。我们在之前peak鉴定时创建过这些文件。如果你没有在<code>ArchRProject</code>加入分组覆盖信息, 那么需要运行如下代码</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addGroupCoverages<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们接着创建一个没有经过Tn5偏好性校正的TSS插入谱。和之前分析一个主要不同是，我们设置了<code>flank=2000</code>, 将TSS向前向后分别延伸2000 bp.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">seTSS <span class="operator">&lt;-</span> getFootprints<span class="punctuation">(</span></span><br><span class="line">  ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">  positions <span class="operator">=</span> GRangesList<span class="punctuation">(</span>TSS <span class="operator">=</span> getTSS<span class="punctuation">(</span>projHeme5<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">  groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span></span><br><span class="line">  flank <span class="operator">=</span> <span class="number">2000</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们接着用<code>plotFootprints()</code>对每个细胞分组绘制TSS插入谱。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">plotFootprints<span class="punctuation">(</span></span><br><span class="line">  seFoot <span class="operator">=</span> seTSS<span class="punctuation">,</span></span><br><span class="line">  ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">  normMethod <span class="operator">=</span> <span class="string">&quot;None&quot;</span><span class="punctuation">,</span></span><br><span class="line">  plotName <span class="operator">=</span> <span class="string">&quot;TSS-No-Normalization&quot;</span><span class="punctuation">,</span></span><br><span class="line">  addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">  flank <span class="operator">=</span> <span class="number">2000</span><span class="punctuation">,</span></span><br><span class="line">  flankNorm <span class="operator">=</span> <span class="number">100</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-b23beabc98094afd8d59ba25139d06e6.png" alt="TSS-No-Normalization"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E5%9B%9B%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter14/" data-id="clm4a30q000e8e6nk2uwm3qyg" data-title="使用ArchR分析单细胞ATAC-seq数据(第十四章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-测试下数据公式-ce-shi-xia-shu-ju-gong-shi" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/11/04/%E6%B5%8B%E8%AF%95%E4%B8%8B%E6%95%B0%E6%8D%AE%E5%85%AC%E5%BC%8F-ce-shi-xia-shu-ju-gong-shi/" class="article-date">
  <time class="dt-published" datetime="2020-11-04T22:35:32.121Z" itemprop="datePublished">2020-11-04</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/11/04/%E6%B5%8B%E8%AF%95%E4%B8%8B%E6%95%B0%E6%8D%AE%E5%85%AC%E5%BC%8F-ce-shi-xia-shu-ju-gong-shi/">测试下数据公式</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>$$<br>a+b<br>$$</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/11/04/%E6%B5%8B%E8%AF%95%E4%B8%8B%E6%95%B0%E6%8D%AE%E5%85%AC%E5%BC%8F-ce-shi-xia-shu-ju-gong-shi/" data-id="clm4a30tl00jhe6nkfdxb9qil" data-title="测试下数据公式" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-基于igv.js的一个小工具-a-simple-wrapper-of-igv-js" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/11/02/%E5%9F%BA%E4%BA%8Eigv.js%E7%9A%84%E4%B8%80%E4%B8%AA%E5%B0%8F%E5%B7%A5%E5%85%B7-a-simple-wrapper-of-igv-js/" class="article-date">
  <time class="dt-published" datetime="2020-11-02T22:05:02.886Z" itemprop="datePublished">2020-11-02</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Python/">Python</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/11/02/%E5%9F%BA%E4%BA%8Eigv.js%E7%9A%84%E4%B8%80%E4%B8%AA%E5%B0%8F%E5%B7%A5%E5%85%B7-a-simple-wrapper-of-igv-js/">基于igv.js的一个小工具</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>当我想查看软件运行结束后得到的BAM, BigWig, BED和GTF文件时，我都需要先把他们下载到本地，然后用IGV打开。每每这个时候，我就会非常痛苦，因为我懒得下载。</p>
<p>为了解决这个问题，我基于igv.js 写了一个Python脚本，可以根据提供的参考基因组，bam文件输出一个网页，然后通过利用 node.js 或者 Python的网页服务器打开一个端口，直接在网页上进行查看。</p>
<p>代码地址: <a target="_blank" rel="noopener" href="https://github.com/xuzhougeng/myscripts/blob/master/igv_web.py">https://github.com/xuzhougeng/myscripts/blob/master/igv_web.py</a></p>
<p>可以下面这行代码进行下载脚本, 例如我放在了家目录</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wget https://raw.githubusercontent.com/xuzhougeng/myscripts/master/igv_web.py</span><br></pre></td></tr></table></figure>

<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python3 ~/igv_web.py  -r ref/genome.fa -m 01-read-align/*.bam -b feature.bed</span><br></pre></td></tr></table></figure>

<p>接着用python的 SimpleHTTPServer 模块启动一个网页服务器</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">/usr/bin/python -m SimpleHTTPServer 9999</span><br></pre></td></tr></table></figure>

<p>当然更加推荐使用npm的http-server，通过网页进行访问</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">npx http-server ./ -p 9999 -a 0.0.0.0</span><br></pre></td></tr></table></figure>

<p>效果如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/11/image-653ecf688d544364bc6498c0725ab215.png" alt="image.png"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/11/02/%E5%9F%BA%E4%BA%8Eigv.js%E7%9A%84%E4%B8%80%E4%B8%AA%E5%B0%8F%E5%B7%A5%E5%85%B7-a-simple-wrapper-of-igv-js/" data-id="clm4a30m1007we6nka9bh4586" data-title="基于igv.js的一个小工具" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-如何安装JCVI-how-to-install-jcvi" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/10/22/%E5%A6%82%E4%BD%95%E5%AE%89%E8%A3%85JCVI-how-to-install-jcvi/" class="article-date">
  <time class="dt-published" datetime="2020-10-22T02:23:07.588Z" itemprop="datePublished">2020-10-22</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/10/22/%E5%A6%82%E4%BD%95%E5%AE%89%E8%A3%85JCVI-how-to-install-jcvi/">如何安装JCVI</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>使用conda安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda create -y -c bioconda -n jcvi jcvi </span><br></pre></td></tr></table></figure>

<p>使用时需要启动环境</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda activate jcvi</span><br></pre></td></tr></table></figure>

<p>在此基础上可以更新到最新版的JCVI (Github版本会不断的增加新功能并修订bug）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">pip install git+git://github.com/tanghaibao/jcvi.git</span><br></pre></td></tr></table></figure>

<p>安装额外的依赖环境</p>
<ul>
<li><p>Kent tools</p>
</li>
<li><p>BEDTOOLS</p>
</li>
<li><p>EMBOSS</p>
</li>
<li><p><a target="_blank" rel="noopener" href="http://last.cbrc.jp/">LAST</a></p>
</li>
<li><p>LaTex</p>
</li>
</ul>
<p>依赖环境中，BEDTOOLS, EMBOSS和LAST可以用conda安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># BEDTOOLS 和 EMBOSS</span></span><br><span class="line">conda install -y -n jcvi -c bioconda bedtools emboss last</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>Kent tools需要编译, 并且依赖MySQL，暂时不需要安装。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Kent tools, 134M</span></span><br><span class="line">wget http://hgdownload.cse.ucsc.edu/admin/jksrc.zip </span><br><span class="line">unzip  jksrc.zip</span><br><span class="line"><span class="built_in">cd</span> kent/src/lib</span><br><span class="line">make </span><br></pre></td></tr></table></figure>

<p>对于Latex，conda的安装版本 ( texlive-core )在我后续的测试中各种出问题, 我最后根据根据官方的<a target="_blank" rel="noopener" href="https://www.tug.org/texlive/quickinstall.html">quickinstall.html</a> 进行了安装，具体的安装过程见 <a href="/archives/install-latex-without-root">LaTex</a></p>
<p>建议通过root进行安装（或者手工安装）。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># ubuntu</span></span><br><span class="line">sudo apt-get install -y texlive texlive-latex-extra texlive-latex-recommended</span><br><span class="line"><span class="comment"># centos</span></span><br><span class="line">sudo yum install -y  texlive texlive-latex texlive-xetex texlive-collection-latexrecommended</span><br></pre></td></tr></table></figure>

<p>最终通过Python版的MCscan对安装结果进行测试 （需要用conda安装seqkit)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p <span class="built_in">test</span> &amp;&amp; <span class="built_in">cd</span> <span class="built_in">test</span></span><br><span class="line"><span class="comment">#Athaliana</span></span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/arabidopsis_thaliana/cdna/Arabidopsis_thaliana.TAIR10.cdna.all.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/arabidopsis_thaliana/pep/Arabidopsis_thaliana.TAIR10.pep.all.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.44.gff3.gz</span><br><span class="line"><span class="comment">#Osativa</span></span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/oryza_sativa/cdna/Oryza_sativa.IRGSP-1.0.cdna.all.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/oryza_sativa/pep/Oryza_sativa.IRGSP-1.0.pep.all.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/gff3/oryza_sativa/Oryza_sativa.IRGSP-1.0.44.gff3.gz </span><br><span class="line"><span class="comment"># convert gff to bed</span></span><br><span class="line">python -m jcvi.formats.gff bed --<span class="built_in">type</span>=mRNA --key=transcript_id Arabidopsis_thaliana.TAIR10.44.gff3.gz &gt; ath.bed</span><br><span class="line">python -m jcvi.formats.gff bed --<span class="built_in">type</span>=mRNA --key=transcript_id Oryza_sativa.IRGSP-1.0.44.gff3.gz &gt; osa.bed </span><br><span class="line"><span class="comment"># deduplication</span></span><br><span class="line">python -m jcvi.formats.bed <span class="built_in">uniq</span> ath.bed</span><br><span class="line">python -m jcvi.formats.bed <span class="built_in">uniq</span> osa.bed </span><br><span class="line"><span class="comment"># Athaliana</span></span><br><span class="line">seqkit grep -f &lt;(<span class="built_in">cut</span> -f 4 ath.uniq.bed ) Arabidopsis_thaliana.TAIR10.cdna.all.fa.gz | seqkit <span class="built_in">seq</span> -i &gt; ath.cds</span><br><span class="line">seqkit grep -f &lt;(<span class="built_in">cut</span> -f 4 ath.uniq.bed ) Arabidopsis_thaliana.TAIR10.pep.all.fa.gz | seqkit <span class="built_in">seq</span> -i &gt; ath.pep </span><br><span class="line"><span class="comment"># Osativa</span></span><br><span class="line">seqkit grep -f &lt;(<span class="built_in">cut</span> -f 4 osa.uniq.bed )  Oryza_sativa.IRGSP-1.0.cdna.all.fa.gz | seqkit <span class="built_in">seq</span> -i  &gt; osa.cds</span><br><span class="line">seqkit grep -f &lt;(<span class="built_in">cut</span> -f 4 osa.uniq.bed ) Oryza_sativa.IRGSP-1.0.pep.all.fa.gz | seqkit <span class="built_in">seq</span> -i  &gt; osa.pep </span><br><span class="line"><span class="comment"># create a new directory</span></span><br><span class="line"><span class="built_in">mkdir</span> -p cds &amp;&amp; <span class="built_in">cd</span> cds</span><br><span class="line"><span class="built_in">ln</span> -s ../ath.cds ath.cds</span><br><span class="line"><span class="built_in">ln</span> -s ../ath.uniq.bed ath.bed</span><br><span class="line"><span class="built_in">ln</span> -s ../osa.cds osa.cds</span><br><span class="line"><span class="built_in">ln</span> -s ../osa.uniq.bed osa.bed </span><br><span class="line"><span class="comment"># compara</span></span><br><span class="line">python -m jcvi.compara.catalog ortholog --no_strip_names ath osa </span><br></pre></td></tr></table></figure>

<p>如果顺利，当前目录下会有一个 ath.osa.pdf 文件</p>
<p>我测试的时候遇到了两个不顺利，都是和LaTex有关的报错。</p>
<p>第一个是latex没有安装导致的报错</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">FileNotFoundError: [Errno 2] No such file or directory: <span class="string">&#x27;latex&#x27;</span>: <span class="string">&#x27;latex&#x27;</span></span><br><span class="line">RuntimeError: Failed to process string with tex because latex could not be found</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>第二个是用conda安装texlive-core后，出现的第二个问题，还是得手工安装LaTex才解决。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">kpathsea: Running mktexfmt latex.fmt</span><br><span class="line">Can<span class="string">&#x27;t locate mktexlsr.pl in @INC  </span></span><br></pre></td></tr></table></figure>


      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/10/22/%E5%A6%82%E4%BD%95%E5%AE%89%E8%A3%85JCVI-how-to-install-jcvi/" data-id="clm4a30s600h4e6nkbq4j6w8h" data-title="如何安装JCVI" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/8/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/7/">7</a><a class="page-number" href="/page/8/">8</a><span class="page-number current">9</span><a class="page-number" href="/page/10/">10</a><a class="page-number" href="/page/11/">11</a><span class="space">&hellip;</span><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/10/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/09/">九月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/09/04/%E5%A6%82%E4%BD%95%E5%88%A4%E6%96%AD%E8%BD%AC%E5%BD%95%E7%BB%84%E6%B5%8B%E5%BA%8F%E6%98%AF%E5%90%A6%E4%B8%BA%E9%93%BE%E7%89%B9%E5%BC%82%E6%80%A7%E5%BB%BA%E5%BA%93/">如何判断转录组测序是否为链特异性建库</a>
          </li>
        
          <li>
            <a href="/2023/09/02/%E5%A6%82%E4%BD%95%E8%AE%A9%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8D%E5%86%8D%E9%BE%9F%E9%80%9F%E4%B8%8B%E8%BD%BDgithub%E7%9A%84%E6%95%B0%E6%8D%AE/">如何让服务器不再龟速下载github的数据</a>
          </li>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>