<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/11/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-Singularity和MPI应用-singularity-and-mpi-applications" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/06/22/Singularity%E5%92%8CMPI%E5%BA%94%E7%94%A8-singularity-and-mpi-applications/" class="article-date">
  <time class="dt-published" datetime="2020-06-22T01:39:27.294Z" itemprop="datePublished">2020-06-22</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/06/22/Singularity%E5%92%8CMPI%E5%BA%94%E7%94%A8-singularity-and-mpi-applications/">Singularity和MPI应用</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>MPI(Message Passin Interface)广泛应用于高性能服务器中，可用于单系统多节点或者多个计算平台间通讯，目前主流的两个开源软件分别是OpenMPI和MPICH。Singularity同时支持这两个开源工具，本篇教程介绍如何在Singularity容器开发和运行MPI程序。</p>
<p>在Singularity容器是执行MPI程序最常见的方法就是使用宿主机中的MPI。由于我们会同时使用宿主机的MPI和容器中的MPI，因此这被称为Host MPI或者混合模型。考虑到大部分高性能计算平台并不都支持宿主机和容器之间的文件系统共享，因此我们不讨论如何将存储盘挂载到容器，从而使用容器中宿主机MPI。</p>
<p>混合途径的基本想法就是，当你要执行SIngularity容器的MPI代码时，你会使用<code>mpiexec</code>类似命令去调用<code>singularity</code>命令。容器外部的MPI进程会和容器内的MPI进行写作，容器后的MPI代码会实例化任务。</p>
<p>Open MPI&#x2F;Singularity的工作流程如下</p>
<ol>
<li>用户在shell中启动MPI(mpirun, mpiexec)</li>
<li>Open MPI接着调用进程管理守护进程(process management daemon, ORTED)</li>
<li>ORTED进程启动所需的Singularity容器</li>
<li>Singularity构建容器和命名空间环境</li>
<li>Singularity启动容器内的MPI应用</li>
<li>MPI进程启动，加载Open MPI库</li>
<li>Open MPI库通过进程管理接口(Process Management, PMI)连接回ORTED进程</li>
</ol>
<p>此时在容器运行MPI就和直接宿主机运行MPI程序一样。优势在于它能够整合到Slurm等资源管理工具，并且和之前运行MPI应用一样。但是你需要保证容器中的MPI和宿主机的MPI版本兼容，也就是在构建MPI容器时保证和系统的MPI一致。</p>
<p>我们举个简单的例子。假设我们已经有了一个叫做<code>mpitest.c</code>MPI应用</p>
<figure class="highlight c"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;mpi.h&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;stdio.h&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;stdlib.h&gt;</span></span></span><br><span class="line"></span><br><span class="line"><span class="type">int</span> <span class="title function_">main</span> <span class="params">(<span class="type">int</span> argc, <span class="type">char</span> **argv)</span> &#123;</span><br><span class="line">        <span class="type">int</span> rc;</span><br><span class="line">        <span class="type">int</span> size;</span><br><span class="line">        <span class="type">int</span> myrank;</span><br><span class="line"></span><br><span class="line">        rc = MPI_Init (&amp;argc, &amp;argv);</span><br><span class="line">        <span class="keyword">if</span> (rc != MPI_SUCCESS) &#123;</span><br><span class="line">                <span class="built_in">fprintf</span> (<span class="built_in">stderr</span>, <span class="string">&quot;MPI_Init() failed&quot;</span>);</span><br><span class="line">                <span class="keyword">return</span> EXIT_FAILURE;</span><br><span class="line">        &#125;</span><br><span class="line"></span><br><span class="line">        rc = MPI_Comm_size (MPI_COMM_WORLD, &amp;size);</span><br><span class="line">        <span class="keyword">if</span> (rc != MPI_SUCCESS) &#123;</span><br><span class="line">                <span class="built_in">fprintf</span> (<span class="built_in">stderr</span>, <span class="string">&quot;MPI_Comm_size() failed&quot;</span>);</span><br><span class="line">                <span class="keyword">goto</span> exit_with_error;</span><br><span class="line">        &#125;</span><br><span class="line"></span><br><span class="line">        rc = MPI_Comm_rank (MPI_COMM_WORLD, &amp;myrank);</span><br><span class="line">        <span class="keyword">if</span> (rc != MPI_SUCCESS) &#123;</span><br><span class="line">                <span class="built_in">fprintf</span> (<span class="built_in">stderr</span>, <span class="string">&quot;MPI_Comm_rank() failed&quot;</span>);</span><br><span class="line">                <span class="keyword">goto</span> exit_with_error;</span><br><span class="line">        &#125;</span><br><span class="line"></span><br><span class="line">        <span class="built_in">fprintf</span> (<span class="built_in">stdout</span>, <span class="string">&quot;Hello, I am rank %d/%d&quot;</span>, myrank, size);</span><br><span class="line"></span><br><span class="line">        MPI_Finalize();</span><br><span class="line"></span><br><span class="line">        <span class="keyword">return</span> EXIT_SUCCESS;</span><br><span class="line"></span><br><span class="line"> exit_with_error:</span><br><span class="line">        MPI_Finalize();</span><br><span class="line">        <span class="keyword">return</span> EXIT_FAILURE;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<blockquote>
<p>MPI只是一个库接口，里面的函数能够被其他编程语言所调用，所以支持Fortran, C, Python, R等。</p>
</blockquote>
<p>接下来根据宿主机的MPI版本，编写一个定义文件。假如我们宿主机的MPI是MPICH，那么定义文件(命名为mpitest)如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br></pre></td><td class="code"><pre><span class="line">Bootstrap: docker</span><br><span class="line">From: ubuntu:latest</span><br><span class="line"></span><br><span class="line">%files</span><br><span class="line">    mpitest.c /opt</span><br><span class="line"></span><br><span class="line">%environment</span><br><span class="line">    <span class="built_in">export</span> MPICH_DIR=/opt/mpich-3.3</span><br><span class="line">    <span class="built_in">export</span> SINGULARITY_MPICH_DIR=<span class="variable">$MPICH_DIR</span></span><br><span class="line">    <span class="built_in">export</span> SINGULARITYENV_APPEND_PATH=<span class="variable">$MPICH_DIR</span>/bin</span><br><span class="line">    <span class="built_in">export</span> SINGULAIRTYENV_APPEND_LD_LIBRARY_PATH=<span class="variable">$MPICH_DIR</span>/lib</span><br><span class="line"></span><br><span class="line">%post</span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Installing required packages...&quot;</span></span><br><span class="line">    apt-get update &amp;&amp; apt-get install -y wget git bash gcc gfortran g++ make</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Information about the version of MPICH to use</span></span><br><span class="line">    <span class="built_in">export</span> MPICH_VERSION=3.3</span><br><span class="line">    <span class="built_in">export</span> MPICH_URL=<span class="string">&quot;http://www.mpich.org/static/downloads/<span class="variable">$MPICH_VERSION</span>/mpich-<span class="variable">$MPICH_VERSION</span>.tar.gz&quot;</span></span><br><span class="line">    <span class="built_in">export</span> MPICH_DIR=/opt/mpich</span><br><span class="line"></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Installing MPICH...&quot;</span></span><br><span class="line">    <span class="built_in">mkdir</span> -p /tmp/mpich</span><br><span class="line">    <span class="built_in">mkdir</span> -p /opt</span><br><span class="line">    <span class="comment"># Download</span></span><br><span class="line">    <span class="built_in">cd</span> /tmp/mpich &amp;&amp; wget -O mpich-<span class="variable">$MPICH_VERSION</span>.tar.gz <span class="variable">$MPICH_URL</span> &amp;&amp; tar xzf mpich-<span class="variable">$MPICH_VERSION</span>.tar.gz</span><br><span class="line">    <span class="comment"># Compile and install</span></span><br><span class="line">    <span class="built_in">cd</span> /tmp/mpich/mpich-<span class="variable">$MPICH_VERSION</span> &amp;&amp; ./configure --prefix=<span class="variable">$MPICH_DIR</span> &amp;&amp; make install</span><br><span class="line">    <span class="comment"># Set env variables so we can compile our application</span></span><br><span class="line">    <span class="built_in">export</span> PATH=<span class="variable">$MPICH_DIR</span>/bin:<span class="variable">$PATH</span></span><br><span class="line">    <span class="built_in">export</span> LD_LIBRARY_PATH=<span class="variable">$MPICH_DIR</span>/lib:<span class="variable">$LD_LIBRARY_PATH</span></span><br><span class="line">    <span class="built_in">export</span> MANPATH=<span class="variable">$MPICH_DIR</span>/share/man:<span class="variable">$MANPATH</span></span><br><span class="line"></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Compiling the MPI application...&quot;</span></span><br><span class="line">    <span class="built_in">cd</span> /opt &amp;&amp; mpicc -o mpitest mpitest.c</span><br></pre></td></tr></table></figure>

<p>如果是Open MPI,</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">Bootstrap: docker</span><br><span class="line">From: ubuntu:latest</span><br><span class="line"></span><br><span class="line">%files</span><br><span class="line">    mpitest.c /opt</span><br><span class="line"></span><br><span class="line">%environment</span><br><span class="line">    <span class="built_in">export</span> OMPI_DIR=/opt/ompi</span><br><span class="line">    <span class="built_in">export</span> SINGULARITY_OMPI_DIR=<span class="variable">$OMPI_DIR</span></span><br><span class="line">    <span class="built_in">export</span> SINGULARITYENV_APPEND_PATH=<span class="variable">$OMPI_DIR</span>/bin</span><br><span class="line">    <span class="built_in">export</span> SINGULAIRTYENV_APPEND_LD_LIBRARY_PATH=<span class="variable">$OMPI_DIR</span>/lib</span><br><span class="line"></span><br><span class="line">%post</span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Installing required packages...&quot;</span></span><br><span class="line">    apt-get update &amp;&amp; apt-get install -y wget git bash gcc gfortran g++ make file</span><br><span class="line"></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Installing Open MPI&quot;</span></span><br><span class="line">    <span class="built_in">export</span> OMPI_DIR=/opt/ompi</span><br><span class="line">    <span class="built_in">export</span> OMPI_VERSION=4.0.1</span><br><span class="line">    <span class="built_in">export</span> OMPI_URL=<span class="string">&quot;https://download.open-mpi.org/release/open-mpi/v4.0/openmpi-<span class="variable">$OMPI_VERSION</span>.tar.bz2&quot;</span></span><br><span class="line">    <span class="built_in">mkdir</span> -p /tmp/ompi</span><br><span class="line">    <span class="built_in">mkdir</span> -p /opt</span><br><span class="line">    <span class="comment"># Download</span></span><br><span class="line">    <span class="built_in">cd</span> /tmp/ompi &amp;&amp; wget -O openmpi-<span class="variable">$OMPI_VERSION</span>.tar.bz2 <span class="variable">$OMPI_URL</span> &amp;&amp; tar -xjf openmpi-<span class="variable">$OMPI_VERSION</span>.tar.bz2</span><br><span class="line">    <span class="comment"># Compile and install</span></span><br><span class="line">    <span class="built_in">cd</span> /tmp/ompi/openmpi-<span class="variable">$OMPI_VERSION</span> &amp;&amp; ./configure --prefix=<span class="variable">$OMPI_DIR</span> &amp;&amp; make install</span><br><span class="line">    <span class="comment"># Set env variables so we can compile our application</span></span><br><span class="line">    <span class="built_in">export</span> PATH=<span class="variable">$OMPI_DIR</span>/bin:<span class="variable">$PATH</span></span><br><span class="line">    <span class="built_in">export</span> LD_LIBRARY_PATH=<span class="variable">$OMPI_DIR</span>/lib:<span class="variable">$LD_LIBRARY_PATH</span></span><br><span class="line">    <span class="built_in">export</span> MANPATH=<span class="variable">$OMPI_DIR</span>/share/man:<span class="variable">$MANPATH</span></span><br><span class="line"></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Compiling the MPI application...&quot;</span></span><br><span class="line">    <span class="built_in">cd</span> /opt &amp;&amp; mpicc -o mpitest mpitest.c</span><br></pre></td></tr></table></figure>

<p>接着我们从定义文件中构建出sig文件, <code>singularity build mpitest.sig mpitest</code></p>
<p>最后，执行Singularity容器的MPI应用的标准方法就是在host中运行<code>mpirun</code>命令，它会启动Singularity容器，并最终运行其中的MPI程序。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">mpirun -n 4 singularity <span class="built_in">exec</span> mpitest.sig /opt/mpitest</span><br><span class="line"><span class="comment"># 输出信息如下</span></span><br><span class="line"><span class="comment"># Hello, I am rank 0/4Hello, I am rank 1/4Hello, I am rank 2/4Hello, I am rank 3/4</span></span><br></pre></td></tr></table></figure>

<p>如果是在SLURM这类任务提交系统中，我们需要批处理脚本来执行MPI引用，举个例子</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#!/bin/bash</span></span><br><span class="line"><span class="comment">#SBATCH --job-name singularity-mpi</span></span><br><span class="line"><span class="comment">#SBATCH -N $NNODES # total number of nodes</span></span><br><span class="line"><span class="comment">#SBATCH --time=00:05:00 # Max execution time</span></span><br><span class="line"></span><br><span class="line">mpirun -n <span class="variable">$NP</span> singularity <span class="built_in">exec</span> /var/nfsshare/gvallee/mpich.sif /opt/mpitest</span><br></pre></td></tr></table></figure>

<p>执行方式如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sbatch my_job.sh</span><br></pre></td></tr></table></figure>

<p>参考资料: <a target="_blank" rel="noopener" href="https://sylabs.io/guides/3.3/user-guide/mpi.html">https://sylabs.io/guides/3.3/user-guide/mpi.html</a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/06/22/Singularity%E5%92%8CMPI%E5%BA%94%E7%94%A8-singularity-and-mpi-applications/" data-id="clm1ywdgg0066ecokbwsf6f5d" data-title="Singularity和MPI应用" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第九章)-analysis-sc-atac-seq-with-archr-chapter9" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/06/04/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B9%9D%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter9/" class="article-date">
  <time class="dt-published" datetime="2020-06-04T09:35:45.570Z" itemprop="datePublished">2020-06-04</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/06/04/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B9%9D%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter9/">使用ArchR分析单细胞ATAC-seq数据(第九章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第9章-ArchR的拟混池重复"><a href="#第9章-ArchR的拟混池重复" class="headerlink" title="第9章 ArchR的拟混池重复"></a>第9章 ArchR的拟混池重复</h1><p>因为scATAC-seq数据本质上只有两种值，也就是说每个位点要么开放要么不开放，所以你会发现某些情况下无法使用单个细胞进行数据分析。此外，许多我们想要做的分析也依赖于重复才能计算统计显著性。对于单细胞数据，我们通过构建拟混池重复(pseudo-bulk replicates)来解决该问题。所谓的拟混池(pesudo-bulk)指的就是将单细胞进行合并模拟成混池测序的ATAC-seq实验得到的数据。ArchR为每个目标细胞分组构建多个拟混池样本，也就得到了拟混池重复。这个模拟过程背后的假设是将单细胞进行合并结果和实际的混池结果非常接近，以至于不需要在乎背后的差异。这些细胞分组通常都是来自于单个细胞类群，或者是一直细胞类型对应的可能聚类。我们这一章会介绍如何使用ArchR生成这些拟混池重复。</p>
<h2 id="9-1-ArchR如何构建拟混池重复"><a href="#9-1-ArchR如何构建拟混池重复" class="headerlink" title="9.1 ArchR如何构建拟混池重复"></a>9.1 ArchR如何构建拟混池重复</h2><p>ArchR使用分级优先法(tiered priority approach)构建拟混池重复。使用者定义: i)最小和最大重复数(minReps&#x2F;maxReps), ii)每个重复的最少和最多细胞数(minCells&#x2F;maxCells), iii)缺少足够细胞用于构建足够重复时的采样率(Sampling Ratio)。举个例子，当采样率等于0.8时，每个重复中80%的细胞来自于无放回抽样(这会导致重复间出现有放回抽样)。在这种情况下，多个重复中可能有一些相同的细胞，但是为了能在缺少足够细胞的分组里得到拟混池重复，这是必要的牺牲。</p>
<p>我们的拟混池重复生成过程可以用如下的决策树进行描述</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/06/pseudobulkReplicate_DecisionTree-47b9cb7311714040afd8afcd14dee720.png" alt="pseudobulkReplicate_DecisionTree"></p>
<p>流程有些复杂，这里概述下这个流程中的几个关键注意事项。</p>
<p>首先，用户确定细胞的分组，ArchR通常会称之为聚类，我们可以在第五章分析分析得到。</p>
<p>其次，对于每一组，ArchR尝试去创建理想的拟混池重复。所谓理想的拟混池重复，指的是每个重复只有一个样本构成，这就要求每个样本得保证足量的细胞数。这样保证了重复间的样本多样性和生物学变异，是ArchR所期望得到的最佳结果，但这一过程实际上会有5种可能的结果，根据ArchR的偏好排序如下</p>
<ol>
<li>数据拥有足够多的样本(至少要等于maxRep定义的重复)，且每个样本的细胞数都大于 <code>minCell</code>, 于是每个样本都可以作为拟混池重复，每个重复的细胞都来自于同一个样本。</li>
<li>一些样本的细胞数超过<code>minCell</code>,因此能够单独形成一个重复。剩下的重复则是通过将余下的细胞进行混合，然后通过无放回抽样得到。</li>
<li>数据里没有一个样本的细胞数超过<code>minCell</code>, 但是总细胞数超过<code>minCells * minReps</code>。因此将所有的细胞进行混合，然后进行无放回抽样，抽样时不考虑细胞来源。</li>
<li>一个细胞分组中的总细胞数低于 <code>minCells * minReps</code>，但是大于<code>minCells / Sample Ratio</code>。此时单个样本的构建采取无放回抽样，重复间则需要有放回抽样，降低多个拟混池重复间的相同细胞数。</li>
<li>一个细胞分组中的总细胞数低于 <code>minCells /  Sample Ratio</code> 。这意味着我们必须在单个重复和跨重复中都采取有放回抽样策略。这是最糟糕的情况，后续在使用这些拟混池重复做下游分析分析要特别小心。后续可以通过设置ArchR的<code>minCells</code>参数进行淘汰。</li>
</ol>
<p>我们使用如下的数据集阐述这一过程</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">Sample  Cluster1  Cluster2  Cluster3  Cluster4  Cluster5</span><br><span class="line">A       800       600       900       100       75</span><br><span class="line">B       1000      50        400       150       25</span><br><span class="line">C       600       900       100       200       50</span><br><span class="line">D       1200      500       50        50        25</span><br><span class="line">E       900       100       50        150       50</span><br><span class="line">F       700       200       100       100       25</span><br></pre></td></tr></table></figure>

<p>我们设置的参数为<code>minRep=3</code>,<code>maxRep=5</code>, <code>minCells=300</code>, <code>maxCells=1000</code>和<code>sampleRatio=0.8</code>，也就是最少有3个重复，最多是5个重复，每个重复至少有300个细胞，最多是1000个细胞，当细胞数不满住要求，抽样率设置为0.8.</p>
<h3 id="9-1-1-Cluster1"><a href="#9-1-1-Cluster1" class="headerlink" title="9.1.1 Cluster1"></a>9.1.1 Cluster1</h3><p>对于Cluster1, 我们的6个样本(大于<code>maxRep</code>)的细胞数都高于<code>minCells</code>(300)。这是最理想的情况，对应上述第一种情况，我们将会得到5个拟混池重复，保证每个重复都来自独立的样本。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">Rep1 = 800 cells from SampleA</span><br><span class="line">Rep2 = 1000 cells from SampleB</span><br><span class="line">Rep3 = 1000 cells from SampleD</span><br><span class="line">Rep4 = 900 cells from SampleE</span><br><span class="line">Rep5 = 700 cells from SampleF</span><br></pre></td></tr></table></figure>

<p>对于这些重复，我们需要注意两个事情:(1) 因为我们的样本数足够多，能够保证每个重复都来自独立的样本，所以可以淘汰其中细胞数最少的SampleC。(2)由于<code>maxCells</code>设置为1000，因此最多只能有1000个细胞。</p>
<h3 id="9-1-2-Cluster2"><a href="#9-1-2-Cluster2" class="headerlink" title="9.1.2 Cluster2"></a>9.1.2 Cluster2</h3><p>对于Cluster2, 我们有3个样本的细胞数超过<code>minCells</code>, 另外3个样本的细胞数都不够。这对应上述第二种情况，我们会以如下的方法构建拟混池重复。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Rep1 = 600 cells from SampleA</span><br><span class="line">Rep2 = 900 cells from SampleC</span><br><span class="line">Rep3 = 500 cells from SampleD</span><br><span class="line">Rep4 = 350 cells [50 cells from SampleB + 100 from SampleE + 200 from SampleF]</span><br></pre></td></tr></table></figure>

<p>在这个例子中，Rep4由其他几个样本的细胞混合后通过无放回抽样得到</p>
<h3 id="9-1-3-Cluster3"><a href="#9-1-3-Cluster3" class="headerlink" title="9.1.3 Cluster3"></a>9.1.3 Cluster3</h3><p>对于Cluster3，我们只有两个样本超过<code>minCells</code>, 不满足<code>minReps</code>。但是如果我们将剩余的样本的细胞进行混合形成额外的重复，它的细胞数就超过了<code>minCells</code>。最终我们得到了3个拟混池重复，对应上述的情况3。我们将得到如下重复</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Rep1 = 900 cells from SampleA</span><br><span class="line">Rep2 = 400 cells from SampleB</span><br><span class="line">Rep3 = 250 cells [100 cells from SampleC + 50 from SampleD + 50 from SampleE + 50 from SampleF]</span><br></pre></td></tr></table></figure>

<p>和Cluster2类似，Cluster3的Rep3由其他几个样本的细胞混合后通过无放回抽样得到</p>
<h3 id="9-1-4-Cluster4"><a href="#9-1-4-Cluster4" class="headerlink" title="9.1.4 Cluster4"></a>9.1.4 Cluster4</h3><p>对于Cluster4，总细胞数是570个，小于<code>minCells * minReps</code>(900). 在这个情况下，我们无法保证有足够多的细胞通过无放回抽样的方式保证每个重复都有最小的细胞数。但是，总的细胞数依旧依旧大于<code>minCells / sampleRatio</code>(375个细胞)，这意味着每个重复中细胞可以来自于无放回抽样，重复之间的细胞需要放回抽样。这对应着上述的情况4，我们将得到如下重复</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Rep1 = 300 cells [250 unique cells + 25 cells overlapping Rep2 + 25 cells overlapping Rep3]</span><br><span class="line">Rep2 = 300 cells [250 unique cells + 25 cells overlapping Rep1 + 25 cells overlapping Rep3]</span><br><span class="line">Rep3 = 300 cells [250 unique cells + 25 cells overlapping Rep1 + 25 cells overlapping Rep2]</span><br></pre></td></tr></table></figure>

<p>在这个情况中，ArchR会尽可能降低任意两个拟混池重复的相同细胞。</p>
<h3 id="9-1-5-Cluster5"><a href="#9-1-5-Cluster5" class="headerlink" title="9.1.5 Cluster5"></a>9.1.5 Cluster5</h3><p>对于Cluster5，总共是250个细胞，同时小于<code>minCells * minReps</code>（900）和<code>minCells / sampleRatio</code>(375). 这意味着每个样本都需要有放回的抽样，重复之间也需要有放回抽样，才能得到拟混池重复。这是上述说到的第5种情况，是其中最糟糕的情况。对于这类拟混池重复，在后续的分析中需要谨慎使用。我们将得到如下的重复:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Rep1 = 300 cells [250 unique cells + 25 cells overlapping Rep2 + 25 cells overlapping Rep3]</span><br><span class="line">Rep2 = 300 cells [250 unique cells + 25 cells overlapping Rep1 + 25 cells overlapping Rep3]</span><br><span class="line">Rep3 = 300 cells [250 unique cells + 25 cells overlapping Rep1 + 25 cells overlapping Rep2]</span><br></pre></td></tr></table></figure>

<h2 id="9-2-构建拟混池重复"><a href="#9-2-构建拟混池重复" class="headerlink" title="9.2 构建拟混池重复"></a>9.2 构建拟混池重复</h2><p>通过上一节了解ArchR构建拟混池重复的逻辑后，我们就可以开始实际操作了。在ArchR中，我们通过调用<code>addGroupCoverages()</code>函数来构建拟混池重复。它的关键参数<code>groupBy</code>,定义了拟混池重复需要使用的分组。这里我们用的是上一章scRNA-seq数据标记的细胞类型，也就是<code>Cluster2</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme4 <span class="operator">&lt;-</span> addGroupCoverages<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme3<span class="punctuation">,</span> groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>得到这些拟混池重复后，我们就能从数据中鉴定peak了。就像之前所说的，我们不希望使用所有的细胞鉴定peak，而是单独根据每一组细胞（例如聚类）单独鉴定peak，这样才有可能分析出不同组的特异性peak。这一章得到数据就为后续鉴定peak提供了良好的开始。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/06/04/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B9%9D%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter9/" data-id="clm1ywdjj00bhecok5b931myu" data-title="使用ArchR分析单细胞ATAC-seq数据(第九章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第八章)-analysis-sc-atac-seq-with-archr-chapter8" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/05/31/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%85%AB%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter8/" class="article-date">
  <time class="dt-published" datetime="2020-05-31T22:27:31.103Z" itemprop="datePublished">2020-05-31</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/05/31/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%85%AB%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter8/">使用ArchR分析单细胞ATAC-seq数据(第八章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第8章-使用scRNA-seq定义cluster类型"><a href="#第8章-使用scRNA-seq定义cluster类型" class="headerlink" title="第8章: 使用scRNA-seq定义cluster类型"></a>第8章: 使用scRNA-seq定义cluster类型</h1><p>除了使用基因得分定义细胞类群以外，ArchR还能整合scRNA-seq数据。通过将scATAC-seq数据里的基因得分矩阵和scRNA-seq数据的基因表达量矩阵进行对比，ArchR就能将scATAC-seq的细胞比对到scRNA-seq的细胞，实现两种数据的整合。之后，我们借助scRNA-seq数据已经定义的细胞类群，或者整合后的scRNA-seq的基因表达量来注释细胞类群。在代码内部，我们调用了<code>Seurat::FindTransferAnchors()</code>，从而实现两种数据集之间的比较。当然，ArchR不是简单地对函数进行封装，而是在此基础上通过对数据的拆分，实现并行计算，从而能将该流程应用到更大规模的细胞中。</p>
<img src="/upload/2020/05/image-296524df11944c0e8571c05b932fb1fa.png" alt="img" style="zoom:10%;" />

<p>该整合过程实际上会寻找scATAC-seq和scRNA-seq两者中最相似的细胞，然后将scRNA-seq中对应的细胞表达量赋值给scATAAC-seq细胞。最后，scATAC-seq中每个细胞都会有基因表达量特征，能被用于下游分析。这一章会阐述如何利用该信息定义细胞类型，之后会介绍如何使用连接的scRNA-seq数据做更加复杂的分析，例如识别预测的顺式调控元件。我们相信由于多组学单细胞谱的商业化，这类整合分析将会越来越多。同时，在ArchR中使用公共数据里匹配细胞类型的scRNA-seq数据或者自己使用目标样本得到的scRNA-seq数据也能加强scATAC-seq分析。</p>
<h2 id="8-1-scATAC-seq细胞和scRNA-seq细胞跨平台连接"><a href="#8-1-scATAC-seq细胞和scRNA-seq细胞跨平台连接" class="headerlink" title="8.1 scATAC-seq细胞和scRNA-seq细胞跨平台连接"></a>8.1 scATAC-seq细胞和scRNA-seq细胞跨平台连接</h2><p>为了能将我们教程中的scATAC-seq数据与其匹配的scRNA-seq数据进行整合，我们将使用 Granja* et al (2019) 里的造血细胞scRNA-seq数据。</p>
<p>scRNA-seq数据以 <code>RangedSummarizedExperiment</code>对象保存，大小为111MB。此外，ArchR还接受未经修改的<code>Seurat</code>对象作为整合流程的输入。我们使用<code>download.file</code>下载数据</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span><span class="punctuation">(</span><span class="operator">!</span>file.exists<span class="punctuation">(</span><span class="string">&quot;scRNA-Hematopoiesis-Granja-2019.rds&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    download.file<span class="punctuation">(</span></span><br><span class="line">        url <span class="operator">=</span> <span class="string">&quot;https://jeffgranja.s3.amazonaws.com/ArchR/TestData/scRNA-Hematopoiesis-Granja-2019.rds&quot;</span><span class="punctuation">,</span></span><br><span class="line">        destfile <span class="operator">=</span> <span class="string">&quot;scRNA-Hematopoiesis-Granja-2019.rds&quot;</span></span><br><span class="line">    <span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>下载之后，我们使用<code>readRDS</code>进行读取，并查看该对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">seRNA <span class="operator">&lt;-</span> readRDS<span class="punctuation">(</span><span class="string">&quot;scRNA-Hematopoiesis-Granja-2019.rds&quot;</span><span class="punctuation">)</span></span><br><span class="line">seRNA</span><br><span class="line"><span class="comment">## class: RangedSummarizedExperiment</span></span><br><span class="line"><span class="comment">## dim: 20287 35582</span></span><br><span class="line"><span class="comment">## metadata(0):</span></span><br><span class="line"><span class="comment">## assays(1): counts</span></span><br><span class="line"><span class="comment">## rownames(20287): FAM138A OR4F5 … S100B PRMT2</span></span><br><span class="line"><span class="comment">## rowData names(3): gene_name gene_id exonLength</span></span><br><span class="line"><span class="comment">## colnames(35582): CD34_32_R5:AAACCTGAGTATCGAA-1</span></span><br><span class="line"><span class="comment">## CD34_32_R5:AAACCTGAGTCGTTTG-1 …</span></span><br><span class="line"><span class="comment">## BMMC_10x_GREENLEAF_REP2:TTTGTTGCATGTGTCA-1</span></span><br><span class="line"><span class="comment">## BMMC_10x_GREENLEAF_REP2:TTTGTTGCATTGAAAG-1</span></span><br><span class="line"><span class="comment">## colData names(10): Group nUMI_pre … BioClassification Barcode</span></span><br></pre></td></tr></table></figure>

<p>从输出信息中，我们可以发现它里面有基因表达量的count矩阵和对应的元信息。</p>
<p>元信息列里的<code>BioClassification</code>记录着scRNA-seq数据中每个细胞对应的细胞类型分类</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">colnames<span class="punctuation">(</span>colData<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># [1] &quot;Group&quot;             &quot;nUMI_pre&quot;          &quot;nUMI&quot;             </span></span><br><span class="line"><span class="comment"># [4] &quot;nGene&quot;             &quot;initialClusters&quot;   &quot;UMAP1&quot;            </span></span><br><span class="line"><span class="comment"># [7] &quot;UMAP2&quot;             &quot;Clusters&quot;          &quot;BioClassification&quot;</span></span><br><span class="line"><span class="comment"># [10] &quot;Barcode&quot;</span></span><br></pre></td></tr></table></figure>

<p>使用<code>table()</code>，我们可以看到scRNA-seq细胞类型每一群的细胞数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">table<span class="punctuation">(</span>colData<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="operator">$</span>BioClassification<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#        01_HSC 02_Early.Eryth  03_Late.Eryth  04_Early.Baso    05_CMP.LMPP </span></span><br><span class="line"><span class="comment">#          1425           1653            446            111           2260 </span></span><br><span class="line"><span class="comment">#      06_CLP.1         07_GMP    08_GMP.Neut         09_pDC         10_cDC </span></span><br><span class="line"><span class="comment">#           903           2097           1050            544            325 </span></span><br><span class="line"><span class="comment">#11_CD14.Mono.1 12_CD14.Mono.2   13_CD16.Mono         14_Unk       15_CLP.2 </span></span><br><span class="line"><span class="comment">#          1800           4222            292            520            377 </span></span><br><span class="line"><span class="comment">#      16_Pre.B           17_B      18_Plasma       19_CD8.N      20_CD4.N1 </span></span><br><span class="line"><span class="comment">#           710           1711             62           1521           2470 </span></span><br><span class="line"><span class="comment">#     21_CD4.N2       22_CD4.M      23_CD8.EM      24_CD8.CM          25_NK </span></span><br><span class="line"><span class="comment">#          2364           3539            796           2080           2143 </span></span><br><span class="line"><span class="comment">#        26_Unk </span></span><br><span class="line"><span class="comment">#           161 </span></span><br></pre></td></tr></table></figure>

<p>我们后续会用到两种整合方法。第一种是<strong>无约束整合</strong>，我们对scATAC-seq实验里的细胞不作任何假设，直接尝试将这些细胞和scRNA-seq实验里的任意细胞进行配对。这是一种初步可行方案，后续会根据这一步得到的结果，对整合步骤进行约束来提升跨平台配对的质量。第二种方法是<strong>约束整合</strong>，即利用对细胞类型的先验知识限制搜索范围。举个例子，如果我们知道scATAC-seq中的Cluster A,B,C对应着三种不同的T细胞，scRNA-seq中的Cluster X,Y对应着两种不同的T细胞，我们告诉ArchR只需要尝试将scATAC-seq中的Cluster A,B,C跟scRNA-seq中的Cluster X,Y进行配对。下面，我们将先以无约束整合初步地鉴定每一种聚类的类型，然后根据分析结果做更加细致的约束整合。</p>
<h3 id="8-1-1-无约束整合"><a href="#8-1-1-无约束整合" class="headerlink" title="8.1.1 无约束整合"></a>8.1.1 无约束整合</h3><p>我们使用<code>addGeneIntegrationMatrix()</code>对scATAC-seq和scRNA-seq数据进行整合。正如之前所提到的，该函数的<code>seRNA</code>参数接受<code>Seurat</code>或<code>RangedSummarizedExperiment</code>对象作为输入。因为第一轮是探索性质的无约束整合，因此，我们不会将结果保存在Arrow文件中(<code>addToArrow=FALSE</code>)。整合后的矩阵将会根据<code>matrixName</code>进行命名，存放在<code>ArchRProject</code>中。该函数的其他参数对应<code>cellColData</code>中列名用于存放额外的信息，<code>nameCell</code>存放scRNA-seq中匹配的细胞ID，<code>nameGroup</code>存放scRNA-seq细胞中的分组ID，<code>nameScore</code>存放跨平台整合得分。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addGeneIntegrationMatrix<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    useMatrix <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    matrixName <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span></span><br><span class="line">    seRNA <span class="operator">=</span> seRNA<span class="punctuation">,</span></span><br><span class="line">    addToArrow <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">    groupRNA <span class="operator">=</span> <span class="string">&quot;BioClassification&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameCell <span class="operator">=</span> <span class="string">&quot;predictedCell_Un&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameGroup <span class="operator">=</span> <span class="string">&quot;predictedGroup_Un&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameScore <span class="operator">=</span> <span class="string">&quot;predictedScore_Un&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>无约束整合的结果可能并不准确，但是为后续更加精细的约束分析奠定了基础。</p>
<h3 id="8-1-2-约束整合"><a href="#8-1-2-约束整合" class="headerlink" title="8.1.2 约束整合"></a>8.1.2 约束整合</h3><p>现在我们已经有了初步的无约束整合结果，我们就有了大致的细胞类型分布情况，接着就是优化整合结果。</p>
<p>因为我们教程里的数据来自于造血细胞，我们将会在理想状态下将类似的细胞整合在一起。首先，我们先确认scRNA-seq里细胞类型在我们的scATAC-seq聚类中分布情况。这一步的目标是使用无约束整合的方法找到scATAC-seq和scRNA-seq中和T细胞和NK细胞对应的聚类，后续会用到该信息进行约束整合。具体操作为，我们创建一个<code>confusionMatrix</code>,并关注<code>Cluster</code>和<code>predictedGroup_Un</code>的交叉部分中scRNA-seq的细胞类型。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cM <span class="operator">&lt;-</span> as.matrix<span class="punctuation">(</span>confusionMatrix<span class="punctuation">(</span>projHeme2<span class="operator">$</span>Clusters<span class="punctuation">,</span> projHeme2<span class="operator">$</span>predictedGroup_Un<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">preClust <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>cM<span class="punctuation">)</span><span class="punctuation">[</span>apply<span class="punctuation">(</span>cM<span class="punctuation">,</span> <span class="number">1</span> <span class="punctuation">,</span> which.max<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">cbind<span class="punctuation">(</span>preClust<span class="punctuation">,</span> rownames<span class="punctuation">(</span>cM<span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">#Assignments</span></span><br></pre></td></tr></table></figure>

<p>输出信息如下，展示了12个scATAC-seq聚类中对应最优可能的scRNA-seq细胞类型。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#      preClust              </span></span><br><span class="line"><span class="comment"># [1,] &quot;03_Late.Eryth&quot;  &quot;C10&quot;</span></span><br><span class="line"><span class="comment"># [2,] &quot;20_CD4.N1&quot;      &quot;C8&quot; </span></span><br><span class="line"><span class="comment"># [3,] &quot;16_Pre.B&quot;       &quot;C3&quot; </span></span><br><span class="line"><span class="comment"># [4,] &quot;08_GMP.Neut&quot;    &quot;C11&quot;</span></span><br><span class="line"><span class="comment"># [5,] &quot;17_B&quot;           &quot;C4&quot; </span></span><br><span class="line"><span class="comment"># [6,] &quot;11_CD14.Mono.1&quot; &quot;C1&quot; </span></span><br><span class="line"><span class="comment"># [7,] &quot;01_HSC&quot;         &quot;C12&quot;</span></span><br><span class="line"><span class="comment"># [8,] &quot;22_CD4.M&quot;       &quot;C9&quot; </span></span><br><span class="line"><span class="comment"># [9,] &quot;09_pDC&quot;         &quot;C5&quot; </span></span><br><span class="line"><span class="comment"># [10,] &quot;25_NK&quot;          &quot;C7&quot; </span></span><br><span class="line"><span class="comment"># [11,] &quot;12_CD14.Mono.2&quot; &quot;C2&quot; </span></span><br><span class="line"><span class="comment"># [12,] &quot;06_CLP.1&quot;       &quot;C6&quot;</span></span><br></pre></td></tr></table></figure>

<p>首先，我们检查在无约束整合中用到的scRNA-seq数据里细胞类型标签。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">unique<span class="punctuation">(</span>unique<span class="punctuation">(</span>projHeme2<span class="operator">$</span>predictedGroup_Un<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># [1] &quot;08_GMP.Neut&quot;    &quot;25_NK&quot;          &quot;16_Pre.B&quot;       &quot;06_CLP.1&quot;      </span></span><br><span class="line"><span class="comment"># [5] &quot;07_GMP&quot;         &quot;11_CD14.Mono.1&quot; &quot;04_Early.Baso&quot;  &quot;22_CD4.M&quot;      </span></span><br><span class="line"><span class="comment"># [9] &quot;03_Late.Eryth&quot;  &quot;05_CMP.LMPP&quot;    &quot;17_B&quot;           &quot;19_CD8.N&quot;      </span></span><br><span class="line"><span class="comment">#[13] &quot;09_pDC&quot;         &quot;13_CD16.Mono&quot;   &quot;23_CD8.EM&quot;      &quot;12_CD14.Mono.2&quot;</span></span><br><span class="line"><span class="comment">#[17] &quot;20_CD4.N1&quot;      &quot;02_Early.Eryth&quot; &quot;21_CD4.N2&quot;      &quot;24_CD8.CM&quot;     </span></span><br><span class="line"><span class="comment">#[21] &quot;01_HSC&quot; </span></span><br></pre></td></tr></table></figure>

<p>从上面的输出中，我们发现scRNA-seq数据中与NK细胞和T细胞对应的聚类是Cluster 19 - 25。</p>
<p>接着我们创建一个字符串模式用来表示这些聚类，后续的约束整合会用到，其中<code>|</code>在正则表达式中表示”或”，我们之后使用<code>grep</code>根据这些字符串模式从scATAC-seq提取和scRNA-seq对应的聚类。。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#From scRNA</span></span><br><span class="line">cTNK <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span>paste0<span class="punctuation">(</span><span class="number">19</span><span class="operator">:</span><span class="number">25</span><span class="punctuation">)</span><span class="punctuation">,</span> collapse<span class="operator">=</span><span class="string">&quot;|&quot;</span><span class="punctuation">)</span></span><br><span class="line">cTNK</span><br><span class="line"><span class="comment"># [1] &quot;19|20|21|22|23|24|25&quot;</span></span><br></pre></td></tr></table></figure>

<p>其余的聚类就称之为”Non-T cell, Non-NK cell”(例如Cluster 1 - 18)，也创建了对应的字符串模式</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cNonTNK <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span>paste0<span class="punctuation">(</span><span class="string">&quot;0&quot;</span><span class="punctuation">,</span> <span class="number">1</span><span class="operator">:</span><span class="number">9</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">10</span><span class="operator">:</span><span class="number">13</span><span class="punctuation">,</span> <span class="number">15</span><span class="operator">:</span><span class="number">18</span><span class="punctuation">)</span><span class="punctuation">,</span> collapse<span class="operator">=</span><span class="string">&quot;|&quot;</span><span class="punctuation">)</span></span><br><span class="line">cNonTNK</span><br><span class="line"><span class="comment">#[1] &quot;01|02|03|04|05|06|07|08|09|10|11|12|13|15|16|17|18&quot;</span></span><br></pre></td></tr></table></figure>

<p>接着再用字符串模式在<code>preClust</code>找到对应的scATAC-seq列名，然后使用列名从混合矩阵提取对应的列。</p>
<p>对于T细胞和NK细胞，scATAC-seq聚类ID就是C7, C8, C9</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#Assign scATAC to these categories</span></span><br><span class="line">clustTNK <span class="operator">&lt;-</span> rownames<span class="punctuation">(</span>cM<span class="punctuation">)</span><span class="punctuation">[</span>grep<span class="punctuation">(</span>cTNK<span class="punctuation">,</span> preClust<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">clustTNK</span><br><span class="line"><span class="comment">#[1] &quot;C8&quot; &quot;C9&quot; &quot;C7&quot;</span></span><br></pre></td></tr></table></figure>

<p>对于” Non-T cells and Non-NK cells”, ID就是scATAC-seq聚类余下的部分</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">clustNonTNK <span class="operator">&lt;-</span> rownames<span class="punctuation">(</span>cM<span class="punctuation">)</span><span class="punctuation">[</span>grep<span class="punctuation">(</span>cNonTNK<span class="punctuation">,</span> preClust<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">clustNonTNK</span><br><span class="line"><span class="comment"># [1] &quot;C10&quot; &quot;C3&quot;  &quot;C11&quot; &quot;C4&quot;  &quot;C1&quot;  &quot;C12&quot; &quot;C5&quot;  &quot;C2&quot;  &quot;C6&quot; </span></span><br></pre></td></tr></table></figure>

<p>接着在scRNA-seq中做相同的操作，筛选出相同的细胞类型。首先，我们鉴定scRNA-seq数据中T细胞和NK细胞</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#RNA get cells in these categories</span></span><br><span class="line">rnaTNK <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="punctuation">[</span>grep<span class="punctuation">(</span>cTNK<span class="punctuation">,</span> colData<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="operator">$</span>BioClassification<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">head<span class="punctuation">(</span>rnaTNK<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#[1] &quot;PBMC_10x_GREENLEAF_REP1:AAACCCAGTCGTCATA-1&quot;</span></span><br><span class="line"><span class="comment">#[2] &quot;PBMC_10x_GREENLEAF_REP1:AAACCCATCCGATGTA-1&quot;</span></span><br><span class="line"><span class="comment">#[3] &quot;PBMC_10x_GREENLEAF_REP1:AAACCCATCTCAACGA-1&quot;</span></span><br><span class="line"><span class="comment">#[4] &quot;PBMC_10x_GREENLEAF_REP1:AAACCCATCTCTCGAC-1&quot;</span></span><br><span class="line"><span class="comment">#[5] &quot;PBMC_10x_GREENLEAF_REP1:AAACGAACAATCGTCA-1&quot;</span></span><br><span class="line"><span class="comment">#[6] &quot;PBMC_10x_GREENLEAF_REP1:AAACGAACACGATTCA-1&quot;</span></span><br></pre></td></tr></table></figure>

<p>然后，鉴定scRNA-seq数据中”Non-T cell Non-NK cell cells”</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">rnaNonTNK <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="punctuation">[</span>grep<span class="punctuation">(</span>cNonTNK<span class="punctuation">,</span> colData<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="operator">$</span>BioClassification<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">head<span class="punctuation">(</span>rnaNonTNK<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#[1] &quot;CD34_32_R5:AAACCTGAGTATCGAA-1&quot; &quot;CD34_32_R5:AAACCTGAGTCGTTTG-1&quot;</span></span><br><span class="line"><span class="comment">#[3] &quot;CD34_32_R5:AAACCTGGTTCCACAA-1&quot; &quot;CD34_32_R5:AAACGGGAGCTTCGCG-1&quot;</span></span><br><span class="line"><span class="comment">#[5] &quot;CD34_32_R5:AAACGGGAGGGAGTAA-1&quot; &quot;CD34_32_R5:AAACGGGAGTTACGGG-1&quot;</span></span><br></pre></td></tr></table></figure>

<p>约束整合需要我们提供一个嵌套list。这是一个由多个<code>SimpleList</code>对象组成的<code>SimpleList</code>, 每一组对应一个约束情况。在案例中，我们有两个组，一个组称之为<code>TNK</code>,包括两个平台的T和NK细胞，另一个组为<code>NonTNK</code>，包括两个平台的”Non-T cell Non-NK cell”细胞。每个<code>SimpleList</code>对象都包含两个细胞ID的向量，一个是ATAC，一个是RNA.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">groupList <span class="operator">&lt;-</span> SimpleList<span class="punctuation">(</span></span><br><span class="line">    TNK <span class="operator">=</span> SimpleList<span class="punctuation">(</span></span><br><span class="line">        ATAC <span class="operator">=</span> projHeme2<span class="operator">$</span>cellNames<span class="punctuation">[</span>projHeme2<span class="operator">$</span>Clusters <span class="operator">%in%</span> clustTNK<span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">        RNA <span class="operator">=</span> rnaTNK</span><br><span class="line">    <span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    NonTNK <span class="operator">=</span> SimpleList<span class="punctuation">(</span></span><br><span class="line">        ATAC <span class="operator">=</span> projHeme2<span class="operator">$</span>cellNames<span class="punctuation">[</span>projHeme2<span class="operator">$</span>Clusters <span class="operator">%in%</span> clustNonTNK<span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">        RNA <span class="operator">=</span> rnaNonTNK</span><br><span class="line">    <span class="punctuation">)</span>    </span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们将该列表传递给<code>addGeneIntegrationMatrix()</code>函数的<code>groupList</code>参数。注意，我们依旧没有将结果添加到Arrow文件中 (<code>addToArrow = FALSE</code>)。我们强烈建议，在保存到Arrow文件前先彻底的检查结果，看结果是否符合预期。在教程的下一节会介绍该操作。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addGeneIntegrationMatrix<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    useMatrix <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    matrixName <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span></span><br><span class="line">    seRNA <span class="operator">=</span> seRNA<span class="punctuation">,</span></span><br><span class="line">    addToArrow <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> </span><br><span class="line">    groupList <span class="operator">=</span> groupList<span class="punctuation">,</span></span><br><span class="line">    groupRNA <span class="operator">=</span> <span class="string">&quot;BioClassification&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameCell <span class="operator">=</span> <span class="string">&quot;predictedCell_Co&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameGroup <span class="operator">=</span> <span class="string">&quot;predictedGroup_Co&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameScore <span class="operator">=</span> <span class="string">&quot;predictedScore_Co&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="8-1-3-约束整合和无约束整合对比"><a href="#8-1-3-约束整合和无约束整合对比" class="headerlink" title="8.1.3 约束整合和无约束整合对比"></a>8.1.3 约束整合和无约束整合对比</h3><p>正如之前所提到的，我们的scATAC-seq数据可以根据整合的scRNA-seq数据进行定义，并且有约束和无约束这两种方式。为了对两者进行对比，我们分别根据这两种整合结果对scATAC-seq的数据进行上色。</p>
<p>首先，使用ArchR内置的<code>paletteDiscrete()</code>函数生成调色板</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">pal <span class="operator">&lt;-</span> paletteDiscrete<span class="punctuation">(</span>values <span class="operator">=</span> colData<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="operator">$</span>BioClassification<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>在ArchR中，调色板本质上一个命名向量，每个十六进制编码的颜色对应着一个名字。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">pal</span><br><span class="line"><span class="comment">#        01_HSC 02_Early.Eryth  03_Late.Eryth  04_Early.Baso    05_CMP.LMPP </span></span><br><span class="line"><span class="comment">#     &quot;#D51F26&quot;      &quot;#502A59&quot;      &quot;#235D55&quot;      &quot;#3D6E57&quot;      &quot;#8D2B8B&quot; </span></span><br><span class="line"><span class="comment">#      06_CLP.1         07_GMP    08_GMP.Neut         09_pDC         10_cDC </span></span><br><span class="line"><span class="comment">#     &quot;#DE6C3E&quot;      &quot;#F9B712&quot;      &quot;#D8CE42&quot;      &quot;#8E9ACD&quot;      &quot;#B774B1&quot; </span></span><br><span class="line"><span class="comment">#11_CD14.Mono.1 12_CD14.Mono.2   13_CD16.Mono         14_Unk       15_CLP.2 </span></span><br><span class="line"><span class="comment">#     &quot;#D69FC8&quot;      &quot;#C7C8DE&quot;      &quot;#8FD3D4&quot;      &quot;#89C86E&quot;      &quot;#CC9672&quot; </span></span><br><span class="line"><span class="comment">#      16_Pre.B           17_B      18_Plasma       19_CD8.N      20_CD4.N1 </span></span><br><span class="line"><span class="comment">#     &quot;#CF7E96&quot;      &quot;#A27AA4&quot;      &quot;#CD4F32&quot;      &quot;#6B977E&quot;      &quot;#518AA3&quot; </span></span><br><span class="line"><span class="comment">#     21_CD4.N2       22_CD4.M      23_CD8.EM      24_CD8.CM          25_NK </span></span><br><span class="line"><span class="comment">#     &quot;#5A5297&quot;      &quot;#0F707D&quot;      &quot;#5E2E32&quot;      &quot;#A95A3C&quot;      &quot;#B28D5C&quot; </span></span><br><span class="line"><span class="comment">#        26_Unk </span></span><br><span class="line"><span class="comment">#     &quot;#3D3D3D&quot;</span></span><br></pre></td></tr></table></figure>

<p>我们在scATAC-seq数据根据无约束整合得到的scRNA-seq细胞类型进行可视化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span></span><br><span class="line">    projHeme2<span class="punctuation">,</span> </span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;predictedGroup_Un&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    pal <span class="operator">=</span> pal</span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">p1</span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-94b3538cffe94b2d9d2945ae01429b89.png" alt="Plot-UMAP-RNA-Integration_1" style="zoom:25%;" />

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">do.call<span class="punctuation">(</span>cowplot<span class="operator">::</span>plot_grid<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">list</span><span class="punctuation">(</span>ncol <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span> p2c<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>同样，我们也可以根据约束整合得到scATAC-seq对应的scRNA-seq的细胞类型进行可视化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">p2 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span></span><br><span class="line">    projHeme2<span class="punctuation">,</span> </span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;predictedGroup_Co&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    pal <span class="operator">=</span> pal</span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">p2</span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-a8c23615cd644c859ac5e80b45c25230.png" alt="Plot-UMAP-RNA-Integration_2" style="zoom:25%;" />

<p>这两者的结果差异其实非常细微，主要是我们感兴趣的细胞类型原本就存在明显的差异。当然，仔细观察还能发现其中的不同之处，尤其是T细胞(Clusters 17-22)</p>
<p>我们用<code>plotPDF()</code>函数保存该图矢量版本。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span>p2<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-UMAP-RNA-Integration.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们现在可以用<code>saveArchRProject()</code>函数保存我们的<code>projHeme2</code>对象。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">saveArchRProject<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> outputDirectory <span class="operator">=</span> <span class="string">&quot;Save-ProjHeme2&quot;</span><span class="punctuation">,</span> load <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="8-2-为每个scATAC-seq细胞增加拟scRNA-seq谱"><a href="#8-2-为每个scATAC-seq细胞增加拟scRNA-seq谱" class="headerlink" title="8.2 为每个scATAC-seq细胞增加拟scRNA-seq谱"></a>8.2 为每个scATAC-seq细胞增加拟scRNA-seq谱</h2><p>既然我们对scATAC-seq和scRNA-seq整合的结果感到满意，我们就能用<code>addToArrow=TRUE</code>重新运行，在Arrow文件中添加相关联的表达量矩阵数据。根据之前所提到的，我们传入<code>groupList</code>约束整合，在<code>nameCell</code>，<code>nameGroup</code>和<code>nameScore</code>中加入列名。这些元信息列都会被添加到<code>cellColData</code>中。</p>
<p>这里，我们新建了一个<code>projHeme3</code>，用于后续教程。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#~5 minutes</span></span><br><span class="line">projHeme3 <span class="operator">&lt;-</span> addGeneIntegrationMatrix<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    useMatrix <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    matrixName <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span></span><br><span class="line">    seRNA <span class="operator">=</span> seRNA<span class="punctuation">,</span></span><br><span class="line">    addToArrow <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span></span><br><span class="line">    force<span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span></span><br><span class="line">    groupList <span class="operator">=</span> groupList<span class="punctuation">,</span></span><br><span class="line">    groupRNA <span class="operator">=</span> <span class="string">&quot;BioClassification&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameCell <span class="operator">=</span> <span class="string">&quot;predictedCell&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameGroup <span class="operator">=</span> <span class="string">&quot;predictedGroup&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameScore <span class="operator">=</span> <span class="string">&quot;predictedScore&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>现在，当我们使用<code>getAvailableMatrices()</code>检查哪些矩阵可用时，我们会发现<code>GeneIntegrationMatrix</code>已经被添加到Arrow文件中</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">getAvailableMatrices<span class="punctuation">(</span>projHeme3<span class="punctuation">)</span></span><br><span class="line"><span class="comment"># [1] &quot;GeneIntegrationMatrix&quot; &quot;GeneScoreMatrix&quot; &quot;TileMatrix&quot;</span></span><br></pre></td></tr></table></figure>

<p>在新的<code>GeneIntegrationMatrix</code>中，我们可以比较连接的基因表达量和根据基因得分推断的基因表达量</p>
<p>我们需要先确保在项目中加入了填充权重值(impute weights)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme3 <span class="operator">&lt;-</span> addImputeWeights<span class="punctuation">(</span>projHeme3<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>现在，我们来生成一些UMAP图，里面的<strong>基因表达量</strong>值来自于<code>GeneIntegrationMatrix</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;CD34&quot;</span><span class="punctuation">,</span> <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span><span class="punctuation">,</span> <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MS4A1&quot;</span><span class="punctuation">,</span> <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span><span class="punctuation">,</span> <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD8A&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TBX21&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IL7R&quot;</span> <span class="comment">#TCells</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p1 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme3<span class="punctuation">,</span> </span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> markerGenes<span class="punctuation">,</span> </span><br><span class="line">    continuousSet <span class="operator">=</span> <span class="string">&quot;horizonExtra&quot;</span><span class="punctuation">,</span></span><br><span class="line">    embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    imputeWeights <span class="operator">=</span> getImputeWeights<span class="punctuation">(</span>projHeme3<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>以及一些相同UMAP图，但是使用<code>GeneScoreMatrix</code>里的<strong>基因得分</strong>值</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">p2 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme3<span class="punctuation">,</span> </span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    continuousSet <span class="operator">=</span> <span class="string">&quot;horizonExtra&quot;</span><span class="punctuation">,</span></span><br><span class="line">    name <span class="operator">=</span> markerGenes<span class="punctuation">,</span> </span><br><span class="line">    embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    imputeWeights <span class="operator">=</span> getImputeWeights<span class="punctuation">(</span>projHeme3<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>最后用<code>cowplot</code>将这些标记基因绘制在一起</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">p1c <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>p1<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    x <span class="operator">+</span> guides<span class="punctuation">(</span>color <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> fill <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">    theme_ArchR<span class="punctuation">(</span>baseSize <span class="operator">=</span> <span class="number">6.5</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span>plot.margin <span class="operator">=</span> unit<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;cm&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span></span><br><span class="line">        axis.text.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.text.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">    <span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p2c <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>p2<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    x <span class="operator">+</span> guides<span class="punctuation">(</span>color <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> fill <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">    theme_ArchR<span class="punctuation">(</span>baseSize <span class="operator">=</span> <span class="number">6.5</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span>plot.margin <span class="operator">=</span> unit<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;cm&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span></span><br><span class="line">        axis.text.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.text.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">    <span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">do.call<span class="punctuation">(</span>cowplot<span class="operator">::</span>plot_grid<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">list</span><span class="punctuation">(</span>ncol <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span> p1c<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-4ed685c296af4038bbb7c47852bfbf1e.png" alt="Plot-UMAP-Markers-RNA-W-Imputation_1" style="zoom:50%;" />

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">do.call<span class="punctuation">(</span>cowplot<span class="operator">::</span>plot_grid<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">list</span><span class="punctuation">(</span>ncol <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span> p2c<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-837cacbfb4534818b998b844551299dd.png" alt="Plot-UMAP-Markers-RNA-W-Imputation_2" style="zoom:50%;" />

<p>和预期的一样，两个方法推测的基因表达量存在相似性，但并不相同。</p>
<p>使用<code>plotPDF()</code>函数保存可编辑的矢量版。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>plotList <span class="operator">=</span> p1<span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;Plot-UMAP-Marker-Genes-RNA-W-Imputation.pdf&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme3<span class="punctuation">,</span> </span><br><span class="line">    addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="8-3-使用scRNA-seq信息标记scATAC-seq聚类"><a href="#8-3-使用scRNA-seq信息标记scATAC-seq聚类" class="headerlink" title="8.3 使用scRNA-seq信息标记scATAC-seq聚类"></a>8.3 使用scRNA-seq信息标记scATAC-seq聚类</h2><p>现在，我们确定了scATAC-seq和scRNA-seq数据间的对应关系，我们就可以使用scRNA-seq数据中细胞类型对我们的scATAC-seq聚类进行定义。</p>
<p>首先，我们会在scATAC-seq和整合分析得到<code>predictedGroup</code>之间构建一个混合矩阵</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cM <span class="operator">&lt;-</span> confusionMatrix<span class="punctuation">(</span>projHeme3<span class="operator">$</span>Clusters<span class="punctuation">,</span> projHeme3<span class="operator">$</span>predictedGroup<span class="punctuation">)</span></span><br><span class="line">labelOld <span class="operator">&lt;-</span> rownames<span class="punctuation">(</span>cM<span class="punctuation">)</span></span><br><span class="line">labelOld</span><br><span class="line"><span class="comment">#  [1] &quot;Cluster11&quot; &quot;Cluster2&quot;  &quot;Cluster12&quot; &quot;Cluster1&quot;  &quot;Cluster8&quot;  &quot;Cluster4&quot; </span></span><br><span class="line"><span class="comment">#  [7] &quot;Cluster9&quot;  &quot;Cluster5&quot;  &quot;Cluster7&quot;  &quot;Cluster14&quot; &quot;Cluster3&quot;  &quot;Cluster10&quot;</span></span><br><span class="line"><span class="comment"># [13] &quot;Cluster6&quot;  &quot;Cluster13&quot;</span></span><br></pre></td></tr></table></figure>

<p>接着，对于每一个scATAC-seq聚类，我们根据<code>predictedGroup</code>确定最能定义聚类的细胞类型。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">labelNew <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>cM<span class="punctuation">)</span><span class="punctuation">[</span>apply<span class="punctuation">(</span>cM<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> which.max<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">labelNew</span><br></pre></td></tr></table></figure>

<p>接着，我们需要对新的聚类标签进行重命名，简化分类系统。对于每一个scRNA-seq的聚类，我们会重新定义标签，以便更好解释。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">remapClust <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;01_HSC&quot;</span> <span class="operator">=</span> <span class="string">&quot;Progenitor&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;02_Early.Eryth&quot;</span> <span class="operator">=</span> <span class="string">&quot;Erythroid&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;03_Late.Eryth&quot;</span> <span class="operator">=</span> <span class="string">&quot;Erythroid&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;04_Early.Baso&quot;</span> <span class="operator">=</span> <span class="string">&quot;Basophil&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;05_CMP.LMPP&quot;</span> <span class="operator">=</span> <span class="string">&quot;Progenitor&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;06_CLP.1&quot;</span> <span class="operator">=</span> <span class="string">&quot;CLP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;07_GMP&quot;</span> <span class="operator">=</span> <span class="string">&quot;GMP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;08_GMP.Neut&quot;</span> <span class="operator">=</span> <span class="string">&quot;GMP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;09_pDC&quot;</span> <span class="operator">=</span> <span class="string">&quot;pDC&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;10_cDC&quot;</span> <span class="operator">=</span> <span class="string">&quot;cDC&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;11_CD14.Mono.1&quot;</span> <span class="operator">=</span> <span class="string">&quot;Mono&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;12_CD14.Mono.2&quot;</span> <span class="operator">=</span> <span class="string">&quot;Mono&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;13_CD16.Mono&quot;</span> <span class="operator">=</span> <span class="string">&quot;Mono&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;15_CLP.2&quot;</span> <span class="operator">=</span> <span class="string">&quot;CLP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;16_Pre.B&quot;</span> <span class="operator">=</span> <span class="string">&quot;PreB&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;17_B&quot;</span> <span class="operator">=</span> <span class="string">&quot;B&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;18_Plasma&quot;</span> <span class="operator">=</span> <span class="string">&quot;Plasma&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;19_CD8.N&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD8.N&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;20_CD4.N1&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD4.N&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;21_CD4.N2&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD4.N&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;22_CD4.M&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD4.M&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;23_CD8.EM&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD8.EM&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;24_CD8.CM&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD8.CM&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;25_NK&quot;</span> <span class="operator">=</span> <span class="string">&quot;NK&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">remapClust <span class="operator">&lt;-</span> remapClust<span class="punctuation">[</span><span class="built_in">names</span><span class="punctuation">(</span>remapClust<span class="punctuation">)</span> <span class="operator">%in%</span> labelNew<span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<p>接着，使用<code>mapLables()</code>函数进行标签转换，将旧的标签映射到新的标签上。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">labelNew2 <span class="operator">&lt;-</span> mapLabels<span class="punctuation">(</span>labelNew<span class="punctuation">,</span> oldLabels <span class="operator">=</span> <span class="built_in">names</span><span class="punctuation">(</span>remapClust<span class="punctuation">)</span><span class="punctuation">,</span> newLabels <span class="operator">=</span> remapClust<span class="punctuation">)</span></span><br><span class="line">labelNew2</span><br><span class="line"><span class="comment">#  [1] &quot;GMP&quot;        &quot;B&quot;          &quot;PreB&quot;       &quot;CD4.N&quot;      &quot;Mono&quot;      </span></span><br><span class="line"><span class="comment">#  [6] &quot;Erythroid&quot;  &quot;Progenitor&quot; &quot;CD4.M&quot;      &quot;pDC&quot;        &quot;NK&quot;        </span></span><br><span class="line"><span class="comment"># [11] &quot;CLP&quot;        &quot;Mono&quot;</span></span><br></pre></td></tr></table></figure>

<p>合并<code>labelsOld</code>和<code>labelsNew2</code>，我们现在可以用<code>mapLables()</code>函数在<code>cellColData</code>里新建聚类标签。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme3<span class="operator">$</span>Clusters2 <span class="operator">&lt;-</span> mapLabels<span class="punctuation">(</span>projHeme3<span class="operator">$</span>Clusters<span class="punctuation">,</span> newLabels <span class="operator">=</span> labelNew2<span class="punctuation">,</span> oldLabels <span class="operator">=</span> labelOld<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>得到新的标签后，我们绘制最新的UMAP展示聚类结果。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>projHeme3<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">)</span></span><br><span class="line">p1</span><br></pre></td></tr></table></figure>

<p>如果被分析scATAC-seq数据对应的细胞系统也有scRNA-seq数据存在，那么这种分析范式将会特别有用。正如之前所说，这种scRNA-seq和scATAC-seq整合分析也为后续更加复杂的基因调控分析提供了出色的框架。</p>
<p>用<code>plotPDF()</code>函数保存该图矢量版本。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-UMAP-Remap-Clusters.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>使用<code>saveArchRProject</code>保存我们最初的projHeme3.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">saveArchRProject<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme3<span class="punctuation">,</span> outputDirectory <span class="operator">=</span> <span class="string">&quot;Save-ProjHeme3&quot;</span><span class="punctuation">,</span> load <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>


      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/05/31/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%85%AB%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter8/" data-id="clm1ywdjn00brecok3myk18gu" data-title="使用ArchR分析单细胞ATAC-seq数据(第八章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第七章)-analysis-sc-atac-seq-with-archr-chapter7" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/05/25/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B8%83%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter7/" class="article-date">
  <time class="dt-published" datetime="2020-05-25T22:58:26.329Z" itemprop="datePublished">2020-05-25</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/05/25/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B8%83%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter7/">使用ArchR分析单细胞ATAC-seq数据(第七章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第7章-ArchR的基因得分和标记基因"><a href="#第7章-ArchR的基因得分和标记基因" class="headerlink" title="第7章: ArchR的基因得分和标记基因"></a>第7章: ArchR的基因得分和标记基因</h1><p>尽管ArchR能够可靠地对细胞进行聚类，但它并不能先验(<em>a priori</em>)地知道每个聚类所代表的细胞类型。通常这个任务也只能靠人工注释，毕竟每个项目都不太一样。</p>
<p>为了实现细胞类型注释，我们会用到已知的细胞类型特异的标记基因。在scATAC-seq数据中，这些基因的表达量其实是根据染色质开放数据估计的基因得分(gene score)。所谓的基因得分，本质上就是用基因附近的调控元件去预测基因的表达量。ArchR的亮点在于，它允许用户提供复杂的自定义的<strong>距离加权开放性模型</strong>去计算这些基因得分。</p>
<h2 id="7-1-在ArchR中计算基因得分"><a href="#7-1-在ArchR中计算基因得分" class="headerlink" title="7.1 在ArchR中计算基因得分"></a>7.1 在ArchR中计算基因得分</h2><p>我们在文章中测试了超过50种不同的基因得分模型，并找到一种在大部分测试情况下都拥有良好表现的模型。这个模型目前是ArchR的默认设置，它包括三个主要部分</p>
<ol>
<li>在整个基因主体内(gene body)的开放部分都对基因得分有贡献</li>
<li>对于潜在的远距调控元件，我们使用一个指数加权函数计算得分</li>
<li>我们加入了基因边界(gene boundary)用于减少无关调控元件对基因得分的贡献</li>
</ol>
<p>那么，ArchR到底是怎么计算基因得分的呢？我建议先看图，再看文字描述。图中标注了两个TSS，分别表示两个基因的转录起始位点(TSS)，其中红色标识的基因是目标基因，我们需要根据其附近的开放情况来预测它的表达量。</p>
<img src="/upload/2020/05/image-86acaf58acfb4d58889acbfd219c9b95.png" alt="基因活跃得分" style="zoom:15%;" />

<p>对于每条染色体，ArchR都会构建一个分块矩阵(tile matrix)，块的大小由用户进行定义，在定义时计算（默认是500 bp），然后分析用户定义的另一个基因分窗(默认是基因两边100kb)和上一个分块矩阵的重叠部分。接着计算每个分块的起始位置或结束位置相对于基因主体(可额外添加上下游延伸)或基因起始位置的距离。我们发现基因表达量最好的预测者是基因区的局部开放性，包括启动子和基因主体。对于之前提到的远距调控元件，为了能正确地处理给定基因的远距离开放状态，ArchR只筛选了在基因分窗内但不跨到另一个基因区的分块。这个过滤标准既能利用远端调控元件提高基因表达量的预测准确性，同时也避免了无关开放状态的影响（例如，邻近基因启动子的开放状态）。每个分块到基因的距离会根据用户定义的开放模型里的权重进行转换，默认是e(-abs(distance)&#x2F;5000) + e-1)。</p>
<p>我们还发现基因长度会明显影响总体的基因得分，这是因为如果基因区包括基因主体，当基因越长时，落在基因主体内的开放区也就越多，这些开放区会和最大的基于距离的权重相乘，也就会计算出更高的得分。为了调整基因长度所带来的差异，ArchR以每个基因长度的倒数(1 &#x2F; gene size)作为权重，然后将该权重线性缩放到1到用户定义的最大值之间(默认是5)。保证了越短的基因对应更大的相对权重，一定程度上避免了长度的影响。</p>
<p>之后每个分块对应的距离权重和基因长度权重会和每个分块内的Tn5 insertion数相乘，接着将基因分窗内的所有分块的得分进行相加，相加时排除邻近基因区的分块。最后计算的总开放度就是基因得分(gene score)，接着所有基因的基因得分会按照用户定义的常数进行深度标准化(默认是10,000)。最后，计算的基因得分会被保存在相应的Arrow文件中用作下游分析。</p>
<p>在创建Arrow文件时，如果设置<code>createArrowFiles</code>的参数<code>addGeneScoreMat = TRUE</code>(默认为TRUE)，那么生成的Arrow文件中会包含基因得分矩阵。此外，我们也可以在任意时候使用<code>addGeneScoreMatrix()</code>在Arrow文件中加入基因得分矩阵。一旦计算完成，每个细胞都可以在嵌入图中根据基因得分进行展示，方便鉴定不同的细胞类型。之后的章节中会逐一介绍基因得分的应用。</p>
<p>当然并不是所有基因都会有基因得分，尤其是基因密度高的区域里的基因，它们更容易出现问题。因此，最好能在基因组浏览器对所有基因得分都做一遍检查，我们也会在后续章节里做相应介绍。</p>
<h2 id="7-2-鉴定标记特征"><a href="#7-2-鉴定标记特征" class="headerlink" title="7.2 鉴定标记特征"></a>7.2 鉴定标记特征</h2><p>除了使用已知的细胞类型相关标记基因用于聚类注释外，ArchR还能无偏地鉴定任何给定细胞分组(如聚类)的标记特征。这些特征包括单不限于peaks, 基于基因得分的基因, 基于chromVAR离差的转录因子motif。这部分工作可通过<code>getMarkerFeatures()</code>函数完成，它的核心参数是<code>useMatrix</code>和<code>groupBy</code>。<code>useMatrix</code>可以是任何的矩阵输入，如果设置为<code>GeneScoreMatrix</code>那么函数就会分析不同细胞类型特异出现的基因。这就能无偏的找到每个聚类的活跃基因，可以用来辅助聚类注释。</p>
<p>正如之前所提到的，<code>getMarkerFeatures()</code>能够接受任何存在Arrow文件里的矩阵作为输入，分析细胞类型的特异特征。例如<code>useMatrix = &quot;TileMatrix&quot;</code>用于鉴定细胞类群的特异基因区间，<code>useMatrix = &quot;PeakMatrix&quot;</code>用于鉴定细胞类群的特异peak。后续章节还会介绍如何使用其他特征类型作为<code>getMarkerFeatures()</code>的输入。</p>
<h3 id="7-2-1-标记特征的鉴定过程"><a href="#7-2-1-标记特征的鉴定过程" class="headerlink" title="7.2.1 标记特征的鉴定过程"></a>7.2.1 标记特征的鉴定过程</h3><p>标记特征的鉴定过程取决于每一组细胞对应的偏好-匹配背景细胞的选择。对于所有特征，每个细胞组都会与它自己的背景细胞组进行比较，判断给定的细胞组有更显著的开放性。</p>
<img src="/upload/2020/05/image-0dc48c66c89a4be598bdc9ccc7b80702.png" alt="markerFeature_schematic" style="zoom:45%;" />

<p>选择合适的背景细胞组对标记特征的鉴定至关重要，取决于<code>getMarkerFeatures)()</code>的<code>bias</code>参数所构建的多维空间。对于给定分组的细胞，ArchR会在提供的多维空间中寻找每个细胞分组外的最近邻细胞。这些细胞和给定分组内细胞极其相似，因此被称为<strong>偏好-匹配</strong>细胞(bias-matched cells)。通过这一方法，即便是在细胞数目比较少的组中，我们都能为每个特征计算稳健的显著性。</p>
<p>ArchR以<code>bias</code>参数中的所有维度作为输入，以分位数对值进行归一化，让每个维度的方差分布在相同的相对标度。举一个简单的例子，如如果<code>bias</code>的输入是<code>TSS</code>和<code>log10(Num Fragments)</code>，未经分位数归一化的值如下图所示</p>
<img src="/upload/2020/05/image-0d6f658c2d8144118dea1fc89eca5f36.png" alt="ackground_preNorm" style="zoom: 25%;" />

<p>这里y轴的相对方差和于x轴的相对方差相比，就显得特别小。如果我们对这些轴进行归一化，将值缩放到0-1之间，x和y的相对方差就近乎一样。<strong>注意</strong>，这个操作还会明显地改变每个细胞的最近邻（见下图右）。</p>
<img src="/upload/2020/05/image-28d4b94d9bed46ad8e08a2844b9c1de0.png" alt="background_postNorm" style="zoom: 25%;" />

<p>ArchR会对所有维度进行归一化，在归一化后的多维空间中以欧氏距离寻找最近邻。</p>
<h2 id="7-3-鉴定标记基因"><a href="#7-3-鉴定标记基因" class="headerlink" title="7.3 鉴定标记基因"></a>7.3 鉴定标记基因</h2><p>为了根据基因得分来鉴定标记基因，我们需要在调用<code>getMarkerFeatures()</code>时设置<code>useMatrix = &quot;GeneScoreMatrix</code>。此外，我们设置了<code>groupBy=&quot;Clusters&quot;</code>告诉ArchR使用<code>cellColData</code>的”Clusters”列对细胞分组，从而得到每个聚类特异的标记基因。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">markersGS <span class="operator">&lt;-</span> getMarkerFeatures<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    useMatrix <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters&quot;</span><span class="punctuation">,</span></span><br><span class="line">    bias <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;TSSEnrichment&quot;</span><span class="punctuation">,</span> <span class="string">&quot;log10(nFrags)&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    testMethod <span class="operator">=</span> <span class="string">&quot;wilcoxon&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>该函数返回一个<code>SummarizedExperiment</code>对象，里面记录着每个标记特征的相关信息。这是ArchR的一个常见输出，是下游数据分析的关键输出格式。<code>SummarizedExperiment</code>对象和矩阵类似，行是感兴趣的特征（例如基因），列表示样本。一个<code>SummarizedExperiment</code>能够记录一个或多个assay矩阵。如果你需要了解更多和<code>SummarizedExperiment</code>相关的内容，建议阅读相关的<a target="_blank" rel="noopener" href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">Bioconductor页面</a>，这里就不深入对这方面内容进行介绍。</p>
<p>我们能够使用<code>getMarkers()</code>函数从<code>SummarizedExperiment</code>对象中提取出一个包含多个<code>DataFrame</code>对象的列表，每个<code>DataFrame</code>对象都对应着一个聚类，记录着该聚类里相关的标记特征。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">markerList <span class="operator">&lt;-</span> getMarkers<span class="punctuation">(</span>markersGS<span class="punctuation">,</span> cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.01 &amp; Log2FC &gt;= 1.25&quot;</span><span class="punctuation">)</span></span><br><span class="line">markerList<span class="operator">$</span>C6</span><br></pre></td></tr></table></figure>

<p>我们使用<code>markerHeatmap()</code>创建含有所有标记特征的可视化热图。通过<code>labelMarkers()</code>在热图中标记部分感兴趣的基因。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;CD34&quot;</span><span class="punctuation">,</span> <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span><span class="punctuation">,</span> <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MS4A1&quot;</span><span class="punctuation">,</span> <span class="string">&quot;EBF1&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MME&quot;</span><span class="punctuation">,</span> <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CEBPB&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MPO&quot;</span><span class="punctuation">,</span> <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;IRF8&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD8A&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TBX21&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IL7R&quot;</span> <span class="comment">#TCells</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">heatmapGS <span class="operator">&lt;-</span> markerHeatmap<span class="punctuation">(</span></span><br><span class="line">  seMarker <span class="operator">=</span> markersGS<span class="punctuation">,</span> </span><br><span class="line">  cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.01 &amp; Log2FC &gt;= 1.25&quot;</span><span class="punctuation">,</span> </span><br><span class="line">  labelMarkers <span class="operator">=</span> markerGenes<span class="punctuation">,</span></span><br><span class="line">  transpose <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们需要用<code>ComplexHeatmap::draw()</code>才能绘制该热图，这是因为<code>heatmapGS</code>实际上是一个热图列表。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ComplexHeatmap<span class="operator">::</span>draw<span class="punctuation">(</span>heatmapGS<span class="punctuation">,</span> heatmap_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">,</span> annotation_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-2fb12c1f7da64e2ca3a990aafa1f2a8d.png" alt="GeneScores Marker heatmap" style="zoom:20%;" />

<p>使用plotPDF()函数保存可编辑的矢量版。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>heatmapGS<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;GeneScores-Marker-Heatmap&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">8</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">6</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="7-4-在嵌入上可视化标记基因"><a href="#7-4-在嵌入上可视化标记基因" class="headerlink" title="7.4 在嵌入上可视化标记基因"></a>7.4 在嵌入上可视化标记基因</h2><p>在之前章节中提到，我们可以在UMAP嵌入上展示每个细胞的基因得分。这可以通过修改<code>plotEmbedding()</code>函数里的<code>colorBy</code>和<code>name</code>参数实现</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;CD34&quot;</span><span class="punctuation">,</span>  <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span><span class="punctuation">,</span> <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MS4A1&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MME&quot;</span><span class="punctuation">,</span> <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MPO&quot;</span><span class="punctuation">,</span> <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD8A&quot;</span><span class="comment">#TCells</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> markerGenes<span class="punctuation">,</span> </span><br><span class="line">    embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    quantCut <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0.01</span><span class="punctuation">,</span> <span class="number">0.95</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    imputeWeights <span class="operator">=</span> <span class="literal">NULL</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>选择绘图列表的其中一个基因进行展示</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p<span class="operator">$</span>CD14</span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-1b321875a5db4095babe00b0391df152.png" alt="Plot-UMAP-Marker-Genes-WO-Imputation" style="zoom:25%;" />

<p>如果是需要绘制所有基因，那么可以使用<code>cowplot</code>将不同的标记基因整合到一张图中</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">p2 <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>p<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    x <span class="operator">+</span> guides<span class="punctuation">(</span>color <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> fill <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">    theme_ArchR<span class="punctuation">(</span>baseSize <span class="operator">=</span> <span class="number">6.5</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span>plot.margin <span class="operator">=</span> unit<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;cm&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span></span><br><span class="line">        axis.text.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.text.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">    <span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line">do.call<span class="punctuation">(</span>cowplot<span class="operator">::</span>plot_grid<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">list</span><span class="punctuation">(</span>ncol <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span>p2<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-cf5de1d7abe94e0ebf711dafa7c69fca.png" alt="UMAP of Marker gene" style="zoom:75%;" />

<p>使用<code>plotPDF()</code>函数保存可编辑的矢量版。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>plotList <span class="operator">=</span> p<span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;Plot-UMAP-Marker-Genes-WO-Imputation.pdf&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="7-5-使用MAGIC填充标记基因"><a href="#7-5-使用MAGIC填充标记基因" class="headerlink" title="7.5 使用MAGIC填充标记基因"></a>7.5 使用MAGIC填充标记基因</h2><p>在上一节中，你可能注意到一些基因得分图变化很大，这是因为scATAC-seq数据太过稀疏。我们使用<a target="_blank" rel="noopener" href="https://github.com/KrishnaswamyLab/MAGIC">MAGIC</a>根据邻近细胞填充基因得分对信号进行平滑化处理。我们发现者能够极大程度地提高基因得分的可视化解读性。要执行此操作，我们需要先在我们的<code>ArchRProject</code>中加入填充权重。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addImputeWeights<span class="punctuation">(</span>projHeme2<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这些填充权重会在之后绘制UMAP嵌入图里的基因得分时传入到<code>plotEmbedding()</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;CD34&quot;</span><span class="punctuation">,</span>  <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span><span class="punctuation">,</span> <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MS4A1&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MME&quot;</span><span class="punctuation">,</span> <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MPO&quot;</span><span class="punctuation">,</span> <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD8A&quot;</span><span class="comment">#TCells</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> markerGenes<span class="punctuation">,</span> </span><br><span class="line">    embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    imputeWeights <span class="operator">=</span> getImputeWeights<span class="punctuation">(</span>projHeme2<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>和之前一样，我们可以只画其中一个基因</p>
<img src="/upload/2020/05/image-7a804bdfceff43b89a6e4ab7464061a7.png" alt="UMAP of imputation" style="zoom:50%;" />

<p>也可以用<code>cowplot</code>绘制所有的标记基因</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#Rearrange for grid plotting</span></span><br><span class="line">p2 <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>p<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    x <span class="operator">+</span> guides<span class="punctuation">(</span>color <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> fill <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">    theme_ArchR<span class="punctuation">(</span>baseSize <span class="operator">=</span> <span class="number">6.5</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span>plot.margin <span class="operator">=</span> unit<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;cm&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span></span><br><span class="line">        axis.text.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.text.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">    <span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line">do.call<span class="punctuation">(</span>cowplot<span class="operator">::</span>plot_grid<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">list</span><span class="punctuation">(</span>ncol <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span>p2<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-9b93a3a6aabd4b99ac6e737dfe17a6a4.png" alt="UMAP of all imputation" style="zoom:50%;" />

<p>使用plotPDF()函数保存可编辑的矢量版。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>plotList <span class="operator">=</span> p<span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;Plot-UMAP-Marker-Genes-W-Imputation.pdf&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>


<h2 id="7-6-使用ArchRBrowser绘制Track"><a href="#7-6-使用ArchRBrowser绘制Track" class="headerlink" title="7.6 使用ArchRBrowser绘制Track"></a>7.6 使用ArchRBrowser绘制Track</h2><p>除了在UMAP上绘制每个细胞的基因得分外，我们还可以在基因组浏览器上浏览这些标记基因的局部染色体开放状态。为了执行该操作，我们使用<code>plotBrowserTrack()</code>函数，它会构建一系列绘图对象，每个对象对应一个标记基因。函数会根据<code>groupBy</code>输入的组别信息在不同track上绘制每组的开放状态。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;CD34&quot;</span><span class="punctuation">,</span> <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span><span class="punctuation">,</span> <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MS4A1&quot;</span><span class="punctuation">,</span> <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span><span class="punctuation">,</span> <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD8A&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TBX21&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IL7R&quot;</span> <span class="comment">#TCells</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p <span class="operator">&lt;-</span> plotBrowserTrack<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    geneSymbol <span class="operator">=</span> markerGenes<span class="punctuation">,</span> </span><br><span class="line">    upstream <span class="operator">=</span> <span class="number">50000</span><span class="punctuation">,</span></span><br><span class="line">    downstream <span class="operator">=</span> <span class="number">50000</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>通过选择列表中的给定基因来绘制最终结果</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grid<span class="operator">::</span>grid.newpage<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">grid<span class="operator">::</span>grid.draw<span class="punctuation">(</span>p<span class="operator">$</span>CD14<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-fa9adf3584ea492484867839f4b64ef7.png" alt="Plot-Tracks-Marker-Genes_5" style="zoom:25%;" />

<p>使用<code>plotPDF()</code>函数可以将多幅基因座位对应的图保存在一个PDF文件中。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>plotList <span class="operator">=</span> p<span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;Plot-Tracks-Marker-Genes.pdf&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="7-7-启动ArchRBrowser"><a href="#7-7-启动ArchRBrowser" class="headerlink" title="7.7 启动ArchRBrowser"></a>7.7 启动ArchRBrowser</h2><p>scATAC-seq数据分析一个与生俱来的挑战就是在基因组浏览器上可视化聚类间的染色质开放水平。传统做法是，先对scATAC-seq的fragments进行分组，然后构建一个基因组覆盖度的bigwig文件，最后对track进行标准化后才能实现定量层面上的可视化。终端用户经常用于测序数据可视化的基因组浏览器包括 <a target="_blank" rel="noopener" href="https://epigenomegateway.wustl.edu/">WashU Epigenome Browser</a>,  <a target="_blank" rel="noopener" href="http://genome.ucsc.edu/">UCSC Genome Browser</a>, <a target="_blank" rel="noopener" href="http://software.broadinstitute.org/software/igv/">IGV browser</a>。这一过程需要用到多个不同软件，并且一旦细胞分组信息改变，或者增加了更多的样本，就得重新分组输出bigiwig文件，相当耗时。</p>
<p>为了避免重复劳动，ArchR开发了基于Shiny的交互式基因组浏览器，只需要一行<code>ArchRBrowser(ArchRProj)</code>命令就可启动。因为数据都存放在Arrow文件中，所以这个交互式浏览器就能动态的更改细胞分组信息，分辨率以及标准化，实现实时的track水平的可视化。ArchR Genome Browser 同样也能生成高质量的PDF格式输出文件用于发表或分享。此外，浏览器支持用户通过<code>features</code>参数传入<code>GenomicRanges</code>对象用来展示特征，或者通过<code>loop</code>参数传入基因组交互文件（ co-accessibility, peak-to-gene linkages, loops from chromatin conformation data）。其中<code>loop</code>的预期格式是<code>GRanges</code>，起点表示其中一个loop的中心，终点表示另一个loop中心。</p>
<p>使用<code>ArchRBrowser()</code>函数启动我们的本地交互基因组浏览器</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-51ae326ad0fb47f0a60e9dc62310245d.png" alt="ArchR_Browser_1"></p>
<p>通过选择”Gene Symbol”，就可以开始浏览了。你可能需要点击”Plot Track”才能强制让你的浏览器刷新</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-941f533e83154ccf92caf016d8fac763.png" alt="ArchR_Browser_2"></p>
<p>当我们绘制一个基因位点后，我们看到不同track代表的是我们数据的不同聚类。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-5c87ee91bd9b47aa95c10686bb19fe2b.png" alt="ArchR_Browser_3"></p>
<p>如果我们点击”Additional Parameters”侧边栏，我们可以选择部分聚类进行展示</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-47321577c4ba422780cb67cc5a9aaf3f.png" alt="ArchR_Browser_4"></p>
<p>通过反选cluster1, 2, 3，我们就在绘图中移除了它们</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-324391d639304514a4b11648489e8932.png" alt="ArchR_Browser_5"></p>
<p>当我们返回到”Plot”时，此时Cluster1,2,3就从图片中消失了。同样你也可能需要点击”Plot Track”才能让浏览器刷新结果</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-db216dfaf3bc4bb5a56b5d41b2696e67.png" alt="ArchR_Browser_6"></p>
<p>无论在哪个阶段，我们都可以点击”Download The Track”来输出当前的绘图结果。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/05/25/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B8%83%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter7/" data-id="clm1ywdjh00baecok1ppqh3wj" data-title="使用ArchR分析单细胞ATAC-seq数据(第七章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第六章)-analysis-sc-atac-seq-with-archr-chapter6" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/05/24/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%85%AD%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter6/" class="article-date">
  <time class="dt-published" datetime="2020-05-24T18:37:29.622Z" itemprop="datePublished">2020-05-24</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/05/24/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%85%AD%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter6/">使用ArchR分析单细胞ATAC-seq数据(第六章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第6章-单细胞嵌入"><a href="#第6章-单细胞嵌入" class="headerlink" title="第6章: 单细胞嵌入"></a>第6章: 单细胞嵌入</h1><p>在ArchR中，类似于UMAP和t-SNE的嵌入方法被用于在降维空间中可视化展示单细胞数据。这些嵌入有各自的优势和缺陷。我们之所以称它们为”嵌入”是因为他们只限于对聚类进行可视化而非用于鉴定聚类(在LSI子空间中的聚类分析)。UMAP和t-SNE的主要区别在于对细胞和聚类间的距离解释，t-SNE用于保留数据的局部结构，而UMAP则是保留局部结构的同时尽可能保留全局结构。从理论上来讲，UMAP中细胞聚类间的距离存在意义，而t-SNE则没有。也就是说，你不能说t-SNE中聚类A比聚类C更接近聚类B，而UMAP在设计的时候允许这种类型的比较。不过需要注意的是，由于UMAP是新出现的方法，因此使用UMAP的文章才刚刚兴起。</p>
<p>需要注意的是，无论是UMAP还是t-SNE，两个的运行结果都不是确定性的，也就是相同输入会得到不完全相同的结果。尽管如此，我们使用样本重复后发现t-SNE比UMAP更加的随机。此外，使用<code>uwot</code>包里UMAP时，设置相同的随机数种子会得到相同的结果。选择使用UMAP还是t-SNE是有细微差别的，但在我们手中，UMAP非常适合各种应用，这是我们对scATAC seq数据的标准选择。UMAP的性能也比t-SNE快。也许最重要的是，使用UMAP可以创建一个嵌入并将新样本投影到该嵌入中，而这在t-SNE中是不可能的，因为数据的拟合和预测是同时发生的。</p>
<p>无论您选择哪种方法，输入参数都会对结果嵌入产生剧烈影响。因此，了解各种输入参数并调整这些参数以最好地满足您自己的数据需要是很重要的。ArchR实现了一组默认的输入参数，这些参数适用于大多数情况，但实际上没有一组参数可以直接套用，我们要根据不同的细胞数、复杂度和质量进行调整。</p>
<h2 id="6-1-Uniform-Manifold-Approximation-and-Projection-UMAP"><a href="#6-1-Uniform-Manifold-Approximation-and-Projection-UMAP" class="headerlink" title="6.1 Uniform Manifold Approximation and Projection (UMAP)"></a>6.1 Uniform Manifold Approximation and Projection (UMAP)</h2><p>我们使用<code>addUMAP</code>函数运行UMAP</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addUMAP<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    nNeighbors <span class="operator">=</span> <span class="number">30</span><span class="punctuation">,</span> </span><br><span class="line">    minDist <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span> </span><br><span class="line">    metric <span class="operator">=</span> <span class="string">&quot;cosine&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>使用<code>@</code>操作符可以从<code>ArchRProject</code>中列出有哪些可用的<code>embedding</code>，如<code>projHeme2@embeddings</code></p>
<p>我们使用<code>plotEmbedding</code>函数绘制UMAP图，设置<code>embedding=&quot;UMAP&quot;</code>。通过修改<code>colorBy</code>和<code>name</code>来告诉ArchR使用给定哪个元信息矩阵的列对细胞进行上色。p1是按照样本进行上色</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Sample&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>除了根据”Sample”外，我们还可以根据上一张鉴定的”Cluster”进行上色</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p2 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Clusters&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>使用<code>ggAlignPlots</code>将这两个图共同展示，”type&#x3D;h”表示两个图是水平排列</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ggAlignPlots<span class="punctuation">(</span>p1<span class="punctuation">,</span> p2<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;h&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-b23887afd69d49d7911e95dda3454337.png" alt="UMAP of Seurat"></p>
<p>用<code>plotPDF()</code>可以将保存图片的矢量版。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span>p2<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-UMAP-Sample-Clusters.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们还可以使用<code>plotEmbedding()</code>可视化之前用<code>scran</code>聚类的结果</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Sample&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">)</span></span><br><span class="line">p2 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;ScranClusters&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">)</span></span><br><span class="line">ggAlignPlots<span class="punctuation">(</span>p1<span class="punctuation">,</span> p2<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;h&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-d2116752bc394216ac398287dea3c486.png" alt="UMAP of scran"></p>
<p>同样用<code>plotPDF()</code>可以将保存图片的矢量版。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span>p2<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-UMAP-Sample-ScranClusters.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="6-2-t-Stocastic-Neighbor-Embedding-t-SNE"><a href="#6-2-t-Stocastic-Neighbor-Embedding-t-SNE" class="headerlink" title="6.2 t-Stocastic Neighbor Embedding (t-SNE)"></a>6.2 t-Stocastic Neighbor Embedding (t-SNE)</h2><p>t-SNE图可以用<code>addTSNE()</code>函数运行得到</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addTSNE<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;TSNE&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    perplexity <span class="operator">=</span> <span class="number">30</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>和之前UMAP一样，我们同样使用<code>plotEmbedding()</code>绘制t-SNE嵌入图。这里不需要考虑嵌入的类型，可以继续使用之前的<code>colorBy</code>和<code>name</code>参数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Sample&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;TSNE&quot;</span><span class="punctuation">)</span></span><br><span class="line">p2 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Clusters&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;TSNE&quot;</span><span class="punctuation">)</span></span><br><span class="line">ggAlignPlots<span class="punctuation">(</span>p1<span class="punctuation">,</span> p2<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;h&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>


<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-5fe75a907040457f8d17d4901853d90e.png" alt="t-SNE of Seurat"></p>
<p>用<code>plotPDF()</code>可以将保存图片的矢量版。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span>p2<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-TSNE-Sample-Clusters.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>和之前UMAP的操作类似，我们可以将<code>scran</code>的结果和<code>Seurat::FindClusters()</code>的结果进行比较</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Sample&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;TSNE&quot;</span><span class="punctuation">)</span></span><br><span class="line">p2 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;ScranClusters&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;TSNE&quot;</span><span class="punctuation">)</span></span><br><span class="line">ggAlignPlots<span class="punctuation">(</span>p1<span class="punctuation">,</span> p2<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;h&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-1afcd4800aa44715b51395f09f109ffc.png" alt="t-SNE of Scran"></p>
<p>用<code>plotPDF()</code>可以将保存图片的矢量版。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span>p2<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-tSNE-Sample-ScranClusters.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="6-3-Harmony后降维"><a href="#6-3-Harmony后降维" class="headerlink" title="6.3 Harmony后降维"></a>6.3 Harmony后降维</h2><p>在第4章中，我们通过<code>addHarmony</code>调用Harmony对数据进行批次校正，创建了一个名为”Harmony”的<code>reducedDims</code>对象。我们通过UMAP或t-SNE对结果进行嵌入可视化，对迭代LSI结果和Harmony校正结果进行比较，评估Harmony的作用。</p>
<p>保持和之前UMAP嵌入一样的参数，只修改<code>reducedDims=&quot;Harmony&quot;</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addUMAP<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;Harmony&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;UMAPHarmony&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    nNeighbors <span class="operator">=</span> <span class="number">30</span><span class="punctuation">,</span> </span><br><span class="line">    minDist <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span> </span><br><span class="line">    metric <span class="operator">=</span> <span class="string">&quot;cosine&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">p3 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Sample&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;UMAPHarmony&quot;</span><span class="punctuation">)</span></span><br><span class="line">p4 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Clusters&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;UMAPHarmony&quot;</span><span class="punctuation">)</span></span><br><span class="line">ggAlignPlots<span class="punctuation">(</span>p3<span class="punctuation">,</span> p4<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;h&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-47977d6cde0b411991a8f6827bc57114.png" alt="UMAP of Harmony"></p>
<p>用<code>plotPDF()</code>可以将保存图片的矢量版。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span>p2<span class="punctuation">,</span>p3<span class="punctuation">,</span>p4<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-UMAP2Harmony-Sample-Clusters.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>相同的方法用t-SNE进行可视化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addTSNE<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;Harmony&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;TSNEHarmony&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    perplexity <span class="operator">=</span> <span class="number">30</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">p3 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Sample&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;TSNEHarmony&quot;</span><span class="punctuation">)</span></span><br><span class="line">p4 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Clusters&quot;</span><span class="punctuation">,</span> embedding <span class="operator">=</span> <span class="string">&quot;TSNEHarmony&quot;</span><span class="punctuation">)</span></span><br><span class="line">ggAlignPlots<span class="punctuation">(</span>p3<span class="punctuation">,</span> p4<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;h&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-63300928b3f249d8805920f0d40a62d0.png" alt="t-SNE of Harmony"></p>
<p>用<code>plotPDF()</code>可以将保存图片的矢量版。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span>p2<span class="punctuation">,</span>p3<span class="punctuation">,</span>p4<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-TSNE2Harmony-Sample-Clusters.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>


      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/05/24/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%85%AD%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter6/" data-id="clm1ywdjo00bvecok81a90hpv" data-title="使用ArchR分析单细胞ATAC-seq数据(第六章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第五章)-analysis-sc-atac-seq-with-archr-chapter5" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/05/24/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%BA%94%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter5/" class="article-date">
  <time class="dt-published" datetime="2020-05-24T17:19:17.385Z" itemprop="datePublished">2020-05-24</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/05/24/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%BA%94%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter5/">使用ArchR分析单细胞ATAC-seq数据(第五章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第5章-使用ArchR聚类"><a href="#第5章-使用ArchR聚类" class="headerlink" title="第5章: 使用ArchR聚类"></a>第5章: 使用ArchR聚类</h1><p>大部分单细胞聚类算法都在降维后空间中计算最近邻图，然后鉴定”社区”或者细胞聚类。这些方法效果表现都特别出色，已经是scRNA-seq的标准策略，所以ArchR直接使用了目前scRNA-seq包中最佳的聚类算法用来对scATAC-seq数据进行聚类。</p>
<h2 id="5-1-使用Seurat的FindClusters-函数进行聚类"><a href="#5-1-使用Seurat的FindClusters-函数进行聚类" class="headerlink" title="5.1 使用Seurat的FindClusters()函数进行聚类"></a>5.1 使用Seurat的<code>FindClusters()</code>函数进行聚类</h2><p>我们发现<a target="_blank" rel="noopener" href="https://github.com/satijalab/seurat">Seurat</a>实现的图聚类方法表现最好，所以在ArchR中，<code>addClusters()</code>函数本质是上将额外的参数通过<code>...</code>传递给<code>Seurat::FindClusters()</code>函数，从而得到聚类结果。在分析中，我们发现<code>Seurat::FindClusters()</code>是一个确定性的聚类算法，也就是相同的输入总是能得到完全相同的输出。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addClusters<span class="punctuation">(</span></span><br><span class="line">    input <span class="operator">=</span> projHeme2<span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span></span><br><span class="line">    method <span class="operator">=</span> <span class="string">&quot;Seurat&quot;</span><span class="punctuation">,</span></span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;Clusters&quot;</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">0.8</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 只展示部分信息</span></span><br><span class="line"><span class="comment"># Maximum modularity in 10 random starts: 0.8568</span></span><br><span class="line"><span class="comment"># Number of communities: 11</span></span><br><span class="line"><span class="comment"># Elapsed time: 1 seconds</span></span><br></pre></td></tr></table></figure>

<p>我们可以使用<code>$</code>符号来获取聚类信息，输出结果是每个细胞对应的cluster</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">head<span class="punctuation">(</span>projHeme2<span class="operator">$</span>Clusters<span class="punctuation">)</span></span><br><span class="line"><span class="comment"># [1] &quot;C10&quot; &quot;C6&quot;  &quot;C1&quot;  &quot;C2&quot;  &quot;C2&quot;  &quot;C10&quot;</span></span><br></pre></td></tr></table></figure>

<p>我们统计下每个cluster的细胞数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">table<span class="punctuation">(</span>projHeme2<span class="operator">$</span>Clusters<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#  C1  C10  C11   C2   C3   C4   C5   C6   C7   C8   C9 </span></span><br><span class="line"><span class="comment"># 310 1247 1436  480  323  379  852 1271  677 2550  726 </span></span><br></pre></td></tr></table></figure>

<p>为了更好了解样本在cluster的分布，我们可以使用<code>confusionMatrix()</code>函数通过每个样本创建一个聚类混淆矩阵(cluster confusion matrix)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">cM <span class="operator">&lt;-</span> confusionMatrix<span class="punctuation">(</span>paste0<span class="punctuation">(</span>projHeme2<span class="operator">$</span>Clusters<span class="punctuation">)</span><span class="punctuation">,</span> paste0<span class="punctuation">(</span>projHeme2<span class="operator">$</span>Sample<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">cM</span><br></pre></td></tr></table></figure>

<p>文字信息太多，这里直接用热图进行展示</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>pheatmap<span class="punctuation">)</span></span><br><span class="line">cM <span class="operator">&lt;-</span> cM <span class="operator">/</span> Matrix<span class="operator">::</span>rowSums<span class="punctuation">(</span>cM<span class="punctuation">)</span></span><br><span class="line">p <span class="operator">&lt;-</span> pheatmap<span class="operator">::</span>pheatmap<span class="punctuation">(</span></span><br><span class="line">    mat <span class="operator">=</span> as.matrix<span class="punctuation">(</span>cM<span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">    color <span class="operator">=</span> paletteContinuous<span class="punctuation">(</span><span class="string">&quot;whiteBlue&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">    border_color <span class="operator">=</span> <span class="string">&quot;black&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">p</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-ecebca2150204ef89108f0b498b4556c.png" alt="混淆矩阵"></p>
<p>细胞有时在二维嵌入中的相对位置与所识别的聚类并不完全一致。更确切的说，单个聚类中的细胞可能出现在嵌入的多个不同区域中。在这些情况下，可以适当地调整聚类参数或嵌入参数，直到两者之间达成一致。</p>
<h2 id="5-2-使用scran聚类"><a href="#5-2-使用scran聚类" class="headerlink" title="5.2 使用scran聚类"></a>5.2 使用<code>scran</code>聚类</h2><p>除了<code>Seurat</code>, ArchR还能够使用<a target="_blank" rel="noopener" href="https://bioconductor.org/packages/release/bioc/html/scran.html">scran</a>进行聚类分析，我们只需要修改<code>addClusters()</code>中的<code>method</code>参数即可。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addClusters<span class="punctuation">(</span></span><br><span class="line">    input <span class="operator">=</span> projHeme2<span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span></span><br><span class="line">    method <span class="operator">=</span> <span class="string">&quot;scran&quot;</span><span class="punctuation">,</span></span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;ScranClusters&quot;</span><span class="punctuation">,</span></span><br><span class="line">    k <span class="operator">=</span> <span class="number">15</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>


      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/05/24/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%BA%94%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter5/" data-id="clm1ywdjm00boecokhlo1f56p" data-title="使用ArchR分析单细胞ATAC-seq数据(第五章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第四章)-sanalysis-sc-atac-seq-with-archr-chapter4" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/05/24/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%9B%9B%E7%AB%A0)-sanalysis-sc-atac-seq-with-archr-chapter4/" class="article-date">
  <time class="dt-published" datetime="2020-05-24T17:16:02.278Z" itemprop="datePublished">2020-05-24</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/05/24/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%9B%9B%E7%AB%A0)-sanalysis-sc-atac-seq-with-archr-chapter4/">使用ArchR分析单细胞ATAC-seq数据(第四章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第4章-ArchR降维分析"><a href="#第4章-ArchR降维分析" class="headerlink" title="第4章: ArchR降维分析"></a>第4章: ArchR降维分析</h1><p>scATAC-seq数据的<strong>稀疏性</strong>让降维这一步充满了挑战性。在scATAC-seq数据中，一个特定的位置存在三种状态，分为不开放，一条链开放，两条链都开放。但即便是在高质量的scATAC-seq数据中，大部分开放区域都没有被转座酶切割，因此大部分座位都是0。此外，如果我们发现一个细胞(A)的一处peak区域有三个Tn5 insertions，而另外一个细胞(B)对应区域只有一个Tn5 insertions，由于数据的稀疏性，我们也无法断定A就比B更开放。鉴于此，许多分析框架都会把scATAC-seq数据矩阵二值化，也就是只有0和1两种情况。即便如此，因为转座酶切割的位置还是很少，所以二值矩阵大部分区域还是0。我们还需要注意的是，scATAC-seq数据中的0既有可能表示”不开放”，也可能是”没取样到”, 从生物学角度上出发，这是两种截然相反的推断。正因为数据大部分都是0，所以1比0更有信息量。scATAC-seq数据的<strong>稀疏性</strong>便是来源于这种低信息量。</p>
<p>如果你直接在稀疏矩阵上使用标准的降维分析，例如主成分分析(Principal Component Analysis, PCA)，然后使用前两个主成分进行绘图，你可能无法得到你想要的结果，这是由于数据的稀疏性会导致细胞间在0的位置有更高的相似性。为了解决这个问题，我们采用了分层的降维策略。</p>
<p>首先，我们会使用隐语义分析(Latent Semantic Indexing, LSI)进行降维。LSI方法最早用在自然语言分析中，根据字数来评估文档相似度。之所以在自然语言分析中发明该方法，是因为文档集中会有许多不同的词，每个词出现的次数也很少，最终的数据非常稀疏且充满噪音。<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pubmed/25953818">Cusanovich et al. (Science 2015)</a>第一次在scATAC-sq数分析中引入了LSI。在scATAC-seq场景下，不同样本就是不同的文档，不同的区域&#x2F;peak就是不同的单词(注1)。计算步骤如下</p>
<ol>
<li>对每个细胞计算根据深度标准化后的词频(Term frequency, TF)</li>
<li>根据逆文档频率(IDF, inverse document frequency)进行标准化，便于后续特征选择</li>
<li>最后对结果矩阵进行对数变换（也就是<code>log(TF-IDF)</code>）</li>
</ol>
<p>接着，奇异值分解( singular value decomposition, SVD)技术能够分析出不同样本间更有价值的信息，从而降低了维度。上面这两步能够将原本成千上万维的稀疏矩阵的降维到数十个或者数百个维度。</p>
<p>最后，我们使用更加常见降维方法，例如UMAP( Uniform Manifold Approximation and Projection ) 或t-SNE( t-distributed stochastic neighbor embedding ) 对数据进行可视化。ArchR称这这些可视化方法为嵌入(embeddings, 注2)</p>
<ul>
<li>注1: 对LSI最直观理解就是它在自然语言分析的用途，也就是为每篇文档找到关键词，这个关键词必须要在当前文档中出现，同时最好别在其他文档里出现。在scATAC-seq的语境中，如果一个区域在所有细胞中都有insertion，那么就不特异，最好是只有几个细胞有insertion，而大部分细胞没有，这才算一个特异的insertion。</li>
<li>注2: 我们可以认为道路是嵌入在三维空间中一维流形。我们用一维道路中的地址号码确定地址，而非三维空间的坐标。</li>
</ul>
<h2 id="4-1-ArchR的LSI实现"><a href="#4-1-ArchR的LSI实现" class="headerlink" title="4.1 ArchR的LSI实现"></a>4.1 ArchR的LSI实现</h2><p>ArchR实现了多种LSI方法，并用不同测数据集对这些方法进行了测试。ArchR默认的LSI实现和Timothy Stuart在<a target="_blank" rel="noopener" href="https://satijalab.org/signac/">Signac</a>引入的方法有关，也就是先将词频(TF, term frequency)根据常数(10,000)进行深度标准化，然后根据逆文档频率(IDF, inverse document frequency)进行标准化，最后对结果矩阵进行对数变换（也就是<code>log(TF-IDF)</code>）</p>
<p>LSI降维的其中一个关键输入是起始矩阵。到目前为止，scATAC-seq主要有两种策略计算矩阵，一是使用peak区域，二是对全基因组分块(genome-wide tiles)。由于在降维前，我们还没有聚类，也没有聚类特异的peak，所以不能使用peak区域作为LSI的输入。而且直接使用所有数据进行peak calling，会导致一些细胞特异性的peak被掩盖。更何况，当你在实验中增加新的样本，那么合并后的peak集就会发生改变，导致整个方法不太稳定。第二种策略通过对全基因组分块，保证了特征集(feature set)的一致性和无偏好性，缓解了第一种方法存在的问题。只不过，使用全基因组分块会得到一个非常大的细胞X特征的矩阵，因此大部分LSI实现都至少用5000 bp对基因组进行分块。计算量降低的同时也导致分辨率急剧下降，因为大部分开放区域只有几百个碱基的长度。</p>
<p>在Arrow文件的设计帮助下，即便是以500-bp作为基因组的分块，ArchR依旧能够快速的计算LSI。这就解决了分辨率的问题，使得在calling peak前就能进行聚类分析。使用500-bp进行分块的更大挑战是这会得到一个拥有数百万个特征的cell-by-tile矩阵。尽管ArchR能够通过读取必要矩阵的方式在R中加载部分数据，但是我们还是实现了一个”近似LSI”算法，使用一部分细胞进行最初的降维分析。近似LSI有两个主要用途，(1)提高降维速度，(2)初始降维时使用的细胞数的减少会降低数据的粒度，而这种粒度的减少有利于减少数据中的批次效应。但要注意，这可能会掩盖真实的生物学现象，因此近似LSI方法需要在人为监督下谨慎的使用。</p>
<h2 id="4-2-隐语义-Latent-Semantic-Indexing-迭代"><a href="#4-2-隐语义-Latent-Semantic-Indexing-迭代" class="headerlink" title="4.2 隐语义(Latent Semantic Indexing)迭代"></a>4.2 隐语义(Latent Semantic Indexing)迭代</h2><p>在scRNA-seq降维分析中(例如PCA)，一般都会鉴定变异基因。之所以这样做，是因为高变异基因通常更有生物学意义，并能够减少实验噪音。scATAC-seq数据只有0&#x2F;1两种情况，所以你不能为降维分析提供变异peak。除了无法使用变异更大的peak，我们也尝试使用了更加开放的特征作为LSI的输入，然而经过多个样本的测试，我们发现处理结果存在较高的噪音，并且很难重复。为了解决该问题，我们提出了”迭代LSI法” (<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pubmed/31375813">Satpathy*, Granja* et al. Nature Biotechnology 2019</a>和 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pubmed/31792411">Granja*, Klemm* and McGinnis* et al. Nature Biotechnology 2019</a>). 这个方法首先在最开放的分块中计算初始的LSI变换，识别未受到批次效应影响的低分辨率聚类。例如，我们可以在PBMC中鉴定出几个主要大群(T细胞、B细胞和单核细胞)。接着，ArchR计算这些聚类在每个特征中的平均开放度。ArchR识别这些聚类中变化最大的peak，用作后续LSI的特征。在第二次迭代中，变化最大的peak就类似于scRAN-seq中高变异基因。用户可以决定进行多次轮的LSI迭代。我们发现这个方法能够降低观察到的批次效应，并能使用更加合理的特征矩阵进行降维。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-4c87a1fbfe6741a8aced892540735271.png" alt="LSI过程"></p>
<p>我们可以是用ArchR的<code>addIterativeLSI()</code>函数执行迭代LSI分析。虽然默认参数适用于绝大部分情况，但是我们还是鼓励你使用<code>?addIterativeLSI</code>看看有哪些可用的参数，以及它们对你的数据集的影响。一般经常修改的参数是<code>iterations</code>,<code>varFeatures</code>和<code>resolution</code>。注意，LSI的结果并不是确定的，也就是即便你用相同参数处理同一个数据，也会得到不一样的结果。当然只是略微不同，总体上还是相似的。因此，一旦你得到一个相对理想的降维结果，请及时保存你的<code>ArchRProject</code>项目或者相关的LSI信息。</p>
<p>处于教学目的，我们会创建一个名为”IterativeLSI”的<code>reduceDims</code>对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addIterativeLSI<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span></span><br><span class="line">    useMatrix <span class="operator">=</span> <span class="string">&quot;TileMatrix&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    iterations <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> </span><br><span class="line">    clusterParams <span class="operator">=</span> <span class="built_in">list</span><span class="punctuation">(</span> <span class="comment">#See Seurat::FindClusters</span></span><br><span class="line">        resolution <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0.2</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        sampleCells <span class="operator">=</span> <span class="number">10000</span><span class="punctuation">,</span> </span><br><span class="line">        n.start <span class="operator">=</span> <span class="number">10</span></span><br><span class="line">    <span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">    varFeatures <span class="operator">=</span> <span class="number">25000</span><span class="punctuation">,</span> </span><br><span class="line">    dimsToUse <span class="operator">=</span> <span class="number">1</span><span class="operator">:</span><span class="number">30</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>如果你在下游分析时看到一些批次效应，其中一个选择就是增加LSI的迭代次数，并以更低的分辨率进行聚类。此外，也可以通过降低特征变量的数目让程序关注变异更大的特征。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addIterativeLSI<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span></span><br><span class="line">    useMatrix <span class="operator">=</span> <span class="string">&quot;TileMatrix&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;IterativeLSI2&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    iterations <span class="operator">=</span> <span class="number">4</span><span class="punctuation">,</span> </span><br><span class="line">    clusterParams <span class="operator">=</span> <span class="built_in">list</span><span class="punctuation">(</span> <span class="comment">#See Seurat::FindClusters</span></span><br><span class="line">        resolution <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0.1</span><span class="punctuation">,</span> <span class="number">0.2</span><span class="punctuation">,</span> <span class="number">0.4</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        sampleCells <span class="operator">=</span> <span class="number">10000</span><span class="punctuation">,</span> </span><br><span class="line">        n.start <span class="operator">=</span> <span class="number">10</span></span><br><span class="line">    <span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">    varFeatures <span class="operator">=</span> <span class="number">15000</span><span class="punctuation">,</span> </span><br><span class="line">    dimsToUse <span class="operator">=</span> <span class="number">1</span><span class="operator">:</span><span class="number">30</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="4-3-近似LSI"><a href="#4-3-近似LSI" class="headerlink" title="4.3 近似LSI"></a>4.3 近似LSI</h2><p>对于特别大的数据集，ArchR使用LSI投影得到近似的LSI降维结果。这个步骤和之前的迭代LSI流程相似，但是在LSI步骤存在一些不同。首先，我们会<strong>随机</strong>选择一部分”路标细胞”用于LSI降维分析。接着，使用”路标细胞”计算的逆文档频率对余下的细胞进行TF-IDF标准化。然后，这些标准化的细胞被投影到”路标细胞”定义的SVD子空间中。最终，我们就将余下的细胞根据少量的”路标细胞”实现了LSI变换和投影。因为ArchR不需要将所有细胞的read都保存在内存中，而是不断的提取每个样本的细胞，将其投影到路标细胞的LSI空间中，所以近似LSI算法非常高效。因此，即便之后用到了更大的数据集，近似LSI也能够减少内存的使用。</p>
<p><strong>注意</strong>，路标细胞的数量取决于数据集中不同细胞的比例。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-6da2a440991f4aeab6baf6ccc24c3a66.png" alt="近似LSI"></p>
<p>近似LSI通过修改ArchR的<code>addIterativeLSI()</code>函数的<code>sampleCellsFinal</code>和<code>projectCellsPre</code>参数来实现.<code>samplesCellsFinal</code>决定了路标细胞的数目，<code>projectCellsPre</code>告诉ArchR使用路标细胞对其余细胞进行投影。</p>
<h2 id="4-4-使用Harmony矫正批次效应"><a href="#4-4-使用Harmony矫正批次效应" class="headerlink" title="4.4 使用Harmony矫正批次效应"></a>4.4 使用Harmony矫正批次效应</h2><p>在某些情况下，迭代的LSI方法并不能矫正严重的批次效应。ArchR使用了最初为了scRNA-seq开发的<a target="_blank" rel="noopener" href="https://github.com/immunogenomics/harmony">Harmony</a>进行批次效应矫正。 ArchR提供了专门的封装函数<code>addHarmony()</code>将降维后的对象直接传递给<code>Harmony::HarmonyMatrix()</code>, 额外参数可以通过<code>...</code>进行传递。更详细的信息可以用<code>?addHarmony()</code>进行了解。用户需要根据他们的特定目的注意批次效应校正的结果。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addHarmony<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span></span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;Harmony&quot;</span><span class="punctuation">,</span></span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Sample&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"><span class="comment">## Harmony 1/10</span></span><br><span class="line"><span class="comment">## Harmony 2/10</span></span><br><span class="line"><span class="comment">## Harmony 3/10</span></span><br><span class="line"><span class="comment">## Harmony converged after 3 iterations</span></span><br></pre></td></tr></table></figure>

<p>这一步会在我们的<code>projHeme2</code>创建一个新的名为<code>Harmony</code>的<code>reduceDims</code>对象，可以通过<code>projHeme2@reducedDims$Harmony</code>访问。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/05/24/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%9B%9B%E7%AB%A0)-sanalysis-sc-atac-seq-with-archr-chapter4/" data-id="clm1ywdjx00cmecok493874wf" data-title="使用ArchR分析单细胞ATAC-seq数据(第四章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第三章)-analysis-sc-atac-seq-with-archr-chapter3" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/05/23/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B8%89%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter3/" class="article-date">
  <time class="dt-published" datetime="2020-05-23T18:24:13.183Z" itemprop="datePublished">2020-05-23</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/05/23/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B8%89%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter3/">使用ArchR分析单细胞ATAC-seq数据(第三章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第3章-创建ArchRProject"><a href="#第3章-创建ArchRProject" class="headerlink" title="第3章: 创建ArchRProject"></a>第3章: 创建ArchRProject</h1><p><code>ArchRProject</code>对象允许我们将多个Arrow文件整理到单个项目之中。<code>ArchRProject</code>比较小能够直接存放在内存中。通过操作<code>ArchRProject</code>，我们能够快速读取Arrow文件里的数据，并将修改后的数据写回到Arrow文件里。因此你几乎可以在这一章中找到后续分析所要用到的所有函数。此外，由于<code>ArchRProject</code>文件能够进行保存并在之后进行读取，那么也就保证了分析的连续性，同时也方便合作者之间的相互交流。这一章主要介绍如何创建<code>ArchRProject</code>对象并和它进行交互。</p>
<h2 id="3-1-创建一个ArchRProject"><a href="#3-1-创建一个ArchRProject" class="headerlink" title="3.1 创建一个ArchRProject"></a>3.1 创建一个ArchRProject</h2><p><code>ArchRProject</code>至少需要设置两个参数，<code>ArrowFiles</code>和<code>outputDirectory</code>. 其中ArrowFilesArrow文件的文件路径，如果没有，清先阅读第一章将BAM&#x2F;Fragments文件转成Arrow文件。<code>outputDirectory</code>指的是下游分析得到的文件的保存路径。ArchR会自动将之前提供的<code>geneAnnotation</code>和<code>genomeAnnotation</code>和新的<code>ArchRProject</code>项目进行关联。这些在我们在之前运行<code>addArchRGenome(&quot;hg19&quot;)</code>的时候被保存在环境变量中。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">projHeme1 <span class="operator">&lt;-</span> ArchRProject<span class="punctuation">(</span></span><br><span class="line">  ArrowFiles <span class="operator">=</span> ArrowFiles<span class="punctuation">,</span></span><br><span class="line">  outputDirectory <span class="operator">=</span> <span class="string">&quot;HemeTutorial&quot;</span><span class="punctuation">,</span></span><br><span class="line">  copyArrows <span class="operator">=</span> <span class="literal">TRUE</span> <span class="comment">#This is recommened so that if you modify the Arrow files you have an original copy for later usage.</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>因为这是我们hematopoiesis项目的第一轮分析，所以我们将<code>ArchRProject</code>命名为”proJHeme1”。在后续的分析中，我们会不断修改并更新这个<code>ArchRProject</code>对象，我们需要使用不同的变量名对项目进行跟踪。如果变量名始终都是同一个，那么就可能出现明明是同一个命令，有些时候能运行成功有些时候运行失败的情况，原因可能就是某些中间步骤被漏掉了，导致同一个变量名的背后的实际内容其实是不一样的。</p>
<p>我们可以直接检查下<code>ArchRProject</code>的内容</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">projHeme1</span><br><span class="line"><span class="comment">#输出信息</span></span><br><span class="line"><span class="comment"># class: ArchRProject</span></span><br><span class="line"><span class="comment"># outputDirectory: /path/to/HemeTutorial</span></span><br><span class="line"><span class="comment"># samples(3): scATAC_BMMC_R1 scATAC_CD34_BMMC_R1 scATAC_PBMC_R1</span></span><br><span class="line"><span class="comment"># sampleColData names(1): ArrowFiles</span></span><br><span class="line"><span class="comment"># cellColData names(15): Sample TSSEnrichment ... DoubletEnrichment BlacklistRatio</span></span><br><span class="line"><span class="comment"># numberOfCells(1): 10661</span></span><br><span class="line"><span class="comment"># medianTSS(1): 16.832</span></span><br><span class="line"><span class="comment"># medianFrags(1): 3050</span></span><br></pre></td></tr></table></figure>

<p>从上面的输入中，我们可以发现<code>ArchRProject</code>在初始化的时候增加了一些额外信息</p>
<ol>
<li>outputDirectory: 输出目录</li>
<li>samples： 样本名</li>
<li>sampleColData: 记录每个样本的元信息</li>
<li>cellColData: 记录每个细胞的元信息，第二章计算的”DoubleEnrichment”,”DoubletScore”就存放在此处</li>
<li>numberOfCells: 项目总的合格细胞数，也就是不包括之前没有通过质控，或者被认为doublet的细胞<br>1.medianTSS&#x2F;medianFrags: 所有细胞的TSS值和Fragments数的中位数</li>
</ol>
<p>我们可以通过下面这行代码了解下<code>ArchRProject</code>在R中会使用多少内存</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">paste0<span class="punctuation">(</span><span class="string">&quot;Memory Size = &quot;</span><span class="punctuation">,</span> <span class="built_in">round</span><span class="punctuation">(</span>object.size<span class="punctuation">(</span>projHeme1<span class="punctuation">)</span> <span class="operator">/</span> <span class="number">10</span><span class="operator">^</span><span class="number">6</span><span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot; MB&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># &quot;Memory Size = 37.477 MB&quot;</span></span><br></pre></td></tr></table></figure>

<p>我们还可以检查当前的<code>ArchRProject</code>中存放了哪些矩阵数据，这些数据一旦增加后就可以在下游分析中使用。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">getAvailableMatrices<span class="punctuation">(</span>projHeme1<span class="punctuation">)</span></span><br><span class="line"><span class="comment"># &quot;GeneScoreMatrix&quot; &quot;TileMatrix&quot;</span></span><br></pre></td></tr></table></figure>

<h2 id="3-2-操作ArchRProject对象"><a href="#3-2-操作ArchRProject对象" class="headerlink" title="3.2 操作ArchRProject对象"></a>3.2 操作ArchRProject对象</h2><p>既然已经在上一节创建了<code>ArchRProject</code>，这一节就来介绍一些比较常用的操作来从该对象中获取我们需要的信息和保存我们的数据</p>
<h3 id="例1-使用-访问cellColData"><a href="#例1-使用-访问cellColData" class="headerlink" title="例1: 使用$访问cellColData"></a>例1: 使用<code>$</code>访问<code>cellColData</code></h3><p>我们可以使用<code>$</code>访问细胞名(cellNames), 样本名(sample)和TSS富集得分(TSS enrichment Scores)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">head<span class="punctuation">(</span>projHeme1<span class="operator">$</span>cellNames<span class="punctuation">)</span></span><br><span class="line">head<span class="punctuation">(</span>projHeme1<span class="operator">$</span>Sample<span class="punctuation">)</span></span><br><span class="line">quantile<span class="punctuation">(</span>projHeme1<span class="operator">$</span>TSSEnrichment<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>除了这些以外，你还可以利用R的补全功能，在输入<code>projHeme1$</code>后看看它还能接哪些内容。</p>
<h3 id="例2-从ArchRProject中提取部分细胞"><a href="#例2-从ArchRProject中提取部分细胞" class="headerlink" title="例2: 从ArchRProject中提取部分细胞"></a>例2: 从<code>ArchRProject</code>中提取部分细胞</h3><p>我们可以将以前学习的R语言向量&#x2F;矩阵&#x2F;数据框提取数据的方法无缝的迁移到<code>ArchRProject</code>上，但区别在于ArchR并不会直接提取数据，而是构建一个新的<code>ArchRProject</code>对象。</p>
<p>最简单的方法是根据位置和细胞名</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 根据位置</span></span><br><span class="line">projHeme1<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">100</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line"><span class="comment"># 根据细胞名</span></span><br><span class="line">projHeme1<span class="punctuation">[</span>projHeme1<span class="operator">$</span>cellNames<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">100</span><span class="punctuation">]</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<p>复杂一些就是先根据逻辑判断提取细胞名，接着根据细胞名提取数据。例如挑选<code>scATAC_BMMC_R1</code>的细胞，或是TSSEnrichment大于8的细胞。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># sample name</span></span><br><span class="line">idxSample <span class="operator">&lt;-</span> BiocGenerics<span class="operator">::</span>which<span class="punctuation">(</span>projHeme1<span class="operator">$</span>Sample <span class="operator">%in%</span> <span class="string">&quot;scATAC_BMMC_R1&quot;</span><span class="punctuation">)</span></span><br><span class="line">cellsSample <span class="operator">&lt;-</span> projHeme1<span class="operator">$</span>cellNames<span class="punctuation">[</span>idxSample<span class="punctuation">]</span></span><br><span class="line">projHeme1<span class="punctuation">[</span>cellsSample<span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line"><span class="comment"># TSS enrichment score</span></span><br><span class="line">idxPass <span class="operator">&lt;-</span> which<span class="punctuation">(</span>projHeme1<span class="operator">$</span>TSSEnrichment <span class="operator">&gt;=</span> <span class="number">8</span><span class="punctuation">)</span></span><br><span class="line">cellsPass <span class="operator">&lt;-</span> projHeme1<span class="operator">$</span>cellNames<span class="punctuation">[</span>idxPass<span class="punctuation">]</span></span><br><span class="line">projHeme1<span class="punctuation">[</span>cellsPass<span class="punctuation">,</span> <span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<h3 id="例3-在ArchRProject增加数据"><a href="#例3-在ArchRProject增加数据" class="headerlink" title="例3: 在ArchRProject增加数据"></a>例3: 在<code>ArchRProject</code>增加数据</h3><p>我们可以通过在<code>cellColData</code>增加新列的方式来为我们项目中的细胞增加额外的元信息。</p>
<p>例如，我们可以从原来的样本名中提取更易阅读的部分添加到<code>cellColData</code>中。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">bioNames <span class="operator">&lt;-</span> gsub<span class="punctuation">(</span><span class="string">&quot;_R2|_R1|scATAC_&quot;</span><span class="punctuation">,</span><span class="string">&quot;&quot;</span><span class="punctuation">,</span>projHeme1<span class="operator">$</span>Sample<span class="punctuation">)</span></span><br><span class="line">head<span class="punctuation">(</span>bioNames<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#[1] &quot;BMMC&quot; &quot;BMMC&quot; &quot;BMMC&quot; &quot;BMMC&quot; &quot;BMMC&quot; &quot;BMMC&quot;</span></span><br></pre></td></tr></table></figure>

<p>一种方式是使用<code>$</code>直接往<code>cellColData</code>添加列</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme1<span class="operator">$</span>bioNames <span class="operator">&lt;-</span> bioNames</span><br></pre></td></tr></table></figure>

<p>或者是用<code>addCellColData()</code>函数往<code>cellColData</code>中添加新的列。使用<code>addCellColData()</code>的好处在于，你可以只为部分增加信息。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">bioNames <span class="operator">&lt;-</span> bioNames<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">10</span><span class="punctuation">]</span></span><br><span class="line">cellNames <span class="operator">&lt;-</span> projHeme1<span class="operator">$</span>cellNames<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">10</span><span class="punctuation">]</span></span><br><span class="line">projHeme1 <span class="operator">&lt;-</span> addCellColData<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme1<span class="punctuation">,</span> data <span class="operator">=</span> paste0<span class="punctuation">(</span>bioNames<span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    cells <span class="operator">=</span> cellNames<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;bioNames2&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><code>ArchR</code>默认会用<code>NA</code>填充其余部分。正因如此，如果我们对比<code>bioNames</code>和<code>bioNames2</code>时，你就可以发现<code>NA</code>填充了那些没有被选择的部分</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">getCellColData<span class="punctuation">(</span>projHeme1<span class="punctuation">,</span> select <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;bioNames&quot;</span><span class="punctuation">,</span> <span class="string">&quot;bioNames2&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">DataFrame with <span class="number">10661</span> rows and <span class="number">2</span> columns</span><br><span class="line"><span class="comment">#                                     bioNames   bioNames2</span></span><br><span class="line"><span class="comment">#                                  &lt;character&gt; &lt;character&gt;</span></span><br><span class="line"><span class="comment">#scATAC_BMMC_R1#TTATGTCAGTGATTAG-1        BMMC        BMMC</span></span><br><span class="line"><span class="comment">#scATAC_BMMC_R1#AAGATAGTCACCGCGA-1        BMMC        BMMC</span></span><br><span class="line"><span class="comment">#scATAC_BMMC_R1#GCATTGAAGATTCCGT-1        BMMC        BMMC</span></span><br><span class="line"><span class="comment">#scATAC_BMMC_R1#TATGTTCAGGGTTCCC-1        BMMC        BMMC</span></span><br><span class="line"><span class="comment">#scATAC_BMMC_R1#TCCATCGGTCCCGTGA-1        BMMC        BMMC</span></span><br><span class="line"><span class="comment">#...                                       ...         ...</span></span><br><span class="line"><span class="comment">#scATAC_PBMC_R1#GCTGCGAAGATCCGAG-1        PBMC          NA</span></span><br><span class="line"><span class="comment">#scATAC_PBMC_R1#GCAGCTGGTGGCCTTG-1        PBMC          NA</span></span><br><span class="line"><span class="comment">#scATAC_PBMC_R1#GCAGATTGTACGCAAG-1        PBMC          NA</span></span><br><span class="line"><span class="comment">#scATAC_PBMC_R1#TTCGTTACATTGAACC-1        PBMC          NA</span></span><br><span class="line"><span class="comment">#scATAC_PBMC_R1#CGCTATCGTGAGGTCA-1        PBMC          N</span></span><br></pre></td></tr></table></figure>


<h3 id="例4-从cellColData中提取列"><a href="#例4-从cellColData中提取列" class="headerlink" title="例4: 从cellColData中提取列"></a>例4: 从<code>cellColData</code>中提取列</h3><p>ArchR提供了<code>getCellColData</code>用于从<code>ArchRProject</code>中提取元信息列。你可以认为这是<code>$</code>的功能增强版，因为<code>$</code>只能提取一列，而<code>getCellColData</code>可以提取多列信息，此外还有许多神奇的操作。</p>
<p>例如我们可以根据列名获取数据，比方说每个细胞的唯一fragment数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">df <span class="operator">&lt;-</span> getCellColData<span class="punctuation">(</span>projHeme1<span class="punctuation">,</span> select <span class="operator">=</span> <span class="string">&quot;nFrags&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>甚至，你还能在列名中添加一些运算操作，这样会直接输出运算后的结果</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">df <span class="operator">&lt;-</span> getCellColData<span class="punctuation">(</span>projHeme1<span class="punctuation">,</span> select <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;log10(nFrags)&quot;</span><span class="punctuation">,</span> <span class="string">&quot;nFrags - 1&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="例5-绘制质控信息和TSS富集得分的对比图"><a href="#例5-绘制质控信息和TSS富集得分的对比图" class="headerlink" title="例5: 绘制质控信息和TSS富集得分的对比图"></a>例5: 绘制质控信息和TSS富集得分的对比图</h3><p>根据上述提供的案例，我们可以很容易获取到每个细胞的标准scATAC-seq质控信息。我们发现目前最靠谱的质控标准是TSS富集的得分(评估ATAC-seq数据的信噪比)和唯一比对数(比对数如果不够，那么该细胞也没有分析的价值)</p>
<p>我们先用<code>getCellColData</code>提取我们需要的两列，其中nFrages需要进行log10运算</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">df <span class="operator">&lt;-</span> getCellColData<span class="punctuation">(</span>projHeme1<span class="punctuation">,</span> select <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;log10(nFrags)&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TSSEnrichment&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>接下来，我们基于TSS富集得分和唯一比对进行绘图，函数<code>ggPoint</code>是ArchR对<code>ggplot2</code>的点图相关函数的封装。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">p <span class="operator">&lt;-</span> ggPoint<span class="punctuation">(</span></span><br><span class="line">    x <span class="operator">=</span> df<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">    y <span class="operator">=</span> df<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">    colorDensity <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span></span><br><span class="line">    continuousSet <span class="operator">=</span> <span class="string">&quot;sambaNight&quot;</span><span class="punctuation">,</span></span><br><span class="line">    xlabel <span class="operator">=</span> <span class="string">&quot;Log10 Unique Fragments&quot;</span><span class="punctuation">,</span></span><br><span class="line">    ylabel <span class="operator">=</span> <span class="string">&quot;TSS Enrichment&quot;</span><span class="punctuation">,</span></span><br><span class="line">    xlim <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span>log10<span class="punctuation">(</span><span class="number">500</span><span class="punctuation">)</span><span class="punctuation">,</span> quantile<span class="punctuation">(</span>df<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span> probs <span class="operator">=</span> <span class="number">0.99</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    ylim <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> quantile<span class="punctuation">(</span>df<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">,</span> probs <span class="operator">=</span> <span class="number">0.99</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span> <span class="operator">+</span> geom_hline<span class="punctuation">(</span>yintercept <span class="operator">=</span> <span class="number">4</span><span class="punctuation">,</span> lty <span class="operator">=</span> <span class="string">&quot;dashed&quot;</span><span class="punctuation">)</span> <span class="operator">+</span> geom_vline<span class="punctuation">(</span>xintercept <span class="operator">=</span> <span class="number">3</span><span class="punctuation">,</span> lty <span class="operator">=</span> <span class="string">&quot;dashed&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-63a17393b7664d39a148f25c1555a3f1.png" alt="dot plot"></p>
<p>该图的主要作用是找到高质量的细胞。你可能会注意到一些细胞已经在我们之前创建Arrow文件时被设置的阈值<code>filterTSS, filterFargs</code>过滤掉了。然而，如果我们觉得之前的质控过滤对于这个样本不太合适，我们可以根据该图调整阈值，有必要的话还可以重新产生Arrow文件。</p>
<p>尽管可以使用PDF的方式保存上图，但是对于这种点特别多的图，还是用PNG形式比较好。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">png<span class="punctuation">(</span><span class="string">&quot;TSS-vs-Frags.png&quot;</span><span class="punctuation">)</span></span><br><span class="line">plot<span class="punctuation">(</span>p<span class="punctuation">)</span></span><br><span class="line">dev.off<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">getwd<span class="punctuation">(</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="3-3-为ArchRProject每个样本绘制统计信息"><a href="#3-3-为ArchRProject每个样本绘制统计信息" class="headerlink" title="3.3 为ArchRProject每个样本绘制统计信息"></a>3.3 为ArchRProject每个样本绘制统计信息</h2><p>如果一个项目中有多个不同的样本，我们很自然地就会想比较这些样本。所谓一图胜千言，ArchR提供了两种小提琴图(violin plot)和山脊图(ridge plot)用来展示不同组之间的信息。这两种类型的图可以用一个函数<code>plotGroups</code>进行绘制。当然这函数除了根据样本进行分组绘图外，还可以使用下游分析得到的分组信息（例如聚类）。</p>
<p><strong>例1</strong>: 根据TSS富集得分为每个样本绘制山脊图。</p>
<p>绘制山脊图时，需要设置<code>plotAs = &quot;ridges&quot;</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotGroups<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme1<span class="punctuation">,</span></span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Sample&quot;</span><span class="punctuation">,</span></span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span></span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;TSSEnrichment&quot;</span><span class="punctuation">,</span></span><br><span class="line">    plotAs <span class="operator">=</span> <span class="string">&quot;ridges&quot;</span></span><br><span class="line">   <span class="punctuation">)</span></span><br><span class="line">p1</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-96a12652a7b74424a156cc0eb8b4c236.png" alt="ridges of TSSEnrichment"></p>
<p><strong>例2</strong>: 根据TSS富集得分为每个样本绘制小提琴图。ArchR绘制的小提琴图额外叠加了一层箱线图(a box-and-whiskers plot in the style of Tukey)，中间水平的三条线表示数据中的1&#x2F;4分位数，中位数和3&#x2F;4分位数。最下方是最小值，最上方是最大值（或者是1.5倍的IQR, interquartile range, 1&#x2F;4分位数与3&#x2F;4分位数的距离）</p>
<p>绘制小提琴图时，需要设置<code>plotAs = &quot;violin&quot;</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">p2 <span class="operator">&lt;-</span> plotGroups<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme1<span class="punctuation">,</span></span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Sample&quot;</span><span class="punctuation">,</span></span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span></span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;TSSEnrichment&quot;</span><span class="punctuation">,</span></span><br><span class="line">    plotAs <span class="operator">=</span> <span class="string">&quot;violin&quot;</span><span class="punctuation">,</span></span><br><span class="line">    alpha <span class="operator">=</span> <span class="number">0.4</span><span class="punctuation">,</span></span><br><span class="line">    addBoxPlot <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line">   <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-1d2e7b23376b474bad74f0dc9926c530.png" alt="Violin plot"></p>
<p><strong>例3</strong>: 根据log10(unique nuclear fragments)为每个样本绘制山脊图。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">p3 <span class="operator">&lt;-</span> plotGroups<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme1<span class="punctuation">,</span></span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Sample&quot;</span><span class="punctuation">,</span></span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span></span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;log10(nFrags)&quot;</span><span class="punctuation">,</span></span><br><span class="line">    plotAs <span class="operator">=</span> <span class="string">&quot;ridges&quot;</span></span><br><span class="line">   <span class="punctuation">)</span></span><br><span class="line">p3</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-dca5dd1a2e4e4293a50dce27b44e7836.png" alt="ridge plot"></p>
<p><strong>例4</strong>:  根据log10(unique nuclear fragments)为每个样本绘制小提琴图。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">p4 <span class="operator">&lt;-</span> plotGroups<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme1<span class="punctuation">,</span></span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Sample&quot;</span><span class="punctuation">,</span></span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span></span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;log10(nFrags)&quot;</span><span class="punctuation">,</span></span><br><span class="line">    plotAs <span class="operator">=</span> <span class="string">&quot;violin&quot;</span><span class="punctuation">,</span></span><br><span class="line">    alpha <span class="operator">=</span> <span class="number">0.4</span><span class="punctuation">,</span></span><br><span class="line">    addBoxPlot <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line">   <span class="punctuation">)</span></span><br><span class="line">p4</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-6c02fbc2c22e4cda8ee65cebcb8f3535.png" alt="Violin Plot"></p>
<p>我们可以使用<code>plotPDF()</code>将上面的图绘制在一个PDF中</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span>p2<span class="punctuation">,</span>p3<span class="punctuation">,</span>p4<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;QC-Sample-Statistics.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme1<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">4</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">4</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="3-4-绘制样本的TSS富集谱和Fragment大小分布"><a href="#3-4-绘制样本的TSS富集谱和Fragment大小分布" class="headerlink" title="3.4 绘制样本的TSS富集谱和Fragment大小分布"></a>3.4 绘制样本的TSS富集谱和Fragment大小分布</h2><p>鉴于数据的存储和读取方式， ArchR能够快速的从Arrow文件中计算出fragment大小分布和TSS富集谱。</p>
<p><strong>Fragments大小分布</strong>:  我们使用<code>plotFragmentSizes</code>函数为绘制所有样本的fragment大小分布。ATAC-seq数据中的fragment大小分布可能会因样本、细胞类型和批次不同，但这和数据质量并不是严格相关。比如说我们的数据在不同组之间就存在一些差别。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotFragmentSizes<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme1<span class="punctuation">)</span></span><br><span class="line">p1</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-da8c354686ba405fab5cd2888ec80d4a.png" alt="Fragment size distribution"></p>
<p><strong>TSS富集谱</strong>: 我们使用<code>plotTSSEnrichment()</code>函数绘制TSS富集谱。TSS富集谱需要在中心位置有一个明显的峰，在峰的右边还会有一个核小体引起的小隆起（约147 bp）。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">p2 <span class="operator">&lt;-</span> plotTSSEnrichment<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme1<span class="punctuation">)</span></span><br><span class="line">p2</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-5aeda9f111d24a6da0762f9395fcb500.png" alt="TSS enrichment profiles"></p>
<p>和之前一样，我们可以使用<code>plotPDF()</code>将图形以可编辑的矢量图进行保存。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span>p2<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;QC-Sample-FragSizes-TSSProfile.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme1<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="3-5-保存和加载ArchRProject"><a href="#3-5-保存和加载ArchRProject" class="headerlink" title="3.5 保存和加载ArchRProject"></a>3.5 保存和加载<code>ArchRProject</code></h2><p>ArchR提供了<code>saveArchRProject</code>函数让我们能将内存中<code>ArchRProject</code>对象的保存到本地，并在之后进行读取，当然保存到本地的数据也就能分享给其他人。从根本上来讲，<code>ArchRProject</code>是一个指向多个Arrow文件的对象，因此使用<code>saveArchRProject()</code>保存<code>ArchRProject</code>的过程实际上分为两步</p>
<ol>
<li>将当前的Arrow文件复制到目标输出目录下<code>outputDirectory</code>,确保它们能和新的<code>ArchRProject</code>关联</li>
<li>将目标<code>ArchRProject</code>保存到<code>outputDirectory</code>目录下</li>
</ol>
<p>我们这里以<code>projHeme1</code>为例介绍<code>saveArchRProject()</code>的用法</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">saveArchRProject<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme1<span class="punctuation">,</span> outputDirectory <span class="operator">=</span> <span class="string">&quot;Save-ProjHeme1&quot;</span><span class="punctuation">,</span> load <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>运行结束后，<code>outputDirectory</code>对应的目录下就会保存相应的Arrow文件和<code>projHeme1</code>的Rds文件。</p>
<p><strong>重要知识点</strong>: 这一步骤<strong>不会</strong>自动更新你当前R session里活跃的<code>ArchRProject</code>对象。也就说，当前R session里的<code>projHeme1</code>对象还是指向原来的Arrow文件，而不会指向拷贝到<code>outputDirectory</code>目录下的Arrow文件。</p>
<p>换言之，如果我们想要将当前的项目复制到新的目录，你可以设置<code>load=TRUE</code>。下面的代码运行结束后得到的<code>projHemeNew</code>就是指向拷贝后Arrow文件的<code>ArchRProject</code>对象。你可以使用原来的<code>projHeme1</code>，这就会覆盖当前R session的<code>projHeme1</code>。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHemeNew <span class="operator">&lt;-</span> saveArchRProject<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme1<span class="punctuation">,</span> outputDirectory <span class="operator">=</span> <span class="string">&quot;Save-ProjHeme1&quot;</span><span class="punctuation">,</span> load <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="3-6-使用ArchRProject过滤doublets"><a href="#3-6-使用ArchRProject过滤doublets" class="headerlink" title="3.6 使用ArchRProject过滤doublets"></a>3.6 使用ArchRProject过滤doublets</h2><p>当我们使用<code>addDOubletScores()</code>增加了预测的doublet信息后，我们就能使用<code>filterDoublets()</code>过滤这些预测的doublets。这个函数里的一个关键参数是<code>filterRatio</code>, 由于不同样本的制备过程中的细胞上样量不同，那么不同样本的doublet也就不同，为了保证不同样本的过滤后结果保持一致，所以引入了<code>filterRatio</code>。该值越高，被认为是doublet而被过滤的细胞也就越多。</p>
<p>了解了<code>filterRatio</code>的作用后，我们再来看它的定义，它是根据未被过滤的细胞数计算的需要被过滤doublet的最大值(the maximum ratio of predicted doublets to filter based on the number of pass-filter cells)。举个例子。假如这里有5000个细胞，需要被过滤doublet的最大值就等于 <code>filterRatio * 5000^2 / (100000)</code>, 也就是<code>filterRatio * 5000 * 0.05</code>。</p>
<p>如果还是不太了解，也不太纠结，毕竟大部分时候我们用的都是默认参数。我们现在只需要知道这个<code>filterRatio</code>参数很重要，得根据后续结果进行调整。这里就直接运行<code>filterDoublets</code>对细胞进行过滤。出于教学目的，我们将处理结果保存为一个新的<code>ArchRProject</code>对象，也就是<code>projHeme2</code>，实际分析中，你可以使用原来的名字覆盖之前的结果，降低内存消耗。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> filterDoublets<span class="punctuation">(</span>projHeme1<span class="punctuation">)</span></span><br><span class="line"><span class="comment"># Filtering 410 cells from ArchRProject!</span></span><br><span class="line"><span class="comment">#	scATAC_BMMC_R1 : 243 of 4932 (4.9%)</span></span><br><span class="line"><span class="comment">#	scATAC_CD34_BMMC_R1 : 107 of 3275 (3.3%)</span></span><br><span class="line"><span class="comment">#	scATAC_PBMC_R1 : 60 of 2454 (2.4%)</span></span><br></pre></td></tr></table></figure>

<p>之前的<code>projHeme1</code>有10,661个细胞，经过上一步的过滤，<code>projHeme2</code>剩下了10,251个细胞，也就意味着有410个细胞(3.85%)是doublet。而且每个样本的过滤比例并不相同，这就是<code>filterRatio</code>的作用。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">projHeme2</span><br><span class="line"><span class="comment"># ...</span></span><br><span class="line"><span class="comment"># numberOfCells(1): 10251</span></span><br><span class="line"><span class="comment"># medianTSS(1): 16.856</span></span><br><span class="line"><span class="comment"># medianFrags(1): 2991</span></span><br></pre></td></tr></table></figure>

<p>如果你想过滤更多<code>ArchRProject</code>里的细胞，你可以设置更高的<code>filterRatio</code>。和参数调整的更多信息可以用<code>?filterDoublets</code>查看。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">projHemeTmp <span class="operator">&lt;-</span> filterDoublets<span class="punctuation">(</span>projHeme1<span class="punctuation">,</span> filterRatio <span class="operator">=</span> <span class="number">1.5</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># Filtering 614 cells from ArchRProject!</span></span><br><span class="line"><span class="comment">#	scATAC_BMMC_R1 : 364 of 4932 (7.4%)</span></span><br><span class="line"><span class="comment">#	scATAC_CD34_BMMC_R1 : 160 of 3275 (4.9%)</span></span><br><span class="line"><span class="comment">#	scATAC_PBMC_R1 : 90 of 2454 (3.7%)</span></span><br></pre></td></tr></table></figure>

<p>最后，既然<code>projHemeTmp</code>只是说明用的临时对象，我们现在可以将其从R session中删除了。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">rm<span class="punctuation">(</span>projHemeTmp<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/05/23/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B8%89%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter3/" data-id="clm1ywdjh00bdecok66mb2vu4" data-title="使用ArchR分析单细胞ATAC-seq数据(第三章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第二章)-analysis-sc-atac-seq-with-archr-chapter2" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/05/22/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%BA%8C%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter2/" class="article-date">
  <time class="dt-published" datetime="2020-05-22T19:07:27.475Z" itemprop="datePublished">2020-05-22</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/05/22/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%BA%8C%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter2/">使用ArchR分析单细胞ATAC-seq数据(第二章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第2章-使用ArchR推断Doublet"><a href="#第2章-使用ArchR推断Doublet" class="headerlink" title="第2章: 使用ArchR推断Doublet"></a>第2章: 使用ArchR推断Doublet</h1><p>单细胞数据分析中的一个重要问题就是”doublet”对分析的影响。”doublet”指的是单个液滴(droplet)捕获了一个条形码珠(barcode bead)和多个细胞核。这会导致原本来自于多个细胞的read结果被当成一个细胞，结果原来两个细胞被平均成一个细胞。我们在这一章中将会介绍如何使用计算的方法鉴定并过滤doublet。</p>
<h2 id="2-1-ArchR是如何鉴定doublet"><a href="#2-1-ArchR是如何鉴定doublet" class="headerlink" title="2.1 ArchR是如何鉴定doublet"></a>2.1 ArchR是如何鉴定doublet</h2><p>几乎所有平台得到的单细胞数据都或多或少的存在doublet。所谓的doublet就是一个液滴中混入了多个细胞的细胞核，导致原本的多个细胞被认为是一个细胞。在10X Genomics平台，doublets的细胞比例和一次反应中的细胞数有关，细胞越多，doublet也就越多。不过即便你按照标准流程，也依旧会存在一定比例的doublet，而只要有5%的doublet数据，就会对后续的聚类分析造成影响，尤其是后续的发育&#x2F;谱系分析。因为doublet看起来像是两类细胞的中间状态，但是当你不知道这是doublet造成的假象时，你就会得到错误的结论。</p>
<p>为了预测哪些细胞才是真的doublet，我们通过对不同细胞进行组合得到模拟的doublet(in silico doublets)，然后将这些模拟的doublet和原来的细胞一同投影到UMAP，接着确定和它们最近的细胞。通过不断地迭代对该步骤，我们就可以找到数据中那些与模拟doublet信号最接近的细胞。这些细胞就是潜在的doublet。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-6d08f01cee2144cb85d5b6093fd6a98b.png" alt="Doublet分析原理"></p>
<p>为了开发ArchR的doublet鉴定算法并验证它的可靠性，我们将10个不同类型的细胞进行混样测序。在没有doublet的情况下，我们的scATAC-seq数据最终应该有10个不同的细胞类型。但是当我们故意在10X Genomics scATAC-seq试剂中加入过量的细胞(25,000&#x2F;per reaction), 我们会得到许多doublet。我们使用<a target="_blank" rel="noopener" href="https://github.com/statgen/demuxlet">demuxlet</a>根据一个细胞里是否包括两种不同细胞的基因型来判断该细胞是否是doublet。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-70166ecd270f42f8a0d55ccbe1d6c19e.png" alt="colored by cell type"></p>
<p>如上图所示，我们的真实结果(ground truth)被预测的doublet覆盖。ROC曲线中的AUC(area under the curve) &gt; 0.90（见注1)</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-4d22f53cb9eb45f985998d899e6b2a90.png" alt="grouded truth"><br>ArchR通过计算的方法移除这些doublets后，我们数据整体结构有了明显的变化，符合预期，也就是有10个不同的细胞类型。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-3abae0bde7cd4e76b092dbe3fd454a63.png" alt="Post-ArchR doublet Removal"></p>
<p>注1: ROC曲线全称为受试者工作特征曲线 （receiver operating characteristic curve），它是根据一系列不同的二分类方式（分界值或决定阈），以真阳性率（敏感性）为纵坐标，假阳性率（1-特异性）为横坐标绘制的曲线。AUC就是衡量学习器优劣的一种性能指标，AUC越接近于1.0，表示检测方法真实性越高， 等于0.5表示真实性越低</p>
<h2 id="2-2-使用ArchR推断scATAC-seq-doublets"><a href="#2-2-使用ArchR推断scATAC-seq-doublets" class="headerlink" title="2.2 使用ArchR推断scATAC-seq doublets"></a>2.2 使用ArchR推断scATAC-seq doublets</h2><p>ArchR默认使用ArchR手稿中的doublet参数。我们建议用户检查移除doublet前后的数据变化，理解移除doulet对细胞的影响，根据结果对已有的参数进行调整，而不是生搬硬我们的参数设置。我们也会展示一些</p>
<p>我们使用ArchR的<code>addDoubleScores()</code>函数来移除doublet。对于教程中使用的数据，每个样本大约需要2到5分钟时间进行处理，最后每个细胞的doublet得分会添加到Arrow文件中。 你可以使用<code>?addDoubletScores</code>了解该函数的每个参数的意义。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">doubScores <span class="operator">&lt;-</span> addDoubletScores<span class="punctuation">(</span></span><br><span class="line">    input <span class="operator">=</span> ArrowFiles<span class="punctuation">,</span></span><br><span class="line">    k <span class="operator">=</span> <span class="number">10</span><span class="punctuation">,</span> <span class="comment">#Refers to how many cells near a &quot;pseudo-doublet&quot; to count.</span></span><br><span class="line">    knnMethod <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">,</span> <span class="comment">#Refers to the embedding to use for nearest neighbor search with doublet projection.</span></span><br><span class="line">    LSIMethod <span class="operator">=</span> <span class="number">1</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们以其中一个样本运行过程的输出日志信息进行介绍，这里的<code>R^2</code>用于描述样本中的细胞异质性，如果该数值非常小（例如小于0.9），说明该样本的细胞都非常相似。那么使用模拟的方法去鉴定doublet就不太合适了。这个很好理解，如果所有细胞都表达一个基因，并且表达量是1，那么你模拟的doublet也会只有一个细胞，且表达量是均值1，结果就是所有细胞都是doublet。在这种情况下，我们推荐跳过doublet预测这一步。或者你可以尝试设置<code>knnMethod = &quot;LSI&quot;</code>,<code>force = TRUE</code>，在LSI子空间中进行投影。（相当于提高分辨率）。无论如何，你都需要手动检查结果，确保运行过程符合预期。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## If there is an issue, please report to github with logFile!</span></span><br><span class="line"><span class="comment">## 2020-04-15 09:28:44 : Batch Execution w/ safelapply!, 0 mins elapsed.</span></span><br><span class="line"><span class="comment">## 2020-04-15 09:28:44 : scATAC_BMMC_R1 (1 of 3) : Computing Doublet Statistics, 0.001 mins elapsed.</span></span><br><span class="line"><span class="comment">## scATAC_BMMC_R1 (1 of 3) : UMAP Projection R^2 = 0.9736</span></span><br><span class="line"><span class="comment">## scATAC_BMMC_R1 (1 of 3) : UMAP Projection R^2 = 0.9736</span></span><br></pre></td></tr></table></figure>

<blockquote>
<p>这里的R^2大于0.9，表示细胞存在异质性。</p>
</blockquote>
<p><code>addDoubletScores</code>在计算doublet得分的时候，还会在”QualityControl”中为每个样本生成三张图（一个样本一个PDF）</p>
<ol>
<li>Doublet Enrichments: 假设doublet符合均匀分布，那么每个细胞附近的模拟doublet数目都差不多。如果一个细胞附近相对于其他细胞有更多的doublet, 就认为它富集了doublet</li>
<li>Doublet Score: 基于均匀分布假设，计算doublet富集显著性，以<code>-log10(binomial adjusted p-value)</code>进行展示。我们更推荐根据doublet enrichment鉴定doublet。</li>
<li>Doublet Density: 模拟的duoblet在二维空间的投影，我们可以直观的了解我们模拟的doublet的分布情况。</li>
</ol>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-59979b1c13a547b58e0e3cf319cdb4d5.png" alt="BMMC"></p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-508cf6034ef84816860d34aed8fd3e10.png" alt="CD34 BMMC"></p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-0447a20a697f46ce9295fa5815687b49.png" alt="PBMC"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/05/22/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%BA%8C%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter2/" data-id="clm1ywdjj00bkecok28jb7tva" data-title="使用ArchR分析单细胞ATAC-seq数据(第二章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第一章)-analysis-sc-atac-seq-with-archr-chapter1" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/05/21/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B8%80%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter1/" class="article-date">
  <time class="dt-published" datetime="2020-05-21T13:08:26.879Z" itemprop="datePublished">2020-05-21</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/05/21/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B8%80%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter1/">使用ArchR分析单细胞ATAC-seq数据(第一章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第1章-ArchR基础入门"><a href="#第1章-ArchR基础入门" class="headerlink" title="第1章: ArchR基础入门"></a>第1章: ArchR基础入门</h1><p>这一章将会介绍如何导入数据，如何构建Arrow文件，这是后续ArchR分析的基础。</p>
<h2 id="1-1-ATAC-seq术语介绍"><a href="#1-1-ATAC-seq术语介绍" class="headerlink" title="1.1 ATAC-seq术语介绍"></a>1.1 ATAC-seq术语介绍</h2><p>“<strong>fragment</strong>“是ATAC-seq实验中的一个重要概念，它指的是通过Tn5转座酶对DNA分子进行酶切，然后经由双端测序得到的序列。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-6e2966763aa0479cb68b6157e1e6dd31.png" alt="fragment产生过程"></p>
<p>我们会根据Tn5插入导致的偏移从read比对得到的位置推断出<strong>fragment</strong>的起始和结束位置。根据之前的报道，Tn5转座酶以同源二聚体的形式结合到DNA上，在两个Tn5分子间隔着9-bp的DNA序列。根据这个情况，每个Tn5同源二聚体的结合事件会产生2个<strong>Insertions</strong>，中间隔着9bp。因此，真实的”开放”位置的中心在Tn5二聚体的正中间，而不是Tn5的插入位置。为了尽可能的还原真实情况，我们对Tn5的<strong>Insertions</strong>进行了校正，即正链的插入结果往右移动4bp(+4 bp), 负链的插入结果往左偏移5bp(-5 bp)。这个和<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pubmed/24097267">最早提出的ATAC-seq</a>里的描述是一致的。因此，在<code>ArchR</code>中，”<strong>fragment</strong>“指的是一个<code>table</code>或<code>genomic ranges</code>对象, 记录在染色体上，经过偏移校正后的单碱基起始位置，以及经过偏移校正后单碱基结束位置，每个fragment都会对应唯一的细胞条形码。类似的，”Insertions”这得是偏移校正后的单碱基位置，它位于开放位置的正中心。</p>
<blockquote>
<p>“fragment”和”insertions”这两个词我并没有将其翻译成中文，我觉得这两个单词可能就和PCR一样，当说到它们的时候，脑中会有一个画面描述它们，而不是局限一个词。</p>
</blockquote>
<h2 id="1-2-为什么是用ArchR"><a href="#1-2-为什么是用ArchR" class="headerlink" title="1.2 为什么是用ArchR"></a>1.2 为什么是用ArchR</h2><p>现在其实已经有一些工具能够处理单细胞ATAC-seq数据，我们为什么要额外造一个轮子，开发一个新的项目呢？主要是ArchR提供了其他工具目前尚不能实现的功能</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-b7bbe4ff672b46e0992416e762c03fb6.png" alt="工具对比"></p>
<p>不仅如此，ArchR通过优化数据结构降低了内存消耗，使用并行提高了运行速度，因此保证其性能优于其他同类型工具。在超过70,000个细胞的分析项目中，一些软件需要超过128G的可用内存。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-5fb674263af341d4a0c11339fcb982fb.png" alt="性能对比"></p>
<p>ArchR最初就是根据Unix的笔记本进行设计（我觉得他说的是MacBook Pro），因此对于中等大小的实验（小于100,000个细胞），ArchR能够保证一些特殊分析的运行速度，并能实时的展示结果，让我们能够更深入的和数据进行互动，给出有意义的生物学解释。当然，如果细胞数更多，你最好使用服务器进行分析。ArchR提供了方便的图形导出功能，在服务器处理完项目之后，可以直接下载到本地进行使用。</p>
<p><strong>目前，ArchR并没有针对Windows进行优化</strong>。这句话的意思是指，ArchR的并行策略是基于Unix系统而非Windows系统，因此上述说的性能提升不包括Windows。</p>
<h2 id="1-3-什么是Arrow文件-ArchRProject"><a href="#1-3-什么是Arrow文件-ArchRProject" class="headerlink" title="1.3 什么是Arrow文件&#x2F;ArchRProject"></a>1.3 什么是Arrow文件&#x2F;ArchRProject</h2><p>正如开头所说，ArchR分析项目的基础是Arrow文件。每个Arrow文件记录着每个<strong>独立样本</strong>的所有相关信息（例如元信息、开放的fragment和数据矩阵）。这里说的<strong>独立样本</strong>最好是最详尽的分析单元（例如，一种特定条件下的单个重复）。在创建Arrow文件以及一些附加分析中，ArchR会编辑和更新相应的Arrow文件，在其中添加额外的信息层。值得注意的是，Arrow文件实际指的是磁盘上的文件路径。更确切的说，Arrow文件并不是一个存放在内存中的R语言对象，而是存放在磁盘上HDF5文件。正因如此，我们使用<code>ArchRProject</code>对象用来关联这些Arrow文件，将其关联到单个分析框架中，从而保证在R中能高效访问它们。而这个<code>ArchRProject</code>对象占用内存不多，因此才是存放在内存中的R语言对象。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-4a871115e6a84b64a4f393f6fc7107a8.png" alt="Arrow File"></p>
<p>有一些操作会直接修改Arrow文件，而一些操作会先作用于<code>ArchRproject</code>，接着反过来更新每个相关Arrow文件。因为Arrow文件是以非常大的HDF5格式存放，所以ArchR的<code>get-er</code>函数通过和<code>ArchRProject</code>进行交互获取数据，而<code>add-er</code>函数既能直接在Arrow文件中添加数据，也能直接在<code>ArchRpoject</code>里添加数据，或者通过<code>ArchRpoject</code>向Arrow文件添加数据。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-ee747106c4aa4f5682afa1536819e170.png" alt="ArchRProject"></p>
<h2 id="1-4-ArchR的输入文件类型"><a href="#1-4-ArchR的输入文件类型" class="headerlink" title="1.4 ArchR的输入文件类型"></a>1.4 ArchR的输入文件类型</h2><p>ArchR主要以scATAC-seq原始数据经上游处理后的两种常见输出文件(BAM, fragment)作为输入。Fragment文件记录着scATAC-seq的fragment以及对应的细胞ID，每一行都是一条记录，该文件需要是tabix(见注1)排序并建立索引保证能被高效读取。BAM文件则是二进制格式下的tabix排序文件，记录着scATAC-seq的fragment、原始数据、细胞条形码和其他信息。具体使用何种文件作为输入取决于你的上游处理流程。以10XGenomics的CellRanger软件为例，它的输出文件是fragments，而sci-ATAC-seq流程则输出BAM文件。</p>
<p>ArchR使用<code>scanTabix</code>函数读取fragment文件，使用<code>scanBAM</code>读取BAM文件。输入过程并不会直接读取所有数据，而是每次读取一大块(chunks)，然后将这一块数据转换成fragment格式(见注2)，经过压缩先暂时以HDF5格式保存到本地磁盘上，避免消耗过多的内存。最后，等所有数据都加载完成之后，该数据相关的所有临时HDF5文件会被重新读取，经过组织之后以单个HDF5形式保存为Arrow文件。ArchR之所以能以较小的内存量高效地读取大文件，就是因为它采用的是这种分块处理的方法，由于每一块数据的处理都互不干扰，因此也就能够并行计算。</p>
<ul>
<li>注1: 当以制表符记录基因组位置信息时，我们可以通过压缩让其体积变小(bgzip&#x2F;gzip)，通过建立<a target="_blank" rel="noopener" href="http://www.htslib.org/doc/tabix.html">tabix</a>索引高效访问给定位置的信息。</li>
<li>注2: fragments一共有5列，分别是chromosome, offset-adjusted chromosome <strong>start</strong> position, offset-adjusted chromosome <strong>end</strong> position, and cellular barcode <strong>ID</strong>, duplicate count</li>
</ul>
<h2 id="1-5-开始设置"><a href="#1-5-开始设置" class="headerlink" title="1.5 开始设置"></a>1.5 开始设置</h2><p>在后续分析之前，请先设置好我们的工作目录，设置将要使用的线程数，加载我们的基因和基因组注释。由于每个人的环境都不太一样，所以你后续可能需要用<code>addArchRThreads()</code>修改线程数。默认情况下，ArchR使用系统一半的可用线程，你可以手动进行调整。如果你用的是Windows，那么默认都是1，这是因为ArchR的多线程是基于Unix操作系统。</p>
<p>第零步，安装ArchR。目前ArchR托管在GitHub上，因此无法直接用CRAN或者Bioconductor中直接下载安装。</p>
<p>方法1：适用于网络状态好的情况</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>requireNamespace<span class="punctuation">(</span><span class="string">&quot;devtools&quot;</span><span class="punctuation">,</span> quietly <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span> install.packages<span class="punctuation">(</span><span class="string">&quot;devtools&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>requireNamespace<span class="punctuation">(</span><span class="string">&quot;BiocManager&quot;</span><span class="punctuation">,</span> quietly <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span> install.packages<span class="punctuation">(</span><span class="string">&quot;BiocManager&quot;</span><span class="punctuation">)</span></span><br><span class="line">devtools<span class="operator">::</span>install_github<span class="punctuation">(</span><span class="string">&quot;GreenleafLab/ArchR&quot;</span><span class="punctuation">,</span> ref<span class="operator">=</span><span class="string">&quot;master&quot;</span><span class="punctuation">,</span> repos <span class="operator">=</span> BiocManager<span class="operator">::</span>repositories<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>方法2: 适用于上述方法多次失败的情况</p>
<p>先从Github上下载项目到本地, 大约有262Mb</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git <span class="built_in">clone</span> https://github.com/GreenleafLab/ArchR.git</span><br></pre></td></tr></table></figure>

<p>然后在 R里面进行安装</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;nabor&quot;</span><span class="punctuation">,</span><span class="string">&quot;motifmatchr&quot;</span><span class="punctuation">,</span><span class="string">&quot;chromVAR&quot;</span><span class="punctuation">,</span><span class="string">&quot;ComplexHeatmap&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">install.packages<span class="punctuation">(</span><span class="string">&quot;./ArchR&quot;</span><span class="punctuation">,</span> repo<span class="operator">=</span><span class="literal">NULL</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>接着安装一些额外包</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ArchR<span class="operator">::</span>installExtraPackages<span class="punctuation">(</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>如果安装失败，就手动安装 Seurat, immunogenomics&#x2F;harmony,  immunogenomics&#x2F;presto,  Cairo，shiny, shinythemes, rhandsontable</p>
<p>第一步，我们加载ArchR包。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>ArchR<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>第二步, 我们需要设置ArchR函数的默认线程数。你需要在每个新的R session中都设置该参数，线程多多益善，但是不要超过总线程的3&#x2F;4。因为线程会和内存挂钩，所以线程数越多，内存相应使用的也越多。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">addArchRThreads<span class="punctuation">(</span>threads <span class="operator">=</span> <span class="number">16</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>第三步: 我们设置需要使用基因和基因组注释。同样，这也是每个新的R session需要设置的参数。当然，我们需要使用和序列比对阶段相同的基因组版本。对于本教程使用的数据，我们会使用hg19参考基因组。ArchR支持多种基因组注释，并且允许自定义基因组注释。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">addArchRGenome<span class="punctuation">(</span><span class="string">&quot;hg19&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>基因和基因组注释信息是后续计算TSS富集得分，核小体含量和基因活跃度得分所必需的。同样，你得保证这里选择的基因组版本得和你上游数据处理时用到的基因组版本一致。我们这里设置”hg19”就是因为上数据处理用的是hg19作为参考基因组。除了hg19外，ArchR还提供了”hg38”, “mm9”, “mm10”, 并且允许用户使用<code>createGeneAnnotation</code>和<code>createGenomeAnnotation</code>函数创建其他物种注释。</p>
<p>我们使用<code>addArchRGenome</code>函数无缝地为ArchR提供这些信息。该函数告诉ArchR，在之后的所有分析中，它应该使用定义在<code>ArchRgenome</code>的相关<code>genomeAnnotation</code>和<code>geneAnnotation</code>。每一个原生支持的基因组都包括四个对象(object)</p>
<ul>
<li>BSgenome: 记录 染色体坐标信息和染色体序列信息</li>
<li>GRanges: 记录blacklist, 即对分析没有用设置可能产生干扰的区域</li>
<li>TxDb: 定义所有基因的位置信息</li>
<li>OrgDb: 提供基因编号，以及不同基因编号之间的转换</li>
</ul>
<p>如下是ArchR自带的物种注释的数据来源</p>
<hr>
<p>ArchR的预编译的<strong>hg19</strong>基因组用的是<code>BSgenome.Hsapiens.UCSC.hg19</code>, <code>TxDb.Hsapiens.UCSC.hg19.knownGene</code>, <code>org.Hs.eg.db</code>。而黑名单(记录着一直打开的区域，对后续分析没有帮助的位置）则来源于 <a target="_blank" rel="noopener" href="https://github.com/Boyle-Lab/Blacklist/blob/master/lists/hg19-blacklist.v2.bed.gz">hg19 v2 blacklist regions</a> 和<a target="_blank" rel="noopener" href="https://github.com/caleblareau/mitoblacklist/blob/master/peaks/hg19_peaks.narrowPeak">mitochondrial regions that show high mappability to the hg19 nuclear genome from  Caleb Lareau and Jason Buenrostro</a>，用<code>ArchR::mergeGR()</code>函数进行合并</p>
<hr>
<p>ArchR的预编译的<strong>hg38</strong>基因组用的是<code>BSgenome.Hsapiens.UCSC.hg38</code>, <code>TxDb.Hsapiens.UCSC.hg38.knownGene</code>, <code>org.Hs.eg.db</code>。而黑名单(记录着一直打开的区域，对后续分析没有帮助的位置）则来源于 <a target="_blank" rel="noopener" href="https://github.com/Boyle-Lab/Blacklist/blob/master/lists/hg38-blacklist.v2.bed.gz">hg19 v2 blacklist regions</a> 和<a target="_blank" rel="noopener" href="https://github.com/caleblareau/mitoblacklist/blob/master/peaks/hg38_peaks.narrowPeak">mitochondrial regions that show high mappability to the hg38 nuclear genome from  Caleb Lareau and Jason Buenrostro</a>，用<code>ArchR::mergeGR()</code>函数进行合并</p>
<hr>
<p>ArchR的预编译的<strong>mm9</strong>基因组用的是<code>BSgenome.Mmusculus.UCSC.mm9</code>, <code>TxDb.Mmusculus.UCSC.mm9.knownGene</code>, <code>org.Mm.eg.db</code>。而黑名单(记录着一直打开的区域，对后续分析没有帮助的位置）则来源于 <a target="_blank" rel="noopener" href="http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm9-mouse/mm9-blacklist.bed.gz">mm9 v1 blacklist regions</a> 和<a target="_blank" rel="noopener" href="https://github.com/caleblareau/mitoblacklist/blob/master/peaks/mm9_peaks.narrowPeak">mitochondrial regions that show high mappability to the mm9 nuclear genome from  Caleb Lareau and Jason Buenrostro</a>，用<code>ArchR::mergeGR()</code>函数进行合并</p>
<hr>
<p>ArchR的预编译的<strong>mm10</strong>基因组用的是<code>BSgenome.Mmusculus.UCSC.mm10</code>, <code>TxDb.Mmusculus.UCSC.mm10.knownGene</code>, <code>org.Mm.eg.db</code>。而黑名单(记录着一直打开的区域，对后续分析没有帮助的位置）则来源于 <a target="_blank" rel="noopener" href="https://github.com/Boyle-Lab/Blacklist/blob/master/lists/mm10-blacklist.v2.bed.gz">mm10 v2 blacklist regions</a> 和<a target="_blank" rel="noopener" href="https://github.com/caleblareau/mitoblacklist/blob/master/peaks/mm10_peaks.narrowPeak">mitochondrial regions that show high mappability to the mm10 nuclear genome from  Caleb Lareau and Jason Buenrostro</a>，用<code>ArchR::mergeGR()</code>函数进行合并</p>
<h3 id="1-5-1-自定义ArchRGenome"><a href="#1-5-1-自定义ArchRGenome" class="headerlink" title="1.5.1: 自定义ArchRGenome"></a>1.5.1: 自定义ArchRGenome</h3><p>如上所述，一个<code>ArchRGenome</code>需要包括基因组注释和基因注释</p>
<p>为了构建自定义的基因组注释，我们会使用<code>createGenomeAnnotation</code>. 他需要如下2个信息</p>
<ul>
<li>BSgenome: 包括基因组的序列信息，你可以通过谷歌查找指定物种的Bioconductor包</li>
<li>GRanges: 记录基因组中的黑名单区域，用来过滤下游分析中不需要的区间。这不是必须的，毕竟不是所有物种都能有这个文件。 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pubmed/31249361">publication on the ENCODE blacklists</a>提供黑名单列表的制作方法。</li>
</ul>
<p>例如，我们如果需要为Drosophila melanogaster创建一个基因组注释，那么我们需要先下载对应的BSgenome</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>requireNamespace<span class="punctuation">(</span><span class="string">&quot;BSgenome.Dmelanogaster.UCSC.dm6&quot;</span><span class="punctuation">,</span> quietly <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&quot;BSgenome.Dmelanogaster.UCSC.dm6&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line">library<span class="punctuation">(</span>BSgenome.Dmelanogaster.UCSC.dm6<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>接着，我们从BSgenome对象中创建基因组注释</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">genomeAnnotation <span class="operator">&lt;-</span> createGenomeAnnotation<span class="punctuation">(</span>genome <span class="operator">=</span> BSgenome.Dmelanogaster.UCSC.dm6<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>查看这个对象，你可以看到一个ArchR基因组注释的组成内容</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">genomeAnnotation</span><br><span class="line"><span class="comment">## List of length 3</span></span><br><span class="line"><span class="comment">## names(3): genome chromSizes blacklist</span></span><br></pre></td></tr></table></figure>

<p>为了构造一个自定义基因注释，我们需要用到<code>createGeneAnnotation()</code>，它需要你提供如下两个信息  </p>
<ul>
<li>TxDb: 记录gene&#x2F;transcript的坐标信息</li>
<li>OrgDb: 用于基因名和其他基因编号的转换</li>
</ul>
<p>继续之前的Drosophila melanogaster案例，我们需要先安装并加载相关的<code>TxDb</code>和<code>OrgDb</code>对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>requireNamespace<span class="punctuation">(</span><span class="string">&quot;TxDb.Dmelanogaster.UCSC.dm6.ensGene&quot;</span><span class="punctuation">,</span> quietly <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&quot;TxDb.Dmelanogaster.UCSC.dm6.ensGene&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>requireNamespace<span class="punctuation">(</span><span class="string">&quot;org.Dm.eg.db&quot;</span><span class="punctuation">,</span> quietly <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&quot;org.Dm.eg.db&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line">library<span class="punctuation">(</span>TxDb.Dmelanogaster.UCSC.dm6.ensGene<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>org.Dm.eg.db<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>记着，我们构建基因注释对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">geneAnnotation <span class="operator">&lt;-</span> createGeneAnnotation<span class="punctuation">(</span>TxDb <span class="operator">=</span> TxDb.Dmelanogaster.UCSC.dm6.ensGene<span class="punctuation">,</span> OrgDb <span class="operator">=</span> org.Dm.eg.db<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>检查该对象，可以了解ArchR关于基因注释的存放形式</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">geneAnnotation</span><br><span class="line"><span class="comment">## List of length 3</span></span><br><span class="line"><span class="comment">## names(3): genes exons TSS</span></span><br></pre></td></tr></table></figure>

<p>如果你没有TxDb和OrgDb对象，你也可以直接根据如下信息创建<code>geneAnnotation</code>对象</p>
<ol>
<li>“genes”: GRanges对象，记录基因的坐标信息，起始位置和结束。必须要有一列是和”exons”对象中其中一列匹配</li>
<li>“exons”: 记录每个基因外显子的坐标。必须要有一列和”genes”对象中的一列匹配</li>
<li>“GRanges”: 记录TSS(转录起始位点)的坐标</li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">geneAnnotation <span class="operator">&lt;-</span> createGeneAnnotation<span class="punctuation">(</span></span><br><span class="line">  TSS <span class="operator">=</span> geneAnnotation<span class="operator">$</span>TSS<span class="punctuation">,</span></span><br><span class="line">  exons <span class="operator">=</span> geneAnnotation<span class="operator">$</span>exons<span class="punctuation">,</span></span><br><span class="line">  genes <span class="operator">=</span> geneAnnotation<span class="operator">$</span>genes</span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="1-5-2-使用非标准基因组"><a href="#1-5-2-使用非标准基因组" class="headerlink" title="1.5.2 使用非标准基因组"></a>1.5.2 使用非标准基因组</h3><p>ArchR会做一些必要的检查，来避免你做一些ArchR觉得”不符合常理”的操作。其中就是检查你的基因组注释的<code>seqnames</code>都需要以”chr”开头。大部分情况下这都没有问题，除了一些基因组的染色体并不都是以”chr”作为前缀（例如玉米）。如果用ArchR分析这些基因组，你需要告知ArchR忽略染色体名前缀，否则会报错停止。你需要在创建Arrow文件前，先运行<code>addArchRChrPrefix(chrPrefix = FALSE) </code>. 它会当前的R session里停止所有对染色体名前缀的检查操作。</p>
<p>此外，ArchR默认会将染色体名&#x2F;seqnames转成字符串，因此如果你的<code>seqnames</code>都是数字，你需要以字符串形式提供这seqnames, 例如，你需要执行<code>useSeqnames = c(&quot;1&quot;, &quot;2&quot;, &quot;3&quot;)</code>，而不是<code>useSeqnames = c(1, 2, 3)</code></p>
<p>你可以用<code>getArchRChrPrefix</code>随时检查当前R session是否需要染色体前缀。</p>
<h2 id="1-6-创建Arrow文件"><a href="#1-6-创建Arrow文件" class="headerlink" title="1.6 创建Arrow文件"></a>1.6 创建Arrow文件</h2><p>在本教程的后续部分，我们会使用<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pubmed/31792411">Granja* et al. Nature Biotechnology 2019</a>文章的数据进行展示，当然不是所有的数据集，我们会使用降抽样的造血细胞数据集，保证大部分人的电脑都能带动。该数据集包括了骨髓单核细胞(BMMC)和外周血单核细胞(PBMC)，以及CD34+造血干细胞和骨髓前体细胞(CD34 BMMC)。</p>
<p>我们下载的数据以fragment格式存放，记录每个比对序列在基因组上的位置。Fragments文件是10X Genomics分析平台的其中一个输出文件，或者你也可以从BAM文件进</p>
<p>一旦我们得到了fragment文件，我们将它们的路径记录在一个向量中，然后作为参数传给<code>createArrowFiles()</code>. 在构建过程中，一些基本的元信息和矩阵会增加到各个Arrow文件中，包括<code>TileMatrix</code>和<code>GeneScoreMatrix</code>. 其中TileMarix记录的以500-bp作为分窗统计染色体上各个位置上是否有insertion（具体见<code>addTileMatrix</code>）, GeneScoreMatrix则是基于邻近基因启动子的insetion数推断的基因表达量（见<code>addGeneScoreMatrix()</code>）。</p>
<p>教程用到的数据，可以用<code>getTutorialData</code>进行下载，大约为0.5G。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">inputFiles <span class="operator">&lt;-</span> getTutorialData<span class="punctuation">(</span><span class="string">&quot;Hematopoiesis&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>当然这一步实际是在本地建立一个HemeFragments目录，并下载指定数据，因此如果网络太差，可以自己尝试下载。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> HemeFragments &amp;&amp; <span class="built_in">cd</span> HemeFragments</span><br><span class="line">wget <span class="string">&#x27;https://jeffgranja.s3.amazonaws.com/ArchR/TestData/HemeFragments/scATAC_BMMC_R1.fragments.tsv.gz&#x27;</span></span><br><span class="line">wget <span class="string">&#x27;https://jeffgranja.s3.amazonaws.com/ArchR/TestData/HemeFragments/scATAC_CD34_BMMC_R1.fragments.tsv.gz&#x27;</span></span><br><span class="line">wget <span class="string">&#x27;https://jeffgranja.s3.amazonaws.com/ArchR/TestData/HemeFragments/scATAC_PBMC_R1.fragments.tsv.gz&#x27;</span></span><br></pre></td></tr></table></figure>

<p>在启动项目前，我们必须得设置<code>ArchRGenome</code>并设置线程数<code>threads</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">addArchRGenome<span class="punctuation">(</span><span class="string">&quot;hg19&quot;</span><span class="punctuation">)</span> <span class="comment"># hg38, mm9, mm10</span></span><br><span class="line">addArchRThreads<span class="punctuation">(</span>threads <span class="operator">=</span> <span class="number">16</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>从现在开始，我们将会用10-15分钟的时间创建Arrow文件。对每一个样本，都会有如下操作</p>
<ol>
<li>从给定文件中读取开放信息(fragments)</li>
<li>为每个细胞计算质控信息(TSS富集得分和核小体信息)</li>
<li>根据质控参数过滤细胞</li>
<li>以500bp为窗口构建全基因组范围的TileMatrix</li>
<li>使用自定义的<code>geneAnnotaiton</code>创建GeneScoreMatrix, 其中<code>geneAnnotation</code>在我们调用<code>addArchRGenome</code>时定义</li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">ArrowFiles <span class="operator">&lt;-</span> createArrowFiles<span class="punctuation">(</span></span><br><span class="line">  inputFiles <span class="operator">=</span> inputFiles<span class="punctuation">,</span></span><br><span class="line">  sampleNames <span class="operator">=</span> <span class="built_in">names</span><span class="punctuation">(</span>inputFiles<span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  filterTSS <span class="operator">=</span> <span class="number">4</span><span class="punctuation">,</span> <span class="comment"># 这个参数不需要过高，后续可以调整</span></span><br><span class="line">  filterFrags <span class="operator">=</span> <span class="number">1000</span><span class="punctuation">,</span></span><br><span class="line">  addTileMat <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span></span><br><span class="line">  addGeneScoreMat <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们可以看下<code>ArrowFiles</code>对象，检查它是否真的是一个存放Arrow文件路径的向量</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ArrowFiles</span><br><span class="line"><span class="comment"># [1] &quot;scATAC_BMMC_R1.arrow&quot;      &quot;scATAC_CD34_BMMC_R1.arrow&quot;</span></span><br><span class="line"><span class="comment"># [3] &quot;scATAC_PBMC_R1.arrow&quot;</span></span><br></pre></td></tr></table></figure>

<h2 id="1-7-细胞质控"><a href="#1-7-细胞质控" class="headerlink" title="1.7 细胞质控"></a>1.7 细胞质控</h2><p>scATAC-seq的严格质控对于剔除低质量细胞至关重要。在ArchR中，我们考虑数据的三种信息</p>
<ol>
<li>每个细胞中唯一比对数(number of unique nuclear fragments)，不包括比对到线粒体的部分(unique nuclear fragments, 类似于单细胞转录组数据的表达的基因量)</li>
<li>信噪比(signal-to-background ratio)。如果是死细胞或者快死的细胞，那么DNA倾向于去染色质化，就会导致全局转座酶随机切割，体现出来就是信噪比低。</li>
<li>fragment大小分布( fragment size distribution)。 由于核小体周期性，我们期望看到在147-bp附近出现一个低谷，因为缠绕一个核小体的DNA序列大约为147bp。</li>
</ol>
<p>第一个参数, 唯一比对数非常的直观，如果一个细胞中的唯一比对太少，显然在后续分析中也没有太多可用价值，可以直接剔除掉。</p>
<p>第二个参数，信噪比是根据TSS富集分数进行计算。传统的混池ATAC-seq分析中，会使用TSS富集得分作为标准流程的一部分，用于确定信噪比（如 <a target="_blank" rel="noopener" href="https://www.encodeproject.org/atac-seq/">ENCODE计划</a>）。我们和其他人通过混池ATAC-seq和scATAC-seq分析发现，TSS富集得分在大部分的细胞类型中都具有代表性。TSS富集得分的背后思想是，由于大蛋白复合体会结合在启动子区域，ATAC-seq数据更多的富集在基因的TSS区域而不基因组其他区域。通过检查这些TSS区域中心的开放水平，我们发现中心相对于两侧(两边的1900-2000 bp处)存在富集现象。因此，我们定义富集中心(TSS中心)相对于两侧区域的比值为TSS富集得分(TSS enrichment score)</p>
<p>传统上，我们会计算每个混池ATAC-seq样本中每个碱基的开放性，之后这个谱会被用于确定TSS富集得分。在scATAC-seq中通过该方法计算每个细胞的TSS富集得分速度较慢，并且需要很强的算力。为了提高计算效率，同时还能得到和传统计算接近的结果，我们以TSS位置为中心，以50-bp作为分窗计算两边的平均开放程度，之后该值除以TSS两侧位置(+-1900-200bp)的平均开放程度。 这种计算方法和原来相对比，两者的相关性大于0.99，并且结果几乎一样。</p>
<p>第三个参数fragment大小分布并不是特别重要，最好是人工检查下。由于DNA缠绕核小体的模式，我们预期在fragment大小分布中看到核小体的周期性分布。这些山峰和低谷的出现正是由于DNA缠绕0，1，2个核小体的结果(Tn5无法切割缠绕在核小体的DNA序列，只能切割两边）</p>
<p>ArchR默认会过滤TSS富集得分低于4或唯一比对数小于1000（也就是保留TSS富集得分大于4且唯一比对数大于1000的细胞）。<strong>切记</strong>，ArchR默认的参数最初是用于人类数据，不能直接用于其他物种。每个数据都有他的独特性，切勿生搬硬套，我们需要按照实际情况设置参数。一定要在<code>createArrowFiles()</code>运行前设置好该参数。</p>
<hr>
<p>创建Arrow文件会在当前目录下生成一个”QualityControl”目录，这里面包括2个和样本相关的图。第一个图展示<code>log10(unique nuclear fragments) vs TSS enrichment score</code>, 虚线表示过滤阈值。第二图注释fragment大小分布图。</p>
<p>对我们的教程数据，我们的三个样本表现如下</p>
<p>BMMC:</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-8391ec1c6ae54aadbe2d7412c7db903e.png" alt="BMMC"></p>
<p>CD34+ BMMC:</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-1956fc3ef8a54484bd157729c3fdaccd.png" alt="CD34 BMMC"></p>
<p>PBMC:</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/05/image-6619fa90e04a43d4926293283d6c77b5.png" alt="PBMC"></p>
<p>我们现在整理完毕我们的Arrow文件，接下来就是创建<code>ArchRProject</code>了。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/05/21/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B8%80%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter1/" data-id="clm1ywdjg00b6ecok17hb56oi" data-title="使用ArchR分析单细胞ATAC-seq数据(第一章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/10/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/9/">9</a><a class="page-number" href="/page/10/">10</a><span class="page-number current">11</span><a class="page-number" href="/page/12/">12</a><a class="page-number" href="/page/13/">13</a><span class="space">&hellip;</span><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/12/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/09/">九月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/09/02/%E5%A6%82%E4%BD%95%E8%AE%A9%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8D%E5%86%8D%E9%BE%9F%E9%80%9F%E4%B8%8B%E8%BD%BDgithub%E7%9A%84%E6%95%B0%E6%8D%AE/">如何让服务器不再龟速下载github的数据</a>
          </li>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>