<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/22/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-LTR_retriever_ 一个更加准的LTR整合分析工具-LTR_retriever-accurate-identification-of-LTR-retrotransposons" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/27/LTR_retriever_%20%E4%B8%80%E4%B8%AA%E6%9B%B4%E5%8A%A0%E5%87%86%E7%9A%84LTR%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7-LTR_retriever-accurate-identification-of-LTR-retrotransposons/" class="article-date">
  <time class="dt-published" datetime="2019-08-27T12:58:17.308Z" itemprop="datePublished">2019-08-27</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/27/LTR_retriever_%20%E4%B8%80%E4%B8%AA%E6%9B%B4%E5%8A%A0%E5%87%86%E7%9A%84LTR%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7-LTR_retriever-accurate-identification-of-LTR-retrotransposons/">LTR_retriever:一个更加准的LTR整合分析工具</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h2 id="背景篇"><a href="#背景篇" class="headerlink" title="背景篇"></a>背景篇</h2><p>在植物基因组中，I类转座因子，LTR-RT(LTR retrotransposons)是基因组扩张的主要原因。完整的LTR长度在85<del>5000 bp之间，下图图A表示的是一个完整的LTR-RT，灰色框表示TSD(target site duplications), 红色三角形表示LTR motif(长度在2bp左右), 蓝色框表示LTR。LTR中间序列长度在1,000</del>15,000之间波动。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/2013053-538f429117b6891f-5b3fc254688246a18ebe845bab37279e.png" alt="LTR-RT结构"></p>
<p>完整的LTR-RT主要归为两大类: Gypsy和Copia。如果LTR中间的序列不包含开放阅读框(ORF), 那么所属的LTR-RT就无法独立的转座。</p>
<h2 id="安装篇"><a href="#安装篇" class="headerlink" title="安装篇"></a>安装篇</h2><p><code>LTR_retriever</code>不是一个独立的工具，他的主要作用就是整合 LTRharvest, LTR_FINDER, MGEScan 3.0.0, LTR_STRUC, 和 LtrDetector的结果，过滤其中的假阳性LTR-RT，得到高质量的LTR-RT库。</p>
<p>LTR_retriever托管在GitHub, <a target="_blank" rel="noopener" href="https://github.com/oushujun/LTR_retriever">https://github.com/oushujun/LTR_retriever</a>, 下载<code>LTR_retriever</code>本体</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git <span class="built_in">clone</span> https://github.com/oushujun/LTR_retriever.git</span><br></pre></td></tr></table></figure>

<p>之后修改<code>LTR_retriever</code>下的<code>paths</code>, 提供BLAST+, RepeatMasker， HMMER， CDHIT这些工具的路径。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">BLAST+=/your_path_to/BLAST+2.2.30/bin/</span><br><span class="line">RepeatMasker=/your_path_to/RepeatMasker4.0.0/</span><br><span class="line">HMMER=/your_path_to/HMMER3.1b2/bin/</span><br><span class="line">CDHIT=/your_path_to/CDHIT4.6.1/</span><br><span class="line">BLAST=/your_path_to/BLAST2.2.26/bin/ <span class="comment">#not required if CDHIT provided</span></span><br></pre></td></tr></table></figure>

<p>更加方便的安装方法用Bioconda安装好cd-hit repeatmasker， 然后下载LTR_retriever:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">conda create -n LTR_retriever</span><br><span class="line"><span class="built_in">source</span> activate LTR_retriever</span><br><span class="line">conda install -c conda-forge perl perl-text-soundex</span><br><span class="line">conda install -c bioconda cd-hit</span><br><span class="line">conda install -c bioconda/label/cf201901 repeatmasker</span><br><span class="line">git <span class="built_in">clone</span> https://github.com/oushujun/LTR_retriever.git</span><br><span class="line">./LTR_retriever/LTR_retriever -h</span><br></pre></td></tr></table></figure>

<p>此外你还需要额外安装<code>LTRharvest</code>, <code>LTR_FINDER</code> 和<code>MGEScan_LTR</code>。</p>
<ul>
<li>LTRharverst: <a target="_blank" rel="noopener" href="http://genometools.org/">http://genometools.org/</a></li>
<li>LTR_FINDER: <a target="_blank" rel="noopener" href="https://github.com/xzhub/LTR_Finder">https://github.com/xzhub/LTR_Finder</a></li>
<li>修改版MGEScan_LTR: <a target="_blank" rel="noopener" href="http://dawgpaws.sourceforge.net/">http://dawgpaws.sourceforge.net/</a></li>
</ul>
<p>由于MGEScan_LTR装起来比我想象中麻烦，所以本文就仅使用LTRharverst和LTR_FINDER</p>
<h2 id="使用篇"><a href="#使用篇" class="headerlink" title="使用篇"></a>使用篇</h2><blockquote>
<p>尽管LTR_retriever支持多个LTR工具的输入，但其实上LTRharverst和LTR_FINDER的结果就已经很不错了。</p>
</blockquote>
<p>以拟南芥的基因组序列为例，分别使用LTRharverst和LTR_FINDER来寻找拟南芥中潜在LTR序列，之后用<code>LTR_retreiver</code>来合并结果。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#LTRharvest</span></span><br><span class="line">gt suffixerator \</span><br><span class="line">  -db TAIR10.fa \</span><br><span class="line">  -indexname TAIR10 \</span><br><span class="line">  -tis -suf -lcp -des -ssp -sds -dna</span><br><span class="line">gt ltrharvest \</span><br><span class="line">  -index TAIR10 \</span><br><span class="line">  -similar 90 -vic 10 -seed 20 -seqids <span class="built_in">yes</span> \</span><br><span class="line">  -minlenltr 100 -maxlenltr 7000 -mintsd 4 -maxtsd 6 \</span><br><span class="line">  -motif TGCA -motifmis 1  &gt; TAIR10.harvest.scn &amp;</span><br><span class="line"><span class="comment"># LTR_FINDER</span></span><br><span class="line">ltr_finder -D 15000 -d 1000 -L 7000 -l 100 -p 20 -C -M 0.9 TAIR10.fa &gt; TAIR10.finder.scn &amp;</span><br></pre></td></tr></table></figure>

<p><code>LTR_retriever</code>支持单个候选的LTR，</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">LTR_retriever -genome TAIR10.fa -inharvest TAIR10.harvest.scn</span><br></pre></td></tr></table></figure>

<p>也支持多个候选LTR输入</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">LTR_retriever -genome TAIR10.fa -inharvest TAIR10.harvest.scn -infinder TAIR10.finder.scn -threads 20</span><br></pre></td></tr></table></figure>

<p>输出文件如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/2013053-37ef5a8181e7e271-2923e29399c34ee4a62f874be8b745e0.png" alt="运行结果"></p>
<h2 id="其他测试"><a href="#其他测试" class="headerlink" title="其他测试"></a>其他测试</h2><p>LAI值是作者提出用于衡量基因组完整度参数。比较2个LTR输入和1个LTR输入的LAI值，后者是15.62，前者是14.47，这也意味这个值其实是受到输入的候选LTR数目影响，但最终结果应该稳定在一个阈值内。</p>
<p>我测试了多个物种在两种软件下找到的LTR，以及最终pass留下的LTR, 发现最终能够pass，数量都相对较少。同时限速步骤就是LTR_finder 和 LTRharvest。</p>
<table>
<thead>
<tr>
<th>物种</th>
<th>基因组大小</th>
<th>LTR_finder</th>
<th>LTRharvest</th>
<th>Pass</th>
<th>LAI</th>
<th>测序技术</th>
</tr>
</thead>
<tbody><tr>
<td>A. lyrata</td>
<td>206M</td>
<td>1456</td>
<td>1017</td>
<td>1044</td>
<td>20.39</td>
<td>Sanger</td>
</tr>
<tr>
<td>A. thaliana (TAIR10)</td>
<td>120 M</td>
<td>207</td>
<td>550</td>
<td>184</td>
<td>15.62</td>
<td>Sanger</td>
</tr>
<tr>
<td>B. rapa (2.5)</td>
<td>391M</td>
<td>1251</td>
<td>3182</td>
<td>520</td>
<td>0</td>
<td>PacBio + 二代20Kb 40Kb文库</td>
</tr>
<tr>
<td>B. rapa (3.0)</td>
<td>353 M</td>
<td>3515</td>
<td>3635</td>
<td>1968</td>
<td>7.16</td>
<td>PacBio + BioNano + Hi-C</td>
</tr>
<tr>
<td>C.rubella</td>
<td>135 M</td>
<td>643</td>
<td>600</td>
<td>144</td>
<td>10.96</td>
<td>454 + Sanger</td>
</tr>
<tr>
<td>A. alpina</td>
<td>336 M</td>
<td>3840</td>
<td>3107</td>
<td>2556</td>
<td>11.01</td>
<td>PacBio + BioNano + Hi-C</td>
</tr>
<tr>
<td>某物种A</td>
<td>454 M</td>
<td>5384</td>
<td>2789</td>
<td>4294</td>
<td>17.89</td>
<td>PacBio</td>
</tr>
</tbody></table>
<p>还有一个有趣的现象，B. rapa 3.0版本尽管是最近用三代加Hi-C组装的基因，但是以LAI的标准，只能算是Draft级别, 当然也比2.5版本好出不少。</p>
<p>当然作者也对很多物种的多个版本组装进行了比较，下图来自于 Assessing genome assembly quality using the LTR Assembly Index (LAI)</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/2013053-48feb2021d9a65e3-f7061838e05b48cab7e47c4db73d3536.png" alt="基因组评估"></p>
<p>如果使用该软件记得引用下面两篇文献</p>
<ul>
<li>LTR_retriever: A Highly Accurate and Sensitive Program for Identification of Long Terminal Repeat Retrotransposons</li>
<li>Assessing genome assembly quality using the LTR Assembly Index (LAI)</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/27/LTR_retriever_%20%E4%B8%80%E4%B8%AA%E6%9B%B4%E5%8A%A0%E5%87%86%E7%9A%84LTR%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7-LTR_retriever-accurate-identification-of-LTR-retrotransposons/" data-id="cllg90xzl001meqp2akhpfetn" data-title="LTR_retriever:一个更加准的LTR整合分析工具" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-HiC-Pro_ Hi-C数据预处理高效工具-hic-pro-an-optimized-and-flexible-pipeline-for-hi-c-data-processing" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/26/HiC-Pro_%20Hi-C%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86%E9%AB%98%E6%95%88%E5%B7%A5%E5%85%B7-hic-pro-an-optimized-and-flexible-pipeline-for-hi-c-data-processing/" class="article-date">
  <time class="dt-published" datetime="2019-08-26T21:29:00.048Z" itemprop="datePublished">2019-08-26</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/26/HiC-Pro_%20Hi-C%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86%E9%AB%98%E6%95%88%E5%B7%A5%E5%85%B7-hic-pro-an-optimized-and-flexible-pipeline-for-hi-c-data-processing/">HiC-Pro:Hi-C数据预处理高效工具</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>HiC-Pro是一个高效率的Hi-C数据预处理工具，能够应用于dilution Hi-C, in situ Hi-C, DNase Hi-C, Micro-C, capture-C, capture Hi-C 和 HiChip 这些数据。</p>
<p>HiC-Pro的工作流程如下, 简单的说就是先双端测序各自比对，然后进行合并，根据合并的结果筛选有效配对。之后有效配对用于构建contact maps.</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/hicpro_wkflow-7cce986ba7d84c05857b509f2beb1c79.png" alt="工作流程"></p>
<h2 id="安装方法"><a href="#安装方法" class="headerlink" title="安装方法"></a>安装方法</h2><p>HiC-Pro依赖于如下的软件</p>
<ul>
<li>Bowtie2(&gt;2.2.2), 用于序列比对</li>
<li>Python2.7, 并安装 Pysam, bx-python, numpy, scipy</li>
<li>R, <em>RColorBrewer</em>  + <em>ggplot2</em> </li>
<li>samtools &gt; 1.1</li>
<li>GNU sort, 支持 -V， 按照version进行排序</li>
</ul>
<p>如果有root权限，更加推荐使用Singularity，比Conda更简单。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 下载</span></span><br><span class="line">wget https://zerkalo.curie.fr/partage/HiC-Pro/singularity_images/hicpro_latest_ubuntu.img</span><br><span class="line"><span class="comment"># 使用</span></span><br><span class="line">singularity <span class="built_in">exec</span> /opt/biosoft/HiC-Pro/hicpro_latest_ubuntu.img HiC-Pro -h</span><br></pre></td></tr></table></figure>


<p>没有Root的安装方法: 为了保证环境的干净，我用conda进行了一个环境进行安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">conda create -y -n hic-pro python=2.7 pysam bx-python numpy scipy samtools bowtie2</span><br><span class="line">conda activate hic-pro</span><br></pre></td></tr></table></figure>

<p>以2.11.1版本为例进行介绍，最新的版本在<a target="_blank" rel="noopener" href="https://github.com/nservant/HiC-Pro/releases%E6%A3%80%E6%9F%A5">https://github.com/nservant/HiC-Pro/releases检查</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">wget https://github.com/nservant/HiC-Pro/archive/v2.11.1.tar.gz</span><br><span class="line">tar -zxvf v2.11.1.tar.gz</span><br><span class="line"><span class="built_in">cd</span> HiC-Pro-2.11.1</span><br></pre></td></tr></table></figure>

<p>因为我希望把HiC-Pro安装到<code>~/opt/bisofot</code>下，所以我需要修改当前目录下的<code>config-install.txt</code>中的PREFIX部分，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">PREFIX =  /home/xzg/opt/bisofot</span><br></pre></td></tr></table></figure>

<p>如果服务器支持任务投递，可以修改CLUSTER_SYS部分, 设置为TORQUE, SGE, SLURM 或 LSF，</p>
<p>我的miniconda的安装目录是<code>~/miniconda3</code>, 所以hic-pro环境的实际路径是<code>~/miniconda3/envs/hic-pro</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">make configure</span><br><span class="line">make install</span><br></pre></td></tr></table></figure>

<p>安装结束之后，<code>/home/xzg/opt/bisofot</code>文件夹下就出现了<code>HiC-Pro_2.11.1</code>，之后的软件调用方式为</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosfot/HiC-Pro_2.11.1/bin/HiC-Pro -h</span><br></pre></td></tr></table></figure>

<p>如果没有出现Error 就说明安装成功了。</p>
<h2 id="处理流程"><a href="#处理流程" class="headerlink" title="处理流程"></a>处理流程</h2><p>让我们新建一个项目文件夹，以一个测试数据集为例进行介绍。</p>
<p>下载测试数据并解压缩，该数据来自于Dixon et al. 2012 ， 使用HindIII 进行酶切</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p hic-pro &amp;&amp; <span class="built_in">cd</span> hic-pro</span><br><span class="line">wget https://zerkalo.curie.fr/partage/HiC-Pro/HiCPro_testdata.tar.gz &amp;&amp; tar -zxvf HiCPro_testdata.tar.gz</span><br></pre></td></tr></table></figure>
<p>之后将测序结果移动或者软连接到fastq文件夹下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p fastq</span><br><span class="line"><span class="built_in">mv</span> test_data/* fastq</span><br><span class="line"><span class="built_in">ls</span> fastq</span><br><span class="line"><span class="comment"># dixon_2M  dixon_2M_2</span></span><br></pre></td></tr></table></figure>

<h3 id="创建注释文件"><a href="#创建注释文件" class="headerlink" title="创建注释文件"></a>创建注释文件</h3><p>为了处理原始数据，HiC-Pro需要三个注释文件</p>
<ul>
<li>BED文件，记录可能的酶切位点</li>
<li>table文件，记录每条contig&#x2F;scaffold&#x2F;chromosome的长度</li>
<li>bowtie2索引</li>
</ul>
<p>其中BED文件和table文件必须要放在<code>HiC-Pro_2.11.1/annotations</code>目录下，该文件夹下已经有了人类hg19和小鼠mm10。 我们以GRCh38为例, 介绍如何创建这三个注释信息 </p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 切换目录</span></span><br><span class="line"><span class="built_in">cd</span> ~/opt/biosfot/HiC-Pro_2.11.1/annotation</span><br><span class="line"><span class="comment"># 下载GRCh38的序列</span></span><br><span class="line">wget ftp://ftp.ensembl.org/pub/release-97/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz</span><br><span class="line">gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz</span><br><span class="line"><span class="comment"># BED</span></span><br><span class="line">~/opt/biosoft/HiC-Pro_2.11.1/bin/utils/digest_genome.py -r HindIII  -o GRCh38_HindIII.bed Homo_sapiens.GRCh38.dna.primary_assembly.fa</span><br><span class="line"><span class="comment"># chromosome size</span></span><br><span class="line">seqkit fx2tab -<span class="built_in">nl</span> Homo_sapiens.GRCh38.dna.primary_assembly.fa | awk <span class="string">&#x27;&#123;print $1&quot;\t&quot;$2&#125;&#x27;</span> &gt; GRCh38.chrom.size</span><br><span class="line"><span class="comment"># bowtie2 index</span></span><br><span class="line">bowtie2-build --threads 20 Homo_sapiens.GRCh38.dna.primary_assembly.fa GRCh38</span><br></pre></td></tr></table></figure>

<h3 id="配置HiC-Pro"><a href="#配置HiC-Pro" class="headerlink" title="配置HiC-Pro"></a>配置HiC-Pro</h3><p>拷贝HiC-Pro的配置文件到项目文件夹下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cp</span> ~/opt/biosfot/HiC-Pro_2.11.1/config-hicpro.txt .</span><br></pre></td></tr></table></figure>

<p>修改配置文件config-hicpro.txt</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 线程数</span></span><br><span class="line">N_CPU = 80</span><br><span class="line"><span class="comment"># bowtie2索引, 绝对路径</span></span><br><span class="line">BOWTIE2_IDX_PATH = /home/xzg/opt/biosfot/HiC-Pro_2.11.1/annotation</span><br><span class="line"><span class="comment"># bowtie2索引时的前缀</span></span><br><span class="line">REFERENCE_GENOME = GRCh38</span><br><span class="line"><span class="comment"># 参考基因组各染色体长度</span></span><br><span class="line"><span class="comment"># 绝对路径</span></span><br><span class="line">GENOME_SIZE = /home/xzg/opt/biosfot/HiC-Pro_2.11.1/annotationGRCh38.chrom.size</span><br><span class="line"><span class="comment"># 酶切位点</span></span><br><span class="line"><span class="comment"># 绝对路径</span></span><br><span class="line">GENOME_FRAGMENT = /home/xzg/opt/biosfot/HiC-Pro_2.11.1/annotationGRCh38_HindIII.bed</span><br><span class="line">LIGATION_SITE = AAGCTAGCTT</span><br></pre></td></tr></table></figure>

<p>对于LIGATION_SITE，不同酶切位点对应的序列为HindIII(AAGCTAGCTT), MboI(GATCGATC) , DpnII(GATCGATC), NcoI(CCATGCATGG)。</p>
<p>我们要修改的参数其实就是上面几个。当然该配置文件还有许多参数可以修改，具体见<a target="_blank" rel="noopener" href="https://github.com/nservant/HiC-Pro/blob/master/doc/MANUAL.md">https://github.com/nservant/HiC-Pro/blob/master/doc/MANUAL.md</a></p>
<p>运行如下代码，启动分析项目</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/HiC-Pro_2.11.1/bin/HiC-Pro -i fastq -o results -c config-hicpro.txt</span><br></pre></td></tr></table></figure>

<p>HiC-Pro会新建一个工作目录，results, 之后会遍历fastq目录，寻找其中的fastq文件，将其软连接到results下的rawdata, 之后就开始用bowtie2比对以及后续的分析。</p>
<blockquote>
<p>测试数据代码运行到<code>Run ICE Normalization</code>就中断了，可能是用的参考基因组和原来的教程(hg19)不一样， 不过我自己的数据集是没有问题的。</p>
</blockquote>
<p>最后的结果如下:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">$ tree -L 2 results </span><br><span class="line">results</span><br><span class="line">├── bowtie_results <span class="comment"># 比对之后的输出 </span></span><br><span class="line">│   ├── bwt2</span><br><span class="line">│   ├── bwt2_global</span><br><span class="line">│   └── bwt2_local</span><br><span class="line">├── config-hicpro.txt</span><br><span class="line">├── hic_results</span><br><span class="line">│   ├── data <span class="comment"># 有效配对</span></span><br><span class="line">│   ├── matrix <span class="comment"># contact maps</span></span><br><span class="line">│   ├── pic <span class="comment"># 可视化质控信息</span></span><br><span class="line">│   └── stats <span class="comment">#文字版质控信息</span></span><br><span class="line">├── logs <span class="comment"># 各种日志</span></span><br><span class="line">│   ├── dixon_2M</span><br><span class="line">│   └── dixon_2M_2</span><br><span class="line">├── rawdata -&gt; /home/xzg/project/Tutorial/hic-pro/fastq</span><br><span class="line">└── tmp</span><br></pre></td></tr></table></figure>

<p>一个关键的结果就是data文件里的以*.validPairs*结尾的文件，有7+1列，</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">read_name chr_reads1  pos_reads1  strand_reads1  chr_reads2  pos_reads2  strand_reads2  fragment_size [allele_specific_tag] </span><br></pre></td></tr></table></figure>

<p>此外，HiC-Pro提供了一个脚本用于将输出的allValidPairs转成<a target="_blank" rel="noopener" href="https://github.com/aidenlab/juicer/wiki/Juicer-Tools-Quick-Start">JABT</a>的输入</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">~/HiC-Pro_2.11.1/bin/utils/hicpro2juicebox.sh \</span><br><span class="line">    -i hic_results/data/dixon_2M/dixon_2M.allValidPairs \</span><br><span class="line">    -g /home/xzg/opt/biosfot/HiC-Pro_2.11.1/annotation/GRCh38.chrom.size \</span><br><span class="line">    -j ~/opt/biosoft/juicer/scripts/common/juicer_tools.jar</span><br></pre></td></tr></table></figure>

<p>最终会输出一个以.hic结尾的文件。</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li>官方手册: <a target="_blank" rel="noopener" href="https://github.com/nservant/HiC-Pro/blob/master/doc/MANUAL.md">https://github.com/nservant/HiC-Pro/blob/master/doc/MANUAL.md</a></li>
<li>官方帮助文档: <a target="_blank" rel="noopener" href="https://nservant.github.io/HiC-Pro/MANUAL.html">https://nservant.github.io/HiC-Pro/MANUAL.html</a></li>
<li><em>Servant N., Varoquaux N., Lajoie BR., Viara E., Chen CJ., Vert JP., Dekker J., Heard E., Barillot E.</em> HiC-Pro: An optimized and flexible pipeline for Hi-C processing. Genome Biology 2015, 16:259 <a target="_blank" rel="noopener" href="https://doi.org/10.1186/s13059-015-0831-x">doi:10.1186&#x2F;s13059-015-0831-x</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/26/HiC-Pro_%20Hi-C%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86%E9%AB%98%E6%95%88%E5%B7%A5%E5%85%B7-hic-pro-an-optimized-and-flexible-pipeline-for-hi-c-data-processing/" data-id="cllg90xzj001eeqp21qu0960w" data-title="HiC-Pro:Hi-C数据预处理高效工具" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-「基因组注释」使用RepeatModeler从头注释基因组的重复序列-Repeat-annotation-with-RepeatModeler" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/26/%E3%80%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%B3%A8%E9%87%8A%E3%80%8D%E4%BD%BF%E7%94%A8RepeatModeler%E4%BB%8E%E5%A4%B4%E6%B3%A8%E9%87%8A%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%9A%84%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97-Repeat-annotation-with-RepeatModeler/" class="article-date">
  <time class="dt-published" datetime="2019-08-26T17:31:51.898Z" itemprop="datePublished">2019-08-26</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/26/%E3%80%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%B3%A8%E9%87%8A%E3%80%8D%E4%BD%BF%E7%94%A8RepeatModeler%E4%BB%8E%E5%A4%B4%E6%B3%A8%E9%87%8A%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%9A%84%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97-Repeat-annotation-with-RepeatModeler/">「基因组注释」使用RepeatModeler从头注释基因组的重复序列</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>RepeatModeler可用来从头对基因组的重复序列家族进行建模注释，它的核心组件是RECON和RepatScout。</p>
<h2 id="使用方法"><a href="#使用方法" class="headerlink" title="使用方法"></a>使用方法</h2><p>以拟南芥的参考基因组为例，假设基因组的名字为”Athaliana.fa”</p>
<p>第一步：为RepeatModeler创建索引数据库</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">BuildDatabase -name ath -engine ncbi Athaliana.fa</span><br><span class="line"><span class="comment"># -engine ncbi： 表示使用rmblast</span></span><br><span class="line"><span class="comment"># -name aht： 表示数据库的名字为ath</span></span><br></pre></td></tr></table></figure>

<p>这一步其实认为调用了<code>makeblastdb</code>，结果文章和<code>makeblastdb</code>一致</p>
<p>第二步：运行RepeatModeler</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">RepeatModeler -database ath -engine ncbi -pa 10 &amp;&gt; ath.out &amp;</span><br><span class="line"><span class="comment"># -database 要和上一步一致</span></span><br><span class="line"><span class="comment"># -engine 要和上一步一致</span></span><br><span class="line"><span class="comment"># -pa 表示线程数</span></span><br></pre></td></tr></table></figure>

<p>运行中的的文件存放在<code>RM_.xxx</code>文件夹下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">RM_82213.MonOct151054032018</span><br><span class="line">├── consensi.fa </span><br><span class="line">├── consensi.fa.classified</span><br><span class="line">├── consensi.fa.masked</span><br><span class="line">├── families-classified.stk</span><br><span class="line">├── families.stk</span><br><span class="line">├── round-1</span><br><span class="line">├── round-2</span><br><span class="line">├── round-3</span><br><span class="line">├── round-4</span><br><span class="line">└── round-5</span><br></pre></td></tr></table></figure>

<p>运行结束后，就得到了<code>ath-families.fa</code>和<code>ath-families.stk</code>。 前者是找到的重复序列，后者是Stockholm格式的种子联配文件(seed alignment file), 可以用<code>util/dfamConsensusTool.pl</code>上传到<code>Dfam_consensus</code>数据库中。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grep -i Unkown ath-families.fa</span><br><span class="line"><span class="comment"># 没有结果，全都归类，毕竟拟南芥</span></span><br></pre></td></tr></table></figure>

<p><code>RM_.xxx</code>的<code>consensi.fa.classified</code>和<code>ath-families.fa</code>内容一样，也是FASTA格式的文件，，只不过每个序列的ID会标注它来自于哪个重复序列家族，如果无法归类，则用”Unkown”标注。</p>
<p>你可以将<code>ath-families.fa</code>分ModelerID.lib和Modelerunknown.lib, </p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">seqkit grep -nrp Unknown ath-families.fa &gt; Modelerunknown.lib</span><br><span class="line">seqkit grep -vnrp Unknown ath-families.fa &gt;  ModelerID.lib</span><br></pre></td></tr></table></figure>

<p>其中Modelerunknown.lib用RepeatMasker或TEclass进一步注释，如果能够被分类则从Modelerunknown.lib,移动到ModelerID.lib</p>
<p>这里用TEclass进行分类，软件安装使用参考<a href="/archives/Classify-TE-consensus-sequence-by-TEclass">使用TEclass对TE一致性序列进行分类</a>.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">TEclassTest Modelerunknown.lib</span><br></pre></td></tr></table></figure>

<blockquote>
<p>这一步其实无法运行，因为没有输入。</p>
</blockquote>
<p>第三步：过滤基因片段</p>
<blockquote>
<p>许多文章都没有这样子做，因此这一步完全可以不用看</p>
</blockquote>
<p>RepeatModeler找到的重复序列进一步在植物蛋白数据库（不包括转座子蛋白）进行搜索，如果和植物蛋白匹配，或者在序列的侧翼50bp以内，就将该重复序列剔除。这一步可用工具是<code>ProtExcluder</code><br>ProtExcluder的安装需要HMMER3， 方法如下</p>
<p>首先一定要装HMMER3.0，然后安装方法为</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">wget http://eddylab.org/software/hmmer/hmmer-3.0.tar.gz</span><br><span class="line">tar xf hmmer-3.0.tar.gz</span><br><span class="line"><span class="built_in">cd</span> hmmer-3.0</span><br><span class="line">./configure --prefix=<span class="variable">$HOME</span>/opt/biosoft/hmmer-3.0</span><br><span class="line">make &amp;&amp; make install</span><br><span class="line"><span class="built_in">cd</span> easel</span><br><span class="line">./configure --prefix=<span class="variable">$HOME</span>/opt/biosoft/hmmer-3.0</span><br><span class="line">make &amp;&amp; make install</span><br></pre></td></tr></table></figure>

<p>后续安装的ProtExcluder1.1非常坑爹，他居然认为hmmer的运行文件是放在binaries目录下，所以你还需要去<code>~/opt/biosoft/hmmer-3.0</code>把bin文件夹改成binaries</p>
<p>然后安装ProtExcluder1.1</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">tar zxf  ProtExcluder1.1.tar.gz</span><br><span class="line"><span class="built_in">mv</span> ProtExcluder1.1 ~/opt/biosoft</span><br><span class="line"><span class="built_in">cd</span>  ~/opt/biosoft/ProtExcluder1.1 </span><br><span class="line">./Installer.pl   -m   ~/opt/biosoft/hmmer-3.0 -p P ~/opt/biosoft/ProtExcluder1.1/</span><br><span class="line"><span class="comment"># 注意运行路径后一定要有&quot;/&quot;</span></span><br></pre></td></tr></table></figure>

<p>然后将找到的重复序列用blastx比对到植物的蛋白数据库中，你可以到RefSeq上进行下载。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">blastx -query ath-families.fa -db /path/to/refseq-plant/db/plant.protein -num_threads 70 &gt; ath-families.blastx &amp;</span><br></pre></td></tr></table></figure>

<p>最后运行ProtExcluder.pl</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">/path/to/ProtExcluder.pl -f 50   ath-families.blastx  ath-families.fa </span><br></pre></td></tr></table></figure>

<p>从理论上说结果是”XXXnoProtFinal”，但这个破软件各种报错，而且报错信息信息量太少，直接放弃这个破软件了。</p>
<p>考虑到目前看了很多文献也没人说要过滤，所以这一步就放弃吧</p>
<p>注1： 一般而言RepeatModeler在一周内就能运行完，物种小一点，服务器好一点，基本上一天就行了<br>注2： 一般而言，同源注释所用的重复序列库的数据比较可靠，但不一定全，而我们自己建立的重复序列库比较全，但是未必准。<br>注3： SwissProt的植物蛋白数据库的下载地址为：<a href="ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions">ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions</a>，而NCBI的Refseq植物蛋白数据库下载地址为<a href="ftp://ftp.ncbi.nih.gov/refseq/release/plant/">ftp://ftp.ncbi.nih.gov/refseq/release/plant/</a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/26/%E3%80%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%B3%A8%E9%87%8A%E3%80%8D%E4%BD%BF%E7%94%A8RepeatModeler%E4%BB%8E%E5%A4%B4%E6%B3%A8%E9%87%8A%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%9A%84%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97-Repeat-annotation-with-RepeatModeler/" data-id="cllg90y2t007meqp26akm1bvl" data-title="「基因组注释」使用RepeatModeler从头注释基因组的重复序列" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-利用3D-DNA流程组装基因组-scaffolding-genome-using-3d-dna-workflow" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/25/%E5%88%A9%E7%94%A83D-DNA%E6%B5%81%E7%A8%8B%E7%BB%84%E8%A3%85%E5%9F%BA%E5%9B%A0%E7%BB%84-scaffolding-genome-using-3d-dna-workflow/" class="article-date">
  <time class="dt-published" datetime="2019-08-25T16:22:03.560Z" itemprop="datePublished">2019-08-25</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/25/%E5%88%A9%E7%94%A83D-DNA%E6%B5%81%E7%A8%8B%E7%BB%84%E8%A3%85%E5%9F%BA%E5%9B%A0%E7%BB%84-scaffolding-genome-using-3d-dna-workflow/">利用3D-DNA流程组装基因组</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="利用3D-DNA流程组装基因组"><a href="#利用3D-DNA流程组装基因组" class="headerlink" title="利用3D-DNA流程组装基因组"></a>利用3D-DNA流程组装基因组</h1><p>使用二代数据或三代数据得到contig后，下一步就是将contig提升到染色体水平。有很多策略可以做到这一点，比如说遗传图谱，BioNano(看运气), HiC, 参考近源物种。</p>
<p>如果利用HiC进行准染色体水平，那么目前常见的组装软件有下面几个</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://github.com/DovetailGenomics/HiRise_July2015_GR">HiRise</a>: 2015年后的GitHub就不再更新</li>
<li><a target="_blank" rel="noopener" href="https://github.com/shendurelab/LACHESIS">LACHESIS</a>: 发表在NBT，2017年后不再更新</li>
<li><a target="_blank" rel="noopener" href="https://github.com/marbl/SALSA">SALSA</a>: 发表在BMC genomics,  仍在更新中</li>
<li><a target="_blank" rel="noopener" href="https://github.com/theaidenlab/3d-dna">3D-DNA</a>: 发表在science，仍在更新中</li>
<li><a target="_blank" rel="noopener" href="https://github.com/tangerzhang/ALLHiC">ALLHiC</a>: 发表在Nature Plants, 用于解决植物多倍体组装问题</li>
</ul>
<p>对于<strong>二倍体</strong>物种而言，目前3D-DNA应该是组装效果最好的一个软件。</p>
<h2 id="工作流程"><a href="#工作流程" class="headerlink" title="工作流程"></a>工作流程</h2><p>使用3D-DNA做基因组组装的整体流程如下图，分别为组装，Juicer分析Hi-C数据，3D-DNA进行scaffolding，使用JBAT对组装结果进行手工纠正，最终得到准染色体水平的基因组。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/3d-dna-pipeline-86ffcd937b294ec8834c704d90bc149b.png" alt="总体流程"></p>
<p>基因组组装可以是二代测序方法，也可以是三代测序组装方法，总之会得到contig。</p>
<p>Juicer的工作流程见下图，输入原始的fastq文件，处理得到中间文件.hic, 之后对.hic文件用于下游分析，包括</p>
<ul>
<li>Arrowhead: 寻找存在关联的区域</li>
<li>HiCCUPS: 分析局部富集peaks</li>
<li>MotifFinder: 用于锚定peaks</li>
<li>Persons: 计算观测&#x2F;期望的皮尔森相关系数矩阵</li>
<li>Eigenvector: 确定分隔</li>
</ul>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/juicer-workflow-28cf57888a514546b1666cc5c101af00.png" alt="juicer工作流程"></p>
<p>之后Juicer的输出结果给3D-DNA，分析流程见下图。3D-DNA先根据Hi-C数据分析contig中的misjoin，对其进行纠错。之后通过四步,分别是Polish, Split, Seal和Merge, 得到最终的基因组序列</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/3d-dna-pipeline-2-4f9ac6cd65b44dcb86e967e3f407da08.png" alt="3d-dna流程"></p>
<h2 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h2><p>在安装之前，确保服务器上有了下面这些依赖软件工具</p>
<ul>
<li>LastZ（仅在杂合基因组的二倍体模式下使用）</li>
<li>Java &gt;&#x3D; 1.7</li>
<li>GNU Awk &gt;&#x3D; 4.02</li>
<li>GNU coreutils sort &gt; 8.11</li>
<li>Python &gt;&#x3D; 2.7</li>
<li>scipy, numpy, matplotlib</li>
<li>GNU Parallel &gt;&#x3D;20150322 (不必要，但是强力推荐)</li>
<li>bwa</li>
</ul>
<p>我们需要安装两个软件，一个是3D-DNA，另一个是juicer。</p>
<p>CPU版本的juicer安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p ~/opt/biosoft/</span><br><span class="line"><span class="built_in">cd</span> ~/opt/biosoft</span><br><span class="line">git <span class="built_in">clone</span> https://github.com/theaidenlab/juicer.git</span><br><span class="line"><span class="built_in">cd</span> juicer</span><br><span class="line"><span class="built_in">ln</span> -s CPU scripts</span><br><span class="line"><span class="built_in">cd</span> scripts/common</span><br><span class="line">wget https://hicfiles.tc4ga.com/public/juicer/juicer_tools.1.9.9_jcuda.0.8.jar</span><br><span class="line"><span class="built_in">ln</span> -s juicer_tools.1.9.9_jcuda.0.8.jar  juicer_tools.jar</span><br></pre></td></tr></table></figure>

<p>然后用<code>~/opt/biosoft/juicer/scripts/juicer.sh -h</code>检查是否有帮助信息输出</p>
<p>3D-DNA安装也很容易，只需要从Github上将内容克隆到本地即可</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ~/opt/biosoft</span><br><span class="line">git <span class="built_in">clone</span> https://github.com/theaidenlab/3d-dna.git</span><br></pre></td></tr></table></figure>

<p>用<code>sh ~/opt/biosoft/3d-dna/run-asm-pipeline.sh -h</code>查看是否有帮助文档输出。</p>
<h2 id="参数详解"><a href="#参数详解" class="headerlink" title="参数详解"></a>参数详解</h2><p>以CPU版本的为例，juicer.sh的参数如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Usage: juicer.sh [-g genomeID] [-d topDir] [-s site] [-a about] [-R end]</span><br><span class="line">                 [-S stage] [-p chrom.sizes path] [-y restriction site file]</span><br><span class="line">                 [-z reference genome file] [-D Juicer scripts directory]</span><br><span class="line">                 [-b ligation] [-t threads] [-r] [-h] [-f] [-j] </span><br></pre></td></tr></table></figure>

<p>参数说明</p>
<ul>
<li><code>-g</code>: 定义一个物种名</li>
<li><code>-s</code>: 酶切类型, HindIII(AAGCTAGCTT),  MboI(GATCGATC) , DpnII(GATCGATC),  NcoI(CCATGCATGG)</li>
<li><code>-z</code> : 参考基因组文件</li>
<li><code>-y</code>: 限制性酶切位点可能出现位置文件</li>
<li><code>-p</code>: 染色体大小文件</li>
<li><code>-C</code>: 将原来的文件进行拆分，必须是4的倍数，默认是90000000, 即22.5M reads</li>
<li><code>-S</code>: 和任务重运行有关，从中途的某一步开始,”merge”, “dedup”, “final”, “postproc” 或 “early”</li>
<li><code>-D</code>: juicer的目录，我们安装在<code>~/opt/biosoft/</code>，所以设置为<code>~/opt/biosoft/juicer</code></li>
<li><code>-a</code>: 实验的描述说明，可以不用设置</li>
<li><code>-t</code>: 线程数</li>
</ul>
<p><code>juicer.sh</code>还有AWS, LSF, PBS, SLURM版本，由于我的服务器是单主机，无法进行测试讲解。</p>
<p>如果你的基因组不是复杂基因组，比如说高杂合，高重复序列，或者Hi-C数据测太少，那么3d-dna的流程更加简单, <code>run-asm-pipeline.sh -h</code>只有四个参数需要改</p>
<ul>
<li><code>-i|--input</code>: 过滤长度低于给定阈值的contig&#x2F;scaffold, 默认是15000</li>
<li><code>-r|--round</code>: 基因组中misjoin的纠错轮数，默认是2，当基因组比较准确时，设置为0，然后在JABT中调整会更好</li>
<li><code>-m|--mode</code>: 是否调用merge模块，当且仅当在杂合度比较高的情况下使用，也就是组装的单倍型基因组明显偏大</li>
<li><code>-s|--stage</code>:  从polish, split, seal, merge 或finalize 的某一个阶段开始</li>
</ul>
<p>但是，一旦基因组复杂起来，那么需要调整的参数就非常多了, <code>run-asm-pipeline.sh --help</code>会输出更多的信息，你需要根据当前结果去确定每个阶段的参数应该如何调整。</p>
<p>最终的输出文件最关键的是下面三类:</p>
<ul>
<li><code>.fasta</code>: 以FINAL标记的是最终结果</li>
<li><code>.hic</code>: 各个阶段都会有输出结果，用于在JABT中展示</li>
<li><code>.assembly</code>: 各个阶段都会有输出，一共两列，存放contig的组装顺序</li>
</ul>
<h2 id="分析过程"><a href="#分析过程" class="headerlink" title="分析过程"></a>分析过程</h2><p>假如你现在目录下有2个文件夹，reference</p>
<ul>
<li>reference: 存放一个genome.fa, 为组装的contigs</li>
<li>fastq: 存放HiC二代双端测序结果，read_R1_fastq.gz, read_R2_fastq.gz</li>
</ul>
<blockquote>
<p>注意，genome.fa中的序列一定得是80个字符分隔的情况，也就是多行FASTA。</p>
</blockquote>
<h3 id="增加一个新的基因组"><a href="#增加一个新的基因组" class="headerlink" title="增加一个新的基因组"></a>增加一个新的基因组</h3><p><strong>第一步</strong>： 为基因组建立BWA索引</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> reference</span><br><span class="line">bwa index genome.fa</span><br></pre></td></tr></table></figure>

<p><strong>第二步</strong>: 根据基因组构建创建可能的酶切位点文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python ~/opt/biosoft/juicer/misc/generate_site_positions.py DpnII genome genome.fa</span><br></pre></td></tr></table></figure>

<p><strong>第三步</strong>: 运行如下命令, 获取每条contig的长度</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">&#x27;BEGIN&#123;OFS=&quot;\t&quot;&#125;&#123;print $1, $NF&#125;&#x27;</span> genome_DpnII.txt &gt; genome.chrom.size</span><br><span class="line"><span class="comment"># 返回上级目录</span></span><br><span class="line"><span class="built_in">cd</span> ..</span><br></pre></td></tr></table></figure>

<h3 id="运行juicer"><a href="#运行juicer" class="headerlink" title="运行juicer"></a>运行juicer</h3><p>保证当前目录下有fastq和reference文件夹，然后运行如下命令，一定要设置<code>-z</code>,<code>-p</code>,<code>-y</code>这三个参数</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/juicer/scripts/juicer.sh \</span><br><span class="line">	-g genome \</span><br><span class="line">	-s MboI \</span><br><span class="line">	-z reference/genome.fa \</span><br><span class="line">	-y reference/genome_DpnII.txt \</span><br><span class="line">	-p reference/genome.chrom.size \</span><br><span class="line">	-D ~/opt/biosoft/juicer \</span><br><span class="line">	-t 40 &amp;&gt; juicer.log &amp;</span><br></pre></td></tr></table></figure>

<p>你可能会好奇为啥这里出现两个酶，DpnII和MboI。这是因为DpnI, DpnII, MboI, Sau3AI, 识别相同的序列，GATC，仅仅是对甲基化敏感度不同。</p>
<p>输出的结果文件都在aligned目录下，其中”merged_nodups.txt”就是下一步3D-DNA的输入文件之一</p>
<h3 id="运行3d-dna"><a href="#运行3d-dna" class="headerlink" title="运行3d-dna"></a>运行3d-dna</h3><p>3d-dna的运行也没有多少参数可以调整，如果对组装的信心高，就用<code>-r 0</code>, 否则用默认的<code>-r 2</code>就行了。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/3d-dna/run-asm-pipeline.sh -r 2 reference/genome.fa aligned/merged_nodups.txt &amp;&gt; 3d.<span class="built_in">log</span> &amp;</span><br></pre></td></tr></table></figure>

<p>然后在<a target="_blank" rel="noopener" href="https://github.com/aidenlab/juicer/wiki/Juicer-Tools-Quick-Start">Juicer-Tools</a>中对结果进行可视化，对可能的错误进行纠正</p>
<p>最后输出文件中，包含FINAL就是我们需要的结果。</p>
<h3 id="使用juicerbox进行手工纠错"><a href="#使用juicerbox进行手工纠错" class="headerlink" title="使用juicerbox进行手工纠错"></a>使用juicerbox进行手工纠错</h3><p>关于juicerbox的用法，我已经将原视频搬运到哔哩哔哩, 见<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/av65134634">https://www.bilibili.com/video/av65134634</a></p>
<p>最常见的几种组装错误:</p>
<ul>
<li>misjoin: 切割</li>
<li>translocations: 移动</li>
<li>inversions:  翻转</li>
<li>chromosome boundaries: 确定染色体的边界</li>
</ul>
<p>这些错误的判断依赖于经验，所以只能靠自己多试试了。</p>
<p>最后输出genome.review.assembly用于下一步的分析</p>
<h3 id="再次运行3d-dna"><a href="#再次运行3d-dna" class="headerlink" title="再次运行3d-dna"></a>再次运行3d-dna</h3><p>根据JABT手工纠正的结果, genome.review.assembly, 使用<code>run-asm-pipeline-post-review.sh</code>重新组装基因组。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/3d-dna/run-asm-pipeline-post-review.sh \</span><br><span class="line">    -r genome.review.assembly genome.fa aligned/merged_nodups.txt &amp;&gt; 3d.<span class="built_in">log</span> &amp;</span><br></pre></td></tr></table></figure>

<h2 id="个人使用评价"><a href="#个人使用评价" class="headerlink" title="个人使用评价"></a>个人使用评价</h2><p>juicer的代码个人感觉不是特别的好，至少以下几个地方都需要改，</p>
<ul>
<li>临时文件不会去及时删除</li>
<li>bwa得到的SAM文件处理方式有待优化，使用BAM能更快的并行计算</li>
<li>参数命令的判断很差，用-z判断字符串是否为0，而不是用-f或-d去判断文件是否存在，这个我已经提了一个issue，希望能改吧</li>
<li>Linux的sort支持多线程，但是没看到用</li>
<li>脚本中有些限速步骤的awk代码，不知道什么时候能改成更高效的处理</li>
</ul>
<p>前两条导致了运行过程中要占用大量的硬盘，所以不准备2T左右的硬盘，很容易出错。第三条是一些报错不会及时停止运算，也不容易排查。估计公司从效率角度出发，应该是写了很多脚本来替换原来的awk脚本了</p>
<p>另外，juicer在多倍体物种上表现很差，建议使用ALLHiC</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li><a target="_blank" rel="noopener" href="https://github.com/theaidenlab/3d-dna">https://github.com/theaidenlab/3d-dna</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/aidenlab/juicer">https://github.com/aidenlab/juicer</a></li>
<li><a target="_blank" rel="noopener" href="http://aidenlab.org/assembly/manual_180322.pdf">http://aidenlab.org/assembly/manual_180322.pdf</a></li>
<li><a target="_blank" rel="noopener" href="https://www.neb.com/faqs/0001/01/01/what-s-the-difference-between-dpni-dpnii-mboi-and-sau3ai">https://www.neb.com/faqs/0001/01/01/what-s-the-difference-between-dpni-dpnii-mboi-and-sau3ai</a></li>
</ul>
<p>假如你不小心设置了错误的<code>-p</code>参数，也不是特别的要紧，因为之后在最后阶段（final） 才会遇到了下面这个报错</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Could not find chromosome sizes file <span class="keyword">for</span>: reference/genome.chrom.size</span><br><span class="line">***! Can<span class="string">&#x27;t find inter.hic in aligned/inter_30.hic</span></span><br><span class="line"><span class="string">***! Error! Either inter.hic or inter_30.hic were not created</span></span><br><span class="line"><span class="string">Either inter.hic or inter_30.hic were not created. Check aligned for results</span></span><br></pre></td></tr></table></figure>

<p>即便遇到了这个报错也不要紧，因为inter.hic 和 inter_30.hic在3d-dna流程中用不到，所以不需要解决。</p>
<p>如果需要解决的话，有两个解决方案，一种重新运行命令，只不过多加一个参数<code>-S final</code>, 就会跳过之前的比对，合并和去重步骤，直接到后面STATISTICS环节。但是这样依旧会有一些不必要的计算工作，所以另一种方法就是运行原脚本必要的代码</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">juiceDir=~/opt/biosoft/juicer</span><br><span class="line">outputdir=aligned</span><br><span class="line">genomePath=reference/genome.chrom.size</span><br><span class="line">site_file=reference/genome_DpnII.txt</span><br><span class="line">ligation=GATCGATC</span><br><span class="line"><span class="comment"># output is inter.hic</span></span><br><span class="line"><span class="variable">$&#123;juiceDir&#125;</span>/scripts/common/juicer_tools pre -f <span class="variable">$site_file</span> -s <span class="variable">$outputdir</span>/inter.txt -g <span class="variable">$outputdir</span>/inter_hists.m -q 1 <span class="variable">$outputdir</span>/merged_nodups.txt <span class="variable">$outputdir</span>/inter.hic <span class="variable">$genomePath</span> </span><br><span class="line"><span class="comment"># output is inter_30.txt</span></span><br><span class="line"><span class="variable">$&#123;juiceDir&#125;</span>/scripts/common/statistics.pl -s <span class="variable">$site_file</span> -l <span class="variable">$ligation</span> -o <span class="variable">$outputdir</span>/inter_30.txt -q 30 <span class="variable">$outputdir</span>/merged_nodups.txt</span><br><span class="line"><span class="comment"># output is inter_30.hic</span></span><br><span class="line"><span class="variable">$&#123;juiceDir&#125;</span>/scripts/common/juicer_tools pre -f <span class="variable">$site_file</span> -s <span class="variable">$outputdir</span>/inter_30.txt -g <span class="variable">$outputdir</span>/inter_30_hists.m -q 30 <span class="variable">$outputdir</span>/merged_nodups.txt <span class="variable">$outputdir</span>/inter_30.hic <span class="variable">$genomePath</span></span><br></pre></td></tr></table></figure>


      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/25/%E5%88%A9%E7%94%A83D-DNA%E6%B5%81%E7%A8%8B%E7%BB%84%E8%A3%85%E5%9F%BA%E5%9B%A0%E7%BB%84-scaffolding-genome-using-3d-dna-workflow/" data-id="cllg90y4k00e5eqp26wo51p4d" data-title="利用3D-DNA流程组装基因组" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-样本足够多时别用DESeq2，用非参数检验都行-DESeq2-vs-wilcox-in-DE-analysis" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/25/%E6%A0%B7%E6%9C%AC%E8%B6%B3%E5%A4%9F%E5%A4%9A%E6%97%B6%E5%88%AB%E7%94%A8DESeq2%EF%BC%8C%E7%94%A8%E9%9D%9E%E5%8F%82%E6%95%B0%E6%A3%80%E9%AA%8C%E9%83%BD%E8%A1%8C-DESeq2-vs-wilcox-in-DE-analysis/" class="article-date">
  <time class="dt-published" datetime="2019-08-25T14:00:16.175Z" itemprop="datePublished">2019-08-25</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/25/%E6%A0%B7%E6%9C%AC%E8%B6%B3%E5%A4%9F%E5%A4%9A%E6%97%B6%E5%88%AB%E7%94%A8DESeq2%EF%BC%8C%E7%94%A8%E9%9D%9E%E5%8F%82%E6%95%B0%E6%A3%80%E9%AA%8C%E9%83%BD%E8%A1%8C-DESeq2-vs-wilcox-in-DE-analysis/">样本足够多时别用DESeq2，用非参数检验都行</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>果子老师做过一个非常惊人的举动，用DESeq2处理1225例样本的TCGA数据，在没有使用<code>DESeq</code>多线程参数<code>parallel</code>的情况下，跑了将近40个小时。</p>
<p>那么问题来了，在那么大的样本量的情况下，应该用DESeq2进行数据处理吗？我的结论是不应该，DESeq2的适用场景是小样本的差异表分析，降低假阳性。当你的样本量足够多的时候，我们其实有更好的选择。</p>
<p>这里以果子老师的数据为例，来对比DESeq2的结果和我的分析结果进行比较.</p>
<p>加载DESeq2结果</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">load(file=<span class="string">&quot;dds_very_long.Rdata&quot;</span>)</span><br><span class="line">library(DESeq2)</span><br><span class="line">deseq2_result &lt;- results(dds)</span><br><span class="line">table(deseq2_result$padj &lt; <span class="number">0.01</span>)</span><br><span class="line"><span class="comment"># FALSE  TRUE </span></span><br><span class="line"><span class="comment"># 23997 25072</span></span><br></pre></td></tr></table></figure>

<p>下面我分析时的数据预处理 部分，</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">options<span class="punctuation">(</span>stringsAsFactors <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 加载数据</span></span><br><span class="line">load<span class="punctuation">(</span>file <span class="operator">=</span> <span class="string">&quot;BRCA_RNASEQ_exprdf.Rdata&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取表达量矩阵</span></span><br><span class="line">expr_mt <span class="operator">&lt;-</span> as.matrix<span class="punctuation">(</span>expr_df<span class="punctuation">[</span><span class="punctuation">,</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">row.names<span class="punctuation">(</span>expr_mt<span class="punctuation">)</span> <span class="operator">&lt;-</span> expr_df<span class="operator">$</span>gene_id</span><br><span class="line">colnames<span class="punctuation">(</span>expr_mt<span class="punctuation">)</span> <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>expr_df<span class="punctuation">)</span><span class="punctuation">[</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 根据文库大小标准化</span></span><br><span class="line">expr_mt <span class="operator">&lt;-</span> expr_mt <span class="operator">/</span> <span class="built_in">rep</span><span class="punctuation">(</span>colSums<span class="punctuation">(</span>expr_mt<span class="punctuation">)</span><span class="punctuation">,</span> each<span class="operator">=</span>nrow<span class="punctuation">(</span>expr_mt<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">*</span> <span class="number">1e6</span></span><br><span class="line"><span class="comment"># 过滤地表达基因</span></span><br><span class="line">expr_mt <span class="operator">&lt;-</span> expr_mt<span class="punctuation">[</span>rowSums<span class="punctuation">(</span>expr_mt <span class="operator">&gt;</span> <span class="number">0</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="punctuation">(</span>ncol<span class="punctuation">(</span>expr_mt<span class="punctuation">)</span> <span class="operator">/</span> <span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 统计癌症和癌旁</span></span><br><span class="line">TCGA_id <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>expr_mt<span class="punctuation">)</span></span><br><span class="line">table<span class="punctuation">(</span>substring<span class="punctuation">(</span>TCGA_id<span class="punctuation">,</span><span class="number">14</span><span class="punctuation">,</span><span class="number">15</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">### 我们发现了7个转移的样本，本次分析，我们关注的是癌症和癌旁，先把转移的样本去掉</span></span><br><span class="line"><span class="comment">### 原发和转移的对比作为家庭作业</span></span><br><span class="line"></span><br><span class="line">TCGA_id <span class="operator">&lt;-</span> TCGA_id<span class="punctuation">[</span>substring<span class="punctuation">(</span>TCGA_id<span class="punctuation">,</span><span class="number">14</span><span class="punctuation">,</span><span class="number">15</span><span class="punctuation">)</span><span class="operator">!=</span><span class="string">&quot;06&quot;</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment">### 创建metadata</span></span><br><span class="line">sample <span class="operator">&lt;-</span> ifelse<span class="punctuation">(</span>substring<span class="punctuation">(</span>TCGA_id<span class="punctuation">,</span><span class="number">14</span><span class="punctuation">,</span><span class="number">15</span><span class="punctuation">)</span><span class="operator">==</span><span class="string">&quot;01&quot;</span><span class="punctuation">,</span><span class="string">&quot;cancer&quot;</span><span class="punctuation">,</span><span class="string">&quot;normal&quot;</span><span class="punctuation">)</span></span><br><span class="line">sample <span class="operator">&lt;-</span> factor<span class="punctuation">(</span>sample<span class="punctuation">,</span>levels <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;normal&quot;</span><span class="punctuation">,</span><span class="string">&quot;cancer&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>ordered <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line">metadata <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>TCGA_id<span class="punctuation">,</span>sample<span class="punctuation">)</span> </span><br></pre></td></tr></table></figure>

<p>下一步，利用非参数检验方法， wilcox.test，关于非参数检验的缘起可以看「女士品茶」的第16章摆脱参数</p>
<blockquote>
<p>威尔科克森注释着计算t检验和方法分析的公式，意识到这些不同寻常的极端数值会对结果产生极大的影响，导致“学生”的t检验偏小。 … 如果异常值体现了某种因素对系统数据的系统性污染，那么使用非参数方法只会让事情变得更糟。</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># wilcox.test差异分析 ---------------------------------------------------------</span></span><br><span class="line">cancer_sample <span class="operator">&lt;-</span> metadata<span class="punctuation">[</span>metadata<span class="operator">$</span>sample <span class="operator">==</span> <span class="string">&quot;cancer&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TCGA_id&quot;</span><span class="punctuation">]</span></span><br><span class="line">normal_sample <span class="operator">&lt;-</span> metadata<span class="punctuation">[</span>metadata<span class="operator">$</span>sample <span class="operator">==</span> <span class="string">&quot;normal&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TCGA_id&quot;</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">cancer_mt <span class="operator">&lt;-</span> expr_mt<span class="punctuation">[</span><span class="punctuation">,</span>colnames<span class="punctuation">(</span>expr_mt<span class="punctuation">)</span> <span class="operator">%in%</span> cancer_sample <span class="punctuation">]</span></span><br><span class="line">normal_mt <span class="operator">&lt;-</span> expr_mt<span class="punctuation">[</span><span class="punctuation">,</span>colnames<span class="punctuation">(</span>expr_mt<span class="punctuation">)</span> <span class="operator">%in%</span> normal_sample <span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 计算logFoldChanges</span></span><br><span class="line">logFC <span class="operator">&lt;-</span> log2<span class="punctuation">(</span>rowMeans<span class="punctuation">(</span>as.matrix<span class="punctuation">(</span>cancer_df<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">/</span> rowMeans<span class="punctuation">(</span>as.matrix<span class="punctuation">(</span>normal_df<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">library<span class="punctuation">(</span>future.apply<span class="punctuation">)</span></span><br><span class="line">plan<span class="punctuation">(</span>multiprocess<span class="punctuation">)</span></span><br><span class="line">p_values <span class="operator">&lt;-</span> future_lapply<span class="punctuation">(</span>seq<span class="punctuation">(</span>nrow<span class="punctuation">(</span>cancer_df<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  res <span class="operator">&lt;-</span> wilcox.test<span class="punctuation">(</span>x <span class="operator">=</span> cancer_mt<span class="punctuation">[</span>x<span class="punctuation">,</span><span class="punctuation">]</span><span class="punctuation">,</span> y <span class="operator">=</span>  normal_mt<span class="punctuation">[</span>x<span class="punctuation">,</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">  res<span class="operator">$</span>p.value</span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p <span class="operator">&lt;-</span> unlist<span class="punctuation">(</span>p_values<span class="punctuation">)</span></span><br><span class="line">p.adj <span class="operator">&lt;-</span> p.adjust<span class="punctuation">(</span>p<span class="punctuation">,</span> method <span class="operator">=</span> <span class="string">&quot;fdr&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">table<span class="punctuation">(</span>p.adj <span class="operator">&lt;</span> <span class="number">0.01</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># FALSE  TRUE </span></span><br><span class="line"><span class="comment"># 10997 24030</span></span><br></pre></td></tr></table></figure>

<p>我们得到了24,030个校正后p值小于0.01的基因，而DESeq2是25,072个。如果比较全部的基因的话，韦恩图上可以发现，绝大部分基因都是相同的。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/compare_total-e4635c9e41a94c169d51af5a4ae35810.png" alt="总体比较"></p>
<p>但是通常情况下，我们会更去关注一些变化比较大且p值显著的基因，用这些基因去做下游的富集分析。所以，下一步就是看看后面富集分析结果两者有什么区别。</p>
<p>我们用Y叔的clusterProfiler，去分析倍数变化4倍，矫正p值小于0.01的基因</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 提取基因</span></span><br><span class="line">library<span class="punctuation">(</span>clusterProfiler<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>org.Hs.eg.db<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">org <span class="operator">&lt;-</span> org.Hs.eg.db</span><br><span class="line">diffgene1 <span class="operator">&lt;-</span> row.names<span class="punctuation">(</span>expr_mt<span class="punctuation">)</span><span class="punctuation">[</span>p.adj <span class="operator">&lt;</span> <span class="number">0.01</span> <span class="operator">&amp;</span> <span class="built_in">abs</span><span class="punctuation">(</span>logFC<span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="number">2</span><span class="punctuation">]</span></span><br><span class="line">diffgene1 <span class="operator">&lt;-</span> substr<span class="punctuation">(</span>diffgene1<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> <span class="number">15</span><span class="punctuation">)</span></span><br><span class="line">diffgene2 <span class="operator">&lt;-</span> row.names<span class="punctuation">(</span>deseq2_result<span class="punctuation">)</span><span class="punctuation">[</span>deseq2_result<span class="operator">$</span>padj <span class="operator">&lt;</span> <span class="number">0.01</span> <span class="operator">&amp;</span> </span><br><span class="line">                                           <span class="operator">!</span> <span class="built_in">is.na</span><span class="punctuation">(</span>deseq2_result<span class="operator">$</span>padj<span class="punctuation">)</span> <span class="operator">&amp;</span></span><br><span class="line">                                           <span class="built_in">abs</span><span class="punctuation">(</span>deseq2_result<span class="operator">$</span>log2FoldChange<span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="number">2</span><span class="punctuation">]</span></span><br><span class="line">diffgene2 <span class="operator">&lt;-</span> substr<span class="punctuation">(</span>diffgene2<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> <span class="number">15</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>GO富集分析</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>clusterProfiler<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>org.Hs.eg.db<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">org <span class="operator">&lt;-</span> org.Hs.eg.db</span><br><span class="line">diffgene1 <span class="operator">&lt;-</span> row.names<span class="punctuation">(</span>expr_mt<span class="punctuation">)</span><span class="punctuation">[</span>p.adj <span class="operator">&lt;</span> <span class="number">0.01</span> <span class="operator">&amp;</span> <span class="built_in">abs</span><span class="punctuation">(</span>logFC<span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="number">2</span><span class="punctuation">]</span></span><br><span class="line">diffgene1 <span class="operator">&lt;-</span> substr<span class="punctuation">(</span>diffgene1<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> <span class="number">15</span><span class="punctuation">)</span></span><br><span class="line">diffgene2 <span class="operator">&lt;-</span> row.names<span class="punctuation">(</span>deseq2_result<span class="punctuation">)</span><span class="punctuation">[</span>deseq2_result<span class="operator">$</span>padj <span class="operator">&lt;</span> <span class="number">0.01</span> <span class="operator">&amp;</span> </span><br><span class="line">                                           <span class="operator">!</span> <span class="built_in">is.na</span><span class="punctuation">(</span>deseq2_result<span class="operator">$</span>padj<span class="punctuation">)</span> <span class="operator">&amp;</span></span><br><span class="line">                                           <span class="built_in">abs</span><span class="punctuation">(</span>deseq2_result<span class="operator">$</span>log2FoldChange<span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="number">2</span><span class="punctuation">]</span></span><br><span class="line">diffgene2 <span class="operator">&lt;-</span> substr<span class="punctuation">(</span>diffgene2<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> <span class="number">15</span><span class="punctuation">)</span></span><br><span class="line">ego1 <span class="operator">&lt;-</span> enrichGO<span class="punctuation">(</span>diffgene1<span class="punctuation">,</span> </span><br><span class="line">                 OrgDb <span class="operator">=</span> org<span class="punctuation">,</span></span><br><span class="line">                 keyType <span class="operator">=</span> <span class="string">&quot;ENSEMBL&quot;</span><span class="punctuation">,</span></span><br><span class="line">                 ont <span class="operator">=</span> <span class="string">&quot;BP&quot;</span></span><br><span class="line">                 <span class="punctuation">)</span></span><br><span class="line">ego2 <span class="operator">&lt;-</span> enrichGO<span class="punctuation">(</span>diffgene2<span class="punctuation">,</span> </span><br><span class="line">                 OrgDb <span class="operator">=</span> org<span class="punctuation">,</span></span><br><span class="line">                 keyType <span class="operator">=</span> <span class="string">&quot;ENSEMBL&quot;</span><span class="punctuation">,</span></span><br><span class="line">                 ont <span class="operator">=</span> <span class="string">&quot;BP&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">merge_result <span class="operator">&lt;-</span> merge_result<span class="punctuation">(</span><span class="built_in">list</span><span class="punctuation">(</span>wilcox<span class="operator">=</span>ego1<span class="punctuation">,</span>DESeq2<span class="operator">=</span>ego2<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">dotplot<span class="punctuation">(</span>merge_result<span class="punctuation">,</span>showCategory<span class="operator">=</span> <span class="number">20</span> <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/compare_dotplot-916bf6b3815d45b288e3a5e3459bf62e.png" alt="比较20个GO词条"></p>
<p>从点图中，你可以认为这两个分析结果是一致。</p>
<p>综上，当你在样本量足够多（两组都不少于10吧），其实没有去用DESeq2这些复杂的工具，用基础的统计学检验方法就能得到很好的结果了。</p>
<p>在样本量比较小的时候，用复杂的模型是无奈之举，它有很多假设成分在，尤其是你还想从无重复的实验设计中算p值。当你样本量够多的时候，用最简单的模型其实就会有很好的结果。</p>
<blockquote>
<p>本次分析用到的数据可以通过在微信公众号搜索 果子学生信 后台回复 “果子学统计” ，就可以拿到了</p>
</blockquote>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/25/%E6%A0%B7%E6%9C%AC%E8%B6%B3%E5%A4%9F%E5%A4%9A%E6%97%B6%E5%88%AB%E7%94%A8DESeq2%EF%BC%8C%E7%94%A8%E9%9D%9E%E5%8F%82%E6%95%B0%E6%A3%80%E9%AA%8C%E9%83%BD%E8%A1%8C-DESeq2-vs-wilcox-in-DE-analysis/" data-id="cllg90y6m00jdeqp24oju7tdl" data-title="样本足够多时别用DESeq2，用非参数检验都行" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何安装perl模块-How-to-install-perl-module" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/23/%E5%A6%82%E4%BD%95%E5%AE%89%E8%A3%85perl%E6%A8%A1%E5%9D%97-How-to-install-perl-module/" class="article-date">
  <time class="dt-published" datetime="2019-08-23T12:57:37.759Z" itemprop="datePublished">2019-08-23</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/23/%E5%A6%82%E4%BD%95%E5%AE%89%E8%A3%85perl%E6%A8%A1%E5%9D%97-How-to-install-perl-module/">如何安装perl模块</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>由于生物信息早期最多用的语言是perl，因此不可避免就要用别人的perl脚本或者基于perl的项目来处理数据。</p>
<p>使用perl脚本和使用其他编程语言的脚本没啥不同，毕竟你只要传入参数，它就能给你结果。因此对于我们这些不用perl写脚本，只需要调用的人而言，唯一要学会的事情就是**如何安装perl的模块”。</p>
<p>关于perl模块安装，最古老的方法就是使用<code>perl -MCPAN -e shell</code>或者是<code>cpan</code>（两者等价），这也是我最先接触的方法，这里介绍如何使用<code>local::lib</code>和<code>cpanm</code>实现非root权限安装perl模块。</p>
<h2 id="使用系统自带的perl"><a href="#使用系统自带的perl" class="headerlink" title="使用系统自带的perl"></a>使用系统自带的perl</h2><p>安装任何软件最怕遇到的问题就是权限问题，因此我们需要先安装<code>local::lib</code>，使得我们能够将perl模块安装到任何地方，简单的说就是安装到我们的家目录下</p>
<p>第一步，下载源代码进行编译安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">wget https://cpan.metacpan.org/authors/id/H/HA/HAARG/local-lib-2.000024.tar.gz</span><br><span class="line">tar xf local-lib-2.000024.tar.gz</span><br><span class="line"><span class="built_in">cd</span> local-lib-2.000024</span><br><span class="line">perl Makefile.PL --bootstrap=~/opt</span><br><span class="line">make <span class="built_in">test</span> &amp;&amp; make install</span><br></pre></td></tr></table></figure>

<p>第二步：<strong>设置环境变量</strong>，使得perl在安装模块的时候会优先使用我们指定的路径</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">echo</span> <span class="string">&#x27;eval &quot;$(perl -I$HOME/opt/lib/perl5 -Mlocal::lib=$HOME/opt)&quot;&#x27;</span> &gt;&gt; ~/.bashrc</span><br></pre></td></tr></table></figure>

<p>先用<code>perl -I$HOME/opt/lib/perl5 -Mlocal::lib=$HOME/opt</code>表示运行前先添加<code>$HOME/opt/lib/perl5</code>到自己的搜索路径<code>@INC</code>中，然后传入参数<code>$HOME/opt</code>执行模块<code>local::lob</code>，这个模块的执行结果会输出如下内容</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Attempting to create directory /home6/wangjw/opt</span><br><span class="line">PATH=<span class="string">&quot;/home/zgxu/opt/bin<span class="variable">$&#123;PATH:+:<span class="variable">$&#123;PATH&#125;</span>&#125;</span>&quot;</span>; <span class="built_in">export</span> PATH;</span><br><span class="line">PERL5LIB=<span class="string">&quot;/home/zgxu/opt/lib/perl5<span class="variable">$&#123;PERL5LIB:+:<span class="variable">$&#123;PERL5LIB&#125;</span>&#125;</span>&quot;</span>; <span class="built_in">export</span> PERL5LIB;</span><br><span class="line">PERL_LOCAL_LIB_ROOT=<span class="string">&quot;/home/zgxu/opt<span class="variable">$&#123;PERL_LOCAL_LIB_ROOT:+:<span class="variable">$&#123;PERL_LOCAL_LIB_ROOT&#125;</span>&#125;</span>&quot;</span>; <span class="built_in">export</span> PERL_LOCAL_LIB_ROOT;</span><br><span class="line">PERL_MB_OPT=<span class="string">&quot;--install_base \&quot;/home/zgxu/opt\&quot;&quot;</span>; <span class="built_in">export</span> PERL_MB_OPT;</span><br><span class="line">PERL_MM_OPT=<span class="string">&quot;INSTALL_BASE=/home/zgxu/opt&quot;</span>; <span class="built_in">export</span> PERL_MM_OPT;</span><br></pre></td></tr></table></figure>

<p>这些就作为<code>eval</code>的参数进行执行，也就是说你重启终端后后，<code>PERL5LIB</code> <code>PERL_LOCAL_LIB_ROOT</code>,<code>PERL_MB_OPT</code>,<code>PERL_MM_OPT</code>这几个变量就会重新设置，以此保证你后续安装perl模块时，会优先安装到自己的选择的目录</p>
<p>第三步：安装cpam. 由于之前已经配置了<code>local::lib</code>，因此perl编译的工具都会默认安装到<code>~/opt</code>目录下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">wget https://cpan.metacpan.org/authors/id/M/MI/MIYAGAWA/App-cpanminus-1.7043.tar.gz</span><br><span class="line">tar xf App-cpanminus-1.7043.tar.gz</span><br><span class="line"><span class="built_in">cd</span> App-cpanminus-1.7043</span><br><span class="line">perl Makefile.PL</span><br><span class="line">make <span class="built_in">test</span> &amp;&amp; make install</span><br></pre></td></tr></table></figure>

<p>第四步：使用国内镜像提高下载速度，可以通过别名的方式实现</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">echo</span> <span class="string">&#x27;alias cpanm=&quot;cpanm --mirror http://mirrors.163.com/cpan --mirror-only&quot;&#x27;</span> &gt;&gt;~/.bashrc</span><br></pre></td></tr></table></figure>

<p>之后便可以使用<code>cpanm Module::Name</code>安装任意的软件了。</p>
<h2 id="自己编译一个perl"><a href="#自己编译一个perl" class="headerlink" title="自己编译一个perl"></a>自己编译一个perl</h2><p>自己编译Perl的好处就在于之后的perl模块都会安装到自己的Perl目录下，而不会对系统造成影响。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget -4 http://www.cpan.org/src/5.0/perl-5.26.1.tar.gz</span><br><span class="line">tar xf perl-5.26.1.tar.gz</span><br><span class="line"><span class="built_in">cd</span> perl-5.26.1</span><br><span class="line">./Configure -des -Dprefix=<span class="variable">$HOME</span>/opt/sysoft/perl-5.26.1</span><br><span class="line">make <span class="built_in">test</span></span><br><span class="line">make install</span><br></pre></td></tr></table></figure>

<p>然后用<code>perl -e &#39;&#123;print &quot;$_\n&quot; foreach @INC&#125;&#39;</code>会发现perl只会在自己的目录<code>~/opt/sysoft/perl-5.26.1</code>下查找模块。那么使用<code>cpanm Module::Name</code>安装的任何包都只会安装到<code>~/opt/sysoft/perl-5.26.1</code>下，你也不需要安装<code>local::lib</code>了</p>
<h2 id="conda的perl和系统的perl冲突"><a href="#conda的perl和系统的perl冲突" class="headerlink" title="conda的perl和系统的perl冲突"></a>conda的perl和系统的perl冲突</h2><p>有一次我遇到这个问题</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl: symbol lookup error: /home/wangjw/perl5/lib/perl5/x86_64-linux-thread-multi/auto/Cwd/Cwd.so: undefined symbol</span><br></pre></td></tr></table></figure>

<p>这个问题是因为用系统perl安装的软件被conda的perl优先查找到导致，用<code>perl -V</code>和<code>perl -e &#39;&#123;print &quot;$_\n&quot; foreach @INC&#125;&#39;</code>可以发现conda的perl查找路径低于我为系统perl安装的路径，解决方案如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">export</span> PERL5LIB=<span class="string">&quot;&quot;</span></span><br></pre></td></tr></table></figure>


      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/23/%E5%A6%82%E4%BD%95%E5%AE%89%E8%A3%85perl%E6%A8%A1%E5%9D%97-How-to-install-perl-module/" data-id="cllg90y5500gbeqp29p8bgmjp" data-title="如何安装perl模块" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-3分的水稻转录组分析实战-Rice-RNA-seq-analysis" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/22/3%E5%88%86%E7%9A%84%E6%B0%B4%E7%A8%BB%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%E5%AE%9E%E6%88%98-Rice-RNA-seq-analysis/" class="article-date">
  <time class="dt-published" datetime="2019-08-22T19:54:49.801Z" itemprop="datePublished">2019-08-22</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/22/3%E5%88%86%E7%9A%84%E6%B0%B4%E7%A8%BB%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%E5%AE%9E%E6%88%98-Rice-RNA-seq-analysis/">3分的水稻转录组分析实战</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="水稻转录组分析实战"><a href="#水稻转录组分析实战" class="headerlink" title="水稻转录组分析实战"></a>水稻转录组分析实战</h1><p>文章标题: Genes, pathways and transcription factors involved in seedling stage chilling stress tolerance in indica rice through RNA-Seq analysis </p>
<p>该文章发表在BMC plant biology，植物科学2区SCI，近四年IF稳定在3分左右。这篇文章的核心部分就是测了好几个转录组，然后做了一点实验验证。</p>
<h2 id="文章简介"><a href="#文章简介" class="headerlink" title="文章简介"></a>文章简介</h2><p>本部分内容由我的一位师弟完成</p>
<h3 id="背景"><a href="#背景" class="headerlink" title="背景"></a>背景</h3><p>世界50％以上的人口将稻米作为饮食中碳水化合物的主要来源。这是因为水稻的种植范围很广，从山地到平原地区均有种植。正因如此，水稻的一生会面临多种生物胁迫（如病原菌，昆虫等）和非生物胁迫（如寒冷，高温，盐碱等），因此解析水稻的抗逆机制，培育具有多种抗逆性状的水稻具有很大的应用价值。而本文的作者是印度人，印度大部分地区的气候具有明显的旱雨两季，旱季的水稻在苗期易受冷（寒）害的影响。随着RNA-seq手段越来越成熟，在比较不同转录组的差异中发挥了重要作用，这对做基因功能的实验室来讲无疑是巨大的福音。作者就想利用RNA-seq的手段，寻找耐寒品种（Geetanjali）和冷敏感品种（Shahabhagidhan）在转录组水平的差异表达的基因，以期在水稻耐寒过程中鉴定到新颖的转录本，并能获得对基因不同表达水平的认识。</p>
<h3 id="方法"><a href="#方法" class="headerlink" title="方法"></a>方法</h3><p>总体思路，对耐寒和冷敏感的品种在冷处理过程中的不同时间点取样，并进行转录组分析，分析完成后，用qRT-PCR来验证表达上调和下调的基因，进一步confirm分析结果。</p>
<h3 id="结果"><a href="#结果" class="headerlink" title="结果"></a>结果</h3><p>基于低温胁迫处理的5个不同时间点的比较转录组分析显示：在冷敏感品种中有更多的下调基因，在耐寒品种中有更多上调的基因。 在冷敏感品种（CSV）中检测到总计13930个差异基因，而在耐寒品种（CTV）中检测到10599个差异基因，且随时间的延长，差异基因的数目也增加。GO富集分析揭示了18个CSV项和28个CTV项显著参与到molecular function, cellular component and biological process。GO分类揭示了耐寒过程中的差异基因显著参与转录调控，活性氧爆发，脂质结合，催化和水解酶活性的作用。KEGG注释通路揭示了更多数量的基因在耐寒过程中调节不同的通路来发挥作用。</p>
<h3 id="结论"><a href="#结论" class="headerlink" title="结论"></a>结论</h3><p>结论真的是简单到吐血，真的是简单到吐血，真的是简单到吐血。就一句话：通过对耐寒品种的转录组分析，我们检测到了涉及到抗逆过程的基因，通路和转录因子。</p>
<h2 id="转录组分析重现"><a href="#转录组分析重现" class="headerlink" title="转录组分析重现"></a>转录组分析重现</h2><blockquote>
<p>转录组分析环境的搭建和一个分析流程全过程可以哔哩哔哩上搜索”徐洲更”, 即可看我录制的视频。对于学习过程中出现的问题，可以参加我们的付费答疑群，活动见<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/w47Hc2BmRCLpQMlRZ4rCCw">https://mp.weixin.qq.com/s/w47Hc2BmRCLpQMlRZ4rCCw</a>。</p>
</blockquote>
<p>本次文章重现涉及到的软件如下:</p>
<ul>
<li>命令行：prefetch, fasterq-dump, hisat2, samtools, featureCounts, gffread</li>
<li>R：DESeq2, clusterProfiler</li>
</ul>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">conda create -n rna-seq sra-tools fastqc fastp  hisat2  samtools subread  gffread multiqc</span><br><span class="line">conda activate rna-seq</span><br></pre></td></tr></table></figure>

<h3 id="数据下载"><a href="#数据下载" class="headerlink" title="数据下载"></a>数据下载</h3><p>文章用到的RNA-seq数据编号，PRJNA288892， 一共有12个样本。我以这个项目编号新建一个文件夹，开始本次的分析</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p PRJNA288892 &amp;&amp; <span class="built_in">cd</span> PRJNA288892</span><br></pre></td></tr></table></figure>

<p>接下来是下载文章用到的数据，编号存在SRR_Acc_List.txt</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">SRR2089751</span><br><span class="line">SRR2089753</span><br><span class="line">SRR2089756</span><br><span class="line">SRR2093937</span><br><span class="line">SRR2093938</span><br><span class="line">SRR2093939</span><br><span class="line">SRR2093948</span><br><span class="line">SRR2093955</span><br><span class="line">SRR2093959</span><br><span class="line">SRR2093960</span><br><span class="line">SRR2093961</span><br><span class="line">SRR2089754</span><br></pre></td></tr></table></figure>

<p>建立一个sra文件夹，使用prefetch将数据下载到该文件夹下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> sra</span><br><span class="line"><span class="built_in">cat</span> SRR_Acc_List.txt | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> prefetch -O sra <span class="variable">$id</span> ; <span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<p>然后将sra文件转成fastq格式</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p 00-raw-data</span><br><span class="line"><span class="built_in">cat</span> SRR_Acc_List.txt | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> fasterq-dump -p -3 -O 00-raw-data sra/<span class="variable">$id</span>.sra ; <span class="keyword">done</span> &amp;</span><br></pre></td></tr></table></figure>

<h3 id="原始数据质控"><a href="#原始数据质控" class="headerlink" title="原始数据质控"></a>原始数据质控</h3><p>原文只说了High quality (QV &gt; 25) 的read用于比对，并没有提到具体的质控的软件。我们这里用fastp做质控，因为fastp跑完之后也会输出一个网页，所以我也就懒得用FastQC再跑一次</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p 01-clean-data</span><br><span class="line"><span class="built_in">mkdir</span> -p qc</span><br><span class="line"><span class="built_in">cat</span> SRR_Acc_List.txt | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> fastp -w 4 -i 00-raw-data/<span class="variable">$id</span>.sra_1.fastq -I 00-raw-data/<span class="variable">$id</span>.sra_2.fastq -o 01-clean-data/<span class="variable">$id</span>.sra_1.fastq.gz -O 01-clean-data/<span class="variable">$id</span>.sra_2.fastq.gz --html qc/<span class="variable">$id</span>.html --json qc/<span class="variable">$id</span>.json ; <span class="keyword">done</span> &amp;</span><br></pre></td></tr></table></figure>

<blockquote>
<p>重申一个观点，对于<strong>普通</strong>转录组，由于测序质量过关，比对软件也能处理开头的接头序列，我基本不做质控。</p>
</blockquote>
<h3 id="建立索引"><a href="#建立索引" class="headerlink" title="建立索引"></a>建立索引</h3><p>文章说它用到水稻参考基因组和注释文件下载自<a href="ftp://ftp.plantbiology.msu.edu/pub/data/EukaryoticProjects/osativa/annotationdbs/pseudomolecules/version_7.0/all.dir/">ftp://ftp.plantbiology.msu.edu/pub/data/EukaryoticProjects/osativa/annotationdbs/pseudomolecules/version_7.0/all.dir/</a>, 而有趣的是这个链接我打不开，目前可用的下载地址其实是: <a target="_blank" rel="noopener" href="http://rice.plantbiology.msu.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/">http://rice.plantbiology.msu.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> reference &amp;&amp; <span class="built_in">cd</span> reference</span><br><span class="line"><span class="comment"># 参考基因组</span></span><br><span class="line">wget http://rice.plantbiology.msu.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/all.con</span><br><span class="line"><span class="comment"># 注释文件</span></span><br><span class="line">wget http://rice.plantbiology.msu.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/all.gff3</span><br></pre></td></tr></table></figure>

<p>此外文章用的是TopHat v1.3.3和Bowtie v0.12.9  默认参数做的比对，我对此不作任何评价。我们这里用HISAT2建立索引</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">hhisat2-build -p 10 reference/all.con reference/Osativa</span><br></pre></td></tr></table></figure>

<h3 id="序列比对"><a href="#序列比对" class="headerlink" title="序列比对"></a>序列比对</h3><p>用HISAT2将过滤后的fastq文件和参考基因组进行比对，得到下游分析所需要的BAM文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p 02-read-align</span><br><span class="line"><span class="built_in">cat</span> SRR_Acc_List.txt | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> hisat2 -p 10 -x reference/Osativa -1 01-clean-data/<span class="variable">$id</span>.sra_1.fastq.gz -2 01-clean-data/<span class="variable">$id</span>.sra_2.fastq.gz --new-summary --summary-file qc/<span class="variable">$id</span>.ht2.txt | samtools <span class="built_in">sort</span> -@ 4 &gt; 02-read-align/<span class="variable">$id</span>.bam ; <span class="keyword">done</span> &amp;</span><br></pre></td></tr></table></figure>

<h3 id="表达量定量"><a href="#表达量定量" class="headerlink" title="表达量定量"></a>表达量定量</h3><p>文章使用的Cufflinks v1.3.0对基因的表达量进行定量，使用FPKM进行了标准化。我这里也不做评价，我们用featureCounts进行表达量定量</p>
<p>先将GFF3转成GTF格式</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gffread reference/all.gff3 -T -o reference/Osativa.gtf</span><br></pre></td></tr></table></figure>

<p>然后用featureCoutns</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p 03-gene-count</span><br><span class="line">featureCounts -a reference/Osativa.gtf -o 03-gene-count/gene_counts 02-read-align/*.bam -T 10 -g gene_id</span><br><span class="line"><span class="built_in">mv</span> 03-gene-count/gene_counts.summary qc/gene_counts.summary</span><br></pre></td></tr></table></figure>

<p>最终的结果，就可以读取到R语言里面进行分析了</p>
<h3 id="MultiQC"><a href="#MultiQC" class="headerlink" title="MultiQC"></a>MultiQC</h3><p>上面每一步的中间文件都被我放到了qc文件夹里，因此可以用multiqc进行整合</p>
<h3 id="差异分析"><a href="#差异分析" class="headerlink" title="差异分析"></a>差异分析</h3><blockquote>
<p>由于前面几步的运算时间比较久，我这里直接准备了表达量矩阵，下载方式为百度网盘链接：<a target="_blank" rel="noopener" href="https://pan.baidu.com/s/1c_mdGlDggV2DH79H-X7J_g">https://pan.baidu.com/s/1c_mdGlDggV2DH79H-X7J_g</a> 提取码：jmi5 </p>
</blockquote>
<p>差异表达分析用但是精确检验(exact test)。</p>
<p>之前的SRR编号在分析的时候最好转成实际的样本编号，方便分析的时候查看</p>
<table>
<thead>
<tr>
<th>编号</th>
<th>品种</th>
<th>处理</th>
</tr>
</thead>
<tbody><tr>
<td>SRR2089751.sra</td>
<td>Sahabhagidhan</td>
<td>0h control（易感品种对照）</td>
</tr>
<tr>
<td>SRR2089753.sra</td>
<td>Sahabhagidhan</td>
<td>6h冷害处理</td>
</tr>
<tr>
<td>SRR2089754.sra</td>
<td>Sahabhagidhan</td>
<td>12h冷害处理</td>
</tr>
<tr>
<td>SRR2089756.sra</td>
<td>Sahabhagidhan</td>
<td>24h冷害处理</td>
</tr>
<tr>
<td>SRR2093937.sra</td>
<td>Sahabhagidhan</td>
<td>48h冷害处理</td>
</tr>
<tr>
<td>SRR2093938.sra</td>
<td>Sahabhagidhan</td>
<td>48h冷害处理后又24h恢复</td>
</tr>
<tr>
<td>SRR2093939.sra</td>
<td>Geetanjali</td>
<td>0h conrol</td>
</tr>
<tr>
<td>SRR2093948.sra</td>
<td>Geetanjali</td>
<td>6h冷害处理</td>
</tr>
<tr>
<td>SRR2093955.sra</td>
<td>Geetanjali</td>
<td>12h冷害处理</td>
</tr>
<tr>
<td>SRR2093959.sra</td>
<td>Geetanjali</td>
<td>24h冷害处理</td>
</tr>
<tr>
<td>SRR2093960.sra</td>
<td>Geetanjali</td>
<td>48h冷害处理</td>
</tr>
<tr>
<td>SRR2093961.sra</td>
<td>Geetanjali</td>
<td>48h冷害处理后又24h恢复</td>
</tr>
</tbody></table>
<p>Sahabhagidhan是不耐冷的品种(cold susceptible variety, CSV) ,而Geetanjali则是耐冷品种 (CTV, cold tolerant variety) 。</p>
<p>我们这里分析的时候把两个物种在6, 12, 24, 48时间点的样本作为重复，仅仅比较不同物种在冷害处理后的差异基因，而不分析冷害处理后随时间点变化的基因。</p>
<p>加载数据:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">df &lt;- read.table(&quot;gene_counts&quot;, header = T,</span><br><span class="line">                 sep = &quot;\t&quot;,</span><br><span class="line">                 stringsAsFactors = F)</span><br><span class="line">expr_df &lt;- df[,c(1,8,9,10,11,14,15,16,17)]</span><br><span class="line">colnames(expr_df) &lt;- c(&quot;geneid&quot;,</span><br><span class="line">                       &quot;CSV_1&quot;,&quot;CSV_2&quot;,&quot;CSV_3&quot;,&quot;CSV_4&quot;,</span><br><span class="line">                       &quot;CTV_1&quot;,&quot;CTV_2&quot;,&quot;CTV_3&quot;,&quot;CTV_4&quot;)</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>DESeq2差异分析</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">library(DESeq2)</span><br><span class="line">coldata &lt;- data.frame(condition=factor(c(&quot;CSV&quot;,&quot;CSV&quot;,&quot;CSV&quot;,&quot;CSV&quot;,</span><br><span class="line">                      &quot;CTV&quot;,&quot;CTV&quot;,&quot;CTV&quot;,&quot;CTV&quot;)))</span><br><span class="line">dds &lt;- DESeqDataSetFromMatrix(expr_df,</span><br><span class="line">                       colData = coldata,</span><br><span class="line">                       design = ~ condition,</span><br><span class="line">                       tidy = TRUE)</span><br><span class="line">dds &lt;- DESeq(dds)</span><br></pre></td></tr></table></figure>

<p>PCA比较下组内差异是不是比组间差异小</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">rld &lt;- rlog(dds)</span><br><span class="line">plotPCA(rld)</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/rice_pca-e000a342aec6400f969c460fdc71c44c.png" alt="PCA"></p>
<p>DESeq2的差异表达分析，只需要一行代码</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">res &lt;- results(dds)</span><br></pre></td></tr></table></figure>

<p>用ggplot2画个火山图,</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">plot_df &lt;- data.frame(res)</span><br><span class="line">plot_df &lt;- cbind(geneid=row.names(plot_df),</span><br><span class="line">                        plot_df)</span><br><span class="line"></span><br><span class="line">plot_df &lt;- plot_df[!is.na(plot_df$padj), ]</span><br><span class="line"></span><br><span class="line">ggplot(data=plot_df, aes(x=log2FoldChange, </span><br><span class="line">                      y =-log10(padj))) +</span><br><span class="line">  geom_point(alpha=0.8, size=1.2)+</span><br><span class="line">  labs(title=&quot;Volcanoplot&quot;, x=&quot;log2 (fold change)&quot;,y=&quot;-log10 (q-value)&quot;)+</span><br><span class="line">  theme(plot.title = element_text(hjust = 0.4))+</span><br><span class="line">  geom_hline(yintercept = -log10(0.05),lty=4,color=&quot;red&quot;)+</span><br><span class="line">  geom_hline(yintercept = -log10(0.01),lty=4,color=&quot;blue&quot;)+</span><br><span class="line">  geom_vline(xintercept = c(1,-1),lty=4,alpha=0.8,color=&quot;blue&quot;)+</span><br><span class="line">  geom_vline(xintercept = c(2,-2),lty=4,alpha=0.8,color=&quot;red&quot;)+</span><br><span class="line">  theme_bw()+</span><br><span class="line">  theme(panel.border = element_blank(),</span><br><span class="line">        panel.grid.major = element_blank(),</span><br><span class="line">        panel.grid.minor = element_blank(),</span><br><span class="line">        axis.line = element_line(colour = &quot;black&quot;))</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/rice-volcano-65deaa78f62b4972a72e0d5129840cdd.png" alt="火山图"></p>
<p>火山图的作用，一是检查你的差异分析是否合理，也就是大部分基因都是没有变化，只有一部分是差异基因，第二个是可以帮你确定阈值。</p>
<p>我发现即便是LogFoldChange &gt; 2（相差4倍）, p.adj &lt; 0.01，依旧是有很多差异基因。所以我这里以这个为标准进行筛选，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">res_sub &lt;- subset(res, abs(log2FoldChange) &gt; 2 &amp; padj &lt;0.01)</span><br><span class="line">gene_up &lt;- row.names(res_sub[res_sub$log2FoldChange &gt; 0, ])</span><br><span class="line">length(gene_up) # 1024</span><br><span class="line">gene_down &lt;- row.names(res_sub[res_sub$log2FoldChange &lt; 0, ])</span><br><span class="line">length(gene_down) # 702</span><br></pre></td></tr></table></figure>

<p>其实这个时候你会有一个问题，所谓的基因上调和下调是谁相对于谁呢？很简单，直接把差异的基因的标准化矩阵提取出来即可。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">gene_counts &lt;- counts(dds, normalized = TRUE)</span><br><span class="line">gene_counts &lt;- gene_counts[row.names(res_sub), ]</span><br><span class="line">head(gene_counts)</span><br></pre></td></tr></table></figure>

<p>当你看到具体的数值时</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">                    CSV_1     CSV_2      CSV_3      CSV_4     CTV_1</span><br><span class="line">LOC_Os01g01430   1.739379   0.00000   5.971535   8.188634  31.70621</span><br><span class="line">                   CTV_2     CTV_3    CTV_4</span><br><span class="line">LOC_Os01g01430  29.60817  17.19996  10.9967</span><br></pre></td></tr></table></figure>

<p>你自然就知道了CTV vs CSV里，CSV是作为分母，而CTV是分子。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">log2 fold change (MLE): condition CTV vs CSV </span><br><span class="line">Wald test p-value: condition CTV vs CSV </span><br><span class="line">DataFrame with 6 rows and 6 columns</span><br><span class="line">                       baseMean    log2FoldChange             lfcSE</span><br><span class="line">                      &lt;numeric&gt;         &lt;numeric&gt;         &lt;numeric&gt;</span><br><span class="line">LOC_Os01g01430 13.1763234817146  2.57839743463268 0.925763093663462</span><br></pre></td></tr></table></figure>


<h3 id="使用clusterProfiler做GO-KEGG富集分析"><a href="#使用clusterProfiler做GO-KEGG富集分析" class="headerlink" title="使用clusterProfiler做GO&#x2F;KEGG富集分析"></a>使用clusterProfiler做GO&#x2F;KEGG富集分析</h3><p>文章用agriGo对DEG进行分类，用singular enrichment analysis (SEA) 做GO富集分析，进一步的注释用Web Gene Ontology Annotation Plot (WEGO) 进行画图。做KEGG富集分析时，用的是KEGG GENES database KAAS (KEGG Automatic Annotation Server) 对差异基因进行功能注释.</p>
<p>这里我也不做评价，我们此处用Y叔的clusterProfiler做GO&#x2F;KEGG富集分析。</p>
<p>在做GO富集分析的时候，我遇到了一个难题，我发现Bioconductor上并没有水稻的物种注释包。解决这个问题有两种方法，</p>
<ul>
<li>一种是利用AnnotationHub在线检索抓取OrgDb, 但是这些包都是用ENTREZID，你需要先将RAP-DB或者MSU转为ENTREZID才行</li>
<li>另一种则是我构建好一个物种注释包，你们只要安装就可以继续使用<code>enrichGO</code>， 参考<a target="_blank" rel="noopener" href="https://github.com/xuzhougeng/org.Osativa.eg.db">https://github.com/xuzhougeng/org.Osativa.eg.db</a></li>
</ul>
<p>参考我的GitHub页面安装，<a target="_blank" rel="noopener" href="https://github.com/xuzhougeng/org.Osativa.eg.db">https://github.com/xuzhougeng/org.Osativa.eg.db</a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">install.packages(&quot;https://github.com/xuzhougeng/org.Osativa.eg.db/releases/download/v0.01/org.Osativa.eg.db.tar.gz&quot;, </span><br><span class="line">    repos = NULL, </span><br><span class="line">    type=&quot;source&quot;)</span><br></pre></td></tr></table></figure>

<blockquote>
<p>目前这个物种包比较简单，就只能做GO富集分析，以及MSU和RAP-DB之间的ID转换</p>
</blockquote>
<p>然后进行GO富集分析</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">library(clusterProfiler)</span><br><span class="line">library(org.Osativa.eg.db)</span><br><span class="line"></span><br><span class="line">org &lt;- org.Osativa.eg.db</span><br><span class="line">ego_up &lt;- enrichGO(gene_up,</span><br><span class="line">                OrgDb = org,</span><br><span class="line">                keyType = &quot;GID&quot;,</span><br><span class="line">                pAdjustMethod = &quot;none&quot;,</span><br><span class="line">                ont=&quot;BP&quot;)</span><br><span class="line">p1 &lt;- dotplot(ego_up)</span><br><span class="line"></span><br><span class="line">ego_down &lt;- enrichGO(gene_down,</span><br><span class="line">                OrgDb = org,</span><br><span class="line">                keyType = &quot;GID&quot;,</span><br><span class="line">                pAdjustMethod = &quot;none&quot;,</span><br><span class="line">                ont=&quot;BP&quot;)</span><br><span class="line">p2 &lt;- dotplot(ego_down)</span><br><span class="line"></span><br><span class="line">library(cowplot)</span><br><span class="line">plot_grid(p1,p2)</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/rice-dotplo-01f5f8e8d85d4d72bf093d13f1cef0e6.png" alt="dotplot"></p>
<p>除了点图，我们还可以用clusterProfiler画下面这个图，展现每个基因在不同GO通路的倍数变化</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">fc &lt;- plot_df$log2FoldChange</span><br><span class="line">names(fc) &lt;-  plot_df$geneid</span><br><span class="line"></span><br><span class="line">heatplot(ego_down, foldChange = fc)</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/riceheatplot-94d1fc4441444f0da0fa52d092e9bc03.png" alt="heatplot"></p>
<blockquote>
<p>讲真，GO富集分析给的信息真的是少，我以为是我的分析出现了问题，所以我去看了文章里GO分析中BP部分，发现信息是一样的少。</p>
</blockquote>
<p>GO富集分析做完之后，KEGG富集分析怎么做呢？参考我之前的这篇<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/UnUPVoaMpfJWCQEpkmdTWA">https://mp.weixin.qq.com/s/UnUPVoaMpfJWCQEpkmdTWA</a></p>
<p>因为KEGG要求的输入编号是RAP-DB，所以我们需要进行一次ID转换，在编号中加上”-01”，将编号中的”g”改成”t”, 就可以做KEGG富集分析了</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">rap_id &lt;- mapIds(x = org, </span><br><span class="line">                 keys = gene_up, </span><br><span class="line">                 column = &quot;RAP&quot;,</span><br><span class="line">                 &quot;GID&quot;)</span><br><span class="line">rap_id &lt;- paste0(rap_id[!is.na(rap_id)], &quot;-01&quot;)</span><br><span class="line">rap_id &lt;- gsub(&quot;g&quot;,&quot;t&quot;,rap_id)</span><br><span class="line">ekegg &lt;- enrichKEGG(rap_id, organism = &quot;dosa&quot;, pAdjustMethod = &quot;none&quot;)</span><br></pre></td></tr></table></figure>

<p>没成想，KEGG的分析结果中只得到了一个结果, 所以也懒得放气泡图了。也难怪文章只是将差异基因按照KEGG进行了分类，而没有做富集。我这里也用Y叔的包对上调差异基因按照KEGG归类</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">rice_kegg &lt;- clusterProfiler::download_KEGG(&quot;dosa&quot;)</span><br><span class="line"></span><br><span class="line">kegg_df &lt;- rice_kegg$KEGGPATHID2EXTID</span><br><span class="line">kegg_df &lt;- kegg_df[kegg_df$to %in% rap_id,]</span><br><span class="line">kegg_df &lt;- merge(kegg_df, rice_kegg$KEGGPATHID2NAME,</span><br><span class="line">                 by.x=&quot;from&quot;,by.y=&quot;from&quot;)</span><br><span class="line"></span><br><span class="line">kegg_class &lt;- as.data.frame(sort(table(kegg_df$to.y), decreasing = T)[1:10])</span><br><span class="line">colnames(kegg_class) &lt;- c(&quot;pathway&quot;,&quot;times&quot;)</span><br><span class="line">ggplot(kegg_class,aes(x=pathway, y = times)) +</span><br><span class="line">  geom_bar(fill=&quot;#ca0020&quot;,stat=&quot;identity&quot;) + coord_flip() +</span><br><span class="line">  theme_bw() + geom_text(aes(y = times+1, label = times))</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/kegg_classfication-538b7daab9234c40a588baa915533b1d.png" alt="KEGG分类1"></p>
<p>当然上下调一起展示也不难，就是代码会长一点</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line">rice_kegg &lt;- clusterProfiler::download_KEGG(&quot;dosa&quot;)</span><br><span class="line"># 获取上调的KEGG PATHWAY</span><br><span class="line">up_rap_id &lt;- mapIds(x = org, </span><br><span class="line">                 keys = gene_up, </span><br><span class="line">                 column = &quot;RAP&quot;,</span><br><span class="line">                 &quot;GID&quot;)</span><br><span class="line">up_rap_id &lt;- paste0(up_rap_id[!is.na(up_rap_id)], &quot;-01&quot;)</span><br><span class="line"></span><br><span class="line">up_rap_id &lt;- gsub(&quot;g&quot;,&quot;t&quot;,up_rap_id)</span><br><span class="line"></span><br><span class="line">up_df &lt;- rice_kegg$KEGGPATHID2EXTID</span><br><span class="line">up_df &lt;- kegg_df[kegg_df$to %in% up_rap_id,]</span><br><span class="line">up_df &lt;- merge(kegg_df, rice_kegg$KEGGPATHID2NAME,</span><br><span class="line">                 by.x=&quot;from&quot;,by.y=&quot;from&quot;)</span><br><span class="line"></span><br><span class="line"># 获取下调的KEGG PATHWAY</span><br><span class="line">down_rap_id &lt;- mapIds(x = org, </span><br><span class="line">                    keys = gene_down, </span><br><span class="line">                    column = &quot;RAP&quot;,</span><br><span class="line">                    &quot;GID&quot;)</span><br><span class="line">down_rap_id &lt;- paste0(down_rap_id[!is.na(down_rap_id)], &quot;-01&quot;)</span><br><span class="line">down_rap_id &lt;- gsub(&quot;g&quot;,&quot;t&quot;,down_rap_id)</span><br><span class="line">down_df &lt;- rice_kegg$KEGGPATHID2EXTID</span><br><span class="line">down_df &lt;- down_df[down_df$to %in% down_rap_id,]</span><br><span class="line">down_df &lt;- merge(down_df, rice_kegg$KEGGPATHID2NAME,</span><br><span class="line">               by.x=&quot;from&quot;,by.y=&quot;from&quot;)</span><br><span class="line"></span><br><span class="line"># 统计</span><br><span class="line">kegg_class_up &lt;- as.data.frame(sort(table(up_df$to.y), </span><br><span class="line">                                 decreasing = T)[1:10])</span><br><span class="line">kegg_class_down &lt;- as.data.frame(sort(table(down_df$to.y), </span><br><span class="line">                                    decreasing = T)[1:10])</span><br><span class="line">#合并</span><br><span class="line">kegg_class &lt;- rbind(kegg_class_up, kegg_class_down)</span><br><span class="line">colnames(kegg_class) &lt;- c(&quot;pathway&quot;,&quot;times&quot;)</span><br><span class="line">kegg_class$source &lt;- rep(c(&quot;up&quot;,&quot;down&quot;),</span><br><span class="line">                         times=c(nrow(kegg_class_up),nrow(kegg_class_down)))</span><br><span class="line">#画图</span><br><span class="line">ggplot(kegg_class,aes(x=pathway, y = times)) +</span><br><span class="line">  geom_bar(aes(fill=source),stat=&quot;identity&quot;,position = &quot;dodge&quot;) + </span><br><span class="line">  scale_fill_manual(values = c(up=&quot;#ca0020&quot;,down=&quot;#2b83ba&quot;)) +</span><br><span class="line">  coord_flip() +</span><br><span class="line">  theme_bw()</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/kegg_classfication-2-99ebb8d178e042debb6dc6cd4f2e706d.png" alt="KEGG分类2"></p>
<p>这里这展示了代码，具体如何讲故事，如何调整参数，其实就靠作者自由发挥了。</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><p>水稻两大权威注释组织</p>
<ul>
<li>RAP-DB: <a target="_blank" rel="noopener" href="https://rapdb.dna.affrc.go.jp/">https://rapdb.dna.affrc.go.jp/</a></li>
<li>RGAP: <a target="_blank" rel="noopener" href="http://rice.plantbiology.msu.edu/">http://rice.plantbiology.msu.edu/</a></li>
<li>clusterPrifiler文档: <a target="_blank" rel="noopener" href="https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html">https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html</a></li>
<li>DESeq2文档: <a target="_blank" rel="noopener" href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/22/3%E5%88%86%E7%9A%84%E6%B0%B4%E7%A8%BB%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%E5%AE%9E%E6%88%98-Rice-RNA-seq-analysis/" data-id="cllg90xyy000aeqp2f1ez8mv6" data-title="3分的水稻转录组分析实战" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何让你的GitHub能被引用-How-to-citate-my-GitHub-repository" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/22/%E5%A6%82%E4%BD%95%E8%AE%A9%E4%BD%A0%E7%9A%84GitHub%E8%83%BD%E8%A2%AB%E5%BC%95%E7%94%A8-How-to-citate-my-GitHub-repository/" class="article-date">
  <time class="dt-published" datetime="2019-08-22T14:01:58.403Z" itemprop="datePublished">2019-08-22</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/22/%E5%A6%82%E4%BD%95%E8%AE%A9%E4%BD%A0%E7%9A%84GitHub%E8%83%BD%E8%A2%AB%E5%BC%95%E7%94%A8-How-to-citate-my-GitHub-repository/">如何让你的GitHub能被引用</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>你需要先登陆到<a target="_blank" rel="noopener" href="https://zenodo.org/">https://zenodo.org/</a>，然后需要创建一个账号，这个账号一定要用GitHub账号进行授权。之后你发现一个Using GitHub， 点击Check out</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-f5d0b1e46c7c41119367e3e9d1cb66a9.png" alt="Step1"></p>
<p>然后你就会发现自己的repository都被列出来了</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-2-ef7546de956a4434ba5bac38d0766673.png" alt="Step2"></p>
<p>之后将你想要被引用的仓库的按钮从off变成ON</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-3-05f1c271e9664247aaaf62682ef323a3.png" alt="Step3"></p>
<p>接着到你想被引用的仓库里, 我们要创建要给release，这样子就有了版本号，如此别人才能知道自己用的版本是哪一版。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-4-56c702f36b334d7c9764a72f35b3132f.png" alt="Step4"></p>
<p>我们就可以创建release，可以设置合适的版本好，比如说我这边设置为0.01</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-5-b8f9f1ed1b0042838b6371df6e48d9e5.png" alt="Step5"></p>
<p>我们就拥有了一个释放版本</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-6-ed976b5018a741b2bcac5a28aae851ea.png" alt="Step6"></p>
<p>之后回到Zenodo，点击Upload</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-7-30755112bfd24f559184141b038b4ef7.png" alt="Step7"></p>
<p>就会出现你刚才释放的软件</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-8-8d759eb87fea4c56b8ccc2e56d25b188.png" alt="Step8"></p>
<p>点击它，你就拥有了可以引用的DOI号</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-9-55fe96b3f5e24d6589f01c9f4a0822df.png" alt="Step9"></p>
<p>参考资料：<a target="_blank" rel="noopener" href="https://guides.github.com/activities/citable-code/">https://guides.github.com/activities/citable-code/</a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/22/%E5%A6%82%E4%BD%95%E8%AE%A9%E4%BD%A0%E7%9A%84GitHub%E8%83%BD%E8%A2%AB%E5%BC%95%E7%94%A8-How-to-citate-my-GitHub-repository/" data-id="cllg90y5i00hkeqp2aonjaqea" data-title="如何让你的GitHub能被引用" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-MECAT2_ 三代高效组装工具-MECAT2-genome-assembly-tools" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/21/MECAT2_%20%E4%B8%89%E4%BB%A3%E9%AB%98%E6%95%88%E7%BB%84%E8%A3%85%E5%B7%A5%E5%85%B7-MECAT2-genome-assembly-tools/" class="article-date">
  <time class="dt-published" datetime="2019-08-21T14:16:16.702Z" itemprop="datePublished">2019-08-21</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/21/MECAT2_%20%E4%B8%89%E4%BB%A3%E9%AB%98%E6%95%88%E7%BB%84%E8%A3%85%E5%B7%A5%E5%85%B7-MECAT2-genome-assembly-tools/">MECAT2:三代高效组装工具</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="MECAT2-三代高效组装工具"><a href="#MECAT2-三代高效组装工具" class="headerlink" title="MECAT2: 三代高效组装工具"></a>MECAT2: 三代高效组装工具</h1><p>MECAT2是三代测序数据PacBio的高效组装工具，是之前<a target="_blank" rel="noopener" href="https://github.com/xiaochuanle/MECAT">MECAT</a>的改进版, 修复了之前的很多bug, 使用基于string graph的<code>fsa</code>替换了之前的<code>mecat2canu</code>。</p>
<p>MECAT2由4个模块组成:</p>
<ul>
<li><code>mecat2pw</code>: SMART reads快速和准确地配对比对工具</li>
<li><code>mecat2ref</code>: SMART reads的参考基因组比对工具</li>
<li><code>mecat2cns</code>: 基于配对的重叠对存在噪音的reads进行纠错</li>
<li><code>fsa</code>: 基于string graph的组装工具</li>
</ul>
<p>目前MECAT2只支持PacBio数据，对于Nanopore数据，肖老师他们开发了<code>NECAT</code></p>
<h2 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h2><p>MECAT2软件安装非常简单，不存在上一代MECAT的安装难问题了</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">mkdir <span class="operator">-</span>p <span class="operator">~</span><span class="operator">/</span>opt<span class="operator">/</span>biosoft</span><br><span class="line">cd <span class="operator">~</span><span class="operator">/</span>opt<span class="operator">/</span>biosoft</span><br><span class="line">git clone https<span class="operator">:</span><span class="operator">/</span><span class="operator">/</span>github.com<span class="operator">/</span>xiaochuanle<span class="operator">/</span>MECAT2.git</span><br><span class="line">cd MECAT2</span><br><span class="line">make <span class="operator">-</span>j <span class="number">4</span></span><br></pre></td></tr></table></figure>

<p>唯一要注意的是尽量要用新版的perl，经测试，Perl 5.26 是可以正常运行脚本。 </p>
<p>最后加入到环境变量<code>PATH</code>中</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">export</span> PATH=~/opt/biosoft/MECAT2/Linux-amd64/bin:<span class="variable">$PATH</span></span><br></pre></td></tr></table></figure>

<h2 id="软件使用"><a href="#软件使用" class="headerlink" title="软件使用"></a>软件使用</h2><p>我们以 <em>E. coli</em> 数据集作为案例，介绍MECAT2应该如何使用。</p>
<p><strong>第一步</strong>，我们要下载数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p ~/ecoli</span><br><span class="line"><span class="built_in">cd</span> ~/ecoli</span><br><span class="line">wget http://gembox.cbcb.umd.edu/mhap/raw/ecoli_filtered.fastq.gz</span><br><span class="line">gunzip ecoli_filtered.fastq.gz</span><br></pre></td></tr></table></figure>

<p><strong>注意</strong>: 目前MECAT2还不支持gz压缩文件，如果你直接用gz作为输入，会提示core dumped。</p>
<p>MECAT2不是根据文件名来确定输入的序列格式, 也就是如果你把 ecoli_filtered.fastq 命名为  ecoli_filtered.fasta, 组装也不会出错。</p>
<p><strong>第二步</strong>, 准备模板的config文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mecat.pl config ecoli_config_file.txt</span><br></pre></td></tr></table></figure>

<p>使用<code>vim</code>修改config文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">PROJECT=ecoli</span><br><span class="line">RAWREADS=ecoli_filtered.fastq</span><br><span class="line">GENOME_SIZE=4800000</span><br><span class="line">THREADS=4</span><br><span class="line">MIN_READ_LENGTH=2000</span><br><span class="line">CNS_OVLP_OPTIONS=<span class="string">&quot;&quot;</span></span><br><span class="line">CNS_OPTIONS=<span class="string">&quot;-r 0.6 -a 1000 -c 4 -l 2000&quot;</span></span><br><span class="line">TRIM_OVLP_OPTIONS=<span class="string">&quot;-B&quot;</span></span><br><span class="line">ASM_OVLP_OPTIONS=<span class="string">&quot;-n 100 -z 10 -b 2000 -e 0.5 -j 1 -u 0 -a 400&quot;</span></span><br><span class="line">FSA_OL_FILTER_OPTIONS=<span class="string">&quot;--max_overhang=-1 --min_identity=-1&quot;</span></span><br><span class="line">FSA_ASSEMBLE_OPTIONS=<span class="string">&quot;&quot;</span></span><br><span class="line">USE_GRID=<span class="literal">false</span></span><br><span class="line">CLEANUP=0</span><br><span class="line">CNS_OUTPUT_COVERAGE=30</span><br><span class="line">GRID_NODE=0</span><br></pre></td></tr></table></figure>

<p><strong>第三步</strong>: 原始数据纠错</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mecat.pl correct ecoli_config_file.txt</span><br></pre></td></tr></table></figure>

<p><strong>第四步</strong>: 对纠错后的reads进行组装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mecat.pl assemble ecoli_config_file.txt</span><br></pre></td></tr></table></figure>

<p>输出结果:</p>
<ul>
<li>纠错后的reads： <code>ecoli/1-consensus/cns_reads.fasta</code>.</li>
<li>最长30X纠错后用于trimming的reads:  <code>ecoli/1-consensus/cns_final.fasta</code>.</li>
<li>trimmed reads:  <code>ecoli/2-trim_bases/trimReads.fasta</code></li>
<li>组装的contigs: <code>ecoli/4-fsa/contigs.fasta</code></li>
</ul>
<p>最终的组装结果为</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Count: 1</span><br><span class="line">Tatal: 4630437</span><br><span class="line">Max: 4630437</span><br><span class="line">Min: 4630437</span><br><span class="line">N25: 4630437</span><br><span class="line">L25: 1</span><br></pre></td></tr></table></figure>

<p>能够完美的就装出一条序列</p>
<h2 id="配置文件介绍"><a href="#配置文件介绍" class="headerlink" title="配置文件介绍"></a>配置文件介绍</h2><p>MECAT2的组装过程的参数都在配置文件中, 也就是之前的<code>ecoli_config_file.txt</code>。我按照不同部分对这些参数进行介绍</p>
<h3 id="基本参数"><a href="#基本参数" class="headerlink" title="基本参数"></a>基本参数</h3><p>这些参数属于改起来不怎么纠结的参数</p>
<ul>
<li><code>PROJECT=ecoli</code>: 项目名，之后的所有输出文件都在该项目名的目录下</li>
<li><code>RAWREADS=</code>: 原始数据的位置，可以是Fasta或者是Fastq， H5格式要先转成Fasta (参考 <a target="_blank" rel="noopener" href="https://github.com/xiaochuanle/MECAT2#S-input-format">Input Format</a>).</li>
<li><code>GENOME_SIZE=</code>: 基因组大小，单位是bp，如果是120M，那么就是120000000.</li>
<li><code>THREADS=</code>: 线程数</li>
<li><code>MIN_READ_LENGTH=</code>: 用于纠错和trim的reads的最低长度. 数据质量好，就长一点，比如说1000到2000</li>
<li><code>USE_GRID=false</code>: 是否有多个计算节点</li>
<li><code>CLEANUP=0</code>: 运行结束后删除MECAT2的中间文件, 大基因组的临时文件很大，所以要设置为1.</li>
<li><code>GRID_NODE=0</code>, 当<code>USE_GRID=1</code>时，设置用到的计算节点数，如果是单节点服务器，不需要设置。</li>
</ul>
<h3 id="纠错和修剪阶段"><a href="#纠错和修剪阶段" class="headerlink" title="纠错和修剪阶段"></a>纠错和修剪阶段</h3><p>纠错阶段(correct stage)和修剪阶段(trimming stage)，MECAT2调用的<code>mecat2pw</code>和<code>mecat2cns</code>， 与之相关的配置如下</p>
<ul>
<li><code>CNS_OVLP_OPTIONS=&quot;&quot;</code>, 在纠错阶段是检测候选重叠的参数, 会传给<code>mecat2pw</code></li>
<li><code>CNS_OPTIONS=&quot;&quot;</code>：原始reads纠错参数，会传递给<code>mecat2cns</code>, </li>
<li><code>TRIM_OVLP_OPTIONS=&quot;&quot;</code>: 在trim阶段，用于检测重叠的参数,会传给<code>v2asmpm</code></li>
</ul>
<p>但是我发现<code>v2asmpm</code>没有文档说明，和软件开发者讨论之后，给出的说明如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">V2asmpm -Pworkpath -Tt -Sx -Ey -B</span><br><span class="line"></span><br><span class="line">-Pworkpath 工作目录是workpath</span><br><span class="line">-Tt 用t个CPU线程</span><br><span class="line">-Sx -Ey 这两个参数一起表示只计算第x到第y卷(包括第y卷), 通常工作目录下会有000001.fasta, 000002.fasta这样的分卷</span><br><span class="line">-B 如果有这个选项, 就输出二进制格式的比对结果; 如果没有这个选项, 就输出文本格式的比对结果</span><br></pre></td></tr></table></figure>
<p>因此用默认的<code>-B</code>就行了。</p>
<h3 id="组装阶段"><a href="#组装阶段" class="headerlink" title="组装阶段"></a>组装阶段</h3><p>组装阶段相关的参数如下</p>
<ul>
<li><code>CNS_OUTPUT_COVERAGE=30</code>: 选择多少覆盖度的最长纠错后reads进行trim和组装。举个例子，4.8M的 <em>E. coli</em> 30X, 是144MB。 一般选择20~40之间就行， 会传递给<code>mecat2elr</code></li>
<li><code>ASM_OVLP_OPTIONS=&quot;&quot;</code>: 在组装阶段，用于检测重叠的参数，传给<code>v2asmpm.sh</code></li>
<li><code>FSA_OL_FILTER_OPTIONS=&quot;&quot;</code>, 过滤重叠的参数，传递给<code>fsa_ol_filter</code></li>
<li><code>FSA_ASSEMBLE_OPTIONS=&quot;&quot;</code>, 组装trimm reads的参数, 传给<code>v2asmpm</code></li>
</ul>
<p>这些参数调整起来都很复杂，只能建议看函数帮助文档了。举个例子<code>FSA_OL_FILTER_OPTIONS=&quot;--max_overhang=-1 --min_identity=-1&quot;</code>根据阅读<code>fsa_ol_filter</code>的帮助发现，<code>-1</code>表示让程序自己决定</p>
<p>而<code>ASM_OVLP_OPTIONS=&quot;-n 100 -z 10 -b 2000 -e 0.5 -j 1 -u 0 -a 400&quot;</code>， 根据和软件开发者的讨论，是<strong>不需要额外调整</strong>，用默认的即可。</p>
<p>MECAT2是基于比对纠错的组装工具，组装速度上比同类型的Canu和Falcon都快。仅看组装结果的N50参数，MECAT2结果和Canu, Falcon结果都是差不多的。</p>
<p>考虑到BUSCO完整度，Canu是三者最高。不过用三代数据和二代数据进行几轮Polish后会提高到Canu相同水平。三个软件上的BUSCO（embryophyta_odb10）值如下:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Canu: C:98.8%[S:96.3%,D:2.5%],F:0.5%,M:0.7%,n:1375</span><br><span class="line">MECAT2: C:93.5%[S:91.6%,D:1.9%],F:4.6%,M:1.9%,n:1375</span><br><span class="line">Falcon: C:96.8%[S:95.1%,D:1.7%],F:2.0%,M:1.2%,n:1375</span><br></pre></td></tr></table></figure>

<p>因此，如果你原本你是用Falcon做组装，那么可以改成MECAT2了，效果差不多，速度还更快了。对于越大的物种，在有限的资源下，更推荐用MECAT2。</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li><a target="_blank" rel="noopener" href="https://github.com/xiaochuanle/MECAT2">https://github.com/xiaochuanle/MECAT2</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/21/MECAT2_%20%E4%B8%89%E4%BB%A3%E9%AB%98%E6%95%88%E7%BB%84%E8%A3%85%E5%B7%A5%E5%85%B7-MECAT2-genome-assembly-tools/" data-id="cllg90xzx002qeqp237jt1w9t" data-title="MECAT2:三代高效组装工具" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-「三代组装」使用新版Falcon进行三代测序基因组组装-Falcon-genome-assembly-tools" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/20/%E3%80%8C%E4%B8%89%E4%BB%A3%E7%BB%84%E8%A3%85%E3%80%8D%E4%BD%BF%E7%94%A8%E6%96%B0%E7%89%88Falcon%E8%BF%9B%E8%A1%8C%E4%B8%89%E4%BB%A3%E6%B5%8B%E5%BA%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-Falcon-genome-assembly-tools/" class="article-date">
  <time class="dt-published" datetime="2019-08-20T15:53:04.991Z" itemprop="datePublished">2019-08-20</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/20/%E3%80%8C%E4%B8%89%E4%BB%A3%E7%BB%84%E8%A3%85%E3%80%8D%E4%BD%BF%E7%94%A8%E6%96%B0%E7%89%88Falcon%E8%BF%9B%E8%A1%8C%E4%B8%89%E4%BB%A3%E6%B5%8B%E5%BA%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-Falcon-genome-assembly-tools/">「三代组装」使用新版Falcon进行三代测序基因组组装</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <blockquote>
<p>这里的新版指的是PacBio公司在2018年9月发布pb-assembly, 而这篇文章是在2018年9月30日发的。</p>
</blockquote>
<p>今年早些时候在参加三代培训时，听说PacBio会在今年对Falcon进行一些改变。前几天我在读 <strong>readthedocs</strong>上的Falcon文档时，发现了文档页面上方出现了这样两栏提醒</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/attentions-d2f8c343963f471f8a2fcd85a64e708e.png" alt="注意"></p>
<p>其中<a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/pb-assembly">pb_assembly</a>就是新的FALCON组装套装在GitHub上的使用文档，经过了几天的探索，我对它有了一点了解，写一篇教程作为官方文档的一些补充吧。</p>
<h2 id="新版亮点"><a href="#新版亮点" class="headerlink" title="新版亮点"></a>新版亮点</h2><ol>
<li>整合了串联重复序列和遍在重复序列的屏蔽（之前没有这一步）</li>
<li>以GFA格式存放graph文件，后续可以用Bandage进行可视化</li>
<li>通过算法和性能优化，提高了Associate Contigs的准确性</li>
<li>分析流程的性能优化</li>
</ol>
<h2 id="软件安装和数据准备"><a href="#软件安装和数据准备" class="headerlink" title="软件安装和数据准备"></a>软件安装和数据准备</h2><p>Falcon终于拥抱了bioconda, 这也就意味着我们再也不需要用到他们原本笨拙的安装脚本，浪费时间在安装软件上。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">conda create -n pb-assembly  pb-assembly</span><br><span class="line"><span class="built_in">source</span> activate pb-assembly</span><br><span class="line"><span class="comment"># 或者</span></span><br><span class="line">conda create -p ~/opt/biosoft/pb-assembly pb-assembly</span><br><span class="line"><span class="built_in">source</span> activate ~/opt/biosoft/pb-assembly </span><br></pre></td></tr></table></figure>

<p>这里使用<a target="_blank" rel="noopener" href="https://pb-falcon.readthedocs.io/en/latest/tutorial.html">https://pb-falcon.readthedocs.io/en/latest/tutorial.html</a>上的所用的E. coli数据集</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wget https://downloads.pacbcloud.com/public/data/git-sym/ecoli.m140913_050931_42139_c100713652400000001823152404301535_s1_p0.subreads.tar.gz</span><br><span class="line">tar -xvzf ecoli.m140913_050931_42139_c100713652400000001823152404301535_s1_p0.subreads.tar.gz</span><br></pre></td></tr></table></figure>

<p>解压缩后的文件夹里有三个300M的Fasta文件, 将他们的实际路径记录到<code>input.fofn</code>中</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ecoli.1.fasta</span><br><span class="line">ecoli.2.fasta</span><br><span class="line">ecoli.3.fasta</span><br></pre></td></tr></table></figure>

<h2 id="准备配置文件"><a href="#准备配置文件" class="headerlink" title="准备配置文件"></a>准备配置文件</h2><p>为了进行组装，需要准备一个配置文件。我的配置文件为<code>fc_run.cfg</code>，内容如下。你们可以先预览一下，后面看我的解释说明。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#### Input</span></span><br><span class="line">[General]</span><br><span class="line">input_fofn=input.fofn</span><br><span class="line">input_type=raw</span><br><span class="line">pa_DBdust_option=</span><br><span class="line">pa_fasta_filter_option=pass</span><br><span class="line">target=assembly</span><br><span class="line">skip_checks=False</span><br><span class="line">LA4Falcon_preload=<span class="literal">false</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#### Data Partitioning</span></span><br><span class="line">pa_DBsplit_option=-x500 -s50</span><br><span class="line">ovlp_DBsplit_option=-x500 -s50</span><br><span class="line"></span><br><span class="line"><span class="comment">#### Repeat Masking</span></span><br><span class="line">pa_HPCTANmask_option=</span><br><span class="line">pa_REPmask_code=1,100;2,80;3,60</span><br><span class="line"></span><br><span class="line"><span class="comment">####Pre-assembly</span></span><br><span class="line">genome_size=0</span><br><span class="line">seed_coverage=20</span><br><span class="line">length_cutoff=3000    </span><br><span class="line">pa_HPCdaligner_option=-v -B128 -M24</span><br><span class="line">pa_daligner_option=-e.7 -l1000 -k18 -h80 -w8 -s100</span><br><span class="line">falcon_sense_option=--output-multi --min-idt 0.70 --min-cov 2 --max-n-read 800</span><br><span class="line">falcon_sense_greedy=False</span><br><span class="line"></span><br><span class="line"><span class="comment">####Pread overlapping</span></span><br><span class="line">ovlp_daligner_option=-e.96 -l2000 -k24 -h1024 -w6 -s100</span><br><span class="line">ovlp_HPCdaligner_option=-v -B128 -M24</span><br><span class="line"></span><br><span class="line"><span class="comment">####Final Assembly</span></span><br><span class="line">overlap_filtering_setting=--max-diff 100 --max-cov 300 --min-cov 2</span><br><span class="line">fc_ovlp_to_graph_option=</span><br><span class="line">length_cutoff_pr=2000</span><br><span class="line"></span><br><span class="line">[job.defaults]</span><br><span class="line">job_type=<span class="built_in">local</span></span><br><span class="line">pwatcher_type=blocking</span><br><span class="line">JOB_QUEUE=default</span><br><span class="line">MB=32768</span><br><span class="line">NPROC=6</span><br><span class="line">njobs=32</span><br><span class="line">submit = /bin/bash -c <span class="string">&quot;<span class="variable">$&#123;JOB_SCRIPT&#125;</span>&quot;</span> &gt; <span class="string">&quot;<span class="variable">$&#123;JOB_STDOUT&#125;</span>&quot;</span> 2&gt; <span class="string">&quot;<span class="variable">$&#123;JOB_STDERR&#125;</span>&quot;</span></span><br><span class="line"></span><br><span class="line">[job.step.da]</span><br><span class="line">NPROC=4</span><br><span class="line">MB=32768</span><br><span class="line">njobs=32</span><br><span class="line">[job.step.la]</span><br><span class="line">NPROC=4</span><br><span class="line">MB=32768</span><br><span class="line">njobs=32</span><br><span class="line">[job.step.cns]</span><br><span class="line">NPROC=8</span><br><span class="line">MB=65536</span><br><span class="line">njobs=5</span><br><span class="line">[job.step.pla]</span><br><span class="line">NPROC=4</span><br><span class="line">MB=32768</span><br><span class="line">njobs=4</span><br><span class="line">[job.step.asm]</span><br><span class="line">NPROC=24</span><br><span class="line">MB=196608</span><br><span class="line">njobs=1</span><br></pre></td></tr></table></figure>

<p>根据注释信息，文件分为”input”, “Data partitioning”, “Repeat Masking”, “Pre-assembly”, “Pread overlapping”, “Final Assembly”, 以及最后的任务调度部分，让我们分别看下这里面的内容</p>
<h3 id="输入-Input"><a href="#输入-Input" class="headerlink" title="输入(Input)"></a>输入(Input)</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#### Input</span></span><br><span class="line">[General]</span><br><span class="line">input_fofn=input.fofn</span><br><span class="line">input_type=raw</span><br><span class="line">pa_DBdust_option=</span><br><span class="line">pa_fasta_filter_option=pass</span><br><span class="line">target=assembly</span><br><span class="line">skip_checks=False</span><br><span class="line">LA4Falcon_preload=<span class="literal">false</span></span><br></pre></td></tr></table></figure>

<p>输入这里参数比较简单，基本不需要做任何改动，除了 pa_fasta_filter_options，用于处理一个ZMW（测序翻译孔）有多条subread时，到底选择哪一条的问题。</p>
<ul>
<li>“pass”: 不做过滤，全部要。</li>
<li>“streamed-median”: 表示选择大于中位数的subread</li>
<li>“streamed-internal-median”: 当一个ZMW里的subread低于3条时选择最长，多于单条则选择大于中位数的subread</li>
</ul>
<blockquote>
<p>0.01版本pb-assembly的<code>pa_DBdust_option</code>有一个bug，也就是里面的参数不会传递给DBdust, DBdust是对read进行soft-masking，一般都用默认参数，因此这个bug问题不大。</p>
</blockquote>
<h3 id="数据分配-Data-Partitioning"><a href="#数据分配-Data-Partitioning" class="headerlink" title="数据分配(Data Partitioning)"></a>数据分配(Data Partitioning)</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pa_DBsplit_option=-x500 -s50</span><br><span class="line">ovlp_DBsplit_option=-x500 -s50</span><br></pre></td></tr></table></figure>

<p>这部分的设置会将参数传递给<code>DBsplit</code>，将数据进行拆分多个block，后续的并行计算都基于block。<code>-s 50</code>表示每个block大小为50M。 这适用于基因组比较小的物种，如果是大基因组则应该设置为<code>-s 200</code>或者<code>-s 400</code></p>
<h3 id="重复屏蔽-Repeat-Masking"><a href="#重复屏蔽-Repeat-Masking" class="headerlink" title="重复屏蔽(Repeat Masking)"></a>重复屏蔽(Repeat Masking)</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#### Repeat Masking</span></span><br><span class="line">pa_HPCTANmask_option=</span><br><span class="line">pa_REPmask_code=1,100;2,80;3,60</span><br></pre></td></tr></table></figure>

<p>屏蔽重复序列可以在不损失组装准确性的同时，提高后续组装的overlap&#x2F;daligner步骤10~20倍速度，见<a target="_blank" rel="noopener" href="https://dazzlerblog.wordpress.com/2016/04/01/detecting-and-soft-masking-repeats/">Detecting and Masking Repeats</a>.</p>
<p>pa_HPCTANmask_option的参数会传给串联重复步骤的<code>HPCTANmask</code>, 而<code>pa_REPmask_code</code>很复杂，它分为三次迭代，因此这里1:100;2,80;3,60 就表示第一次迭代检测每个block中出现超过100次的序列，第二次迭代将2个block合并一起检测超过80次的序列，第三次将3个block进行合并检测超过60次的序列。</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/pbbioconda/issues/20">https://github.com/PacificBiosciences/pbbioconda/issues/20</a></p>
</blockquote>
<h3 id="预组装（纠错）pre-assembly"><a href="#预组装（纠错）pre-assembly" class="headerlink" title="预组装（纠错）pre-assembly"></a>预组装（纠错）pre-assembly</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">####Pre-assembly</span></span><br><span class="line">genome_size=0</span><br><span class="line">seed_coverage=20</span><br><span class="line">length_cutoff=3000    </span><br><span class="line">pa_HPCdaligner_option=-v -B128 -M24</span><br><span class="line">pa_daligner_option=-e.7 -l1000 -k18 -h80 -w8 -s100</span><br><span class="line">falcon_sense_option=--output-multi --min-idt 0.70 --min-cov 2 --max-n-read 800</span><br><span class="line">falcon_sense_greedy=False</span><br></pre></td></tr></table></figure>

<p>当<code>length_cutoff=-1</code>时，设置<code>genome_size</code>和<code>seed_coverage</code>会自动计算要过滤的序列。否则是过滤低于一定长度的read。</p>
<p>pa_HPCdalinger_option参数不需要调整，-M表示每个进程的内存为24G，一般200M的block对应16G。</p>
<p>pa_daligner_option的参数比较重要：</p>
<ul>
<li><code>-e</code>:错误率，低质量序列设置为0.70，高质量设置为0.80。 值越高避免单倍型的坍缩</li>
<li><code>-l</code>:  最低overlap的长度，文库比较短时为1000， 文库比较长为5000.</li>
<li><code>-k</code>:  低质量数据为14，高质量数据为18</li>
<li><code>-h</code>: 表示完全match的k-mer所覆盖的碱基数。和<code>-l</code>, <code>-e</code>有关，越大越严格。预组装时最大也不要超过最低overlap长度的1&#x2F;4. 最低就设置为80</li>
</ul>
<h3 id="纠错后相互比对"><a href="#纠错后相互比对" class="headerlink" title="纠错后相互比对"></a>纠错后相互比对</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">####Pread overlapping</span></span><br><span class="line">ovlp_daligner_option=-e.96 -l2000 -k24 -h1024 -w6 -s100</span><br><span class="line">ovlp_HPCdaligner_option=-v -B128 -M24</span><br></pre></td></tr></table></figure>

<p>和上面的参数类似，但是-e的范围调整为0.93-0.96，-l范围调整为1800-6000， -k调整为 18-24</p>
<h3 id="最后组装"><a href="#最后组装" class="headerlink" title="最后组装"></a>最后组装</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">overlap_filtering_setting=--max-diff 100 --max-cov 300 --min-cov 2</span><br><span class="line">fc_ovlp_to_graph_option=</span><br><span class="line">length_cutoff_pr=2000</span><br></pre></td></tr></table></figure>

<p>这里的参数就可以随便调整了，因为这一步速度很快。 例如<code>length_cutoff_pr</code>就可以从2000，提高到15000.</p>
<p>最后还有一部分是任务投递系统，如果是单节点运行，需要注意设置 njobs，这是同时投递的任务数。假如你将[job.step.cns]按照如下的方式设置，那么同时会出现 $ 8 \times 50 &#x3D; 400 $ 个任务，如果你的内存只有128G，运行一段时间后你的所有内存就会被耗尽，那么基本上你就只能重启服务器了。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">[job.step.cns]</span><br><span class="line">NPROC=8</span><br><span class="line">MB=65536</span><br><span class="line">njobs=50</span><br></pre></td></tr></table></figure>

<h2 id="运行结果"><a href="#运行结果" class="headerlink" title="运行结果"></a>运行结果</h2><p>用上述的配置文件，以<code>fc_run fc_run.cfg</code>运行后，最后的<code>2-asm-falcon/p_ctg.fa</code>的序列数有4条，最长为4,685,024, 之后我将<code>length_cutoff_pr</code>调整为15k，<code>2-asm-falcon/p_ctg.fa</code>序列只有一条，长度为<code>4,638,411</code></p>
<p>下载<code>asm.gfa</code>到本地，用<a target="_blank" rel="noopener" href="https://rrwick.github.io/Bandage/">Bandage</a>可视化，可以发现组装效果不错。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/bandage-2c760dfa117940389cedc639a8aa9906.png" alt="Bandage"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/20/%E3%80%8C%E4%B8%89%E4%BB%A3%E7%BB%84%E8%A3%85%E3%80%8D%E4%BD%BF%E7%94%A8%E6%96%B0%E7%89%88Falcon%E8%BF%9B%E8%A1%8C%E4%B8%89%E4%BB%A3%E6%B5%8B%E5%BA%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-Falcon-genome-assembly-tools/" data-id="cllg90y2s007ieqp2e3yqh35v" data-title="「三代组装」使用新版Falcon进行三代测序基因组组装" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/21/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/20/">20</a><a class="page-number" href="/page/21/">21</a><span class="page-number current">22</span><a class="page-number" href="/page/23/">23</a><a class="page-number" href="/page/24/">24</a><a class="extend next" rel="next" href="/page/23/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
          <li>
            <a href="/2023/06/22/%E5%9C%A8Windows%E4%B8%AD%E6%90%AD%E5%BB%BA%E4%B8%80%E4%B8%AA%E5%9F%BA%E4%BA%8EWSL%E7%9A%84Python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6%E7%8E%AF%E5%A2%83-building-a-wsl-based-python-data-science-environment-in-windows/">在Windows中搭建一个基于WSL的Python数据科学环境</a>
          </li>
        
          <li>
            <a href="/2023/06/16/%E7%94%A8Colab%E7%BC%93%E8%A7%A3%E6%88%91%E7%9A%84GPU%E7%84%A6%E8%99%91-%E7%94%A8colab%E7%BC%93%E8%A7%A3%E6%88%91%E7%9A%84gpu%E7%84%A6%E8%99%91/">用Colab缓解我的GPU焦虑</a>
          </li>
        
          <li>
            <a href="/2023/05/31/%E6%8E%8C%E6%8F%A1%E4%BA%86%E4%B8%80%E4%B8%AA%E5%85%B3%E4%BA%8EBcrypt%E5%8A%A0%E5%AF%86%E7%9A%84%E7%9F%A5%E8%AF%86-acquired-a-knowledge-of-bcrypt-encryption/">掌握了一个关于Bcrypt加密的知识</a>
          </li>
        
          <li>
            <a href="/2023/05/28/%E6%88%91%E4%B8%8D%E7%9F%A5%E9%81%93%E8%AF%B4%E4%BA%9B%E4%BB%80%E4%B9%88-what-to-say/">我不知道说些什么</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>