<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/8/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-「C++名词解析」指针、常量指针、指向常量的指针、指向常量的常量指针-c名词解析指针常量指针指向常量的指针指向常量的常量指针" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/05/29/%E3%80%8CC++%E5%90%8D%E8%AF%8D%E8%A7%A3%E6%9E%90%E3%80%8D%E6%8C%87%E9%92%88%E3%80%81%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88%E3%80%81%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E6%8C%87%E9%92%88%E3%80%81%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88-c%E5%90%8D%E8%AF%8D%E8%A7%A3%E6%9E%90%E6%8C%87%E9%92%88%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E6%8C%87%E9%92%88%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88/" class="article-date">
  <time class="dt-published" datetime="2021-05-29T14:49:32.764Z" itemprop="datePublished">2021-05-29</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/05/29/%E3%80%8CC++%E5%90%8D%E8%AF%8D%E8%A7%A3%E6%9E%90%E3%80%8D%E6%8C%87%E9%92%88%E3%80%81%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88%E3%80%81%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E6%8C%87%E9%92%88%E3%80%81%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88-c%E5%90%8D%E8%AF%8D%E8%A7%A3%E6%9E%90%E6%8C%87%E9%92%88%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E6%8C%87%E9%92%88%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88/">「C++名词解析」指针、常量指针、指向常量的指针、指向常量的常量指针</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>指针(pointer), 一个有点特殊的变量，该变量记录着它所指向的对象的内存地址。</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="type">int</span> a = <span class="number">1</span>;</span><br><span class="line"><span class="type">int</span> *ptr = &amp;a;</span><br></pre></td></tr></table></figure>

<p>常量指针(const pointer), 指针本质上是个变量，因此可以用 <code>const</code> 进行修饰，表示该指针记录的值是一个常量。常量指针在声明时需要初始化值。</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="type">int</span> a=<span class="number">1</span>;</span><br><span class="line"><span class="type">int</span> *<span class="type">const</span> ptr = &amp;a;</span><br></pre></td></tr></table></figure>

<p>指向常量的指针(pointer refer to const)</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"><span class="type">int</span> <span class="type">const</span> a = <span class="number">1</span>;</span><br><span class="line"><span class="type">int</span> cont *ptr = &amp;a;</span><br></pre></td></tr></table></figure>

<p>指向常量的常量指针</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"><span class="type">int</span> <span class="type">const</span> *<span class="type">const</span> ptr </span><br><span class="line"></span><br></pre></td></tr></table></figure>
      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/05/29/%E3%80%8CC++%E5%90%8D%E8%AF%8D%E8%A7%A3%E6%9E%90%E3%80%8D%E6%8C%87%E9%92%88%E3%80%81%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88%E3%80%81%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E6%8C%87%E9%92%88%E3%80%81%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88-c%E5%90%8D%E8%AF%8D%E8%A7%A3%E6%9E%90%E6%8C%87%E9%92%88%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E6%8C%87%E9%92%88%E6%8C%87%E5%90%91%E5%B8%B8%E9%87%8F%E7%9A%84%E5%B8%B8%E9%87%8F%E6%8C%87%E9%92%88/" data-id="clm1xkn9f00e6dpnib8hbbmoc" data-title="「C++名词解析」指针、常量指针、指向常量的指针、指向常量的常量指针" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-如何在Visual Studio上配置OpenCV 4.5.2用于C++项目开发-visual-studio-configuration-of-opencv-for-cpp-development" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/05/26/%E5%A6%82%E4%BD%95%E5%9C%A8Visual%20Studio%E4%B8%8A%E9%85%8D%E7%BD%AEOpenCV%204.5.2%E7%94%A8%E4%BA%8EC++%E9%A1%B9%E7%9B%AE%E5%BC%80%E5%8F%91-visual-studio-configuration-of-opencv-for-cpp-development/" class="article-date">
  <time class="dt-published" datetime="2021-05-26T23:16:07.675Z" itemprop="datePublished">2021-05-26</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/05/26/%E5%A6%82%E4%BD%95%E5%9C%A8Visual%20Studio%E4%B8%8A%E9%85%8D%E7%BD%AEOpenCV%204.5.2%E7%94%A8%E4%BA%8EC++%E9%A1%B9%E7%9B%AE%E5%BC%80%E5%8F%91-visual-studio-configuration-of-opencv-for-cpp-development/">如何在Visual Studio上配置OpenCV 4.5.2用于C++项目开发</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>这是我第一次在号称宇宙最强的IDE Visual Studio上调用非标准库编译<code>C++</code>程序，中间遇到了很多的报错，让我怀疑人生。 </p>
<p>首先，我们需要从<a target="_blank" rel="noopener" href="http://sourceforge.net/projects/opencvlibrary/files/opencv-win/">Sourceforge</a>下载OpenCV, 目前最新版是4.5.2</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-51a11bc996ed4923ba22cb81eb4891fa.png" alt="image.png"></p>
<p>下载的exe文件双击之后会出现解压界面</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-db97fe58403f431081103684dc9af996.png" alt="解压中"></p>
<p>解压缩后得到如下文件</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-0ed0562fe5fa410c90109dfa60ffaad6.png" alt="解压内容"></p>
<p>我们需要依次系统属性-&gt;高级-&gt;环境变量，找到并设置设置环境变量Path</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-f9cec893255d4403a6cfe7295b6d1c9e.png" alt="环境变量"></p>
<p>如果不在环境变量中设置opencv中的bin路径，会出现如下报错 【由于找不到opencv_world452d.dll, 无法继续执行代码】</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-d79dd7544b154cc289b8c35bbfcd2387.png" alt="error1"></p>
<p>在Visual Studio 2019中使用OpenCV构建项目的流程如下</p>
<p>第一步：新建一个C++的控制台项目。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-654548fdb7414748b468f464b1d3a27a.png" alt="新建项目"></p>
<p>第二步：配置项目的项目名称，例如opencv-test</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-ad4394ad531d4675bd3d08c0fe732dff.png" alt="image.png"></p>
<p>创建项目之后，先将Debug 改为x64，因为后续构建的是x64的项目。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-5234740fcd2d469caf3f511a4e14d6fc.png" alt="修改Debug"></p>
<p>第三步：为了能让代码正常运行，我们需要配置这个OpenCV项目的所需的include和library路径，以及依赖的lib文件。</p>
<p>在菜单栏的项目(P)中选择属性(P)</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-151bb1e5a6d84d9b9a489adf37b5e54b.png" alt="项目属性"></p>
<p>选择VC++目录(VC++ Directories)的包含目录(Include Directories)，点击编辑</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-135ed28733ac450093f5aa87bd5d22b0.png" alt="image.png"></p>
<p>添加opencv的include的路径(头文件）</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-0ac77d7dd74f43e4bcb8980d9555c3c6.png" alt="include"></p>
<p>同样的操作，也用于库目录(Library Directories)的设置</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-d83597426001403087700e7263bafe27.png" alt="library"></p>
<p>添加library路径（路径比较深）</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-1e3da7a7e9c243bca11ae07bc87d326f.png" alt="library路径"></p>
<p>检查下，刚刚修改的include和library</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-d7c389f081cc4ab6b68ef0ff0fc9c24b.png" alt="检查状态"></p>
<p>此外还需要增加一个包括所有模块的lib文件。选择链接器(linker)中的输入(input), 接着编辑附加依赖项</p>
<p>![linker(<a target="_blank" rel="noopener" href="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-75a058ac43214ee9bfb9fb18f884df75.png">https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-75a058ac43214ee9bfb9fb18f884df75.png</a>)</p>
<p>添加<strong>opencv_world452d.lib</strong>，该文件包括opencv里的所有模块（lib文件和opencv版本有关，可从lib目录中确认）。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-0ab452f73b074436bea8273df0612707.png" alt="符加依赖库"></p>
<p>如果不设置该变量，会出现【无法解析外部符号】的报错</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/05/image-2796e7747e1a4d09b3cb2d520973b6cb.png" alt="error2"></p>
<p>如果写错成 opencv_world452d.dll，就会出现【 LNK1104  无法打开文件”opencv_world452d.dll”】报错（之所以写成dll，是因为之前出现的dll找不到报错，导致我以为增加的是这个文件）。</p>
<p>如下是测试代码</p>
<figure class="highlight c++"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;opencv2/core/core.hpp&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;opencv2/highgui/highgui.hpp&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;opencv2/imgproc.hpp&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;iostream&gt;</span></span></span><br><span class="line"></span><br><span class="line"><span class="keyword">using</span> <span class="keyword">namespace</span> cv;</span><br><span class="line"><span class="keyword">using</span> <span class="keyword">namespace</span> std;</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="type">int</span> <span class="title">main</span><span class="params">()</span></span></span><br><span class="line"><span class="function"></span>&#123;</span><br><span class="line">  Mat image = Mat::<span class="built_in">zeros</span>(<span class="number">600</span>, <span class="number">600</span>, CV_8UC3);</span><br><span class="line">  <span class="built_in">putText</span>(image, <span class="string">&quot;xuzhougeng&quot;</span>, <span class="built_in">Point</span>(<span class="number">100</span>,<span class="number">300</span>), FONT_HERSHEY_COMPLEX, <span class="number">2.0</span>, <span class="built_in">Scalar</span>(<span class="number">100</span>,<span class="number">200</span>,<span class="number">200</span> ), <span class="number">5</span> );</span><br><span class="line">  <span class="built_in">imshow</span>(<span class="string">&quot;Display Window&quot;</span>, image);</span><br><span class="line">  <span class="built_in">waitKey</span>(<span class="number">0</span>);</span><br><span class="line">  <span class="keyword">return</span> <span class="number">0</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>能正常出图表示能正确在项目中调用OpenCV。</p>
<p>参考资料：</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://docs.opencv.org/master/d3/d52/tutorial_windows_install.html">https://docs.opencv.org/master/d3/d52/tutorial_windows_install.html</a></li>
<li><a target="_blank" rel="noopener" href="https://medium.com/@subwaymatch/opencv-410-with-vs-2019-3d0bc0c81d96">https:&#x2F;&#x2F;medium.com&#x2F;@subwaymatch&#x2F;opencv-410-with-vs-2019-3d0bc0c81d96</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/05/26/%E5%A6%82%E4%BD%95%E5%9C%A8Visual%20Studio%E4%B8%8A%E9%85%8D%E7%BD%AEOpenCV%204.5.2%E7%94%A8%E4%BA%8EC++%E9%A1%B9%E7%9B%AE%E5%BC%80%E5%8F%91-visual-studio-configuration-of-opencv-for-cpp-development/" data-id="clm1xkn6q006gdpni9iznb25x" data-title="如何在Visual Studio上配置OpenCV 4.5.2用于C++项目开发" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-Centos的Slurm安装笔记-install-slurm-in-centos" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/04/21/Centos%E7%9A%84Slurm%E5%AE%89%E8%A3%85%E7%AC%94%E8%AE%B0-install-slurm-in-centos/" class="article-date">
  <time class="dt-published" datetime="2021-04-21T11:49:05.177Z" itemprop="datePublished">2021-04-21</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/04/21/Centos%E7%9A%84Slurm%E5%AE%89%E8%A3%85%E7%AC%94%E8%AE%B0-install-slurm-in-centos/">Centos的Slurm安装笔记</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <blockquote>
<p>因为有一些软件必须要用Slurm，所以不得不在我的主机上配置slurm。</p>
</blockquote>
<p>Slurm的安装依赖于root权限</p>
<h3 id="munge配置"><a href="#munge配置" class="headerlink" title="munge配置"></a>munge配置</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">wget https://github.com/dun/munge/releases/download/munge-0.5.14/munge-0.5.14.tar.xz</span><br><span class="line">rpmbuild -tb --without verify munge-0.5.14.tar.xz</span><br><span class="line"><span class="built_in">cd</span> /root/rpmbuild/RPMS/x86_64</span><br><span class="line">rpm -ivh munge-0.5.14-1.el7.x86_64.rpm \</span><br><span class="line">    munge-devel-0.5.14-1.el7.x86_64.rpm munge-libs-0.5.14-1.el7.x86_64.rpm</span><br></pre></td></tr></table></figure>

<p>创建密钥</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sudo -u munge /usr/sbin/mungekey -v</span><br><span class="line"><span class="comment"># mungekey: Info: Created &quot;/etc/munge/munge.key&quot; with 1024-bit key </span></span><br></pre></td></tr></table></figure>

<p>生成的 munge.key 文件需要分发到所有的计算节点。</p>
<p>启动守护进程(daemon)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">systemctl <span class="built_in">enable</span> munge</span><br><span class="line">systemctl start  munge</span><br><span class="line"><span class="comment"># 检查状态</span></span><br><span class="line">systemctl status munge </span><br></pre></td></tr></table></figure>

<h3 id="方法1-RPM安装"><a href="#方法1-RPM安装" class="headerlink" title="方法1: RPM安装"></a>方法1: RPM安装</h3><p>下载页面， <a target="_blank" rel="noopener" href="https://www.schedmd.com/downloads.php">https://www.schedmd.com/downloads.php</a></p>
<p>因为是CentOS7, 因此我下载的是19.05版本。 而20.11可能不再支持Python2。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget https://download.schedmd.com/slurm/slurm-19.05.8.tar.bz2</span><br><span class="line">yum install pam-devel perl-Switch -y</span><br><span class="line">rpmbuild -ta slurm-19.05.8.tar.bz2</span><br><span class="line"><span class="built_in">cd</span> /root/rpmbuild/RPMS/x86_64</span><br><span class="line">rpm --install slurm-*.rpm</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>创建用户 slurm</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">adduser slurm</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>创建配置文件（非常关键）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p /etc/slurm</span><br><span class="line"><span class="built_in">touch</span> /etc/slurm/slurm.conf </span><br></pre></td></tr></table></figure>

<p>etc中slurm.conf文件里面的配置信息来自于<a target="_blank" rel="noopener" href="https://slurm.schedmd.com/configurator.html">https://slurm.schedmd.com/configurator.html</a> 生成，需要配置如下选项</p>
<ul>
<li><p>SlurmctldHost: 信息来自于 <code>hostname -f </code></p>
</li>
<li><p>NodeName: 信息来自于 <code>hostname -f</code>, 只不过是子节点的服务器信息，如果只有单个主机，那么同上</p>
</li>
<li><p>ComputeNodeAddress: 计算节点的IP地址，仅有单个节点时，信息为空</p>
</li>
<li><p>PartitionName: 任务分配名，改成batch</p>
</li>
<li><p>CPUs: 设置为空</p>
</li>
<li><p>CoresPerSocket: 实际的物理CPU数，例如96</p>
</li>
<li><p>ThreadsPerCore: 如果超线程，设置为2</p>
</li>
<li><p>RealMemory: 服务器内存大小，单位为Mb</p>
</li>
<li><p>SlurmUser: slurm要求有一个专门的用户，</p>
</li>
<li><p>StateSaveLocation: 一定要改成 &#x2F;var&#x2F;spool&#x2F;slurmd， 否则会出现权限问题</p>
</li>
</ul>
<p>最后还需要增加一行 <code>CgroupMountpoint=/sys/fs/cgroup</code></p>
<p>启动 slurmctld, slurmd 的守护进程(deamon)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 控制节点</span></span><br><span class="line">systemctl <span class="built_in">enable</span> slurmctld</span><br><span class="line">systemctl start slurmctld</span><br><span class="line">systemctl status slurmctld </span><br><span class="line"><span class="comment"># 计算节点 </span></span><br><span class="line">systemctl <span class="built_in">enable</span> slurmd</span><br><span class="line">systemctl start slurmd</span><br><span class="line">systemctl status slurmd  </span><br></pre></td></tr></table></figure>

<h3 id="方法2-通过OpenHPC仓库"><a href="#方法2-通过OpenHPC仓库" class="headerlink" title="方法2: 通过OpenHPC仓库"></a>方法2: 通过OpenHPC仓库</h3><h3 id="测试安装"><a href="#测试安装" class="headerlink" title="测试安装"></a>测试安装</h3><p>安装结果后，我们创建一个 test.sbatch, 信息如下，用于测试</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#!/bin/bash</span></span><br><span class="line"><span class="comment">#SBATCH -J test # Job name</span></span><br><span class="line"><span class="comment">#SBATCH -o job.%j.out # Name of stdout output file (%j expands to %jobId)</span></span><br><span class="line"><span class="comment">#SBATCH -N 1 # Total number of nodes requested</span></span><br><span class="line"><span class="comment">#SBATCH -n 2 # Total number of mpi tasks #requested</span></span><br><span class="line"><span class="comment">#SBATCH -t 01:30:00 # Run time (hh:mm:ss) - 1.5 hours</span></span><br><span class="line"><span class="comment"># Launch MPI-based executable</span></span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;Test output from Slurm Testjob&quot;</span></span><br><span class="line">NODEFILE=`generate_pbs_nodefile`</span><br><span class="line"><span class="built_in">cat</span> <span class="variable">$NODEFILE</span></span><br><span class="line"><span class="built_in">sleep</span> 20</span><br></pre></td></tr></table></figure>

<p>递交任务</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sbatch ./test.sbatch </span><br><span class="line"><span class="comment"># Submitted batch job 2</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>查看状态</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">squeue</span><br></pre></td></tr></table></figure>

<p>如果能输出一个job.X.out 文件，说明我们的SLURM已经配置成功。</p>
<h2 id="可能报错和解决方案"><a href="#可能报错和解决方案" class="headerlink" title="可能报错和解决方案"></a>可能报错和解决方案</h2><p>使用 <code>rpm --install</code>的时候可能会遇到如下的报错。这表示你需要安装perl的Switch模块</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">error: Failed dependencies:</span><br><span class="line">  perl(Switch) is needed by slurm-openlava-19.05.8-1.el7.x86_64</span><br><span class="line">  perl(Switch) is needed by slurm-torque-19.05.8-1.el7.x86_64</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>启动 slurmd的deamon失败</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># systemctl start slurmd</span></span><br><span class="line">Job <span class="keyword">for</span> slurmd.service failed because the control process exited with error code. </span><br><span class="line">See <span class="string">&quot;systemctl status slurmd.service&quot;</span> and <span class="string">&quot;journalctl -xe&quot;</span> <span class="keyword">for</span> details.</span><br></pre></td></tr></table></figure>

<p>按照提示运行 <code>systemctl status slurmd.service</code> 发现error信息如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">error: Node configuration differs from hardware: Procs=1:192(hw) Boards=1:1(hw) SocketsPerBoard=1:4(hw) ...e=1:2(hw)</span><br><span class="line">error: cgroup namespace <span class="string">&#x27;freezer&#x27;</span> not mounted. aborting</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>第一个error原因是在<a target="_blank" rel="noopener" href="https://slurm.schedmd.com/configurator.html">https://slurm.schedmd.com/configurator.html</a> 填写 “Compute Machines” 的硬件信息出现错误</p>
<p>第二个error原因是配置文件的默认配置表现不佳，需要做如下替换</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">echo</span> CgroupMountpoint=/sys/fs/cgroup &gt;&gt; /etc/slurm/cgroup.conf</span><br></pre></td></tr></table></figure>

<blockquote>
<p>参考: <a target="_blank" rel="noopener" href="https://stackoverflow.com/questions/62641323/error-cgroup-namespace-freezer-not-mounted-aborting">https://stackoverflow.com/questions/62641323/error-cgroup-namespace-freezer-not-mounted-aborting</a></p>
</blockquote>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><p>配置slurm: <a target="_blank" rel="noopener" href="https://slurm.schedmd.com/configurator.html">https://slurm.schedmd.com/configurator.html</a></p>
<p>单节点slurm: <a target="_blank" rel="noopener" href="http://docs.nanomatch.de/technical/SimStackRequirements/SingleNodeSlurm.html">http://docs.nanomatch.de/technical/SimStackRequirements/SingleNodeSlurm.html</a></p>
<p>munge配置:<a target="_blank" rel="noopener" href="https://github.com/dun/munge/wiki/Installation-Guide">https://github.com/dun/munge/wiki/Installation-Guide</a></p>
<p>Slurm安装与使用: <a target="_blank" rel="noopener" href="http://wiki.casjc.com/?p=378">http://wiki.casjc.com/?p=378</a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/04/21/Centos%E7%9A%84Slurm%E5%AE%89%E8%A3%85%E7%AC%94%E8%AE%B0-install-slurm-in-centos/" data-id="clm1xknag00hwdpni8vlfebyc" data-title="Centos的Slurm安装笔记" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用Biopython调用EMBOSS的needle-biopython-emboss-needle" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/04/14/%E4%BD%BF%E7%94%A8Biopython%E8%B0%83%E7%94%A8EMBOSS%E7%9A%84needle-biopython-emboss-needle/" class="article-date">
  <time class="dt-published" datetime="2021-04-14T20:32:13.198Z" itemprop="datePublished">2021-04-14</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/04/14/%E4%BD%BF%E7%94%A8Biopython%E8%B0%83%E7%94%A8EMBOSS%E7%9A%84needle-biopython-emboss-needle/">使用Biopython调用EMBOSS的needle</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>最近有一个需求，计算250条序列两两之间的相似度，之后根据相似度继续进行序列过滤，即A和B如果相似度高于一个阈值，那么就把B去掉。</p>
<p>我最初想到的就是在Python里面调用EMBOSS的needle，准备使用Python的os或者subprocess模块，但是突然想到之前看WGDI源代码时，发现作者时通过biopython的API进行MAFFT的调用，于是就决定改用Biopython.</p>
<p>通过查找文档，我找到了Biopython对这部分内容的介绍，见 <a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec97">http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec97</a>,示例代码如下</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> Bio.Emboss.Applications <span class="keyword">import</span> NeedleCommandline</span><br><span class="line">needle_cline = NeedleCommandline(asequence=<span class="string">&quot;alpha.faa&quot;</span>, </span><br><span class="line">                                 bsequence=<span class="string">&quot;beta.faa&quot;</span>, </span><br><span class="line">                                 gapopen=<span class="number">10</span>, gapextend=<span class="number">0.5</span>, </span><br><span class="line">                                 outfile=<span class="string">&quot;needle.txt&quot;</span>)</span><br><span class="line">stdout, stderr = needle_cline()</span><br><span class="line">align = AlignIO.read(<span class="string">&quot;needle.txt&quot;</span>, <span class="string">&quot;emboss&quot;</span>)</span><br></pre></td></tr></table></figure>

<p>但是这个代码距离我最终代码还有很长的路要走，因为我需要先将序列写出到文件，然后调用命令行，然后再读取输出文件。我的问题是，能不能省掉这几次I&#x2F;O操作。在Emboss文档的下面，我发现Biopython提供了两个模块，分别是Bio.pairwise2和Bio.Align.PairwiseAligner, 可以直接使用Biopython的.SeqRecord对象。按照文档描述，pairwise2和PairwiseAligner能够达到和调用EMBOSS相同的速度，经过我测试，发现两者速度的确差别不大，这说明直接在Python里面进行配对序列联配并不比先输出运行所需文件，然后读取运行生成文件这个流程快，所以经过一番折腾之后，我还是决定使用NeedleCommandline。</p>
<p>为了方便后续调用，我写了一个函数，函数的前3个参数是用来从SeqRecord列表中提取序列用于写出，第4个参数用来提供needle程序的路径。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">def</span> <span class="title function_">run_needle</span>(<span class="params">aseq_pos, bseq_pos, seq, needle_path, verbose=<span class="literal">False</span></span>):</span><br><span class="line">    <span class="string">&quot;&quot;&quot;run EMBOSS needle retrun align </span></span><br><span class="line"><span class="string">    &quot;&quot;&quot;</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># get the Bio object</span></span><br><span class="line">    aseq = seq[aseq_pos]</span><br><span class="line">    bseq = seq[bseq_pos]</span><br><span class="line"></span><br><span class="line">    <span class="comment"># create tempfile for write fasta and needle result</span></span><br><span class="line">    aseq_file = mkstemp(suffix=<span class="string">&quot;.fas&quot;</span>)[<span class="number">1</span>]</span><br><span class="line">    bseq_file =  mkstemp(suffix=<span class="string">&quot;.fas&quot;</span>)[<span class="number">1</span>]</span><br><span class="line"></span><br><span class="line">    needle_file =  mkstemp(suffix=<span class="string">&quot;.needle&quot;</span>)[<span class="number">1</span>]</span><br><span class="line"></span><br><span class="line">    <span class="comment"># write the sequence</span></span><br><span class="line">    SeqIO.write(aseq, aseq_file, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line">    SeqIO.write(bseq, bseq_file, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># build needle commnad</span></span><br><span class="line">    <span class="keyword">if</span> needle_path <span class="keyword">is</span> <span class="literal">None</span>:</span><br><span class="line">        needle_cmd = NeedleCommandline( asequence=aseq_file, bsequence=bseq_file,</span><br><span class="line">                   gapopen=<span class="number">10</span>, gapextend=<span class="number">0.5</span>, outfile=needle_file)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="keyword">assert</span> os.path.isfile(needle_path)</span><br><span class="line">        needle_cmd = NeedleCommandline( needle_path, asequence=aseq_file, bsequence=bseq_file,</span><br><span class="line">                   gapopen=<span class="number">10</span>, gapextend=<span class="number">0.5</span>, outfile=needle_file)</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># running needle_cmd</span></span><br><span class="line">    stdout, stderr = needle_cmd()</span><br><span class="line">    <span class="keyword">if</span> verbose:</span><br><span class="line">        <span class="built_in">print</span>(stdout + stderr, file = sys.stderr)</span><br><span class="line">    </span><br><span class="line">    align = AlignIO.read(needle_file, <span class="string">&quot;emboss&quot;</span>)</span><br><span class="line">    <span class="comment"># deleting the tempfile</span></span><br><span class="line">    os.unlink(aseq_file)</span><br><span class="line">    os.unlink(bseq_file)</span><br><span class="line">    os.unlink(needle_file)</span><br><span class="line">    <span class="keyword">return</span> align</span><br></pre></td></tr></table></figure>

<p>在写这个函数时，我在输出文件的文件名上折腾了很长时间，最初想使用原序列的ID，但ID包括一些奇奇怪怪的字符，例如<code>(|:</code>, 而NeedleCommandline不会自动在构建的命令中增加引号，使得运行时会出错。(你可以试试<code>echo a(b</code>会出现什么情况 )</p>
<p>后来，我使用了随机数作为文件名。这个方法在单个进程时问题不大，当时当我为了加速使用多进程时，却发现发现随机数重复了。进程A输出的文件A由于和进程B的输入文件名一样，导致被跑完的进程B删了。当然，解决方法也很简单，加上一些判断文件是否存在的ifelse语句即可。</p>
<p>不过，我还是选择调用tempfile标准库，使用mkstemp获取临时文件名，才解决了输出文件名的危机。</p>
<p>最终我就可以在循环中调用这个函数，然后将结果保存在列表中，用于后续处理。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">fasta = <span class="string">&quot;/path/to/your/sequence.fasta&quot;</span></span><br><span class="line">seqs = [ fas <span class="keyword">for</span> fas <span class="keyword">in</span> SeqIO.parse(fasta, <span class="string">&quot;fasta&quot;</span>) ]</span><br><span class="line">needle_path = <span class="string">&quot;/opt/biosoft/EMBOSS-6.6.0/bin/needle&quot;</span></span><br><span class="line"></span><br><span class="line">align_res = []</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(<span class="built_in">len</span>(seqs)):</span><br><span class="line">    <span class="keyword">for</span> j <span class="keyword">in</span> <span class="built_in">range</span>(i+<span class="number">1</span>, <span class="built_in">len</span>(seqs)):</span><br><span class="line">        align = run_needle(i, j, seqs , needle_path)</span><br><span class="line">        align_res.append(align )</span><br><span class="line"><span class="comment"># 以pickle保存python对象</span></span><br><span class="line">outfile = <span class="built_in">open</span>(<span class="string">&quot;pairwise_stats.txt&quot;</span>, <span class="string">&quot;w&quot;</span>)</span><br><span class="line"><span class="keyword">for</span> align <span class="keyword">in</span> align_res:</span><br><span class="line">    line = get_psa_stat(align)</span><br><span class="line">    line = <span class="string">&quot;\t&quot;</span>.join(line) + <span class="string">&quot;\n&quot;</span></span><br><span class="line">    outfile.write(line)</span><br><span class="line"></span><br><span class="line">outfile.close()</span><br></pre></td></tr></table></figure>

<p>根据我的计算，循环要运行3万多次，按照平均1s一个循环，只要500多分钟，小于9个小时，我只要在回去前运行这个程序（写完第一版大概是晚上10点），然后第二天早上过来收结果就行了。</p>
<p>但是，正如前文提及到的，「该程序会在多个进程运行时报错」，你也就知道我没有跑循环，而是选择了通过并行对循环进行加速。在一篇中，我将会讲一个，如何把一个简单的问题复杂化，最终又回到最初代码的故事</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/04/14/%E4%BD%BF%E7%94%A8Biopython%E8%B0%83%E7%94%A8EMBOSS%E7%9A%84needle-biopython-emboss-needle/" data-id="clm1xkn5q001rdpni4e3z4e07" data-title="使用Biopython调用EMBOSS的needle" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-将博客搬至CSDN-backuptocsdn" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/04/06/%E5%B0%86%E5%8D%9A%E5%AE%A2%E6%90%AC%E8%87%B3CSDN-backuptocsdn/" class="article-date">
  <time class="dt-published" datetime="2021-04-06T09:39:33.136Z" itemprop="datePublished">2021-04-06</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/04/06/%E5%B0%86%E5%8D%9A%E5%AE%A2%E6%90%AC%E8%87%B3CSDN-backuptocsdn/">将博客搬至CSDN</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>个人博客上的所有内容，也会自动搬家到CSDN上。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/04/06/%E5%B0%86%E5%8D%9A%E5%AE%A2%E6%90%AC%E8%87%B3CSDN-backuptocsdn/" data-id="clm1xkn7g008udpni51f74xmq" data-title="将博客搬至CSDN" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-如何提高Github的访问速度-how-to-accelerate-access-to-github" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/04/06/%E5%A6%82%E4%BD%95%E6%8F%90%E9%AB%98Github%E7%9A%84%E8%AE%BF%E9%97%AE%E9%80%9F%E5%BA%A6-how-to-accelerate-access-to-github/" class="article-date">
  <time class="dt-published" datetime="2021-04-06T09:09:19.160Z" itemprop="datePublished">2021-04-06</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/04/06/%E5%A6%82%E4%BD%95%E6%8F%90%E9%AB%98Github%E7%9A%84%E8%AE%BF%E9%97%AE%E9%80%9F%E5%BA%A6-how-to-accelerate-access-to-github/">如何提高Github的访问速度</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>最近总觉得Github的访问速度变慢了，导致我的工作效率也肉眼可见的降低了，主要体现在代码数量和质量的双重降低。</p>
<p>为了解决这一问题，我通过网络检索找到了一个非常好的工具，叫做dev-sidecar (<a target="_blank" rel="noopener" href="https://github.com/docmirror/dev-sidecar">https://github.com/docmirror/dev-sidecar</a>), 这个工具的好处在于，图形化界面，完美解决了windows和MacOS上的问题，但是问题也出在图形化界面上，这意味着没有配置图形界面的Linux就用不了。</p>
<p>但是这难不倒我们，因为我看到了这个工具的参考部分。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/04/image-faa3d53abb1a485eb99513971e70ed3e.png" alt="reference"></p>
<p>从中, 我们可以看到一个非常重要的地址，fastgit, 这是一个对于 GitHub.com 的镜像加速器。它的使用方法及其简单，也就是将原来的github.com替换成 hub.fastgit.org即可。</p>
<p>以 bioconda&#x2F;bioconda-recipes为例，在使用原版的GitHub时，我们的下载速度基本上维持在2MiB&#x2F;s，某些时候可能不到100Kb.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">$ time git <span class="built_in">clone</span> https://github.com/bioconda/bioconda-recipes</span><br><span class="line">Cloning into <span class="string">&#x27;bioconda-recipes&#x27;</span>...</span><br><span class="line">remote: Enumerating objects: 64, <span class="keyword">done</span>.</span><br><span class="line">remote: Counting objects: 100% (64/64), <span class="keyword">done</span>.</span><br><span class="line">remote: Compressing objects: 100% (50/50), <span class="keyword">done</span>.</span><br><span class="line">remote: Total 276716 (delta 30), reused 31 (delta 11), pack-reused 276652</span><br><span class="line">Receiving objects: 100% (276716/276716), 303.49 MiB | 3.72 MiB/s, <span class="keyword">done</span>.</span><br><span class="line">Resolving deltas: 100% (152345/152345), <span class="keyword">done</span>.</span><br><span class="line">Checking out files: 100% (17906/17906), <span class="keyword">done</span>.</span><br><span class="line">git <span class="built_in">clone</span> https://github.com/bioconda/bioconda-recipes  46.51s user 7.04s system 53% cpu 1:39.24 total</span><br></pre></td></tr></table></figure>

<p>但是使用fastgit加速之后，我的下载速度直接飙升到10Mib&#x2F;s以上，峰值可以达到30Mib&#x2F;s. </p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">$ time git <span class="built_in">clone</span> https://hub.fastgit.org//bioconda/bioconda-recipes</span><br><span class="line">Cloning into <span class="string">&#x27;bioconda-recipes&#x27;</span>...</span><br><span class="line">remote: Enumerating objects: 64, <span class="keyword">done</span>.</span><br><span class="line">remote: Counting objects: 100% (64/64), <span class="keyword">done</span>.</span><br><span class="line">remote: Compressing objects: 100% (50/50), <span class="keyword">done</span>.</span><br><span class="line">remote: Total 276716 (delta 30), reused 31 (delta 11), pack-reused 276652</span><br><span class="line">Receiving objects: 100% (276716/276716), 303.23 MiB | 21.16 MiB/s, <span class="keyword">done</span>.</span><br><span class="line">Resolving deltas: 100% (152348/152348), <span class="keyword">done</span>.</span><br><span class="line">Checking out files: 100% (17906/17906), <span class="keyword">done</span>.</span><br><span class="line">git <span class="built_in">clone</span> https://hub.fastgit.org//bioconda/bioconda-recipes  44.88s user 6.09s system 101% cpu 50.367 total</span><br></pre></td></tr></table></figure>

<p>速度的提升可能是其次的，最重要的是原本因为网络问题的fatal error导致根本下载不了的repos，现在起码能保证能克隆到本地了，实现了从0到1的进步。</p>
<p>当然，如果觉得每次都需要替换URL太过麻烦，fastgit还支持直接修改Git的配置，即</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">git config --global url.<span class="string">&quot;https://hub.fastgit.org/&quot;</span>.insteadOf <span class="string">&quot;https://github.com/&quot;</span></span><br><span class="line">git config --global protocol.http.allow always</span><br></pre></td></tr></table></figure>

<p>之后，原本需要从Github上克隆的资源都会被定向到fastgit上，不再需要手动进行修改了。</p>
<p>不过这一缺陷在于，部分时候镜像站点可能会出现不可用的情况，此时你从Github克隆时依旧会被改向到镜像站点，你会误以为原站点出现了问题。不过只要我们人人都去支持下这个项目，随着可用节点的增多，这一缺陷也不是问题了。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/04/06/%E5%A6%82%E4%BD%95%E6%8F%90%E9%AB%98Github%E7%9A%84%E8%AE%BF%E9%97%AE%E9%80%9F%E5%BA%A6-how-to-accelerate-access-to-github/" data-id="clm1xkn6y0078dpnic2ebdpt5" data-title="如何提高Github的访问速度" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何用WGDI进行共线性分析(三)-comparative-genome-analysis-with-wgdi-3" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/02/23/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%B8%89)-comparative-genome-analysis-with-wgdi-3/" class="article-date">
  <time class="dt-published" datetime="2021-02-23T17:18:00.310Z" itemprop="datePublished">2021-02-23</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/02/23/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%B8%89)-comparative-genome-analysis-with-wgdi-3/">如何用WGDI进行共线性分析(三)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>在上一篇教程的最后，我写道「我们能够更加直观的观察到2个peak，基本确定2个多倍化事件」。这里就引出了一个问题，为什么我们可以根据peak来推断多倍化时间？在Lynch和Conery在2000年发表在Science的论文中，他们证明了小规模基因复制的Ks分布是L型，而在L型分布背景上叠加的峰则是来自于演化历史中某个突然的大规模复制事件。例如下图a中，实线是小规模复制的L型分布（呈指数分布, exponential distribution), 最初的峰可能是近期的串联复制引起，随着时间推移基因丢失，形成一个向下的坡。另一条虚线中的峰（呈正态分布， normal distribution）则是由全基因组复制引起。而b-h是一些物种的ks分布。</p>
<blockquote>
<p>注，Lynch和Conery这两人是最早一批研究真核基因组的复制基因的总体保留和丢失情况的研究者。</p>
</blockquote>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-fd0dc6527c7e416d9e5ba7b7f033a66f.png" alt="Kevin Vanneste et al. 2012 Molecular Biology and Evolution"></p>
<p>也就是说，我们可以通过对Ks频率分布图的观察来判断物种在历史上发生的基因组复制事件。同时又因为WGD在Ks频率分布图中表现为正态分布，那么我们可以通过对峰进行拟合来得到WGD事件所对应的Ks值。</p>
<p>使用WGDI对峰进行拟合需要注意的是，<strong>它一次只能拟合一个峰</strong>，因此我们需要先通过之前Ks可视化中所用到kspeaks模块(-kp)来过滤，接着用<strong>PeaksFit</strong>(-pf)模块对峰进行拟合得到模型参数，最后用KsFigures(-kf)将拟合结果绘制到一张图上。</p>
<p>过滤的目的是筛选出同一个WGD事件所形成的共线性区块，这可以通过设置kspeaks模块中和过滤有关的参数来实现</p>
<ul>
<li>pvalue: 共线性区块的显著性</li>
<li>tandem: 是否过滤串联重复基因</li>
<li>block_length: 共线性区块的基因对的数目</li>
<li>ks_area: 根据 ks_median 筛选给定区间的共线性区块</li>
<li>multiple: 选择homo的哪列作为筛选标准</li>
<li>homo: 根据homo，筛选给定区间内的共线性区块对</li>
</ul>
<p>其中比较关键的是ks_area 和homo,mutiple, 前者确定ks的大致区间，后者则是更精细地筛选共线性区块。block_length可以过滤掉一些比较小的共线性区块，毕竟基因数越多，就越不可能是随机事件所产生的共线性。</p>
<p>通过观察之前Ks频率分布图，我们大致确定两个峰所对应的区间分别是0<del>1.25，1.25</del>2.5。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-b0c9264f436447709adeca16093f7dd9.png" alt="figure 1"></p>
<p>我们分别用 <code>wgdi -kp \? &gt; peak1.conf</code>和 <code>wgdi -kp \? &gt; peak2.conf</code>创建两个文件，主要修改其中的ks_area, homo都设置为-1,1 表示不筛选</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># peak1.conf</span></span><br><span class="line"><span class="section">[kspeaks]</span></span><br><span class="line"><span class="attr">blockinfo</span>  =  ath_block_information.csv</span><br><span class="line"><span class="attr">pvalue</span>  =  <span class="number">0.05</span></span><br><span class="line"><span class="attr">tandem</span>  =  <span class="literal">true</span></span><br><span class="line"><span class="attr">block_length</span>  =  <span class="number">5</span></span><br><span class="line"><span class="attr">ks_area</span>  =  <span class="number">0</span>,<span class="number">1.25</span></span><br><span class="line"><span class="attr">multiple</span>  =  <span class="number">1</span></span><br><span class="line"><span class="attr">homo</span>  =  -<span class="number">1</span>,<span class="number">1</span></span><br><span class="line"><span class="attr">fontsize</span>  =  <span class="number">9</span></span><br><span class="line"><span class="attr">area</span>  =  <span class="number">0</span>,<span class="number">2</span></span><br><span class="line"><span class="attr">figsize</span>  =  <span class="number">10</span>,<span class="number">6.18</span></span><br><span class="line"><span class="attr">savefig</span>  =  ath.ks_peak1.distri.pdf</span><br><span class="line"><span class="attr">savefile</span>  =  ath.ks_peak1.distri.csv</span><br><span class="line"></span><br><span class="line"><span class="comment"># peak2</span></span><br><span class="line"><span class="section">[kspeaks]</span></span><br><span class="line"><span class="attr">blockinfo</span>  =  ath_block_information.csv</span><br><span class="line"><span class="attr">pvalue</span>  =  <span class="number">0.05</span></span><br><span class="line"><span class="attr">tandem</span>  =  <span class="literal">true</span></span><br><span class="line"><span class="attr">block_length</span>  =  <span class="number">5</span></span><br><span class="line"><span class="attr">ks_area</span>  =  <span class="number">1.25</span>,<span class="number">2.5</span></span><br><span class="line"><span class="attr">multiple</span>  =  <span class="number">1</span></span><br><span class="line"><span class="attr">homo</span>  =  -<span class="number">1</span>,<span class="number">1</span></span><br><span class="line"><span class="attr">fontsize</span>  =  <span class="number">9</span></span><br><span class="line"><span class="attr">area</span>  =  <span class="number">0</span>,<span class="number">2</span></span><br><span class="line"><span class="attr">figsize</span>  =  <span class="number">10</span>,<span class="number">6.18</span></span><br><span class="line"><span class="attr">savefig</span>  =  ath.ks_peak2.distri.pdf</span><br><span class="line"><span class="attr">savefile</span>  =  ath.ks_peak2.distri.csv</span><br><span class="line"> </span><br></pre></td></tr></table></figure>

<p>运行 <code>wgdi -kp peak1.conf</code>  和 <code>wgdi -kp peak.conf</code>输出结果。下图中左是peak1, 右是peak2.</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-a7bf68cd528647cba17c1bcbb141ac41.png" alt="figure 2 "></p>
<p>得到这个图之后，我们还可以进一步绘制Ks点阵图， 使用 <code>wgdi -bk &gt;&gt;peak1.conf</code>  和 <code>wgdi -bk &gt;&gt; peak2.conf</code>增加配置信息，然后对其进行修改</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># peak1</span></span><br><span class="line"><span class="section">[blockks]</span></span><br><span class="line"><span class="attr">lens1</span>  =  ath.len</span><br><span class="line"><span class="attr">lens2</span>  =  ath.len</span><br><span class="line"><span class="attr">genome1_name</span>  =  Athaliana</span><br><span class="line"><span class="attr">genome2_name</span>  =  Athaliana</span><br><span class="line"><span class="attr">blockinfo</span>  =  ath.ks_peak1.distri.csv</span><br><span class="line"><span class="attr">pvalue</span>  =  <span class="number">0.05</span></span><br><span class="line"><span class="attr">tandem</span>  =  <span class="literal">true</span></span><br><span class="line"><span class="attr">tandem_length</span>  =  <span class="number">200</span></span><br><span class="line"><span class="attr">markersize</span>  =  <span class="number">1</span></span><br><span class="line"><span class="attr">area</span>  =  <span class="number">0</span>,<span class="number">2</span></span><br><span class="line"><span class="attr">block_length</span>  =  <span class="number">5</span></span><br><span class="line"><span class="attr">figsize</span>  =  <span class="number">8</span>,<span class="number">8</span></span><br><span class="line"><span class="attr">savefig</span>  =  ath.ks_peak1_dotplot.pdf</span><br><span class="line"><span class="comment"># peak2</span></span><br><span class="line"><span class="section">[blockks]</span></span><br><span class="line"><span class="attr">lens1</span>  =  ath.len</span><br><span class="line"><span class="attr">lens2</span>  =  ath.len</span><br><span class="line"><span class="attr">genome1_name</span>  =  Athaliana</span><br><span class="line"><span class="attr">genome2_name</span>  =  Athaliana</span><br><span class="line"><span class="attr">blockinfo</span>  =  ath.ks_peak2.distri.csv</span><br><span class="line"><span class="attr">pvalue</span>  =  <span class="number">0.05</span></span><br><span class="line"><span class="attr">tandem</span>  =  <span class="literal">true</span></span><br><span class="line"><span class="attr">tandem_length</span>  =  <span class="number">200</span></span><br><span class="line"><span class="attr">markersize</span>  =  <span class="number">1</span></span><br><span class="line"><span class="attr">area</span>  =  <span class="number">0</span>,<span class="number">2</span></span><br><span class="line"><span class="attr">block_length</span>  =  <span class="number">5</span></span><br><span class="line"><span class="attr">figsize</span>  =  <span class="number">8</span>,<span class="number">8</span></span><br><span class="line"><span class="attr">savefig</span>  =  ath.ks_peak2_dotplot.pdf</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>运行 <code>wgdi -bk peak1.conf</code>，和 <code>wgdi -bk peak2.conf</code> 输出结果。下图中，左边是peak1，右边是peak2.</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-fd1f77c2359c4879869ad5865dc22c28.png" alt="figure 3"></p>
<p>由于我们设置的是 homo&#x3D;-1,1 因此只是根据共线性块的Ks中位数筛选不同的peak，如果需要更加精细的挑选给定Ks中位数范围内的共线性区块，可以尝试修改homo值再观察结果。</p>
<blockquote>
<p>homo值表示的共线性区块中基因对的匹配情况，越接近于1，越有可能是近期WGD事件得到的基因对，越接近-1，越有可能是古老WGD事件得到的基因对。</p>
</blockquote>
<p>接着，我们将用blockks步骤输出的ath.ks_peak1.distri.csv和tah.ks_peak2.distri.csv 作为peakfit模块的输入，用于拟合。</p>
<p>分别用 <code>wgdi -pf \? &gt;&gt; peak1.conf</code> 和 <code>wgdi -pf \? &gt;&gt; peak2.conf</code> 创建配置文件, 主要修改blockinfo</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># peak1</span></span><br><span class="line"><span class="section">[peaksfit]</span></span><br><span class="line"><span class="attr">blockinfo</span>  =  ath.ks_peak1.distri.csv</span><br><span class="line"><span class="attr">mode</span>  =  median</span><br><span class="line"><span class="attr">bins_number</span>  =  <span class="number">200</span></span><br><span class="line"><span class="attr">ks_area</span>  =  <span class="number">0</span>,<span class="number">10</span></span><br><span class="line"><span class="attr">fontsize</span>  =  <span class="number">9</span></span><br><span class="line"><span class="attr">area</span>  =  <span class="number">0</span>,<span class="number">3</span></span><br><span class="line"><span class="attr">figsize</span>  =  <span class="number">10</span>,<span class="number">6.18</span></span><br><span class="line"><span class="attr">savefig</span>  =  ath.ks_peak1_peaksfit.pdf </span><br><span class="line"></span><br><span class="line"><span class="comment"># peak2</span></span><br><span class="line"><span class="section">[peaksfit]</span></span><br><span class="line"><span class="attr">blockinfo</span>  =  ath.ks_peak2.distri.csv</span><br><span class="line"><span class="attr">mode</span>  =  median</span><br><span class="line"><span class="attr">bins_number</span>  =  <span class="number">200</span></span><br><span class="line"><span class="attr">ks_area</span>  =  <span class="number">0</span>,<span class="number">10</span></span><br><span class="line"><span class="attr">fontsize</span>  =  <span class="number">9</span></span><br><span class="line"><span class="attr">area</span>  =  <span class="number">0</span>,<span class="number">3</span></span><br><span class="line"><span class="attr">figsize</span>  =  <span class="number">10</span>,<span class="number">6.18</span></span><br><span class="line"><span class="attr">savefig</span>  =  ath.ks_peak2_peaksfit.pdf </span><br></pre></td></tr></table></figure>

<p>分别运行 <code>wgdi -pf peak1.conf</code> 和 <code>wgdi -pf peak2.conf</code>, 会得到拟合的图，以及拟合得到参数</p>
<p>peak1拟合参数</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">R-square: 0.9804748699350867</span><br><span class="line">The gaussian fitting curve parameters are :</span><br><span class="line">5.02360835744403  |  0.8319599832352784  |  0.10382203381206191</span><br></pre></td></tr></table></figure>

<p>peak2拟合参数</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">R-square: 0.9188261142099129</span><br><span class="line">The gaussian fitting curve parameters are :</span><br><span class="line">2.084853812322066  |  1.8332872127128195  |  0.2506813629824027</span><br></pre></td></tr></table></figure>

<p>拟合参数将会用于后续的 ksfigure 模块。</p>
<p>我们新建一个 all_ks.csv文件, 该文件的第一行为标题行，第二行以后为数据行。一共有4+3n列，其中第一列是样本信息，第2-3列对应线条的属性，后面都是拟合参数</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">,color,linewidth,linestyle,,,,,,</span><br><span class="line">ath_ath,green,1,--,5.02360835744403,0.8319599832352784,0.10382203381206191,2.084853812322066,1.8332872127128195,0.2506813629824027</span><br></pre></td></tr></table></figure>

<p>注意，在终端里面编辑时，需要记住linestyle后面的逗号数量是 3n个, 其中n是最大peak数，例如拟南芥有2个peak，那么就得有6个逗号。</p>
<p>假如，我们之前用wgdi拟合过其他物种的peak，也得到了一些拟合参数，那么添加到这个文件中，例如</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"> ,color,linewidth,linestyle,,,,,,</span><br><span class="line">ath_ath,red,1,--,5.02360835744403,0.8319599832352784,0.10382203381206191,2.084853812322066,1.8332872127128195,0.2506813629824027</span><br><span class="line">vvi_vvi,yellow,1,-,3.00367275,1.288717936,0.177816426</span><br><span class="line">vvi_oin_gamma,orange,2,-,1.910418336,1.328469514,0.262257112</span><br><span class="line">vvi_oin,green,1,-,4.948194212,0.882608858,0.10426</span><br></pre></td></tr></table></figure>

<p>最后我们用 <code>wgdi -kf &gt;&gt; ath.conf</code> , 创建配置文件并进行修改</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="section">[ksfigure]</span></span><br><span class="line"><span class="attr">ksfit</span> = all_ks.csv</span><br><span class="line"><span class="attr">labelfontsize</span> = <span class="number">15</span></span><br><span class="line"><span class="attr">legendfontsize</span> = <span class="number">15</span></span><br><span class="line"><span class="attr">xlabel</span> = none</span><br><span class="line"><span class="attr">ylabel</span> = none</span><br><span class="line"><span class="attr">title</span> = none</span><br><span class="line"><span class="attr">area</span> = <span class="number">0</span>,<span class="number">4</span></span><br><span class="line"><span class="attr">figsize</span> = <span class="number">10</span>,<span class="number">6.18</span></span><br><span class="line"><span class="attr">savefig</span> =  all_ks.pdf</span><br></pre></td></tr></table></figure>

<p>运行 <code>wgdi -kf  ath.conf</code> 得到下图</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-b70bff05f977435c9d22b76bd98c69ea.png" alt="figure 4"></p>
<p>最后，对这篇教程进行总结。</p>
<ol>
<li>我们依次使用 wgdi -kp 过滤共线性区块， wgdi -pf 对峰进行拟合， wgdi -kf 展示最后的拟合结果</li>
<li>共线性区块的过滤是一个精细的工作，可以尝试设置不同的homo范围观察峰的变化来确定参数是否合理</li>
</ol>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/02/23/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%B8%89)-comparative-genome-analysis-with-wgdi-3/" data-id="clm1xkn70007idpnic751fcrs" data-title="如何用WGDI进行共线性分析(三)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-WGDI之深入理解blockinfo输出结果-wgdi-explaining-blockinfo-result" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/02/20/WGDI%E4%B9%8B%E6%B7%B1%E5%85%A5%E7%90%86%E8%A7%A3blockinfo%E8%BE%93%E5%87%BA%E7%BB%93%E6%9E%9C-wgdi-explaining-blockinfo-result/" class="article-date">
  <time class="dt-published" datetime="2021-02-20T16:59:36.979Z" itemprop="datePublished">2021-02-20</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/02/20/WGDI%E4%B9%8B%E6%B7%B1%E5%85%A5%E7%90%86%E8%A7%A3blockinfo%E8%BE%93%E5%87%BA%E7%BB%93%E6%9E%9C-wgdi-explaining-blockinfo-result/">WGDI之深入理解blockinfo输出结果</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>blockinfo模块输出文件以csv格式进行存放，共23列，可以用EXCEL直接打开。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-d50c94541b844111b750a4d7fd6aef40.png" alt="block info"></p>
<p>其中16列非常容易裂解，描述如下</p>
<ol>
<li><p><code>id</code> 即共线性的结果的唯一标识</p>
</li>
<li><p><code>chr1</code>,<code>start1</code>,<code>end1</code> 即参考基因组（点图的左边）的共线性范围（对应GFF1的位置）</p>
</li>
<li><p><code>chr2</code>,<code>start2</code>,<code>end2</code> 即参考基因组（点图的上边）的共线性范围（对应GFF2的位置)</p>
</li>
<li><p><code>pvalue</code> 即共线性结果评估，常常认为小于0.01的更合理些</p>
</li>
<li><p><code>length</code> 即共线性片段中基因对数目</p>
</li>
<li><p><code>ks_median</code> 即共线性片段上所有基因对<code>ks</code>的中位数（主要用来评判ks分布的）</p>
</li>
<li><p><code>ks_average</code> 即共线性片段上所有基因对<code>ks</code>的平均值</p>
</li>
<li><p><code>block1</code>,<code>block2</code>分别为共线性片段上基因<code>order</code>的位置。</p>
</li>
<li><p><code>ks</code>共线性片段上所有基因对的<code>ks</code>值</p>
</li>
<li><p><code>density1</code>,<code>density2</code> 共线性片段的基因分布密集程度。值越小表示稀疏。</p>
</li>
</ol>
<p>最后两列，<code>class1</code>和 <code>class2</code>会在 alignment 模块中用到，对应的是两个block分组，默认值是0表示两个block是同一组。这两列后期需要自己根据覆盖率，染色体核型等多个方面进行确定。举个例子，我们可以根据 homo1 的取值范围对class1进行赋值，例如-1<del>-0.5 是 1，-0.5 ~ 0.5 是2，0.5</del>1是3，最后在alignment中会就会用三种颜色来展示，例如下图的1,2,3分别对应red,blue,green.</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-d164d05aa6414c6ca054ea747952028d.png" alt="alignment"></p>
<p>中间的<code>homo1</code>,<code>homo2</code>,<code>homo3</code>,<code>homo4</code>,<code>homo5</code>并非那么直观，先说结论：</p>
<ul>
<li><p>这里的homoN(N&#x3D;1,2,3,4,5) 表示一个基因有N个最佳匹配时的取值</p>
</li>
<li><p>N由mutiple参数确定，对应点阵图(dotplot)中的红点</p>
</li>
<li><p>multiple的取值一般取1即可，表示最近一次的WGD可能是一次二倍化事件，因此每个基因只会有一个最佳匹配。如果设置为2，可能是一次3倍化，每个基因由两个最佳匹配。当然实际情况可能会更加复杂，比如说异源四倍体，或者异源六倍体，或者没有多倍化只是小规模的基因复制(small-scale gene duplication) 等情况，也会影响multiple的设置。</p>
</li>
<li><p>homoN会在后面过滤共线性区块时用到，一般最近的WGD事件所产生的共线性区块会比较接近1，而古老的WGD产生的共线性区块则接近-1.</p>
</li>
</ul>
<p>接着，我们将根据源代码 <a target="_blank" rel="noopener" href="https://github.com/SunPengChuan/wgdi/blob/53914c317a457c8a629f66912295425ffb3f7c63/wgdi/block_info.py#L78">blast_homo</a>和<a target="_blank" rel="noopener" href="https://github.com/SunPengChuan/wgdi/blob/53914c317a457c8a629f66912295425ffb3f7c63/wgdi/block_info.py#L15">blast_position</a> 来说明结算过程。</p>
<p>首先需要用到blast_homo函数，用来输出每个基因对在不同最佳匹配情况下的取值（-1,0,1)。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">def</span> <span class="title function_">blast_homo</span>(<span class="params">self, blast, gff1, gff2, repeat_number</span>):</span><br><span class="line">    index = [group.sort_values(by=<span class="number">11</span>, ascending=<span class="literal">False</span>)[:repeat_number].index.tolist()</span><br><span class="line">             <span class="keyword">for</span> name, group <span class="keyword">in</span> blast.groupby([<span class="number">0</span>])]</span><br><span class="line">    blast = blast.loc[np.concatenate(</span><br><span class="line">        np.array([k[:repeat_number] <span class="keyword">for</span> k <span class="keyword">in</span> index], dtype=<span class="built_in">object</span>)), [<span class="number">0</span>, <span class="number">1</span>]]</span><br><span class="line">    blast = blast.assign(homo1=np.nan, homo2=np.nan,</span><br><span class="line">                         homo3=np.nan, homo4=np.nan, homo5=np.nan)</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">1</span>, <span class="number">6</span>):</span><br><span class="line">        bluenum = i+<span class="number">5</span></span><br><span class="line">        redindex = np.concatenate(</span><br><span class="line">            np.array([k[:i] <span class="keyword">for</span> k <span class="keyword">in</span> index], dtype=<span class="built_in">object</span>))</span><br><span class="line">        blueindex = np.concatenate(</span><br><span class="line">            np.array([k[i:bluenum] <span class="keyword">for</span> k <span class="keyword">in</span> index], dtype=<span class="built_in">object</span>))</span><br><span class="line">        grayindex = np.concatenate(</span><br><span class="line">            np.array([k[bluenum:repeat_number] <span class="keyword">for</span> k <span class="keyword">in</span> index], dtype=<span class="built_in">object</span>))</span><br><span class="line">        blast.loc[redindex, <span class="string">&#x27;homo&#x27;</span>+<span class="built_in">str</span>(i)] = <span class="number">1</span></span><br><span class="line">        blast.loc[blueindex, <span class="string">&#x27;homo&#x27;</span>+<span class="built_in">str</span>(i)] = <span class="number">0</span></span><br><span class="line">        blast.loc[grayindex, <span class="string">&#x27;homo&#x27;</span>+<span class="built_in">str</span>(i)] = -<span class="number">1</span></span><br><span class="line">    <span class="keyword">return</span> blast</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>for循环前的代码作用是提取每个基因BLAST后的前N个最佳结果。循环的作用基因对进行赋值，主要规则是基因对如果在点图中为<strong>红色</strong>，赋值为1，<strong>蓝色</strong>赋值为0，<strong>灰色</strong>赋值为-1。</p>
<ul>
<li><p>homo1 对应 redindex &#x3D; 0:1, bluenum &#x3D; 1:6, grayindex &#x3D; 6:repeat_number</p>
</li>
<li><p>homo2 对应redindex &#x3D; 0:2, bluenum &#x3D; 2:7, grayindex &#x3D; 7:repeat_number</p>
</li>
<li><p>…</p>
</li>
<li><p>homo5对应redindex&#x3D;0:5, bluenum&#x3D;5:10, grayindex &#x3D; 10:repeat_number</p>
</li>
</ul>
<p>最终函数返回的就是每个基因对，在不同最佳匹配数下的赋值结果。</p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">                0          1  homo1  homo2  homo3  homo4  homo5</span><br><span class="line">185893  AT1G01010  AT4G01550    1.0    1.0    1.0    1.0    1.0</span><br><span class="line">185894  AT1G01010  AT1G02230    0.0    1.0    1.0    1.0    1.0</span><br><span class="line">185899  AT1G01010  AT4G35580   -1.0    0.0    0.0    0.0    0.0</span><br><span class="line">185900  AT1G01010  AT1G33060   -1.0   -1.0    0.0    0.0    0.0</span><br><span class="line">185901  AT1G01010  AT3G49530   -1.0   -1.0   -1.0    0.0    0.0</span><br><span class="line">185902  AT1G01010  AT5G24590   -1.0   -1.0   -1.0   -1.0    0.0</span><br><span class="line">250822  AT1G01030  AT1G13260    0.0    0.0    0.0    1.0    1.0</span><br><span class="line">250823  AT1G01030  AT1G68840    0.0    0.0    0.0    0.0    1.0</span><br><span class="line">250825  AT1G01030  AT1G25560    0.0    0.0    0.0    0.0    0.0</span><br><span class="line">250826  AT1G01030  AT3G25730   -1.0    0.0    0.0    0.0    0.0</span><br><span class="line">250824  AT1G01030  AT5G06250   -1.0   -1.0    0.0    0.0    0.0</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>然后block_position函数, 会用 <code>for k in block[1]</code>的循环提取每个共线性区块中每个基因对的homo值，然后用 <code>df = pd.DataFrame(blk_homo)</code> 和 <code> homo = df.mean().values</code>求均值。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">def</span> <span class="title function_">block_position</span>(<span class="params">self, collinearity, blast, gff1, gff2, ks</span>):</span><br><span class="line">    data = []</span><br><span class="line">    <span class="keyword">for</span> block <span class="keyword">in</span> collinearity:</span><br><span class="line">        blk_homo, blk_ks = [],  []</span><br><span class="line">        <span class="keyword">if</span> block[<span class="number">1</span>][<span class="number">0</span>][<span class="number">0</span>] <span class="keyword">not</span> <span class="keyword">in</span> gff1.index <span class="keyword">or</span> block[<span class="number">1</span>][<span class="number">0</span>][<span class="number">2</span>] <span class="keyword">not</span> <span class="keyword">in</span> gff2.index:</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line">        chr1, chr2 = gff1.loc[block[<span class="number">1</span>][<span class="number">0</span>][<span class="number">0</span>],</span><br><span class="line">                              <span class="string">&#x27;chr&#x27;</span>], gff2.loc[block[<span class="number">1</span>][<span class="number">0</span>][<span class="number">2</span>], <span class="string">&#x27;chr&#x27;</span>]</span><br><span class="line">        array1, array2 = [<span class="built_in">float</span>(i[<span class="number">1</span>]) <span class="keyword">for</span> i <span class="keyword">in</span> block[<span class="number">1</span>]], [</span><br><span class="line">            <span class="built_in">float</span>(i[<span class="number">3</span>]) <span class="keyword">for</span> i <span class="keyword">in</span> block[<span class="number">1</span>]]</span><br><span class="line">        start1, end1 = array1[<span class="number">0</span>], array1[-<span class="number">1</span>]</span><br><span class="line">        start2, end2 = array2[<span class="number">0</span>], array2[-<span class="number">1</span>]</span><br><span class="line">        block1, block2 = [], []</span><br><span class="line">        <span class="comment">## 提取block中对应基因对的homo值</span></span><br><span class="line">        <span class="keyword">for</span> k <span class="keyword">in</span> block[<span class="number">1</span>]:</span><br><span class="line">            block1.append(<span class="built_in">int</span>(<span class="built_in">float</span>(k[<span class="number">1</span>])))</span><br><span class="line">            block2.append(<span class="built_in">int</span>(<span class="built_in">float</span>(k[<span class="number">3</span>])))</span><br><span class="line">            <span class="keyword">if</span> k[<span class="number">0</span>]+<span class="string">&quot;,&quot;</span>+k[<span class="number">2</span>] <span class="keyword">in</span> ks.index:</span><br><span class="line">                pair_ks = ks[<span class="built_in">str</span>(k[<span class="number">0</span>])+<span class="string">&quot;,&quot;</span>+<span class="built_in">str</span>(k[<span class="number">2</span>])]</span><br><span class="line">                blk_ks.append(pair_ks)</span><br><span class="line">            <span class="keyword">elif</span> k[<span class="number">2</span>]+<span class="string">&quot;,&quot;</span>+k[<span class="number">0</span>] <span class="keyword">in</span> ks.index:</span><br><span class="line">                pair_ks = ks[<span class="built_in">str</span>(k[<span class="number">2</span>])+<span class="string">&quot;,&quot;</span>+<span class="built_in">str</span>(k[<span class="number">0</span>])]</span><br><span class="line">                blk_ks.append(pair_ks)</span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                blk_ks.append(-<span class="number">1</span>)</span><br><span class="line">            <span class="keyword">if</span> k[<span class="number">0</span>]+<span class="string">&quot;,&quot;</span>+k[<span class="number">2</span>] <span class="keyword">not</span> <span class="keyword">in</span> blast.index:</span><br><span class="line">                <span class="keyword">continue</span></span><br><span class="line">            blk_homo.append(</span><br><span class="line">                blast.loc[k[<span class="number">0</span>]+<span class="string">&quot;,&quot;</span>+k[<span class="number">2</span>], [<span class="string">&#x27;homo&#x27;</span>+<span class="built_in">str</span>(i) <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">1</span>, <span class="number">6</span>)]].values.tolist())</span><br><span class="line">        ks_arr = [k <span class="keyword">for</span> k <span class="keyword">in</span> blk_ks <span class="keyword">if</span> k &gt;= <span class="number">0</span>]</span><br><span class="line">        <span class="keyword">if</span> <span class="built_in">len</span>(ks_arr) == <span class="number">0</span>:</span><br><span class="line">            ks_median = -<span class="number">1</span></span><br><span class="line">            ks_average = -<span class="number">1</span></span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            arr_ks = [k <span class="keyword">for</span> k <span class="keyword">in</span> blk_ks <span class="keyword">if</span> k &gt;= <span class="number">0</span>]</span><br><span class="line">            ks_median = base.get_median(arr_ks)</span><br><span class="line">            ks_average = <span class="built_in">sum</span>(arr_ks)/<span class="built_in">len</span>(arr_ks)</span><br><span class="line">        <span class="comment"># 对5列homo值求均值    </span></span><br><span class="line">        df = pd.DataFrame(blk_homo)</span><br><span class="line">        homo = df.mean().values</span><br><span class="line">        <span class="keyword">if</span> <span class="built_in">len</span>(homo) == <span class="number">0</span>:</span><br><span class="line">            homo = [-<span class="number">1</span>, -<span class="number">1</span>, -<span class="number">1</span>, -<span class="number">1</span>, -<span class="number">1</span>]</span><br><span class="line">        blkks = <span class="string">&#x27;_&#x27;</span>.join([<span class="built_in">str</span>(k) <span class="keyword">for</span> k <span class="keyword">in</span> blk_ks])</span><br><span class="line">        block1 = <span class="string">&#x27;_&#x27;</span>.join([<span class="built_in">str</span>(k) <span class="keyword">for</span> k <span class="keyword">in</span> block1])</span><br><span class="line">        block2 = <span class="string">&#x27;_&#x27;</span>.join([<span class="built_in">str</span>(k) <span class="keyword">for</span> k <span class="keyword">in</span> block2])</span><br><span class="line">        data.append([block[<span class="number">0</span>], chr1, chr2, start1, end1, start2, end2, block[<span class="number">2</span>], <span class="built_in">len</span>(</span><br><span class="line">            block[<span class="number">1</span>]), ks_median, ks_average, homo[<span class="number">0</span>], homo[<span class="number">1</span>], homo[<span class="number">2</span>], homo[<span class="number">3</span>], homo[<span class="number">4</span>], block1, block2, blkks])</span><br><span class="line">    data = pd.DataFrame(data, columns=[<span class="string">&#x27;id&#x27;</span>, <span class="string">&#x27;chr1&#x27;</span>, <span class="string">&#x27;chr2&#x27;</span>, <span class="string">&#x27;start1&#x27;</span>, <span class="string">&#x27;end1&#x27;</span>, <span class="string">&#x27;start2&#x27;</span>, <span class="string">&#x27;end2&#x27;</span>,</span><br><span class="line">                                       <span class="string">&#x27;pvalue&#x27;</span>, <span class="string">&#x27;length&#x27;</span>, <span class="string">&#x27;ks_median&#x27;</span>, <span class="string">&#x27;ks_average&#x27;</span>, <span class="string">&#x27;homo1&#x27;</span>, <span class="string">&#x27;homo2&#x27;</span>, <span class="string">&#x27;homo3&#x27;</span>,</span><br><span class="line">                                       <span class="string">&#x27;homo4&#x27;</span>, <span class="string">&#x27;homo5&#x27;</span>, <span class="string">&#x27;block1&#x27;</span>, <span class="string">&#x27;block2&#x27;</span>, <span class="string">&#x27;ks&#x27;</span>])</span><br><span class="line">    data[<span class="string">&#x27;density1&#x27;</span>] = data[<span class="string">&#x27;length&#x27;</span>] / \</span><br><span class="line">        ((data[<span class="string">&#x27;end1&#x27;</span>]-data[<span class="string">&#x27;start1&#x27;</span>]).<span class="built_in">abs</span>()+<span class="number">1</span>)</span><br><span class="line">    data[<span class="string">&#x27;density2&#x27;</span>] = data[<span class="string">&#x27;length&#x27;</span>] / \</span><br><span class="line">        ((data[<span class="string">&#x27;end2&#x27;</span>]-data[<span class="string">&#x27;start2&#x27;</span>]).<span class="built_in">abs</span>()+<span class="number">1</span>)</span><br><span class="line">    <span class="keyword">return</span> data     </span><br></pre></td></tr></table></figure>

<p>最终得到的homo1的homo5，是不同最佳匹配基因数下计算的值。如果共线性的点大部分<strong>为红色</strong>，那么该值接近于1；如果共线性的点大部分<strong>为蓝色</strong>，那么该值接近于0；如果共线性的点大部分<strong>为灰色</strong>，那么该值接近于-1。也就是我们可以根据最初的点图中的颜色来确定将来筛选不同WGD事件所产生共线性区块。</p>
<p>这也就是为什么homoN可以作为共线性片段的筛选标准。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/02/20/WGDI%E4%B9%8B%E6%B7%B1%E5%85%A5%E7%90%86%E8%A7%A3blockinfo%E8%BE%93%E5%87%BA%E7%BB%93%E6%9E%9C-wgdi-explaining-blockinfo-result/" data-id="clm1xkn9500cudpnicnea3fpg" data-title="WGDI之深入理解blockinfo输出结果" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何用WGDI进行共线性分析(二)-comparative-genome-analysis-with-wgdi-2" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/02/02/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%BA%8C)-comparative-genome-analysis-with-wgdi-2/" class="article-date">
  <time class="dt-published" datetime="2021-02-02T15:56:46.659Z" itemprop="datePublished">2021-02-02</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/02/02/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%BA%8C)-comparative-genome-analysis-with-wgdi-2/">如何用WGDI进行共线性分析(二)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h2 id="Ks可视化"><a href="#Ks可视化" class="headerlink" title="Ks可视化"></a>Ks可视化</h2><p>我们在上一篇<a href="/archives/comparative-genome-analysis-with-wgdi-1">如何用WGDI进行共线性分析(一)</a>得到共线性分析结果和Ks值输出结果的整合表格文件后，我们就可以绘制Ks点阵图和Ks频率分布图对共线性区的Ks进行探索性分析，从而确定可能的Ks峰值，用于后续分析。</p>
<h3 id="Ks点阵图"><a href="#Ks点阵图" class="headerlink" title="Ks点阵图"></a>Ks点阵图</h3><p>最初绘制的点图信息量很大，基本上涵盖了历史上发生的加倍事件所产生的共线性区。我们能大致的判断片段是否存在复制以及复制了多少次，至于这些片段是否来自于同一次加倍事件则不太好确定。借助于Ks信息 (Ks值可以反应一定尺度内的演化时间)，我们就可以较为容易地根据点阵图上共线性区域的颜色来区分多倍化事件。</p>
<p>首先，创建配置文件（这次是 <code>-bk</code>模块，BlockKs）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgdi -bk \? &gt;&gt; ath.conf</span><br></pre></td></tr></table></figure>

<p>然后，修改配置文件</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="section">[blockks]</span></span><br><span class="line"><span class="attr">lens1</span> = ath.len</span><br><span class="line"><span class="attr">lens2</span> = ath.len</span><br><span class="line"><span class="attr">genome1_name</span> =  A. thaliana</span><br><span class="line"><span class="attr">genome2_name</span> =  A. thaliana</span><br><span class="line"><span class="attr">blockinfo</span> =  ath_block_information.csv</span><br><span class="line"><span class="attr">pvalue</span> = <span class="number">0.05</span></span><br><span class="line"><span class="attr">tandem</span> = <span class="literal">true</span></span><br><span class="line"><span class="attr">tandem_length</span> = <span class="number">200</span></span><br><span class="line"><span class="attr">markersize</span> = <span class="number">1</span></span><br><span class="line"><span class="attr">area</span> = <span class="number">0</span>,<span class="number">3</span></span><br><span class="line"><span class="attr">block_length</span> =  <span class="number">5</span></span><br><span class="line"><span class="attr">figsize</span> = <span class="number">8</span>,<span class="number">8</span></span><br><span class="line"><span class="attr">savefig</span> = ath.ks.dotplot.pdf</span><br></pre></td></tr></table></figure>

<p>blockinfo 是前面共线性分析和Ks分析的整合结果，是绘图的基础。根据下面几个标准进行过滤</p>
<ul>
<li>pvalue: 共线性区的显著性， 对应blockinfo中的pvalue列</li>
<li>tandem: 是否过滤由串联基因所形成的共线性区，即点阵图中对角线部分</li>
<li>tandem_length: 如果过滤，那么评估tandem的标准就是两个区块的物理距离</li>
<li>block_length: 一个共线区的最小基因对的数量，对应blockinfo中的length列</li>
</ul>
<p>最后运行wgdi，输出图片。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgdi -bk ath.conf</span><br></pre></td></tr></table></figure>

<p>输出图片中的Ks的取值范围由参数area控制。图中的每个点都是各个共线性区内的基因对的Ks值。不难发现每个共线区内的Ks的颜色都差不多，意味着Ks值波动不大，基于这个现象，我们在判定Ks峰值的时候，采用<strong>共线性区的中位数</strong>就比其他方式要准确的多。在图中，我们大致能观察到2种颜色，绿色和蓝色，对应着两次比较近的加倍事件。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-8b9639b4eef24c2eafb7e691494e9a90.png" alt="Ks点阵图"></p>
<h3 id="Ks频率分布图"><a href="#Ks频率分布图" class="headerlink" title="Ks频率分布图"></a>Ks频率分布图</h3><p>除了用点阵图展示Ks外，我们还可以绘制Ks频率的分布情况。假如一个物种在历史上发生过多倍化，那么从那个时间点增加的基因经过相同时间的演化后，基因对之间的Ks值应该相差不多，即，归属于某个Ks区间的频率会明显高于其他区间。</p>
<p>首先，创建配置文件（-kp, ksPeak）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wgdi -kp ? &gt;&gt; ath.conf</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>然后，修改配置文件</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="section">[kspeaks]</span></span><br><span class="line"><span class="attr">blockinfo</span> = ath_block_information.csv</span><br><span class="line"><span class="attr">pvalue</span> = <span class="number">0.05</span></span><br><span class="line"><span class="attr">tandem</span> = <span class="literal">true</span></span><br><span class="line"><span class="attr">block_length</span> = <span class="number">5</span></span><br><span class="line"><span class="attr">ks_area</span> = <span class="number">0</span>,<span class="number">10</span></span><br><span class="line"><span class="attr">multiple</span>  = <span class="number">1</span></span><br><span class="line"><span class="attr">homo</span> = -<span class="number">1</span>,<span class="number">1</span></span><br><span class="line"><span class="attr">fontsize</span> = <span class="number">9</span></span><br><span class="line"><span class="attr">area</span> = <span class="number">0</span>,<span class="number">3</span></span><br><span class="line"><span class="attr">figsize</span> = <span class="number">10</span>,<span class="number">6.18</span></span><br><span class="line"><span class="attr">savefig</span> = ath.ks_median.distri.pdf</span><br><span class="line"><span class="attr">savefile</span> = ath.ks_median.distri.csv</span><br></pre></td></tr></table></figure>

<p>这一步除了输出Ks峰图(savefig)外，还会输出一个根据输入文件( blockinfo )进行过滤后的输出文件(savefile)。<strong>过滤标准</strong>除了之前Ks点阵图提及的 tandem, pvalue, block_length 外，还多了三个参数, ks_area, multiple, homo.</p>
<ul>
<li><p>pvalue: 共线性区的显著性， 对应blockinfo中的pvalue列，pvalue设置为0.05时会保留看着很不错的共线性片段，但是会导致古老片段的减少。</p>
</li>
<li><p>ks_area对应blockinfo中的ks列，该列记录了共线区所有基因对的ks值。ks_area&#x3D;0,10 表示只保留ks值在0到10之间的基因对。</p>
</li>
<li><p>multiple和blockinfo中的homo1,homo2,homo3,home4,homo5…列有关。一般默认为1, 表示选择homo1列用于后续的过滤。如果改成multiple&#x3D;2, 表示选择homo2</p>
</li>
<li><p>homo用于根据共线性中基因对的总体得分(取值范围为-1到1，值越高表明最佳匹配的基因对越多)对共线性区域进行过滤。当multiple&#x3D;1, homo&#x3D;-1,1时，表示根据homo1进行过滤，只保留取值范围在-1到1之间的共线性区。</p>
</li>
</ul>
<p>最后运行程序</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgdi -kp ath.conf</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-18a05fbb7f6746c1ac675b3d92797d48.png" alt="Ks频率分布图"></p>
<p>从图中，我们能够更加直观的观察到2个peak，基本确定2个多倍化事件。</p>
<p>既然得到了一个Ks的peak图，我们可以和另一款工具wgd的拟南芥分析结果进行对比。wgd的Ks值计算流程为，先进行所有蛋白之间的相互比对，根据基因之间的同源性进行聚类，然后构建系统发育树确定同源基因，最后调用PAML的codeml计算Ks，对应下图A的A.thaliana。如果存在参考基因组，那么根据共线性锚点（对应下图D）对Ks分布进行更新, 对应下图A的A.thaliana anchors。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-9d4ed8a3f6b648cd94e673cf95aeca7d.png" alt="wgd的分析结果"></p>
<p>同样是拟南芥的分析，wgd的点图（上图D）信息比较杂乱，存在较多的噪声点，而WGDI的Ks点阵图能有效的反应出不同共线性区域的Ks特点。wgd的Ks分布图中的只能看到一个比较明显的峰，而WGDI的分析结果能明显的观察到两个。wgd在Ks上存在的问题很大一部分原因是它们是直接采用旁系同源基因计算ks，容易受到串联重复基因积累的影响。而wgd则是基于共线性区计算Ks，结果更加可靠，尤其是后续还可以通过不断的调整参数，来确保Ks的峰值正确判断，这也是为什么在绘制Ks频率分布图的同时还要生成一个过滤后的文件。</p>
<blockquote>
<p>题外话: wgd是我在2019年学习WGD相关分析找到的一个流程工具，那个时候虽然也看了文章里面关于软件的细节介绍，但是由于对比较基因组学这一领域并不熟悉，所以也不好评判结果的可靠性。最近在学习wgdi时，一直和开发者反复讨论软件的一些参数细节，这才知道这里面的很多细节。这也警醒我，不能追求软件数量上的多，只求用软件跑出自己想要的结果发表文章，失去了科研的严谨性。</p>
</blockquote>
<p>我们会在下一节介绍如何利用Ks点阵图和Ks频率分布图更可靠的拟合Ks峰值。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/02/02/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%BA%8C)-comparative-genome-analysis-with-wgdi-2/" data-id="clm1xkn71007mdpniara3h0dk" data-title="如何用WGDI进行共线性分析(二)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何用WGDI进行共线性分析(一)-comparative-genome-analysis-with-wgdi-1" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/02/01/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%B8%80)-comparative-genome-analysis-with-wgdi-1/" class="article-date">
  <time class="dt-published" datetime="2021-02-01T13:57:32.701Z" itemprop="datePublished">2021-02-01</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/02/01/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%B8%80)-comparative-genome-analysis-with-wgdi-1/">如何用WGDI进行共线性分析(一)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>多倍化以及后续的基因丢失和二倍化现象存在于大部分的物种中, 是物种进化的重要动力。如果一个物种在演化过程中发生过多倍化，那么在基因组上就会存在一些共线性区域（即两个区域之间的基因是旁系同源基因，其基因的排布顺序基本一致）。</p>
<p>例如拟南芥经历了3次古多倍化，包括2次二倍化，一次3倍化 (Tang. et.al 2008 Science).  </p>
<blockquote>
<p>..For example,  Arabidopsis thaliana (thale cress) has undergone   three paleo-polyploidies, including two doublings  (5) and one tripling (12), resulting in ~12 copies of  its ancestral chromosome set in a ~160-Mb genome…</p>
</blockquote>
<p>当然之前只是记住了结论，现在我想的是，如何复现出这个分析结果呢？</p>
<h2 id="前期准备"><a href="#前期准备" class="headerlink" title="前期准备"></a>前期准备</h2><p>首先，我们需要使用conda安装wgdi。 我一般会新建一个环境避免新装软件和其他软件产生冲突</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 安装软件</span></span><br><span class="line">conda create -c bioconda -c conda-forge -n wgdi wgdi</span><br><span class="line"><span class="comment"># 启动环境</span></span><br><span class="line">conda activate wgdi </span><br></pre></td></tr></table></figure>

<p>之后，我们从ENSEMBL上下载基因组，CDS, 蛋白和GFF文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#Athaliana</span></span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz</span><br><span class="line">gunzip Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/arabidopsis_thaliana/cds/Arabidopsis_thaliana.TAIR10.cds.all.fa.gz</span><br><span class="line">gunzip Arabidopsis_thaliana.TAIR10.cds.all.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/arabidopsis_thaliana/pep/Arabidopsis_thaliana.TAIR10.pep.all.fa.gz</span><br><span class="line">gunzip Arabidopsis_thaliana.TAIR10.pep.all.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.44.gff3.gz</span><br><span class="line">gunzip Arabidopsis_thaliana.TAIR10.44.gff3.gz</span><br></pre></td></tr></table></figure>

<p>注意: cDNA和CDS并不相同，CDS只包括起始密码子到终止密码子之间的序列，而cDNA还会包括UTR.</p>
<h2 id="数据预处理"><a href="#数据预处理" class="headerlink" title="数据预处理"></a>数据预处理</h2><p>数据的预处理是用时最长的步骤，因为软件要求的输入格式并非是你手头所拥有的格式，你通常都需要进行一些转换才能得到它所需的形式。</p>
<p>wgdi需要三种信息，分别是BLAST, 基因的位置信息和染色体长度信息，要求格式如下</p>
<ol>
<li><p>BLAST 的 <code>-outfmt 6</code>输出的文件</p>
</li>
<li><p>基因的位置信息: 以tab分隔，分别为chr，id，start，end，strand，order，old_id。（并非真正意义上的GFF格式）</p>
</li>
<li><p>染色体长度信息和染色体上的基因个数，格式为 chr, length, gene number</p>
</li>
</ol>
<p>同时，对于每个基因我们只需要一个转录本，我通常使用最长的转录本作为该基因的代表。之前我写过一个脚本 (get_the_longest_transcripts.py) 提取每个基因的最长转录本，我在此处上写了一个新的脚本用来根据参考基因组和注释的GFF文件生成wgdi的两个输入文件，脚本地址为<a target="_blank" rel="noopener" href="https://github.com/xuzhougeng/myscripts/blob/master/comparative/generate_conf.py">https://github.com/xuzhougeng/myscripts/blob/master/comparative/generate_conf.py</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">pytho generate_conf.py -p ath Arabidopsis_thaliana.TAIR10.dna.toplevel.fa Arabidopsis_thaliana.TAIR10.44.gff3</span><br></pre></td></tr></table></figure>

<p>输出文件是ath.gff 和 ath.len.</p>
<p>由于ENSEMBL上GFF的ID编号和pep.fa 和cds.fa 不太一致，简单的说就是，编号之前中有一个 “gene:” 和 “transcript:” . 如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">head</span> ath.gff </span><br><span class="line">1  gene:AT1G01010  3631  5899  +  1  transcript:AT1G01010.1</span><br><span class="line">1  gene:AT1G01020  6788  9130  -  2  transcript:AT1G01020.1</span><br><span class="line">1  gene:AT1G01030  11649  13714  -  3  transcript:AT1G01030.1</span><br><span class="line">1  gene:AT1G01040  23416  31120  +  4  transcript:AT1G01040.2</span><br><span class="line">1  gene:AT1G01050  31170  33171  -  5  transcript:AT1G01050.1</span><br><span class="line">1  gene:AT1G01060  33379  37757  -  6  transcript:AT1G01060.3</span><br><span class="line">1  gene:AT1G01070  38444  41017  -  7  transcript:AT1G01070.1</span><br><span class="line">1  gene:AT1G01080  45296  47019  -  8  transcript:AT1G01080.2</span><br><span class="line">1  gene:AT1G01090  47234  49304  -  9  transcript:AT1G01090.1</span><br><span class="line">1  gene:AT1G01100  49909  51210  -  10  transcript:AT1G01100.2</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>于是我用 sed 删除了这些信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sed -i -e <span class="string">&#x27;s/gene://&#x27;</span> -e <span class="string">&#x27;s/transcript://&#x27;</span> ath.gff</span><br></pre></td></tr></table></figure>

<p>另外，pep.fa, cds.fa里面包含所有的基因，且命名特别的长， 同时我们也不需要基因中的 “.1”, “.2” 这部分信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">head</span> -n 5 Arabidopsis_thaliana.TAIR10.cds.all.fa </span><br><span class="line">&gt;AT3G05780.1 cds chromosome:TAIR10:3:1714941:1719608:-1 gene:AT3G05780 gene_biotype:protein_coding transcript_biotype:protein_coding gene_symbol:LON3 description:Lon protease homolog 3, mitochondrial [Source:UniProtKB/Swiss-Prot;Acc:Q9M9L8]</span><br><span class="line">ATGATGCCTAAACGGTTTAACACGAGTGGCTTTGACACGACTCTTCGTCTACCTTCGTAC</span><br><span class="line">TACGGTTTCTTGCATCTTACACAAAGCTTAACCCTAAATTCCCGCGTTTTCTACGGTGCC</span><br><span class="line">CGCCATGTGACTCCTCCGGCTATTCGGATCGGATCCAATCCGGTTCAGAGTCTACTACTC</span><br><span class="line">TTCAGGGCACCGACTCAGCTTACCGGATGGAACCGGAGTTCTCGCGATTTATTGGGTCGTb</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>借助于seqkit,  我对原来的数据进行了筛选和重命名</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">seqkit grep -f &lt;(<span class="built_in">cut</span> -f 7 ath.gff ) Arabidopsis_thaliana.TAIR10.cds.all.fa | seqkit <span class="built_in">seq</span> --id-regexp <span class="string">&quot;^(.*?)\\.\\d&quot;</span> -i &gt; ath.cds.fa</span><br><span class="line">seqkit grep -f &lt;(<span class="built_in">cut</span> -f 7 ath.gff ) Arabidopsis_thaliana.TAIR10.pep.all.fa | seqkit <span class="built_in">seq</span> --id-regexp <span class="string">&quot;^(.*?)\\.\\d&quot;</span> -i &gt; ath.pep.fa  </span><br></pre></td></tr></table></figure>

<p>通过NCBI的BLASTP 或者DIAMOND 进行蛋白之间相互比对，输出格式为 <code>-outfmt 6</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">makeblastdb -<span class="keyword">in</span> ath.pep.fa -dbtype prot</span><br><span class="line">blastp -num_threads 50 -db ath.pep.fa -query ath.pep.fa -outfmt 6 -evalue 1e-5 -num_alignments 20  -out ath.blastp.txt &amp; </span><br></pre></td></tr></table></figure>

<h2 id="共线性分析"><a href="#共线性分析" class="headerlink" title="共线性分析"></a>共线性分析</h2><h3 id="绘制点阵图"><a href="#绘制点阵图" class="headerlink" title="绘制点阵图"></a>绘制点阵图</h3><p>在上面步骤结束后，其实绘制点阵图就非常简单了，只需要创建配置文件，然后修改配置文件，最后运行wgdi。</p>
<blockquote>
<p>基本wgdi的其他分析也都是三部曲，创建配置，修改配置，运行程序。</p>
</blockquote>
<p>第一步，创建配置文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wgdi -d \? &gt; ath.conf</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>配置文件的信息如下，</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="section">[dotplot]</span></span><br><span class="line"><span class="attr">blast</span> = blast file</span><br><span class="line"><span class="attr">gff1</span> =  gff1 file</span><br><span class="line"><span class="attr">gff2</span> =  gff2 file</span><br><span class="line"><span class="attr">lens1</span> = lens1 file</span><br><span class="line"><span class="attr">lens2</span> = lens2 file</span><br><span class="line"><span class="attr">genome1_name</span> =  Genome1 name</span><br><span class="line"><span class="attr">genome2_name</span> =  Genome2 name</span><br><span class="line"><span class="attr">multiple</span>  = <span class="number">1</span>   <span class="comment"># 最好的同源基因数, 用输出结果中会用红点表示</span></span><br><span class="line"><span class="attr">score</span> = <span class="number">100</span>     <span class="comment"># blast输出的score 过滤 </span></span><br><span class="line"><span class="attr">evalue</span> = <span class="number">1</span>e-<span class="number">5</span>   <span class="comment"># blast输出的evalue 过滤 </span></span><br><span class="line"><span class="attr">repeat_number</span> = <span class="number">20</span>  <span class="comment"># genome2相对于genome1的最多同源基因数</span></span><br><span class="line"><span class="attr">position</span> = order</span><br><span class="line"><span class="attr">blast_reverse</span> = <span class="literal">false</span></span><br><span class="line"><span class="attr">ancestor_left</span> = none</span><br><span class="line"><span class="attr">ancestor_top</span> = none</span><br><span class="line"><span class="attr">markersize</span> = <span class="number">0.5</span>  <span class="comment"># 点的大小</span></span><br><span class="line"><span class="attr">figsize</span> = <span class="number">10</span>,<span class="number">10</span>   <span class="comment"># 图片大小</span></span><br><span class="line"><span class="attr">savefig</span> = savefile(.png,.pdf)</span><br></pre></td></tr></table></figure>

<p>第二步，修改配置文件。因为是自我比对，所以 gff1和gff2内容一样， lens1和lens2内容一样</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">[dotplot]</span><br><span class="line">blast = ath.blastp.txt</span><br><span class="line">gff1 =  ath.gff</span><br><span class="line">gff2 =  ath.gff</span><br><span class="line">lens1 = ath.len</span><br><span class="line">lens2 = ath.len</span><br><span class="line">genome1_name =  A. thaliana</span><br><span class="line">genome2_name =  A. thaliana</span><br><span class="line">multiple  = 1</span><br><span class="line">score = 100</span><br><span class="line">evalue = 1e-5</span><br><span class="line">repeat_number = 5 </span><br><span class="line">position = order</span><br><span class="line">blast_reverse = <span class="literal">false</span></span><br><span class="line">ancestor_left = none</span><br><span class="line">ancestor_top = none</span><br><span class="line">markersize = 0.5</span><br><span class="line">figsize = 10,10</span><br><span class="line">savefig = ath.dot.png</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>最后运行如下命令</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgdi -d ath.conf</span><br></pre></td></tr></table></figure>

<p>输出的png如下所示</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-aca03c4f647d421a9523c285ff8752c6.png" alt="点阵图"></p>
<p>从图中，我们很容易地就能观察到三种颜色，分别是红色，蓝色，和灰色。在WGDI中，红色表示genome2的基因在genome1中的最优同源（相似度最高）的匹配，次好的四个基因是蓝色，而剩余部分是为灰色。图中对角线出现的片段并非是自身比对，因为WGDI已经过滤掉自身比自身的结果。那么问题来了，这些基因是什么呢？这个问题会在后续的分析中进行解答。</p>
<p>如果基因组存在加倍事件，那么对于一个基因组的某个区域可能会存在另一个区域与其有着相似的基因（同源基因），并且这些基因的排布顺序较为一致。反应到点阵图上，就是能够观察到有多个点所组成的”线”。这里的线需要有一个引号，因为当你放大之后，你会发现其实还是点而已。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-799f63357317488d98484b761446b1ea.png" alt="部分区域放大"></p>
<p>由于进一步鉴定共线性区域就建立在这些”点”的基础上，那么影响这些点是否为同源基因的参数就非常重要了，即配置文件中的score,evalue,repeat_number。</p>
<h3 id="共线性分析和Ks计算"><a href="#共线性分析和Ks计算" class="headerlink" title="共线性分析和Ks计算"></a>共线性分析和Ks计算</h3><p>对于同线性(synteny)和共线性(collinearity)，一般认为同线性指的是两个区域有一定数量的同源基因，对基因顺序排布无要求。共线性是同线性的特殊形式，要求同源基因的排布顺序也相似。</p>
<p>wgdi开发了<code>-icl</code>模块（<strong>I</strong>mproved version of <strong>C</strong>olinear<strong>S</strong>can）用于进行共线性分析，使用起来非常简单，也是先建立配置文件。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgdi -icl \? &gt;&gt; ath.conf</span><br></pre></td></tr></table></figure>

<p>然后对配置文件进行修改.</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="section">[collinearity]</span></span><br><span class="line"><span class="attr">gff1</span> = ath.gff</span><br><span class="line"><span class="attr">gff2</span> = ath.gff</span><br><span class="line"><span class="attr">lens1</span> = ath.len</span><br><span class="line"><span class="attr">lens2</span> = ath.len</span><br><span class="line"><span class="attr">blast</span> = ath.blastp.txt</span><br><span class="line"><span class="attr">blast_reverse</span> = <span class="literal">false</span></span><br><span class="line"><span class="attr">multiple</span>  = <span class="number">1</span></span><br><span class="line"><span class="attr">process</span> = <span class="number">8</span></span><br><span class="line"><span class="attr">evalue</span> = <span class="number">1</span>e-<span class="number">5</span></span><br><span class="line"><span class="attr">score</span> = <span class="number">100</span></span><br><span class="line"><span class="attr">grading</span> = <span class="number">50</span>,<span class="number">40</span>,<span class="number">25</span></span><br><span class="line"><span class="attr">mg</span> = <span class="number">40</span>,<span class="number">40</span></span><br><span class="line"><span class="attr">pvalue</span> = <span class="number">0.2</span></span><br><span class="line"><span class="attr">repeat_number</span> = <span class="number">10</span></span><br><span class="line"><span class="attr">positon</span> = order</span><br><span class="line"><span class="attr">savefile</span> = ath.collinearity.txt</span><br></pre></td></tr></table></figure>

<p>其中的evalue, score和BLAST文件的过滤有关，用于筛选同源基因对。multiple确定最佳比对基因，repeat_number表示最多允许多少基因是潜在的同源基因, grading根据同源基因的匹配程度进行打分, mg 指的是共线性区域中所允许最大空缺基因数。</p>
<p>运行结果后，会得到ath.collinearity.txt，记录共线性区域。以 # 起始的行记录共线性区域的元信息，例如得分(score)，显著性(pvalue)，基因对数(N)等。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grep <span class="string">&#x27;^#&#x27;</span> ath.collinearity.txt | <span class="built_in">tail</span> -n 1</span><br><span class="line"><span class="comment"># Alignment 959: score=778.0 pvalue=0.0003 N=16 Pt&amp;Pt minus </span></span><br></pre></td></tr></table></figure>

<p>紧接着，我们就可以根据共线性结果来计算Ks。</p>
<p>同样的也是先生成配置文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgdi -ks \? &gt;&gt; ath.conf</span><br></pre></td></tr></table></figure>

<p>然后修改配置文件。我们需要提供cds和pep文件，共线性分析输出文件(支持MCScanX的共线性分析结果）。WGDI会用muscle根据蛋白序列进行联配，然后使用pal2pal.pl 基于cds序列将蛋白联配转成密码子联配，最后用paml中的yn00计算ka和ks。</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="section">[ks]</span></span><br><span class="line"><span class="attr">cds_file</span> =   ath.cds.fa</span><br><span class="line"><span class="attr">pep_file</span> =   ath.pep.fa</span><br><span class="line"><span class="attr">align_software</span> = muscle</span><br><span class="line"><span class="attr">pairs_file</span> = ath.collinearity.txt</span><br><span class="line"><span class="attr">ks_file</span> = ath.ks</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>ks计算需要一段时间，那么不妨在计算时了解下Ka和Ks。Ka和Ks分别指的是非同义替换位点数，和同义替换位点数。根据中性演化假说，基因的变化大多是中性突变，不会影响生物的生存，那么当一对同源基因分开的时间越早，不影响生存的碱基替换数就越多(Ks)，反之越多。</p>
<p>运行结果后，输出的Ks文件共有6列，对应的是每个基因对的Ka和Ks值。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">head</span> ath.ks</span><br><span class="line">id1  id2  ka_NG86  ks_NG86  ka_YN00  ks_YN00</span><br><span class="line">AT1G72300  AT1G17240  0.1404  0.629  0.1467  0.5718</span><br><span class="line">AT1G72330  AT1G17290  0.0803  0.5442  0.079  0.6386</span><br><span class="line">AT1G72350  AT1G17310  0.3709  0.9228  0.3745  1.071</span><br><span class="line">AT1G72410  AT1G17360  0.2636  0.875  0.2634  1.1732</span><br><span class="line">AT1G72450  AT1G17380  0.2828  1.1068  0.2857  1.5231</span><br><span class="line">AT1G72490  AT1G17400  0.255  1.2113  0.2597  2.0862</span><br><span class="line">AT1G72520  AT1G17420  0.1006  0.9734  0.1025  1.0599</span><br><span class="line">AT1G72620  AT1G17430  0.1284  0.7328  0.1375  0.603</span><br><span class="line">AT1G72630  AT1G17455  0.0643  0.7608  0.0613  1.2295</span><br></pre></td></tr></table></figure>

<p>鉴于共线性和Ks值输出结果信息量太大，不方便使用，更好的方法是将两者进行聚合，得到关于各个共线性区的汇总信息。</p>
<p>WGDI提供 <code>-bi</code> 参数帮助我们进行数据整合。同样也是生成配置文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgdi -bi ? &gt;&gt; ath.conf</span><br></pre></td></tr></table></figure>

<p>修改配置文件. 其中collinearity 和ks是我们前两步的输出文件，而 ks_col 则是声明使用ks文件里的哪一列。</p>
<figure class="highlight ini"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="section">[blockinfo]</span></span><br><span class="line"><span class="attr">blast</span> = ath.blastp.txt</span><br><span class="line"><span class="attr">gff1</span> =  ath.gff</span><br><span class="line"><span class="attr">gff2</span> =  ath.gff</span><br><span class="line"><span class="attr">lens1</span> = ath.len</span><br><span class="line"><span class="attr">lens2</span> = ath.len</span><br><span class="line"><span class="attr">collinearity</span> = ath.collinearity.txt</span><br><span class="line"><span class="attr">score</span> = <span class="number">100</span></span><br><span class="line"><span class="attr">evalue</span> = <span class="number">1</span>e-<span class="number">5</span></span><br><span class="line"><span class="attr">repeat_number</span> = <span class="number">20</span></span><br><span class="line"><span class="attr">position</span> = order</span><br><span class="line"><span class="attr">ks</span> = ath.ks</span><br><span class="line"><span class="attr">ks_col</span> = ks_NG86</span><br><span class="line"><span class="attr">savefile</span> = ath_block_information.csv</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>运行即可</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wgdi -bi ath.conf</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>输出文件以csv格式进行存放，因此可以用EXCEL直接打开，共11列。</p>
<ol>
<li><code>id</code> 即共线性的结果的唯一标识</li>
<li><code>chr1</code>,<code>start1</code>,<code>end1</code> 即参考基因组（点图的左边）的共线性范围</li>
<li><code>chr2</code>,<code>start2</code>,<code>end2</code> 即参考基因组（点图的上边）的共线性范围</li>
<li><code>pvalue</code> 即共线性结果评估，常常认为小于0.01的更合理些</li>
<li><code>length</code> 即共线性片段长度</li>
<li><code>ks_median</code> 即共线性片段上所有基因对<code>ks</code>的中位数（主要用来评判ks分布的）</li>
<li><code>ks_average</code> 即共线性片段上所有基因对<code>ks</code>的平均值</li>
<li><code>homo1</code>,<code>homo2</code>,<code>homo3</code>,<code>homo4</code>,<code>homo5</code>与<code>multiple</code>参数有关，即一共有homo+multiple列</li>
</ol>
<p>  主要规则是：基因对如果在点图中为红色，赋值为1，蓝色赋值为0，灰色赋值为-1（颜色参考前述 <code>wgdi</code> 点图 相关推文）。共线性片段上所有基因对赋值后求平均值，这样就赋予该共线性一个-1,1的值。如果共线性的点大部分为红色，那么该值接近于1。可以作为共线性片段的筛选。</p>
<ol>
<li><code>block1</code>,<code>block2</code>分别为共线性片段上基因<code>order</code>的位置。</li>
<li><code>ks</code>共线性片段的上基因对的<code>ks</code>值</li>
<li><code>density1</code>,<code>density2</code> 共线性片段的基因分布密集程度。值越小表示稀疏</li>
</ol>
<p>这个表格是后续探索分析的一个重点，比如说我们可以用它来分析我们点图中出现在对角线上的基因。下面的代码就是提取了第一个共线性区域的所有基因，然后进行富集分析，绘制了一个气泡图。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">df <span class="operator">&lt;-</span> read.csv<span class="punctuation">(</span><span class="string">&quot;ath_block_information.csv&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">tandem_length <span class="operator">&lt;-</span> 200</span><br><span class="line"></span><br><span class="line">df<span class="operator">$</span>start <span class="operator">&lt;-</span> df<span class="operator">$</span>start2 <span class="operator">-</span> df<span class="operator">$</span>start1</span><br><span class="line">df<span class="operator">$</span>end <span class="operator">&lt;-</span> df<span class="operator">$</span>end2 <span class="operator">-</span> df<span class="operator">$</span>end1</span><br><span class="line"></span><br><span class="line">df_tandem <span class="operator">&lt;-</span> df<span class="punctuation">[</span><span class="built_in">abs</span><span class="punctuation">(</span>df<span class="operator">$</span>start<span class="punctuation">)</span> <span class="operator">&lt;=</span> tandem_length <span class="operator">|</span></span><br><span class="line">     <span class="built_in">abs</span><span class="punctuation">(</span>df<span class="operator">$</span>end<span class="punctuation">)</span> <span class="operator">&lt;=</span> tandem_length<span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">gff <span class="operator">&lt;-</span> read.table<span class="punctuation">(</span><span class="string">&quot;ath.gff&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">syn_block1 <span class="operator">&lt;-</span>  df_tandem<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;block1&quot;</span><span class="punctuation">,</span> <span class="string">&quot;block2&quot;</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">gene_order <span class="operator">&lt;-</span> unlist<span class="punctuation">(</span></span><br><span class="line">  lapply<span class="punctuation">(</span>syn_block1<span class="punctuation">,</span> strsplit<span class="punctuation">,</span> split<span class="operator">=</span><span class="string">&quot;_&quot;</span><span class="punctuation">,</span> fixed<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">,</span> useBytes<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">gene_order <span class="operator">&lt;-</span> unique<span class="punctuation">(</span><span class="built_in">as.numeric</span><span class="punctuation">(</span>gene_order<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">syn_block1_gene <span class="operator">&lt;-</span> gff<span class="punctuation">[</span> gff<span class="operator">$</span>V6 <span class="operator">%in%</span> gene_order<span class="punctuation">,</span> <span class="string">&quot;V2&quot;</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#BiocManager::install(&quot;org.At.tair.db&quot;)</span></span><br><span class="line">library<span class="punctuation">(</span>org.At.tair.db<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>clusterProfiler<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ego <span class="operator">&lt;-</span> enrichGO<span class="punctuation">(</span>syn_block1_gene<span class="punctuation">,</span> </span><br><span class="line">                OrgDb <span class="operator">=</span> org.At.tair.db<span class="punctuation">,</span> </span><br><span class="line">                keyType <span class="operator">=</span> <span class="string">&quot;TAIR&quot;</span><span class="punctuation">,</span>ont <span class="operator">=</span> <span class="string">&quot;BP&quot;</span></span><br><span class="line">                <span class="punctuation">)</span></span><br><span class="line">dotplot<span class="punctuation">(</span>ego<span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/02/image-4a707a066ab745c2b566d55ab28b765d.png" alt="气泡图"></p>
<p>进一步观察这些基因，可以发现这些基因的编号都是前后相连，说明这个基因簇和花粉识别有一定关系。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> as.data.frame<span class="punctuation">(</span>ego<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line">                   ID           Description GeneRatio  BgRatio       pvalue</span><br><span class="line">GO<span class="operator">:</span><span class="number">0048544</span> GO<span class="operator">:</span><span class="number">0048544</span> recognition of pollen    <span class="number">11</span><span class="operator">/</span><span class="number">560</span> <span class="number">43</span><span class="operator">/</span><span class="number">21845</span> <span class="number">7.794138e-09</span></span><br><span class="line">               p.adjust      qvalue</span><br><span class="line">GO<span class="operator">:</span><span class="number">0048544</span> <span class="number">8.121073e-06</span> <span class="number">7.95418e-06</span></span><br><span class="line">                                                                                                                  geneID</span><br><span class="line">GO<span class="operator">:</span><span class="number">0048544</span> AT1G61360<span class="operator">/</span>AT1G61370<span class="operator">/</span>AT1G61380<span class="operator">/</span>AT1G61390<span class="operator">/</span>AT1G61400<span class="operator">/</span>AT1G61420<span class="operator">/</span>AT1G61440<span class="operator">/</span>AT1G61480<span class="operator">/</span>AT1G61490<span class="operator">/</span>AT1G61500<span class="operator">/</span>AT1G61550</span><br><span class="line">           Count</span><br><span class="line">GO<span class="operator">:</span><span class="number">0048544</span>    <span class="number">11</span></span><br></pre></td></tr></table></figure>

<blockquote>
<p>A <strong>gene cluster</strong> is a group of two or more genes found within an organism’s DNA that encode similar polypeptides, or proteins, which collectively share a generalized function and are **often located within a few thousand base pairs **of each other. –<a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Gene_cluster">维基百科</a></p>
</blockquote>
<p>接下来，我们还可以根据这个表格中的Ks对点阵图进行上色， 以及正确的计算Ks峰值。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/02/01/%E5%A6%82%E4%BD%95%E7%94%A8WGDI%E8%BF%9B%E8%A1%8C%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90(%E4%B8%80)-comparative-genome-analysis-with-wgdi-1/" data-id="clm1xkn70007fdpni33wlf9e2" data-title="如何用WGDI进行共线性分析(一)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/7/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/6/">6</a><a class="page-number" href="/page/7/">7</a><span class="page-number current">8</span><a class="page-number" href="/page/9/">9</a><a class="page-number" href="/page/10/">10</a><span class="space">&hellip;</span><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/9/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/09/">九月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/09/02/%E5%A6%82%E4%BD%95%E8%AE%A9%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8D%E5%86%8D%E9%BE%9F%E9%80%9F%E4%B8%8B%E8%BD%BDgithub%E7%9A%84%E6%95%B0%E6%8D%AE/">如何让服务器不再龟速下载github的数据</a>
          </li>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>