<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>如何对基因组序列进行注释 | 洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="基因组组装完成后，或者是完成了草图，就不可避免遇到一个问题，需要对基因组序列进行注释。注释之前首先得构建基因模型，有三种策略：  从头注释(de novo prediction)：通过已有的概率模型来预测基因结构，在预测剪切位点和UTR区准确性较低 同源预测(homology-based prediction)：有一些基因蛋白在相近物种间的保守型搞，所以可以使用已有的高质量近缘物种注释信息通过序列">
<meta property="og:type" content="article">
<meta property="og:title" content="如何对基因组序列进行注释">
<meta property="og:url" content="http://xuzhougeng.top/2019/10/11/%E5%A6%82%E4%BD%95%E5%AF%B9%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%BA%8F%E5%88%97%E8%BF%9B%E8%A1%8C%E6%B3%A8%E9%87%8A-How-to-annotate-your-genome/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:description" content="基因组组装完成后，或者是完成了草图，就不可避免遇到一个问题，需要对基因组序列进行注释。注释之前首先得构建基因模型，有三种策略：  从头注释(de novo prediction)：通过已有的概率模型来预测基因结构，在预测剪切位点和UTR区准确性较低 同源预测(homology-based prediction)：有一些基因蛋白在相近物种间的保守型搞，所以可以使用已有的高质量近缘物种注释信息通过序列">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-8c9601ef6651e157-2c2a2743d4d64b31a902dbaf23b62ac5.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-f5472a3dada24b75-4acf9e3d46e44f58a96f8843d10f5458.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-b5a07df1dedaa354-e340ae1c7a7040049ac93c04d89b2a49.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-82b4f41b84eba54b-2a776bfefb084c85a2b9c2d75905c4a3.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-d7184dedba04ef7b-8f36de4ffea447dea61de1765882f4e8.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-693396ba4693efa0-9a4534b6b3924f3284a117698b926893.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-0200855851b844ce-bb010eed70444f82ab494eb706e92087.png">
<meta property="og:image" content="http://xuzhougeng.top/assets/2013053-b5a07df1dedaa354.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-562fefad5dec9b07-6a137a307f494a6385192c6d270ef644.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-da3cb7a46b5327a5-38978a3654cd461e96b6208b54c8fef6.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-65f7c80123c3f586-b52d6c6326d14440886443690aabbdef.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-2bb888cdb1983838-84e79099d9a14a70a7e3c37ed3508c13.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-d30be93f484be77d-28a3dcd689154b4caf797a2dcf0ea2fe.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-e86990227c77f511-03651bbbee1b4847a35bb983407dedfc.png">
<meta property="article:published_time" content="2019-10-11T14:56:46.108Z">
<meta property="article:modified_time" content="2019-10-11T14:56:46.108Z">
<meta property="article:author" content="xuzhougeng">
<meta property="article:tag" content="注释 | 重复序列">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-8c9601ef6651e157-2c2a2743d4d64b31a902dbaf23b62ac5.png">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main"><article id="post-如何对基因组序列进行注释-How-to-annotate-your-genome" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/11/%E5%A6%82%E4%BD%95%E5%AF%B9%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%BA%8F%E5%88%97%E8%BF%9B%E8%A1%8C%E6%B3%A8%E9%87%8A-How-to-annotate-your-genome/" class="article-date">
  <time class="dt-published" datetime="2019-10-11T14:56:46.108Z" itemprop="datePublished">2019-10-11</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 class="p-name article-title" itemprop="headline name">
      如何对基因组序列进行注释
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>基因组组装完成后，或者是完成了草图，就不可避免遇到一个问题，需要对基因组序列进行注释。注释之前首先得构建基因模型，有三种策略：</p>
<ul>
<li>从头注释(<em>de novo</em> prediction)：通过已有的概率模型来预测基因结构，在预测剪切位点和UTR区准确性较低</li>
<li>同源预测(homology-based prediction)：有一些基因蛋白在相近物种间的保守型搞，所以可以使用已有的高质量近缘物种注释信息通过序列联配的方式确定外显子边界和剪切位点</li>
<li>基于转录组预测(transcriptome-based prediction)：通过物种的RNA-seq数据辅助注释，能够较为准确的确定剪切位点和外显子区域。</li>
</ul>
<p>每一种方法都有自己的优缺点，所以最后需要用EvidenceModeler(EVM)和GLEAN工具进行整合，合并成完整的基因结构。基于可靠的基因结构，后续可才是功能注释，蛋白功能域注释，基因本体论注释，通路注释等。</p>
<p>那么基因注释重要吗？可以说是非常重要了，尤其是高通量测序非常便宜的现在。你可以花不到一万的价格对600M的物种进行100X的普通文库测序，然后拼接出草图。但是这个草图的价值还需要你进行注释后才能显现出来。有可能你和诺贝尔奖就差一个注释的基因组。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-8c9601ef6651e157-2c2a2743d4d64b31a902dbaf23b62ac5.png" alt="基因注释的重要性"></p>
<h2 id="从案例中学习套路"><a href="#从案例中学习套路" class="headerlink" title="从案例中学习套路"></a>从案例中学习套路</h2><h3 id="陆地棉基因组注释"><a href="#陆地棉基因组注释" class="headerlink" title="陆地棉基因组注释"></a>陆地棉基因组注释</h3><p>文章标题为“Sequencing of allotetraploid cotton (Gossypium hirsutum L. acc. TM-1) provides a resource for fiber improvement”.</p>
<p><strong>同源注释</strong>：从Phytozome上下载了7个植物的基因组蛋白序列(Arabidopsis thaliana, Carica papaya, Glycine max, G. raimondii, Populus trichocarpa, Theobroma cacao and Vitis vinifera), 使用 <em>TblastN</em> 将蛋白序列比对到组装序列上，E-value的阈值为1e-5. 将不同蛋白的BLAST的hits用 <em>Solar</em> 软件进行合并。<em>GeneWise</em> 根据每个BLAST hit的对应基因区域预测完整的基因结构。</p>
<p><strong>从头预测</strong>：先得构建repeat-mask genome， 在这个基础上就用 <em>August</em>, <em>Genescan</em>, <em>GlimmerHMM</em>, <em>Geneid</em> 和 <em>SNAP</em> 预测编码区</p>
<p><strong>转录组预测</strong>：用Tophat将RNA-seq数据比对到组装序列上，然后用cufflinks组装转录本形成基因模型。</p>
<p>综上，使用 <em>EvidenceModeler(EVM)</em> 将上面的结果组装成非冗余的基因结构。进一步根据Cscore &gt; 0.5，peptide coverage &gt; 0.5 和CDS overlaping with TE进行筛选。还有过滤掉超过30%编码区被Pfam或Interprot TE domain的注释的基因模型。</p>
<p>这些基因模型使用BLASTP进行功能注释，所用数据库为SWiss-Prot和TrEMBL.蛋白功能使用InterProScan和HMMER注释，数据库为InterPro和Pfam。GO注释则是直接雇佣InterPro和Pfam注释得到的对应entry。通路注释使用KEGG数据库。</p>
<h3 id="Cardamine-hirsuta基因组注释"><a href="#Cardamine-hirsuta基因组注释" class="headerlink" title="Cardamine hirsuta基因组注释"></a>Cardamine hirsuta基因组注释</h3><p>文章标题为“The Cardamine hirsuta genome offers insight into the evolution of morphological diversity”。</p>
<p><strong>同源注释</strong>：使用 <em>GenomeThreader</em> 以拟南芥为剪切模型，以及PlantsGDB resourc上 <em>Brassica rapa</em> (v1.1), <em>A. thaliana</em>(TAIR10), <em>A. lyrata</em> (v6), <em>tomato</em> (v3.6), <em>poplar</em> (v2) 和 <em>A. thaliana</em> (version PUT-169), <em>B. napus</em> (version PUT-172) EST assemblies 的完整的代表性蛋白集。</p>
<p><strong>转录本预测</strong>： 将 <em>C. hirsuta</em> RNA-seq数据比对到基因序列，然后用cufflinks拼接</p>
<p><strong>从头预测</strong>：转录本预测得到的潜在蛋白编码转录本使用网页工具 <em>ORFpredictor</em> 进行预测， 同时用 <em>blastx</em> 和 <em>A. thalina</em> 进行比较，选择90%序列相似度和最高5%长度差异的部分从而保证保留完整的编码框(有启动子和终止子)。 这些基因模型根据相互之间的相似度和重叠度进行聚类，高度相似(&gt;95)从聚类中剔除，保证非冗余训练集。为了训练gene finder, 它们选随机选取了2000个位点，20%是单个外显子基因。从头预测工具为 <em>August</em> , <em>GlimmerHMM</em>, <em>Geneid</em> 和 <em>SNAP</em> . 此外还用了Fgenesh+, 以双子叶特异矩阵为参数进行预测。</p>
<p>最后使用JIGSAW算法根据以上结果进行训练，随后再次用JIGSAW对每个基因模型计算统计学权重。</p>
<p>可变剪切模型则是基于苗、叶、花和果实的RNA-seq比对组装结果。</p>
<p>GO注释使用<a target="_blank" rel="noopener" href="https://github.com/groupschoof/AHRD/">AHRD流程</a></p>
<h3 id="小结"><a href="#小结" class="headerlink" title="小结"></a>小结</h3><p>举的2个例子都是植物，主要是植物基因组不仅是组装，注释都是一大难题。因为植物基因组有大量的重复区，假基因，还有很多新的蛋白编码基因和非编码基因，比如说玉米基因组80%以上都是重复区域。然后当我检索这两篇文章所用工具的时候，我不经意或者说不可避免就遇到了这个网站 <a target="_blank" rel="noopener" href="http://www.plantgdb.org/">http://www.plantgdb.org/</a> , 一个整合植物基因组学工具和资源的网站，但是这个网站似乎2年没有更新了。当然这个网站也挺不错,<a target="_blank" rel="noopener" href="http://bioservices.usd.edu/gsap.html">http://bioservices.usd.edu/gsap.html</a>, 他给出了一套完整的注释流程以及每一步的输入和输出情况。</p>
<p>此外，2017年在《Briefings in Bioinformatics》发表的”Plant genome and transcriptome annotations: from misconceptions to simple solution” 则是从五个角度对植物基因组注释做了很完整的总结</p>
<ul>
<li>植物科学的常见本体</li>
<li>功能注释的常用数据库和资源</li>
<li>已注释的植物基因组意味着什么</li>
<li>一个自动化注释流程</li>
<li>一个参考流程图，用来说明使用公用数据库注释植物基因组&#x2F;转录组的常规步骤</li>
</ul>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-f5472a3dada24b75-4acf9e3d46e44f58a96f8843d10f5458.png" alt="注释流程图"></p>
<p>以上，通过套路我们对整个基因组注释有一个大概的了解，后续就需要通过实际操作来理解细节。</p>
<h2 id="基因组注释"><a href="#基因组注释" class="headerlink" title="基因组注释"></a>基因组注释</h2><p>当我们谈到基因注释的时候，我们通常认为注释是指“对基因功能的描述”，比如说A基因在细胞的那个部分，通过招募B来调控C，从而引起病变。但是基因结构也是注释的一种形式，而且是先决条件，也就是在看似随机的ATCG的碱基排列中找到特殊的部分，而这些特殊的区域有着不一样的功能。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-b5a07df1dedaa354-e340ae1c7a7040049ac93c04d89b2a49.png" alt="gene structure"></p>
<p>在正式启动基因组注释项目之前，需要先检查组装是否合格，比如contig N50的长度是否大于基因的平均长度，使用BUSCO&#x2F;CEGMA检查基因的完整性，如果不满足要求，可能输出结果中大部分的contig中都不存在一个完整的基因结构。当组装得到的contig符合要求时，就可以开始基因组注释环节，这一步分为三步：基因结构预测，基因功能注释，可视化和质控。</p>
<h3 id="基因组结构注释"><a href="#基因组结构注释" class="headerlink" title="基因组结构注释"></a>基因组结构注释</h3><p>基因结构注释应是功能注释的先决条件，完整的真核生物基因组注释流程需要如下步骤：</p>
<ol>
<li>必要的基因组重复序列屏蔽</li>
<li>从头寻找基因, 可用工具为: GeneMarkHMM, FGENESH, Augustus, SNAP, GlimmerHMM, Genscan</li>
<li>同源蛋白预测, 内含子分析: GeneWIse, Exonerate, GenomeThreader</li>
<li>将EST序列，全长cDNA序列和Trinity&#x2F;Cufflinks&#x2F;Stringtie组装的转录组和基因组联配</li>
<li>如果第4步用到了多个数据来源，使用PASA基于重叠情况进行联配</li>
<li>使用EvidenceModler根据上述结果进行整合</li>
<li>使用PASA更新EVM的一致性预测，增加UTR注释和可变剪切注释</li>
<li>必要的人工检查</li>
</ol>
<p>基本上是套路化的分析流程，也就有一些工具通过整合几步开发了流程管理工具，比如说BRAKER结合了GeneMark和Augustus，MAKER2整合了SNAP,Exonerate，虽然BRAKER说自己的效果比MAKER2好，但是用的人似乎不多，根据web of knowledge统计，两者的引用率分别是44,283, 当然BRAKER是2016，MAKER2是2011，后者在时间上有优势。</p>
<p>这里准备先按部就班的按照流程进行注释，所用的数据是 <em>Cardamine hirsuta</em> , 数据下载方式如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Cardamine hirsutat基因组数据</span></span><br><span class="line"><span class="built_in">mkdir</span> chi_annotation &amp;&amp; <span class="built_in">cd</span> chi_annotation</span><br><span class="line">wget http://chi.mpipz.mpg.de/download/sequences/chi_v1.fa</span><br><span class="line"><span class="built_in">cat</span> chi_v1.fa | <span class="built_in">tr</span> <span class="string">&#x27;atcg&#x27;</span> <span class="string">&#x27;ATCG&#x27;</span> &gt; chi_unmasked.fa</span><br><span class="line"><span class="comment"># 注释结果</span></span><br><span class="line">wget http://chi.mpipz.mpg.de/download/annotations/carhr38.gff</span><br><span class="line"><span class="comment"># Cardamine hirsutat转录组数据</span></span><br><span class="line"><span class="built_in">mkdir</span> rna-seq &amp;&amp; <span class="built_in">cd</span> rna-seq</span><br><span class="line">wget -4 -q -A <span class="string">&#x27;*.fastq.gz&#x27;</span> -np -nd -r 2 http://chi.mpipz.mpg.de/download/fruit_rnaseq/cardamine_hirsuta/ &amp;</span><br><span class="line">wget -4 -q -A <span class="string">&#x27;*.fastq.gz&#x27;</span> -np -nd -r 2 http://chi.mpipz.mpg.de/download/leaf_rnaseq/cardamine_hirsuta/ &amp;</span><br></pre></td></tr></table></figure>

<p>软件安装不在正文中出现，会放在附录中，除了某些特别复杂的软件。</p>
<h4 id="01-重复序列屏蔽"><a href="#01-重复序列屏蔽" class="headerlink" title="01-重复序列屏蔽"></a>01-重复序列屏蔽</h4><p><strong>重复屏蔽</strong>：真核生物的基因组存在大量的重复序列，植物基因组的重复序列甚至可以高达80%。尽管重复序列对维持染色体的空间结构、基因的表达调控、遗传重组等都具有重要作用，但是却会导致BLAST的结果出现大量假阳性，增加基因结构的预测的计算压力甚至影响注释正确性。基因组中的重复按照序列特征可以分为两类：串联重复(tandem repeats)和散在重复(interspersed repeats).</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-82b4f41b84eba54b-2a776bfefb084c85a2b9c2d75905c4a3.png" alt="人类中的重复序列划分"></p>
<p>鉴定基因组重复区域的方法有两种：一种基于文库(library)的同源(homology)方法，该文库收集了其他物种的某一种重复的一致性序列，通过相似性来鉴定重复；另一种是从头预测(<em>de novo</em>)，将序列和自己比较或者是高频K-mer来鉴定重复。</p>
<p>目前重复序列注释主要软件就是RepeatMasker和RepeatModel。这里要注意分析的fasta的ID不能过长，不然会报错。如果序列ID过长可以使用bioawk进行转换，后续用到RepatModel不支持多行存放序列的fasta格式。</p>
<p>直接使用同源注释工具RepeatMasker寻找重复序列：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> 00-RepeatMask</span><br><span class="line">~/opt/biosoft/RepeatMasker/RepeatMasker -e ncbi -species arabidopsis -pa 40 -gff -<span class="built_in">dir</span> 00-RepeatMask/ chi_unmasked.fa</span><br><span class="line"><span class="comment"># -e ncbi</span></span><br><span class="line"><span class="comment"># -species 选择物种 用~/opt/biosoft/RepeatMasker/util/queryRepeatDatabase.pl -tree 了解</span></span><br><span class="line"><span class="comment"># -lib 增加额外数据库,</span></span><br><span class="line"><span class="comment"># -pa 并行计算</span></span><br><span class="line"><span class="comment"># -gff 输出gff注释</span></span><br><span class="line"><span class="comment"># -dir 输出路径</span></span><br><span class="line"><span class="comment"># annotation with the library produced by RepeatModel</span></span><br></pre></td></tr></table></figure>

<p>输出结果中主要关注如下三个(其中xxx表示一类文件名)</p>
<ul>
<li>xxx.fa.masked, 将重复序列用N代替</li>
<li>xxx.fa.out.gff, 以gff2形式存放重复序列出现的位置</li>
<li>xxx.fa.tbl, 该文件记录着分类信息</li>
</ul>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cat</span> 00-RepeatMask/chi_unmasked.fa.tbl</span><br><span class="line">==================================================</span><br><span class="line">file name: chi_unmasked.fa</span><br><span class="line">sequences:           624</span><br><span class="line">total length:  198654690 bp  (191241357 bp excl N/X-runs)</span><br><span class="line">GC level:         35.24 %</span><br><span class="line">bases masked:   35410625 bp ( 17.83 %)</span><br><span class="line">==================================================</span><br></pre></td></tr></table></figure>

<p>也就是说该物种198M中有将近18%的重复序列，作为参考，拟南芥125Mb 14%重复序列, 水稻389M，36%重复，人类基因组是3G，50%左右的重复序列。</p>
<p>使用最后的<code>chi_unmasked.fa.masked</code>用于下一步的基因结构预测。</p>
<p>注：当然也可以用RepeatModel进行从头预测，得到的预测结果后续可以整合到RepeatMasker</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># de novo predict</span></span><br><span class="line">~/opt/biosoft/RepeatModeler-open-1.0.11/BuildDatabase -name <span class="built_in">test</span> -engine ncbi output.fa</span><br><span class="line">~/opt/biosoft/RepeatModeler-open-1.0.11/RepeatModeler -database <span class="built_in">test</span></span><br></pre></td></tr></table></figure>

<p>这一步速度极其慢，由于我们的目的只是获取屏蔽后序列降低后续从头预测的压力，所以可以先不做这一步。在后续分析重复序列在基因组进化上的作用时可以做这一步。下</p>
<blockquote>
<p>如果从头预测的结果与同源预测的结果有30%以上的overlap，并且分类不一致，会把从头预测的结果过滤掉。从头预测与同源预测结果有overlap，但是分类一致的，都会保留。但是统计的时候不会重复统计。</p>
</blockquote>
<h4 id="02-从头-ab-initio-预测基因"><a href="#02-从头-ab-initio-预测基因" class="headerlink" title="02-从头(ab initio)预测基因"></a>02-从头(ab initio)预测基因</h4><h5 id="基于已有模型或无监督训练"><a href="#基于已有模型或无监督训练" class="headerlink" title="基于已有模型或无监督训练"></a>基于已有模型或无监督训练</h5><p>目前的从头预测软件大多是基于HMM(隐马尔科夫链)和贝叶斯理论，通过已有物种的注释信息对软件进行训练，从训练结果中去推断一段基因序列中可能的结构，在这方面做的最好的工具是<strong>AUGUSTUS</strong> 它可以仅使用序列信息进行预测，也可以整合EST, cDNA, RNA-seq数据作为先验模型进行预测。</p>
<p>AUGUSTUS的无root安装比较麻烦，我折腾了好几天最后卒，不过辛亏有bioconda，<code>conda create -n annotation augustus=3.3</code>.</p>
<p>它的使用看起来很简单，我们可以尝试使用一段拟南芥已知的基因序列让其预测，比如前8k序列</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">seqkit faidx TAIR10.fa Chr1:1-8000 &gt; test.fa</span><br><span class="line">augustus --speices=arabidopsis test.fa &gt; test.gff</span><br></pre></td></tr></table></figure>

<p>如果仅仅看两者的CDS区，结果完全一致，相当于看过一遍参考答案去做题目，题目都做对了。</p>
<blockquote>
<p>注:已经被训练的物种信息可以用<code>augustus --species=help</code>查看。</p>
</blockquote>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-d7184dedba04ef7b-8f36de4ffea447dea61de1765882f4e8.png" alt="结果比较"></p>
<p>在<strong>不使用RNA-seq数据</strong>的情况下，可以基于拟南芥的训练模型进行预测，采用下面的方式多条染色体并行augustus</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> 01-augustsus &amp;&amp; <span class="built_in">cd</span> 01-augustsus</span><br><span class="line"><span class="built_in">ln</span> ../00-RepeatMask/chi_unmasked.fa.masked genome.fa</span><br><span class="line">seqkit <span class="built_in">split</span> genome.fa <span class="comment">#结果文件在genome.fa.split</span></span><br><span class="line">find genome.fa.split/ -<span class="built_in">type</span> f -name <span class="string">&quot;*.fa&quot;</span> | parallel -j 30 augustus --species=arabidopsis --gff3=on &gt;&gt; temp.gff <span class="comment">#并行处理</span></span><br><span class="line">join_aug_pred.pl &lt; temp.gff  | grep -v <span class="string">&#x27;^#&#x27;</span> &gt; temp.joined.gff</span><br><span class="line">bedtools <span class="built_in">sort</span> -i temp.joined.gff &gt; augustsus.gff</span><br></pre></td></tr></table></figure>

<p>AUGUSTUS依赖于已有的模型，而<strong>GeneMark-ES&#x2F;ET</strong>则是唯一一款支持无监督训练模型，之后再识别真核基因组蛋白编码区的工具。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gmes_petap.pl --ES --sequence genome.fa --cores 50</span><br></pre></td></tr></table></figure>

<p>最后得到的是genemark.gtf，是标准的GTF格式，可以使用Sequence Ontology Project提供的gtf2gff3.pl进行转换</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">wget http://genes.mit.edu/burgelab/miso/scripts/gtf2gff3.pl</span><br><span class="line"><span class="built_in">chmod</span> 755 gtf2gff3.pl</span><br><span class="line">gtf2gff3.pl genemark.gtf | bedtools <span class="built_in">sort</span> -i - &gt; genemark.gff</span><br></pre></td></tr></table></figure>

<p>不同从头预测软件的实际效果可以通过在IGV中加载文章提供的gff文件和预测后的gff文件进行比较，一般会存在如下几个问题：</p>
<ul>
<li>基因多了，或者少了，也就是假阳性和假阴性现象</li>
<li>UTR区域难以预测，这个比较正常</li>
<li>未正确识别可变剪切位点，导致前后几个基因识别成一个基因</li>
</ul>
<p>考虑到转录组测序已经非常便宜，可以通过该物种的RNA-seq提供覆盖度信息进行预测。</p>
<h5 id="基于转录组数据预测"><a href="#基于转录组数据预测" class="headerlink" title="基于转录组数据预测"></a>基于转录组数据预测</h5><p>根据已有的模型或者自训练可以正确预测很大一部分的基因，但如果需要提高预测的正确性，还需要额外的信息。在过去就需要提供物种本身的cDNA, EST，而现在更多的是基于转录组序列进行训练。尽管RNA-seq数据在基因组上的比对情况能够推测出内含子位置，根据覆盖度可以推测出外显子和非编码区的边界，但是仅仅依赖于RNA-seq的覆盖不能可信地推测出蛋白编码区(Hoff K.J. Stanke M. 2015).</p>
<p>AUGUSTUS可以利用转录组比对数据中的位置信息来训练模型，GeneMark-ET可以利用RNA-seq得到的内含子位点信息自我训练HMM参数，进行基因预测。BRAKER2将两者进行整合，使用GeneMark-ET根据RNA-seq无监督训练模型寻找基因，然后用AUGUSTUS进行模型训练，最后完成基因预测</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-693396ba4693efa0-9a4534b6b3924f3284a117698b926893.png" alt="BRAKER流程"></p>
<p>首先使用hisat2根据屏蔽后的参考序列建立索引，进行比对。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 项目根目录</span></span><br><span class="line"><span class="built_in">mkdir</span> index</span><br><span class="line">hisat2-build 01-augustus/genome.fa index/chi_masked</span><br><span class="line">hisat2 -p 20 -x index/chi_masked -1 rna-seq/leaf_ox_r1_1.fastq.gz -2 rna-seq/leaf_ox_r1_2.fastq.gz | samtools <span class="built_in">sort</span> -@ 10 &gt; 02-barker/leaf_ox_r1.bam &amp;</span><br><span class="line">isat2 -p 20 -x index/chi_masked -1 rna-seq/ox_flower9_rep1_1.fastq.gz -2 rna-seq/ox_flower9_rep1_2.fastq.gz | samtools <span class="built_in">sort</span> -@ 10 &gt; 02-barker/ox_flower9.bam &amp;</span><br><span class="line">hisat2 -p 20 -x index/chi_masked -1 rna-seq/ox_flower16_rep1_1.fastq.gz -2 rna-seq/ox_flower16_rep1_2.fastq.gz | samtools <span class="built_in">sort</span> -@ 10 &gt; 02-barker/ox_flower16.bam &amp;</span><br></pre></td></tr></table></figure>

<p>然后，以未屏蔽重复序列的参考序列和BAM文件作为输入，让BRAKER2（安装会稍显麻烦，因为依赖许多软件）进行预测。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">braker.pl --gff3 --cores 50 --species=carhr --genome=chi_unmasked.fa --bam=02-barker/leaf_ox_r1.bam,02-barker/ox_flower16.bam,02-barker/ox_flower9.bam</span><br><span class="line"><span class="comment"># --gff3: 输出GFF3格式</span></span><br><span class="line"><span class="comment"># --genome: 基因组序列</span></span><br><span class="line"><span class="comment"># --bam: 比对后的BAM文件，允许多个</span></span><br><span class="line"><span class="comment"># --cores: 处理核心数</span></span><br></pre></td></tr></table></figure>

<p>最后会得到如下输出文件</p>
<ul>
<li>hintsfile.gff: 从RNA-seq比对结果的BAM文件中提取，其中内含子用于训练GeneMark-EX, 使用所有特征训练AUGUSTUS</li>
<li>GeneMark-ET&#x2F;genemark.gtf: GeneMark-EX根据RNA-seq数据训练后预测的基因</li>
<li>augustus.hints.gff: AUGUSTUS输出文件</li>
</ul>
<p>将augustus.hints.gff3和文章的注释文件(carhr38.gtf)比较，见下图：</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-0200855851b844ce-bb010eed70444f82ab494eb706e92087.png" alt="结果比较"></p>
<p>其实不难发现，在不考虑UTR区域情况下，两者的差别其实更多表现是基因数目上，其实也就是利用转录组数据推测结构的问题所在，没有覆盖的区域到底是真的没有基因，还是有基因结构只不过所用组织没有表达，或者说那个区域其实是假基因？此外，如果基因间隔区域很短，有时候还会错误地把两个不同的基因预测为一个基因。因此，应该注重RNA-seq数据在<strong>剪切位点识别</strong>和<strong>外显子边界确定</strong>的优势。</p>
<h4 id="03-同源预测基因结构"><a href="#03-同源预测基因结构" class="headerlink" title="03-同源预测基因结构"></a>03-同源预测基因结构</h4><p>同源预测(homology prediction)利用近缘物种已知基因进行序列比对，找到同源序列。然后在同源序列的基础上，根据基因信号如剪切信号、基因起始和终止密码子对基因结构进行预测，如下示意图：</p>
<p><img src="/assets/2013053-b5a07df1dedaa354.png" alt="同源注释"></p>
<p>相对于从头预测的“大海捞针”，同源预测相当于先用一块磁铁在基因组大海中缩小了可能区域，然后从可能区域中鉴定基因结构。在10年之前，当时RNA-seq还没有普及, 只有少部分物种才有EST序列和cDNA序列的情况下，这的确是一个比较好的策略，那么问题来了，现在还需要进行这一步吗，如果需要是出于那种角度考虑呢?</p>
<p>在同源预测上，目前看到的大部分基因组文章都是基于TBLASTN + GeneWise，这可能是因为大部分基因组文章都是国内做的，这些注释自然而言用的就是公司的流程，然后目前国内的公司大多数又和某一家公司有一些关系。不过最近的3010水稻泛基因组用的是MAKER, 感谢部分提到这部分工作是由M. Roa(Philippine Genome Center Core Facilities for Bioinformatics, Department of Science)做的，算是一股清流吧。当然我在看Cardamine hirsuta基因组注释文章，发现它们同源注释部分用的是GenomeThreader, 该工具在本篇文章成文时的3月之前又更新了。</p>
<p>GeneWise的网站说它目前由Ewan Birney维护，只不过不继续开发了，因为Guy Slater开发Exonerate解决了GeneWise存在的很多问题，并且速度快了1000倍。考虑到目前只有GeneWise能利用HMM根据蛋白找DNA，而且ENSEMBL的注释流程也有一些核心模块用到了它，所以作者依旧在缓慢的开发这个工具(自2.4.1已经10多年没有更新了)，当然这个工具也是非常的慢。尽管这一步不会用到GeneWise作为我们的同源注释选项，但是我们可以尝试用GeneWise手工注释一个基因，主要步骤如下</p>
<ul>
<li>第一步： 使用BLASTX，根据dna序列搜索到蛋白序列，只需要第一个最佳比对结果</li>
<li>第二步： 选择最佳比对的氨基酸序列</li>
<li>第三步： 将dna序列前后延长2kb，与氨基酸序列一并传入给genewise进行同源预测</li>
</ul>
<p>提取前5K序列，然后选择在TAIR上用BLASTX进行比对</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">seqkit faidx chi_unmasked.fa Chr1:1-5000 &gt; chr1_5k.fa</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-562fefad5dec9b07-6a137a307f494a6385192c6d270ef644.png" alt="BLASTX"></p>
<p>选择第一个比对结果中的氨基酸序列，和前5k的DNA序列一并作为GeneWise的输入</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-da3cb7a46b5327a5-38978a3654cd461e96b6208b54c8fef6.png" alt="GeneWise2"></p>
<p>最后的结果出乎了我的意料</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-65f7c80123c3f586-b52d6c6326d14440886443690aabbdef.png" alt="预测结果"></p>
<p>让我们跳过这个尴尬的环节，毕竟很可能是我不太熟练使用工作所致。这里说点我的看法，除非你真的没有转录组数据，必须要用到同源物种的蛋白进行预测，或者你手动处理几个基因，否则不建议使用这个工具，因为你可能连安装都搞不定。</p>
<p>让我们用GenomeThreader基于上面的DNA序列和氨基酸序列进行同源基因结构预测吧</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">gth -genomic chr1_5k.fa -protein cer.fa -intermediate -gff3out</span><br><span class="line"><span class="comment"># 其中cer.fa就是AT1G02205.2的氨基酸序列</span></span><br></pre></td></tr></table></figure>

<p>结果一致，并且从RNA-seq的覆盖情况也符合预期</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">Chr1	gth	exon	1027	1197	Parent=gene1	Chr1    MIPS_CARH_v3.8  exon    1027    1197</span><br><span class="line">Chr1	gth	exon	1275	1448	Parent=gene1	Chr1    MIPS_CARH_v3.8  exon    1275    1448</span><br><span class="line">Chr1	gth	exon	1541	1662	Parent=gene1	Chr1    MIPS_CARH_v3.8  exon    1555    1662</span><br><span class="line">Chr1	gth	exon	1807	2007	Parent=gene1	Chr1    MIPS_CARH_v3.8  exon    1807    2007</span><br><span class="line">Chr1	gth	exon	2085	2192	Parent=gene1	Chr1    MIPS_CARH_v3.8  exon    2085    2192</span><br><span class="line">Chr1	gth	exon	2294	2669	Parent=gene1	Chr1    MIPS_CARH_v3.8  exon    2294    2669</span><br><span class="line">Chr1	gth	exon	3636	3855	Parent=gene1	Chr1    MIPS_CARH_v3.8  exon    3636    3855</span><br><span class="line">Chr1	gth	exon	3971	4203	Parent=gene1	Chr1    MIPS_CARH_v3.8  exon    3971    4203</span><br><span class="line">Chr1	gth	exon	4325	4548	Parent=gene1	Chr1    MIPS_CARH_v3.8  exon    4325    4548</span><br><span class="line">Chr1	gth	exon	4676	4735	Parent=gene1	Chr1    MIPS_CARH_v3.8  exon    4676    4735</span><br></pre></td></tr></table></figure>

<p>全基因组范围预测流程如下：</p>
<p>准备cDNA和或protein序列：在<a target="_blank" rel="noopener" href="https://phytozome.jgi.doe.gov/p">https://phytozome.jgi.doe.gov/p</a>下载靠谱的物种的蛋白质序列，如 <em>Arabidopsis thaliana</em>, <em>Oryza sativa</em>, <em>Brassica rapa</em>, 查找文献寻找目前该物种的已有EST&#x2F;cDNA序列，或者RNA-seq从头组装转录组。这里仅考虑用同源物种的蛋白序列进行比对分析，转录组从头组装数据用于PASA整体比对到参考基因组和更新已有的基因解雇。</p>
<p>分别测试下不同物种的同源注释结果</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#run seperately</span></span><br><span class="line">gth -species arabidopsis -translationtable 1 -gff3 -intermediate -protein ~/db/protein_db/Athaliana_167_TAIR10.protein.fa.gz -genomic chi_unmasked.fa -o 03-genomethreader/Athaliana.gff3 &amp;</span><br><span class="line">gth -species arabidopsis -translationtable 1 -gff3 -intermediate -protein ~/db/protein_db/BrapaFPsc_277_v1.3.protein.fa.gz -genomic chi_unmasked.fa -o 03-genomethreader/Brapa.gff3 &amp;</span><br><span class="line">gth -species arabidopsis -translationtable 1 -gff3 -intermediate -protein ~/db/protein_db/Osativa_323_v7.0.protein.fa.gz -genomic chi_unmasked.fa -o 03-genomethreader/Osativa.gff3 &amp;</span><br></pre></td></tr></table></figure>

<p>在定性角度上来看，同源注释的结果和从头预测的没啥差别, 其中B. rapa和A. thaliana和C. hirsuta都属于十字花科，而O. sativa是禾本科, 所以前两者预测的效果好。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-2bb888cdb1983838-84e79099d9a14a70a7e3c37ed3508c13.png" alt="IGV展示"></p>
<p>当然实际的同源注释流程中不能是单个物种分别预测，应该是将所有的蛋白序列进行合并，然后用BLASTX找到最优的联配，之后用GenomeThreader进行预测。PASA流程提到的<strong>UniRef90</strong>作为同源注释的搜索数据库可能是更好的选择，由于UniRef优先选择哪些人工审查、注释质量高、来源于模式动植物的蛋白，所以可靠性相对于直接使用同源物中可能更高。</p>
<blockquote>
<p>BLASTX + GenomeThreader的代码探索中</p>
</blockquote>
<h4 id="04-RNA-seq的两种使用策略"><a href="#04-RNA-seq的两种使用策略" class="headerlink" title="04-RNA-seq的两种使用策略"></a>04-RNA-seq的两种使用策略</h4><p>对于RNA-seq数据，有两种使用策略，一种是使用HISAT2 + StringTie先比对再组装, 一种是从头组装，然后使用PASA将转录本比对到基因组上。</p>
<h5 id="基于HISAT2-StringTie"><a href="#基于HISAT2-StringTie" class="headerlink" title="基于HISAT2 + StringTie"></a>基于HISAT2 + StringTie</h5><p>首先，使用HISAT2将RNA-seq数据比对到参考基因组, 这一步和之前相似，但是要增加一个参数<code>--dta</code>，使得StingTie能更好的利用双端信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">hisat2-build 01-augustus/genome.fa index/chi_masked</span><br><span class="line">hisat2 --dta -p 20 -x index/chi_masked -1 rna-seq/leaf_ox_r1_1.fastq.gz -2 rna-seq/leaf_ox_r1_2.fastq.gz | samtools <span class="built_in">sort</span> -@ 10 &gt; rna-seq/leaf_ox_r1.bam &amp;</span><br><span class="line">hisat2 --dta -p 20 -x index/chi_masked -1 rna-seq/ox_flower9_rep1_1.fastq.gz -2 rna-seq/ox_flower9_rep1_2.fastq.gz | samtools <span class="built_in">sort</span> -@ 10 &gt; rna-seq/ox_flower9.bam &amp;</span><br><span class="line">hisat2 --dta -p 20 -x index/chi_masked -1 rna-seq/ox_flower16_rep1_1.fastq.gz -2 rna-seq/ox_flower16_rep1_2.fastq.gz | samtools <span class="built_in">sort</span> -@ 10 &gt; rna-seq/ox_flower16.bam &amp;</span><br><span class="line">samtools merge -@ 10 rna-seq/merged.bam rna-seq/leaf_ox_r1.bam rna-seq/ox_flower9.bam rna-seq/ox_flower16.bam</span><br></pre></td></tr></table></figure>

<p>然后用StringTie进行转录本预测</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">stringtie -p 10 -o rna-seq/merged.gtf rna-seq/merged.bam</span><br></pre></td></tr></table></figure>

<p>对于后续的EvidenceModeler而言，它不需要UTR信息，只需要编码区CDS，需要用TransDecoder进行编码区预测</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">util/cufflinks_gtf_genome_to_cdna_fasta.pl merged.gtf input/chi_masked.fa &gt; transcripts.fasta</span><br><span class="line">util/cufflinks_gtf_to_alignment_gff3.pl merged.gtf &gt; transcripts.gff3</span><br><span class="line">TransDecoder.LongOrfs -t transcripts.fasta</span><br><span class="line">TransDecoder.Predict -t transcripts.fasta</span><br><span class="line">util/cdna_alignment_orf_to_genome_orf.pl \</span><br><span class="line">     transcripts.fasta.transdecoder.gff3 \</span><br><span class="line">     transcripts.gff3 \</span><br><span class="line">     transcripts.fasta &gt; transcripts.fasta.transdecoder.genome.gff3</span><br></pre></td></tr></table></figure>

<p>最后结果<code>transcripts.fasta.transdecoder.gff3</code>用于提供给EvidenceModeler</p>
<h5 id="基于PASA"><a href="#基于PASA" class="headerlink" title="基于PASA"></a>基于PASA</h5><p>在多年以前，那个基因组组装还没有白菜价，只有几个模式物种基因组的时代，对于一个未测序的基因组，研究者如果要研究某一个基因的功能，大多会通过同源物种相似基因设计PCR引物，然后去扩增cDNA. 如果是一个已知基因组的物种，如果要大规模识别基因, 研究者通常会使用EST(expressed sequence tags)序列。</p>
<p>相对于基于算法的从头预测，cDNA和EST序列更能够真实的反应出一个基因的真实结构，如可变剪切、UTR和Poly-A位点。PASA(Progam to Assemble Spliced Alignments)流程最早用于拟南芥基因组注释，最初的设计是通过将全长(full-length)cDNA和EST比对到参考基因组上，去发现和更新基因组注释。其中FL-cDNA和EST序列对最后结果的权重不同。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-d30be93f484be77d-28a3dcd689154b4caf797a2dcf0ea2fe.png" alt="PASA流程示意"></p>
<p>这是以前的故事，现在的故事是二代转录组以及一些三代转录组数据，那么如何处理这些数据呢？我认为三代转录组相对于过去的FL-cDNA，而二代转录组数据经过拼接后可以看作是更长的EST序列。由于目前最普及的还是普通的mRNA-seq, 也就只介绍这部分流程。</p>
<blockquote>
<p>考虑到我还没有研究过三代的全长转录组，分析过数据，这里的思考极有可能出错，后续可能会修改这一部分思考。</p>
</blockquote>
<p>转录组组装使用Trinity(conda安装)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> rna-seq</span><br><span class="line">Trinity --seqType fq --CPU 50 --max_memory 64G --left leaf_ox_r1_1.fastq.gz,ox_flower16_rep1_1.fastq.gz,ox_flower9_rep1_1.fastq.gz --right leaf_ox_r1_2.fastq.gz,ox_flower16_rep1_2.fastq.gz,ox_flower9_rep1_2.fastq.gz &amp;</span><br></pre></td></tr></table></figure>

<p>PASA是由30多个命令组成的流程，相关命令位于<code>PASApipeline/scripts</code>，为了适应不同的分析，有些参数需要通过修改配置文件更改,</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cp</span> ~/opt/biosoft/PASApipeline/pasa_conf/pasa.alignAssembly.Template.txt alignAssembly.config</span><br><span class="line"><span class="comment"># 修改如下内容</span></span><br><span class="line">DATABASE=database.sqlite</span><br><span class="line">validate_alignments_in_db.dbi:--MIN_PERCENT_ALIGNED=80</span><br><span class="line">validate_alignments_in_db.dbi:--MIN_AVG_PER_ID=80</span><br></pre></td></tr></table></figure>

<p>上述几行配置文件表明SQLite3数据库的名字，设置了<code>scripts/validate_alignments_in_db.dbi</code>的几个参数, 表示联配程度和相似程度。后续以Trinity组装结果和参考基因组作为输入，运行程序：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/PASApipeline/scripts/Launch_PASA_pipeline.pl -c alignAssembly.config -C -R -g ../chi_unmasked.fa -t ../rna-seq/trinity_out_dir/Trinity.fasta --ALIGNERS blat,gmap</span><br></pre></td></tr></table></figure>

<p>最后结果如下：</p>
<ul>
<li>database.sqlite.pasa_assemblies_described.txt</li>
<li>database.sqlite.pasa_assemblies.gff3</li>
<li>database.sqlite.pasa_assemblies.gtf</li>
<li>database.sqlite.pasa_assemblies.bed</li>
</ul>
<p>其中gff3格式用于后续的分析。</p>
<blockquote>
<p>目前的一些想法， 将从头组装的转录本比对到参考基因组上很大依赖组装结果，所以和EST序列和cDNA相比，质量上还有一点差距。</p>
</blockquote>
<h4 id="05-整合预测结果"><a href="#05-整合预测结果" class="headerlink" title="05-整合预测结果"></a>05-整合预测结果</h4><p>从头预测，同源注释和转录组整合都会得到一个预测结果，相当于收集了大量证据，下一步就是通过这些证据定义出更加可靠的基因结构，这一步可以通过人工排查，也可以使用EVidenceModeler(EVM). EVM只接受三类输入文件：</p>
<ul>
<li><code>gene_prediction.gff3</code>: 标准的GFF3格式，必须要有gene, mRNA, exon, CDS这些特征，用<code>EVidenceModeler-1.1.1/EvmUtils/gff3_gene_prediction_file_validator.pl</code>验证</li>
<li><code>protein_alignments.gff3</code>: 标准的GFF3格式，第9列要有ID信和和target信息, 标明是比对结果</li>
<li><code>transcript_alignments.gff3</code>:标准的GFF3格式，第9列要有ID信和和target信息，标明是比对结果</li>
</ul>
<p>EVM对<code>gene_prediction.gff3</code>有特殊的要求，就是GFF文件需要反映出一个基因的结构，gene-&gt;(mRNA -&gt; (exon-&gt;cds(?))(+))(+), 表示一个基因可以有多个mRNA，即基因的可变剪接, 一个mRNA都可以由一个或者多个exon(外显子), 外显子可以是非翻译区(UTR),也可以是编码区(CDS). 而GlimmerHMM, SNAP等</p>
<p>这三类根据人为经验来确定其可信度，从直觉上就是用PASA根据mRNA得到的结果高于从头预测。</p>
<p>第一步：创建权重文件,第一列是来源类型(ABINITIO_PREDICTION, PROTEIN, TRANSCRIPT), 第二列对应着GFF3文件的第二列，第三列则是权重.我这里用了<strong>三</strong>个来源的数据。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> 05-EVM &amp;&amp; <span class="built_in">cd</span> 05-EVM</span><br><span class="line"><span class="comment">#vim weights.txt</span></span><br><span class="line">ABINITIO_PREDICTION      augustus       4</span><br><span class="line">TRANSCRIPT      assembler-database.sqlite      7</span><br><span class="line">OTHER_PREDICTION  transdecoder  8</span><br></pre></td></tr></table></figure>

<blockquote>
<p>我觉得根据基因组引导组装的ORF的可信度高于组装后比对，所以得分和PASA差不多一样高。从头预测权重一般都是1，但是BRAKER可信度稍微高一点，可以在2~5之间。</p>
</blockquote>
<p>第二步：分割原始数据, 用于后续并行. 为了降低内存消耗，–segmentsSize设置的大小需要少于1Mb(这里是100k)， –overlapSize的不能太小，如果数学好，可用设置成基因平均长度加上2个标准差，数学不好，就设置成10K吧</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cat</span> transcripts.fasta.transdecoder.genome.gff3 ../braker/carhr/augustus.hints.gff3 &gt; gene_predictions.gff3</span><br><span class="line"><span class="built_in">ln</span> ../04-align-transcript/database.sqlite.pasa_assemblies.gff3 transcript_alignments.gff3</span><br><span class="line">~/opt/biosoft/EVidenceModeler-1.1.1/EvmUtils/partition_EVM_inputs.pl --genome ../chi_unmasked.fa --gene_predictions gene_predictions.gff3 --transcript_alignments transcript_alignments.gff3 --segmentSize 100000 --overlapSize 10000 --partition_listing partitions_list.out</span><br></pre></td></tr></table></figure>

<p>第三步：创建并行运算命令并且执行</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/EVidenceModeler-1.1.1/EvmUtils/write_EVM_commands.pl --genome ../chi_unmasked.fa --weights `<span class="built_in">pwd</span>`/weights.txt \</span><br><span class="line">      --gene_predictions gene_predictions.gff3 \</span><br><span class="line">      --transcript_alignments transcript_alignments.gff3 \</span><br><span class="line">      --output_file_name evm.out  --partitions partitions_list.out &gt;  commands.list</span><br><span class="line">parallel --<span class="built_in">jobs</span> 10 &lt; commands.list</span><br></pre></td></tr></table></figure>

<p>第四步：合并并行结果</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/EVidenceModeler-1.1.1/EvmUtils/recombine_EVM_partial_outputs.pl --partitions partitions_list.out --output_file_name evm.out</span><br></pre></td></tr></table></figure>

<p>第五步：结果转换成GFF3</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/EVidenceModeler-1.1.1/EvmUtils/convert_EVM_outputs_to_GFF3.pl  --partitions partitions_list.out --output evm.out  --genome ../chi_unmasked.fa</span><br><span class="line">find . -regex <span class="string">&quot;.*evm.out.gff3&quot;</span> -<span class="built_in">exec</span> <span class="built_in">cat</span> &#123;&#125; \; | bedtools <span class="built_in">sort</span> -i - &gt; EVM.all.gff</span><br></pre></td></tr></table></figure>

<p>当前权重设置下，EVM的结果更加严格，需要按照实际情况调整，增加其他证据。</p>
<h4 id="06-可选步骤"><a href="#06-可选步骤" class="headerlink" title="06-可选步骤"></a>06-可选步骤</h4><p><strong>注释过滤</strong>：对于初步预测得到的基因，还可以稍微优化一下，例如剔除编码少于50个AA的预测结果，将转座子单独放到一个文件中(软件有TransposonPSI)。</p>
<p>这里基于<code>gffread</code>先根据注释信息提取所有的CDS序列，过滤出长度不足50AA的序列，基于这些序列过滤原来的的注释</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">gffread EVM.all.gff -g input/genome.fa -y tr_cds.fa</span><br><span class="line">bioawk -c fastx <span class="string">&#x27;$seq &lt; 50 &#123;print $comment&#125;&#x27;</span> tr_cds.fa | <span class="built_in">cut</span> -d <span class="string">&#x27;=&#x27;</span> -f 2 &gt; short_aa_gene_list.txt</span><br><span class="line">grep -v -w -f short_aa_gene_list.txt EvM.all.gff &gt; filter.gff</span><br></pre></td></tr></table></figure>

<p><strong>使用PASA更新EVM结果</strong>：EVM结果不包括UTR区域和可变剪切的注释信息，可以使用PASA进行更新。然而这部分已经无法逃避MySQL, 服务器上并没有MySQL的权限，我需要学习Perl脚本进行修改。因此基因结构注释到此先放一放。</p>
<h4 id="07-基因编号"><a href="#07-基因编号" class="headerlink" title="07-基因编号"></a>07-基因编号</h4><p>对每个基因实现编号，形如ABCD000010的效果，方便后续分析。如下代码是基于EVM.all.gff，使用方法为<code>python gffrename.py EVM_output.gff prefix &gt; renamed.gff</code>.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#!/usr/bin/env python3</span></span><br><span class="line">import re</span><br><span class="line">import sys</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> len(sys.argv) &lt; 3:</span><br><span class="line">    sys.exit()</span><br><span class="line"></span><br><span class="line">gff = open(sys.argv[1])</span><br><span class="line">prf = sys.argv[2]</span><br><span class="line"></span><br><span class="line">count = 0</span><br><span class="line">mRNA  = 0</span><br><span class="line">cds   = 0</span><br><span class="line">exon  = 0</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;##gff-version 3.2.1&quot;</span>)</span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> gff:</span><br><span class="line">    <span class="keyword">if</span> not line.startswith(<span class="string">&quot;\n&quot;</span>):</span><br><span class="line">        records = line.split(<span class="string">&quot;\t&quot;</span>)</span><br><span class="line">        records[1] = <span class="string">&quot;.&quot;</span></span><br><span class="line">    <span class="keyword">if</span> re.search(r<span class="string">&quot;\tgene\t&quot;</span>, line):</span><br><span class="line">        count = count + 10</span><br><span class="line">        mRNA  = 0</span><br><span class="line">        gene_id = prf + str(count).zfill(6)</span><br><span class="line">        records[8] = <span class="string">&quot;ID=&#123;&#125;&quot;</span>.format(gene_id)</span><br><span class="line">    <span class="keyword">elif</span> re.search(r<span class="string">&quot;\tmRNA\t&quot;</span>, line):</span><br><span class="line">        cds   = 0</span><br><span class="line">        exon  = 0</span><br><span class="line">        mRNA  = mRNA + 1</span><br><span class="line">        mRNA_id    = gene_id + <span class="string">&quot;.&quot;</span> + str(mRNA)</span><br><span class="line">        records[8] = <span class="string">&quot;ID=&#123;&#125;;Parent=&#123;&#125;&quot;</span>.format(mRNA_id, gene_id)</span><br><span class="line">    <span class="keyword">elif</span> re.search(r<span class="string">&quot;\texon\t&quot;</span>, line):</span><br><span class="line">        exon     = exon + 1</span><br><span class="line">        exon_id  = mRNA_id + <span class="string">&quot;_exon_&quot;</span> + str(exon)</span><br><span class="line">        records[8] = <span class="string">&quot;ID=&#123;&#125;;Parent=&#123;&#125;&quot;</span>.format(exon_id, mRNA_id)</span><br><span class="line">    <span class="keyword">elif</span> re.search(r<span class="string">&quot;\tCDS\t&quot;</span>, line):</span><br><span class="line">        cds     = cds + 1</span><br><span class="line">        cds_id  = mRNA_id + <span class="string">&quot;_cds_&quot;</span> + str(cds)</span><br><span class="line">        records[8] = <span class="string">&quot;ID=&#123;&#125;;Parent=&#123;&#125;&quot;</span>.format(cds_id, mRNA_id)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="built_in">continue</span></span><br><span class="line"></span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&quot;\t&quot;</span>.<span class="built_in">join</span>(records))</span><br><span class="line"></span><br><span class="line">gff.close()</span><br></pre></td></tr></table></figure>

<h4 id="一些经验"><a href="#一些经验" class="headerlink" title="一些经验"></a>一些经验</h4><p>如果有转录组数据，没必须要使用太多的从头预测工具，braker2 加 GlimmerHMM可能就够用了, 更多是使用PASA和StringTie利用好转录组数据进行注释。</p>
<h3 id="基因功能注释"><a href="#基因功能注释" class="headerlink" title="基因功能注释"></a>基因功能注释</h3><p>基因功能的注释依赖于上一步的基因结构预测，根据预测结果从基因组上提取翻译后的 <strong>蛋白序列</strong> 和主流的数据库进行比对，完成功能注释。常用数据库一共有以几种：</p>
<ul>
<li>Nr：NCBI官方非冗余蛋白数据库，包括PDB, Swiss-Prot, PIR, PRF; 如果要用DNA序列，就是nt库</li>
<li>Pfam: 蛋白结构域注释的分类系统</li>
<li>Swiss-Prot: 高质量的蛋白数据库，蛋白序列得到实验的验证</li>
<li>KEGG: 代谢通路注释数据库.</li>
<li>GO: 基因本体论注释数据库</li>
</ul>
<p>除了以上几个比较通用的数据库外，其实还有很多小众数据库，应该根据课题研究和背景进行选择。<strong>注意</strong>，数据库本身并不能进行注释，你只是通过序列相似性进行搜索，而返回的结果你称之为注释。因此数据库和搜索工具要进行区分，所以你需要单独下载数据库和搜索工具，或者是同时下载包含数据库和搜索工具的安装包。</p>
<blockquote>
<p>注意，后续分析中一定要保证你的蛋白序列中不能有代表氨基酸字符以外的字符，比如说有些软件会把最后一个终止密码子翻译成”.”或者”*“</p>
</blockquote>
<h4 id="BLASTP"><a href="#BLASTP" class="headerlink" title="BLASTP"></a>BLASTP</h4><p>这一部分用到的数据库都是用BLASTP进行检索，基本都是四步发：下载数据库，构建BLASTP索引，数据库检索，结果整理。其中结果整理需要根据BLASTP的输出格式调整。</p>
<p><a href="ftp://ftp.ncbi.nlm.nih.gov/blast/db/">Nr</a>的NCBI收集的最全的蛋白序列数据库，但是无论是用NCBI的BLAST还是用速度比较快DIAMOND对nr进行搜索，其实都没有利用好物种本身的信息。因此在RefSeq上下载对应物种的蛋白序列, 用BLASTP进行注释即可。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># download</span></span><br><span class="line">wget -4  -nd -np -r 1 -A *.faa.gz ftp://ftp.ncbi.nlm.nih.gov/refseq/release/plant/</span><br><span class="line"><span class="built_in">mkdir</span> -p ~/db/RefSeq</span><br><span class="line">zcat *.gz &gt; ~/db/RefSeq/plant.protein.faa</span><br><span class="line"><span class="comment"># build index</span></span><br><span class="line">~/opt/biosoft/ncbi-blast-2.7.1+/bin/makeblastdb -<span class="keyword">in</span> plant.protein.faa -dbtype prot -parse_seqids -title RefSeq_plant -out plant</span><br><span class="line"><span class="comment"># search</span></span><br><span class="line">~/opt/biosoft/ncbi-blast-2.7.1+/bin/blastp -query protein.fa -out RefSeq_plant_blastp.xml -db ~/db/RefSeq/uniprot_sprot.fasta -evalue 1e-5 -outfmt 5 -num_threads 50 &amp;</span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="http://www.uniprot.org/downloads">Swiss-Prot</a>里收集了目前可信度最高的蛋白序列，一共有55w条记录，数据量比较小，</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># download</span></span><br><span class="line">wget -4 -q ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz</span><br><span class="line">gzip -d uniprot_sprot.fasta.gz</span><br><span class="line"><span class="comment"># builid index</span></span><br><span class="line">~/opt/biosoft/ncbi-blast-2.7.1+/bin/makeblastdb -<span class="keyword">in</span> uniprot_sprot.fasta -dbtype prot -title swiss_prot -parse_seqids</span><br><span class="line"><span class="comment"># search</span></span><br><span class="line">~/opt/biosoft/ncbi-blast-2.7.1+/bin/blastp -query protein.fa -out swiss_prot.xml -db ~/db/swiss_prot/uniprot_sprot.fasta -evalue 1e-5 -outfmt 5 -num_threads 50 &amp;</span><br></pre></td></tr></table></figure>

<p>关于结果整理，已经有很多人写了脚本，比如说我搜索BLAST XML CSV，就找到了<a target="_blank" rel="noopener" href="https://github.com/Sunhh/NGS_data_processing/blob/master/annot_tools/blast_xml_parse.py">https://github.com/Sunhh/NGS_data_processing/blob/master/annot_tools/blast_xml_parse.py</a>, 所以就不过多介绍。</p>
<h4 id="InterProScan"><a href="#InterProScan" class="headerlink" title="InterProScan"></a>InterProScan</h4><p>下面介绍的工具是<a target="_blank" rel="noopener" href="https://www.ebi.ac.uk/interpro/interproscan.html">InterProScan</a>, 从它的9G的体量就可以感受它的强大之处，一次运行同时实现多个信息注释。</p>
<ul>
<li>InterPro注释</li>
<li>Pfam数据库注释(可以通过hmmscan搜索pfam数据库完成)</li>
<li>GO注释(可以基于NR和Pfam等数据库，然后BLAST2GO完成,)</li>
<li>Reactome通路注释，不同于KEGG</li>
</ul>
<p>命令如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">./interproscan-5.29-68.0/interproscan.sh -appl Pfam  -f TSV -i sample.fa -cpu 50 -b sample -goterms -iprlookup -pa</span><br></pre></td></tr></table></figure>

<p><code>-appl</code>告诉软件要执行哪些数据分析，勾选的越多，分析速度越慢，Pfam就行。</p>
<h4 id="KEGG"><a href="#KEGG" class="headerlink" title="KEGG"></a>KEGG</h4><p><strong>KEGG数据库</strong>目前本地版收费，在线版收费，所以只能将蛋白序列在KEGG服务器上运行。因此你需要在<a target="_blank" rel="noopener" href="http://www.genome.jp/tools/kaas/">http://www.genome.jp/tools/kaas/</a>选择合适的工具进行后续的分析。我上传的50M大小蛋白序列，在KEGG服务器上只需要运行8个小时，也就是晚上提交任务，白天回来干活。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-e86990227c77f511-03651bbbee1b4847a35bb983407dedfc.png" alt="运行时间"></p>
<h2 id="附录"><a href="#附录" class="headerlink" title="附录"></a>附录</h2><p>基因组注释的常用软件：</p>
<ul>
<li>重复区域<ul>
<li>RepeatMasker：识别基因组中的可能重复</li>
<li>RepeatModeler: 识别新的重复序列</li>
<li>LTR-FINDER: <a target="_blank" rel="noopener" href="http://tlife.fudan.edu.cn/ltr_finder/">http://tlife.fudan.edu.cn/ltr_finder/</a></li>
</ul>
</li>
<li>从头预测<ul>
<li>Augustus</li>
<li>Fgenesh</li>
</ul>
</li>
<li>同源预测<ul>
<li>GeneWise</li>
<li>Exonerate</li>
<li>Trinity</li>
<li>GenomeThreader</li>
</ul>
</li>
<li>注释合并<ul>
<li>GLEAN：已经落伍于时代了</li>
<li>EvidenceModeler： 与时俱进</li>
</ul>
</li>
<li>流程<ul>
<li>PASA：真核生物基因的转录本可变剪切自动化注释项目，需要提供物种的EST或RNA-seq数据</li>
<li>MAKER</li>
<li>BRAKER1: 使用GeneMark-ET和AUGUSTUS基于RNA-Seq注释基因结构</li>
<li>EuGene</li>
</ul>
</li>
<li>可视化<ul>
<li>IGV</li>
<li>JBrowse&#x2F;GBrowse</li>
</ul>
</li>
</ul>
<p>参考文献和推荐阅读：</p>
<ul>
<li>NCBI真核生物基因组注释流程<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/genome/annotation_euk/process/">https://www.ncbi.nlm.nih.gov/genome/annotation_euk/process/</a></li>
<li>真核基因组注释入门: “A beginner’s guide to eukaryotic genome annotation”</li>
<li>二代测序注释流程:Comparative Gene Finding: “Annotation Pipelines for Next-Generation Sequencing Projects”</li>
<li>基因组转录组注释策略: “Plant genome and transcriptome annotations: from misconceptions to simple solution”</li>
<li>重复序列综述: “Repetitive DNA and next-generation sequencing: computational challenges and solutions”</li>
<li>MAKER2教程: <a target="_blank" rel="noopener" href="http://weatherby.genetics.utah.edu/MAKER/wiki/index.php/MAKER_Tutorial_for_WGS_Assembly_and_Annotation_Winter_School_2018">http://weatherby.genetics.utah.edu/MAKER/wiki/index.php/MAKER_Tutorial_for_WGS_Assembly_and_Annotation_Winter_School_2018</a></li>
<li>《生物信息学》 樊龙江: 第1-5章: 基因预测与功能注释</li>
<li>《NGS生物信息分析》 陈连福： 真核生物基因组基因注释</li>
<li>JGS流程: <a target="_blank" rel="noopener" href="https://genome.jgi.doe.gov/programs/fungi/FungalGenomeAnnotationSOP.pdf">https://genome.jgi.doe.gov/programs/fungi/FungalGenomeAnnotationSOP.pdf</a></li>
</ul>
<h2 id="环境准备"><a href="#环境准备" class="headerlink" title="环境准备"></a>环境准备</h2><h3 id="数据下载"><a href="#数据下载" class="headerlink" title="数据下载"></a>数据下载</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Cardamine hirsutat基因组数据</span></span><br><span class="line"><span class="built_in">mkdir</span> chi_annotation &amp;&amp; <span class="built_in">cd</span> chi_annotation</span><br><span class="line">wget http://chi.mpipz.mpg.de/download/sequences/chi_v1.fa</span><br><span class="line"><span class="built_in">cat</span> chi_v1.fa | <span class="built_in">tr</span> <span class="string">&#x27;atcg&#x27;</span> <span class="string">&#x27;ATCG&#x27;</span> &gt; chi_unmasked.fa</span><br><span class="line"><span class="comment"># Cardamine hirsutat转录组数据</span></span><br><span class="line">wget -4 -q -A <span class="string">&#x27;*.fastq.gz&#x27;</span> -np -nd -r 2 http://chi.mpipz.mpg.de/download/fruit_rnaseq/cardamine_hirsuta/ &amp;</span><br><span class="line">wget -4 -q -A <span class="string">&#x27;*.fastq.gz&#x27;</span> -np -nd -r 2 http://chi.mpipz.mpg.de/download/leaf_rnaseq/cardamine_hirsuta/ &amp;</span><br></pre></td></tr></table></figure>

<h3 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h3><p><strong>RepeatMasker</strong>: 用于注释基因组的重复区，需要安装RMBlast, TRF，以及在<a target="_blank" rel="noopener" href="http://www.girinst.org/">http://www.girinst.org</a>注册以下载Repbase</p>
<p>安装RepeatMasker</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget http://tandem.bu.edu/trf/downloadstrf409.linux64</span><br><span class="line"><span class="built_in">mv</span> trf409.linux64 ~/opt/bin/trf</span><br><span class="line"><span class="built_in">chmod</span> a+x ~/opt/bin/trf</span><br><span class="line"><span class="comment"># RMBlast</span></span><br><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/2.6.0/ncbi-blast-2.6.0+-src.tar.gz</span><br><span class="line">wget http://www.repeatmasker.org/isb-2.6.0+-changes-vers2.patch.gz</span><br><span class="line">tar xf ncbi-blast-2.6.0+-src</span><br><span class="line">gunzip isb-2.6.0+-changes-vers2.patch.gz</span><br><span class="line"><span class="built_in">cd</span> ncbi-blast-2.6.0+-src</span><br><span class="line">patch -p1 &lt; ../isb-2.6.0+-changes-vers2.patch</span><br><span class="line"><span class="built_in">cd</span> c++</span><br><span class="line">./configure --with-mt --prefix=~/opt/biosoft/rmblast --without-debug &amp;&amp; make &amp;&amp; make install</span><br><span class="line"><span class="comment"># RepeatMasker</span></span><br><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget http://repeatmasker.org/RepeatMasker-open-4-0-7.tar.gz</span><br><span class="line">tar xf RepeatMasker-open-4-0-7.tar.gz</span><br><span class="line"><span class="built_in">mv</span> RepeatMasker ~/opt/biosoft/</span><br><span class="line"><span class="built_in">cd</span> ~/opt/biosoft/RepeatMasker</span><br><span class="line"><span class="comment">## 解压repbase数据到Libraries下</span></span><br><span class="line"><span class="comment">## 配置RepatMasker</span></span><br><span class="line">perl ./configure</span><br></pre></td></tr></table></figure>

<p>在上面的基础上安装<strong>RepeatModel</strong></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># RECON</span></span><br><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget -4 http://repeatmasker.org/RepeatModeler/RECON-1.08.tar.gz</span><br><span class="line">tar xf RECON-1.08.tar.gz</span><br><span class="line"><span class="built_in">cd</span> RECON-1.08/src</span><br><span class="line">make &amp;&amp; make install</span><br><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line"><span class="built_in">mv</span> RECON-1.08 ~/opt/biosoft</span><br><span class="line"><span class="comment"># nesg</span></span><br><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line"><span class="built_in">mkdir</span> nesg &amp;&amp; <span class="built_in">cd</span> nesg</span><br><span class="line">wget -4 ftp://ftp.ncbi.nih.gov/pub/seg/nseg/*</span><br><span class="line">make</span><br><span class="line"><span class="built_in">mv</span> nmerge nseg ~/opt/bin/</span><br><span class="line"><span class="comment"># RepeatScout</span></span><br><span class="line">http://www.repeatmasker.org/RepeatScout-1.0.5.tar.gz</span><br><span class="line"><span class="comment"># RepeatModel</span></span><br><span class="line">wget -4 http://repeatmasker.org/RepeatModeler/RepeatModeler-open-1.0.11.tar.gz</span><br><span class="line">tar xf RepeatModeler-open-1.0.11.tar.gz</span><br><span class="line"><span class="built_in">mv</span> RepeatModeler-open-1.0.11 ~/opt/biosoft/</span><br><span class="line"><span class="built_in">cd</span> ~/opt/biosoft/RepeatModeler-open-1.0.11</span><br><span class="line"><span class="comment"># 配置</span></span><br><span class="line">perl ./configure</span><br><span class="line"><span class="built_in">export</span> PATH=~/opt/biosoft/maker:<span class="variable">$PATH</span></span><br></pre></td></tr></table></figure>

<p><strong>BLAST</strong>，BLAST有两个版本可供选择, WuBLAST或者NCBI-BLAST，我个人倾向于NCBI-BLAST，并且推荐使用编译后二进制版本，因为编译实在是太花时间了</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ncbi-blast-2.7.1+-x64-linux.tar.gz</span><br><span class="line">tar xf ncbi-blast-2.7.1+-x64-linux.tar.gz -C ~/opt/biosoft</span><br><span class="line"><span class="comment"># 环境变量</span></span><br><span class="line"><span class="built_in">export</span> PATH=~/opt/biosoft/ncbi-blast-2.7.1+/bin:<span class="variable">$PATH</span></span><br><span class="line"><span class="comment"># 用于后续的BRAKER2</span></span><br><span class="line">conda create -n annotation blast=2.2.31</span><br></pre></td></tr></table></figure>

<p><strong>AUGUSTUS</strong>: 可以说是最好的预测软件，使用conda安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">source</span> activate annotation</span><br><span class="line">conda install augustus=3.3</span><br></pre></td></tr></table></figure>

<p><strong>GeneMark-ES&#x2F;ET</strong>则是唯一一款支持无监督训练模型, 软件下载需要登记</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget http://topaz.gatech.edu/GeneMark/tmp/GMtool_Qg87n/gm_et_linux_64.tar.gz</span><br><span class="line">tar xf gm_et_linux_64.tar.gz</span><br><span class="line"><span class="built_in">mv</span> gm_et_linux_64/gmes_petap/ ~/opt/biosoft</span><br><span class="line">wget http://topaz.gatech.edu/GeneMark/tmp/GMtool_Qg87n/gm_key_64.gz</span><br><span class="line">gzip -dc gm_key_64.gz &gt; ~/.gm_key</span><br><span class="line">cpan YAML Hash::Merge Logger::Simple Parallel::ForkManager</span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;export PATH=<span class="variable">$PATH</span>:~/opt/biosoft/gmes_petap/&quot;</span> &gt;&gt; ~/.bashrc</span><br></pre></td></tr></table></figure>

<p><strong>GlimmerHMM</strong>:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget -4 ftp://ccb.jhu.edu/pub/software/glimmerhmm/GlimmerHMM-3.0.4.tar.gz</span><br><span class="line">tar xf GlimmerHMM-3.0.4.tar.gz -C ~/opt/biosoft</span><br></pre></td></tr></table></figure>

<p><strong>SNAP</strong>: 基因从头预测工具，在处理含有长内含子上的基因组上表现欠佳</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 安装</span></span><br><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">git <span class="built_in">clone</span> https://github.com/KorfLab/SNAP.git</span><br><span class="line"><span class="built_in">cd</span> SNP</span><br><span class="line">make</span><br><span class="line"><span class="built_in">cd</span> ..</span><br><span class="line"><span class="built_in">mv</span> SNAP ~/opt/biosoft</span><br><span class="line"><span class="comment"># 环境变量</span></span><br><span class="line"><span class="built_in">export</span> Zoe=~/opt/biosoft/SNAP/Zoe</span><br><span class="line"><span class="built_in">export</span> PATH=~/opt/biosoft/SNAP:<span class="variable">$PATH</span></span><br></pre></td></tr></table></figure>

<p><strong>Exnerate 2.2</strong>: 配对序列比对工具，提供二进制版本, 功能类似于GeneWise，能够将cDNA或蛋白以gao align的方式和基因组序列联配。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget http://ftp.ebi.ac.uk/pub/software/vertebrategenomics/exonerate/exonerate-2.2.0-x86_64.tar.gz</span><br><span class="line">tar xf exonerate-2.2.0-x86_64.tar.gz</span><br><span class="line"><span class="built_in">mv</span> exonerate-2.2.0-x86_64 ~/opt/biosoft/exonerate-2.2.0</span><br><span class="line"><span class="comment"># .bashrc添加环境变量</span></span><br><span class="line"><span class="built_in">export</span> PATH=~/opt/biosoft/exonerate-2.2.0:<span class="variable">$PATH</span></span><br><span class="line"><span class="comment"># 或</span></span><br><span class="line">conda install -c bioconda exonerate</span><br></pre></td></tr></table></figure>

<p><strong>GenomeThreader 1.70</strong>: 同源预测软件，1.7.0版本更新于2018年2月</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget -4 http://genomethreader.org/distributions/gth-1.7.0-Linux_x86_64-64bit.tar.gz</span><br><span class="line">tar xf gth-1.7.0-Linux_x86_64-64bit.tar.gz -C ~/opt/biosoft</span><br><span class="line"><span class="comment"># 修改.bashrc增加如下行</span></span><br><span class="line"><span class="built_in">export</span> PATH=<span class="variable">$PATH</span>:<span class="variable">$HOME</span>/opt/biosoft/gth-1.7.0-Linux_x86_64-64bit/bin</span><br><span class="line"><span class="built_in">export</span> BSSMDIR=<span class="string">&quot;<span class="variable">$HOME</span>/opt/biosoft/gth-1.7.0-Linux_x86_64-64bit/bin/bssm&quot;</span></span><br><span class="line"><span class="built_in">export</span> GTHATADIR=<span class="string">&quot;<span class="variable">$HOME</span>/opt/biosoft/gth-1.7.0-Linux_x86_64-64bit/bin/gthdata&quot;</span></span><br></pre></td></tr></table></figure>

<p><strong>BRAKER2</strong>: 依赖AUGUSTUS 3.3, GeneMark-EX 4.33, BAMTOOLS 2.5.1, NCBI BLAST+ 2.2.31+(可选 SAMTOOLS 1.74+, GenomeThreader 1.70)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">cpan File::Spec::Functions Module::Load::Conditional POSIX Scalar::Util::Numeric YAML File::Which Logger::Simple Parallel::ForkManager</span><br><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget -4 http://exon.biology.gatech.edu/GeneMark/Braker/BRAKER2.tar.gz</span><br><span class="line">tar xf BRAKER2.tar.gz -C ~/opt/biosoft</span><br><span class="line"><span class="built_in">echo</span> <span class="string">&quot;export PATH=<span class="variable">$PATH</span>:<span class="variable">$HOME</span>/opt/biosoft/BRAKER_v2.1.0/&quot;</span> &gt;&gt; ~/.bashrc</span><br><span class="line"><span class="comment"># 在~/.bashrc设置如下软件所在环境变量</span></span><br><span class="line"><span class="built_in">export</span> AUGUSTUS_CONFIG_PATH=<span class="variable">$HOME</span>/miniconda3/envs/annotation/config/</span><br><span class="line"><span class="built_in">export</span> AUGUSTUS_SCRIPTS_PATH=<span class="variable">$HOME</span>/miniconda3/envs/annotation/bin/</span><br><span class="line"><span class="built_in">export</span> BAMTOOLS_PATH=<span class="variable">$HOME</span>/miniconda3/envs/annotation/bin/</span><br><span class="line"><span class="built_in">export</span> GENEMARK_PATH=<span class="variable">$HOME</span>/opt/biosoft/gmes_petap/</span><br><span class="line"><span class="built_in">export</span> SAMTOOLS_PATH=<span class="variable">$HOME</span>/miniconda3/envs/annotation/bin/</span><br><span class="line"><span class="built_in">export</span> ALIGNMENT_TOOL_PATH=<span class="variable">$HOME</span>/opt/biosoft/gth-1.7.0-Linux_x86_64-64bit/bin/</span><br></pre></td></tr></table></figure>

<p><strong>TransDecoder</strong> 编码区域预测工具，需要预先安装NCBI-BLAST</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cpan URI::Escape</span><br><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget -4 https://github.com/TransDecoder/TransDecoder/archive/TransDecoder-v5.3.0.zip</span><br><span class="line">unzip TransDecoder-v5.3.0.zip</span><br><span class="line"><span class="built_in">cd</span> TransDecoder-v5.3.0</span><br><span class="line">make <span class="built_in">test</span></span><br></pre></td></tr></table></figure>

<p><strong>MARKER</strong>: 使用conda安装会特别的方便，最好新建环境</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda create -n marker marker</span><br></pre></td></tr></table></figure>

<p><strong>PASA</strong>: 依赖于一个数据库(MySQL或SQLite), Perl模块(DBD::mysql或DBD::SQLite), GMAP, BLAT, Fasta3。由于MySQL在HPC集群中的表现不如SQLite，以及安装MySQL还需要各种管理员权限，于是就有人进行了修改，增加了feature&#x2F;sqlite分支, 见<a target="_blank" rel="noopener" href="https://github.com/PASApipeline/PASApipeline/pull/43">Add support for SQLite</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">cpan DB_File URI::Escape DBI DBD::SQLite</span><br><span class="line"><span class="comment"># GMAP</span></span><br><span class="line">wget http://research-pub.gene.com/gmap/src/gmap-gsnap-2017-11-15.tar.gz</span><br><span class="line">tar xf gmap-gsnap-2017-11-15.tar.gz</span><br><span class="line"><span class="built_in">cd</span> gmap-2017-11-15</span><br><span class="line">./configure --prefix=<span class="variable">$HOME</span>/opt/gmap</span><br><span class="line">make &amp;&amp; make install</span><br><span class="line"><span class="comment"># BLAT</span></span><br><span class="line"><span class="built_in">cd</span> ~/opt/bin</span><br><span class="line">wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/blat/blat &amp;&amp; <span class="built_in">chmod</span> 755 ./blat</span><br><span class="line"><span class="comment"># Fasta3</span></span><br><span class="line">wget -4 http://faculty.virginia.edu/wrpearson/fasta/fasta36/fasta-36.3.8g.tar.gz &amp;&amp; \</span><br><span class="line">        tar zxvf fasta-36.3.8g.tar.gz &amp;&amp; \</span><br><span class="line">        <span class="built_in">cd</span> ./fasta-36.3.8g/src &amp;&amp; \</span><br><span class="line">        make -f ../make/Makefile.linux_sse2 all &amp;&amp; \</span><br><span class="line">        <span class="built_in">cp</span> ../bin/fasta36 ~/opt/bin</span><br><span class="line"><span class="comment"># 以上程序需添加到环境变量中</span></span><br><span class="line"><span class="comment"># PASApipeline</span></span><br><span class="line"><span class="built_in">cd</span> ~/opt/biosoft</span><br><span class="line">git <span class="built_in">clone</span> https://github.com/PASApipeline/PASApipeline.git</span><br><span class="line"><span class="built_in">cd</span> PASApipeline &amp;&amp; \</span><br><span class="line">git checkout feature/sqlite &amp;&amp; \</span><br><span class="line">git submodule init &amp;&amp; git submodule update &amp;&amp; \</span><br><span class="line">make</span><br></pre></td></tr></table></figure>

<p><strong>EVidenceModeler</strong>: 整合不同来源的注释结果，找到可靠的基因结构</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ~/src</span><br><span class="line">wget -4 https://github.com/EVidenceModeler/EVidenceModeler/archive/v1.1.1.tar.gz</span><br><span class="line">tar xf v1.1.1.tar.gz</span><br><span class="line"><span class="built_in">mv</span> EVidenceModeler-1.1.1 ~/opt/biosoft/</span><br></pre></td></tr></table></figure>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/11/%E5%A6%82%E4%BD%95%E5%AF%B9%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%BA%8F%E5%88%97%E8%BF%9B%E8%A1%8C%E6%B3%A8%E9%87%8A-How-to-annotate-your-genome/" data-id="clm1z22hb00f3ehn509a3fuzx" data-title="如何对基因组序列进行注释" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li></ul>

    </footer>
  </div>
  
    
<nav id="article-nav">
  
    <a href="/2019/10/13/%E7%BB%99R%E4%BD%BF%E7%94%A8%E8%80%85%E7%9A%84C++%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9F%A5%E8%AF%86-C++_For_R_User/" id="article-nav-newer" class="article-nav-link-wrap">
      <strong class="article-nav-caption">前一篇</strong>
      <div class="article-nav-title">
        
          给R使用者的C++最少必要知识
        
      </div>
    </a>
  
  
    <a href="/2019/10/10/%E5%A6%82%E4%BD%95%E5%B1%95%E7%A4%BAMUMMER%E7%9A%84%E7%BB%93%E6%9E%9C-Visualization-of-MUMMER-Result/" id="article-nav-older" class="article-nav-link-wrap">
      <strong class="article-nav-caption">后一篇</strong>
      <div class="article-nav-title">如何展示MUMMER的结果</div>
    </a>
  
</nav>

  
</article>


</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/09/">九月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/09/02/%E5%A6%82%E4%BD%95%E8%AE%A9%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8D%E5%86%8D%E9%BE%9F%E9%80%9F%E4%B8%8B%E8%BD%BDgithub%E7%9A%84%E6%95%B0%E6%8D%AE/">如何让服务器不再龟速下载github的数据</a>
          </li>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>