<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/5/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-MacOS原生Arm64架构的R解决依赖编译问题-how-to-compile-r-package-in-arm64-macos" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2022/02/21/MacOS%E5%8E%9F%E7%94%9FArm64%E6%9E%B6%E6%9E%84%E7%9A%84R%E8%A7%A3%E5%86%B3%E4%BE%9D%E8%B5%96%E7%BC%96%E8%AF%91%E9%97%AE%E9%A2%98-how-to-compile-r-package-in-arm64-macos/" class="article-date">
  <time class="dt-published" datetime="2022-02-21T09:51:46.786Z" itemprop="datePublished">2022-02-21</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2022/02/21/MacOS%E5%8E%9F%E7%94%9FArm64%E6%9E%B6%E6%9E%84%E7%9A%84R%E8%A7%A3%E5%86%B3%E4%BE%9D%E8%B5%96%E7%BC%96%E8%AF%91%E9%97%AE%E9%A2%98-how-to-compile-r-package-in-arm64-macos/">MacOS原生Arm64架构的R解决依赖编译问题</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>目前(截止到2022年2月份), Arm64架构的R能够直接安装CRAN绝大部分的R包，但是Bioconductor的R包都需要编译（这也就为什么我推荐用Intel架构的R，可以避免遇到要编译R包里的坑）。</p>
<p>编译R包最怕遇到的问题，就是遇到R以外的底层库的依赖，比如说Rhtslib, 就要求编译htslib, 而htslib依赖lzma是MacOS默认没有安装，因此我们会因为缺少xz库的支持，导致htslib编译失败，从而导致Rhtslib安装失败，间接导致所有依赖于Rhtslib的R包无法使用。还有一些R包编译需要gsl库，就会因为找不到 <code>gsl-config</code> 而报错。</p>
<p>经过我不断踩坑后，我整理了解决方案。后续，如果你也打算在Arm64的Mac上安装原生的R，就按照我的思路来配置</p>
<p>首先，你需要安装homebrew, 国内用户推荐按照清华镜像里的方案来配置，<a target="_blank" rel="noopener" href="https://mirrors.tuna.tsinghua.edu.cn/help/homebrew/">https://mirrors.tuna.tsinghua.edu.cn/help/homebrew/</a> , 能够避免网络问题导致的坑。在海外的朋友，就按照官方站点, brew.sh 的命令来配置。</p>
<p>homebrew会安装在 <code>/opt/homebrew</code>下，该目录里面包括后续安装的软件(bin), 头文件(include)和动态链接库(lib). </p>
<p>之后，我们需要为R配置相关环节，使得R能够调用到homebrew安装的环境。</p>
<p>我们需要打开R，然后运行如下命令</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">file.edit<span class="punctuation">(</span>file.path<span class="punctuation">(</span>Sys.getenv<span class="punctuation">(</span><span class="string">&quot;R_HOME&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;etc&quot;</span><span class="punctuation">,</span> <span class="string">&quot;Makeconf&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>如此就能打开R的Makeconf文件，R就通过该文件里面的参数来进行R包和底层依赖的编译。我们搜索搜索 CFLAG, 添加  <code>-I/opt/homebrew/inlucde</code>， 搜索LDFLAGS, 添加 <code>-L/opt/homebrew/lib</code>, 效果如下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">CFLAGS = -I/opt/homebrew/include -falign-functions=64 -Wall -g -O2 $(LTO)</span><br><span class="line">LDFLAGS = -L/opt/R/arm64/lib -L/opt/homebrew/lib</span><br></pre></td></tr></table></figure>

<p>重启R&#x2F;RStudio之后，你就能编译安装Rhtslib（当然你需要先运行 <code>brew install xz</code>，把lzma.h 装好&#96;)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&quot;Rhtslib&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>此外你可能还会遇到<code>gsl-config not found</code>报错，尽管你明明用<code>brew install gsl</code>安装了软件，也能够在命令行中成功调用 <code>gsl-config</code>，但是R就是视而不见。 这是因为R读取的变量，默认<strong>就来自于</strong>MacOS的环境变量来自于 <code>/etc/path</code>文件和 <code>/ect/path.d</code>目录里。 因此，你即便在.zshrc, 或者.bashrc里面加入了 homebrew 的路径，R也是不认的。</p>
<p>解决方案，打开R，然后运行如下命令,</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">file.edit<span class="punctuation">(</span><span class="string">&quot;~/.Rprofile&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>添加如下内容，修改R启动时的环境变量</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Sys.setenv(&quot;PATH&quot;=paste0( &quot;/opt/homebrew/bin:/opt/homebrew/sbin:&quot;,  Sys.getenv(&quot;PATH&quot;)))</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>重启R&#x2F;Rstudio之后，R就能够用到homebrew安装的程序了。</p>
<p>或者，我们也可以下载到本地，用 <code>R CMD INSTALL</code>进行编译安装，以DirichletMultinomial为例</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wget https://mirrors.tuna.tsinghua.edu.cn/bioconductor/packages/3.14/bioc/src/contrib/DirichletMultinomial_1.36.0.tar.gz</span><br><span class="line">R CMD INSTALL DirichletMultinomial_1.36.0.tar.gz</span><br></pre></td></tr></table></figure>

<p>总结一下:</p>
<ol>
<li>通过homebrew来安装依赖环境</li>
<li>编译R的Makeconf使得R能够调用homebrew里的include和lib</li>
<li>编译R的.Rprofile修改PATH，使得R能够用到hombrew的bin的程序</li>
</ol>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2022/02/21/MacOS%E5%8E%9F%E7%94%9FArm64%E6%9E%B6%E6%9E%84%E7%9A%84R%E8%A7%A3%E5%86%B3%E4%BE%9D%E8%B5%96%E7%BC%96%E8%AF%91%E9%97%AE%E9%A2%98-how-to-compile-r-package-in-arm64-macos/" data-id="clm1xknbi00jrdpnibxw6e4pz" data-title="MacOS原生Arm64架构的R解决依赖编译问题" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-吐槽下biopython的SeqIO.write导出fasta-biopython-seqio-write-are-not-good" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2022/02/17/%E5%90%90%E6%A7%BD%E4%B8%8Bbiopython%E7%9A%84SeqIO.write%E5%AF%BC%E5%87%BAfasta-biopython-seqio-write-are-not-good/" class="article-date">
  <time class="dt-published" datetime="2022-02-17T08:40:38.045Z" itemprop="datePublished">2022-02-17</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Python/">Python</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2022/02/17/%E5%90%90%E6%A7%BD%E4%B8%8Bbiopython%E7%9A%84SeqIO.write%E5%AF%BC%E5%87%BAfasta-biopython-seqio-write-are-not-good/">吐槽下biopython的SeqIO.write导出fasta</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>当使用biopython的SeqIO模块加载fasta之后，每条fasta都会记录成 <code>Bio.SeqRecord.SeqRecord</code>, 会有id, name, description, dbxrefs 等属性.</p>
<p>举个例子，输入数据是 test.fa</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># test.fa</span></span><br><span class="line"><span class="comment">#&gt;a|b c</span></span><br><span class="line"><span class="comment">#ATCG</span></span><br><span class="line">seq = SeqIO.read(<span class="string">&quot;test.fa&quot;</span>, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line">seq</span><br><span class="line">SeqRecord(seq=Seq(<span class="string">&#x27;ATCG&#x27;</span>), <span class="built_in">id</span>=<span class="string">&#x27;a|b&#x27;</span>, name=<span class="string">&#x27;a|b&#x27;</span>, description=<span class="string">&#x27;a|b c&#x27;</span>, dbxrefs=[])</span><br></pre></td></tr></table></figure>

<p>我们可以将其保存成另一个文件</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">with</span> <span class="built_in">open</span>(<span class="string">&quot;test2.fa&quot;</span>, <span class="string">&quot;w&quot;</span>) <span class="keyword">as</span> f:</span><br><span class="line">    SeqIO.write(seq, f, <span class="string">&quot;fasta&quot;</span>)</span><br></pre></td></tr></table></figure>


<p>输出的header和输入的header是完全一样的。</p>
<p>假如，你想修改他的名字, 比如说把 “a|b” 改成 “a”, 你会发现一个非常诡异的事情，就是它完全不会按照你想的来.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">seq.<span class="built_in">id</span> = <span class="string">&quot;a&quot;</span></span><br><span class="line"><span class="comment"># 输出 a a|b c</span></span><br><span class="line">seq.name = <span class="string">&quot;a&quot;</span></span><br><span class="line"><span class="comment"># 输出a|b c</span></span><br><span class="line">seq.description = <span class="string">&quot;a c&quot;</span></span><br><span class="line"><span class="comment"># 输出 a|b a c</span></span><br></pre></td></tr></table></figure>

<p>如果你改了他的ID，他会输出 ID + description, 如果你改了name，相当于没改，如果你改了description，就真的只有description（即header中空格后的信息）会发生变化。</p>
<p>假如我想输出 “&gt;a c”, 那我就必须下面这个形式</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">seq.<span class="built_in">id</span> = <span class="string">&quot;a&quot;</span></span><br><span class="line">seq.description = <span class="string">&quot;c&quot;</span></span><br></pre></td></tr></table></figure>

<p>我实在是没想懂他背后的逻辑, 修改输出的header居然如此不符合预期</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2022/02/17/%E5%90%90%E6%A7%BD%E4%B8%8Bbiopython%E7%9A%84SeqIO.write%E5%AF%BC%E5%87%BAfasta-biopython-seqio-write-are-not-good/" data-id="clm1xkn6f004wdpnid8cw1c94" data-title="吐槽下biopython的SeqIO.write导出fasta" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-我的生信自学心得分享-my-experience-bioinformatic-learning" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2022/02/16/%E6%88%91%E7%9A%84%E7%94%9F%E4%BF%A1%E8%87%AA%E5%AD%A6%E5%BF%83%E5%BE%97%E5%88%86%E4%BA%AB-my-experience-bioinformatic-learning/" class="article-date">
  <time class="dt-published" datetime="2022-02-16T14:26:33.456Z" itemprop="datePublished">2022-02-16</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2022/02/16/%E6%88%91%E7%9A%84%E7%94%9F%E4%BF%A1%E8%87%AA%E5%AD%A6%E5%BF%83%E5%BE%97%E5%88%86%E4%BA%AB-my-experience-bioinformatic-learning/">我的生信自学心得分享</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        
      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2022/02/16/%E6%88%91%E7%9A%84%E7%94%9F%E4%BF%A1%E8%87%AA%E5%AD%A6%E5%BF%83%E5%BE%97%E5%88%86%E4%BA%AB-my-experience-bioinformatic-learning/" data-id="clm1xkn7j0097dpnihfhadtnx" data-title="我的生信自学心得分享" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-RAP-DB版本的水稻注释文件的优化-optimizeannotationgffofrice" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2022/02/14/RAP-DB%E7%89%88%E6%9C%AC%E7%9A%84%E6%B0%B4%E7%A8%BB%E6%B3%A8%E9%87%8A%E6%96%87%E4%BB%B6%E7%9A%84%E4%BC%98%E5%8C%96-optimizeannotationgffofrice/" class="article-date">
  <time class="dt-published" datetime="2022-02-14T07:17:21.226Z" itemprop="datePublished">2022-02-14</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2022/02/14/RAP-DB%E7%89%88%E6%9C%AC%E7%9A%84%E6%B0%B4%E7%A8%BB%E6%B3%A8%E9%87%8A%E6%96%87%E4%BB%B6%E7%9A%84%E4%BC%98%E5%8C%96-optimizeannotationgffofrice/">RAP-DB版本的水稻注释文件的优化</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <blockquote>
<p>代码和数据都在GitHub上，见 <a target="_blank" rel="noopener" href="https://github.com/xuzhougeng/rice_annotation">https://github.com/xuzhougeng/rice_annotation</a></p>
</blockquote>
<p>水稻有两个主要的注释项目组</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://rapdb.dna.affrc.go.jp/">RAP-DB</a>: 编号格式形如Os01g0100100</li>
<li><a target="_blank" rel="noopener" href="http://rice.uga.edu/">RGAP</a>:  编号格式形如LOC_Os0101010</li>
</ul>
<p>虽然这两个注释项目组使用的是同一个水稻基因组，但是注释结果却存在一些差异。你可以使用任意一套注释进行分析，通过ID的对应关系，将其中一套编号转换成另外一套编号，我用的是RAP-DB版本，这是因为KEGG数据库使用的就是RAP-DB的编号形式。</p>
<p>RAP-DB的网站上提供了很多数据方便我们使用，比如说光基因序列文件就至少提供了为UTR+CDS, UTR+CDS+Intron, CDS三类，非常贴心。美中不足的一点是，他们提供的GFF文件并不规整。</p>
<p>他们提供了2个下载链接，分别对应GFF和GTF</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://rapdb.dna.affrc.go.jp/download/archive/irgsp1/IRGSP-1.0_representative_2021-11-11.tar.gz">Gene structure and function information in GFF format</a>.</li>
<li><a target="_blank" rel="noopener" href="https://rapdb.dna.affrc.go.jp/download/archive/irgsp1/IRGSP-1.0_representative_transcript_exon_2021-11-11.gtf.gz">Gene structure (only exon) information in GTF format</a>.</li>
</ul>
<p>GFF下载之后是一个压缩包，解压能得到locus.gff, transcripts_exon.gff, transcripts.gff. GTF下载之后只有一个gtf文件。</p>
<p>然而，无论是GFF还是GTF文件，他们都没有提供完整信息。GTF里面只有transcript和exon. GFF要么只有gene(locus.gff), 要么只有mRNA和exon(transcripts_exon.gff), 要么是mRNA,UTR和CDS(transcripts.gff). 我就非常纳闷了，为什么没有一个文件，同时有gene, mRNA, UTR, CDS, exon这些完整信息呢？</p>
<p>算了，既然他们没有给，我就自己动手，于是我写了一些代码，基于他们提供loucs.gff和transcripts.gff进行整理，代码我上传到GitHub。</p>
<p>我在代码中，将chr01-12改成了Chr1-Chr12, 所以对应的基因组文件，我也手动调整了。整理好的FASTA, GTF, GFF文件，我都一并上传到GitHub上，方便后续使用，地址为 <a target="_blank" rel="noopener" href="https://github.com/xuzhougeng/rice_annotation">https://github.com/xuzhougeng/rice_annotation</a></p>
<p>既然整理好了FASTA和GFF文件，为了方便后续在R语言里面使用，我还生成了对应BSgenome和TxDb，安装方法如下（生成代码和输出R包都在GitHub上）</p>
<p>安装方式(以BSgenome为例):</p>
<ol>
<li>在R里安装：<code>install.pakcages(&quot;/路径/到/BSgenome.Osativa.MSU.xzg_1.0.tar.gz&quot;, repos=NULL, type=&quot;source) </code></li>
<li>在命令行里安装: <code>R CMD INSTALL BSgenome.Osativa.MSU.xzg_1.0.tar.gz</code></li>
</ol>
<p>为了避免和Biocondutor其他MSU的包冲突，所以我用自己名字的缩写xzg作为后缀。</p>
<p>安装之后，就可以通过library进行加载</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>BSgenome.Osativa.MSU.xzg<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>TxDb.Osativa.MSU.xzg<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>



      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2022/02/14/RAP-DB%E7%89%88%E6%9C%AC%E7%9A%84%E6%B0%B4%E7%A8%BB%E6%B3%A8%E9%87%8A%E6%96%87%E4%BB%B6%E7%9A%84%E4%BC%98%E5%8C%96-optimizeannotationgffofrice/" data-id="clm1xknbm00k8dpni5rp86nw1" data-title="RAP-DB版本的水稻注释文件的优化" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-Mac键盘上符号是什么含义？-mac-jian-pan-shang-fu-hao-shi-shen-me-han-yi-" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2022/02/10/Mac%E9%94%AE%E7%9B%98%E4%B8%8A%E7%AC%A6%E5%8F%B7%E6%98%AF%E4%BB%80%E4%B9%88%E5%90%AB%E4%B9%89%EF%BC%9F-mac-jian-pan-shang-fu-hao-shi-shen-me-han-yi-/" class="article-date">
  <time class="dt-published" datetime="2022-02-10T06:17:11.860Z" itemprop="datePublished">2022-02-10</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2022/02/10/Mac%E9%94%AE%E7%9B%98%E4%B8%8A%E7%AC%A6%E5%8F%B7%E6%98%AF%E4%BB%80%E4%B9%88%E5%90%AB%E4%B9%89%EF%BC%9F-mac-jian-pan-shang-fu-hao-shi-shen-me-han-yi-/">Mac键盘上符号是什么含义？</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2022/02/image-62c5b421712f444abca2a35021d4c6f2.png" alt="keyboard"></p>
<p>在我用mac的时候，我一直有一个问题，那就是mac键盘上一些按键为什么会用哪个符号表示？比如说，最常见的command键用的⌘表示。通过一番检索，我找到了一个视频<a target="_blank" rel="noopener" href="https://www.youtube.com/watch?v=3QYKsf0qzKY">The Secret History of Mac Keyboard Keys</a> ，这里简单的整理成文字，进行介绍下。</p>
<p>command(⌘),最初用的是苹果的logo, 称为apple key. 但是乔布斯觉得随处使用苹果的logo不利于用户区分, 就让设计师重新找logo, 最终设计师看到这个最早用在Scandinavia的高速公路旁的露营地上使用的图标, 替换了原来的apple logo。</p>
<p>option(⌥), 在windows上对应alt键，两者都有切换的含义, 所以这个图标可以想象成铁道上两条轨道。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2022/02/image-1c09f531ac9b41b9a36546d27a4c4bc4.png" alt="轨道"></p>
<p>control(^)用的符号叫做脱字符(caret). windows和mac都有control键，但是windows的control通常对应mac上command键, 这就让我们在不同系统间切换变得头疼了. 通常而言在终端上, windows和mac的control键时相同含义，但是对于其他跨平台的软件, windows用的control键到mac上都要改成command键。之所以用脱字符, 是因为这个符号指向上方, 让你去脱离当下掌控(control)全局。</p>
<blockquote>
<p>脱字符在我们小时候写作文时比较常用,我们写完一句话之后,发现漏掉一个或几个字,于是就用这个符号加我们需要补充的内容, 把脱落掉的文字补充回去。</p>
</blockquote>
<p>Caps Lock(⇪) 和shift(⇧) 键, 在英文输入模式下, 前者用于锁定大写输入, 后者用于临时输入大写字母, 在中文输入模式下, 前者默认功能是切换中英文, 后者还是临时输入大写字母. 基于英文输入语境, 这两个功能的符号就可以理解成保持大写和临时大写了.</p>
<p>delete(⌫) 对应windows上backspace键, 功能和符号我觉得是比较对应的, 就是删除(叉掉)内容. 有一点要注意到是 windows系统可以用键盘上的Del删除文件, 但是mac删除文件的快捷键默认是command + del, 对应windows键盘的win + backspace, 而不能用win + Del. 这是因为Del对应的是⌦, 表示向后删除.</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2022/02/10/Mac%E9%94%AE%E7%9B%98%E4%B8%8A%E7%AC%A6%E5%8F%B7%E6%98%AF%E4%BB%80%E4%B9%88%E5%90%AB%E4%B9%89%EF%BC%9F-mac-jian-pan-shang-fu-hao-shi-shen-me-han-yi-/" data-id="clm1xknbj00jxdpni1vpb53jc" data-title="Mac键盘上符号是什么含义？" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-SSH登录服务器小技巧-ssh-small-trick" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2022/02/10/SSH%E7%99%BB%E5%BD%95%E6%9C%8D%E5%8A%A1%E5%99%A8%E5%B0%8F%E6%8A%80%E5%B7%A7-ssh-small-trick/" class="article-date">
  <time class="dt-published" datetime="2022-02-10T02:04:41.519Z" itemprop="datePublished">2022-02-10</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2022/02/10/SSH%E7%99%BB%E5%BD%95%E6%9C%8D%E5%8A%A1%E5%99%A8%E5%B0%8F%E6%8A%80%E5%B7%A7-ssh-small-trick/">SSH登录服务器小技巧</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>最近切换到了MacOS平台进行办公，就不能用Windows下好用的XShell，用上了传统在命令行输入 <code>ssh -p port user@address</code>的方式进行登录了。</p>
<p>作为一个‘懒惰’的人，我肯定是要避免重复的运行登录命令了。回溯用过的命令进行复用是一种方式，但还是需要输入密码，所以我的操作方式如下</p>
<p>第一步: 通过编辑 <code>~/.ssh/config</code>文件， 为指定服务器增加别名</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Host 别名</span><br><span class="line">    HostName 服务器地址</span><br><span class="line">    User 用户名</span><br><span class="line">    Port 端口</span><br></pre></td></tr></table></figure>

<p>这样子就能用 <code>ssh 目标服务器的别名</code>的方式登陆指定服务器，不必写后面的内容</p>
<p>第二步: 将本地的ssh公钥上传到目标服务器，实现免密登陆。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 生成密钥</span></span><br><span class="line">ssh-keygen</span><br><span class="line"><span class="comment"># 上传公钥</span></span><br><span class="line">ssh-copy-id 目标服务器的别名</span><br></pre></td></tr></table></figure>

<p>通过上述两步的配置，后续登陆服务器就就能节约不少时间，同时基于SSH的scp在服务器之间传送也不需要输入密码了。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2022/02/10/SSH%E7%99%BB%E5%BD%95%E6%9C%8D%E5%8A%A1%E5%99%A8%E5%B0%8F%E6%8A%80%E5%B7%A7-ssh-small-trick/" data-id="clm1xkn8z00chdpniaaag83v0" data-title="SSH登录服务器小技巧" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-M1芯片如何配置R分析环境「Intel版」-configure-r-analysis-environment-in-m1-macos" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2022/02/01/M1%E8%8A%AF%E7%89%87%E5%A6%82%E4%BD%95%E9%85%8D%E7%BD%AER%E5%88%86%E6%9E%90%E7%8E%AF%E5%A2%83%E3%80%8CIntel%E7%89%88%E3%80%8D-configure-r-analysis-environment-in-m1-macos/" class="article-date">
  <time class="dt-published" datetime="2022-02-01T12:17:34.540Z" itemprop="datePublished">2022-02-01</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2022/02/01/M1%E8%8A%AF%E7%89%87%E5%A6%82%E4%BD%95%E9%85%8D%E7%BD%AER%E5%88%86%E6%9E%90%E7%8E%AF%E5%A2%83%E3%80%8CIntel%E7%89%88%E3%80%8D-configure-r-analysis-environment-in-m1-macos/">M1芯片如何配置R分析环境「Intel版」</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <blockquote>
<p>视频地址: <a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1F3411E7td/">https://www.bilibili.com/video/BV1F3411E7td/</a></p>
</blockquote>
<p>各位好，我是洲更，欢迎收看本期教程。</p>
<p>在这一期视频中，我将会根据我的个人经验介绍如何在M1芯片的Mac上配置R语言的分析环境。</p>
<p>自从苹果在2021年10月份发布搭载M1 Pro或者是M1 Max芯片的14寸和16寸的MacBook Pro之后，目前在苹果官方网站就只能够购买M1芯片的MacBook Pro了。也就意味着，往后我们购买到的Mac都是ARM加购，而不是原来intel的x86架构。</p>
<p>我在付完钱，等待2个星期之后，终于也在2021年的11月的最后一天，到我手上这台搭载M1 Max芯片的16寸的MacBook Pro 。</p>
<p>于是，我终于有设备可以让我介绍，如何在M1芯片的电脑上配置R语言的分析环境了。</p>
<blockquote>
<p>插句题外话：不得不说，苹果的供应链系统真的强，只要你愿意等，你就能用它标的价格购买，而不像我之前想买的PS5和英伟达的显卡，要么就是几万人抢几百台，要么就得要多花个几百几千从其他店铺买。</p>
</blockquote>
<p>在这次教程中，我会介绍如何在M1芯片的Mac电脑上搭建R的分析环境，主要包括如下几个部分</p>
<ul>
<li>R语言的安装</li>
<li>RStudio的安装</li>
<li>R包安装所需的环境，</li>
<li>R包运行环境的配置</li>
<li>R包安装</li>
</ul>
<blockquote>
<p>尽管，我主要是以M1芯片的Mac介绍如何配置R语言环境，但由于目前许多软件都没有跟上，所以下载软件大部分还是为intel芯片开发的，需要通过rosetta2转译后才能在M1芯片的mac使用，所以这篇教程也适用于搭载intel芯片的mac电脑。</p>
</blockquote>
<h2 id="R语言的安装"><a href="#R语言的安装" class="headerlink" title="R语言的安装"></a>R语言的安装</h2><p>首先，我们需要安装最核心的组件，也就是R语言。</p>
<p>我们先打开浏览器，搜索R，找到R语言的官网,<a target="_blank" rel="noopener" href="http://r-project.org/">http://r-project.org</a>, 选择Download R，国外用户选择主站，国内用户推荐选择清华镜像源</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2022/02/image-e7059bac0bf4463082a13f288ca6cbf4.png" alt="R-Project"></p>
<p>然后点击   Download R for macOS， 此处，我们就会有两个选择， 一个是intel版本的R，一个是arm64版本，也就是专门给M1芯片开发的R。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2022/02/image-b035de5eb5164a88ba0766fcab4c2eda.png" alt="R-下载"></p>
<p>那么到底要选择哪个呢？</p>
<p>因为我这个视频是面向初学者的，比较基础，所以我无脑推荐使用Intel 64-bit的R语言，牺牲一点性能去避免不必要的折腾，是非常值得的。</p>
<p>后续，等到M1芯片相关的R语言环境变好了，我再出一期视频介绍。</p>
<p>下载完之后，我们只需要不断的下一步，就可以了。</p>
<h2 id="RStudio安装"><a href="#RStudio安装" class="headerlink" title="RStudio安装"></a>RStudio安装</h2><p>说完了R语言的安装，RStudio的安装就非常的容易了，因为他没有选择，目前就只有一个intel的版本。</p>
<p>同样通过搜索找到Rstudio的官方网站，然后找到他的下载方式，然后点击下载Mac版本的Rstudio。下载地址为<a target="_blank" rel="noopener" href="https://www.rstudio.com/products/rstudio/download/#download">https://www.rstudio.com/products/rstudio/download/#download</a> </p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2022/02/image-f0f9a435cdce4fd28bbdffdd444b71e9.png" alt="Rstudio"></p>
<p>下载完成后，我们双击打开，将其拖动到Applications中。</p>
<p>然后打开启动台，等待Rstudio出现，然后点击它打开它。</p>
<p>这里会出现一个弹窗，问我们是不是需要安装一个命令行开发工具，我们选择取消。后续我们会专门去安装这个工具。</p>
<p>我们简单的打一个 print(“hello world”)测试即可。</p>
<p>在安装完R和Rstudio后，我还想给大家推荐一个非常好用的工具， <a target="_blank" rel="noopener" href="https://rud.is/rswitch/">Rswitch</a></p>
<p>MacOS版本的RStudio是不支持切换不同的版本的R的，安装了这个工具后，我们就能够实现类似于Windows那种不同版本R语言切换的功能，</p>
<p>然而，这个软件比较小众，百度上我们暂时搜索不到（不过微软的必应可以），不过，这里我们选择直接输入网址， (<a target="_blank" rel="noopener" href="https://rud.is/rswitch/">https://rud.is/rswitch/</a>)。</p>
<p>我们点击下载，等待下载完成，最后将其拖动到应用程序中就算安装完成了。</p>
<p>我们在启动台里将其打开，就可以在屏幕的右上角找到它。点击它，就能够用它来切换不同版本的R，以及用它来启动Rstudio。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2022/02/image-2e79b6309ba74b7ca55d77d51be1e1a0.png" alt="Rswitch说明"></p>
<h3 id="编译环境"><a href="#编译环境" class="headerlink" title="编译环境"></a>编译环境</h3><p>在编译环境上，我们需要安装2个工具， xcode和gfortran。</p>
<p>xcode是苹果提供的开发工具，里面包括编译C和C++代码所需的工具。安装过程很简单，</p>
<ul>
<li>首先，我们打开启动台，</li>
<li>然后，在其他里面，找到终端，并打开</li>
<li>使用快捷键cmd + + 就可以放大终端</li>
<li>在终端里输入 <code>sudo xcode-select --install</code>，敲回车键</li>
<li>此时提示我们需要输入密码，注意，为了保证安全，所以在输入密码的时候，是看不到我们输入内容，我们输入结束后，敲回车就可以确定了。</li>
<li>接着就有一个窗口弹出来，我们点击同意。</li>
</ul>
<p>安装时间取决于网络环境，所以你的时间可能和我的不一样。我们等待它完成。</p>
<p>安装成功之后，我们在终端里面输入如下命令，如果没有提示找不到，就说明安装成功了。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">which</span> clang</span><br><span class="line"><span class="built_in">which</span> clang++</span><br></pre></td></tr></table></figure>



<p>除了xcode外，R包编译还会用到gfortran。跟xcode不同，gfortran的安装需要分两种情况，一种是为intel版本的R准备的版本，一种是为M1芯片版本的R准备的版本。</p>
<p>那么gfortran从哪里下载呢？我们先在浏览器中输入这个地址，<a target="_blank" rel="noopener" href="https://mac.r-project.org/tools/index.html">https://mac.r-project.org/tools/index.html</a>，打开之后，就可以看到下载方式了。</p>
<p>因为我们安装的是intel 64位的R，所以我们需要选择intel的版本下载就行。</p>
<p>如果你发现自己无法打开，可以尝试<a target="_blank" rel="noopener" href="https://download.csdn.net/download/u012110870/78708981">https://download.csdn.net/download/u012110870/78708981</a></p>
<p>intel的版本安装就是不断的下一步。</p>
<h3 id="运行环境"><a href="#运行环境" class="headerlink" title="运行环境"></a>运行环境</h3><p>接下来，我们来准备R语言的运行环境。</p>
<p>在运行环境上，我们需要安装两个工具，一个是Java,一个是，XQuartz, </p>
<p>有些包在安装的时候看起来没啥问题，但是在运行的时候却会报错，比如说rJava，究其原因，就是因为我们没有安装java环境。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>rJava<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">The operation couldn’t be completed. Unable to locate a Java Runtime.</span><br><span class="line">Please visit http<span class="operator">:</span><span class="operator">/</span><span class="operator">/</span>www.java.com <span class="keyword">for</span> information on installing Java.</span><br><span class="line"></span><br><span class="line">错误<span class="operator">:</span> package or namespace load failed <span class="keyword">for</span> ‘rJava’<span class="operator">:</span></span><br><span class="line"> loadNamespace<span class="punctuation">(</span><span class="punctuation">)</span>里算<span class="string">&#x27;rJava&#x27;</span>时.onLoad失败了，详细内容：</span><br><span class="line">  调用<span class="operator">:</span> fun<span class="punctuation">(</span>libname<span class="punctuation">,</span> pkgname<span class="punctuation">)</span></span><br><span class="line">  错误<span class="operator">:</span> JVM could not be found</span><br><span class="line">此外<span class="operator">:</span> Warning messages<span class="operator">:</span></span><br><span class="line"><span class="number">1</span><span class="operator">:</span> In system<span class="punctuation">(</span><span class="string">&quot;/usr/libexec/java_home&quot;</span><span class="punctuation">,</span> intern <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span> <span class="operator">:</span></span><br><span class="line">  运行命令<span class="string">&#x27;/usr/libexec/java_home&#x27;</span>的状态是<span class="number">1</span></span><br><span class="line"><span class="number">2</span><span class="operator">:</span> In fun<span class="punctuation">(</span>libname<span class="punctuation">,</span> pkgname<span class="punctuation">)</span> <span class="operator">:</span></span><br><span class="line">  Cannot find JVM library <span class="string">&#x27;NA/lib/server/libjvm.dylib&#x27;</span></span><br><span class="line">Install Java and<span class="operator">/</span>or check JAVA_HOME <span class="punctuation">(</span><span class="keyword">if</span> <span class="keyword">in</span> doubt<span class="punctuation">,</span> do NOT set it<span class="punctuation">,</span> it will be detected<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>那如何安装Java呢？这时候我们又需要借助于清华镜像了。</p>
<p>我们打开浏览器，搜索清华镜像，然后打开，选择 AdoptOpenJDK，我们选择最新的17，选择JDK，我们装的是intel版本的R，因此点击x64，点击mac，点击这个tar.gz结尾的文件进行下载。</p>
<p>下载结束之后，我们需要安装我们之前下载的安装包。这一步需要用到终端，请跟着视频输入命令。</p>
<p>这里，我选择将文件拖动到终端中，来获取文件路径</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sudo <span class="built_in">mkdir</span> -p /opt/jdk_x64</span><br><span class="line">sudo tar xf OpenJDK17U-jdk_x64_mac_hotspot_17.0.1_12.tar -C /opt/jdk_x64</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>按照上述步骤装完Java只完成一半，另外一半，我们需要在RStudio里操作。</p>
<p>我们打开Rstudio，输入如下命令 <code>file.edit(&quot;~/.Rprofile&quot;)</code> 打开R的配置文件，然后属于如下内容。由于你下载的版本可能和我的不一样，所以在输入过程中，要不断的使用tab 补全功能，确保路径正确。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Sys.setenv<span class="punctuation">(</span><span class="string">&quot;JAVA_HOME&quot;</span><span class="operator">=</span><span class="string">&quot;/opt/jdk_x64/jdk-17.0.1+12/Contents/Home&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>重启Rstudio后，R语言就会根据我们的配置来确定Java所在目录下。</p>
<p>除了java外，XQuartz，是后续安装的R包rgl不可或缺的一部分。如果没有安装XQuartz，即便安装了rgl，加载也会出现如下报错。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> library<span class="punctuation">(</span>rgl<span class="punctuation">)</span></span><br><span class="line">Error <span class="keyword">in</span> dyn.load<span class="punctuation">(</span>dynlib <span class="operator">&lt;-</span> getDynlib<span class="punctuation">(</span>dir<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">:</span> </span><br><span class="line">  无法载入共享目标对象‘<span class="operator">/</span>Library<span class="operator">/</span>Frameworks<span class="operator">/</span>R.framework<span class="operator">/</span>Versions<span class="operator">/</span><span class="number">4.1</span><span class="operator">-</span>arm64<span class="operator">/</span>Resources<span class="operator">/</span>library<span class="operator">/</span>rgl<span class="operator">/</span>libs<span class="operator">/</span>rgl.so’：<span class="operator">:</span></span><br><span class="line">  dlopen<span class="punctuation">(</span><span class="operator">/</span>Library<span class="operator">/</span>Frameworks<span class="operator">/</span>R.framework<span class="operator">/</span>Versions<span class="operator">/</span><span class="number">4.1</span><span class="operator">-</span>arm64<span class="operator">/</span>Resources<span class="operator">/</span>library<span class="operator">/</span>rgl<span class="operator">/</span>libs<span class="operator">/</span>rgl.so<span class="punctuation">,</span> <span class="number">0x0006</span><span class="punctuation">)</span><span class="operator">:</span></span><br><span class="line">   Library not loaded<span class="operator">:</span> <span class="operator">/</span>opt<span class="operator">/</span>X11<span class="operator">/</span>lib<span class="operator">/</span>libGLU.1.dylib</span><br><span class="line">  Referenced from<span class="operator">:</span> <span class="operator">/</span>Library<span class="operator">/</span>Frameworks<span class="operator">/</span>R.framework<span class="operator">/</span>Versions<span class="operator">/</span><span class="number">4.1</span><span class="operator">-</span>arm64<span class="operator">/</span>Resources<span class="operator">/</span>library<span class="operator">/</span>rgl<span class="operator">/</span>libs<span class="operator">/</span>rgl.so</span><br><span class="line">  Reason<span class="operator">:</span> tried<span class="operator">:</span> <span class="string">&#x27;/opt/X11/lib/libGLU.1.dylib&#x27;</span> <span class="punctuation">(</span>no such file<span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&#x27;/usr/lib/libGLU.1.dylib&#x27;</span> <span class="punctuation">(</span>no such file<span class="punctuation">)</span></span><br><span class="line">错误<span class="operator">:</span> package or namespace load failed <span class="keyword">for</span> ‘export’<span class="operator">:</span></span><br><span class="line"> loadNamespace<span class="punctuation">(</span><span class="punctuation">)</span>里算<span class="string">&#x27;rgl&#x27;</span>时.onLoad失败了，详细内容：</span><br><span class="line">  调用<span class="operator">:</span> rgl.init<span class="punctuation">(</span>initValue<span class="punctuation">,</span> onlyNULL<span class="punctuation">)</span></span><br><span class="line">  错误<span class="operator">:</span> OpenGL is not available <span class="keyword">in</span> this build</span><br><span class="line">此外<span class="operator">:</span> Warning messages<span class="operator">:</span></span><br><span class="line"><span class="number">1</span><span class="operator">:</span>   Loading rgl<span class="string">&#x27;s DLL failed. </span></span><br><span class="line"><span class="string">  This build of rgl depends on XQuartz, which failed to load.</span></span><br><span class="line"><span class="string"> See the discussion in https://stackoverflow.com/a/66127391/2554330 </span></span><br><span class="line"><span class="string">2: Trying without OpenGL...</span></span><br></pre></td></tr></table></figure>

<p>我们打开浏览器，输入它的官方站点: <a target="_blank" rel="noopener" href="https://www.xquartz.org/">https://www.xquartz.org/</a> </p>
<p>XQuartz目前是2.8.1版本，我们选择下载，双击打开，开始安装。安装结束之后，mac会注销一下。之后，我们打开启动台，打开其他，就能看到它了。</p>
<h2 id="R包安装"><a href="#R包安装" class="headerlink" title="R包安装"></a>R包安装</h2><p>在完成环境配置之后，我们终于可以安装R包了。</p>
<p>我们打开Rstudio，为了提供安装速度，我建议国内的用户修改一下镜像，提高安装速度，推荐国内的用户，按照如下方式修改修改。</p>
<p>首先，我们打开R的配置文件，然后我们打开清华镜像，点击CRAN旁边的问号，然后复制这一行到 R的配置文件里；接着我们找到Bioconductor，复制这一行到R的配置文件里。注意用cmd + s进行保存。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">file.edit<span class="punctuation">(</span><span class="string">&quot;~/.Rprofile&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">options<span class="punctuation">(</span><span class="string">&quot;repos&quot;</span> <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span>CRAN<span class="operator">=</span><span class="string">&quot;https://mirrors.tuna.tsinghua.edu.cn/CRAN/&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">options<span class="punctuation">(</span>BioC_mirror<span class="operator">=</span><span class="string">&quot;https://mirrors.tuna.tsinghua.edu.cn/bioconductor&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>之后，我们关掉配置文件，并重启Rstudio。</p>
<p>我们新建一个R脚本。</p>
<p>我们首先安装一个tidyverse，来测试CRAN的包的安装。</p>
<p>输入install.packages(“tidyverse”), 然后用快捷键 cmd + 回车运行。</p>
<p>首先蹦跶出一堆红字，这是它依赖的其他R包。</p>
<p>然后提示 有一个R包，它虽然有二进制版本，但源代码是最新的，问我们是不是要编译最新的。因为我们已经配置好了编译环境，所以这里输入yes。</p>
<p>后续，又出现了大量的红字，虽然很红，但是不是报错，都是正常信息。此时出现了黑字，这些事编译信息。</p>
<p>等他安装结束之后，我们输入 <code>library(tidyverse)</code> 来确认是否安装成功。 虽然有很多信息，但是没有看到error就说明成功了</p>
<p>接着，我们安装xlsx 来测试java环境。 没有报错说明，java环境正常。</p>
<p>然后是 rgl；我这里提示R出现fatal error，这并不是我们的环境配置出问题了，而是因为我的操作是在虚拟机里面执行的，是虚拟机性能不行，在加载时出错了。</p>
<p>接着，我们安装， BiocManager 用来安装 Bioconductor上的包。安装成功后，我们安装DESeq2，来测试Bioconductor，同样它也有很多依赖。</p>
<p>安装结束之后，提醒我们是否需要升级包，我这里输入n 表示不想升级，你们可以试试输入y。</p>
<p>最后我们试试安装github上的包，我们以ComplexHeat为例。我们输入如下命令， 运行的时候提示devtools没有找到，这说明我们没有安装这个包。</p>
<p>我们先去安装这个包； 这里有提示没有 这个函数， 原因是我输入出错了。改正就行了。</p>
<p>安装devtools后，我们再次运行之前的命令，这次就没有问题了。 我们加载也是成功的。</p>
<p>参考资料</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://mac.r-project.org/">https://mac.r-project.org/</a></li>
<li><a target="_blank" rel="noopener" href="https://mac.r-project.org/tools/index.html">https://mac.r-project.org/tools/index.html</a></li>
<li><a target="_blank" rel="noopener" href="https://kaopubear.top/blog/2021-07-06-macos-arm64-R-tips/">https://kaopubear.top/blog/2021-07-06-macos-arm64-R-tips/</a></li>
<li><a target="_blank" rel="noopener" href="https://code2care.org/q/install-native-java-jdk-jre-on-apple-silicon-m1-mac">https://code2care.org/q/install-native-java-jdk-jre-on-apple-silicon-m1-mac</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2022/02/01/M1%E8%8A%AF%E7%89%87%E5%A6%82%E4%BD%95%E9%85%8D%E7%BD%AER%E5%88%86%E6%9E%90%E7%8E%AF%E5%A2%83%E3%80%8CIntel%E7%89%88%E3%80%8D-configure-r-analysis-environment-in-m1-macos/" data-id="clm1xknbb00j9dpni2uoh5ax4" data-title="M1芯片如何配置R分析环境「Intel版」" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何使用GMAP和GSNAP进行转录组序列比对-align-transcript-with-gmap" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2022/01/18/%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8GMAP%E5%92%8CGSNAP%E8%BF%9B%E8%A1%8C%E8%BD%AC%E5%BD%95%E7%BB%84%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9-align-transcript-with-gmap/" class="article-date">
  <time class="dt-published" datetime="2022-01-18T08:06:16.614Z" itemprop="datePublished">2022-01-18</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2022/01/18/%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8GMAP%E5%92%8CGSNAP%E8%BF%9B%E8%A1%8C%E8%BD%AC%E5%BD%95%E7%BB%84%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9-align-transcript-with-gmap/">如何使用GMAP/GSNAP进行转录组序列比对</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>GMAP最早用于讲EST&#x2F;cDNA序列比对到参考基因组上，可以用于基因组结构注释。后来高通量测序时代，又开发了GSNAP支持高通量数据比对，这篇文章主要介绍GMAP，毕竟高通量转录组数据比对大家更喜欢用STAR, HISTA2等软件。</p>
<h2 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h2><p>下面是我源码安装的代码</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">wget http://research-pub.gene.com/gmap/src/gmap-gsnap-2018-07-04.tar.gz</span><br><span class="line">tar xf gmap-gsnap-2018-07-04.tar.gz</span><br><span class="line"><span class="built_in">cd</span> gmap-2018-07-04/</span><br><span class="line">./configure --prefix=<span class="variable">$HOME</span>/opt/biosoft/gmap</span><br><span class="line">make -j 20</span><br></pre></td></tr></table></figure>

<h2 id="软件使用"><a href="#软件使用" class="headerlink" title="软件使用"></a>软件使用</h2><p>如下步骤假设你有一个物种的基因组序列和对应的CDS序列，分别命名为”reference.fa”和”cds.fa”</p>
<h3 id="第一步-构建GMAP-GSNAP索引数据库"><a href="#第一步-构建GMAP-GSNAP索引数据库" class="headerlink" title="第一步:构建GMAP&#x2F;GSNAP索引数据库"></a>第一步:构建GMAP&#x2F;GSNAP索引数据库</h3><p>GMAP&#x2F;GSNAP对FASTA文件中每个记录下的序列的长度有一定限制, 每一条不能超过4G, 能应付的了大部分物种了。</p>
<p>构建索引分为两种情况考虑，第一种是一个fasta文件包含所有的序列</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/gmap/bin/gmap_build -d reference reference.fa</span><br></pre></td></tr></table></figure>

<p>第二种则是每个染色体的序列都单独存放在一个文件夹里，比如说你下载人类参考基因组序列解压后发现有N多个fasta文件, 然后你就想用其中几条染色体构建索引</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/gmap/bin/gmap_build -d reference Chr1.fa Chr2.fa Chr3.fa ...</span><br></pre></td></tr></table></figure>

<p>注: 这里的<code>-d</code>表示数据K库的名字，默认把索引存放在gmap安装路径下的share里，可以用<code>-D</code>更改.此外还有一个参数<code>-k</code>用于设置K-mer的长度, 默认是15, 理论上只有大于4GB基因组才会有两条一摸一样的15bp序列(当然是完全随机情况下)。</p>
<h3 id="第二步-正式比对"><a href="#第二步-正式比对" class="headerlink" title="第二步:正式比对"></a>第二步:正式比对</h3><p>建立完索引之后就可以将已有的CDS或者EST序列和参考基因组序列进行比较。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/opt/biosoft/gmap/bin/gmap -t 10 -d reference -f gff3_gene cds.fa &gt; cds_gene.gff3</span><br></pre></td></tr></table></figure>

<p>其中<code>-t</code>设置线程数, <code>-d</code>表示参考基因组数据库的名字, 都是常规参数。我比较感兴趣的参数是如何将序列输出成GFF格式. GMAP允许多种格式的输出，比如说<code>-S</code>只看联配的总体情况，而<code>-A</code>会显示每个比对上序列的联配情况, 还可以输出蛋白序列(-P)或者是genomic序列(-E). 但是做结构注释要的gff文件，参数就是<code>-f gff3_gene</code>, <code>-f gff3_match_cdna</code>, <code>-f gff3_match_est</code>。</p>
<h2 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h2><p>要想对一个软件有更好的认识，最好还是看看他们文章是怎么说的。</p>
<ul>
<li>GMAP: a genomic mapping and alignment program for mRNA and EST sequences<br>Bioinformatics 2005 21:1859-1875 <a target="_blank" rel="noopener" href="http://bioinformatics.oupjournals.org/cgi/content/abstract/21/9/1859">Abstract</a> <a target="_blank" rel="noopener" href="http://bioinformatics.oupjournals.org/cgi/content/full/21/9/1859">Full Text</a>, Thomas D. Wu and Colin K. Watanabe</li>
<li>Fast and SNP-tolerant detection of complex variants and splicing in short reads<br>Bioinformatics 2010 26:873-881 <a target="_blank" rel="noopener" href="http://bioinformatics.oupjournals.org/cgi/content/abstract/26/7/873">Abstract</a><a target="_blank" rel="noopener" href="http://bioinformatics.oupjournals.org/cgi/content/full/26/7/873">Full Text</a>, Thomas D. Wu and Serban Nacu</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2022/01/18/%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8GMAP%E5%92%8CGSNAP%E8%BF%9B%E8%A1%8C%E8%BD%AC%E5%BD%95%E7%BB%84%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9-align-transcript-with-gmap/" data-id="clm1xkn6m005xdpnifi072srm" data-title="如何使用GMAP/GSNAP进行转录组序列比对" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-深入探索biopython的配对序列联配模块-explore-biopythonpairwise-alignment" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2022/01/02/%E6%B7%B1%E5%85%A5%E6%8E%A2%E7%B4%A2biopython%E7%9A%84%E9%85%8D%E5%AF%B9%E5%BA%8F%E5%88%97%E8%81%94%E9%85%8D%E6%A8%A1%E5%9D%97-explore-biopythonpairwise-alignment/" class="article-date">
  <time class="dt-published" datetime="2022-01-02T08:34:48.384Z" itemprop="datePublished">2022-01-02</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Python/">Python</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2022/01/02/%E6%B7%B1%E5%85%A5%E6%8E%A2%E7%B4%A2biopython%E7%9A%84%E9%85%8D%E5%AF%B9%E5%BA%8F%E5%88%97%E8%81%94%E9%85%8D%E6%A8%A1%E5%9D%97-explore-biopythonpairwise-alignment/">深入探索biopython的配对序列联配模块</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <blockquote>
<p>本文基于biopython&gt;&#x3D;1.79</p>
</blockquote>
<p>配对序列联配是一个非常基础的生信分析，biopython提供了Bio.Align.PairwiseAligner基于动态规划处理两个序列的联配问题。</p>
<p>如下是一个简单的案例：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> Align</span><br><span class="line">aligner = Align.PairwiseAligner()</span><br><span class="line">alignments = aligner.align(<span class="string">&quot;ACCGGTTT&quot;</span>, <span class="string">&quot;ACGGGTT&quot;</span>)</span><br></pre></td></tr></table></figure>

<p>它首先初始化了一个PairwiseAligner对象，然后用该对象对两条序列进行联配，联配结果记录在了alignments。</p>
<p>alignments里面记录得分最高的联配结果（不止一个），我们可以通过循环来遍历其联配结果。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">for</span> alignment <span class="keyword">in</span> alignments:</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&quot;Score = %.1f:&quot;</span> % alignment.score)</span><br><span class="line">    <span class="built_in">print</span>(alignment)</span><br></pre></td></tr></table></figure>

<p>PairwiseAligner的运行速度极快，这是因为他底层通过C语言实现，而非Python代码，如下是他的结构体。</p>
<figure class="highlight c"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">typedef</span> <span class="class"><span class="keyword">struct</span> &#123;</span></span><br><span class="line">    PyObject_HEAD</span><br><span class="line">    Mode mode;</span><br><span class="line">    Algorithm algorithm;</span><br><span class="line">    <span class="type">double</span> match;</span><br><span class="line">    <span class="type">double</span> mismatch;</span><br><span class="line">    <span class="type">double</span> epsilon;</span><br><span class="line">    <span class="type">double</span> target_internal_open_gap_score;</span><br><span class="line">    <span class="type">double</span> target_internal_extend_gap_score;</span><br><span class="line">    <span class="type">double</span> target_left_open_gap_score;</span><br><span class="line">    <span class="type">double</span> target_left_extend_gap_score;</span><br><span class="line">    <span class="type">double</span> target_right_open_gap_score;</span><br><span class="line">    <span class="type">double</span> target_right_extend_gap_score;</span><br><span class="line">    <span class="type">double</span> query_internal_open_gap_score;</span><br><span class="line">    <span class="type">double</span> query_internal_extend_gap_score;</span><br><span class="line">    <span class="type">double</span> query_left_open_gap_score;</span><br><span class="line">    <span class="type">double</span> query_left_extend_gap_score;</span><br><span class="line">    <span class="type">double</span> query_right_open_gap_score;</span><br><span class="line">    <span class="type">double</span> query_right_extend_gap_score;</span><br><span class="line">    PyObject* target_gap_function;</span><br><span class="line">    PyObject* query_gap_function;</span><br><span class="line">    Py_buffer substitution_matrix;</span><br><span class="line">    PyObject* alphabet;</span><br><span class="line">    <span class="type">int</span>* mapping;</span><br><span class="line">    <span class="type">int</span> wildcard;</span><br><span class="line">&#125; Aligner;</span><br></pre></td></tr></table></figure>

<p>在结构体中，有一个非常关键的algorithm, 决定了联配时所需的算法。主要分为三个大类 <strong>NeedlemanWunschSmithWaterman</strong>, <strong>WatermanSmithBeyer</strong>和<strong>Gotoh</strong>.算法无法手动指定，而是根据match, mismatch, gap等参数的得分来确定(共6种)，具体的决策过程如下</p>
<p>当gap得分与gap长度有关，也就是 gap_score 赋值一个<strong>函数</strong>时，在WatermanSmithBeyer基础上，当mode是global, 使用 “Waterman-Smith-Beyer global alignment algorithm”；当mode是local，使用 “Waterman-Smith-Beyer local alignment algorithm”</p>
<p>当gap打开的得分和gap延伸的得分相同时，在NeedlemanWunschSmithWaterman基础上，如果mode是global， 使用Needleman-Wunsch；如果mode是local， 使用Smith-Waterman</p>
<p>其他情况下，也就是得分不同，则是在Gotoh基础上，如果mode是global，选择 “Gotoh global alignment algorithm”，如果mode是local，则选择”Gotoh local alignment algorithm”</p>
<p>我们举几个简单的例子来说明</p>
<p>场景1: 通过函数定义为gap打分</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> Align</span><br><span class="line">aligner = Align.PairwiseAligner()</span><br><span class="line">aligner.gap_score = <span class="keyword">lambda</span> x: x*<span class="number">2</span></span><br><span class="line">aligner.algorithm</span><br><span class="line"><span class="comment"># &#x27;Waterman-Smith-Beyer global alignment algorithm&#x27;</span></span><br><span class="line">aligner.mode = <span class="string">&quot;local&quot;</span></span><br><span class="line">aligner.algorithm</span><br><span class="line"><span class="comment"># &#x27;Waterman-Smith-Beyer local alignment algorithm&#x27;</span></span><br></pre></td></tr></table></figure>

<p>场景2: gap打开和延伸得分相同</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> Align</span><br><span class="line">aligner = Align.PairwiseAligner()</span><br><span class="line">aligner.open_gap_score = -<span class="number">1</span></span><br><span class="line">aligner.extend_gap_score = -<span class="number">1</span></span><br><span class="line">aligner.algorithm</span><br><span class="line"><span class="comment"># &#x27;Needleman-Wunsch&#x27;</span></span><br></pre></td></tr></table></figure>

<p>场景3: gap打开和延伸得分不同</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> Align</span><br><span class="line">aligner = Align.PairwiseAligner()</span><br><span class="line">aligner.open_gap_score = -<span class="number">1</span></span><br><span class="line">aligner.extend_gap_score = -<span class="number">2</span></span><br><span class="line">aligner.algorithm</span><br><span class="line"><span class="comment"># &#x27;Gotoh global alignment algorithm&#x27;</span></span><br></pre></td></tr></table></figure>


<p>除了为match和mismatch设置固定值外，我们还可以考虑使用得分矩阵为不同碱基或者氨基酸设置不同的得分，例如BLOSUM62</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> Align</span><br><span class="line">aligner = Align.PairwiseAligner()</span><br><span class="line"><span class="comment"># add matrix </span></span><br><span class="line"><span class="keyword">from</span> Bio.Align <span class="keyword">import</span> substitution_matrices</span><br><span class="line">aligner.substitution_matrix = substitution_matrices.load(<span class="string">&quot;BLOSUM62&quot;</span>)</span><br><span class="line"><span class="comment"># align</span></span><br><span class="line">alignments = aligner.align(<span class="string">&quot;ACCGGTTT&quot;</span>, <span class="string">&quot;ACGGGTT&quot;</span>)</span><br></pre></td></tr></table></figure>

<p>biopython的替换矩阵来自于<a href="ftp://ftp.ncbi.nih.gov/blast/matrices/">ftp://ftp.ncbi.nih.gov/blast/matrices/</a>, 保存在Bio&#x2F;Align&#x2F;substitution_matrices&#x2F;data 目录下</p>
<p>在搞定联配之后，我们如何获取联配后的结果？实际上对于一个 PairwiseAlignment 对象，例如 alignment, 我们调用 <code>print(alignment)</code>实际等价于 <code>print(format(alignment))</code>, 因此我们可以通过如下代码，获取联配后的query和target序列</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">seqs = <span class="built_in">format</span>(alignment).split(<span class="string">&quot;\n&quot;</span>)</span><br><span class="line">target = seqs[<span class="number">0</span>]</span><br><span class="line">query = seqs[<span class="number">2</span>]</span><br></pre></td></tr></table></figure>

<p>最后总结下本文内容：</p>
<ol>
<li>biopython的配对序列联配算法是C编写，运行速度极快</li>
<li>联配算法会根据 gap score和mode 自动确定</li>
<li>基于format函数可以获取联配后的序列</li>
</ol>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2022/01/02/%E6%B7%B1%E5%85%A5%E6%8E%A2%E7%B4%A2biopython%E7%9A%84%E9%85%8D%E5%AF%B9%E5%BA%8F%E5%88%97%E8%81%94%E9%85%8D%E6%A8%A1%E5%9D%97-explore-biopythonpairwise-alignment/" data-id="clm1xkn7v00amdpni0c8p5xfh" data-title="深入探索biopython的配对序列联配模块" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-snapATAC使用后的主观评测-snapatac-subjective-evaluation" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/12/30/snapATAC%E4%BD%BF%E7%94%A8%E5%90%8E%E7%9A%84%E4%B8%BB%E8%A7%82%E8%AF%84%E6%B5%8B-snapatac-subjective-evaluation/" class="article-date">
  <time class="dt-published" datetime="2021-12-30T06:05:11.034Z" itemprop="datePublished">2021-12-30</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2021/12/30/snapATAC%E4%BD%BF%E7%94%A8%E5%90%8E%E7%9A%84%E4%B8%BB%E8%A7%82%E8%AF%84%E6%B5%8B-snapatac-subjective-evaluation/">snapATAC使用后的主观评测</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>2019年，一篇发表在Genome Biology上题为「Assessment of computational methods for the analysis of single-cell ATAC-seq data」的文章对目前一些常见的单细胞ATAC-seq分析工具进行了评测，其中SnapATAC是当时文章认为比较好的工具</p>
<blockquote>
<p>Despite variation across methods and datasets, SnapATAC, Cusanovich2018, and cisTopic outperform other methods in separating cell populations of different coverages and noise levels in both synthetic and real datasets. Notably, SnapATAC is the only method able to analyze a large dataset (&gt; 80,000 cells).</p>
</blockquote>
<p>我曾经测试过这个工具，发现了其中的一个bug，后面就没有继续在用了。在2021年的最后几天，我又尝试使用了这个工具。如果你也在分析single cell ATAC-seq数据，想尝试snapATAC的话，希望这篇文章对你有所帮助。</p>
<ol>
<li>snapATAC有一个配套的工具, snapTools, 它能够将fastq文件处理成所需的snap文件，或者将CellRanger-ATAC输出BAM&#x2F;TSV文件转成snap文件。对于CellRanger-ATAC的数据转换，我不推荐使用BAM文件，因为他需要先给bam文件里read name添加barcode信息，然后按照read name进行排序，消耗的时间非常惊人。</li>
<li>snapATAC虽然能够处理大于80k的细胞，但是这和你选择的bin-size有关。snapATAC在分析的时候还是会将bin-by-cell的矩阵加载到内存中，因此内存还是一个比较重要的因素。我分析的时候，当bin-size比较小时，就会因为内存占用过多被kill。</li>
<li>snapATAC的可视化基于plot3D, 而不是ggplot2, 如果想要在原先图基础上进行修改，不如ggplot2那么方便。</li>
<li>snapATAC的聚类分析默认用的是R-igraph, leiden算法需要安装Python模块。R-igraph无法通过修改resolution来改变类群的数目，而leiden可以。不过，我好奇，为什么snapATAC不考虑使用Seurat的聚类算法？</li>
</ol>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/12/30/snapATAC%E4%BD%BF%E7%94%A8%E5%90%8E%E7%9A%84%E4%B8%BB%E8%A7%82%E8%AF%84%E6%B5%8B-snapatac-subjective-evaluation/" data-id="clm1xkn9d00dwdpniaavs92os" data-title="snapATAC使用后的主观评测" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/4/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/3/">3</a><a class="page-number" href="/page/4/">4</a><span class="page-number current">5</span><a class="page-number" href="/page/6/">6</a><a class="page-number" href="/page/7/">7</a><span class="space">&hellip;</span><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/6/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/09/">九月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/09/02/%E5%A6%82%E4%BD%95%E8%AE%A9%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8D%E5%86%8D%E9%BE%9F%E9%80%9F%E4%B8%8B%E8%BD%BDgithub%E7%9A%84%E6%95%B0%E6%8D%AE/">如何让服务器不再龟速下载github的数据</a>
          </li>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>