<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/10/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第十二章)-analysis-sc-atac-seq-with-archr-chapter12" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/07/19/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%BA%8C%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter12/" class="article-date">
  <time class="dt-published" datetime="2020-07-19T07:08:56.959Z" itemprop="datePublished">2020-07-19</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/07/19/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%BA%8C%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter12/">使用ArchR分析单细胞ATAC-seq数据(第十二章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第12章-使用ArchR进行motif和特征富集分析"><a href="#第12章-使用ArchR进行motif和特征富集分析" class="headerlink" title="第12章 使用ArchR进行motif和特征富集分析"></a>第12章 使用ArchR进行motif和特征富集分析</h1><p>在鉴定到可靠的peak集之后，我们也会想预测有哪些转录因子参与了结合事件(binding events)，从而产生了这些染色质开放位点。在分析标记peak或差异peak时，这能帮助我们更好的理解为什么某组的peak会富集某一类转录因子的结合位点。举个例子，我们想在细胞特异的染色质开放区域中找到定义谱系的关键转录因子。同样，我们也想根据其他已知特征对不同的peak进行富集分析。比如说，我们像知道是否细胞类型A的细胞特异性ATAC-seq peak对于另一组基因组区域（如ChIP-seq peak）也富集。这一章会详细介绍ArchR中的富集分析原理。</p>
<h2 id="12-1-差异peak中的motif富集"><a href="#12-1-差异peak中的motif富集" class="headerlink" title="12.1 差异peak中的motif富集"></a>12.1 差异peak中的motif富集</h2><p>继续上一章的差异peak分析，我们可以寻找在不同类型细胞富集的peak中的motif。我们需要先将motif的注释信息加入到我们的<code>ArchRProject</code>中。我们调用<code>addMotifAnnotations()</code>函数分析<code>ArchRProject</code>的peak中是否存在motif。运行结束后会在<code>ArchRProject</code>对象中加入一个新的二值矩阵，用于判断peak是否包括motif。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addMotifAnnotations<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> motifSet <span class="operator">=</span> <span class="string">&quot;cisbp&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Motif&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>接着我们使用上一章差异检验得到的<code>markerTest</code>分析motif的富集情况，这是一个<code>SummarizedExperiment</code>对象。我们用<code>peakAnnoEnrichments()</code>函数分析这些差异开放peak是否富集某一类moitf。可以设置<code>cutOff</code>来过滤peak，例如<code>PDR &lt;= 0.1 &amp; Log2FC &gt;=0.5</code>记录的是”Erythroid”比”Progenitor”更开放的peak。</p>
<p><strong>注</strong>: <code>peakAnnoEnrichment()</code>能用于多种差异富集检验，在后续章节还会介绍。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">motifsUp <span class="operator">&lt;-</span> peakAnnoEnrichment<span class="punctuation">(</span></span><br><span class="line">    seMarker <span class="operator">=</span> markerTest<span class="punctuation">,</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    peakAnnotation <span class="operator">=</span> <span class="string">&quot;Motif&quot;</span><span class="punctuation">,</span></span><br><span class="line">    cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; Log2FC &gt;= 0.5&quot;</span></span><br><span class="line">  <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>输出的<code>peakAnnoEnrichment()</code>是一个<code>SummarizedExperiment</code>对象，里面存放着多个<code>assays</code>, 记录着超几何检验的富集结果。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">motifsUp</span><br><span class="line"><span class="comment"># class: SummarizedExperiment</span></span><br><span class="line"><span class="comment"># dim: 870 1</span></span><br><span class="line"><span class="comment"># metadata(0):</span></span><br><span class="line"><span class="comment"># assays(10): mlog10Padj mlog10p … CompareFrequency feature</span></span><br><span class="line"><span class="comment"># rownames(870): TFAP2B_1 TFAP2D_2 … TBX18_869 TBX22_870</span></span><br><span class="line"><span class="comment"># rowData names(0):</span></span><br><span class="line"><span class="comment"># colnames(1): Erythroid</span></span><br><span class="line"><span class="comment"># colData names(0):</span></span><br></pre></td></tr></table></figure>

<p>然后，我们创建一个<code>data.frame</code>对象用于<code>ggplot</code>作图，包括motif名，矫正的p值和显著性排序。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">df <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>TF <span class="operator">=</span> rownames<span class="punctuation">(</span>motifsUp<span class="punctuation">)</span><span class="punctuation">,</span> mlog10Padj <span class="operator">=</span> assay<span class="punctuation">(</span>motifsUp<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">df <span class="operator">&lt;-</span> df<span class="punctuation">[</span>order<span class="punctuation">(</span>df<span class="operator">$</span>mlog10Padj<span class="punctuation">,</span> decreasing <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line">df<span class="operator">$</span>rank <span class="operator">&lt;-</span> <span class="built_in">seq_len</span><span class="punctuation">(</span>nrow<span class="punctuation">(</span>df<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>正如我们所预期的那样，”Erythroid”里开放的peak富集的motif主要是GATA转录因子，符合以往研究中”GATA1”在erythroid分化中发挥的作用。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">head<span class="punctuation">(</span>df<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>使用<code>ggplot</code>展示结果，以<code>ggrepel</code>来标识每个TF motif名。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">ggUp <span class="operator">&lt;-</span> ggplot<span class="punctuation">(</span>df<span class="punctuation">,</span> aes<span class="punctuation">(</span>rank<span class="punctuation">,</span> mlog10Padj<span class="punctuation">,</span> color <span class="operator">=</span> mlog10Padj<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  geom_point<span class="punctuation">(</span>size <span class="operator">=</span> <span class="number">1</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  ggrepel<span class="operator">::</span>geom_label_repel<span class="punctuation">(</span></span><br><span class="line">        data <span class="operator">=</span> df<span class="punctuation">[</span>rev<span class="punctuation">(</span><span class="built_in">seq_len</span><span class="punctuation">(</span><span class="number">30</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span><span class="punctuation">,</span> aes<span class="punctuation">(</span>x <span class="operator">=</span> rank<span class="punctuation">,</span> y <span class="operator">=</span> mlog10Padj<span class="punctuation">,</span> label <span class="operator">=</span> TF<span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        size <span class="operator">=</span> <span class="number">1.5</span><span class="punctuation">,</span></span><br><span class="line">        nudge_x <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span></span><br><span class="line">        color <span class="operator">=</span> <span class="string">&quot;black&quot;</span></span><br><span class="line">  <span class="punctuation">)</span> <span class="operator">+</span> theme_ArchR<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  ylab<span class="punctuation">(</span><span class="string">&quot;-log10(P-adj) Motif Enrichment&quot;</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  xlab<span class="punctuation">(</span><span class="string">&quot;Rank Sorted TFs Enriched&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  scale_color_gradientn<span class="punctuation">(</span>colors <span class="operator">=</span> paletteContinuous<span class="punctuation">(</span>set <span class="operator">=</span> <span class="string">&quot;comet&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ggUp</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-7ac9fda249bb48c29e75c423cba7902d.png" alt="Erythroid-vs-Progenitor-Markers-Motifs-Enriched_1"></p>
<p>通过设置<code>Log2FC &lt;= 0.5</code>我们可以挑选出在”Progenitor”里更加开放的peak，然后分析其中富集的motif。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">motifsDo <span class="operator">&lt;-</span> peakAnnoEnrichment<span class="punctuation">(</span></span><br><span class="line">    seMarker <span class="operator">=</span> markerTest<span class="punctuation">,</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    peakAnnotation <span class="operator">=</span> <span class="string">&quot;Motif&quot;</span><span class="punctuation">,</span></span><br><span class="line">    cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; Log2FC &lt;= -0.5&quot;</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line">motifsDo</span><br></pre></td></tr></table></figure>

<p>准备绘图所需数据框</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">df <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>TF <span class="operator">=</span> rownames<span class="punctuation">(</span>motifsDo<span class="punctuation">)</span><span class="punctuation">,</span> mlog10Padj <span class="operator">=</span> assay<span class="punctuation">(</span>motifsDo<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">df <span class="operator">&lt;-</span> df<span class="punctuation">[</span>order<span class="punctuation">(</span>df<span class="operator">$</span>mlog10Padj<span class="punctuation">,</span> decreasing <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line">df<span class="operator">$</span>rank <span class="operator">&lt;-</span> <span class="built_in">seq_len</span><span class="punctuation">(</span>nrow<span class="punctuation">(</span>df<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>此时，我们会发现在”Progenitor”细胞更加开放的peak中，更多富集RUNX, ELF和CBFB。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">head<span class="punctuation">(</span>df<span class="punctuation">)</span></span><br><span class="line"><span class="comment"># TF mlog10Padj rank</span></span><br><span class="line"><span class="comment"># 326 ELF2_326 88.68056 1</span></span><br><span class="line"><span class="comment"># 733 RUNX1_733 64.00586 2</span></span><br><span class="line"><span class="comment"># 801 CBFB_801 53.55426 3</span></span><br><span class="line"><span class="comment"># 732 RUNX2_732 53.14766 4</span></span><br><span class="line"><span class="comment"># 734 ENSG00000250096_734 53.14766 5</span></span><br><span class="line"><span class="comment"># 336 SPIB_336 52.79666 6</span></span><br></pre></td></tr></table></figure>

<p>使用<code>ggplot</code>展示结果。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">ggDo <span class="operator">&lt;-</span> ggplot<span class="punctuation">(</span>df<span class="punctuation">,</span> aes<span class="punctuation">(</span>rank<span class="punctuation">,</span> mlog10Padj<span class="punctuation">,</span> color <span class="operator">=</span> mlog10Padj<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  geom_point<span class="punctuation">(</span>size <span class="operator">=</span> <span class="number">1</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  ggrepel<span class="operator">::</span>geom_label_repel<span class="punctuation">(</span></span><br><span class="line">        data <span class="operator">=</span> df<span class="punctuation">[</span>rev<span class="punctuation">(</span><span class="built_in">seq_len</span><span class="punctuation">(</span><span class="number">30</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span><span class="punctuation">,</span> aes<span class="punctuation">(</span>x <span class="operator">=</span> rank<span class="punctuation">,</span> y <span class="operator">=</span> mlog10Padj<span class="punctuation">,</span> label <span class="operator">=</span> TF<span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        size <span class="operator">=</span> <span class="number">1.5</span><span class="punctuation">,</span></span><br><span class="line">        nudge_x <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span></span><br><span class="line">        color <span class="operator">=</span> <span class="string">&quot;black&quot;</span></span><br><span class="line">  <span class="punctuation">)</span> <span class="operator">+</span> theme_ArchR<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  ylab<span class="punctuation">(</span><span class="string">&quot;-log10(FDR) Motif Enrichment&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  xlab<span class="punctuation">(</span><span class="string">&quot;Rank Sorted TFs Enriched&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  scale_color_gradientn<span class="punctuation">(</span>colors <span class="operator">=</span> paletteContinuous<span class="punctuation">(</span>set <span class="operator">=</span> <span class="string">&quot;comet&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ggDo</span><br></pre></td></tr></table></figure>

<p><code>plotFDF()</code>函数能够以可编辑的矢量版本保存图片。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>ggUp<span class="punctuation">,</span> ggDo<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Erythroid-vs-Progenitor-Markers-Motifs-Enriched&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="12-2-标记Peak的motif富集分析"><a href="#12-2-标记Peak的motif富集分析" class="headerlink" title="12.2 标记Peak的motif富集分析"></a>12.2 标记Peak的motif富集分析</h2><p>和之前利用差异peak的motif富集分析类似，我们同样能用<code>getMarkerFeatures()</code>分析标记peak里富集的motif。</p>
<p>我们向函数<code>peakAnnotationEnrichment()</code>传入存放标记peak的<code>SummarizedExperiment</code>对象，即<code>markersPeaks</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">enrichMotifs <span class="operator">&lt;-</span> peakAnnoEnrichment<span class="punctuation">(</span></span><br><span class="line">    seMarker <span class="operator">=</span> markersPeaks<span class="punctuation">,</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    peakAnnotation <span class="operator">=</span> <span class="string">&quot;Motif&quot;</span><span class="punctuation">,</span></span><br><span class="line">    cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; Log2FC &gt;= 0.5&quot;</span></span><br><span class="line">  <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>输出的<code>peakAnnoEnrichment()</code>是一个<code>SummarizedExperiment</code>对象，里面存放着多个<code>assays</code>, 记录着超几何检验的富集结果。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">enrichMotifs</span><br><span class="line"><span class="comment"># class: SummarizedExperiment</span></span><br><span class="line"><span class="comment"># dim: 870 11</span></span><br><span class="line"><span class="comment"># metadata(0):</span></span><br><span class="line"><span class="comment"># assays(10): mlog10Padj mlog10p … CompareFrequency feature</span></span><br><span class="line"><span class="comment"># rownames(870): TFAP2B_1 TFAP2D_2 … TBX18_869 TBX22_870</span></span><br><span class="line"><span class="comment"># rowData names(0):</span></span><br><span class="line"><span class="comment"># colnames(11): B CD4.M … PreB Progenitor</span></span><br><span class="line"><span class="comment"># colData names(0):</span></span><br></pre></td></tr></table></figure>

<p>直接用<code>plotEnrichHeatmap()</code>函数绘制不同细胞组的富集的motif。通过设置参数<code>n</code>限制每个细胞分组中展示的motif。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">heatmapEM <span class="operator">&lt;-</span> plotEnrichHeatmap<span class="punctuation">(</span>enrichMotifs<span class="punctuation">,</span> n <span class="operator">=</span> <span class="number">7</span><span class="punctuation">,</span> transpose <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>使用<code>ComplexHeatmap::draw()</code>函数展示结果</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ComplexHeatmap<span class="operator">::</span>draw<span class="punctuation">(</span>heatmapEM<span class="punctuation">,</span> heatmap_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">,</span> annotation_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-2f3eb61ae5a64bb297b86f3ec0b3e06c.png" alt="Motifs-Enriched-Marker-Heatmap"></p>
<p><code>plotFDF()</code>函数能够以可编辑的矢量版本保存图片。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>heatmapEM<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Motifs-Enriched-Marker-Heatmap&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">8</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">6</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="12-3-ArchR富集分析"><a href="#12-3-ArchR富集分析" class="headerlink" title="12.3 ArchR富集分析"></a>12.3 ArchR富集分析</h2><p>除了分析peak中富集motif, ArchR还能进行个性化的富集分析。为了方便这类数据探索，我们已人工确定了一些特征数据集，它们能比较容易地在我们感兴趣的peak区间进行检验。我们接下来将会逐个介绍这些特征数据集。该分析最初受<a target="_blank" rel="noopener" href="http://code.databio.org/LOLA/">LOLA</a>启发。</p>
<h3 id="12-3-1-Encode-TF-结合位点"><a href="#12-3-1-Encode-TF-结合位点" class="headerlink" title="12.3.1 Encode TF 结合位点"></a>12.3.1 Encode TF 结合位点</h3><p>ENCODE协会已经将TF结合位点(TFBS)匹配到多种细胞类型和因子中。我们可以利用这些TFBS去更好地理解聚类结果。例如，我们可以根据富集结果去判断未知细胞类型的可能类型。为了能够使用ENCODE TFBS特征集进行分析，我们需要调用<code>addArchRAnnotations()</code>函数，设置<code>collection = &quot;EncodeTFBS&quot;</code>. 和使用<code>addPeakAnnotations()</code>类似，这会创建一个二值矩阵，记录我们的标记peak是否和ENCODE TFBS有重叠。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addArchRAnnotations<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> collection <span class="operator">=</span> <span class="string">&quot;EncodeTFBS&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们接着使用<code>peakAnnoEnrichment()</code>函数分析这些 ENCODE TFBS是否在我们的peak中富集。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">enrichEncode <span class="operator">&lt;-</span> peakAnnoEnrichment<span class="punctuation">(</span></span><br><span class="line">    seMarker <span class="operator">=</span> markersPeaks<span class="punctuation">,</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    peakAnnotation <span class="operator">=</span> <span class="string">&quot;EncodeTFBS&quot;</span><span class="punctuation">,</span></span><br><span class="line">    cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; Log2FC &gt;= 0.5&quot;</span></span><br><span class="line">  <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>和之前一样，该函数返回一个<code>SummarizedExperiment</code>对象。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">enrichEncode</span><br><span class="line"><span class="comment"># class: SummarizedExperiment</span></span><br><span class="line"><span class="comment"># dim: 689 11</span></span><br><span class="line"><span class="comment"># metadata(0):</span></span><br><span class="line"><span class="comment"># assays(10): mlog10Padj mlog10p … CompareFrequency feature</span></span><br><span class="line"><span class="comment"># rownames(689): 1.CTCF-Dnd41… 2.EZH2_39-Dnd41… …</span></span><br><span class="line"><span class="comment"># 688.CTCF-WERI_Rb_1… 689.CTCF-WI_38…</span></span><br><span class="line"><span class="comment"># rowData names(0):</span></span><br><span class="line"><span class="comment"># colnames(11): B CD4.M … PreB Progenitor</span></span><br><span class="line"><span class="comment"># colData names(0):</span></span><br></pre></td></tr></table></figure>

<p>我们可以使用<code>plotEnrichHeatmap</code>函数从富集结果中创建热图。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">heatmapEncode <span class="operator">&lt;-</span> plotEnrichHeatmap<span class="punctuation">(</span>enrichEncode<span class="punctuation">,</span> n <span class="operator">=</span> <span class="number">7</span><span class="punctuation">,</span> transpose <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>然后用<code>ComplexHeatmap::draw()</code>绘制热图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ComplexHeatmap<span class="operator">::</span>draw<span class="punctuation">(</span>heatmapEncode<span class="punctuation">,</span> heatmap_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">,</span> annotation_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-32de05c1269845a08a690ad44420dc3c.png" alt="EncodeTFBS-Enriched-Marker-Heatmap"></p>
<p><code>plotFDF()</code>函数能够以可编辑的矢量版本保存图片。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>heatmapEncode<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;EncodeTFBS-Enriched-Marker-Heatmap&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">8</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">6</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="12-3-2-混池ATAC-seq"><a href="#12-3-2-混池ATAC-seq" class="headerlink" title="12.3.2 混池ATAC-seq"></a>12.3.2 混池ATAC-seq</h3><p>和ENCODE TFBS类似，我们还可以使用混池ATAC-seq实验鉴定的peak，分析两者的重叠情况。通过设置<code>collection=&quot;ATAC&quot;</code>来调用混池ATAC-seqpeak数据集。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addArchRAnnotations<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> collection <span class="operator">=</span> <span class="string">&quot;ATAC&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>接着通过设置<code>peakAnnotation = &quot;ATAC&quot;</code>检验我们的标记peak是否富集了混池ATAC-seq的peak。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">enrichATAC <span class="operator">&lt;-</span> peakAnnoEnrichment<span class="punctuation">(</span></span><br><span class="line">    seMarker <span class="operator">=</span> markersPeaks<span class="punctuation">,</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    peakAnnotation <span class="operator">=</span> <span class="string">&quot;ATAC&quot;</span><span class="punctuation">,</span></span><br><span class="line">    cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; Log2FC &gt;= 0.5&quot;</span></span><br><span class="line">  <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>和之前一样，该函数会输出<code>SummarizedExperiment</code>对象，记录着富集结果</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">enrichATAC</span><br><span class="line"><span class="comment"># class: SummarizedExperiment</span></span><br><span class="line"><span class="comment"># dim: 96 11</span></span><br><span class="line"><span class="comment"># metadata(0):</span></span><br><span class="line"><span class="comment"># assays(10): mlog10Padj mlog10p … CompareFrequency feature</span></span><br><span class="line"><span class="comment"># rownames(96): Brain_Astrocytes Brain_Excitatory_neurons … Heme_MPP</span></span><br><span class="line"><span class="comment"># Heme_NK</span></span><br><span class="line"><span class="comment"># rowData names(0):</span></span><br><span class="line"><span class="comment"># colnames(11): B CD4.M … PreB Progenitor</span></span><br><span class="line"><span class="comment"># colData names(0):</span></span><br></pre></td></tr></table></figure>

<p>我们用<code>plotEnrichHeatmap()</code>函数基于<code>SummarizedExperiment</code>绘制富集热图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">heatmapATAC <span class="operator">&lt;-</span> plotEnrichHeatmap<span class="punctuation">(</span>enrichATAC<span class="punctuation">,</span> n <span class="operator">=</span> <span class="number">7</span><span class="punctuation">,</span> transpose <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>使用<code>ComplexHeatmap::draw()</code>绘制结果</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ComplexHeatmap<span class="operator">::</span>draw<span class="punctuation">(</span>heatmapATAC<span class="punctuation">,</span> heatmap_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">,</span> annotation_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>



<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-4ed3908df6d5445eb0fabf2cedc35067.png" alt="ATAC-Enriched-Marker-Heatmap"></p>
<p><code>plotFDF()</code>函数能够以可编辑的矢量版本保存图片。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>heatmapATAC<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;ATAC-Enriched-Marker-Heatmap&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">8</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">6</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="12-3-3-Codex-TFBS"><a href="#12-3-3-Codex-TFBS" class="headerlink" title="12.3.3 Codex TFBS"></a>12.3.3 Codex TFBS</h3><p>相同类型的分析还能用于 <a target="_blank" rel="noopener" href="http://codex.stemcells.cam.ac.uk/">CODEX</a> TFBS，只要设置<code>collection = &quot;Codex&quot;</code>即可</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addArchRAnnotations<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> collection <span class="operator">=</span> <span class="string">&quot;Codex&quot;</span><span class="punctuation">)</span></span><br><span class="line">enrichCodex <span class="operator">&lt;-</span> peakAnnoEnrichment<span class="punctuation">(</span></span><br><span class="line">    seMarker <span class="operator">=</span> markersPeaks<span class="punctuation">,</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    peakAnnotation <span class="operator">=</span> <span class="string">&quot;Codex&quot;</span><span class="punctuation">,</span></span><br><span class="line">    cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; Log2FC &gt;= 0.5&quot;</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line">heatmapCodex <span class="operator">&lt;-</span> plotEnrichHeatmap<span class="punctuation">(</span>enrichCodex<span class="punctuation">,</span> n <span class="operator">=</span> <span class="number">7</span><span class="punctuation">,</span> transpose <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">ComplexHeatmap<span class="operator">::</span>draw<span class="punctuation">(</span>heatmapCodex<span class="punctuation">,</span> heatmap_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">,</span> annotation_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>


<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-bf96a8e558384c578ea9922aa41d045f.png" alt="Codex-Enriched-Marker-Heatmap"></p>
<p>于是我们就保存图片了</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>heatmapCodex<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Codex-Enriched-Marker-Heatmap&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">8</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">6</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="12-4-自定义富集"><a href="#12-4-自定义富集" class="headerlink" title="12.4 自定义富集"></a>12.4 自定义富集</h2><p>除了之前这些经过人工审核的注释数据集，ArchR还能处理用户自定义注释信息来执行富集分析。接下来，我们会介绍如何根据ENCODE ChIP-seq实验来创建自定义的注释信息。</p>
<p>首先，我们先提供后续将被使用并下载的数据集，也可以提供本地文件。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">EncodePeaks <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">  Encode_K562_GATA1 <span class="operator">=</span> <span class="string">&quot;https://www.encodeproject.org/files/ENCFF632NQI/@@download/ENCFF632NQI.bed.gz&quot;</span><span class="punctuation">,</span></span><br><span class="line">  Encode_GM12878_CEBPB <span class="operator">=</span> <span class="string">&quot;https://www.encodeproject.org/files/ENCFF761MGJ/@@download/ENCFF761MGJ.bed.gz&quot;</span><span class="punctuation">,</span></span><br><span class="line">  Encode_K562_Ebf1 <span class="operator">=</span> <span class="string">&quot;https://www.encodeproject.org/files/ENCFF868VSY/@@download/ENCFF868VSY.bed.gz&quot;</span><span class="punctuation">,</span></span><br><span class="line">  Encode_K562_Pax5 <span class="operator">=</span> <span class="string">&quot;https://www.encodeproject.org/files/ENCFF339KUO/@@download/ENCFF339KUO.bed.gz&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>然后，我们用<code>addPeakAnnotation()</code>函数在<code>ArchRProject</code>函数中增加自定义注释。我们这里将其命名为”ChIP”</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addPeakAnnotations<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> regions <span class="operator">=</span> EncodePeaks<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;ChIP&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>和之前一样，我们使用<code>peakAnnoEnrichment()</code>函数根据自定义的注释信息执行peak注释富集分析</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">enrichRegions <span class="operator">&lt;-</span> peakAnnoEnrichment<span class="punctuation">(</span></span><br><span class="line">    seMarker <span class="operator">=</span> markersPeaks<span class="punctuation">,</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    peakAnnotation <span class="operator">=</span> <span class="string">&quot;ChIP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; Log2FC &gt;= 0.5&quot;</span></span><br><span class="line">  <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>并以相同的步骤生成注释热图。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">heatmapRegions <span class="operator">&lt;-</span> plotEnrichHeatmap<span class="punctuation">(</span>enrichRegions<span class="punctuation">,</span> n <span class="operator">=</span> <span class="number">7</span><span class="punctuation">,</span> transpose <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">ComplexHeatmap<span class="operator">::</span>draw<span class="punctuation">(</span>heatmapRegions<span class="punctuation">,</span> heatmap_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">,</span> annotation_legend_side <span class="operator">=</span> <span class="string">&quot;bot&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-78f7176114f249eb9a295863d2788e45.png" alt="Regions-Enriched-Marker-Heatmap"></p>
<p><code>plotFDF()</code>函数能够以可编辑的矢量版本保存图片。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>heatmapRegions<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Regions-Enriched-Marker-Heatmap&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">8</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">6</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>


      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/07/19/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%BA%8C%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter12/" data-id="cllgasdsx00afdkovcqt8c9kw" data-title="使用ArchR分析单细胞ATAC-seq数据(第十二章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第十一章)-analysis-sc-atac-seq-with-archr-chapter11" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/07/19/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%B8%80%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter11/" class="article-date">
  <time class="dt-published" datetime="2020-07-19T06:58:51.204Z" itemprop="datePublished">2020-07-19</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/07/19/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%B8%80%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter11/">使用ArchR分析单细胞ATAC-seq数据(第十一章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第11章-使用ArchR鉴定标记Peak"><a href="#第11章-使用ArchR鉴定标记Peak" class="headerlink" title="第11章: 使用ArchR鉴定标记Peak"></a>第11章: 使用ArchR鉴定标记Peak</h1><p>在讨论基因得分(gene score)这一章中，我们已经介绍了标记特征的鉴定。相同的函数<code>getMakerFeatures()</code>也能够用于从<code>ArchRProject</code>任意矩阵中鉴定标记特征。所谓的标记特征指的是相对于其他细胞分组唯一的特征。这些特征能帮助我们理解类群或者细胞类型特异的生物学现象。在这一章中，我们会讨论如何使用该功能鉴定标记peak。</p>
<h2 id="11-1-使用ArchR鉴定标记Peak"><a href="#11-1-使用ArchR鉴定标记Peak" class="headerlink" title="11.1: 使用ArchR鉴定标记Peak"></a>11.1: 使用ArchR鉴定标记Peak</h2><p>通常而言，我们想知道哪些peak是某个聚类或者某一些聚类所特有的。在ArchR中，这可以通过设置<code>addMarkFeatures()</code>函数的<code>useMatrix=&quot;PeakMatrix&quot;</code>来实现（无需监督）。</p>
<p>首先，我们需要再看一眼<code>projHeme5</code>中有哪些细胞类型，以及它们的相对比例</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">table<span class="punctuation">(</span>projHeme5<span class="operator">$</span>Cluster2<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>现在，让我们调用<code>getMarkerFeatures</code>参数，并设置<code>useMatrix=&quot;PeakMatrix&quot;</code>. 此外，为了降低不同细胞组之间的数据质量对结果的影响，我们可以设置<code>bias</code>参数，其中<code>bias = c(&quot;TSSEnrichment&quot;, &quot;log10(nFrags)&quot;)</code>就是用来避免TSS富集和每个细胞的fragment数对结果的影响。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">markersPeaks <span class="operator">&lt;-</span> getMarkerFeatures<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    useMatrix <span class="operator">=</span> <span class="string">&quot;PeakMatrix&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span></span><br><span class="line">  bias <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;TSSEnrichment&quot;</span><span class="punctuation">,</span> <span class="string">&quot;log10(nFrags)&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  testMethod <span class="operator">=</span> <span class="string">&quot;wilcoxon&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><code>getMarkerFeatures()</code>函数返回一个<code>SummarizedExperiment</code>对象，该对象包含一些不同的<code>assays</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">markerPeaks</span><br></pre></td></tr></table></figure>

<p>接着，我们可以用<code>getMarkers</code>函数从输出的<code>SummarizedExperiment</code>对象中提取我们感兴趣的部分。默认情况下，它会返回一个包含多个<code>DataFrame</code>的列表，不同的<code>DataFrame</code>表示来自不同的细胞分组。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">markerList <span class="operator">&lt;-</span> getMarkers<span class="punctuation">(</span>markersPeaks<span class="punctuation">,</span> cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.01 &amp; Log2FC &gt;= 1&quot;</span><span class="punctuation">)</span></span><br><span class="line">markerList</span><br></pre></td></tr></table></figure>

<p>如果我们对特定的一个细胞分组感兴趣，我们可以用<code>$</code>进行提取。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">markerList<span class="operator">$</span>Erythroid</span><br></pre></td></tr></table></figure>

<p>除了返回一个包含多个<code>DataFrame</code>的列表外，我们还可以用<code>getMarkers()</code>返回一个<code>GRangesList</code>，只要设置<code>returnGR=TRUE</code>即可。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">markerList <span class="operator">&lt;-</span> getMarkers<span class="punctuation">(</span>markersPeaks<span class="punctuation">,</span> cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.01 &amp; Log2FC &gt;= 1&quot;</span><span class="punctuation">,</span> returnGR <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">markerList</span><br></pre></td></tr></table></figure>

<p>这个<code>GRangesList</code>同样可以用<code>$</code>提取特定细胞组的结果，返回的是一个<code>GRanges</code>对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">markerList<span class="operator">$</span>Erythroid</span><br></pre></td></tr></table></figure>

<h2 id="11-2-在ArchR中绘制Marker-Peaks"><a href="#11-2-在ArchR中绘制Marker-Peaks" class="headerlink" title="11.2 在ArchR中绘制Marker Peaks"></a>11.2 在ArchR中绘制Marker Peaks</h2><p>ArchR提供了许多绘图函数用于<code>getMarkerfeatuers()</code>返回的<code>SummarizedExperiment</code>对象的可视化。</p>
<h3 id="11-2-1-Marker-Peak-Heatmap"><a href="#11-2-1-Marker-Peak-Heatmap" class="headerlink" title="11.2.1 Marker Peak Heatmap"></a>11.2.1 Marker Peak Heatmap</h3><p><code>markerHeatmap</code>能以热图的形式展示标记Marker Peak（或者其他<code>getMarkerFeatures()</code>输出的特征）</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">heatmapPeaks <span class="operator">&lt;-</span> markerHeatmap<span class="punctuation">(</span></span><br><span class="line">  seMarker <span class="operator">=</span> markersPeaks<span class="punctuation">,</span> </span><br><span class="line">  cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; Log2FC &gt;= 0.5&quot;</span><span class="punctuation">,</span></span><br><span class="line">  transpose <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>使用<code>draw</code>函数绘制结果</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>heatmapPeaks<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Peak-Marker-Heatmap&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">8</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">6</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-1459d659891f41b4a8f488ceb1fc8ee1.png" alt="Marker Peak Heatmap"></p>
<p><code>plotFDF()</code>函数能够以可编辑的矢量版本保存图片</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>heatmapPeaks<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Peak-Marker-Heatmap&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">8</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">6</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="11-2-2-Marker-Peak-MA和火山图"><a href="#11-2-2-Marker-Peak-MA和火山图" class="headerlink" title="11.2.2 Marker Peak MA和火山图"></a>11.2.2 Marker Peak MA和火山图</h3><p>除了绘制热图，我们也可以为每个细胞分组绘制MA或者火山图(volcano)。这些图可以用<code>markerPlot()</code>函数绘制。对于MA图，需要设置参数<code>plotAs=&quot;MA&quot;</code>. 我们以”Erythroid”细胞分组为例，设置参数<code>name = &quot;Erythroid&quot;</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pma <span class="operator">&lt;-</span> markerPlot<span class="punctuation">(</span>seMarker <span class="operator">=</span> markersPeaks<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Erythroid&quot;</span><span class="punctuation">,</span> cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; Log2FC &gt;= 1&quot;</span><span class="punctuation">,</span> plotAs <span class="operator">=</span> <span class="string">&quot;MA&quot;</span><span class="punctuation">)</span></span><br><span class="line">pma</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-904ba33c14ee4f6bab904eb1958a9732.png" alt="MA图"></p>
<p>同样的，只要设置<code>plotAs=&quot;Volcano&quot;</code>就可以绘制火山图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pv <span class="operator">&lt;-</span> markerPlot<span class="punctuation">(</span>seMarker <span class="operator">=</span> markersPeaks<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Erythroid&quot;</span><span class="punctuation">,</span> cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; Log2FC &gt;= 1&quot;</span><span class="punctuation">,</span> plotAs <span class="operator">=</span> <span class="string">&quot;Volcano&quot;</span><span class="punctuation">)</span></span><br><span class="line">pv</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-07e8fc062e764598a6ed6035b234baa5.png" alt="volcano plot"></p>
<p><code>plotFDF()</code>函数能够以可编辑的矢量版本保存图片。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>pma<span class="punctuation">,</span> pv<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Erythroid-Markers-MA-Volcano&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="11-2-3-Browser-Tracks的Marker-Peak"><a href="#11-2-3-Browser-Tracks的Marker-Peak" class="headerlink" title="11.2.3 Browser Tracks的Marker Peak"></a>11.2.3 Browser Tracks的Marker Peak</h3><p>此外，我们在基因组浏览器上检查这些peak区域，只需要为<code>plotBrowserTrack()</code>函数的<code>features</code>参数传入 相应的peak区间。这会额外在我们的ArchR track图的下方以BED形式展示marker peak区域。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">p <span class="operator">&lt;-</span> plotBrowserTrack<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    geneSymbol <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;GATA1&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    features <span class="operator">=</span>  getMarkers<span class="punctuation">(</span>markersPeaks<span class="punctuation">,</span> cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; Log2FC &gt;= 1&quot;</span><span class="punctuation">,</span> returnGR <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="string">&quot;Erythroid&quot;</span><span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">    upstream <span class="operator">=</span> <span class="number">50000</span><span class="punctuation">,</span></span><br><span class="line">    downstream <span class="operator">=</span> <span class="number">50000</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们使用<code>grid::grid.draw()</code>绘制结果</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grid<span class="operator">::</span>grid.newpage<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">grid<span class="operator">::</span>grid.draw<span class="punctuation">(</span>p<span class="operator">$</span>GATA1<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-f86bc96dc4354f56aeba7db49c4fb4c9.png" alt="browser track"></p>
<p><code>plotFDF()</code>函数能够以可编辑的矢量版本保存图片。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-Tracks-With-Features&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="11-3-组间配对检验"><a href="#11-3-组间配对检验" class="headerlink" title="11.3 组间配对检验"></a>11.3 组间配对检验</h2><p>标记特征鉴定是一种特别的差异表达检验。此外，使用相同的<code>getMarkerFeatures()</code>函数也能实现标准化的差异分析。我们只需要设置<code>useGroup</code>为一组细胞，然后设置<code>bgdGroup</code>为另一组细胞即可。这就可以对给定两组进行差异分析。在这些差异分析中，在<code>useGroups</code>比较高的peak的倍数变化值是正数，在<code>bgdGroups</code>比较高的peak则是由负的倍数变化值。</p>
<p>这里，我们对”Erythroid”与”Progenitor”细胞组进行配对检验。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">markerTest <span class="operator">&lt;-</span> getMarkerFeatures<span class="punctuation">(</span></span><br><span class="line">  ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">  useMatrix <span class="operator">=</span> <span class="string">&quot;PeakMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">  groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span></span><br><span class="line">  testMethod <span class="operator">=</span> <span class="string">&quot;wilcoxon&quot;</span><span class="punctuation">,</span></span><br><span class="line">  bias <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;TSSEnrichment&quot;</span><span class="punctuation">,</span> <span class="string">&quot;log10(nFrags)&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  useGroups <span class="operator">=</span> <span class="string">&quot;Erythroid&quot;</span><span class="punctuation">,</span></span><br><span class="line">  bgdGroups <span class="operator">=</span> <span class="string">&quot;Progenitor&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>使用<code>markerPlot()</code>函数可以绘制MA或者火山图。MA图需要设置<code>plotAs=&quot;MA&quot;</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pma <span class="operator">&lt;-</span> markerPlot<span class="punctuation">(</span>seMarker <span class="operator">=</span> markerTest<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Erythroid&quot;</span><span class="punctuation">,</span> cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; abs(Log2FC) &gt;= 1&quot;</span><span class="punctuation">,</span> plotAs <span class="operator">=</span> <span class="string">&quot;MA&quot;</span><span class="punctuation">)</span></span><br><span class="line">pma</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-e9db62e1bdbc4c709d979c08dbef8cb6.png" alt="group marker peak MA plot"></p>
<p>火山图需要设置<code>plotAs=&quot;Volcano&quot;</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pv <span class="operator">&lt;-</span> markerPlot<span class="punctuation">(</span>seMarker <span class="operator">=</span> markerTest<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Erythroid&quot;</span><span class="punctuation">,</span> cutOff <span class="operator">=</span> <span class="string">&quot;FDR &lt;= 0.1 &amp; abs(Log2FC) &gt;= 1&quot;</span><span class="punctuation">,</span> plotAs <span class="operator">=</span> <span class="string">&quot;Volcano&quot;</span><span class="punctuation">)</span></span><br><span class="line">pv</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-e4610c5ece3c47718f4514cdc17cd85b.png" alt="group marker peak volcano"></p>
<p><code>plotFDF()</code>函数能够以可编辑的矢量版本保存图片。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>pma<span class="punctuation">,</span> pv<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Erythroid-vs-Progenitor-Markers-MA-Volcano&quot;</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>在后续章节中，我们还会进行差异分析，因为会在我们的差异开放的peak中搜索富集的motif。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/07/19/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%B8%80%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter11/" data-id="cllgasdsv00a8dkovfkdr1k8x" data-title="使用ArchR分析单细胞ATAC-seq数据(第十一章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第十章)-analysis-sc-atac-seq-with-archr-chapter10" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/07/12/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter10/" class="article-date">
  <time class="dt-published" datetime="2020-07-12T23:15:33.294Z" itemprop="datePublished">2020-07-12</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/07/12/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter10/">使用ArchR分析单细胞ATAC-seq数据(第十章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第10章-使用ArchR识别Peak"><a href="#第10章-使用ArchR识别Peak" class="headerlink" title="第10章: 使用ArchR识别Peak"></a>第10章: 使用ArchR识别Peak</h1><p>识别peak(Peak calling)是ATAC-seq数据分析的常规操作之一。因为每个细胞的scATAC-seq数据都只有两种状态（开放或不开放），所以我们不能从单个细胞中识别peak。因此，我们在之前的章节中先对细胞进行了分组（细胞聚类）。此外，我们还构建了拟混池重复(pseduo-bulk replicates)，用于评估被识别的peak的可重复性。</p>
<h2 id="10-1-迭代重叠Peak合并流程"><a href="#10-1-迭代重叠Peak合并流程" class="headerlink" title="10.1 迭代重叠Peak合并流程"></a>10.1 迭代重叠Peak合并流程</h2><p>我们先介绍<a target="_blank" rel="noopener" href="https://science.sciencemag.org/content/362/6413/eaav1898">2018年的迭代重叠Peak合并流程</a>，然后逐一介绍其他策略（这些策略都存在着一些问题）。</p>
<h3 id="10-1-1-等宽和不定宽peak"><a href="#10-1-1-等宽和不定宽peak" class="headerlink" title="10.1.1 等宽和不定宽peak"></a>10.1.1 等宽和不定宽peak</h3><p>我们之所以选择501-bp的等宽peak，一方面是下游处理不再需要根据peak宽度进行标准化从而简化了计算，另一方面是ATAC-seq的大部分peak宽度都低于501-bp。如果使用不定宽peak，后续不同样本的peak合并就会变得特别复杂。而且，我们也没有觉得使用不定宽peak带来的潜在好处值得我们花费那么多精力。更进一步的说，无论你选择哪种，大部分的分析结果也都不会受到影响。</p>
<p>下面，我们会以几个含有不同peak的细胞类型为例，用来讲解不同peak合并方法的差异。</p>
<h3 id="10-1-2-Raw-Peak-Overlap"><a href="#10-1-2-Raw-Peak-Overlap" class="headerlink" title="10.1.2: Raw Peak Overlap"></a>10.1.2: Raw Peak Overlap</h3><p><strong>raw peak overlap</strong>就是分析不同细胞类型的peak是否有重叠，如果有将其合并成单个较大的peak。在如下的图解中，我们会发现尽管细胞类型A和C的前两个peak没有直接相连，但是由于细胞类型B中第一个peak和细胞类型A和C的前两个peak相互重叠，最终使得三个peak合并成一个更大peak。除了这个问题外，如果你想跟踪peak的顶点，你要么记录每一个合并后的新peak的顶点，要么记录每个合并后peak对应的原peak的顶点。</p>
<p>最后，这种类型的peak合并方法通常用<code>bedtools merge</code>命令实现。</p>
<img src="/upload/2020/07/peakCalling_RawOverlap-a52800f5365a473a9626ccd63aca0755.png" alt="img" style="zoom: 25%;" />

<h3 id="10-1-3-Clustered-Overlap"><a href="#10-1-3-Clustered-Overlap" class="headerlink" title="10.1.3: Clustered Overlap"></a>10.1.3: Clustered Overlap</h3><p><strong>clustered overlap</strong>方法将peak进行聚类，然后根据规则挑选其中一个胜者。我们可以使用<code>bedtools cluster</code>命令对peak进行聚类，然后挑选每个peak聚类中最显著的peak。根据我们的经验，它会漏掉附近较小的peak，导致总体peak数减少。</p>
<img src="/upload/2020/07/peakCalling_ClusteredOverlap-fb6ee0d46e45431785f5b5a11ef1ebaf.png" alt="img" style="zoom:25%;" />

<h3 id="10-1-4-Iterative-Overlap"><a href="#10-1-4-Iterative-Overlap" class="headerlink" title="10.1.4 Iterative Overlap"></a>10.1.4 Iterative Overlap</h3><p><strong>iterative overlap</strong>尽量避免了以上提到的问题。首先根据显著性对peak进行排序。接着我们保留其中最显著的peak，并移除掉所有于最显著peak发生重叠的peak。然后，在所有余下的peak中，我们会重复上述步骤，直到没有peak为止。这个能够避免10.1.2提到的问题，同时还能使用固定宽度peak。</p>
<img src="/upload/2020/07/peakCalling_IterativeOverlap-e01c21c57f4a4493adef39a398beb44b.png" style="zoom:25%;" />

<h3 id="10-1-5-不同方法对比"><a href="#10-1-5-不同方法对比" class="headerlink" title="10.1.5 不同方法对比"></a>10.1.5 不同方法对比</h3><p>我们很容易从最终的peak集中看到以上这些方法的区别。我们认为，<strong>iterative overlap</strong>策略能够较少的代价得到最好的peak集。</p>
<img src="/upload/2020/07/peakCalling_Comparison-c8b214a7b733438298af50c3ff7dca3c.png" alt="comparison" style="zoom:25%;" />

<h3 id="10-1-6-ArchR的工作方式"><a href="#10-1-6-ArchR的工作方式" class="headerlink" title="10.1.6 ArchR的工作方式"></a>10.1.6 ArchR的工作方式</h3><p><strong>iterative overlap</strong>这种以分级形式对数据进行合并，尽可能保留所有细胞类型特异的peak。</p>
<p>想象一下，你有3种细胞类型，A，B和C，每一种细胞类型多有3个拟混池重复。ArchR使用函数<code>addReproduciblePeakSet()</code>实现<strong>iterative overlap</strong>的peak合并流程。首先，ArchR单独为每个拟混池重复鉴定peak。然后ArchR同时分析每个细胞类型的所有拟混池重复，进行第一次的<strong>iterative overlap</strong>过滤。需要注意的是，ArchR使用标准化的peak显著性分值矩阵去比较不同样本的peak的显著性。因为有报道称MACS2的显著性和测序深度成比例关系，所以无法直接比较不同样本的peak显著性。在第一轮的<strong>iterative overlap</strong>过滤后，ArchR检查所有拟混池重复中每个peak的可重复性，只保留符合参数<code>reproducibility</code>的peak。最后，3种细胞类型(A, B, C)每一个都有单独的合并后的peak集。</p>
<p>接着，我们重复以上步骤，用于合并细胞类型A, B 和C的peak集。在这一步，我们还会根据不同的细胞类型对peak的显著性进行标准化，然后执行<strong>iterative overlap</strong>过滤。</p>
<p>最终我们会得到单个固定宽度的peak数据集。</p>
<h3 id="10-1-7-如何更改策略"><a href="#10-1-7-如何更改策略" class="headerlink" title="10.1.7 如何更改策略"></a>10.1.7 如何更改策略</h3><p>加入你不想用<strong>iterative overlap</strong>合并策略，那么你可以跳过<code>addReproduciblePeakSet()</code>，或者使用<code>ArchRProj &lt;- addPeakSet()</code>添加自己的peak数据集</p>
<h2 id="10-2-使用w-MACS2鉴定peak"><a href="#10-2-使用w-MACS2鉴定peak" class="headerlink" title="10.2 使用w&#x2F;MACS2鉴定peak"></a>10.2 使用w&#x2F;MACS2鉴定peak</h2><p>如上所述，我们使用ArchR的<code>addReproduciblePeakSet()</code>得到可重复的peak。默认情况下，ArchR会使用MACS2鉴定peak，此外ArchR也实现它原生的peak鉴定工具，用在MACS2无法安装的情况（例如，我们无法在Windows上安装MACS2），该可选peak鉴定方法会在下一节介绍。</p>
<p>为了使用MACS2鉴定peak，ArchR必须能够找到MACS2的执行路径。ArchR会先在你的<code>PATH</code>环境变量进行查找。如果找不到，ArchR会尝试确认你是否用<code>pip</code>或<code>pip3</code>安装过MACS2. 如果以上都没有陈宫，ArchR就会放弃并输出报错信息。如果你安装了MACS2，但是ArchR找不到它，那么你需要在<code>addReproduciblePeakSet()</code>函数中设置<code>pathToMacs2</code>参数。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">pathToMacs2 <span class="operator">&lt;-</span> findMacs2<span class="punctuation">(</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这是成功的信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## Searching For MACS2..</span></span><br><span class="line"><span class="comment">## Found with $path!</span></span><br></pre></td></tr></table></figure>

<p>如下是失败信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Searching For MACS2..</span><br><span class="line">Not Found <span class="keyword">in</span> <span class="variable">$PATH</span></span><br><span class="line">Not Found with pip</span><br><span class="line">Not Found with pip3</span><br><span class="line">Error <span class="keyword">in</span> findMacs2() : </span><br><span class="line">  Could Not Find Macs2! Please install w/ pip, add to your <span class="variable">$PATH</span>, or just supply the macs2 path directly and avoid this <span class="keyword">function</span>!</span><br></pre></td></tr></table></figure>

<p>当你有MACS2的位置信息后，我们接着就可以构建可重复的合并peak数据集(约5到10分钟)。为了避免细胞过少的拟混池重复的影响，我们可以通过设置<code>peaksPerCell</code>参数来限制每个细胞的peak数上限。这能避免细胞数过少的peak给最终合并的peak集贡献低质量的peak。此外，<code>addReproduciblePeakSet()</code>还有许多其他的参数可以调整，用<code>?addReproduciblePeakSet</code>可以看到更多信息。</p>
<p>每个<code>ArchRProject</code>项目只能包括一个peak集。我们最后将<code>addReproduciblePeakSet()</code>的输出赋值给我们想要的<code>ArchRProject()</code>。如果你想要测试不同的peak集，你必须先保存一份<code>ArchRProject()</code>并拷贝Arrow文件。尽管这会消耗更多的硬盘存储，但这是Arrow文件的会保存peak矩阵信息，因此无法避免。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">projHeme4 <span class="operator">&lt;-</span> addReproduciblePeakSet<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme4<span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    pathToMacs2 <span class="operator">=</span> pathToMacs2</span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们使用<code>getPeakSet()</code>函数以<code>GRanges</code>对象提取peak集。该peak集对象包括每个peak最初来源信息。然而这并不是意味着该peak只是在那个组中出现，而是意味着那组中的该peak拥有最高的标准化后显著性。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">getPeakSet<span class="punctuation">(</span>projHeme4<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="10-3鉴定Peaks-w-TileMatrix"><a href="#10-3鉴定Peaks-w-TileMatrix" class="headerlink" title="10.3鉴定Peaks w&#x2F;TileMatrix"></a>10.3鉴定Peaks w&#x2F;TileMatrix</h2><p>如上所述，ArchR提供了原生的peak鉴定工具。尽管经过测试，该工具效果和MACS2表现差不多，但除非实在是没有办法，否则都不建议使用这个原生方法。</p>
<p>ArchR的原生peak鉴定工具基于500-bp的<code>TileMatrix</code>鉴定peak，可以通过设置<code>addReproduciblePeakSet()</code>函数的<code>peakMethod</code>参数来进行调用。注意，我们没有将输出保存在<code>projHeme4</code>对象中，因为我们这里只是测试而已，所以不打算保留这个peak集。将其保存在<code>ArchRProject</code>对象会覆盖之前已经存放在<code>projHeme4</code>的peak集。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">projHemeTmp <span class="operator">&lt;-</span> addReproduciblePeakSet<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme4<span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span></span><br><span class="line">    peakMethod <span class="operator">=</span> <span class="string">&quot;Tiles&quot;</span><span class="punctuation">,</span></span><br><span class="line">    method <span class="operator">=</span> <span class="string">&quot;p&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们同样可以检查这个合并后的peak集</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">getPeakSet<span class="punctuation">(</span>projHemeTmp<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="10-3-1-比较两种方法"><a href="#10-3-1-比较两种方法" class="headerlink" title="10.3.1 比较两种方法"></a>10.3.1 比较两种方法</h3><p>我们可以根据重叠peak的百分比等指标来比较MACS2和ArchR自带的<code>TileMatrix</code>方法的差异。</p>
<p>1，我们检查MACS2鉴定的peak和<code>TileMatrix</code>鉴定的peak的重叠情况</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">length</span><span class="punctuation">(</span>subsetByOverlaps<span class="punctuation">(</span>getPeakSet<span class="punctuation">(</span>projHeme4<span class="punctuation">)</span><span class="punctuation">,</span> getPeakSet<span class="punctuation">(</span>projHemeTmp<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">/</span> <span class="built_in">length</span><span class="punctuation">(</span>getPeakSet<span class="punctuation">(</span>projHeme4<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">### 0.9627246</span></span><br></pre></td></tr></table></figure>

<p>2，我们检查<code>TileMatrix</code>鉴定的peak和MACS2鉴定的peak的重叠情况。我们发现这个重叠比例并不高。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">length</span><span class="punctuation">(</span>subsetByOverlaps<span class="punctuation">(</span>getPeakSet<span class="punctuation">(</span>projHemeTmp<span class="punctuation">)</span><span class="punctuation">,</span> getPeakSet<span class="punctuation">(</span>projHeme4<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">/</span> <span class="built_in">length</span><span class="punctuation">(</span>getPeakSet<span class="punctuation">(</span>projHemeTmp<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">### 0.7533365</span></span><br></pre></td></tr></table></figure>

<p>如果我们增加peak的边界(从500-bp提高到1000)，MACS2鉴定的peak和<code>TileMatrix</code>鉴定的peak的重叠比例几乎没变化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">length</span><span class="punctuation">(</span>subsetByOverlaps<span class="punctuation">(</span>resize<span class="punctuation">(</span>getPeakSet<span class="punctuation">(</span>projHeme4<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">1000</span><span class="punctuation">,</span> <span class="string">&quot;center&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> getPeakSet<span class="punctuation">(</span>projHemeTmp<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">/</span> <span class="built_in">length</span><span class="punctuation">(</span>getPeakSet<span class="punctuation">(</span>projHeme4<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">### 0.9676687</span></span><br></pre></td></tr></table></figure>

<p>但是增加了<code>TileMatrix</code>的peak和MACS2找到的peak的重叠比例。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">length</span><span class="punctuation">(</span>subsetByOverlaps<span class="punctuation">(</span>getPeakSet<span class="punctuation">(</span>projHemeTmp<span class="punctuation">)</span><span class="punctuation">,</span> resize<span class="punctuation">(</span>getPeakSet<span class="punctuation">(</span>projHeme4<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">1000</span><span class="punctuation">,</span> <span class="string">&quot;center&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">/</span> <span class="built_in">length</span><span class="punctuation">(</span>getPeakSet<span class="punctuation">(</span>projHemeTmp<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">### 0.8287639</span></span><br></pre></td></tr></table></figure>

<p>来自译者的话: 我们从中可以得出，<code>TileMatrix</code>能鉴定的peak，MACS2也差不多都能鉴定到，但是MACS2鉴定出来的peak，未必<code>TileMatrix</code>也能鉴定。此外，<code>TileMatrix</code>找到的一些peak和MACS2找到的peak也比较近，因此通过延长peak的宽度，就能提高比例。</p>
<blockquote>
<p>在10.1这个这个章节中，作者用到了菊花链拓扑(daisy chain)这个名词用来描述不同peak相互连接的情况。考虑到这个菊花链并非是一个常用名词，因此我通过意译来翻译这些名词。</p>
</blockquote>
<h2 id="10-4-增加Peak矩阵"><a href="#10-4-增加Peak矩阵" class="headerlink" title="10.4 增加Peak矩阵"></a>10.4 增加Peak矩阵</h2><p>我们可以用<code>saveArchRProject()</code>函数保存原来的<code>projHeme4</code>对象。该<code>ArchRProject</code>包含来源于MACS2的合并后peak集。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">saveArchRProject<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme4<span class="punctuation">,</span> outputDirectory <span class="operator">=</span> <span class="string">&quot;Save-ProjHeme4&quot;</span><span class="punctuation">,</span> load <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>为了方便下游分析，我们可以创建新的<code>ArchRProject</code>命名为<code>projHeme5</code>，并增加新的矩阵，该矩阵记录这新的合并后peak集的insertion计数。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addPeakMatrix<span class="punctuation">(</span>projHeme4<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>于是，<code>projHeme5</code>现在又多了一个新的矩阵，名为”Peakmatrix”，这是和”GeneScoreMatrix”和”TileMatrix”类似的<strong>专用矩阵变量名</strong>。正如之前所说，每个<code>ArchRProject</code>对象只能有一个peak集和一个<code>PeakMatrix</code>。当然，你可以不同命名构建无限个自定义的特征矩阵，但是<code>PeakMatrix</code>只能有一个，它就是用来保存来自于peak集的insertion计数矩阵。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">getAvailableMatrices<span class="punctuation">(</span>projHeme5<span class="punctuation">)</span></span><br><span class="line"><span class="comment">## [1] &quot;GeneIntegrationMatrix&quot; &quot;GeneScoreMatrix&quot; &quot;PeakMatrix&quot;</span></span><br><span class="line"><span class="comment">## [4] &quot;TileMatrix&quot;</span></span><br></pre></td></tr></table></figure>







      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/07/12/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter10/" data-id="cllgasdt100atdkov7mv7d7kr" data-title="使用ArchR分析单细胞ATAC-seq数据(第十章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-MAKER避免重复运算-maker-redue-redundant-computations" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/07/07/MAKER%E9%81%BF%E5%85%8D%E9%87%8D%E5%A4%8D%E8%BF%90%E7%AE%97-maker-redue-redundant-computations/" class="article-date">
  <time class="dt-published" datetime="2020-07-07T07:15:44.060Z" itemprop="datePublished">2020-07-07</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/07/07/MAKER%E9%81%BF%E5%85%8D%E9%87%8D%E5%A4%8D%E8%BF%90%E7%AE%97-maker-redue-redundant-computations/">MAKER避免重复运算</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="MAKER深入篇-如何避免重复运算"><a href="#MAKER深入篇-如何避免重复运算" class="headerlink" title="MAKER深入篇-如何避免重复运算"></a>MAKER深入篇-如何避免重复运算</h1><p>通常而言，我们会运行不只一轮的MAKER。如果参考组序列没有变化，那么有一些计算只需要做一次就行了，例如将EST, Repeat和Protein序列比对到参考基因组，得到它们对应的位置。</p>
<p>我们有三种方法可以避免不必要的运算，第一种方法是直接修改配置文件，让MAKER重复利用之前的运行结果；第二种方式是利用之前输出的GFF文件，通过配置”Re-annotation Using MAKER Derived GFF3”里的选项来跳过对应的计算；第三种方法于是利用之前输出的GFF文件，从中提取EST&#x2F;Repeat&#x2F;Protein的位置信息保存为GFF文件，通过配置”est_gff”, “protein_gff”, “rm_gff”来避免重新计算位置信息。</p>
<p>后续分析建立<a href="/archives/maker-train-snap-model">MAKER高级篇-SNAP模型训练</a>基础上，也就是通过protein和est序列直接输出基因模型，然后训练出初步的HMM模型</p>
<h2 id="方法1"><a href="#方法1" class="headerlink" title="方法1"></a>方法1</h2><p>方法1最为简单，我们只需要修改之前的<code>maker_opts.ctl</code>里的参数，然后重新运行即可。运行时会输出如下的警告信息。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-eca66a656c5e4256ae3418fca282d92c.png" alt="警告信息"></p>
<p><strong>注意</strong>: MAKER是通过对比<code>maker_opt.ctl</code>里的配置信息和自己运行时记录的<code>maker_opts.log</code>来判断哪些参数发生了改变。因此，如果SNAP第二次训练生成的文件，要是和上一次命名相同，那么它会认为你这次输入的模型文件和上次相同，就会跳过SNAP预测这一步。</p>
<p>实际运行时，MAKER会跳过BLAST步骤，但是依旧会调用”exonerate”来处理BLAST结果。</p>
<h2 id="方法2"><a href="#方法2" class="headerlink" title="方法2"></a>方法2</h2><p>如果你不小心把maker的输出文件删掉了，但是你保留着之前<code>gff3_merge</code>默认参数输出的文件，那么你可以使用该文件来跳过BLAST和Exonerate运算。</p>
<p>Step1: 配置”Re-annotation Using MAKER Derived GFF3”里的参数</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#-----Re-annotation Using MAKER Derived GFF3</span></span><br><span class="line">maker_gff=round1.gff <span class="comment">#MAKER derived GFF3 file</span></span><br><span class="line">est_pass=1 <span class="comment">#use ESTs in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">altest_pass=1 <span class="comment">#use alternate organism ESTs in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">protein_pass=1 <span class="comment">#use protein alignments in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">rm_pass=1 <span class="comment">#use repeats in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">model_pass=1 <span class="comment">#use gene models in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">pred_pass=1 <span class="comment">#use ab-initio predictions in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">other_pass=1 <span class="comment">#passthrough anyything else in maker_gff: 1 = yes, 0 = no</span></span><br></pre></td></tr></table></figure>

<p>此处的<code>round1.gff</code>通过<code>gff3_merge</code>从上一论的maker输出中提取，代码如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gff3_merge -d genome.maker.output/genome_master_datastore_index.log -o round1.gff</span><br></pre></td></tr></table></figure>

<p>Step2: 将”EST Evidence”和”Protein Homology Evidence”里的配置清空，如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#-----EST Evidence (for best results provide a file for at least one)</span></span><br><span class="line">est= <span class="comment">#set of ESTs or assembled mRNA-seq in fasta format</span></span><br><span class="line">altest= <span class="comment">#EST/cDNA sequence file in fasta format from an alternate organismest_gff= #aligned ESTs or mRNA-seq from an external GFF3 file</span></span><br><span class="line">altest_gff= <span class="comment">#aligned ESTs from a closly relate species in GFF3 format</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#-----Protein Homology Evidence (for best results provide a file for at least one)</span></span><br><span class="line">protein=  <span class="comment">#protein sequence file in fasta format (i.e. from mutiple organisms)</span></span><br><span class="line">protein_gff=  <span class="comment">#aligned protein homology evidence from an external GFF3 file</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#-----Repeat Masking (leave values blank to skip repeat masking)</span></span><br><span class="line">model_org= <span class="comment">#select a model organism for RepBase masking in RepeatMasker</span></span><br><span class="line">rmlib= <span class="comment">#provide an organism specific repeat library in fasta format for RepeatMasker</span></span><br><span class="line">repeat_protein= <span class="comment">#provide a fasta file of transposable element proteins for RepeatRunner</span></span><br><span class="line">rm_gff= <span class="comment">#pre-identified repeat elements from an external GFF3 file</span></span><br><span class="line">prok_rm=0 <span class="comment">#forces MAKER to repeatmask prokaryotes (no reason to change this), 1 = yes, 0 = no</span></span><br><span class="line">softmask=1 <span class="comment">#use soft-masking rather than hard-masking in BLAST (i.e. seg and dust filtering)</span></span><br></pre></td></tr></table></figure>

<p>Step3: 配置”Gene Prediction”，例如SNAP, 同时将”est2genome”和”protein2genome”设置为0</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#-----Gene Prediction</span></span><br><span class="line">snaphmm=snap.hmm <span class="comment">#SNAP HMM file</span></span><br><span class="line">gmhmm= <span class="comment">#GeneMark HMM file</span></span><br><span class="line">augustus_species= <span class="comment">#Augustus gene prediction species model</span></span><br><span class="line"><span class="comment"># 略过其他参数</span></span><br><span class="line">est2genome=0 <span class="comment">#infer gene predictions directly from ESTs, 1 = yes, 0 = no</span></span><br><span class="line">protein2genome=0 <span class="comment">#infer predictions from protein homology, 1 = yes, 0 = no</span></span><br><span class="line"><span class="comment"># 略过其他参数</span></span><br></pre></td></tr></table></figure>

<p>会跳过exonerate步骤，直接从snap预测开始。</p>
<h2 id="方法3"><a href="#方法3" class="headerlink" title="方法3"></a>方法3</h2><p>我们还可以通过设置<code>est2gff</code>, <code>protein_gff</code>和<code>rm_gff</code>，来避免重复序列屏蔽和BLAST+Exonerate运算</p>
<p>Step1: 从之前的MAKER输出的GFF文件种提取EST&#x2F;Protein&#x2F;Repeat的位置信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># transcript alignment</span></span><br><span class="line">awk <span class="string">&#x27;&#123; if ($2 ~ &quot;est&quot;) print $0 &#125;&#x27;</span> round1.gff &gt; est.gff</span><br><span class="line"><span class="comment"># protein alignments</span></span><br><span class="line">awk <span class="string">&#x27;&#123; if ($2 == &quot;protein2genome&quot;) print $0 &#125;&#x27;</span> round1.gff &gt; protein2genome.gff</span><br><span class="line"><span class="comment"># repeat alignments</span></span><br><span class="line">awk <span class="string">&#x27;&#123; if ($2 ~ &quot;repeat&quot;) print $0 &#125;&#x27;</span> round1.gff &gt; repeats.gff</span><br></pre></td></tr></table></figure>

<p>Step2: 修改EST Evidence &#x2F; rotein Homology Evidence &#x2F;Repeat Masking里的配置参数</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#-----EST Evidence (for best results provide a file for at least one)</span></span><br><span class="line">est= <span class="comment">#set of ESTs or assembled mRNA-seq in fasta format</span></span><br><span class="line">altest= <span class="comment">#EST/cDNA sequence file in fasta format from an alternate organismest_gff=est.gff #aligned ESTs or mRNA-seq from an external GFF3 file</span></span><br><span class="line">est_gff=./est.gff</span><br><span class="line">altest_gff= <span class="comment">#aligned ESTs from a closly relate species in GFF3 format</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#-----Protein Homology Evidence (for best results provide a file for at least one)</span></span><br><span class="line">protein=  <span class="comment">#protein sequence file in fasta format (i.e. from mutiple organisms)</span></span><br><span class="line">protein_gff=protein2genome.gff  <span class="comment">#aligned protein homology evidence from an external GFF3 file</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#-----Repeat Masking (leave values blank to skip repeat masking)</span></span><br><span class="line">model_org= <span class="comment">#select a model organism for RepBase masking in RepeatMasker</span></span><br><span class="line">rmlib= <span class="comment">#provide an organism specific repeat library in fasta format for RepeatMasker</span></span><br><span class="line">repeat_protein= <span class="comment">#provide a fasta file of transposable element proteins for RepeatRunner</span></span><br><span class="line">rm_gff=repeats.gff <span class="comment">#pre-identified repeat elements from an external GFF3 file</span></span><br><span class="line">prok_rm=0 <span class="comment">#forces MAKER to repeatmask prokaryotes (no reason to change this), 1 = yes, 0 = no</span></span><br><span class="line">softmask=1 <span class="comment">#use soft-masking rather than hard-masking in BLAST (i.e. seg and dust filtering)</span></span><br></pre></td></tr></table></figure>

<p>Step3: 配置”Gene Prediction”，例如SNAP, 同时将”est2genome”和”protein2genome”设置为0</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#-----Gene Prediction</span></span><br><span class="line">snaphmm=snap.hmm <span class="comment">#SNAP HMM file</span></span><br><span class="line">gmhmm= <span class="comment">#GeneMark HMM file</span></span><br><span class="line">augustus_species= <span class="comment">#Augustus gene prediction species model</span></span><br><span class="line"><span class="comment"># 略过其他参数</span></span><br><span class="line">est2genome=0 <span class="comment">#infer gene predictions directly from ESTs, 1 = yes, 0 = no</span></span><br><span class="line">protein2genome=0 <span class="comment">#infer predictions from protein homology, 1 = yes, 0 = no</span></span><br><span class="line"><span class="comment"># 略过其他参数</span></span><br></pre></td></tr></table></figure>

<p>同样也会跳过exonerate步骤，直接从snap开始。</p>
<h2 id="结果比较"><a href="#结果比较" class="headerlink" title="结果比较"></a>结果比较</h2><p>对于这三种方法，从运行日志中看，三者都会跳过重复序列屏蔽，将EST和蛋白序列回帖到参考基因组的步骤，然而最终预测的基因数却不一致。</p>
<p>分析方法2和方法1的输出GFF文件时，发现方法2输出包括<code>exonerate_protein2genome-gene</code>和<code>exonerate_est2genome-gene</code>。推测其原因在第二种方法的<code>model_pass</code>, <code>pred_passs</code>参数在设置为1时会使用之前est2genome和protein2genome输出的基因模型，而由于模型本身就来自于EST和Protein，就变成自我验证，于是输出结果就变多了。当设置<code>model_pass</code>, <code>pred_passs</code>参数为0时，最终保证方法2和方法1输出结果一致。</p>
<p>之后设置<code>model_pass</code>, <code>pred_passs</code>参数为0，然后比较方法1，方法2和方法3输出的GFF。我发现方法1和方法2的第二列信息完全相同，是blastn, blastx, est2genome, maker, protein2genome, repeatmasker, snap_masked, 而方法3的第二列为est_gff:est2genome, maker, protein_gff:protein2genome, repeat_gff:repeatmasker, snap_masked.  目前只能推测是MAKER对这些证据使用方式不同引起了最终输出结果的差异，但具体的原理我没有分析清楚，不过不妨碍使用。</p>
<p>最后，三种方法使用优先级分别是方法1  &gt; 方法2 &gt; 方法3，其中方法2要注意设置<code>model_pass</code>, <code>pred_passs</code>的设置。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/07/07/MAKER%E9%81%BF%E5%85%8D%E9%87%8D%E5%A4%8D%E8%BF%90%E7%AE%97-maker-redue-redundant-computations/" data-id="cllgasdpl002mdkov2mbkfqpp" data-title="MAKER避免重复运算" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-MAKER训练SNAP基因模型-maker-train-snap-model" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/07/06/MAKER%E8%AE%AD%E7%BB%83SNAP%E5%9F%BA%E5%9B%A0%E6%A8%A1%E5%9E%8B-maker-train-snap-model/" class="article-date">
  <time class="dt-published" datetime="2020-07-06T00:09:21.893Z" itemprop="datePublished">2020-07-06</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/07/06/MAKER%E8%AE%AD%E7%BB%83SNAP%E5%9F%BA%E5%9B%A0%E6%A8%A1%E5%9E%8B-maker-train-snap-model/">MAKER训练SNAP基因模型</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h2 id="准备阶段"><a href="#准备阶段" class="headerlink" title="准备阶段"></a>准备阶段</h2><p>训练SNAP模型，需要准备三个文件，分别是参考基因组序列，组装的转录本序列和同源蛋白序列。</p>
<p>对于参考基因组序列，我们要保证N50需要超过预期基因长度的中位数，否则注释效果不好。不过目前的基因组在三代测序的加持下，基本上都不是问题。</p>
<p>对于同源蛋白， 建议只用UniProt&#x2F;Sprot的人工检查过的高质量蛋白序列，而不是盲目参考文献，使用近源物种的所有蛋白。除非你的近源物种都是模式物种，蛋白序列可靠性高，否则用错误的输入进行训练，数据越多反而错的越多。</p>
<p>我们可以在<a href="ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/">ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/</a>选择合适的uniprot_sprot数据,  然后将其输出为fasta格式。以植物为例</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_sprot_plants.dat.gz</span><br><span class="line">zcat uniprot_sprot_plants.dat.gz |\</span><br><span class="line">    awk <span class="string">&#x27;&#123;if (/^ /) &#123;gsub(/ /, &quot;&quot;); print&#125; else if (/^AC/) print &quot;&gt;&quot; $2&#125;&#x27;</span> |\</span><br><span class="line">    sed <span class="string">&#x27;s/;$//&#x27;</span>&gt; protein.fa</span><br></pre></td></tr></table></figure>

<p>对于转录本，我们通常会测一些转录组数据，有三种策略可以得到转录本。（这里暂时不考虑三代全长转录本）</p>
<ol>
<li>Trinity重头组装转录本</li>
<li>使用STAR + Trinity 获取转录本</li>
<li>使用STAR + StringTie  + gffread 获取转录本</li>
</ol>
<p>对于这三种策略，不推荐策略一，因为在有参考基因组的情况下，策略二不但计算效率高，而且能避免组装错误（多倍体等位基因之间相似度高）。对于策略二和策略三，我会推荐策略三。因为对于靠的比较近的基因，Trinity很可能会把这两个基因装成一个。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-a2927f2e13ff4a24b7fab8da45207278.png" alt="Fig1"></p>
<p>并且利用该转录本作为输入训练SNAP模型，之后以SNAP模型作为输入，将转录组和同源蛋白作为证据而不是直接用作模型，我们再检查maker的结果, 也会发现使用StringTie进行组装的结果才是对的。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/07/image-5197490a14354bdd9e58e75a59468114.png" alt="Fig2"></p>
<p>因此使用策略二不但计算量大，而且有些情况下还会导致过近的转录本错误融合，反而影响了最终效果，因此我最终推荐<strong>策略三</strong>。当然，这是我定性通过IGV浏览结果得出的结论，样本小，结论未必可靠，仅供参考。</p>
<h2 id="训练阶段"><a href="#训练阶段" class="headerlink" title="训练阶段"></a>训练阶段</h2><p>假设我们准备的三个文件分别命名为, genome.fa, Trinity-GG.fasta 和 protein.fa</p>
<p>接着使用<code>maker -CTL</code>新建配置文件, 设置如下选项</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">genome=genome.fa</span><br><span class="line">est=Trinity-GG.fasta</span><br><span class="line">protein=protein.fa</span><br><span class="line">est2genome=1</span><br><span class="line">protein2genome=1</span><br></pre></td></tr></table></figure>

<p>然后使用<code>mpiexec -n 线程数 maker &amp;&gt; run.log</code>运行程序。</p>
<p>处理结果后，我们新建一个snap目录训练模型</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> snap &amp;&amp; <span class="built_in">cd</span> snap</span><br><span class="line">gff3_merge -d ../genome.maker.output/genome_master_datastore_index.log</span><br></pre></td></tr></table></figure>

<p>使用makerzff构建输入文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">maker2zff -c 0.8 -e 0.8 -o 0.8 -x 0.2 genome.all.gff</span><br></pre></td></tr></table></figure>

<p><code>maker2zff</code>会根据AED(-x)和QI值进行过滤，其中QI值一共有9项，每一项的含义如下</p>
<ol>
<li>Length of the 5’ UTR</li>
<li>Fraction of splice sites confirmed by an EST alignment (<code>-c</code>)</li>
<li>Fraction of exons that overlap an EST alignmetn(<code>-e</code>)</li>
<li>Fraction of exons that overlap EST or Protein alignments(<code>-o</code>)</li>
<li>Fraction of splice site confrimed by a ab-initio prediction(<code>-a</code>)</li>
<li>Fraction of exons that overlap a ab-initio prediction(<code>-t</code>)</li>
<li>Number of exons in the mRNA</li>
<li>length of the 3’ UTR</li>
<li>Length of the protein sequence produced by the mRNA (<code>-l</code>)</li>
</ol>
<p>如果QI值第二项为-1，表示没有支持该剪切位点的EST证据.</p>
<p>接着构建模型</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">fathom -categorize 1000 genome.ann genome.dna</span><br><span class="line">fathom -<span class="built_in">export</span> 1000 -plus uni.ann uni.dna</span><br><span class="line">forge export.ann export.dna</span><br><span class="line">hmm-assembler.pl snap . &gt; ../snap.hmm</span><br></pre></td></tr></table></figure>

<p>修改配置，然后重新运行。MAKER会自动处理冲突的部分，避免重复序列屏蔽等的一些重复计算。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">genome=genome.fa</span><br><span class="line">est=Trinity-GG.fasta</span><br><span class="line">protein=protein.fa</span><br><span class="line">snap=snap.hmm</span><br><span class="line">est2genome=0</span><br><span class="line">protein2genome=0</span><br></pre></td></tr></table></figure>

<p>根据输出结果再一次训练模型</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> snap2 &amp;&amp; <span class="built_in">cd</span> snap2</span><br><span class="line">gff3_merge -d ../genome.maker.output/genome_master_datastore_index.log</span><br><span class="line">fathom -categorize 1000 genome.ann genome.dna</span><br><span class="line">fathom -<span class="built_in">export</span> 1000 -plus uni.ann uni.dna</span><br><span class="line">forge export.ann export.dna</span><br><span class="line">hmm-assembler.pl snap . &gt; ../snap2.hmm</span><br></pre></td></tr></table></figure>

<p>通常迭代2-3次就够了，毕竟我们可能还会训练AUGUSTUS和GeneMark模型，通过比较多个模型来得到最终结果。</p>
<p>参考资料: <a target="_blank" rel="noopener" href="http://weatherby.genetics.utah.edu/MAKER/wiki/index.php/MAKER_Tutorial_for_WGS_Assembly_and_Annotation_Winter_School_2018#Training_ab_initio_Gene_Predictors">http://weatherby.genetics.utah.edu/MAKER/wiki/index.php/MAKER_Tutorial_for_WGS_Assembly_and_Annotation_Winter_School_2018#Training_ab_initio_Gene_Predictors</a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/07/06/MAKER%E8%AE%AD%E7%BB%83SNAP%E5%9F%BA%E5%9B%A0%E6%A8%A1%E5%9E%8B-maker-train-snap-model/" data-id="cllgasdpj002hdkovcs8jdds9" data-title="MAKER训练SNAP基因模型" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-MAKER配置文件详解-explain-maker-control-files" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/07/01/MAKER%E9%85%8D%E7%BD%AE%E6%96%87%E4%BB%B6%E8%AF%A6%E8%A7%A3-explain-maker-control-files/" class="article-date">
  <time class="dt-published" datetime="2020-07-01T06:26:12.241Z" itemprop="datePublished">2020-07-01</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/07/01/MAKER%E9%85%8D%E7%BD%AE%E6%96%87%E4%BB%B6%E8%AF%A6%E8%A7%A3-explain-maker-control-files/">MAKER配置文件详解</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>本文翻译自<a target="_blank" rel="noopener" href="http://weatherby.genetics.utah.edu/MAKER/wiki/index.php/The_MAKER_control_files_explained">http://weatherby.genetics.utah.edu/MAKER/wiki/index.php/The_MAKER_control_files_explained</a></p>
<p>MAKER会生成三个配置文件，如下</p>
<ul>
<li>maker_opts.ctl: 控制MAKER分析行为的主要配置文件</li>
<li>maker_bopts.ctl:  BLAST和Exnerate的过滤阈值</li>
<li>maker_exe.ctl: MAKER运行过程中所依赖软件的路径</li>
</ul>
<h2 id="maker-opts-ctl"><a href="#maker-opts-ctl" class="headerlink" title="maker_opts.ctl"></a>maker_opts.ctl</h2><h3 id="基因组"><a href="#基因组" class="headerlink" title="基因组"></a>基因组</h3><p>用于设置被注释的基因组序列的位置和物种类型，包括<code>genome</code>和<code>organisam_type</code>两项</p>
<ul>
<li>genome: FASTA序列路径</li>
<li>organism_type: eukaryotic,prokaryotic二选一</li>
</ul>
<p>需要注意的是，基因组序列的N50需要超过预期基因长度的中位数，否则注释效果不好。另外最好保证基因组序列只包括A,T,C,G,N, 对于其他类型兼并碱基可以都改成N.</p>
<h3 id="使用MAKER得到GFF3进行重注释"><a href="#使用MAKER得到GFF3进行重注释" class="headerlink" title="使用MAKER得到GFF3进行重注释"></a>使用MAKER得到GFF3进行重注释</h3><p>这一项基本上我们用不上，它是在当你把MAKER的中间输出文件都删除了，仅保留了输出的GFF3文件时，你可以用之前相同的输入设置重新运行流程得到相同的输出。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#-----Re-annotation Using MAKER Derived GFF3</span></span><br><span class="line">maker_gff= <span class="comment">#MAKER derived GFF3 file</span></span><br><span class="line">est_pass=0 <span class="comment">#use ESTs in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">altest_pass=0 <span class="comment">#use alternate organism ESTs in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">protein_pass=0 <span class="comment">#use protein alignments in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">rm_pass=0 <span class="comment">#use repeats in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">model_pass=0 <span class="comment">#use gene models in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">pred_pass=0 <span class="comment">#use ab-initio predictions in maker_gff: 1 = yes, 0 = no</span></span><br><span class="line">other_pass=0 <span class="comment">#passthrough anyything else in maker_gff: 1 = yes, 0 = no</span></span><br></pre></td></tr></table></figure>

<h3 id="EST-转录本证据"><a href="#EST-转录本证据" class="headerlink" title="EST&#x2F;转录本证据"></a>EST&#x2F;转录本证据</h3><p>出于历史原因, MAKER还是用EST代表了之前的EST数据和目前的转录组数据。 此处不只是使用EST数据，而是可以使用组装的mRNA-seq, 组装的全长cDNA。 我们预期他们能够正确的组装，并联配到正确的剪切位点(对于FASTA格式，MAKER使用<code>exonerate</code>找到剪切位点）。用途如下:</p>
<ul>
<li>直接<strong>推断</strong>基因模型(est2genome)</li>
<li>作为预测结果的支持证据</li>
<li>修改结果和增加UTR</li>
<li>鉴定可变转录本</li>
<li>在某些情况下，这些数据和其他证据能帮助MAKER推断基因边界</li>
<li>在预测步骤中辅助基因预测工具推断剪切位点</li>
</ul>
<p>设置项如下:</p>
<ul>
<li><code>est</code>: EST, 全长cDNA, 组装的mRNA序列，可以通过逗号分隔多个文件。</li>
<li><code>altest</code>: 如果真的没有当前物种的转录组数据，也可以使用同源物种的序列。这些序列会通过tblstx进行比对，会消耗大量的运算时间。</li>
<li><code>est_gff</code>: 预比对的转录本，以GFF3格式存放，通常来自于CuffLinks&#x2F;StringTie的组装结果，或者是上一步的MAKER注释</li>
<li><code>altest_gff</code>: 和altest一样，只不过是比对后以GFF3存放，基本上也用不到</li>
</ul>
<h3 id="同源蛋白证据"><a href="#同源蛋白证据" class="headerlink" title="同源蛋白证据"></a>同源蛋白证据</h3><p>和之前的转录组数据类似，用途如下</p>
<ul>
<li>直接推断基因模型(protein2genome), 仅在它们能够正确的联配到剪切位点附件。</li>
<li>作为预测结果的支持证据(MAKER会检查CDS，保证基因预测结果和蛋白联配是相同的阅读框)</li>
<li>某些情况下，用于推断基因边界</li>
<li>在预测步骤中，使用从蛋白推断的ORF辅助从头预测软件</li>
</ul>
<p><strong>建议</strong>使用 uniprot&#x2F;swiss-prot 或 RefSeq上的NP数据，因为经过人工审查，可信度较高。不建议是用UniProt&#x2F;tremble或者Genbank上的数据，这些数据的可信度较低。你可以挑选几个同源物种的高可信度蛋白。或者使用MAKER注释的其他物种AED小于0.5的转录本产物。</p>
<p>由于许多注释里包含一些死亡转座子(dead transposons)或伪基因(pseudogenes)，因此<strong>不建议</strong>使用临近物种的所有注释蛋白。我们想象一个比较糟糕的情况，如果你有邻近物种的死亡转座子，当你构建你的重复序列屏蔽文库时，你发现其中一个条目和该序列匹配。 于是你假设这是一个真实的基因，于是你从屏蔽文库中删除了该条目。吸纳子啊，当你注释基因组的时候，该基因变成了注释集中的一整个基因组家族，但这其实是糟糕的证据和重复序列屏蔽所导致的后果。</p>
<p>需要设置的就两项：</p>
<ul>
<li><code>protein</code>: 以FASTA存放的蛋白序列位置</li>
<li><code>protein_gff</code>: 以GFF3格式记录的蛋白序列预先比对结果，通常来自于之前MAKER输出</li>
</ul>
<h3 id="重复序列重复屏蔽"><a href="#重复序列重复屏蔽" class="headerlink" title="重复序列重复屏蔽"></a>重复序列重复屏蔽</h3><p>我们可以通过屏蔽重复序列来避免EST和蛋白比对到重复区域，防止基因预测算法在这些区域预测外显子。由于许多重复序列会编码真实的蛋白(例如反转座子等)，基因预测工具和比对工具会被他们所迷惑(会在一个基因中错误的加上外显子)</p>
<ul>
<li><code>model_org</code>: 默认是all, 可以设置成RepeatMasker中RepBase数据库的其中一个</li>
<li><code>rmlib</code>: 自己构建的重复序列文库</li>
<li><code>repeat_protein</code>: 转座因子的蛋白序列</li>
<li><code>rm_gff</code>: 以GFF3格式记录的重复序列位置信息</li>
<li><code>prok_rm</code>: 不需要修改，因为原核生物不需要考虑重复序列</li>
<li><code>softmask</code>: 不需要修改</li>
</ul>
<h3 id="从头基因预测"><a href="#从头基因预测" class="headerlink" title="从头基因预测"></a>从头基因预测</h3><p>如果你需要从MAKER以外获取基因模型，则需要在这一节添加相应的配置。根据可信度高低，MAKER会对这些基因模型采取不同的行为。</p>
<ul>
<li><p>通过软件预测的基因结构可信度低，它们不会影响证据簇(evidence cluster). MAKER会保留预测结果，或者根据EST证据调整外显子，如果有证据支持，那么他们会保留在最终的注释集中。如果有多个注释结果，MAKER会对其进行比较，从中挑选出最优结果。</p>
</li>
<li><p><code>model_gff</code>提供的基因模型的可信度最高，会影响证据簇。在一些基因边界判定中，MAKER在证据簇的影响下，更倾向于保留之前的基因模型而非替换。它们也会保留名字。MAKER不会修改模型，要么删除要么保留。最后，即便没有证据支持，MAKER还是会保留他们，而不会删除他们，只不过最终的AED会设置为1.</p>
</li>
<li><p>snaphmm: SNAP的HMM文件路径，允许多个输入，以逗号分隔</p>
</li>
<li><p>gmhmm: GeneMark的HMM文件文件路径，允许多个输入，以逗号分隔</p>
</li>
<li><p>augustus_species: Augustus的基因模型命名, 不容易训练，但是效果很好</p>
</li>
<li><p>fgenesh_par_file: FGENESH的HMM参数文件，收费工具，基本上用不到</p>
</li>
<li><p>pred_gff: 其他预测工具的输出结果，以GFF3格式保存</p>
</li>
<li><p>model_gff: 最高可信度的GFF输入</p>
</li>
<li><p>run_evm&#x3D;0: 是否让MAKER运行EVM，速度会变慢</p>
</li>
<li><p>est2genome: 让MAKER根据EST推测基因模型</p>
</li>
<li><p>protein2genome: 让MAKER根据蛋白序列推测基因模型</p>
</li>
<li><p>trna: 使用tRNAscan分析tRNA #find tRNAs with tRNAscan, 1 &#x3D; yes, 0 &#x3D; no</p>
</li>
<li><p>snoscan_rrna: Snoscan分析snoRNA所需的rRNA文件&#x3D; #rRNA file to have Snoscan find snoRNAs</p>
</li>
<li><p>snoscan_meth：Snoscan分析snoRNA所需的O-methylation site 文件</p>
</li>
<li><p>unmask: 是否在屏蔽序列中进行基因预测，默认是0</p>
</li>
<li><p>allow_overlap: 允许的基因重叠比例，从0到1，空白表示使用默认值</p>
</li>
</ul>
<h3 id="其他类型的注释"><a href="#其他类型的注释" class="headerlink" title="其他类型的注释"></a>其他类型的注释</h3><p>这一项功能很简单，就是提供一个GFF文件，在MAKER运行结束后增加里面的信息</p>
<ul>
<li><code>other_gff</code>: 其他类型注释的GFF文件路径</li>
</ul>
<h3 id="外部程序选项"><a href="#外部程序选项" class="headerlink" title="外部程序选项"></a>外部程序选项</h3><p>这里的两个参数用于影响外部程序，即BLAST的行为</p>
<ul>
<li><code>alt_peptide</code>: 对于非标准氨基酸的替换方法，默认是C(cysteine)</li>
<li><code>cpus</code>: BLAST的线程数，如果使用MPI，该值可以设置的小一些，默认是1.</li>
</ul>
<h3 id="MAKER行为选项"><a href="#MAKER行为选项" class="headerlink" title="MAKER行为选项"></a>MAKER行为选项</h3><p>这里的选项用于调整MAKER的行为，使其符合你的基因组特性</p>
<ul>
<li><code>mad_dna_len</code>至少要3倍于预期的最大内含子长度。在内存足够的情况下，对于<strong>脊椎动物</strong>可以考虑设置为<code>300000</code>，植物一般没有那么大的内含子</li>
<li><code>min_contig</code>: 低于该值的contig会被过滤掉，建议设置为10k.</li>
<li><code>pred_flank</code>: 在基因预测时，将证据簇在两端进行扩展，默认是200 bp. 对于比较紧凑的基因组，降低该值能够避免基因错误合并。对于比较稀疏的基因组，提高该值可以避免外显子缺失。</li>
<li><code>pred_stats</code>: 默认是0，只计算MAKER预测基因的AED和QI值，设置为1则计算所有从头预测的基因结构。</li>
<li><code>AED_threashold</code>:  根据AED(0-1)值来过滤输出的基因，默认是1，表示保留所有预测结果。</li>
<li><code>min_protein</code>:  一些时候，基因预测工具会生成许多短预测结果，而由于一些证据类型（例如mRNA-seq）存在噪音，导致这些预测结果看起来有证据支持，于是保留在最终的输出结果中(AED&gt;1). 通过限制预测蛋白的氨基酸数(amino acides, AA)，可以减少预测结果。</li>
<li><code>alt_splice</code>: 是否计算可变转录本</li>
<li><code>always_complete</code>: 这个是MAKER开发者在合作者的要求下加上的参数，用来确保基因模型始终有起始密码子和终止密码子，默认是0</li>
<li><code>map_forward</code>: 用于保留老版本的GFF文件的信息，映射到新的版本GFF中。</li>
<li><code>keep_preds</code>:  设置为1时表示保留所有的预测结果，默认是0.</li>
<li><code>split_hit</code>:   新版本的MAKER(&gt;2.28)不需要考虑该项。因为之前版本拆分后的contig是互不重叠，于是就有可能有外显子被刚好被拆成两端，设置该项可以保留该信息。</li>
<li><code>single_exon</code>: 如果一个EST只有只有单个外显子，默认情况下MAKER并不会把它当做支持基因模型的证据, 除非还有同源蛋白作为支持。单外显子EST和组装的mRNA-seq转录本通常是RNA制备过程中的基因组序列污染。关闭时会降低MAKER的敏感度(sensitivity), 但是当你打开它的时候，命中的特异性会比整体的准确度(accuracy)差得多。</li>
<li><code>single_length</code>: 在启用<code>single_exon</code>时，设置该项用于保留一些比较小的序列，但即便注释了也可能不是有功能的蛋白。</li>
<li><code>correct_est_fusion</code>用来避免因为UTR的重叠导致将基因模型的错误合并，在<strong>真菌基因组</strong>中比较常见。它会检查基因模型的5’ UTR长度是否超过基因长度的一半，如果是的话，那么MAKER会在起始密码位置打断基因，然后在5’UTR区重新预测基因。</li>
<li><code>tries</code>: 尝试次数，默认是2</li>
<li><code>clean_try</code>: 重新尝试时，是否删除之前的文件，默认是0，也就是不删除。建议设置为1</li>
<li><code>clean_up</code>: 删除分析过程中的文件，默认不删除</li>
<li><code>TMP</code>: 临时文件的目录，默认存放在<code>/tmp</code>下，建议设置一个容量比较大的目录</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/07/01/MAKER%E9%85%8D%E7%BD%AE%E6%96%87%E4%BB%B6%E8%AF%A6%E8%A7%A3-explain-maker-control-files/" data-id="cllgasdpl002pdkovaut0ege9" data-title="MAKER配置文件详解" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-MAKER的并行化初探-parallelization-of-maker" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/06/29/MAKER%E7%9A%84%E5%B9%B6%E8%A1%8C%E5%8C%96%E5%88%9D%E6%8E%A2-parallelization-of-maker/" class="article-date">
  <time class="dt-published" datetime="2020-06-29T04:28:13.909Z" itemprop="datePublished">2020-06-29</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/06/29/MAKER%E7%9A%84%E5%B9%B6%E8%A1%8C%E5%8C%96%E5%88%9D%E6%8E%A2-parallelization-of-maker/">MAKER的并行化初探</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>MAKER并行分为两种，一种基于MPI，运行方式为<code>mpiexec -n 线程数 maker</code>, 一种是在同一个项目中运行多次maker。前者需要在安装MAKER时进行设置，后者相当于你手动按照染色体数目进行拆分，然后分开运行MAKER。本片主要介绍基于MPI的并行策略。</p>
<p>下图是MAKER是基于MPI的并行化流程。分为三个层次，contig水平，更小的片段水平，不同程序的并行。</p>
<p>contig水平就是对每条contig进行分别注释。而由于每个contig的长度不一，当较短的contig运行结束后，我们还需要等待较长的contig结束。因此，我们还需要将contig继续拆分，让每条contig分成更小的片段，成为基本的分析单元。对于每个分析单元，还可以考虑将不同程序进行并行。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/06/image-5624eadd6cc346a4b93a63be53fea0fb.png" alt="Fig1"></p>
<p>在MAKER实际运行时，如果用gotop进行查看你会发现系统中会出现很多maker进程(以<code>mpiexec -n 10</code>为例)</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/06/image-004af64038c349fb99dbb3ee58531b58.png" alt="Fig2"></p>
<p>但是你发现只有少量的和注释有关的程序(如blastx)。如果用<code>ps aux | grep maker</code>查找和maker有关的进程，你会发现大量的<code>maintain.pl</code>进程，并且这些进程后面跟着一大串你不认识的符号</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">/usr/bin/perl /opt/biosoft/maker/bin/../lib/File/maintain.pl 46535 30 %04%09%08%31%3...</span><br></pre></td></tr></table></figure>

<p>为了理解MAKER的并行，你需要通过<code>pstree -ap [mpiexec进程ID]</code>更细致的了解MAKER的进程树</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/06/image-e6e682747a3a443e9a2b0e9810f0177c.png" alt="Fig3"></p>
<p>从中你可以发现，因为我们用<code>mpiexec</code>启动了10个maker进程，所以<code>hydra_pmi_proxy</code>有10个maker子进程。对于这10个maker子进程，每个进程都至少会有一个<code>maintain.pl</code>。通过阅读<code>maintain.pl</code>的源代码，我们可以得知该程序后接参数中的46535是PID(进程号), 30是sleep time, 最后一个是URI编码字符，是利用<code>Storable::freeze</code>持久化的数据结构，可以通过如下的代码进行解码（命名为decode.pl)</p>
<figure class="highlight perl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#!/usr/bin/env perl</span></span><br><span class="line"><span class="keyword">use</span> warnings;</span><br><span class="line"><span class="keyword">use</span> strict;</span><br><span class="line"></span><br><span class="line"><span class="keyword">use</span> URI::Escape;</span><br><span class="line"><span class="keyword">use</span> Storable;</span><br><span class="line"><span class="keyword">use</span> vars <span class="string">qw($LOCK)</span>;</span><br><span class="line"></span><br><span class="line"><span class="keyword">my</span> $serial = <span class="keyword">shift</span>;</span><br><span class="line"></span><br><span class="line">$serial = uri_unescape($serial);</span><br><span class="line">$LOCK = Storable::thaw($serial);</span><br><span class="line"></span><br><span class="line"><span class="keyword">print</span> <span class="string">&quot;$LOCK-&gt;&#123;lock_file&#125; \n&quot;</span></span><br></pre></td></tr></table></figure>

<p>可以批量对<code>maintain.pl</code>里的信息进行解码</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ps aux | grep maintain.pl | grep -v grep | awk <span class="string">&#x27;&#123;print $15&#125;&#x27;</span> | xargs -i perl decode.pl &#123;&#125;</span><br></pre></td></tr></table></figure>

<p>因此，那么多的maker并不是在话说，而是在负责任务调度，差不多一个任务会有3个maker进程保驾护航。</p>
<p>回到之前的进程数，这里有一些maker进程下会跟着一些正在运行的命令，blastx, exonerate</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/06/image-01a3786169a74d278d804a79c9e36e91.png" alt="Fig4"></p>
<p>而有些只有maker和maintain.pl进程。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/06/image-6d38421ce79d4c129fe92e78d71a926d.png" alt="Fig5"></p>
<p>如果等待一会，再次运行<code>pstree -ap [mpiexec进程ID]</code>，你会发现这两种状态会发生转换，这说明任务启动也需要一段时间。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/06/29/MAKER%E7%9A%84%E5%B9%B6%E8%A1%8C%E5%8C%96%E5%88%9D%E6%8E%A2-parallelization-of-maker/" data-id="cllgasdpi002edkoves5e5ftq" data-title="MAKER的并行化初探" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-Singularity和MPI应用-singularity-and-mpi-applications" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/06/22/Singularity%E5%92%8CMPI%E5%BA%94%E7%94%A8-singularity-and-mpi-applications/" class="article-date">
  <time class="dt-published" datetime="2020-06-22T01:39:27.294Z" itemprop="datePublished">2020-06-22</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/06/22/Singularity%E5%92%8CMPI%E5%BA%94%E7%94%A8-singularity-and-mpi-applications/">Singularity和MPI应用</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>MPI(Message Passin Interface)广泛应用于高性能服务器中，可用于单系统多节点或者多个计算平台间通讯，目前主流的两个开源软件分别是OpenMPI和MPICH。Singularity同时支持这两个开源工具，本篇教程介绍如何在Singularity容器开发和运行MPI程序。</p>
<p>在Singularity容器是执行MPI程序最常见的方法就是使用宿主机中的MPI。由于我们会同时使用宿主机的MPI和容器中的MPI，因此这被称为Host MPI或者混合模型。考虑到大部分高性能计算平台并不都支持宿主机和容器之间的文件系统共享，因此我们不讨论如何将存储盘挂载到容器，从而使用容器中宿主机MPI。</p>
<p>混合途径的基本想法就是，当你要执行SIngularity容器的MPI代码时，你会使用<code>mpiexec</code>类似命令去调用<code>singularity</code>命令。容器外部的MPI进程会和容器内的MPI进行写作，容器后的MPI代码会实例化任务。</p>
<p>Open MPI&#x2F;Singularity的工作流程如下</p>
<ol>
<li>用户在shell中启动MPI(mpirun, mpiexec)</li>
<li>Open MPI接着调用进程管理守护进程(process management daemon, ORTED)</li>
<li>ORTED进程启动所需的Singularity容器</li>
<li>Singularity构建容器和命名空间环境</li>
<li>Singularity启动容器内的MPI应用</li>
<li>MPI进程启动，加载Open MPI库</li>
<li>Open MPI库通过进程管理接口(Process Management, PMI)连接回ORTED进程</li>
</ol>
<p>此时在容器运行MPI就和直接宿主机运行MPI程序一样。优势在于它能够整合到Slurm等资源管理工具，并且和之前运行MPI应用一样。但是你需要保证容器中的MPI和宿主机的MPI版本兼容，也就是在构建MPI容器时保证和系统的MPI一致。</p>
<p>我们举个简单的例子。假设我们已经有了一个叫做<code>mpitest.c</code>MPI应用</p>
<figure class="highlight c"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;mpi.h&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;stdio.h&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;stdlib.h&gt;</span></span></span><br><span class="line"></span><br><span class="line"><span class="type">int</span> <span class="title function_">main</span> <span class="params">(<span class="type">int</span> argc, <span class="type">char</span> **argv)</span> &#123;</span><br><span class="line">        <span class="type">int</span> rc;</span><br><span class="line">        <span class="type">int</span> size;</span><br><span class="line">        <span class="type">int</span> myrank;</span><br><span class="line"></span><br><span class="line">        rc = MPI_Init (&amp;argc, &amp;argv);</span><br><span class="line">        <span class="keyword">if</span> (rc != MPI_SUCCESS) &#123;</span><br><span class="line">                <span class="built_in">fprintf</span> (<span class="built_in">stderr</span>, <span class="string">&quot;MPI_Init() failed&quot;</span>);</span><br><span class="line">                <span class="keyword">return</span> EXIT_FAILURE;</span><br><span class="line">        &#125;</span><br><span class="line"></span><br><span class="line">        rc = MPI_Comm_size (MPI_COMM_WORLD, &amp;size);</span><br><span class="line">        <span class="keyword">if</span> (rc != MPI_SUCCESS) &#123;</span><br><span class="line">                <span class="built_in">fprintf</span> (<span class="built_in">stderr</span>, <span class="string">&quot;MPI_Comm_size() failed&quot;</span>);</span><br><span class="line">                <span class="keyword">goto</span> exit_with_error;</span><br><span class="line">        &#125;</span><br><span class="line"></span><br><span class="line">        rc = MPI_Comm_rank (MPI_COMM_WORLD, &amp;myrank);</span><br><span class="line">        <span class="keyword">if</span> (rc != MPI_SUCCESS) &#123;</span><br><span class="line">                <span class="built_in">fprintf</span> (<span class="built_in">stderr</span>, <span class="string">&quot;MPI_Comm_rank() failed&quot;</span>);</span><br><span class="line">                <span class="keyword">goto</span> exit_with_error;</span><br><span class="line">        &#125;</span><br><span class="line"></span><br><span class="line">        <span class="built_in">fprintf</span> (<span class="built_in">stdout</span>, <span class="string">&quot;Hello, I am rank %d/%d&quot;</span>, myrank, size);</span><br><span class="line"></span><br><span class="line">        MPI_Finalize();</span><br><span class="line"></span><br><span class="line">        <span class="keyword">return</span> EXIT_SUCCESS;</span><br><span class="line"></span><br><span class="line"> exit_with_error:</span><br><span class="line">        MPI_Finalize();</span><br><span class="line">        <span class="keyword">return</span> EXIT_FAILURE;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<blockquote>
<p>MPI只是一个库接口，里面的函数能够被其他编程语言所调用，所以支持Fortran, C, Python, R等。</p>
</blockquote>
<p>接下来根据宿主机的MPI版本，编写一个定义文件。假如我们宿主机的MPI是MPICH，那么定义文件(命名为mpitest)如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br></pre></td><td class="code"><pre><span class="line">Bootstrap: docker</span><br><span class="line">From: ubuntu:latest</span><br><span class="line"></span><br><span class="line">%files</span><br><span class="line">    mpitest.c /opt</span><br><span class="line"></span><br><span class="line">%environment</span><br><span class="line">    <span class="built_in">export</span> MPICH_DIR=/opt/mpich-3.3</span><br><span class="line">    <span class="built_in">export</span> SINGULARITY_MPICH_DIR=<span class="variable">$MPICH_DIR</span></span><br><span class="line">    <span class="built_in">export</span> SINGULARITYENV_APPEND_PATH=<span class="variable">$MPICH_DIR</span>/bin</span><br><span class="line">    <span class="built_in">export</span> SINGULAIRTYENV_APPEND_LD_LIBRARY_PATH=<span class="variable">$MPICH_DIR</span>/lib</span><br><span class="line"></span><br><span class="line">%post</span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Installing required packages...&quot;</span></span><br><span class="line">    apt-get update &amp;&amp; apt-get install -y wget git bash gcc gfortran g++ make</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Information about the version of MPICH to use</span></span><br><span class="line">    <span class="built_in">export</span> MPICH_VERSION=3.3</span><br><span class="line">    <span class="built_in">export</span> MPICH_URL=<span class="string">&quot;http://www.mpich.org/static/downloads/<span class="variable">$MPICH_VERSION</span>/mpich-<span class="variable">$MPICH_VERSION</span>.tar.gz&quot;</span></span><br><span class="line">    <span class="built_in">export</span> MPICH_DIR=/opt/mpich</span><br><span class="line"></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Installing MPICH...&quot;</span></span><br><span class="line">    <span class="built_in">mkdir</span> -p /tmp/mpich</span><br><span class="line">    <span class="built_in">mkdir</span> -p /opt</span><br><span class="line">    <span class="comment"># Download</span></span><br><span class="line">    <span class="built_in">cd</span> /tmp/mpich &amp;&amp; wget -O mpich-<span class="variable">$MPICH_VERSION</span>.tar.gz <span class="variable">$MPICH_URL</span> &amp;&amp; tar xzf mpich-<span class="variable">$MPICH_VERSION</span>.tar.gz</span><br><span class="line">    <span class="comment"># Compile and install</span></span><br><span class="line">    <span class="built_in">cd</span> /tmp/mpich/mpich-<span class="variable">$MPICH_VERSION</span> &amp;&amp; ./configure --prefix=<span class="variable">$MPICH_DIR</span> &amp;&amp; make install</span><br><span class="line">    <span class="comment"># Set env variables so we can compile our application</span></span><br><span class="line">    <span class="built_in">export</span> PATH=<span class="variable">$MPICH_DIR</span>/bin:<span class="variable">$PATH</span></span><br><span class="line">    <span class="built_in">export</span> LD_LIBRARY_PATH=<span class="variable">$MPICH_DIR</span>/lib:<span class="variable">$LD_LIBRARY_PATH</span></span><br><span class="line">    <span class="built_in">export</span> MANPATH=<span class="variable">$MPICH_DIR</span>/share/man:<span class="variable">$MANPATH</span></span><br><span class="line"></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Compiling the MPI application...&quot;</span></span><br><span class="line">    <span class="built_in">cd</span> /opt &amp;&amp; mpicc -o mpitest mpitest.c</span><br></pre></td></tr></table></figure>

<p>如果是Open MPI,</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">Bootstrap: docker</span><br><span class="line">From: ubuntu:latest</span><br><span class="line"></span><br><span class="line">%files</span><br><span class="line">    mpitest.c /opt</span><br><span class="line"></span><br><span class="line">%environment</span><br><span class="line">    <span class="built_in">export</span> OMPI_DIR=/opt/ompi</span><br><span class="line">    <span class="built_in">export</span> SINGULARITY_OMPI_DIR=<span class="variable">$OMPI_DIR</span></span><br><span class="line">    <span class="built_in">export</span> SINGULARITYENV_APPEND_PATH=<span class="variable">$OMPI_DIR</span>/bin</span><br><span class="line">    <span class="built_in">export</span> SINGULAIRTYENV_APPEND_LD_LIBRARY_PATH=<span class="variable">$OMPI_DIR</span>/lib</span><br><span class="line"></span><br><span class="line">%post</span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Installing required packages...&quot;</span></span><br><span class="line">    apt-get update &amp;&amp; apt-get install -y wget git bash gcc gfortran g++ make file</span><br><span class="line"></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Installing Open MPI&quot;</span></span><br><span class="line">    <span class="built_in">export</span> OMPI_DIR=/opt/ompi</span><br><span class="line">    <span class="built_in">export</span> OMPI_VERSION=4.0.1</span><br><span class="line">    <span class="built_in">export</span> OMPI_URL=<span class="string">&quot;https://download.open-mpi.org/release/open-mpi/v4.0/openmpi-<span class="variable">$OMPI_VERSION</span>.tar.bz2&quot;</span></span><br><span class="line">    <span class="built_in">mkdir</span> -p /tmp/ompi</span><br><span class="line">    <span class="built_in">mkdir</span> -p /opt</span><br><span class="line">    <span class="comment"># Download</span></span><br><span class="line">    <span class="built_in">cd</span> /tmp/ompi &amp;&amp; wget -O openmpi-<span class="variable">$OMPI_VERSION</span>.tar.bz2 <span class="variable">$OMPI_URL</span> &amp;&amp; tar -xjf openmpi-<span class="variable">$OMPI_VERSION</span>.tar.bz2</span><br><span class="line">    <span class="comment"># Compile and install</span></span><br><span class="line">    <span class="built_in">cd</span> /tmp/ompi/openmpi-<span class="variable">$OMPI_VERSION</span> &amp;&amp; ./configure --prefix=<span class="variable">$OMPI_DIR</span> &amp;&amp; make install</span><br><span class="line">    <span class="comment"># Set env variables so we can compile our application</span></span><br><span class="line">    <span class="built_in">export</span> PATH=<span class="variable">$OMPI_DIR</span>/bin:<span class="variable">$PATH</span></span><br><span class="line">    <span class="built_in">export</span> LD_LIBRARY_PATH=<span class="variable">$OMPI_DIR</span>/lib:<span class="variable">$LD_LIBRARY_PATH</span></span><br><span class="line">    <span class="built_in">export</span> MANPATH=<span class="variable">$OMPI_DIR</span>/share/man:<span class="variable">$MANPATH</span></span><br><span class="line"></span><br><span class="line">    <span class="built_in">echo</span> <span class="string">&quot;Compiling the MPI application...&quot;</span></span><br><span class="line">    <span class="built_in">cd</span> /opt &amp;&amp; mpicc -o mpitest mpitest.c</span><br></pre></td></tr></table></figure>

<p>接着我们从定义文件中构建出sig文件, <code>singularity build mpitest.sig mpitest</code></p>
<p>最后，执行Singularity容器的MPI应用的标准方法就是在host中运行<code>mpirun</code>命令，它会启动Singularity容器，并最终运行其中的MPI程序。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">mpirun -n 4 singularity <span class="built_in">exec</span> mpitest.sig /opt/mpitest</span><br><span class="line"><span class="comment"># 输出信息如下</span></span><br><span class="line"><span class="comment"># Hello, I am rank 0/4Hello, I am rank 1/4Hello, I am rank 2/4Hello, I am rank 3/4</span></span><br></pre></td></tr></table></figure>

<p>如果是在SLURM这类任务提交系统中，我们需要批处理脚本来执行MPI引用，举个例子</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#!/bin/bash</span></span><br><span class="line"><span class="comment">#SBATCH --job-name singularity-mpi</span></span><br><span class="line"><span class="comment">#SBATCH -N $NNODES # total number of nodes</span></span><br><span class="line"><span class="comment">#SBATCH --time=00:05:00 # Max execution time</span></span><br><span class="line"></span><br><span class="line">mpirun -n <span class="variable">$NP</span> singularity <span class="built_in">exec</span> /var/nfsshare/gvallee/mpich.sif /opt/mpitest</span><br></pre></td></tr></table></figure>

<p>执行方式如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sbatch my_job.sh</span><br></pre></td></tr></table></figure>

<p>参考资料: <a target="_blank" rel="noopener" href="https://sylabs.io/guides/3.3/user-guide/mpi.html">https://sylabs.io/guides/3.3/user-guide/mpi.html</a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/06/22/Singularity%E5%92%8CMPI%E5%BA%94%E7%94%A8-singularity-and-mpi-applications/" data-id="cllgasdqd004vdkovcgx5b3ni" data-title="Singularity和MPI应用" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第九章)-analysis-sc-atac-seq-with-archr-chapter9" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/06/04/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B9%9D%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter9/" class="article-date">
  <time class="dt-published" datetime="2020-06-04T09:35:45.570Z" itemprop="datePublished">2020-06-04</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/06/04/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B9%9D%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter9/">使用ArchR分析单细胞ATAC-seq数据(第九章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第9章-ArchR的拟混池重复"><a href="#第9章-ArchR的拟混池重复" class="headerlink" title="第9章 ArchR的拟混池重复"></a>第9章 ArchR的拟混池重复</h1><p>因为scATAC-seq数据本质上只有两种值，也就是说每个位点要么开放要么不开放，所以你会发现某些情况下无法使用单个细胞进行数据分析。此外，许多我们想要做的分析也依赖于重复才能计算统计显著性。对于单细胞数据，我们通过构建拟混池重复(pseudo-bulk replicates)来解决该问题。所谓的拟混池(pesudo-bulk)指的就是将单细胞进行合并模拟成混池测序的ATAC-seq实验得到的数据。ArchR为每个目标细胞分组构建多个拟混池样本，也就得到了拟混池重复。这个模拟过程背后的假设是将单细胞进行合并结果和实际的混池结果非常接近，以至于不需要在乎背后的差异。这些细胞分组通常都是来自于单个细胞类群，或者是一直细胞类型对应的可能聚类。我们这一章会介绍如何使用ArchR生成这些拟混池重复。</p>
<h2 id="9-1-ArchR如何构建拟混池重复"><a href="#9-1-ArchR如何构建拟混池重复" class="headerlink" title="9.1 ArchR如何构建拟混池重复"></a>9.1 ArchR如何构建拟混池重复</h2><p>ArchR使用分级优先法(tiered priority approach)构建拟混池重复。使用者定义: i)最小和最大重复数(minReps&#x2F;maxReps), ii)每个重复的最少和最多细胞数(minCells&#x2F;maxCells), iii)缺少足够细胞用于构建足够重复时的采样率(Sampling Ratio)。举个例子，当采样率等于0.8时，每个重复中80%的细胞来自于无放回抽样(这会导致重复间出现有放回抽样)。在这种情况下，多个重复中可能有一些相同的细胞，但是为了能在缺少足够细胞的分组里得到拟混池重复，这是必要的牺牲。</p>
<p>我们的拟混池重复生成过程可以用如下的决策树进行描述</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2020/06/pseudobulkReplicate_DecisionTree-47b9cb7311714040afd8afcd14dee720.png" alt="pseudobulkReplicate_DecisionTree"></p>
<p>流程有些复杂，这里概述下这个流程中的几个关键注意事项。</p>
<p>首先，用户确定细胞的分组，ArchR通常会称之为聚类，我们可以在第五章分析分析得到。</p>
<p>其次，对于每一组，ArchR尝试去创建理想的拟混池重复。所谓理想的拟混池重复，指的是每个重复只有一个样本构成，这就要求每个样本得保证足量的细胞数。这样保证了重复间的样本多样性和生物学变异，是ArchR所期望得到的最佳结果，但这一过程实际上会有5种可能的结果，根据ArchR的偏好排序如下</p>
<ol>
<li>数据拥有足够多的样本(至少要等于maxRep定义的重复)，且每个样本的细胞数都大于 <code>minCell</code>, 于是每个样本都可以作为拟混池重复，每个重复的细胞都来自于同一个样本。</li>
<li>一些样本的细胞数超过<code>minCell</code>,因此能够单独形成一个重复。剩下的重复则是通过将余下的细胞进行混合，然后通过无放回抽样得到。</li>
<li>数据里没有一个样本的细胞数超过<code>minCell</code>, 但是总细胞数超过<code>minCells * minReps</code>。因此将所有的细胞进行混合，然后进行无放回抽样，抽样时不考虑细胞来源。</li>
<li>一个细胞分组中的总细胞数低于 <code>minCells * minReps</code>，但是大于<code>minCells / Sample Ratio</code>。此时单个样本的构建采取无放回抽样，重复间则需要有放回抽样，降低多个拟混池重复间的相同细胞数。</li>
<li>一个细胞分组中的总细胞数低于 <code>minCells /  Sample Ratio</code> 。这意味着我们必须在单个重复和跨重复中都采取有放回抽样策略。这是最糟糕的情况，后续在使用这些拟混池重复做下游分析分析要特别小心。后续可以通过设置ArchR的<code>minCells</code>参数进行淘汰。</li>
</ol>
<p>我们使用如下的数据集阐述这一过程</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">Sample  Cluster1  Cluster2  Cluster3  Cluster4  Cluster5</span><br><span class="line">A       800       600       900       100       75</span><br><span class="line">B       1000      50        400       150       25</span><br><span class="line">C       600       900       100       200       50</span><br><span class="line">D       1200      500       50        50        25</span><br><span class="line">E       900       100       50        150       50</span><br><span class="line">F       700       200       100       100       25</span><br></pre></td></tr></table></figure>

<p>我们设置的参数为<code>minRep=3</code>,<code>maxRep=5</code>, <code>minCells=300</code>, <code>maxCells=1000</code>和<code>sampleRatio=0.8</code>，也就是最少有3个重复，最多是5个重复，每个重复至少有300个细胞，最多是1000个细胞，当细胞数不满住要求，抽样率设置为0.8.</p>
<h3 id="9-1-1-Cluster1"><a href="#9-1-1-Cluster1" class="headerlink" title="9.1.1 Cluster1"></a>9.1.1 Cluster1</h3><p>对于Cluster1, 我们的6个样本(大于<code>maxRep</code>)的细胞数都高于<code>minCells</code>(300)。这是最理想的情况，对应上述第一种情况，我们将会得到5个拟混池重复，保证每个重复都来自独立的样本。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">Rep1 = 800 cells from SampleA</span><br><span class="line">Rep2 = 1000 cells from SampleB</span><br><span class="line">Rep3 = 1000 cells from SampleD</span><br><span class="line">Rep4 = 900 cells from SampleE</span><br><span class="line">Rep5 = 700 cells from SampleF</span><br></pre></td></tr></table></figure>

<p>对于这些重复，我们需要注意两个事情:(1) 因为我们的样本数足够多，能够保证每个重复都来自独立的样本，所以可以淘汰其中细胞数最少的SampleC。(2)由于<code>maxCells</code>设置为1000，因此最多只能有1000个细胞。</p>
<h3 id="9-1-2-Cluster2"><a href="#9-1-2-Cluster2" class="headerlink" title="9.1.2 Cluster2"></a>9.1.2 Cluster2</h3><p>对于Cluster2, 我们有3个样本的细胞数超过<code>minCells</code>, 另外3个样本的细胞数都不够。这对应上述第二种情况，我们会以如下的方法构建拟混池重复。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Rep1 = 600 cells from SampleA</span><br><span class="line">Rep2 = 900 cells from SampleC</span><br><span class="line">Rep3 = 500 cells from SampleD</span><br><span class="line">Rep4 = 350 cells [50 cells from SampleB + 100 from SampleE + 200 from SampleF]</span><br></pre></td></tr></table></figure>

<p>在这个例子中，Rep4由其他几个样本的细胞混合后通过无放回抽样得到</p>
<h3 id="9-1-3-Cluster3"><a href="#9-1-3-Cluster3" class="headerlink" title="9.1.3 Cluster3"></a>9.1.3 Cluster3</h3><p>对于Cluster3，我们只有两个样本超过<code>minCells</code>, 不满足<code>minReps</code>。但是如果我们将剩余的样本的细胞进行混合形成额外的重复，它的细胞数就超过了<code>minCells</code>。最终我们得到了3个拟混池重复，对应上述的情况3。我们将得到如下重复</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Rep1 = 900 cells from SampleA</span><br><span class="line">Rep2 = 400 cells from SampleB</span><br><span class="line">Rep3 = 250 cells [100 cells from SampleC + 50 from SampleD + 50 from SampleE + 50 from SampleF]</span><br></pre></td></tr></table></figure>

<p>和Cluster2类似，Cluster3的Rep3由其他几个样本的细胞混合后通过无放回抽样得到</p>
<h3 id="9-1-4-Cluster4"><a href="#9-1-4-Cluster4" class="headerlink" title="9.1.4 Cluster4"></a>9.1.4 Cluster4</h3><p>对于Cluster4，总细胞数是570个，小于<code>minCells * minReps</code>(900). 在这个情况下，我们无法保证有足够多的细胞通过无放回抽样的方式保证每个重复都有最小的细胞数。但是，总的细胞数依旧依旧大于<code>minCells / sampleRatio</code>(375个细胞)，这意味着每个重复中细胞可以来自于无放回抽样，重复之间的细胞需要放回抽样。这对应着上述的情况4，我们将得到如下重复</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Rep1 = 300 cells [250 unique cells + 25 cells overlapping Rep2 + 25 cells overlapping Rep3]</span><br><span class="line">Rep2 = 300 cells [250 unique cells + 25 cells overlapping Rep1 + 25 cells overlapping Rep3]</span><br><span class="line">Rep3 = 300 cells [250 unique cells + 25 cells overlapping Rep1 + 25 cells overlapping Rep2]</span><br></pre></td></tr></table></figure>

<p>在这个情况中，ArchR会尽可能降低任意两个拟混池重复的相同细胞。</p>
<h3 id="9-1-5-Cluster5"><a href="#9-1-5-Cluster5" class="headerlink" title="9.1.5 Cluster5"></a>9.1.5 Cluster5</h3><p>对于Cluster5，总共是250个细胞，同时小于<code>minCells * minReps</code>（900）和<code>minCells / sampleRatio</code>(375). 这意味着每个样本都需要有放回的抽样，重复之间也需要有放回抽样，才能得到拟混池重复。这是上述说到的第5种情况，是其中最糟糕的情况。对于这类拟混池重复，在后续的分析中需要谨慎使用。我们将得到如下的重复:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Rep1 = 300 cells [250 unique cells + 25 cells overlapping Rep2 + 25 cells overlapping Rep3]</span><br><span class="line">Rep2 = 300 cells [250 unique cells + 25 cells overlapping Rep1 + 25 cells overlapping Rep3]</span><br><span class="line">Rep3 = 300 cells [250 unique cells + 25 cells overlapping Rep1 + 25 cells overlapping Rep2]</span><br></pre></td></tr></table></figure>

<h2 id="9-2-构建拟混池重复"><a href="#9-2-构建拟混池重复" class="headerlink" title="9.2 构建拟混池重复"></a>9.2 构建拟混池重复</h2><p>通过上一节了解ArchR构建拟混池重复的逻辑后，我们就可以开始实际操作了。在ArchR中，我们通过调用<code>addGroupCoverages()</code>函数来构建拟混池重复。它的关键参数<code>groupBy</code>,定义了拟混池重复需要使用的分组。这里我们用的是上一章scRNA-seq数据标记的细胞类型，也就是<code>Cluster2</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme4 <span class="operator">&lt;-</span> addGroupCoverages<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme3<span class="punctuation">,</span> groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>得到这些拟混池重复后，我们就能从数据中鉴定peak了。就像之前所说的，我们不希望使用所有的细胞鉴定peak，而是单独根据每一组细胞（例如聚类）单独鉴定peak，这样才有可能分析出不同组的特异性peak。这一章得到数据就为后续鉴定peak提供了良好的开始。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/06/04/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E4%B9%9D%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter9/" data-id="cllgasdsr009rdkov60vp7286" data-title="使用ArchR分析单细胞ATAC-seq数据(第九章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用ArchR分析单细胞ATAC-seq数据(第八章)-analysis-sc-atac-seq-with-archr-chapter8" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/05/31/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%85%AB%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter8/" class="article-date">
  <time class="dt-published" datetime="2020-05-31T22:27:31.103Z" itemprop="datePublished">2020-05-31</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/05/31/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%85%AB%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter8/">使用ArchR分析单细胞ATAC-seq数据(第八章)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第8章-使用scRNA-seq定义cluster类型"><a href="#第8章-使用scRNA-seq定义cluster类型" class="headerlink" title="第8章: 使用scRNA-seq定义cluster类型"></a>第8章: 使用scRNA-seq定义cluster类型</h1><p>除了使用基因得分定义细胞类群以外，ArchR还能整合scRNA-seq数据。通过将scATAC-seq数据里的基因得分矩阵和scRNA-seq数据的基因表达量矩阵进行对比，ArchR就能将scATAC-seq的细胞比对到scRNA-seq的细胞，实现两种数据的整合。之后，我们借助scRNA-seq数据已经定义的细胞类群，或者整合后的scRNA-seq的基因表达量来注释细胞类群。在代码内部，我们调用了<code>Seurat::FindTransferAnchors()</code>，从而实现两种数据集之间的比较。当然，ArchR不是简单地对函数进行封装，而是在此基础上通过对数据的拆分，实现并行计算，从而能将该流程应用到更大规模的细胞中。</p>
<img src="/upload/2020/05/image-296524df11944c0e8571c05b932fb1fa.png" alt="img" style="zoom:10%;" />

<p>该整合过程实际上会寻找scATAC-seq和scRNA-seq两者中最相似的细胞，然后将scRNA-seq中对应的细胞表达量赋值给scATAAC-seq细胞。最后，scATAC-seq中每个细胞都会有基因表达量特征，能被用于下游分析。这一章会阐述如何利用该信息定义细胞类型，之后会介绍如何使用连接的scRNA-seq数据做更加复杂的分析，例如识别预测的顺式调控元件。我们相信由于多组学单细胞谱的商业化，这类整合分析将会越来越多。同时，在ArchR中使用公共数据里匹配细胞类型的scRNA-seq数据或者自己使用目标样本得到的scRNA-seq数据也能加强scATAC-seq分析。</p>
<h2 id="8-1-scATAC-seq细胞和scRNA-seq细胞跨平台连接"><a href="#8-1-scATAC-seq细胞和scRNA-seq细胞跨平台连接" class="headerlink" title="8.1 scATAC-seq细胞和scRNA-seq细胞跨平台连接"></a>8.1 scATAC-seq细胞和scRNA-seq细胞跨平台连接</h2><p>为了能将我们教程中的scATAC-seq数据与其匹配的scRNA-seq数据进行整合，我们将使用 Granja* et al (2019) 里的造血细胞scRNA-seq数据。</p>
<p>scRNA-seq数据以 <code>RangedSummarizedExperiment</code>对象保存，大小为111MB。此外，ArchR还接受未经修改的<code>Seurat</code>对象作为整合流程的输入。我们使用<code>download.file</code>下载数据</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span><span class="punctuation">(</span><span class="operator">!</span>file.exists<span class="punctuation">(</span><span class="string">&quot;scRNA-Hematopoiesis-Granja-2019.rds&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    download.file<span class="punctuation">(</span></span><br><span class="line">        url <span class="operator">=</span> <span class="string">&quot;https://jeffgranja.s3.amazonaws.com/ArchR/TestData/scRNA-Hematopoiesis-Granja-2019.rds&quot;</span><span class="punctuation">,</span></span><br><span class="line">        destfile <span class="operator">=</span> <span class="string">&quot;scRNA-Hematopoiesis-Granja-2019.rds&quot;</span></span><br><span class="line">    <span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>下载之后，我们使用<code>readRDS</code>进行读取，并查看该对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">seRNA <span class="operator">&lt;-</span> readRDS<span class="punctuation">(</span><span class="string">&quot;scRNA-Hematopoiesis-Granja-2019.rds&quot;</span><span class="punctuation">)</span></span><br><span class="line">seRNA</span><br><span class="line"><span class="comment">## class: RangedSummarizedExperiment</span></span><br><span class="line"><span class="comment">## dim: 20287 35582</span></span><br><span class="line"><span class="comment">## metadata(0):</span></span><br><span class="line"><span class="comment">## assays(1): counts</span></span><br><span class="line"><span class="comment">## rownames(20287): FAM138A OR4F5 … S100B PRMT2</span></span><br><span class="line"><span class="comment">## rowData names(3): gene_name gene_id exonLength</span></span><br><span class="line"><span class="comment">## colnames(35582): CD34_32_R5:AAACCTGAGTATCGAA-1</span></span><br><span class="line"><span class="comment">## CD34_32_R5:AAACCTGAGTCGTTTG-1 …</span></span><br><span class="line"><span class="comment">## BMMC_10x_GREENLEAF_REP2:TTTGTTGCATGTGTCA-1</span></span><br><span class="line"><span class="comment">## BMMC_10x_GREENLEAF_REP2:TTTGTTGCATTGAAAG-1</span></span><br><span class="line"><span class="comment">## colData names(10): Group nUMI_pre … BioClassification Barcode</span></span><br></pre></td></tr></table></figure>

<p>从输出信息中，我们可以发现它里面有基因表达量的count矩阵和对应的元信息。</p>
<p>元信息列里的<code>BioClassification</code>记录着scRNA-seq数据中每个细胞对应的细胞类型分类</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">colnames<span class="punctuation">(</span>colData<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># [1] &quot;Group&quot;             &quot;nUMI_pre&quot;          &quot;nUMI&quot;             </span></span><br><span class="line"><span class="comment"># [4] &quot;nGene&quot;             &quot;initialClusters&quot;   &quot;UMAP1&quot;            </span></span><br><span class="line"><span class="comment"># [7] &quot;UMAP2&quot;             &quot;Clusters&quot;          &quot;BioClassification&quot;</span></span><br><span class="line"><span class="comment"># [10] &quot;Barcode&quot;</span></span><br></pre></td></tr></table></figure>

<p>使用<code>table()</code>，我们可以看到scRNA-seq细胞类型每一群的细胞数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">table<span class="punctuation">(</span>colData<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="operator">$</span>BioClassification<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#        01_HSC 02_Early.Eryth  03_Late.Eryth  04_Early.Baso    05_CMP.LMPP </span></span><br><span class="line"><span class="comment">#          1425           1653            446            111           2260 </span></span><br><span class="line"><span class="comment">#      06_CLP.1         07_GMP    08_GMP.Neut         09_pDC         10_cDC </span></span><br><span class="line"><span class="comment">#           903           2097           1050            544            325 </span></span><br><span class="line"><span class="comment">#11_CD14.Mono.1 12_CD14.Mono.2   13_CD16.Mono         14_Unk       15_CLP.2 </span></span><br><span class="line"><span class="comment">#          1800           4222            292            520            377 </span></span><br><span class="line"><span class="comment">#      16_Pre.B           17_B      18_Plasma       19_CD8.N      20_CD4.N1 </span></span><br><span class="line"><span class="comment">#           710           1711             62           1521           2470 </span></span><br><span class="line"><span class="comment">#     21_CD4.N2       22_CD4.M      23_CD8.EM      24_CD8.CM          25_NK </span></span><br><span class="line"><span class="comment">#          2364           3539            796           2080           2143 </span></span><br><span class="line"><span class="comment">#        26_Unk </span></span><br><span class="line"><span class="comment">#           161 </span></span><br></pre></td></tr></table></figure>

<p>我们后续会用到两种整合方法。第一种是<strong>无约束整合</strong>，我们对scATAC-seq实验里的细胞不作任何假设，直接尝试将这些细胞和scRNA-seq实验里的任意细胞进行配对。这是一种初步可行方案，后续会根据这一步得到的结果，对整合步骤进行约束来提升跨平台配对的质量。第二种方法是<strong>约束整合</strong>，即利用对细胞类型的先验知识限制搜索范围。举个例子，如果我们知道scATAC-seq中的Cluster A,B,C对应着三种不同的T细胞，scRNA-seq中的Cluster X,Y对应着两种不同的T细胞，我们告诉ArchR只需要尝试将scATAC-seq中的Cluster A,B,C跟scRNA-seq中的Cluster X,Y进行配对。下面，我们将先以无约束整合初步地鉴定每一种聚类的类型，然后根据分析结果做更加细致的约束整合。</p>
<h3 id="8-1-1-无约束整合"><a href="#8-1-1-无约束整合" class="headerlink" title="8.1.1 无约束整合"></a>8.1.1 无约束整合</h3><p>我们使用<code>addGeneIntegrationMatrix()</code>对scATAC-seq和scRNA-seq数据进行整合。正如之前所提到的，该函数的<code>seRNA</code>参数接受<code>Seurat</code>或<code>RangedSummarizedExperiment</code>对象作为输入。因为第一轮是探索性质的无约束整合，因此，我们不会将结果保存在Arrow文件中(<code>addToArrow=FALSE</code>)。整合后的矩阵将会根据<code>matrixName</code>进行命名，存放在<code>ArchRProject</code>中。该函数的其他参数对应<code>cellColData</code>中列名用于存放额外的信息，<code>nameCell</code>存放scRNA-seq中匹配的细胞ID，<code>nameGroup</code>存放scRNA-seq细胞中的分组ID，<code>nameScore</code>存放跨平台整合得分。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addGeneIntegrationMatrix<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    useMatrix <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    matrixName <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span></span><br><span class="line">    seRNA <span class="operator">=</span> seRNA<span class="punctuation">,</span></span><br><span class="line">    addToArrow <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span></span><br><span class="line">    groupRNA <span class="operator">=</span> <span class="string">&quot;BioClassification&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameCell <span class="operator">=</span> <span class="string">&quot;predictedCell_Un&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameGroup <span class="operator">=</span> <span class="string">&quot;predictedGroup_Un&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameScore <span class="operator">=</span> <span class="string">&quot;predictedScore_Un&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>无约束整合的结果可能并不准确，但是为后续更加精细的约束分析奠定了基础。</p>
<h3 id="8-1-2-约束整合"><a href="#8-1-2-约束整合" class="headerlink" title="8.1.2 约束整合"></a>8.1.2 约束整合</h3><p>现在我们已经有了初步的无约束整合结果，我们就有了大致的细胞类型分布情况，接着就是优化整合结果。</p>
<p>因为我们教程里的数据来自于造血细胞，我们将会在理想状态下将类似的细胞整合在一起。首先，我们先确认scRNA-seq里细胞类型在我们的scATAC-seq聚类中分布情况。这一步的目标是使用无约束整合的方法找到scATAC-seq和scRNA-seq中和T细胞和NK细胞对应的聚类，后续会用到该信息进行约束整合。具体操作为，我们创建一个<code>confusionMatrix</code>,并关注<code>Cluster</code>和<code>predictedGroup_Un</code>的交叉部分中scRNA-seq的细胞类型。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cM <span class="operator">&lt;-</span> as.matrix<span class="punctuation">(</span>confusionMatrix<span class="punctuation">(</span>projHeme2<span class="operator">$</span>Clusters<span class="punctuation">,</span> projHeme2<span class="operator">$</span>predictedGroup_Un<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">preClust <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>cM<span class="punctuation">)</span><span class="punctuation">[</span>apply<span class="punctuation">(</span>cM<span class="punctuation">,</span> <span class="number">1</span> <span class="punctuation">,</span> which.max<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">cbind<span class="punctuation">(</span>preClust<span class="punctuation">,</span> rownames<span class="punctuation">(</span>cM<span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">#Assignments</span></span><br></pre></td></tr></table></figure>

<p>输出信息如下，展示了12个scATAC-seq聚类中对应最优可能的scRNA-seq细胞类型。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#      preClust              </span></span><br><span class="line"><span class="comment"># [1,] &quot;03_Late.Eryth&quot;  &quot;C10&quot;</span></span><br><span class="line"><span class="comment"># [2,] &quot;20_CD4.N1&quot;      &quot;C8&quot; </span></span><br><span class="line"><span class="comment"># [3,] &quot;16_Pre.B&quot;       &quot;C3&quot; </span></span><br><span class="line"><span class="comment"># [4,] &quot;08_GMP.Neut&quot;    &quot;C11&quot;</span></span><br><span class="line"><span class="comment"># [5,] &quot;17_B&quot;           &quot;C4&quot; </span></span><br><span class="line"><span class="comment"># [6,] &quot;11_CD14.Mono.1&quot; &quot;C1&quot; </span></span><br><span class="line"><span class="comment"># [7,] &quot;01_HSC&quot;         &quot;C12&quot;</span></span><br><span class="line"><span class="comment"># [8,] &quot;22_CD4.M&quot;       &quot;C9&quot; </span></span><br><span class="line"><span class="comment"># [9,] &quot;09_pDC&quot;         &quot;C5&quot; </span></span><br><span class="line"><span class="comment"># [10,] &quot;25_NK&quot;          &quot;C7&quot; </span></span><br><span class="line"><span class="comment"># [11,] &quot;12_CD14.Mono.2&quot; &quot;C2&quot; </span></span><br><span class="line"><span class="comment"># [12,] &quot;06_CLP.1&quot;       &quot;C6&quot;</span></span><br></pre></td></tr></table></figure>

<p>首先，我们检查在无约束整合中用到的scRNA-seq数据里细胞类型标签。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">unique<span class="punctuation">(</span>unique<span class="punctuation">(</span>projHeme2<span class="operator">$</span>predictedGroup_Un<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># [1] &quot;08_GMP.Neut&quot;    &quot;25_NK&quot;          &quot;16_Pre.B&quot;       &quot;06_CLP.1&quot;      </span></span><br><span class="line"><span class="comment"># [5] &quot;07_GMP&quot;         &quot;11_CD14.Mono.1&quot; &quot;04_Early.Baso&quot;  &quot;22_CD4.M&quot;      </span></span><br><span class="line"><span class="comment"># [9] &quot;03_Late.Eryth&quot;  &quot;05_CMP.LMPP&quot;    &quot;17_B&quot;           &quot;19_CD8.N&quot;      </span></span><br><span class="line"><span class="comment">#[13] &quot;09_pDC&quot;         &quot;13_CD16.Mono&quot;   &quot;23_CD8.EM&quot;      &quot;12_CD14.Mono.2&quot;</span></span><br><span class="line"><span class="comment">#[17] &quot;20_CD4.N1&quot;      &quot;02_Early.Eryth&quot; &quot;21_CD4.N2&quot;      &quot;24_CD8.CM&quot;     </span></span><br><span class="line"><span class="comment">#[21] &quot;01_HSC&quot; </span></span><br></pre></td></tr></table></figure>

<p>从上面的输出中，我们发现scRNA-seq数据中与NK细胞和T细胞对应的聚类是Cluster 19 - 25。</p>
<p>接着我们创建一个字符串模式用来表示这些聚类，后续的约束整合会用到，其中<code>|</code>在正则表达式中表示”或”，我们之后使用<code>grep</code>根据这些字符串模式从scATAC-seq提取和scRNA-seq对应的聚类。。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#From scRNA</span></span><br><span class="line">cTNK <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span>paste0<span class="punctuation">(</span><span class="number">19</span><span class="operator">:</span><span class="number">25</span><span class="punctuation">)</span><span class="punctuation">,</span> collapse<span class="operator">=</span><span class="string">&quot;|&quot;</span><span class="punctuation">)</span></span><br><span class="line">cTNK</span><br><span class="line"><span class="comment"># [1] &quot;19|20|21|22|23|24|25&quot;</span></span><br></pre></td></tr></table></figure>

<p>其余的聚类就称之为”Non-T cell, Non-NK cell”(例如Cluster 1 - 18)，也创建了对应的字符串模式</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cNonTNK <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span>paste0<span class="punctuation">(</span><span class="string">&quot;0&quot;</span><span class="punctuation">,</span> <span class="number">1</span><span class="operator">:</span><span class="number">9</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">10</span><span class="operator">:</span><span class="number">13</span><span class="punctuation">,</span> <span class="number">15</span><span class="operator">:</span><span class="number">18</span><span class="punctuation">)</span><span class="punctuation">,</span> collapse<span class="operator">=</span><span class="string">&quot;|&quot;</span><span class="punctuation">)</span></span><br><span class="line">cNonTNK</span><br><span class="line"><span class="comment">#[1] &quot;01|02|03|04|05|06|07|08|09|10|11|12|13|15|16|17|18&quot;</span></span><br></pre></td></tr></table></figure>

<p>接着再用字符串模式在<code>preClust</code>找到对应的scATAC-seq列名，然后使用列名从混合矩阵提取对应的列。</p>
<p>对于T细胞和NK细胞，scATAC-seq聚类ID就是C7, C8, C9</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#Assign scATAC to these categories</span></span><br><span class="line">clustTNK <span class="operator">&lt;-</span> rownames<span class="punctuation">(</span>cM<span class="punctuation">)</span><span class="punctuation">[</span>grep<span class="punctuation">(</span>cTNK<span class="punctuation">,</span> preClust<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">clustTNK</span><br><span class="line"><span class="comment">#[1] &quot;C8&quot; &quot;C9&quot; &quot;C7&quot;</span></span><br></pre></td></tr></table></figure>

<p>对于” Non-T cells and Non-NK cells”, ID就是scATAC-seq聚类余下的部分</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">clustNonTNK <span class="operator">&lt;-</span> rownames<span class="punctuation">(</span>cM<span class="punctuation">)</span><span class="punctuation">[</span>grep<span class="punctuation">(</span>cNonTNK<span class="punctuation">,</span> preClust<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">clustNonTNK</span><br><span class="line"><span class="comment"># [1] &quot;C10&quot; &quot;C3&quot;  &quot;C11&quot; &quot;C4&quot;  &quot;C1&quot;  &quot;C12&quot; &quot;C5&quot;  &quot;C2&quot;  &quot;C6&quot; </span></span><br></pre></td></tr></table></figure>

<p>接着在scRNA-seq中做相同的操作，筛选出相同的细胞类型。首先，我们鉴定scRNA-seq数据中T细胞和NK细胞</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#RNA get cells in these categories</span></span><br><span class="line">rnaTNK <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="punctuation">[</span>grep<span class="punctuation">(</span>cTNK<span class="punctuation">,</span> colData<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="operator">$</span>BioClassification<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">head<span class="punctuation">(</span>rnaTNK<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#[1] &quot;PBMC_10x_GREENLEAF_REP1:AAACCCAGTCGTCATA-1&quot;</span></span><br><span class="line"><span class="comment">#[2] &quot;PBMC_10x_GREENLEAF_REP1:AAACCCATCCGATGTA-1&quot;</span></span><br><span class="line"><span class="comment">#[3] &quot;PBMC_10x_GREENLEAF_REP1:AAACCCATCTCAACGA-1&quot;</span></span><br><span class="line"><span class="comment">#[4] &quot;PBMC_10x_GREENLEAF_REP1:AAACCCATCTCTCGAC-1&quot;</span></span><br><span class="line"><span class="comment">#[5] &quot;PBMC_10x_GREENLEAF_REP1:AAACGAACAATCGTCA-1&quot;</span></span><br><span class="line"><span class="comment">#[6] &quot;PBMC_10x_GREENLEAF_REP1:AAACGAACACGATTCA-1&quot;</span></span><br></pre></td></tr></table></figure>

<p>然后，鉴定scRNA-seq数据中”Non-T cell Non-NK cell cells”</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">rnaNonTNK <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="punctuation">[</span>grep<span class="punctuation">(</span>cNonTNK<span class="punctuation">,</span> colData<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="operator">$</span>BioClassification<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">head<span class="punctuation">(</span>rnaNonTNK<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#[1] &quot;CD34_32_R5:AAACCTGAGTATCGAA-1&quot; &quot;CD34_32_R5:AAACCTGAGTCGTTTG-1&quot;</span></span><br><span class="line"><span class="comment">#[3] &quot;CD34_32_R5:AAACCTGGTTCCACAA-1&quot; &quot;CD34_32_R5:AAACGGGAGCTTCGCG-1&quot;</span></span><br><span class="line"><span class="comment">#[5] &quot;CD34_32_R5:AAACGGGAGGGAGTAA-1&quot; &quot;CD34_32_R5:AAACGGGAGTTACGGG-1&quot;</span></span><br></pre></td></tr></table></figure>

<p>约束整合需要我们提供一个嵌套list。这是一个由多个<code>SimpleList</code>对象组成的<code>SimpleList</code>, 每一组对应一个约束情况。在案例中，我们有两个组，一个组称之为<code>TNK</code>,包括两个平台的T和NK细胞，另一个组为<code>NonTNK</code>，包括两个平台的”Non-T cell Non-NK cell”细胞。每个<code>SimpleList</code>对象都包含两个细胞ID的向量，一个是ATAC，一个是RNA.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">groupList <span class="operator">&lt;-</span> SimpleList<span class="punctuation">(</span></span><br><span class="line">    TNK <span class="operator">=</span> SimpleList<span class="punctuation">(</span></span><br><span class="line">        ATAC <span class="operator">=</span> projHeme2<span class="operator">$</span>cellNames<span class="punctuation">[</span>projHeme2<span class="operator">$</span>Clusters <span class="operator">%in%</span> clustTNK<span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">        RNA <span class="operator">=</span> rnaTNK</span><br><span class="line">    <span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    NonTNK <span class="operator">=</span> SimpleList<span class="punctuation">(</span></span><br><span class="line">        ATAC <span class="operator">=</span> projHeme2<span class="operator">$</span>cellNames<span class="punctuation">[</span>projHeme2<span class="operator">$</span>Clusters <span class="operator">%in%</span> clustNonTNK<span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">        RNA <span class="operator">=</span> rnaNonTNK</span><br><span class="line">    <span class="punctuation">)</span>    </span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们将该列表传递给<code>addGeneIntegrationMatrix()</code>函数的<code>groupList</code>参数。注意，我们依旧没有将结果添加到Arrow文件中 (<code>addToArrow = FALSE</code>)。我们强烈建议，在保存到Arrow文件前先彻底的检查结果，看结果是否符合预期。在教程的下一节会介绍该操作。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">projHeme2 <span class="operator">&lt;-</span> addGeneIntegrationMatrix<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    useMatrix <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    matrixName <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span></span><br><span class="line">    seRNA <span class="operator">=</span> seRNA<span class="punctuation">,</span></span><br><span class="line">    addToArrow <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> </span><br><span class="line">    groupList <span class="operator">=</span> groupList<span class="punctuation">,</span></span><br><span class="line">    groupRNA <span class="operator">=</span> <span class="string">&quot;BioClassification&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameCell <span class="operator">=</span> <span class="string">&quot;predictedCell_Co&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameGroup <span class="operator">=</span> <span class="string">&quot;predictedGroup_Co&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameScore <span class="operator">=</span> <span class="string">&quot;predictedScore_Co&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="8-1-3-约束整合和无约束整合对比"><a href="#8-1-3-约束整合和无约束整合对比" class="headerlink" title="8.1.3 约束整合和无约束整合对比"></a>8.1.3 约束整合和无约束整合对比</h3><p>正如之前所提到的，我们的scATAC-seq数据可以根据整合的scRNA-seq数据进行定义，并且有约束和无约束这两种方式。为了对两者进行对比，我们分别根据这两种整合结果对scATAC-seq的数据进行上色。</p>
<p>首先，使用ArchR内置的<code>paletteDiscrete()</code>函数生成调色板</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">pal <span class="operator">&lt;-</span> paletteDiscrete<span class="punctuation">(</span>values <span class="operator">=</span> colData<span class="punctuation">(</span>seRNA<span class="punctuation">)</span><span class="operator">$</span>BioClassification<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>在ArchR中，调色板本质上一个命名向量，每个十六进制编码的颜色对应着一个名字。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">pal</span><br><span class="line"><span class="comment">#        01_HSC 02_Early.Eryth  03_Late.Eryth  04_Early.Baso    05_CMP.LMPP </span></span><br><span class="line"><span class="comment">#     &quot;#D51F26&quot;      &quot;#502A59&quot;      &quot;#235D55&quot;      &quot;#3D6E57&quot;      &quot;#8D2B8B&quot; </span></span><br><span class="line"><span class="comment">#      06_CLP.1         07_GMP    08_GMP.Neut         09_pDC         10_cDC </span></span><br><span class="line"><span class="comment">#     &quot;#DE6C3E&quot;      &quot;#F9B712&quot;      &quot;#D8CE42&quot;      &quot;#8E9ACD&quot;      &quot;#B774B1&quot; </span></span><br><span class="line"><span class="comment">#11_CD14.Mono.1 12_CD14.Mono.2   13_CD16.Mono         14_Unk       15_CLP.2 </span></span><br><span class="line"><span class="comment">#     &quot;#D69FC8&quot;      &quot;#C7C8DE&quot;      &quot;#8FD3D4&quot;      &quot;#89C86E&quot;      &quot;#CC9672&quot; </span></span><br><span class="line"><span class="comment">#      16_Pre.B           17_B      18_Plasma       19_CD8.N      20_CD4.N1 </span></span><br><span class="line"><span class="comment">#     &quot;#CF7E96&quot;      &quot;#A27AA4&quot;      &quot;#CD4F32&quot;      &quot;#6B977E&quot;      &quot;#518AA3&quot; </span></span><br><span class="line"><span class="comment">#     21_CD4.N2       22_CD4.M      23_CD8.EM      24_CD8.CM          25_NK </span></span><br><span class="line"><span class="comment">#     &quot;#5A5297&quot;      &quot;#0F707D&quot;      &quot;#5E2E32&quot;      &quot;#A95A3C&quot;      &quot;#B28D5C&quot; </span></span><br><span class="line"><span class="comment">#        26_Unk </span></span><br><span class="line"><span class="comment">#     &quot;#3D3D3D&quot;</span></span><br></pre></td></tr></table></figure>

<p>我们在scATAC-seq数据根据无约束整合得到的scRNA-seq细胞类型进行可视化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span></span><br><span class="line">    projHeme2<span class="punctuation">,</span> </span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;predictedGroup_Un&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    pal <span class="operator">=</span> pal</span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">p1</span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-94b3538cffe94b2d9d2945ae01429b89.png" alt="Plot-UMAP-RNA-Integration_1" style="zoom:25%;" />

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">do.call<span class="punctuation">(</span>cowplot<span class="operator">::</span>plot_grid<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">list</span><span class="punctuation">(</span>ncol <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span> p2c<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>同样，我们也可以根据约束整合得到scATAC-seq对应的scRNA-seq的细胞类型进行可视化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">p2 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span></span><br><span class="line">    projHeme2<span class="punctuation">,</span> </span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;predictedGroup_Co&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    pal <span class="operator">=</span> pal</span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">p2</span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-a8c23615cd644c859ac5e80b45c25230.png" alt="Plot-UMAP-RNA-Integration_2" style="zoom:25%;" />

<p>这两者的结果差异其实非常细微，主要是我们感兴趣的细胞类型原本就存在明显的差异。当然，仔细观察还能发现其中的不同之处，尤其是T细胞(Clusters 17-22)</p>
<p>我们用<code>plotPDF()</code>函数保存该图矢量版本。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span>p2<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-UMAP-RNA-Integration.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们现在可以用<code>saveArchRProject()</code>函数保存我们的<code>projHeme2</code>对象。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">saveArchRProject<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> outputDirectory <span class="operator">=</span> <span class="string">&quot;Save-ProjHeme2&quot;</span><span class="punctuation">,</span> load <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="8-2-为每个scATAC-seq细胞增加拟scRNA-seq谱"><a href="#8-2-为每个scATAC-seq细胞增加拟scRNA-seq谱" class="headerlink" title="8.2 为每个scATAC-seq细胞增加拟scRNA-seq谱"></a>8.2 为每个scATAC-seq细胞增加拟scRNA-seq谱</h2><p>既然我们对scATAC-seq和scRNA-seq整合的结果感到满意，我们就能用<code>addToArrow=TRUE</code>重新运行，在Arrow文件中添加相关联的表达量矩阵数据。根据之前所提到的，我们传入<code>groupList</code>约束整合，在<code>nameCell</code>，<code>nameGroup</code>和<code>nameScore</code>中加入列名。这些元信息列都会被添加到<code>cellColData</code>中。</p>
<p>这里，我们新建了一个<code>projHeme3</code>，用于后续教程。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#~5 minutes</span></span><br><span class="line">projHeme3 <span class="operator">&lt;-</span> addGeneIntegrationMatrix<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> </span><br><span class="line">    useMatrix <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    matrixName <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span><span class="punctuation">,</span></span><br><span class="line">    seRNA <span class="operator">=</span> seRNA<span class="punctuation">,</span></span><br><span class="line">    addToArrow <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span></span><br><span class="line">    force<span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span></span><br><span class="line">    groupList <span class="operator">=</span> groupList<span class="punctuation">,</span></span><br><span class="line">    groupRNA <span class="operator">=</span> <span class="string">&quot;BioClassification&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameCell <span class="operator">=</span> <span class="string">&quot;predictedCell&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameGroup <span class="operator">=</span> <span class="string">&quot;predictedGroup&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nameScore <span class="operator">=</span> <span class="string">&quot;predictedScore&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>现在，当我们使用<code>getAvailableMatrices()</code>检查哪些矩阵可用时，我们会发现<code>GeneIntegrationMatrix</code>已经被添加到Arrow文件中</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">getAvailableMatrices<span class="punctuation">(</span>projHeme3<span class="punctuation">)</span></span><br><span class="line"><span class="comment"># [1] &quot;GeneIntegrationMatrix&quot; &quot;GeneScoreMatrix&quot; &quot;TileMatrix&quot;</span></span><br></pre></td></tr></table></figure>

<p>在新的<code>GeneIntegrationMatrix</code>中，我们可以比较连接的基因表达量和根据基因得分推断的基因表达量</p>
<p>我们需要先确保在项目中加入了填充权重值(impute weights)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme3 <span class="operator">&lt;-</span> addImputeWeights<span class="punctuation">(</span>projHeme3<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>现在，我们来生成一些UMAP图，里面的<strong>基因表达量</strong>值来自于<code>GeneIntegrationMatrix</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;CD34&quot;</span><span class="punctuation">,</span> <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span><span class="punctuation">,</span> <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MS4A1&quot;</span><span class="punctuation">,</span> <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span><span class="punctuation">,</span> <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD8A&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TBX21&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IL7R&quot;</span> <span class="comment">#TCells</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p1 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme3<span class="punctuation">,</span> </span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> markerGenes<span class="punctuation">,</span> </span><br><span class="line">    continuousSet <span class="operator">=</span> <span class="string">&quot;horizonExtra&quot;</span><span class="punctuation">,</span></span><br><span class="line">    embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    imputeWeights <span class="operator">=</span> getImputeWeights<span class="punctuation">(</span>projHeme3<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>以及一些相同UMAP图，但是使用<code>GeneScoreMatrix</code>里的<strong>基因得分</strong>值</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">p2 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme3<span class="punctuation">,</span> </span><br><span class="line">    colorBy <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    continuousSet <span class="operator">=</span> <span class="string">&quot;horizonExtra&quot;</span><span class="punctuation">,</span></span><br><span class="line">    name <span class="operator">=</span> markerGenes<span class="punctuation">,</span> </span><br><span class="line">    embedding <span class="operator">=</span> <span class="string">&quot;UMAP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    imputeWeights <span class="operator">=</span> getImputeWeights<span class="punctuation">(</span>projHeme3<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>最后用<code>cowplot</code>将这些标记基因绘制在一起</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">p1c <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>p1<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    x <span class="operator">+</span> guides<span class="punctuation">(</span>color <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> fill <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">    theme_ArchR<span class="punctuation">(</span>baseSize <span class="operator">=</span> <span class="number">6.5</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span>plot.margin <span class="operator">=</span> unit<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;cm&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span></span><br><span class="line">        axis.text.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.text.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">    <span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p2c <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>p2<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    x <span class="operator">+</span> guides<span class="punctuation">(</span>color <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> fill <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">    theme_ArchR<span class="punctuation">(</span>baseSize <span class="operator">=</span> <span class="number">6.5</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span>plot.margin <span class="operator">=</span> unit<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;cm&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">    theme<span class="punctuation">(</span></span><br><span class="line">        axis.text.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.x<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.text.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">        axis.ticks.y<span class="operator">=</span>element_blank<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">    <span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">do.call<span class="punctuation">(</span>cowplot<span class="operator">::</span>plot_grid<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">list</span><span class="punctuation">(</span>ncol <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span> p1c<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-4ed685c296af4038bbb7c47852bfbf1e.png" alt="Plot-UMAP-Markers-RNA-W-Imputation_1" style="zoom:50%;" />

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">do.call<span class="punctuation">(</span>cowplot<span class="operator">::</span>plot_grid<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">list</span><span class="punctuation">(</span>ncol <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span> p2c<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<img src="/upload/2020/05/image-837cacbfb4534818b998b844551299dd.png" alt="Plot-UMAP-Markers-RNA-W-Imputation_2" style="zoom:50%;" />

<p>和预期的一样，两个方法推测的基因表达量存在相似性，但并不相同。</p>
<p>使用<code>plotPDF()</code>函数保存可编辑的矢量版。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>plotList <span class="operator">=</span> p1<span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;Plot-UMAP-Marker-Genes-RNA-W-Imputation.pdf&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme3<span class="punctuation">,</span> </span><br><span class="line">    addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="8-3-使用scRNA-seq信息标记scATAC-seq聚类"><a href="#8-3-使用scRNA-seq信息标记scATAC-seq聚类" class="headerlink" title="8.3 使用scRNA-seq信息标记scATAC-seq聚类"></a>8.3 使用scRNA-seq信息标记scATAC-seq聚类</h2><p>现在，我们确定了scATAC-seq和scRNA-seq数据间的对应关系，我们就可以使用scRNA-seq数据中细胞类型对我们的scATAC-seq聚类进行定义。</p>
<p>首先，我们会在scATAC-seq和整合分析得到<code>predictedGroup</code>之间构建一个混合矩阵</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cM <span class="operator">&lt;-</span> confusionMatrix<span class="punctuation">(</span>projHeme3<span class="operator">$</span>Clusters<span class="punctuation">,</span> projHeme3<span class="operator">$</span>predictedGroup<span class="punctuation">)</span></span><br><span class="line">labelOld <span class="operator">&lt;-</span> rownames<span class="punctuation">(</span>cM<span class="punctuation">)</span></span><br><span class="line">labelOld</span><br><span class="line"><span class="comment">#  [1] &quot;Cluster11&quot; &quot;Cluster2&quot;  &quot;Cluster12&quot; &quot;Cluster1&quot;  &quot;Cluster8&quot;  &quot;Cluster4&quot; </span></span><br><span class="line"><span class="comment">#  [7] &quot;Cluster9&quot;  &quot;Cluster5&quot;  &quot;Cluster7&quot;  &quot;Cluster14&quot; &quot;Cluster3&quot;  &quot;Cluster10&quot;</span></span><br><span class="line"><span class="comment"># [13] &quot;Cluster6&quot;  &quot;Cluster13&quot;</span></span><br></pre></td></tr></table></figure>

<p>接着，对于每一个scATAC-seq聚类，我们根据<code>predictedGroup</code>确定最能定义聚类的细胞类型。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">labelNew <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>cM<span class="punctuation">)</span><span class="punctuation">[</span>apply<span class="punctuation">(</span>cM<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> which.max<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">labelNew</span><br></pre></td></tr></table></figure>

<p>接着，我们需要对新的聚类标签进行重命名，简化分类系统。对于每一个scRNA-seq的聚类，我们会重新定义标签，以便更好解释。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">remapClust <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;01_HSC&quot;</span> <span class="operator">=</span> <span class="string">&quot;Progenitor&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;02_Early.Eryth&quot;</span> <span class="operator">=</span> <span class="string">&quot;Erythroid&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;03_Late.Eryth&quot;</span> <span class="operator">=</span> <span class="string">&quot;Erythroid&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;04_Early.Baso&quot;</span> <span class="operator">=</span> <span class="string">&quot;Basophil&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;05_CMP.LMPP&quot;</span> <span class="operator">=</span> <span class="string">&quot;Progenitor&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;06_CLP.1&quot;</span> <span class="operator">=</span> <span class="string">&quot;CLP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;07_GMP&quot;</span> <span class="operator">=</span> <span class="string">&quot;GMP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;08_GMP.Neut&quot;</span> <span class="operator">=</span> <span class="string">&quot;GMP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;09_pDC&quot;</span> <span class="operator">=</span> <span class="string">&quot;pDC&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;10_cDC&quot;</span> <span class="operator">=</span> <span class="string">&quot;cDC&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;11_CD14.Mono.1&quot;</span> <span class="operator">=</span> <span class="string">&quot;Mono&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;12_CD14.Mono.2&quot;</span> <span class="operator">=</span> <span class="string">&quot;Mono&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;13_CD16.Mono&quot;</span> <span class="operator">=</span> <span class="string">&quot;Mono&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;15_CLP.2&quot;</span> <span class="operator">=</span> <span class="string">&quot;CLP&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;16_Pre.B&quot;</span> <span class="operator">=</span> <span class="string">&quot;PreB&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;17_B&quot;</span> <span class="operator">=</span> <span class="string">&quot;B&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;18_Plasma&quot;</span> <span class="operator">=</span> <span class="string">&quot;Plasma&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;19_CD8.N&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD8.N&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;20_CD4.N1&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD4.N&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;21_CD4.N2&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD4.N&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;22_CD4.M&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD4.M&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;23_CD8.EM&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD8.EM&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;24_CD8.CM&quot;</span> <span class="operator">=</span> <span class="string">&quot;CD8.CM&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;25_NK&quot;</span> <span class="operator">=</span> <span class="string">&quot;NK&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">remapClust <span class="operator">&lt;-</span> remapClust<span class="punctuation">[</span><span class="built_in">names</span><span class="punctuation">(</span>remapClust<span class="punctuation">)</span> <span class="operator">%in%</span> labelNew<span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<p>接着，使用<code>mapLables()</code>函数进行标签转换，将旧的标签映射到新的标签上。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">labelNew2 <span class="operator">&lt;-</span> mapLabels<span class="punctuation">(</span>labelNew<span class="punctuation">,</span> oldLabels <span class="operator">=</span> <span class="built_in">names</span><span class="punctuation">(</span>remapClust<span class="punctuation">)</span><span class="punctuation">,</span> newLabels <span class="operator">=</span> remapClust<span class="punctuation">)</span></span><br><span class="line">labelNew2</span><br><span class="line"><span class="comment">#  [1] &quot;GMP&quot;        &quot;B&quot;          &quot;PreB&quot;       &quot;CD4.N&quot;      &quot;Mono&quot;      </span></span><br><span class="line"><span class="comment">#  [6] &quot;Erythroid&quot;  &quot;Progenitor&quot; &quot;CD4.M&quot;      &quot;pDC&quot;        &quot;NK&quot;        </span></span><br><span class="line"><span class="comment"># [11] &quot;CLP&quot;        &quot;Mono&quot;</span></span><br></pre></td></tr></table></figure>

<p>合并<code>labelsOld</code>和<code>labelsNew2</code>，我们现在可以用<code>mapLables()</code>函数在<code>cellColData</code>里新建聚类标签。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">projHeme3<span class="operator">$</span>Clusters2 <span class="operator">&lt;-</span> mapLabels<span class="punctuation">(</span>projHeme3<span class="operator">$</span>Clusters<span class="punctuation">,</span> newLabels <span class="operator">=</span> labelNew2<span class="punctuation">,</span> oldLabels <span class="operator">=</span> labelOld<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>得到新的标签后，我们绘制最新的UMAP展示聚类结果。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">p1 <span class="operator">&lt;-</span> plotEmbedding<span class="punctuation">(</span>projHeme3<span class="punctuation">,</span> colorBy <span class="operator">=</span> <span class="string">&quot;cellColData&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">)</span></span><br><span class="line">p1</span><br></pre></td></tr></table></figure>

<p>如果被分析scATAC-seq数据对应的细胞系统也有scRNA-seq数据存在，那么这种分析范式将会特别有用。正如之前所说，这种scRNA-seq和scATAC-seq整合分析也为后续更加复杂的基因调控分析提供了出色的框架。</p>
<p>用<code>plotPDF()</code>函数保存该图矢量版本。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>p1<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;Plot-UMAP-Remap-Clusters.pdf&quot;</span><span class="punctuation">,</span> ArchRProj <span class="operator">=</span> projHeme2<span class="punctuation">,</span> addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>使用<code>saveArchRProject</code>保存我们最初的projHeme3.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">saveArchRProject<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme3<span class="punctuation">,</span> outputDirectory <span class="operator">=</span> <span class="string">&quot;Save-ProjHeme3&quot;</span><span class="punctuation">,</span> load <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>


      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2020/05/31/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%85%AB%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter8/" data-id="cllgasdsu00a1dkov7sfi9ucg" data-title="使用ArchR分析单细胞ATAC-seq数据(第八章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/9/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/8/">8</a><a class="page-number" href="/page/9/">9</a><span class="page-number current">10</span><a class="page-number" href="/page/11/">11</a><a class="page-number" href="/page/12/">12</a><span class="space">&hellip;</span><a class="page-number" href="/page/24/">24</a><a class="extend next" rel="next" href="/page/11/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
          <li>
            <a href="/2023/06/22/%E5%9C%A8Windows%E4%B8%AD%E6%90%AD%E5%BB%BA%E4%B8%80%E4%B8%AA%E5%9F%BA%E4%BA%8EWSL%E7%9A%84Python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6%E7%8E%AF%E5%A2%83-building-a-wsl-based-python-data-science-environment-in-windows/">在Windows中搭建一个基于WSL的Python数据科学环境</a>
          </li>
        
          <li>
            <a href="/2023/06/16/%E7%94%A8Colab%E7%BC%93%E8%A7%A3%E6%88%91%E7%9A%84GPU%E7%84%A6%E8%99%91-%E7%94%A8colab%E7%BC%93%E8%A7%A3%E6%88%91%E7%9A%84gpu%E7%84%A6%E8%99%91/">用Colab缓解我的GPU焦虑</a>
          </li>
        
          <li>
            <a href="/2023/05/31/%E6%8E%8C%E6%8F%A1%E4%BA%86%E4%B8%80%E4%B8%AA%E5%85%B3%E4%BA%8EBcrypt%E5%8A%A0%E5%AF%86%E7%9A%84%E7%9F%A5%E8%AF%86-acquired-a-knowledge-of-bcrypt-encryption/">掌握了一个关于Bcrypt加密的知识</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>