<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/19/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-如何使用BSA方法进行遗传定位（水稻篇）-QTL-analysis-using-BSA-seq-in-rice" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/04/%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8BSA%E6%96%B9%E6%B3%95%E8%BF%9B%E8%A1%8C%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D%EF%BC%88%E6%B0%B4%E7%A8%BB%E7%AF%87%EF%BC%89-QTL-analysis-using-BSA-seq-in-rice/" class="article-date">
  <time class="dt-published" datetime="2019-10-04T10:45:49.860Z" itemprop="datePublished">2019-10-04</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/04/%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8BSA%E6%96%B9%E6%B3%95%E8%BF%9B%E8%A1%8C%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D%EF%BC%88%E6%B0%B4%E7%A8%BB%E7%AF%87%EF%BC%89-QTL-analysis-using-BSA-seq-in-rice/">如何使用BSA方法进行遗传定位（水稻篇）</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>BSA虽然不是我最早接触的高通量数据分析项目（最早的是RNA-seq），但是却是我最早独立开展的数据分析项目， 我曾经专门写过一篇文章介绍如何使用SHOREMap做<a href="/archives/Mapping-by-sequencing-Using-SHOREmap">拟南芥的EMS诱变群体的BSA分析</a></p>
<p>在遗传定位上，相对于GWAS和binmap，BSA是一个比较省钱的策略，它只需要测两个亲本和后代中两个极端差异群体即可，但是它对实验设计，表型考察，样本挑选都有比较高的要求。如果你的表型差异并不是泾渭分明，那么还是不要用BSA比较合适。这方面知识，建议阅读徐云碧老师2016年发表在PBJ上的”Bulked sample analysis in genetics, genomics and crop improvement”, 这篇教程侧重于实际分析，而非理论讲解。</p>
<p>用于讲解的文章题为”Identification of a cold-tolerant locus in rice (Oryza sativa L.) using bulked segregant analysis with a next-generation sequencing strategy”, 发表在Rice上。 文章用的耐寒(Kongyu131)和不耐寒(Dongnong422)的两个亲本进行杂交，得到的F2后代利用单粒传法得到RIL(重组自交系)群体.之后用BWA+GATK识别SNP，计算ED和SNP-index作图。</p>
<p>这次实战的目标也就是得到文章关键的两张图：</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/image-20191003181641928-89fb8e679a084735851669aea22bbf70.png" alt="ED"></p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/image-20191003181653883-32d9596aaef14375a83085d72436618a.png" alt="SNP-index"></p>
<h2 id="环境准备"><a href="#环境准备" class="headerlink" title="环境准备"></a>环境准备</h2><p>新建一个项目，用于存放本次分析的数据和结果</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p rice-bsa </span><br><span class="line"><span class="built_in">cd</span> rice-bsa</span><br></pre></td></tr></table></figure>

<p>后续分析还需要用到如下软件:</p>
<ul>
<li>wget: 一般Linux系统会自带，用于下载数据</li>
<li>seqkit: 多能的序列处理软件</li>
<li>fastp: 数据质控</li>
<li>bwa: 数据比对到参考基因组</li>
<li>samtools: 处理sam&#x2F;bam文件</li>
<li>sambamba:  标记重复序列</li>
<li>bcftools: 处理vcf&#x2F;bcf，能用于变异检测</li>
<li>R: 数据分析</li>
</ul>
<p>我们需要分别下载参考基因组(IRGSP)数据和测序数据。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p ref &amp;&amp; <span class="built_in">cd</span> ref</span><br><span class="line">wget https://rapdb.dna.affrc.go.jp/download/archive/irgsp1/IRGSP-1.0_genome.fasta.gz</span><br><span class="line">gunzip IRGSP-1.0_genome.fasta.gz</span><br><span class="line"><span class="comment"># 建立索引</span></span><br><span class="line">bwa index IRGSP-1.0_genome.fasta</span><br></pre></td></tr></table></figure>

<p>之后 根据文章提供的编号，SRR6327815, SRR6327816, SRR6327817, SRR6327818 ，我们到<a target="_blank" rel="noopener" href="https://www.ebi.ac.uk/ena/">ENA</a> 查找对应的下载链接进行数据下载。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p data &amp;&amp; <span class="built_in">cd</span> data</span><br><span class="line">wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR632/005/SRR6327815/SRR6327815_&#123;1,2&#125;.fastq.gz</span><br><span class="line">wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR632/006/SRR6327816/SRR6327816_&#123;1,2&#125;.fastq.gz</span><br><span class="line">wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR632/007/SRR6327817/SRR6327817_&#123;1,2&#125;.fastq.gz</span><br><span class="line">wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR632/008/SRR6327818/SRR6327818_&#123;1,2&#125;.fastq.gz</span><br></pre></td></tr></table></figure>

<blockquote>
<p>因为在NCBI上下载的是SRA格式，需要做数据转换，而从ENA上可以直接下载压缩过的fastq文件，所以我更偏好于ENA。</p>
</blockquote>
<h2 id="上游预处理"><a href="#上游预处理" class="headerlink" title="上游预处理"></a>上游预处理</h2><p>对于测序数据而言，为了能够获得后续用于计算SNP-index或者ED的SNP信息，我们需要将二代测序得到的数据回贴到参考基因组上，然后利用变异检测软件找到每个样本和参考基因组的区别。</p>
<h3 id="数据质控"><a href="#数据质控" class="headerlink" title="数据质控"></a>数据质控</h3><p>原始数据中可能有一部分序列质量不够高，会影响后续分析，比如说测序质量不够，或者说存在接头。通常我们会都会对原始数据进行一波过滤，这里用的是fastp，优点就是快。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p 01-clean-data</span><br><span class="line">fastp -i data/SRR6327815_1.fastq.gz -I data/SRR6327815_2.fastq.gz -o 01-clean-data/SRR6327815_1.fastq.gz -O 01-clean-data/SRR6327815_2.fastq.gz</span><br><span class="line">fastp -i data/SRR6327816_1.fastq.gz -I data/SRR6327816_2.fastq.gz -o 01-clean-data/SRR6327816_1.fastq.gz -O 01-clean-data/SRR6327816_2.fastq.gz</span><br><span class="line">fastp -i data/SRR6327817_1.fastq.gz -I data/SRR6327817_2.fastq.gz -o 01-clean-data/SRR6327817_1.fastq.gz -O 01-clean-data/SRR6327817_2.fastq.gz</span><br><span class="line">fastp -i data/SRR6327818_1.fastq.gz -I data/SRR6327818_2.fastq.gz -o 01-clean-data/SRR6327818_1.fastq.gz -O 01-clean-data/SRR6327818_2.fastq.gz</span><br></pre></td></tr></table></figure>

<h3 id="序列比对"><a href="#序列比对" class="headerlink" title="序列比对"></a>序列比对</h3><p>之后将各个样本的序列回帖到参考基因组</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p 02-read-align</span><br><span class="line">NUMBER_THREADS=80</span><br><span class="line">REFERENCE=ref/IRGSP-1.0_genome.fasta</span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> `<span class="built_in">ls</span> 01-clean-data/`; <span class="keyword">do</span> </span><br><span class="line">    sample=<span class="variable">$&#123;i%%_*&#125;</span></span><br><span class="line">    (bwa mem -M -R <span class="string">&quot;@RG\\tID:<span class="variable">$&#123;sample&#125;</span>\\tSM:<span class="variable">$&#123;sample&#125;</span>\\tPL:ILLUMINA&quot;</span> \</span><br><span class="line">        -t <span class="variable">$NUMBER_THREADS</span> <span class="variable">$REFERENCE</span> 01-clean-data/<span class="variable">$&#123;sample&#125;</span>_1.fastq.gz 01-clean-data/<span class="variable">$&#123;sample&#125;</span>_2.fastq.gz  || <span class="built_in">echo</span> -n <span class="string">&#x27;error&#x27;</span> ) \</span><br><span class="line">        | samtools <span class="built_in">sort</span> -@ 20 -o 02-read-align/<span class="variable">$&#123;sample&#125;</span>_sort.bam -</span><br><span class="line">        samtools index -@ 20 02-read-align/<span class="variable">$&#123;sample&#125;</span>_sort.bam</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<p>这里用到了稍微比较高级的shell操作</p>
<ul>
<li>变量名替换<code>sample=$&#123;i%%_*&#125;</code></li>
<li>逻辑判断: <code>||</code></li>
<li>子shell: <code>()</code></li>
<li>for循环: <code>for do done</code></li>
</ul>
<p>接着我门需要去除重复序列，这里的重复序列指的是拥有相同位置信息，且序列也一模一样的read，通常是由PCR扩增引起。过滤重复序列的目的是为了提高变异检测的准确性，如果一条read上出现的“变异”其实是在第一轮PCR扩增时引入的错误，那么后续的扩增只会让这个错误一直保留着，随后测序的时候这条许多又被多次测到，那么在后续的分析中由于多次出现，就有可能会变异检测软件当作真实变异。</p>
<p>这里用sambamba，因为它的速度比较快，且结果和picard一模一样。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> `<span class="built_in">ls</span> 02-read-align/*_sort.bam`; <span class="keyword">do</span></span><br><span class="line">    sample=<span class="variable">$&#123;i%%_*&#125;</span></span><br><span class="line">    sambamba markdup -r -t 10 <span class="variable">$&#123;sample&#125;</span>_sort.bam <span class="variable">$&#123;sample&#125;</span>_mkdup.bam </span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<h3 id="变异检测"><a href="#变异检测" class="headerlink" title="变异检测"></a>变异检测</h3><p>变异检测最常见的就是bcftools, freebayes和GATK. 这里用的是BCFtools，主要原因还是它的速度比较快。</p>
<p>这里为了让他的速度更快，我用了<code>--region</code>参数分染色体并行，由于水稻有12条染色体，相当于提速了12倍</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p 03-variants</span><br><span class="line"><span class="built_in">ls</span> -1 02-read-align/*_mkdup.bam &gt; bam_list.txt</span><br><span class="line">seqkit <span class="built_in">seq</span> -n ref/IRGSP-1.0_genome.fasta | \</span><br><span class="line"><span class="keyword">while</span> <span class="built_in">read</span> region</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">bcftools mpileup -f ref/IRGSP-1.0_genome.fasta  \</span><br><span class="line">	--redo-BAQ --min-BQ 30 \</span><br><span class="line">	--per-sample-mF \</span><br><span class="line">	--annotate FORMAT/AD,FORMAT/DP \</span><br><span class="line">	--regions <span class="variable">$&#123;region&#125;</span> \</span><br><span class="line">	-Ou --bam-list bam_list.txt  | \</span><br><span class="line">	bcftools call -<span class="built_in">mv</span> -Ob -o 03-variants/<span class="variable">$&#123;region&#125;</span>.bcf &amp;</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<p>之后将拆分运行的结果合并到单个文件中</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p 04-variant-filter</span><br><span class="line">bcftools concat --naive -o 04-variant-filter/merged.bcf 03-variants/*.bcf</span><br></pre></td></tr></table></figure>

<h3 id="变异过滤"><a href="#变异过滤" class="headerlink" title="变异过滤"></a>变异过滤</h3><p>得到VCF文件还需要进行一些过滤，来提高变异的准确性. 这个需要根据具体的项目来进行, 但是有一些固定的指标可以用来对结果进行过滤，例如</p>
<ul>
<li>测序深度</li>
<li>非参考基因组的高质量read数</li>
<li>是否和indel紧邻，通常和indel比较近的snp都不可靠</li>
<li>平均的比对质量,</li>
</ul>
<p>举个例子:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">cd</span> 04-variant-filter</span><br><span class="line"></span><br><span class="line">bcftools filter  -g3 -G10 -e<span class="string">&#x27;%QUAL&lt;10 || (RPB&lt;0.1 &amp;&amp; %QUAL&lt;15) || (AC&lt;2 &amp;&amp; %QUAL&lt;15) || MQ &lt; 30 || MQSB &lt;=0.1&#x27;</span>  merged.bcf &gt; filter.vcf</span><br></pre></td></tr></table></figure>

<p>之后，我只选择了snp用于后续分析</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 04-variant-filter目录下</span></span><br><span class="line">bcftools view -i <span class="string">&#x27;TYPE=&quot;snp&quot; &amp; N_ALT =1 &amp; STRLEN(ALT) = 1&#x27;</span> filter.vcf &gt; snps.vcf</span><br></pre></td></tr></table></figure>

<p>最终留下了将近80w的SNP。这就是后续R语言下游分析的基础。</p>
<blockquote>
<p>上游处理后的数据可以从链接:<a target="_blank" rel="noopener" href="https://pan.baidu.com/s/1n6CX33E6Qrpf85INMxGGAQ">https://pan.baidu.com/s/1n6CX33E6Qrpf85INMxGGAQ</a> ( 密码:q9u3 )下载</p>
</blockquote>
<h2 id="下游分析"><a href="#下游分析" class="headerlink" title="下游分析"></a>下游分析</h2><p>下游分析可能是比较成熟化的流程分析，对服务器的要求比较高一些，下游分析则是对背景知识要求高一些，例如VCF的格式，遗传学的三大定律等。</p>
<p>让我们先安装并加载所需的R包</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">install.pacakges<span class="punctuation">(</span><span class="string">&quot;devtools&quot;</span><span class="punctuation">)</span></span><br><span class="line">install.packages<span class="punctuation">(</span><span class="string">&quot;vcfR&quot;</span><span class="punctuation">)</span></span><br><span class="line">devtools<span class="operator">:</span>install_github<span class="punctuation">(</span><span class="string">&quot;xuzhougeng/binmapr&quot;</span><span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span><span class="string">&quot;vcfR&quot;</span><span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span><span class="string">&quot;binmapr&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>然后我们需要利用vcfR读取VCF文件</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">vcf <span class="operator">&lt;-</span> read.vcfR<span class="punctuation">(</span><span class="string">&quot;04-variant-filter/snps.vcf&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>接着从VCF对象中提取两个关键信息，AD(Allele Depth)和GT(Genotype)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">gt <span class="operator">&lt;-</span> extract.gt<span class="punctuation">(</span>vcf<span class="punctuation">)</span></span><br><span class="line">ad <span class="operator">&lt;-</span> extract.gt<span class="punctuation">(</span>vcf<span class="punctuation">,</span> <span class="string">&quot;AD&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>gt是一个基因型矩阵，基于之的前过滤操作，所以这里只会有”0&#x2F;0”, “0&#x2F;1”,”1&#x2F;1”这三种情况，而ad则是等位基因的count数. 我们用<code>head</code>查看前10行来了解下情况， </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> head<span class="punctuation">(</span>gt<span class="punctuation">)</span></span><br><span class="line">            SRR6327815 SRR6327816 SRR6327817 SRR6327818</span><br><span class="line">chr01_1151  <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/1&quot;</span>     </span><br><span class="line">chr01_6918  <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/1&quot;</span>     </span><br><span class="line">chr01_17263 <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/1&quot;</span>     </span><br><span class="line">chr01_21546 <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/1&quot;</span>     </span><br><span class="line">chr01_24732 <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/0&quot;</span>     </span><br><span class="line">chr01_33667 <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/1&quot;</span> </span><br><span class="line"></span><br><span class="line"><span class="operator">&gt;</span> head<span class="punctuation">(</span>ad<span class="punctuation">)</span></span><br><span class="line">            SRR6327815 SRR6327816 SRR6327817 SRR6327818</span><br><span class="line">chr01_1151  <span class="string">&quot;14,0&quot;</span>     <span class="string">&quot;0,18&quot;</span>     <span class="string">&quot;10,0&quot;</span>     <span class="string">&quot;11,3&quot;</span>    </span><br><span class="line">chr01_6918  <span class="string">&quot;31,0&quot;</span>     <span class="string">&quot;0,30&quot;</span>     <span class="string">&quot;32,0&quot;</span>     <span class="string">&quot;21,5&quot;</span>    </span><br><span class="line">chr01_17263 <span class="string">&quot;46,0&quot;</span>     <span class="string">&quot;0,34&quot;</span>     <span class="string">&quot;20,0&quot;</span>     <span class="string">&quot;21,8&quot;</span>    </span><br><span class="line">chr01_21546 <span class="string">&quot;0,25&quot;</span>     <span class="string">&quot;0,20&quot;</span>     <span class="string">&quot;17,0&quot;</span>     <span class="string">&quot;16,3&quot;</span>    </span><br><span class="line">chr01_24732 <span class="string">&quot;0,29&quot;</span>     <span class="string">&quot;36,0&quot;</span>     <span class="string">&quot;21,0&quot;</span>     <span class="string">&quot;31,0&quot;</span>    </span><br><span class="line">chr01_33667 <span class="string">&quot;0,26&quot;</span>     <span class="string">&quot;0,31&quot;</span>     <span class="string">&quot;28,0&quot;</span>     <span class="string">&quot;29,11&quot;</span> </span><br></pre></td></tr></table></figure>

<p>其中SRR6327815, SRR6327816, SRR6327817, SRR6327818 分别对应着 KY131, DN422, T-pool 和 S-pool</p>
<p>仔细观察的话，你会发现一个chr01_21546是一个有趣的位置，因为双亲都是纯合情况下，SRR6327818居然是 “16,3”, 基因型是”0&#x2F;1”, 这既有可能是亲本是杂合但没有测到，也有可能是后代测错了，一个简单粗暴的方法就是删掉它。此外我们还需要考虑是选择KY131是”1&#x2F;1”且 “DN422” 是 “0&#x2F;0”的位点。还是选择KY131是”0&#x2F;0”,  且 “DN422” 是 “1&#x2F;1”位点。最好的方法就是两种都测试一下。</p>
<p>首先测试KY131是”1&#x2F;1”且 “DN422” 是 “0&#x2F;0”的位点</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">mask <span class="operator">&lt;-</span> which<span class="punctuation">(</span>gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327815&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;1/1&quot;</span> <span class="operator">&amp;</span>  gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327816&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;0/0&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ad_flt <span class="operator">&lt;-</span> ad<span class="punctuation">[</span>mask<span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;SRR6327817&quot;</span><span class="punctuation">,</span> <span class="string">&quot;SRR6327818&quot;</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">colnames<span class="punctuation">(</span>ad_flt<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;T_Pool&quot;</span><span class="punctuation">,</span> <span class="string">&quot;S_Pool&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这一波过滤，我们从原来的80w位点中留下了20w个位点，因此<strong>测双亲很重要</strong>，能够极大的降低噪音。</p>
<p>然后，我们就可以根据AD计算SNP-index，即 ALT_COUNT &#x2F; (REF_COUNT + ALT_COUNT),  在0-1之间。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">freq <span class="operator">&lt;-</span> calcFreqFromAd<span class="punctuation">(</span>ad_flt<span class="punctuation">,</span> min.depth <span class="operator">=</span> <span class="number">10</span><span class="punctuation">,</span> max.depth <span class="operator">=</span> <span class="number">100</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这里设了一个最大和最小深度用于计算频率，太大的深度可能是同源基因或者是重复序列，太低的深度在计算的时候不太准确。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">freq2 <span class="operator">&lt;-</span> freq<span class="punctuation">[</span>Matrix<span class="operator">::</span>rowSums<span class="punctuation">(</span><span class="built_in">is.na</span><span class="punctuation">(</span>freq<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="number">0</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<p>接着我们就可以尝试去重现文章的结果了</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">par<span class="punctuation">(</span>mfrow <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span> formatC<span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">12</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> flag<span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  freq_flt <span class="operator">&lt;-</span> freq2<span class="punctuation">[</span>grepl<span class="punctuation">(</span>i<span class="punctuation">,</span>row.names<span class="punctuation">(</span>freq2<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">  pos <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>substring<span class="punctuation">(</span>row.names<span class="punctuation">(</span>freq_flt<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  plot<span class="punctuation">(</span>pos<span class="punctuation">,</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">]</span> <span class="operator">-</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span> ylim <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">       pch <span class="operator">=</span> <span class="number">20</span><span class="punctuation">,</span> cex <span class="operator">=</span> <span class="number">0.2</span><span class="punctuation">,</span></span><br><span class="line">       xlab <span class="operator">=</span> i<span class="punctuation">,</span></span><br><span class="line">       ylab <span class="operator">=</span> <span class="built_in">expression</span><span class="punctuation">(</span>paste<span class="punctuation">(</span>Delta<span class="punctuation">,</span> <span class="string">&quot; &quot;</span> <span class="punctuation">,</span><span class="string">&quot;SNP index&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/snp-index-1-9357d7610a5540ce980fe807c647fb14.png" alt="snp-index-1"></p>
<p>从图中，我惊讶的发现一些染色体上的部分区域居然都没有标记了，所以我们选择KY131是”0&#x2F;0”,  而 “DN422” 是 “1&#x2F;1”的位点进行分析</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">mask <span class="operator">&lt;-</span> which<span class="punctuation">(</span>gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327815&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;0/0&quot;</span> <span class="operator">&amp;</span>  gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327816&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;1/1&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ad_flt <span class="operator">&lt;-</span> ad<span class="punctuation">[</span>mask<span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;SRR6327817&quot;</span><span class="punctuation">,</span> <span class="string">&quot;SRR6327818&quot;</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">colnames<span class="punctuation">(</span>ad_flt<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;T_Pool&quot;</span><span class="punctuation">,</span> <span class="string">&quot;S_Pool&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">freq <span class="operator">&lt;-</span> calcFreqFromAd<span class="punctuation">(</span>ad_flt<span class="punctuation">,</span> min.depth <span class="operator">=</span> <span class="number">10</span><span class="punctuation">,</span> max.depth <span class="operator">=</span> <span class="number">100</span><span class="punctuation">)</span></span><br><span class="line">freq2 <span class="operator">&lt;-</span> freq<span class="punctuation">[</span>Matrix<span class="operator">::</span>rowSums<span class="punctuation">(</span><span class="built_in">is.na</span><span class="punctuation">(</span>freq<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="number">0</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">par<span class="punctuation">(</span>mfrow <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span> formatC<span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">12</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> flag<span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  freq_flt <span class="operator">&lt;-</span> freq2<span class="punctuation">[</span>grepl<span class="punctuation">(</span>i<span class="punctuation">,</span>row.names<span class="punctuation">(</span>freq2<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">  pos <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>substring<span class="punctuation">(</span>row.names<span class="punctuation">(</span>freq_flt<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  plot<span class="punctuation">(</span>pos<span class="punctuation">,</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span> <span class="operator">-</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">,</span> ylim <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">       pch <span class="operator">=</span> <span class="number">20</span><span class="punctuation">,</span> cex <span class="operator">=</span> <span class="number">0.2</span><span class="punctuation">,</span></span><br><span class="line">       xlab <span class="operator">=</span> i<span class="punctuation">,</span></span><br><span class="line">       ylab <span class="operator">=</span> <span class="built_in">expression</span><span class="punctuation">(</span>paste<span class="punctuation">(</span>Delta<span class="punctuation">,</span> <span class="string">&quot; &quot;</span> <span class="punctuation">,</span><span class="string">&quot;SNP index&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/snp-index-2-9496feba458a4b5682385962586f07c5.png" alt="snp index-2"></p>
<p>从结果中，我们能够发现一个有趣的结果，不同的筛选标记标准会导致标记在染色体上分布发生变化，这可能意味着在某种的筛选标准下，会使得和性状有关的位点明显的被过滤掉。</p>
<p>根据文章里的结果，候选基因落在6号染色体的20-25M中，也就是选择KY131是”0&#x2F;0”,  而 “DN422” 是 “1&#x2F;1”的位点结果和原文比较类似。那么如果我们原本不知道这个结果应该怎么办？这其实也不是什么问题，像我这样把两幅图都做出来，然后和实验设计者交流下，也就差不多知道答案了。</p>
<p>除了SNP-index外，文章还有一个ED(Euclidean distance)方法用于定位，我根据文章的公式和自己的理解写了下代码</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">mask <span class="operator">&lt;-</span> which<span class="punctuation">(</span>gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327815&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;0/0&quot;</span> <span class="operator">&amp;</span>  gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327816&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;1/1&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ad_flt <span class="operator">&lt;-</span> ad<span class="punctuation">[</span>mask<span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;SRR6327817&quot;</span><span class="punctuation">,</span> <span class="string">&quot;SRR6327818&quot;</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">ED_list <span class="operator">&lt;-</span> apply<span class="punctuation">(</span>ad_flt<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  count <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>unlist<span class="punctuation">(</span>strsplit<span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="string">&quot;,&quot;</span><span class="punctuation">,</span>fixed <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span>useBytes <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  depth1 <span class="operator">&lt;-</span> count<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span> <span class="operator">+</span> count<span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span></span><br><span class="line">  depth2 <span class="operator">&lt;-</span> count<span class="punctuation">[</span><span class="number">3</span><span class="punctuation">]</span> <span class="operator">+</span> count<span class="punctuation">[</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">  </span><br><span class="line">  ED <span class="operator">&lt;-</span> <span class="built_in">sqrt</span><span class="punctuation">(</span><span class="punctuation">(</span>count<span class="punctuation">[</span><span class="number">3</span><span class="punctuation">]</span> <span class="operator">/</span> depth2 <span class="operator">-</span> count<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span> <span class="operator">/</span> depth1<span class="punctuation">)</span><span class="operator">^</span><span class="number">2</span> <span class="operator">+</span> </span><br><span class="line">               <span class="punctuation">(</span>count<span class="punctuation">[</span><span class="number">4</span><span class="punctuation">]</span> <span class="operator">/</span> depth2<span class="operator">-</span> count<span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span> <span class="operator">/</span>depth1<span class="punctuation">)</span><span class="operator">^</span><span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>ED<span class="operator">^</span><span class="number">5</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">par<span class="punctuation">(</span>mfrow <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span> formatC<span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">12</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> flag<span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  ED_flt <span class="operator">&lt;-</span> ED_list<span class="punctuation">[</span>grepl<span class="punctuation">(</span>i<span class="punctuation">,</span><span class="built_in">names</span><span class="punctuation">(</span>ED_list<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">  pos <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>substring<span class="punctuation">(</span><span class="built_in">names</span><span class="punctuation">(</span>ED_flt<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  plot<span class="punctuation">(</span>pos<span class="punctuation">,</span> ED_flt<span class="punctuation">,</span></span><br><span class="line">       pch <span class="operator">=</span> <span class="number">20</span><span class="punctuation">,</span> cex <span class="operator">=</span> <span class="number">0.2</span><span class="punctuation">,</span></span><br><span class="line">       xlab <span class="operator">=</span> i<span class="punctuation">,</span></span><br><span class="line">       ylab <span class="operator">=</span> <span class="string">&quot;ED&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/ED-b396b941cc0b4b95ae7850935ce92f40.png" alt="ED"></p>
<p>我复现的结果和文章的区别在于，少一个拟合线，也就是以1Mb为区间，每次滑动10Kb计算均值。以KY131是”0&#x2F;0”,  而 “DN422” 是 “1&#x2F;1”的位点结果为例，我们来增加这条线</p>
<p>我写了一个函数用于根据窗口来计算均值，输入是之前snp-index的位置和对应的值，以及确定窗口的大小和步长，</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line">calcValueByWindow <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>pos<span class="punctuation">,</span> value<span class="punctuation">,</span></span><br><span class="line">                              window_size <span class="operator">=</span> <span class="number">1000000</span><span class="punctuation">,</span></span><br><span class="line">                              step_size <span class="operator">=</span> <span class="number">100000</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  <span class="comment"># get the max position in the postion</span></span><br><span class="line">  max_pos <span class="operator">&lt;-</span> <span class="built_in">max</span><span class="punctuation">(</span>pos<span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># construct the window</span></span><br><span class="line">  window_start <span class="operator">&lt;-</span> seq<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> max_pos <span class="operator">+</span> window_size<span class="punctuation">,</span>step_size<span class="punctuation">)</span></span><br><span class="line">  window_end <span class="operator">&lt;-</span> window_start <span class="operator">+</span> step_size</span><br><span class="line">  mean_value <span class="operator">&lt;-</span> vector<span class="punctuation">(</span>mode <span class="operator">=</span> <span class="string">&quot;numeric&quot;</span><span class="punctuation">,</span> <span class="built_in">length</span> <span class="operator">=</span> <span class="built_in">length</span><span class="punctuation">(</span>window_start<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># select the value inside the window</span></span><br><span class="line">  <span class="keyword">for</span> <span class="punctuation">(</span>j <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>window_start<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    </span><br><span class="line">    pos_in_window <span class="operator">&lt;-</span> which<span class="punctuation">(</span>pos <span class="operator">&gt;</span> window_start<span class="punctuation">[</span>j<span class="punctuation">]</span> <span class="operator">&amp;</span></span><br><span class="line">                             pos <span class="operator">&lt;</span> window_end<span class="punctuation">[</span>j<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">    value_in_window <span class="operator">&lt;-</span> value<span class="punctuation">[</span>pos_in_window<span class="punctuation">]</span></span><br><span class="line">    </span><br><span class="line">    mean_value<span class="punctuation">[</span>j<span class="punctuation">]</span> <span class="operator">&lt;-</span> mean<span class="punctuation">(</span>value_in_window<span class="punctuation">)</span></span><br><span class="line">    </span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  <span class="comment"># remove the Not A Number position</span></span><br><span class="line">  nan_pos <span class="operator">&lt;-</span>  <span class="built_in">is.nan</span><span class="punctuation">(</span>mean_value<span class="punctuation">)</span></span><br><span class="line">  mean_value <span class="operator">&lt;-</span> mean_value<span class="punctuation">[</span><span class="operator">!</span> nan_pos<span class="punctuation">]</span></span><br><span class="line">  window_pos <span class="operator">&lt;-</span> <span class="punctuation">(</span><span class="punctuation">(</span>window_start <span class="operator">+</span> window_end<span class="punctuation">)</span><span class="operator">/</span> <span class="number">2</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="operator">!</span>nan_pos<span class="punctuation">]</span></span><br><span class="line">  df <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>pos   <span class="operator">=</span> window_pos<span class="punctuation">,</span></span><br><span class="line">                   value <span class="operator">=</span> mean_value<span class="punctuation">)</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>df<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>得到结果就可以用用<code>lines</code>在之前结果上加上均值线</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">par<span class="punctuation">(</span>mfrow <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span> formatC<span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">12</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> flag<span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  </span><br><span class="line">  freq_flt <span class="operator">&lt;-</span> freq2<span class="punctuation">[</span>grepl<span class="punctuation">(</span>i<span class="punctuation">,</span>row.names<span class="punctuation">(</span>freq2<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">  pos <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>substring<span class="punctuation">(</span>row.names<span class="punctuation">(</span>freq_flt<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  snp_index <span class="operator">&lt;-</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span> <span class="operator">-</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">]</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># bin</span></span><br><span class="line">  df <span class="operator">&lt;-</span> calcValueByWindow<span class="punctuation">(</span>pos <span class="operator">=</span> pos<span class="punctuation">,</span> value <span class="operator">=</span> snp_index<span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  plot<span class="punctuation">(</span>x <span class="operator">=</span> pos<span class="punctuation">,</span> y <span class="operator">=</span>snp_index<span class="punctuation">,</span> </span><br><span class="line">       ylim <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">       pch <span class="operator">=</span> <span class="number">20</span><span class="punctuation">,</span> cex <span class="operator">=</span> <span class="number">0.2</span><span class="punctuation">,</span></span><br><span class="line">       xlab <span class="operator">=</span> i<span class="punctuation">,</span></span><br><span class="line">       ylab <span class="operator">=</span> <span class="built_in">expression</span><span class="punctuation">(</span>paste<span class="punctuation">(</span>Delta<span class="punctuation">,</span> <span class="string">&quot; &quot;</span> <span class="punctuation">,</span><span class="string">&quot;SNP index&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  lines<span class="punctuation">(</span>x <span class="operator">=</span> df<span class="operator">$</span>pos<span class="punctuation">,</span> y <span class="operator">=</span> df<span class="operator">$</span>value<span class="punctuation">,</span> col <span class="operator">=</span> <span class="string">&quot;red&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>最后再对文章总结一下。文章并不是只用了BSA的方法进行定位，他们花了几年的时间用SSR分子标记确定了候选基因可能区间，用BSA的方法在原有基础上缩小了定位区间。当然即便如此，候选基因也有上百个，作者通过BLAST的方式，对这些基因进行了注释。尽管中间还加了一些GO富集分析的内容，说这些基因富集在某个词条里，有一个是DNA metabolic processes(GO:0006259)，但我觉得如果作者用clusterProfiler做富集分析，它肯定无法得到任何富集结果。他做富集分析的目的是其实下面这个描述，也就是找到和抗冻相关的基因</p>
<blockquote>
<p>LOC_Os06g39740 and LOC_Os06g39750,were annotated as the function of “response to cold (GO: 0009409)”, suggesting their key roles in regulating cold tolerance in rice. “</p>
</blockquote>
<p>当然他还做了qRT-PCR进行了验证，最后推测LOC_Os06g39750应该是目标基因，这个基因里还有8个SNP位点。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/04/%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8BSA%E6%96%B9%E6%B3%95%E8%BF%9B%E8%A1%8C%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D%EF%BC%88%E6%B0%B4%E7%A8%BB%E7%AF%87%EF%BC%89-QTL-analysis-using-BSA-seq-in-rice/" data-id="clm1ywdl700giecokc6ks5drx" data-title="如何使用BSA方法进行遗传定位（水稻篇）" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用SHOREmap做mapping-by-sequencing-Mapping-by-sequencing-Using-SHOREmap" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/03/%E4%BD%BF%E7%94%A8SHOREmap%E5%81%9Amapping-by-sequencing-Mapping-by-sequencing-Using-SHOREmap/" class="article-date">
  <time class="dt-published" datetime="2019-10-03T16:43:58.610Z" itemprop="datePublished">2019-10-03</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/03/%E4%BD%BF%E7%94%A8SHOREmap%E5%81%9Amapping-by-sequencing-Mapping-by-sequencing-Using-SHOREmap/">使用SHOREmap做mapping-by-sequencing</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="简介"><a href="#简介" class="headerlink" title="简介"></a>简介</h1><p>SHOREmap可以用来分析传统作图群体（自然系natural strains和分化系,diverged accession杂交，或outcrossing)或近等作图群体（isogenic mapping population, 诱变后代与未诱变亲本进行杂交，即会交,backcrossing)所产生的重测序数据。根据作图群体构建方式的不同，SHOREmap的outcross或backcross采用不同基于滑窗(sliding)方式对等位基因频率进行分析。</p>
<p>SHOREmap的backcross和outcross都需要从突变重组库中获得的一致的碱基识别信息</p>
<h1 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h1><h2 id="前置安装"><a href="#前置安装" class="headerlink" title="前置安装"></a>前置安装</h2><p>SHOREmap需要<a target="_blank" rel="noopener" href="http://www.mps.mpg.de/dislin/">DISLIN</a>科学库进行数据可视化</p>
<p>但是在安装DISLIN之前还需要保证存在<code>/usr/lib/libXm.so*</code>和<code>/usr/lib/libXm.so*</code>，这两者的安全需要root权限，所以要么联系管理员，要么想办法绕开(这个办法，我还没有想到).</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sudo apt-get update</span><br><span class="line">sudo apt-get install libmotif4</span><br><span class="line">sudo apt-get install libxt-dev</span><br></pre></td></tr></table></figure>

<p>开始安装doslin库</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">cd /path/to/src</span><br><span class="line"># 下载</span><br><span class="line">wget ftp://ftp.gwdg.de/pub/grafik/dislin/linux/i586_64/dislin-11.0.linux.i586_64.tar.gz</span><br><span class="line"># 解压缩</span><br><span class="line">tar -zxvf dislin-11.0.linux.i586_64.tar.gz</span><br><span class="line">cd dislin-11.0</span><br><span class="line"># 加入系统路径</span><br><span class="line">mkdir -p $HOME/biosoft/dislin</span><br><span class="line">DISLIN=$HOME/biosoft/dislin</span><br><span class="line">export DISLIN</span><br><span class="line"># 安装</span><br><span class="line">./INSTALL</span><br><span class="line"># 复制dislin_d.h 到dislin的文件下</span><br><span class="line">cp ./example/dislin_d.h $DISLIN</span><br><span class="line"># 删除安装文件(可选)</span><br><span class="line">rm -rf dislin-11.0</span><br></pre></td></tr></table></figure>

<h2 id="安装SHOREmap-v3-x"><a href="#安装SHOREmap-v3-x" class="headerlink" title="安装SHOREmap v3.x"></a>安装SHOREmap v3.x</h2><p>我这次安装的是当前最新的3.4版本，其他版本估计换汤不换药。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">cd $HOME/biosoft</span><br><span class="line">wget http://bioinfo.mpipz.mpg.de/shoremap/SHOREmap_v3.4.tar.gz</span><br><span class="line"># 替换SHOREmap下的dislin的一些文件</span><br><span class="line">tar -zxvf SHOREmap_v3,4</span><br><span class="line">rm dislin/*dislin_d.*</span><br><span class="line">cp $DISLIN/*dislin_d.* dislin</span><br><span class="line"># 编辑/etc/profile或.bashrc</span><br><span class="line">vi .bashrc</span><br><span class="line">export LD_LIBRARY_PATH=$HOME/src/SHOREmap_v3.4/dislin</span><br><span class="line"># 退出保存.bashrc: Esc+:wq</span><br><span class="line">source .bashrc</span><br><span class="line"># 到之前安装的文件夹下</span><br><span class="line">cd &amp; cd src/SHOREmap_v3.4</span><br><span class="line">(可选，如果没有g++)sudo apt-get install build-essential</span><br><span class="line">make</span><br><span class="line"># 将编译文件拷贝到习惯的文件夹中，然后添加执行路径</span><br><span class="line">cp SHOREmap ../../biosoft/SHOREmap_v3.4</span><br><span class="line">echo &quot;export $HOME/bisoft/SHOREmap_v3.4&quot; &gt;&gt; ~/.bashrc</span><br></pre></td></tr></table></figure>

<p>最后，可以重新启动一下bash验证</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/shoremap-6bbbdce394e14f3fa6072d2938bf6ed8.png" alt="shoremap"></p>
<h2 id="官方网站提供的两个常见问题的解答"><a href="#官方网站提供的两个常见问题的解答" class="headerlink" title="官方网站提供的两个常见问题的解答"></a>官方网站提供的两个常见问题的解答</h2><p>Note 1: if the compilation complains like “&#x2F;usr&#x2F;bin&#x2F;ld: cannot find -lXt” (or “&#x2F;usr&#x2F;bin&#x2F;ld: cannot find -ldislin_d”), please open the makefile with the command</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">vi makefile</span><br></pre></td></tr></table></figure>
<p>Press keys ‘esc’ and ‘i’ on the keyboard to edit makefile; move the cursor with arrow keys to the position before -lXt, and edit -L&#x2F;path&#x2F;to&#x2F;libXt.so&#x2F;; if ‘-ldislin_d’ is not found, edit -L&#x2F;path&#x2F;to&#x2F;dislin_d&#x2F; before -ldislin_d). After that, press keys ‘esc’, type in <strong>:wq</strong>, and press enter to save editing and quit vi. (‘-L’ tells the linker where to find the library given by -l)</p>
<p>Note 2: if ‘&#x2F;usr&#x2F;lib&#x2F;ld: warning: libXm.so.3, needed by .&#x2F;dislin&#x2F;libdislin_d.so, not found (try using -rpath or -rpath-link)’ occurs, and you have installed libmotif4, do the following:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cp /usr/lib/libXm.so.4 /usr/lib/libXm.so.3</span><br></pre></td></tr></table></figure>
<p>We can make SHOREmap avaiable for general use by inserting the following command into &#x2F;etc&#x2F;profile</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">export PATH=$PATH:/path/to/SHOREmap_v3.x</span><br></pre></td></tr></table></figure>
<p>and</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">source /etc/profile</span><br></pre></td></tr></table></figure>

<h1 id="总体流程"><a href="#总体流程" class="headerlink" title="总体流程"></a>总体流程</h1><h2 id="OUTCROSS"><a href="#OUTCROSS" class="headerlink" title="OUTCROSS"></a>OUTCROSS</h2><table>
<thead>
<tr>
<th>outcross的基本步骤</th>
<th>描述</th>
</tr>
</thead>
<tbody><tr>
<td>SHOREmap extract</td>
<td>提取与SNP突变相关的重测序的一致的识别</td>
</tr>
<tr>
<td>SHOREmap create</td>
<td>根据背景&#x2F;亲本系的重测序质量创建SNP标记列表</td>
</tr>
<tr>
<td>SHOREmap outcross</td>
<td>进行等位基因频率分析并定义mapping interval（也就是找到突变所在的大致区域)</td>
</tr>
<tr>
<td>SHOREmap annotate</td>
<td>对mapping interval中的突变基因效应进行注释</td>
</tr>
</tbody></table>
<h2 id="BACKCROSS"><a href="#BACKCROSS" class="headerlink" title="BACKCROSS"></a>BACKCROSS</h2><table>
<thead>
<tr>
<th>backcross的基本步骤</th>
<th>描述</th>
</tr>
</thead>
<tbody><tr>
<td>SHOREmap extract</td>
<td>提取与SNP突变相关的重测序的一致的识别</td>
</tr>
<tr>
<td>SHOREmap backcross</td>
<td>进行等位基因分析</td>
</tr>
<tr>
<td>SHOREmap annotate</td>
<td>对mapping interval中的突变基因效应进行注</td>
</tr>
</tbody></table>
<h1 id="具体步骤"><a href="#具体步骤" class="headerlink" title="具体步骤"></a>具体步骤</h1><h2 id="下载数据"><a href="#下载数据" class="headerlink" title="下载数据"></a>下载数据</h2><p>只安装软件，却没有数据，我们也只能干瞪眼。</p>
<h3 id="oucross分析所需数据"><a href="#oucross分析所需数据" class="headerlink" title="oucross分析所需数据"></a>oucross分析所需数据</h3><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># OCF2 </span><br><span class="line">wget -4 -q http://bioinfo.mpipz.mpg.de/shoremap/data/software/OC.fg.reads1.fq.gz &amp;</span><br><span class="line">wget -4 -q http://bioinfo.mpipz.mpg.de/shoremap/data/software/OC.fg.reads2.fq.gz &amp;</span><br><span class="line"># Ler </span><br><span class="line">wget -4 -q http://bioinfo.mpipz.mpg.de/shoremap/data/software/OC.bg.reads1.fq.gz &amp;</span><br><span class="line">wget -4 -qh ttp://bioinfo.mpipz.mpg.de/shoremap/data/software/OC.bg.reads2.fq.gz &amp;</span><br></pre></td></tr></table></figure>
<h3 id="backcross分析所需数据"><a href="#backcross分析所需数据" class="headerlink" title="backcross分析所需数据"></a>backcross分析所需数据</h3><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># BCF2</span><br><span class="line">wget -4 -q http://bioinfo.mpipz.mpg.de/shoremap/data/software/BC.fg.reads1.fq.gz &amp;</span><br><span class="line">wget -4 -q http://bioinfo.mpipz.mpg.de/shoremap/data/software/BC.fg.reads2.fq.gz &amp;</span><br><span class="line"># mir159a (Col)</span><br><span class="line">wget -4 -q http://bioinfo.mpipz.mpg.de/shoremap/data/software/BC.bg.reads1.fq.gz</span><br><span class="line">wget -4 -q http://bioinfo.mpipz.mpg.de/shoremap/data/software/BC.bg.reads2.fq.gz</span><br></pre></td></tr></table></figure>

<h3 id="其他数据"><a href="#其他数据" class="headerlink" title="其他数据"></a>其他数据</h3><p>除了最基本的测序数据外，我们可能还需要参考基因组，已有的注释数据等</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># 参考基因组</span><br><span class="line">wget -4 -q http://bioinfo.mpipz.mpg.de/shoremap/data/software/TAIR10_chr_all.fas &amp;</span><br><span class="line"># 基因注释</span><br><span class="line">wget -4 -q http://bioinfo.mpipz.mpg.de/shoremap/data/software/TAIR10_GFF3_genes.gff &amp;</span><br><span class="line"># SHORE操作的结果数据</span><br><span class="line">wget -4 -q http://bioinfo.mpipz.mpg.de/shoremap/data/software/scoring_matrix_het.txt &amp;</span><br></pre></td></tr></table></figure>

<h2 id="重测序"><a href="#重测序" class="headerlink" title="重测序"></a>重测序</h2><p>首先使用<code>bwa</code>,<code>bowtie2</code>等read比对工具将得到的数据比对到参考基因组上。<br>假设你当前处在MBS文件夹下，该文件下有如下文件</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"># 混池测序结果，双端</span><br><span class="line">BC.fg.reads1.fq.gz</span><br><span class="line">BC.fg.reads2.fq.gz</span><br><span class="line"># 背景信息测序结果</span><br><span class="line">BC.bg.reads1.fq.gz</span><br><span class="line">BC.bg.reads2.fa.gz</span><br><span class="line"># 拟南芥参考基因组</span><br><span class="line">TAIR10_chr_all.fas</span><br><span class="line"># 拟南芥注释信息</span><br><span class="line">TAIR10_GFF3_genes.gff</span><br></pre></td></tr></table></figure>
<p>以下操作都是基于上述文件进行。</p>
<h3 id="第一步-序列比对，产生SAM文件"><a href="#第一步-序列比对，产生SAM文件" class="headerlink" title="第一步: 序列比对，产生SAM文件"></a>第一步: 序列比对，产生SAM文件</h3><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">mkdir index</span><br><span class="line"># 创建比对所需索引</span><br><span class="line">bowtie2-build TAIR10_chr_all.fas index/TAIR10</span><br><span class="line"># 序列比对</span><br><span class="line">bowtie2 -x index/TAIR10 -1 BC.fg.reads1.fq.gz -2 BC.fg.reads2.fq.gz -S FG.sam</span><br><span class="line">bowtie2 -x index/TAIR10 -1 BC.bg.reads1.fq.gz -2 BC.bg.reads2.fq.gz -S BG.sam</span><br></pre></td></tr></table></figure>
<h3 id="第二步：-SAMtools预测突变位点"><a href="#第二步：-SAMtools预测突变位点" class="headerlink" title="第二步： SAMtools预测突变位点"></a>第二步： SAMtools预测突变位点</h3><p>为了加快运算速度，可以先转换格式，并排序</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">samtools view -b -o BG.bam BG.sam</span><br><span class="line">samtools view -b -o FG.bam FG.sam</span><br><span class="line">samtools sort -o BG.sorted.bam BG.bam </span><br><span class="line">samtools sort -o FG.sorted.bam FG.bam </span><br><span class="line">samtools index BG.sorted.bam</span><br><span class="line">samtools index FG.sorted.bam</span><br></pre></td></tr></table></figure>
<p>consensus-calling program 寻找可能的变异位点<br> 由于官方说明中使用samtools版本为0.1.19，先要解释一下参数 mpileup -uD -f，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"># -f:faidx indexed reference sequence file 前后版本一致</span><br><span class="line"># -u:generate uncompress BCF output 前后版本一致</span><br><span class="line"># -D:output per-sample DP in BCF (require -g/-u).与输出格式有关，目前改为-t</span><br></pre></td></tr></table></figure>
<p>因此，对于1.4版本的samtools，相应的参数为</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mpileup -u -t DP -f </span><br></pre></td></tr></table></figure>

<p>官方说明的bcftools也是0.1.19，参数为bcftools view -vcg 旧版本的view在当前的版本用于过滤，功能被call替代</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"># -v Output variant sites only (force -c)</span><br><span class="line"># -c属于Call variants using Bayesian inference. This option automatically invokes option -e.When -v is in use, skip loci where the fraction of samples covered by reads is below FLOAT. [0],目前被m取代</span><br><span class="line"># -g Call per-sample genotypes at variant sites (force -c)，这个没有找到合适的替代参数</span><br></pre></td></tr></table></figure>

<p>综上，推荐使用如下命令行</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">samtools mpileup -u -t DP -f ../../../index/TAIR10_chr_all.fa ../../align/bwa/default/BG.sorted.bam | bcftools call -vm -Ov &gt; BG.vcf &amp;</span><br><span class="line">samtools mpileup -u -t DP -f ../../../index/TAIR10_chr_all.fa ../../align/bwa/default/FG.sorted.bam | bcftools call -vm -Ov &gt; FG.vcf &amp;</span><br></pre></td></tr></table></figure>
<h3 id="额外步骤：VCF格式转换"><a href="#额外步骤：VCF格式转换" class="headerlink" title="额外步骤：VCF格式转换"></a>额外步骤：VCF格式转换</h3><p>由于vcftools工具版本，所以最后的文件版本是4.2，而SHOREmap要求4.1。通过biostar找到高人写的降级工具（其实就是把一些字符替换一下，但是不了解vcf不同版本的差异话，是不知道怎么写）<br>把下面的代码存为<code>vcf_dowgrade.sh</code></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"># If you are trying to view VCF 4.2 files in IGV - you may run into issues. This function might help you.</span><br><span class="line"># This script will:</span><br><span class="line"># 1. Rename the file as version 4.1</span><br><span class="line"># 2. Replace parentheses in the INFO lines (IGV doesn&#x27;t like these!)</span><br><span class="line"></span><br><span class="line">function vcf_downgrade() &#123;</span><br><span class="line">  outfile=$&#123;1/.bcf/&#125;</span><br><span class="line">  outfile=$&#123;outfile/.gz/&#125;</span><br><span class="line">  outfile=$&#123;outfile/.vcf/&#125;</span><br><span class="line">  bcftools view --max-alleles 2 -O v $1 | \</span><br><span class="line">  sed &quot;s/##fileformat=VCFv4.2/##fileformat=VCFv4.1/&quot; | \</span><br><span class="line">  sed &quot;s/(//&quot; | \</span><br><span class="line">  sed &quot;s/)//&quot; | \</span><br><span class="line">  sed &quot;s/,Version=\&quot;3\&quot;&gt;/&gt;/&quot; | \</span><br><span class="line">  bcftools view -O z &gt; $&#123;outfile&#125;.dg.vcf.gz</span><br><span class="line">  tabix $&#123;outfile&#125;.dg.vcf.gz</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>其实对于单个文件而言，可以直接用以下命令</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">infile=BG.vcf</span><br><span class="line">outfile=BG.vcf</span><br><span class="line">bcftools  view --max-alleles 2 -O v $&#123;infile&#125; | \</span><br><span class="line">sed &quot;s/##fileformat=VCFv4.2/##fileformat=VCFv4.1/&quot; | \</span><br><span class="line">  sed &quot;s/(//&quot; | \</span><br><span class="line">  sed &quot;s/)//&quot; | \</span><br><span class="line">  sed &quot;s/,Version=\&quot;3\&quot;&gt;/&gt;/&quot; | \</span><br><span class="line">  bcftools view -O z &gt; $&#123;outfile&#125;.dg.vcf.gz</span><br></pre></td></tr></table></figure>

<h2 id="使用SHOREmap寻找突变所在区"><a href="#使用SHOREmap寻找突变所在区" class="headerlink" title="使用SHOREmap寻找突变所在区"></a>使用SHOREmap寻找突变所在区</h2><p>第一步：需要把bcf文件通过SHOREmap convert转换成SHOREmap能认识的格式</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">SHOREmap convert --marker samtools.vcf --folder path/to/folder -runid int</span><br></pre></td></tr></table></figure>
<p>会生成三个文件3_converted_consen.txt, 3_converted_variant.txt and 3_converted_reference.txt.</p>
<p>第二步：提取候选分子标记的consensus information(mapping pool)</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">SHOREmap extract --chrsizes chromsize.txt --folder ../SHOREmap_analysis --marker 11_converted_variant.txt --consen 11_converted_consen.txt -verbose</span><br></pre></td></tr></table></figure>

<p>第三步：<br>然后使用SHOREmap backcross分析。SHOREmap backcross可用来分析回交作图群体所得到重组后代混池数据。相对于传统作图群体，只有诱变剂产生的突变会分离，也只有这些才会用于突变定位。</p>
<p>SHOREmap backcross会尝试过滤出所有参考基因组和测序池之间不同部分用于找到突变点特异部分。为了保证不是自然变异或者是测序错误，测序池选择的部分要多次出现在亲本或背景中。然后根据前景和&#x2F;或背景的(识别碱基,base calls,质量&#x2F;覆盖率&#x2F;等位基因)信息，确定是否把保留的SNP位点作为分子标记。在正确的筛选后（拟南芥大概有上百个标记），SHOREmap backcross就能在分析marker的AF后识别大致的峰。进一步对变异注释后，就能找到目标性状的候选基因了。</p>
<p>SHOREmap backcross所需的输入文件如下：</p>
<ul>
<li>染色体大小文件，--chrsizes。分为两行，一行是染色体位置，一行是染色体大小。scaffold同理</li>
<li>候选marker文件。也就是使用SHOREmap convert通过vcf生成的converted_variant.txt，每一列的含义如下。<br>Column    Description<br>1    Project name<br>2    Identity of chromosome<br>3    Position of the SNP-marker<br>4    Reference base<br>5    Alternative base (or mutant base)<br>6    Quality of the alternative base (ranging from 0 to 40)<br>7    Number of reads supporting the predicted base<br>8    Ratio of reads supporting the predicted base to total coverage</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">SHOREmap backcross --chrsizes chromsize.txt --marker ../convert/11_converted_variant.txt --consen extracted_consensus_0.txt --folder ../BC_analysis -plot-bc --marker-score 40 --marker-freq 0.0 --min-converage 10 --max-coverage 80 -bg ../convert/12_converted_variant.txt  --bg-cov 1 --bg-freq 0.0 --bg-score 1 -non-EMS --cluster 1 --marker-hit 1 -verbose</span><br></pre></td></tr></table></figure>

<p>第四步：对结果进行注释</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">SHOREmap annotate --chrsizes chromsize.txt --folder ../BC_analysis/ann --snp ../convert/11_converted_variant.txt --chrom 2 --start 1 --end 4000000 --genome ../../TAIR10_chr_all.fas --gff ../../TAIR10_GFF3_genes.gff</span><br></pre></td></tr></table></figure>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/03/%E4%BD%BF%E7%94%A8SHOREmap%E5%81%9Amapping-by-sequencing-Mapping-by-sequencing-Using-SHOREmap/" data-id="clm1ywdke00dtecok2etsfw7v" data-title="使用SHOREmap做mapping-by-sequencing" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何用binmapr进行遗传定位-build-binmap-with-binmapr" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/01/%E5%A6%82%E4%BD%95%E7%94%A8binmapr%E8%BF%9B%E8%A1%8C%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D-build-binmap-with-binmapr/" class="article-date">
  <time class="dt-published" datetime="2019-10-01T21:27:06.707Z" itemprop="datePublished">2019-10-01</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/01/%E5%A6%82%E4%BD%95%E7%94%A8binmapr%E8%BF%9B%E8%A1%8C%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D-build-binmap-with-binmapr/">如何用binmapr进行遗传定位</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <blockquote>
<p>binmapr目前代码正在重构，目前无法从CRAN进行下载，本篇教程仅为存档用。</p>
</blockquote>
<p>binmapr是我折腾的一个R包，它能够将NGS测序得到SNP数据基于binmap进行纠错，用于更好的遗传定位。</p>
<p>在阅读本文之前，请先花点时间看看<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/x6zRylSiPn0LmNtmEGbdAA">Bin, Bin, Bin！Map, Map, Map Now!</a>, 我只是将里面的步骤整理成R包方便调用而已。</p>
<p>首先你得安装并加载R包。因为这个R包目前主要是方便自己使用，所以托管在GitHub上，需要用devtools进行安装</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">devtools<span class="operator">::</span>install_github<span class="punctuation">(</span><span class="string">&quot;xuzhougeng/binmapr&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>之后就可以和其他R包一样正常使用</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>binmapr<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>R包的使用非常简单，就是调用<code>batchCallGeno</code> 将原本的genotype矩阵按照15bp对snp进行纠正</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">geno <span class="operator">&lt;-</span> batchCallGeno<span class="punctuation">(</span>GT_flt<span class="punctuation">,</span> CHROM <span class="operator">=</span> CHROM<span class="punctuation">,</span> </span><br><span class="line">                      outdir <span class="operator">=</span> <span class="string">&quot;.&quot;</span><span class="punctuation">,</span></span><br><span class="line">                      pos.start <span class="operator">=</span> <span class="number">7</span><span class="punctuation">,</span> fix.size <span class="operator">=</span> <span class="number">10</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>因此，你只要提供一个符合要求的输入即可。</p>
<p>以李广伟师兄的数据为例，我已经将其整理成示例数据，因此可以可以通过下面两行命令读取</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">data<span class="punctuation">(</span>geno<span class="punctuation">)</span></span><br><span class="line">data<span class="punctuation">(</span>pheno<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这两个都是数据框，其中geno存放的是基因型数据，而pheno存放的是表型数据</p>
<p>为了能够让batchCallGeno运行，我们需要将<code>geno</code>数据转成一个矩阵，其中行名是位置信息，列名是样本信息</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">GT <span class="operator">&lt;-</span> as.matrix<span class="punctuation">(</span>geno<span class="punctuation">[</span><span class="punctuation">,</span><span class="operator">-</span><span class="number">1</span><span class="operator">:</span><span class="operator">-</span><span class="number">4</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">row.names<span class="punctuation">(</span>GT<span class="punctuation">)</span> <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span>geno<span class="operator">$</span>CHR<span class="punctuation">,</span> <span class="string">&quot;_&quot;</span><span class="punctuation">,</span> geno<span class="operator">$</span>POS<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>由于每个位点都在所有样本中都不一定存在，因此可以考虑过滤一些缺失比较多的位点.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">miss_ratio <span class="operator">&lt;-</span> rowSums<span class="punctuation">(</span><span class="built_in">is.na</span><span class="punctuation">(</span>GT<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">/</span> ncol<span class="punctuation">(</span>GT<span class="punctuation">)</span></span><br><span class="line">GT_flt <span class="operator">&lt;-</span> GT<span class="punctuation">[</span>miss_ratio <span class="operator">&lt;</span> <span class="number">0.20</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<p>我这里过滤掉缺失大于20%的位点，原本是打算用5%，未曾想到这个标准过滤下去，90%的数据都快没了，吓得我赶紧用<code>summary(miss_ratio)</code>看了一波分布，改了下标准。</p>
<p>有了合适的数据类型后，就可以调用batchCallGeno函数了。运行结束后，除了得到一个列表外存放基因型外，还会在当前目录下输出csv和pdf文件。其中csv是表型数据，而pdf则是纠错前后的基因型分布。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">CHROM <span class="operator">&lt;-</span> unique<span class="punctuation">(</span>geno<span class="operator">$</span>CHR<span class="punctuation">)</span></span><br><span class="line">geno_out <span class="operator">&lt;-</span> batchCallGeno<span class="punctuation">(</span>GT_flt<span class="punctuation">,</span> CHROM <span class="operator">=</span> CHROM<span class="punctuation">,</span> </span><br><span class="line">                      pos.start <span class="operator">=</span> <span class="number">7</span><span class="punctuation">,</span> fix.size <span class="operator">=</span> <span class="number">10</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>介绍下几个参数</p>
<ul>
<li><code>CHROM</code>:  用于构建binmap的染色体</li>
<li><code>pos.start</code>， 该参数用于绘图时从行名中提取坐标，例如<code>Chr1_1245</code>需要设置为pos.start&#x3D;6, 这里命名为<code>chr01_12345</code>所以要设置为pos.start&#x3D;7</li>
<li><code>fix.size</code>, 它和纠错有关，比如说你的基因型是00000100000, 中间的1很可能是错误的，因为它只出现了一次。<code>fix.size</code>需要根据具体数据来调整。</li>
</ul>
<p>下一步，我们就可以用方差分析的方法来寻找和表型相关的区间</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Load phenotype</span></span><br><span class="line">anova_analysis <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">,</span> phenotype<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  </span><br><span class="line">  p.lmout <span class="operator">&lt;-</span> lm<span class="punctuation">(</span>phenotype <span class="operator">~</span> x<span class="punctuation">,</span> na.action <span class="operator">=</span> na.exclude<span class="punctuation">)</span></span><br><span class="line">  p <span class="operator">&lt;-</span> anova<span class="punctuation">(</span>p.lmout<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span><span class="number">5</span><span class="punctuation">]</span><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>p<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">geno_mt <span class="operator">&lt;-</span> geno_out<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">geno_mt_reorder <span class="operator">&lt;-</span> geno_mt<span class="punctuation">[</span><span class="punctuation">,</span>pheno<span class="operator">$</span>ID<span class="punctuation">]</span></span><br><span class="line">pvals <span class="operator">&lt;-</span> apply<span class="punctuation">(</span>geno_mt_reorder<span class="punctuation">,</span> </span><br><span class="line">               <span class="number">1</span><span class="punctuation">,</span> </span><br><span class="line">               anova_analysis<span class="punctuation">,</span> </span><br><span class="line">               phenotype <span class="operator">=</span> pheno<span class="operator">$</span>PH<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>接着画图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">pos <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>substring<span class="punctuation">(</span>row.names<span class="punctuation">(</span>geno_mt_reorder<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">plotQtl<span class="punctuation">(</span>pos <span class="operator">=</span> pos<span class="punctuation">,</span> </span><br><span class="line">        pvalue <span class="operator">=</span> pvals<span class="punctuation">,</span> </span><br><span class="line">        chr.name <span class="operator">=</span> <span class="string">&quot;chr_01&quot;</span><span class="punctuation">,</span></span><br><span class="line">        ymax <span class="operator">=</span> <span class="number">11</span><span class="punctuation">,</span></span><br><span class="line">        threshold <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/QTL-mapping-e8e24a5d5f2b460c9892908c7c1e38eb.png" alt="QTL mapping"></p>
<p>我们可以从中看到一个非常显著的峰，里面的基因中就有一个碰巧是水稻里的绿色革命基因，sd1 LOC_Os01g66100 物理区间是 38382382-38385504，距QTL最显著位置只有258kb的距离，对于一个只有172个个体的F2群体而言，结果是不是已经很不错了。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/01/%E5%A6%82%E4%BD%95%E7%94%A8binmapr%E8%BF%9B%E8%A1%8C%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D-build-binmap-with-binmapr/" data-id="clm1ywdlx00i8ecokegp5cvmw" data-title="如何用binmapr进行遗传定位" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何在服务器上安装最新的R-install-r-in-linux" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/09/23/%E5%A6%82%E4%BD%95%E5%9C%A8%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E5%AE%89%E8%A3%85%E6%9C%80%E6%96%B0%E7%9A%84R-install-r-in-linux/" class="article-date">
  <time class="dt-published" datetime="2019-09-23T11:52:13.832Z" itemprop="datePublished">2019-09-23</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/09/23/%E5%A6%82%E4%BD%95%E5%9C%A8%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E5%AE%89%E8%A3%85%E6%9C%80%E6%96%B0%E7%9A%84R-install-r-in-linux/">如何在服务器上安装最新的R</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>R语言在服务器上安装是一个比较可麻烦可简单的事情，这里记录下R语言在两个比较常见的Linux发行版的安装方法，分别是CentOS和Ubuntu。</p>
<h2 id="通用方法（无需Root）"><a href="#通用方法（无需Root）" class="headerlink" title="通用方法（无需Root）"></a>通用方法（无需Root）</h2><p>只要你的服务器能够安装conda，那么你就可以用conda去安装你的R语言。conda已经不再局限于最早的Python的环境管理了，而是扩展到R, Java, C&#x2F;C++等编程语言。</p>
<blockquote>
<p>Package, dependency and environment management for any language—Python, R, Ruby, Lua, Scala, Java, JavaScript, C&#x2F; C++, FORTRAN, and more.</p>
</blockquote>
<p>让我们使用<code>conda search r-base</code>在conda的频道中检索，</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">...</span><br><span class="line">r-base                         3.5.1   hfb2a302_1009  anaconda/cloud/conda-forge</span><br><span class="line">r-base                         3.5.1   hfb2a302_1010  anaconda/cloud/conda-forge</span><br><span class="line">r-base                         3.6.0      hce969dd_0  pkgs/r</span><br><span class="line">r-base                         3.6.1      h6e652e1_3  anaconda/cloud/conda-forge</span><br><span class="line">r-base                         3.6.1      h8900bf8_0  anaconda/cloud/conda-forge</span><br><span class="line">r-base                         3.6.1      h8900bf8_1  anaconda/cloud/conda-forge</span><br><span class="line">r-base                         3.6.1      h8900bf8_2  anaconda/cloud/conda-forge</span><br><span class="line">r-base                         3.6.1      hba50c9b_4  anaconda/cloud/conda-forge</span><br><span class="line">r-base                         3.6.1      hce969dd_0  pkgs/r</span><br></pre></td></tr></table></figure>

<p>发现能找到好几个版本的R语言，我推荐通过新建环境的方式安装不同版本的R语言，这样就能在不同环境间切换。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda create -n r-3.6.1 r-base=3.6.1</span><br></pre></td></tr></table></figure>

<p>之后用<code>conda activate r-3.6.1</code>调用R的环境即可。</p>
<p>这个方法的优点是不需要root权限，安装方便，不过听过在使用的时候或许会出现一些bug，我还没有遇到。</p>
<p>除了conda外，我们可以通过手工解决R语言的依赖环境，通过源码安装最新的R语言，这个方法也不依赖于平台。不过你需要看下这篇<a href="/archives/Compile-Software-in-Linux-Without-Root">无root权限解决编译时的依赖问题</a>, 但是很麻烦。还是建议用conda比较合适。</p>
<p>假如需要用Jupyter notebook调用R，那么安装方式为</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">conda install -c r r-irkernel </span><br><span class="line"><span class="comment"># or</span></span><br><span class="line">conda install -c r r-essentials</span><br></pre></td></tr></table></figure>

<p>因为 conda install -c r r&#x3D;3.6.x&#x2F;r-base 默认不会安装 irkernel，而且先安装的 r&#x3D;3.6.x&#x2F;r-base 可能与后安装的 r-irkernel&#x2F;r-essentials 产生冲突。</p>
<h2 id="CentOS"><a href="#CentOS" class="headerlink" title="CentOS"></a>CentOS</h2><p>CentOS&#x2F;RedHat是可以通过<code>sudo yum install R</code>的方式安装R语言，解决一切依赖问题，并且安装比较新的R版本</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">================================================================</span><br><span class="line"> Package      Arch       Version     Repository    Size</span><br><span class="line">================================================================</span><br><span class="line">Installing:</span><br><span class="line"> R            x86_64     3.6.0-1.el7  epel          30 k</span><br><span class="line">Installing <span class="keyword">for</span> dependencies:</span><br><span class="line"> R-core       x86_64     3.6.0-1.el7  epel          57 M</span><br><span class="line"> R-core-devel x86_64     3.6.0-1.el7  epel          109 k</span><br><span class="line"> R-devel      x86_64     3.6.0-1.el7  epel           30 k</span><br><span class="line"> R-java       x86_64     3.6.0-1.el7  epel           31 k</span><br><span class="line"> R-java-devel x86_64     3.6.0-1.el7  epel           30 K</span><br><span class="line"> ....</span><br><span class="line"> Transaction Summary</span><br><span class="line">=================================================================</span><br><span class="line">Install  1 Package  (+373 Dependent packages)</span><br><span class="line">Upgrade             (  14 Dependent packages)</span><br></pre></td></tr></table></figure>

<p>这个方法安装的是比较新的R，基本上所有新版本能装的R包，它也能装了。但是如果你需要安装最新的R，那么就需要从头编译</p>
<p>先运行<code>sudo yum install R</code>搞定大部分依赖问题，然后你可能还得手动解决几个依赖问题，我遇到的是X11和libcurl</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># X11</span></span><br><span class="line">yum install xorg-x11-server-devel libX11-devel libXt-devel</span><br><span class="line"><span class="comment"># libcurl</span></span><br><span class="line">yum install libcurl-devel</span><br></pre></td></tr></table></figure>

<p>之后下载源代码编译安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">wget https://mirrors.tuna.tsinghua.edu.cn/CRAN/src/base/R-3/R-3.6.1.tar.gz</span><br><span class="line">tar -zxvf R-3.6.1.tar.gz</span><br><span class="line"><span class="built_in">cd</span> R-3.6.1</span><br><span class="line"><span class="comment"># --enable-R-shlib for Rstudio server</span></span><br><span class="line"><span class="built_in">mkdir</span> -p /opt/sysoft</span><br><span class="line">./configure --enable-R-shlib --prefix=/opt/sysoft/R-3.6.1</span><br><span class="line">make -j 8</span><br><span class="line">make install </span><br></pre></td></tr></table></figure>

<p>其中<code>--enable-R-shlib</code>是Rsutdio-Server安装需要，而<code>--prefix</code>是指定安装路径</p>
<h2 id="Ubuntu"><a href="#Ubuntu" class="headerlink" title="Ubuntu"></a>Ubuntu</h2><p>在Ubuntu上直接通过<code>sudo apt-get install r-base</code>的方式安装的不是最新版本的R，而是R-3.4版本。</p>
<blockquote>
<p>R 3.4 packages for Ubuntu on i386 and amd64 are available for all stable Desktop releases of Ubuntu prior to Bionic Beaver (18.04) until their official end of life date.</p>
</blockquote>
<p>不过我们可以通过增加安装源的方式，使得能够通过<code>apt-get</code>的方式安装最新的R。</p>
<p>第一步，确认你的Ubuntu版本，是Xenial Xerus(16.04; LTS), Trusty Tahr (14.04; LTS), Bionic Beaver (18.04;LTS), Cosmic Cuttlefish (18.10), Disco Dingo (19.04)的哪一种。</p>
<p>第二步，根据你的服务器Ubuntu版本，按照需求复制下面的<strong>其中</strong>一行代码（一定要注意，是一行，不是全部复制）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">deb https://cloud.r-project.org/bin/linux/ubuntu disco-cran35/</span><br><span class="line">deb https://cloud.r-project.org/bin/linux/ubuntu cosmic-cran35/</span><br><span class="line">deb https://cloud.r-project.org/bin/linux/ubuntu bionic-cran35/</span><br><span class="line">deb https://cloud.r-project.org/bin/linux/ubuntu xenial-cran35/</span><br><span class="line">deb https://cloud.r-project.org/bin/linux/ubuntu trusty-cran35/</span><br></pre></td></tr></table></figure>

<p>然后用vim编译<code>/etc/apt/sources.list</code>, 添加你复制的内容到最后一行，我的服务器是 xenial，所以增加的是</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">deb https://cloud.r-project.org/bin/linux/ubuntu xenial-cran35/</span><br></pre></td></tr></table></figure>

<p>除了修改<code>/etc/apt/sources.list</code>，还需要增加APT</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys E298A3A825C0D65DFD57CBB651716619E084DAB9</span><br></pre></td></tr></table></figure>

<p>之后，用下面的命令就可以安装最新的R</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sudo apt-get update</span><br><span class="line">sudo apt-get install r-base</span><br><span class="line">sudo apt-get install r-base-dev</span><br></pre></td></tr></table></figure>

<p>这个方法稍微麻烦些，据说通过这样子安装的R存在一些bug，不过我没有遇到。</p>
<p>假如你需要安装不同版本的R语言，那么就需要下载源代码进行编译安装。根据我的经验，你至少先得用下面这些命令安装R的依赖环境（可能还不够）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 设置环境变量</span></span><br><span class="line"><span class="built_in">export</span> CFLAGS=<span class="string">&quot; -fPIC&quot;</span> CXXFLAGS=<span class="string">&quot; -fPIC&quot;</span></span><br><span class="line"><span class="comment">## build-essential</span></span><br><span class="line">sudo apt-get install -y build-essential</span><br><span class="line"><span class="comment">## java</span></span><br><span class="line">sudo apt install -y openjdk-9-jdk</span><br><span class="line"><span class="comment">## 各种包</span></span><br><span class="line">sudo apt install -y autoconf libz-dev libbz2-dev liblzma-dev libssl-dev </span><br><span class="line"><span class="comment"># solve  libcurl problem</span></span><br><span class="line"><span class="comment">#sudo apt install -y libcurl4-openssl-dev  # not works for Ubuntu 16.04</span></span><br><span class="line">sudo apt install -y libcurl4-gnutls-dev </span><br><span class="line"><span class="comment">### curses</span></span><br><span class="line">sudo apt-get install -y libncurses5-dev</span><br><span class="line"><span class="comment">### solve X11 problem</span></span><br><span class="line">sudo apt-get install -y xorg-dev</span><br><span class="line"><span class="comment">### zlib2</span></span><br><span class="line">wget http://zlib.net/zlib-1.2.11.tar.gz</span><br><span class="line">tar -zxvf zlib-1.2.11.tar.gz &amp;&amp; <span class="built_in">cd</span> zlib-1.2.11 &amp;&amp; ./configure &amp;&amp; make &amp;&amp; sudo make install &amp;&amp; <span class="built_in">cd</span> .. &amp;&amp; <span class="built_in">rm</span> -rf zlib-1.2.11</span><br><span class="line"><span class="comment">### bzip2</span></span><br><span class="line">wget https://fossies.org/linux/misc/bzip2-1.0.8.tar.gz</span><br><span class="line">tar -zxvf bzip2-1.0.8.tar.gz &amp;&amp; <span class="built_in">cd</span> bzip2-1.0.8  </span><br><span class="line"><span class="comment"># add -fPIC</span></span><br><span class="line">sed -i <span class="string">&#x27;s/CFLAGS=/CFLAGS=-fPIC /&#x27;</span> Makefile</span><br><span class="line">make &amp;&amp; sudo make install &amp;&amp; <span class="built_in">cd</span> .. &amp;&amp; <span class="built_in">rm</span> -rf  bzip2-1.0.8</span><br></pre></td></tr></table></figure>

<blockquote>
<p>假如你使用的是conda用户，那么安装之前，你需要用先退出conda环境，不然后续libcurl可能会一直提示出错</p>
</blockquote>
<p>下载R的源代码，进行编译安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">wget https://mirrors.tuna.tsinghua.edu.cn/CRAN/src/base/R-3/R-3.6.1.tar.gz</span><br><span class="line">tar -zxvf R-3.6.1.tar.gz</span><br><span class="line"><span class="built_in">cd</span> R-3.6.1</span><br><span class="line"><span class="comment"># --enable-R-shlib for Rstudio server</span></span><br><span class="line">./configure --enable-R-shlib --prefix=/opt/sysoft/R-3.6.1</span><br><span class="line">make -j 8</span><br><span class="line">make install </span><br></pre></td></tr></table></figure>

<h2 id="其他"><a href="#其他" class="headerlink" title="其他"></a>其他</h2><p>此外，我还推荐大家安装OpenSSL&#x2F;Curl&#x2F;XML&#x2F;HD5F，可以节约之后的一点时间</p>
<p>Fedora, CentOS, RHEL</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">yum install openssl-devel</span><br><span class="line">yum install libcurl-devel</span><br><span class="line">yum install libxml2-devel</span><br><span class="line">yum install hdf5-devel</span><br></pre></td></tr></table></figure>

<p>Debian, Ubuntu等</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">sudo apt install libssl-dev </span><br><span class="line">sudo apt install libcurl4-openssl-dev</span><br><span class="line">sudo apt install libxml2-dev</span><br><span class="line">sudo apt-get install libhdf5-dev</span><br></pre></td></tr></table></figure>


      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/09/23/%E5%A6%82%E4%BD%95%E5%9C%A8%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E5%AE%89%E8%A3%85%E6%9C%80%E6%96%B0%E7%9A%84R-install-r-in-linux/" data-id="clm1ywdlg00h6ecok55zhfge8" data-title="如何在服务器上安装最新的R" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-无root权限解决编译时的依赖问题-Compile-Software-in-Linux-Without-Root" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/09/23/%E6%97%A0root%E6%9D%83%E9%99%90%E8%A7%A3%E5%86%B3%E7%BC%96%E8%AF%91%E6%97%B6%E7%9A%84%E4%BE%9D%E8%B5%96%E9%97%AE%E9%A2%98-Compile-Software-in-Linux-Without-Root/" class="article-date">
  <time class="dt-published" datetime="2019-09-23T11:02:01.221Z" itemprop="datePublished">2019-09-23</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/09/23/%E6%97%A0root%E6%9D%83%E9%99%90%E8%A7%A3%E5%86%B3%E7%BC%96%E8%AF%91%E6%97%B6%E7%9A%84%E4%BE%9D%E8%B5%96%E9%97%AE%E9%A2%98-Compile-Software-in-Linux-Without-Root/">无root权限解决编译时的依赖问题</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>如果你拥有最高权限，如果你只管理一台服务器，那么系统自带的包管理工具就帮你解决了所有问题。但是真实世界没有那么美好，所以我花了那么久时间去学习如何从源码开始编译一个软件。</p>
<p><strong>环境</strong>为CentOS Linux release 7.4.1708 (Core), Linux内核version 3.10.0-693.el7.x86_64， GCC版本为4.8.5 20150623 (Red Hat 4.8.5-16) (GCC)，</p>
<h2 id="Linux的编译体系"><a href="#Linux的编译体系" class="headerlink" title="Linux的编译体系"></a>Linux的编译体系</h2><p>无管理员权限编译的常规三部曲是<code>./configure --prefix=$HOME/usr &amp;&amp; make &amp;&amp; make install</code>，其中最重要的一步就是<code>configure</code>，它所做的任务如下</p>
<ul>
<li>检查GCC版本以及是否安装了编译所需工具</li>
<li>如果需要头文件，则默认去<code>/usr/include</code>查找</li>
<li>如果涉及到动态编译库，则默认去<code>/usr/lib</code>和<code>/usr/lib64</code>查找. 注：<code>lib</code>的函数库仅用于开机时用,提供给&#x2F;bin和&#x2F;sbin.</li>
</ul>
<p>那为何需要配置？配置主要解决软件开发和软件实际安装时平台不同所导致的问题。</p>
<p>首先，一个C&#x2F;C++工程不可能只用到标准库，很多已有的轮子就不需要重复制造。其次，由于很多软件都重复用到相同的依赖库，那么如果把这些依赖库独立成单独的模块，在调用的时候加载，也能节省空间。早期为了适配多个平台，开发人员需要手写条件语句来检查环境依赖，后来GNU专门开发了Autotools辅助构建源码编译所需要的关键文件。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/9/2013053-36054cdebaf2ec9d-d308d285e8734e5eb72f778be7e33c98.png" alt="Autotools"></p>
<h2 id="编译环境变量"><a href="#编译环境变量" class="headerlink" title="编译环境变量"></a>编译环境变量</h2><p>用<code>./configure -h</code>查看帮助文档的时候，会提示几个和编译相关非常重要的环境变量。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_"># </span><span class="language-bash">编译器</span></span><br><span class="line">CC          指定C编译器(compiler command)路径</span><br><span class="line">CXX         指定C++编译器</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">编译器选项</span></span><br><span class="line">CFLAGS      用于C编译器的选项</span><br><span class="line">CXXFLAGS    用于C++编译器的选项</span><br><span class="line">LDFLAGS     链接相关选项,如果你有自定义的函数库(lib dir)，即可以用 -L&lt;lib dir&gt;指定</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">预编译器</span></span><br><span class="line">CXXCPP      C++ 预处理器(preprocessor)</span><br><span class="line">CPP         C 预处理器(preprocessor)</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">预编译器选项</span></span><br><span class="line">CPPFLAGS    C/C++预处理器选项, 如果你自定义的头文件，可以用-I&lt;include dir&gt;</span><br></pre></td></tr></table></figure>

<p>Makfile规则中的编译命令通常遵守如下规范：</p>
<p>1,首先从源代码生成目标文件( <strong>预处理,编译,汇编</strong> )，”-c”选项表示不执行链接步骤；</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$</span><span class="language-bash">(CC) $(CPPFLAGS) $(CFLAGS) example.c   -c   -o example.o</span></span><br></pre></td></tr></table></figure>

<p>2,然后将目标文件链接为最终的结果( <strong>链接</strong> )，”-o”选项用于指定输出文件的名字。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$</span><span class="language-bash">(CC) $(LDFLAGS) example.o   -o example</span></span><br></pre></td></tr></table></figure>

<p>这些只是约定俗成的惯例，所以有些人会“随性而为”，你也拿他没有办法。尽管将源代码编译为二进制文件的四个步骤由不同的程序(cpp,gcc&#x2F;g++,as,ld)完成，但是事实上 cpp, as, ld 都是由 gcc&#x2F;g++ 进行间接调用的。换句话说，<strong>控制了 gcc&#x2F;g++ 就等于控制了所有四个步骤</strong>。从 Makefile 规则中的编译命令可以看出，编译工具的行为全靠 <strong>CC&#x2F;CXX CPPFLAGS CFLAGS&#x2F;CXXFLAGS LDFLAGS</strong> 这几个变量在控制。所以控制这些变量最简单的做法是首先设置与这些 Makefile 变量同名的环境变量并将它们 export 为 <strong>环境变量</strong>（全局），然后运行 configure 脚本，大多数 configure 脚本会使用这同名的环境变量代替 Makefile 中的值</p>
<ul>
<li>CC&#x2F;CXX: 指定C&#x2F;C++编译所在路径，即可以选择不同的版本的编译器进行编译。</li>
<li>CPPFLAGS: 这是用于预处理阶段的选项。用于添加不在标准路径<code>/usr/include</code>下的头文件。如<code>CPPFLAGS=&quot;-I$HOME/usr/include -I$HOME/usr/include/ncurses&quot;</code></li>
<li>CFLAGS&#x2F;CXXFLAGS： 指定头文件（.h文件）的路径，如：<code>CFLAGS=-I/usr/include -I/path/include</code>。同样地，安装一个包时会在安装路径下建立一个include目录，当安装过程中出现问题时，试着把以前安装的包的include目录加入到该变量中来。</li>
</ul>
<blockquote>
<p>CPPFLAG和CFLAGS&#x2F;CXXFLAGS这两个变量可以认为是等价关系，都用在预处理阶段，也就是都能用于指定头文件所在位置。</p>
</blockquote>
<ul>
<li>LDFLAGS：gcc 等编译器会用到的一些优化参数，也可以在里面指定库文件的位置。用法：<code>LDFLAGS=-L/usr/lib -L/path/to/your/lib</code>。每安装一个包都几乎一定的会在安装目录里建立一个lib目录。如果明明安装了某个包，而安装另一个包时，它愣是说找不到，可以抒那个包的lib路径加入的LDFALGS中试一下。</li>
</ul>
<p>有时候LDFLAGS指定-L虽然能让链接器找到库进行链接，但是运行时链接器却找不到这个库，如果要让软件运行时库文件的路径也得到扩展，那么我们需要增加这两个库给”-Wl,R”：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">LDFLAGS = -L/var/xxx/lib -L/opt/mysql/lib -Wl,R/var/xxx/lib -Wl,R/opt/mysql/lib</span><br></pre></td></tr></table></figure>

<p>如在执行.&#x2F;configure以前设置环境变量 <code>export LDFLAGS=&quot;-L/var/xxx/lib -L/opt/mysql/lib -Wl,R/var/xxx/lib -Wl,R/opt/mysql/lib&quot;</code>，注意设置环境变量等号两边不可以有空格，而且要加上引号（shell的用法）。那么执行configure以后，Makefile将会设置这个选项，链接时会有这个参数，编译出来的可执行程序的库文件搜索路径就得到扩展了</p>
<p>除了通过以上几种环境变量为编译器提供头文件和静态和动态库文件的位置信息外，还有一种变量叫做 <strong>PKG_CONFIG_PATH</strong> , 它从<code>xx.pc</code>文件获取读取相应的环境环境。</p>
<p><strong>注意</strong>:Linux下编译共享库时，必须加上-fPIC参数，即<code>export CFLAGS=&quot; -fPIC&quot; CXXFLAGS=&quot; -fPIC&quot;</code>否则在链接时会有错误提示.这是在编译zsh时候发现明明装了ncurse却还是不能用的共享库的坑。</p>
<blockquote>
<p>fPIC的目的是什么？共享对象可能会被不同的进程加载到不同的位置上，如果共享对象中的指令使用了绝对地址、外部模块地址，那么在共享对象被加载时就必须根据相关模块的加载位置对这个地址做调整，也就是修改这些地址，让它在对应进程中能正确访问，而被修改到的段就不能实现多进程共享一份物理内存，它们在每个进程中都必须有一份物理内存的拷贝。fPIC指令就是为了让使用到同一个共享对象的多个进程能尽可能多的共享物理内存，它背后把那些涉及到绝对地址、外部模块地址访问的地方都抽离出来，保证代码段的内容可以多进程相同，实现共享。</p>
</blockquote>
<p><strong>动态库路径问题</strong>: 由前面可以知道许多大型软件为了减少体积不会完全编译所有功能，而是借助于动态连接技术在运行时加载封装好的动态连接库内的功能。这就涉及一个非常重要的问题，软件如何知道动态链接库所处的位置。动态库搜索路径分两种情况，一种是编译生成可执行文件时，另外一种是运行可执行文件时。</p>
<p>编译生成可执行文件时，动态库的搜索路径顺序如下：</p>
<ul>
<li>首先gcc会找-L选项；</li>
<li>然后再找gcc的环境变量<code>LIBRARY_PATH</code>，可以在.profile设置这个环境变量，并且可以通过选项-v查看gcc最终编译时LIBRARY_PATH的值；</li>
<li>再找内定目录: <code>/lib：/usr/lib：/usr/local/lib</code>，这些都是当初compile gcc时写在程序内的。</li>
</ul>
<p>注意上面索顺序是<strong>不会递归</strong>在目录下搜索的。</p>
<p>生成可执行文件后，运行文件时，动态库的搜索路径顺序如下：</p>
<ul>
<li>首先编译目标代码时指定的动态库搜索路径，就是用选项 <code>-Wl,rpath</code> 指定程序在运行时动态库的搜索路径，比如gcc -Wl,-rpath,include -L. -ldltest hello.c，在执行文件时会搜索路径.&#x2F;include；</li>
<li>环境变量<code>LD_LIBRARY_PATH</code>指定的动态库搜索路径；</li>
<li>配置文件<code>/etc/ld.so.conf</code>中指定的动态库搜索路径，即在配置文件中添加动态库的绝对路径，然后运行指令ldconfig是配置文件生效；</li>
<li>默认的动态库搜索路径<code>/lib:/usr/lib</code>。</li>
</ul>
<p>同样上面索顺序是不会递归在目录下搜索的。通常使用动态库简单做法是：把生成的so文件拷贝到&#x2F;usr&#x2F;lib中，这样不管是生成可以执行文件时，还是执行程序时，都能找到需要的so文件。但是普通用户没有&#x2F;usr&#x2F;lib的写入权限，所有要指定<code>LD_LIBRARY_PATH</code>.ls</p>
<p>参考资料:</p>
<ul>
<li><a target="_blank" rel="noopener" href="http://blog.csdn.net/xinyuan510214/article/details/50457433">CFLAGS详解</a></li>
<li><a target="_blank" rel="noopener" href="http://blog.csdn.net/hjwang1/article/details/44497489">Makefile编译选项CC与CXX&#x2F;CPPFLAGS、CFLAGS与CXXFLAGS&#x2F;LDFLAGS</a></li>
<li><a target="_blank" rel="noopener" href="http://blog.csdn.net/MaximusZhou/article/details/38559963">使用gcc时头文件路径和动态链接库路径</a></li>
</ul>
<h2 id="GCC安装-非必要"><a href="#GCC安装-非必要" class="headerlink" title="GCC安装(非必要)"></a>GCC安装(非必要)</h2><p>首先让我们利用系统原来老旧的GCC编译器编译出最新版本的gcc吧，毕竟安装软件的时候，GCC的版本一定要过最低要求。</p>
<p><strong>第一步</strong>： 下载gcc源码</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">mkdir -p ~/src &amp;&amp; cd ~/src</span><br><span class="line">wget https://mirrors.tuna.tsinghua.edu.cn/gnu/gcc/gcc-7.2.0/gcc-7.2.0.tar.gz</span><br><span class="line">tar -zxvf gcc-7.2.0.tar.gz &amp;&amp; cd gcc-7.2.0</span><br><span class="line">ls</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/9/2013053-c5d4c17d7e569961-e411ca5a623a4f4ca75bd3596b0b5530.png" alt="check"></p>
<p><strong>第二步</strong>， 检查系统是否已经具备前置软件, 主要是GMP，MPFR, MPC。这些软件可以到<a href="ftp://gcc.gnu.org/pub/gcc/infrastructure/">ftp://gcc.gnu.org/pub/gcc/infrastructure/</a>找到，然后下载后解压缩，并移动到gcc源码文件夹下。 可以在配置的时候用<code>--with-gmp, --with-mpfr --with-mpc</code>指定具体所在路径。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">cd src</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">GNU Multiple precision Library</span></span><br><span class="line">wget ftp://gcc.gnu.org/pub/gcc/infrastructure/gmp-6.1.0.tar.bz2 \</span><br><span class="line">&amp;&amp; tar -jxvf gmp-6.1.0.tar.bz2 &amp;&amp; mv gmp-6.1.0 gcc-7.2.0/gmp</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">isl library</span></span><br><span class="line">wget ftp://gcc.gnu.org/pub/gcc/infrastructure/isl-0.18.tar.bz2 \</span><br><span class="line">&amp;&amp; tar -jxvf isl-0.18.tar.bz2 &amp;&amp; mv isl-0.18 gcc-7.2.0/isl</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">MPFR Library</span></span><br><span class="line">wget ftp://gcc.gnu.org/pub/gcc/infrastructure/mpfr-3.1.4.tar.bz2 \</span><br><span class="line">&amp;&amp; tar -jxvf mpfr-3.1.4.tar.bz2 &amp;&amp; mv mpfr-3.1.4 gcc-7.2.0/mpfr</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">MPC Library</span></span><br><span class="line">wget ftp://gcc.gnu.org/pub/gcc/infrastructure/mpc-1.0.3.tar.gz \</span><br><span class="line">&amp;&amp; tar -zxvf mpc-1.0.3.tar.gz &amp;&amp; mv mpc-1.0.3 gcc-7.2.0/mpc</span><br></pre></td></tr></table></figure>

<p>不过更加人性化的方法是在GCC源码根目录下运行<code>./contrib/download_prerequisites</code>，可以自动搞定。</p>
<p><strong>第三步</strong>：使用<code>./configure</code>进行配置。官方<strong>强烈</strong>建议, 配置所在文件夹一定要和源码所在文件夹区分开，此外configure还可以配置很多参数，我的代码如下：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">mkdir build &amp;&amp; cd build</span><br><span class="line">../configure\</span><br><span class="line">	--prefix=$HOME/usr \ # 指定安装路径</span><br><span class="line">	--disable-multilib \ # 取消32位库编译</span><br><span class="line">	--enable-threads=posix \ # 使用POSIX/Unix98作为线程支持库</span><br></pre></td></tr></table></figure>

<p>基本上这一步不会出现太多的报错，都能够顺利生成Makefile.</p>
<p><strong>第四步</strong>： 编译. 这步有一个小技巧就是利用多核处理器进行加速，例如<code>make -j2</code> 就是双核。</p>
<p>这一部分很慢很慢，因为默认设置下是3个阶段的引导(3-stage bootstrap), 以保证能够编译出完整的GCC系统并且还不会出错，你可以在配置的时候用<code>--disable-bootstrap</code>进行关闭。</p>
<p><strong>第五步</strong>： 安装。如果你编译都成功了，那么安装也不会存在问题了， <code>make install</code>.</p>
<p>那么我们编译的GCC和系统自带的有什么<strong>区别</strong>吗？</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_"># </span><span class="language-bash">从头编译</span></span><br><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="variable">$HOME</span>/usr/bin/gcc -v</span></span><br><span class="line">Using built-in specs.</span><br><span class="line">COLLECT_GCC=/home/zgxu/usr/bin/gcc</span><br><span class="line">COLLECT_LTO_WRAPPER=/home/zgxu/usr/libexec/gcc/x86_64-pc-linux-gnu/7.2.0/lto-wrapper</span><br><span class="line">Target: x86_64-pc-linux-gnu</span><br><span class="line">Configured with: ../configure --prefix=/home/zgxu/usr --disable-multilib --enable-threads=posix</span><br><span class="line">Thread model: posix</span><br><span class="line">gcc version 7.2.0 (GCC)</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">系统自带</span></span><br><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">gcc -v</span></span><br><span class="line">Using built-in specs.</span><br><span class="line">COLLECT_GCC=gcc</span><br><span class="line">COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/lto-wrapper</span><br><span class="line">Target: x86_64-redhat-linux</span><br><span class="line">Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux</span><br><span class="line">Thread model: posix</span><br><span class="line">gcc version 4.8.5 20150623 (Red Hat 4.8.5-16) (GCC)</span><br></pre></td></tr></table></figure>

<p>不谈安装路径和版本，基本上 <strong>差别</strong> 就是在配置这一步，而这些参数就需要仔细研究了。</p>
<p>一个 <strong>错误</strong> : ‘Link tests are not allowed after GCC_NO_EXECUTABLES.’ 后来发现是第三步没有在独立的文件下构建Makefile.</p>
<p>参考资料：</p>
<ul>
<li>installing GCC: <a target="_blank" rel="noopener" href="https://gcc.gnu.org/install/">https://gcc.gnu.org/install/</a></li>
<li>linux下编译gcc6.2.0: <a target="_blank" rel="noopener" href="https://www.cnblogs.com/oloroso/p/5984985.html">https://www.cnblogs.com/oloroso/p/5984985.html</a></li>
</ul>
<h2 id="CMake-平台无关的编译软件"><a href="#CMake-平台无关的编译软件" class="headerlink" title="CMake: 平台无关的编译软件"></a>CMake: 平台无关的编译软件</h2><p>不同平台有着不同的Make工具用于编译,例如 GNU Make ，QT 的 qmake ，微软的 MS nmake，BSD Make（pmake），Makepp，等等。这些 Make 工具遵循着不同的规范和标准，所执行的 Makefile 格式也千差万别。这样就带来了一个严峻的问题：如果软件想跨平台，必须要保证能够在不同平台编译。而如果使用上面的 Make 工具，就得为每一种标准写一次 Makefile ，这将是一件让人抓狂的工作。</p>
<p>CMake就是针对上面问题所设计的工具：它首先允许开发者编写一种平台无关的 CMakeList.txt 文件来定制整个编译流程，然后再根据目标用户的平台进一步生成所需的本地化 Makefile 和工程文件，如 Unix 的 Makefile 或 Windows 的 Visual Studio 工程。从而做到“Write once, run everywhere”。显然，CMake 是一个比上述几种 make 更高级的编译配置工具。一些使用 CMake 作为项目架构系统的知名开源项目有 VTK、ITK、KDE、OpenCV、OSG 等.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">wget https://cmake.org/files/v3.10/cmake-3.10.2.tar.gz</span><br><span class="line">tar xf cmake-3.10.2.tar.gz</span><br><span class="line"><span class="built_in">cd</span> cmake-3.10.2</span><br></pre></td></tr></table></figure>

<p>参考资料：</p>
<ul>
<li><a target="_blank" rel="noopener" href="http://www.hahack.com/codes/cmake/">http://www.hahack.com/codes/cmake/</a></li>
<li><a target="_blank" rel="noopener" href="https://www.cnblogs.com/d-blog/p/4617208.html">https://www.cnblogs.com/d-blog/p/4617208.html</a></li>
</ul>
<h2 id="几个必须要装的函数库"><a href="#几个必须要装的函数库" class="headerlink" title="几个必须要装的函数库"></a>几个必须要装的函数库</h2><p>在安装之前需要先声明几个环境变量，可以直接添加在配置文件中。这都是后面遇到共享库的问题得到的经验教训。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">export CFLAGS=&quot; -fPIC&quot;</span><br><span class="line">export CXXFLAGS=&quot; -fPIC&quot;</span><br><span class="line">export CPPFLAGS=&quot;-I$HOME/usr/include -I$HOME/usr/include/ncurses -I$HOME/usr/include/X11&quot;</span><br><span class="line">export LDFLAGS=&quot;-L$HOME/usr/lib -L$HOME/usr/lib64&quot;</span><br><span class="line">export LD_LIBRARY_PATH=$HOME/usr/lib:$HOME/usr/lib64</span><br><span class="line">export PKG_CONFIG_PATH=$HOME/usr/lib/pkgconfig:$HOME/usr/share/pkgconfig</span><br></pre></td></tr></table></figure>

<p><strong>ncurses</strong>提供了一系列的函数以便使用者调用它们去生成基于文本的用户界面，许多大名鼎鼎的软件都用到了ncurses，例如vim, screen,tmux,zsh等。并且<strong>samtools</strong>如果需要tview可视化BAM文件，也需要这个库做支持。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">wget ftp://ftp.invisible-island.net/ncurses/ncurses.tar.gz &amp;&amp; tar -zxvf ncurses.tar.gz</span><br><span class="line">./configure --enable-shared --prefix=$HOME/usr</span><br><span class="line">make &amp;&amp; make install</span><br></pre></td></tr></table></figure>

<p><strong>Libevent</strong>是一个用C语言编写的、轻量级的开源高性能事件通知库, 后续安装tmux时候需要这个依赖库。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_"># </span><span class="language-bash">libevent</span></span><br><span class="line">cd src</span><br><span class="line">wget https://github.com/libevent/libevent/releases/download/release-2.1.8-stable/libevent-2.1.8-stable.tar.gz</span><br><span class="line">tar -zxvf libevent-2.1.8-stable.tar.gz &amp;&amp; cd  libevent-2.1.8</span><br><span class="line">./configure prefix=$HOME/usr &amp;&amp; make &amp;&amp; make install</span><br></pre></td></tr></table></figure>

<p><strong>bzip2, xz, zlib</strong>: 文件压缩相关函数库，后续samtools编译时需要。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget http://www.zlib.net/zlib-1.2.11.tar.gz</span><br><span class="line">tar -zxvf zlib-1.2.11.tar.gz &amp;&amp; cd zlib-1.2.11 &amp;&amp; ./configure --prefix=$HOME/usr &amp;&amp; make &amp;&amp; make install</span><br><span class="line">wget http://www.bzip.org/1.0.6/bzip2-1.0.6.tar.gz</span><br><span class="line">tar -zxvf bzip2-1.0.6.tar.gz &amp;&amp; cd bzip2-1.0.6 &amp;&amp; ./configure --prefix=$HOME/usr &amp;&amp; make &amp;&amp; make install</span><br><span class="line">wget https://tukaani.org/xz/xz-5.2.3.tar.gz</span><br><span class="line">tar -zxvf xz-5.2.3.tar.gz &amp;&amp; cd xz-5.2.3 &amp;&amp; ./configure --prefix=$HOME/usr &amp;&amp; make &amp;&amp; make install</span><br></pre></td></tr></table></figure>

<p><strong>openssl, libssh2, libcurl</strong>: 计算机之间文件传输访问相关库。其中OpenSSL是一个安全套接字层密码库，囊括主要的密码算法、常用的密钥和证书封装管理功能及SSL协议，并提供丰富的应用程序供测试或其它目的使用。libssh2是一个C 函数库，用来实现SSH2协议。libcurl主要功能就是用不同的协议连接和沟通不同的服务器.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_"># </span><span class="language-bash">安装有先后</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">openssl</span></span><br><span class="line">wget https://www.openssl.org/source/openssl-1.0.2m.tar.gz</span><br><span class="line">tar -zxvf openssl-1.0.2m.tar.gz &amp;&amp; cd openssl-1.0.2m</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">这里非常神奇的居然是config，添加shared生成动态库</span></span><br><span class="line">./config prefix=$HOME/usr shared</span><br><span class="line">make &amp;&amp; make install</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">卸载使用 make clean</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">libssh2</span></span><br><span class="line">wget https://www.libssh2.org/download/libssh2-1.8.0.tar.gz</span><br><span class="line">tar -zxvf libssh2-1.8.0.tar.gz &amp;&amp; cd libssh2-1.8.0</span><br><span class="line">./configure --with-libssl-prefix=$HOME/usr/ssl --prefix=$HOME/usr</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">libcurl</span></span><br><span class="line">wget https://curl.haxx.se/download/curl-7.56.1.tar.gz</span><br><span class="line">tar -zxvf curl-7.56.1.tar.gz &amp;&amp; cd curl-7.56.1</span><br><span class="line">./configure --prefix=$HOME/usr --enable-http --enable-ftp --enable-file --enable-proxy --enable-telnet --enable-libcurl-option --enable-ipv6 --with-lib --with-ssl</span><br></pre></td></tr></table></figure>

<p><strong>readline</strong>: GNU提供用于这些命令补全、搜索历史命令、行编辑快捷键等等这些人性化的交互方式的函数库，缺少这个标准库，编译的R就缺少自动补全的功能。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">wget http://ftp.gnu.org/gnu/readline/readline-7.0.tar.gz</span><br><span class="line">tar -zxvf readline-7.0.tar.gz &amp;&amp; cd readline-7.0</span><br><span class="line">./configure --prefix=$HOME/usr &amp;&amp; make &amp;&amp; make install</span><br></pre></td></tr></table></figure>

<p><strong>PCRE</strong>: 提供和Perl5相同语法和语义正则表达式的函数库，后续安装R用到。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">wget ftp://ftp.csx.cam.ac.uk/pub/software/programming/pcre/pcre-8.41.tar.gz</span><br><span class="line">tar -zxvf pcre-8.41.tar.gz &amp;&amp; cd pcre-8.41</span><br><span class="line">./configure --enable-utf --enable-pcregrep-libz --enable-pcregrep-libbz2 --prefix=$HOME/usr</span><br></pre></td></tr></table></figure>

<p><strong>x11</strong>：X11也叫做X Window系统，X Window系统 (X11或X)是一种位图显示的视窗系统,是在 Unix 和 类Unix 操作系统，以及OpenVMS上建立图形用户界面的标准工具包和协议，并可用于几乎所有已有的现代操作系统。主要是R编译的时候要用，具体用途有待开发。</p>
<p>x11安装比较复杂，有很多的依赖库，因此需要先安装xtrans, xextproto, xcb(lib,xcb-proto, libpthread-subs), kbproto,xproto,inputproto。但是编译很容易，仅提供下载链接</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">https://www.x.org/releases/X11R7.7/src/lib/xtrans-1.2.7.tar.gz</span><br><span class="line">https://www.x.org/releases/X11R7.7/src/proto/xextproto-7.2.1.tar.gz</span><br><span class="line">https://www.x.org/releases/X11R7.7/src/proto/kbproto-1.0.6.tar.gz</span><br><span class="line">https://www.x.org/releases/X11R7.7/src/proto/xproto-7.0.23.tar.gz</span><br><span class="line">https://www.x.org/releases/X11R7.7/src/proto/inputproto-2.2.tar.gz</span><br><span class="line">https://www.x.org/releases/X11R7.7/src/xcb/libpthread-stubs-0.3.tar.gz</span><br><span class="line">https://www.x.org/releases/X11R7.7/src/xcb/xcb-proto-1.7.1.tar.gz</span><br><span class="line">https://www.x.org/releases/X11R7.7/src/xcb/libxcb-1.8.1.tar.gz</span><br><span class="line">https://www.x.org/releases/X11R7.7/src/lib/libSM-1.2.1.tar.gz</span><br><span class="line">https://www.x.org/releases/X11R7.7/src/lib/libICE-1.0.8.tar.gz</span><br><span class="line">https://www.x.org/releases/X11R7.7/src/lib/libXt-1.1.3.tar.gz</span><br></pre></td></tr></table></figure>

<p>相当于人工检查依赖环境，仅仅是繁琐而已，并不困难</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_"># </span><span class="language-bash">安装X11</span></span><br><span class="line">wget -4 https://www.x.org/releases/X11R7.7/src/lib/libX11-1.5.0.tar.gz</span><br><span class="line">tar -zxvf libX11-1.5.0.tar.gz &amp;&amp; cd libX11-1.5.0</span><br><span class="line">./configure --prefix=$HOME/usr &amp;&amp; make &amp;&amp; make install</span><br></pre></td></tr></table></figure>

<h2 id="编译案例"><a href="#编译案例" class="headerlink" title="编译案例"></a>编译案例</h2><h3 id="安装zsh"><a href="#安装zsh" class="headerlink" title="安装zsh"></a>安装zsh</h3><p>zsh或许可以认为是最好的shell，用过zsh的人都不会想着bash了。不过zsh自定义配置，可供选择的插件以及主题实在是太多，因此一定要搭配oh-my-zsh。zsh依赖ncurses.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">wget -O zsh.tar.gz https://sourceforge.net/projects/zsh/files/latest/download</span><br><span class="line">tar -zxvf zsh.tar.gz &amp;&amp; cd zsh</span><br><span class="line">export CPPFLAGS=&quot;-I$HOME/usr/include/&quot; LDFLAGS=&quot;-L$HOME/usr/lib&quot;</span><br><span class="line">../configure --prefix=$HOME/usr --enable-shared</span><br><span class="line">make &amp;&amp; make install</span><br></pre></td></tr></table></figure>

<p>由于没有root权限，无法使用<code>chsh</code>，只能通过在<code>~/.bashrc</code>添加<code>exec $HOME/usr/bin/zsh -l</code>保证登陆的时候自动切换成zsh。其次, zsh搭配oh-my-zsh才完整, 只不过这里只能手动安装了。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_"># </span><span class="language-bash">从github上克隆oh-my-zsh</span></span><br><span class="line">git clone git://github.com/robbyrussell/oh-my-zsh.git ~/.oh-my-zsh</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">用oh-my-zsh的zsh配置文件替代</span></span><br><span class="line">cp ~/.oh-my-zsh/templates/zshrc.zsh-template ~/.zhsrc</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">安装一些字体, 不然一些主题会显示异常</span></span><br><span class="line">cd src</span><br><span class="line">git clone https://github.com/powerline/fonts.git --depth=1</span><br><span class="line">cd fonts &amp;&amp; ./install.sh</span><br></pre></td></tr></table></figure>

<p>重启一下终端，后面根据需要调整配置文件里的参数。</p>
<h3 id="编译tmux"><a href="#编译tmux" class="headerlink" title="编译tmux"></a>编译tmux</h3><p>tmux和screen类似，也是文本终端神器, 依赖于libevent和ncurses.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">export CPPFLAGS=&quot;-I$HOME/usr/include -I$HOME/usr/include/ncurses&quot;</span><br><span class="line">export LDFLAGS=&quot;-L$HOME/usr/lib -L$HOME/usr/lib64&quot;</span><br><span class="line">mkdir -p src &amp;&amp; cd src</span><br><span class="line">git clone https://github.com/tmux/tmux.git</span><br><span class="line">cd tmux</span><br><span class="line">sh autogen.sh</span><br><span class="line">./configure --prefix=$HOME/usr</span><br><span class="line">make &amp;&amp; make install</span><br></pre></td></tr></table></figure>

<h3 id="编译R语言"><a href="#编译R语言" class="headerlink" title="编译R语言"></a>编译R语言</h3><p>由于我自己编译完全版的GCC套餐，很多之前的gfortran不存在的问题也就不存在了（管理员安装了Java）。此外，R还需要gnu readline, pcre &gt; 8.2, x11。当然这些函数包都在之前安装好了。</p>
<p>一些依赖库</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_"># </span><span class="language-bash">安装tidyverse发现xm12需要libiconv的libiconv.so.2</span></span><br><span class="line">https://ftp.gnu.org/pub/gnu/libiconv/libiconv-1.15.tar.gz</span><br></pre></td></tr></table></figure>

<p>正式安装</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">wget https://cran.r-project.org/src/base/R-3/R-3.4.2.tar.gz</span><br><span class="line">tar -zxvf R-3.4.2.tar.gz  &amp;&amp; cd R-3.4.2/</span><br><span class="line">./configure --prefix=$HOME/R</span><br><span class="line">make &amp;&amp; make install</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/9/2013053-2bae1c5cdc39de29-b1856615b47a43b692e8da39655adcdc.png" alt="R configure"></p>
<p>到此，我可以说Linux平台下即便我没有root权限，也没有多少软件包是我所不能手工编译。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/09/23/%E6%97%A0root%E6%9D%83%E9%99%90%E8%A7%A3%E5%86%B3%E7%BC%96%E8%AF%91%E6%97%B6%E7%9A%84%E4%BE%9D%E8%B5%96%E9%97%AE%E9%A2%98-Compile-Software-in-Linux-Without-Root/" data-id="clm1ywdnu00lcecokhvjyhdeu" data-title="无root权限解决编译时的依赖问题" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用RepeatModeler和RepeatMasker注释基因组重复序列-Annotate-repeat-in-genome-with-RepeatModeler-and-RepeatMasker" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/09/18/%E4%BD%BF%E7%94%A8RepeatModeler%E5%92%8CRepeatMasker%E6%B3%A8%E9%87%8A%E5%9F%BA%E5%9B%A0%E7%BB%84%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97-Annotate-repeat-in-genome-with-RepeatModeler-and-RepeatMasker/" class="article-date">
  <time class="dt-published" datetime="2019-09-18T10:58:40.865Z" itemprop="datePublished">2019-09-18</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/09/18/%E4%BD%BF%E7%94%A8RepeatModeler%E5%92%8CRepeatMasker%E6%B3%A8%E9%87%8A%E5%9F%BA%E5%9B%A0%E7%BB%84%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97-Annotate-repeat-in-genome-with-RepeatModeler-and-RepeatMasker/">使用RepeatModeler和RepeatMasker注释基因组重复序列</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>重复序列注释有两种常用策略，基于同源序列相似性和基于重复序列结构特征。<code>RepeatMasker</code>是基于同源序列相似性注释序列的常用工具, <code>RepeatModeler</code>可用来从头对基因组的重复序列家族进行建模注释，它的核心组件是RECON和RepatScout。</p>
<p>这篇教程介绍如何使用<code>RepeatModeler</code>从头鉴定基因组的重复序列，之后用<code>RepeatMasker</code>根据自定义的重复序列库注释基因组的重复序列。</p>
<h2 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h2><p>原本的<code>RepeatMasker</code>和<code>RepeatModeler</code>的手动安装需要配置很多文件，但是利用bioconda就只用一行命令。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda create -n repeat repeatmasker repeatmodeler</span><br></pre></td></tr></table></figure>

<p>之后在RepeatMasker环境下配置运行环境。由于我的miniconda装在<code>~/opt</code>路径下，因此对应的RepeatMasker路径为<code>~/opt/miniconda3/envs/repeat/share/RepeatMasker/</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">conda activate repeat</span><br><span class="line"><span class="built_in">cd</span> ~/opt/miniconda3/envs/repeat/share/RepeatMasker/</span><br><span class="line">perl ./configure</span><br></pre></td></tr></table></figure>

<p>这一步只需要配置好比对软件</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/9/2013053-9aa560fcadf94885-cb7b3c03d6a644d5be1ebca74a1c1fa4.png" alt="配置比对工具"></p>
<p>之后就会显示RepeatMasker已经配置完毕，其中Dfam_3.0是用于注释的数据库。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/9/1568772932341-2431026ea88645559eda6daf7ef96cd0.png" alt="配置完成"></p>
<h2 id="软件运行"><a href="#软件运行" class="headerlink" title="软件运行"></a>软件运行</h2><p>以拟南芥的参考基因组为例，基因组命名为”Athaliana.fa”</p>
<p>第一步：为RepeatModeler创建索引数据库</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">BuildDatabase -name ath -engine ncbi Athaliana.fa</span><br><span class="line"><span class="comment"># -engine ncbi： 表示使用rmblast</span></span><br><span class="line"><span class="comment"># -name aht： 表示数据库的名字为ath</span></span><br></pre></td></tr></table></figure>

<p>第二步：运行RepeatModeler从头预测</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">RepeatModeler -database ath -engine ncbi -pa 20 &amp;&gt; ath.out &amp;</span><br><span class="line"><span class="comment"># -database 要和上一步一致</span></span><br><span class="line"><span class="comment"># -engine 要和上一步一致</span></span><br><span class="line"><span class="comment"># -pa 表示线程数</span></span><br></pre></td></tr></table></figure>

<p>这一步运行时间相对比较久，和线程数有关。运行中的的文件存放在<code>RM_.xxx</code>文件夹下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">RM_100741.WedSep181006282019</span><br><span class="line">├── consensi.fa </span><br><span class="line">├── consensi.fa.classified</span><br><span class="line">├── consensi.fa.masked</span><br><span class="line">├── families-classified.stk</span><br><span class="line">├── families.stk</span><br><span class="line">├── round-1</span><br><span class="line">├── round-2</span><br><span class="line">├── round-3</span><br><span class="line">├── round-4</span><br><span class="line">└── round-5</span><br></pre></td></tr></table></figure>

<p>运行结束后，就得到了<code>ath-families.fa</code>和<code>ath-families.stk</code>。 前者是找到的重复序列，后者是Stockholm格式的种子联配文件(seed alignment file), 可以用<code>util/dfamConsensusTool.pl</code>上传到<code>Dfam_consensus</code>数据库中。</p>
<p><code>ath-families.fa</code>的fasta的序列部分格式为<code>&gt;repeatname#class/subclass</code>，用于表明每个重复序列的归类。</p>
<p>第三步：根据自定义的重复序列数据库注释基因组</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">RepeatMasker -lib ath-families.fa -e ncbi -<span class="built_in">dir</span> . Athaliana.fa</span><br></pre></td></tr></table></figure>

<p>RepeatMasker比较常用的参数如下</p>
<ul>
<li><code>-e</code>: 搜索引擎，默认都选择ncbi</li>
<li><code>-pa</code>: 并行计算，多线程</li>
<li><code>-s</code>, <code>-q</code>, <code>-qq</code>: 搜索速度，速度和敏感度成反比</li>
<li><code>-lib</code>: 自定义重复数据库</li>
<li><code>-species</code>: 指定物种，例如human, mouse, arabidopsis</li>
<li><code>-gff</code>: 额外输出GFF文件</li>
</ul>
<p>输出结果中, 以.masked结尾的是用N屏蔽后的序列，以tal结尾的则是统计各种重复序列的比例。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/09/18/%E4%BD%BF%E7%94%A8RepeatModeler%E5%92%8CRepeatMasker%E6%B3%A8%E9%87%8A%E5%9F%BA%E5%9B%A0%E7%BB%84%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97-Annotate-repeat-in-genome-with-RepeatModeler-and-RepeatMasker/" data-id="clm1ywdkc00doecokc9xl3jij" data-title="使用RepeatModeler和RepeatMasker注释基因组重复序列" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-使用wgd进行全基因组复制分析-wgd-simple-command-line-tools-for-the-analysis-of-ancient-whole-genome duplications" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/09/15/%E4%BD%BF%E7%94%A8wgd%E8%BF%9B%E8%A1%8C%E5%85%A8%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%A4%8D%E5%88%B6%E5%88%86%E6%9E%90-wgd-simple-command-line-tools-for-the-analysis-of-ancient-whole-genome%20duplications/" class="article-date">
  <time class="dt-published" datetime="2019-09-15T15:42:17.554Z" itemprop="datePublished">2019-09-15</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/09/15/%E4%BD%BF%E7%94%A8wgd%E8%BF%9B%E8%A1%8C%E5%85%A8%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%A4%8D%E5%88%B6%E5%88%86%E6%9E%90-wgd-simple-command-line-tools-for-the-analysis-of-ancient-whole-genome%20duplications/">使用wgd进行全基因组复制分析</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="使用wgd进行全基因组复制分析"><a href="#使用wgd进行全基因组复制分析" class="headerlink" title="使用wgd进行全基因组复制分析"></a>使用wgd进行全基因组复制分析</h1><p>因为全基因组复制(Whole genome duplications, WGD)是生物进化的重要因素之一, 所以WGD分析也是基因组分析经常用到的一种分析方法。举个例子，我们之所以能在多条染色体之间发现一些古老基因连锁现象，是因为被子植物在过去2亿年时间里就出现了多次的全基因组复制和基因丢失事件（见下图，Tang et al., 2008）</p>
<img src="/upload/2019/9/1568514073578-2e96a960afd241c1ac261f52c9ed3124.png" alt="基因组进化" style="zoom:50%;" />

<p>古老WGD检测有两种方法，一种是共线性分析，另一种则是根据Ks分布图。其中Ks定义为平均每个同义位点上的同义置换数，与其对应的还有一个Ka，指的是平均每个非同义位点上的非同义置换数。</p>
<p>如果没有WGD或是大片段重复，那么基因组中的旁系同源基因的同义置换符合指数分布(exponential distribution), 反之，Ks分布图中就会出现一个由于WGD导致的正态分布峰(normal distributed peak). 而古老WGD的年龄则可通过分析这些峰中的同源置换数目来预测(Tiley et al., 2018)。</p>
<p>以发表在Science上的_Papaver somniferum_ L. 基因组文章中的图Fig 1C为例，文章分别分析了_P. somniferum_ 和其他几个物种的Ks分布。从Ks分布图可以看到_P. somniferum_经历了一次比较近的WGD事件(Guo et al., 2018)。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/9/1568531170622-f836464d9d6b46b5896ba091cb4a9b2e.png" alt="Ks plot"></p>
<p>文章中关于WGD的分析流程参考附录8.1 Whole genome duplication in opium poppy genome, 我总结了关键的几点</p>
<ul>
<li>使用megablast进行自比对，寻找基因组中共线性的区块</li>
<li>使用BLASTP基于RBH( reciprocal best hits )进行蛋白之间的相互比对</li>
<li>使用KaKs_Calculator基于YN模型计算RBH基因对中的Ks(synonymous substitution rate)</li>
<li>为了区分Ks中peak是WGD事件还是小规模重复，作者使用MCScanX进行共线性分析，发现93.9%的RBH基因都是基因组内共线性</li>
</ul>
<p>目前WGD的分析流程也有人发了文章，我通过关键字”wgd pipeline”搜索找到了如下几个:</p>
<ul>
<li>GenoDup:  <a target="_blank" rel="noopener" href="https://github.com/MaoYafei/GenoDup-Pipeline">https://github.com/MaoYafei/GenoDup-Pipeline</a></li>
<li>WGDdetector: <a target="_blank" rel="noopener" href="https://github.com/yongzhiyang2012/WGDdetector">https://github.com/yongzhiyang2012/WGDdetector</a></li>
<li>wgd:  <a target="_blank" rel="noopener" href="https://github.com/arzwa/wgd">https://github.com/arzwa/wgd</a></li>
</ul>
<p>这一篇介绍的是wgd的用法。</p>
<h2 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h2><p>wgd目前无法用bioconda直接安装，所以安装起来稍显麻烦，这是因为它的依赖软件有点多。wgd依赖于以下软件</p>
<ul>
<li>BLAST</li>
<li>MCL</li>
<li>MUSCLE&#x2F;MAFFT&#x2F;PRANK</li>
<li>PAML</li>
<li>PhyML&#x2F;FastTree</li>
<li>i-ADHoRe</li>
</ul>
<p>但是好消息是它依赖的软件大部分都可以用bioconda进行安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">conda create -n wgd python=3.5 blast mcl muscle mafft prank paml  fasttree cmake libpng mpi=1.0=mpich</span><br><span class="line">conda activate wgd</span><br></pre></td></tr></table></figure>

<blockquote>
<p>这里选择了mpi&#x3D;1.0&#x3D;mpich, 原因是i-adhore依赖于mpich. 如果安装了openmpi就会出现error while loading shared libraries: libmpi_cxx.so.40: cannot open shared object file: No such file or directory</p>
</blockquote>
<p>之后的安装就简单多了</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">git <span class="built_in">clone</span> https://github.com/arzwa/wgd.git</span><br><span class="line"><span class="built_in">cd</span> wgd</span><br><span class="line">pip install .</span><br><span class="line"><span class="comment"># 或者一行命令</span></span><br><span class="line">pip install git+https://github.com/arzwa/wgd.git</span><br></pre></td></tr></table></figure>

<p>对于i-ADHoRe，需要先在<a target="_blank" rel="noopener" href="http://bioinformatics.psb.ugent.be/webtools/i-adhore/licensing/">http://bioinformatics.psb.ugent.be/webtools/i-adhore/licensing/</a>同意许可，才能下载i-ADHoRe-3.0</p>
<p>由于我的miniconda3安装在<code>~/opt/</code>下，所以安装路径为<code>~/opt/miniconda3/envs/wgd/</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">tar -zxvf i-adhore-3.0.01.tar.gz</span><br><span class="line"><span class="built_in">cd</span> i-adhore-3.0.01</span><br><span class="line"><span class="built_in">mkdir</span> -p build &amp;&amp; <span class="built_in">cd</span> build</span><br><span class="line">cmake .. -DCMAKE_INSTALL_PREFIX=~/opt/miniconda3/envs/wgd/</span><br><span class="line">make -j 4 </span><br><span class="line">make insatall</span><br></pre></td></tr></table></figure>

<h2 id="软件介绍"><a href="#软件介绍" class="headerlink" title="软件介绍"></a>软件介绍</h2><p><a target="_blank" rel="noopener" href="https://wgd.readthedocs.io/en/latest/index.html">WGD</a>一共有9个子模块</p>
<ul>
<li>kde: 对Ks分布进行KDE拟合</li>
<li>ksd : Ks分布构建</li>
<li>mcl:All-vs-ALl的BLASP比对 + MCL分类分析.</li>
<li>mix: Ks分布的混合建模.</li>
<li>pre: 对CDS文件进行预处理</li>
<li>syn: 调用I-ADHoRe 3.0利用GFF文件进行共线性分析</li>
<li>viz: 绘制柱状图和密度图</li>
<li>wf1: 全基因组旁系同源基因组(paranome)的Ks标准分析流程，调用mcl, ksd和syn</li>
<li>wf2: one-vs-one 同源基因(ortholog)的Ks标准分析流程，调用wcl和ksd</li>
</ul>
<p>流程示意图如下:</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/9/wgd-flowchart-d39960ae52c44ba9a56464f63716bf63.png" alt="工作流程"></p>
<h2 id="使用方法"><a href="#使用方法" class="headerlink" title="使用方法"></a>使用方法</h2><p>以甘蔗的基因组 <em>Saccharum spontaneum</em> L 为例，基因组为8倍体，共32条染色体(2n &#x3D; 4x8 &#x3D; 32)</p>
<p>下载CDS和GFF注释文件 tutorial</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p wgd_tutorial &amp;&amp; <span class="built_in">cd</span> wgd_tutorial</span><br><span class="line">wget http://www.life.illinois.edu/ming/downloads/Spontaneum_genome/Sspon.v20190103.cds.fasta.gz</span><br><span class="line">wget http://www.life.illinois.edu/ming/downloads/Spontaneum_genome/Sspon.v20190103.gff3.gz</span><br><span class="line">gunzip *.gz</span><br></pre></td></tr></table></figure>

<p>先用<code>conda activate wgd</code>启动我们的分析环境，然后就开始分析了</p>
<p><strong>第一步</strong>: 使用<code>wgd mcl</code>鉴定基因组内的同源基因</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">wgd mcl -n 20 --cds --mcl -s Sspon.v20190103.cds.fasta -o Sspon_cds.out</span><br><span class="line"><span class="comment"># -n: 线程</span></span><br><span class="line"><span class="comment"># --cds: 输入为cds</span></span><br><span class="line"><span class="comment"># --mcl: mcl聚类</span></span><br></pre></td></tr></table></figure>

<p>这一步运行过程中，wgd会先检查cds序列是否有效，也就是是否以ATG(起始密码子)开头，且以TAA&#x2F;TAG&#x2F;TGA(终止密码子)结尾，然后将cds转成氨基酸序列，之后用Blastp进行相互比对，然后根据blastp结果用mcl聚类的方式寻找旁系同源基因。</p>
<p>输出结果在Sspon_cds.out，有两个结果输出</p>
<ul>
<li>blast.tsv: BLASTP的outfmt6输出结果</li>
<li>blast.tsv.mcl: MCL聚类结果，每一行可以认为是一个基因家族(gene family)</li>
</ul>
<p><strong>第二步</strong>: 使用<code>wgd ksd</code>构建Ks分布</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wgd ksd --n_threads 80 Sspon_cds.out/Sspon.v20190103.cds.fasta.blast.tsv.mcl Sspon.v20190103.cds.fasta</span><br></pre></td></tr></table></figure>

<p>这一步也是先过滤cds中的无效数据，然后用mafft(默认)&#x2F;muscle&#x2F;prank对每个基因家族进行多重序列联配，用codeml计算dN&#x2F;dS, 用alc&#x2F;fasttree(默认)&#x2F;phyml建树</p>
<p>输出结果在wgd_ksd目录下</p>
<ul>
<li>ks.tsv: 每个基因家族中基因对的各项统计，其中包括Ka和Ks</li>
<li>ks.svg: Ks分布，见下图</li>
</ul>
<img src="/upload/2019/9/wgd_ks_distribution-5d632164b20f419c959267c751efc93a.png" alt="Ks分布" style="zoom:50%;" />

<p><strong>第三步</strong>: 如果基因组质量不错，那么可以使用<code>wgd syn</code>进行共线性分析。它能帮我们找到基因组内的共线性区块，以及相应的锚定点</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">wgd syn --feature gene --gene_attribute ID \</span><br><span class="line">    -ks wgd_ksd/Sspon.v20190103.cds.fasta.ks.tsv \</span><br><span class="line">    Sspon.v20190103.gff3 Sspon_cds.out/Sspon.v20190103.cds.fasta.blast.tsv.mcl</span><br><span class="line"><span class="comment">#--feature: 从第三列选择特征</span></span><br><span class="line"><span class="comment">#--gene_attribute: 从第九列提取编号</span></span><br></pre></td></tr></table></figure>

<p>输出图片以.svg结尾，如下所示，图中颜色红蓝代表Ks得分。</p>
<img src="/upload/2019/9/wgd_syn-169aae6bd15d4ed192d245da884c88a0.png" alt="wgd syn" style="zoom:50%;" />

<p>Ks的下游分析通常还包括对Ks分布的统计建模，这可以使用<code>wgd kde</code>进行核密度拟合(Kernel density estimate, KDE)或用<code>wgd mix</code>建立高斯混合模型(Gaussian mixture models)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># KDE</span></span><br><span class="line">wgd kde wgd_ksd/Sspon.v20190103.cds.fasta.ks.tsv</span><br><span class="line"><span class="comment"># Gaussian</span></span><br><span class="line">wgd mix wgd_ksd/Sspon.v20190103.cds.fasta.ks.tsv</span><br></pre></td></tr></table></figure>

<p><code>wgd kde</code>输出kde.svg, 而<code>wdg mix</code>则生成一个 wgd_mix文件夹。</p>
<p><strong>混合模型</strong>常用于基于Ks分布研究WGD。基于一些基本的分子进化假设，我们预期Ks分布中由WGD导致的peak应该符合高斯分布，能够用log-normal分布近似。但是考虑到混合模型容易过拟合，特别是Ks分布，因此作者不建议使用混合模型作为多次WGD假设的正式统计测试。</p>
<h2 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h2><p>Tang, H., Bowers, J.E., Wang, X., Ming, R., Alam, M., and Paterson, A.H. (2008). Synteny and Collinearity in Plant Genomes. Science <em>320</em>, 486–488.</p>
<p>Tiley, G.P., Barker, M.S., and Burleigh, J.G. (2018). Assessing the Performance of Ks Plots for Detecting Ancient Whole Genome Duplications. Genome Biol Evol <em>10</em>, 2882–2898.</p>
<p>Guo, L., Winzer, T., Yang, X., Li, Y., Ning, Z., He, Z., Teodor, R., Lu, Y., Bowser, T.A., Graham, I.A., et al. (2018). The opium poppy genome and morphinan production. Science <em>362</em>, 343–347.</p>
<p>Zwaenepoel, A., and Van de Peer, Y. (2019). wgd—simple command line tools for the analysis of ancient whole-genome duplications. Bioinformatics <em>35</em>, 2153–2155.</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/09/15/%E4%BD%BF%E7%94%A8wgd%E8%BF%9B%E8%A1%8C%E5%85%A8%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%A4%8D%E5%88%B6%E5%88%86%E6%9E%90-wgd-simple-command-line-tools-for-the-analysis-of-ancient-whole-genome%20duplications/" data-id="clm1ywdko00evecokh35dhf8q" data-title="使用wgd进行全基因组复制分析" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2019-09-10T05_45_43.377Z-2019-09-10T05_45_43.377Z" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/09/10/2019-09-10T05_45_43.377Z-2019-09-10T05_45_43.377Z/" class="article-date">
  <time class="dt-published" datetime="2019-09-10T13:45:44.037Z" itemprop="datePublished">2019-09-10</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/09/10/2019-09-10T05_45_43.377Z-2019-09-10T05_45_43.377Z/">Tue Sep 10 2019 05:45:43 GMT+0000 (Coordinated Universal Time)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free/</span><br><span class="line">conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main/</span><br><span class="line">conda config --<span class="built_in">set</span> show_channel_urls <span class="built_in">yes</span></span><br><span class="line">conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/bioconda/</span><br><span class="line">conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge/</span><br></pre></td></tr></table></figure>
      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/09/10/2019-09-10T05_45_43.377Z-2019-09-10T05_45_43.377Z/" data-id="clm1ywddx001gecok94jnazmk" data-title="Tue Sep 10 2019 05:45:43 GMT+0000 (Coordinated Universal Time)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-使用refgenie管理你的参考基因组-Manage-reference-genome-with-refgenie" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/09/09/%E4%BD%BF%E7%94%A8refgenie%E7%AE%A1%E7%90%86%E4%BD%A0%E7%9A%84%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84-Manage-reference-genome-with-refgenie/" class="article-date">
  <time class="dt-published" datetime="2019-09-09T20:41:45.089Z" itemprop="datePublished">2019-09-09</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/09/09/%E4%BD%BF%E7%94%A8refgenie%E7%AE%A1%E7%90%86%E4%BD%A0%E7%9A%84%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84-Manage-reference-genome-with-refgenie/">使用refgenie管理你的参考基因组</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>在服务器管理初期，我管理参考基因组的方法非常简单，就是以物种名建立一个文件夹，然后把和该物种有关的FASTA文件，GFF文件都放到该文件中，之后在文件夹中建立不同软件的索引。</p>
<p>当然目前已经有一些软件可以帮助你进行管理，比如说<code>refgenie</code>。它是一个Python编写的参考基因组管理工具，软件的设计思路如下:</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/9/image-799c4ee8c367485080b144fe4bec4c74.png" alt="设计思路"></p>
<p>你既可以在本地自己构建，也可以从它的服务器上下载已有的物种。</p>
<h2 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装"></a>软件安装</h2><p>refgenie的安装非常简单，只需要一行代码</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 建议先安装miniconda</span></span><br><span class="line">pip install refgenie</span><br></pre></td></tr></table></figure>

<p>之后，你需要初始化一个文件夹，之后下载的基因组都会在该目录下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p ~/reference/</span><br><span class="line">refgenie init -c ~/reference/genome_config.yaml</span><br></pre></td></tr></table></figure>

<p>将这一行<code>export REFGENIE=~/reference/genome_config.yaml</code>根据所用的SHELL加入到对应的<code>.bashrc</code>或<code>.zshrc</code></p>
<h2 id="软件使用"><a href="#软件使用" class="headerlink" title="软件使用"></a>软件使用</h2><p>我们可以使用<code>refgenie listr</code>去查看目前refgenomes服务器中已经有的参考基因组，输出信息如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">Querying available assets from server: http://refgenomes.databio.org/assets</span><br><span class="line">Remote genomes: hg19, hg19_cdna, hg38, hg38_cdna</span><br><span class="line">Remote assets:</span><br><span class="line">  hg19: bismark_bt1_index; bismark_bt2_index; bowtie2_index; bwa_index; fasta; hisat2_index</span><br><span class="line">  hg19_cdna: bowtie2_index; hisat2_index; kallisto_index; salmon_index</span><br><span class="line">  hg38: bismark_bt1_index; bismark_bt2_index; bowtie2_index; bwa_index; fasta; hisat2_index</span><br><span class="line">  hg38_cdna: bowtie2_index; hisat2_index; kallisto_index; salmon_index</span><br></pre></td></tr></table></figure>

<p>我们可以用<code>refgenie pull</code>来下载数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">refgenie pull --genome hg38 --asset bowtie2_index</span><br></pre></td></tr></table></figure>

<p>当然更常见的情况是，你的物种并不在已有的列表中，以及这是一个国外服务器，你甚至都无法拉取列表，所以我们需要用<code>refgenie build</code>来构建参考基因组。</p>
<p>我们以拟南芥参考基因组为例，我们需要先从<a target="_blank" rel="noopener" href="http://plants.ensembl.org/index.html">EnsemblPlants</a>上下载参考基因组序列</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wget <span class="string">&#x27;ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz&#x27;</span></span><br></pre></td></tr></table></figure>

<p>先导入fasta文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">refgenie build --genome TAIR10 --asset fasta --fasta Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz</span><br></pre></td></tr></table></figure>

<p>建立bwa的索引</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">refgenie build --genome TAIR10 --asset bwa_index --fasta Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz</span><br></pre></td></tr></table></figure>

<p>此时检查<code>~/reference/genome_config.yaml</code>文件，里面记录了刚才新建文件的位置</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">config_version: 0.2</span><br><span class="line">genome_folder: /path/to/reference</span><br><span class="line">genome_server: http://refgenomes.databio.org</span><br><span class="line">genomes:</span><br><span class="line">  TAIR10:</span><br><span class="line">    assets:</span><br><span class="line">      fasta:</span><br><span class="line">        path: fasta/TAIR10.fa</span><br><span class="line">        asset_description: Sequences <span class="keyword">in</span> the FASTA format</span><br><span class="line">      fai:</span><br><span class="line">        path: fasta/TAIR10.fa.fai</span><br><span class="line">        asset_description: Indexed fasta file, produced with samtools faidx</span><br><span class="line">      chrom_sizes:</span><br><span class="line">        path: fasta/TAIR10.chrom.sizes</span><br><span class="line">        asset_description: Chromosome sizes file</span><br><span class="line">      bwa_index:</span><br><span class="line">        path: bwa_index</span><br><span class="line">        asset_description: Genome index <span class="keyword">for</span> Burrows-Wheeler Alignment Tool, produced with bwa index</span><br></pre></td></tr></table></figure>

<p>和build相关的详细信息参考<a target="_blank" rel="noopener" href="http://refgenie.databio.org/en/latest/build/">http://refgenie.databio.org/en/latest/build/</a></p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><p>官方文档: <a target="_blank" rel="noopener" href="http://refgenie.databio.org/en/latest/">http://refgenie.databio.org/en/latest/</a><img src="/1" alt="refgenie_interfaces.svg"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/09/09/%E4%BD%BF%E7%94%A8refgenie%E7%AE%A1%E7%90%86%E4%BD%A0%E7%9A%84%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84-Manage-reference-genome-with-refgenie/" data-id="clm1ywdkn00erecok9hcicjtj" data-title="使用refgenie管理你的参考基因组" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2019-09-09T04_04_28.482Z-2019-09-09T04_04_28.482Z" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/09/09/2019-09-09T04_04_28.482Z-2019-09-09T04_04_28.482Z/" class="article-date">
  <time class="dt-published" datetime="2019-09-09T12:04:29.805Z" itemprop="datePublished">2019-09-09</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/09/09/2019-09-09T04_04_28.482Z-2019-09-09T04_04_28.482Z/">Mon Sep 09 2019 04:04:28 GMT+0000 (Coordinated Universal Time)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>同一个R包，不同的安装界面</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/9/image-3d183b98eb664adeb47bd24a57a881c5.png" alt="image.png"></p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/9/image-4d7d61c6894a40e08be8eec421f26936.png" alt="image.png"></p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/9/image-614d0ad8ca894ef4b5a16012a36c9305.png" alt="image.png"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/09/09/2019-09-09T04_04_28.482Z-2019-09-09T04_04_28.482Z/" data-id="clm1ywddw001becokff8032uw" data-title="Mon Sep 09 2019 04:04:28 GMT+0000 (Coordinated Universal Time)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/18/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/17/">17</a><a class="page-number" href="/page/18/">18</a><span class="page-number current">19</span><a class="page-number" href="/page/20/">20</a><a class="page-number" href="/page/21/">21</a><span class="space">&hellip;</span><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/20/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/09/">九月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/09/02/%E5%A6%82%E4%BD%95%E8%AE%A9%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8D%E5%86%8D%E9%BE%9F%E9%80%9F%E4%B8%8B%E8%BD%BDgithub%E7%9A%84%E6%95%B0%E6%8D%AE/">如何让服务器不再龟速下载github的数据</a>
          </li>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>