<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>使用ASMap构建高密度遗传图谱 | 洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="在我大学的时候，构建遗传图谱靠的是人工跑胶，然后看胶图统计基因型。当时我用的SSCP(单分子构象多态性)技术区分单个碱基存在差异的等位基因，要放在4度过夜12小时，然后第二天银染显色，放在一个医学看片的设备上读条带。 现在测序便宜了，简化基因组测序随随便便就能获得成千上万的分子标记。然而标记多有标记多的烦恼，就是以前的作图软件不好用了。以前的暴力穷举的方法在海量标记面前，几乎不可能在有限的时间里完">
<meta property="og:type" content="article">
<meta property="og:title" content="使用ASMap构建高密度遗传图谱">
<meta property="og:url" content="http://xuzhougeng.top/2019/10/06/%E4%BD%BF%E7%94%A8ASMap%E6%9E%84%E5%BB%BA%E9%AB%98%E5%AF%86%E5%BA%A6%E9%81%97%E4%BC%A0%E5%9B%BE%E8%B0%B1-Build-high-density-genetic-map-with-ASMap/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:description" content="在我大学的时候，构建遗传图谱靠的是人工跑胶，然后看胶图统计基因型。当时我用的SSCP(单分子构象多态性)技术区分单个碱基存在差异的等位基因，要放在4度过夜12小时，然后第二天银染显色，放在一个医学看片的设备上读条带。 现在测序便宜了，简化基因组测序随随便便就能获得成千上万的分子标记。然而标记多有标记多的烦恼，就是以前的作图软件不好用了。以前的暴力穷举的方法在海量标记面前，几乎不可能在有限的时间里完">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-ad1e894f837bc6bf-d81b41040f9046dcb12319381ebf878a.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-22c2c7322e096282-d319341714d84b26b8003d08c476da91.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-1efac32ed7ebb769-a00a8710ac764a4c974c690cb4b9d951.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-69c90cc209e6bd42-6f3e0572c8a745e2ba51beaca2948b09.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-610dda9e54155b86-f8545e5d22724effa9f263d737bbd37e.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-d926007182146247-ebc143a4f0094c2195f53cb507c5d904.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-ce51881d19790f5c-1635ca96e79c46fa87b01cb2c1877356.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-d3416167e333d503-2784180654e141a1a7ed3e26d46862d5.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-3e48bdf7c43a5dba-c5573d9940e94409b5ce3b28414f23f5.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-a0b8dbb07f72bbb9-ad71118e0ff44791977a31cacc9bf118.png">
<meta property="article:published_time" content="2019-10-06T17:03:13.955Z">
<meta property="article:modified_time" content="2019-10-06T17:03:13.955Z">
<meta property="article:author" content="xuzhougeng">
<meta property="article:tag" content="NGS">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-ad1e894f837bc6bf-d81b41040f9046dcb12319381ebf878a.png">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main"><article id="post-使用ASMap构建高密度遗传图谱-Build-high-density-genetic-map-with-ASMap" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/06/%E4%BD%BF%E7%94%A8ASMap%E6%9E%84%E5%BB%BA%E9%AB%98%E5%AF%86%E5%BA%A6%E9%81%97%E4%BC%A0%E5%9B%BE%E8%B0%B1-Build-high-density-genetic-map-with-ASMap/" class="article-date">
  <time class="dt-published" datetime="2019-10-06T17:03:13.955Z" itemprop="datePublished">2019-10-06</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 class="p-name article-title" itemprop="headline name">
      使用ASMap构建高密度遗传图谱
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>在我大学的时候，构建遗传图谱靠的是人工跑胶，然后看胶图统计基因型。当时我用的SSCP(单分子构象多态性)技术区分单个碱基存在差异的等位基因，要放在4度过夜12小时，然后第二天银染显色，放在一个医学看片的设备上读条带。</p>
<p>现在测序便宜了，简化基因组测序随随便便就能获得成千上万的分子标记。然而标记多有标记多的烦恼，就是以前的作图软件不好用了。以前的暴力穷举的方法在海量标记面前，几乎不可能在有限的时间里完成构图任务，而且更加麻烦的是高通量的SNP标记还容易出错。</p>
<p>R语言中有一些建图软件包，不过都有一些问题。<code>qtl</code>包的建图算法太老，<code>onemap</code>的算法是SERIATION, RECORD, RCD, UG, 但是代码不是C&#x2F;C++写的，所以运行速度太让人着急。</p>
<p>在2017年，出现了一个新的R包，ASMap，它其实就是将MSTmap移植到了R语言中。MSTmap是很久之前就提出的高通量作图方法，只不过需要编译，有点不太友好。</p>
<p>ASMap支持如下群体:</p>
<ul>
<li>BC: 回交群体</li>
<li>DH: 双单倍体群体</li>
<li>ARIL: 高世代重组近交系, 理论上是已经纯合的遗传群体(除了部分剩余杂合位点), 属于永久性群体</li>
<li>RILn: 低世代重组近交系，n&#x3D;2时，就是F2群体。</li>
</ul>
<p>算法说明简分为两个部分:</p>
<ul>
<li><p><strong>聚类算法</strong>: 如果分子标记 $m_j$ 和$m_k$ 来自于两个不同的连锁群，那么$P_{jk}&#x3D;0.5$, 并且标记间的加权距离(hamming distance)为 $E(d_{djk}) &#x3D; n&#x2F;2$。 根据定义，MSTmap使用<a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Hoeffding%27s_inequality">Hoeffding不等式</a>确定标记是否来自于同一个连锁群</p>
</li>
<li><p><strong>标记排序</strong>: 先根据遗传距离对标记进行分箱，也就是将共分离的标记当做一个对象。之后排序问题就视作 TSP(旅行商问题), 也就是寻找所有标记间最短的路径。</p>
</li>
</ul>
<h3 id="可用函数"><a href="#可用函数" class="headerlink" title="可用函数"></a>可用函数</h3><p>图谱构建函数: 一共有两个，注意p.value的调整</p>
<ul>
<li>mstmap.data.frame()</li>
<li>mstmap.cross(): 从qtl包的对象进行转换，对于高世代的RIL群体，要用<code>conver2riself</code>转换类型, 对于BCnFn群体要用<code>conver2bcsft</code>转换类型</li>
</ul>
<p>标记调整函数:</p>
<ul>
<li>pullCross(): 抽取异常分子标记</li>
<li>pushCross(): 检查完毕后，将符合要求的标记放回原处</li>
<li>pp.init(): 调整偏分离的阈值</li>
</ul>
<p>可视化诊断: </p>
<ul>
<li>profileGen():  统计基因型中交换数，双交换数，缺失情况。 适用于对遗传图谱质量进行评估</li>
<li>profileMark: 统计标记的偏分离，缺失比例，等位基因分布，双交换数</li>
<li>heatMap(): 以标记间RF(重组率)绘制热图</li>
</ul>
<h3 id="使用案例"><a href="#使用案例" class="headerlink" title="使用案例"></a>使用案例</h3><p>内容翻译自: <em>R Package ASMap: Effcient Genetic Linkage Map Construction and Diagnosis</em></p>
<h4 id="第一步-环境准备"><a href="#第一步-环境准备" class="headerlink" title="第一步: 环境准备"></a>第一步: 环境准备</h4><p>加载环境和数据，其中数据集为300个群体，3023个标记的回交群体</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span><span class="string">&quot;qtl&quot;</span><span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span><span class="string">&quot;ASMap&quot;</span><span class="punctuation">)</span></span><br><span class="line">data<span class="punctuation">(</span><span class="string">&quot;mapBCu&quot;</span><span class="punctuation">)</span></span><br><span class="line">summary<span class="punctuation">(</span>mapBCu<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>如果是自己的数据集，有两种方法可以构建，一种是用<code>mstmap.cross</code> 转换<code>qtl</code>包中<code>read.cross</code>得到的对象，如下</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">mapthis <span class="operator">&lt;-</span> read.cross<span class="punctuation">(</span><span class="string">&quot;csv&quot;</span><span class="punctuation">,</span> <span class="string">&quot;http://www.rqtl.org/tutorials&quot;</span><span class="punctuation">,</span> <span class="string">&quot;mapthis.csv&quot;</span><span class="punctuation">,</span></span><br><span class="line">                      estimate.map<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">mapthis <span class="operator">&lt;-</span> convert2bcsft<span class="punctuation">(</span>mapthis<span class="punctuation">,</span> BC.gen <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> F.gen <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> estimate.map <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">mapthis <span class="operator">&lt;-</span> mstmap.cross<span class="punctuation">(</span>mapthis<span class="punctuation">,</span> id <span class="operator">=</span> <span class="string">&quot;id&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>或者用<code>mstmap.data.frame</code>, 要调整如下的参数:</p>
<ul>
<li>object: 数据框, 行对应的标记名, 列是每个个体对应的基因型. <strong>注</strong>, 构建数据框时，一定要保证字符串不能被当做是因子(stringsAsFactors&#x3D;FALSE)。</li>
<li>pop.type: 群体类型，“BH”,”BC”, “RILn”(低世代的近交系,RIL2可以认为是F2), “ARIL”(高世代的近交系)</li>
<li>dist.fun: 计算遗传距离的函数, 默认是”kosambi”, 不需要更改。</li>
<li>p.value: 和群体大小有关，群体越大，p值要越小，也就是要更严格。</li>
</ul>
<blockquote>
<p>对于低世代的近交系，亲本基因型编码为”A”或”a”, “B”或”b”, 杂合基因型编码为”X”. 所谓的缺失值编码为”U”或”-“</p>
</blockquote>
<p><code>mstmap.data.frame</code>这一步运行会比较慢，这是由于它会预先对数据进行分箱，减少标记量，提高后续的运算效率。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-ad1e894f837bc6bf-d81b41040f9046dcb12319381ebf878a.png" alt="运行案例"></p>
<h4 id="第二步-预处理"><a href="#第二步-预处理" class="headerlink" title="第二步: 预处理"></a>第二步: 预处理</h4><p>在正式构建遗传图谱之间，我们需要谨慎的按照如下清单检查我们用于构图的基因型&#x2F;分子标记的质量，包括但不限于</p>
<ul>
<li>检查每个个体对应的标记缺失比例，以及每个标记在所有个体中的缺失失比例。缺失比例过高说明标记或个体存在问题</li>
<li>检查群体间是否存在标记高度相似的两个个体。</li>
<li>检查过度偏分离(segregation distortion)的标记。 高度偏分离的标记可能不会定位到唯一的座位上。</li>
<li>检查交换等位基因的标记。这些标记在构建图谱期间，难以聚类，或者和其他标记有关联，因此需要在分析前修复他们的匹配。</li>
<li>检查共分离的标记。 共分离的标记会严重影响构图时的计算效率，所以可以先暂时把它们忽略掉。</li>
</ul>
<p>检查每个个体的基因型缺失情况</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plot.missing<span class="punctuation">(</span>mapBcu<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-22c2c7322e096282-d319341714d84b26b8003d08c476da91.png" alt="缺失分布"></p>
<p>水平的黑线表示一些个体中，大部分的标记都没有对应的基因型，因此我们将这些个体移除。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sg <span class="operator">&lt;-</span> statGen<span class="punctuation">(</span>mapBCu<span class="punctuation">,</span> bychr<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">,</span> stat.type <span class="operator">=</span> <span class="string">&quot;miss&quot;</span><span class="punctuation">)</span></span><br><span class="line">mapBC1 <span class="operator">&lt;-</span> subset<span class="punctuation">(</span>mapBCu<span class="punctuation">,</span> ind<span class="operator">=</span>sg<span class="operator">$</span>miss <span class="operator">&lt;</span> <span class="number">1600</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>从图谱构建的角度来看，高度相关的个体会增强标记的偏分离. 因此对于基因型几乎一模一样的个体，最好在建立图谱前移除</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">gc <span class="operator">&lt;-</span> genClones<span class="punctuation">(</span>mapBC1<span class="punctuation">,</span> tol<span class="operator">=</span><span class="number">0.95</span><span class="punctuation">)</span></span><br><span class="line">gc<span class="operator">$</span>cgd</span><br></pre></td></tr></table></figure>

<p>从R返回的结果中，可以发现有将近11组的基因型几乎一样。但是第一组的BC045和BC039由于存在大量的基因型缺失，不足以说明两者相同，同样情况的还有BC052和BC045,以及BC168和BC045, 可以排除这三项。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">cgd <span class="operator">&lt;-</span> gc<span class="operator">$</span>cgd<span class="punctuation">[</span><span class="operator">-</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">,</span><span class="number">5</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">mapBC2 <span class="operator">&lt;-</span> fixClones<span class="punctuation">(</span>mapBC1<span class="punctuation">,</span> cgd<span class="punctuation">,</span> consensus <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>接下来，可以进一步检查特定位点的观测等位基因频率和期望频率的偏差。这有可能是<strong>基因型考察错误</strong>，也有可能意味着该区域受到<strong>潜在的生物学和遗传学机制</strong>影响。这一步，同时检查分子标记的偏分离情况，不同基因型的占比，和缺失率。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">profileMark<span class="punctuation">(</span>mapBC2<span class="punctuation">,</span> stat.type <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;seg.dist&quot;</span><span class="punctuation">,</span> <span class="string">&quot;prop&quot;</span><span class="punctuation">,</span><span class="string">&quot;miss&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">            crit.val <span class="operator">=</span> <span class="string">&quot;bonf&quot;</span><span class="punctuation">,</span> layout <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">)</span><span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&quot;l&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-1efac32ed7ebb769-a00a8710ac764a4c974c690cb4b9d951.png" alt="基因型总体情况"></p>
<p>对于高度的偏分离的标记，我们应该直接忽略掉</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">mm <span class="operator">&lt;-</span> statMark<span class="punctuation">(</span>mapBC2<span class="punctuation">,</span> stat.type <span class="operator">=</span> <span class="string">&quot;marker&quot;</span><span class="punctuation">)</span><span class="operator">$</span>marker<span class="operator">$</span>AB</span><br><span class="line">dm <span class="operator">&lt;-</span> markernames<span class="punctuation">(</span>mapBC2<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">(</span>mm <span class="operator">&gt;</span> <span class="number">0.98</span><span class="punctuation">)</span> <span class="operator">|</span> <span class="punctuation">(</span>mm <span class="operator">&lt;</span> <span class="number">0.2</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">mapBC3 <span class="operator">&lt;-</span> drop.markers<span class="punctuation">(</span>mapBC2<span class="punctuation">,</span> dm<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>而对于不偏分离不怎么严重的标记，最好的方法是先把他们放在一边，等图谱构建完成之后再把它们放回到遗传图谱中。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">pp <span class="operator">&lt;-</span> pp.init<span class="punctuation">(</span>miss.thresh <span class="operator">=</span> <span class="number">0.1</span><span class="punctuation">,</span> seg.thresh <span class="operator">=</span> <span class="string">&quot;bonf&quot;</span><span class="punctuation">)</span></span><br><span class="line">mapBC3 <span class="operator">&lt;-</span> pullCross<span class="punctuation">(</span>mapBC3<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;missing&quot;</span><span class="punctuation">,</span> pars<span class="operator">=</span>pp<span class="punctuation">)</span></span><br><span class="line">mapBC3 <span class="operator">&lt;-</span> pullCross<span class="punctuation">(</span>mapBC3<span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&quot;seg.distortion&quot;</span><span class="punctuation">,</span> pars<span class="operator">=</span>pp<span class="punctuation">)</span></span><br><span class="line">mapBC3 <span class="operator">&lt;-</span> pullCross<span class="punctuation">(</span>mapBC3<span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&quot;co.located&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h4 id="第三步-构图"><a href="#第三步-构图" class="headerlink" title="第三步: 构图"></a>第三步: 构图</h4><p>构图通常不是一次性完成的工作，需要多次迭代。先进行第一次尝试，运行时间取决于计算机性能和标记数目</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">mapBC4 <span class="operator">&lt;-</span> mstmap<span class="punctuation">(</span>mapBC3<span class="punctuation">,</span> bychr <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> trace <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> p.value <span class="operator">=</span> <span class="number">1e-12</span><span class="punctuation">)</span></span><br><span class="line">chrlen<span class="punctuation">(</span>mapBC4<span class="punctuation">)</span> <span class="comment">#每个连锁群的长度</span></span><br></pre></td></tr></table></figure>

<p>对构建出图谱进行可视化检查</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">heatMap<span class="punctuation">(</span>mapBC4<span class="punctuation">,</span> lmax <span class="operator">=</span> <span class="number">70</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-69c90cc209e6bd42-6f3e0572c8a745e2ba51beaca2948b09.png" alt="热图"></p>
<p>热图分为两个部分看。上三角展示的是标记间的重组率，下三角则是标记间的LOD值，两者模式要是一样，那就说明图谱还是靠谱的。中间的正方形为连锁不平衡区(LD block), 一般在连锁不平衡区极少发生交换，而不同的连锁不平衡区交换频繁。</p>
<p>但是光看热图还是不够的，我们还得保证遗传图谱里个体的交换次数不应该太多。也就是，每个个体能发生的交换是有限的，比如说如果染色体长度为200cM，那么在回交后代中，每个个体的重组次数应该不多于14次。</p>
<blockquote>
<p>水稻每条染色体平均0.5~1次交换。人类每次减数分裂平均30次。</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pg <span class="operator">&lt;-</span> profileGen<span class="punctuation">(</span>mapBC4<span class="punctuation">,</span> bychr <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> stat.type <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;xo&quot;</span><span class="punctuation">,</span> <span class="string">&quot;dxo&quot;</span><span class="punctuation">,</span> <span class="string">&quot;miss&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">                 id<span class="operator">=</span><span class="string">&quot;Genotype&quot;</span><span class="punctuation">,</span> xo.lambda <span class="operator">=</span> <span class="number">14</span><span class="punctuation">,</span> layout<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span>lty<span class="operator">=</span><span class="number">2</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-610dda9e54155b86-f8545e5d22724effa9f263d737bbd37e.png" alt="交换数分析"></p>
<p>从上图可以发现，一些个体的重组次数明显过高，而且这些个体的双交换次数明显多于其他个体，缺失率也是如此。</p>
<p>我们需要将这些个体剔除，对遗传图谱进行升级</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">mapBC5 <span class="operator">&lt;-</span> subsetCross<span class="punctuation">(</span>mapBC4<span class="punctuation">,</span> ind <span class="operator">=</span> <span class="operator">!</span>pg<span class="operator">$</span>xo.lambda<span class="punctuation">)</span></span><br><span class="line">mapBC6 <span class="operator">&lt;-</span> mstmap<span class="punctuation">(</span>mapBC5<span class="punctuation">,</span> bychr <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> trace <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> p.value <span class="operator">=</span> <span class="number">1e-12</span><span class="punctuation">)</span></span><br><span class="line">chrlen<span class="punctuation">(</span>mapBC6<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>和上次相比，每个连锁群的长度明显减少，一般而言遗传图谱越小，图谱越可靠。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">profileMark<span class="punctuation">(</span>mapBC6<span class="punctuation">,</span> stat.type <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;seg.dist&quot;</span><span class="punctuation">,</span> <span class="string">&quot;prop&quot;</span><span class="punctuation">,</span> <span class="string">&quot;dxo&quot;</span><span class="punctuation">,</span> <span class="string">&quot;recomb&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">            layout <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">)</span><span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;l&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-d926007182146247-ebc143a4f0094c2195f53cb507c5d904.png" alt="交换数分析"></p>
<p>最后结果表明，每个个体的双交换次数都没有超过1 个。</p>
<h4 id="第四步-加入前提剔除的标记"><a href="#第四步-加入前提剔除的标记" class="headerlink" title="第四步: 加入前提剔除的标记"></a>第四步: 加入前提剔除的标记</h4><p>对于前期放在一边的标记，可以在图谱构建完成后放回到完成的图谱中，同时要仔细的检查。</p>
<p>先将515个缺失率在10%~20%的标记塞到图谱上。这一步是将当前的标记和图谱上的标记计算重组率，选择最近位置插入。然后展示长度明显小于其他连锁群的4个连锁群，判断是否能够将他们进行合并。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">pp <span class="operator">&lt;-</span> pp.init<span class="punctuation">(</span>miss.thresh <span class="operator">=</span> <span class="number">0.22</span><span class="punctuation">,</span> max.rf <span class="operator">=</span> <span class="number">0.3</span><span class="punctuation">)</span></span><br><span class="line">mapBC6 <span class="operator">&lt;-</span> pushCross<span class="punctuation">(</span>mapBC6<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;missing&quot;</span><span class="punctuation">,</span> pars <span class="operator">=</span> pp<span class="punctuation">)</span></span><br><span class="line">heatMap<span class="punctuation">(</span>mapBC6<span class="punctuation">,</span> chr <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;L.3&quot;</span><span class="punctuation">,</span> <span class="string">&quot;L.5&quot;</span><span class="punctuation">,</span> <span class="string">&quot;L.8&quot;</span><span class="punctuation">,</span> <span class="string">&quot;L.9&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> lmax <span class="operator">=</span> <span class="number">70</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-ce51881d19790f5c-1635ca96e79c46fa87b01cb2c1877356.png" alt="热图诊断"></p>
<p>从图中，我们可以发现，L.3和L.5之间有明显的连锁，L8和L.9有明显的连锁。我们使用<code>mergeCross()</code>将这两个连锁群进行合并，这个时候的<code>mstmap</code>要保证<code>bychr=TRUE</code>, 对不同的连锁群分别计算标记距离，而不是从聚类开始。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">mlist <span class="operator">&lt;-</span> <span class="built_in">list</span><span class="punctuation">(</span><span class="string">&quot;L.3&quot;</span> <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;L.3&quot;</span><span class="punctuation">,</span> <span class="string">&quot;L.5&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;L.8&quot;</span> <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;L.8&quot;</span><span class="punctuation">,</span> <span class="string">&quot;L.9&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">mapBC6 <span class="operator">&lt;-</span> mergeCross<span class="punctuation">(</span>mapBC6<span class="punctuation">,</span> merge <span class="operator">=</span> mlist<span class="punctuation">)</span></span><br><span class="line"><span class="built_in">names</span><span class="punctuation">(</span>mapBC6<span class="operator">$</span>geno<span class="punctuation">)</span> <span class="operator">&lt;-</span> paste<span class="punctuation">(</span><span class="string">&quot;L.&quot;</span><span class="punctuation">,</span> <span class="number">1</span><span class="operator">:</span><span class="number">7</span><span class="punctuation">,</span> sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span></span><br><span class="line">mapBC7 <span class="operator">&lt;-</span> mstmap<span class="punctuation">(</span>mapBC6<span class="punctuation">,</span> bychr <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> trace <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> p.value <span class="operator">=</span> <span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">chrlen<span class="punctuation">(</span>mapBC7<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这样子就得到了更加优化的图谱。但是L1, L2, L.4的连锁群距离有一些轻微的提高，说明可能存在过多的交换现象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pg1 <span class="operator">&lt;-</span> profileGen<span class="punctuation">(</span>mapBC7<span class="punctuation">,</span> bychr <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> stat.type <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;xo&quot;</span><span class="punctuation">,</span> <span class="string">&quot;dxo&quot;</span><span class="punctuation">,</span><span class="string">&quot;miss&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">                  id <span class="operator">=</span> <span class="string">&quot;Genotype&quot;</span><span class="punctuation">,</span> xo.lambda <span class="operator">=</span> <span class="number">14</span><span class="punctuation">,</span> layout <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span> lty <span class="operator">=</span> <span class="number">2</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-d3416167e333d503-2784180654e141a1a7ed3e26d46862d5.png" alt="交换分析"></p>
<p>我们发现新引入的部分个体是导致该现象的元凶，这就需要我们将这些株系次移除掉，然后重建图谱</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">mapBC8 <span class="operator">&lt;-</span> subsetCross<span class="punctuation">(</span>mapBC7<span class="punctuation">,</span> ind<span class="operator">=</span><span class="operator">!</span>pg1<span class="operator">$</span>xo.lambda<span class="punctuation">)</span></span><br><span class="line">mapBC9 <span class="operator">&lt;-</span> mstmap<span class="punctuation">(</span>mapBC8<span class="punctuation">,</span> bychr <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> trace <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> p.value <span class="operator">=</span> <span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">chrlen<span class="punctuation">(</span>mapBC9<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>移除的株系不会对连锁群的数目造成影响，连锁图谱只会在组内进行重建。大部分连锁群的长度都有了一定程度的下降。继续观察一下标记的遗传图谱的表现情况</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">profileMark<span class="punctuation">(</span>mapBC9<span class="punctuation">,</span> stat.type <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;seg.dist&quot;</span><span class="punctuation">,</span> <span class="string">&quot;prop&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> layout <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">            type <span class="operator">=</span> <span class="string">&quot;l&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-3e48bdf7c43a5dba-c5573d9940e94409b5ce3b28414f23f5.png" alt="频率变化"></p>
<p>从图中可以发现，在L.2里面有一个诡异的突起(偏分离), 需要被移除。当然我们也能看到在L.3, L.5, L.7中，新增的缺失率10%~20%的标记，有一定程度的偏分离。剩下的295个标记为”seg.distortion”的标记说不定能解释这个情况，所以下一步就是把它们塞回去</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">sm <span class="operator">&lt;-</span> statMark<span class="punctuation">(</span>mapBC9<span class="punctuation">,</span> chr <span class="operator">=</span> <span class="string">&quot;L.2&quot;</span><span class="punctuation">,</span> stat.type <span class="operator">=</span> <span class="string">&quot;marker&quot;</span><span class="punctuation">)</span></span><br><span class="line">dm <span class="operator">&lt;-</span> markernames<span class="punctuation">(</span>mapBC9<span class="punctuation">,</span> <span class="string">&quot;L.2&quot;</span><span class="punctuation">)</span><span class="punctuation">[</span>sm<span class="operator">$</span>marker<span class="operator">$</span>neglog10P <span class="operator">&gt;</span> <span class="number">6</span><span class="punctuation">]</span></span><br><span class="line">mapBC10 <span class="operator">&lt;-</span> drop.markers<span class="punctuation">(</span>mapBC9<span class="punctuation">,</span> dm<span class="punctuation">)</span></span><br><span class="line">pp <span class="operator">&lt;-</span> pp.init<span class="punctuation">(</span>seg.ratio <span class="operator">=</span> <span class="string">&quot;70:30&quot;</span><span class="punctuation">)</span></span><br><span class="line">mapBC11 <span class="operator">&lt;-</span> pushCross<span class="punctuation">(</span>mapBC10<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;seg.distortion&quot;</span><span class="punctuation">,</span> pars <span class="operator">=</span> pp<span class="punctuation">)</span></span><br><span class="line">mapBC12 <span class="operator">&lt;-</span> mstmap<span class="punctuation">(</span>mapBC11<span class="punctuation">,</span> bychr <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> trace <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> p.value <span class="operator">=</span> <span class="number">2</span><span class="punctuation">)</span></span><br><span class="line"><span class="built_in">round</span><span class="punctuation">(</span>chrlen<span class="punctuation">(</span>mapBC12<span class="punctuation">)</span> <span class="operator">-</span> chrlen<span class="punctuation">(</span>mapBC9<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">)</span></span><br><span class="line">nmar<span class="punctuation">(</span>mapBC12<span class="punctuation">)</span> <span class="operator">-</span> nmar<span class="punctuation">(</span>mapBC10<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>最后一步就是加入共分离的标记</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mapBC <span class="operator">&lt;-</span> pushCross<span class="punctuation">(</span>mapBC12<span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;co.located&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>最后结果如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/2013053-a0b8dbb07f72bbb9-ad71118e0ff44791977a31cacc9bf118.png" alt="结果统计"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/06/%E4%BD%BF%E7%94%A8ASMap%E6%9E%84%E5%BB%BA%E9%AB%98%E5%AF%86%E5%BA%A6%E9%81%97%E4%BC%A0%E5%9B%BE%E8%B0%B1-Build-high-density-genetic-map-with-ASMap/" data-id="clm4a30hw001se6nkf8tx6k64" data-title="使用ASMap构建高密度遗传图谱" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li></ul>

    </footer>
  </div>
  
    
<nav id="article-nav">
  
    <a href="/2019/10/09/%E3%80%8C%E6%96%87%E7%8C%AE%E3%80%8D%E5%A4%9A%E5%80%8D%E4%BD%93%E6%A4%8D%E7%89%A9%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%B5%8B%E5%BA%8F%E7%BB%84%E8%A3%85%E5%BD%93%E5%89%8D%E7%AD%96%E7%95%A5-Current-Strategies-of-Polyploid-Plant-Genome-Sequence-Assembly/" id="article-nav-newer" class="article-nav-link-wrap">
      <strong class="article-nav-caption">前一篇</strong>
      <div class="article-nav-title">
        
          「文献」多倍体植物基因组测序组装当前策略
        
      </div>
    </a>
  
  
    <a href="/2019/10/04/%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8BSA%E6%96%B9%E6%B3%95%E8%BF%9B%E8%A1%8C%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D%EF%BC%88%E6%B0%B4%E7%A8%BB%E7%AF%87%EF%BC%89-QTL-analysis-using-BSA-seq-in-rice/" id="article-nav-older" class="article-nav-link-wrap">
      <strong class="article-nav-caption">后一篇</strong>
      <div class="article-nav-title">如何使用BSA方法进行遗传定位（水稻篇）</div>
    </a>
  
</nav>

  
</article>


</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/09/">九月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/09/04/%E5%A6%82%E4%BD%95%E5%88%A4%E6%96%AD%E8%BD%AC%E5%BD%95%E7%BB%84%E6%B5%8B%E5%BA%8F%E6%98%AF%E5%90%A6%E4%B8%BA%E9%93%BE%E7%89%B9%E5%BC%82%E6%80%A7%E5%BB%BA%E5%BA%93/">如何判断转录组测序是否为链特异性建库</a>
          </li>
        
          <li>
            <a href="/2023/09/02/%E5%A6%82%E4%BD%95%E8%AE%A9%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8D%E5%86%8D%E9%BE%9F%E9%80%9F%E4%B8%8B%E8%BD%BDgithub%E7%9A%84%E6%95%B0%E6%8D%AE/">如何让服务器不再龟速下载github的数据</a>
          </li>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>