<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/24/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-「文章重现」2019发表在NBT的10x sc-ATAC-seq分析重现-2019-NBT-10x-sc-ATAC-seq-analysis" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/10/%E3%80%8C%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0%E3%80%8D2019%E5%8F%91%E8%A1%A8%E5%9C%A8NBT%E7%9A%8410x%20sc-ATAC-seq%E5%88%86%E6%9E%90%E9%87%8D%E7%8E%B0-2019-NBT-10x-sc-ATAC-seq-analysis/" class="article-date">
  <time class="dt-published" datetime="2019-08-10T17:22:32.875Z" itemprop="datePublished">2019-08-10</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/10/%E3%80%8C%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0%E3%80%8D2019%E5%8F%91%E8%A1%A8%E5%9C%A8NBT%E7%9A%8410x%20sc-ATAC-seq%E5%88%86%E6%9E%90%E9%87%8D%E7%8E%B0-2019-NBT-10x-sc-ATAC-seq-analysis/">「文章重现」2019发表在NBT的10x sc-ATAC-seq分析重现</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>最近(2019.8)在NBT上发表了一篇单细胞ATAC-seq的文章，题为 “Massively parallel single-cell chromatin landscapes of human immune cell development and intratumoral T cell exhaustion”。这篇文章由ATAC-seq的发明者所在的实验室用了10X Genomic 公司开发的sc-ATAC-seq技术分析而来。</p>
<p>文章一共测了20w个细胞，GEO编号是<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE129785">GSE129785</a>, 里面是分析后的数据，而原始的FASTQ数据则在<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA532774">https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA532774</a>, 数据量400G。此外，原始数据和处理过的数据还可以从<a target="_blank" rel="noopener" href="https://support.10xgenomics.com/single-cell-atac/datasets">https://support.10xgenomics.com/single-cell-atac/datasets</a>进行下载。</p>
<p>我下载了其中66.7G的tar文件，解压后放在了百度网盘(链接：<a target="_blank" rel="noopener" href="https://pan.baidu.com/s/1fu4MOhhW3aFtInsiJ5zBWg">https://pan.baidu.com/s/1fu4MOhhW3aFtInsiJ5zBWg</a> 提取码：6j7d ).</p>
<p>最为更重要的是，文章提供了核心分析部分用到的代码，代码地址为：<a target="_blank" rel="noopener" href="https://github.com/GreenleafLab/10x-scATAC-2019">https://github.com/GreenleafLab/10x-scATAC-2019</a></p>
<p>因此，我们就可以根据文章的数据和代码来学习10X sc-ATAC-seq的数据分析方法, 我们的目标是尽量重复出文章里出现的图，</p>
<h2 id="文章数据"><a href="#文章数据" class="headerlink" title="文章数据"></a>文章数据</h2><p>这篇文章能发到NBT，我觉得很大一部分的原因是它用大量的数据证明了scATAC-seq的可靠性，GSE129785编号中里放了67个样本，其中的GSE129785_RAW.tar(里面的文件是处理后的fragments.tsv.gz)有66.7G，解压后之后都是Cell Ranger ATAC (v 1.1.0) 处理后的fragments.gz文件, 我根据文章内容分门别类介绍这些数据：</p>
<p><strong>第零类</strong>: 人类GM12878细胞系和小鼠A20细胞系(B lymphocytes)混合测序，通过滴定的方式来获取不同的细胞量，从而评估双&#x2F;多细胞率(原文计算的结果是约1%)。一共有4个样本，如下</p>
<p>BY_500, BY_1k, BY_5k, BY_10k</p>
<p><strong>第一类</strong>：人类PMBC中不同比例的不同类型细胞混合。作者的目的是为了分析多少细胞量会被鉴定出真实的类群，作者发现，百分之一或者千分之一的比例就可以鉴定出细胞类群，有14个样本</p>
<p>0p1_99p9_CD4Mem_CD8Naive, 0p1_99p9_Mono_T, 0p5_99p5_CD4Mem_CD8Naive, 0p5_99p5_Mono_T, 1_99_CD4Mem_CD8Naive, 1_99_Mono_T, 50_50_CD4Mem_CD8Naive, 50_50_Mono_T, 99_1_CD4Mem_CD8Naive, 99_1_MonoT, 99p5_0p5_CD4Mem_CD8Naive, 99p5_0p5_MonoT, 99p9_0p1_CD4Mem_CD8Naive, 99p9_0p1_Mono_T</p>
<p>其中，p是point的简写，0p1_99p9_CD4Mem_CD8Naive就是1 CD4Mem : 999 CD8Naive. 其中MonoT是作者手误，应该是Mono_T。文章里面只提到了monocytes 和 T cells，但是这里面的数据表明还有不同比例的CD细胞混合。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/sup_fig2-0b1a68fbd8b54b93abdf70b956282c3e.jpg" alt="sup_fig2"></p>
<p><strong>第二类</strong>：fresh PBMC和viably frozen PBMC，后者分为分选和未分选。作者的目的是为了确定即便是冻存的细胞也是能够做sc-ATAC-seq。有3个样本</p>
<p>Fresh_pbmc_5k, Frozen_sorted_pbmc_5k, Frozen_unsorted_pbmc_5k。</p>
<p><strong>第三类</strong>：16个人健康人的peripheral blood, 4个重复</p>
<p>PBMC_Rep1, PBMC_Rep2, PBMC_Rep3, PBMC_Rep4</p>
<p><strong>第四类</strong>：16个人健康人的 bone marrow cells, 就1个样本</p>
<p>Bone_Marrow_Rep1</p>
<p><strong>第五类</strong>：16个人健康人的富集特定表面标记的细胞, </p>
<p>Dendritic_Cells, Monocytes, B_Cells, CD34_Progenitors_Rep1, Regulatory_T_Cells, Naive_CD4_T_Cells_Rep1, Memory_CD4_T_Cells_Rep1, CD4_HelperT, CD4_Memory, CD4_Naive, NK_Cells, Naive_CD8_T_Cells, Memory_CD8_T_Cells, Dendritic_all_cells, CD34_Progenitors_Rep2, Memory_CD4_T_Cells_Rep2, Naive_CD4_T_Cells_Rep2</p>
<p>第三，四，五类其实都是健康人的免疫细胞, 文章过滤后有61,806个细胞, 聚类结果见附录sup fig. 3a</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/sup-fig3a-76590ea376ac4366972e08c88ab290b4.jpg" alt="sup-fig3a"></p>
<p><strong>第六类</strong>：免疫疗法前后的细胞，<strong>site-matched</strong> serial tumor biopsies <strong>pre-</strong> and <strong>post</strong>-PD-1 blockade (pembrolizumab) from five patients, plus post-therapy biopsies from two additional patients, 一共有21个样本，过滤后有37,818个细胞</p>
<ul>
<li>SU001_Immune_Post2, SU001_Tcell_Post2, SU001_Tcell_Post, SU001_Total_Post2, SU001_Total_Pre, SU001_Tumor_Immune_Post</li>
<li>SU006_Immune_Pre, SU006_Tcell_Pre, SU006_Total_Post, SU006_Tumor_Pre</li>
<li>SU008_Immune_Post, SU008_Immune_Pre, SU008_Tcell_Post, SU008_Tcell_Pre, SU008_Tumor_Post, SU008_Tumor_Pre</li>
<li>SU009_Tcell_Post, SU009_Tcell_Pre, SU009_Tumor_Immune_Post, SU009_Tumor_Immune_Pre</li>
<li>SU010_Total_Post, SU010_Total_Pre</li>
<li>SU005_Total_Post</li>
<li>SU007_Total_Post</li>
</ul>
<p><strong>注</strong>: 这里前5个病人还可以分类，一种是直接测Total，也就是unbiased fashion，另一种区分了Immune, Tcell, Tumor_Immune, 也就是cell sorting后分开测, T cells (CD45 + CD3 + ), non-T immune cells (CD45 + CD3 − ), stromal and tumor cells (CD45 − ), 参考附录sup fig. 6a</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/sup-fig6a-5b31ed3f70bd4d39b2e4e54bf3b6dd2b.jpg" alt="sup-fig6a"></p>
<p>对于这些数据，文章用的是下面这套流程来进行处理</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/sc-ATAC-seq-analysis-flow-73caa485a12a4ec5a8f76c842b5ab383.jpg" alt="sc-ATAC-seq-analysis-flow"></p>
<h2 id="数据预处理"><a href="#数据预处理" class="headerlink" title="数据预处理"></a>数据预处理</h2><p>文章的数据预处理交给了Cell Ranger ATAC (v 1.1.0) ，用的是hg19 作为参考基因组，并不是最新GRCh38。 </p>
<blockquote>
<p>经@Jimmy @王诗翔 @伊现富老师 解释，是因为hg19的惯性太大，”周边”太多, 整个生信软件生态没有一致性更新，所以很多分析只能用hg19做。</p>
</blockquote>
<p>Cell Ranger ATAC (v 1.1.0)  能做很多工作，</p>
<ul>
<li>Read过滤和比对(Read filtering and alignment)</li>
<li>Barcode记数(Barcode counting)</li>
<li>鉴定转座酶切割位点(Identification of transposase cut sites)</li>
<li>发现可及染色质峰(Detection of accessible chromatin peaks)</li>
<li>定义细胞(Cell calling)</li>
<li>peak和TF的count矩阵(Count matrix generation for peaks and transcription factors)</li>
<li>降维(Dimensionality reduction)</li>
<li>细胞聚类(Cell clustering)</li>
<li>聚类间差异可及性分析(Cluster differential accessibility)</li>
</ul>
<p>原文用了很长一段描述(见<strong>Data processing using Cell Ranger ATAC software</strong>)来说明了Cell Ranger ATAC的具体过程。当然我们不在乎过程，我们只在乎结果。而结果就是下面这些:</p>
<ul>
<li>singlecell.csv: 每个细胞的信息，如是否在TSS</li>
<li>possorted_bam.bam,possorted_bam.bam.bai: 处理后的BAM与其索引</li>
<li>raw_peak_bc_matrix.h5: 原始的peak-cell矩阵, h5格式存放</li>
<li>raw_peak_bc_matrix: 原始的peak-cell矩阵</li>
<li>analysis: 各种分析结果，如聚类，富集，降维等</li>
<li>filtered_peak_bc_matrix.h5: 过滤后的peak-cell矩阵, h5格式存放</li>
<li>filtered_peak_bc_matrix: 过滤后的peak-cell矩阵</li>
<li>fragments.tsv.gz, fragments.tsv.gz.tbi: 每个barcode的序列和它对应的基因组位置和数目</li>
<li>filtered_tf_bc_matrix.h5: 过滤后的TF-cell矩阵, h5格式存放</li>
<li>filtered_tf_bc_matrix: 过滤后的TF-cell矩阵</li>
<li>cloupe.cloupe: Loupe Cell Browser的输入文件</li>
<li>summary.csv, summary.json: 数据的统计结果, 以csv和json格式存放</li>
<li>web_summary.html: 网页总结信息</li>
<li>peaks.bed: 所有的peak汇总</li>
<li>peak_annotation.csv: peak的注释结果</li>
</ul>
<p>虽然结果很多，但是文章只用到了<strong>framgments.tsv.gz</strong>，就如同我们单细胞转录组分析也只要表达量矩阵。</p>
<h2 id="scATAC-seq-数据分析"><a href="#scATAC-seq-数据分析" class="headerlink" title="scATAC-seq 数据分析"></a>scATAC-seq 数据分析</h2><p>之后文章的方法部分就开始描述scATAC-seq的分析内容，一共有19项，描述多达3.5页。如果让作者在附录尽情发挥的话，那么每一项都会有1到2页内容，所以这将会一项非常持久的更新计划。</p>
<ul>
<li>Filtering cells by TSS enrichment and unique fragments</li>
<li>Generating a counts matrix</li>
<li>Generating union peak sets with LSI</li>
<li>Reads-in-peaks-normalized bigwigs and sequencing tracks</li>
<li>ATAC-seq-centric LSI clustering and visualization</li>
<li>Inferring copy number amplification</li>
<li>TF footprinting</li>
<li>ChromVAR</li>
<li>Computing gene activity scores using Cicero co-accessibility</li>
<li>Analysis of autoimmune variants using Cicero co-accessibility and chromVAR.</li>
<li>HiChIP meta-virtual 4C (metav4C) analysis for Cicero co-accessibility links</li>
<li>Overlap of Cicero co-accessibility links with GTEx eQTLs</li>
<li>Constructing ATAC-seq pseudo-bulk replicates of maximal variance</li>
<li>Constructing gene score pseudo-bulk replicates of maximal variance</li>
<li>Identification of cluster-specific peaks and gene scores through feature binarization</li>
<li>Pseudotime analysis</li>
<li>Barnyard mixing analysis</li>
<li>Analysis of fresh versus frozen PBMCs</li>
<li>Spike-in analysis</li>
</ul>
<h3 id="根据TSS富集和唯一片段过滤细胞"><a href="#根据TSS富集和唯一片段过滤细胞" class="headerlink" title="根据TSS富集和唯一片段过滤细胞"></a>根据TSS富集和唯一片段过滤细胞</h3><p>这一步用到的脚本是<code>01_Filter_Cells_v2.R</code>，读取的fragments为四列。前三列是fragment在染色体的位置，第四列是barcode信息，第五列则是该fragments被测了几次。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#chr	start	end	barcode	number</span></span><br><span class="line">chr1    10059   10114   TTATGTCAGGTTGTTC-1      7</span><br></pre></td></tr></table></figure>

<p>分析者先过滤不到那些fragments少于100的细胞。</p>
<p>之后，按照文章里提到过滤低质量细胞的标准进行过滤</p>
<ul>
<li>cut-offs of 1,000 unique nuclear fragments per cell </li>
<li>a transcription start site (TSS) enrichment score of 8</li>
</ul>
<p>每个细胞的唯一片段数计算比较简单也很好理解，而TSS enrichment score需要看他们之前在2018年发表在science文章的附录(Corces et. al 2018 science)</p>
<blockquote>
<p>To get the fragment length distribution, the width of each<br>fragment&#x2F;GRange was plotted. To get the TSS enrichment profile, each TSS from the R package “TxDb.Hsapiens.UCSC.hg38.knownGene” (accessed by transcripts(TxDb)) was extended 2000 bp in each direction and overlapped with the insertions (each end of a fragment) using “findOverlaps”. Next, the distance between the insertions and the strand-corrected TSS was calculated and the number of insertions occurring in each single-base bin was summed. To normalize this value to the local background, the accessibility at each position +&#x2F;- 2000 bp from the TSS was normalized to the mean of the accessibility at positions +&#x2F;-1900-2000 bp from the TSS. The final TSS enrichment reported was the maximum enrichment value within +&#x2F;- 50 bp of the TSS after smoothing with a rolling mean every 51 bp.</p>
</blockquote>
<p>这依赖于一个核心函数，<code>insertionProfileSingles</code>, 作用是分析每个细胞的插入谱</p>
<p>最后分析者为了避免潜在的doublets，还过滤了含有大于45,000的唯一fragmeng的细胞</p>
<h3 id="构建counts矩阵"><a href="#构建counts矩阵" class="headerlink" title="构建counts矩阵"></a>构建counts矩阵</h3><p>这部分描述包括后续的各种矩阵产生，包括cell-by-window, cell-by-peak, cell-by-tf。主要涉及<code>02_Get_Peak_Set_hg19_v2.R</code> 里的<code>countInsertions</code>函数。</p>
<p> <strong>第一步</strong>:获取 <strong>fragment GenomicRanges</strong>(数据结构如下)，也就是<code>01_Filter_Cells_v2.R</code>过滤后的结果</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">GRanges object with <span class="number">6</span> ranges and <span class="number">2</span> metadata columns<span class="operator">:</span></span><br><span class="line">      seqnames      ranges strand <span class="operator">|</span>                 RG         N</span><br><span class="line">         <span class="operator">&lt;</span>Rle<span class="operator">&gt;</span>   <span class="operator">&lt;</span>IRanges<span class="operator">&gt;</span>  <span class="operator">&lt;</span>Rle<span class="operator">&gt;</span> <span class="operator">|</span>        <span class="operator">&lt;</span>character<span class="operator">&gt;</span> <span class="operator">&lt;</span>numeric<span class="operator">&gt;</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span>     chr1 <span class="number">10060</span><span class="operator">-</span><span class="number">10114</span>      <span class="operator">*</span> <span class="operator">|</span> TTATGTCAGGTTGTTC<span class="operator">-</span><span class="number">1</span>         <span class="number">7</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span>     chr1 <span class="number">10060</span><span class="operator">-</span><span class="number">10126</span>      <span class="operator">*</span> <span class="operator">|</span> GAGGTCCCAGCGTCGT<span class="operator">-</span><span class="number">1</span>         <span class="number">4</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">3</span><span class="punctuation">]</span>     chr1 <span class="number">10061</span><span class="operator">-</span><span class="number">10126</span>      <span class="operator">*</span> <span class="operator">|</span> AACGTACAGTCTCTAG<span class="operator">-</span><span class="number">1</span>         <span class="number">5</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">4</span><span class="punctuation">]</span>     chr1 <span class="number">10061</span><span class="operator">-</span><span class="number">10529</span>      <span class="operator">*</span> <span class="operator">|</span> TAGCCGGAGGATGTAT<span class="operator">-</span><span class="number">1</span>         <span class="number">1</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">5</span><span class="punctuation">]</span>     chr1 <span class="number">10062</span><span class="operator">-</span><span class="number">10114</span>      <span class="operator">*</span> <span class="operator">|</span> TGAATCGTCTGACTTT<span class="operator">-</span><span class="number">1</span>        <span class="number">14</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">6</span><span class="punctuation">]</span>     chr1 <span class="number">10062</span><span class="operator">-</span><span class="number">10126</span>      <span class="operator">*</span> <span class="operator">|</span> CCGTAGGGTCACAGTT<span class="operator">-</span><span class="number">1</span>        <span class="number">13</span></span><br></pre></td></tr></table></figure>

<p><strong>第二步</strong>: 然后通过将fragments的起始位置和结束位置进行合并, 就得到了Tn5 insertion GenomicRange</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">by <span class="operator">&lt;-</span> <span class="string">&quot;RG&quot;</span></span><br><span class="line">inserts <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    GRanges<span class="punctuation">(</span>seqnames <span class="operator">=</span> seqnames<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">,</span> ranges <span class="operator">=</span> IRanges<span class="punctuation">(</span>start<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">,</span> start<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> RG <span class="operator">=</span> mcols<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span>by<span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    GRanges<span class="punctuation">(</span>seqnames <span class="operator">=</span> seqnames<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">,</span> ranges <span class="operator">=</span> IRanges<span class="punctuation">(</span>end<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">,</span> end<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> RG <span class="operator">=</span> mcols<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span>by<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><strong>第三步</strong>: 用<code>findOverlaps</code>找和insertion重叠的feature(如window, TSS, peak, TF)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">overlapDF <span class="operator">&lt;-</span> DataFrame<span class="punctuation">(</span>findOverlaps<span class="punctuation">(</span>query<span class="punctuation">,</span> </span><br><span class="line">                                    inserts<span class="punctuation">,</span> </span><br><span class="line">                                    ignore.strand <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> </span><br><span class="line">                                    maxgap<span class="operator">=</span><span class="operator">-</span><span class="number">1L</span><span class="punctuation">,</span> minoverlap<span class="operator">=</span><span class="number">0L</span><span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;any&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><strong>第四步</strong>: 增加新的一列(id)到重叠输出对象中，这一列的内容是细胞barcode的ID，然后转成<code>sparseMatrix</code>提高存储和计算效率</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">overlapDF<span class="operator">$</span>name <span class="operator">&lt;-</span> mcols<span class="punctuation">(</span>inserts<span class="punctuation">)</span><span class="punctuation">[</span>overlapDF<span class="punctuation">[</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">]</span><span class="punctuation">,</span> by<span class="punctuation">]</span></span><br><span class="line"><span class="comment"># transform类似于dplyr:mutate</span></span><br><span class="line">overlapTDF <span class="operator">&lt;-</span> transform<span class="punctuation">(</span>overlapDF<span class="punctuation">,</span> id <span class="operator">=</span> match<span class="punctuation">(</span>name<span class="punctuation">,</span> unique<span class="punctuation">(</span>name<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 构建稀疏矩阵对象</span></span><br><span class="line">sparseM <span class="operator">&lt;-</span> Matrix<span class="operator">::</span>sparseMatrix<span class="punctuation">(</span></span><br><span class="line">    i <span class="operator">=</span> overlapTDF<span class="punctuation">[</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span> </span><br><span class="line">    j <span class="operator">=</span> overlapTDF<span class="punctuation">[</span><span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">    x <span class="operator">=</span> <span class="built_in">rep</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> nrow<span class="punctuation">(</span>overlapTDF<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">    dims <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">length</span><span class="punctuation">(</span>query<span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">(</span>unique<span class="punctuation">(</span>overlapDF<span class="operator">$</span>name<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>sparseM<span class="punctuation">)</span> <span class="operator">&lt;-</span> unique<span class="punctuation">(</span>overlapDF<span class="operator">$</span>name<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><strong>第五步</strong>: Tn5插入数在peak的比例 (frip)计算方法为: 稀疏矩阵的列和和每个细胞的fragment数相除</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">inPeaks <span class="operator">&lt;-</span> table<span class="punctuation">(</span>overlapDF<span class="operator">$</span>name<span class="punctuation">)</span></span><br><span class="line">total <span class="operator">&lt;-</span> table<span class="punctuation">(</span>mcols<span class="punctuation">(</span>inserts<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span> by<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">total <span class="operator">&lt;-</span> total<span class="punctuation">[</span><span class="built_in">names</span><span class="punctuation">(</span>inPeaks<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">frip <span class="operator">&lt;-</span> inPeaks <span class="operator">/</span> total</span><br></pre></td></tr></table></figure>

<p><strong>第六步</strong>(可选): counts矩阵用edgeR的<code>cpm(matrix, log = TRUE, prior.count = 3)</code>进行log标准化。这个矩阵用于降低那些低count值元素对变异的影响。该标准化<strong>假设</strong>不同细胞类型中所有染色质开放区的差异是比较小的。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">logMat <span class="operator">&lt;-</span> edgeR<span class="operator">::</span>cpm<span class="punctuation">(</span>clusterSums<span class="punctuation">,</span> <span class="built_in">log</span> <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> prior.count <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="利用LSI生成peak合集"><a href="#利用LSI生成peak合集" class="headerlink" title="利用LSI生成peak合集"></a>利用LSI生成peak合集</h3><p>方法参考自Cusanovich, D. A. et al (Cell 2018), 代码见<code>02_Get_Peak_Set_hg19_v2.R</code>, 步骤如下</p>
<p><strong>第一步</strong>: 利用tile函数，沿着基因组构建2.5kb的window</p>
 <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">genome <span class="operator">&lt;-</span> BSgenome.Hsapiens.UCSC.hg19</span><br><span class="line">chromSizes <span class="operator">&lt;-</span> GRanges<span class="punctuation">(</span><span class="built_in">names</span><span class="punctuation">(</span>seqlengths<span class="punctuation">(</span>genome<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> IRanges<span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> seqlengths<span class="punctuation">(</span>genome<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">chromSizes <span class="operator">&lt;-</span> GenomeInfoDb<span class="operator">::</span>keepStandardChromosomes<span class="punctuation">(</span>chromSizes<span class="punctuation">,</span> pruning.mode <span class="operator">=</span> <span class="string">&quot;coarse&quot;</span><span class="punctuation">)</span></span><br><span class="line">windows <span class="operator">&lt;-</span> unlist<span class="punctuation">(</span>tile<span class="punctuation">(</span>chromSizes<span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">2500</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><strong>第二步</strong>: 利用<code>countInsertions</code>基于windows和fragments获取cell-by-window矩阵。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># lapply遍历所有的fragments, 然后合并</span></span><br><span class="line">fragmentFiles <span class="operator">&lt;-</span> list.files<span class="punctuation">(</span><span class="string">&quot;data&quot;</span><span class="punctuation">,</span> pattern <span class="operator">=</span> <span class="string">&quot;.rds&quot;</span><span class="punctuation">,</span> full.names <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">countsList <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span><span class="built_in">seq_along</span><span class="punctuation">(</span>fragmentFiles<span class="punctuation">)</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>i<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">	message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;%s of %s&quot;</span><span class="punctuation">,</span> i<span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">(</span>fragmentFiles<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">	counts <span class="operator">&lt;-</span> countInsertions<span class="punctuation">(</span>windows<span class="punctuation">,</span> readRDS<span class="punctuation">(</span>fragmentFiles<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span> by <span class="operator">=</span> <span class="string">&quot;RG&quot;</span><span class="punctuation">)</span><span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">  counts</span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line">mat <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>countsList<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> x<span class="punctuation">)</span> <span class="operator">%&gt;%</span> Reduce<span class="punctuation">(</span><span class="string">&quot;cbind&quot;</span><span class="punctuation">,</span>.<span class="punctuation">)</span></span><br><span class="line">remove<span class="punctuation">(</span>countsList<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p> <strong>第三步</strong>: 矩阵二项化，并且只要所有细胞中前20,000最开放位点</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># seuratLSI</span></span><br><span class="line">mat<span class="operator">@</span>x<span class="punctuation">[</span>mat<span class="operator">@</span>x <span class="operator">&gt;</span> <span class="number">0</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="number">1</span> </span><br><span class="line">mat <span class="operator">&lt;-</span> mat<span class="punctuation">[</span>head<span class="punctuation">(</span>order<span class="punctuation">(</span>Matrix<span class="operator">::</span>rowSums<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">,</span>decreasing <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">,</span>nFeatures<span class="punctuation">)</span><span class="punctuation">,</span><span class="punctuation">]</span> </span><br></pre></td></tr></table></figure>

<p><strong>第四步</strong>: 利用TF-IDF(term frequency-inverse document frequency)降维</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 计算单个细胞中每个开放位点的频率</span></span><br><span class="line">freqs <span class="operator">&lt;-</span> t<span class="punctuation">(</span>t<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="operator">/</span>Matrix<span class="operator">::</span>colSums<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 计算逆文档频率</span></span><br><span class="line">idf   <span class="operator">&lt;-</span> as<span class="punctuation">(</span><span class="built_in">log</span><span class="punctuation">(</span><span class="number">1</span> <span class="operator">+</span> ncol<span class="punctuation">(</span>mat<span class="punctuation">)</span> <span class="operator">/</span> Matrix<span class="operator">::</span>rowSums<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;sparseVector&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 计算TF-IDF = TF * IDF</span></span><br><span class="line">tfidf <span class="operator">&lt;-</span> as<span class="punctuation">(</span>Matrix<span class="operator">::</span>Diagonal<span class="punctuation">(</span>x<span class="operator">=</span>as.vector<span class="punctuation">(</span>idf<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;sparseMatrix&quot;</span><span class="punctuation">)</span> <span class="operator">%*%</span> freqs</span><br></pre></td></tr></table></figure>

<blockquote>
<p>TF-IDF的思想: 如果某个词比较少见，但是它在这篇文章中多次出现，那么它很可能就反映了这篇文章的特性，正是我们所需要的关键词 – <a target="_blank" rel="noopener" href="http://www.ruanyifeng.com/blog/2013/03/tf-idf.html">http://www.ruanyifeng.com/blog/2013/03/tf-idf.html</a></p>
</blockquote>
<p><strong>第五步</strong>: 利用R语言中的irlba对标准化TF-IDF进行奇异值分解(SVD) , 保留第2-25个维度(第一个维度通常和cell read depth高度相关)作为Seurat的输入</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># SVD降维</span></span><br><span class="line">svd <span class="operator">&lt;-</span> irlba<span class="operator">::</span>irlba<span class="punctuation">(</span>tfidf<span class="punctuation">,</span> nComponents<span class="punctuation">,</span> nComponents<span class="punctuation">)</span></span><br><span class="line">svdDiag <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> nrow<span class="operator">=</span>nComponents<span class="punctuation">,</span> ncol<span class="operator">=</span>nComponents<span class="punctuation">)</span></span><br><span class="line">diag<span class="punctuation">(</span>svdDiag<span class="punctuation">)</span> <span class="operator">&lt;-</span> svd<span class="operator">$</span>d</span><br><span class="line">matSVD <span class="operator">&lt;-</span> t<span class="punctuation">(</span>svdDiag <span class="operator">%*%</span> t<span class="punctuation">(</span>svd<span class="operator">$</span>v<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">rownames<span class="punctuation">(</span>matSVD<span class="punctuation">)</span> <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>mat<span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>matSVD<span class="punctuation">)</span> <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span><span class="string">&quot;PC&quot;</span><span class="punctuation">,</span><span class="built_in">seq_len</span><span class="punctuation">(</span>ncol<span class="punctuation">(</span>matSVD<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 构建Seurat对象</span></span><br><span class="line">obj <span class="operator">&lt;-</span> CreateSeuratObject<span class="punctuation">(</span>mat<span class="punctuation">,</span> project<span class="operator">=</span><span class="string">&#x27;scATAC&#x27;</span><span class="punctuation">,</span> min.cells<span class="operator">=</span><span class="number">0</span><span class="punctuation">,</span> min.genes<span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span></span><br><span class="line">obj <span class="operator">&lt;-</span> SetDimReduction<span class="punctuation">(</span>object <span class="operator">=</span> obj<span class="punctuation">,</span> reduction.type <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> slot <span class="operator">=</span> <span class="string">&quot;cell.embeddings&quot;</span><span class="punctuation">,</span> new.data <span class="operator">=</span> matSVD<span class="punctuation">)</span></span><br><span class="line">obj <span class="operator">&lt;-</span> SetDimReduction<span class="punctuation">(</span>object <span class="operator">=</span> obj<span class="punctuation">,</span> reduction.type <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> slot <span class="operator">=</span> <span class="string">&quot;key&quot;</span><span class="punctuation">,</span> new.data <span class="operator">=</span> <span class="string">&quot;PC&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>第三，四，五步被整合到一个函数<code>seuratLSI</code>中， 文章用的是Seurat V2.3</p>
<p><strong>第六步</strong>: 并用<code>FindClusters</code>进行SNN图聚类(默认0.8分辨率),  如果最小的细胞类群细胞数不够200，降低分辨率重新聚类， 一个函数<code>addClusters</code>实现。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">addClusters <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>obj<span class="punctuation">,</span> minGroupSize <span class="operator">=</span> <span class="number">50</span><span class="punctuation">,</span> dims.use <span class="operator">=</span> <span class="built_in">seq_len</span><span class="punctuation">(</span><span class="number">50</span><span class="punctuation">)</span><span class="punctuation">,</span> initialResolution <span class="operator">=</span> <span class="number">0.8</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  <span class="comment"># 初步聚类</span></span><br><span class="line">  currentResolution <span class="operator">&lt;-</span> initialResolution</span><br><span class="line">  obj <span class="operator">&lt;-</span> FindClusters<span class="punctuation">(</span>object <span class="operator">=</span> obj<span class="punctuation">,</span> reduction.type <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> dims.use <span class="operator">=</span> dims.use<span class="punctuation">,</span> resolution <span class="operator">=</span> currentResolution<span class="punctuation">,</span> print.output <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">  <span class="comment"># 获取最少聚类的细胞数和聚类数</span></span><br><span class="line">  minSize <span class="operator">&lt;-</span> <span class="built_in">min</span><span class="punctuation">(</span>table<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">[[</span>paste0<span class="punctuation">(</span><span class="string">&quot;res.&quot;</span><span class="punctuation">,</span>currentResolution<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  nClust <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>unique<span class="punctuation">(</span>paste0<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">[[</span>paste0<span class="punctuation">(</span><span class="string">&quot;res.&quot;</span><span class="punctuation">,</span>currentResolution<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;Current Resolution = %s, No of Clusters = %s, Minimum Cluster Size = %s&quot;</span><span class="punctuation">,</span> currentResolution<span class="punctuation">,</span> nClust<span class="punctuation">,</span> minSize<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  <span class="comment"># 判断最少聚类细胞数是否大于200，如果低于200，降低分辨率(0.8 × N)</span></span><br><span class="line">  <span class="keyword">while</span><span class="punctuation">(</span>minSize <span class="operator">&lt;=</span> minGroupSize<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    obj<span class="operator">@</span>meta.data <span class="operator">&lt;-</span> obj<span class="operator">@</span>meta.data<span class="punctuation">[</span><span class="punctuation">,</span><span class="operator">-</span>which<span class="punctuation">(</span>colnames<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">)</span><span class="operator">==</span>paste0<span class="punctuation">(</span><span class="string">&quot;res.&quot;</span><span class="punctuation">,</span>currentResolution<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">    currentResolution <span class="operator">&lt;-</span> currentResolution<span class="operator">*</span>initialResolution</span><br><span class="line">    obj <span class="operator">&lt;-</span> FindClusters<span class="punctuation">(</span>object <span class="operator">=</span> obj<span class="punctuation">,</span> reduction.type <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> dims.use <span class="operator">=</span> dims.use<span class="punctuation">,</span> resolution <span class="operator">=</span> currentResolution<span class="punctuation">,</span> print.output <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> force.recalc <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">    minSize <span class="operator">&lt;-</span> <span class="built_in">min</span><span class="punctuation">(</span>table<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">[[</span>paste0<span class="punctuation">(</span><span class="string">&quot;res.&quot;</span><span class="punctuation">,</span>currentResolution<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    nClust <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>unique<span class="punctuation">(</span>paste0<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">[[</span>paste0<span class="punctuation">(</span><span class="string">&quot;res.&quot;</span><span class="punctuation">,</span>currentResolution<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;Current Resolution = %s, No of Clusters = %s, Minimum Cluster Size = %s&quot;</span><span class="punctuation">,</span> currentResolution<span class="punctuation">,</span> nClust<span class="punctuation">,</span> minSize<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>obj<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>需要<strong>注意</strong>，文章用的Seurat版本V2.3, 所以如果安装3.0版本, 上面的代码就失效了但是有解决方案，参考如下博客</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://divingintogeneticsandgenomics.rbind.io/post/clustering-scatacseq-data-the-tf-idf-way/">https://divingintogeneticsandgenomics.rbind.io/post/clustering-scatacseq-data-the-tf-idf-way/</a></li>
</ul>
<p><strong>第七步</strong>: 之后用MACS2对每个聚类里 Tn5-corrected singlebase insertions (each end of the Tn5-corrected fragments) 都进行peak calling，参数设置为<code>–shift -75–extsize 150–nomodel–callsummits–nolambda–keep-dup all -q 0.05</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 保存BED文件</span></span><br><span class="line">dirClusters <span class="operator">&lt;-</span> <span class="string">&quot;results/LSI-Cluster-Beds/&quot;</span></span><br><span class="line">dir.create<span class="punctuation">(</span>dirClusters<span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span><span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>fragmentFiles<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">	fragments <span class="operator">&lt;-</span>readRDS<span class="punctuation">(</span>fragmentFiles<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">	<span class="keyword">for</span><span class="punctuation">(</span>j <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">	  message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;%s of %s&quot;</span><span class="punctuation">,</span> j<span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">	  fragmentsj <span class="operator">&lt;-</span> fragments<span class="punctuation">[</span>fragments<span class="operator">$</span>RG <span class="operator">%in%</span> clusterResults<span class="punctuation">[[</span>j<span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">	  <span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">length</span><span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">	    out <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span></span><br><span class="line">	      chr <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span>seqnames<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span><span class="punctuation">,</span> seqnames<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">	      start <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">as.integer</span><span class="punctuation">(</span>start<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span> <span class="operator">-</span> <span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">as.integer</span><span class="punctuation">(</span>end<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span> <span class="operator">-</span> <span class="number">1</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">	      end <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">as.integer</span><span class="punctuation">(</span>start<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">as.integer</span><span class="punctuation">(</span>end<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">	      <span class="punctuation">)</span> <span class="operator">%&gt;%</span> readr<span class="operator">::</span>write_tsv<span class="punctuation">(</span></span><br><span class="line">          x <span class="operator">=</span> .<span class="punctuation">,</span> </span><br><span class="line">          append <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> </span><br><span class="line">          path <span class="operator">=</span> paste0<span class="punctuation">(</span>dirClusters<span class="punctuation">,</span> paste0<span class="punctuation">(</span><span class="built_in">names</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">[</span>j<span class="punctuation">]</span><span class="punctuation">,</span> <span class="string">&quot;.bed&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">          col_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">	    <span class="punctuation">&#125;</span></span><br><span class="line">	<span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"><span class="comment"># 运行MACS2</span></span><br><span class="line">dirPeaks <span class="operator">&lt;-</span> <span class="string">&quot;results/LSI-Cluster-Peaks/&quot;</span></span><br><span class="line">method <span class="operator">&lt;-</span> <span class="string">&quot;q&quot;</span></span><br><span class="line">cutoff <span class="operator">&lt;-</span> 0.05</span><br><span class="line">shift <span class="operator">&lt;-</span> <span class="operator">-</span><span class="number">75</span></span><br><span class="line">extsize <span class="operator">&lt;-</span> 150</span><br><span class="line">genome_size <span class="operator">&lt;-</span> 2.7e9</span><br><span class="line"><span class="keyword">for</span><span class="punctuation">(</span>j <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">	message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;%s of %s&quot;</span><span class="punctuation">,</span> j<span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">	clusterBedj <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span>dirClusters<span class="punctuation">,</span><span class="built_in">names</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">[</span>j<span class="punctuation">]</span><span class="punctuation">,</span><span class="string">&quot;.bed&quot;</span><span class="punctuation">)</span></span><br><span class="line">	cmdPeaks <span class="operator">&lt;-</span> sprintf<span class="punctuation">(</span></span><br><span class="line">	    <span class="string">&quot;macs2 callpeak -g %s --name %s --treatment %s --outdir %s --format BED --nomodel --call-summits --nolambda --keep-dup all&quot;</span><span class="punctuation">,</span> </span><br><span class="line">	    genome_size<span class="punctuation">,</span> </span><br><span class="line">	    <span class="built_in">names</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">[</span>j<span class="punctuation">]</span><span class="punctuation">,</span> </span><br><span class="line">	    clusterBedj<span class="punctuation">,</span> </span><br><span class="line">	    dirPeaks</span><br><span class="line">	  <span class="punctuation">)</span></span><br><span class="line">	<span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span><span class="built_in">is.null</span><span class="punctuation">(</span>shift<span class="punctuation">)</span> <span class="operator">&amp;</span> <span class="operator">!</span><span class="built_in">is.null</span><span class="punctuation">(</span>extsize<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">	  cmdPeaks <span class="operator">&lt;-</span> sprintf<span class="punctuation">(</span><span class="string">&quot;%s --shift %s --extsize %s&quot;</span><span class="punctuation">,</span> cmdPeaks<span class="punctuation">,</span> shift<span class="punctuation">,</span> extsize<span class="punctuation">)</span></span><br><span class="line">	<span class="punctuation">&#125;</span></span><br><span class="line">	<span class="keyword">if</span> <span class="punctuation">(</span>tolower<span class="punctuation">(</span>method<span class="punctuation">)</span> <span class="operator">==</span> <span class="string">&quot;p&quot;</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">	  cmdPeaks <span class="operator">&lt;-</span> sprintf<span class="punctuation">(</span><span class="string">&quot;%s -p %s&quot;</span><span class="punctuation">,</span> cmdPeaks<span class="punctuation">,</span> cutoff<span class="punctuation">)</span></span><br><span class="line">	<span class="punctuation">&#125;</span><span class="keyword">else</span> <span class="punctuation">&#123;</span></span><br><span class="line">	  cmdPeaks <span class="operator">&lt;-</span> sprintf<span class="punctuation">(</span><span class="string">&quot;%s -q %s&quot;</span><span class="punctuation">,</span> cmdPeaks<span class="punctuation">,</span> cutoff<span class="punctuation">)</span></span><br><span class="line">	<span class="punctuation">&#125;</span></span><br><span class="line">	message<span class="punctuation">(</span><span class="string">&quot;Running Macs2...&quot;</span><span class="punctuation">)</span></span><br><span class="line">	message<span class="punctuation">(</span>cmdPeaks<span class="punctuation">)</span></span><br><span class="line">	system<span class="punctuation">(</span>cmdPeaks<span class="punctuation">,</span> intern <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>之后读取peak summits文件，前后延伸250 bp, 最终宽度为501bp， 并基于ENCODE hg19 blacklist(<a target="_blank" rel="noopener" href="https://www.encodeproject.org/annotations/ENCSR636HFF/)%E8%BF%87%E6%BB%A4%EF%BC%8C%E4%BB%A5%E5%8F%8A%E9%82%A3%E4%BA%9B%E5%BB%B6%E4%BC%B8%E5%90%8E%E4%BC%9A%E8%B6%85%E5%87%BA%E6%9F%93%E8%89%B2%E4%BD%93%E7%9A%84peak%E3%80%82">https://www.encodeproject.org/annotations/ENCSR636HFF/)过滤，以及那些延伸后会超出染色体的peak。</a></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 读取</span></span><br><span class="line">df <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span></span><br><span class="line">  samples <span class="operator">=</span> gsub<span class="punctuation">(</span><span class="string">&quot;\\_summits.bed&quot;</span><span class="punctuation">,</span><span class="string">&quot;&quot;</span><span class="punctuation">,</span>list.files<span class="punctuation">(</span>dirPeaks<span class="punctuation">,</span> pattern <span class="operator">=</span> <span class="string">&quot;\\_summits.bed&quot;</span><span class="punctuation">,</span> full.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  groups <span class="operator">=</span> <span class="string">&quot;scATAC&quot;</span><span class="punctuation">,</span></span><br><span class="line">  summits <span class="operator">=</span> list.files<span class="punctuation">(</span>dirPeaks<span class="punctuation">,</span> pattern <span class="operator">=</span> <span class="string">&quot;\\_summits.bed&quot;</span><span class="punctuation">,</span> full.names <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 合并peak, 延伸, 过滤</span></span><br><span class="line">unionPeaks <span class="operator">&lt;-</span> extendedPeakSet<span class="punctuation">(</span></span><br><span class="line">    df <span class="operator">=</span> df<span class="punctuation">,</span></span><br><span class="line">    BSgenome <span class="operator">=</span> genome<span class="punctuation">,</span> </span><br><span class="line">    extend <span class="operator">=</span> <span class="number">250</span><span class="punctuation">,</span></span><br><span class="line">    blacklist <span class="operator">=</span> <span class="string">&quot;data/hg19.blacklist.bed&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nSummits <span class="operator">=</span> <span class="number">200000</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 过滤Y染色体和线粒体</span></span><br><span class="line">unionPeaks <span class="operator">&lt;-</span> unionPeaks<span class="punctuation">[</span>seqnames<span class="punctuation">(</span>unionPeaks<span class="punctuation">)</span> <span class="operator">%in%</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">22</span><span class="punctuation">,</span><span class="string">&quot;X&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">unionPeaks <span class="operator">&lt;-</span> keepSeqlevels<span class="punctuation">(</span>unionPeaks<span class="punctuation">,</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">22</span><span class="punctuation">,</span><span class="string">&quot;X&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>其中extendPeakSet里的函数<code>nonOverlappingGRanges</code> 会对单个样本中相互重叠的peak进行合并处理。步骤如下</p>
<ul>
<li>如果存在有直接重叠的peak，就只保留最显著的peak，过滤不怎么显著的peak</li>
<li>迭代上一步，处理下一个重叠的peak，直到这类直接性重叠的peak要么都保留，要么都移除。这一步用于处理每个聚类的peak集，保留前200,000延伸的summit, 得到了每个cluster的’cluster-specific’ peak集</li>
<li>接着对每个样本的MACS2 peak score进行标准化，即<code>−log10(Q value)</code>, 然后通过将每个个体的score转成中位数<code>trunc(rank(v))/length(v)’</code>, 其中v表示MACS2 peak scores 向量, 等到”score quantitle”。标准化之后peak就能够在聚类之间进行相互比较，能够产生每个数据集的peak合集</li>
<li>合并所有的cluster的peak，使用之前的迭代方法处理重叠，得到了最显著的peak，抛弃了不怎么显著的peak</li>
<li>接着将那些和基因组gap区(也就是序列是N)重叠peak移掉，以及Y染色体上的peak</li>
</ul>
<p><strong>第八步</strong>: 构建cell-by-peak矩阵.  用到的是<code>countInsertions</code>函数，只不过第一个参数改为了unionPeaks.</p>
 <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#Create Counts list</span></span><br><span class="line">countsPeaksList <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span><span class="built_in">seq_along</span><span class="punctuation">(</span>fragmentFiles<span class="punctuation">)</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>i<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;%s of %s&quot;</span><span class="punctuation">,</span> i<span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">(</span>fragmentFiles<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  gc<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">  countInsertions<span class="punctuation">(</span>unionPeaks<span class="punctuation">,</span> readRDS<span class="punctuation">(</span>fragmentFiles<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span> by <span class="operator">=</span> <span class="string">&quot;RG&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>做一些总结性分析</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#CountsMatrix</span></span><br><span class="line">mat <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>countsPeaksList<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> x<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> Reduce<span class="punctuation">(</span><span class="string">&quot;cbind&quot;</span><span class="punctuation">,</span>.<span class="punctuation">)</span></span><br><span class="line">frip <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>countsPeaksList<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> x<span class="punctuation">[[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> unlist</span><br><span class="line">total <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>countsPeaksList<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> x<span class="punctuation">[[</span><span class="number">3</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> unlist</span><br><span class="line"></span><br><span class="line">se <span class="operator">&lt;-</span> SummarizedExperiment<span class="punctuation">(</span></span><br><span class="line">  assays <span class="operator">=</span> SimpleList<span class="punctuation">(</span>counts <span class="operator">=</span> mat<span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">  rowRanges <span class="operator">=</span> unionPeaks</span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line">rownames<span class="punctuation">(</span>se<span class="punctuation">)</span> <span class="operator">&lt;-</span> paste<span class="punctuation">(</span>seqnames<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">,</span>start<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">,</span>end<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;_&quot;</span><span class="punctuation">)</span></span><br><span class="line">colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="operator">$</span>FRIP <span class="operator">&lt;-</span> frip</span><br><span class="line">colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="operator">$</span>uniqueFrags <span class="operator">&lt;-</span> total <span class="operator">/</span> <span class="number">2</span></span><br></pre></td></tr></table></figure>


<h3 id="Reads-in-peaks-normalized-bigwigs-and-sequencing-tracks"><a href="#Reads-in-peaks-normalized-bigwigs-and-sequencing-tracks" class="headerlink" title="Reads-in-peaks-normalized bigwigs and sequencing tracks."></a>Reads-in-peaks-normalized bigwigs and sequencing tracks.</h3><p>这部分内容在GitHub上没有代码，所以我根据文章描述编写代码实现</p>
<p>…</p>
<h3 id="ATAC-seq-centric-LSI-聚类和可视化"><a href="#ATAC-seq-centric-LSI-聚类和可视化" class="headerlink" title="ATAC-seq-centric LSI 聚类和可视化"></a>ATAC-seq-centric LSI 聚类和可视化</h3><p>以Cusanovich et. al (Nature 2018) 提出的策略计算TF-IDF转换, 然后irlba进行奇异值分解(SVD), 接着用前50维作为Seurat的输入，实现SNN图聚类，然后用<code>FindCluster</code>寻找聚类。 你会发现这个内容和上一节类似，只不过上一节是基于window, 这一节是基于peak。</p>
 <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># TF-IDF, SVD </span></span><br><span class="line">obj <span class="operator">&lt;-</span> seuratLSI<span class="punctuation">(</span>assay<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">,</span> nComponents <span class="operator">=</span> <span class="built_in">max</span><span class="punctuation">(</span>nPCs1<span class="punctuation">)</span><span class="punctuation">,</span> nFeatures <span class="operator">=</span> <span class="literal">NULL</span><span class="punctuation">)</span></span><br><span class="line">obj<span class="operator">@</span>meta.data <span class="operator">&lt;-</span> as.data.frame<span class="punctuation">(</span>cbind<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">,</span> colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 聚类分析</span></span><br><span class="line">obj <span class="operator">&lt;-</span> FindClusters<span class="punctuation">(</span>object <span class="operator">=</span> obj<span class="punctuation">,</span> dims.use <span class="operator">=</span> nPCs1<span class="punctuation">,</span> print.output <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> n.start <span class="operator">=</span> <span class="number">10</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>分析者发现这里存在能被检测到的批次效应，它会混淆后续的分析。因此为了减缓影响，他们从二项化后到可及矩阵中计算每个聚类的和，然后对edgeR的<code>cpm(matrix, log = TRUE, prior.count = 3)</code>结果进行log标准化 </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">mat <span class="operator">&lt;-</span> assay<span class="punctuation">(</span>se<span class="punctuation">)</span></span><br><span class="line">mat<span class="operator">@</span>x<span class="punctuation">[</span>mat<span class="operator">@</span>x <span class="operator">&gt;</span> <span class="number">0</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="number">1</span></span><br><span class="line">clusterSums <span class="operator">&lt;-</span> groupSums<span class="punctuation">(</span>mat <span class="operator">=</span> mat<span class="punctuation">,</span> groups <span class="operator">=</span> paste0<span class="punctuation">(</span><span class="string">&quot;C&quot;</span><span class="punctuation">,</span>obj<span class="operator">@</span>meta.data<span class="operator">$</span>res.0.8<span class="punctuation">)</span><span class="punctuation">,</span> sparse <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">logMat <span class="operator">&lt;-</span> edgeR<span class="operator">::</span>cpm<span class="punctuation">(</span>clusterSums<span class="punctuation">,</span> <span class="built_in">log</span> <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> prior.count <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>然后用<code>rowVars</code>挑选前25,000个波动peak，这是基于上一步标准化的矩阵 </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">varPeaks <span class="operator">&lt;-</span> head<span class="punctuation">(</span>order<span class="punctuation">(</span>matrixStats<span class="operator">::</span>rowVars<span class="punctuation">(</span>logMat<span class="punctuation">)</span><span class="punctuation">,</span> decreasing <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">,</span> nTop<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>为什么不用原来的稀疏二项矩阵，而用log标准化的CPM矩阵，原因有亮点</p>
<ul>
<li>能够降低不同聚类间因细胞数不同导致的误差</li>
<li>count转换到标准化空间后的均值-变异关系会减弱</li>
</ul>
<p>之后重新用25,000波动peak提取稀疏的二项化矩阵，然后重新做TF-IDF, SVD, SNN graph聚类等运算。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#Re-run Seurat LSI</span></span><br><span class="line">message<span class="punctuation">(</span><span class="string">&quot;Making Seurat LSI Object...&quot;</span><span class="punctuation">)</span></span><br><span class="line">obj2 <span class="operator">&lt;-</span> seuratLSI<span class="punctuation">(</span>assay<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">[</span>varPeaks<span class="punctuation">,</span><span class="punctuation">]</span><span class="punctuation">,</span> nComponents <span class="operator">=</span> <span class="built_in">max</span><span class="punctuation">(</span>nPCs2<span class="punctuation">)</span><span class="punctuation">,</span> nFeatures <span class="operator">=</span> <span class="literal">NULL</span><span class="punctuation">)</span></span><br><span class="line">stopifnot<span class="punctuation">(</span>identical<span class="punctuation">(</span>rownames<span class="punctuation">(</span>obj2<span class="operator">@</span>meta.data<span class="punctuation">)</span><span class="punctuation">,</span> colnames<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">obj2<span class="operator">@</span>meta.data <span class="operator">&lt;-</span> as.data.frame<span class="punctuation">(</span>cbind<span class="punctuation">(</span>obj2<span class="operator">@</span>meta.data<span class="punctuation">,</span> colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">message<span class="punctuation">(</span><span class="string">&quot;Adding Graph Clusters...&quot;</span><span class="punctuation">)</span></span><br><span class="line">obj2 <span class="operator">&lt;-</span> FindClusters<span class="punctuation">(</span>object <span class="operator">=</span> obj2<span class="punctuation">,</span> reduction.type <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> dims.use <span class="operator">=</span> nPCs2<span class="punctuation">,</span> print.output <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> n.start <span class="operator">=</span> <span class="number">10</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>用<code>Seurat::RunUMAP</code>函数进行可视化展示</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">message<span class="punctuation">(</span><span class="string">&quot;Running UMAP&quot;</span><span class="punctuation">)</span></span><br><span class="line">obj2 <span class="operator">&lt;-</span> RunUMAP<span class="punctuation">(</span>object <span class="operator">=</span> obj2<span class="punctuation">,</span> reduction.use <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> dims.use <span class="operator">=</span> nPCs2<span class="punctuation">)</span></span><br><span class="line">plotUMAP <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>GetCellEmbeddings<span class="punctuation">(</span>obj2<span class="punctuation">,</span>reduction.type<span class="operator">=</span><span class="string">&quot;umap&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> obj2<span class="operator">@</span>meta.data<span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>plotUMAP<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;x&quot;</span><span class="punctuation">,</span><span class="string">&quot;y&quot;</span><span class="punctuation">,</span>colnames<span class="punctuation">(</span>plotUMAP<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">3</span><span class="operator">:</span>ncol<span class="punctuation">(</span>plotUMAP<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">clustCol <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>plotUMAP<span class="punctuation">)</span><span class="punctuation">[</span>grep<span class="punctuation">(</span><span class="string">&quot;res&quot;</span><span class="punctuation">,</span>colnames<span class="punctuation">(</span>plotUMAP<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="operator">$</span>Clusters <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span><span class="string">&quot;Cluster&quot;</span><span class="punctuation">,</span><span class="built_in">as.integer</span><span class="punctuation">(</span>plotUMAP<span class="punctuation">[</span><span class="punctuation">,</span>clustCol<span class="punctuation">]</span><span class="punctuation">)</span> <span class="operator">+</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="operator">$</span>UMAP1 <span class="operator">&lt;-</span> plotUMAP<span class="operator">$</span>x</span><br><span class="line">colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="operator">$</span>UMAP2 <span class="operator">&lt;-</span> plotUMAP<span class="operator">$</span>y</span><br><span class="line"></span><br><span class="line">pdf<span class="punctuation">(</span><span class="string">&quot;results/LSI-Clustering-Peaks.pdf&quot;</span><span class="punctuation">)</span></span><br><span class="line">ggplot<span class="punctuation">(</span>plotUMAP<span class="punctuation">,</span> aes<span class="punctuation">(</span>x<span class="operator">=</span>x<span class="punctuation">,</span>y<span class="operator">=</span>y<span class="punctuation">,</span>color<span class="operator">=</span>res.0.8<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span> geom_point<span class="punctuation">(</span>size <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  theme_bw<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span> xlab<span class="punctuation">(</span><span class="string">&quot;UMAP1&quot;</span><span class="punctuation">)</span> <span class="operator">+</span> ylab<span class="punctuation">(</span><span class="string">&quot;UMAP2&quot;</span><span class="punctuation">)</span></span><br><span class="line">dev.off<span class="punctuation">(</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>对于亚群分析(hematopoiesis: CD34 + bone marrow 和 DCs; tumor: T cells), 分析者重新计算了聚类的和以及log标准化的CPM值，分别从CD34+ cells 和 T cells中鉴定出前10,00和5,00个波动peak。这些波动peak用于提取稀疏二项化可及矩阵，然后做TF-IDF转换。对TF-IDF转换结果进行奇异值分解(SVD),得到更低的维度, 即用前25个维度表征原来的数据集。分贝用1-25和2-25个维度作为Seurat的输入，进行SNN graph clustering（默认resolution&#x3D;0.8). 最后的结果用<code>Seurat::RunUMAP</code>函数进行可视化展示</p>
<h3 id="细胞类群定义"><a href="#细胞类群定义" class="headerlink" title="细胞类群定义"></a>细胞类群定义</h3><p>以human hematopoiesis为例，介绍作者定义细胞类群的三种方法。</p>
<p> (1) 顺式作用元件(cis-lements, 即ATAC-seq peaks)的染色质可及性</p>
<p> (2) 利用和单个基因启动子相关的几个增强子的可及性计算基因活跃得分(GA)</p>
<p> (3) TF活跃度，通过计算每个细胞中全基因组范围内的TF结合位点可及性得出</p>
<p>这三种方法最大的特点就是不需要你再去测一个单细胞RNA-seq或者要一个已有的Bulk ATAC-seq谱，仅仅用当前的单细胞ATAC-seq数据即可。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/10/%E3%80%8C%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0%E3%80%8D2019%E5%8F%91%E8%A1%A8%E5%9C%A8NBT%E7%9A%8410x%20sc-ATAC-seq%E5%88%86%E6%9E%90%E9%87%8D%E7%8E%B0-2019-NBT-10x-sc-ATAC-seq-analysis/" data-id="cllszh643007ee2qe0xxvco1z" data-title="「文章重现」2019发表在NBT的10x sc-ATAC-seq分析重现" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-「生信基础课」初学者入门Linux最少必要的知识点-「生信基础课」初学者入门Linux最少必要的知识点" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/07/%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9-%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9/" class="article-date">
  <time class="dt-published" datetime="2019-08-07T11:22:24.381Z" itemprop="datePublished">2019-08-07</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/07/%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9-%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9/">「生信基础课」初学者入门Linux最少必要的知识点</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <blockquote>
<p>这是我在<a target="_blank" rel="noopener" href="https://space.bilibili.com/249108235">哔哩哔哩</a>上课程的文字稿版本。我有点完美主义的倾向，所以为了避免自己因为觉得自己某个部分的说话不利索，不清楚，所以总是会写稿子。<br><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/Linux%E5%85%A5%E9%97%A8%E8%A7%86%E9%A2%91%E7%B4%A0%E6%9D%90-e7a9613d7c074c9dbd0d20be3ecadd3b.jpg" alt="素材库"></p>
</blockquote>
<p>这次课程的主题是生信入门必须学习的几个Linux操作。</p>
<h2 id="为什么要学习Linux"><a href="#为什么要学习Linux" class="headerlink" title="为什么要学习Linux"></a>为什么要学习Linux</h2><p>如果你要学习生物信息学，那么你一定要学会使用Linux。这是因为，绝大部分的生信软件都需要运行在Linux平台，并且有些时候，你会要用到服务器去处理大量的数据，而这些服务器几乎都是Linux环境。</p>
<p>当然，你可以说，我可以用云平台呀，毕竟现在很多公司都开发了云平台，我们拖拖拽拽就可以搞分析了呀。是的，现在的云平台也来越成熟，使用体验也很棒，我也很喜欢。不过我还是建议大家学Linux，因为当你学会Linux之后，你会变得更加自由，不再受限于别人的平台，而且能让日常工作变得高效起来。</p>
<h2 id="如何学习Linux"><a href="#如何学习Linux" class="headerlink" title="如何学习Linux"></a>如何学习Linux</h2><p>那么如何学习呢？通常你问别人如何学Linux的时候，对方十有八九会给你推荐鸟哥的书，这没啥毛病，因为我也是靠鸟哥的书入门的，那本书已经被我翻成了这副模样。</p>
<p>学习没有捷径，但是可以少走几步路。我们不需要精通Linux，成为Linux大牛，靠Linux吃饭，我们只需要掌握一些基本操作，搞得定日常基本操作就行。就像绝大部分人用Windows的时候，也不知道如何配置Windows里的环境变量，但是一点不妨碍基本使用呀。</p>
<p>通过我对自己平时操作的总结，我认为学Linux起码要掌握下面这三方面内容。</p>
<ul>
<li>操作迁移，就是把Windows上的常见操作都在Linux里找到对应的操作</li>
<li>TAB补全，不但可以让你的操作更快，而且还能让你避免大部分的报错</li>
<li>环境变量，学习Linux一定要理解环境变量，否则你就装不好软件。</li>
</ul>
<p>不过，最重要的是，你得先有一个Linux环境。我在哔哩哔哩上录制了两个视频，教大家如何在Windows平台上配置你的Linux环境，如何在Mac上配置Linux环境。</p>
<p>先看Vim视频:</p>
<ul>
<li><p><a target="_blank" rel="noopener" href="https://www.bilibili.com/video/av58673256">https://www.bilibili.com/video/av58673256</a><br>  再配置Linux环境：</p>
</li>
<li><p>Windows10&#x2F;Linux: <a target="_blank" rel="noopener" href="https://www.bilibili.com/video/av58541835">https://www.bilibili.com/video/av58541835</a></p>
</li>
<li><p>MacOS: <a target="_blank" rel="noopener" href="https://www.bilibili.com/video/av58541835">https://www.bilibili.com/video/av58541835</a></p>
</li>
</ul>
<p>本次视频的操作虽然是在mac上录制，但是适用于大部分的Linux环境，包括Ubuntu和CentOS，如果有什么问题，欢迎在留言区提问交流。</p>
<h2 id="TAB补全"><a href="#TAB补全" class="headerlink" title="TAB补全"></a>TAB补全</h2><p>在演示Linux操作之前，我一定要先强调下tab补全的重要性，它是我认为最重要的一个操作。</p>
<p>初学者在学习Linux的时候，最常见的状态就是，按照教程内容，一个一个字符的敲。由于刚学习，敲代码还不熟练，因此，十有八九会出现敲错的情况，那么结果就是代码运行失败。运行失败怎么办？可能也不会去看报错，然后检查自己的输入，估计就是截图或者拍照发到群里提问了。</p>
<p>因此，在正式开始敲代码之前，大家先在键盘上找到tab，多敲几次，感受到它的存在。</p>
<p>后面，我还会不断强调它，直到它刻在你的脑子里。</p>
<h2 id="操作迁移"><a href="#操作迁移" class="headerlink" title="操作迁移"></a>操作迁移</h2><p>接下来，我们来讲讲这次课程的主要内容，<strong>操作迁移</strong>。</p>
<p>我们在Windows上最常做的操作是什么呢？我觉得，最常见的操作应该是下面这些吧</p>
<ul>
<li>浏览目录</li>
<li>切换目录</li>
<li>新建文件夹&#x2F;删除文件夹</li>
<li>新建文件&#x2F;删除文件</li>
<li>查看文本</li>
<li>了解资源使用情况</li>
<li>数据下载</li>
<li>安装软件</li>
</ul>
<p>那么这些操作在Linux里的对应命令是什么呢？我们先浏览一下命令，然后通过一个模拟的项目来一个个了解他们。</p>
<ul>
<li>浏览目录: ls</li>
<li>切换目录: cd, pwd</li>
<li>新建文件夹&#x2F;删除文件夹: mkdir &#x2F; rmdir &#x2F; rm -r</li>
<li>新建文件&#x2F;删除文件: vim &#x2F; rm</li>
<li>查看文本: less, head, tail, cat</li>
<li>了解资源使用情况: top</li>
<li>数据下载: wget &#x2F; curl</li>
<li>安装软件: apt-get&#x2F;yum&#x2F;conda</li>
</ul>
<p>那么这些命令应该怎么用呢？其实很简单，就是输入命令名，后面的参数，看情况加。强调一点，如果要用到参数，那么命令名和参数之间是需要空格分隔的。</p>
<p>我们以一个项目为例，去学习使用最常见的Linux命令。这个任务内容如下</p>
<p>1, 在家目录下<strong>创建</strong>文件夹，名为 abc</p>
<p>2, 将abc<strong>重命名</strong>为 study</p>
<p>3, 下载拟南芥的注释GFF文件， 解压缩</p>
<p>4, 使用less<strong>查看</strong>文件内容</p>
<p>5, 删除gff 文件</p>
<p>6, <strong>删除</strong> study<strong>文件夹</strong></p>
<p>具体操作阅读原文看我的视频。大约在5分20秒处</p>
<h2 id="环境变量"><a href="#环境变量" class="headerlink" title="环境变量"></a>环境变量</h2><p>环境变量也是Linux学习中非常重要的知识点，不了解他甚至都不能用好软件。</p>
<p>不过环境变量属于哪种你不知道，你觉得很高级，一旦知道后，却发现很简单的存在。其实我们从小学或者初中开始就通过数学了解到它。比如说x  + 3 &#x3D;5， x+ 4&#x3D;5，其中的x就是变量，就是会变的量。</p>
<p>在计算机里，它就用来存储其他的值一个名字，相当于一个中介。</p>
<p>变量有可以分为环境变量和局部变量，环境变量就是比局部变量作用更广泛的一个变量。举个例子，当我说到爱因斯坦这个名字时，这会让你想到的是一个物理学家，但是如果你家里有一只猫，它也叫爱因斯坦，他就是一个局部变量。局部变量可以覆盖环境变量。</p>
<p>接下来我们将会通过一个例子，通过PATH这个环境变量来理解。</p>
<ul>
<li><p>查看环境变量PATH</p>
</li>
<li><p>新建一个目录存放软件</p>
</li>
<li><p>将目录加入环境变量PATH</p>
</li>
<li><p>下载软件</p>
</li>
<li><p>给软件添加执行权限</p>
</li>
<li><p>测试软件</p>
</li>
</ul>
<p>具体操作阅读原文看我的视频。在15分钟的时候。</p>
<p>视频我传到了哔哩哔哩，假如你不喜欢哔哩哔哩，想要下载我的2880 x 1800（大约3G）的视频，这是百度盘地址，链接: <a target="_blank" rel="noopener" href="https://pan.baidu.com/s/17yxk78_kpeL5BA-3P0i06Q">https://pan.baidu.com/s/17yxk78_kpeL5BA-3P0i06Q</a>  密码:7ro0。相信我，你哔哩哔哩已经够高清了，你真的不需要20分钟3G的视频。</p>
<p>对于入门而言，知道的越少越好，因为细节无穷无尽。你先要上手，会敲命令了。那么后期继续深入反而就只是时间问题了。</p>
<p>之后，我会更新一系列入门提高的视频，介绍Linux的各种细节，比如说，</p>
<ul>
<li>括号的用法</li>
<li>变量操作</li>
<li>shell脚本</li>
<li>…</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/07/%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9-%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9/" data-id="cllszh645007ne2qe0fnzh5xy" data-title="「生信基础课」初学者入门Linux最少必要的知识点" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-从零开始学CIRCOS绘制圈图(五)-从零开始学CIRCOS绘制圈图(五)" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/31/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)/" class="article-date">
  <time class="dt-published" datetime="2019-07-31T13:11:18.509Z" itemprop="datePublished">2019-07-31</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/31/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)/">从零开始学CIRCOS绘制圈图(五)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>这一部分承接<a href="/archives/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E5%9B%9B)">从零开始学CIRCOS绘制圈图(四)</a>，对之前的布局进行调整，使结构更加适合于发表。</p>
<h2 id="染色体的位置顺序"><a href="#染色体的位置顺序" class="headerlink" title="染色体的位置顺序"></a>染色体的位置顺序</h2><p>默认情况下是会显示所有的染色体，而且会先从ath1到ath5,然后从aly1到aly8, 那么如何调整顺序呢？</p>
<p>需要设置的参数是<code>chromosomes_order</code>,  按照自己的需求调整位置。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chromosomes_order = aly8,aly7,aly6,aly5,aly4,aly3,aly2,aly1</span><br></pre></td></tr></table></figure>

<p>虽然circos提供了一些快捷操作，比如说<code>-</code>,<code>$</code>,<code>^</code>, 但是远不如直接输入直观。</p>
<h2 id="反转染色体的坐标顺序"><a href="#反转染色体的坐标顺序" class="headerlink" title="反转染色体的坐标顺序"></a>反转染色体的坐标顺序</h2><p>默认都是从0到最大值，可以通过正则匹配的方式，将aly的染色体改成从最大值到0</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chromosomes_reverse = /aly/</span><br></pre></td></tr></table></figure>

<h2 id="调整染色体缩放"><a href="#调整染色体缩放" class="headerlink" title="调整染色体缩放"></a>调整染色体缩放</h2><p>为了让两个物种的基因组能够占Circos的两边，需要设置缩放。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chromosomes_scale = /aly/=0.5rn,/ath/=0.5rn</span><br></pre></td></tr></table></figure>

<p><code>/aly/=0.5rn</code>表示aly的所有染色体总共占据50%的空间</p>
<h2 id="定义染色体的间隔"><a href="#定义染色体的间隔" class="headerlink" title="定义染色体的间隔"></a>定义染色体的间隔</h2><p> <code>&lt;spcaing&gt;</code>控制的是不同染色体之间间隔的空隙大小，比如说下面的参数表示所有染色体之间的间隔是0.005r。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&lt;spacing&gt;</span><br><span class="line">default = 0.005r</span><br><span class="line">&lt;/spacing&gt;</span><br></pre></td></tr></table></figure>

<p>我们设置某些染色体之间的距离大一些，比如说两个物种的起始染色体和两个物种的终止染色体</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">&lt;spacing&gt;</span><br><span class="line">default = 1u</span><br><span class="line"></span><br><span class="line">  &lt;pairwise aly1 ath1&gt;</span><br><span class="line">   spacing = 10u</span><br><span class="line">  &lt;/pairwise&gt;</span><br><span class="line">  </span><br><span class="line">  &lt;pairwise aly8 ath5&gt;</span><br><span class="line">   spacing = 10u</span><br><span class="line">  &lt;/pairwise&gt;</span><br><span class="line">  </span><br><span class="line">&lt;/spacing&gt;</span><br></pre></td></tr></table></figure>

<h2 id="调整起始位置"><a href="#调整起始位置" class="headerlink" title="调整起始位置"></a>调整起始位置</h2><p>默认情况下，ath1从-90度位置开始，由于设置了aly1和ath1之间的空隙为10u，于是看起来就不对称了。为了让结果图能够对错，需要调整角度。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">angle_offset* = -85</span><br></pre></td></tr></table></figure>

<p>最终效果</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/circos-final-da04fc42d5984d8e8205ce9d95a849de.png" alt="circos-final"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/07/31/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)/" data-id="cllszh64k0095e2qegent4un4" data-title="从零开始学CIRCOS绘制圈图(五)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-从零开始学CIRCOS绘制圈图(四)-从零开始学CIRCOS绘制圈图(四)" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/31/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E5%9B%9B)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E5%9B%9B)/" class="article-date">
  <time class="dt-published" datetime="2019-07-31T10:04:13.139Z" itemprop="datePublished">2019-07-31</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/31/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E5%9B%9B)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E5%9B%9B)/">从零开始学CIRCOS绘制圈图(四)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>通常circos的中间部分不是空白区域，会用一条条线进行连接，表示两个染色体部分区域有关系。</p>
<h2 id="数据格式"><a href="#数据格式" class="headerlink" title="数据格式"></a>数据格式</h2><p>对于link，circos要求输入数据至少有6列，分别是<code>chr1 start1 end1 chr2 start2 end2 [options]</code></p>
<p>举个例子</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chr1	1000000	2000000	chr5	3000000	4000000</span><br></pre></td></tr></table></figure>

<h2 id="构建输入"><a href="#构建输入" class="headerlink" title="构建输入"></a>构建输入</h2><p>这次会以A. lyrata 和 A.thalina的基因组为例，利用JCVI来构建Circos的输入.</p>
<p>新建一个项目文件夹</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p ath_aly &amp;&amp; <span class="built_in">cd</span> ath_aly</span><br></pre></td></tr></table></figure>

<p>如果没有安装jcvi，用conda进行安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">conda create -n jcvi jcvi</span><br><span class="line">conda activate jcvi</span><br><span class="line">conda install last scipy</span><br><span class="line">wget https://raw.githubusercontent.com/tanghaibao/jcvi-bin/master/bin/scip</span><br><span class="line"><span class="built_in">chmod</span> 755 scip</span><br><span class="line"><span class="built_in">mv</span> scip ~/miniconda3/bin</span><br></pre></td></tr></table></figure>

<h3 id="数据下载"><a href="#数据下载" class="headerlink" title="数据下载"></a>数据下载</h3><p>从<a target="_blank" rel="noopener" href="http://plants.ensembl.org/index.html">http://plants.ensembl.org/index.html</a>分别下载这两个物种的GFF文件和CDS序列， </p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Athaliana</span></span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/arabidopsis_thaliana/cds/Arabidopsis_thaliana.TAIR10.cds.all.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.44.gff3.gz</span><br><span class="line"><span class="comment"># Alyrata</span></span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/fasta/arabidopsis_lyrata/cds/Arabidopsis_lyrata.v.1.0.cds.all.fa.gz</span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/gff3/arabidopsis_lyrata/Arabidopsis_lyrata.v.1.0.44.gff3.gz</span><br></pre></td></tr></table></figure>

<h3 id="数据预处理"><a href="#数据预处理" class="headerlink" title="数据预处理"></a>数据预处理</h3><p>将GFF3转成BED格式</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.formats.gff bed --<span class="built_in">type</span>=mRNA --key=transcript_id Arabidopsis_thaliana.TAIR10.44.gff3.gz &gt; ath.bed</span><br><span class="line">python -m jcvi.formats.gff bed --<span class="built_in">type</span>=mRNA --key=transcript_id Arabidopsis_lyrata.v.1.0.44.gff3.gz  &gt; aly.bed</span><br></pre></td></tr></table></figure>

<p>将BED去重复</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.formats.bed <span class="built_in">uniq</span> ath.bed</span><br><span class="line">python -m jcvi.formats.bed <span class="built_in">uniq</span> aly.bed</span><br></pre></td></tr></table></figure>

<p>提取CDS序列</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">seqkit grep -f &lt;(<span class="built_in">cut</span> -f 4 ath.bed ) Athaliana_167_TAIR10.cds.fa.gz | seqkit <span class="built_in">seq</span> -i &gt; ath.cds</span><br><span class="line">seqkit grep -f &lt;(<span class="built_in">cut</span> -f 4 aly.bed )  Arabidopsis_lyrata.v.1.0.cds.all.fa.gz | seqkit <span class="built_in">seq</span> -i &gt; aly.cds</span><br></pre></td></tr></table></figure>

<h3 id="karyotype"><a href="#karyotype" class="headerlink" title="karyotype"></a>karyotype</h3><p> 从ENSEMBLE下载的GFF文件中，已经包含了每个基因组的大小, 当然也可以各种工具从基因组序列序列中获取大小。</p>
<p>karyotype.aly.txt</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">chr	-	aly1	aly1	0	33132539	rdylbu-11-div-1</span><br><span class="line">chr	-	aly2	aly2	0	19320864	rdylbu-11-div-2</span><br><span class="line">chr	-	aly3	aly3	0	24464547	rdylbu-11-div-3</span><br><span class="line">chr	-	aly4	aly4	0	23328337	rdylbu-11-div-4</span><br><span class="line">chr	-	aly5	aly5	0	21221946	rdylbu-11-div-5</span><br><span class="line">chr	-	aly6	aly6	0	25113588	rdylbu-11-div-6</span><br><span class="line">chr	-	aly7	aly7	0	24649197	rdylbu-11-div-7</span><br><span class="line">chr	-	aly8	aly8	0	22951293	rdylbu-11-div-8</span><br></pre></td></tr></table></figure>

<p>karyotype.tair10.txt</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">chr	-	ath1	ath1	0	30427671	brbg-10-div-1</span><br><span class="line">chr	-	ath2	ath2	0	19698289	brbg-10-div-3</span><br><span class="line">chr	-	ath3 	ath3	0	23459830	brbg-10-div-5</span><br><span class="line">chr	-	ath4	ath4	0	18585056	brbg-10-div-7</span><br><span class="line">chr	-	ath5 	ath5	0	26975502	brbg-10-div-9</span><br></pre></td></tr></table></figure>

<p>因为ENSMEBLE上Athalina和Alyrata的染色体命名都是1,2,3…，就会导致CIRCOS无法正确的区分来源，因此在原本的命名前加上了物种名缩写做为标签。</p>
<p>同时。我们要修改之前的bed文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sed -e <span class="string">&#x27;s/^/ath/&#x27;</span> ath.uniq.bed  &gt; ath.bed</span><br><span class="line">sed -e <span class="string">&#x27;s/^/aly/&#x27;</span> aly.uniq.bed  &gt; aly.bed</span><br></pre></td></tr></table></figure>

<h3 id="links"><a href="#links" class="headerlink" title="links"></a>links</h3><p>为了构建links文件，需要利用JCVI进行共线性分析.</p>
<p>确保有4个文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">ls</span> ???.???</span><br><span class="line">aly.bed aly.cds ath.bed ath.cds</span><br></pre></td></tr></table></figure>

<p>用jcvi进行分析</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.compara.catalog ortholog --no_strip_names ath aly</span><br><span class="line">python -m jcvi.compara.synteny screen --minspan=30 --simple ath.aly.anchors ath.aly.anchors.new</span><br></pre></td></tr></table></figure>

<p>其中<code>ath.aly.anchors.simple</code>是我们后续要用到的文件。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">head</span> -n 1 ath.aly.anchors.simple</span><br><span class="line">AT1G24260.1	AT1G27280.1	fgenesh1_pm.C_scaffold_1002045	fgenesh2_kg.1__2877__AT1G24260.1	225	-</span><br></pre></td></tr></table></figure>

<p>我们需要将<code>ath.aly.anchors.simple</code>里的基因名替换成实际的位置信息，将其变成符合Circos的输入信息。</p>
<p>我写了一个simple2links.py脚本，代码在我的GitHub上，<a target="_blank" rel="noopener" href="https://github.com/xuzhougeng/myscripts">https://github.com/xuzhougeng/myscripts</a>。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python ~/myscripts/simple2links.py ath.aly.anchors.simple</span><br></pre></td></tr></table></figure>

<p>最终会输出<code>ath.aly.anchors.simple_link.txt</code></p>
<h2 id="配置circos"><a href="#配置circos" class="headerlink" title="配置circos"></a>配置circos</h2><p>接下来做如下配置。 新建一个etc文件夹，在里面建立一个<code>links.conf</code>用来配置links, 建立一个<code>ticks.conf</code>配置ticks</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> etc</span><br><span class="line"><span class="comment"># vim etc/links.conf</span></span><br><span class="line">&lt;links&gt;</span><br><span class="line"></span><br><span class="line">&lt;<span class="built_in">link</span>&gt;</span><br><span class="line">file          = ath.aly.anchors.simple_link.txt</span><br><span class="line">radius        = 0.61r</span><br><span class="line">color         = blue_a4</span><br><span class="line">ribbon = <span class="built_in">yes</span></span><br><span class="line">&lt;/link&gt;</span><br><span class="line"></span><br><span class="line">&lt;/links&gt;</span><br><span class="line"></span><br><span class="line"><span class="comment"># vim etc/ticks.conf</span></span><br><span class="line"></span><br><span class="line">show_ticks          = <span class="built_in">yes</span></span><br><span class="line">show_tick_labels    = <span class="built_in">yes</span></span><br><span class="line"></span><br><span class="line">&lt;ticks&gt;</span><br><span class="line"></span><br><span class="line">radius           = 1r</span><br><span class="line">color            = black</span><br><span class="line">thickness        = 2p</span><br><span class="line">multiplier       = 1e-6</span><br><span class="line">format           = %d</span><br><span class="line"></span><br><span class="line">&lt;tick&gt;</span><br><span class="line">spacing        = 1u</span><br><span class="line">size           = 5p</span><br><span class="line">&lt;/tick&gt;</span><br><span class="line"></span><br><span class="line">&lt;tick&gt;</span><br><span class="line">thickness      = 4p</span><br><span class="line">spacing        = 5u</span><br><span class="line">size           = 10p</span><br><span class="line">show_label     = <span class="built_in">yes</span></span><br><span class="line">label_size     = 10p</span><br><span class="line">label_offset   = 10p</span><br><span class="line">format         = %d</span><br><span class="line">&lt;/tick&gt;</span><br><span class="line"></span><br><span class="line">&lt;/ticks&gt;</span><br></pre></td></tr></table></figure>

<p>编辑<code>circos.conf</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line">karyotype = karyotype.tair10.txt,karyotype.aly.txt</span><br><span class="line"></span><br><span class="line">chromosomes_color = chr1=rdylbu-11-div-1,chr2=rdylbu-11-div-3,chr3=rdylbu-11-div-5,chr4=rdylbu-11-div-7,chr5=rdylbu-11-div-9</span><br><span class="line"></span><br><span class="line">chromosomes_units = 1000000</span><br><span class="line">&lt;&lt;<span class="string">include ./etc/ticks.conf&gt;&gt;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">&lt;ideogram&gt;</span></span><br><span class="line"><span class="string">&lt;spacing&gt;</span></span><br><span class="line"><span class="string">default = 0.005r</span></span><br><span class="line"><span class="string">&lt;/spacing&gt;</span></span><br><span class="line"><span class="string">radius           = 0.90r</span></span><br><span class="line"><span class="string">thickness        = 20p</span></span><br><span class="line"><span class="string">fill             = yes</span></span><br><span class="line"><span class="string">stroke_color     = dgrey</span></span><br><span class="line"><span class="string">stroke_thickness = 2p</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">show_label     = yes #展示label</span></span><br><span class="line"><span class="string">label_font     = default # 字体</span></span><br><span class="line"><span class="string">label_radius   = dims(ideogram,radius) + 0.08r #位置</span></span><br><span class="line"><span class="string">label_size     = 16 # 字体大小</span></span><br><span class="line"><span class="string">label_parallel = yes # 是否平行</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">label_format   = eval(sprintf(&quot;%s&quot;,var(chr))) # 格式</span></span><br><span class="line"><span class="string">&lt;/ideogram&gt;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">&lt;&lt;include</span> ./etc/links.conf&gt;&gt;</span><br><span class="line"></span><br><span class="line">&lt;image&gt;</span><br><span class="line"><span class="built_in">dir</span>*    = .    <span class="comment"># 输出文件夹</span></span><br><span class="line">radius* = 500p <span class="comment"># 图片半径</span></span><br><span class="line">svg*    = no   <span class="comment"># 是否输出svg</span></span><br><span class="line">&lt;&lt;<span class="string">include etc/image.conf&gt;&gt;</span></span><br><span class="line"><span class="string">&lt;/image&gt;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">&lt;&lt;include</span> etc/colors_fonts_patterns.conf&gt;&gt;</span><br><span class="line">&lt;&lt;<span class="string">include etc/housekeeping.conf&gt;&gt;</span></span><br></pre></td></tr></table></figure>

<p>运行<code>circos -conf circos.conf</code>的效果如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/circos-links-1-90873e10d3304f45a9a8dd715b925588.png" alt="circos-links-1"></p>
<p>不难发现一个问题，里面线的颜色一模一样，不容易进行区分。虽然可以在输入<code>ath.aly.anchors.simple_link.txt</code>里增加颜色，但是circos提供了一个动态规则，可以更加方便的直接在配置文件里修改。</p>
<h2 id="建立规则-rules"><a href="#建立规则-rules" class="headerlink" title="建立规则(rules)"></a>建立规则(rules)</h2><p>circos的配置格式为</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">&lt;rules&gt;</span><br><span class="line"></span><br><span class="line">&lt;rule&gt;</span><br><span class="line">...</span><br><span class="line">&lt;/rule&gt;</span><br><span class="line"></span><br><span class="line">&lt;rule&gt;</span><br><span class="line">...</span><br><span class="line">&lt;/rule&gt;</span><br><span class="line">...</span><br><span class="line"></span><br><span class="line">&lt;/rules&gt;</span><br></pre></td></tr></table></figure>

<p>circos的规则可以很复杂，但是最简单的情况就是下面这种</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&lt;rule&gt;</span><br><span class="line">condition = var(chr1) eq <span class="string">&quot;ath1&quot;</span></span><br><span class="line">color=rdylgn-5-div-1</span><br><span class="line">&lt;/rule&gt;</span><br></pre></td></tr></table></figure>

<p><code>condition = var(chr1) eq &quot;ath1&quot;</code>表示，判断link文件中左侧染色体的名字(<code>var(chr1)</code>)是不是(<code>eq</code>)”ath1”，如果是的话，那么颜色就是rdylgn-5-div-1</p>
<p>我们可以在<code>etc/links.conf</code>中增加五个条件，修改后的<code>links.conf</code>如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">&lt;<span class="built_in">link</span>&gt;</span><br><span class="line">file          = ath.aly.anchors.simple_link.txt</span><br><span class="line">radius        = 0.61r</span><br><span class="line">color         = blue_a4</span><br><span class="line">ribbon = <span class="built_in">yes</span></span><br><span class="line"></span><br><span class="line">&lt;rules&gt;</span><br><span class="line">&lt;rule&gt;</span><br><span class="line">condition = var(chr1) eq <span class="string">&quot;ath1&quot;</span></span><br><span class="line">color=rdylgn-5-div-1</span><br><span class="line">&lt;/rule&gt;</span><br><span class="line">&lt;rule&gt;</span><br><span class="line">condition = var(chr1) eq <span class="string">&quot;ath2&quot;</span></span><br><span class="line">color=rdylgn-5-div-2</span><br><span class="line">&lt;/rule&gt;</span><br><span class="line">&lt;rule&gt;</span><br><span class="line">condition = var(chr1) eq <span class="string">&quot;ath3&quot;</span></span><br><span class="line">color=rdylgn-5-div-3</span><br><span class="line">&lt;/rule&gt;</span><br><span class="line">&lt;rule&gt;</span><br><span class="line">condition = var(chr1) eq <span class="string">&quot;ath4&quot;</span></span><br><span class="line">color=rdylgn-5-div-4</span><br><span class="line">&lt;/rule&gt;</span><br><span class="line">&lt;rule&gt;</span><br><span class="line">condition = var(chr1) eq <span class="string">&quot;ath5&quot;</span></span><br><span class="line">color=rdylgn-5-div-5</span><br><span class="line">&lt;/rule&gt;</span><br><span class="line">&lt;/rules&gt;</span><br><span class="line"></span><br><span class="line">&lt;/link&gt;</span><br><span class="line">&lt;/links&gt;</span><br></pre></td></tr></table></figure>

<p>运行<code>circos -conf circos.conf</code>的效果如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/circos-link-2-2222e6d11bd049c0b5e9a200c5e50508.png" alt="circos-link-2"></p>
<p>这张图还有一个问题是，通常别人的Circos染色体都是对称排列，右边第一个是ath1，那么左边第一个也最好是aly1，那么应该如何配置呢？以及两套基因组会分别占据两侧，中间有一个明显的空隙，如何做到的呢？</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/07/31/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E5%9B%9B)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E5%9B%9B)/" data-id="cllszh64l0098e2qe6o6x6krj" data-title="从零开始学CIRCOS绘制圈图(四)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-从零开始学CIRCOS绘制圈图(三)-从零开始学CIRCOS绘制圈图(三)" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/30/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%89)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%89)/" class="article-date">
  <time class="dt-published" datetime="2019-07-30T16:41:42.657Z" itemprop="datePublished">2019-07-30</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/30/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%89)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%89)/">从零开始学CIRCOS绘制圈图(三)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>这一篇会在之前的基础上开始在circos绘制基因密度信息。为了保证一致性，可以新建如下几个文件</p>
<p><code>circos.conf</code>:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br></pre></td><td class="code"><pre><span class="line">karyotype = karyotype.tair10.txt</span><br><span class="line"></span><br><span class="line">chromosomes_color = chr1=rdylbu-11-div-1,chr2=rdylbu-11-div-3,chr3=rdylbu-11-div-5,chr4=rdylbu-11-div-7,chr5=rdylbu-11-div-9</span><br><span class="line"></span><br><span class="line">chromosomes_units = 1000000</span><br><span class="line">&lt;&lt;<span class="string">include ticks.conf&gt;&gt;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">&lt;ideogram&gt;</span></span><br><span class="line"><span class="string">&lt;spacing&gt;</span></span><br><span class="line"><span class="string">default = 0.005r</span></span><br><span class="line"><span class="string">&lt;/spacing&gt;</span></span><br><span class="line"><span class="string">radius           = 0.90r</span></span><br><span class="line"><span class="string">thickness        = 20p</span></span><br><span class="line"><span class="string">fill             = yes</span></span><br><span class="line"><span class="string">stroke_color     = dgrey</span></span><br><span class="line"><span class="string">stroke_thickness = 2p</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">show_label     = yes #展示label</span></span><br><span class="line"><span class="string">label_font     = default # 字体</span></span><br><span class="line"><span class="string">label_radius   = dims(ideogram,radius) + 0.05r #位置</span></span><br><span class="line"><span class="string">label_size     = 16 # 字体大小</span></span><br><span class="line"><span class="string">label_parallel = yes # 是否平行</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">label_format   = eval(sprintf(&quot;%s&quot;,var(chr))) # 格式</span></span><br><span class="line"><span class="string">&lt;/ideogram&gt;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">&lt;image&gt;</span></span><br><span class="line"><span class="string">dir*    = .    # 输出文件夹</span></span><br><span class="line"><span class="string">radius* = 500p # 图片半径</span></span><br><span class="line"><span class="string">svg*    = no   # 是否输出svg</span></span><br><span class="line"><span class="string">&lt;&lt;include</span> etc/image.conf&gt;&gt;</span><br><span class="line">&lt;/image&gt;</span><br><span class="line"></span><br><span class="line">&lt;&lt;<span class="string">include etc/colors_fonts_patterns.conf&gt;&gt;</span></span><br><span class="line"><span class="string">&lt;&lt;include</span> etc/housekeeping.conf&gt;&gt;</span><br></pre></td></tr></table></figure>

<p><code>karyotype.tair10.txt</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">chr - chr1 chr1 0 30427617 chr1</span><br><span class="line">chr - chr2 chr2 0 19698289 chr2</span><br><span class="line">chr - chr3 chr3 0 23459830 chr3</span><br><span class="line">chr - chr4 chr4 0 18585056 chr4</span><br><span class="line">chr - chr5 chr5 0 26975502 chr5</span><br></pre></td></tr></table></figure>

<p><code>ticks.conf</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">show_ticks          = <span class="built_in">yes</span></span><br><span class="line">show_tick_labels    = <span class="built_in">yes</span></span><br><span class="line"></span><br><span class="line">&lt;ticks&gt;</span><br><span class="line"></span><br><span class="line">radius           = 1r</span><br><span class="line">color            = black</span><br><span class="line">thickness        = 2p</span><br><span class="line">multiplier       = 1e-6</span><br><span class="line">format           = %d</span><br><span class="line"></span><br><span class="line">&lt;tick&gt;</span><br><span class="line">spacing        = 1u</span><br><span class="line">size           = 5p</span><br><span class="line">&lt;/tick&gt;</span><br><span class="line"></span><br><span class="line">&lt;tick&gt;</span><br><span class="line">thickness      = 4p</span><br><span class="line">spacing        = 5u</span><br><span class="line">size           = 10p</span><br><span class="line">show_label     = <span class="built_in">yes</span></span><br><span class="line">label_size     = 10p</span><br><span class="line">label_offset   = 10p</span><br><span class="line">format         = %d</span><br><span class="line">&lt;/tick&gt;</span><br><span class="line"></span><br><span class="line">&lt;/ticks&gt;</span><br></pre></td></tr></table></figure>

<p>在开始之前，请确保已经安装了bedtools，如果没有的话，用conda安装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 安装bedtools</span></span><br><span class="line">conda install -c bioconda bedtools</span><br></pre></td></tr></table></figure>

<h2 id="数据格式"><a href="#数据格式" class="headerlink" title="数据格式"></a>数据格式</h2><p>为了能够在circos绘制基因密度信息，需要先知道circos要求的输入数据格式是什么。</p>
<p>对于折线图(line)，散点图(scatter)，柱状图(histogram)和热图(heatmap)，Circos要求的数据输入格式相同，也就是<code>chr start end value [options]</code>,  如果熟悉BED格式定义的话，你就会发现除了可选(options)外，Circos要求的格式就是4列的BED。</p>
<p>对于可选列，可以和<code>circos.conf</code>里的<code>&lt;rule&gt;</code>搭配使用，属于比较高级的用法。</p>
<h2 id="数据预处理"><a href="#数据预处理" class="headerlink" title="数据预处理"></a>数据预处理</h2><p>为了能够获得所需的基因密度信息，我们需要下载拟南芥的GFF文件。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># download</span></span><br><span class="line">wget ftp://ftp.ensemblgenomes.org/pub/plants/release-44/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.44.gff3.gz</span><br></pre></td></tr></table></figure>

<p>提取基因的位置信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">zgrep <span class="string">&#x27;[[:blank:]]gene[[:blank:]]&#x27;</span> Arabidopsis_thaliana.TAIR10.44.gff3.gz | <span class="built_in">cut</span> -f 1,4,5 | awk <span class="string">&#x27;&#123;print &quot;chr&quot;$1&quot;\t&quot;$2&quot;\t&quot;$3&#125;&#x27;</span> &gt; genes.bed</span><br></pre></td></tr></table></figure>

<p>接着用bedtools以500kb为滑窗，沿染色体创建窗口</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cut</span> -d <span class="string">&#x27; &#x27;</span> -f 3,6 karyotype.tair10.txt | <span class="built_in">tr</span> <span class="string">&#x27; &#x27;</span> <span class="string">&#x27;\t&#x27;</span> &gt; tair10.genome</span><br><span class="line">bedtools makewindows -g tair10.genome -w 500000 &gt; tair10.windows</span><br></pre></td></tr></table></figure>

<p>最后统计信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bedtools coverage -a tair10.windows -b genes.bed | <span class="built_in">cut</span> -f 1-4 &gt; genes_num.txt</span><br></pre></td></tr></table></figure>

<h2 id="展示信息"><a href="#展示信息" class="headerlink" title="展示信息"></a>展示信息</h2><p>我们可以先用最少的参数，同时展示不同的图形。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br></pre></td><td class="code"><pre><span class="line">...</span><br><span class="line">&lt;plots&gt;</span><br><span class="line"></span><br><span class="line">&lt;plot&gt;</span><br><span class="line">type      = line</span><br><span class="line">thickness = 2</span><br><span class="line">max_gap = 1u</span><br><span class="line">file = genes_num.txt</span><br><span class="line">color   = redv</span><br><span class="line">r0      = 0.51r</span><br><span class="line">r1      = 0.60r</span><br><span class="line">&lt;/plot&gt;</span><br><span class="line"></span><br><span class="line">&lt;plot&gt;</span><br><span class="line">type    = heatmap</span><br><span class="line">file    = genes_num.txt</span><br><span class="line">color   = spectral-5-div</span><br><span class="line">r1      = 0.70r</span><br><span class="line">r0      = 0.61r</span><br><span class="line">&lt;/plot&gt;</span><br><span class="line"></span><br><span class="line">&lt;plot&gt;</span><br><span class="line">type = scatter</span><br><span class="line">fill_color       = grey</span><br><span class="line">stroke_color     = black</span><br><span class="line">glyph            = circle</span><br><span class="line">glyph_size       = 10</span><br><span class="line">file = genes_num.txt</span><br><span class="line">r1   = 0.80r</span><br><span class="line">r0   = 0.71r</span><br><span class="line">&lt;/plot&gt;</span><br><span class="line"></span><br><span class="line">&lt;plot&gt;</span><br><span class="line">type = histogram</span><br><span class="line">file = genes_num.txt</span><br><span class="line">r1   = 0.89r</span><br><span class="line">r0   = 0.81r</span><br><span class="line">&lt;/plot&gt;</span><br><span class="line"></span><br><span class="line">&lt;/plots&gt;</span><br><span class="line">...</span><br></pre></td></tr></table></figure>

<p>运行之后，效果如下</p>
<p> <img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/circos-gene-num-1-6fb138709d0a4cc6b7e1eab37e639c6f.png" alt="初步效果"></p>
<p>这张图展示了巨大的进步空间，至少可以从如下几个角度进行调整，</p>
<ul>
<li>柱状图有点丑，需要填充颜色</li>
<li>热图的颜色不符合直觉，基因数越少反而颜色越深</li>
<li>散点图的点太大</li>
<li>有些图形可能需要加上背景色。</li>
<li>对于一些过大的数值，最好用一种颜色表示。</li>
</ul>
<p>当知道自己的目标后，后续的事情就是找对应参数和调整参数. 部分我用来调整参数如下</p>
<ul>
<li><code>show</code> - 是否展示该图形</li>
<li><code>type</code> - 展示的图形类型</li>
<li><code>file</code> - 输入的数据文件所在路径</li>
<li><code>min/max</code> - 数据范围</li>
<li><code>r0/r1</code> - 内径和外径，在圈图中的位置</li>
<li><code>glyph</code> - 对于<strong>散点图</strong>而言，还可以选择符号的类型，是circle, rectangle, 还是 triangle</li>
<li><code>glyph_size</code> - 符号的大小，单位为p</li>
<li><code>color</code> 散点图符号颜色，柱状图外部线的颜色</li>
<li><code>stroke_color</code> - 对于散点图，符号外部是否也要颜色</li>
<li><code>stroke_thickness</code> - 对于散点图，符号外部线的厚度</li>
<li><code>fill_color</code>:柱状图填充色</li>
</ul>
<p>调整后的参数为</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">...</span><br><span class="line">&lt;plots&gt;</span><br><span class="line"></span><br><span class="line">&lt;plot&gt;</span><br><span class="line"><span class="built_in">type</span>    = heatmap</span><br><span class="line">file    = genes_num.txt</span><br><span class="line">color   = blues-9-seq</span><br><span class="line">r1      = 0.70r</span><br><span class="line">r0      = 0.61r</span><br><span class="line">&lt;/plot&gt;</span><br><span class="line"></span><br><span class="line">&lt;plot&gt;</span><br><span class="line"><span class="built_in">type</span> = scatter</span><br><span class="line">fill_color       = black <span class="comment"># 填充色</span></span><br><span class="line">stroke_color     = black</span><br><span class="line">glyph            = circle</span><br><span class="line">glyph_size       = 5 <span class="comment"># 元素大小</span></span><br><span class="line">file = genes_num.txt</span><br><span class="line">r1   = 0.80r</span><br><span class="line">r0   = 0.71r</span><br><span class="line">&lt;/plot&gt;</span><br><span class="line"></span><br><span class="line">&lt;plot&gt;</span><br><span class="line"><span class="built_in">type</span> = histogram</span><br><span class="line">file = genes_num.txt</span><br><span class="line">fill_color = blue <span class="comment"># 填充色</span></span><br><span class="line">r1   = 0.89r</span><br><span class="line">r0   = 0.81r</span><br><span class="line">&lt;/plot&gt;</span><br><span class="line"></span><br><span class="line">&lt;/plots&gt;</span><br><span class="line">...</span><br></pre></td></tr></table></figure>

<p>图形效果如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/circos-data-vis-068b36c26fdc4663bf79b977830f6cdf.png" alt="circos-data-vis"></p>
<p>这个效果我个人还是比较满意的。更加复杂的设置，目前还不适合我，需要一步一步来。</p>
<hr>
<p>下一部分将会介绍如何在circos展示两个物种的共线性，也就是circos不同染色体之间的连线。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/07/30/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%89)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%89)/" data-id="cllszh64j008ze2qe4l343ttb" data-title="从零开始学CIRCOS绘制圈图(三)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-从零开始学CIRCOS绘制圈图(二)-从零开始学CIRCOS绘制圈图(二)" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/30/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%8C)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%8C)/" class="article-date">
  <time class="dt-published" datetime="2019-07-30T10:52:23.083Z" itemprop="datePublished">2019-07-30</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/30/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%8C)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%8C)/">从零开始学CIRCOS绘制圈图(二)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>在<a href="/archives/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6Circos%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%80)">从零开始学CIRCOS绘制圈图(一)</a>， 我们已经绘制出一个比较丑的circos图，这一部分是讲解一些细节。</p>
<p>这一部分会从上一步的两个文件开始，分别是</p>
<p><code>karyotype.tair10.txt</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">chr - chr1 chr1 0 30427617 chr1</span><br><span class="line">chr - chr2 chr2 0 19698289 chr2</span><br><span class="line">chr - chr3 chr3 0 23459830 chr3</span><br><span class="line">chr - chr4 chr4 0 18585056 chr4</span><br><span class="line">chr - chr5 chr5 0 26975502 chr5</span><br></pre></td></tr></table></figure>

<p><code>circos.conf</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">karyotype = karyotype.tair10.txt</span><br><span class="line"></span><br><span class="line">&lt;ideogram&gt;</span><br><span class="line">&lt;spacing&gt;</span><br><span class="line">default = 0.005r</span><br><span class="line">&lt;/spacing&gt;</span><br><span class="line">radius           = 0.90r</span><br><span class="line">thickness        = 20p</span><br><span class="line">fill             = <span class="built_in">yes</span></span><br><span class="line">stroke_color     = dgrey</span><br><span class="line">stroke_thickness = 2p</span><br><span class="line">&lt;/ideogram&gt;</span><br><span class="line"></span><br><span class="line">&lt;image&gt;</span><br><span class="line">&lt;&lt;<span class="string">include etc/image.conf&gt;&gt;</span></span><br><span class="line"><span class="string">&lt;/image&gt;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">&lt;&lt;include</span> etc/colors_fonts_patterns.conf&gt;&gt;</span><br><span class="line">&lt;&lt;<span class="string">include etc/housekeeping.conf&gt;&gt;</span></span><br></pre></td></tr></table></figure>

<h2 id="配色"><a href="#配色" class="headerlink" title="配色"></a>配色</h2><p>第一个要解决的问题是配色问题，其实上一部分的配色定义依旧有点丑。如果不想重新打开文件修改的话，其实除了在karyotype文件里定义颜色外，我们还可以直接在circos.conf文件定义颜色.</p>
<p>举个例子，在<code>karyotype = karyotype.tair10.txt</code>后加一行</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chromosomes_color = chr1=rdylbu-11-div-1,chr2=rdylbu-11-div-3,chr3=rdylbu-11-div-5,chr4=rdylbu-11-div-7,chr5=rdylbu-11-div-9</span><br></pre></td></tr></table></figure>

<p>问题是，我们怎们知道这些名字后所代表的颜色呢？</p>
<p>Circos中颜色的命名格式为<code>PALETTE-NUMCOLORS-TYPE-IDX</code>:</p>
<ul>
<li>PALETTE:调色版名，如rdylbu</li>
<li>NUMCOLORS: 颜色数目,  11</li>
<li>调色版类型: div(diverging), seq(sequential), qual(qualitative)</li>
<li>IDX: 调色版中的颜色索引</li>
</ul>
<p>而Circos颜色来自于<a target="_blank" rel="noopener" href="http://colorbrewer2.org/">http://colorbrewer2.org</a></p>
<p>因此，<code>gnbu-9-seq</code>对应的是就是下图的<code>9-class GnBu</code></p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/circos-colors-88bba063fa264c9f9825a9b9fd3baad0.png" alt="颜色"></p>
<p>配色不仅仅用在karyotype中，在后续的热图和柱状图等还会涉及到它，毕竟一张好看的图，配色占了很大的比例。</p>
<h2 id="显示标签"><a href="#显示标签" class="headerlink" title="显示标签"></a>显示标签</h2><p>默认输出图片是没有染色体名字的标签，需要在<code>&lt;ideogram&gt;</code>里添加和<code>label</code>有关的参数</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">...</span><br><span class="line">&lt;/spacing&gt;</span><br><span class="line">...</span><br><span class="line">show_label     = <span class="built_in">yes</span> <span class="comment">#展示label</span></span><br><span class="line">label_font     = default <span class="comment"># 字体</span></span><br><span class="line">label_radius   = dims(ideogram,radius) + 0.05r <span class="comment">#位置</span></span><br><span class="line">label_size     = 16 <span class="comment"># 字体大小</span></span><br><span class="line">label_parallel = <span class="built_in">yes</span> <span class="comment"># 是否平行</span></span><br><span class="line"></span><br><span class="line">label_format   = <span class="built_in">eval</span>(sprintf(<span class="string">&quot;%s&quot;</span>,var(chr))) <span class="comment"># 格式</span></span><br><span class="line">&lt;/ideogram&gt;</span><br></pre></td></tr></table></figure>

<p>关于标签(label)， 更详细的介绍在<a target="_blank" rel="noopener" href="http://circos.ca/documentation/tutorials/ideograms/labels/">http://circos.ca/documentation/tutorials/ideograms/labels/</a></p>
<p>重新运行之后，发现字相对而言太小了。有两种结局方案，一种是调整<code>label_size</code>，比如说48，另一种是调整图片整体大小。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/circos-small-label-8fa0fec44a434954be54e07f8ce85ba8.png" alt="显示标签"></p>
<h2 id="输出设置"><a href="#输出设置" class="headerlink" title="输出设置"></a>输出设置</h2><p>默认情况下，输出图片的 半径是1500p, 所以设置的<code>label_size=16</code>就会显得特别小。我们可以在配置文件中调整图片的班级，以及其他设置。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">&lt;image&gt;</span><br><span class="line"><span class="built_in">dir</span>*    = .    <span class="comment"># 输出文件夹</span></span><br><span class="line">radius* = 500p <span class="comment"># 图片半径</span></span><br><span class="line">svg*    = no   <span class="comment"># 是否输出svg</span></span><br><span class="line">&lt;&lt;<span class="string">include etc/image.conf&gt;&gt;</span></span><br><span class="line"><span class="string">&lt;/image&gt;</span></span><br></pre></td></tr></table></figure>

<p><strong>注</strong>: 在参数后加一个<code>*</code>表示覆盖已有的设置，比如说<code>svg*=no</code>就是覆盖已有的<code>svg=yes</code>。</p>
<p>运行结果如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/circos-large-label-3f6597855c3641b7baf918b7cc5fd558.png" alt="修改输出图片大小"></p>
<h2 id="刻度-ticks"><a href="#刻度-ticks" class="headerlink" title="刻度(ticks)"></a>刻度(ticks)</h2><p>大部分的circos还会有刻度来展示染色体的大小。由于ticks的定制比较复杂，所以一般会单独搞一个配置文件，<code>ticks.conf</code>存放ticks相关的参数设置.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">chromosomes_units = 1000000</span><br><span class="line">show_ticks          = <span class="built_in">yes</span></span><br><span class="line">show_tick_labels    = <span class="built_in">yes</span></span><br><span class="line"></span><br><span class="line">&lt;ticks&gt;</span><br><span class="line"></span><br><span class="line">radius           = 1r</span><br><span class="line">color            = black</span><br><span class="line">thickness        = 2p</span><br><span class="line"></span><br><span class="line">multiplier       = 1e-6 <span class="comment">#输出的标签为实际长度与其相乘</span></span><br><span class="line"></span><br><span class="line">format           = %d <span class="comment"># %d表示显示整数</span></span><br><span class="line"></span><br><span class="line">&lt;tick&gt;</span><br><span class="line">spacing        = 1u</span><br><span class="line">size           = 5p</span><br><span class="line">&lt;/tick&gt;</span><br><span class="line"></span><br><span class="line">&lt;tick&gt;</span><br><span class="line">spacing        = 5u</span><br><span class="line">size           = 10p</span><br><span class="line">show_label     = <span class="built_in">yes</span></span><br><span class="line">label_size     = 10p</span><br><span class="line">label_offset   = 10p</span><br><span class="line">format         = %d</span><br><span class="line">&lt;/tick&gt;</span><br><span class="line"></span><br><span class="line">&lt;/ticks&gt;</span><br></pre></td></tr></table></figure>

<p>用 <code>show_ticks</code>和 <code>show_tick_labels</code>控制是否展示刻度，以及刻度对应的标签。</p>
<p><code>&lt;ticks</code>与<code>&lt;/ticks&gt;</code>里控制刻度的全局参数，例如位置为1r(radius&#x3D;1r), 颜色为黑色(color&#x3D;black), 厚度为2p(thickness&#x3D;2p),   由于默认直接展示染色体的实际位置，因此会显示1000000这种结果，所以定义<code>multiplier=1e-6</code>,  实际显示结果为 1000000 * 1e-6 &#x3D; 1。</p>
<p>后面就可以通过<code>&lt;tick&gt;</code>和 <code>&lt;/tick&gt;</code>来分别绘制不同类型的tick，重要的参数如下：</p>
<ul>
<li><code>spacing</code>表示刻度之间的距离，1u表示一个长度单位，需要在<code>circos.conf</code>文件里通过<code>chromosome_unit</code>来定义，通常都是<code>chromosome_unit=1000000</code>。</li>
<li><code>size</code>表示tick的长度</li>
<li><code>show_label</code>: 控制是否展示标签，默认不展示。</li>
<li><code>label_offset</code>: 则是让label往外在偏移一点距离</li>
</ul>
<p><code>circos -conf circos.conf</code>运行结果如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/circos-ticks-c284af9cff484548983579dcee75c154.png" alt="增加刻度"></p>
<p>我们发现有些刻度的标签和染色体标签发生了重叠，这个可以通过<code>label_radius</code>进行调整。</p>
<h2 id="单位"><a href="#单位" class="headerlink" title="单位"></a>单位</h2><p>上面出现了控制图形不同元素大小的三个单位，p,r,u。p(pixels), 表示绝对大小， r(relative), 相对大小， u(chromosome unit)。 如果使用p作为单位，需要考虑最终输出图形<code>&lt;image&gt;</code>定义的radius。 而r是相对大小，会随着最终图形大小而发生变换。u一般在显示刻度时使用。</p>
<hr>
<p>这一部分是在原来简陋的输出上进行了美化，没有用到除了染色体长度以外的信息。下一部分介绍如何展示基因密度等信息。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/07/30/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%8C)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%8C)/" data-id="cllszh64k0092e2qe7jk61ug9" data-title="从零开始学CIRCOS绘制圈图(二)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何高效地从BAM文件中提取fastq-如何高效地从BAM文件中提取fastq" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/29/%E5%A6%82%E4%BD%95%E9%AB%98%E6%95%88%E5%9C%B0%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq-%E5%A6%82%E4%BD%95%E9%AB%98%E6%95%88%E5%9C%B0%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq/" class="article-date">
  <time class="dt-published" datetime="2019-07-29T10:05:32.347Z" itemprop="datePublished">2019-07-29</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/29/%E5%A6%82%E4%BD%95%E9%AB%98%E6%95%88%E5%9C%B0%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq-%E5%A6%82%E4%BD%95%E9%AB%98%E6%95%88%E5%9C%B0%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq/">如何高效地从BAM文件中提取fastq</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>在<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/368b9471656e">如何从BAM文件中提取fastq</a>文章里其实也发现了从BAM里面提取Fastq是有些麻烦，只不过最后通过<code>samtools</code>的子命令实现了数据提取，实现功能之后也没有再去思考如何提高效率。</p>
<p>最近读到<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/e866ae780e79">每周文献-190419-植物单细胞BAM重比对以及假基因研究</a>时，发现里面提到了一个工具叫做 <strong>bazam</strong>, 功能就是提取Fastq文件，文章发表在 Genome Biology 上。</p>
<p>软件地址为<a target="_blank" rel="noopener" href="https://github.com/ssadedin/bazam">https://github.com/ssadedin/bazam</a>，因为他是个Groovy工具（据说Groovy比Java好写）所以安装很方便，</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">git <span class="built_in">clone</span> https://github.com/ssadedin/bazam.git</span><br><span class="line"><span class="built_in">cd</span> bazam</span><br><span class="line">git submodule update --init --recursive</span><br><span class="line">./gradlew clean jar</span><br></pre></td></tr></table></figure>

<p>可以通过如下命令检查是否安装成功</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java -jar build/libs/bazam.jar -bam  test.bam  &gt; tmp.fastq</span><br></pre></td></tr></table></figure>

<p>假如你原来的BAM文件是双端测序，想提取成两个文件，那么可以用如下命令。<strong>注</strong>：这里的your.bam 指的是你实际BAM文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">java -Xmx16G -jar build/libs/bazam.jar \</span><br><span class="line">    -bam your.bam -r1 your_R1.fq -r2 your_R2.fq &amp;</span><br></pre></td></tr></table></figure>

<p>此外，bazam还支持使用<code>-L</code>指定区间来提取某个区间的数据。</p>
<blockquote>
<p>尽管是可以从BAM文件中提取Fastq文件，但如果原始BAM文件过滤了未必对的read，那么你还是会损失掉这部分信息，建议保存好原来的Fastq文件。</p>
</blockquote>
<p>这里补充下为啥从BAM文件中提取双端序列那么麻烦，这是因为一般而言BAM文件都是按照位置信息排序，想要找到配对的reads，要么是根据read的编号进行排序（这个方法要求额外的内存和存储空间），或者就是在提取的时候记录当前的read的ID，再找到另一端ID后释放内存空间。</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li><a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1688-1">Bazam: a rapid method for read extraction and realignment of high-throughput sequencing data</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/ssadedin/bazam">https://github.com/ssadedin/bazam</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/07/29/%E5%A6%82%E4%BD%95%E9%AB%98%E6%95%88%E5%9C%B0%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq-%E5%A6%82%E4%BD%95%E9%AB%98%E6%95%88%E5%9C%B0%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq/" data-id="cllszh68g00k9e2qe9zl5b499" data-title="如何高效地从BAM文件中提取fastq" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-如何从BAM文件中提取fastq-如何从BAM文件中提取fastq" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/29/%E5%A6%82%E4%BD%95%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq-%E5%A6%82%E4%BD%95%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq/" class="article-date">
  <time class="dt-published" datetime="2019-07-29T10:03:14.149Z" itemprop="datePublished">2019-07-29</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/29/%E5%A6%82%E4%BD%95%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq-%E5%A6%82%E4%BD%95%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq/">如何从BAM文件中提取fastq</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>虽然高通量测序分析最常用的操作是将fastq比对到参考基因组得到BAM文件，但偶尔我们也需要提取BAM文件中特定区域中fastq。最开始我认为这是一个非常简单的操作，因为samtools其实已经提供了相应的工具<code>samtools fastq</code>. </p>
<p>以biostar handbook的Ebola病毒数据为例，首先获取比对得到的BAM文件。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 建立文件夹</span></span><br><span class="line"><span class="built_in">mkdir</span> -p refs</span><br><span class="line"><span class="comment"># 根据Accession下载</span></span><br><span class="line">ACC=AF086833</span><br><span class="line">REF=refs/<span class="variable">$ACC</span>.fa</span><br><span class="line">efetch -db=nuccore -format=fasta -<span class="built_in">id</span>=<span class="variable">$ACC</span> | seqret -filter -sid <span class="variable">$ACC</span> &gt; <span class="variable">$REF</span></span><br><span class="line">bwa index <span class="variable">$REF</span></span><br><span class="line">SRR=SRR1553500</span><br><span class="line">fastq-dump -X 100000 --split-files <span class="variable">$SRR</span></span><br><span class="line">BAM=<span class="variable">$SRR</span>.bam</span><br><span class="line">R1=<span class="variable">$&#123;SRR&#125;</span>_1.fastq</span><br><span class="line">R2=<span class="variable">$&#123;SRR&#125;</span>_2.fastq</span><br><span class="line">TAG=<span class="string">&quot;@RG\tID:<span class="variable">$SRR</span>\tSM:<span class="variable">$SRR</span>\tLB:<span class="variable">$SRR</span>&quot;</span></span><br><span class="line">bwa mem -R <span class="variable">$TAG</span> <span class="variable">$REF</span> <span class="variable">$R1</span> <span class="variable">$R2</span> | samtools <span class="built_in">sort</span> &gt; <span class="variable">$BAM</span></span><br><span class="line">samtools index <span class="variable">$BAM</span></span><br></pre></td></tr></table></figure>

<p>让我们尝试提取下前1kb比对的fastq文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">samtools view -h SRR1553500.bam KJ660346.1:1-1000 | samtools fastq -1 read_1.fq -2 read_2.fq -</span><br><span class="line"><span class="comment">#[M::bam2fq_mainloop] discarded 0 singletons</span></span><br><span class="line"><span class="comment">#[M::bam2fq_mainloop] processed 8702 reads</span></span><br></pre></td></tr></table></figure>

<p>看起来似乎我们成功了，那让我们把这些序列比对回去吧</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">bwa mem refs/AF086833.fa read_1.fq read_2.fq</span><br><span class="line">[mem_sam_pe] paired reads have different names: &quot;SRR1553500.11772&quot;, &quot;SRR1553500.43775&quot;</span><br></pre></td></tr></table></figure>

<p>你会发现这里产生了错误。这个错误说的是配对的read名字不同。导致问题的原因是：samtools fastq会按照fastq在bam里原本的顺序进行提取，而按照位置排序的bam的两个配对read并不是前后紧挨着的。所以我们需要进行一波排序。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cat</span> read_1.fq \</span><br><span class="line">| <span class="built_in">paste</span> - - - - \</span><br><span class="line">| <span class="built_in">sort</span> -k1,1 -S 3G \</span><br><span class="line">| <span class="built_in">tr</span> <span class="string">&#x27;\t&#x27;</span> <span class="string">&#x27;\n&#x27;</span> \</span><br><span class="line">| gzip &gt; read_1.fq.gz</span><br><span class="line"><span class="built_in">cat</span> read_2.fq \</span><br><span class="line">| <span class="built_in">paste</span> - - - - \</span><br><span class="line">| <span class="built_in">sort</span> -k1,1 -S 3G \</span><br><span class="line">| <span class="built_in">tr</span> <span class="string">&#x27;\t&#x27;</span> <span class="string">&#x27;\n&#x27;</span> \</span><br><span class="line">| gzip &gt; read_2.fq.gz</span><br></pre></td></tr></table></figure>

<p>这里将5个shell命令用管道连起来完成这个任务，咋看起来比较复杂，但是理解每一步的目的，每个参数的作用，以后就能活用起来了。</p>
<p>首先用cat<strong>逐行</strong>读取read，传入到paste中。paste的作用是将多个文件合并成一个文件，做法比较简单粗暴，举个例子</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cat</span> A.txt</span><br><span class="line">A</span><br><span class="line">B</span><br><span class="line">C</span><br><span class="line"><span class="built_in">cat</span> B.txt</span><br><span class="line">D</span><br><span class="line">E</span><br><span class="line">F</span><br><span class="line"><span class="built_in">paste</span> A.txt B.txt</span><br><span class="line">A	D</span><br><span class="line">B	E</span><br><span class="line">C	F</span><br></pre></td></tr></table></figure>

<p><code>cat read_1.fq | paste - - - - </code>表示读取4行合并成一行。相当于fastq里的read都会用一行而不是四行表示。接着用<code>sort</code>按照第一行(-k1,1)排序，内存的缓冲数据为3G(-S 3G). 然后将制表符替换成换行符，让一行又变回到4行，后续是压缩成gz，降低磁盘占用。</p>
<p>这样做其实还没又解决问题，因为这两个文件里的reads数其实不相同，我们需要进一步从里面剔除掉只有一方才有的read。这里使用的<code>seqkit</code>，一个非常强大的序列处理轮子。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">seqkit grep -f &lt;(seqkit <span class="built_in">seq</span> -n -i read_1.fq.gz ) read_2.fq.gz  &gt; read_flt_2.fq</span><br><span class="line">seqkit grep -f &lt;(seqkit <span class="built_in">seq</span> -n -i read_2.fq.gz ) read_1.fq.gz  &gt; read_flt_1.fq</span><br><span class="line"><span class="built_in">wc</span> -l read_flt_1.fq</span><br><span class="line">   15028 read_flt_1.fq</span><br><span class="line"><span class="built_in">wc</span> -l read_flt_2.fq</span><br><span class="line">   15028 read_flt_2.fq</span><br></pre></td></tr></table></figure>

<p>终于，你得到了配对存放的两个fastq文件。</p>
<p>通过这个案例我想表达的观点是：虽然有些时候没有现成的工具可以解决你的问题，但是你可以通过组合几个现有的工具来完成。也就是将一个大问题拆解成几个独立的小问题，然后逐个击破。</p>
<p>当然，你首先得熟悉一些已有的工具，比如说常用unix命令:sort, paste, tr等，和一些非常强大生信命令行工具：samtools, bedtools, seqkit等。</p>
<h2 id="一个更佳方便的方法"><a href="#一个更佳方便的方法" class="headerlink" title="一个更佳方便的方法"></a>一个更佳方便的方法</h2><p>前面只是问题的一种方法，可能还不是最好的，但是我当初最先想到的一个比较粗暴的方法。当然更佳巧妙的方法是，提取序列之后可以按照read name排序，然后提取</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools view -h SRR1553500.bam KJ660346.1:1-1000 | samtools <span class="built_in">sort</span> -n | samtools fastq -1 read_1.fq -2 read_2.fq -s singleton.fq -</span><br></pre></td></tr></table></figure>

<p>是不是更好理解了。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/07/29/%E5%A6%82%E4%BD%95%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq-%E5%A6%82%E4%BD%95%E4%BB%8EBAM%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96fastq/" data-id="cllszh67400hse2qe2r3w5ti5" data-title="如何从BAM文件中提取fastq" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-R语言的字符串操作-R语言的字符串操作" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/28/R%E8%AF%AD%E8%A8%80%E7%9A%84%E5%AD%97%E7%AC%A6%E4%B8%B2%E6%93%8D%E4%BD%9C-R%E8%AF%AD%E8%A8%80%E7%9A%84%E5%AD%97%E7%AC%A6%E4%B8%B2%E6%93%8D%E4%BD%9C/" class="article-date">
  <time class="dt-published" datetime="2019-07-28T10:25:03.264Z" itemprop="datePublished">2019-07-28</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/28/R%E8%AF%AD%E8%A8%80%E7%9A%84%E5%AD%97%E7%AC%A6%E4%B8%B2%E6%93%8D%E4%BD%9C-R%E8%AF%AD%E8%A8%80%E7%9A%84%E5%AD%97%E7%AC%A6%E4%B8%B2%E6%93%8D%E4%BD%9C/">R语言的字符串操作</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>R语言主要擅长于数值向量和矩阵操作，但是让他去做字符串操作也可以。</p>
<p>字符串的基本操作类型：</p>
<ul>
<li>查找和替换</li>
<li>大小写转换</li>
<li>字符数统计</li>
<li>字符串连接和拆分</li>
</ul>
<p>就我所知，有两套处理函数，一套是Hadley大神的stringr,一套是R自带的。</p>
<h2 id="stringr使用指南"><a href="#stringr使用指南" class="headerlink" title="stringr使用指南"></a>stringr使用指南</h2><p>stringr函数主要分为四类：</p>
<p>1. 字符操作：操作字符向量中的单个符 str_length, str_sub, str_dup</p>
<p>2. 添加，移除和操作空白符 str_pad, str_trim, str_wrap</p>
<p>3. 大小写转换处理 str_to_lower, str_to_upper, str_to_title</p>
<p>4. 模式匹配函数 str_detect, str_subset, str_count, str_locate, str_locate_all, str_match, str_match_all, str_replace, str_replace_all, str_split_fix, str_split, str_extract, str_extract_all </p>
<h3 id="单个字符的处理"><a href="#单个字符的处理" class="headerlink" title="单个字符的处理"></a>单个字符的处理</h3><p>字符长度<code>str_length</code>，等价于<code>nchar</code></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">str_length(&quot;abc&quot;)</span><br></pre></td></tr></table></figure>

<p>根据位置信息提取或替换字符, 类似于substr()</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">x &lt;- c(&quot;abcdef&quot;,&quot;ghijk&quot;)</span><br><span class="line"># 第三个</span><br><span class="line">str_sub(x,3,3)</span><br><span class="line"># 第二个到倒数第二个</span><br><span class="line">str_sub(x,2, -2)</span><br><span class="line"># 替换</span><br><span class="line">str_sub(x,3,3) &lt;- X</span><br></pre></td></tr></table></figure>

<p>重复字符串，不同于rep</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">str_dup(x,c(2,3))</span><br></pre></td></tr></table></figure>

<h3 id="空白符"><a href="#空白符" class="headerlink" title="空白符"></a>空白符</h3><p>前后增加空白字符, str_pad()</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">x &lt;- c(&quot;abc&quot;, &quot;defghi&quot;)</span><br><span class="line">str_pad(x, 10)</span><br><span class="line">str_pad(x, 10, &quot;both&quot;)</span><br></pre></td></tr></table></figure>

<p>移除空白字符, str_trim()</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">x &lt;- c(&quot;   b   &quot;, &quot;c   &quot;, &quot;   d&quot;)</span><br><span class="line">str_trim(x)</span><br><span class="line">str_trim(x,left)</span><br></pre></td></tr></table></figure>

<p>更好的排版，让每一行的看起来一样长， str_wrap()</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">nature &lt;- c(&quot;Nature Methods&#x27; Points of Significance column &quot;,</span><br><span class="line">&quot;on statistics explains many key statistical and &quot;,&quot;experimental design concepts. Other resources include an online&quot;,</span><br><span class="line">&quot; plotting tool and links to statistics guides from other publishers.&quot;)</span><br><span class="line">cat(str_wrap(nature, width=40))</span><br></pre></td></tr></table></figure>

<h3 id="大小写转换"><a href="#大小写转换" class="headerlink" title="大小写转换"></a>大小写转换</h3><ul>
<li>全部大写 str_to_upper, 类似于基础R的toupper()</li>
<li>全部小写 str_to_lower, 类似于基础R的tolower()</li>
<li>title形式 str_to_title</li>
</ul>
<h3 id="模式匹配"><a href="#模式匹配" class="headerlink" title="模式匹配"></a>模式匹配</h3><p>功能： decect, locate, extract, match, replace, split</p>
<p>测试数据：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">strings &lt;- c(</span><br><span class="line">    &quot;apple&quot;,</span><br><span class="line">    &quot;219 733 8965&quot;,</span><br><span class="line">    &quot;329-293-8753&quot;,</span><br><span class="line">    &quot;Work: 579-499-7527; Home: 543.355.3679&quot;</span><br><span class="line">)</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>号码的正则形式:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">phone &lt;- &quot;([2-9][0-9]&#123;2&#125;)[- .]([0-9]&#123;3&#125;)[- .]([0-9]&#123;4&#125;)&quot;</span><br></pre></td></tr></table></figure>

<p>检测字符串是否符合特定模式， str_detect</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">str_detect(strings, phone)</span><br></pre></td></tr></table></figure>

<p>提取字符串(全部内容）， str_subset</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">str_subset(strings, phone)</span><br></pre></td></tr></table></figure>

<p>统计匹配次数，str_count</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">str_count(strings, phone)</span><br></pre></td></tr></table></figure>

<p>定位首个匹配的位置,str_locate(返回matrix), 或所有匹配的位置, str_locate_all（返回list）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">str_locate(strings, phone)</span><br><span class="line">str_locate_all(strings, phone)</span><br></pre></td></tr></table></figure>

<p>str_extract提取首个匹配，返回字符串向量, str_extract_all()提取所有匹配，返回list</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">str_extract(strings, phone)</span><br><span class="line">str_extract_all(strings,phone)</span><br><span class="line">str_extract_all(strings, phone, simplify=TRUE)</span><br></pre></td></tr></table></figure>

<p>str_match提取首个匹配中()分组内的内容， str_match_all()则是全部，因此是list</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">str_match(strings, phone)</span><br><span class="line">str_match_all(strings, phone)</span><br></pre></td></tr></table></figure>

<p>str_replace()替代第一个匹配， str_replace_all替代所有匹配</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">str_replace(strings, phone, &quot;XXX-XXX-XXXX&quot;)</span><br><span class="line">str_replace_all(strings, phone, &quot;XXX-XXX-XXXX&quot;)</span><br></pre></td></tr></table></figure>

<p>str_split_fixed()返回固定数目，全部拆分，str_split（)可变拆分</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">str_split_fixed(&quot;a-b-c&quot;, &quot;-&quot;)</span><br><span class="line">str_spilt(&quot;a-b-c&quot;,&quot;-&quot;, n=2)</span><br></pre></td></tr></table></figure>

<h3 id="4类字符串描述引擎"><a href="#4类字符串描述引擎" class="headerlink" title="4类字符串描述引擎"></a>4类字符串描述引擎</h3><p>1. 默认是正则表达式&#96;vignette(“regular-expression”)</p>
<p>2. 逐byte固定匹配, fixed()</p>
<p>3. locale-sensive 字符匹配, coll()</p>
<p>4. 字符边界分析， boundary()</p>
<h2 id="正则表达式练习"><a href="#正则表达式练习" class="headerlink" title="正则表达式练习"></a>正则表达式练习</h2><h3 id="基本匹配"><a href="#基本匹配" class="headerlink" title="基本匹配"></a>基本匹配</h3><p>最简单的模式就是匹配某个完整的字符</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">x &lt;- c(&quot;apple&quot;, &quot;banana&quot;, &quot;pear&quot;)</span><br><span class="line">str_extract(x, &quot;an&quot;)</span><br></pre></td></tr></table></figure>

<p>如果需要忽略大小写, ignore_case &#x3D;TRUE</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">bananas &lt;- c(&quot;banana&quot;, &quot;Banana&quot;, &quot;BANANA&quot;)</span><br><span class="line">str_dectect(bananas, regrex(&quot;banana&quot;, ignore_case=TRUE))</span><br></pre></td></tr></table></figure>

<p>可以用点<code>.</code>匹配任意字符，但是默认不包括<code>\n</code>,需要用dotall&#x3D;TURE开启</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">str_detect(&quot;\nX\n&quot;, &quot;.X.&quot;)</span><br><span class="line">str_detect(&quot;\nX\n&quot;, regex(&quot;.X.&quot;, dotall=TRUE))</span><br></pre></td></tr></table></figure>

<h3 id="转义（R中一坑）"><a href="#转义（R中一坑）" class="headerlink" title="转义（R中一坑）"></a>转义（R中一坑）</h3><p>正则表达式中有一些是特殊字符，比如说刚才的顿号，因此为了匹配这些特殊字符，我们需要对其转义。在Linux命令行里，转义用的是<code>\</code>, 所以可以直接用<code>\.</code>.</p>
<p>但是R里面的坑就出现了，我们用字符表示正则表达式，<code>\</code> 在字符里被用作转义符号。然后我们需要先把<code>\</code>转义成字符，然后才能进一步转义,<code>\\.</code></p>
<p>如果要匹配<code>\</code> ,就需要<code>\\\\</code>，不可以思议，难以释怀，不知道被坑了多少次。</p>
<p>或者你用<code>\Q...\E</code> 类似于Python的 r’….’， 原意匹配</p>
<h3 id="特殊字符"><a href="#特殊字符" class="headerlink" title="特殊字符"></a>特殊字符</h3><p>一些比较常用的字符匹配</p>
<ul>
<li>\d: 任意数字， \D：任意非数字.</li>
<li>\s: 任意空白字符，\S：任意非空白字符</li>
<li>\w: 匹配单词</li>
<li>\b: 匹配字符边界， \B：非字符边界</li>
<li>[abc], [a-z], [^abc], [^-]：匹配字母，和不匹配</li>
</ul>
<p>在R里面，需要对”&quot;进行转义，所以上面的\在R里都要写成，\<br>下面是一些预编译好的字符集，顾名思义</p>
<ul>
<li>[:punct:]</li>
<li>[:alpha:]</li>
<li>[:lower:]</li>
<li>[:upper:]</li>
<li>[:digit:]</li>
<li>[:xdigit:]</li>
<li>[:alnum:]</li>
<li>[:cntrl:]</li>
<li>[:graph:]</li>
<li>[:print:]</li>
<li>[:space:]</li>
<li>[:blank:]</li>
</ul>
<h3 id="或"><a href="#或" class="headerlink" title="或"></a>或</h3><p>匹配abc或def</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">str_detect(c(&quot;abc&quot;,&quot;def&quot;,&quot;ghi&quot;), &quot;abc|def&quot;)</span><br></pre></td></tr></table></figure>

<h3 id="分组"><a href="#分组" class="headerlink" title="分组"></a>分组</h3><p>匹配grey或gray</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">str_extract(c(&quot;grey&quot;,&quot;gray&quot;), &quot;gr(e|a)y&quot;)</span><br></pre></td></tr></table></figure>

<p>分组可以用\1, \2进行提取，</p>
<h3 id="定位"><a href="#定位" class="headerlink" title="定位"></a>定位</h3><ul>
<li><code>^</code>: 字符串开始， 如<code>^a</code></li>
<li><code>$</code>: 字符串结束， 如<code>a$</code></li>
</ul>
<p>如果字符串有多行，那么就需要regex(multiline&#x3D;TRUE)。此时，</p>
<ul>
<li>\A: 输入开头</li>
<li>\z: 输入结尾</li>
<li>\Z: 头尾</li>
</ul>
<h3 id="重复"><a href="#重复" class="headerlink" title="重复"></a>重复</h3><ul>
<li>?: 0或1</li>
<li>+: 大于等于1</li>
<li>*： 大于等于0</li>
<li>{n}: n次</li>
<li>{n,m}： n到m次</li>
<li>{n,}: 大于那次</li>
</ul>
<p>默认是贪婪模式， 在上述字符后添加”?” 则为非贪婪模式。</p>
<p>PS: 下面是R语言自带的字符处理函数，我已经放弃他们了。</p>
<h2 id="基础R包函数"><a href="#基础R包函数" class="headerlink" title="基础R包函数"></a>基础R包函数</h2><p><code>nchar()</code>: 函数返回字符串长度<br><code>paste()</code>, <code>paste0()</code>: 连接若干个字符串<br><code>sprintf()</code>：格式化输出，下面举例</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">sprintf(&quot;%f&quot;, pi)</span><br><span class="line">sprintf(&quot;%.3f&quot;, pi)</span><br><span class="line">sprintf(&quot;%1.0f&quot;, pi)</span><br><span class="line">sprintf(&quot;%5.1f&quot;, pi)</span><br><span class="line">sprintf(&quot;%05.1f&quot;, pi)</span><br><span class="line">sprintf(&quot;%+f&quot;, pi)</span><br><span class="line">sprintf(&quot;% f&quot;, pi)</span><br><span class="line">sprintf(&quot;%-10f&quot;, pi) # left justified</span><br><span class="line">sprintf(&quot;%e&quot;, pi)</span><br><span class="line">sprintf(&quot;%E&quot;, pi)</span><br><span class="line">sprintf(&quot;%g&quot;, pi)</span><br><span class="line">sprintf(&quot;%g&quot;,   1e6 * pi) # -&gt; exponential</span><br><span class="line">sprintf(&quot;%.9g&quot;, 1e6 * pi) # -&gt; &quot;fixed&quot;</span><br><span class="line">sprintf(&quot;%G&quot;, 1e-6 * pi)</span><br></pre></td></tr></table></figure>

<p><code>toupper()</code>: 大写转换<br><code>tolower()</code>: 小写转换<br><code>substr()</code>: 提取或替换一个字符串向量的子串</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">x &lt;- &quot;abcde&quot;</span><br><span class="line">substr(x,1,2)</span><br><span class="line"># ab</span><br><span class="line">substr(x,1,2) &lt;- 2333</span><br><span class="line"># 233cde</span><br></pre></td></tr></table></figure>

<p>上面都是一些普通的函数，很好理解，下面都是一些和正则表达式相关的函数，如grep, grepl, regexpr, gregexpr, sub, gsub, strsplit<br>因此必须介绍一下R语言的正则表达式写法了。</p>
<ul>
<li>R语言是用的扩展正则表达式（Extended Regular Expressions)</li>
<li>元字符：\ | ( ) [ { ^ $ * + ?</li>
<li>非元字符转义后：\a as BEL, \e as ESC, \f as FF, \n as LF, \r as CR and \t as TAB</li>
<li>一些定义字符集合[:alnum:], [:alpha:], [:blank:], [:cntrl:], [:digit:], [:graph:], [:lower:], [:print:], [:punct:], [:space:], [:upper:],[:xdigit:]</li>
<li>找出“组”字符串 </li>
<li>默认是贪婪模式，可以通过用?改变为非贪婪模式</li>
</ul>
<p>这些是基本知识，可以百度到每个字符的具体解释，或者看文档<code>?regexp</code><br>不说基础知识了，看下应用吧。我常用的操作一般是找到某个字符串，或者对字符串进行替换</p>
<p>比如说，我想找到所有以P开头，且不是P结尾的字符，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">test &lt;- c(&quot;Python&quot;, &quot;Perl&quot;, &quot;PHP&quot;, &quot;JAVA&quot;, &quot;C&quot;, &quot;C++&quot;)</span><br><span class="line">grep(&quot;^P.*?[^P]$&quot;, test)</span><br><span class="line">[1] 1 2</span><br><span class="line">grep(&quot;^P.*?[^P]$&quot;, test,value=TRUE)</span><br><span class="line">[1] &quot;Python&quot; &quot;Perl&quot;</span><br><span class="line">grepl(&quot;^P.*?[^P]$&quot;, test)</span><br><span class="line">[1]  TRUE  TRUE FALSE FALSE FALSE FALSE</span><br><span class="line">regexpr(&quot;^P.*?[^P]$&quot;, test)</span><br><span class="line">[1]  1  1 -1 -1 -1 -1</span><br><span class="line">attr(,&quot;match.length&quot;)</span><br><span class="line">[1]  6  4 -1 -1 -1 -1</span><br><span class="line">attr(,&quot;useBytes&quot;)</span><br><span class="line">&gt; gregexpr(&quot;^P.*?[^P]$&quot;, test)</span><br><span class="line">[[1]]</span><br><span class="line">[1] 1</span><br><span class="line">attr(,&quot;match.length&quot;)</span><br><span class="line">[1] 6</span><br><span class="line">attr(,&quot;useBytes&quot;)</span><br><span class="line">[1] TRUE</span><br><span class="line"></span><br><span class="line">[[2]]</span><br><span class="line">[1] 1</span><br><span class="line">attr(,&quot;match.length&quot;)</span><br><span class="line">[1] 4</span><br><span class="line">attr(,&quot;useBytes&quot;)</span><br><span class="line">[1] TRUE</span><br></pre></td></tr></table></figure>

<p>其中<code>grep()</code>默认是返回下标，如果设置value&#x3D;TRUE，则返回字符串，<code>grepl()</code>返回是否配对的逻辑判断， <code>regexpr</code>则是返回匹配范围，如果不匹配结果是-1，<code>gregexpr</code>和前者功能一致，只不过返回的是列表形式。<br><strong>注</strong>：忽略大小写ignore.case &#x3D; TRUE</p>
<p>现在我想把<code>C++</code>替换成<code>C--</code>。我先试着找到<code>C++</code></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; grep(&quot;\+\+&quot;,test)</span><br><span class="line">错误: 由&quot;&quot;\+&quot;开头的字符串中存在&#x27;\+&#x27;，但没有这种逸出号</span><br></pre></td></tr></table></figure>

<p>什么情况，为什么<code>\+</code>不能把<code>+</code>这个元字符转义？难不成<code>+</code>在R里面不是元字符？我测试下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">grep(&quot;++&quot;,test,value=TRUE)</span><br><span class="line">Error in grep(&quot;++&quot;, test, value = TRUE) : </span><br><span class="line">  正规表现’++&#x27;不对，原因是&#x27;Invalid use of repetition operators&#x27;</span><br></pre></td></tr></table></figure>

<p>啊！看来<code>+</code>还是元字符，难道是<code>\</code> 叛变革命了，我试试看。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&gt; grep(&quot;\23&quot;,&quot;test\23&quot;,value=TRUE)</span><br><span class="line">[1] &quot;test\023&quot;</span><br><span class="line">&gt; grep(&quot;\\23&quot;,&quot;test\23&quot;,value=TRUE)</span><br><span class="line">Error in grep(&quot;\\23&quot;, &quot;test\023&quot;, value = TRUE) : </span><br><span class="line">  正规表现’\23&#x27;不对，原因是&#x27;Invalid back reference&#x27;</span><br></pre></td></tr></table></figure>

<p>看来<code>\</code> 是主要任务是把非元字符转义，如果想把元字符转义成普通字符，只能是<code>\\</code>元字符</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grep(&quot;\\+\\+&quot;,test,value=TRUE)</span><br><span class="line">[1] &quot;C++&quot;</span><br></pre></td></tr></table></figure>

<p>回到我们之前的替换任务sub只对第一个匹配进行替换，gsub对所有匹配替换。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"> sub(&quot;\\+\\+&quot;,&quot;--&quot;,test)</span><br><span class="line">[1] &quot;Python&quot; &quot;Perl&quot;   &quot;PHP&quot;    &quot;JAVA&quot;   &quot;C&quot;      &quot;C--&quot; </span><br></pre></td></tr></table></figure>

<p>最后还可以用<code>strsplit</code>对字符串进行分割，返回的是一个列表</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">x &lt;- c(as = &quot;asfef&quot;, qu = &quot;qwerty&quot;, &quot;yuiop[&quot;, &quot;b&quot;, &quot;stuff.blah.yech&quot;)</span><br><span class="line">strsplit(x, &quot;e&quot;)</span><br></pre></td></tr></table></figure>



      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/07/28/R%E8%AF%AD%E8%A8%80%E7%9A%84%E5%AD%97%E7%AC%A6%E4%B8%B2%E6%93%8D%E4%BD%9C-R%E8%AF%AD%E8%A8%80%E7%9A%84%E5%AD%97%E7%AC%A6%E4%B8%B2%E6%93%8D%E4%BD%9C/" data-id="cllszh634004ge2qefgtp0xwr" data-title="R语言的字符串操作" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-从零开始学Circos绘制圈图(一)-从零开始学Circos绘制圈图(一)" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/26/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6Circos%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%80)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6Circos%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%80)/" class="article-date">
  <time class="dt-published" datetime="2019-07-26T13:45:44.293Z" itemprop="datePublished">2019-07-26</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/26/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6Circos%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%80)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6Circos%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%80)/">从零开始学Circos绘制圈图(一)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>一般基因组文章都会有下面这种酷炫图，用来描述基因组的基因密度分布，转座子的密度分布，和其他物种或者多倍体的多套染色体间的共线性关系，以及其他各种你只要测序就能加上的信息，比如说你要是测了ATAC-seq，加上全基因组开放状态，要是测了多个组织，多个时期的RNA-seq，那就加上热图展现这种变化关系。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/1564118925560-3386fb85ab5a48d8b92b9635a30563e4.png" alt="circos绘制基因组"></p>
<p>当然除了基因组文章，其他类型的文章也可以考虑这种图。接下来我将会写一些列教程（可能有视频），通过教别人学Circos的方式来自学Circos。</p>
<h2 id="环境配置"><a href="#环境配置" class="headerlink" title="环境配置"></a>环境配置</h2><p>建议在Linux环境下配置Circos，之后只要用conda就能配置好分析环境</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 安装</span></span><br><span class="line"><span class="comment">## circos</span></span><br><span class="line">conda create -c bioconda -n circos circos</span><br></pre></td></tr></table></figure>

<p>测试软件安装结果</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 测试circos</span></span><br><span class="line">conda activate circos</span><br><span class="line"><span class="comment"># 确认安装</span></span><br><span class="line">circos -V</span><br><span class="line"><span class="comment"># 显示如下</span></span><br><span class="line"><span class="comment"># circos | v 0.69-8 | 15 Jun 2019 | Perl 5.026002</span></span><br></pre></td></tr></table></figure>

<p>可以从<a target="_blank" rel="noopener" href="http://circos.ca/software/download/">http://circos.ca/software/download/</a>下载官方的教程文件，分别是</p>
<ul>
<li><a target="_blank" rel="noopener" href="http://circos.ca/distribution/circos-course-2017.tgz">circos-course-2017.tgz</a> </li>
<li><a target="_blank" rel="noopener" href="http://circos.ca/distribution/circos-tutorials-current.tgz">circos-tutorials-current.tgz</a></li>
</ul>
<h2 id="处理过程"><a href="#处理过程" class="headerlink" title="处理过程"></a>处理过程</h2><p>Circos依赖于一些列的配置文件，用来定义复杂图形的各个部分，最终加工成图形。</p>
<p>因此，用Circos画图是一个不断增添内容的过程，你要不断根据输出结果来调整输入参数。</p>
<p>并且整个分析中，你还要拥有过关的数据预处理的能力，这是因为Circos不是数据处理工具，它只是展示你已有的数据。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/1564114612076-b1639241644b45c8b802a39659340ed9.png" alt="处理过程"></p>
<h2 id="快速开始"><a href="#快速开始" class="headerlink" title="快速开始"></a>快速开始</h2><p>不管怎么样，先快速绘制出一个Circos图再说。</p>
<blockquote>
<p>果子老师说过，我们不是先成为了老司机才开车，而是开车多了才成为了老司机。</p>
</blockquote>
<p>第一步，先新建一个文件夹，用于存放本次分析的所有数据和配置文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p my_first_circos &amp;&amp; <span class="built_in">cd</span> my_first_circos</span><br></pre></td></tr></table></figure>

<p>然后用<code>vim karyotype.tair10.txt</code>编辑文本，新增如下内容</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">chr - chr1 chr1 0 30427617 black</span><br><span class="line">chr - chr2 chr2 0 19698289 black</span><br><span class="line">chr - chr3 chr3 0 23459830 black</span><br><span class="line">chr - chr4 chr4 0 18585056 black</span><br><span class="line">chr - chr5 chr5 0 26975502 black</span><br></pre></td></tr></table></figure>

<p>之后创建一个<code>circos.conf</code>文件，用于增加各类配置参数</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">touch</span> circos.conf</span><br></pre></td></tr></table></figure>

<p>用<code>vim circos.conf</code>，增加我们的第一条记录，染色体信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">karyotype = karyotype.tair10.txt</span><br></pre></td></tr></table></figure>

<p>然而要想真正的出图，还需要增加至少以下配置语句才行</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">&lt;ideogram&gt;</span><br><span class="line">&lt;spacing&gt;</span><br><span class="line">default = 0.005r</span><br><span class="line">&lt;/spacing&gt;</span><br><span class="line">radius           = 0.90r</span><br><span class="line">thickness        = 20p</span><br><span class="line">fill             = <span class="built_in">yes</span></span><br><span class="line">stroke_color     = dgrey</span><br><span class="line">stroke_thickness = 2p</span><br><span class="line">&lt;/ideogram&gt;</span><br><span class="line"></span><br><span class="line">&lt;image&gt;</span><br><span class="line">&lt;&lt;<span class="string">include etc/image.conf&gt;&gt;</span></span><br><span class="line"><span class="string">&lt;/image&gt;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">&lt;&lt;include</span> etc/colors_fonts_patterns.conf&gt;&gt;</span><br><span class="line">&lt;&lt;<span class="string">include etc/housekeeping.conf&gt;&gt;</span></span><br></pre></td></tr></table></figure>

<p>在当前路径下运行<code>circos -conf circos.conf</code>, 最终效果图如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/1564116740621-63c9fc5b9c0c4cddafd3eb7763653642.png" alt="第一张图"></p>
<p>虽然图比较丑，但是至少我们成功运行了人生第一次的circos, 这就相当于买了一套毛坯房，后面要做的事情就是不断装修。</p>
<p>比如说，我们至少可以让不同染色体拥有不同的颜色，修改之前的<code>karyotype.tair10.txt</code>中的最后一列</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">chr - chr1 chr1 0 30427617 chr1</span><br><span class="line">chr - chr2 chr2 0 19698289 chr2</span><br><span class="line">chr - chr3 chr3 0 23459830 chr3</span><br><span class="line">chr - chr4 chr4 0 18585056 chr4</span><br><span class="line">chr - chr5 chr5 0 26975502 chr5</span><br></pre></td></tr></table></figure>

<p>在当前路径下运行<code>circos -conf circos.conf</code>, 最终效果图如下</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/1564117790558-25d1cec4a31e4dd1beecde09134b1c98.png" alt="1564117790558"></p>
<p>这就引出了第一个知识点，<strong>配色</strong>。</p>
<p>为了实现配色，需要<code>circos.conf</code>文件了有一个和配色有关的语句</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">&lt;&lt;<span class="string">include etc/colors_fonts_patterns.conf&gt;&gt;</span></span><br></pre></td></tr></table></figure>

<p>这里<code>&lt;&lt;&gt;&gt;</code>表示通过<strong>相对路径</strong>的方式加载另外一个配置文件，它的实际路径是和<code>circos</code>所在目录同级的<code>etc</code>,可用下面语句看到<code>colors_fonts_patterns.conf</code>的内容</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">circos_path=$(<span class="built_in">dirname</span> `<span class="built_in">which</span> circos`)</span><br><span class="line">less <span class="variable">$&#123;circos_path%bin&#125;</span>/etc/colors_fonts_patterns.conf</span><br></pre></td></tr></table></figure>

<p>你会发现，这个文件里还嵌套其他的配置文件。最终通过层层排查，你才知道<code>etc/colors.ucsc.conf</code>才是实际定义我们填写的颜色名的文件，而颜色的定义如下：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">chr1  = 153,102,0</span><br><span class="line">chr2  = 102,102,0</span><br><span class="line">chr3  = 153,153,30</span><br><span class="line">chr4  = 204,0,0</span><br><span class="line">chr5  = 255,0,0</span><br></pre></td></tr></table></figure>

<p>还有一个问题，为什么这里用的是两个尖括号<code>&lt;&lt;</code>，而不是一个尖括号<code>&lt;</code>呢？这是因为<code>&lt;</code>已经被用于分隔不同的语句块，如下语句就表示<code>etc/image.conf</code>里的配置信息是用来调整和<code>image</code>有关的配置，而不是去调整<code>ideogram</code>的配置。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&lt;image&gt;</span><br><span class="line">&lt;&lt;<span class="string">include etc/image.conf&gt;&gt;</span></span><br><span class="line"><span class="string">&lt;/image&gt;</span></span><br></pre></td></tr></table></figure>

<hr>
<p>以上是快速开始部分，后续将会在此基础上，做出发表级别的图。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/07/26/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6Circos%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%80)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6Circos%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%B8%80)/" data-id="cllszh64m009be2qe6ljm7nqe" data-title="从零开始学Circos绘制圈图(一)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/23/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/22/">22</a><a class="page-number" href="/page/23/">23</a><span class="page-number current">24</span><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/25/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
          <li>
            <a href="/2023/06/22/%E5%9C%A8Windows%E4%B8%AD%E6%90%AD%E5%BB%BA%E4%B8%80%E4%B8%AA%E5%9F%BA%E4%BA%8EWSL%E7%9A%84Python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6%E7%8E%AF%E5%A2%83-building-a-wsl-based-python-data-science-environment-in-windows/">在Windows中搭建一个基于WSL的Python数据科学环境</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>