<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>使用ArchR分析单细胞ATAC-seq数据(第十五章) | 洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="第15章 使用ArchR进行整合分析ArchR的一个优势能够整合多种水平的信息从而提供新的洞见。我们可以只用ATAC-seq数据进行分析，如识别peak之间的共开放性来预测调控相互作用，或整合scRNA-seq数据，如通过peak-基因的连锁分析预测增性子活性。无论是哪种情况，ArchR都可以很容易地从scATAC seq数据中获得更深入的见解。 15.1 创建细胞低重叠聚集ArchR能方便许多特">
<meta property="og:type" content="article">
<meta property="og:title" content="使用ArchR分析单细胞ATAC-seq数据(第十五章)">
<meta property="og:url" content="http://xuzhougeng.top/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%BA%94%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter15/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:description" content="第15章 使用ArchR进行整合分析ArchR的一个优势能够整合多种水平的信息从而提供新的洞见。我们可以只用ATAC-seq数据进行分析，如识别peak之间的共开放性来预测调控相互作用，或整合scRNA-seq数据，如通过peak-基因的连锁分析预测增性子活性。无论是哪种情况，ArchR都可以很容易地从scATAC seq数据中获得更深入的见解。 15.1 创建细胞低重叠聚集ArchR能方便许多特">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-0631fdf68e6a4ce19cf81b93afaf2cb9.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-02a7ffc28a7d4065b4dd774a14279192.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-bf1a61a336e444f89bd51d621d4080da.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-2d84a27f73194fd28f3839eb319404f6.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-5381cc535f49440cb162687e3420637b.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-3ce86da565734c6696875e677cf1e21e.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-82eb8df58b6848ab9a63c70efb798251.png">
<meta property="article:published_time" content="2021-01-17T16:34:56.985Z">
<meta property="article:modified_time" content="2021-01-17T16:34:56.985Z">
<meta property="article:author" content="xuzhougeng">
<meta property="article:tag" content="ATAC-seq | 单细胞">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-0631fdf68e6a4ce19cf81b93afaf2cb9.png">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main"><article id="post-使用ArchR分析单细胞ATAC-seq数据(第十五章)-analysis-sc-atac-seq-with-archr-chapter15" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%BA%94%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter15/" class="article-date">
  <time class="dt-published" datetime="2021-01-17T16:34:56.985Z" itemprop="datePublished">2021-01-17</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 class="p-name article-title" itemprop="headline name">
      使用ArchR分析单细胞ATAC-seq数据(第十五章)
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="第15章-使用ArchR进行整合分析"><a href="#第15章-使用ArchR进行整合分析" class="headerlink" title="第15章 使用ArchR进行整合分析"></a>第15章 使用ArchR进行整合分析</h1><p>ArchR的一个优势能够整合多种水平的信息从而提供新的洞见。我们可以只用ATAC-seq数据进行分析，如识别peak之间的共开放性来预测调控相互作用，或整合scRNA-seq数据，如通过peak-基因的连锁分析预测增性子活性。无论是哪种情况，ArchR都可以很容易地从scATAC seq数据中获得更深入的见解。</p>
<h2 id="15-1-创建细胞低重叠聚集"><a href="#15-1-创建细胞低重叠聚集" class="headerlink" title="15.1 创建细胞低重叠聚集"></a>15.1 创建细胞低重叠聚集</h2><p>ArchR能方便许多特征间相关性的分析。在这些相关分析中，使用稀疏的单细胞数据进行这些计算会导致大量的噪声。为规避这一挑战，我们采用了一种由<a target="_blank" rel="noopener" href="https://cole-trapnell-lab.github.io/cicero-release/">Cicero</a>引入的方法，在这些分析之前创建单细胞的低重叠聚集。我们过滤与任何其他聚集重叠超过80%的聚集以减少偏差。同时为提升该方法的速度，我们开发了一个优化的迭代重叠检查流程，借由”Rcpp”包通过<code>C++</code>实现了快速特征相关性运算。在ArchR中，这些优化方法被用于计算peak共开放性、peak-基因连锁和其他连锁分析。这些低重叠聚集运算都是在内部完成，为清楚起见，我们先在这里对其介绍。</p>
<h2 id="15-2-ArchR共开放分析"><a href="#15-2-ArchR共开放分析" class="headerlink" title="15.2 ArchR共开放分析"></a>15.2 ArchR共开放分析</h2><p>共开放是多个单细胞种的两个peak之间开放性的相关性。换句话说，当peak A在某个单细胞中是开放状态时，peak B通常也是开放状态。下图可以直观地说明了这一概念，说明增强子E3通常与启动子P是共开放的。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-0631fdf68e6a4ce19cf81b93afaf2cb9.png" alt="ArchR_Coaccessibility"></p>
<p>有一点需要注意，共开放分析找到的peak通常都是细胞类型特异的peak。这是因为这些peak在一种细胞类型中都是开放的，而另一种细胞类型通常都是关闭的。虽然这些peak之间有很强的相关性，但是不意味着这些peak之间存在调控关系。</p>
<p>在ArchR中，我们使用<code>addCoAccessibility()</code>函数计算共开放性，计算结束得到的开放性信息保存在<code>ArchRProject</code>中</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addCoAccessibility<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>使用<code>getCoAccessibility()</code>函数可以从<code>ArchRProject</code>对象中提取共开放性信息，当设置<code>returnLoops=FALSE</code>时会返回一个<code>DataFrame</code>对象。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cA <span class="operator">&lt;-</span> getCoAccessibility<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">FALSE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>该<code>DataFrame</code>对象包含几个重要的信息。<code>queryHits</code>和<code>subjectHits</code>列记录的是两个存在相关性的peak的索引。<code>correlation</code>则是两个peak之间的开放状态相关的数值。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">cA</span><br><span class="line"><span class="comment"># DataFrame with 64824 rows and 4 columns</span></span><br><span class="line"><span class="comment"># queryHits subjectHits seqnames correlation</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 5 10 chr1 0.63855416236086</span></span><br><span class="line"><span class="comment"># … … … … …</span></span><br><span class="line"><span class="comment"># 64823 143977 143978 chrX 0.550319862774772</span></span><br></pre></td></tr></table></figure>

<p>这个共开放<code>DataFrame</code>还有一个元数据成员，包含一个相关peak的<code>GRanges</code>对象。上面提到的<code>queryHits</code>和<code>subjectHits</code>的索引适用于这个<code>GRanges</code>对象。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">metadata<span class="punctuation">(</span>cA<span class="punctuation">)</span><span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 144009 ranges and 0 metadata columns:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># Mono chr1 752499-752999 *</span></span><br><span class="line"><span class="comment"># NK chr1 762651-763151 *</span></span><br><span class="line"><span class="comment"># B chr1 801006-801506 *</span></span><br><span class="line"><span class="comment"># B chr1 805039-805539 *</span></span><br><span class="line"><span class="comment"># CLP chr1 845325-845825 *</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># Erythroid chrX 154664540-154665040 *</span></span><br><span class="line"><span class="comment"># NK chrX 154807324-154807824 *</span></span><br><span class="line"><span class="comment"># PreB chrX 154840785-154841285 *</span></span><br><span class="line"><span class="comment"># PreB chrX 154842404-154842904 *</span></span><br><span class="line"><span class="comment"># NK chrX 154862017-154862517 *</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>如果我们设置<code>retureLoops=TRUE</code>, 那么<code>getCoAccessibility()</code>会以loop track的形式返回共开放数据。在这个<code>GRanges</code>对象中，<code>IRanges</code>的起始和结束对应的是每个存在互作关系中两个共开放peak的位置。<code>resolution</code>参数设置这些loop的碱基对分辨率。当<code>resolution=1</code>时，它会输出连接每个peak中心的loop.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cA <span class="operator">&lt;-</span> getCoAccessibility<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们比较下<code>GRanges</code>和<code>DataFrame</code>这两个对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">cA<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 32412 ranges and 1 metadata column:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># [1] chr1 845575-856640 * | 0.63855416236086</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [32412] chrX 153980218-153990364 * | 0.550319862774772</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>如果我们将loops的分辨率调整为<code>resolution = 1000</code>, 这能够避免绘制过多的共开放互作事件。从下面的输出的<code>GRanges</code>对象中，我们可以看到条目少了很多</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">cA <span class="operator">&lt;-</span> getCoAccessibility<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1000</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cA<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 30997 ranges and 1 metadata column:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># |</span></span><br><span class="line"><span class="comment"># [1] chr1 845500-856500 * | 0.63855416236086</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [30997] chrX 153980500-153990500 * | 0.550319862774772</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>同样的，如果我们进一步降低分辨率<code>resolution = 10000</code>，我们会得到更少共开放交互事件。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">cA <span class="operator">&lt;-</span> getCoAccessibility<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">10000</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cA<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 21142 ranges and 1 metadata column:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># |</span></span><br><span class="line"><span class="comment"># [1] chr1 845000-855000 * | 0.63855416236086</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [21142] chrX 153985000-153995000 * | 0.550319862774772</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<h3 id="15-2-1-在browser-track中绘制共开放"><a href="#15-2-1-在browser-track中绘制共开放" class="headerlink" title="15.2.1 在browser track中绘制共开放"></a>15.2.1 在browser track中绘制共开放</h3><p>当我们在<code>ArchRProject</code>里增加共开放信息后，我们可以用这个信息在基因组浏览器里绘制loop track。具体做法就是在<code>plotBrowserTrack()</code>函数中设置<code>loops</code>参数。我们这里以<code>getCoAccessibility()</code>默认参数输出结果为例，即<code>corCutOff = 0.5</code>, <code>resolution = 1000</code>, <code>returnLoops = TRUE</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;CD34&quot;</span><span class="punctuation">,</span> <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span><span class="punctuation">,</span> <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MS4A1&quot;</span><span class="punctuation">,</span> <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span><span class="punctuation">,</span> <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD8A&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TBX21&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IL7R&quot;</span> <span class="comment">#TCells</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p <span class="operator">&lt;-</span> plotBrowserTrack<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    geneSymbol <span class="operator">=</span> markerGenes<span class="punctuation">,</span> </span><br><span class="line">    upstream <span class="operator">=</span> <span class="number">50000</span><span class="punctuation">,</span></span><br><span class="line">    downstream <span class="operator">=</span> <span class="number">50000</span><span class="punctuation">,</span></span><br><span class="line">    loops <span class="operator">=</span> getCoAccessibility<span class="punctuation">(</span>projHeme5<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>最后用<code>grid.draw</code>函数绘制结果，通过<code>$</code>选择特定的标记基因。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grid<span class="operator">::</span>grid.newpage<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">grid<span class="operator">::</span>grid.draw<span class="punctuation">(</span>p<span class="operator">$</span>CD14<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-02a7ffc28a7d4065b4dd774a14279192.png" alt="Plot-Tracks-Marker-Genes-with-CoAccessibility"></p>
<p>使用<code>plotPDF()</code>保存可编辑的矢量图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>plotList <span class="operator">=</span> p<span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;Plot-Tracks-Marker-Genes-with-CoAccessibility.pdf&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="15-3-ArchR-Peak2GeneLinkages分析"><a href="#15-3-ArchR-Peak2GeneLinkages分析" class="headerlink" title="15.3 ArchR Peak2GeneLinkages分析"></a>15.3 ArchR Peak2GeneLinkages分析</h2><p>和共开放分析类似，ArchR也能分析所谓的”peak-to-gene”关联, 也就是分析peak和基因的相关性。共开放分析和peak-to-gene关联分析的主要不同在于，共开放分析只需要用到ATAC-seq数据，寻找的是peak之间的共开放关系，而peak-to-gene关联分析则会整合scRNA-seq数据，寻找peak开放状态和基因表达量的关系。这其实是相似问题的两种方法。只不过因为peak-to-gene关联分析使用的是scATAC-seq和scRNA-seq数据，我们会认为这个关联更能反应基因调控关系。</p>
<p>在ArchR中，我们使用<code>addPeak2GeneLinks()</code>函数鉴定peak-to-gene的关联。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">projHeme5 <span class="operator">&lt;-</span> addPeak2GeneLinks<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>我们以上一节类似方式提取这些peak-to-gene的关联信息，只不过这里用的是<code>getPeak2GeneLinks()</code>函数。和之前一样，用户可以设置相关性阈值和关联之间的分辨率。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">p2g <span class="operator">&lt;-</span> getPeak2GeneLinks<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.45</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">FALSE</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>当<code>returnLoops=FALSE</code>, 该函数返回一个<code>DataFrame</code>对象，和之前<code>getCoAccessibility()</code>返回的<code>DataFrame</code>相似。主要的不同在于scATAC-seq的索引 peak存放在<code>idxATAC</code>列，而scRNA-seq基因的索引是存放在<code>idxRNA</code>列中</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">p2g</span><br><span class="line"><span class="comment"># DataFrame with 43754 rows and 6 columns</span></span><br><span class="line"><span class="comment"># idxATAC idxRNA Correlation FDR</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 47 5 0.549552663393716 1.34094093110629e-38</span></span><br><span class="line"><span class="comment"># 2 3 6 0.487418258348982 2.1460798658766e-29</span></span><br><span class="line">...</span><br><span class="line"><span class="comment"># VarQATAC VarQRNA</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 0.948753202924817 0.793290683296597</span></span><br><span class="line"><span class="comment"># 2 0.253206396822421 0.445890005913661</span></span><br></pre></td></tr></table></figure>

<p>peak-to-gene关联<code>DataFrame</code>对象同样也有一个<code>GRanges</code>对象记录对应peak的元信息。上面提到的<code>idxATAC</code>索引能用于<code>GRanges</code>对象中。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">metadata<span class="punctuation">(</span>p2g<span class="punctuation">)</span><span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 144009 ranges and 0 metadata columns: ## seqnames ranges strand</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># [1] chr1 752499-752999 *</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># [144009] chrX 154862017-154862517 *</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>如果们设置<code>returnLoop=TRUE</code>，那么<code>getPeak2GeneLinks()</code>会返回一个记录loop track的<code>GRanges</code>对象，链接peak和基因。和之前共开放分析一样，<code>IRanges</code>对象里记录着关联着的peak和基因的位置。当<code>resolution=1</code>, 这连接的是peak的中心和基因的单碱基TSS。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">p2g <span class="operator">&lt;-</span> getPeak2GeneLinks<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.45</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p2g<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 43695 ranges and 2 metadata columns: ## seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># |</span></span><br><span class="line"><span class="comment"># [1] chr1 762901-948847 * | 0.533350285896763</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [43695] chrX 154444701-154664790 * | 0.493389088498317</span></span><br><span class="line"><span class="comment"># FDR</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># [1] 5.20045729958651e-36</span></span><br><span class="line"><span class="comment"># … …</span></span><br><span class="line"><span class="comment"># [43695] 3.37643947034054e-30</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>我们可以将loops的分辨率调整为<code>resolution = 1000</code>, 输出结果主要用于在基因组浏览器中绘制这些连接。因为相同的基因对应的peak过多时，可视化效果会很差。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">p2g <span class="operator">&lt;-</span> getPeak2GeneLinks<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.45</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">1000</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p2g<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 42126 ranges and 2 metadata columns:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># |</span></span><br><span class="line"><span class="comment"># [1] chr1 762500-948500 * | 0.533350285896763</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [42126] chrX 154444500-154664500 * | 0.493389088498317</span></span><br><span class="line"><span class="comment"># FDR</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># [1] 5.20045729958651e-36</span></span><br><span class="line"><span class="comment"># … …</span></span><br><span class="line"><span class="comment"># [42126] 3.37643947034054e-30</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<p>进一步降低分辨率，也会降低peak-to-gene连接的总数。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">p2g <span class="operator">&lt;-</span> getPeak2GeneLinks<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    corCutOff <span class="operator">=</span> <span class="number">0.45</span><span class="punctuation">,</span></span><br><span class="line">    resolution <span class="operator">=</span> <span class="number">10000</span><span class="punctuation">,</span></span><br><span class="line">    returnLoops <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p2g<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># GRanges object with 33645 ranges and 2 metadata columns:</span></span><br><span class="line"><span class="comment"># seqnames ranges strand | value</span></span><br><span class="line"><span class="comment"># |</span></span><br><span class="line"><span class="comment"># [1] chr1 765000-945000 * | 0.533350285896763</span></span><br><span class="line"><span class="comment"># … … … … . …</span></span><br><span class="line"><span class="comment"># [33645] chrX 154445000-154665000 * | 0.493389088498317</span></span><br><span class="line"><span class="comment"># FDR</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># [1] 5.20045729958651e-36</span></span><br><span class="line"><span class="comment"># … …</span></span><br><span class="line"><span class="comment"># [33645] 3.37643947034054e-30</span></span><br><span class="line"><span class="comment"># ——-</span></span><br><span class="line"><span class="comment"># seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span><br></pre></td></tr></table></figure>

<h3 id="15-3-1-在browser-track中绘制peak-to-gene连接"><a href="#15-3-1-在browser-track中绘制peak-to-gene连接" class="headerlink" title="15.3.1 在browser track中绘制peak-to-gene连接"></a>15.3.1 在browser track中绘制peak-to-gene连接</h3><p>我们可以采用之前共开放分析流程中的相同方法，使用<code>plotBrowserTrack()</code>函数在基因组浏览器上绘制peak-to-gene 连接。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">markerGenes  <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;CD34&quot;</span><span class="punctuation">,</span> <span class="comment">#Early Progenitor</span></span><br><span class="line">    <span class="string">&quot;GATA1&quot;</span><span class="punctuation">,</span> <span class="comment">#Erythroid</span></span><br><span class="line">    <span class="string">&quot;PAX5&quot;</span><span class="punctuation">,</span> <span class="string">&quot;MS4A1&quot;</span><span class="punctuation">,</span> <span class="comment">#B-Cell Trajectory</span></span><br><span class="line">    <span class="string">&quot;CD14&quot;</span><span class="punctuation">,</span> <span class="comment">#Monocytes</span></span><br><span class="line">    <span class="string">&quot;CD3D&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD8A&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TBX21&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IL7R&quot;</span> <span class="comment">#TCells</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p <span class="operator">&lt;-</span> plotBrowserTrack<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    geneSymbol <span class="operator">=</span> markerGenes<span class="punctuation">,</span> </span><br><span class="line">    upstream <span class="operator">=</span> <span class="number">50000</span><span class="punctuation">,</span></span><br><span class="line">    downstream <span class="operator">=</span> <span class="number">50000</span><span class="punctuation">,</span></span><br><span class="line">    loops <span class="operator">=</span> getPeak2GeneLinks<span class="punctuation">(</span>projHeme5<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>用<code>grid.draw</code>函数绘制结果，通过<code>$</code>选择特定的标记基因。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grid<span class="operator">::</span>grid.newpage<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">grid<span class="operator">::</span>grid.draw<span class="punctuation">(</span>p<span class="operator">$</span>CD14<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-bf1a61a336e444f89bd51d621d4080da.png" alt="Plot-Tracks-Marker-Genes-with-Peak2GeneLinks"></p>
<p>使用<code>plotPDF()</code>保存可编辑的矢量图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotPDF<span class="punctuation">(</span>plotList <span class="operator">=</span> p<span class="punctuation">,</span> </span><br><span class="line">    name <span class="operator">=</span> <span class="string">&quot;Plot-Tracks-Marker-Genes-with-Peak2GeneLinks.pdf&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> </span><br><span class="line">    addDOC <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">5</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="15-3-2-绘制Peak-to-gene连接热图"><a href="#15-3-2-绘制Peak-to-gene连接热图" class="headerlink" title="15.3.2 绘制Peak-to-gene连接热图"></a>15.3.2 绘制Peak-to-gene连接热图</h3><p>我们还可以通过绘制peak-to-gene热图的方式展示我们的peak-to-gene连接。热图分为两个部分，一个是scATAC-seq数据，一个是scRNA-seq数据。绘图使用<code>plotPeak2GeneHeatmap()</code>函数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p <span class="operator">&lt;-</span> plotPeak2GeneHeatmap<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>热图的行是k-means聚类结果，其中k由我们进行设置，默认是25.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-2d84a27f73194fd28f3839eb319404f6.png" alt="Peak2GeneHeatmap"></p>
<h2 id="15-4-正向TF-调控因子鉴定"><a href="#15-4-正向TF-调控因子鉴定" class="headerlink" title="15.4 正向TF-调控因子鉴定"></a>15.4 正向TF-调控因子鉴定</h2><p>ATAC-seq还可以无偏好的鉴定TF，这些TF所对应的染色质区域(存在该TF结合的motif DNA序列 )表现出很大改变。此外，从聚集的位置权重矩阵(Position Weight Matrices, PWM)中也能看到这些TF家族在其集合位点上有相似的特征(例如 GATA因子)。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-5381cc535f49440cb162687e3420637b.png" alt="GATA_Motif_PWMs"></p>
<p>这种motif的相似性使得识别特定的TF充满了挑战性，这些TF可能在其预测的结合位点驱动染色质开放状态发生改变。为了规避这一挑战，我们之前用ATAC-seq和RNA-seq来鉴定TF，这些TF的基因表达和对应motif的开放状态呈正相关。我们将这些TF命名为”正向调节因子”。然而，这种分析依赖于匹配的基因表达数据，而这些数据并不是所有实验中都有。ArchR可以识别其推断基因得分与其chromVAR TF偏差z-score相关的TF来克服这种依赖性。实现方式为，ArchR利用低重叠细胞聚集将TF motif的chromVAR deviation z-scroe和TF基因的基因活跃得分相关联。当ArchR整合scRNA-seq时，TF的基因表达可以代替推测的基因活性评分。</p>
<h3 id="15-4-1-第一步-鉴定偏离TF-motif"><a href="#15-4-1-第一步-鉴定偏离TF-motif" class="headerlink" title="15.4.1 第一步: 鉴定偏离TF motif"></a>15.4.1 第一步: 鉴定偏离TF motif</h3><p>鉴定正向TF调控因子的第一步就是鉴定存在偏离的TF motif。我们在之前的章节进行过该分析，为所有的motif创建了一个<code>MotifMatrix</code>，记录着chromVAR deviations 和deviation z-scores。使用<code>getGroupSE()</code>函数可以获取基于聚类平均后的数据，返回的是一个<code>SummarizedExperiment</code>.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">seGroupMotif <span class="operator">&lt;-</span> getGroupSE<span class="punctuation">(</span>ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span> useMatrix <span class="operator">=</span> <span class="string">&quot;MotifMatrix&quot;</span><span class="punctuation">,</span> groupBy <span class="operator">=</span> <span class="string">&quot;Clusters2&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>因为<code>SummarizedExperiment</code>对象来自于<code>MotifMatrix</code>, 所以它有两个<code>seqnames</code>, “deviation”和”z”，对应chromVAR里原始的偏离值和偏离值的z-scores.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">seGroupMotif</span><br><span class="line"><span class="comment"># class: SummarizedExperiment</span></span><br><span class="line"><span class="comment"># dim: 1740 11</span></span><br><span class="line"><span class="comment"># metadata(0):</span></span><br><span class="line"><span class="comment"># assays(1): MotifMatrix</span></span><br><span class="line"><span class="comment"># rownames(1740): f1 f2 … f1739 f1740</span></span><br><span class="line"><span class="comment"># rowData names(3): seqnames idx name</span></span><br><span class="line"><span class="comment"># colnames(11): B CD4.M … PreB Progenitor</span></span><br><span class="line"><span class="comment"># colData names(16): TSSEnrichment ReadsInTSS … FRIP nCells</span></span><br></pre></td></tr></table></figure>

<p>我们可以从<code>SummarizedExperiment</code>中只提取deviation z-scores.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">seZ <span class="operator">&lt;-</span> seGroupMotif<span class="punctuation">[</span>rowData<span class="punctuation">(</span>seGroupMotif<span class="punctuation">)</span><span class="operator">$</span>seqnames<span class="operator">==</span><span class="string">&quot;z&quot;</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<p>然后，我们计算该z-score在所有聚类中最大变化值。根据motif在不同聚类中观察到的差异程度，我们可以对其分层。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">rowData<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="operator">$</span>maxDelta <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span><span class="built_in">seq_len</span><span class="punctuation">(</span>ncol<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  rowMaxs<span class="punctuation">(</span>assay<span class="punctuation">(</span>seZ<span class="punctuation">)</span> <span class="operator">-</span> assay<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span>x<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> Reduce<span class="punctuation">(</span><span class="string">&quot;cbind&quot;</span><span class="punctuation">,</span> .<span class="punctuation">)</span> <span class="operator">%&gt;%</span> rowMaxs</span><br></pre></td></tr></table></figure>

<h3 id="15-4-2-第二步-鉴定相关的TF-motif和TF基因得分-表达值"><a href="#15-4-2-第二步-鉴定相关的TF-motif和TF基因得分-表达值" class="headerlink" title="15.4.2 第二步: 鉴定相关的TF motif和TF基因得分&#x2F;表达值"></a>15.4.2 第二步: 鉴定相关的TF motif和TF基因得分&#x2F;表达值</h3><p>接着我们使用<code>correlateMatrices()</code>函数来获取motif开放状态矩阵和基因活跃矩阵(基因得分或基因表达量)，在这里就是<code>GeneScoreMatrix</code>和<code>MotifMatrix</code>. 它们将用于鉴定一类motif开放状态和其自身基因活跃相关的TF。正如之前所提到，我们使用许多低重叠细胞聚集在低维空间里(对应<code>reducedDims</code>参数)来计算相关性。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">corGSM_MM <span class="operator">&lt;-</span> correlateMatrices<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    useMatrix1 <span class="operator">=</span> <span class="string">&quot;GeneScoreMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    useMatrix2 <span class="operator">=</span> <span class="string">&quot;MotifMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>该函数返回一个<code>DataFrame</code>对象，记录着<code>GeneScoreMatrix</code>和<code>MotifMatrix</code>，以及它们在低重叠细胞聚类中的相关性。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">corGSM_MM</span><br><span class="line"><span class="comment"># DataFrame with 825 rows and 14 columns</span></span><br><span class="line"><span class="comment"># GeneScoreMatrix_name MotifMatrix_name cor</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 HES4 HES4_95 0.154455056755803</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># 825 MECP2 MECP2_645 0.0629645419544807</span></span><br><span class="line"><span class="comment"># padj pval GeneScoreMatrix_seqnames</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 0.48577773775473 0.000593860315103582 chr1</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># 825 1 0.163611488297825 chrX</span></span><br><span class="line"><span class="comment"># GeneScoreMatrix_start GeneScoreMatrix_end GeneScoreMatrix_strand</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 935552 934342 2</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># 825 153363188 153287264 2</span></span><br><span class="line"><span class="comment"># GeneScoreMatrix_idx GeneScoreMatrix_matchName MotifMatrix_seqnames</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 15 HES4 z</span></span><br><span class="line"><span class="comment"># … … … …</span></span><br><span class="line"><span class="comment"># 825 874 MECP2 z</span></span><br><span class="line"><span class="comment"># MotifMatrix_idx MotifMatrix_matchName</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># 1 95 HES4</span></span><br><span class="line"><span class="comment"># … … …</span></span><br><span class="line"><span class="comment"># 825 645 MECP2</span></span><br></pre></td></tr></table></figure>

<p>同样的分析还可以使用<code>GeneIntegrationMatrix</code>, 用于替代<code>GeneScoreMatrix</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">corGIM_MM <span class="operator">&lt;-</span> correlateMatrices<span class="punctuation">(</span></span><br><span class="line">    ArchRProj <span class="operator">=</span> projHeme5<span class="punctuation">,</span></span><br><span class="line">    useMatrix1 <span class="operator">=</span> <span class="string">&quot;GeneIntegrationMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    useMatrix2 <span class="operator">=</span> <span class="string">&quot;MotifMatrix&quot;</span><span class="punctuation">,</span></span><br><span class="line">    reducedDims <span class="operator">=</span> <span class="string">&quot;IterativeLSI&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="15-4-3-第三步-在相关性DataFrame中添加极大偏差值"><a href="#15-4-3-第三步-在相关性DataFrame中添加极大偏差值" class="headerlink" title="15.4.3 第三步: 在相关性DataFrame中添加极大偏差值"></a>15.4.3 第三步: 在相关性DataFrame中添加极大偏差值</h3><p>对于每个相关性分析，我们使用第一步计算的极大偏差值(maximum delta )来注释每个motif</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">corGSM_MM<span class="operator">$</span>maxDelta <span class="operator">&lt;-</span> rowData<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="punctuation">[</span>match<span class="punctuation">(</span>corGSM_MM<span class="operator">$</span>MotifMatrix_name<span class="punctuation">,</span> rowData<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="operator">$</span>name<span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;maxDelta&quot;</span><span class="punctuation">]</span></span><br><span class="line">corGIM_MM<span class="operator">$</span>maxDelta <span class="operator">&lt;-</span> rowData<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="punctuation">[</span>match<span class="punctuation">(</span>corGIM_MM<span class="operator">$</span>MotifMatrix_name<span class="punctuation">,</span> rowData<span class="punctuation">(</span>seZ<span class="punctuation">)</span><span class="operator">$</span>name<span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;maxDelta&quot;</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<h3 id="15-4-4-第四步-鉴定正向TF调控因子"><a href="#15-4-4-第四步-鉴定正向TF调控因子" class="headerlink" title="15.4.4 第四步: 鉴定正向TF调控因子"></a>15.4.4 第四步: 鉴定正向TF调控因子</h3><p>我们可以利用所有这些信息来识别正向TF调控因子。在下面的例子中，正向调控因子的标准为</p>
<ul>
<li>motif和基因得分（或基因表达）之间的相关性大于0.5</li>
<li>调整后的p值小于0.01</li>
<li>deviation z-score的最大组间差异位于前四分位。</li>
</ul>
<p>我们应用这些过滤标准，并且做一些文本调整来分离TF名字。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">corGSM_MM <span class="operator">&lt;-</span> corGSM_MM<span class="punctuation">[</span>order<span class="punctuation">(</span><span class="built_in">abs</span><span class="punctuation">(</span>corGSM_MM<span class="operator">$</span>cor<span class="punctuation">)</span><span class="punctuation">,</span> decreasing <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">corGSM_MM <span class="operator">&lt;-</span> corGSM_MM<span class="punctuation">[</span>which<span class="punctuation">(</span><span class="operator">!</span>duplicated<span class="punctuation">(</span>gsub<span class="punctuation">(</span><span class="string">&quot;\\-.*&quot;</span><span class="punctuation">,</span><span class="string">&quot;&quot;</span><span class="punctuation">,</span>corGSM_MM<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;MotifMatrix_name&quot;</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">corGSM_MM<span class="operator">$</span>TFRegulator <span class="operator">&lt;-</span> <span class="string">&quot;NO&quot;</span></span><br><span class="line">corGSM_MM<span class="operator">$</span>TFRegulator<span class="punctuation">[</span>which<span class="punctuation">(</span>corGSM_MM<span class="operator">$</span>cor <span class="operator">&gt;</span> <span class="number">0.5</span> <span class="operator">&amp;</span> corGSM_MM<span class="operator">$</span>padj <span class="operator">&lt;</span> <span class="number">0.01</span> <span class="operator">&amp;</span> corGSM_MM<span class="operator">$</span>maxDelta <span class="operator">&gt;</span> quantile<span class="punctuation">(</span>corGSM_MM<span class="operator">$</span>maxDelta<span class="punctuation">,</span> <span class="number">0.75</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&quot;YES&quot;</span></span><br><span class="line">sort<span class="punctuation">(</span>corGSM_MM<span class="punctuation">[</span>corGSM_MM<span class="operator">$</span>TFRegulator<span class="operator">==</span><span class="string">&quot;YES&quot;</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># [1] “ATOH1” “BCL11A” “CEBPA-DT” “CEBPB” “CEBPD” “CREB1”</span></span><br><span class="line"><span class="comment"># [7] “CREB3L4” “EBF1” “EGR2” “EOMES” “ERF” “ESR1”</span></span><br><span class="line"><span class="comment"># [13] “ETS1” “ETV3” “FUBP1” “GATA1” “GATA2” “GATA5”</span></span><br><span class="line"><span class="comment"># [19] “GATA6” “IRF1” “JDP2” “KLF11” “KLF2” “LYL1”</span></span><br><span class="line"><span class="comment"># [25] “MECOM” “MITF” “NFE2” “NFIA” “NFIB” “NFIC”</span></span><br><span class="line"><span class="comment"># [31] “NFIX” “NHLH1” “POU2F1” “RUNX2” “SIX5” “SMAD1”</span></span><br><span class="line"><span class="comment"># [37] “SMAD9” “SP4” “SPI1” “SPIB” “TAL1” “TCF15”</span></span><br><span class="line"><span class="comment"># [43] “TCF23” “TCF4” “TFAP2C” “TWIST1” “TWIST2” “YY1”</span></span><br><span class="line"><span class="comment"># [49] “ZEB1-AS1”</span></span><br></pre></td></tr></table></figure>

<p>在从基因得分和motif deviation z-scores鉴定到正向TF调控因子后，我们可以在点图中突出展示</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">p <span class="operator">&lt;-</span> ggplot<span class="punctuation">(</span>data.frame<span class="punctuation">(</span>corGSM_MM<span class="punctuation">)</span><span class="punctuation">,</span> aes<span class="punctuation">(</span>cor<span class="punctuation">,</span> maxDelta<span class="punctuation">,</span> color <span class="operator">=</span> TFRegulator<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  geom_point<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  theme_ArchR<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  geom_vline<span class="punctuation">(</span>xintercept <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> lty <span class="operator">=</span> <span class="string">&quot;dashed&quot;</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  scale_color_manual<span class="punctuation">(</span>values <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;NO&quot;</span><span class="operator">=</span><span class="string">&quot;darkgrey&quot;</span><span class="punctuation">,</span> <span class="string">&quot;YES&quot;</span><span class="operator">=</span><span class="string">&quot;firebrick3&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  xlab<span class="punctuation">(</span><span class="string">&quot;Correlation To Gene Score&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  ylab<span class="punctuation">(</span><span class="string">&quot;Max TF Motif Delta&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  scale_y_continuous<span class="punctuation">(</span></span><br><span class="line">    expand <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">    limits <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="built_in">max</span><span class="punctuation">(</span>corGSM_MM<span class="operator">$</span>maxDelta<span class="punctuation">)</span><span class="operator">*</span><span class="number">1.05</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-3ce86da565734c6696875e677cf1e21e.png" alt="TF-Regulators-GeneScores"></p>
<p>使用相同的分析策略处理<code>GeneIntegrationMatrix</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">corGIM_MM <span class="operator">&lt;-</span> corGIM_MM<span class="punctuation">[</span>order<span class="punctuation">(</span><span class="built_in">abs</span><span class="punctuation">(</span>corGIM_MM<span class="operator">$</span>cor<span class="punctuation">)</span><span class="punctuation">,</span> decreasing <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">corGIM_MM <span class="operator">&lt;-</span> corGIM_MM<span class="punctuation">[</span>which<span class="punctuation">(</span><span class="operator">!</span>duplicated<span class="punctuation">(</span>gsub<span class="punctuation">(</span><span class="string">&quot;\\-.*&quot;</span><span class="punctuation">,</span><span class="string">&quot;&quot;</span><span class="punctuation">,</span>corGIM_MM<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;MotifMatrix_name&quot;</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">corGIM_MM<span class="operator">$</span>TFRegulator <span class="operator">&lt;-</span> <span class="string">&quot;NO&quot;</span></span><br><span class="line">corGIM_MM<span class="operator">$</span>TFRegulator<span class="punctuation">[</span>which<span class="punctuation">(</span>corGIM_MM<span class="operator">$</span>cor <span class="operator">&gt;</span> <span class="number">0.5</span> <span class="operator">&amp;</span> corGIM_MM<span class="operator">$</span>padj <span class="operator">&lt;</span> <span class="number">0.01</span> <span class="operator">&amp;</span> corGIM_MM<span class="operator">$</span>maxDelta <span class="operator">&gt;</span> quantile<span class="punctuation">(</span>corGIM_MM<span class="operator">$</span>maxDelta<span class="punctuation">,</span> <span class="number">0.75</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&quot;YES&quot;</span></span><br><span class="line">sort<span class="punctuation">(</span>corGIM_MM<span class="punctuation">[</span>corGIM_MM<span class="operator">$</span>TFRegulator<span class="operator">==</span><span class="string">&quot;YES&quot;</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p <span class="operator">&lt;-</span> ggplot<span class="punctuation">(</span>data.frame<span class="punctuation">(</span>corGIM_MM<span class="punctuation">)</span><span class="punctuation">,</span> aes<span class="punctuation">(</span>cor<span class="punctuation">,</span> maxDelta<span class="punctuation">,</span> color <span class="operator">=</span> TFRegulator<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  geom_point<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  theme_ArchR<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  geom_vline<span class="punctuation">(</span>xintercept <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> lty <span class="operator">=</span> <span class="string">&quot;dashed&quot;</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  scale_color_manual<span class="punctuation">(</span>values <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;NO&quot;</span><span class="operator">=</span><span class="string">&quot;darkgrey&quot;</span><span class="punctuation">,</span> <span class="string">&quot;YES&quot;</span><span class="operator">=</span><span class="string">&quot;firebrick3&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  xlab<span class="punctuation">(</span><span class="string">&quot;Correlation To Gene Expression&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  ylab<span class="punctuation">(</span><span class="string">&quot;Max TF Motif Delta&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  scale_y_continuous<span class="punctuation">(</span></span><br><span class="line">    expand <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">    limits <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="built_in">max</span><span class="punctuation">(</span>corGIM_MM<span class="operator">$</span>maxDelta<span class="punctuation">)</span><span class="operator">*</span><span class="number">1.05</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">p</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2021/01/image-82eb8df58b6848ab9a63c70efb798251.png" alt="TF-Regulators-GeneExpression"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E4%BA%94%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter15/" data-id="clm1ywdjs00c9ecok3tgt64n2" data-title="使用ArchR分析单细胞ATAC-seq数据(第十五章)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
    
<nav id="article-nav">
  
    <a href="/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E5%85%AD%E7%AB%A0)-shi-yong-a-r-c-h-r-fen-xi-dan-xi-bao-a-t-a-c---s-e-q-shu-ju--di-shi-liu-zhang-/" id="article-nav-newer" class="article-nav-link-wrap">
      <strong class="article-nav-caption">前一篇</strong>
      <div class="article-nav-title">
        
          使用ArchR分析单细胞ATAC-seq数据(第十六章)
        
      </div>
    </a>
  
  
    <a href="/2021/01/17/%E4%BD%BF%E7%94%A8ArchR%E5%88%86%E6%9E%90%E5%8D%95%E7%BB%86%E8%83%9EATAC-seq%E6%95%B0%E6%8D%AE(%E7%AC%AC%E5%8D%81%E5%9B%9B%E7%AB%A0)-analysis-sc-atac-seq-with-archr-chapter14/" id="article-nav-older" class="article-nav-link-wrap">
      <strong class="article-nav-caption">后一篇</strong>
      <div class="article-nav-title">使用ArchR分析单细胞ATAC-seq数据(第十四章)</div>
    </a>
  
</nav>

  
</article>


</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/09/">九月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/09/02/%E5%A6%82%E4%BD%95%E8%AE%A9%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8D%E5%86%8D%E9%BE%9F%E9%80%9F%E4%B8%8B%E8%BD%BDgithub%E7%9A%84%E6%95%B0%E6%8D%AE/">如何让服务器不再龟速下载github的数据</a>
          </li>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>