<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>如何使用BSA方法进行遗传定位（水稻篇） | 洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="BSA虽然不是我最早接触的高通量数据分析项目（最早的是RNA-seq），但是却是我最早独立开展的数据分析项目， 我曾经专门写过一篇文章介绍如何使用SHOREMap做拟南芥的EMS诱变群体的BSA分析 在遗传定位上，相对于GWAS和binmap，BSA是一个比较省钱的策略，它只需要测两个亲本和后代中两个极端差异群体即可，但是它对实验设计，表型考察，样本挑选都有比较高的要求。如果你的表型差异并不是泾渭">
<meta property="og:type" content="article">
<meta property="og:title" content="如何使用BSA方法进行遗传定位（水稻篇）">
<meta property="og:url" content="http://xuzhougeng.top/2019/10/04/%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8BSA%E6%96%B9%E6%B3%95%E8%BF%9B%E8%A1%8C%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D%EF%BC%88%E6%B0%B4%E7%A8%BB%E7%AF%87%EF%BC%89-QTL-analysis-using-BSA-seq-in-rice/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:description" content="BSA虽然不是我最早接触的高通量数据分析项目（最早的是RNA-seq），但是却是我最早独立开展的数据分析项目， 我曾经专门写过一篇文章介绍如何使用SHOREMap做拟南芥的EMS诱变群体的BSA分析 在遗传定位上，相对于GWAS和binmap，BSA是一个比较省钱的策略，它只需要测两个亲本和后代中两个极端差异群体即可，但是它对实验设计，表型考察，样本挑选都有比较高的要求。如果你的表型差异并不是泾渭">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/image-20191003181641928-89fb8e679a084735851669aea22bbf70.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/image-20191003181653883-32d9596aaef14375a83085d72436618a.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/snp-index-1-9357d7610a5540ce980fe807c647fb14.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/snp-index-2-9496feba458a4b5682385962586f07c5.png">
<meta property="og:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/ED-b396b941cc0b4b95ae7850935ce92f40.png">
<meta property="article:published_time" content="2019-10-04T10:45:49.860Z">
<meta property="article:modified_time" content="2019-10-05T19:15:22.189Z">
<meta property="article:author" content="xuzhougeng">
<meta property="article:tag" content="水稻 | 转录组 | 遗传定位">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/image-20191003181641928-89fb8e679a084735851669aea22bbf70.png">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main"><article id="post-如何使用BSA方法进行遗传定位（水稻篇）-QTL-analysis-using-BSA-seq-in-rice" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/04/%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8BSA%E6%96%B9%E6%B3%95%E8%BF%9B%E8%A1%8C%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D%EF%BC%88%E6%B0%B4%E7%A8%BB%E7%AF%87%EF%BC%89-QTL-analysis-using-BSA-seq-in-rice/" class="article-date">
  <time class="dt-published" datetime="2019-10-04T10:45:49.860Z" itemprop="datePublished">2019-10-04</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 class="p-name article-title" itemprop="headline name">
      如何使用BSA方法进行遗传定位（水稻篇）
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>BSA虽然不是我最早接触的高通量数据分析项目（最早的是RNA-seq），但是却是我最早独立开展的数据分析项目， 我曾经专门写过一篇文章介绍如何使用SHOREMap做<a href="/archives/Mapping-by-sequencing-Using-SHOREmap">拟南芥的EMS诱变群体的BSA分析</a></p>
<p>在遗传定位上，相对于GWAS和binmap，BSA是一个比较省钱的策略，它只需要测两个亲本和后代中两个极端差异群体即可，但是它对实验设计，表型考察，样本挑选都有比较高的要求。如果你的表型差异并不是泾渭分明，那么还是不要用BSA比较合适。这方面知识，建议阅读徐云碧老师2016年发表在PBJ上的”Bulked sample analysis in genetics, genomics and crop improvement”, 这篇教程侧重于实际分析，而非理论讲解。</p>
<p>用于讲解的文章题为”Identification of a cold-tolerant locus in rice (Oryza sativa L.) using bulked segregant analysis with a next-generation sequencing strategy”, 发表在Rice上。 文章用的耐寒(Kongyu131)和不耐寒(Dongnong422)的两个亲本进行杂交，得到的F2后代利用单粒传法得到RIL(重组自交系)群体.之后用BWA+GATK识别SNP，计算ED和SNP-index作图。</p>
<p>这次实战的目标也就是得到文章关键的两张图：</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/image-20191003181641928-89fb8e679a084735851669aea22bbf70.png" alt="ED"></p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/image-20191003181653883-32d9596aaef14375a83085d72436618a.png" alt="SNP-index"></p>
<h2 id="环境准备"><a href="#环境准备" class="headerlink" title="环境准备"></a>环境准备</h2><p>新建一个项目，用于存放本次分析的数据和结果</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p rice-bsa </span><br><span class="line"><span class="built_in">cd</span> rice-bsa</span><br></pre></td></tr></table></figure>

<p>后续分析还需要用到如下软件:</p>
<ul>
<li>wget: 一般Linux系统会自带，用于下载数据</li>
<li>seqkit: 多能的序列处理软件</li>
<li>fastp: 数据质控</li>
<li>bwa: 数据比对到参考基因组</li>
<li>samtools: 处理sam&#x2F;bam文件</li>
<li>sambamba:  标记重复序列</li>
<li>bcftools: 处理vcf&#x2F;bcf，能用于变异检测</li>
<li>R: 数据分析</li>
</ul>
<p>我们需要分别下载参考基因组(IRGSP)数据和测序数据。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p ref &amp;&amp; <span class="built_in">cd</span> ref</span><br><span class="line">wget https://rapdb.dna.affrc.go.jp/download/archive/irgsp1/IRGSP-1.0_genome.fasta.gz</span><br><span class="line">gunzip IRGSP-1.0_genome.fasta.gz</span><br><span class="line"><span class="comment"># 建立索引</span></span><br><span class="line">bwa index IRGSP-1.0_genome.fasta</span><br></pre></td></tr></table></figure>

<p>之后 根据文章提供的编号，SRR6327815, SRR6327816, SRR6327817, SRR6327818 ，我们到<a target="_blank" rel="noopener" href="https://www.ebi.ac.uk/ena/">ENA</a> 查找对应的下载链接进行数据下载。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p data &amp;&amp; <span class="built_in">cd</span> data</span><br><span class="line">wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR632/005/SRR6327815/SRR6327815_&#123;1,2&#125;.fastq.gz</span><br><span class="line">wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR632/006/SRR6327816/SRR6327816_&#123;1,2&#125;.fastq.gz</span><br><span class="line">wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR632/007/SRR6327817/SRR6327817_&#123;1,2&#125;.fastq.gz</span><br><span class="line">wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR632/008/SRR6327818/SRR6327818_&#123;1,2&#125;.fastq.gz</span><br></pre></td></tr></table></figure>

<blockquote>
<p>因为在NCBI上下载的是SRA格式，需要做数据转换，而从ENA上可以直接下载压缩过的fastq文件，所以我更偏好于ENA。</p>
</blockquote>
<h2 id="上游预处理"><a href="#上游预处理" class="headerlink" title="上游预处理"></a>上游预处理</h2><p>对于测序数据而言，为了能够获得后续用于计算SNP-index或者ED的SNP信息，我们需要将二代测序得到的数据回贴到参考基因组上，然后利用变异检测软件找到每个样本和参考基因组的区别。</p>
<h3 id="数据质控"><a href="#数据质控" class="headerlink" title="数据质控"></a>数据质控</h3><p>原始数据中可能有一部分序列质量不够高，会影响后续分析，比如说测序质量不够，或者说存在接头。通常我们会都会对原始数据进行一波过滤，这里用的是fastp，优点就是快。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p 01-clean-data</span><br><span class="line">fastp -i data/SRR6327815_1.fastq.gz -I data/SRR6327815_2.fastq.gz -o 01-clean-data/SRR6327815_1.fastq.gz -O 01-clean-data/SRR6327815_2.fastq.gz</span><br><span class="line">fastp -i data/SRR6327816_1.fastq.gz -I data/SRR6327816_2.fastq.gz -o 01-clean-data/SRR6327816_1.fastq.gz -O 01-clean-data/SRR6327816_2.fastq.gz</span><br><span class="line">fastp -i data/SRR6327817_1.fastq.gz -I data/SRR6327817_2.fastq.gz -o 01-clean-data/SRR6327817_1.fastq.gz -O 01-clean-data/SRR6327817_2.fastq.gz</span><br><span class="line">fastp -i data/SRR6327818_1.fastq.gz -I data/SRR6327818_2.fastq.gz -o 01-clean-data/SRR6327818_1.fastq.gz -O 01-clean-data/SRR6327818_2.fastq.gz</span><br></pre></td></tr></table></figure>

<h3 id="序列比对"><a href="#序列比对" class="headerlink" title="序列比对"></a>序列比对</h3><p>之后将各个样本的序列回帖到参考基因组</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p 02-read-align</span><br><span class="line">NUMBER_THREADS=80</span><br><span class="line">REFERENCE=ref/IRGSP-1.0_genome.fasta</span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> `<span class="built_in">ls</span> 01-clean-data/`; <span class="keyword">do</span> </span><br><span class="line">    sample=<span class="variable">$&#123;i%%_*&#125;</span></span><br><span class="line">    (bwa mem -M -R <span class="string">&quot;@RG\\tID:<span class="variable">$&#123;sample&#125;</span>\\tSM:<span class="variable">$&#123;sample&#125;</span>\\tPL:ILLUMINA&quot;</span> \</span><br><span class="line">        -t <span class="variable">$NUMBER_THREADS</span> <span class="variable">$REFERENCE</span> 01-clean-data/<span class="variable">$&#123;sample&#125;</span>_1.fastq.gz 01-clean-data/<span class="variable">$&#123;sample&#125;</span>_2.fastq.gz  || <span class="built_in">echo</span> -n <span class="string">&#x27;error&#x27;</span> ) \</span><br><span class="line">        | samtools <span class="built_in">sort</span> -@ 20 -o 02-read-align/<span class="variable">$&#123;sample&#125;</span>_sort.bam -</span><br><span class="line">        samtools index -@ 20 02-read-align/<span class="variable">$&#123;sample&#125;</span>_sort.bam</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<p>这里用到了稍微比较高级的shell操作</p>
<ul>
<li>变量名替换<code>sample=$&#123;i%%_*&#125;</code></li>
<li>逻辑判断: <code>||</code></li>
<li>子shell: <code>()</code></li>
<li>for循环: <code>for do done</code></li>
</ul>
<p>接着我门需要去除重复序列，这里的重复序列指的是拥有相同位置信息，且序列也一模一样的read，通常是由PCR扩增引起。过滤重复序列的目的是为了提高变异检测的准确性，如果一条read上出现的“变异”其实是在第一轮PCR扩增时引入的错误，那么后续的扩增只会让这个错误一直保留着，随后测序的时候这条许多又被多次测到，那么在后续的分析中由于多次出现，就有可能会变异检测软件当作真实变异。</p>
<p>这里用sambamba，因为它的速度比较快，且结果和picard一模一样。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> `<span class="built_in">ls</span> 02-read-align/*_sort.bam`; <span class="keyword">do</span></span><br><span class="line">    sample=<span class="variable">$&#123;i%%_*&#125;</span></span><br><span class="line">    sambamba markdup -r -t 10 <span class="variable">$&#123;sample&#125;</span>_sort.bam <span class="variable">$&#123;sample&#125;</span>_mkdup.bam </span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<h3 id="变异检测"><a href="#变异检测" class="headerlink" title="变异检测"></a>变异检测</h3><p>变异检测最常见的就是bcftools, freebayes和GATK. 这里用的是BCFtools，主要原因还是它的速度比较快。</p>
<p>这里为了让他的速度更快，我用了<code>--region</code>参数分染色体并行，由于水稻有12条染色体，相当于提速了12倍</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p 03-variants</span><br><span class="line"><span class="built_in">ls</span> -1 02-read-align/*_mkdup.bam &gt; bam_list.txt</span><br><span class="line">seqkit <span class="built_in">seq</span> -n ref/IRGSP-1.0_genome.fasta | \</span><br><span class="line"><span class="keyword">while</span> <span class="built_in">read</span> region</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">bcftools mpileup -f ref/IRGSP-1.0_genome.fasta  \</span><br><span class="line">	--redo-BAQ --min-BQ 30 \</span><br><span class="line">	--per-sample-mF \</span><br><span class="line">	--annotate FORMAT/AD,FORMAT/DP \</span><br><span class="line">	--regions <span class="variable">$&#123;region&#125;</span> \</span><br><span class="line">	-Ou --bam-list bam_list.txt  | \</span><br><span class="line">	bcftools call -<span class="built_in">mv</span> -Ob -o 03-variants/<span class="variable">$&#123;region&#125;</span>.bcf &amp;</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<p>之后将拆分运行的结果合并到单个文件中</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">mkdir</span> -p 04-variant-filter</span><br><span class="line">bcftools concat --naive -o 04-variant-filter/merged.bcf 03-variants/*.bcf</span><br></pre></td></tr></table></figure>

<h3 id="变异过滤"><a href="#变异过滤" class="headerlink" title="变异过滤"></a>变异过滤</h3><p>得到VCF文件还需要进行一些过滤，来提高变异的准确性. 这个需要根据具体的项目来进行, 但是有一些固定的指标可以用来对结果进行过滤，例如</p>
<ul>
<li>测序深度</li>
<li>非参考基因组的高质量read数</li>
<li>是否和indel紧邻，通常和indel比较近的snp都不可靠</li>
<li>平均的比对质量,</li>
</ul>
<p>举个例子:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rice-bsa目录下</span></span><br><span class="line"><span class="built_in">cd</span> 04-variant-filter</span><br><span class="line"></span><br><span class="line">bcftools filter  -g3 -G10 -e<span class="string">&#x27;%QUAL&lt;10 || (RPB&lt;0.1 &amp;&amp; %QUAL&lt;15) || (AC&lt;2 &amp;&amp; %QUAL&lt;15) || MQ &lt; 30 || MQSB &lt;=0.1&#x27;</span>  merged.bcf &gt; filter.vcf</span><br></pre></td></tr></table></figure>

<p>之后，我只选择了snp用于后续分析</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 04-variant-filter目录下</span></span><br><span class="line">bcftools view -i <span class="string">&#x27;TYPE=&quot;snp&quot; &amp; N_ALT =1 &amp; STRLEN(ALT) = 1&#x27;</span> filter.vcf &gt; snps.vcf</span><br></pre></td></tr></table></figure>

<p>最终留下了将近80w的SNP。这就是后续R语言下游分析的基础。</p>
<blockquote>
<p>上游处理后的数据可以从链接:<a target="_blank" rel="noopener" href="https://pan.baidu.com/s/1n6CX33E6Qrpf85INMxGGAQ">https://pan.baidu.com/s/1n6CX33E6Qrpf85INMxGGAQ</a> ( 密码:q9u3 )下载</p>
</blockquote>
<h2 id="下游分析"><a href="#下游分析" class="headerlink" title="下游分析"></a>下游分析</h2><p>下游分析可能是比较成熟化的流程分析，对服务器的要求比较高一些，下游分析则是对背景知识要求高一些，例如VCF的格式，遗传学的三大定律等。</p>
<p>让我们先安装并加载所需的R包</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">install.pacakges<span class="punctuation">(</span><span class="string">&quot;devtools&quot;</span><span class="punctuation">)</span></span><br><span class="line">install.packages<span class="punctuation">(</span><span class="string">&quot;vcfR&quot;</span><span class="punctuation">)</span></span><br><span class="line">devtools<span class="operator">:</span>install_github<span class="punctuation">(</span><span class="string">&quot;xuzhougeng/binmapr&quot;</span><span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span><span class="string">&quot;vcfR&quot;</span><span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span><span class="string">&quot;binmapr&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>然后我们需要利用vcfR读取VCF文件</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">vcf <span class="operator">&lt;-</span> read.vcfR<span class="punctuation">(</span><span class="string">&quot;04-variant-filter/snps.vcf&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>接着从VCF对象中提取两个关键信息，AD(Allele Depth)和GT(Genotype)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">gt <span class="operator">&lt;-</span> extract.gt<span class="punctuation">(</span>vcf<span class="punctuation">)</span></span><br><span class="line">ad <span class="operator">&lt;-</span> extract.gt<span class="punctuation">(</span>vcf<span class="punctuation">,</span> <span class="string">&quot;AD&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>gt是一个基因型矩阵，基于之的前过滤操作，所以这里只会有”0&#x2F;0”, “0&#x2F;1”,”1&#x2F;1”这三种情况，而ad则是等位基因的count数. 我们用<code>head</code>查看前10行来了解下情况， </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> head<span class="punctuation">(</span>gt<span class="punctuation">)</span></span><br><span class="line">            SRR6327815 SRR6327816 SRR6327817 SRR6327818</span><br><span class="line">chr01_1151  <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/1&quot;</span>     </span><br><span class="line">chr01_6918  <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/1&quot;</span>     </span><br><span class="line">chr01_17263 <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/1&quot;</span>     </span><br><span class="line">chr01_21546 <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/1&quot;</span>     </span><br><span class="line">chr01_24732 <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/0&quot;</span>     </span><br><span class="line">chr01_33667 <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;1/1&quot;</span>      <span class="string">&quot;0/0&quot;</span>      <span class="string">&quot;0/1&quot;</span> </span><br><span class="line"></span><br><span class="line"><span class="operator">&gt;</span> head<span class="punctuation">(</span>ad<span class="punctuation">)</span></span><br><span class="line">            SRR6327815 SRR6327816 SRR6327817 SRR6327818</span><br><span class="line">chr01_1151  <span class="string">&quot;14,0&quot;</span>     <span class="string">&quot;0,18&quot;</span>     <span class="string">&quot;10,0&quot;</span>     <span class="string">&quot;11,3&quot;</span>    </span><br><span class="line">chr01_6918  <span class="string">&quot;31,0&quot;</span>     <span class="string">&quot;0,30&quot;</span>     <span class="string">&quot;32,0&quot;</span>     <span class="string">&quot;21,5&quot;</span>    </span><br><span class="line">chr01_17263 <span class="string">&quot;46,0&quot;</span>     <span class="string">&quot;0,34&quot;</span>     <span class="string">&quot;20,0&quot;</span>     <span class="string">&quot;21,8&quot;</span>    </span><br><span class="line">chr01_21546 <span class="string">&quot;0,25&quot;</span>     <span class="string">&quot;0,20&quot;</span>     <span class="string">&quot;17,0&quot;</span>     <span class="string">&quot;16,3&quot;</span>    </span><br><span class="line">chr01_24732 <span class="string">&quot;0,29&quot;</span>     <span class="string">&quot;36,0&quot;</span>     <span class="string">&quot;21,0&quot;</span>     <span class="string">&quot;31,0&quot;</span>    </span><br><span class="line">chr01_33667 <span class="string">&quot;0,26&quot;</span>     <span class="string">&quot;0,31&quot;</span>     <span class="string">&quot;28,0&quot;</span>     <span class="string">&quot;29,11&quot;</span> </span><br></pre></td></tr></table></figure>

<p>其中SRR6327815, SRR6327816, SRR6327817, SRR6327818 分别对应着 KY131, DN422, T-pool 和 S-pool</p>
<p>仔细观察的话，你会发现一个chr01_21546是一个有趣的位置，因为双亲都是纯合情况下，SRR6327818居然是 “16,3”, 基因型是”0&#x2F;1”, 这既有可能是亲本是杂合但没有测到，也有可能是后代测错了，一个简单粗暴的方法就是删掉它。此外我们还需要考虑是选择KY131是”1&#x2F;1”且 “DN422” 是 “0&#x2F;0”的位点。还是选择KY131是”0&#x2F;0”,  且 “DN422” 是 “1&#x2F;1”位点。最好的方法就是两种都测试一下。</p>
<p>首先测试KY131是”1&#x2F;1”且 “DN422” 是 “0&#x2F;0”的位点</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">mask <span class="operator">&lt;-</span> which<span class="punctuation">(</span>gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327815&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;1/1&quot;</span> <span class="operator">&amp;</span>  gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327816&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;0/0&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ad_flt <span class="operator">&lt;-</span> ad<span class="punctuation">[</span>mask<span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;SRR6327817&quot;</span><span class="punctuation">,</span> <span class="string">&quot;SRR6327818&quot;</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">colnames<span class="punctuation">(</span>ad_flt<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;T_Pool&quot;</span><span class="punctuation">,</span> <span class="string">&quot;S_Pool&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这一波过滤，我们从原来的80w位点中留下了20w个位点，因此<strong>测双亲很重要</strong>，能够极大的降低噪音。</p>
<p>然后，我们就可以根据AD计算SNP-index，即 ALT_COUNT &#x2F; (REF_COUNT + ALT_COUNT),  在0-1之间。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">freq <span class="operator">&lt;-</span> calcFreqFromAd<span class="punctuation">(</span>ad_flt<span class="punctuation">,</span> min.depth <span class="operator">=</span> <span class="number">10</span><span class="punctuation">,</span> max.depth <span class="operator">=</span> <span class="number">100</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>这里设了一个最大和最小深度用于计算频率，太大的深度可能是同源基因或者是重复序列，太低的深度在计算的时候不太准确。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">freq2 <span class="operator">&lt;-</span> freq<span class="punctuation">[</span>Matrix<span class="operator">::</span>rowSums<span class="punctuation">(</span><span class="built_in">is.na</span><span class="punctuation">(</span>freq<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="number">0</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<p>接着我们就可以尝试去重现文章的结果了</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">par<span class="punctuation">(</span>mfrow <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span> formatC<span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">12</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> flag<span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  freq_flt <span class="operator">&lt;-</span> freq2<span class="punctuation">[</span>grepl<span class="punctuation">(</span>i<span class="punctuation">,</span>row.names<span class="punctuation">(</span>freq2<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">  pos <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>substring<span class="punctuation">(</span>row.names<span class="punctuation">(</span>freq_flt<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  plot<span class="punctuation">(</span>pos<span class="punctuation">,</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">]</span> <span class="operator">-</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span> ylim <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">       pch <span class="operator">=</span> <span class="number">20</span><span class="punctuation">,</span> cex <span class="operator">=</span> <span class="number">0.2</span><span class="punctuation">,</span></span><br><span class="line">       xlab <span class="operator">=</span> i<span class="punctuation">,</span></span><br><span class="line">       ylab <span class="operator">=</span> <span class="built_in">expression</span><span class="punctuation">(</span>paste<span class="punctuation">(</span>Delta<span class="punctuation">,</span> <span class="string">&quot; &quot;</span> <span class="punctuation">,</span><span class="string">&quot;SNP index&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/snp-index-1-9357d7610a5540ce980fe807c647fb14.png" alt="snp-index-1"></p>
<p>从图中，我惊讶的发现一些染色体上的部分区域居然都没有标记了，所以我们选择KY131是”0&#x2F;0”,  而 “DN422” 是 “1&#x2F;1”的位点进行分析</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">mask <span class="operator">&lt;-</span> which<span class="punctuation">(</span>gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327815&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;0/0&quot;</span> <span class="operator">&amp;</span>  gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327816&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;1/1&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ad_flt <span class="operator">&lt;-</span> ad<span class="punctuation">[</span>mask<span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;SRR6327817&quot;</span><span class="punctuation">,</span> <span class="string">&quot;SRR6327818&quot;</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">colnames<span class="punctuation">(</span>ad_flt<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;T_Pool&quot;</span><span class="punctuation">,</span> <span class="string">&quot;S_Pool&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">freq <span class="operator">&lt;-</span> calcFreqFromAd<span class="punctuation">(</span>ad_flt<span class="punctuation">,</span> min.depth <span class="operator">=</span> <span class="number">10</span><span class="punctuation">,</span> max.depth <span class="operator">=</span> <span class="number">100</span><span class="punctuation">)</span></span><br><span class="line">freq2 <span class="operator">&lt;-</span> freq<span class="punctuation">[</span>Matrix<span class="operator">::</span>rowSums<span class="punctuation">(</span><span class="built_in">is.na</span><span class="punctuation">(</span>freq<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="number">0</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">par<span class="punctuation">(</span>mfrow <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span> formatC<span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">12</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> flag<span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  freq_flt <span class="operator">&lt;-</span> freq2<span class="punctuation">[</span>grepl<span class="punctuation">(</span>i<span class="punctuation">,</span>row.names<span class="punctuation">(</span>freq2<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">  pos <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>substring<span class="punctuation">(</span>row.names<span class="punctuation">(</span>freq_flt<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  plot<span class="punctuation">(</span>pos<span class="punctuation">,</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span> <span class="operator">-</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">,</span> ylim <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">       pch <span class="operator">=</span> <span class="number">20</span><span class="punctuation">,</span> cex <span class="operator">=</span> <span class="number">0.2</span><span class="punctuation">,</span></span><br><span class="line">       xlab <span class="operator">=</span> i<span class="punctuation">,</span></span><br><span class="line">       ylab <span class="operator">=</span> <span class="built_in">expression</span><span class="punctuation">(</span>paste<span class="punctuation">(</span>Delta<span class="punctuation">,</span> <span class="string">&quot; &quot;</span> <span class="punctuation">,</span><span class="string">&quot;SNP index&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/snp-index-2-9496feba458a4b5682385962586f07c5.png" alt="snp index-2"></p>
<p>从结果中，我们能够发现一个有趣的结果，不同的筛选标记标准会导致标记在染色体上分布发生变化，这可能意味着在某种的筛选标准下，会使得和性状有关的位点明显的被过滤掉。</p>
<p>根据文章里的结果，候选基因落在6号染色体的20-25M中，也就是选择KY131是”0&#x2F;0”,  而 “DN422” 是 “1&#x2F;1”的位点结果和原文比较类似。那么如果我们原本不知道这个结果应该怎么办？这其实也不是什么问题，像我这样把两幅图都做出来，然后和实验设计者交流下，也就差不多知道答案了。</p>
<p>除了SNP-index外，文章还有一个ED(Euclidean distance)方法用于定位，我根据文章的公式和自己的理解写了下代码</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">mask <span class="operator">&lt;-</span> which<span class="punctuation">(</span>gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327815&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;0/0&quot;</span> <span class="operator">&amp;</span>  gt<span class="punctuation">[</span><span class="punctuation">,</span><span class="string">&quot;SRR6327816&quot;</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;1/1&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">ad_flt <span class="operator">&lt;-</span> ad<span class="punctuation">[</span>mask<span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;SRR6327817&quot;</span><span class="punctuation">,</span> <span class="string">&quot;SRR6327818&quot;</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line">ED_list <span class="operator">&lt;-</span> apply<span class="punctuation">(</span>ad_flt<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  count <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>unlist<span class="punctuation">(</span>strsplit<span class="punctuation">(</span>x<span class="punctuation">,</span> <span class="string">&quot;,&quot;</span><span class="punctuation">,</span>fixed <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span>useBytes <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  depth1 <span class="operator">&lt;-</span> count<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span> <span class="operator">+</span> count<span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span></span><br><span class="line">  depth2 <span class="operator">&lt;-</span> count<span class="punctuation">[</span><span class="number">3</span><span class="punctuation">]</span> <span class="operator">+</span> count<span class="punctuation">[</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">  </span><br><span class="line">  ED <span class="operator">&lt;-</span> <span class="built_in">sqrt</span><span class="punctuation">(</span><span class="punctuation">(</span>count<span class="punctuation">[</span><span class="number">3</span><span class="punctuation">]</span> <span class="operator">/</span> depth2 <span class="operator">-</span> count<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span> <span class="operator">/</span> depth1<span class="punctuation">)</span><span class="operator">^</span><span class="number">2</span> <span class="operator">+</span> </span><br><span class="line">               <span class="punctuation">(</span>count<span class="punctuation">[</span><span class="number">4</span><span class="punctuation">]</span> <span class="operator">/</span> depth2<span class="operator">-</span> count<span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span> <span class="operator">/</span>depth1<span class="punctuation">)</span><span class="operator">^</span><span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>ED<span class="operator">^</span><span class="number">5</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">par<span class="punctuation">(</span>mfrow <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span> formatC<span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">12</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> flag<span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  ED_flt <span class="operator">&lt;-</span> ED_list<span class="punctuation">[</span>grepl<span class="punctuation">(</span>i<span class="punctuation">,</span><span class="built_in">names</span><span class="punctuation">(</span>ED_list<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">  pos <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>substring<span class="punctuation">(</span><span class="built_in">names</span><span class="punctuation">(</span>ED_flt<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  plot<span class="punctuation">(</span>pos<span class="punctuation">,</span> ED_flt<span class="punctuation">,</span></span><br><span class="line">       pch <span class="operator">=</span> <span class="number">20</span><span class="punctuation">,</span> cex <span class="operator">=</span> <span class="number">0.2</span><span class="punctuation">,</span></span><br><span class="line">       xlab <span class="operator">=</span> i<span class="punctuation">,</span></span><br><span class="line">       ylab <span class="operator">=</span> <span class="string">&quot;ED&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/10/ED-b396b941cc0b4b95ae7850935ce92f40.png" alt="ED"></p>
<p>我复现的结果和文章的区别在于，少一个拟合线，也就是以1Mb为区间，每次滑动10Kb计算均值。以KY131是”0&#x2F;0”,  而 “DN422” 是 “1&#x2F;1”的位点结果为例，我们来增加这条线</p>
<p>我写了一个函数用于根据窗口来计算均值，输入是之前snp-index的位置和对应的值，以及确定窗口的大小和步长，</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line">calcValueByWindow <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>pos<span class="punctuation">,</span> value<span class="punctuation">,</span></span><br><span class="line">                              window_size <span class="operator">=</span> <span class="number">1000000</span><span class="punctuation">,</span></span><br><span class="line">                              step_size <span class="operator">=</span> <span class="number">100000</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  <span class="comment"># get the max position in the postion</span></span><br><span class="line">  max_pos <span class="operator">&lt;-</span> <span class="built_in">max</span><span class="punctuation">(</span>pos<span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># construct the window</span></span><br><span class="line">  window_start <span class="operator">&lt;-</span> seq<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> max_pos <span class="operator">+</span> window_size<span class="punctuation">,</span>step_size<span class="punctuation">)</span></span><br><span class="line">  window_end <span class="operator">&lt;-</span> window_start <span class="operator">+</span> step_size</span><br><span class="line">  mean_value <span class="operator">&lt;-</span> vector<span class="punctuation">(</span>mode <span class="operator">=</span> <span class="string">&quot;numeric&quot;</span><span class="punctuation">,</span> <span class="built_in">length</span> <span class="operator">=</span> <span class="built_in">length</span><span class="punctuation">(</span>window_start<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># select the value inside the window</span></span><br><span class="line">  <span class="keyword">for</span> <span class="punctuation">(</span>j <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>window_start<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    </span><br><span class="line">    pos_in_window <span class="operator">&lt;-</span> which<span class="punctuation">(</span>pos <span class="operator">&gt;</span> window_start<span class="punctuation">[</span>j<span class="punctuation">]</span> <span class="operator">&amp;</span></span><br><span class="line">                             pos <span class="operator">&lt;</span> window_end<span class="punctuation">[</span>j<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">    value_in_window <span class="operator">&lt;-</span> value<span class="punctuation">[</span>pos_in_window<span class="punctuation">]</span></span><br><span class="line">    </span><br><span class="line">    mean_value<span class="punctuation">[</span>j<span class="punctuation">]</span> <span class="operator">&lt;-</span> mean<span class="punctuation">(</span>value_in_window<span class="punctuation">)</span></span><br><span class="line">    </span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  <span class="comment"># remove the Not A Number position</span></span><br><span class="line">  nan_pos <span class="operator">&lt;-</span>  <span class="built_in">is.nan</span><span class="punctuation">(</span>mean_value<span class="punctuation">)</span></span><br><span class="line">  mean_value <span class="operator">&lt;-</span> mean_value<span class="punctuation">[</span><span class="operator">!</span> nan_pos<span class="punctuation">]</span></span><br><span class="line">  window_pos <span class="operator">&lt;-</span> <span class="punctuation">(</span><span class="punctuation">(</span>window_start <span class="operator">+</span> window_end<span class="punctuation">)</span><span class="operator">/</span> <span class="number">2</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="operator">!</span>nan_pos<span class="punctuation">]</span></span><br><span class="line">  df <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>pos   <span class="operator">=</span> window_pos<span class="punctuation">,</span></span><br><span class="line">                   value <span class="operator">=</span> mean_value<span class="punctuation">)</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>df<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>得到结果就可以用用<code>lines</code>在之前结果上加上均值线</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">par<span class="punctuation">(</span>mfrow <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">3</span><span class="punctuation">,</span><span class="number">4</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span> formatC<span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">12</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> flag<span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  </span><br><span class="line">  freq_flt <span class="operator">&lt;-</span> freq2<span class="punctuation">[</span>grepl<span class="punctuation">(</span>i<span class="punctuation">,</span>row.names<span class="punctuation">(</span>freq2<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">  pos <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>substring<span class="punctuation">(</span>row.names<span class="punctuation">(</span>freq_flt<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  snp_index <span class="operator">&lt;-</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span> <span class="operator">-</span> freq_flt<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">]</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># bin</span></span><br><span class="line">  df <span class="operator">&lt;-</span> calcValueByWindow<span class="punctuation">(</span>pos <span class="operator">=</span> pos<span class="punctuation">,</span> value <span class="operator">=</span> snp_index<span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  plot<span class="punctuation">(</span>x <span class="operator">=</span> pos<span class="punctuation">,</span> y <span class="operator">=</span>snp_index<span class="punctuation">,</span> </span><br><span class="line">       ylim <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">       pch <span class="operator">=</span> <span class="number">20</span><span class="punctuation">,</span> cex <span class="operator">=</span> <span class="number">0.2</span><span class="punctuation">,</span></span><br><span class="line">       xlab <span class="operator">=</span> i<span class="punctuation">,</span></span><br><span class="line">       ylab <span class="operator">=</span> <span class="built_in">expression</span><span class="punctuation">(</span>paste<span class="punctuation">(</span>Delta<span class="punctuation">,</span> <span class="string">&quot; &quot;</span> <span class="punctuation">,</span><span class="string">&quot;SNP index&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  lines<span class="punctuation">(</span>x <span class="operator">=</span> df<span class="operator">$</span>pos<span class="punctuation">,</span> y <span class="operator">=</span> df<span class="operator">$</span>value<span class="punctuation">,</span> col <span class="operator">=</span> <span class="string">&quot;red&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>最后再对文章总结一下。文章并不是只用了BSA的方法进行定位，他们花了几年的时间用SSR分子标记确定了候选基因可能区间，用BSA的方法在原有基础上缩小了定位区间。当然即便如此，候选基因也有上百个，作者通过BLAST的方式，对这些基因进行了注释。尽管中间还加了一些GO富集分析的内容，说这些基因富集在某个词条里，有一个是DNA metabolic processes(GO:0006259)，但我觉得如果作者用clusterProfiler做富集分析，它肯定无法得到任何富集结果。他做富集分析的目的是其实下面这个描述，也就是找到和抗冻相关的基因</p>
<blockquote>
<p>LOC_Os06g39740 and LOC_Os06g39750,were annotated as the function of “response to cold (GO: 0009409)”, suggesting their key roles in regulating cold tolerance in rice. “</p>
</blockquote>
<p>当然他还做了qRT-PCR进行了验证，最后推测LOC_Os06g39750应该是目标基因，这个基因里还有8个SNP位点。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/10/04/%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8BSA%E6%96%B9%E6%B3%95%E8%BF%9B%E8%A1%8C%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D%EF%BC%88%E6%B0%B4%E7%A8%BB%E7%AF%87%EF%BC%89-QTL-analysis-using-BSA-seq-in-rice/" data-id="cllpwkk0000f0e0p21ior7urw" data-title="如何使用BSA方法进行遗传定位（水稻篇）" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li></ul>

    </footer>
  </div>
  
    
<nav id="article-nav">
  
    <a href="/2019/10/06/%E4%BD%BF%E7%94%A8ASMap%E6%9E%84%E5%BB%BA%E9%AB%98%E5%AF%86%E5%BA%A6%E9%81%97%E4%BC%A0%E5%9B%BE%E8%B0%B1-Build-high-density-genetic-map-with-ASMap/" id="article-nav-newer" class="article-nav-link-wrap">
      <strong class="article-nav-caption">前一篇</strong>
      <div class="article-nav-title">
        
          使用ASMap构建高密度遗传图谱
        
      </div>
    </a>
  
  
    <a href="/2019/10/03/%E4%BD%BF%E7%94%A8SHOREmap%E5%81%9Amapping-by-sequencing-Mapping-by-sequencing-Using-SHOREmap/" id="article-nav-older" class="article-nav-link-wrap">
      <strong class="article-nav-caption">后一篇</strong>
      <div class="article-nav-title">使用SHOREmap做mapping-by-sequencing</div>
    </a>
  
</nav>

  
</article>


</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
          <li>
            <a href="/2023/06/22/%E5%9C%A8Windows%E4%B8%AD%E6%90%AD%E5%BB%BA%E4%B8%80%E4%B8%AA%E5%9F%BA%E4%BA%8EWSL%E7%9A%84Python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6%E7%8E%AF%E5%A2%83-building-a-wsl-based-python-data-science-environment-in-windows/">在Windows中搭建一个基于WSL的Python数据科学环境</a>
          </li>
        
          <li>
            <a href="/2023/06/16/%E7%94%A8Colab%E7%BC%93%E8%A7%A3%E6%88%91%E7%9A%84GPU%E7%84%A6%E8%99%91-%E7%94%A8colab%E7%BC%93%E8%A7%A3%E6%88%91%E7%9A%84gpu%E7%84%A6%E8%99%91/">用Colab缓解我的GPU焦虑</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>