<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/23/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-3分的水稻转录组分析实战-Rice-RNA-seq-analysis" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/22/3%E5%88%86%E7%9A%84%E6%B0%B4%E7%A8%BB%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%E5%AE%9E%E6%88%98-Rice-RNA-seq-analysis/" class="article-date">
  <time class="dt-published" datetime="2019-08-22T19:54:49.801Z" itemprop="datePublished">2019-08-22</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/22/3%E5%88%86%E7%9A%84%E6%B0%B4%E7%A8%BB%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%E5%AE%9E%E6%88%98-Rice-RNA-seq-analysis/">3分的水稻转录组分析实战</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="水稻转录组分析实战"><a href="#水稻转录组分析实战" class="headerlink" title="水稻转录组分析实战"></a>水稻转录组分析实战</h1><p>文章标题: Genes, pathways and transcription factors involved in seedling stage chilling stress tolerance in indica rice through RNA-Seq analysis </p>
<p>该文章发表在BMC plant biology，植物科学2区SCI，近四年IF稳定在3分左右。这篇文章的核心部分就是测了好几个转录组，然后做了一点实验验证。</p>
<h2 id="文章简介"><a href="#文章简介" class="headerlink" title="文章简介"></a>文章简介</h2><p>本部分内容由我的一位师弟完成</p>
<h3 id="背景"><a href="#背景" class="headerlink" title="背景"></a>背景</h3><p>世界50％以上的人口将稻米作为饮食中碳水化合物的主要来源。这是因为水稻的种植范围很广，从山地到平原地区均有种植。正因如此，水稻的一生会面临多种生物胁迫（如病原菌，昆虫等）和非生物胁迫（如寒冷，高温，盐碱等），因此解析水稻的抗逆机制，培育具有多种抗逆性状的水稻具有很大的应用价值。而本文的作者是印度人，印度大部分地区的气候具有明显的旱雨两季，旱季的水稻在苗期易受冷（寒）害的影响。随着RNA-seq手段越来越成熟，在比较不同转录组的差异中发挥了重要作用，这对做基因功能的实验室来讲无疑是巨大的福音。作者就想利用RNA-seq的手段，寻找耐寒品种（Geetanjali）和冷敏感品种（Shahabhagidhan）在转录组水平的差异表达的基因，以期在水稻耐寒过程中鉴定到新颖的转录本，并能获得对基因不同表达水平的认识。</p>
<h3 id="方法"><a href="#方法" class="headerlink" title="方法"></a>方法</h3><p>总体思路，对耐寒和冷敏感的品种在冷处理过程中的不同时间点取样，并进行转录组分析，分析完成后，用qRT-PCR来验证表达上调和下调的基因，进一步confirm分析结果。</p>
<h3 id="结果"><a href="#结果" class="headerlink" title="结果"></a>结果</h3><p>基于低温胁迫处理的5个不同时间点的比较转录组分析显示：在冷敏感品种中有更多的下调基因，在耐寒品种中有更多上调的基因。 在冷敏感品种（CSV）中检测到总计13930个差异基因，而在耐寒品种（CTV）中检测到10599个差异基因，且随时间的延长，差异基因的数目也增加。GO富集分析揭示了18个CSV项和28个CTV项显著参与到molecular function, cellular component and biological process。GO分类揭示了耐寒过程中的差异基因显著参与转录调控，活性氧爆发，脂质结合，催化和水解酶活性的作用。KEGG注释通路揭示了更多数量的基因在耐寒过程中调节不同的通路来发挥作用。</p>
<h3 id="结论"><a href="#结论" class="headerlink" title="结论"></a>结论</h3><p>结论真的是简单到吐血，真的是简单到吐血，真的是简单到吐血。就一句话：通过对耐寒品种的转录组分析，我们检测到了涉及到抗逆过程的基因，通路和转录因子。</p>
<h2 id="转录组分析重现"><a href="#转录组分析重现" class="headerlink" title="转录组分析重现"></a>转录组分析重现</h2><blockquote>
<p>转录组分析环境的搭建和一个分析流程全过程可以哔哩哔哩上搜索”徐洲更”, 即可看我录制的视频。对于学习过程中出现的问题，可以参加我们的付费答疑群，活动见<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/w47Hc2BmRCLpQMlRZ4rCCw">https://mp.weixin.qq.com/s/w47Hc2BmRCLpQMlRZ4rCCw</a>。</p>
</blockquote>
<p>本次文章重现涉及到的软件如下:</p>
<ul>
<li>命令行：prefetch, fasterq-dump, hisat2, samtools, featureCounts, gffread</li>
<li>R：DESeq2, clusterProfiler</li>
</ul>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">conda create -n rna-seq sra-tools fastqc fastp  hisat2  samtools subread  gffread multiqc</span><br><span class="line">conda activate rna-seq</span><br></pre></td></tr></table></figure>

<h3 id="数据下载"><a href="#数据下载" class="headerlink" title="数据下载"></a>数据下载</h3><p>文章用到的RNA-seq数据编号，PRJNA288892， 一共有12个样本。我以这个项目编号新建一个文件夹，开始本次的分析</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p PRJNA288892 &amp;&amp; <span class="built_in">cd</span> PRJNA288892</span><br></pre></td></tr></table></figure>

<p>接下来是下载文章用到的数据，编号存在SRR_Acc_List.txt</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">SRR2089751</span><br><span class="line">SRR2089753</span><br><span class="line">SRR2089756</span><br><span class="line">SRR2093937</span><br><span class="line">SRR2093938</span><br><span class="line">SRR2093939</span><br><span class="line">SRR2093948</span><br><span class="line">SRR2093955</span><br><span class="line">SRR2093959</span><br><span class="line">SRR2093960</span><br><span class="line">SRR2093961</span><br><span class="line">SRR2089754</span><br></pre></td></tr></table></figure>

<p>建立一个sra文件夹，使用prefetch将数据下载到该文件夹下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> sra</span><br><span class="line"><span class="built_in">cat</span> SRR_Acc_List.txt | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> prefetch -O sra <span class="variable">$id</span> ; <span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<p>然后将sra文件转成fastq格式</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p 00-raw-data</span><br><span class="line"><span class="built_in">cat</span> SRR_Acc_List.txt | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> fasterq-dump -p -3 -O 00-raw-data sra/<span class="variable">$id</span>.sra ; <span class="keyword">done</span> &amp;</span><br></pre></td></tr></table></figure>

<h3 id="原始数据质控"><a href="#原始数据质控" class="headerlink" title="原始数据质控"></a>原始数据质控</h3><p>原文只说了High quality (QV &gt; 25) 的read用于比对，并没有提到具体的质控的软件。我们这里用fastp做质控，因为fastp跑完之后也会输出一个网页，所以我也就懒得用FastQC再跑一次</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p 01-clean-data</span><br><span class="line"><span class="built_in">mkdir</span> -p qc</span><br><span class="line"><span class="built_in">cat</span> SRR_Acc_List.txt | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> fastp -w 4 -i 00-raw-data/<span class="variable">$id</span>.sra_1.fastq -I 00-raw-data/<span class="variable">$id</span>.sra_2.fastq -o 01-clean-data/<span class="variable">$id</span>.sra_1.fastq.gz -O 01-clean-data/<span class="variable">$id</span>.sra_2.fastq.gz --html qc/<span class="variable">$id</span>.html --json qc/<span class="variable">$id</span>.json ; <span class="keyword">done</span> &amp;</span><br></pre></td></tr></table></figure>

<blockquote>
<p>重申一个观点，对于<strong>普通</strong>转录组，由于测序质量过关，比对软件也能处理开头的接头序列，我基本不做质控。</p>
</blockquote>
<h3 id="建立索引"><a href="#建立索引" class="headerlink" title="建立索引"></a>建立索引</h3><p>文章说它用到水稻参考基因组和注释文件下载自<a href="ftp://ftp.plantbiology.msu.edu/pub/data/EukaryoticProjects/osativa/annotationdbs/pseudomolecules/version_7.0/all.dir/">ftp://ftp.plantbiology.msu.edu/pub/data/EukaryoticProjects/osativa/annotationdbs/pseudomolecules/version_7.0/all.dir/</a>, 而有趣的是这个链接我打不开，目前可用的下载地址其实是: <a target="_blank" rel="noopener" href="http://rice.plantbiology.msu.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/">http://rice.plantbiology.msu.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> reference &amp;&amp; <span class="built_in">cd</span> reference</span><br><span class="line"><span class="comment"># 参考基因组</span></span><br><span class="line">wget http://rice.plantbiology.msu.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/all.con</span><br><span class="line"><span class="comment"># 注释文件</span></span><br><span class="line">wget http://rice.plantbiology.msu.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/all.gff3</span><br></pre></td></tr></table></figure>

<p>此外文章用的是TopHat v1.3.3和Bowtie v0.12.9  默认参数做的比对，我对此不作任何评价。我们这里用HISAT2建立索引</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">hhisat2-build -p 10 reference/all.con reference/Osativa</span><br></pre></td></tr></table></figure>

<h3 id="序列比对"><a href="#序列比对" class="headerlink" title="序列比对"></a>序列比对</h3><p>用HISAT2将过滤后的fastq文件和参考基因组进行比对，得到下游分析所需要的BAM文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p 02-read-align</span><br><span class="line"><span class="built_in">cat</span> SRR_Acc_List.txt | <span class="keyword">while</span> <span class="built_in">read</span> <span class="built_in">id</span>; <span class="keyword">do</span> hisat2 -p 10 -x reference/Osativa -1 01-clean-data/<span class="variable">$id</span>.sra_1.fastq.gz -2 01-clean-data/<span class="variable">$id</span>.sra_2.fastq.gz --new-summary --summary-file qc/<span class="variable">$id</span>.ht2.txt | samtools <span class="built_in">sort</span> -@ 4 &gt; 02-read-align/<span class="variable">$id</span>.bam ; <span class="keyword">done</span> &amp;</span><br></pre></td></tr></table></figure>

<h3 id="表达量定量"><a href="#表达量定量" class="headerlink" title="表达量定量"></a>表达量定量</h3><p>文章使用的Cufflinks v1.3.0对基因的表达量进行定量，使用FPKM进行了标准化。我这里也不做评价，我们用featureCounts进行表达量定量</p>
<p>先将GFF3转成GTF格式</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gffread reference/all.gff3 -T -o reference/Osativa.gtf</span><br></pre></td></tr></table></figure>

<p>然后用featureCoutns</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p 03-gene-count</span><br><span class="line">featureCounts -a reference/Osativa.gtf -o 03-gene-count/gene_counts 02-read-align/*.bam -T 10 -g gene_id</span><br><span class="line"><span class="built_in">mv</span> 03-gene-count/gene_counts.summary qc/gene_counts.summary</span><br></pre></td></tr></table></figure>

<p>最终的结果，就可以读取到R语言里面进行分析了</p>
<h3 id="MultiQC"><a href="#MultiQC" class="headerlink" title="MultiQC"></a>MultiQC</h3><p>上面每一步的中间文件都被我放到了qc文件夹里，因此可以用multiqc进行整合</p>
<h3 id="差异分析"><a href="#差异分析" class="headerlink" title="差异分析"></a>差异分析</h3><blockquote>
<p>由于前面几步的运算时间比较久，我这里直接准备了表达量矩阵，下载方式为百度网盘链接：<a target="_blank" rel="noopener" href="https://pan.baidu.com/s/1c_mdGlDggV2DH79H-X7J_g">https://pan.baidu.com/s/1c_mdGlDggV2DH79H-X7J_g</a> 提取码：jmi5 </p>
</blockquote>
<p>差异表达分析用但是精确检验(exact test)。</p>
<p>之前的SRR编号在分析的时候最好转成实际的样本编号，方便分析的时候查看</p>
<table>
<thead>
<tr>
<th>编号</th>
<th>品种</th>
<th>处理</th>
</tr>
</thead>
<tbody><tr>
<td>SRR2089751.sra</td>
<td>Sahabhagidhan</td>
<td>0h control（易感品种对照）</td>
</tr>
<tr>
<td>SRR2089753.sra</td>
<td>Sahabhagidhan</td>
<td>6h冷害处理</td>
</tr>
<tr>
<td>SRR2089754.sra</td>
<td>Sahabhagidhan</td>
<td>12h冷害处理</td>
</tr>
<tr>
<td>SRR2089756.sra</td>
<td>Sahabhagidhan</td>
<td>24h冷害处理</td>
</tr>
<tr>
<td>SRR2093937.sra</td>
<td>Sahabhagidhan</td>
<td>48h冷害处理</td>
</tr>
<tr>
<td>SRR2093938.sra</td>
<td>Sahabhagidhan</td>
<td>48h冷害处理后又24h恢复</td>
</tr>
<tr>
<td>SRR2093939.sra</td>
<td>Geetanjali</td>
<td>0h conrol</td>
</tr>
<tr>
<td>SRR2093948.sra</td>
<td>Geetanjali</td>
<td>6h冷害处理</td>
</tr>
<tr>
<td>SRR2093955.sra</td>
<td>Geetanjali</td>
<td>12h冷害处理</td>
</tr>
<tr>
<td>SRR2093959.sra</td>
<td>Geetanjali</td>
<td>24h冷害处理</td>
</tr>
<tr>
<td>SRR2093960.sra</td>
<td>Geetanjali</td>
<td>48h冷害处理</td>
</tr>
<tr>
<td>SRR2093961.sra</td>
<td>Geetanjali</td>
<td>48h冷害处理后又24h恢复</td>
</tr>
</tbody></table>
<p>Sahabhagidhan是不耐冷的品种(cold susceptible variety, CSV) ,而Geetanjali则是耐冷品种 (CTV, cold tolerant variety) 。</p>
<p>我们这里分析的时候把两个物种在6, 12, 24, 48时间点的样本作为重复，仅仅比较不同物种在冷害处理后的差异基因，而不分析冷害处理后随时间点变化的基因。</p>
<p>加载数据:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">df &lt;- read.table(&quot;gene_counts&quot;, header = T,</span><br><span class="line">                 sep = &quot;\t&quot;,</span><br><span class="line">                 stringsAsFactors = F)</span><br><span class="line">expr_df &lt;- df[,c(1,8,9,10,11,14,15,16,17)]</span><br><span class="line">colnames(expr_df) &lt;- c(&quot;geneid&quot;,</span><br><span class="line">                       &quot;CSV_1&quot;,&quot;CSV_2&quot;,&quot;CSV_3&quot;,&quot;CSV_4&quot;,</span><br><span class="line">                       &quot;CTV_1&quot;,&quot;CTV_2&quot;,&quot;CTV_3&quot;,&quot;CTV_4&quot;)</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>DESeq2差异分析</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">library(DESeq2)</span><br><span class="line">coldata &lt;- data.frame(condition=factor(c(&quot;CSV&quot;,&quot;CSV&quot;,&quot;CSV&quot;,&quot;CSV&quot;,</span><br><span class="line">                      &quot;CTV&quot;,&quot;CTV&quot;,&quot;CTV&quot;,&quot;CTV&quot;)))</span><br><span class="line">dds &lt;- DESeqDataSetFromMatrix(expr_df,</span><br><span class="line">                       colData = coldata,</span><br><span class="line">                       design = ~ condition,</span><br><span class="line">                       tidy = TRUE)</span><br><span class="line">dds &lt;- DESeq(dds)</span><br></pre></td></tr></table></figure>

<p>PCA比较下组内差异是不是比组间差异小</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">rld &lt;- rlog(dds)</span><br><span class="line">plotPCA(rld)</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/rice_pca-e000a342aec6400f969c460fdc71c44c.png" alt="PCA"></p>
<p>DESeq2的差异表达分析，只需要一行代码</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">res &lt;- results(dds)</span><br></pre></td></tr></table></figure>

<p>用ggplot2画个火山图,</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">plot_df &lt;- data.frame(res)</span><br><span class="line">plot_df &lt;- cbind(geneid=row.names(plot_df),</span><br><span class="line">                        plot_df)</span><br><span class="line"></span><br><span class="line">plot_df &lt;- plot_df[!is.na(plot_df$padj), ]</span><br><span class="line"></span><br><span class="line">ggplot(data=plot_df, aes(x=log2FoldChange, </span><br><span class="line">                      y =-log10(padj))) +</span><br><span class="line">  geom_point(alpha=0.8, size=1.2)+</span><br><span class="line">  labs(title=&quot;Volcanoplot&quot;, x=&quot;log2 (fold change)&quot;,y=&quot;-log10 (q-value)&quot;)+</span><br><span class="line">  theme(plot.title = element_text(hjust = 0.4))+</span><br><span class="line">  geom_hline(yintercept = -log10(0.05),lty=4,color=&quot;red&quot;)+</span><br><span class="line">  geom_hline(yintercept = -log10(0.01),lty=4,color=&quot;blue&quot;)+</span><br><span class="line">  geom_vline(xintercept = c(1,-1),lty=4,alpha=0.8,color=&quot;blue&quot;)+</span><br><span class="line">  geom_vline(xintercept = c(2,-2),lty=4,alpha=0.8,color=&quot;red&quot;)+</span><br><span class="line">  theme_bw()+</span><br><span class="line">  theme(panel.border = element_blank(),</span><br><span class="line">        panel.grid.major = element_blank(),</span><br><span class="line">        panel.grid.minor = element_blank(),</span><br><span class="line">        axis.line = element_line(colour = &quot;black&quot;))</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/rice-volcano-65deaa78f62b4972a72e0d5129840cdd.png" alt="火山图"></p>
<p>火山图的作用，一是检查你的差异分析是否合理，也就是大部分基因都是没有变化，只有一部分是差异基因，第二个是可以帮你确定阈值。</p>
<p>我发现即便是LogFoldChange &gt; 2（相差4倍）, p.adj &lt; 0.01，依旧是有很多差异基因。所以我这里以这个为标准进行筛选，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">res_sub &lt;- subset(res, abs(log2FoldChange) &gt; 2 &amp; padj &lt;0.01)</span><br><span class="line">gene_up &lt;- row.names(res_sub[res_sub$log2FoldChange &gt; 0, ])</span><br><span class="line">length(gene_up) # 1024</span><br><span class="line">gene_down &lt;- row.names(res_sub[res_sub$log2FoldChange &lt; 0, ])</span><br><span class="line">length(gene_down) # 702</span><br></pre></td></tr></table></figure>

<p>其实这个时候你会有一个问题，所谓的基因上调和下调是谁相对于谁呢？很简单，直接把差异的基因的标准化矩阵提取出来即可。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">gene_counts &lt;- counts(dds, normalized = TRUE)</span><br><span class="line">gene_counts &lt;- gene_counts[row.names(res_sub), ]</span><br><span class="line">head(gene_counts)</span><br></pre></td></tr></table></figure>

<p>当你看到具体的数值时</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">                    CSV_1     CSV_2      CSV_3      CSV_4     CTV_1</span><br><span class="line">LOC_Os01g01430   1.739379   0.00000   5.971535   8.188634  31.70621</span><br><span class="line">                   CTV_2     CTV_3    CTV_4</span><br><span class="line">LOC_Os01g01430  29.60817  17.19996  10.9967</span><br></pre></td></tr></table></figure>

<p>你自然就知道了CTV vs CSV里，CSV是作为分母，而CTV是分子。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">log2 fold change (MLE): condition CTV vs CSV </span><br><span class="line">Wald test p-value: condition CTV vs CSV </span><br><span class="line">DataFrame with 6 rows and 6 columns</span><br><span class="line">                       baseMean    log2FoldChange             lfcSE</span><br><span class="line">                      &lt;numeric&gt;         &lt;numeric&gt;         &lt;numeric&gt;</span><br><span class="line">LOC_Os01g01430 13.1763234817146  2.57839743463268 0.925763093663462</span><br></pre></td></tr></table></figure>


<h3 id="使用clusterProfiler做GO-KEGG富集分析"><a href="#使用clusterProfiler做GO-KEGG富集分析" class="headerlink" title="使用clusterProfiler做GO&#x2F;KEGG富集分析"></a>使用clusterProfiler做GO&#x2F;KEGG富集分析</h3><p>文章用agriGo对DEG进行分类，用singular enrichment analysis (SEA) 做GO富集分析，进一步的注释用Web Gene Ontology Annotation Plot (WEGO) 进行画图。做KEGG富集分析时，用的是KEGG GENES database KAAS (KEGG Automatic Annotation Server) 对差异基因进行功能注释.</p>
<p>这里我也不做评价，我们此处用Y叔的clusterProfiler做GO&#x2F;KEGG富集分析。</p>
<p>在做GO富集分析的时候，我遇到了一个难题，我发现Bioconductor上并没有水稻的物种注释包。解决这个问题有两种方法，</p>
<ul>
<li>一种是利用AnnotationHub在线检索抓取OrgDb, 但是这些包都是用ENTREZID，你需要先将RAP-DB或者MSU转为ENTREZID才行</li>
<li>另一种则是我构建好一个物种注释包，你们只要安装就可以继续使用<code>enrichGO</code>， 参考<a target="_blank" rel="noopener" href="https://github.com/xuzhougeng/org.Osativa.eg.db">https://github.com/xuzhougeng/org.Osativa.eg.db</a></li>
</ul>
<p>参考我的GitHub页面安装，<a target="_blank" rel="noopener" href="https://github.com/xuzhougeng/org.Osativa.eg.db">https://github.com/xuzhougeng/org.Osativa.eg.db</a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">install.packages(&quot;https://github.com/xuzhougeng/org.Osativa.eg.db/releases/download/v0.01/org.Osativa.eg.db.tar.gz&quot;, </span><br><span class="line">    repos = NULL, </span><br><span class="line">    type=&quot;source&quot;)</span><br></pre></td></tr></table></figure>

<blockquote>
<p>目前这个物种包比较简单，就只能做GO富集分析，以及MSU和RAP-DB之间的ID转换</p>
</blockquote>
<p>然后进行GO富集分析</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">library(clusterProfiler)</span><br><span class="line">library(org.Osativa.eg.db)</span><br><span class="line"></span><br><span class="line">org &lt;- org.Osativa.eg.db</span><br><span class="line">ego_up &lt;- enrichGO(gene_up,</span><br><span class="line">                OrgDb = org,</span><br><span class="line">                keyType = &quot;GID&quot;,</span><br><span class="line">                pAdjustMethod = &quot;none&quot;,</span><br><span class="line">                ont=&quot;BP&quot;)</span><br><span class="line">p1 &lt;- dotplot(ego_up)</span><br><span class="line"></span><br><span class="line">ego_down &lt;- enrichGO(gene_down,</span><br><span class="line">                OrgDb = org,</span><br><span class="line">                keyType = &quot;GID&quot;,</span><br><span class="line">                pAdjustMethod = &quot;none&quot;,</span><br><span class="line">                ont=&quot;BP&quot;)</span><br><span class="line">p2 &lt;- dotplot(ego_down)</span><br><span class="line"></span><br><span class="line">library(cowplot)</span><br><span class="line">plot_grid(p1,p2)</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/rice-dotplo-01f5f8e8d85d4d72bf093d13f1cef0e6.png" alt="dotplot"></p>
<p>除了点图，我们还可以用clusterProfiler画下面这个图，展现每个基因在不同GO通路的倍数变化</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">fc &lt;- plot_df$log2FoldChange</span><br><span class="line">names(fc) &lt;-  plot_df$geneid</span><br><span class="line"></span><br><span class="line">heatplot(ego_down, foldChange = fc)</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/riceheatplot-94d1fc4441444f0da0fa52d092e9bc03.png" alt="heatplot"></p>
<blockquote>
<p>讲真，GO富集分析给的信息真的是少，我以为是我的分析出现了问题，所以我去看了文章里GO分析中BP部分，发现信息是一样的少。</p>
</blockquote>
<p>GO富集分析做完之后，KEGG富集分析怎么做呢？参考我之前的这篇<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/UnUPVoaMpfJWCQEpkmdTWA">https://mp.weixin.qq.com/s/UnUPVoaMpfJWCQEpkmdTWA</a></p>
<p>因为KEGG要求的输入编号是RAP-DB，所以我们需要进行一次ID转换，在编号中加上”-01”，将编号中的”g”改成”t”, 就可以做KEGG富集分析了</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">rap_id &lt;- mapIds(x = org, </span><br><span class="line">                 keys = gene_up, </span><br><span class="line">                 column = &quot;RAP&quot;,</span><br><span class="line">                 &quot;GID&quot;)</span><br><span class="line">rap_id &lt;- paste0(rap_id[!is.na(rap_id)], &quot;-01&quot;)</span><br><span class="line">rap_id &lt;- gsub(&quot;g&quot;,&quot;t&quot;,rap_id)</span><br><span class="line">ekegg &lt;- enrichKEGG(rap_id, organism = &quot;dosa&quot;, pAdjustMethod = &quot;none&quot;)</span><br></pre></td></tr></table></figure>

<p>没成想，KEGG的分析结果中只得到了一个结果, 所以也懒得放气泡图了。也难怪文章只是将差异基因按照KEGG进行了分类，而没有做富集。我这里也用Y叔的包对上调差异基因按照KEGG归类</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">rice_kegg &lt;- clusterProfiler::download_KEGG(&quot;dosa&quot;)</span><br><span class="line"></span><br><span class="line">kegg_df &lt;- rice_kegg$KEGGPATHID2EXTID</span><br><span class="line">kegg_df &lt;- kegg_df[kegg_df$to %in% rap_id,]</span><br><span class="line">kegg_df &lt;- merge(kegg_df, rice_kegg$KEGGPATHID2NAME,</span><br><span class="line">                 by.x=&quot;from&quot;,by.y=&quot;from&quot;)</span><br><span class="line"></span><br><span class="line">kegg_class &lt;- as.data.frame(sort(table(kegg_df$to.y), decreasing = T)[1:10])</span><br><span class="line">colnames(kegg_class) &lt;- c(&quot;pathway&quot;,&quot;times&quot;)</span><br><span class="line">ggplot(kegg_class,aes(x=pathway, y = times)) +</span><br><span class="line">  geom_bar(fill=&quot;#ca0020&quot;,stat=&quot;identity&quot;) + coord_flip() +</span><br><span class="line">  theme_bw() + geom_text(aes(y = times+1, label = times))</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/kegg_classfication-538b7daab9234c40a588baa915533b1d.png" alt="KEGG分类1"></p>
<p>当然上下调一起展示也不难，就是代码会长一点</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line">rice_kegg &lt;- clusterProfiler::download_KEGG(&quot;dosa&quot;)</span><br><span class="line"># 获取上调的KEGG PATHWAY</span><br><span class="line">up_rap_id &lt;- mapIds(x = org, </span><br><span class="line">                 keys = gene_up, </span><br><span class="line">                 column = &quot;RAP&quot;,</span><br><span class="line">                 &quot;GID&quot;)</span><br><span class="line">up_rap_id &lt;- paste0(up_rap_id[!is.na(up_rap_id)], &quot;-01&quot;)</span><br><span class="line"></span><br><span class="line">up_rap_id &lt;- gsub(&quot;g&quot;,&quot;t&quot;,up_rap_id)</span><br><span class="line"></span><br><span class="line">up_df &lt;- rice_kegg$KEGGPATHID2EXTID</span><br><span class="line">up_df &lt;- kegg_df[kegg_df$to %in% up_rap_id,]</span><br><span class="line">up_df &lt;- merge(kegg_df, rice_kegg$KEGGPATHID2NAME,</span><br><span class="line">                 by.x=&quot;from&quot;,by.y=&quot;from&quot;)</span><br><span class="line"></span><br><span class="line"># 获取下调的KEGG PATHWAY</span><br><span class="line">down_rap_id &lt;- mapIds(x = org, </span><br><span class="line">                    keys = gene_down, </span><br><span class="line">                    column = &quot;RAP&quot;,</span><br><span class="line">                    &quot;GID&quot;)</span><br><span class="line">down_rap_id &lt;- paste0(down_rap_id[!is.na(down_rap_id)], &quot;-01&quot;)</span><br><span class="line">down_rap_id &lt;- gsub(&quot;g&quot;,&quot;t&quot;,down_rap_id)</span><br><span class="line">down_df &lt;- rice_kegg$KEGGPATHID2EXTID</span><br><span class="line">down_df &lt;- down_df[down_df$to %in% down_rap_id,]</span><br><span class="line">down_df &lt;- merge(down_df, rice_kegg$KEGGPATHID2NAME,</span><br><span class="line">               by.x=&quot;from&quot;,by.y=&quot;from&quot;)</span><br><span class="line"></span><br><span class="line"># 统计</span><br><span class="line">kegg_class_up &lt;- as.data.frame(sort(table(up_df$to.y), </span><br><span class="line">                                 decreasing = T)[1:10])</span><br><span class="line">kegg_class_down &lt;- as.data.frame(sort(table(down_df$to.y), </span><br><span class="line">                                    decreasing = T)[1:10])</span><br><span class="line">#合并</span><br><span class="line">kegg_class &lt;- rbind(kegg_class_up, kegg_class_down)</span><br><span class="line">colnames(kegg_class) &lt;- c(&quot;pathway&quot;,&quot;times&quot;)</span><br><span class="line">kegg_class$source &lt;- rep(c(&quot;up&quot;,&quot;down&quot;),</span><br><span class="line">                         times=c(nrow(kegg_class_up),nrow(kegg_class_down)))</span><br><span class="line">#画图</span><br><span class="line">ggplot(kegg_class,aes(x=pathway, y = times)) +</span><br><span class="line">  geom_bar(aes(fill=source),stat=&quot;identity&quot;,position = &quot;dodge&quot;) + </span><br><span class="line">  scale_fill_manual(values = c(up=&quot;#ca0020&quot;,down=&quot;#2b83ba&quot;)) +</span><br><span class="line">  coord_flip() +</span><br><span class="line">  theme_bw()</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/kegg_classfication-2-99ebb8d178e042debb6dc6cd4f2e706d.png" alt="KEGG分类2"></p>
<p>这里这展示了代码，具体如何讲故事，如何调整参数，其实就靠作者自由发挥了。</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><p>水稻两大权威注释组织</p>
<ul>
<li>RAP-DB: <a target="_blank" rel="noopener" href="https://rapdb.dna.affrc.go.jp/">https://rapdb.dna.affrc.go.jp/</a></li>
<li>RGAP: <a target="_blank" rel="noopener" href="http://rice.plantbiology.msu.edu/">http://rice.plantbiology.msu.edu/</a></li>
<li>clusterPrifiler文档: <a target="_blank" rel="noopener" href="https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html">https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html</a></li>
<li>DESeq2文档: <a target="_blank" rel="noopener" href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/22/3%E5%88%86%E7%9A%84%E6%B0%B4%E7%A8%BB%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%E5%AE%9E%E6%88%98-Rice-RNA-seq-analysis/" data-id="clm1z22eo0057ehn5h5cn7m13" data-title="3分的水稻转录组分析实战" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-如何让你的GitHub能被引用-How-to-citate-my-GitHub-repository" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/22/%E5%A6%82%E4%BD%95%E8%AE%A9%E4%BD%A0%E7%9A%84GitHub%E8%83%BD%E8%A2%AB%E5%BC%95%E7%94%A8-How-to-citate-my-GitHub-repository/" class="article-date">
  <time class="dt-published" datetime="2019-08-22T14:01:58.403Z" itemprop="datePublished">2019-08-22</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/22/%E5%A6%82%E4%BD%95%E8%AE%A9%E4%BD%A0%E7%9A%84GitHub%E8%83%BD%E8%A2%AB%E5%BC%95%E7%94%A8-How-to-citate-my-GitHub-repository/">如何让你的GitHub能被引用</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>你需要先登陆到<a target="_blank" rel="noopener" href="https://zenodo.org/">https://zenodo.org/</a>，然后需要创建一个账号，这个账号一定要用GitHub账号进行授权。之后你发现一个Using GitHub， 点击Check out</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-f5d0b1e46c7c41119367e3e9d1cb66a9.png" alt="Step1"></p>
<p>然后你就会发现自己的repository都被列出来了</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-2-ef7546de956a4434ba5bac38d0766673.png" alt="Step2"></p>
<p>之后将你想要被引用的仓库的按钮从off变成ON</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-3-05f1c271e9664247aaaf62682ef323a3.png" alt="Step3"></p>
<p>接着到你想被引用的仓库里, 我们要创建要给release，这样子就有了版本号，如此别人才能知道自己用的版本是哪一版。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-4-56c702f36b334d7c9764a72f35b3132f.png" alt="Step4"></p>
<p>我们就可以创建release，可以设置合适的版本好，比如说我这边设置为0.01</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-5-b8f9f1ed1b0042838b6371df6e48d9e5.png" alt="Step5"></p>
<p>我们就拥有了一个释放版本</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-6-ed976b5018a741b2bcac5a28aae851ea.png" alt="Step6"></p>
<p>之后回到Zenodo，点击Upload</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-7-30755112bfd24f559184141b038b4ef7.png" alt="Step7"></p>
<p>就会出现你刚才释放的软件</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-8-8d759eb87fea4c56b8ccc2e56d25b188.png" alt="Step8"></p>
<p>点击它，你就拥有了可以引用的DOI号</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/zenodo-github-9-55fe96b3f5e24d6589f01c9f4a0822df.png" alt="Step9"></p>
<p>参考资料：<a target="_blank" rel="noopener" href="https://guides.github.com/activities/citable-code/">https://guides.github.com/activities/citable-code/</a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/22/%E5%A6%82%E4%BD%95%E8%AE%A9%E4%BD%A0%E7%9A%84GitHub%E8%83%BD%E8%A2%AB%E5%BC%95%E7%94%A8-How-to-citate-my-GitHub-repository/" data-id="clm1z22hn00g7ehn5hx8df977" data-title="如何让你的GitHub能被引用" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-MECAT2_ 三代高效组装工具-MECAT2-genome-assembly-tools" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/21/MECAT2_%20%E4%B8%89%E4%BB%A3%E9%AB%98%E6%95%88%E7%BB%84%E8%A3%85%E5%B7%A5%E5%85%B7-MECAT2-genome-assembly-tools/" class="article-date">
  <time class="dt-published" datetime="2019-08-21T14:16:16.702Z" itemprop="datePublished">2019-08-21</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/21/MECAT2_%20%E4%B8%89%E4%BB%A3%E9%AB%98%E6%95%88%E7%BB%84%E8%A3%85%E5%B7%A5%E5%85%B7-MECAT2-genome-assembly-tools/">MECAT2:三代高效组装工具</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="MECAT2-三代高效组装工具"><a href="#MECAT2-三代高效组装工具" class="headerlink" title="MECAT2: 三代高效组装工具"></a>MECAT2: 三代高效组装工具</h1><p>MECAT2是三代测序数据PacBio的高效组装工具，是之前<a target="_blank" rel="noopener" href="https://github.com/xiaochuanle/MECAT">MECAT</a>的改进版, 修复了之前的很多bug, 使用基于string graph的<code>fsa</code>替换了之前的<code>mecat2canu</code>。</p>
<p>MECAT2由4个模块组成:</p>
<ul>
<li><code>mecat2pw</code>: SMART reads快速和准确地配对比对工具</li>
<li><code>mecat2ref</code>: SMART reads的参考基因组比对工具</li>
<li><code>mecat2cns</code>: 基于配对的重叠对存在噪音的reads进行纠错</li>
<li><code>fsa</code>: 基于string graph的组装工具</li>
</ul>
<p>目前MECAT2只支持PacBio数据，对于Nanopore数据，肖老师他们开发了<code>NECAT</code></p>
<h2 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h2><p>MECAT2软件安装非常简单，不存在上一代MECAT的安装难问题了</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">mkdir <span class="operator">-</span>p <span class="operator">~</span><span class="operator">/</span>opt<span class="operator">/</span>biosoft</span><br><span class="line">cd <span class="operator">~</span><span class="operator">/</span>opt<span class="operator">/</span>biosoft</span><br><span class="line">git clone https<span class="operator">:</span><span class="operator">/</span><span class="operator">/</span>github.com<span class="operator">/</span>xiaochuanle<span class="operator">/</span>MECAT2.git</span><br><span class="line">cd MECAT2</span><br><span class="line">make <span class="operator">-</span>j <span class="number">4</span></span><br></pre></td></tr></table></figure>

<p>唯一要注意的是尽量要用新版的perl，经测试，Perl 5.26 是可以正常运行脚本。 </p>
<p>最后加入到环境变量<code>PATH</code>中</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">export</span> PATH=~/opt/biosoft/MECAT2/Linux-amd64/bin:<span class="variable">$PATH</span></span><br></pre></td></tr></table></figure>

<h2 id="软件使用"><a href="#软件使用" class="headerlink" title="软件使用"></a>软件使用</h2><p>我们以 <em>E. coli</em> 数据集作为案例，介绍MECAT2应该如何使用。</p>
<p><strong>第一步</strong>，我们要下载数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> -p ~/ecoli</span><br><span class="line"><span class="built_in">cd</span> ~/ecoli</span><br><span class="line">wget http://gembox.cbcb.umd.edu/mhap/raw/ecoli_filtered.fastq.gz</span><br><span class="line">gunzip ecoli_filtered.fastq.gz</span><br></pre></td></tr></table></figure>

<p><strong>注意</strong>: 目前MECAT2还不支持gz压缩文件，如果你直接用gz作为输入，会提示core dumped。</p>
<p>MECAT2不是根据文件名来确定输入的序列格式, 也就是如果你把 ecoli_filtered.fastq 命名为  ecoli_filtered.fasta, 组装也不会出错。</p>
<p><strong>第二步</strong>, 准备模板的config文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mecat.pl config ecoli_config_file.txt</span><br></pre></td></tr></table></figure>

<p>使用<code>vim</code>修改config文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">PROJECT=ecoli</span><br><span class="line">RAWREADS=ecoli_filtered.fastq</span><br><span class="line">GENOME_SIZE=4800000</span><br><span class="line">THREADS=4</span><br><span class="line">MIN_READ_LENGTH=2000</span><br><span class="line">CNS_OVLP_OPTIONS=<span class="string">&quot;&quot;</span></span><br><span class="line">CNS_OPTIONS=<span class="string">&quot;-r 0.6 -a 1000 -c 4 -l 2000&quot;</span></span><br><span class="line">TRIM_OVLP_OPTIONS=<span class="string">&quot;-B&quot;</span></span><br><span class="line">ASM_OVLP_OPTIONS=<span class="string">&quot;-n 100 -z 10 -b 2000 -e 0.5 -j 1 -u 0 -a 400&quot;</span></span><br><span class="line">FSA_OL_FILTER_OPTIONS=<span class="string">&quot;--max_overhang=-1 --min_identity=-1&quot;</span></span><br><span class="line">FSA_ASSEMBLE_OPTIONS=<span class="string">&quot;&quot;</span></span><br><span class="line">USE_GRID=<span class="literal">false</span></span><br><span class="line">CLEANUP=0</span><br><span class="line">CNS_OUTPUT_COVERAGE=30</span><br><span class="line">GRID_NODE=0</span><br></pre></td></tr></table></figure>

<p><strong>第三步</strong>: 原始数据纠错</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mecat.pl correct ecoli_config_file.txt</span><br></pre></td></tr></table></figure>

<p><strong>第四步</strong>: 对纠错后的reads进行组装</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mecat.pl assemble ecoli_config_file.txt</span><br></pre></td></tr></table></figure>

<p>输出结果:</p>
<ul>
<li>纠错后的reads： <code>ecoli/1-consensus/cns_reads.fasta</code>.</li>
<li>最长30X纠错后用于trimming的reads:  <code>ecoli/1-consensus/cns_final.fasta</code>.</li>
<li>trimmed reads:  <code>ecoli/2-trim_bases/trimReads.fasta</code></li>
<li>组装的contigs: <code>ecoli/4-fsa/contigs.fasta</code></li>
</ul>
<p>最终的组装结果为</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Count: 1</span><br><span class="line">Tatal: 4630437</span><br><span class="line">Max: 4630437</span><br><span class="line">Min: 4630437</span><br><span class="line">N25: 4630437</span><br><span class="line">L25: 1</span><br></pre></td></tr></table></figure>

<p>能够完美的就装出一条序列</p>
<h2 id="配置文件介绍"><a href="#配置文件介绍" class="headerlink" title="配置文件介绍"></a>配置文件介绍</h2><p>MECAT2的组装过程的参数都在配置文件中, 也就是之前的<code>ecoli_config_file.txt</code>。我按照不同部分对这些参数进行介绍</p>
<h3 id="基本参数"><a href="#基本参数" class="headerlink" title="基本参数"></a>基本参数</h3><p>这些参数属于改起来不怎么纠结的参数</p>
<ul>
<li><code>PROJECT=ecoli</code>: 项目名，之后的所有输出文件都在该项目名的目录下</li>
<li><code>RAWREADS=</code>: 原始数据的位置，可以是Fasta或者是Fastq， H5格式要先转成Fasta (参考 <a target="_blank" rel="noopener" href="https://github.com/xiaochuanle/MECAT2#S-input-format">Input Format</a>).</li>
<li><code>GENOME_SIZE=</code>: 基因组大小，单位是bp，如果是120M，那么就是120000000.</li>
<li><code>THREADS=</code>: 线程数</li>
<li><code>MIN_READ_LENGTH=</code>: 用于纠错和trim的reads的最低长度. 数据质量好，就长一点，比如说1000到2000</li>
<li><code>USE_GRID=false</code>: 是否有多个计算节点</li>
<li><code>CLEANUP=0</code>: 运行结束后删除MECAT2的中间文件, 大基因组的临时文件很大，所以要设置为1.</li>
<li><code>GRID_NODE=0</code>, 当<code>USE_GRID=1</code>时，设置用到的计算节点数，如果是单节点服务器，不需要设置。</li>
</ul>
<h3 id="纠错和修剪阶段"><a href="#纠错和修剪阶段" class="headerlink" title="纠错和修剪阶段"></a>纠错和修剪阶段</h3><p>纠错阶段(correct stage)和修剪阶段(trimming stage)，MECAT2调用的<code>mecat2pw</code>和<code>mecat2cns</code>， 与之相关的配置如下</p>
<ul>
<li><code>CNS_OVLP_OPTIONS=&quot;&quot;</code>, 在纠错阶段是检测候选重叠的参数, 会传给<code>mecat2pw</code></li>
<li><code>CNS_OPTIONS=&quot;&quot;</code>：原始reads纠错参数，会传递给<code>mecat2cns</code>, </li>
<li><code>TRIM_OVLP_OPTIONS=&quot;&quot;</code>: 在trim阶段，用于检测重叠的参数,会传给<code>v2asmpm</code></li>
</ul>
<p>但是我发现<code>v2asmpm</code>没有文档说明，和软件开发者讨论之后，给出的说明如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">V2asmpm -Pworkpath -Tt -Sx -Ey -B</span><br><span class="line"></span><br><span class="line">-Pworkpath 工作目录是workpath</span><br><span class="line">-Tt 用t个CPU线程</span><br><span class="line">-Sx -Ey 这两个参数一起表示只计算第x到第y卷(包括第y卷), 通常工作目录下会有000001.fasta, 000002.fasta这样的分卷</span><br><span class="line">-B 如果有这个选项, 就输出二进制格式的比对结果; 如果没有这个选项, 就输出文本格式的比对结果</span><br></pre></td></tr></table></figure>
<p>因此用默认的<code>-B</code>就行了。</p>
<h3 id="组装阶段"><a href="#组装阶段" class="headerlink" title="组装阶段"></a>组装阶段</h3><p>组装阶段相关的参数如下</p>
<ul>
<li><code>CNS_OUTPUT_COVERAGE=30</code>: 选择多少覆盖度的最长纠错后reads进行trim和组装。举个例子，4.8M的 <em>E. coli</em> 30X, 是144MB。 一般选择20~40之间就行， 会传递给<code>mecat2elr</code></li>
<li><code>ASM_OVLP_OPTIONS=&quot;&quot;</code>: 在组装阶段，用于检测重叠的参数，传给<code>v2asmpm.sh</code></li>
<li><code>FSA_OL_FILTER_OPTIONS=&quot;&quot;</code>, 过滤重叠的参数，传递给<code>fsa_ol_filter</code></li>
<li><code>FSA_ASSEMBLE_OPTIONS=&quot;&quot;</code>, 组装trimm reads的参数, 传给<code>v2asmpm</code></li>
</ul>
<p>这些参数调整起来都很复杂，只能建议看函数帮助文档了。举个例子<code>FSA_OL_FILTER_OPTIONS=&quot;--max_overhang=-1 --min_identity=-1&quot;</code>根据阅读<code>fsa_ol_filter</code>的帮助发现，<code>-1</code>表示让程序自己决定</p>
<p>而<code>ASM_OVLP_OPTIONS=&quot;-n 100 -z 10 -b 2000 -e 0.5 -j 1 -u 0 -a 400&quot;</code>， 根据和软件开发者的讨论，是<strong>不需要额外调整</strong>，用默认的即可。</p>
<p>MECAT2是基于比对纠错的组装工具，组装速度上比同类型的Canu和Falcon都快。仅看组装结果的N50参数，MECAT2结果和Canu, Falcon结果都是差不多的。</p>
<p>考虑到BUSCO完整度，Canu是三者最高。不过用三代数据和二代数据进行几轮Polish后会提高到Canu相同水平。三个软件上的BUSCO（embryophyta_odb10）值如下:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Canu: C:98.8%[S:96.3%,D:2.5%],F:0.5%,M:0.7%,n:1375</span><br><span class="line">MECAT2: C:93.5%[S:91.6%,D:1.9%],F:4.6%,M:1.9%,n:1375</span><br><span class="line">Falcon: C:96.8%[S:95.1%,D:1.7%],F:2.0%,M:1.2%,n:1375</span><br></pre></td></tr></table></figure>

<p>因此，如果你原本你是用Falcon做组装，那么可以改成MECAT2了，效果差不多，速度还更快了。对于越大的物种，在有限的资源下，更推荐用MECAT2。</p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li><a target="_blank" rel="noopener" href="https://github.com/xiaochuanle/MECAT2">https://github.com/xiaochuanle/MECAT2</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/21/MECAT2_%20%E4%B8%89%E4%BB%A3%E9%AB%98%E6%95%88%E7%BB%84%E8%A3%85%E5%B7%A5%E5%85%B7-MECAT2-genome-assembly-tools/" data-id="clm1z22fb007pehn5a2wq402l" data-title="MECAT2:三代高效组装工具" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-「三代组装」使用新版Falcon进行三代测序基因组组装-Falcon-genome-assembly-tools" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/20/%E3%80%8C%E4%B8%89%E4%BB%A3%E7%BB%84%E8%A3%85%E3%80%8D%E4%BD%BF%E7%94%A8%E6%96%B0%E7%89%88Falcon%E8%BF%9B%E8%A1%8C%E4%B8%89%E4%BB%A3%E6%B5%8B%E5%BA%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-Falcon-genome-assembly-tools/" class="article-date">
  <time class="dt-published" datetime="2019-08-20T15:53:04.991Z" itemprop="datePublished">2019-08-20</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/20/%E3%80%8C%E4%B8%89%E4%BB%A3%E7%BB%84%E8%A3%85%E3%80%8D%E4%BD%BF%E7%94%A8%E6%96%B0%E7%89%88Falcon%E8%BF%9B%E8%A1%8C%E4%B8%89%E4%BB%A3%E6%B5%8B%E5%BA%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-Falcon-genome-assembly-tools/">「三代组装」使用新版Falcon进行三代测序基因组组装</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <blockquote>
<p>这里的新版指的是PacBio公司在2018年9月发布pb-assembly, 而这篇文章是在2018年9月30日发的。</p>
</blockquote>
<p>今年早些时候在参加三代培训时，听说PacBio会在今年对Falcon进行一些改变。前几天我在读 <strong>readthedocs</strong>上的Falcon文档时，发现了文档页面上方出现了这样两栏提醒</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/attentions-d2f8c343963f471f8a2fcd85a64e708e.png" alt="注意"></p>
<p>其中<a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/pb-assembly">pb_assembly</a>就是新的FALCON组装套装在GitHub上的使用文档，经过了几天的探索，我对它有了一点了解，写一篇教程作为官方文档的一些补充吧。</p>
<h2 id="新版亮点"><a href="#新版亮点" class="headerlink" title="新版亮点"></a>新版亮点</h2><ol>
<li>整合了串联重复序列和遍在重复序列的屏蔽（之前没有这一步）</li>
<li>以GFA格式存放graph文件，后续可以用Bandage进行可视化</li>
<li>通过算法和性能优化，提高了Associate Contigs的准确性</li>
<li>分析流程的性能优化</li>
</ol>
<h2 id="软件安装和数据准备"><a href="#软件安装和数据准备" class="headerlink" title="软件安装和数据准备"></a>软件安装和数据准备</h2><p>Falcon终于拥抱了bioconda, 这也就意味着我们再也不需要用到他们原本笨拙的安装脚本，浪费时间在安装软件上。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">conda create -n pb-assembly  pb-assembly</span><br><span class="line"><span class="built_in">source</span> activate pb-assembly</span><br><span class="line"><span class="comment"># 或者</span></span><br><span class="line">conda create -p ~/opt/biosoft/pb-assembly pb-assembly</span><br><span class="line"><span class="built_in">source</span> activate ~/opt/biosoft/pb-assembly </span><br></pre></td></tr></table></figure>

<p>这里使用<a target="_blank" rel="noopener" href="https://pb-falcon.readthedocs.io/en/latest/tutorial.html">https://pb-falcon.readthedocs.io/en/latest/tutorial.html</a>上的所用的E. coli数据集</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wget https://downloads.pacbcloud.com/public/data/git-sym/ecoli.m140913_050931_42139_c100713652400000001823152404301535_s1_p0.subreads.tar.gz</span><br><span class="line">tar -xvzf ecoli.m140913_050931_42139_c100713652400000001823152404301535_s1_p0.subreads.tar.gz</span><br></pre></td></tr></table></figure>

<p>解压缩后的文件夹里有三个300M的Fasta文件, 将他们的实际路径记录到<code>input.fofn</code>中</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ecoli.1.fasta</span><br><span class="line">ecoli.2.fasta</span><br><span class="line">ecoli.3.fasta</span><br></pre></td></tr></table></figure>

<h2 id="准备配置文件"><a href="#准备配置文件" class="headerlink" title="准备配置文件"></a>准备配置文件</h2><p>为了进行组装，需要准备一个配置文件。我的配置文件为<code>fc_run.cfg</code>，内容如下。你们可以先预览一下，后面看我的解释说明。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#### Input</span></span><br><span class="line">[General]</span><br><span class="line">input_fofn=input.fofn</span><br><span class="line">input_type=raw</span><br><span class="line">pa_DBdust_option=</span><br><span class="line">pa_fasta_filter_option=pass</span><br><span class="line">target=assembly</span><br><span class="line">skip_checks=False</span><br><span class="line">LA4Falcon_preload=<span class="literal">false</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#### Data Partitioning</span></span><br><span class="line">pa_DBsplit_option=-x500 -s50</span><br><span class="line">ovlp_DBsplit_option=-x500 -s50</span><br><span class="line"></span><br><span class="line"><span class="comment">#### Repeat Masking</span></span><br><span class="line">pa_HPCTANmask_option=</span><br><span class="line">pa_REPmask_code=1,100;2,80;3,60</span><br><span class="line"></span><br><span class="line"><span class="comment">####Pre-assembly</span></span><br><span class="line">genome_size=0</span><br><span class="line">seed_coverage=20</span><br><span class="line">length_cutoff=3000    </span><br><span class="line">pa_HPCdaligner_option=-v -B128 -M24</span><br><span class="line">pa_daligner_option=-e.7 -l1000 -k18 -h80 -w8 -s100</span><br><span class="line">falcon_sense_option=--output-multi --min-idt 0.70 --min-cov 2 --max-n-read 800</span><br><span class="line">falcon_sense_greedy=False</span><br><span class="line"></span><br><span class="line"><span class="comment">####Pread overlapping</span></span><br><span class="line">ovlp_daligner_option=-e.96 -l2000 -k24 -h1024 -w6 -s100</span><br><span class="line">ovlp_HPCdaligner_option=-v -B128 -M24</span><br><span class="line"></span><br><span class="line"><span class="comment">####Final Assembly</span></span><br><span class="line">overlap_filtering_setting=--max-diff 100 --max-cov 300 --min-cov 2</span><br><span class="line">fc_ovlp_to_graph_option=</span><br><span class="line">length_cutoff_pr=2000</span><br><span class="line"></span><br><span class="line">[job.defaults]</span><br><span class="line">job_type=<span class="built_in">local</span></span><br><span class="line">pwatcher_type=blocking</span><br><span class="line">JOB_QUEUE=default</span><br><span class="line">MB=32768</span><br><span class="line">NPROC=6</span><br><span class="line">njobs=32</span><br><span class="line">submit = /bin/bash -c <span class="string">&quot;<span class="variable">$&#123;JOB_SCRIPT&#125;</span>&quot;</span> &gt; <span class="string">&quot;<span class="variable">$&#123;JOB_STDOUT&#125;</span>&quot;</span> 2&gt; <span class="string">&quot;<span class="variable">$&#123;JOB_STDERR&#125;</span>&quot;</span></span><br><span class="line"></span><br><span class="line">[job.step.da]</span><br><span class="line">NPROC=4</span><br><span class="line">MB=32768</span><br><span class="line">njobs=32</span><br><span class="line">[job.step.la]</span><br><span class="line">NPROC=4</span><br><span class="line">MB=32768</span><br><span class="line">njobs=32</span><br><span class="line">[job.step.cns]</span><br><span class="line">NPROC=8</span><br><span class="line">MB=65536</span><br><span class="line">njobs=5</span><br><span class="line">[job.step.pla]</span><br><span class="line">NPROC=4</span><br><span class="line">MB=32768</span><br><span class="line">njobs=4</span><br><span class="line">[job.step.asm]</span><br><span class="line">NPROC=24</span><br><span class="line">MB=196608</span><br><span class="line">njobs=1</span><br></pre></td></tr></table></figure>

<p>根据注释信息，文件分为”input”, “Data partitioning”, “Repeat Masking”, “Pre-assembly”, “Pread overlapping”, “Final Assembly”, 以及最后的任务调度部分，让我们分别看下这里面的内容</p>
<h3 id="输入-Input"><a href="#输入-Input" class="headerlink" title="输入(Input)"></a>输入(Input)</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#### Input</span></span><br><span class="line">[General]</span><br><span class="line">input_fofn=input.fofn</span><br><span class="line">input_type=raw</span><br><span class="line">pa_DBdust_option=</span><br><span class="line">pa_fasta_filter_option=pass</span><br><span class="line">target=assembly</span><br><span class="line">skip_checks=False</span><br><span class="line">LA4Falcon_preload=<span class="literal">false</span></span><br></pre></td></tr></table></figure>

<p>输入这里参数比较简单，基本不需要做任何改动，除了 pa_fasta_filter_options，用于处理一个ZMW（测序翻译孔）有多条subread时，到底选择哪一条的问题。</p>
<ul>
<li>“pass”: 不做过滤，全部要。</li>
<li>“streamed-median”: 表示选择大于中位数的subread</li>
<li>“streamed-internal-median”: 当一个ZMW里的subread低于3条时选择最长，多于单条则选择大于中位数的subread</li>
</ul>
<blockquote>
<p>0.01版本pb-assembly的<code>pa_DBdust_option</code>有一个bug，也就是里面的参数不会传递给DBdust, DBdust是对read进行soft-masking，一般都用默认参数，因此这个bug问题不大。</p>
</blockquote>
<h3 id="数据分配-Data-Partitioning"><a href="#数据分配-Data-Partitioning" class="headerlink" title="数据分配(Data Partitioning)"></a>数据分配(Data Partitioning)</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pa_DBsplit_option=-x500 -s50</span><br><span class="line">ovlp_DBsplit_option=-x500 -s50</span><br></pre></td></tr></table></figure>

<p>这部分的设置会将参数传递给<code>DBsplit</code>，将数据进行拆分多个block，后续的并行计算都基于block。<code>-s 50</code>表示每个block大小为50M。 这适用于基因组比较小的物种，如果是大基因组则应该设置为<code>-s 200</code>或者<code>-s 400</code></p>
<h3 id="重复屏蔽-Repeat-Masking"><a href="#重复屏蔽-Repeat-Masking" class="headerlink" title="重复屏蔽(Repeat Masking)"></a>重复屏蔽(Repeat Masking)</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#### Repeat Masking</span></span><br><span class="line">pa_HPCTANmask_option=</span><br><span class="line">pa_REPmask_code=1,100;2,80;3,60</span><br></pre></td></tr></table></figure>

<p>屏蔽重复序列可以在不损失组装准确性的同时，提高后续组装的overlap&#x2F;daligner步骤10~20倍速度，见<a target="_blank" rel="noopener" href="https://dazzlerblog.wordpress.com/2016/04/01/detecting-and-soft-masking-repeats/">Detecting and Masking Repeats</a>.</p>
<p>pa_HPCTANmask_option的参数会传给串联重复步骤的<code>HPCTANmask</code>, 而<code>pa_REPmask_code</code>很复杂，它分为三次迭代，因此这里1:100;2,80;3,60 就表示第一次迭代检测每个block中出现超过100次的序列，第二次迭代将2个block合并一起检测超过80次的序列，第三次将3个block进行合并检测超过60次的序列。</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/pbbioconda/issues/20">https://github.com/PacificBiosciences/pbbioconda/issues/20</a></p>
</blockquote>
<h3 id="预组装（纠错）pre-assembly"><a href="#预组装（纠错）pre-assembly" class="headerlink" title="预组装（纠错）pre-assembly"></a>预组装（纠错）pre-assembly</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">####Pre-assembly</span></span><br><span class="line">genome_size=0</span><br><span class="line">seed_coverage=20</span><br><span class="line">length_cutoff=3000    </span><br><span class="line">pa_HPCdaligner_option=-v -B128 -M24</span><br><span class="line">pa_daligner_option=-e.7 -l1000 -k18 -h80 -w8 -s100</span><br><span class="line">falcon_sense_option=--output-multi --min-idt 0.70 --min-cov 2 --max-n-read 800</span><br><span class="line">falcon_sense_greedy=False</span><br></pre></td></tr></table></figure>

<p>当<code>length_cutoff=-1</code>时，设置<code>genome_size</code>和<code>seed_coverage</code>会自动计算要过滤的序列。否则是过滤低于一定长度的read。</p>
<p>pa_HPCdalinger_option参数不需要调整，-M表示每个进程的内存为24G，一般200M的block对应16G。</p>
<p>pa_daligner_option的参数比较重要：</p>
<ul>
<li><code>-e</code>:错误率，低质量序列设置为0.70，高质量设置为0.80。 值越高避免单倍型的坍缩</li>
<li><code>-l</code>:  最低overlap的长度，文库比较短时为1000， 文库比较长为5000.</li>
<li><code>-k</code>:  低质量数据为14，高质量数据为18</li>
<li><code>-h</code>: 表示完全match的k-mer所覆盖的碱基数。和<code>-l</code>, <code>-e</code>有关，越大越严格。预组装时最大也不要超过最低overlap长度的1&#x2F;4. 最低就设置为80</li>
</ul>
<h3 id="纠错后相互比对"><a href="#纠错后相互比对" class="headerlink" title="纠错后相互比对"></a>纠错后相互比对</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">####Pread overlapping</span></span><br><span class="line">ovlp_daligner_option=-e.96 -l2000 -k24 -h1024 -w6 -s100</span><br><span class="line">ovlp_HPCdaligner_option=-v -B128 -M24</span><br></pre></td></tr></table></figure>

<p>和上面的参数类似，但是-e的范围调整为0.93-0.96，-l范围调整为1800-6000， -k调整为 18-24</p>
<h3 id="最后组装"><a href="#最后组装" class="headerlink" title="最后组装"></a>最后组装</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">overlap_filtering_setting=--max-diff 100 --max-cov 300 --min-cov 2</span><br><span class="line">fc_ovlp_to_graph_option=</span><br><span class="line">length_cutoff_pr=2000</span><br></pre></td></tr></table></figure>

<p>这里的参数就可以随便调整了，因为这一步速度很快。 例如<code>length_cutoff_pr</code>就可以从2000，提高到15000.</p>
<p>最后还有一部分是任务投递系统，如果是单节点运行，需要注意设置 njobs，这是同时投递的任务数。假如你将[job.step.cns]按照如下的方式设置，那么同时会出现 $ 8 \times 50 &#x3D; 400 $ 个任务，如果你的内存只有128G，运行一段时间后你的所有内存就会被耗尽，那么基本上你就只能重启服务器了。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">[job.step.cns]</span><br><span class="line">NPROC=8</span><br><span class="line">MB=65536</span><br><span class="line">njobs=50</span><br></pre></td></tr></table></figure>

<h2 id="运行结果"><a href="#运行结果" class="headerlink" title="运行结果"></a>运行结果</h2><p>用上述的配置文件，以<code>fc_run fc_run.cfg</code>运行后，最后的<code>2-asm-falcon/p_ctg.fa</code>的序列数有4条，最长为4,685,024, 之后我将<code>length_cutoff_pr</code>调整为15k，<code>2-asm-falcon/p_ctg.fa</code>序列只有一条，长度为<code>4,638,411</code></p>
<p>下载<code>asm.gfa</code>到本地，用<a target="_blank" rel="noopener" href="https://rrwick.github.io/Bandage/">Bandage</a>可视化，可以发现组装效果不错。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/bandage-2c760dfa117940389cedc639a8aa9906.png" alt="Bandage"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/20/%E3%80%8C%E4%B8%89%E4%BB%A3%E7%BB%84%E8%A3%85%E3%80%8D%E4%BD%BF%E7%94%A8%E6%96%B0%E7%89%88Falcon%E8%BF%9B%E8%A1%8C%E4%B8%89%E4%BB%A3%E6%B5%8B%E5%BA%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85-Falcon-genome-assembly-tools/" data-id="clm1z22j900koehn5fqyebawm" data-title="「三代组装」使用新版Falcon进行三代测序基因组组装" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-Cicero_ 单细胞共开放分析- Cicero-single-cell-ATAC-seq-analyzing-tools" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/18/Cicero_%20%E5%8D%95%E7%BB%86%E8%83%9E%E5%85%B1%E5%BC%80%E6%94%BE%E5%88%86%E6%9E%90-%20Cicero-single-cell-ATAC-seq-analyzing-tools/" class="article-date">
  <time class="dt-published" datetime="2019-08-18T23:20:21.487Z" itemprop="datePublished">2019-08-18</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/18/Cicero_%20%E5%8D%95%E7%BB%86%E8%83%9E%E5%85%B1%E5%BC%80%E6%94%BE%E5%88%86%E6%9E%90-%20Cicero-single-cell-ATAC-seq-analyzing-tools/">Cicero:单细胞共开放分析</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>Cicero是一个单细胞染色质可及性实验可视化R包。Cicero的<strong>主要功能</strong>就是使用单细胞染色质可及性数据通过分析共开放去预测基因组上<strong>顺式调节作用</strong>(cis-regulatory interactions)，例如增强子和启动子。Cicero能够利用染色质开放性进行单细胞聚类，排序和差异可及性分析。关于Cicero的算法原理，参考<a target="_blank" rel="noopener" href="https://cole-trapnell-lab.github.io/cicero-release/papers/">他们发表的文章</a>.</p>
<h2 id="简介"><a href="#简介" class="headerlink" title="简介"></a>简介</h2><p>Cicero的主要目标是使用单细胞染色质可及性数据去预测基因组上那些在细胞核中存在物理邻近的区域。这能够用于鉴定潜在的<strong>增强子-启动子对</strong>，对基因组顺式作用有一个总体了解。</p>
<p>由于单细胞数据的稀疏性，细胞必须根据相似度进行聚合，才能够对数据中的多种技术因素进行可靠纠正。</p>
<p>最终，Cicero给出了”Cicero 共开放”得分，分数在-1和1之间，用于评估用于给定距离中每个可及peak对之间的共开放性，分数越高，越可能是共开放。</p>
<p>此外，Monocle3的框架让Cicero能对单细胞ATAC-seq实验完成Monocle3上的聚类，拟时间等分析。</p>
<p>Cicero主要提供了两种核心分析:</p>
<ul>
<li><strong>构建和分析顺式调控网络</strong>.</li>
<li><strong>常规单细胞染色质可及性分析</strong>:</li>
</ul>
<h2 id="Cicero安装和加载"><a href="#Cicero安装和加载" class="headerlink" title="Cicero安装和加载"></a>Cicero安装和加载</h2><p>Cicero运行在R语言分析环境中。尽管能够从Bioconductor安装，但是推荐从GitHub上安装。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">install.packages(&quot;BiocManager&quot;)</span><br><span class="line">BiocManager::install(c(&quot;Gviz&quot;, &quot;GenomicRanges&quot;, &quot;rtracklayer&quot;))</span><br><span class="line">install.packages(&quot;devtools&quot;)</span><br><span class="line">devtools::install_github(&quot;cole-trapnell-lab/cicero-release&quot;, ref = &quot;monocle3&quot;)</span><br></pre></td></tr></table></figure>

<p>加载Cicero</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">library(cicero)</span><br></pre></td></tr></table></figure>

<h2 id="数据加载"><a href="#数据加载" class="headerlink" title="数据加载"></a>数据加载</h2><p>Cicero以<code>cell_data_set(CDS)</code>对象存放数据，该对象继承自Bioconductor的<code>SingleCellExperiment</code>对象。我们可以用下面三个函数来操纵数据</p>
<ul>
<li>fData: 获取feature的元信息, 这里的featureI指的是peak</li>
<li>pData: 获取cell&#x2F;sample的元信息</li>
<li>assay: 获取cell-by-peak的count矩阵</li>
</ul>
<p>除了CDS外，还需要基因坐标信息和基因注释信息。</p>
<p>关于坐标信息，以mouse.mm9.genome为例，它是一个数据框，只有两列，一列是染色体编号，另一列对应的长度</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">data(&quot;mouse.mm9.genome&quot;)</span><br></pre></td></tr></table></figure>

<p>基因组注释信息可以用<code>rtracklayer::readGFF</code>加载</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">temp &lt;- tempfile()</span><br><span class="line">download.file(&quot;ftp://ftp.ensembl.org/pub/release-65/gtf/mus_musculus/Mus_musculus.NCBIM37.65.gtf.gz&quot;, temp)</span><br><span class="line">gene_anno &lt;- rtracklayer::readGFF(temp)</span><br><span class="line">unlink(temp)</span><br></pre></td></tr></table></figure>

<p>Bioconductor也有现成的TxDb，可以从中提取注释信息。</p>
<h3 id="简单稀疏矩阵格式"><a href="#简单稀疏矩阵格式" class="headerlink" title="简单稀疏矩阵格式"></a>简单稀疏矩阵格式</h3><p>以一个测试数据集介绍如何加载简单稀疏矩阵格式</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">temp &lt;- textConnection(readLines(gzcon(url(&quot;http://staff.washington.edu/hpliner/data/kidney_data.txt.gz&quot;))))</span><br><span class="line">cicero_data &lt;- read.table(temp)</span><br></pre></td></tr></table></figure>

<p>这里的简单稀疏矩阵格式指的是数据以三列进行存放，第一列是peak位置信息，第二列是样本信息，第三列则是样本中有多少fragment和该peak重叠。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chr1_4756552_4757256    GAATTCGTACCAGGCGCATTGGTAGTCGGGCTCTGA    1</span><br></pre></td></tr></table></figure>

<p>对于这种格式，Cicero提供了一个函数<code>make_atac_cds</code>, 用于构建一个有效的<code>cell_data_set</code>对象，用于下游分析，输入既可以是一个数据框，也可以是一个文件路径。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">input_cds &lt;- make_atac_cds(cicero_data, binarize = TRUE)</span><br></pre></td></tr></table></figure>

<h3 id="10X-scATAC-seq-数据"><a href="#10X-scATAC-seq-数据" class="headerlink" title="10X scATAC-seq 数据"></a>10X scATAC-seq 数据</h3><p>如果数据已经经过10X scATAC-seq处理，那么结果中的<strong>filtered_peak_bc_matrix</strong>里就有我们需要的信息</p>
<p>加载cell-by-peak count矩阵，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># read in matrix data using the Matrix package</span><br><span class="line">indata &lt;- Matrix::readMM(&quot;filtered_peak_bc_matrix/matrix.mtx&quot;) </span><br><span class="line"># binarize the matrix</span><br><span class="line">indata@x[indata@x &gt; 0] &lt;- 1</span><br></pre></td></tr></table></figure>

<p>加载cell元数据</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># format cell info</span><br><span class="line">cellinfo &lt;- read.table(&quot;filtered_peak_bc_matrix/barcodes.tsv&quot;)</span><br><span class="line">row.names(cellinfo) &lt;- cellinfo$V1</span><br><span class="line">names(cellinfo) &lt;- &quot;cells&quot;</span><br></pre></td></tr></table></figure>

<p>加载peak元数据</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># format peak info</span><br><span class="line">peakinfo &lt;- read.table(&quot;filtered_peak_bc_matrix/peaks.bed&quot;)</span><br><span class="line">names(peakinfo) &lt;- c(&quot;chr&quot;, &quot;bp1&quot;, &quot;bp2&quot;)</span><br><span class="line">peakinfo$site_name &lt;- paste(peakinfo$chr, peakinfo$bp1, peakinfo$bp2, sep=&quot;_&quot;)</span><br><span class="line">row.names(peakinfo) &lt;- peakinfo$site_name</span><br></pre></td></tr></table></figure>

<p>为cell-by-peak的count矩阵增加行名(peak元数据)和列名(cell元数据)</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">row.names(indata) &lt;- row.names(peakinfo)</span><br><span class="line">colnames(indata) &lt;- row.names(cellinfo)</span><br></pre></td></tr></table></figure>

<p>然后用<code>new_cell_data_set</code>根据peak元数据，cell元数据构建出一个<code>cell_data_set</code>对象</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># make CDS</span><br><span class="line">input_cds &lt;-  suppressWarnings(new_cell_data_set(indata,</span><br><span class="line">    cell_metadata = cellinfo,</span><br><span class="line">    gene_metadata = peakinfo))</span><br></pre></td></tr></table></figure>

<p>之后用<code>detect_genes</code>统计，对于每个基因在多少细胞中的表达量超过了给定阈值。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">input_cds &lt;- monocle3::detect_genes(input_cds)</span><br></pre></td></tr></table></figure>

<p>过滤没有细胞开放的peak</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">input_cds &lt;- input_cds[Matrix::rowSums(exprs(input_cds)) != 0,] </span><br></pre></td></tr></table></figure>

<h2 id="构建顺式作用网络"><a href="#构建顺式作用网络" class="headerlink" title="构建顺式作用网络"></a>构建顺式作用网络</h2><h3 id="运行Cicero"><a href="#运行Cicero" class="headerlink" title="运行Cicero"></a>运行Cicero</h3><h3 id="构建Cicero-CDS对象"><a href="#构建Cicero-CDS对象" class="headerlink" title="构建Cicero CDS对象"></a>构建Cicero CDS对象</h3><p>因为单细胞染色质开放数据特别稀疏，为了能够比较准确的估计共开放得分，需要将一些比较相近的细胞进行聚合，得到一个相对比较致密的count数据。Cicero使用KNN算法构建细胞的相互重叠集。而细胞之间距离关系则是根据降维后坐标信息。降维方法有很多种，这里以UMAP为例</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">input_cds &lt;- detect_genes(input_cds)</span><br><span class="line">input_cds &lt;- estimate_size_factors(input_cds)</span><br><span class="line"># PCA或LSI线性降维</span><br><span class="line">input_cds &lt;- preprocess_cds(input_cds, method = &quot;LSI&quot;)</span><br><span class="line"># UMAP非线性降维</span><br><span class="line">input_cds &lt;- reduce_dimension(input_cds, reduction_method = &#x27;UMAP&#x27;, </span><br><span class="line">                              preprocess_method = &quot;LSI&quot;)</span><br></pre></td></tr></table></figure>

<p>此时用<code>reducedDimNames(input_cds)</code>就会发现它多了LSI和UMAP两个信息，我们可以用<code>reducedDims</code>来提取UMAP坐标信息。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">umap_coords &lt;- reducedDims(input_cds)$UMAP</span><br></pre></td></tr></table></figure>

<p><code>make_cicero_cds</code>函数就需要其中UMAP坐标信息。</p>
<p><strong>注1</strong>: 假如你已经有了UMAP的坐标信息，那么就不需要运行<code>preprocess_cds</code>和<code>reduce_dimension</code>， 直接提供UMAP坐标信息给<code>make_cicero_cds</code>就可以了。</p>
<p><strong>注2</strong>: umap_coords的行名需要和<code>pData</code>得到表格中的行名一样, 即<code>all(row.names(umap_coords) == row.names(pData(input_cds)))</code>结果为TRUE。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cicero_cds &lt;- make_cicero_cds(input_cds, reduced_coordinates = umap_coords)</span><br></pre></td></tr></table></figure>

<p>此时的cicero_cds是没有UMAP降维信息。</p>
<h3 id="运行Cicero-1"><a href="#运行Cicero-1" class="headerlink" title="运行Cicero"></a>运行Cicero</h3><p>Cicero包的主要功能就是预测基因组上位点之间的共开放. 有两种方式可以获取该信息</p>
<ul>
<li><code>run_cicero</code>: 一步到位。默认参数适用于人类和小鼠，但是其他物种，参考<a target="_blank" rel="noopener" href="https://cole-trapnell-lab.github.io/cicero-release/docs_m3//#important-considerations-for-non-human-data">此处</a></li>
<li>单独运行每个函数，这样子获取中间的信息。</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">#测试部分数据</span><br><span class="line">data(&quot;mouse.mm9.genome&quot;)</span><br><span class="line">sample_genome &lt;- subset(mouse.mm9.genome, V1 == &quot;chr2&quot;)</span><br><span class="line">sample_genome$V2[1] &lt;- 10000000</span><br><span class="line">conns &lt;- run_cicero(cicero_cds, sample_genome, sample_num = 2) </span><br><span class="line">nrow(conns) # 212302</span><br><span class="line">## 全部数据, 时间真久</span><br><span class="line">conns &lt;- run_cicero(cicero_cds, mouse.mm9.genome, sample_num = 2)</span><br><span class="line">nrows(conns) #59741738</span><br></pre></td></tr></table></figure>

<p>其中conns存放的是peak之间共开放得分。</p>
<h3 id="Cicero-Connection可视化"><a href="#Cicero-Connection可视化" class="headerlink" title="Cicero Connection可视化"></a>Cicero Connection可视化</h3><p>有了peak之间共开放的可能性得分后，就可以画一些很酷的图了。用到的函数是<code>plot_connections</code>, 这个函数有很多参数，所以最终呈现结果取决于你的<a target="_blank" rel="noopener" href="https://cole-trapnell-lab.github.io/cicero-release/docs/#advanced-visualizaton">参数调整</a>。</p>
<p>获取基因组注释信息</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">temp &lt;- tempfile()</span><br><span class="line">download.file(&quot;ftp://ftp.ensembl.org/pub/release-65/gtf/mus_musculus/Mus_musculus.NCBIM37.65.gtf.gz&quot;, temp)</span><br><span class="line">gene_anno &lt;- rtracklayer::readGFF(temp)</span><br><span class="line">unlink(temp)</span><br><span class="line"></span><br><span class="line">gene_anno$chromosome &lt;- paste0(&quot;chr&quot;, gene_anno$seqid)</span><br><span class="line">gene_anno$gene &lt;- gene_anno$gene_id</span><br><span class="line">gene_anno$transcript &lt;- gene_anno$transcript_id</span><br><span class="line">gene_anno$symbol &lt;- gene_anno$gene_name</span><br></pre></td></tr></table></figure>

<p>除了从GFF文件中提取外，还可以利用<code>txdb</code>提取转录本位置信息</p>
<p>举例子，下面的函数就是绘制染色体上<code>&quot;chr2:9773451-9848598</code>的共开放情况，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">plot_connections(conns, &quot;chr2&quot;, 9773451, 9848598,</span><br><span class="line">                 gene_model = gene_anno, </span><br><span class="line">                 coaccess_cutoff = .25, </span><br><span class="line">                 connection_width = .5, </span><br><span class="line">                 collapseTranscripts = &quot;longest&quot; )</span><br></pre></td></tr></table></figure>

<p>假如之前用的全基因组来算connections，那么这一步也会很久才能出图，所以建议先自己提取部分数据，这样子作图的速度就会快很多(因为源代码没有优化好矩阵过滤)</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">conns_sub &lt;- conns[conns_matrix[,1] == &quot;chr2&quot; &amp; </span><br><span class="line">                     as.numeric(conns_matrix[,2]) &gt; 8000000 &amp;</span><br><span class="line">                     as.numeric(conns_matrix[,2]) &lt; 10000000 </span><br><span class="line">                     , ]</span><br><span class="line"></span><br><span class="line">plot_connections(conns_sub, &quot;chr2&quot;, 9773451, 9848598,</span><br><span class="line">                 gene_model = gene_anno, </span><br><span class="line">                 coaccess_cutoff = .25, </span><br><span class="line">                 connection_width = .5, </span><br><span class="line">                 collapseTranscripts = &quot;longest&quot; )</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/connection_plot-a71e2def92334614add0e4ce2280bbd8.png" alt="connections"></p>
<h2 id="计算基因活跃度得分"><a href="#计算基因活跃度得分" class="headerlink" title="计算基因活跃度得分"></a>计算基因活跃度得分</h2><p>在启动子区的开放度并不是基因表达很好的预测子。使用Cicero links，我们就能得到启动子和它相关的远距位点的总体开放情况。这种区域间开放度联合得分和基因表达的相关度更高。作者称之为<strong>基因活跃度得分</strong>（Cicero gene activity score)。 可以用两个函数进行计算</p>
<p><code>build_gene_activity_matrix</code>函数接受CDS和Cicero connection list作为输入，返回一个标准化的基因得分。<strong>注意</strong>， CDS必须在fData 表中有一列”gene”, 如果该peak在启动子区就用对应的基因名进行注释, 否则为NA. 我们可以用<code>annotate_cds_by_site</code>进行注释。</p>
<p><code>build_gene_activitity_matrix</code>是未标准化的数据。需要用<code>normalize_gene_activities</code>进行标准化。如果你想比较不同数据集中部分数据的基因活跃度得分, 那么所有的gene activity 部分数据都需要一起标准化。如果只是想标准化其中一个数据集，那么传递一个结果即可。<code>normalize_gene_activities</code>也需要一个每个细胞所有开放位点的命名向量。这存放在CDS对象中的<code>pData</code>，在<code>num_genes_expressed</code>列。</p>
<p>可能还是不是明白，看代码吧。</p>
<p>第一步，是用<code>annotate_cds_by_site</code>对CDS对象进行注释在启动子区域的peak。</p>
<p>我们先从GFF文件中提取启动子的位置，也就是每个转录本的第一个外显子</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"># 分别提取正链和负链的注释, 仅保留第一个外显子</span><br><span class="line">pos &lt;- subset(gene_anno, strand == &quot;+&quot;)</span><br><span class="line">pos &lt;- pos[order(pos$start),] </span><br><span class="line"></span><br><span class="line"># remove all but the first exons per transcript</span><br><span class="line">pos &lt;- pos[!duplicated(pos$transcript),] </span><br><span class="line"># make a 1 base pair marker of the TSS</span><br><span class="line">pos$end &lt;- pos$start + 1 </span><br><span class="line"></span><br><span class="line">neg &lt;- subset(gene_anno, strand == &quot;-&quot;)</span><br><span class="line">neg &lt;- neg[order(neg$start, decreasing = TRUE),] </span><br><span class="line"># remove all but the first exons per transcript</span><br><span class="line">neg &lt;- neg[!duplicated(neg$transcript),] </span><br><span class="line">neg$start &lt;- neg$end - 1</span><br><span class="line"></span><br><span class="line"># 合并正链和负链</span><br><span class="line">gene_annotation_sub &lt;- rbind(pos, neg)</span><br><span class="line"></span><br><span class="line"># 只需要4列信息</span><br><span class="line">gene_annotation_sub &lt;- gene_annotation_sub[,c(&quot;chromosome&quot;, &quot;start&quot;, &quot;end&quot;, &quot;symbol&quot;)]</span><br><span class="line"></span><br><span class="line"># 第四列命名为gene</span><br><span class="line">names(gene_annotation_sub)[4] &lt;- &quot;gene&quot;</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>用<code>annotate_cds_by_site</code>对cds按照gene_annotation_sub注释。会多处两列，overlap和gene。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">input_cds &lt;- annotate_cds_by_site(input_cds, gene_annotation_sub)</span><br><span class="line">tail(fData(input_cds))</span><br></pre></td></tr></table></figure>

<p>第二步: 计算未标准化的基因活跃度得分</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># generate unnormalized gene activity matrix</span><br><span class="line">unnorm_ga &lt;- build_gene_activity_matrix(input_cds, conns)</span><br><span class="line"></span><br><span class="line"># 过滤全为0的行和列</span><br><span class="line">unnorm_ga &lt;- unnorm_ga[!Matrix::rowSums(unnorm_ga) == 0, </span><br><span class="line">                       !Matrix::colSums(unnorm_ga) == 0]</span><br></pre></td></tr></table></figure>

<p>第三步: 标准化基因活跃度得分</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># 你需要一个命名列表</span><br><span class="line">num_genes &lt;- pData(input_cds)$num_genes_expressed</span><br><span class="line">names(num_genes) &lt;- row.names(pData(input_cds))</span><br><span class="line"></span><br><span class="line"># 标准化</span><br><span class="line">cicero_gene_activities &lt;- normalize_gene_activities(unnorm_ga, num_genes)</span><br></pre></td></tr></table></figure>

<p>如果你有多个数据集需要进行标注化，那你就需要传递一个列表给<code>normalize_gene_activities</code>, 并且保证<code>num_genes</code>里包括所有数据集中的所有样本。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># if you had two datasets to normalize, you would pass both:</span><br><span class="line"># num_genes should then include all cells from both sets</span><br><span class="line">unnorm_ga2 &lt;- unnorm_ga</span><br><span class="line">cicero_gene_activities &lt;- normalize_gene_activities(list(unnorm_ga, unnorm_ga2), </span><br><span class="line">                                                    num_genes)</span><br></pre></td></tr></table></figure>

<p>标注化后的基因活跃度得分在0到1之间,</p>
<p>第四步: 在UMAP上可视化基因得分。由于值比较小，要先扩大1000000倍，即log2(GA * 1000000 + 1)，这样子在可视化的时候比较好看。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">plot_umap_ga &lt;- function(cds,</span><br><span class="line">                         cicero_gene_activities,</span><br><span class="line">                         gene,</span><br><span class="line">                         dotSize = 1,</span><br><span class="line">                         log = TRUE)&#123;</span><br><span class="line">  #umap_df &lt;- as.data.frame(colData(cds))</span><br><span class="line">  umap_df &lt;- as.data.frame(reducedDims(input_cds)$UMAP)</span><br><span class="line">  colnames(umap_df) &lt;- c(&quot;UMAP1&quot;,&quot;UMAP2&quot;)</span><br><span class="line">  gene_pos &lt;- which(row.names(cicero_gene_activities) %in% gene)</span><br><span class="line"></span><br><span class="line">  umap_df$score &lt;- log2(cicero_gene_activities[gene_pos,] * 1000000 + 1)</span><br><span class="line">  if (log)&#123;</span><br><span class="line">    umap_df$score &lt;- log(umap_df$score + 1)</span><br><span class="line">  &#125;</span><br><span class="line"></span><br><span class="line">  p &lt;- ggplot(umap_df, aes(x=UMAP1, y=UMAP2)) +</span><br><span class="line">    geom_point(aes(color=score), size = dotSize) + scale_color_viridis(option=&quot;magma&quot;) +</span><br><span class="line">    theme_bw()</span><br><span class="line">  return(p)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h2 id="拟时序分析"><a href="#拟时序分析" class="headerlink" title="拟时序分析"></a>拟时序分析</h2><p>这部分的分析其实比较简单，因为仅仅是将feature从mRNA变成了peak而已。步骤如下</p>
<ol>
<li>数据预处理</li>
<li>数据降维</li>
<li>细胞聚类</li>
<li>轨迹图学习</li>
<li>在拟时间上对细胞进行排序</li>
</ol>
<p>代码如下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"># 读取数据</span><br><span class="line">temp &lt;- textConnection(readLines(gzcon(url(&quot;http://staff.washington.edu/hpliner/data/kidney_data.txt.gz&quot;))))set.seed(2017)</span><br><span class="line">input_cds &lt;- estimate_size_factors(input_cds)</span><br><span class="line">#1 数据预处理</span><br><span class="line">input_cds &lt;- preprocess_cds(input_cds, method = &quot;LSI&quot;)</span><br><span class="line">#2 降维</span><br><span class="line">input_cds &lt;- reduce_dimension(input_cds, reduction_method = &#x27;UMAP&#x27;, </span><br><span class="line">preprocess_method = &quot;LSI&quot;)</span><br><span class="line">#3 细胞聚类</span><br><span class="line">input_cds &lt;- cluster_cells(input_cds)</span><br><span class="line">#4 轨迹图学习</span><br><span class="line">input_cds &lt;- learn_graph(input_cds)</span><br><span class="line">#5 对细胞进行排序</span><br><span class="line">input_cds &lt;- order_cells(input_cds, </span><br><span class="line">                         root_cells = &quot;GAGATTCCAGTTGAATCACTCCATCGAGATAGAGGC&quot;)</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>最后可视化</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plot_cells(input_cds, color_cells_by = &quot;pseudotime&quot;)</span><br></pre></td></tr></table></figure>

<h3 id="差异开放分析"><a href="#差异开放分析" class="headerlink" title="差异开放分析"></a>差异开放分析</h3><p>当你的细胞在拟时间中排列之后，你可能会想知道基因组哪些区域会随着时间发生变化。</p>
<p>如果你已经知道哪些区域会随着时间发生变化，你可以用<code>plot_accessibility_in_pseudotime</code>进行可视化</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">input_cds_lin &lt;- input_cds[,is.finite(pseudotime(input_cds))]</span><br><span class="line">plot_accessibility_in_pseudotime(input_cds_lin[c(&quot;chr1_3238849_3239700&quot;, </span><br><span class="line">                                                 &quot;chr1_3406155_3407044&quot;, </span><br><span class="line">                                                 &quot;chr1_3397204_3397842&quot;)])</span><br></pre></td></tr></table></figure>

<p>当然我们也可以从头分析. 先用<code>aggregate_by_cell_bin</code>对细胞根据相似度进行聚合</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># First, assign a column in the pData table to umap pseudotime</span><br><span class="line">pData(input_cds_lin)$Pseudotime &lt;- pseudotime(input_cds_lin)</span><br><span class="line">pData(input_cds_lin)$cell_subtype &lt;- cut(pseudotime(input_cds_lin), 10)</span><br><span class="line">binned_input_lin &lt;- aggregate_by_cell_bin(input_cds_lin, &quot;cell_subtype&quot;)</span><br></pre></td></tr></table></figure>

<p>之后用Monocle3的<code>fit_models</code>函数寻找差异开发的区域。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"># For speed, run fit_models on 1000 randomly chosen genes</span><br><span class="line">set.seed(1000)</span><br><span class="line">acc_fits &lt;- fit_models(binned_input_lin[sample(1:nrow(fData(binned_input_lin)), 1000),], </span><br><span class="line">                       model_formula_str = &quot;~Pseudotime + num_genes_expressed&quot; )</span><br><span class="line">fit_coefs &lt;- coefficient_table(acc_fits)</span><br><span class="line"></span><br><span class="line"># Subset out the differentially accessible sites with respect to Pseudotime</span><br><span class="line">pseudotime_terms &lt;- subset(fit_coefs, term == &quot;Pseudotime&quot; &amp; q_value &lt; .05)</span><br><span class="line">head(pseudotime_terms)</span><br></pre></td></tr></table></figure>

<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li>官方教程:<a target="_blank" rel="noopener" href="https://cole-trapnell-lab.github.io/cicero-release/docs_m3">https://cole-trapnell-lab.github.io/cicero-release/docs_m3</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/18/Cicero_%20%E5%8D%95%E7%BB%86%E8%83%9E%E5%85%B1%E5%BC%80%E6%94%BE%E5%88%86%E6%9E%90-%20Cicero-single-cell-ATAC-seq-analyzing-tools/" data-id="clm1z22ev005wehn5a7sj87e5" data-title="Cicero:单细胞共开放分析" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-motifmatchr_ 在R语言中分析peak中里是否有motif匹配-motifmatchr-peak-motif-search-in-R" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/17/motifmatchr_%20%E5%9C%A8R%E8%AF%AD%E8%A8%80%E4%B8%AD%E5%88%86%E6%9E%90peak%E4%B8%AD%E9%87%8C%E6%98%AF%E5%90%A6%E6%9C%89motif%E5%8C%B9%E9%85%8D-motifmatchr-peak-motif-search-in-R/" class="article-date">
  <time class="dt-published" datetime="2019-08-17T15:22:37.235Z" itemprop="datePublished">2019-08-17</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/17/motifmatchr_%20%E5%9C%A8R%E8%AF%AD%E8%A8%80%E4%B8%AD%E5%88%86%E6%9E%90peak%E4%B8%AD%E9%87%8C%E6%98%AF%E5%90%A6%E6%9C%89motif%E5%8C%B9%E9%85%8D-motifmatchr-peak-motif-search-in-R/">motifmatchr:在R语言中分析peak中里是否有motif匹配</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p><code>motifmatchr</code>的作用就是分析众多的序列和众多的motifs, 从中找到哪个序列包含哪个motif. 它的核心函数就是<code>matchMotifs</code>，最大特点就是快，因为它用的是MOODS C++库用于motif匹配。</p>
<p>尽管Bioconductor上也有很多工具能够做motif匹配，比如说<code>Biostrings::mathcPWM</code>, <code>TFBSTools::searchSeq</code>，但是motifmatchr更适合于分析许多不同的序列包含许多不同的motif。例如，当分析ChIP-seq或者ATAC-seq数据时, 你可能会想知道在哪个peak里有哪种类型的motif.</p>
<h2 id="R包安装和加载"><a href="#R包安装和加载" class="headerlink" title="R包安装和加载"></a>R包安装和加载</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>motifmatchr<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>GenomicRanges<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="matchMotifs"><a href="#matchMotifs" class="headerlink" title="matchMotifs"></a>matchMotifs</h2><p>motifmatchr的核心函数是<code>matchMotifs</code>，因此了解这个函数的输入数据是什么格式就行了。必须的两个输入是</p>
<ul>
<li>位置权重矩阵(position weight matrices, PWM)或位置频率矩阵(position frequency matrices, PFM), 保存在PWMatrix, PFMatrix, PWMatrixList或PFMatrixList</li>
<li>一组基因组范围(GenomicRanges或RangedSUmmarizedExperiment对象)或一组序列(DNAStringSet, DNAString 或 简单的字符串向量)</li>
</ul>
<h3 id="PWM或PFM矩阵"><a href="#PWM或PFM矩阵" class="headerlink" title="PWM或PFM矩阵"></a>PWM或PFM矩阵</h3><p>Motif的PWM或PFM矩阵可以从<a target="_blank" rel="noopener" href="http://jaspar.genereg.net/">JASPAR</a>, <a target="_blank" rel="noopener" href="http://cisbp.ccbr.utoronto.ca/">CIS-BP</a>下载。</p>
<p>例如，我从CIS-BP下载拟南芥Arabidopsis_thaliana_2019_08_17_2_32_am.zip，解压缩后里有一个<code>pwm_all_motifs</code>文件夹，里面的文本文件存放的就是我所需要的PWM矩阵,  下一步根据这些文件构建出matchMotifs的输入对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">motif_dir <span class="operator">&lt;-</span> <span class="string">&quot;Arabidopsis_thaliana_2019_08_17_2_32_am/pwms_all_motifs&quot;</span></span><br><span class="line">PWList <span class="operator">&lt;-</span> <span class="built_in">list</span><span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>file <span class="keyword">in</span> list.files<span class="punctuation">(</span>motif_dir<span class="punctuation">,</span> pattern <span class="operator">=</span> <span class="string">&quot;.txt&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  df <span class="operator">&lt;-</span> read.table<span class="punctuation">(</span>file.path<span class="punctuation">(</span>motif_dir<span class="punctuation">,</span> file<span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">                   header <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">,</span></span><br><span class="line">                   row.names <span class="operator">=</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">  mt <span class="operator">&lt;-</span> as.matrix<span class="punctuation">(</span>df<span class="punctuation">)</span></span><br><span class="line">  <span class="keyword">if</span> <span class="punctuation">(</span>nrow<span class="punctuation">(</span>mt<span class="punctuation">)</span> <span class="operator">==</span><span class="number">0</span><span class="punctuation">)</span> <span class="keyword">next</span></span><br><span class="line">  motif_id <span class="operator">&lt;-</span> substr<span class="punctuation">(</span>file<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span><span class="number">6</span><span class="punctuation">)</span></span><br><span class="line">  PWList<span class="punctuation">[[</span>motif_id<span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> PWMatrix<span class="punctuation">(</span>ID <span class="operator">=</span> motif_id<span class="punctuation">,</span> profileMatrix <span class="operator">=</span> t<span class="punctuation">(</span>mt<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">PWMatrixLists <span class="operator">&lt;-</span> do.call<span class="punctuation">(</span>PWMatrixList<span class="punctuation">,</span>PWList<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>对于JASPAR，我们有更加方便的提取方法</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>JASPAR2018<span class="punctuation">)</span></span><br><span class="line">species <span class="operator">&lt;-</span> <span class="string">&quot;Arabidopsis thaliana&quot;</span></span><br><span class="line">collection <span class="operator">&lt;-</span> <span class="string">&quot;CORE&quot;</span></span><br><span class="line">opts <span class="operator">&lt;-</span> <span class="built_in">list</span><span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">opts<span class="punctuation">[</span><span class="string">&quot;species&quot;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> species</span><br><span class="line">opts<span class="punctuation">[</span><span class="string">&quot;collection&quot;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> collection</span><br><span class="line">out  <span class="operator">&lt;-</span> TFBSTools<span class="operator">::</span>getMatrixSet<span class="punctuation">(</span>JASPAR2018<span class="punctuation">,</span> opts<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>isTRUE<span class="punctuation">(</span>all.equal<span class="punctuation">(</span>TFBSTools<span class="operator">::</span>name<span class="punctuation">(</span>out<span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">names</span><span class="punctuation">(</span>out<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> </span><br><span class="line">  <span class="built_in">names</span><span class="punctuation">(</span>out<span class="punctuation">)</span> <span class="operator">&lt;-</span> paste<span class="punctuation">(</span><span class="built_in">names</span><span class="punctuation">(</span>out<span class="punctuation">)</span><span class="punctuation">,</span> TFBSTools<span class="operator">::</span>name<span class="punctuation">(</span>out<span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">                      sep <span class="operator">=</span> <span class="string">&quot;_&quot;</span><span class="punctuation">)</span></span><br><span class="line">motif <span class="operator">&lt;-</span> out</span><br></pre></td></tr></table></figure>

<p>JASPAR的motif比较可靠，因此motif相对比较少。</p>
<h3 id="给定范围或序列"><a href="#给定范围或序列" class="headerlink" title="给定范围或序列"></a>给定范围或序列</h3><p>这部分的输入就比较容易获取，你可以提供MACS2的输出BED，利用<code>rtracklayer::import.bed()</code>读取构建成一个GRanges对象。因为是提供的GRanges对象，那么还需要额外设置一个参数<code>genome</code>, 利用<code>Biostrings::readDNAStringSet()</code>读取一个参考基因组序列就行了。</p>
<p>或者用bedtools先根据bed文件提取数据，然后利用<code>Biostrings::readDNAStringSet()</code>读取</p>
<h3 id="示例数据"><a href="#示例数据" class="headerlink" title="示例数据"></a>示例数据</h3><p>我们以包中提供的数据为例，进行演示</p>
<p>加载实例的motif和GRanges对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># load some example motifs</span></span><br><span class="line">data<span class="punctuation">(</span>example_motifs<span class="punctuation">,</span> package <span class="operator">=</span> <span class="string">&quot;motifmatchr&quot;</span><span class="punctuation">)</span> </span><br><span class="line"></span><br><span class="line"><span class="comment"># Make a set of peaks</span></span><br><span class="line">peaks <span class="operator">&lt;-</span> GRanges<span class="punctuation">(</span>seqnames <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;chr1&quot;</span><span class="punctuation">,</span><span class="string">&quot;chr2&quot;</span><span class="punctuation">,</span><span class="string">&quot;chr2&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">                 ranges <span class="operator">=</span> IRanges<span class="punctuation">(</span>start <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">76585873</span><span class="punctuation">,</span><span class="number">42772928</span><span class="punctuation">,</span><span class="number">100183786</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">                                  width <span class="operator">=</span> <span class="number">500</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>获取motif在peak中的位置</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">motif_ix <span class="operator">&lt;-</span> matchMotifs<span class="punctuation">(</span>example_motifs<span class="punctuation">,</span> peaks<span class="punctuation">,</span> genome <span class="operator">=</span> <span class="string">&quot;hg19&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>用<code>motifMatches</code>函数提取匹配矩阵</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">motifMatches<span class="punctuation">(</span>motif_ix<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>输出结果是一个稀疏矩阵</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="number">3</span> x <span class="number">3</span> sparse Matrix of <span class="built_in">class</span> <span class="string">&quot;lgCMatrix&quot;</span></span><br><span class="line">     MA0599.1_KLF5 MA0107.1_RELA MA0137.3_STAT1</span><br><span class="line"><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">,</span><span class="punctuation">]</span>             <span class="operator">|</span>             <span class="operator">|</span>              .</span><br><span class="line"><span class="punctuation">[</span><span class="number">2</span><span class="punctuation">,</span><span class="punctuation">]</span>             <span class="operator">|</span>             .              <span class="operator">|</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">3</span><span class="punctuation">,</span><span class="punctuation">]</span>             <span class="operator">|</span>             .              <span class="operator">|</span></span><br></pre></td></tr></table></figure>

<p>其中的<code>.</code>就表示存在motif匹配。</p>
<p>我们还可以提取motif在peak中的位置</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Get motif positions within peaks for example motifs in peaks </span></span><br><span class="line">motif_ix <span class="operator">&lt;-</span> matchMotifs<span class="punctuation">(</span>example_motifs<span class="punctuation">,</span> peaks<span class="punctuation">,</span> genome <span class="operator">=</span> <span class="string">&quot;hg19&quot;</span><span class="punctuation">,</span></span><br><span class="line">                         out <span class="operator">=</span> <span class="string">&quot;positions&quot;</span><span class="punctuation">)</span> </span><br></pre></td></tr></table></figure>

<h3 id="其他参数"><a href="#其他参数" class="headerlink" title="其他参数"></a>其他参数</h3><p>除了必须的motif信息和你的序列信息输入外，还有一些其他的参数可以做一些设置。</p>
<ul>
<li>背景核苷酸频率, 默认是<code>bg=subject</code>, 也就是你的输入数据作为背景，也可设置成<code>genome</code>或<code>even</code></li>
<li>P值: 用于判断匹配是否足够可信的参数，默认是0.00005，基本上不用修改</li>
<li>输出信息: <code>matchMotifs</code>可以返回三种可能输出，matches, scores 和 positions</li>
</ul>
<p>总的来说，这个R包是一个比较简单的工具，比较容易上手使用。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/17/motifmatchr_%20%E5%9C%A8R%E8%AF%AD%E8%A8%80%E4%B8%AD%E5%88%86%E6%9E%90peak%E4%B8%AD%E9%87%8C%E6%98%AF%E5%90%A6%E6%9C%89motif%E5%8C%B9%E9%85%8D-motifmatchr-peak-motif-search-in-R/" data-id="clm1z22bk001wehn58c210xxt" data-title="motifmatchr:在R语言中分析peak中里是否有motif匹配" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-R语言的稀疏矩阵学习记录-R-Sparse-Matrix-Note" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/17/R%E8%AF%AD%E8%A8%80%E7%9A%84%E7%A8%80%E7%96%8F%E7%9F%A9%E9%98%B5%E5%AD%A6%E4%B9%A0%E8%AE%B0%E5%BD%95-R-Sparse-Matrix-Note/" class="article-date">
  <time class="dt-published" datetime="2019-08-17T10:18:28.869Z" itemprop="datePublished">2019-08-17</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/17/R%E8%AF%AD%E8%A8%80%E7%9A%84%E7%A8%80%E7%96%8F%E7%9F%A9%E9%98%B5%E5%AD%A6%E4%B9%A0%E8%AE%B0%E5%BD%95-R-Sparse-Matrix-Note/">R语言的稀疏矩阵学习记录</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>一个很大的矩阵， 320127 行,  8189列，假如用一个全为0的普通矩阵来存储，需要用到9.8Gb</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">cols <span class="operator">&lt;-</span> 8189</span><br><span class="line">rows <span class="operator">&lt;-</span> 320127</span><br><span class="line">mat <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span>data <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> nrow<span class="operator">=</span><span class="number">320127</span><span class="punctuation">,</span> ncol <span class="operator">=</span> <span class="number">8189</span><span class="punctuation">)</span></span><br><span class="line">print<span class="punctuation">(</span>object.size<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">,</span> unit<span class="operator">=</span><span class="string">&quot;GB&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 19.5 Gb</span></span><br><span class="line">mat <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span>data <span class="operator">=</span> <span class="number">0L</span><span class="punctuation">,</span> nrow<span class="operator">=</span><span class="number">320127</span><span class="punctuation">,</span> ncol <span class="operator">=</span> <span class="number">8189</span><span class="punctuation">)</span></span><br><span class="line">print<span class="punctuation">(</span>object.size<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">,</span> unit<span class="operator">=</span><span class="string">&quot;GB&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 9.8 Gb这里的0其实也要区分</span></span><br></pre></td></tr></table></figure>

<p>这里的<code>0L</code>表示数据类型是<code>integer</code>，默认是<code>numeric</code>. 这两者最大的区别在于，当你用<code>320127L * 8189L</code>，你会得到一个NA，而<code>320127 * 8189</code>不会</p>
<p>如果用稀疏矩阵保存的话</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">mat <span class="operator">&lt;-</span> Matrix<span class="punctuation">(</span>data <span class="operator">=</span> <span class="number">0L</span><span class="punctuation">,</span> nrow<span class="operator">=</span><span class="number">320127</span><span class="punctuation">,</span> ncol <span class="operator">=</span> <span class="number">8189</span><span class="punctuation">,</span> sparse <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">print<span class="punctuation">(</span>object.size<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">,</span> unit<span class="operator">=</span><span class="string">&quot;GB&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">#0 Gb</span></span><br><span class="line"><span class="built_in">dim</span><span class="punctuation">(</span>mat<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#[1] 320127   8189</span></span><br></pre></td></tr></table></figure>

<p>虽然行列数一样，但是稀疏矩阵几乎不占用任何内存。而且普通矩阵支持的运算，比如说求行和，求列和，提取元素的操作，在稀疏矩阵矩阵也是可以的，只不过会多花一点点时间而已。同时还有很多R包支持稀疏矩阵，比如说<code>glmnet</code>，一个做lasso回归的R包。</p>
<p>虽然看起来稀疏矩阵很美好，但是在R语言中那么大的稀疏矩阵的部分操作会出错</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> mat2 <span class="operator">&lt;-</span> mat <span class="operator">+</span> <span class="number">1</span></span><br><span class="line">Error <span class="keyword">in</span> asMethod<span class="punctuation">(</span>object<span class="punctuation">)</span> <span class="operator">:</span> </span><br><span class="line">  Cholmod error <span class="string">&#x27;problem too large&#x27;</span> at file ..<span class="operator">/</span>Core<span class="operator">/</span>cholmod_dense.c<span class="punctuation">,</span> line <span class="number">105</span></span><br></pre></td></tr></table></figure>

<p>即便是我想把它用<code>as.matrix</code>转回普通矩阵，它也报错了</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> mat3 <span class="operator">&lt;-</span> Matrix<span class="operator">::</span>as.matrix<span class="punctuation">(</span>mat<span class="punctuation">)</span></span><br><span class="line">Error <span class="keyword">in</span> asMethod<span class="punctuation">(</span>object<span class="punctuation">)</span> <span class="operator">:</span> </span><br><span class="line">  Cholmod error <span class="string">&#x27;problem too large&#x27;</span> at file ..<span class="operator">/</span>Core<span class="operator">/</span>cholmod_dense.c<span class="punctuation">,</span> line <span class="number">105</span></span><br></pre></td></tr></table></figure>

<p>既然现成的<code>as.matrix</code>无法处理，那怎么办呢？最简单粗暴的方法就是新建一个普通矩阵，然后对稀疏矩阵进行遍历，将稀疏矩阵的值挨个放回到的普通矩阵上。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">mat2 <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span>data <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> nrow<span class="operator">=</span><span class="number">320127</span><span class="punctuation">,</span> ncol <span class="operator">=</span> <span class="number">8189</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_len</span><span class="punctuation">(</span>nrow<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    <span class="keyword">for</span> <span class="punctuation">(</span>j <span class="keyword">in</span> <span class="built_in">seq_len</span><span class="punctuation">(</span>ncol<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">        mat2<span class="punctuation">[</span>i<span class="punctuation">,</span>j<span class="punctuation">]</span> <span class="operator">&lt;-</span> mat<span class="punctuation">[</span>i<span class="punctuation">,</span>j<span class="punctuation">]</span></span><br><span class="line">    <span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>那么这大概要多少时间呢？反正我的电脑跑了2个小时也没有跑完，所以你也别测试了。</p>
<p>那有没有办法可以加速呢？加速的方法就是减少for循环的次数，因为我们是一个稀疏矩阵，大部分的空间都是0，我们只需要将不为0的部分赋值给新矩阵即可。</p>
<p>这需要我们去了解下稀疏矩阵的数据结构</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> str<span class="punctuation">(</span>mat<span class="punctuation">)</span></span><br><span class="line">Formal <span class="built_in">class</span> <span class="string">&#x27;dgCMatrix&#x27;</span> <span class="punctuation">[</span>package <span class="string">&quot;Matrix&quot;</span><span class="punctuation">]</span> with <span class="number">6</span> slots</span><br><span class="line">  ..<span class="operator">@</span> i       <span class="operator">:</span> int<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">)</span> </span><br><span class="line">  ..<span class="operator">@</span> p       <span class="operator">:</span> int <span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">8190</span><span class="punctuation">]</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> ...</span><br><span class="line">  ..<span class="operator">@</span> Dim     <span class="operator">:</span> int <span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">2</span><span class="punctuation">]</span> <span class="number">320127</span> <span class="number">8189</span></span><br><span class="line">  ..<span class="operator">@</span> Dimnames<span class="operator">:</span>List of <span class="number">2</span></span><br><span class="line">  .. ..<span class="operator">$</span> <span class="operator">:</span> <span class="literal">NULL</span></span><br><span class="line">  .. ..<span class="operator">$</span> <span class="operator">:</span> <span class="literal">NULL</span></span><br><span class="line">  ..<span class="operator">@</span> x       <span class="operator">:</span> num<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">)</span> </span><br><span class="line">  ..<span class="operator">@</span> factors <span class="operator">:</span> <span class="built_in">list</span><span class="punctuation">(</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><code>@Dim</code>记录矩阵的维度信息, <code>@Dimnames</code>记录行名和列名, <code>@x</code>记录不为0的数值。<code>@i</code>记录不为0的行索引，和<code>@x</code>对应，这里全为0，所以不记录。<code>@p</code>比较复杂，并不是简单的记录不为0值的列索引，看文档也不知道是啥，不过通过检索可以找到它和不为0值的列索引的换算关系。</p>
<p>因此代码优化为</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">row_pos <span class="operator">&lt;-</span> mat<span class="operator">@</span>i<span class="operator">+</span><span class="number">1</span></span><br><span class="line">col_pos <span class="operator">&lt;-</span> findInterval<span class="punctuation">(</span>seq<span class="punctuation">(</span>mat<span class="operator">@</span>x<span class="punctuation">)</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span>mat<span class="operator">@</span>p<span class="punctuation">[</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="operator">+</span><span class="number">1</span></span><br><span class="line">val <span class="operator">&lt;-</span> mat<span class="operator">@</span>x</span><br><span class="line">    </span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>val<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    tmp<span class="punctuation">[</span>row_pos<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">,</span>col_pos<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> val<span class="punctuation">[</span>i<span class="punctuation">]</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>可以将其封装为一个函数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">as_matrix <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line"></span><br><span class="line">  tmp <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span>data<span class="operator">=</span><span class="number">0L</span><span class="punctuation">,</span> nrow <span class="operator">=</span> mat<span class="operator">@</span>Dim<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span> ncol <span class="operator">=</span> mat<span class="operator">@</span>Dim<span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  row_pos <span class="operator">&lt;-</span> mat<span class="operator">@</span>i<span class="operator">+</span><span class="number">1</span></span><br><span class="line">  col_pos <span class="operator">&lt;-</span> findInterval<span class="punctuation">(</span>seq<span class="punctuation">(</span>mat<span class="operator">@</span>x<span class="punctuation">)</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span>mat<span class="operator">@</span>p<span class="punctuation">[</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="operator">+</span><span class="number">1</span></span><br><span class="line">  val <span class="operator">&lt;-</span> mat<span class="operator">@</span>x</span><br><span class="line">    </span><br><span class="line">  <span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>val<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">      tmp<span class="punctuation">[</span>row_pos<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">,</span>col_pos<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> val<span class="punctuation">[</span>i<span class="punctuation">]</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">    </span><br><span class="line">  row.names<span class="punctuation">(</span>tmp<span class="punctuation">)</span> <span class="operator">&lt;-</span> mat<span class="operator">@</span>Dimnames<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">  colnames<span class="punctuation">(</span>tmp<span class="punctuation">)</span> <span class="operator">&lt;-</span> mat<span class="operator">@</span>Dimnames<span class="punctuation">[[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>tmp<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>如果速度还需要提高，那么可能就需要Rcpp上场了. 我参考着<a target="_blank" rel="noopener" href="http://adv-r.had.co.nz/Rcpp.html">http://adv-r.had.co.nz/Rcpp.html</a>写了一个简单的代码</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line">Rcpp<span class="operator">::</span>sourceCpp<span class="punctuation">(</span>code<span class="operator">=</span><span class="string">&#x27;</span></span><br><span class="line"><span class="string">#include &lt;Rcpp.h&gt;</span></span><br><span class="line"><span class="string">using namespace Rcpp;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">// [[Rcpp::export]]</span></span><br><span class="line"><span class="string">IntegerMatrix asMatrix(NumericVector rp,</span></span><br><span class="line"><span class="string">                       NumericVector cp,</span></span><br><span class="line"><span class="string">                       NumericVector z,</span></span><br><span class="line"><span class="string">                       int nrows,</span></span><br><span class="line"><span class="string">                       int ncols)&#123;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  int k = z.size() ;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  IntegerMatrix  mat(nrows, ncols);</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  for (int i = 0; i &lt; k; i++)&#123;</span></span><br><span class="line"><span class="string">      mat(rp[i],cp[i]) = z[i];</span></span><br><span class="line"><span class="string">  &#125;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  return mat;</span></span><br><span class="line"><span class="string">&#125;</span></span><br><span class="line"><span class="string">&#x27;</span> <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">as_matrix <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line"></span><br><span class="line">  row_pos <span class="operator">&lt;-</span> mat<span class="operator">@</span>i</span><br><span class="line">  col_pos <span class="operator">&lt;-</span> findInterval<span class="punctuation">(</span>seq<span class="punctuation">(</span>mat<span class="operator">@</span>x<span class="punctuation">)</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span>mat<span class="operator">@</span>p<span class="punctuation">[</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  tmp <span class="operator">&lt;-</span> asMatrix<span class="punctuation">(</span>rp <span class="operator">=</span> row_pos<span class="punctuation">,</span> cp <span class="operator">=</span> col_pos<span class="punctuation">,</span> z <span class="operator">=</span> mat<span class="operator">@</span>x<span class="punctuation">,</span></span><br><span class="line">                 nrows <span class="operator">=</span>  mat<span class="operator">@</span>Dim<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span> ncols <span class="operator">=</span> mat<span class="operator">@</span>Dim<span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">  row.names<span class="punctuation">(</span>tmp<span class="punctuation">)</span> <span class="operator">&lt;-</span> mat<span class="operator">@</span>Dimnames<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">  colnames<span class="punctuation">(</span>tmp<span class="punctuation">)</span> <span class="operator">&lt;-</span> mat<span class="operator">@</span>Dimnames<span class="punctuation">[[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>tmp<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>


<p>如果之前的矩阵有78945836个元素，<code>system.time</code>显示只需要40s。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/17/R%E8%AF%AD%E8%A8%80%E7%9A%84%E7%A8%80%E7%96%8F%E7%9F%A9%E9%98%B5%E5%AD%A6%E4%B9%A0%E8%AE%B0%E5%BD%95-R-Sparse-Matrix-Note/" data-id="clm1z22b0000eehn5d0sjgwyq" data-title="R语言的稀疏矩阵学习记录" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-chromVAR_ 从单细胞表观数据中推断转录因子相关开放区域-chromVAR-inferring-transcription-factor-associated-accessibility-from-single-cell-epigenomic-data" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/16/chromVAR_%20%E4%BB%8E%E5%8D%95%E7%BB%86%E8%83%9E%E8%A1%A8%E8%A7%82%E6%95%B0%E6%8D%AE%E4%B8%AD%E6%8E%A8%E6%96%AD%E8%BD%AC%E5%BD%95%E5%9B%A0%E5%AD%90%E7%9B%B8%E5%85%B3%E5%BC%80%E6%94%BE%E5%8C%BA%E5%9F%9F-chromVAR-inferring-transcription-factor-associated-accessibility-from-single-cell-epigenomic-data/" class="article-date">
  <time class="dt-published" datetime="2019-08-16T13:47:16.858Z" itemprop="datePublished">2019-08-16</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/16/chromVAR_%20%E4%BB%8E%E5%8D%95%E7%BB%86%E8%83%9E%E8%A1%A8%E8%A7%82%E6%95%B0%E6%8D%AE%E4%B8%AD%E6%8E%A8%E6%96%AD%E8%BD%AC%E5%BD%95%E5%9B%A0%E5%AD%90%E7%9B%B8%E5%85%B3%E5%BC%80%E6%94%BE%E5%8C%BA%E5%9F%9F-chromVAR-inferring-transcription-factor-associated-accessibility-from-single-cell-epigenomic-data/">chromVAR:从单细胞表观数据中推断转录因子相关开放区域</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>chromVAR是一个用于分析稀疏染色质开放的R包。chromVAR的输入文件包括，ATAC-seq处理后的fragments文件（过滤重复和低质量数据）, DNAse-seq实验结果，以及基因组注释（例如motif位置）</p>
<p>chromVAR先根据所有细胞或者样本的平均情况来计算<strong>期望开放性</strong>， 然后用它来计算每个注释，每个细胞或样本的偏差，最后对开放进行纠正。</p>
<h2 id="安装加载"><a href="#安装加载" class="headerlink" title="安装加载"></a>安装加载</h2><p>安装R包</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">BiocManager::install(&quot;GreenleafLab/chromVAR&quot;)</span><br><span class="line">BiocManager::install(&quot;motifmatchr&quot;)</span><br><span class="line">BiocManager::install(SummarizedExperiment)</span><br><span class="line">BiocManager::install(BiocParallel)</span><br><span class="line">BiocManager::install(&quot;JASPAR2018&quot;) # JASPAR 2018数据库</span><br></pre></td></tr></table></figure>

<p>chromVAR的运行需要先加载下面这些R包</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">library(chromVAR)</span><br><span class="line">library(motifmatchr)</span><br><span class="line">library(Matrix)</span><br><span class="line">library(SummarizedExperiment)</span><br><span class="line">library(BiocParallel)</span><br><span class="line">set.seed(2019)</span><br></pre></td></tr></table></figure>

<p>chromVAR能够使用多核进行并行运算，调用方法如下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># 全部用户</span><br><span class="line">register(MulticoreParam(8, progressbar = TRUE))</span><br><span class="line"># Windows用户，MulticoreParam不好用就用SnowParam</span><br><span class="line">register(SnowParam(workers = 1, type = &quot;SOCK&quot;))</span><br></pre></td></tr></table></figure>

<p><strong>注意</strong>: 不运行的话，后续代码可能会报错</p>
<h2 id="读取输入"><a href="#读取输入" class="headerlink" title="读取输入"></a>读取输入</h2><p>chromVAR接受的输入是落入开放区域的read数统计表。有许多软件可以做到，chromVAR也提供了相应的方法。</p>
<p>首先要提供一个peak文件，文档建议这个peak文件存放的peak为等宽非重叠，建议宽度在<strong>250-500</strong> bp之间。peak文件可以利用已有的bulk ATAC-seq或DNAse-seq数据来获取。对于来自于多个样本的peak，需要先用<code>filterPeaks</code>函数保证peak之间不重合。</p>
<p>使用<code>getPeaks</code>读取peak</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">peakfile &lt;- system.file(&quot;extdata/test_bed.txt&quot;, package = &quot;chromVAR&quot;)</span><br><span class="line">peaks &lt;- getPeaks(peakfile, sort_peaks = TRUE)</span><br></pre></td></tr></table></figure>

<p>MACS2分析结果里会提供narrowpeak格式文件，chromVAR提供了<code>readNarrowpeaks</code>函数进行读取。</p>
<p>随后用<code>getCounts</code>函数基于BAM文件和加载的peak获取count</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">bamfile &lt;- system.file(&quot;extdata/test_RG.bam&quot;, package = &quot;chromVAR&quot;)</span><br><span class="line">fragment_counts &lt;- getCounts(bamfile, </span><br><span class="line">                             peaks, </span><br><span class="line">                             paired =  TRUE, </span><br><span class="line">                             by_rg = TRUE, </span><br><span class="line">                             format = &quot;bam&quot;, </span><br><span class="line">                             colData = DataFrame(celltype = &quot;GM&quot;))</span><br></pre></td></tr></table></figure>

<p>bamfile可以有多个bam文件路径，bam文件的类型和<code>colData</code>对应。此外<code>by_rg</code>定义是否要根据BAM文件中的RG对输入分组。</p>
<p>实际演示的时候，我们官方提供的示例数据</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">data(example_counts, package = &quot;chromVAR&quot;)</span><br></pre></td></tr></table></figure>

<p>因为这是一个<code>*SummarizedExperiment </code>对象，因此我们就能用<code>assay</code>,<code>colData</code>和<code>rowData</code> 了解这个数据集</p>
<p>主体是一个存放count的dgCMatrix</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">assay(example_counts)[1:5,1:2]</span><br><span class="line"></span><br><span class="line">5 x 2 sparse Matrix of class &quot;dgCMatrix&quot;</span><br><span class="line">     singles-GM12878-140905-1 singles-GM12878-140905-2</span><br><span class="line">[1,]                        .                        1</span><br><span class="line">[2,]                        .                        .</span><br><span class="line">[3,]                        .                        .</span><br><span class="line">[4,]                        .                        .</span><br><span class="line">[5,]                        .                        .</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>行是特征(feature), 这里就是peak</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">head(rowData(example_counts), n=2)</span><br><span class="line"></span><br><span class="line">DataFrame with 2 rows and 3 columns</span><br><span class="line">      score      qval        name</span><br><span class="line">  &lt;integer&gt; &lt;numeric&gt; &lt;character&gt;</span><br><span class="line">1       259  25.99629   GM_peak_6</span><br><span class="line">2        82   8.21494   H1_peak_</span><br></pre></td></tr></table></figure>

<p>列表示样本，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">head(colData(example_counts), n=2)</span><br><span class="line"></span><br><span class="line">DataFrame with 2 rows and 2 columns</span><br><span class="line">                           Cell_Type     depth</span><br><span class="line">                         &lt;character&gt; &lt;integer&gt;</span><br><span class="line">singles-GM12878-140905-1          GM      9220</span><br><span class="line">singles-GM12878-140905-2          GM      9401</span><br></pre></td></tr></table></figure>

<h2 id="前期准备"><a href="#前期准备" class="headerlink" title="前期准备"></a>前期准备</h2><h3 id="分析peak中的GC含量"><a href="#分析peak中的GC含量" class="headerlink" title="分析peak中的GC含量"></a>分析peak中的GC含量</h3><p>GC含量信息用于确定哪些peak可能是背景， <code>addGCBias</code>返回更新后的<code>SummarizedExperiment </code>。其中<code>genome</code>支持BSgenome, FaFile和DNAStringSet对象。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">library(BSgenome.Hsapiens.UCSC.hg19)</span><br><span class="line">example_counts &lt;- addGCBias(example_counts, </span><br><span class="line">                            genome = BSgenome.Hsapiens.UCSC.hg19)</span><br><span class="line">head(rowData(example_counts))</span><br></pre></td></tr></table></figure>

<h3 id="过滤输入-单细胞"><a href="#过滤输入-单细胞" class="headerlink" title="过滤输入(单细胞)"></a>过滤输入(单细胞)</h3><p>如果是处理单细胞数据，建议过滤测序深度不够，或者是信噪比低不够(也就是peak中所占read比例低)的数据，主要靠两个参数,<code>min_in_peaks</code>和<code>min_depth</code>.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">counts_filtered &lt;- filterSamples(example_counts, min_depth = 1500, </span><br><span class="line">                                 min_in_peaks = 0.15, shiny = FALSE)</span><br></pre></td></tr></table></figure>

<p>然后我们用<code>filterSamplesPlot</code>对过滤情况进行可视化</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">filtering_plot &lt;- filterSamplesPlot(example_counts, min_depth = 1500, </span><br><span class="line">                                    min_in_peaks = 0.15, use_plotly = FALSE)</span><br><span class="line">filtering_plot</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/filtering_plot-47e10a7514e04f8dbbbecdb27511e577.jpeg"></p>
<p>确认标准后，就可以过滤</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">counts_filtered &lt;- filterPeaks(counts_filtered, non_overlapping = TRUE)</span><br></pre></td></tr></table></figure>

<h3 id="获取Motifs和分析包含motif的peak"><a href="#获取Motifs和分析包含motif的peak" class="headerlink" title="获取Motifs和分析包含motif的peak"></a>获取Motifs和分析包含motif的peak</h3><p>可以用<code>getJasparMotifs</code>在JASPAR数据库中提取motif信息，但是其中<code>getJasparMotifs</code>只是<code>TFBSTools::getMatrixSet</code>一个简单的封装而已,  默认用的是JASPAR2016, 建议通过下面的代码获取最新版本的数据。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"># BiocManager::install(&quot;JASPAR2018&quot;)</span><br><span class="line">library(JASPAR2018)</span><br><span class="line">species &lt;- &quot;Homo sapiens&quot;</span><br><span class="line">collection &lt;- &quot;CORE&quot;</span><br><span class="line">opts &lt;- list()</span><br><span class="line">opts[&quot;species&quot;] &lt;- species</span><br><span class="line">opts[&quot;collection&quot;] &lt;- collection</span><br><span class="line">out  &lt;- TFBSTools::getMatrixSet(JASPAR2018, opts)</span><br><span class="line"></span><br><span class="line">if (!isTRUE(all.equal(TFBSTools::name(out), names(out)))) </span><br><span class="line">  names(out) &lt;- paste(names(out), TFBSTools::name(out), </span><br><span class="line">                      sep = &quot;_&quot;)</span><br><span class="line">motif &lt;- out</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><code>collection</code>参数中接受: “CORE”, “CNE”, “PHYLOFACTS”, “SPLICE”, “POLII”, “FAM”, “PBM”, “PBM_HOMEO”, “PBM_HLH” 选项。</p>
<p>获取Motif的另外一种方式是用<code>chromVARmotifs</code>, 这个R包的主要功能就是整合了一些motif, 通过<code>devtools::install_github(&quot;GreenleafLab/chromVARmotifs&quot;)</code>进行安装。</p>
<p>之后用<code>motifmatchr</code>中的<code>macthMotifs</code>去分析peak中motif包含情况。默认会返回一个SummarizedExperiment 对象，就是一个稀疏矩阵，来提示是不是匹配了motif</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">library(motifmatchr)</span><br><span class="line">motif_ix &lt;- matchMotifs(motifs, counts_filtered, </span><br><span class="line">                        genome = BSgenome.Hsapiens.UCSC.hg19)</span><br></pre></td></tr></table></figure>

<p>关键参数是<code>p.cutoff</code>用于设置严格度，默认是<code>0.00005</code>其实能够返回比较合理的结果。如果需要一些额外信息，可以通过参数<code>out</code>来调整，例如<code>out=positions</code>表示返回实际的匹配位置</p>
<h2 id="偏离度和变异度分析-Deviations-and-Variability"><a href="#偏离度和变异度分析-Deviations-and-Variability" class="headerlink" title="偏离度和变异度分析(Deviations and Variability )"></a>偏离度和变异度分析(Deviations and Variability )</h2><p>上面都是准备阶段，计算deviations和Variability才是这个软件的重点。</p>
<h3 id="计算偏离度"><a href="#计算偏离度" class="headerlink" title="计算偏离度"></a>计算偏离度</h3><p>函数<code>computeDeviations</code>返回的SummarizedExperiment 对象, 包含两个”assays”, </p>
<ul>
<li>偏差纠正后的开放偏离度:<code>assays(dev)$deviations</code></li>
<li>偏离度的Z-score: <code>assays(deviations)$z</code>: 考虑到了当从基因组上随机抽取一些GC含量类似的片段分析时，有多大概率会得到该结果。</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">dev &lt;- computeDeviations(object = counts_filtered, annotations = motif_ix)</span><br></pre></td></tr></table></figure>

<p><code>compuateDeviations</code>支持背景peak的输入，用于的偏移度得分进行标准化，默认会自动计算不会返回结果。你可以选择用<code>getBackgroundPeaks</code>分析背景，然后传递给<code>computeDeviations</code></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">bg &lt;- getBackgroundPeaks(object = counts_filtered)</span><br><span class="line">dev &lt;- computeDeviations(object = counts_filtered, annotations = motif_ix,</span><br><span class="line">                         background_peaks = bg)</span><br></pre></td></tr></table></figure>

<h2 id="变异度"><a href="#变异度" class="headerlink" title="变异度"></a>变异度</h2><p><code>computeVariability</code>会返回一个数据框<code>data.frame</code>，里面有变异度, 该变异度的bootstrap置信区间和衡量拒绝原假设的概率(即 &gt; 1)</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">variability &lt;- computeVariability(dev)</span><br><span class="line">plotVariability(variability, use_plotly = FALSE) </span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/Variability-2c06b2c8bfa447159f32cd4dd0b478c3.jpeg"></p>
<h3 id="偏离度可视化"><a href="#偏离度可视化" class="headerlink" title="偏离度可视化"></a>偏离度可视化</h3><p>可以利用tSNE可视化偏离度</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">tsne_results &lt;- deviationsTsne(dev, threshold = 1.5, perplexity = 10)</span><br><span class="line">tsne_plots &lt;- plotDeviationsTsne(dev, tsne_results, </span><br><span class="line">                                 annotation_name = &quot;TEAD3&quot;, </span><br><span class="line">                                 sample_column = &quot;Cell_Type&quot;, </span><br><span class="line">                                 shiny = FALSE)</span><br><span class="line">cowplot::plot_grid(tsne_plots[[1]],tsne_plots[[1]] )</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/Deviations_tSNE-7fcaa606061241ffb3686d8f8cc8e024.jpeg"></p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li>数据库<a target="_blank" rel="noopener" href="http://jaspar.genereg.net/">JASPAR</a>是一个开放性数据库，用于存放人工审查的非冗余转录因子(TF)结合谱，数据格式为PFM(position frequency matrices ), TFFM( TF flexible models)</li>
<li><a target="_blank" rel="noopener" href="https://greenleaflab.github.io/chromVAR/articles/Introduction.html#get-motifs-and-what-peaks-contain-motifs">chromVAR</a>官方文档</li>
<li><a target="_blank" rel="noopener" href="https://www.nature.com/articles/nmeth.4401">chromVAR: inferring transcription-factor-associated accessibility from single-cell epigenomic data</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/16/chromVAR_%20%E4%BB%8E%E5%8D%95%E7%BB%86%E8%83%9E%E8%A1%A8%E8%A7%82%E6%95%B0%E6%8D%AE%E4%B8%AD%E6%8E%A8%E6%96%AD%E8%BD%AC%E5%BD%95%E5%9B%A0%E5%AD%90%E7%9B%B8%E5%85%B3%E5%BC%80%E6%94%BE%E5%8C%BA%E5%9F%9F-chromVAR-inferring-transcription-factor-associated-accessibility-from-single-cell-epigenomic-data/" data-id="clm1z22bd001behn59kh5facy" data-title="chromVAR:从单细胞表观数据中推断转录因子相关开放区域" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-生物信息学格式和转换工具整理-生物信息学格式和转换整理" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/15/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E5%B7%A5%E5%85%B7%E6%95%B4%E7%90%86-%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E6%95%B4%E7%90%86/" class="article-date">
  <time class="dt-published" datetime="2019-08-15T14:04:57.682Z" itemprop="datePublished">2019-08-15</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/15/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E5%B7%A5%E5%85%B7%E6%95%B4%E7%90%86-%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E6%95%B4%E7%90%86/">生物信息学格式和转换工具整理</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>生信会用到很多格式，不同软件会要求不同的输入格式，所以得要自己整理下。</p>
<h2 id="FASTA和FASTQ"><a href="#FASTA和FASTQ" class="headerlink" title="FASTA和FASTQ"></a>FASTA和FASTQ</h2><h2 id="SAM和BAM"><a href="#SAM和BAM" class="headerlink" title="SAM和BAM"></a>SAM和BAM</h2><h2 id="GTF和GFF"><a href="#GTF和GFF" class="headerlink" title="GTF和GFF"></a>GTF和GFF</h2><h2 id="BED"><a href="#BED" class="headerlink" title="BED"></a>BED</h2><h2 id="BedGraph"><a href="#BedGraph" class="headerlink" title="BedGraph"></a>BedGraph</h2><h2 id="Wig和BigWig"><a href="#Wig和BigWig" class="headerlink" title="Wig和BigWig"></a>Wig和BigWig</h2><p>WIG（wiggle）设计的目的是在基因组浏览器上展示连续性数据。Wiggle数据要求元素等长。对于稀疏或者元素之间不等长，建议用BedGraph格式。而BigWig则是二进制版，体积小，适合传输。</p>
<p>举个例子</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">variableStep chrom=chr2</span><br><span class="line">300701  12.5</span><br><span class="line">300702  12.5</span><br><span class="line">300703  12.5</span><br><span class="line">300704  12.5</span><br><span class="line">300705  12.5</span><br></pre></td></tr></table></figure>

<p>和</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">variableStep chrom=chr2 span=5</span><br><span class="line">300701  12.5</span><br></pre></td></tr></table></figure>

<p>在基因组浏览器上展示时都是2号染色体的300701-300705的值为12.5</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/15/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E5%B7%A5%E5%85%B7%E6%95%B4%E7%90%86-%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E6%95%B4%E7%90%86/" data-id="clm1z22ik00ihehn585ht46bs" data-title="生物信息学格式和转换工具整理" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-实现Seurat 2 和 Seurat3 在同一个环境下共存-Seuarat2-Seurat3-together" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/14/%E5%AE%9E%E7%8E%B0Seurat%202%20%E5%92%8C%20Seurat3%20%E5%9C%A8%E5%90%8C%E4%B8%80%E4%B8%AA%E7%8E%AF%E5%A2%83%E4%B8%8B%E5%85%B1%E5%AD%98-Seuarat2-Seurat3-together/" class="article-date">
  <time class="dt-published" datetime="2019-08-14T10:31:49.412Z" itemprop="datePublished">2019-08-14</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/14/%E5%AE%9E%E7%8E%B0Seurat%202%20%E5%92%8C%20Seurat3%20%E5%9C%A8%E5%90%8C%E4%B8%80%E4%B8%AA%E7%8E%AF%E5%A2%83%E4%B8%8B%E5%85%B1%E5%AD%98-Seuarat2-Seurat3-together/">实现Seurat 2 和 Seurat3 在同一个环境下共存</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>软件升级虽然是一件值得高兴的是，但是代码变化太大却不是一件好消息。比如说Seurat，这个单细胞分析最常用的R包，它的2.x版本和3.x版本的变化就是翻天覆地。</p>
<p>为了能够重现别人的代码，你可能需要重装2.3.4版本的Seurat，官方提供了安装脚本</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">source<span class="punctuation">(</span><span class="string">&quot;https://gist.githubusercontent.com/satijalab/beb9bb50dedc75ee023bd5d9be5fe684/raw/e103577735a2fba9da2ccca14ce1ac33e46c1bc4/old_seurat.R&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>但是在window下安装会出一个报错, 提示无法在Windows下安装MacOS的版本</p>
<p>这个时候需要手动安装Seurat，注意，这里要求有Rtools才能进行编译</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">install.packages<span class="punctuation">(</span><span class="string">&#x27;Seurat&#x27;</span><span class="punctuation">,</span> repos <span class="operator">=</span> <span class="string">&#x27;https://satijalab.org/ran&#x27;</span><span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;source&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>可以设置<code>install.packages</code>里的参数lib.loc，让Seurat安装到其他的文件加，就不会替换原来的Seurat，同时用<code>library</code>加载的时候，也需要设置<code>lib.loc</code>.</p>
<p>上面是简单可行易操作的方法，唯一的问题是你不能同时加载两个Seurat版本，当然你也不会想这样子做，所以这是一个伪需求。</p>
<p>下面就是瞎折腾环节， 我要将Seurat的2.3.4版本单独搞出一个R包，Seurat2，这样子就可以同时加载这两个R包。</p>
<p>下载Seurat 2.3.4</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wget https://satijalab.org/ran/src/contrib/Seurat_2.3.4.tar.gz</span><br></pre></td></tr></table></figure>

<p>解压缩它</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">tar xf Seurat_2.3.4.tar.gz</span><br><span class="line"><span class="built_in">cd</span> Seurat</span><br></pre></td></tr></table></figure>

<p>删除MD5文件，因为它会做文件检验</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">rm</span> MD5</span><br></pre></td></tr></table></figure>

<p>修改里面所有的Seurat替换成Seurat2, seurat替换成seurat2</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">find . -<span class="built_in">type</span> f -print0 | xargs -0 sed -i <span class="string">&quot;s/Seurat/Seurat2/g&quot;</span></span><br><span class="line">find . -<span class="built_in">type</span> f -print0 | xargs -0 sed -i <span class="string">&quot;s/seurat/seurat2/g&quot;</span></span><br></pre></td></tr></table></figure>

<p>这种无差别的替换会有一个问题，会把一些这类<code>http://www.satijalab.org/seurat</code>非代码信息中的seurat替换成seurat2，不过这并不影响实际函数的使用。</p>
<p>将<code>R/seurat.R</code>重名为<code>R/seurat2.R</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mv</span> R/seurat.R R/seurat2.R</span><br></pre></td></tr></table></figure>

<p>之后打包修改后的文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">tar -czf Seurat2.tar.gz Seurat</span><br></pre></td></tr></table></figure>

<p>于是我们就可以安装我们自己修改后的R包了</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">install.packages<span class="punctuation">(</span><span class="string">&quot;Seurat2.tar.gz&quot;</span><span class="punctuation">,</span> repos<span class="operator">=</span><span class="literal">NULL</span><span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&quot;source&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><strong>注意</strong>: 目前未修改测试数据集的对象，所以不能用Seurat2来运行pbmc_small的例子</p>
<p>由于我已经修改好了，所以你们也不需要自己再去修改了。我将其上传到GitHub，所以可以用devtools进行安装。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">devtools<span class="operator">::</span>install_github<span class="punctuation">(</span><span class="string">&quot;xuzhougeng/Seurat2&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>由于代码只是简单修改，存在bug，其实不建议尝试使用。而且安装完Seurat2之后，我<strong>定性</strong>的评估了下Seurat2和Seurat3，发现两者在聚类分析上并没有太大区别, 所以更推荐大家使用最新的<code>Seurat</code>.</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/14/%E5%AE%9E%E7%8E%B0Seurat%202%20%E5%92%8C%20Seurat3%20%E5%9C%A8%E5%90%8C%E4%B8%80%E4%B8%AA%E7%8E%AF%E5%A2%83%E4%B8%8B%E5%85%B1%E5%AD%98-Seuarat2-Seurat3-together/" data-id="clm1z22hs00gjehn5er2n1mxe" data-title="实现Seurat 2 和 Seurat3 在同一个环境下共存" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/22/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/21/">21</a><a class="page-number" href="/page/22/">22</a><span class="page-number current">23</span><a class="page-number" href="/page/24/">24</a><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/24/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" rel="tag">个人博客</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E4%B8%AA%E4%BA%BA%E5%8D%9A%E5%AE%A2/" style="font-size: 10px;">个人博客</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/09/">九月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/08/">八月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/09/02/%E5%A6%82%E4%BD%95%E8%AE%A9%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8D%E5%86%8D%E9%BE%9F%E9%80%9F%E4%B8%8B%E8%BD%BDgithub%E7%9A%84%E6%95%B0%E6%8D%AE/">如何让服务器不再龟速下载github的数据</a>
          </li>
        
          <li>
            <a href="/2023/08/27/%E7%BC%96%E8%AF%91%E5%AE%89%E8%A3%85%E5%AE%B9%E5%99%A8singularity/">编译安装容器singularity</a>
          </li>
        
          <li>
            <a href="/2023/08/25/%E5%85%B3%E4%BA%8E%E7%94%B3%E8%AF%B7OpenAI%E7%9A%84API%E7%9A%84%E5%87%A0%E7%82%B9%E8%A1%A5%E5%85%85/">关于申请OpenAI的API的几点补充</a>
          </li>
        
          <li>
            <a href="/2023/08/18/%E4%BB%8Ehalo%E8%BF%81%E7%A7%BB%E5%88%B0hexo/">从halo迁移到hexo</a>
          </li>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>