<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>洲更的第二大脑</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="洲更的第二大脑">
<meta property="og:url" content="http://xuzhougeng.top/page/23/index.html">
<meta property="og:site_name" content="洲更的第二大脑">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="xuzhougeng">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="洲更的第二大脑" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">洲更的第二大脑</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS 订阅"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="搜索"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜索"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://xuzhougeng.top"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-Cicero_ 单细胞共开放分析- Cicero-single-cell-ATAC-seq-analyzing-tools" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/18/Cicero_%20%E5%8D%95%E7%BB%86%E8%83%9E%E5%85%B1%E5%BC%80%E6%94%BE%E5%88%86%E6%9E%90-%20Cicero-single-cell-ATAC-seq-analyzing-tools/" class="article-date">
  <time class="dt-published" datetime="2019-08-18T23:20:21.487Z" itemprop="datePublished">2019-08-18</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/18/Cicero_%20%E5%8D%95%E7%BB%86%E8%83%9E%E5%85%B1%E5%BC%80%E6%94%BE%E5%88%86%E6%9E%90-%20Cicero-single-cell-ATAC-seq-analyzing-tools/">Cicero:单细胞共开放分析</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>Cicero是一个单细胞染色质可及性实验可视化R包。Cicero的<strong>主要功能</strong>就是使用单细胞染色质可及性数据通过分析共开放去预测基因组上<strong>顺式调节作用</strong>(cis-regulatory interactions)，例如增强子和启动子。Cicero能够利用染色质开放性进行单细胞聚类，排序和差异可及性分析。关于Cicero的算法原理，参考<a target="_blank" rel="noopener" href="https://cole-trapnell-lab.github.io/cicero-release/papers/">他们发表的文章</a>.</p>
<h2 id="简介"><a href="#简介" class="headerlink" title="简介"></a>简介</h2><p>Cicero的主要目标是使用单细胞染色质可及性数据去预测基因组上那些在细胞核中存在物理邻近的区域。这能够用于鉴定潜在的<strong>增强子-启动子对</strong>，对基因组顺式作用有一个总体了解。</p>
<p>由于单细胞数据的稀疏性，细胞必须根据相似度进行聚合，才能够对数据中的多种技术因素进行可靠纠正。</p>
<p>最终，Cicero给出了”Cicero 共开放”得分，分数在-1和1之间，用于评估用于给定距离中每个可及peak对之间的共开放性，分数越高，越可能是共开放。</p>
<p>此外，Monocle3的框架让Cicero能对单细胞ATAC-seq实验完成Monocle3上的聚类，拟时间等分析。</p>
<p>Cicero主要提供了两种核心分析:</p>
<ul>
<li><strong>构建和分析顺式调控网络</strong>.</li>
<li><strong>常规单细胞染色质可及性分析</strong>:</li>
</ul>
<h2 id="Cicero安装和加载"><a href="#Cicero安装和加载" class="headerlink" title="Cicero安装和加载"></a>Cicero安装和加载</h2><p>Cicero运行在R语言分析环境中。尽管能够从Bioconductor安装，但是推荐从GitHub上安装。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">install.packages(&quot;BiocManager&quot;)</span><br><span class="line">BiocManager::install(c(&quot;Gviz&quot;, &quot;GenomicRanges&quot;, &quot;rtracklayer&quot;))</span><br><span class="line">install.packages(&quot;devtools&quot;)</span><br><span class="line">devtools::install_github(&quot;cole-trapnell-lab/cicero-release&quot;, ref = &quot;monocle3&quot;)</span><br></pre></td></tr></table></figure>

<p>加载Cicero</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">library(cicero)</span><br></pre></td></tr></table></figure>

<h2 id="数据加载"><a href="#数据加载" class="headerlink" title="数据加载"></a>数据加载</h2><p>Cicero以<code>cell_data_set(CDS)</code>对象存放数据，该对象继承自Bioconductor的<code>SingleCellExperiment</code>对象。我们可以用下面三个函数来操纵数据</p>
<ul>
<li>fData: 获取feature的元信息, 这里的featureI指的是peak</li>
<li>pData: 获取cell&#x2F;sample的元信息</li>
<li>assay: 获取cell-by-peak的count矩阵</li>
</ul>
<p>除了CDS外，还需要基因坐标信息和基因注释信息。</p>
<p>关于坐标信息，以mouse.mm9.genome为例，它是一个数据框，只有两列，一列是染色体编号，另一列对应的长度</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">data(&quot;mouse.mm9.genome&quot;)</span><br></pre></td></tr></table></figure>

<p>基因组注释信息可以用<code>rtracklayer::readGFF</code>加载</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">temp &lt;- tempfile()</span><br><span class="line">download.file(&quot;ftp://ftp.ensembl.org/pub/release-65/gtf/mus_musculus/Mus_musculus.NCBIM37.65.gtf.gz&quot;, temp)</span><br><span class="line">gene_anno &lt;- rtracklayer::readGFF(temp)</span><br><span class="line">unlink(temp)</span><br></pre></td></tr></table></figure>

<p>Bioconductor也有现成的TxDb，可以从中提取注释信息。</p>
<h3 id="简单稀疏矩阵格式"><a href="#简单稀疏矩阵格式" class="headerlink" title="简单稀疏矩阵格式"></a>简单稀疏矩阵格式</h3><p>以一个测试数据集介绍如何加载简单稀疏矩阵格式</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">temp &lt;- textConnection(readLines(gzcon(url(&quot;http://staff.washington.edu/hpliner/data/kidney_data.txt.gz&quot;))))</span><br><span class="line">cicero_data &lt;- read.table(temp)</span><br></pre></td></tr></table></figure>

<p>这里的简单稀疏矩阵格式指的是数据以三列进行存放，第一列是peak位置信息，第二列是样本信息，第三列则是样本中有多少fragment和该peak重叠。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chr1_4756552_4757256    GAATTCGTACCAGGCGCATTGGTAGTCGGGCTCTGA    1</span><br></pre></td></tr></table></figure>

<p>对于这种格式，Cicero提供了一个函数<code>make_atac_cds</code>, 用于构建一个有效的<code>cell_data_set</code>对象，用于下游分析，输入既可以是一个数据框，也可以是一个文件路径。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">input_cds &lt;- make_atac_cds(cicero_data, binarize = TRUE)</span><br></pre></td></tr></table></figure>

<h3 id="10X-scATAC-seq-数据"><a href="#10X-scATAC-seq-数据" class="headerlink" title="10X scATAC-seq 数据"></a>10X scATAC-seq 数据</h3><p>如果数据已经经过10X scATAC-seq处理，那么结果中的<strong>filtered_peak_bc_matrix</strong>里就有我们需要的信息</p>
<p>加载cell-by-peak count矩阵，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># read in matrix data using the Matrix package</span><br><span class="line">indata &lt;- Matrix::readMM(&quot;filtered_peak_bc_matrix/matrix.mtx&quot;) </span><br><span class="line"># binarize the matrix</span><br><span class="line">indata@x[indata@x &gt; 0] &lt;- 1</span><br></pre></td></tr></table></figure>

<p>加载cell元数据</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># format cell info</span><br><span class="line">cellinfo &lt;- read.table(&quot;filtered_peak_bc_matrix/barcodes.tsv&quot;)</span><br><span class="line">row.names(cellinfo) &lt;- cellinfo$V1</span><br><span class="line">names(cellinfo) &lt;- &quot;cells&quot;</span><br></pre></td></tr></table></figure>

<p>加载peak元数据</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># format peak info</span><br><span class="line">peakinfo &lt;- read.table(&quot;filtered_peak_bc_matrix/peaks.bed&quot;)</span><br><span class="line">names(peakinfo) &lt;- c(&quot;chr&quot;, &quot;bp1&quot;, &quot;bp2&quot;)</span><br><span class="line">peakinfo$site_name &lt;- paste(peakinfo$chr, peakinfo$bp1, peakinfo$bp2, sep=&quot;_&quot;)</span><br><span class="line">row.names(peakinfo) &lt;- peakinfo$site_name</span><br></pre></td></tr></table></figure>

<p>为cell-by-peak的count矩阵增加行名(peak元数据)和列名(cell元数据)</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">row.names(indata) &lt;- row.names(peakinfo)</span><br><span class="line">colnames(indata) &lt;- row.names(cellinfo)</span><br></pre></td></tr></table></figure>

<p>然后用<code>new_cell_data_set</code>根据peak元数据，cell元数据构建出一个<code>cell_data_set</code>对象</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># make CDS</span><br><span class="line">input_cds &lt;-  suppressWarnings(new_cell_data_set(indata,</span><br><span class="line">    cell_metadata = cellinfo,</span><br><span class="line">    gene_metadata = peakinfo))</span><br></pre></td></tr></table></figure>

<p>之后用<code>detect_genes</code>统计，对于每个基因在多少细胞中的表达量超过了给定阈值。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">input_cds &lt;- monocle3::detect_genes(input_cds)</span><br></pre></td></tr></table></figure>

<p>过滤没有细胞开放的peak</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">input_cds &lt;- input_cds[Matrix::rowSums(exprs(input_cds)) != 0,] </span><br></pre></td></tr></table></figure>

<h2 id="构建顺式作用网络"><a href="#构建顺式作用网络" class="headerlink" title="构建顺式作用网络"></a>构建顺式作用网络</h2><h3 id="运行Cicero"><a href="#运行Cicero" class="headerlink" title="运行Cicero"></a>运行Cicero</h3><h3 id="构建Cicero-CDS对象"><a href="#构建Cicero-CDS对象" class="headerlink" title="构建Cicero CDS对象"></a>构建Cicero CDS对象</h3><p>因为单细胞染色质开放数据特别稀疏，为了能够比较准确的估计共开放得分，需要将一些比较相近的细胞进行聚合，得到一个相对比较致密的count数据。Cicero使用KNN算法构建细胞的相互重叠集。而细胞之间距离关系则是根据降维后坐标信息。降维方法有很多种，这里以UMAP为例</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">input_cds &lt;- detect_genes(input_cds)</span><br><span class="line">input_cds &lt;- estimate_size_factors(input_cds)</span><br><span class="line"># PCA或LSI线性降维</span><br><span class="line">input_cds &lt;- preprocess_cds(input_cds, method = &quot;LSI&quot;)</span><br><span class="line"># UMAP非线性降维</span><br><span class="line">input_cds &lt;- reduce_dimension(input_cds, reduction_method = &#x27;UMAP&#x27;, </span><br><span class="line">                              preprocess_method = &quot;LSI&quot;)</span><br></pre></td></tr></table></figure>

<p>此时用<code>reducedDimNames(input_cds)</code>就会发现它多了LSI和UMAP两个信息，我们可以用<code>reducedDims</code>来提取UMAP坐标信息。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">umap_coords &lt;- reducedDims(input_cds)$UMAP</span><br></pre></td></tr></table></figure>

<p><code>make_cicero_cds</code>函数就需要其中UMAP坐标信息。</p>
<p><strong>注1</strong>: 假如你已经有了UMAP的坐标信息，那么就不需要运行<code>preprocess_cds</code>和<code>reduce_dimension</code>， 直接提供UMAP坐标信息给<code>make_cicero_cds</code>就可以了。</p>
<p><strong>注2</strong>: umap_coords的行名需要和<code>pData</code>得到表格中的行名一样, 即<code>all(row.names(umap_coords) == row.names(pData(input_cds)))</code>结果为TRUE。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cicero_cds &lt;- make_cicero_cds(input_cds, reduced_coordinates = umap_coords)</span><br></pre></td></tr></table></figure>

<p>此时的cicero_cds是没有UMAP降维信息。</p>
<h3 id="运行Cicero-1"><a href="#运行Cicero-1" class="headerlink" title="运行Cicero"></a>运行Cicero</h3><p>Cicero包的主要功能就是预测基因组上位点之间的共开放. 有两种方式可以获取该信息</p>
<ul>
<li><code>run_cicero</code>: 一步到位。默认参数适用于人类和小鼠，但是其他物种，参考<a target="_blank" rel="noopener" href="https://cole-trapnell-lab.github.io/cicero-release/docs_m3//#important-considerations-for-non-human-data">此处</a></li>
<li>单独运行每个函数，这样子获取中间的信息。</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">#测试部分数据</span><br><span class="line">data(&quot;mouse.mm9.genome&quot;)</span><br><span class="line">sample_genome &lt;- subset(mouse.mm9.genome, V1 == &quot;chr2&quot;)</span><br><span class="line">sample_genome$V2[1] &lt;- 10000000</span><br><span class="line">conns &lt;- run_cicero(cicero_cds, sample_genome, sample_num = 2) </span><br><span class="line">nrow(conns) # 212302</span><br><span class="line">## 全部数据, 时间真久</span><br><span class="line">conns &lt;- run_cicero(cicero_cds, mouse.mm9.genome, sample_num = 2)</span><br><span class="line">nrows(conns) #59741738</span><br></pre></td></tr></table></figure>

<p>其中conns存放的是peak之间共开放得分。</p>
<h3 id="Cicero-Connection可视化"><a href="#Cicero-Connection可视化" class="headerlink" title="Cicero Connection可视化"></a>Cicero Connection可视化</h3><p>有了peak之间共开放的可能性得分后，就可以画一些很酷的图了。用到的函数是<code>plot_connections</code>, 这个函数有很多参数，所以最终呈现结果取决于你的<a target="_blank" rel="noopener" href="https://cole-trapnell-lab.github.io/cicero-release/docs/#advanced-visualizaton">参数调整</a>。</p>
<p>获取基因组注释信息</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">temp &lt;- tempfile()</span><br><span class="line">download.file(&quot;ftp://ftp.ensembl.org/pub/release-65/gtf/mus_musculus/Mus_musculus.NCBIM37.65.gtf.gz&quot;, temp)</span><br><span class="line">gene_anno &lt;- rtracklayer::readGFF(temp)</span><br><span class="line">unlink(temp)</span><br><span class="line"></span><br><span class="line">gene_anno$chromosome &lt;- paste0(&quot;chr&quot;, gene_anno$seqid)</span><br><span class="line">gene_anno$gene &lt;- gene_anno$gene_id</span><br><span class="line">gene_anno$transcript &lt;- gene_anno$transcript_id</span><br><span class="line">gene_anno$symbol &lt;- gene_anno$gene_name</span><br></pre></td></tr></table></figure>

<p>除了从GFF文件中提取外，还可以利用<code>txdb</code>提取转录本位置信息</p>
<p>举例子，下面的函数就是绘制染色体上<code>&quot;chr2:9773451-9848598</code>的共开放情况，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">plot_connections(conns, &quot;chr2&quot;, 9773451, 9848598,</span><br><span class="line">                 gene_model = gene_anno, </span><br><span class="line">                 coaccess_cutoff = .25, </span><br><span class="line">                 connection_width = .5, </span><br><span class="line">                 collapseTranscripts = &quot;longest&quot; )</span><br></pre></td></tr></table></figure>

<p>假如之前用的全基因组来算connections，那么这一步也会很久才能出图，所以建议先自己提取部分数据，这样子作图的速度就会快很多(因为源代码没有优化好矩阵过滤)</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">conns_sub &lt;- conns[conns_matrix[,1] == &quot;chr2&quot; &amp; </span><br><span class="line">                     as.numeric(conns_matrix[,2]) &gt; 8000000 &amp;</span><br><span class="line">                     as.numeric(conns_matrix[,2]) &lt; 10000000 </span><br><span class="line">                     , ]</span><br><span class="line"></span><br><span class="line">plot_connections(conns_sub, &quot;chr2&quot;, 9773451, 9848598,</span><br><span class="line">                 gene_model = gene_anno, </span><br><span class="line">                 coaccess_cutoff = .25, </span><br><span class="line">                 connection_width = .5, </span><br><span class="line">                 collapseTranscripts = &quot;longest&quot; )</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/connection_plot-a71e2def92334614add0e4ce2280bbd8.png" alt="connections"></p>
<h2 id="计算基因活跃度得分"><a href="#计算基因活跃度得分" class="headerlink" title="计算基因活跃度得分"></a>计算基因活跃度得分</h2><p>在启动子区的开放度并不是基因表达很好的预测子。使用Cicero links，我们就能得到启动子和它相关的远距位点的总体开放情况。这种区域间开放度联合得分和基因表达的相关度更高。作者称之为<strong>基因活跃度得分</strong>（Cicero gene activity score)。 可以用两个函数进行计算</p>
<p><code>build_gene_activity_matrix</code>函数接受CDS和Cicero connection list作为输入，返回一个标准化的基因得分。<strong>注意</strong>， CDS必须在fData 表中有一列”gene”, 如果该peak在启动子区就用对应的基因名进行注释, 否则为NA. 我们可以用<code>annotate_cds_by_site</code>进行注释。</p>
<p><code>build_gene_activitity_matrix</code>是未标准化的数据。需要用<code>normalize_gene_activities</code>进行标准化。如果你想比较不同数据集中部分数据的基因活跃度得分, 那么所有的gene activity 部分数据都需要一起标准化。如果只是想标准化其中一个数据集，那么传递一个结果即可。<code>normalize_gene_activities</code>也需要一个每个细胞所有开放位点的命名向量。这存放在CDS对象中的<code>pData</code>，在<code>num_genes_expressed</code>列。</p>
<p>可能还是不是明白，看代码吧。</p>
<p>第一步，是用<code>annotate_cds_by_site</code>对CDS对象进行注释在启动子区域的peak。</p>
<p>我们先从GFF文件中提取启动子的位置，也就是每个转录本的第一个外显子</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"># 分别提取正链和负链的注释, 仅保留第一个外显子</span><br><span class="line">pos &lt;- subset(gene_anno, strand == &quot;+&quot;)</span><br><span class="line">pos &lt;- pos[order(pos$start),] </span><br><span class="line"></span><br><span class="line"># remove all but the first exons per transcript</span><br><span class="line">pos &lt;- pos[!duplicated(pos$transcript),] </span><br><span class="line"># make a 1 base pair marker of the TSS</span><br><span class="line">pos$end &lt;- pos$start + 1 </span><br><span class="line"></span><br><span class="line">neg &lt;- subset(gene_anno, strand == &quot;-&quot;)</span><br><span class="line">neg &lt;- neg[order(neg$start, decreasing = TRUE),] </span><br><span class="line"># remove all but the first exons per transcript</span><br><span class="line">neg &lt;- neg[!duplicated(neg$transcript),] </span><br><span class="line">neg$start &lt;- neg$end - 1</span><br><span class="line"></span><br><span class="line"># 合并正链和负链</span><br><span class="line">gene_annotation_sub &lt;- rbind(pos, neg)</span><br><span class="line"></span><br><span class="line"># 只需要4列信息</span><br><span class="line">gene_annotation_sub &lt;- gene_annotation_sub[,c(&quot;chromosome&quot;, &quot;start&quot;, &quot;end&quot;, &quot;symbol&quot;)]</span><br><span class="line"></span><br><span class="line"># 第四列命名为gene</span><br><span class="line">names(gene_annotation_sub)[4] &lt;- &quot;gene&quot;</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>用<code>annotate_cds_by_site</code>对cds按照gene_annotation_sub注释。会多处两列，overlap和gene。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">input_cds &lt;- annotate_cds_by_site(input_cds, gene_annotation_sub)</span><br><span class="line">tail(fData(input_cds))</span><br></pre></td></tr></table></figure>

<p>第二步: 计算未标准化的基因活跃度得分</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># generate unnormalized gene activity matrix</span><br><span class="line">unnorm_ga &lt;- build_gene_activity_matrix(input_cds, conns)</span><br><span class="line"></span><br><span class="line"># 过滤全为0的行和列</span><br><span class="line">unnorm_ga &lt;- unnorm_ga[!Matrix::rowSums(unnorm_ga) == 0, </span><br><span class="line">                       !Matrix::colSums(unnorm_ga) == 0]</span><br></pre></td></tr></table></figure>

<p>第三步: 标准化基因活跃度得分</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># 你需要一个命名列表</span><br><span class="line">num_genes &lt;- pData(input_cds)$num_genes_expressed</span><br><span class="line">names(num_genes) &lt;- row.names(pData(input_cds))</span><br><span class="line"></span><br><span class="line"># 标准化</span><br><span class="line">cicero_gene_activities &lt;- normalize_gene_activities(unnorm_ga, num_genes)</span><br></pre></td></tr></table></figure>

<p>如果你有多个数据集需要进行标注化，那你就需要传递一个列表给<code>normalize_gene_activities</code>, 并且保证<code>num_genes</code>里包括所有数据集中的所有样本。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># if you had two datasets to normalize, you would pass both:</span><br><span class="line"># num_genes should then include all cells from both sets</span><br><span class="line">unnorm_ga2 &lt;- unnorm_ga</span><br><span class="line">cicero_gene_activities &lt;- normalize_gene_activities(list(unnorm_ga, unnorm_ga2), </span><br><span class="line">                                                    num_genes)</span><br></pre></td></tr></table></figure>

<p>标注化后的基因活跃度得分在0到1之间,</p>
<p>第四步: 在UMAP上可视化基因得分。由于值比较小，要先扩大1000000倍，即log2(GA * 1000000 + 1)，这样子在可视化的时候比较好看。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">plot_umap_ga &lt;- function(cds,</span><br><span class="line">                         cicero_gene_activities,</span><br><span class="line">                         gene,</span><br><span class="line">                         dotSize = 1,</span><br><span class="line">                         log = TRUE)&#123;</span><br><span class="line">  #umap_df &lt;- as.data.frame(colData(cds))</span><br><span class="line">  umap_df &lt;- as.data.frame(reducedDims(input_cds)$UMAP)</span><br><span class="line">  colnames(umap_df) &lt;- c(&quot;UMAP1&quot;,&quot;UMAP2&quot;)</span><br><span class="line">  gene_pos &lt;- which(row.names(cicero_gene_activities) %in% gene)</span><br><span class="line"></span><br><span class="line">  umap_df$score &lt;- log2(cicero_gene_activities[gene_pos,] * 1000000 + 1)</span><br><span class="line">  if (log)&#123;</span><br><span class="line">    umap_df$score &lt;- log(umap_df$score + 1)</span><br><span class="line">  &#125;</span><br><span class="line"></span><br><span class="line">  p &lt;- ggplot(umap_df, aes(x=UMAP1, y=UMAP2)) +</span><br><span class="line">    geom_point(aes(color=score), size = dotSize) + scale_color_viridis(option=&quot;magma&quot;) +</span><br><span class="line">    theme_bw()</span><br><span class="line">  return(p)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h2 id="拟时序分析"><a href="#拟时序分析" class="headerlink" title="拟时序分析"></a>拟时序分析</h2><p>这部分的分析其实比较简单，因为仅仅是将feature从mRNA变成了peak而已。步骤如下</p>
<ol>
<li>数据预处理</li>
<li>数据降维</li>
<li>细胞聚类</li>
<li>轨迹图学习</li>
<li>在拟时间上对细胞进行排序</li>
</ol>
<p>代码如下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"># 读取数据</span><br><span class="line">temp &lt;- textConnection(readLines(gzcon(url(&quot;http://staff.washington.edu/hpliner/data/kidney_data.txt.gz&quot;))))set.seed(2017)</span><br><span class="line">input_cds &lt;- estimate_size_factors(input_cds)</span><br><span class="line">#1 数据预处理</span><br><span class="line">input_cds &lt;- preprocess_cds(input_cds, method = &quot;LSI&quot;)</span><br><span class="line">#2 降维</span><br><span class="line">input_cds &lt;- reduce_dimension(input_cds, reduction_method = &#x27;UMAP&#x27;, </span><br><span class="line">preprocess_method = &quot;LSI&quot;)</span><br><span class="line">#3 细胞聚类</span><br><span class="line">input_cds &lt;- cluster_cells(input_cds)</span><br><span class="line">#4 轨迹图学习</span><br><span class="line">input_cds &lt;- learn_graph(input_cds)</span><br><span class="line">#5 对细胞进行排序</span><br><span class="line">input_cds &lt;- order_cells(input_cds, </span><br><span class="line">                         root_cells = &quot;GAGATTCCAGTTGAATCACTCCATCGAGATAGAGGC&quot;)</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>最后可视化</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plot_cells(input_cds, color_cells_by = &quot;pseudotime&quot;)</span><br></pre></td></tr></table></figure>

<h3 id="差异开放分析"><a href="#差异开放分析" class="headerlink" title="差异开放分析"></a>差异开放分析</h3><p>当你的细胞在拟时间中排列之后，你可能会想知道基因组哪些区域会随着时间发生变化。</p>
<p>如果你已经知道哪些区域会随着时间发生变化，你可以用<code>plot_accessibility_in_pseudotime</code>进行可视化</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">input_cds_lin &lt;- input_cds[,is.finite(pseudotime(input_cds))]</span><br><span class="line">plot_accessibility_in_pseudotime(input_cds_lin[c(&quot;chr1_3238849_3239700&quot;, </span><br><span class="line">                                                 &quot;chr1_3406155_3407044&quot;, </span><br><span class="line">                                                 &quot;chr1_3397204_3397842&quot;)])</span><br></pre></td></tr></table></figure>

<p>当然我们也可以从头分析. 先用<code>aggregate_by_cell_bin</code>对细胞根据相似度进行聚合</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># First, assign a column in the pData table to umap pseudotime</span><br><span class="line">pData(input_cds_lin)$Pseudotime &lt;- pseudotime(input_cds_lin)</span><br><span class="line">pData(input_cds_lin)$cell_subtype &lt;- cut(pseudotime(input_cds_lin), 10)</span><br><span class="line">binned_input_lin &lt;- aggregate_by_cell_bin(input_cds_lin, &quot;cell_subtype&quot;)</span><br></pre></td></tr></table></figure>

<p>之后用Monocle3的<code>fit_models</code>函数寻找差异开发的区域。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"># For speed, run fit_models on 1000 randomly chosen genes</span><br><span class="line">set.seed(1000)</span><br><span class="line">acc_fits &lt;- fit_models(binned_input_lin[sample(1:nrow(fData(binned_input_lin)), 1000),], </span><br><span class="line">                       model_formula_str = &quot;~Pseudotime + num_genes_expressed&quot; )</span><br><span class="line">fit_coefs &lt;- coefficient_table(acc_fits)</span><br><span class="line"></span><br><span class="line"># Subset out the differentially accessible sites with respect to Pseudotime</span><br><span class="line">pseudotime_terms &lt;- subset(fit_coefs, term == &quot;Pseudotime&quot; &amp; q_value &lt; .05)</span><br><span class="line">head(pseudotime_terms)</span><br></pre></td></tr></table></figure>

<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li>官方教程:<a target="_blank" rel="noopener" href="https://cole-trapnell-lab.github.io/cicero-release/docs_m3">https://cole-trapnell-lab.github.io/cicero-release/docs_m3</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/18/Cicero_%20%E5%8D%95%E7%BB%86%E8%83%9E%E5%85%B1%E5%BC%80%E6%94%BE%E5%88%86%E6%9E%90-%20Cicero-single-cell-ATAC-seq-analyzing-tools/" data-id="cllg90xz8000veqp261qy85i1" data-title="Cicero:单细胞共开放分析" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-motifmatchr_ 在R语言中分析peak中里是否有motif匹配-motifmatchr-peak-motif-search-in-R" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/17/motifmatchr_%20%E5%9C%A8R%E8%AF%AD%E8%A8%80%E4%B8%AD%E5%88%86%E6%9E%90peak%E4%B8%AD%E9%87%8C%E6%98%AF%E5%90%A6%E6%9C%89motif%E5%8C%B9%E9%85%8D-motifmatchr-peak-motif-search-in-R/" class="article-date">
  <time class="dt-published" datetime="2019-08-17T15:22:37.235Z" itemprop="datePublished">2019-08-17</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/17/motifmatchr_%20%E5%9C%A8R%E8%AF%AD%E8%A8%80%E4%B8%AD%E5%88%86%E6%9E%90peak%E4%B8%AD%E9%87%8C%E6%98%AF%E5%90%A6%E6%9C%89motif%E5%8C%B9%E9%85%8D-motifmatchr-peak-motif-search-in-R/">motifmatchr:在R语言中分析peak中里是否有motif匹配</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p><code>motifmatchr</code>的作用就是分析众多的序列和众多的motifs, 从中找到哪个序列包含哪个motif. 它的核心函数就是<code>matchMotifs</code>，最大特点就是快，因为它用的是MOODS C++库用于motif匹配。</p>
<p>尽管Bioconductor上也有很多工具能够做motif匹配，比如说<code>Biostrings::mathcPWM</code>, <code>TFBSTools::searchSeq</code>，但是motifmatchr更适合于分析许多不同的序列包含许多不同的motif。例如，当分析ChIP-seq或者ATAC-seq数据时, 你可能会想知道在哪个peak里有哪种类型的motif.</p>
<h2 id="R包安装和加载"><a href="#R包安装和加载" class="headerlink" title="R包安装和加载"></a>R包安装和加载</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>motifmatchr<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>GenomicRanges<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="matchMotifs"><a href="#matchMotifs" class="headerlink" title="matchMotifs"></a>matchMotifs</h2><p>motifmatchr的核心函数是<code>matchMotifs</code>，因此了解这个函数的输入数据是什么格式就行了。必须的两个输入是</p>
<ul>
<li>位置权重矩阵(position weight matrices, PWM)或位置频率矩阵(position frequency matrices, PFM), 保存在PWMatrix, PFMatrix, PWMatrixList或PFMatrixList</li>
<li>一组基因组范围(GenomicRanges或RangedSUmmarizedExperiment对象)或一组序列(DNAStringSet, DNAString 或 简单的字符串向量)</li>
</ul>
<h3 id="PWM或PFM矩阵"><a href="#PWM或PFM矩阵" class="headerlink" title="PWM或PFM矩阵"></a>PWM或PFM矩阵</h3><p>Motif的PWM或PFM矩阵可以从<a target="_blank" rel="noopener" href="http://jaspar.genereg.net/">JASPAR</a>, <a target="_blank" rel="noopener" href="http://cisbp.ccbr.utoronto.ca/">CIS-BP</a>下载。</p>
<p>例如，我从CIS-BP下载拟南芥Arabidopsis_thaliana_2019_08_17_2_32_am.zip，解压缩后里有一个<code>pwm_all_motifs</code>文件夹，里面的文本文件存放的就是我所需要的PWM矩阵,  下一步根据这些文件构建出matchMotifs的输入对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">motif_dir <span class="operator">&lt;-</span> <span class="string">&quot;Arabidopsis_thaliana_2019_08_17_2_32_am/pwms_all_motifs&quot;</span></span><br><span class="line">PWList <span class="operator">&lt;-</span> <span class="built_in">list</span><span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>file <span class="keyword">in</span> list.files<span class="punctuation">(</span>motif_dir<span class="punctuation">,</span> pattern <span class="operator">=</span> <span class="string">&quot;.txt&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  df <span class="operator">&lt;-</span> read.table<span class="punctuation">(</span>file.path<span class="punctuation">(</span>motif_dir<span class="punctuation">,</span> file<span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">                   header <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">,</span></span><br><span class="line">                   row.names <span class="operator">=</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">  mt <span class="operator">&lt;-</span> as.matrix<span class="punctuation">(</span>df<span class="punctuation">)</span></span><br><span class="line">  <span class="keyword">if</span> <span class="punctuation">(</span>nrow<span class="punctuation">(</span>mt<span class="punctuation">)</span> <span class="operator">==</span><span class="number">0</span><span class="punctuation">)</span> <span class="keyword">next</span></span><br><span class="line">  motif_id <span class="operator">&lt;-</span> substr<span class="punctuation">(</span>file<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span><span class="number">6</span><span class="punctuation">)</span></span><br><span class="line">  PWList<span class="punctuation">[[</span>motif_id<span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> PWMatrix<span class="punctuation">(</span>ID <span class="operator">=</span> motif_id<span class="punctuation">,</span> profileMatrix <span class="operator">=</span> t<span class="punctuation">(</span>mt<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">PWMatrixLists <span class="operator">&lt;-</span> do.call<span class="punctuation">(</span>PWMatrixList<span class="punctuation">,</span>PWList<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>对于JASPAR，我们有更加方便的提取方法</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>JASPAR2018<span class="punctuation">)</span></span><br><span class="line">species <span class="operator">&lt;-</span> <span class="string">&quot;Arabidopsis thaliana&quot;</span></span><br><span class="line">collection <span class="operator">&lt;-</span> <span class="string">&quot;CORE&quot;</span></span><br><span class="line">opts <span class="operator">&lt;-</span> <span class="built_in">list</span><span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">opts<span class="punctuation">[</span><span class="string">&quot;species&quot;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> species</span><br><span class="line">opts<span class="punctuation">[</span><span class="string">&quot;collection&quot;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> collection</span><br><span class="line">out  <span class="operator">&lt;-</span> TFBSTools<span class="operator">::</span>getMatrixSet<span class="punctuation">(</span>JASPAR2018<span class="punctuation">,</span> opts<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>isTRUE<span class="punctuation">(</span>all.equal<span class="punctuation">(</span>TFBSTools<span class="operator">::</span>name<span class="punctuation">(</span>out<span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">names</span><span class="punctuation">(</span>out<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> </span><br><span class="line">  <span class="built_in">names</span><span class="punctuation">(</span>out<span class="punctuation">)</span> <span class="operator">&lt;-</span> paste<span class="punctuation">(</span><span class="built_in">names</span><span class="punctuation">(</span>out<span class="punctuation">)</span><span class="punctuation">,</span> TFBSTools<span class="operator">::</span>name<span class="punctuation">(</span>out<span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">                      sep <span class="operator">=</span> <span class="string">&quot;_&quot;</span><span class="punctuation">)</span></span><br><span class="line">motif <span class="operator">&lt;-</span> out</span><br></pre></td></tr></table></figure>

<p>JASPAR的motif比较可靠，因此motif相对比较少。</p>
<h3 id="给定范围或序列"><a href="#给定范围或序列" class="headerlink" title="给定范围或序列"></a>给定范围或序列</h3><p>这部分的输入就比较容易获取，你可以提供MACS2的输出BED，利用<code>rtracklayer::import.bed()</code>读取构建成一个GRanges对象。因为是提供的GRanges对象，那么还需要额外设置一个参数<code>genome</code>, 利用<code>Biostrings::readDNAStringSet()</code>读取一个参考基因组序列就行了。</p>
<p>或者用bedtools先根据bed文件提取数据，然后利用<code>Biostrings::readDNAStringSet()</code>读取</p>
<h3 id="示例数据"><a href="#示例数据" class="headerlink" title="示例数据"></a>示例数据</h3><p>我们以包中提供的数据为例，进行演示</p>
<p>加载实例的motif和GRanges对象</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># load some example motifs</span></span><br><span class="line">data<span class="punctuation">(</span>example_motifs<span class="punctuation">,</span> package <span class="operator">=</span> <span class="string">&quot;motifmatchr&quot;</span><span class="punctuation">)</span> </span><br><span class="line"></span><br><span class="line"><span class="comment"># Make a set of peaks</span></span><br><span class="line">peaks <span class="operator">&lt;-</span> GRanges<span class="punctuation">(</span>seqnames <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;chr1&quot;</span><span class="punctuation">,</span><span class="string">&quot;chr2&quot;</span><span class="punctuation">,</span><span class="string">&quot;chr2&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">                 ranges <span class="operator">=</span> IRanges<span class="punctuation">(</span>start <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">76585873</span><span class="punctuation">,</span><span class="number">42772928</span><span class="punctuation">,</span><span class="number">100183786</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">                                  width <span class="operator">=</span> <span class="number">500</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>获取motif在peak中的位置</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">motif_ix <span class="operator">&lt;-</span> matchMotifs<span class="punctuation">(</span>example_motifs<span class="punctuation">,</span> peaks<span class="punctuation">,</span> genome <span class="operator">=</span> <span class="string">&quot;hg19&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>用<code>motifMatches</code>函数提取匹配矩阵</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">motifMatches<span class="punctuation">(</span>motif_ix<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>输出结果是一个稀疏矩阵</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="number">3</span> x <span class="number">3</span> sparse Matrix of <span class="built_in">class</span> <span class="string">&quot;lgCMatrix&quot;</span></span><br><span class="line">     MA0599.1_KLF5 MA0107.1_RELA MA0137.3_STAT1</span><br><span class="line"><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">,</span><span class="punctuation">]</span>             <span class="operator">|</span>             <span class="operator">|</span>              .</span><br><span class="line"><span class="punctuation">[</span><span class="number">2</span><span class="punctuation">,</span><span class="punctuation">]</span>             <span class="operator">|</span>             .              <span class="operator">|</span></span><br><span class="line"><span class="punctuation">[</span><span class="number">3</span><span class="punctuation">,</span><span class="punctuation">]</span>             <span class="operator">|</span>             .              <span class="operator">|</span></span><br></pre></td></tr></table></figure>

<p>其中的<code>.</code>就表示存在motif匹配。</p>
<p>我们还可以提取motif在peak中的位置</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Get motif positions within peaks for example motifs in peaks </span></span><br><span class="line">motif_ix <span class="operator">&lt;-</span> matchMotifs<span class="punctuation">(</span>example_motifs<span class="punctuation">,</span> peaks<span class="punctuation">,</span> genome <span class="operator">=</span> <span class="string">&quot;hg19&quot;</span><span class="punctuation">,</span></span><br><span class="line">                         out <span class="operator">=</span> <span class="string">&quot;positions&quot;</span><span class="punctuation">)</span> </span><br></pre></td></tr></table></figure>

<h3 id="其他参数"><a href="#其他参数" class="headerlink" title="其他参数"></a>其他参数</h3><p>除了必须的motif信息和你的序列信息输入外，还有一些其他的参数可以做一些设置。</p>
<ul>
<li>背景核苷酸频率, 默认是<code>bg=subject</code>, 也就是你的输入数据作为背景，也可设置成<code>genome</code>或<code>even</code></li>
<li>P值: 用于判断匹配是否足够可信的参数，默认是0.00005，基本上不用修改</li>
<li>输出信息: <code>matchMotifs</code>可以返回三种可能输出，matches, scores 和 positions</li>
</ul>
<p>总的来说，这个R包是一个比较简单的工具，比较容易上手使用。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/17/motifmatchr_%20%E5%9C%A8R%E8%AF%AD%E8%A8%80%E4%B8%AD%E5%88%86%E6%9E%90peak%E4%B8%AD%E9%87%8C%E6%98%AF%E5%90%A6%E6%9C%89motif%E5%8C%B9%E9%85%8D-motifmatchr-peak-motif-search-in-R/" data-id="cllg90y1g006beqp27iba0jpw" data-title="motifmatchr:在R语言中分析peak中里是否有motif匹配" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-R语言的稀疏矩阵学习记录-R-Sparse-Matrix-Note" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/17/R%E8%AF%AD%E8%A8%80%E7%9A%84%E7%A8%80%E7%96%8F%E7%9F%A9%E9%98%B5%E5%AD%A6%E4%B9%A0%E8%AE%B0%E5%BD%95-R-Sparse-Matrix-Note/" class="article-date">
  <time class="dt-published" datetime="2019-08-17T10:18:28.869Z" itemprop="datePublished">2019-08-17</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/17/R%E8%AF%AD%E8%A8%80%E7%9A%84%E7%A8%80%E7%96%8F%E7%9F%A9%E9%98%B5%E5%AD%A6%E4%B9%A0%E8%AE%B0%E5%BD%95-R-Sparse-Matrix-Note/">R语言的稀疏矩阵学习记录</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>一个很大的矩阵， 320127 行,  8189列，假如用一个全为0的普通矩阵来存储，需要用到9.8Gb</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">cols <span class="operator">&lt;-</span> 8189</span><br><span class="line">rows <span class="operator">&lt;-</span> 320127</span><br><span class="line">mat <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span>data <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> nrow<span class="operator">=</span><span class="number">320127</span><span class="punctuation">,</span> ncol <span class="operator">=</span> <span class="number">8189</span><span class="punctuation">)</span></span><br><span class="line">print<span class="punctuation">(</span>object.size<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">,</span> unit<span class="operator">=</span><span class="string">&quot;GB&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 19.5 Gb</span></span><br><span class="line">mat <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span>data <span class="operator">=</span> <span class="number">0L</span><span class="punctuation">,</span> nrow<span class="operator">=</span><span class="number">320127</span><span class="punctuation">,</span> ncol <span class="operator">=</span> <span class="number">8189</span><span class="punctuation">)</span></span><br><span class="line">print<span class="punctuation">(</span>object.size<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">,</span> unit<span class="operator">=</span><span class="string">&quot;GB&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 9.8 Gb这里的0其实也要区分</span></span><br></pre></td></tr></table></figure>

<p>这里的<code>0L</code>表示数据类型是<code>integer</code>，默认是<code>numeric</code>. 这两者最大的区别在于，当你用<code>320127L * 8189L</code>，你会得到一个NA，而<code>320127 * 8189</code>不会</p>
<p>如果用稀疏矩阵保存的话</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">mat <span class="operator">&lt;-</span> Matrix<span class="punctuation">(</span>data <span class="operator">=</span> <span class="number">0L</span><span class="punctuation">,</span> nrow<span class="operator">=</span><span class="number">320127</span><span class="punctuation">,</span> ncol <span class="operator">=</span> <span class="number">8189</span><span class="punctuation">,</span> sparse <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">print<span class="punctuation">(</span>object.size<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">,</span> unit<span class="operator">=</span><span class="string">&quot;GB&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">#0 Gb</span></span><br><span class="line"><span class="built_in">dim</span><span class="punctuation">(</span>mat<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#[1] 320127   8189</span></span><br></pre></td></tr></table></figure>

<p>虽然行列数一样，但是稀疏矩阵几乎不占用任何内存。而且普通矩阵支持的运算，比如说求行和，求列和，提取元素的操作，在稀疏矩阵矩阵也是可以的，只不过会多花一点点时间而已。同时还有很多R包支持稀疏矩阵，比如说<code>glmnet</code>，一个做lasso回归的R包。</p>
<p>虽然看起来稀疏矩阵很美好，但是在R语言中那么大的稀疏矩阵的部分操作会出错</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> mat2 <span class="operator">&lt;-</span> mat <span class="operator">+</span> <span class="number">1</span></span><br><span class="line">Error <span class="keyword">in</span> asMethod<span class="punctuation">(</span>object<span class="punctuation">)</span> <span class="operator">:</span> </span><br><span class="line">  Cholmod error <span class="string">&#x27;problem too large&#x27;</span> at file ..<span class="operator">/</span>Core<span class="operator">/</span>cholmod_dense.c<span class="punctuation">,</span> line <span class="number">105</span></span><br></pre></td></tr></table></figure>

<p>即便是我想把它用<code>as.matrix</code>转回普通矩阵，它也报错了</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> mat3 <span class="operator">&lt;-</span> Matrix<span class="operator">::</span>as.matrix<span class="punctuation">(</span>mat<span class="punctuation">)</span></span><br><span class="line">Error <span class="keyword">in</span> asMethod<span class="punctuation">(</span>object<span class="punctuation">)</span> <span class="operator">:</span> </span><br><span class="line">  Cholmod error <span class="string">&#x27;problem too large&#x27;</span> at file ..<span class="operator">/</span>Core<span class="operator">/</span>cholmod_dense.c<span class="punctuation">,</span> line <span class="number">105</span></span><br></pre></td></tr></table></figure>

<p>既然现成的<code>as.matrix</code>无法处理，那怎么办呢？最简单粗暴的方法就是新建一个普通矩阵，然后对稀疏矩阵进行遍历，将稀疏矩阵的值挨个放回到的普通矩阵上。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">mat2 <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span>data <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> nrow<span class="operator">=</span><span class="number">320127</span><span class="punctuation">,</span> ncol <span class="operator">=</span> <span class="number">8189</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_len</span><span class="punctuation">(</span>nrow<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    <span class="keyword">for</span> <span class="punctuation">(</span>j <span class="keyword">in</span> <span class="built_in">seq_len</span><span class="punctuation">(</span>ncol<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">        mat2<span class="punctuation">[</span>i<span class="punctuation">,</span>j<span class="punctuation">]</span> <span class="operator">&lt;-</span> mat<span class="punctuation">[</span>i<span class="punctuation">,</span>j<span class="punctuation">]</span></span><br><span class="line">    <span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>那么这大概要多少时间呢？反正我的电脑跑了2个小时也没有跑完，所以你也别测试了。</p>
<p>那有没有办法可以加速呢？加速的方法就是减少for循环的次数，因为我们是一个稀疏矩阵，大部分的空间都是0，我们只需要将不为0的部分赋值给新矩阵即可。</p>
<p>这需要我们去了解下稀疏矩阵的数据结构</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="operator">&gt;</span> str<span class="punctuation">(</span>mat<span class="punctuation">)</span></span><br><span class="line">Formal <span class="built_in">class</span> <span class="string">&#x27;dgCMatrix&#x27;</span> <span class="punctuation">[</span>package <span class="string">&quot;Matrix&quot;</span><span class="punctuation">]</span> with <span class="number">6</span> slots</span><br><span class="line">  ..<span class="operator">@</span> i       <span class="operator">:</span> int<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">)</span> </span><br><span class="line">  ..<span class="operator">@</span> p       <span class="operator">:</span> int <span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">8190</span><span class="punctuation">]</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> <span class="number">0</span> ...</span><br><span class="line">  ..<span class="operator">@</span> Dim     <span class="operator">:</span> int <span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">2</span><span class="punctuation">]</span> <span class="number">320127</span> <span class="number">8189</span></span><br><span class="line">  ..<span class="operator">@</span> Dimnames<span class="operator">:</span>List of <span class="number">2</span></span><br><span class="line">  .. ..<span class="operator">$</span> <span class="operator">:</span> <span class="literal">NULL</span></span><br><span class="line">  .. ..<span class="operator">$</span> <span class="operator">:</span> <span class="literal">NULL</span></span><br><span class="line">  ..<span class="operator">@</span> x       <span class="operator">:</span> num<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">)</span> </span><br><span class="line">  ..<span class="operator">@</span> factors <span class="operator">:</span> <span class="built_in">list</span><span class="punctuation">(</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><code>@Dim</code>记录矩阵的维度信息, <code>@Dimnames</code>记录行名和列名, <code>@x</code>记录不为0的数值。<code>@i</code>记录不为0的行索引，和<code>@x</code>对应，这里全为0，所以不记录。<code>@p</code>比较复杂，并不是简单的记录不为0值的列索引，看文档也不知道是啥，不过通过检索可以找到它和不为0值的列索引的换算关系。</p>
<p>因此代码优化为</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">row_pos <span class="operator">&lt;-</span> mat<span class="operator">@</span>i<span class="operator">+</span><span class="number">1</span></span><br><span class="line">col_pos <span class="operator">&lt;-</span> findInterval<span class="punctuation">(</span>seq<span class="punctuation">(</span>mat<span class="operator">@</span>x<span class="punctuation">)</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span>mat<span class="operator">@</span>p<span class="punctuation">[</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="operator">+</span><span class="number">1</span></span><br><span class="line">val <span class="operator">&lt;-</span> mat<span class="operator">@</span>x</span><br><span class="line">    </span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>val<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    tmp<span class="punctuation">[</span>row_pos<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">,</span>col_pos<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> val<span class="punctuation">[</span>i<span class="punctuation">]</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>可以将其封装为一个函数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">as_matrix <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line"></span><br><span class="line">  tmp <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span>data<span class="operator">=</span><span class="number">0L</span><span class="punctuation">,</span> nrow <span class="operator">=</span> mat<span class="operator">@</span>Dim<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span> ncol <span class="operator">=</span> mat<span class="operator">@</span>Dim<span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  row_pos <span class="operator">&lt;-</span> mat<span class="operator">@</span>i<span class="operator">+</span><span class="number">1</span></span><br><span class="line">  col_pos <span class="operator">&lt;-</span> findInterval<span class="punctuation">(</span>seq<span class="punctuation">(</span>mat<span class="operator">@</span>x<span class="punctuation">)</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span>mat<span class="operator">@</span>p<span class="punctuation">[</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="operator">+</span><span class="number">1</span></span><br><span class="line">  val <span class="operator">&lt;-</span> mat<span class="operator">@</span>x</span><br><span class="line">    </span><br><span class="line">  <span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>val<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">      tmp<span class="punctuation">[</span>row_pos<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">,</span>col_pos<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> val<span class="punctuation">[</span>i<span class="punctuation">]</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">    </span><br><span class="line">  row.names<span class="punctuation">(</span>tmp<span class="punctuation">)</span> <span class="operator">&lt;-</span> mat<span class="operator">@</span>Dimnames<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">  colnames<span class="punctuation">(</span>tmp<span class="punctuation">)</span> <span class="operator">&lt;-</span> mat<span class="operator">@</span>Dimnames<span class="punctuation">[[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>tmp<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>如果速度还需要提高，那么可能就需要Rcpp上场了. 我参考着<a target="_blank" rel="noopener" href="http://adv-r.had.co.nz/Rcpp.html">http://adv-r.had.co.nz/Rcpp.html</a>写了一个简单的代码</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line">Rcpp<span class="operator">::</span>sourceCpp<span class="punctuation">(</span>code<span class="operator">=</span><span class="string">&#x27;</span></span><br><span class="line"><span class="string">#include &lt;Rcpp.h&gt;</span></span><br><span class="line"><span class="string">using namespace Rcpp;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">// [[Rcpp::export]]</span></span><br><span class="line"><span class="string">IntegerMatrix asMatrix(NumericVector rp,</span></span><br><span class="line"><span class="string">                       NumericVector cp,</span></span><br><span class="line"><span class="string">                       NumericVector z,</span></span><br><span class="line"><span class="string">                       int nrows,</span></span><br><span class="line"><span class="string">                       int ncols)&#123;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  int k = z.size() ;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  IntegerMatrix  mat(nrows, ncols);</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  for (int i = 0; i &lt; k; i++)&#123;</span></span><br><span class="line"><span class="string">      mat(rp[i],cp[i]) = z[i];</span></span><br><span class="line"><span class="string">  &#125;</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  return mat;</span></span><br><span class="line"><span class="string">&#125;</span></span><br><span class="line"><span class="string">&#x27;</span> <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">as_matrix <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line"></span><br><span class="line">  row_pos <span class="operator">&lt;-</span> mat<span class="operator">@</span>i</span><br><span class="line">  col_pos <span class="operator">&lt;-</span> findInterval<span class="punctuation">(</span>seq<span class="punctuation">(</span>mat<span class="operator">@</span>x<span class="punctuation">)</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span>mat<span class="operator">@</span>p<span class="punctuation">[</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  tmp <span class="operator">&lt;-</span> asMatrix<span class="punctuation">(</span>rp <span class="operator">=</span> row_pos<span class="punctuation">,</span> cp <span class="operator">=</span> col_pos<span class="punctuation">,</span> z <span class="operator">=</span> mat<span class="operator">@</span>x<span class="punctuation">,</span></span><br><span class="line">                 nrows <span class="operator">=</span>  mat<span class="operator">@</span>Dim<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span> ncols <span class="operator">=</span> mat<span class="operator">@</span>Dim<span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">  row.names<span class="punctuation">(</span>tmp<span class="punctuation">)</span> <span class="operator">&lt;-</span> mat<span class="operator">@</span>Dimnames<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">  colnames<span class="punctuation">(</span>tmp<span class="punctuation">)</span> <span class="operator">&lt;-</span> mat<span class="operator">@</span>Dimnames<span class="punctuation">[[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>tmp<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>


<p>如果之前的矩阵有78945836个元素，<code>system.time</code>显示只需要40s。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/17/R%E8%AF%AD%E8%A8%80%E7%9A%84%E7%A8%80%E7%96%8F%E7%9F%A9%E9%98%B5%E5%AD%A6%E4%B9%A0%E8%AE%B0%E5%BD%95-R-Sparse-Matrix-Note/" data-id="cllg90y0v004ceqp2dwhrhe7i" data-title="R语言的稀疏矩阵学习记录" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-chromVAR_ 从单细胞表观数据中推断转录因子相关开放区域-chromVAR-inferring-transcription-factor-associated-accessibility-from-single-cell-epigenomic-data" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/16/chromVAR_%20%E4%BB%8E%E5%8D%95%E7%BB%86%E8%83%9E%E8%A1%A8%E8%A7%82%E6%95%B0%E6%8D%AE%E4%B8%AD%E6%8E%A8%E6%96%AD%E8%BD%AC%E5%BD%95%E5%9B%A0%E5%AD%90%E7%9B%B8%E5%85%B3%E5%BC%80%E6%94%BE%E5%8C%BA%E5%9F%9F-chromVAR-inferring-transcription-factor-associated-accessibility-from-single-cell-epigenomic-data/" class="article-date">
  <time class="dt-published" datetime="2019-08-16T13:47:16.858Z" itemprop="datePublished">2019-08-16</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/16/chromVAR_%20%E4%BB%8E%E5%8D%95%E7%BB%86%E8%83%9E%E8%A1%A8%E8%A7%82%E6%95%B0%E6%8D%AE%E4%B8%AD%E6%8E%A8%E6%96%AD%E8%BD%AC%E5%BD%95%E5%9B%A0%E5%AD%90%E7%9B%B8%E5%85%B3%E5%BC%80%E6%94%BE%E5%8C%BA%E5%9F%9F-chromVAR-inferring-transcription-factor-associated-accessibility-from-single-cell-epigenomic-data/">chromVAR:从单细胞表观数据中推断转录因子相关开放区域</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>chromVAR是一个用于分析稀疏染色质开放的R包。chromVAR的输入文件包括，ATAC-seq处理后的fragments文件（过滤重复和低质量数据）, DNAse-seq实验结果，以及基因组注释（例如motif位置）</p>
<p>chromVAR先根据所有细胞或者样本的平均情况来计算<strong>期望开放性</strong>， 然后用它来计算每个注释，每个细胞或样本的偏差，最后对开放进行纠正。</p>
<h2 id="安装加载"><a href="#安装加载" class="headerlink" title="安装加载"></a>安装加载</h2><p>安装R包</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">BiocManager::install(&quot;GreenleafLab/chromVAR&quot;)</span><br><span class="line">BiocManager::install(&quot;motifmatchr&quot;)</span><br><span class="line">BiocManager::install(SummarizedExperiment)</span><br><span class="line">BiocManager::install(BiocParallel)</span><br><span class="line">BiocManager::install(&quot;JASPAR2018&quot;) # JASPAR 2018数据库</span><br></pre></td></tr></table></figure>

<p>chromVAR的运行需要先加载下面这些R包</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">library(chromVAR)</span><br><span class="line">library(motifmatchr)</span><br><span class="line">library(Matrix)</span><br><span class="line">library(SummarizedExperiment)</span><br><span class="line">library(BiocParallel)</span><br><span class="line">set.seed(2019)</span><br></pre></td></tr></table></figure>

<p>chromVAR能够使用多核进行并行运算，调用方法如下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># 全部用户</span><br><span class="line">register(MulticoreParam(8, progressbar = TRUE))</span><br><span class="line"># Windows用户，MulticoreParam不好用就用SnowParam</span><br><span class="line">register(SnowParam(workers = 1, type = &quot;SOCK&quot;))</span><br></pre></td></tr></table></figure>

<p><strong>注意</strong>: 不运行的话，后续代码可能会报错</p>
<h2 id="读取输入"><a href="#读取输入" class="headerlink" title="读取输入"></a>读取输入</h2><p>chromVAR接受的输入是落入开放区域的read数统计表。有许多软件可以做到，chromVAR也提供了相应的方法。</p>
<p>首先要提供一个peak文件，文档建议这个peak文件存放的peak为等宽非重叠，建议宽度在<strong>250-500</strong> bp之间。peak文件可以利用已有的bulk ATAC-seq或DNAse-seq数据来获取。对于来自于多个样本的peak，需要先用<code>filterPeaks</code>函数保证peak之间不重合。</p>
<p>使用<code>getPeaks</code>读取peak</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">peakfile &lt;- system.file(&quot;extdata/test_bed.txt&quot;, package = &quot;chromVAR&quot;)</span><br><span class="line">peaks &lt;- getPeaks(peakfile, sort_peaks = TRUE)</span><br></pre></td></tr></table></figure>

<p>MACS2分析结果里会提供narrowpeak格式文件，chromVAR提供了<code>readNarrowpeaks</code>函数进行读取。</p>
<p>随后用<code>getCounts</code>函数基于BAM文件和加载的peak获取count</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">bamfile &lt;- system.file(&quot;extdata/test_RG.bam&quot;, package = &quot;chromVAR&quot;)</span><br><span class="line">fragment_counts &lt;- getCounts(bamfile, </span><br><span class="line">                             peaks, </span><br><span class="line">                             paired =  TRUE, </span><br><span class="line">                             by_rg = TRUE, </span><br><span class="line">                             format = &quot;bam&quot;, </span><br><span class="line">                             colData = DataFrame(celltype = &quot;GM&quot;))</span><br></pre></td></tr></table></figure>

<p>bamfile可以有多个bam文件路径，bam文件的类型和<code>colData</code>对应。此外<code>by_rg</code>定义是否要根据BAM文件中的RG对输入分组。</p>
<p>实际演示的时候，我们官方提供的示例数据</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">data(example_counts, package = &quot;chromVAR&quot;)</span><br></pre></td></tr></table></figure>

<p>因为这是一个<code>*SummarizedExperiment </code>对象，因此我们就能用<code>assay</code>,<code>colData</code>和<code>rowData</code> 了解这个数据集</p>
<p>主体是一个存放count的dgCMatrix</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">assay(example_counts)[1:5,1:2]</span><br><span class="line"></span><br><span class="line">5 x 2 sparse Matrix of class &quot;dgCMatrix&quot;</span><br><span class="line">     singles-GM12878-140905-1 singles-GM12878-140905-2</span><br><span class="line">[1,]                        .                        1</span><br><span class="line">[2,]                        .                        .</span><br><span class="line">[3,]                        .                        .</span><br><span class="line">[4,]                        .                        .</span><br><span class="line">[5,]                        .                        .</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>行是特征(feature), 这里就是peak</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">head(rowData(example_counts), n=2)</span><br><span class="line"></span><br><span class="line">DataFrame with 2 rows and 3 columns</span><br><span class="line">      score      qval        name</span><br><span class="line">  &lt;integer&gt; &lt;numeric&gt; &lt;character&gt;</span><br><span class="line">1       259  25.99629   GM_peak_6</span><br><span class="line">2        82   8.21494   H1_peak_</span><br></pre></td></tr></table></figure>

<p>列表示样本，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">head(colData(example_counts), n=2)</span><br><span class="line"></span><br><span class="line">DataFrame with 2 rows and 2 columns</span><br><span class="line">                           Cell_Type     depth</span><br><span class="line">                         &lt;character&gt; &lt;integer&gt;</span><br><span class="line">singles-GM12878-140905-1          GM      9220</span><br><span class="line">singles-GM12878-140905-2          GM      9401</span><br></pre></td></tr></table></figure>

<h2 id="前期准备"><a href="#前期准备" class="headerlink" title="前期准备"></a>前期准备</h2><h3 id="分析peak中的GC含量"><a href="#分析peak中的GC含量" class="headerlink" title="分析peak中的GC含量"></a>分析peak中的GC含量</h3><p>GC含量信息用于确定哪些peak可能是背景， <code>addGCBias</code>返回更新后的<code>SummarizedExperiment </code>。其中<code>genome</code>支持BSgenome, FaFile和DNAStringSet对象。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">library(BSgenome.Hsapiens.UCSC.hg19)</span><br><span class="line">example_counts &lt;- addGCBias(example_counts, </span><br><span class="line">                            genome = BSgenome.Hsapiens.UCSC.hg19)</span><br><span class="line">head(rowData(example_counts))</span><br></pre></td></tr></table></figure>

<h3 id="过滤输入-单细胞"><a href="#过滤输入-单细胞" class="headerlink" title="过滤输入(单细胞)"></a>过滤输入(单细胞)</h3><p>如果是处理单细胞数据，建议过滤测序深度不够，或者是信噪比低不够(也就是peak中所占read比例低)的数据，主要靠两个参数,<code>min_in_peaks</code>和<code>min_depth</code>.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">counts_filtered &lt;- filterSamples(example_counts, min_depth = 1500, </span><br><span class="line">                                 min_in_peaks = 0.15, shiny = FALSE)</span><br></pre></td></tr></table></figure>

<p>然后我们用<code>filterSamplesPlot</code>对过滤情况进行可视化</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">filtering_plot &lt;- filterSamplesPlot(example_counts, min_depth = 1500, </span><br><span class="line">                                    min_in_peaks = 0.15, use_plotly = FALSE)</span><br><span class="line">filtering_plot</span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/filtering_plot-47e10a7514e04f8dbbbecdb27511e577.jpeg"></p>
<p>确认标准后，就可以过滤</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">counts_filtered &lt;- filterPeaks(counts_filtered, non_overlapping = TRUE)</span><br></pre></td></tr></table></figure>

<h3 id="获取Motifs和分析包含motif的peak"><a href="#获取Motifs和分析包含motif的peak" class="headerlink" title="获取Motifs和分析包含motif的peak"></a>获取Motifs和分析包含motif的peak</h3><p>可以用<code>getJasparMotifs</code>在JASPAR数据库中提取motif信息，但是其中<code>getJasparMotifs</code>只是<code>TFBSTools::getMatrixSet</code>一个简单的封装而已,  默认用的是JASPAR2016, 建议通过下面的代码获取最新版本的数据。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"># BiocManager::install(&quot;JASPAR2018&quot;)</span><br><span class="line">library(JASPAR2018)</span><br><span class="line">species &lt;- &quot;Homo sapiens&quot;</span><br><span class="line">collection &lt;- &quot;CORE&quot;</span><br><span class="line">opts &lt;- list()</span><br><span class="line">opts[&quot;species&quot;] &lt;- species</span><br><span class="line">opts[&quot;collection&quot;] &lt;- collection</span><br><span class="line">out  &lt;- TFBSTools::getMatrixSet(JASPAR2018, opts)</span><br><span class="line"></span><br><span class="line">if (!isTRUE(all.equal(TFBSTools::name(out), names(out)))) </span><br><span class="line">  names(out) &lt;- paste(names(out), TFBSTools::name(out), </span><br><span class="line">                      sep = &quot;_&quot;)</span><br><span class="line">motif &lt;- out</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><code>collection</code>参数中接受: “CORE”, “CNE”, “PHYLOFACTS”, “SPLICE”, “POLII”, “FAM”, “PBM”, “PBM_HOMEO”, “PBM_HLH” 选项。</p>
<p>获取Motif的另外一种方式是用<code>chromVARmotifs</code>, 这个R包的主要功能就是整合了一些motif, 通过<code>devtools::install_github(&quot;GreenleafLab/chromVARmotifs&quot;)</code>进行安装。</p>
<p>之后用<code>motifmatchr</code>中的<code>macthMotifs</code>去分析peak中motif包含情况。默认会返回一个SummarizedExperiment 对象，就是一个稀疏矩阵，来提示是不是匹配了motif</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">library(motifmatchr)</span><br><span class="line">motif_ix &lt;- matchMotifs(motifs, counts_filtered, </span><br><span class="line">                        genome = BSgenome.Hsapiens.UCSC.hg19)</span><br></pre></td></tr></table></figure>

<p>关键参数是<code>p.cutoff</code>用于设置严格度，默认是<code>0.00005</code>其实能够返回比较合理的结果。如果需要一些额外信息，可以通过参数<code>out</code>来调整，例如<code>out=positions</code>表示返回实际的匹配位置</p>
<h2 id="偏离度和变异度分析-Deviations-and-Variability"><a href="#偏离度和变异度分析-Deviations-and-Variability" class="headerlink" title="偏离度和变异度分析(Deviations and Variability )"></a>偏离度和变异度分析(Deviations and Variability )</h2><p>上面都是准备阶段，计算deviations和Variability才是这个软件的重点。</p>
<h3 id="计算偏离度"><a href="#计算偏离度" class="headerlink" title="计算偏离度"></a>计算偏离度</h3><p>函数<code>computeDeviations</code>返回的SummarizedExperiment 对象, 包含两个”assays”, </p>
<ul>
<li>偏差纠正后的开放偏离度:<code>assays(dev)$deviations</code></li>
<li>偏离度的Z-score: <code>assays(deviations)$z</code>: 考虑到了当从基因组上随机抽取一些GC含量类似的片段分析时，有多大概率会得到该结果。</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">dev &lt;- computeDeviations(object = counts_filtered, annotations = motif_ix)</span><br></pre></td></tr></table></figure>

<p><code>compuateDeviations</code>支持背景peak的输入，用于的偏移度得分进行标准化，默认会自动计算不会返回结果。你可以选择用<code>getBackgroundPeaks</code>分析背景，然后传递给<code>computeDeviations</code></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">bg &lt;- getBackgroundPeaks(object = counts_filtered)</span><br><span class="line">dev &lt;- computeDeviations(object = counts_filtered, annotations = motif_ix,</span><br><span class="line">                         background_peaks = bg)</span><br></pre></td></tr></table></figure>

<h2 id="变异度"><a href="#变异度" class="headerlink" title="变异度"></a>变异度</h2><p><code>computeVariability</code>会返回一个数据框<code>data.frame</code>，里面有变异度, 该变异度的bootstrap置信区间和衡量拒绝原假设的概率(即 &gt; 1)</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">variability &lt;- computeVariability(dev)</span><br><span class="line">plotVariability(variability, use_plotly = FALSE) </span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/Variability-2c06b2c8bfa447159f32cd4dd0b478c3.jpeg"></p>
<h3 id="偏离度可视化"><a href="#偏离度可视化" class="headerlink" title="偏离度可视化"></a>偏离度可视化</h3><p>可以利用tSNE可视化偏离度</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">tsne_results &lt;- deviationsTsne(dev, threshold = 1.5, perplexity = 10)</span><br><span class="line">tsne_plots &lt;- plotDeviationsTsne(dev, tsne_results, </span><br><span class="line">                                 annotation_name = &quot;TEAD3&quot;, </span><br><span class="line">                                 sample_column = &quot;Cell_Type&quot;, </span><br><span class="line">                                 shiny = FALSE)</span><br><span class="line">cowplot::plot_grid(tsne_plots[[1]],tsne_plots[[1]] )</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/Deviations_tSNE-7fcaa606061241ffb3686d8f8cc8e024.jpeg"></p>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ul>
<li>数据库<a target="_blank" rel="noopener" href="http://jaspar.genereg.net/">JASPAR</a>是一个开放性数据库，用于存放人工审查的非冗余转录因子(TF)结合谱，数据格式为PFM(position frequency matrices ), TFFM( TF flexible models)</li>
<li><a target="_blank" rel="noopener" href="https://greenleaflab.github.io/chromVAR/articles/Introduction.html#get-motifs-and-what-peaks-contain-motifs">chromVAR</a>官方文档</li>
<li><a target="_blank" rel="noopener" href="https://www.nature.com/articles/nmeth.4401">chromVAR: inferring transcription-factor-associated accessibility from single-cell epigenomic data</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/16/chromVAR_%20%E4%BB%8E%E5%8D%95%E7%BB%86%E8%83%9E%E8%A1%A8%E8%A7%82%E6%95%B0%E6%8D%AE%E4%B8%AD%E6%8E%A8%E6%96%AD%E8%BD%AC%E5%BD%95%E5%9B%A0%E5%AD%90%E7%9B%B8%E5%85%B3%E5%BC%80%E6%94%BE%E5%8C%BA%E5%9F%9F-chromVAR-inferring-transcription-factor-associated-accessibility-from-single-cell-epigenomic-data/" data-id="cllg90y1b005seqp2bv5o5el3" data-title="chromVAR:从单细胞表观数据中推断转录因子相关开放区域" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-生物信息学格式和转换工具整理-生物信息学格式和转换整理" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/15/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E5%B7%A5%E5%85%B7%E6%95%B4%E7%90%86-%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E6%95%B4%E7%90%86/" class="article-date">
  <time class="dt-published" datetime="2019-08-15T14:04:57.682Z" itemprop="datePublished">2019-08-15</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/15/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E5%B7%A5%E5%85%B7%E6%95%B4%E7%90%86-%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E6%95%B4%E7%90%86/">生物信息学格式和转换工具整理</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>生信会用到很多格式，不同软件会要求不同的输入格式，所以得要自己整理下。</p>
<h2 id="FASTA和FASTQ"><a href="#FASTA和FASTQ" class="headerlink" title="FASTA和FASTQ"></a>FASTA和FASTQ</h2><h2 id="SAM和BAM"><a href="#SAM和BAM" class="headerlink" title="SAM和BAM"></a>SAM和BAM</h2><h2 id="GTF和GFF"><a href="#GTF和GFF" class="headerlink" title="GTF和GFF"></a>GTF和GFF</h2><h2 id="BED"><a href="#BED" class="headerlink" title="BED"></a>BED</h2><h2 id="BedGraph"><a href="#BedGraph" class="headerlink" title="BedGraph"></a>BedGraph</h2><h2 id="Wig和BigWig"><a href="#Wig和BigWig" class="headerlink" title="Wig和BigWig"></a>Wig和BigWig</h2><p>WIG（wiggle）设计的目的是在基因组浏览器上展示连续性数据。Wiggle数据要求元素等长。对于稀疏或者元素之间不等长，建议用BedGraph格式。而BigWig则是二进制版，体积小，适合传输。</p>
<p>举个例子</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">variableStep chrom=chr2</span><br><span class="line">300701  12.5</span><br><span class="line">300702  12.5</span><br><span class="line">300703  12.5</span><br><span class="line">300704  12.5</span><br><span class="line">300705  12.5</span><br></pre></td></tr></table></figure>

<p>和</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">variableStep chrom=chr2 span=5</span><br><span class="line">300701  12.5</span><br></pre></td></tr></table></figure>

<p>在基因组浏览器上展示时都是2号染色体的300701-300705的值为12.5</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/15/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E5%B7%A5%E5%85%B7%E6%95%B4%E7%90%86-%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E6%A0%BC%E5%BC%8F%E5%92%8C%E8%BD%AC%E6%8D%A2%E6%95%B4%E7%90%86/" data-id="cllg90y6q00jteqp2474v8n2i" data-title="生物信息学格式和转换工具整理" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
    </footer>
  </div>
  
</article>



  
    <article id="post-实现Seurat 2 和 Seurat3 在同一个环境下共存-Seuarat2-Seurat3-together" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/14/%E5%AE%9E%E7%8E%B0Seurat%202%20%E5%92%8C%20Seurat3%20%E5%9C%A8%E5%90%8C%E4%B8%80%E4%B8%AA%E7%8E%AF%E5%A2%83%E4%B8%8B%E5%85%B1%E5%AD%98-Seuarat2-Seurat3-together/" class="article-date">
  <time class="dt-published" datetime="2019-08-14T10:31:49.412Z" itemprop="datePublished">2019-08-14</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/14/%E5%AE%9E%E7%8E%B0Seurat%202%20%E5%92%8C%20Seurat3%20%E5%9C%A8%E5%90%8C%E4%B8%80%E4%B8%AA%E7%8E%AF%E5%A2%83%E4%B8%8B%E5%85%B1%E5%AD%98-Seuarat2-Seurat3-together/">实现Seurat 2 和 Seurat3 在同一个环境下共存</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>软件升级虽然是一件值得高兴的是，但是代码变化太大却不是一件好消息。比如说Seurat，这个单细胞分析最常用的R包，它的2.x版本和3.x版本的变化就是翻天覆地。</p>
<p>为了能够重现别人的代码，你可能需要重装2.3.4版本的Seurat，官方提供了安装脚本</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">source<span class="punctuation">(</span><span class="string">&quot;https://gist.githubusercontent.com/satijalab/beb9bb50dedc75ee023bd5d9be5fe684/raw/e103577735a2fba9da2ccca14ce1ac33e46c1bc4/old_seurat.R&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>但是在window下安装会出一个报错, 提示无法在Windows下安装MacOS的版本</p>
<p>这个时候需要手动安装Seurat，注意，这里要求有Rtools才能进行编译</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">install.packages<span class="punctuation">(</span><span class="string">&#x27;Seurat&#x27;</span><span class="punctuation">,</span> repos <span class="operator">=</span> <span class="string">&#x27;https://satijalab.org/ran&#x27;</span><span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;source&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>可以设置<code>install.packages</code>里的参数lib.loc，让Seurat安装到其他的文件加，就不会替换原来的Seurat，同时用<code>library</code>加载的时候，也需要设置<code>lib.loc</code>.</p>
<p>上面是简单可行易操作的方法，唯一的问题是你不能同时加载两个Seurat版本，当然你也不会想这样子做，所以这是一个伪需求。</p>
<p>下面就是瞎折腾环节， 我要将Seurat的2.3.4版本单独搞出一个R包，Seurat2，这样子就可以同时加载这两个R包。</p>
<p>下载Seurat 2.3.4</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wget https://satijalab.org/ran/src/contrib/Seurat_2.3.4.tar.gz</span><br></pre></td></tr></table></figure>

<p>解压缩它</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">tar xf Seurat_2.3.4.tar.gz</span><br><span class="line"><span class="built_in">cd</span> Seurat</span><br></pre></td></tr></table></figure>

<p>删除MD5文件，因为它会做文件检验</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">rm</span> MD5</span><br></pre></td></tr></table></figure>

<p>修改里面所有的Seurat替换成Seurat2, seurat替换成seurat2</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">find . -<span class="built_in">type</span> f -print0 | xargs -0 sed -i <span class="string">&quot;s/Seurat/Seurat2/g&quot;</span></span><br><span class="line">find . -<span class="built_in">type</span> f -print0 | xargs -0 sed -i <span class="string">&quot;s/seurat/seurat2/g&quot;</span></span><br></pre></td></tr></table></figure>

<p>这种无差别的替换会有一个问题，会把一些这类<code>http://www.satijalab.org/seurat</code>非代码信息中的seurat替换成seurat2，不过这并不影响实际函数的使用。</p>
<p>将<code>R/seurat.R</code>重名为<code>R/seurat2.R</code></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mv</span> R/seurat.R R/seurat2.R</span><br></pre></td></tr></table></figure>

<p>之后打包修改后的文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">tar -czf Seurat2.tar.gz Seurat</span><br></pre></td></tr></table></figure>

<p>于是我们就可以安装我们自己修改后的R包了</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">install.packages<span class="punctuation">(</span><span class="string">&quot;Seurat2.tar.gz&quot;</span><span class="punctuation">,</span> repos<span class="operator">=</span><span class="literal">NULL</span><span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&quot;source&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><strong>注意</strong>: 目前未修改测试数据集的对象，所以不能用Seurat2来运行pbmc_small的例子</p>
<p>由于我已经修改好了，所以你们也不需要自己再去修改了。我将其上传到GitHub，所以可以用devtools进行安装。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">devtools<span class="operator">::</span>install_github<span class="punctuation">(</span><span class="string">&quot;xuzhougeng/Seurat2&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>由于代码只是简单修改，存在bug，其实不建议尝试使用。而且安装完Seurat2之后，我<strong>定性</strong>的评估了下Seurat2和Seurat3，发现两者在聚类分析上并没有太大区别, 所以更推荐大家使用最新的<code>Seurat</code>.</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/14/%E5%AE%9E%E7%8E%B0Seurat%202%20%E5%92%8C%20Seurat3%20%E5%9C%A8%E5%90%8C%E4%B8%80%E4%B8%AA%E7%8E%AF%E5%A2%83%E4%B8%8B%E5%85%B1%E5%AD%98-Seuarat2-Seurat3-together/" data-id="cllg90y6500hxeqp2h91rgmxx" data-title="实现Seurat 2 和 Seurat3 在同一个环境下共存" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-用rtracklayer读取和输出BigWig-rtracklayer-BigWig" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/13/%E7%94%A8rtracklayer%E8%AF%BB%E5%8F%96%E5%92%8C%E8%BE%93%E5%87%BABigWig-rtracklayer-BigWig/" class="article-date">
  <time class="dt-published" datetime="2019-08-13T19:58:24.434Z" itemprop="datePublished">2019-08-13</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/R/">R</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/13/%E7%94%A8rtracklayer%E8%AF%BB%E5%8F%96%E5%92%8C%E8%BE%93%E5%87%BABigWig-rtracklayer-BigWig/">用rtracklayer读取和输出BigWig</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>BigWig是一个能用于加载到基因组浏览器上展示的格式。它的格式比较复杂，不适合直接阅读，通常由BedGraph文件转换而来。</p>
<p>在R语言中可以通过<code>rtracklayer</code>的<code>export.bw</code>输入和输出BigWig文件。</p>
<p>默认情况下，我们导入的数据集是GRanges格式</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">test_path <span class="operator">&lt;-</span> system.file<span class="punctuation">(</span><span class="string">&quot;tests&quot;</span><span class="punctuation">,</span> package <span class="operator">=</span> <span class="string">&quot;rtracklayer&quot;</span><span class="punctuation">)</span></span><br><span class="line">test_bw <span class="operator">&lt;-</span> file.path<span class="punctuation">(</span>test_path<span class="punctuation">,</span> <span class="string">&quot;test.bw&quot;</span><span class="punctuation">)</span></span><br><span class="line">gr <span class="operator">&lt;-</span> import<span class="punctuation">(</span>test_bw<span class="punctuation">)</span></span><br><span class="line">gr </span><br></pre></td></tr></table></figure>

<p>输出信息如下</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">gr</span><br><span class="line">GRanges object with <span class="number">9</span> ranges and <span class="number">1</span> metadata column<span class="operator">:</span></span><br><span class="line">      seqnames    ranges strand <span class="operator">|</span>     score</span><br><span class="line">         <span class="operator">&lt;</span>Rle<span class="operator">&gt;</span> <span class="operator">&lt;</span>IRanges<span class="operator">&gt;</span>  <span class="operator">&lt;</span>Rle<span class="operator">&gt;</span> <span class="operator">|</span> <span class="operator">&lt;</span>numeric<span class="operator">&gt;</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span>     chr2     <span class="number">1</span><span class="operator">-</span><span class="number">300</span>      <span class="operator">*</span> <span class="operator">|</span>        <span class="operator">-</span><span class="number">1</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span>     chr2   <span class="number">301</span><span class="operator">-</span><span class="number">600</span>      <span class="operator">*</span> <span class="operator">|</span>     <span class="operator">-</span><span class="number">0.75</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">3</span><span class="punctuation">]</span>     chr2   <span class="number">601</span><span class="operator">-</span><span class="number">900</span>      <span class="operator">*</span> <span class="operator">|</span>      <span class="operator">-</span><span class="number">0.5</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">4</span><span class="punctuation">]</span>     chr2  <span class="number">901</span><span class="operator">-</span><span class="number">1200</span>      <span class="operator">*</span> <span class="operator">|</span>     <span class="operator">-</span><span class="number">0.25</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">5</span><span class="punctuation">]</span>     chr2 <span class="number">1201</span><span class="operator">-</span><span class="number">1500</span>      <span class="operator">*</span> <span class="operator">|</span>         <span class="number">0</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">6</span><span class="punctuation">]</span>    chr19 <span class="number">1501</span><span class="operator">-</span><span class="number">1800</span>      <span class="operator">*</span> <span class="operator">|</span>      <span class="number">0.25</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">7</span><span class="punctuation">]</span>    chr19 <span class="number">1801</span><span class="operator">-</span><span class="number">2100</span>      <span class="operator">*</span> <span class="operator">|</span>       <span class="number">0.5</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">8</span><span class="punctuation">]</span>    chr19 <span class="number">2101</span><span class="operator">-</span><span class="number">2400</span>      <span class="operator">*</span> <span class="operator">|</span>      <span class="number">0.75</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">9</span><span class="punctuation">]</span>    chr19 <span class="number">2401</span><span class="operator">-</span><span class="number">2700</span>      <span class="operator">*</span> <span class="operator">|</span>         <span class="number">1</span></span><br><span class="line">  <span class="operator">-</span><span class="operator">-</span><span class="operator">-</span><span class="operator">-</span><span class="operator">-</span><span class="operator">-</span><span class="operator">-</span></span><br><span class="line">  seqinfo<span class="operator">:</span> <span class="number">2</span> sequences from an unspecified genome</span><br></pre></td></tr></table></figure>

<p>我们可以通过设定<code>import</code>里的<code>as</code>参数，使其读取为RleList</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">rle <span class="operator">&lt;-</span> import<span class="punctuation">(</span>test_bw<span class="punctuation">,</span> as <span class="operator">=</span> <span class="string">&quot;RleList&quot;</span><span class="punctuation">)</span></span><br><span class="line">rle</span><br></pre></td></tr></table></figure>

<p>输出信息如下</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">RleList of <span class="built_in">length</span> <span class="number">2</span></span><br><span class="line"><span class="operator">$</span>chr2</span><br><span class="line">numeric<span class="operator">-</span>Rle of <span class="built_in">length</span> <span class="number">243199373</span> with <span class="number">5</span> runs</span><br><span class="line">  Lengths<span class="operator">:</span>       <span class="number">300</span>       <span class="number">300</span>       <span class="number">300</span>       <span class="number">300</span> <span class="number">243198173</span></span><br><span class="line">  Values <span class="operator">:</span>        <span class="operator">-</span><span class="number">1</span>     <span class="operator">-</span><span class="number">0.75</span>      <span class="operator">-</span><span class="number">0.5</span>     <span class="operator">-</span><span class="number">0.25</span>         <span class="number">0</span></span><br><span class="line"></span><br><span class="line"><span class="operator">$</span>chr19</span><br><span class="line">numeric<span class="operator">-</span>Rle of <span class="built_in">length</span> <span class="number">59128983</span> with <span class="number">6</span> runs</span><br><span class="line">  Lengths<span class="operator">:</span>     <span class="number">1500</span>      <span class="number">300</span>      <span class="number">300</span>      <span class="number">300</span>      <span class="number">300</span> <span class="number">59126283</span></span><br><span class="line">  Values <span class="operator">:</span>        <span class="number">0</span>     <span class="number">0.25</span>      <span class="number">0.5</span>     <span class="number">0.75</span>        <span class="number">1</span>        <span class="number">0</span></span><br></pre></td></tr></table></figure>

<p>如果要输出BigWig文件，也只要保证提供的对象是GRanges或RleList， 只不过要保证一定要有score列表示信号强度。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">export.bw<span class="punctuation">(</span>rle<span class="punctuation">,</span> <span class="string">&quot;test.bw&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>假如要用R语言输出下面这个GRanges对象为BigWig</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">gr <span class="operator">&lt;-</span> GRanges<span class="punctuation">(</span></span><br><span class="line">        seqnames<span class="operator">=</span>Rle<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;chr1&quot;</span><span class="punctuation">,</span> <span class="string">&quot;chr2&quot;</span><span class="punctuation">,</span> <span class="string">&quot;chr1&quot;</span><span class="punctuation">,</span> <span class="string">&quot;chr3&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">        ranges<span class="operator">=</span>IRanges<span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">10</span><span class="punctuation">,</span> end<span class="operator">=</span><span class="number">10</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">        strand<span class="operator">=</span>Rle<span class="punctuation">(</span>strand<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;-&quot;</span><span class="punctuation">,</span> <span class="string">&quot;+&quot;</span><span class="punctuation">,</span> <span class="string">&quot;*&quot;</span><span class="punctuation">,</span> <span class="string">&quot;+&quot;</span><span class="punctuation">,</span> <span class="string">&quot;-&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">        seqlengths<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span>chr1<span class="operator">=</span><span class="number">11</span><span class="punctuation">,</span> chr2<span class="operator">=</span><span class="number">12</span><span class="punctuation">,</span> chr3<span class="operator">=</span><span class="number">13</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>那么思路就是</p>
<ol>
<li>获取染色体的长度</li>
<li>将染色体按照一定宽度分窗, 可用函数为<code>tile</code>和<code>slidingWindows</code></li>
<li>统计每个窗口的read数, 需要用到函数<code>coverage</code>, <code>Views</code>和<code>ViewSums</code></li>
<li>输出BigWig</li>
</ol>
<p>代码实现:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 获取染色体长度</span></span><br><span class="line">chromSizes <span class="operator">&lt;-</span> GRanges<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;chr1&quot;</span><span class="punctuation">,</span><span class="string">&quot;chr2&quot;</span><span class="punctuation">,</span><span class="string">&quot;chr3&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> IRanges<span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">11</span><span class="punctuation">,</span><span class="number">12</span><span class="punctuation">,</span><span class="number">13</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 按照宽度2进行分窗</span></span><br><span class="line">windows <span class="operator">&lt;-</span> tile<span class="punctuation">(</span>chromSizes<span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">2</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 统计每个窗口的read数</span></span><br><span class="line">cvg <span class="operator">&lt;-</span> coverage<span class="punctuation">(</span>gr<span class="punctuation">)</span></span><br><span class="line">cov_by_wnd <span class="operator">&lt;-</span> Views<span class="punctuation">(</span>cvg<span class="punctuation">,</span> windows<span class="punctuation">)</span></span><br><span class="line">sum_by_wnd <span class="operator">&lt;-</span> viewSums<span class="punctuation">(</span>cov_by_wnd<span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span><span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>windows<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  mcols<span class="punctuation">(</span>windows<span class="punctuation">[[</span>i<span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="string">&#x27;score&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> sum_by_wnd<span class="punctuation">[[</span>i<span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"><span class="comment"># 输出</span></span><br><span class="line">seqlengths<span class="punctuation">(</span>windows<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">11</span><span class="punctuation">,</span><span class="number">12</span><span class="punctuation">,</span><span class="number">13</span><span class="punctuation">)</span></span><br><span class="line">export.bw<span class="punctuation">(</span>unlist<span class="punctuation">(</span>windows<span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;tmp.bw&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>


      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/13/%E7%94%A8rtracklayer%E8%AF%BB%E5%8F%96%E5%92%8C%E8%BE%93%E5%87%BABigWig-rtracklayer-BigWig/" data-id="cllg90y6t00k1eqp2013n5de6" data-title="用rtracklayer读取和输出BigWig" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-「文章重现」2019发表在NBT的10x sc-ATAC-seq分析重现-2019-NBT-10x-sc-ATAC-seq-analysis" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/10/%E3%80%8C%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0%E3%80%8D2019%E5%8F%91%E8%A1%A8%E5%9C%A8NBT%E7%9A%8410x%20sc-ATAC-seq%E5%88%86%E6%9E%90%E9%87%8D%E7%8E%B0-2019-NBT-10x-sc-ATAC-seq-analysis/" class="article-date">
  <time class="dt-published" datetime="2019-08-10T17:22:32.875Z" itemprop="datePublished">2019-08-10</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/10/%E3%80%8C%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0%E3%80%8D2019%E5%8F%91%E8%A1%A8%E5%9C%A8NBT%E7%9A%8410x%20sc-ATAC-seq%E5%88%86%E6%9E%90%E9%87%8D%E7%8E%B0-2019-NBT-10x-sc-ATAC-seq-analysis/">「文章重现」2019发表在NBT的10x sc-ATAC-seq分析重现</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>最近(2019.8)在NBT上发表了一篇单细胞ATAC-seq的文章，题为 “Massively parallel single-cell chromatin landscapes of human immune cell development and intratumoral T cell exhaustion”。这篇文章由ATAC-seq的发明者所在的实验室用了10X Genomic 公司开发的sc-ATAC-seq技术分析而来。</p>
<p>文章一共测了20w个细胞，GEO编号是<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE129785">GSE129785</a>, 里面是分析后的数据，而原始的FASTQ数据则在<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA532774">https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA532774</a>, 数据量400G。此外，原始数据和处理过的数据还可以从<a target="_blank" rel="noopener" href="https://support.10xgenomics.com/single-cell-atac/datasets">https://support.10xgenomics.com/single-cell-atac/datasets</a>进行下载。</p>
<p>我下载了其中66.7G的tar文件，解压后放在了百度网盘(链接：<a target="_blank" rel="noopener" href="https://pan.baidu.com/s/1fu4MOhhW3aFtInsiJ5zBWg">https://pan.baidu.com/s/1fu4MOhhW3aFtInsiJ5zBWg</a> 提取码：6j7d ).</p>
<p>最为更重要的是，文章提供了核心分析部分用到的代码，代码地址为：<a target="_blank" rel="noopener" href="https://github.com/GreenleafLab/10x-scATAC-2019">https://github.com/GreenleafLab/10x-scATAC-2019</a></p>
<p>因此，我们就可以根据文章的数据和代码来学习10X sc-ATAC-seq的数据分析方法, 我们的目标是尽量重复出文章里出现的图，</p>
<h2 id="文章数据"><a href="#文章数据" class="headerlink" title="文章数据"></a>文章数据</h2><p>这篇文章能发到NBT，我觉得很大一部分的原因是它用大量的数据证明了scATAC-seq的可靠性，GSE129785编号中里放了67个样本，其中的GSE129785_RAW.tar(里面的文件是处理后的fragments.tsv.gz)有66.7G，解压后之后都是Cell Ranger ATAC (v 1.1.0) 处理后的fragments.gz文件, 我根据文章内容分门别类介绍这些数据：</p>
<p><strong>第零类</strong>: 人类GM12878细胞系和小鼠A20细胞系(B lymphocytes)混合测序，通过滴定的方式来获取不同的细胞量，从而评估双&#x2F;多细胞率(原文计算的结果是约1%)。一共有4个样本，如下</p>
<p>BY_500, BY_1k, BY_5k, BY_10k</p>
<p><strong>第一类</strong>：人类PMBC中不同比例的不同类型细胞混合。作者的目的是为了分析多少细胞量会被鉴定出真实的类群，作者发现，百分之一或者千分之一的比例就可以鉴定出细胞类群，有14个样本</p>
<p>0p1_99p9_CD4Mem_CD8Naive, 0p1_99p9_Mono_T, 0p5_99p5_CD4Mem_CD8Naive, 0p5_99p5_Mono_T, 1_99_CD4Mem_CD8Naive, 1_99_Mono_T, 50_50_CD4Mem_CD8Naive, 50_50_Mono_T, 99_1_CD4Mem_CD8Naive, 99_1_MonoT, 99p5_0p5_CD4Mem_CD8Naive, 99p5_0p5_MonoT, 99p9_0p1_CD4Mem_CD8Naive, 99p9_0p1_Mono_T</p>
<p>其中，p是point的简写，0p1_99p9_CD4Mem_CD8Naive就是1 CD4Mem : 999 CD8Naive. 其中MonoT是作者手误，应该是Mono_T。文章里面只提到了monocytes 和 T cells，但是这里面的数据表明还有不同比例的CD细胞混合。</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/sup_fig2-0b1a68fbd8b54b93abdf70b956282c3e.jpg" alt="sup_fig2"></p>
<p><strong>第二类</strong>：fresh PBMC和viably frozen PBMC，后者分为分选和未分选。作者的目的是为了确定即便是冻存的细胞也是能够做sc-ATAC-seq。有3个样本</p>
<p>Fresh_pbmc_5k, Frozen_sorted_pbmc_5k, Frozen_unsorted_pbmc_5k。</p>
<p><strong>第三类</strong>：16个人健康人的peripheral blood, 4个重复</p>
<p>PBMC_Rep1, PBMC_Rep2, PBMC_Rep3, PBMC_Rep4</p>
<p><strong>第四类</strong>：16个人健康人的 bone marrow cells, 就1个样本</p>
<p>Bone_Marrow_Rep1</p>
<p><strong>第五类</strong>：16个人健康人的富集特定表面标记的细胞, </p>
<p>Dendritic_Cells, Monocytes, B_Cells, CD34_Progenitors_Rep1, Regulatory_T_Cells, Naive_CD4_T_Cells_Rep1, Memory_CD4_T_Cells_Rep1, CD4_HelperT, CD4_Memory, CD4_Naive, NK_Cells, Naive_CD8_T_Cells, Memory_CD8_T_Cells, Dendritic_all_cells, CD34_Progenitors_Rep2, Memory_CD4_T_Cells_Rep2, Naive_CD4_T_Cells_Rep2</p>
<p>第三，四，五类其实都是健康人的免疫细胞, 文章过滤后有61,806个细胞, 聚类结果见附录sup fig. 3a</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/sup-fig3a-76590ea376ac4366972e08c88ab290b4.jpg" alt="sup-fig3a"></p>
<p><strong>第六类</strong>：免疫疗法前后的细胞，<strong>site-matched</strong> serial tumor biopsies <strong>pre-</strong> and <strong>post</strong>-PD-1 blockade (pembrolizumab) from five patients, plus post-therapy biopsies from two additional patients, 一共有21个样本，过滤后有37,818个细胞</p>
<ul>
<li>SU001_Immune_Post2, SU001_Tcell_Post2, SU001_Tcell_Post, SU001_Total_Post2, SU001_Total_Pre, SU001_Tumor_Immune_Post</li>
<li>SU006_Immune_Pre, SU006_Tcell_Pre, SU006_Total_Post, SU006_Tumor_Pre</li>
<li>SU008_Immune_Post, SU008_Immune_Pre, SU008_Tcell_Post, SU008_Tcell_Pre, SU008_Tumor_Post, SU008_Tumor_Pre</li>
<li>SU009_Tcell_Post, SU009_Tcell_Pre, SU009_Tumor_Immune_Post, SU009_Tumor_Immune_Pre</li>
<li>SU010_Total_Post, SU010_Total_Pre</li>
<li>SU005_Total_Post</li>
<li>SU007_Total_Post</li>
</ul>
<p><strong>注</strong>: 这里前5个病人还可以分类，一种是直接测Total，也就是unbiased fashion，另一种区分了Immune, Tcell, Tumor_Immune, 也就是cell sorting后分开测, T cells (CD45 + CD3 + ), non-T immune cells (CD45 + CD3 − ), stromal and tumor cells (CD45 − ), 参考附录sup fig. 6a</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/sup-fig6a-5b31ed3f70bd4d39b2e4e54bf3b6dd2b.jpg" alt="sup-fig6a"></p>
<p>对于这些数据，文章用的是下面这套流程来进行处理</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/sc-ATAC-seq-analysis-flow-73caa485a12a4ec5a8f76c842b5ab383.jpg" alt="sc-ATAC-seq-analysis-flow"></p>
<h2 id="数据预处理"><a href="#数据预处理" class="headerlink" title="数据预处理"></a>数据预处理</h2><p>文章的数据预处理交给了Cell Ranger ATAC (v 1.1.0) ，用的是hg19 作为参考基因组，并不是最新GRCh38。 </p>
<blockquote>
<p>经@Jimmy @王诗翔 @伊现富老师 解释，是因为hg19的惯性太大，”周边”太多, 整个生信软件生态没有一致性更新，所以很多分析只能用hg19做。</p>
</blockquote>
<p>Cell Ranger ATAC (v 1.1.0)  能做很多工作，</p>
<ul>
<li>Read过滤和比对(Read filtering and alignment)</li>
<li>Barcode记数(Barcode counting)</li>
<li>鉴定转座酶切割位点(Identification of transposase cut sites)</li>
<li>发现可及染色质峰(Detection of accessible chromatin peaks)</li>
<li>定义细胞(Cell calling)</li>
<li>peak和TF的count矩阵(Count matrix generation for peaks and transcription factors)</li>
<li>降维(Dimensionality reduction)</li>
<li>细胞聚类(Cell clustering)</li>
<li>聚类间差异可及性分析(Cluster differential accessibility)</li>
</ul>
<p>原文用了很长一段描述(见<strong>Data processing using Cell Ranger ATAC software</strong>)来说明了Cell Ranger ATAC的具体过程。当然我们不在乎过程，我们只在乎结果。而结果就是下面这些:</p>
<ul>
<li>singlecell.csv: 每个细胞的信息，如是否在TSS</li>
<li>possorted_bam.bam,possorted_bam.bam.bai: 处理后的BAM与其索引</li>
<li>raw_peak_bc_matrix.h5: 原始的peak-cell矩阵, h5格式存放</li>
<li>raw_peak_bc_matrix: 原始的peak-cell矩阵</li>
<li>analysis: 各种分析结果，如聚类，富集，降维等</li>
<li>filtered_peak_bc_matrix.h5: 过滤后的peak-cell矩阵, h5格式存放</li>
<li>filtered_peak_bc_matrix: 过滤后的peak-cell矩阵</li>
<li>fragments.tsv.gz, fragments.tsv.gz.tbi: 每个barcode的序列和它对应的基因组位置和数目</li>
<li>filtered_tf_bc_matrix.h5: 过滤后的TF-cell矩阵, h5格式存放</li>
<li>filtered_tf_bc_matrix: 过滤后的TF-cell矩阵</li>
<li>cloupe.cloupe: Loupe Cell Browser的输入文件</li>
<li>summary.csv, summary.json: 数据的统计结果, 以csv和json格式存放</li>
<li>web_summary.html: 网页总结信息</li>
<li>peaks.bed: 所有的peak汇总</li>
<li>peak_annotation.csv: peak的注释结果</li>
</ul>
<p>虽然结果很多，但是文章只用到了<strong>framgments.tsv.gz</strong>，就如同我们单细胞转录组分析也只要表达量矩阵。</p>
<h2 id="scATAC-seq-数据分析"><a href="#scATAC-seq-数据分析" class="headerlink" title="scATAC-seq 数据分析"></a>scATAC-seq 数据分析</h2><p>之后文章的方法部分就开始描述scATAC-seq的分析内容，一共有19项，描述多达3.5页。如果让作者在附录尽情发挥的话，那么每一项都会有1到2页内容，所以这将会一项非常持久的更新计划。</p>
<ul>
<li>Filtering cells by TSS enrichment and unique fragments</li>
<li>Generating a counts matrix</li>
<li>Generating union peak sets with LSI</li>
<li>Reads-in-peaks-normalized bigwigs and sequencing tracks</li>
<li>ATAC-seq-centric LSI clustering and visualization</li>
<li>Inferring copy number amplification</li>
<li>TF footprinting</li>
<li>ChromVAR</li>
<li>Computing gene activity scores using Cicero co-accessibility</li>
<li>Analysis of autoimmune variants using Cicero co-accessibility and chromVAR.</li>
<li>HiChIP meta-virtual 4C (metav4C) analysis for Cicero co-accessibility links</li>
<li>Overlap of Cicero co-accessibility links with GTEx eQTLs</li>
<li>Constructing ATAC-seq pseudo-bulk replicates of maximal variance</li>
<li>Constructing gene score pseudo-bulk replicates of maximal variance</li>
<li>Identification of cluster-specific peaks and gene scores through feature binarization</li>
<li>Pseudotime analysis</li>
<li>Barnyard mixing analysis</li>
<li>Analysis of fresh versus frozen PBMCs</li>
<li>Spike-in analysis</li>
</ul>
<h3 id="根据TSS富集和唯一片段过滤细胞"><a href="#根据TSS富集和唯一片段过滤细胞" class="headerlink" title="根据TSS富集和唯一片段过滤细胞"></a>根据TSS富集和唯一片段过滤细胞</h3><p>这一步用到的脚本是<code>01_Filter_Cells_v2.R</code>，读取的fragments为四列。前三列是fragment在染色体的位置，第四列是barcode信息，第五列则是该fragments被测了几次。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#chr	start	end	barcode	number</span></span><br><span class="line">chr1    10059   10114   TTATGTCAGGTTGTTC-1      7</span><br></pre></td></tr></table></figure>

<p>分析者先过滤不到那些fragments少于100的细胞。</p>
<p>之后，按照文章里提到过滤低质量细胞的标准进行过滤</p>
<ul>
<li>cut-offs of 1,000 unique nuclear fragments per cell </li>
<li>a transcription start site (TSS) enrichment score of 8</li>
</ul>
<p>每个细胞的唯一片段数计算比较简单也很好理解，而TSS enrichment score需要看他们之前在2018年发表在science文章的附录(Corces et. al 2018 science)</p>
<blockquote>
<p>To get the fragment length distribution, the width of each<br>fragment&#x2F;GRange was plotted. To get the TSS enrichment profile, each TSS from the R package “TxDb.Hsapiens.UCSC.hg38.knownGene” (accessed by transcripts(TxDb)) was extended 2000 bp in each direction and overlapped with the insertions (each end of a fragment) using “findOverlaps”. Next, the distance between the insertions and the strand-corrected TSS was calculated and the number of insertions occurring in each single-base bin was summed. To normalize this value to the local background, the accessibility at each position +&#x2F;- 2000 bp from the TSS was normalized to the mean of the accessibility at positions +&#x2F;-1900-2000 bp from the TSS. The final TSS enrichment reported was the maximum enrichment value within +&#x2F;- 50 bp of the TSS after smoothing with a rolling mean every 51 bp.</p>
</blockquote>
<p>这依赖于一个核心函数，<code>insertionProfileSingles</code>, 作用是分析每个细胞的插入谱</p>
<p>最后分析者为了避免潜在的doublets，还过滤了含有大于45,000的唯一fragmeng的细胞</p>
<h3 id="构建counts矩阵"><a href="#构建counts矩阵" class="headerlink" title="构建counts矩阵"></a>构建counts矩阵</h3><p>这部分描述包括后续的各种矩阵产生，包括cell-by-window, cell-by-peak, cell-by-tf。主要涉及<code>02_Get_Peak_Set_hg19_v2.R</code> 里的<code>countInsertions</code>函数。</p>
<p> <strong>第一步</strong>:获取 <strong>fragment GenomicRanges</strong>(数据结构如下)，也就是<code>01_Filter_Cells_v2.R</code>过滤后的结果</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">GRanges object with <span class="number">6</span> ranges and <span class="number">2</span> metadata columns<span class="operator">:</span></span><br><span class="line">      seqnames      ranges strand <span class="operator">|</span>                 RG         N</span><br><span class="line">         <span class="operator">&lt;</span>Rle<span class="operator">&gt;</span>   <span class="operator">&lt;</span>IRanges<span class="operator">&gt;</span>  <span class="operator">&lt;</span>Rle<span class="operator">&gt;</span> <span class="operator">|</span>        <span class="operator">&lt;</span>character<span class="operator">&gt;</span> <span class="operator">&lt;</span>numeric<span class="operator">&gt;</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span>     chr1 <span class="number">10060</span><span class="operator">-</span><span class="number">10114</span>      <span class="operator">*</span> <span class="operator">|</span> TTATGTCAGGTTGTTC<span class="operator">-</span><span class="number">1</span>         <span class="number">7</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span>     chr1 <span class="number">10060</span><span class="operator">-</span><span class="number">10126</span>      <span class="operator">*</span> <span class="operator">|</span> GAGGTCCCAGCGTCGT<span class="operator">-</span><span class="number">1</span>         <span class="number">4</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">3</span><span class="punctuation">]</span>     chr1 <span class="number">10061</span><span class="operator">-</span><span class="number">10126</span>      <span class="operator">*</span> <span class="operator">|</span> AACGTACAGTCTCTAG<span class="operator">-</span><span class="number">1</span>         <span class="number">5</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">4</span><span class="punctuation">]</span>     chr1 <span class="number">10061</span><span class="operator">-</span><span class="number">10529</span>      <span class="operator">*</span> <span class="operator">|</span> TAGCCGGAGGATGTAT<span class="operator">-</span><span class="number">1</span>         <span class="number">1</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">5</span><span class="punctuation">]</span>     chr1 <span class="number">10062</span><span class="operator">-</span><span class="number">10114</span>      <span class="operator">*</span> <span class="operator">|</span> TGAATCGTCTGACTTT<span class="operator">-</span><span class="number">1</span>        <span class="number">14</span></span><br><span class="line">  <span class="punctuation">[</span><span class="number">6</span><span class="punctuation">]</span>     chr1 <span class="number">10062</span><span class="operator">-</span><span class="number">10126</span>      <span class="operator">*</span> <span class="operator">|</span> CCGTAGGGTCACAGTT<span class="operator">-</span><span class="number">1</span>        <span class="number">13</span></span><br></pre></td></tr></table></figure>

<p><strong>第二步</strong>: 然后通过将fragments的起始位置和结束位置进行合并, 就得到了Tn5 insertion GenomicRange</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">by <span class="operator">&lt;-</span> <span class="string">&quot;RG&quot;</span></span><br><span class="line">inserts <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    GRanges<span class="punctuation">(</span>seqnames <span class="operator">=</span> seqnames<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">,</span> ranges <span class="operator">=</span> IRanges<span class="punctuation">(</span>start<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">,</span> start<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> RG <span class="operator">=</span> mcols<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span>by<span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    GRanges<span class="punctuation">(</span>seqnames <span class="operator">=</span> seqnames<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">,</span> ranges <span class="operator">=</span> IRanges<span class="punctuation">(</span>end<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">,</span> end<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> RG <span class="operator">=</span> mcols<span class="punctuation">(</span>fragments<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span>by<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><strong>第三步</strong>: 用<code>findOverlaps</code>找和insertion重叠的feature(如window, TSS, peak, TF)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">overlapDF <span class="operator">&lt;-</span> DataFrame<span class="punctuation">(</span>findOverlaps<span class="punctuation">(</span>query<span class="punctuation">,</span> </span><br><span class="line">                                    inserts<span class="punctuation">,</span> </span><br><span class="line">                                    ignore.strand <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> </span><br><span class="line">                                    maxgap<span class="operator">=</span><span class="operator">-</span><span class="number">1L</span><span class="punctuation">,</span> minoverlap<span class="operator">=</span><span class="number">0L</span><span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&quot;any&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><strong>第四步</strong>: 增加新的一列(id)到重叠输出对象中，这一列的内容是细胞barcode的ID，然后转成<code>sparseMatrix</code>提高存储和计算效率</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">overlapDF<span class="operator">$</span>name <span class="operator">&lt;-</span> mcols<span class="punctuation">(</span>inserts<span class="punctuation">)</span><span class="punctuation">[</span>overlapDF<span class="punctuation">[</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">]</span><span class="punctuation">,</span> by<span class="punctuation">]</span></span><br><span class="line"><span class="comment"># transform类似于dplyr:mutate</span></span><br><span class="line">overlapTDF <span class="operator">&lt;-</span> transform<span class="punctuation">(</span>overlapDF<span class="punctuation">,</span> id <span class="operator">=</span> match<span class="punctuation">(</span>name<span class="punctuation">,</span> unique<span class="punctuation">(</span>name<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 构建稀疏矩阵对象</span></span><br><span class="line">sparseM <span class="operator">&lt;-</span> Matrix<span class="operator">::</span>sparseMatrix<span class="punctuation">(</span></span><br><span class="line">    i <span class="operator">=</span> overlapTDF<span class="punctuation">[</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span> </span><br><span class="line">    j <span class="operator">=</span> overlapTDF<span class="punctuation">[</span><span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">    x <span class="operator">=</span> <span class="built_in">rep</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> nrow<span class="punctuation">(</span>overlapTDF<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">    dims <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">length</span><span class="punctuation">(</span>query<span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">(</span>unique<span class="punctuation">(</span>overlapDF<span class="operator">$</span>name<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>sparseM<span class="punctuation">)</span> <span class="operator">&lt;-</span> unique<span class="punctuation">(</span>overlapDF<span class="operator">$</span>name<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><strong>第五步</strong>: Tn5插入数在peak的比例 (frip)计算方法为: 稀疏矩阵的列和和每个细胞的fragment数相除</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">inPeaks <span class="operator">&lt;-</span> table<span class="punctuation">(</span>overlapDF<span class="operator">$</span>name<span class="punctuation">)</span></span><br><span class="line">total <span class="operator">&lt;-</span> table<span class="punctuation">(</span>mcols<span class="punctuation">(</span>inserts<span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span> by<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">total <span class="operator">&lt;-</span> total<span class="punctuation">[</span><span class="built_in">names</span><span class="punctuation">(</span>inPeaks<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">frip <span class="operator">&lt;-</span> inPeaks <span class="operator">/</span> total</span><br></pre></td></tr></table></figure>

<p><strong>第六步</strong>(可选): counts矩阵用edgeR的<code>cpm(matrix, log = TRUE, prior.count = 3)</code>进行log标准化。这个矩阵用于降低那些低count值元素对变异的影响。该标准化<strong>假设</strong>不同细胞类型中所有染色质开放区的差异是比较小的。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">logMat <span class="operator">&lt;-</span> edgeR<span class="operator">::</span>cpm<span class="punctuation">(</span>clusterSums<span class="punctuation">,</span> <span class="built_in">log</span> <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> prior.count <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="利用LSI生成peak合集"><a href="#利用LSI生成peak合集" class="headerlink" title="利用LSI生成peak合集"></a>利用LSI生成peak合集</h3><p>方法参考自Cusanovich, D. A. et al (Cell 2018), 代码见<code>02_Get_Peak_Set_hg19_v2.R</code>, 步骤如下</p>
<p><strong>第一步</strong>: 利用tile函数，沿着基因组构建2.5kb的window</p>
 <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">genome <span class="operator">&lt;-</span> BSgenome.Hsapiens.UCSC.hg19</span><br><span class="line">chromSizes <span class="operator">&lt;-</span> GRanges<span class="punctuation">(</span><span class="built_in">names</span><span class="punctuation">(</span>seqlengths<span class="punctuation">(</span>genome<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> IRanges<span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> seqlengths<span class="punctuation">(</span>genome<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">chromSizes <span class="operator">&lt;-</span> GenomeInfoDb<span class="operator">::</span>keepStandardChromosomes<span class="punctuation">(</span>chromSizes<span class="punctuation">,</span> pruning.mode <span class="operator">=</span> <span class="string">&quot;coarse&quot;</span><span class="punctuation">)</span></span><br><span class="line">windows <span class="operator">&lt;-</span> unlist<span class="punctuation">(</span>tile<span class="punctuation">(</span>chromSizes<span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">2500</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><strong>第二步</strong>: 利用<code>countInsertions</code>基于windows和fragments获取cell-by-window矩阵。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># lapply遍历所有的fragments, 然后合并</span></span><br><span class="line">fragmentFiles <span class="operator">&lt;-</span> list.files<span class="punctuation">(</span><span class="string">&quot;data&quot;</span><span class="punctuation">,</span> pattern <span class="operator">=</span> <span class="string">&quot;.rds&quot;</span><span class="punctuation">,</span> full.names <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">countsList <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span><span class="built_in">seq_along</span><span class="punctuation">(</span>fragmentFiles<span class="punctuation">)</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>i<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">	message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;%s of %s&quot;</span><span class="punctuation">,</span> i<span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">(</span>fragmentFiles<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">	counts <span class="operator">&lt;-</span> countInsertions<span class="punctuation">(</span>windows<span class="punctuation">,</span> readRDS<span class="punctuation">(</span>fragmentFiles<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span> by <span class="operator">=</span> <span class="string">&quot;RG&quot;</span><span class="punctuation">)</span><span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">  counts</span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line">mat <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>countsList<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> x<span class="punctuation">)</span> <span class="operator">%&gt;%</span> Reduce<span class="punctuation">(</span><span class="string">&quot;cbind&quot;</span><span class="punctuation">,</span>.<span class="punctuation">)</span></span><br><span class="line">remove<span class="punctuation">(</span>countsList<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p> <strong>第三步</strong>: 矩阵二项化，并且只要所有细胞中前20,000最开放位点</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># seuratLSI</span></span><br><span class="line">mat<span class="operator">@</span>x<span class="punctuation">[</span>mat<span class="operator">@</span>x <span class="operator">&gt;</span> <span class="number">0</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="number">1</span> </span><br><span class="line">mat <span class="operator">&lt;-</span> mat<span class="punctuation">[</span>head<span class="punctuation">(</span>order<span class="punctuation">(</span>Matrix<span class="operator">::</span>rowSums<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">,</span>decreasing <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">,</span>nFeatures<span class="punctuation">)</span><span class="punctuation">,</span><span class="punctuation">]</span> </span><br></pre></td></tr></table></figure>

<p><strong>第四步</strong>: 利用TF-IDF(term frequency-inverse document frequency)降维</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 计算单个细胞中每个开放位点的频率</span></span><br><span class="line">freqs <span class="operator">&lt;-</span> t<span class="punctuation">(</span>t<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="operator">/</span>Matrix<span class="operator">::</span>colSums<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 计算逆文档频率</span></span><br><span class="line">idf   <span class="operator">&lt;-</span> as<span class="punctuation">(</span><span class="built_in">log</span><span class="punctuation">(</span><span class="number">1</span> <span class="operator">+</span> ncol<span class="punctuation">(</span>mat<span class="punctuation">)</span> <span class="operator">/</span> Matrix<span class="operator">::</span>rowSums<span class="punctuation">(</span>mat<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;sparseVector&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 计算TF-IDF = TF * IDF</span></span><br><span class="line">tfidf <span class="operator">&lt;-</span> as<span class="punctuation">(</span>Matrix<span class="operator">::</span>Diagonal<span class="punctuation">(</span>x<span class="operator">=</span>as.vector<span class="punctuation">(</span>idf<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;sparseMatrix&quot;</span><span class="punctuation">)</span> <span class="operator">%*%</span> freqs</span><br></pre></td></tr></table></figure>

<blockquote>
<p>TF-IDF的思想: 如果某个词比较少见，但是它在这篇文章中多次出现，那么它很可能就反映了这篇文章的特性，正是我们所需要的关键词 – <a target="_blank" rel="noopener" href="http://www.ruanyifeng.com/blog/2013/03/tf-idf.html">http://www.ruanyifeng.com/blog/2013/03/tf-idf.html</a></p>
</blockquote>
<p><strong>第五步</strong>: 利用R语言中的irlba对标准化TF-IDF进行奇异值分解(SVD) , 保留第2-25个维度(第一个维度通常和cell read depth高度相关)作为Seurat的输入</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># SVD降维</span></span><br><span class="line">svd <span class="operator">&lt;-</span> irlba<span class="operator">::</span>irlba<span class="punctuation">(</span>tfidf<span class="punctuation">,</span> nComponents<span class="punctuation">,</span> nComponents<span class="punctuation">)</span></span><br><span class="line">svdDiag <span class="operator">&lt;-</span> matrix<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> nrow<span class="operator">=</span>nComponents<span class="punctuation">,</span> ncol<span class="operator">=</span>nComponents<span class="punctuation">)</span></span><br><span class="line">diag<span class="punctuation">(</span>svdDiag<span class="punctuation">)</span> <span class="operator">&lt;-</span> svd<span class="operator">$</span>d</span><br><span class="line">matSVD <span class="operator">&lt;-</span> t<span class="punctuation">(</span>svdDiag <span class="operator">%*%</span> t<span class="punctuation">(</span>svd<span class="operator">$</span>v<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">rownames<span class="punctuation">(</span>matSVD<span class="punctuation">)</span> <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>mat<span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>matSVD<span class="punctuation">)</span> <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span><span class="string">&quot;PC&quot;</span><span class="punctuation">,</span><span class="built_in">seq_len</span><span class="punctuation">(</span>ncol<span class="punctuation">(</span>matSVD<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 构建Seurat对象</span></span><br><span class="line">obj <span class="operator">&lt;-</span> CreateSeuratObject<span class="punctuation">(</span>mat<span class="punctuation">,</span> project<span class="operator">=</span><span class="string">&#x27;scATAC&#x27;</span><span class="punctuation">,</span> min.cells<span class="operator">=</span><span class="number">0</span><span class="punctuation">,</span> min.genes<span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span></span><br><span class="line">obj <span class="operator">&lt;-</span> SetDimReduction<span class="punctuation">(</span>object <span class="operator">=</span> obj<span class="punctuation">,</span> reduction.type <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> slot <span class="operator">=</span> <span class="string">&quot;cell.embeddings&quot;</span><span class="punctuation">,</span> new.data <span class="operator">=</span> matSVD<span class="punctuation">)</span></span><br><span class="line">obj <span class="operator">&lt;-</span> SetDimReduction<span class="punctuation">(</span>object <span class="operator">=</span> obj<span class="punctuation">,</span> reduction.type <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> slot <span class="operator">=</span> <span class="string">&quot;key&quot;</span><span class="punctuation">,</span> new.data <span class="operator">=</span> <span class="string">&quot;PC&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>第三，四，五步被整合到一个函数<code>seuratLSI</code>中， 文章用的是Seurat V2.3</p>
<p><strong>第六步</strong>: 并用<code>FindClusters</code>进行SNN图聚类(默认0.8分辨率),  如果最小的细胞类群细胞数不够200，降低分辨率重新聚类， 一个函数<code>addClusters</code>实现。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">addClusters <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>obj<span class="punctuation">,</span> minGroupSize <span class="operator">=</span> <span class="number">50</span><span class="punctuation">,</span> dims.use <span class="operator">=</span> <span class="built_in">seq_len</span><span class="punctuation">(</span><span class="number">50</span><span class="punctuation">)</span><span class="punctuation">,</span> initialResolution <span class="operator">=</span> <span class="number">0.8</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  <span class="comment"># 初步聚类</span></span><br><span class="line">  currentResolution <span class="operator">&lt;-</span> initialResolution</span><br><span class="line">  obj <span class="operator">&lt;-</span> FindClusters<span class="punctuation">(</span>object <span class="operator">=</span> obj<span class="punctuation">,</span> reduction.type <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> dims.use <span class="operator">=</span> dims.use<span class="punctuation">,</span> resolution <span class="operator">=</span> currentResolution<span class="punctuation">,</span> print.output <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">  <span class="comment"># 获取最少聚类的细胞数和聚类数</span></span><br><span class="line">  minSize <span class="operator">&lt;-</span> <span class="built_in">min</span><span class="punctuation">(</span>table<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">[[</span>paste0<span class="punctuation">(</span><span class="string">&quot;res.&quot;</span><span class="punctuation">,</span>currentResolution<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  nClust <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>unique<span class="punctuation">(</span>paste0<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">[[</span>paste0<span class="punctuation">(</span><span class="string">&quot;res.&quot;</span><span class="punctuation">,</span>currentResolution<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;Current Resolution = %s, No of Clusters = %s, Minimum Cluster Size = %s&quot;</span><span class="punctuation">,</span> currentResolution<span class="punctuation">,</span> nClust<span class="punctuation">,</span> minSize<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  <span class="comment"># 判断最少聚类细胞数是否大于200，如果低于200，降低分辨率(0.8 × N)</span></span><br><span class="line">  <span class="keyword">while</span><span class="punctuation">(</span>minSize <span class="operator">&lt;=</span> minGroupSize<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    obj<span class="operator">@</span>meta.data <span class="operator">&lt;-</span> obj<span class="operator">@</span>meta.data<span class="punctuation">[</span><span class="punctuation">,</span><span class="operator">-</span>which<span class="punctuation">(</span>colnames<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">)</span><span class="operator">==</span>paste0<span class="punctuation">(</span><span class="string">&quot;res.&quot;</span><span class="punctuation">,</span>currentResolution<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">    currentResolution <span class="operator">&lt;-</span> currentResolution<span class="operator">*</span>initialResolution</span><br><span class="line">    obj <span class="operator">&lt;-</span> FindClusters<span class="punctuation">(</span>object <span class="operator">=</span> obj<span class="punctuation">,</span> reduction.type <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> dims.use <span class="operator">=</span> dims.use<span class="punctuation">,</span> resolution <span class="operator">=</span> currentResolution<span class="punctuation">,</span> print.output <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> force.recalc <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">    minSize <span class="operator">&lt;-</span> <span class="built_in">min</span><span class="punctuation">(</span>table<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">[[</span>paste0<span class="punctuation">(</span><span class="string">&quot;res.&quot;</span><span class="punctuation">,</span>currentResolution<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    nClust <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>unique<span class="punctuation">(</span>paste0<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">[[</span>paste0<span class="punctuation">(</span><span class="string">&quot;res.&quot;</span><span class="punctuation">,</span>currentResolution<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;Current Resolution = %s, No of Clusters = %s, Minimum Cluster Size = %s&quot;</span><span class="punctuation">,</span> currentResolution<span class="punctuation">,</span> nClust<span class="punctuation">,</span> minSize<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  <span class="built_in">return</span><span class="punctuation">(</span>obj<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p>需要<strong>注意</strong>，文章用的Seurat版本V2.3, 所以如果安装3.0版本, 上面的代码就失效了但是有解决方案，参考如下博客</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://divingintogeneticsandgenomics.rbind.io/post/clustering-scatacseq-data-the-tf-idf-way/">https://divingintogeneticsandgenomics.rbind.io/post/clustering-scatacseq-data-the-tf-idf-way/</a></li>
</ul>
<p><strong>第七步</strong>: 之后用MACS2对每个聚类里 Tn5-corrected singlebase insertions (each end of the Tn5-corrected fragments) 都进行peak calling，参数设置为<code>–shift -75–extsize 150–nomodel–callsummits–nolambda–keep-dup all -q 0.05</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 保存BED文件</span></span><br><span class="line">dirClusters <span class="operator">&lt;-</span> <span class="string">&quot;results/LSI-Cluster-Beds/&quot;</span></span><br><span class="line">dir.create<span class="punctuation">(</span>dirClusters<span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span><span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>fragmentFiles<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">	fragments <span class="operator">&lt;-</span>readRDS<span class="punctuation">(</span>fragmentFiles<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">	<span class="keyword">for</span><span class="punctuation">(</span>j <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">	  message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;%s of %s&quot;</span><span class="punctuation">,</span> j<span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">	  fragmentsj <span class="operator">&lt;-</span> fragments<span class="punctuation">[</span>fragments<span class="operator">$</span>RG <span class="operator">%in%</span> clusterResults<span class="punctuation">[[</span>j<span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">	  <span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">length</span><span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">	    out <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span></span><br><span class="line">	      chr <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span>seqnames<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span><span class="punctuation">,</span> seqnames<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">	      start <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">as.integer</span><span class="punctuation">(</span>start<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span> <span class="operator">-</span> <span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">as.integer</span><span class="punctuation">(</span>end<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span> <span class="operator">-</span> <span class="number">1</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">	      end <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">as.integer</span><span class="punctuation">(</span>start<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">as.integer</span><span class="punctuation">(</span>end<span class="punctuation">(</span>fragmentsj<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">	      <span class="punctuation">)</span> <span class="operator">%&gt;%</span> readr<span class="operator">::</span>write_tsv<span class="punctuation">(</span></span><br><span class="line">          x <span class="operator">=</span> .<span class="punctuation">,</span> </span><br><span class="line">          append <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> </span><br><span class="line">          path <span class="operator">=</span> paste0<span class="punctuation">(</span>dirClusters<span class="punctuation">,</span> paste0<span class="punctuation">(</span><span class="built_in">names</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">[</span>j<span class="punctuation">]</span><span class="punctuation">,</span> <span class="string">&quot;.bed&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">          col_names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">	    <span class="punctuation">&#125;</span></span><br><span class="line">	<span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"><span class="comment"># 运行MACS2</span></span><br><span class="line">dirPeaks <span class="operator">&lt;-</span> <span class="string">&quot;results/LSI-Cluster-Peaks/&quot;</span></span><br><span class="line">method <span class="operator">&lt;-</span> <span class="string">&quot;q&quot;</span></span><br><span class="line">cutoff <span class="operator">&lt;-</span> 0.05</span><br><span class="line">shift <span class="operator">&lt;-</span> <span class="operator">-</span><span class="number">75</span></span><br><span class="line">extsize <span class="operator">&lt;-</span> 150</span><br><span class="line">genome_size <span class="operator">&lt;-</span> 2.7e9</span><br><span class="line"><span class="keyword">for</span><span class="punctuation">(</span>j <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">	message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;%s of %s&quot;</span><span class="punctuation">,</span> j<span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">	clusterBedj <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span>dirClusters<span class="punctuation">,</span><span class="built_in">names</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">[</span>j<span class="punctuation">]</span><span class="punctuation">,</span><span class="string">&quot;.bed&quot;</span><span class="punctuation">)</span></span><br><span class="line">	cmdPeaks <span class="operator">&lt;-</span> sprintf<span class="punctuation">(</span></span><br><span class="line">	    <span class="string">&quot;macs2 callpeak -g %s --name %s --treatment %s --outdir %s --format BED --nomodel --call-summits --nolambda --keep-dup all&quot;</span><span class="punctuation">,</span> </span><br><span class="line">	    genome_size<span class="punctuation">,</span> </span><br><span class="line">	    <span class="built_in">names</span><span class="punctuation">(</span>clusterResults<span class="punctuation">)</span><span class="punctuation">[</span>j<span class="punctuation">]</span><span class="punctuation">,</span> </span><br><span class="line">	    clusterBedj<span class="punctuation">,</span> </span><br><span class="line">	    dirPeaks</span><br><span class="line">	  <span class="punctuation">)</span></span><br><span class="line">	<span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span><span class="built_in">is.null</span><span class="punctuation">(</span>shift<span class="punctuation">)</span> <span class="operator">&amp;</span> <span class="operator">!</span><span class="built_in">is.null</span><span class="punctuation">(</span>extsize<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">	  cmdPeaks <span class="operator">&lt;-</span> sprintf<span class="punctuation">(</span><span class="string">&quot;%s --shift %s --extsize %s&quot;</span><span class="punctuation">,</span> cmdPeaks<span class="punctuation">,</span> shift<span class="punctuation">,</span> extsize<span class="punctuation">)</span></span><br><span class="line">	<span class="punctuation">&#125;</span></span><br><span class="line">	<span class="keyword">if</span> <span class="punctuation">(</span>tolower<span class="punctuation">(</span>method<span class="punctuation">)</span> <span class="operator">==</span> <span class="string">&quot;p&quot;</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">	  cmdPeaks <span class="operator">&lt;-</span> sprintf<span class="punctuation">(</span><span class="string">&quot;%s -p %s&quot;</span><span class="punctuation">,</span> cmdPeaks<span class="punctuation">,</span> cutoff<span class="punctuation">)</span></span><br><span class="line">	<span class="punctuation">&#125;</span><span class="keyword">else</span> <span class="punctuation">&#123;</span></span><br><span class="line">	  cmdPeaks <span class="operator">&lt;-</span> sprintf<span class="punctuation">(</span><span class="string">&quot;%s -q %s&quot;</span><span class="punctuation">,</span> cmdPeaks<span class="punctuation">,</span> cutoff<span class="punctuation">)</span></span><br><span class="line">	<span class="punctuation">&#125;</span></span><br><span class="line">	message<span class="punctuation">(</span><span class="string">&quot;Running Macs2...&quot;</span><span class="punctuation">)</span></span><br><span class="line">	message<span class="punctuation">(</span>cmdPeaks<span class="punctuation">)</span></span><br><span class="line">	system<span class="punctuation">(</span>cmdPeaks<span class="punctuation">,</span> intern <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>之后读取peak summits文件，前后延伸250 bp, 最终宽度为501bp， 并基于ENCODE hg19 blacklist(<a target="_blank" rel="noopener" href="https://www.encodeproject.org/annotations/ENCSR636HFF/)%E8%BF%87%E6%BB%A4%EF%BC%8C%E4%BB%A5%E5%8F%8A%E9%82%A3%E4%BA%9B%E5%BB%B6%E4%BC%B8%E5%90%8E%E4%BC%9A%E8%B6%85%E5%87%BA%E6%9F%93%E8%89%B2%E4%BD%93%E7%9A%84peak%E3%80%82">https://www.encodeproject.org/annotations/ENCSR636HFF/)过滤，以及那些延伸后会超出染色体的peak。</a></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 读取</span></span><br><span class="line">df <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span></span><br><span class="line">  samples <span class="operator">=</span> gsub<span class="punctuation">(</span><span class="string">&quot;\\_summits.bed&quot;</span><span class="punctuation">,</span><span class="string">&quot;&quot;</span><span class="punctuation">,</span>list.files<span class="punctuation">(</span>dirPeaks<span class="punctuation">,</span> pattern <span class="operator">=</span> <span class="string">&quot;\\_summits.bed&quot;</span><span class="punctuation">,</span> full.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  groups <span class="operator">=</span> <span class="string">&quot;scATAC&quot;</span><span class="punctuation">,</span></span><br><span class="line">  summits <span class="operator">=</span> list.files<span class="punctuation">(</span>dirPeaks<span class="punctuation">,</span> pattern <span class="operator">=</span> <span class="string">&quot;\\_summits.bed&quot;</span><span class="punctuation">,</span> full.names <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 合并peak, 延伸, 过滤</span></span><br><span class="line">unionPeaks <span class="operator">&lt;-</span> extendedPeakSet<span class="punctuation">(</span></span><br><span class="line">    df <span class="operator">=</span> df<span class="punctuation">,</span></span><br><span class="line">    BSgenome <span class="operator">=</span> genome<span class="punctuation">,</span> </span><br><span class="line">    extend <span class="operator">=</span> <span class="number">250</span><span class="punctuation">,</span></span><br><span class="line">    blacklist <span class="operator">=</span> <span class="string">&quot;data/hg19.blacklist.bed&quot;</span><span class="punctuation">,</span></span><br><span class="line">    nSummits <span class="operator">=</span> <span class="number">200000</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 过滤Y染色体和线粒体</span></span><br><span class="line">unionPeaks <span class="operator">&lt;-</span> unionPeaks<span class="punctuation">[</span>seqnames<span class="punctuation">(</span>unionPeaks<span class="punctuation">)</span> <span class="operator">%in%</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">22</span><span class="punctuation">,</span><span class="string">&quot;X&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">unionPeaks <span class="operator">&lt;-</span> keepSeqlevels<span class="punctuation">(</span>unionPeaks<span class="punctuation">,</span> paste0<span class="punctuation">(</span><span class="string">&quot;chr&quot;</span><span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="operator">:</span><span class="number">22</span><span class="punctuation">,</span><span class="string">&quot;X&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>其中extendPeakSet里的函数<code>nonOverlappingGRanges</code> 会对单个样本中相互重叠的peak进行合并处理。步骤如下</p>
<ul>
<li>如果存在有直接重叠的peak，就只保留最显著的peak，过滤不怎么显著的peak</li>
<li>迭代上一步，处理下一个重叠的peak，直到这类直接性重叠的peak要么都保留，要么都移除。这一步用于处理每个聚类的peak集，保留前200,000延伸的summit, 得到了每个cluster的’cluster-specific’ peak集</li>
<li>接着对每个样本的MACS2 peak score进行标准化，即<code>−log10(Q value)</code>, 然后通过将每个个体的score转成中位数<code>trunc(rank(v))/length(v)’</code>, 其中v表示MACS2 peak scores 向量, 等到”score quantitle”。标准化之后peak就能够在聚类之间进行相互比较，能够产生每个数据集的peak合集</li>
<li>合并所有的cluster的peak，使用之前的迭代方法处理重叠，得到了最显著的peak，抛弃了不怎么显著的peak</li>
<li>接着将那些和基因组gap区(也就是序列是N)重叠peak移掉，以及Y染色体上的peak</li>
</ul>
<p><strong>第八步</strong>: 构建cell-by-peak矩阵.  用到的是<code>countInsertions</code>函数，只不过第一个参数改为了unionPeaks.</p>
 <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#Create Counts list</span></span><br><span class="line">countsPeaksList <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span><span class="built_in">seq_along</span><span class="punctuation">(</span>fragmentFiles<span class="punctuation">)</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>i<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  message<span class="punctuation">(</span>sprintf<span class="punctuation">(</span><span class="string">&quot;%s of %s&quot;</span><span class="punctuation">,</span> i<span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">(</span>fragmentFiles<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  gc<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">  countInsertions<span class="punctuation">(</span>unionPeaks<span class="punctuation">,</span> readRDS<span class="punctuation">(</span>fragmentFiles<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span> by <span class="operator">=</span> <span class="string">&quot;RG&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>做一些总结性分析</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#CountsMatrix</span></span><br><span class="line">mat <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>countsPeaksList<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> x<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> Reduce<span class="punctuation">(</span><span class="string">&quot;cbind&quot;</span><span class="punctuation">,</span>.<span class="punctuation">)</span></span><br><span class="line">frip <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>countsPeaksList<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> x<span class="punctuation">[[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> unlist</span><br><span class="line">total <span class="operator">&lt;-</span> lapply<span class="punctuation">(</span>countsPeaksList<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> x<span class="punctuation">[[</span><span class="number">3</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> unlist</span><br><span class="line"></span><br><span class="line">se <span class="operator">&lt;-</span> SummarizedExperiment<span class="punctuation">(</span></span><br><span class="line">  assays <span class="operator">=</span> SimpleList<span class="punctuation">(</span>counts <span class="operator">=</span> mat<span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">  rowRanges <span class="operator">=</span> unionPeaks</span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line">rownames<span class="punctuation">(</span>se<span class="punctuation">)</span> <span class="operator">&lt;-</span> paste<span class="punctuation">(</span>seqnames<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">,</span>start<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">,</span>end<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;_&quot;</span><span class="punctuation">)</span></span><br><span class="line">colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="operator">$</span>FRIP <span class="operator">&lt;-</span> frip</span><br><span class="line">colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="operator">$</span>uniqueFrags <span class="operator">&lt;-</span> total <span class="operator">/</span> <span class="number">2</span></span><br></pre></td></tr></table></figure>


<h3 id="Reads-in-peaks-normalized-bigwigs-and-sequencing-tracks"><a href="#Reads-in-peaks-normalized-bigwigs-and-sequencing-tracks" class="headerlink" title="Reads-in-peaks-normalized bigwigs and sequencing tracks."></a>Reads-in-peaks-normalized bigwigs and sequencing tracks.</h3><p>这部分内容在GitHub上没有代码，所以我根据文章描述编写代码实现</p>
<p>…</p>
<h3 id="ATAC-seq-centric-LSI-聚类和可视化"><a href="#ATAC-seq-centric-LSI-聚类和可视化" class="headerlink" title="ATAC-seq-centric LSI 聚类和可视化"></a>ATAC-seq-centric LSI 聚类和可视化</h3><p>以Cusanovich et. al (Nature 2018) 提出的策略计算TF-IDF转换, 然后irlba进行奇异值分解(SVD), 接着用前50维作为Seurat的输入，实现SNN图聚类，然后用<code>FindCluster</code>寻找聚类。 你会发现这个内容和上一节类似，只不过上一节是基于window, 这一节是基于peak。</p>
 <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># TF-IDF, SVD </span></span><br><span class="line">obj <span class="operator">&lt;-</span> seuratLSI<span class="punctuation">(</span>assay<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">,</span> nComponents <span class="operator">=</span> <span class="built_in">max</span><span class="punctuation">(</span>nPCs1<span class="punctuation">)</span><span class="punctuation">,</span> nFeatures <span class="operator">=</span> <span class="literal">NULL</span><span class="punctuation">)</span></span><br><span class="line">obj<span class="operator">@</span>meta.data <span class="operator">&lt;-</span> as.data.frame<span class="punctuation">(</span>cbind<span class="punctuation">(</span>obj<span class="operator">@</span>meta.data<span class="punctuation">,</span> colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 聚类分析</span></span><br><span class="line">obj <span class="operator">&lt;-</span> FindClusters<span class="punctuation">(</span>object <span class="operator">=</span> obj<span class="punctuation">,</span> dims.use <span class="operator">=</span> nPCs1<span class="punctuation">,</span> print.output <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> n.start <span class="operator">=</span> <span class="number">10</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>分析者发现这里存在能被检测到的批次效应，它会混淆后续的分析。因此为了减缓影响，他们从二项化后到可及矩阵中计算每个聚类的和，然后对edgeR的<code>cpm(matrix, log = TRUE, prior.count = 3)</code>结果进行log标准化 </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">mat <span class="operator">&lt;-</span> assay<span class="punctuation">(</span>se<span class="punctuation">)</span></span><br><span class="line">mat<span class="operator">@</span>x<span class="punctuation">[</span>mat<span class="operator">@</span>x <span class="operator">&gt;</span> <span class="number">0</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="number">1</span></span><br><span class="line">clusterSums <span class="operator">&lt;-</span> groupSums<span class="punctuation">(</span>mat <span class="operator">=</span> mat<span class="punctuation">,</span> groups <span class="operator">=</span> paste0<span class="punctuation">(</span><span class="string">&quot;C&quot;</span><span class="punctuation">,</span>obj<span class="operator">@</span>meta.data<span class="operator">$</span>res.0.8<span class="punctuation">)</span><span class="punctuation">,</span> sparse <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">logMat <span class="operator">&lt;-</span> edgeR<span class="operator">::</span>cpm<span class="punctuation">(</span>clusterSums<span class="punctuation">,</span> <span class="built_in">log</span> <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> prior.count <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>然后用<code>rowVars</code>挑选前25,000个波动peak，这是基于上一步标准化的矩阵 </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">varPeaks <span class="operator">&lt;-</span> head<span class="punctuation">(</span>order<span class="punctuation">(</span>matrixStats<span class="operator">::</span>rowVars<span class="punctuation">(</span>logMat<span class="punctuation">)</span><span class="punctuation">,</span> decreasing <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">,</span> nTop<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>为什么不用原来的稀疏二项矩阵，而用log标准化的CPM矩阵，原因有亮点</p>
<ul>
<li>能够降低不同聚类间因细胞数不同导致的误差</li>
<li>count转换到标准化空间后的均值-变异关系会减弱</li>
</ul>
<p>之后重新用25,000波动peak提取稀疏的二项化矩阵，然后重新做TF-IDF, SVD, SNN graph聚类等运算。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#Re-run Seurat LSI</span></span><br><span class="line">message<span class="punctuation">(</span><span class="string">&quot;Making Seurat LSI Object...&quot;</span><span class="punctuation">)</span></span><br><span class="line">obj2 <span class="operator">&lt;-</span> seuratLSI<span class="punctuation">(</span>assay<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">[</span>varPeaks<span class="punctuation">,</span><span class="punctuation">]</span><span class="punctuation">,</span> nComponents <span class="operator">=</span> <span class="built_in">max</span><span class="punctuation">(</span>nPCs2<span class="punctuation">)</span><span class="punctuation">,</span> nFeatures <span class="operator">=</span> <span class="literal">NULL</span><span class="punctuation">)</span></span><br><span class="line">stopifnot<span class="punctuation">(</span>identical<span class="punctuation">(</span>rownames<span class="punctuation">(</span>obj2<span class="operator">@</span>meta.data<span class="punctuation">)</span><span class="punctuation">,</span> colnames<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">obj2<span class="operator">@</span>meta.data <span class="operator">&lt;-</span> as.data.frame<span class="punctuation">(</span>cbind<span class="punctuation">(</span>obj2<span class="operator">@</span>meta.data<span class="punctuation">,</span> colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">message<span class="punctuation">(</span><span class="string">&quot;Adding Graph Clusters...&quot;</span><span class="punctuation">)</span></span><br><span class="line">obj2 <span class="operator">&lt;-</span> FindClusters<span class="punctuation">(</span>object <span class="operator">=</span> obj2<span class="punctuation">,</span> reduction.type <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> dims.use <span class="operator">=</span> nPCs2<span class="punctuation">,</span> print.output <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> n.start <span class="operator">=</span> <span class="number">10</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>用<code>Seurat::RunUMAP</code>函数进行可视化展示</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">message<span class="punctuation">(</span><span class="string">&quot;Running UMAP&quot;</span><span class="punctuation">)</span></span><br><span class="line">obj2 <span class="operator">&lt;-</span> RunUMAP<span class="punctuation">(</span>object <span class="operator">=</span> obj2<span class="punctuation">,</span> reduction.use <span class="operator">=</span> <span class="string">&quot;pca&quot;</span><span class="punctuation">,</span> dims.use <span class="operator">=</span> nPCs2<span class="punctuation">)</span></span><br><span class="line">plotUMAP <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>GetCellEmbeddings<span class="punctuation">(</span>obj2<span class="punctuation">,</span>reduction.type<span class="operator">=</span><span class="string">&quot;umap&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> obj2<span class="operator">@</span>meta.data<span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>plotUMAP<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;x&quot;</span><span class="punctuation">,</span><span class="string">&quot;y&quot;</span><span class="punctuation">,</span>colnames<span class="punctuation">(</span>plotUMAP<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">3</span><span class="operator">:</span>ncol<span class="punctuation">(</span>plotUMAP<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">clustCol <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>plotUMAP<span class="punctuation">)</span><span class="punctuation">[</span>grep<span class="punctuation">(</span><span class="string">&quot;res&quot;</span><span class="punctuation">,</span>colnames<span class="punctuation">(</span>plotUMAP<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="operator">$</span>Clusters <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span><span class="string">&quot;Cluster&quot;</span><span class="punctuation">,</span><span class="built_in">as.integer</span><span class="punctuation">(</span>plotUMAP<span class="punctuation">[</span><span class="punctuation">,</span>clustCol<span class="punctuation">]</span><span class="punctuation">)</span> <span class="operator">+</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="operator">$</span>UMAP1 <span class="operator">&lt;-</span> plotUMAP<span class="operator">$</span>x</span><br><span class="line">colData<span class="punctuation">(</span>se<span class="punctuation">)</span><span class="operator">$</span>UMAP2 <span class="operator">&lt;-</span> plotUMAP<span class="operator">$</span>y</span><br><span class="line"></span><br><span class="line">pdf<span class="punctuation">(</span><span class="string">&quot;results/LSI-Clustering-Peaks.pdf&quot;</span><span class="punctuation">)</span></span><br><span class="line">ggplot<span class="punctuation">(</span>plotUMAP<span class="punctuation">,</span> aes<span class="punctuation">(</span>x<span class="operator">=</span>x<span class="punctuation">,</span>y<span class="operator">=</span>y<span class="punctuation">,</span>color<span class="operator">=</span>res.0.8<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span> geom_point<span class="punctuation">(</span>size <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  theme_bw<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span> xlab<span class="punctuation">(</span><span class="string">&quot;UMAP1&quot;</span><span class="punctuation">)</span> <span class="operator">+</span> ylab<span class="punctuation">(</span><span class="string">&quot;UMAP2&quot;</span><span class="punctuation">)</span></span><br><span class="line">dev.off<span class="punctuation">(</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>对于亚群分析(hematopoiesis: CD34 + bone marrow 和 DCs; tumor: T cells), 分析者重新计算了聚类的和以及log标准化的CPM值，分别从CD34+ cells 和 T cells中鉴定出前10,00和5,00个波动peak。这些波动peak用于提取稀疏二项化可及矩阵，然后做TF-IDF转换。对TF-IDF转换结果进行奇异值分解(SVD),得到更低的维度, 即用前25个维度表征原来的数据集。分贝用1-25和2-25个维度作为Seurat的输入，进行SNN graph clustering（默认resolution&#x3D;0.8). 最后的结果用<code>Seurat::RunUMAP</code>函数进行可视化展示</p>
<h3 id="细胞类群定义"><a href="#细胞类群定义" class="headerlink" title="细胞类群定义"></a>细胞类群定义</h3><p>以human hematopoiesis为例，介绍作者定义细胞类群的三种方法。</p>
<p> (1) 顺式作用元件(cis-lements, 即ATAC-seq peaks)的染色质可及性</p>
<p> (2) 利用和单个基因启动子相关的几个增强子的可及性计算基因活跃得分(GA)</p>
<p> (3) TF活跃度，通过计算每个细胞中全基因组范围内的TF结合位点可及性得出</p>
<p>这三种方法最大的特点就是不需要你再去测一个单细胞RNA-seq或者要一个已有的Bulk ATAC-seq谱，仅仅用当前的单细胞ATAC-seq数据即可。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/10/%E3%80%8C%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0%E3%80%8D2019%E5%8F%91%E8%A1%A8%E5%9C%A8NBT%E7%9A%8410x%20sc-ATAC-seq%E5%88%86%E6%9E%90%E9%87%8D%E7%8E%B0-2019-NBT-10x-sc-ATAC-seq-analysis/" data-id="cllg90y2y0084eqp20i409zt1" data-title="「文章重现」2019发表在NBT的10x sc-ATAC-seq分析重现" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-「生信基础课」初学者入门Linux最少必要的知识点-「生信基础课」初学者入门Linux最少必要的知识点" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/08/07/%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9-%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9/" class="article-date">
  <time class="dt-published" datetime="2019-08-07T11:22:24.381Z" itemprop="datePublished">2019-08-07</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/Linux/">Linux</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/08/07/%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9-%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9/">「生信基础课」初学者入门Linux最少必要的知识点</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <blockquote>
<p>这是我在<a target="_blank" rel="noopener" href="https://space.bilibili.com/249108235">哔哩哔哩</a>上课程的文字稿版本。我有点完美主义的倾向，所以为了避免自己因为觉得自己某个部分的说话不利索，不清楚，所以总是会写稿子。<br><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/8/Linux%E5%85%A5%E9%97%A8%E8%A7%86%E9%A2%91%E7%B4%A0%E6%9D%90-e7a9613d7c074c9dbd0d20be3ecadd3b.jpg" alt="素材库"></p>
</blockquote>
<p>这次课程的主题是生信入门必须学习的几个Linux操作。</p>
<h2 id="为什么要学习Linux"><a href="#为什么要学习Linux" class="headerlink" title="为什么要学习Linux"></a>为什么要学习Linux</h2><p>如果你要学习生物信息学，那么你一定要学会使用Linux。这是因为，绝大部分的生信软件都需要运行在Linux平台，并且有些时候，你会要用到服务器去处理大量的数据，而这些服务器几乎都是Linux环境。</p>
<p>当然，你可以说，我可以用云平台呀，毕竟现在很多公司都开发了云平台，我们拖拖拽拽就可以搞分析了呀。是的，现在的云平台也来越成熟，使用体验也很棒，我也很喜欢。不过我还是建议大家学Linux，因为当你学会Linux之后，你会变得更加自由，不再受限于别人的平台，而且能让日常工作变得高效起来。</p>
<h2 id="如何学习Linux"><a href="#如何学习Linux" class="headerlink" title="如何学习Linux"></a>如何学习Linux</h2><p>那么如何学习呢？通常你问别人如何学Linux的时候，对方十有八九会给你推荐鸟哥的书，这没啥毛病，因为我也是靠鸟哥的书入门的，那本书已经被我翻成了这副模样。</p>
<p>学习没有捷径，但是可以少走几步路。我们不需要精通Linux，成为Linux大牛，靠Linux吃饭，我们只需要掌握一些基本操作，搞得定日常基本操作就行。就像绝大部分人用Windows的时候，也不知道如何配置Windows里的环境变量，但是一点不妨碍基本使用呀。</p>
<p>通过我对自己平时操作的总结，我认为学Linux起码要掌握下面这三方面内容。</p>
<ul>
<li>操作迁移，就是把Windows上的常见操作都在Linux里找到对应的操作</li>
<li>TAB补全，不但可以让你的操作更快，而且还能让你避免大部分的报错</li>
<li>环境变量，学习Linux一定要理解环境变量，否则你就装不好软件。</li>
</ul>
<p>不过，最重要的是，你得先有一个Linux环境。我在哔哩哔哩上录制了两个视频，教大家如何在Windows平台上配置你的Linux环境，如何在Mac上配置Linux环境。</p>
<p>先看Vim视频:</p>
<ul>
<li><p><a target="_blank" rel="noopener" href="https://www.bilibili.com/video/av58673256">https://www.bilibili.com/video/av58673256</a><br>  再配置Linux环境：</p>
</li>
<li><p>Windows10&#x2F;Linux: <a target="_blank" rel="noopener" href="https://www.bilibili.com/video/av58541835">https://www.bilibili.com/video/av58541835</a></p>
</li>
<li><p>MacOS: <a target="_blank" rel="noopener" href="https://www.bilibili.com/video/av58541835">https://www.bilibili.com/video/av58541835</a></p>
</li>
</ul>
<p>本次视频的操作虽然是在mac上录制，但是适用于大部分的Linux环境，包括Ubuntu和CentOS，如果有什么问题，欢迎在留言区提问交流。</p>
<h2 id="TAB补全"><a href="#TAB补全" class="headerlink" title="TAB补全"></a>TAB补全</h2><p>在演示Linux操作之前，我一定要先强调下tab补全的重要性，它是我认为最重要的一个操作。</p>
<p>初学者在学习Linux的时候，最常见的状态就是，按照教程内容，一个一个字符的敲。由于刚学习，敲代码还不熟练，因此，十有八九会出现敲错的情况，那么结果就是代码运行失败。运行失败怎么办？可能也不会去看报错，然后检查自己的输入，估计就是截图或者拍照发到群里提问了。</p>
<p>因此，在正式开始敲代码之前，大家先在键盘上找到tab，多敲几次，感受到它的存在。</p>
<p>后面，我还会不断强调它，直到它刻在你的脑子里。</p>
<h2 id="操作迁移"><a href="#操作迁移" class="headerlink" title="操作迁移"></a>操作迁移</h2><p>接下来，我们来讲讲这次课程的主要内容，<strong>操作迁移</strong>。</p>
<p>我们在Windows上最常做的操作是什么呢？我觉得，最常见的操作应该是下面这些吧</p>
<ul>
<li>浏览目录</li>
<li>切换目录</li>
<li>新建文件夹&#x2F;删除文件夹</li>
<li>新建文件&#x2F;删除文件</li>
<li>查看文本</li>
<li>了解资源使用情况</li>
<li>数据下载</li>
<li>安装软件</li>
</ul>
<p>那么这些操作在Linux里的对应命令是什么呢？我们先浏览一下命令，然后通过一个模拟的项目来一个个了解他们。</p>
<ul>
<li>浏览目录: ls</li>
<li>切换目录: cd, pwd</li>
<li>新建文件夹&#x2F;删除文件夹: mkdir &#x2F; rmdir &#x2F; rm -r</li>
<li>新建文件&#x2F;删除文件: vim &#x2F; rm</li>
<li>查看文本: less, head, tail, cat</li>
<li>了解资源使用情况: top</li>
<li>数据下载: wget &#x2F; curl</li>
<li>安装软件: apt-get&#x2F;yum&#x2F;conda</li>
</ul>
<p>那么这些命令应该怎么用呢？其实很简单，就是输入命令名，后面的参数，看情况加。强调一点，如果要用到参数，那么命令名和参数之间是需要空格分隔的。</p>
<p>我们以一个项目为例，去学习使用最常见的Linux命令。这个任务内容如下</p>
<p>1, 在家目录下<strong>创建</strong>文件夹，名为 abc</p>
<p>2, 将abc<strong>重命名</strong>为 study</p>
<p>3, 下载拟南芥的注释GFF文件， 解压缩</p>
<p>4, 使用less<strong>查看</strong>文件内容</p>
<p>5, 删除gff 文件</p>
<p>6, <strong>删除</strong> study<strong>文件夹</strong></p>
<p>具体操作阅读原文看我的视频。大约在5分20秒处</p>
<h2 id="环境变量"><a href="#环境变量" class="headerlink" title="环境变量"></a>环境变量</h2><p>环境变量也是Linux学习中非常重要的知识点，不了解他甚至都不能用好软件。</p>
<p>不过环境变量属于哪种你不知道，你觉得很高级，一旦知道后，却发现很简单的存在。其实我们从小学或者初中开始就通过数学了解到它。比如说x  + 3 &#x3D;5， x+ 4&#x3D;5，其中的x就是变量，就是会变的量。</p>
<p>在计算机里，它就用来存储其他的值一个名字，相当于一个中介。</p>
<p>变量有可以分为环境变量和局部变量，环境变量就是比局部变量作用更广泛的一个变量。举个例子，当我说到爱因斯坦这个名字时，这会让你想到的是一个物理学家，但是如果你家里有一只猫，它也叫爱因斯坦，他就是一个局部变量。局部变量可以覆盖环境变量。</p>
<p>接下来我们将会通过一个例子，通过PATH这个环境变量来理解。</p>
<ul>
<li><p>查看环境变量PATH</p>
</li>
<li><p>新建一个目录存放软件</p>
</li>
<li><p>将目录加入环境变量PATH</p>
</li>
<li><p>下载软件</p>
</li>
<li><p>给软件添加执行权限</p>
</li>
<li><p>测试软件</p>
</li>
</ul>
<p>具体操作阅读原文看我的视频。在15分钟的时候。</p>
<p>视频我传到了哔哩哔哩，假如你不喜欢哔哩哔哩，想要下载我的2880 x 1800（大约3G）的视频，这是百度盘地址，链接: <a target="_blank" rel="noopener" href="https://pan.baidu.com/s/17yxk78_kpeL5BA-3P0i06Q">https://pan.baidu.com/s/17yxk78_kpeL5BA-3P0i06Q</a>  密码:7ro0。相信我，你哔哩哔哩已经够高清了，你真的不需要20分钟3G的视频。</p>
<p>对于入门而言，知道的越少越好，因为细节无穷无尽。你先要上手，会敲命令了。那么后期继续深入反而就只是时间问题了。</p>
<p>之后，我会更新一系列入门提高的视频，介绍Linux的各种细节，比如说，</p>
<ul>
<li>括号的用法</li>
<li>变量操作</li>
<li>shell脚本</li>
<li>…</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/08/07/%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9-%E3%80%8C%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80%E8%AF%BE%E3%80%8D%E5%88%9D%E5%AD%A6%E8%80%85%E5%85%A5%E9%97%A8Linux%E6%9C%80%E5%B0%91%E5%BF%85%E8%A6%81%E7%9A%84%E7%9F%A5%E8%AF%86%E7%82%B9/" data-id="cllg90y30008ceqp26t1n0ff2" data-title="「生信基础课」初学者入门Linux最少必要的知识点" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-从零开始学CIRCOS绘制圈图(五)-从零开始学CIRCOS绘制圈图(五)" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/31/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)/" class="article-date">
  <time class="dt-published" datetime="2019-07-31T13:11:18.509Z" itemprop="datePublished">2019-07-31</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/31/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)/">从零开始学CIRCOS绘制圈图(五)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>这一部分承接<a href="/archives/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E5%9B%9B)">从零开始学CIRCOS绘制圈图(四)</a>，对之前的布局进行调整，使结构更加适合于发表。</p>
<h2 id="染色体的位置顺序"><a href="#染色体的位置顺序" class="headerlink" title="染色体的位置顺序"></a>染色体的位置顺序</h2><p>默认情况下是会显示所有的染色体，而且会先从ath1到ath5,然后从aly1到aly8, 那么如何调整顺序呢？</p>
<p>需要设置的参数是<code>chromosomes_order</code>,  按照自己的需求调整位置。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chromosomes_order = aly8,aly7,aly6,aly5,aly4,aly3,aly2,aly1</span><br></pre></td></tr></table></figure>

<p>虽然circos提供了一些快捷操作，比如说<code>-</code>,<code>$</code>,<code>^</code>, 但是远不如直接输入直观。</p>
<h2 id="反转染色体的坐标顺序"><a href="#反转染色体的坐标顺序" class="headerlink" title="反转染色体的坐标顺序"></a>反转染色体的坐标顺序</h2><p>默认都是从0到最大值，可以通过正则匹配的方式，将aly的染色体改成从最大值到0</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chromosomes_reverse = /aly/</span><br></pre></td></tr></table></figure>

<h2 id="调整染色体缩放"><a href="#调整染色体缩放" class="headerlink" title="调整染色体缩放"></a>调整染色体缩放</h2><p>为了让两个物种的基因组能够占Circos的两边，需要设置缩放。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chromosomes_scale = /aly/=0.5rn,/ath/=0.5rn</span><br></pre></td></tr></table></figure>

<p><code>/aly/=0.5rn</code>表示aly的所有染色体总共占据50%的空间</p>
<h2 id="定义染色体的间隔"><a href="#定义染色体的间隔" class="headerlink" title="定义染色体的间隔"></a>定义染色体的间隔</h2><p> <code>&lt;spcaing&gt;</code>控制的是不同染色体之间间隔的空隙大小，比如说下面的参数表示所有染色体之间的间隔是0.005r。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&lt;spacing&gt;</span><br><span class="line">default = 0.005r</span><br><span class="line">&lt;/spacing&gt;</span><br></pre></td></tr></table></figure>

<p>我们设置某些染色体之间的距离大一些，比如说两个物种的起始染色体和两个物种的终止染色体</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">&lt;spacing&gt;</span><br><span class="line">default = 1u</span><br><span class="line"></span><br><span class="line">  &lt;pairwise aly1 ath1&gt;</span><br><span class="line">   spacing = 10u</span><br><span class="line">  &lt;/pairwise&gt;</span><br><span class="line">  </span><br><span class="line">  &lt;pairwise aly8 ath5&gt;</span><br><span class="line">   spacing = 10u</span><br><span class="line">  &lt;/pairwise&gt;</span><br><span class="line">  </span><br><span class="line">&lt;/spacing&gt;</span><br></pre></td></tr></table></figure>

<h2 id="调整起始位置"><a href="#调整起始位置" class="headerlink" title="调整起始位置"></a>调整起始位置</h2><p>默认情况下，ath1从-90度位置开始，由于设置了aly1和ath1之间的空隙为10u，于是看起来就不对称了。为了让结果图能够对错，需要调整角度。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">angle_offset* = -85</span><br></pre></td></tr></table></figure>

<p>最终效果</p>
<p><img src="https://halo-1252249331.cos.ap-shanghai.myqcloud.com/upload/2019/7/circos-final-da04fc42d5984d8e8205ce9d95a849de.png" alt="circos-final"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://xuzhougeng.top/2019/07/31/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)-%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E5%AD%A6CIRCOS%E7%BB%98%E5%88%B6%E5%9C%88%E5%9B%BE(%E4%BA%94)/" data-id="cllg90y3f009oeqp28v860kks" data-title="从零开始学CIRCOS绘制圈图(五)" class="article-share-link"><span class="fa fa-share">分享</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/22/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/21/">21</a><a class="page-number" href="/page/22/">22</a><span class="page-number current">23</span><a class="page-number" href="/page/24/">24</a><a class="extend next" rel="next" href="/page/24/">下一页 &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分类</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/C/">C++</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Java/">Java</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Rust/">Rust</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%85%B6%E4%BB%96/">其他</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6/">数据科学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/">数据结构与算法</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB/">文献阅读</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%96%87%E7%AB%A0%E9%87%8D%E7%8E%B0/">文章重现</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1/">生信软件工具箱</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6%E5%B7%A5%E5%85%B7%E7%AE%B1-%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/">生信软件工具箱 | 基因组学</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">转录组学</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" rel="tag">ATAC-seq | 差异分析</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" rel="tag">ATAC-seq | 表观组 | ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">ATAC-seq | 表观组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BigWig/" rel="tag">BigWig</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/BioNano/" rel="tag">BioNano</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C/" rel="tag">C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/C-C-%E7%AE%97%E6%B3%95/" rel="tag">C&#x2F;C++ | 算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/ChIP-seq/" rel="tag">ChIP-seq</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/GitHub/" rel="tag">GitHub</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Hi-C/" rel="tag">Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/JCVI/" rel="tag">JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/MacOS/" rel="tag">MacOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS/" rel="tag">NGS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">NGS | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/PBMC/" rel="tag">PBMC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">Perl | 软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/QT-MSVC/" rel="tag">QT | MSVC</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat/" rel="tag">Seurat</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">Seurat | 单细胞 | 数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">Seurat | 转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Web%E5%BC%80%E5%8F%91/" rel="tag">Web开发</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" rel="tag">Zotero | 文献 | 坚果云</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/biocondutor/" rel="tag">biocondutor</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/conda/" rel="tag">conda</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/samtools/" rel="tag">samtools</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/typora/" rel="tag">typora</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" rel="tag">可视化 | CIRCOS</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" rel="tag">可视化 | JCVI</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">可视化 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">可视化 | 比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" rel="tag">基因组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" rel="tag">小技巧</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" rel="tag">小技巧 | SSH</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" rel="tag">数据挖掘</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" rel="tag">数据结构</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" rel="tag">数据结构 | C&#x2F;C++</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" rel="tag">正则表达式</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" rel="tag">正则表达式 | 字符串处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" rel="tag">比较基因组学</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">水稻 | 转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">水稻 | 转录组 | 遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A/" rel="tag">注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" rel="tag">注释 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">注释 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">注释 | 流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" rel="tag">注释 | 流程工具 | MAKER</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">注释 | 重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" rel="tag">流程工具</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" rel="tag">流程工具 | 基因家族</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" rel="tag">流程工具 | 服务器</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" rel="tag">源码解读</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" rel="tag">源码解读 | hash</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%88%AC%E8%99%AB/" rel="tag">爬虫</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" rel="tag">环境变量</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" rel="tag">环境配置 | WSL</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%AE%97%E6%B3%95/" rel="tag">算法</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85/" rel="tag">组装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" rel="tag">组装 | Hi-C</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" rel="tag">组装 | 转录组 | 注释</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" rel="tag">编程语言</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" rel="tag">自然语言 | 深度学习</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" rel="tag">表观组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%BB%E4%B9%A6/" rel="tag">读书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="tag">转录组</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" rel="tag">转录组 | 单细胞</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" rel="tag">转录组 | 差异分析 | TCGA</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag">转录组 | 序列比对</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" rel="tag">软件安装</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" rel="tag">遗传定位</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" rel="tag">重复序列</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/ATAC-seq-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 20px;">ATAC-seq | 单细胞</a> <a href="/tags/ATAC-seq-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" style="font-size: 10px;">ATAC-seq | 差异分析</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-ChIP-seq/" style="font-size: 10px;">ATAC-seq | 表观组 | ChIP-seq</a> <a href="/tags/ATAC-seq-%E8%A1%A8%E8%A7%82%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">ATAC-seq | 表观组 | 单细胞</a> <a href="/tags/BigWig/" style="font-size: 10px;">BigWig</a> <a href="/tags/BioNano/" style="font-size: 10px;">BioNano</a> <a href="/tags/C-C/" style="font-size: 19px;">C/C++</a> <a href="/tags/C-C-%E7%AE%97%E6%B3%95/" style="font-size: 10px;">C/C++ | 算法</a> <a href="/tags/ChIP-seq/" style="font-size: 10px;">ChIP-seq</a> <a href="/tags/GitHub/" style="font-size: 11px;">GitHub</a> <a href="/tags/Hi-C/" style="font-size: 10px;">Hi-C</a> <a href="/tags/JCVI/" style="font-size: 10px;">JCVI</a> <a href="/tags/MacOS/" style="font-size: 16px;">MacOS</a> <a href="/tags/NGS/" style="font-size: 11px;">NGS</a> <a href="/tags/NGS-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">NGS | 遗传定位</a> <a href="/tags/PBMC/" style="font-size: 10px;">PBMC</a> <a href="/tags/Perl-%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 10px;">Perl | 软件安装</a> <a href="/tags/QT-MSVC/" style="font-size: 10px;">QT | MSVC</a> <a href="/tags/Seurat/" style="font-size: 10px;">Seurat</a> <a href="/tags/Seurat-%E5%8D%95%E7%BB%86%E8%83%9E-%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 10px;">Seurat | 单细胞 | 数据挖掘</a> <a href="/tags/Seurat-%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">Seurat | 转录组 | 单细胞</a> <a href="/tags/Web%E5%BC%80%E5%8F%91/" style="font-size: 11px;">Web开发</a> <a href="/tags/Zotero-%E6%96%87%E7%8C%AE-%E5%9D%9A%E6%9E%9C%E4%BA%91/" style="font-size: 11px;">Zotero | 文献 | 坚果云</a> <a href="/tags/biocondutor/" style="font-size: 10px;">biocondutor</a> <a href="/tags/conda/" style="font-size: 11px;">conda</a> <a href="/tags/samtools/" style="font-size: 10px;">samtools</a> <a href="/tags/typora/" style="font-size: 10px;">typora</a> <a href="/tags/%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 13px;">单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12px;">可视化</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-CIRCOS/" style="font-size: 14px;">可视化 | CIRCOS</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-JCVI/" style="font-size: 12px;">可视化 | JCVI</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 10px;">可视化 | 单细胞</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96-%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">可视化 | 比较基因组学</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 11px;">基因家族</a> <a href="/tags/%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 11px;">基因组</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7/" style="font-size: 17px;">小技巧</a> <a href="/tags/%E5%B0%8F%E6%8A%80%E5%B7%A7-SSH/" style="font-size: 10px;">小技巧 | SSH</a> <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">序列比对</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98/" style="font-size: 11px;">数据挖掘</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84/" style="font-size: 12px;">数据结构</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84-C-C/" style="font-size: 16px;">数据结构 | C/C++</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 12px;">服务器</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F/" style="font-size: 10px;">正则表达式</a> <a href="/tags/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F-%E5%AD%97%E7%AC%A6%E4%B8%B2%E5%A4%84%E7%90%86/" style="font-size: 10px;">正则表达式 | 字符串处理</a> <a href="/tags/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AD%A6/" style="font-size: 10px;">比较基因组学</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 10px;">水稻 | 转录组</a> <a href="/tags/%E6%B0%B4%E7%A8%BB-%E8%BD%AC%E5%BD%95%E7%BB%84-%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">水稻 | 转录组 | 遗传定位</a> <a href="/tags/%E6%B3%A8%E9%87%8A/" style="font-size: 12px;">注释</a> <a href="/tags/%E6%B3%A8%E9%87%8A-MAKER/" style="font-size: 13px;">注释 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">注释 | 序列比对</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 10px;">注释 | 流程工具</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-MAKER/" style="font-size: 10px;">注释 | 流程工具 | MAKER</a> <a href="/tags/%E6%B3%A8%E9%87%8A-%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">注释 | 重复序列</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7/" style="font-size: 17px;">流程工具</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F/" style="font-size: 10px;">流程工具 | 基因家族</a> <a href="/tags/%E6%B5%81%E7%A8%8B%E5%B7%A5%E5%85%B7-%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">流程工具 | 服务器</a> <a href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 10px;">深度学习</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB/" style="font-size: 10px;">源码解读</a> <a href="/tags/%E6%BA%90%E7%A0%81%E8%A7%A3%E8%AF%BB-hash/" style="font-size: 10px;">源码解读 | hash</a> <a href="/tags/%E7%88%AC%E8%99%AB/" style="font-size: 10px;">爬虫</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/" style="font-size: 10px;">环境变量</a> <a href="/tags/%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE-WSL/" style="font-size: 10px;">环境配置 | WSL</a> <a href="/tags/%E7%AE%97%E6%B3%95/" style="font-size: 11px;">算法</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 12px;">系统发育</a> <a href="/tags/%E7%BB%84%E8%A3%85/" style="font-size: 18px;">组装</a> <a href="/tags/%E7%BB%84%E8%A3%85-Hi-C/" style="font-size: 16px;">组装 | Hi-C</a> <a href="/tags/%E7%BB%84%E8%A3%85-%E8%BD%AC%E5%BD%95%E7%BB%84-%E6%B3%A8%E9%87%8A/" style="font-size: 10px;">组装 | 转录组 | 注释</a> <a href="/tags/%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80/" style="font-size: 11px;">编程语言</a> <a href="/tags/%E8%87%AA%E7%84%B6%E8%AF%AD%E8%A8%80-%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/" style="font-size: 11px;">自然语言 | 深度学习</a> <a href="/tags/%E8%A1%A8%E8%A7%82%E7%BB%84/" style="font-size: 10px;">表观组</a> <a href="/tags/%E8%AF%BB%E4%B9%A6/" style="font-size: 11px;">读书</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 11px;">转录组</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%8D%95%E7%BB%86%E8%83%9E/" style="font-size: 11px;">转录组 | 单细胞</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90-TCGA/" style="font-size: 10px;">转录组 | 差异分析 | TCGA</a> <a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84-%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" style="font-size: 10px;">转录组 | 序列比对</a> <a href="/tags/%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85/" style="font-size: 15px;">软件安装</a> <a href="/tags/%E9%81%97%E4%BC%A0%E5%AE%9A%E4%BD%8D/" style="font-size: 10px;">遗传定位</a> <a href="/tags/%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" style="font-size: 12px;">重复序列</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/06/">六月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/05/">五月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/04/">四月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/03/">三月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/02/">二月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2023/01/">一月 2023</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/12/">十二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/11/">十一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/10/">十月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/09/">九月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/08/">八月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/07/">七月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/06/">六月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">四月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">三月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/02/">二月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">一月 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/12/">十二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">十一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/10/">十月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">九月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/08/">八月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">七月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/06/">六月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">五月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">四月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">二月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">一月 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">十一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/10/">十月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/08/">八月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/07/">七月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">六月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">五月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/04/">四月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/03/">三月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">二月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/11/">十一月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">十月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">九月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/08/">八月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">七月 2019</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2023/06/24/Oatk%EF%BC%9A%E5%88%A9%E7%94%A8HiFi%20Read%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E5%99%A8%E7%BB%84%E8%A3%85-oatk-assembly-of-organelles-using-hifi-read/">Oatk：利用HiFi Read进行细胞器组装</a>
          </li>
        
          <li>
            <a href="/2023/06/22/%E5%9C%A8Windows%E4%B8%AD%E6%90%AD%E5%BB%BA%E4%B8%80%E4%B8%AA%E5%9F%BA%E4%BA%8EWSL%E7%9A%84Python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6%E7%8E%AF%E5%A2%83-building-a-wsl-based-python-data-science-environment-in-windows/">在Windows中搭建一个基于WSL的Python数据科学环境</a>
          </li>
        
          <li>
            <a href="/2023/06/16/%E7%94%A8Colab%E7%BC%93%E8%A7%A3%E6%88%91%E7%9A%84GPU%E7%84%A6%E8%99%91-%E7%94%A8colab%E7%BC%93%E8%A7%A3%E6%88%91%E7%9A%84gpu%E7%84%A6%E8%99%91/">用Colab缓解我的GPU焦虑</a>
          </li>
        
          <li>
            <a href="/2023/05/31/%E6%8E%8C%E6%8F%A1%E4%BA%86%E4%B8%80%E4%B8%AA%E5%85%B3%E4%BA%8EBcrypt%E5%8A%A0%E5%AF%86%E7%9A%84%E7%9F%A5%E8%AF%86-acquired-a-knowledge-of-bcrypt-encryption/">掌握了一个关于Bcrypt加密的知识</a>
          </li>
        
          <li>
            <a href="/2023/05/28/%E6%88%91%E4%B8%8D%E7%9F%A5%E9%81%93%E8%AF%B4%E4%BA%9B%E4%BB%80%E4%B9%88-what-to-say/">我不知道说些什么</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2023 xuzhougeng<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>